{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6d9e95d4",
   "metadata": {},
   "source": [
    "* **2024-09-13:** Modified notebook for Kevin to have a single heatmap with top and right annotations\n",
    "* **ComplexHeatmap reference:** https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fb36f61",
   "metadata": {},
   "source": [
    "#### Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3faa1d14",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "------------------------------------------------------------------------------\n",
      "\n",
      "You have loaded plyr after dplyr - this is likely to cause problems.\n",
      "If you need functions from both plyr and dplyr, please load plyr first, then dplyr:\n",
      "library(plyr); library(dplyr)\n",
      "\n",
      "------------------------------------------------------------------------------\n",
      "\n",
      "\n",
      "Attaching package: ‘plyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ────────────────────────────────────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.5.1     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "── \u001b[1mConflicts\u001b[22m ──────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32marrange()\u001b[39m   masks \u001b[34mdplyr\u001b[39m::arrange()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompact()\u001b[39m  masks \u001b[34mplyr\u001b[39m::compact()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32mcount()\u001b[39m     masks \u001b[34mdplyr\u001b[39m::count()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32mdesc()\u001b[39m      masks \u001b[34mdplyr\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32mfailwith()\u001b[39m  masks \u001b[34mdplyr\u001b[39m::failwith()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m   masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32mid()\u001b[39m        masks \u001b[34mdplyr\u001b[39m::id()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m      masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32mmutate()\u001b[39m    masks \u001b[34mdplyr\u001b[39m::mutate()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32mrename()\u001b[39m    masks \u001b[34mdplyr\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32msummarise()\u001b[39m masks \u001b[34mdplyr\u001b[39m::summarise()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mplyr\u001b[39m::\u001b[32msummarize()\u001b[39m masks \u001b[34mdplyr\u001b[39m::summarize()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.18.0\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite either one:\n",
      "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
      "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "    genomic data. Bioinformatics 2016.\n",
      "\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.16\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘gridtext’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ComplexHeatmap’:\n",
      "\n",
      "    textbox_grob\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(dplyr); library(plyr); library(glue)\n",
    "library(tidyverse)\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)  # colorRamp2()\n",
    "library(GetoptLong)\n",
    "library(ggsci)  # pal_jco()\n",
    "library(gridtext)\n",
    "options(stringsAsFactors=F)\n",
    "library(IRdisplay)  # display() for pretty display of R outputs\n",
    "options(repr.matrix.max.rows=600, repr.matrix.max.cols=200)  # View more table cells (Default: 60, 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "672fb82b",
   "metadata": {},
   "source": [
    "#### Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "e34c36e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp = \"cutie_df.txt\"\n",
    "cutie_df = read.table(fp, sep='\\t', header=T, stringsAsFactors=F,\n",
    "                      check.names=F, quote='', comment.char='')\n",
    "fp = \"lefse_df.txt\"\n",
    "lefse_df = read.table(fp, sep='\\t', header=T, check.names=F,\n",
    "                      quote='', comment.char='')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b74ffd95",
   "metadata": {},
   "source": [
    "#### Plotting functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "370b3c04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat Metaphlan full taxonomy strings\n",
    "rename_taxa = function(x) {\n",
    "    strsplit(as.character(x), ';')[[1]] %>% \n",
    "    .[length(.)] %>% \n",
    "    {gsub('^s__', '', .)} %>%\n",
    "    setNames(x)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cd4afa28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Will convert Cutie output from long to wide format\n",
    "# p, r, cutie, var1, var2: Column names in Cutie df,\n",
    "#    replace with your Cutie columns of interest\n",
    "create_cutie_matrices = function(cutie.df) {\n",
    "    output = lapply(list(p='p', r='r', c='cutie'), function(v) {\n",
    "        cutie.df[, c('var1', 'var2', v)] %>%\n",
    "        reshape(idvar='var1', timevar='var2', direction='wide') %>%\n",
    "        'rownames<-'(.$var1) %>% .[,-1] %>%\n",
    "        'colnames<-'(gsub(glue(\"^{v}[.]\"), \"\", colnames(.))) %>%\n",
    "        .[order(rownames(.)), order(colnames(.))] %>%\n",
    "        as.matrix(check.names=F)\n",
    "    })\n",
    "    return(output)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ac1f17d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Converts output from create_cutie_matrices into ComplexHeatmap function input\n",
    "create_heatmap_matrices = function(cutie.matrices, p.filt=T, \n",
    "                                   lefse.filt=NULL, r.filt=NULL) {\n",
    "    # Filter function\n",
    "    filter_df = function(key, df, rowfilt) {\n",
    "        if (sum(rowfilt)>0) {\n",
    "            df = df[rowfilt,]\n",
    "        } else {\n",
    "            print(glue(\"{key} filter error: Zero row features\"))\n",
    "        }\n",
    "        return(df)\n",
    "    }\n",
    "    # Filter by correlation strength\n",
    "    if (!is.null(r.filt)) {\n",
    "        rowfilt = apply(cutie.matrices$r, 1, function(x) any(abs(x)>=r.filt, na.rm=T))\n",
    "        print(glue(\"r filt: {sum(rowfilt)}/{length(rowfilt)} features\"))\n",
    "        cutie.matrices = lapply(cutie.matrices, function(x) filter_df('r', x, rowfilt))\n",
    "    }\n",
    "    # Filter by significant Lefse features\n",
    "    if (!is.null(lefse.filt)) {\n",
    "        rowfilt = rownames(cutie.matrices$r) %in% lefse.filt\n",
    "        print(glue(\"Lefse filt: {sum(rowfilt)}/{length(rowfilt)} features\"))\n",
    "        cutie.matrices = lapply(cutie.matrices, function(x) filter_df('lefse', x, rowfilt))\n",
    "    }\n",
    "    # Filter significant correlations\n",
    "    check.rowfilt = c()\n",
    "    if (p.filt) {\n",
    "        rowfilt = apply(cutie.matrices$p, 1, function(x) any(x<0.05, na.rm=T))\n",
    "        cutie.matrices = lapply(cutie.matrices, function(x) filter_df('pval', x, rowfilt))\n",
    "    }\n",
    "    # Omit all NA rows/cols\n",
    "    rowfilt = apply(cutie.matrices$r, 1, function(x) all(is.na(x)))\n",
    "    print(glue(\"NA filt: {sum(rowfilt)}/{length(rowfilt)} features\"))\n",
    "    colfilt = apply(cutie.matrices$r, 2, function(x) !all(is.na(x)))\n",
    "    if (sum(rowfilt)>0 & all(lapply(colfilt, sum)>0)) {\n",
    "        output = lapply(cutie.matrices, function(x) x[rowfilt, colfilt[[i]]])\n",
    "    } else {\n",
    "        print(glue(\"Omit NAs filter error: (row={sum(rowfilt)}, col={sum(colfilt)}) features\"))\n",
    "        output = cutie.matrices\n",
    "    }\n",
    "    print(glue(\"Total: {nrow(cutie.matrices$r)} -> {nrow(output$r)} features\"))      \n",
    "            \n",
    "    return(output)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "08bed93d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot heatmap\n",
    "plot_heatmap = function(hm.inputs, lefse.df,\n",
    "                          fp, f.w, f.h, plot.title=NULL,\n",
    "                          hm.title, hm.w, hm.h, \n",
    "                          annot.r.lefse.ymax,\n",
    "                          hm.rownames=NULL, hm.colnames=NULL,\n",
    "                          pch.size = 6,\n",
    "                          row.km=1, col.km=1, km.gap=1.5) {\n",
    "    # Annotations\n",
    "    annot_top = HeatmapAnnotation(\n",
    "        `Top Annotation` = annotmap_top[colnames(hm_mat$r)],\n",
    "        col = list(`Top Annotation`=annotcol_top),\n",
    "        simple_anno_size = unit(3, \"mm\"), \n",
    "        show_annotation_name = F\n",
    "    )\n",
    "    annotmap_right = lefse.df[match(rownames(hm_mat$r), lefse.df$feature), ]\n",
    "    annot_right = rowAnnotation(\n",
    "        `Right Annotation` = anno_simple(\n",
    "            annotmap_right$class,\n",
    "            col = annotcol_right,\n",
    "            width = unit(3,'mm')\n",
    "        ),\n",
    "        Lefse = anno_barplot(\n",
    "            round(annotmap_right$log_LDA, digits=1),\n",
    "            gp=gpar(fill=annotcol_right[annotmap_right$class], col=NA),\n",
    "            width=unit(1.5, 'cm'),\n",
    "            bar_width=0.8, \n",
    "            add_numbers=TRUE,\n",
    "            axis_param = list(labels_rot=0),\n",
    "            numbers_rot = 0,\n",
    "            numbers_gp = gpar(col='white', cex=0.8),\n",
    "            numbers_offset = unit(-0.55, 'cm'),\n",
    "            ylim = c(0, annot.r.lefse.ymax)  # transposed\n",
    "        ),\n",
    "        annotation_label = c('', gt_render(\"log10(LDA)\", gp = gpar(cex=0.9))),\n",
    "        gap = unit(1.7, \"mm\")  # Between 2 annotations\n",
    "    )\n",
    "    # Legends\n",
    "    pch.shape.p = 0\n",
    "    pch.shape.c = 20\n",
    "    pch.color = 'black'\n",
    "    pch.lwd = 1.7\n",
    "    lgd1 = Legend(  # anno_simple() legend has to be manually created\n",
    "        title = 'Right Annotation',\n",
    "        labels = c('Psychotic Case', 'Healthy Control'),\n",
    "        legend_gp = gpar(fill = annotcol_right)\n",
    "    )\n",
    "    lgd2 = Legend(\n",
    "        title = 'LEfSe Annotation',\n",
    "        labels = c('Psychotic Case', 'Healthy Control'),\n",
    "        legend_gp = gpar(fill = annotcol_right)\n",
    "    )\n",
    "    hm_color = colorRamp2(\n",
    "        c(-1,-0.5,0,0.5,1), \n",
    "        c(\"turquoise4\", 'turquoise3', \"white\", \"sienna1\", 'sienna2')\n",
    "    )  # Alternative color scheme: c(\"dodgerblue3\", 'dodgerblue1', \"white\", \"brown1\", 'brown3')\n",
    "    lgd3 = Legend(\n",
    "        title=\"Correlation r\",\n",
    "        direction='horizontal',\n",
    "        col_fun=hm_color,\n",
    "        at=c(-1,0,1), \n",
    "        legend_width=unit(2.5, \"cm\"),\n",
    "        grid_height=unit(3, \"mm\")\n",
    "    )\n",
    "    lgd4 = Legend(\n",
    "        title = \"Significant correlations\",\n",
    "        labels = c('P < 0.05', 'True Positive'),\n",
    "        type = \"points\", \n",
    "        pch = c(pch.shape.p, pch.shape.c),\n",
    "        legend_gp = gpar(lwd=1.5, col=1), \n",
    "        background = NULL, size = unit(c(4,3), \"mm\")\n",
    "    )\n",
    "    lgd = packLegend(lgd1, lgd2, lgd3, lgd4)\n",
    "    # Heatmaps\n",
    "    ht_opt(\n",
    "        heatmap_row_names_gp = gpar(fontsize=11),\n",
    "        heatmap_column_names_gp = gpar(fontsize=11),\n",
    "        heatmap_column_title_gp = gpar(fontsize=11),\n",
    "        ROW_ANNO_PADDING=unit(2.7, 'mm'),\n",
    "        COLUMN_ANNO_PADDING=unit(2, 'mm')\n",
    "    )\n",
    "    hm.rownames = if (is.null(hm.rownames)) rownames(hm.inputs$r) else hm.rownames\n",
    "    hm.colnames = if (is.null(hm.colnames)) colnames(hm.inputs$r) else hm.colnames\n",
    "    hm_main = Heatmap(\n",
    "        hm.inputs$r, col = hm_color, na_col = 'black',\n",
    "        column_title = hm.title,        \n",
    "        column_title_gp = gpar(fontsize = 12),\n",
    "        column_km = col.km,\n",
    "        column_gap = unit(km.gap, \"mm\"),\n",
    "        row_labels = gt_render(\n",
    "            hm.rownames, padding = unit(c(0,0,0,1.2), \"mm\") # top,right,bottom,left\n",
    "        ),\n",
    "        column_labels = gt_render(hm.colnames, padding=unit(1.2, 'mm')),\n",
    "        right_annotation = annot_right,\n",
    "        top_annotation = annot_top,\n",
    "        width = unit(hm.w, \"cm\"),\n",
    "        show_heatmap_legend = F,\n",
    "        layer_fun = function(j, i, x, y, width, height, fill) {\n",
    "            v = pindex(hm.inputs$p, i, j)\n",
    "            l = v<0.05\n",
    "            if (any(l, na.rm=T)) {\n",
    "                grid.points(x[l], y[l], pch=pch.shape.p, size=unit(pch.size, \"mm\"),\n",
    "                            gp=gpar(lwd=pch.lwd, col=pch.color))\n",
    "            }\n",
    "            v = pindex(hm.inputs$c, i, j)\n",
    "            l = v==1\n",
    "            if (any(l, na.rm=T)) {\n",
    "                grid.points(x[l], y[l], pch=pch.shape.c, size=unit(pch.size-3.5, \"mm\"),\n",
    "                            gp=gpar(lwd=pch.lwd, col=pch.color))\n",
    "            }\n",
    "        }\n",
    "    )\n",
    "    # Full heatmap\n",
    "    png(qq(fp), width=f.w, height=f.h, units='in', res=300)\n",
    "    draw(\n",
    "        hm_main, \n",
    "        height = unit(hm.h, \"cm\"),\n",
    "        ht_gap = unit(2.5, \"mm\"),\n",
    "        row_km = row.km, row_gap = unit(km.gap, \"mm\"),\n",
    "        column_title = plot.title,\n",
    "        column_title_gp = gpar(fontsize=16),\n",
    "        annotation_legend_list = lgd,\n",
    "        annotation_legend_side = \"left\"\n",
    "    )\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "774cd735",
   "metadata": {},
   "source": [
    "#### Create annotation color and label maps\n",
    "Color names: https://r-charts.com/colors/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a87fb9b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat data\n",
    "cutie_df = filter(cutie_df, grepl('^WB_|^HC_', cutie_df$var2))\n",
    "lefse_df = filter(lefse_df, p<0.05)\n",
    "lefse_df$class = mapvalues(lefse_df$class,\n",
    "                           c('PsychoticCase', 'HealthyControl'),\n",
    "                           c('Psychotic Case', 'Healthy Control'),\n",
    "                           warn_missing=F)\n",
    "\n",
    "# Create annotation label and color maps\n",
    "imaging_vars = cutie_df$var2\n",
    "annotmap_top = ifelse(grepl('^WB_|vol', imaging_vars), 'Volume', 'Metabolite') %>%\n",
    "    setNames(imaging_vars)\n",
    "annotcol_top = c(Volume='orange', Metabolite='navy')\n",
    "annotcol_right = c(\"Psychotic Case\"='#dd443c', \n",
    "                 \"Healthy Control\"='#0073C2FF')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a67ff932",
   "metadata": {},
   "source": [
    "#### Plot heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5c67dbc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lefse filt: 118/1771 features\n",
      "NA filt: 0/32 features\n",
      "Omit NAs filter error: (row=0, col=10) features\n",
      "Total: 32 -> 32 features\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Create heatmap input matrices\n",
    "cutie_matrices = create_cutie_matrices(cutie_df)\n",
    "hm_mat = create_heatmap_matrices(cutie.matrices=cutie_matrices,\n",
    "                                 lefse.filt=lefse_df$feature)\n",
    "\n",
    "# Create row/col name maps\n",
    "M_rownames = sapply(rownames(hm_mat$r), rename_taxa)\n",
    "M_colnames = gsub('_', ' ', colnames(hm_mat$r))\n",
    "\n",
    "# Plot and save heatmap\n",
    "fp = glue('heatmap.png')\n",
    "plot_heatmap(\n",
    "    hm.inputs=hm_mat, lefse.df=lefse_df,\n",
    "    fp=fp, f.w=8.5, f.h=7, plot.title=glue(\"Plot Title\"),\n",
    "    hm.w=4, hm.h=12, hm.title='Heatmap Title', \n",
    "    annot.r.lefse.ymax=6,\n",
    "    hm.rownames=M_rownames,\n",
    "    hm.colnames=M_colnames,\n",
    "    row.km=2, col.km=1,\n",
    "    pch.size=5\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d426a62",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
