{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3f62dc73-df18-4548-bc0d-aea2a125fef0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "import pandas as pd\n",
    "import scipy.stats\n",
    "import statsmodels.stats.multitest\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "\n",
    "# disable warnings, use w caution\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "# project specific libs\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import pathlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a8b42930-bcde-48f3-9753-8779e1aa11fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# project specific path\n",
    "path = '/Users/KevinBu/Desktop/clemente_lab/Projects/oa/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0343d6a9-1c06-4ed7-90a1-ed193002a1dd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Process Mapping Files ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "a3013747-2e0e-412a-b719-b2a6a28afbf2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>Separate</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Together</th>\n",
       "      <th>ContactEmail</th>\n",
       "      <th>ContactName</th>\n",
       "      <th>PrimaryInvestigator</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>RawDataNotes</th>\n",
       "      <th>...</th>\n",
       "      <th>broccoli</th>\n",
       "      <th>Garbanzo_beans</th>\n",
       "      <th>pork</th>\n",
       "      <th>beef</th>\n",
       "      <th>burger</th>\n",
       "      <th>Total_omega3</th>\n",
       "      <th>Adherence_omega3</th>\n",
       "      <th>Total_omega6</th>\n",
       "      <th>Adherence_omega6</th>\n",
       "      <th>Total_o3_o6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.plasma</th>\n",
       "      <td>TTCAGTTCGTTA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-001.pre.plasma.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.plasma</th>\n",
       "      <td>CGGCCAGAAGCA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>post</td>\n",
       "      <td>OAD-001.post.plasma.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001.post.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>72.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>131.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.plasma</th>\n",
       "      <td>GACGTTAAGAAT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-003.pre.plasma.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>53.2</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.plasma</th>\n",
       "      <td>TCGCTACAGATG</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>post</td>\n",
       "      <td>OAD-003.post.plasma.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003.post.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>108.0</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>171.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.plasma</th>\n",
       "      <td>ATGGGACCTTCA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-004.pre.plasma.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-004.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>2.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>28.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>35.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 188 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    BarcodeSequence    LinkerPrimerSequence Separate  \\\n",
       "#SampleID                                                              \n",
       "OAD-001.pre.plasma     TTCAGTTCGTTA  CCGGACTACHVGGGTWTCTAAT      All   \n",
       "OAD-001.post.plasma    CGGCCAGAAGCA  CCGGACTACHVGGGTWTCTAAT      All   \n",
       "OAD-003.pre.plasma     GACGTTAAGAAT  CCGGACTACHVGGGTWTCTAAT      All   \n",
       "OAD-003.post.plasma    TCGCTACAGATG  CCGGACTACHVGGGTWTCTAAT      All   \n",
       "OAD-004.pre.plasma     ATGGGACCTTCA  CCGGACTACHVGGGTWTCTAAT      All   \n",
       "\n",
       "                    Timepoint                           Together  \\\n",
       "#SampleID                                                          \n",
       "OAD-001.pre.plasma        pre   OAD-001.pre.plasma.guma.plate313   \n",
       "OAD-001.post.plasma      post  OAD-001.post.plasma.guma.plate313   \n",
       "OAD-003.pre.plasma        pre   OAD-003.pre.plasma.guma.plate313   \n",
       "OAD-003.post.plasma      post  OAD-003.post.plasma.guma.plate313   \n",
       "OAD-004.pre.plasma        pre   OAD-004.pre.plasma.guma.plate313   \n",
       "\n",
       "                                     ContactEmail    ContactName  \\\n",
       "#SampleID                                                          \n",
       "OAD-001.pre.plasma   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-001.post.plasma  rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-003.pre.plasma   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-003.post.plasma  rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-004.pre.plasma   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "\n",
       "                    PrimaryInvestigator Cohort  \\\n",
       "#SampleID                                        \n",
       "OAD-001.pre.plasma           Jose Scher  NonVA   \n",
       "OAD-001.post.plasma          Jose Scher  NonVA   \n",
       "OAD-003.pre.plasma           Jose Scher  NonVA   \n",
       "OAD-003.post.plasma          Jose Scher  NonVA   \n",
       "OAD-004.pre.plasma           Jose Scher  NonVA   \n",
       "\n",
       "                                         RawDataNotes  ... broccoli  \\\n",
       "#SampleID                                              ...            \n",
       "OAD-001.pre.plasma    OAD-001.pre.stool.guma.plate313  ...      0.0   \n",
       "OAD-001.post.plasma  OAD-001.post.stool.guma.plate313  ...      4.0   \n",
       "OAD-003.pre.plasma    OAD-003.pre.stool.guma.plate313  ...      0.0   \n",
       "OAD-003.post.plasma  OAD-003.post.stool.guma.plate313  ...      4.0   \n",
       "OAD-004.pre.plasma    OAD-004.pre.stool.guma.plate313  ...      2.8   \n",
       "\n",
       "                    Garbanzo_beans pork  beef burger Total_omega3  \\\n",
       "#SampleID                                                           \n",
       "OAD-001.pre.plasma             0.0  0.0   0.0    0.0          NaN   \n",
       "OAD-001.post.plasma            0.0  0.0   0.0    0.0         48.0   \n",
       "OAD-003.pre.plasma             0.0  0.0  14.0    0.0         16.8   \n",
       "OAD-003.post.plasma            0.0  0.0   0.0    0.0         48.0   \n",
       "OAD-004.pre.plasma             0.0  0.0   0.0    0.0          2.8   \n",
       "\n",
       "                    Adherence_omega3 Total_omega6 Adherence_omega6 Total_o3_o6  \n",
       "#SampleID                                                                       \n",
       "OAD-001.pre.plasma     Low adherence          NaN    Low adherence         0.0  \n",
       "OAD-001.post.plasma    Low adherence         72.0    Low adherence       131.0  \n",
       "OAD-003.pre.plasma     Low adherence         53.2   High adherence        75.0  \n",
       "OAD-003.post.plasma    Low adherence        108.0   High adherence       171.0  \n",
       "OAD-004.pre.plasma     Low adherence         28.0    Low adherence        35.8  \n",
       "\n",
       "[5 rows x 188 columns]"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# from AC Q2 run of merged saliva stool\n",
    "df_map = pd.read_csv(path + 'inputs/Qiime2_0/qiime_mapping_file.tsv', sep='\\t', index_col=0)\n",
    "q2_row = df_map.loc['#q2:types',:]\n",
    "df_map = df_map.drop('#q2:types')\n",
    "\n",
    "# change index so it matches metadata file\n",
    "df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "\n",
    "# drop MOC and elution buffer\n",
    "df_map = df_map.drop(['MOC.320','elutionbuffer.plate313'])\n",
    "\n",
    "# grab metadata\n",
    "df_meta = pd.read_csv(path + 'inputs/Metadata_OA.csv')\n",
    "\n",
    "# rename 'Run_ID_Saliva' to be correct\n",
    "df_meta['Timepoints'] = df_meta['Timepoints'].apply(lambda x: 'pre' if x == '0' else 'post')\n",
    "df_meta['Patient_ID'] = df_meta['Patient_ID'].apply(lambda x: x[:-3])  \n",
    "df_meta['Study_ID'] = df_meta['Study_ID'].apply(lambda x: x.split('_')[0][-3:]) \n",
    "df_outcomes = df_meta.copy()\n",
    "\n",
    "# create per sample type mapping files\n",
    "type_to_ST = {'saliva':'Saliva','stool':'fecal'}\n",
    "type_to_df_map = {}\n",
    "\n",
    "# split into specimen type\n",
    "for t in type_to_ST:\n",
    "    # subset on specimen type\n",
    "    df_map_type = df_map[df_map['SpecimenType'] == type_to_ST[t]]\n",
    "\n",
    "    # as to not overwrite df meta\n",
    "    df_meta_type = df_meta.copy()\n",
    "\n",
    "    # create new sample ID for specimen type and set as index\n",
    "    df_meta_type['#SampleID'] = df_meta['Patient_ID'] + '-' + df_meta['Study_ID'] + '.' + df_meta['Timepoints'] + '.' + t\n",
    "    df_meta_type = df_meta_type.set_index('#SampleID')\n",
    "\n",
    "    # create full mapping file\n",
    "    df_map_type = pd.concat([df_map_type, df_meta_type],axis=1)\n",
    "\n",
    "    # use only sequenced samples\n",
    "    df_map_type = df_map_type.dropna(how='any',subset='BarcodeSequence')\n",
    "\n",
    "    # drop all na columns\n",
    "    df_map_type = df_map_type.dropna(how='all',axis=1)\n",
    "\n",
    "    # drop VAD OA 015 because misdx with PsA not OA\n",
    "    # drop OAD007 in saliva bc possible contaminants\n",
    "    if t == 'saliva':\n",
    "        df_map_type = df_map_type.drop(['VAOAD-015.pre.saliva','VAOAD-015.post.saliva'])\n",
    "        # df_map_type = df_map_type.drop(['OAD-007.pre.saliva','OAD-007.post.saliva'])\n",
    "\n",
    "    if t == 'stool':\n",
    "        df_map_type = df_map_type.drop(['VAOAD-015.pre.stool','VAOAD-015.post.stool'])\n",
    "\n",
    "    # populate dict of mapping files\n",
    "    type_to_df_map[t] = df_map_type\n",
    "\n",
    "    # export for q2\n",
    "    df_q2_type = pd.concat([q2_row.to_frame().T, df_map_type])\n",
    "    df_q2_type.index.name = '#SampleID'\n",
    "    df_q2_type.iloc[0,:] = 'categorical'\n",
    "    df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + t + '.tsv', sep='\\t')\n",
    "    df_q2_type = df_q2_type[df_q2_type['Adherece_antiinflam'].isin(['Moderate adherence', 'High adherence','categorical'])]\n",
    "    df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh.tsv', sep='\\t')\n",
    "\n",
    "    # split by response category\n",
    "    df_q2_resp = df_q2_type[df_q2_type['WOMAC_P_Response'].isin(['categorical', 'Response'])]\n",
    "    df_q2_resp.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_response.tsv', sep='\\t')\n",
    "    df_q2_nonresp = df_q2_type[df_q2_type['WOMAC_P_Response'].isin(['categorical', 'No response'])]\n",
    "    df_q2_nonresp.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_noresponse.tsv', sep='\\t')\n",
    "\n",
    "    # split by timepoint\n",
    "    df_q2_pre = df_q2_type[df_q2_type['Timepoint'].isin(['pre','categorical'])]\n",
    "    df_q2_pre = df_q2_pre.dropna(subset='WOMAC_P_Response')\n",
    "    df_q2_pre.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_pre.tsv', sep='\\t')\n",
    "    df_q2_post = df_q2_type[df_q2_type['Timepoint'].isin(['post', 'categorical'])]\n",
    "    df_q2_post = df_q2_post.dropna(subset='WOMAC_P_Response')\n",
    "    df_q2_post.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_post.tsv', sep='\\t')\n",
    "\n",
    "# create new fake df map for plasma \n",
    "df_plasma = type_to_df_map['stool'].copy()\n",
    "df_plasma.index = df_plasma.index.map(lambda x: x.replace('.stool','.plasma'))\n",
    "df_plasma['Together'] = df_plasma['Together'].map(lambda x: x.replace('.stool','.plasma'))\n",
    "type_to_df_map['plasma'] = df_plasma\n",
    "\n",
    "type_to_df_map['plasma'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "248b05fc-430d-4a56-9d93-bfc3334b840a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Assess Outcomes ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "95e4ef90-0416-4803-a5b2-d730031630c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n",
      "['VAS_Pt', 'VAS_overall', 'WOMAC_pain', 'WOMAC_stiffness', 'WOMAC_activity', 'WOMAC_total', 'Pain_DETECT', 'CES_D', 'Helplesness', 'Magnification', 'Rumination', 'PCS_EN', 'Sleep_distrubance', 'PASE_walk', 'PASE_light', 'BMI']\n",
      "32\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var</th>\n",
       "      <th>effect</th>\n",
       "      <th>pval</th>\n",
       "      <th>stat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VAS_Pt</td>\n",
       "      <td>198.5</td>\n",
       "      <td>0.008316</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VAS_overall</td>\n",
       "      <td>168.5</td>\n",
       "      <td>0.131314</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_pain</td>\n",
       "      <td>170.0</td>\n",
       "      <td>0.017900</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_stiffness</td>\n",
       "      <td>172.5</td>\n",
       "      <td>0.013052</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_activity</td>\n",
       "      <td>164.5</td>\n",
       "      <td>0.032592</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               var  effect      pval stat\n",
       "0           VAS_Pt   198.5  0.008316  MWU\n",
       "0      VAS_overall   168.5  0.131314  MWU\n",
       "0       WOMAC_pain   170.0  0.017900  MWU\n",
       "0  WOMAC_stiffness   172.5  0.013052  MWU\n",
       "0   WOMAC_activity   164.5  0.032592  MWU"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###\n",
    "# Hypothesis 1: There will be a measurable difference in WOMAC pain response scores and \n",
    "# other outcomes from baseline to after the dietary intervention.\n",
    "###\n",
    "\n",
    "# outcome variables\n",
    "outcomes = ['VAS_Pt', 'VAS_overall', 'WOMAC_pain', 'WOMAC_stiffness', 'WOMAC_activity', 'WOMAC_total', 'Pain_DETECT', \n",
    "            'CES_D', 'Helplesness', 'Magnification', 'Rumination', 'PCS_EN', 'Sleep_distrubance', 'PASE_walk', 'PASE_light', \n",
    "            #'PASE_gardening', # Where did this go? gardening_improve is binary\n",
    "            'BMI']\n",
    "\n",
    "\n",
    "# hypothesis 1\n",
    "# create a new df_meta\n",
    "df_meta = pd.read_csv(path + 'inputs/Metadata_OA.csv')\n",
    "\n",
    "# rename 'Run_ID_Saliva' to be correct\n",
    "df_meta['Timepoints'] = df_meta['Timepoints'].apply(lambda x: 'pre' if x == '0' else 'post')\n",
    "df_meta['Patient_ID'] = df_meta['Patient_ID'].apply(lambda x: x[:-3])  \n",
    "df_meta['Study_ID'] = df_meta['Study_ID'].apply(lambda x: x.split('_')[0][-3:]) \n",
    "df_meta['#SampleID'] = df_meta['Patient_ID'] + '-' + df_meta['Study_ID'] + '.' + df_meta['Timepoints'] + '.stool'\n",
    "df_meta = df_meta.set_index('#SampleID')\n",
    "\n",
    "# convert % to floats for calculations down the road\n",
    "bin = []\n",
    "cont = []\n",
    "df_paired_os = []\n",
    "for w in outcomes:\n",
    "    df_w = df_meta[w]\n",
    "    if df_w.nunique() > 2: # do spearman\n",
    "        df_meta[w] = df_meta[w].astype(str).str.replace('%','').astype(float).values\n",
    "        cont.append(w)\n",
    "    else:\n",
    "        bin.append(w)\n",
    "\n",
    "    # compute difference and store it\n",
    "    df_md = df_meta.copy()\n",
    "    df_md['SubjectID'] = df_md['Patient_ID'] + df_md['Study_ID']\n",
    "    \n",
    "    # first drop unpaired samples\n",
    "    s_remove = []\n",
    "    for s in list(df_md['SubjectID'].values):\n",
    "        if len(df_md[df_md['SubjectID'] == s]) != 2:\n",
    "            s_remove.append(s)\n",
    "    df_md = df_md.loc[~df_md['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "    \n",
    "    # set vars\n",
    "    group_var = 'Timepoints'\n",
    "    pair_var = 'SubjectID'\n",
    "    groups = ['pre','post']\n",
    "    \n",
    "    # get paired per indiv pair\n",
    "    pair_to_diff = {}\n",
    "    for p in list(df_md[pair_var].values):\n",
    "        df = df_md[df_md[pair_var] == p]\n",
    "        t0 = float(df[df[group_var] == groups[0]][w].values)\n",
    "        tf = float(df[df[group_var] == groups[1]][w].values)\n",
    "        pair_to_diff[p] = tf - t0\n",
    "    \n",
    "    df_paired_o = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[w + '_diff'])\n",
    "    df_paired_os.append(df_paired_o)\n",
    "\n",
    "df_meta_paired = pd.concat([*df_paired_os], axis=1)    \n",
    "\n",
    "print(bin)\n",
    "print(cont)\n",
    "\n",
    "# split into all and mod high only\n",
    "a = 'modhigh'\n",
    "job = 'jobs02'\n",
    "df_meta = df_meta[df_meta['Adherece_antiinflam'].isin(['Moderate adherence', 'High adherence'])]\n",
    "        \n",
    "print(len(df_meta))\n",
    "\n",
    "df_results = pd.DataFrame(columns=['var','effect','pval','stat'])\n",
    "# do post treatment vals of binary vars differ from pre treatment 'unpaired'\n",
    "for b in bin:\n",
    "    ct_table_ind=pd.crosstab(df_meta[\"Timepoints\"],df_meta[b])\n",
    "    chi2_stat, p, dof, expected = scipy.stats.chi2_contingency(ct_table_ind)\n",
    "    row=pd.DataFrame.from_dict({'var': [b],'effect':[chi2_stat],'pval':[p],'stat':['chi2']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "# fishers exact\n",
    "for b in bin:\n",
    "    ct_table_ind=pd.crosstab(df_meta[\"Timepoints\"],df_meta[b])\n",
    "    fisher, p = scipy.stats.fisher_exact(ct_table_ind)\n",
    "    row=pd.DataFrame.from_dict({'var': [b],'effect':[t],'pval':[p],'stat':['fisher']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "    \n",
    "# do post treatment vals of continuous vars differ from pre treatment unpaired\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    try:\n",
    "        W,p = scipy.stats.mannwhitneyu(x=df_pre[c].values,y=df_post[c].values, nan_policy='omit')\n",
    "    except:\n",
    "        W,p = 0, 1\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['mwu']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v')\n",
    "        sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette='dark:grey', hue=None, orient='v')\n",
    "    \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/mwu_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['MWU']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "    \n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    t,p = scipy.stats.ttest_ind(a=df_pre[c].values,b=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['ttest']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v')\n",
    "        sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette='dark:grey', hue=None, orient='v')\n",
    "    \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/tt_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['ttest']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "# unpaired and then paired\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    W,p = scipy.stats.wilcoxon(x=df_pre[c].values,y=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['WSR']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.swarmplot(data=df_meta, x='Timepoints', y=c, hue='WOMAC_P_Response', palette=['b','r'], orient='v')\n",
    "        for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "            ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "        sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v', hue=None, boxprops=dict(alpha=.3))\n",
    "        \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/wsr_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['WSR']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    t,p = scipy.stats.ttest_rel(a=df_pre[c].values,b=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['pairedt']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette=['b','r'], hue='WOMAC_P_Response',orient='v')\n",
    "        for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "            ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "        sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v', hue=None, boxprops=dict(alpha=0.3))\n",
    "        \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/pairedt_' + c  + '.pdf')\n",
    "        plt.close()          \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['pairedt']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "df_results.to_csv(path + 'outputs/' + job + '/outcome_testing.tsv', sep='\\t')\n",
    "df_results.head()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0379f4d-6b07-489a-bfae-75b6085b9a15",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Grab Paired Alpha ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "298b4824-3c24-485e-8705-5daf4a85d6d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bray_Curtis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.6672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>0.2837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>0.4237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>0.3257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>0.1022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Bray_Curtis\n",
       "OAD001       0.6672\n",
       "OAD003       0.2837\n",
       "OAD004       0.4237\n",
       "OAD005       0.3257\n",
       "OAD006       0.1022"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###\n",
    "# Hypothesis 2: There will be an association between oral and gut microbiome and pain outcomes\n",
    "###\n",
    "# construct alpha, beta and paired alpha dataframes\n",
    "g_to_dfd = {}\n",
    "# g_test = ['stool','saliva_adh', 'stool_adh', 'saliva']\n",
    "g_test = ['saliva_adh', 'stool_adh']\n",
    "\n",
    "for g in g_test:\n",
    "    # maps diversity type to dataframe\n",
    "    g_to_dfd[g] = {}\n",
    "\n",
    "    # get alpha diversities\n",
    "    df_alpha = pd.read_csv(path + 'outputs/Qiime2_' + g + '/metadata.tsv', sep='\\t', index_col=0)\n",
    "    df_alpha = df_alpha.drop('#q2:types')\n",
    "    df_alpha['SubjectID'] = df_alpha['Patient_ID'] + df_alpha['Study_ID']\n",
    "    df_alpha = df_alpha[['SubjectID', 'Timepoints', 'shannon_entropy']]\n",
    "    g_to_dfd[g]['alpha'] = df_alpha\n",
    "\n",
    "    # get paired alpha div, first drop unpaired samples\n",
    "    s_remove = []\n",
    "    for s in list(df_alpha['SubjectID'].values):\n",
    "        if len(df_alpha[df_alpha['SubjectID'] == s]) != 2:\n",
    "            s_remove.append(s)\n",
    "    df_alpha = df_alpha.loc[~df_alpha['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "    \n",
    "    # set vars\n",
    "    alpha_metric = 'shannon_entropy'\n",
    "    group_var = 'Timepoints'\n",
    "    pair_var = 'SubjectID'\n",
    "    groups = ['pre','post']\n",
    "    \n",
    "    # get paired per indiv pair\n",
    "    pair_to_diff = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        df = df_alpha[df_alpha[pair_var] == p]\n",
    "        alpha_0 = float(df[df[group_var] == groups[0]][alpha_metric].values)\n",
    "        alpha_1 = float(df[df[group_var] == groups[1]][alpha_metric].values)\n",
    "        pair_to_diff[p] = alpha_1 - alpha_0\n",
    "    \n",
    "    df_paired_alpha = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[alpha_metric + '_diff'])\n",
    "    g_to_dfd[g]['paired_alpha'] = df_paired_alpha\n",
    "\n",
    "    # get beta div\n",
    "    df_beta = pd.read_csv(path + 'outputs/Qiime2_' + g + '/core_metrics_results/distance-matrix.tsv',\n",
    "                              sep='\\t', index_col=0)\n",
    "        \n",
    "    # grab twin to pair dict\n",
    "    pair_to_ids = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        df = df_alpha[df_alpha[pair_var] == p]\n",
    "        id_0 = str(df[df[group_var] == groups[0]].index.values[0])\n",
    "        id_1 = str(df[df[group_var] == groups[1]].index.values[0])\n",
    "        pair_to_ids[p] = (id_0, id_1)\n",
    "    \n",
    "    # get distances for each twin pair per beta div matrix    \n",
    "    pair_to_dist = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        id_0, id_1 = pair_to_ids[p]\n",
    "        pair_to_dist[p] = df_beta.loc[id_0, id_1]\n",
    "    \n",
    "    df_paired_beta = pd.DataFrame.from_dict(pair_to_dist, orient='index', columns=['Bray_Curtis'])\n",
    "    g_to_dfd[g]['paired_beta'] = df_paired_beta\n",
    "\n",
    "g_to_dfd['stool_adh']['paired_beta'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e018facb-88c3-4c10-be22-dbe70ee41351",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Chi2 Association of Outcome with Alpha ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9d4f38ff-41d6-4f31-94ab-69879de56109",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>div</th>\n",
       "      <th>outcome</th>\n",
       "      <th>statistic</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>pval</th>\n",
       "      <th>timepoint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>0.955089</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.313871</td>\n",
       "      <td>0.758599</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.244898</td>\n",
       "      <td>0.620691</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>17.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.682542</td>\n",
       "      <td>0.510397</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>0.396892</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>-0.563791</td>\n",
       "      <td>0.582492</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.244898</td>\n",
       "      <td>0.620691</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>21.000000</td>\n",
       "      <td>0.638889</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.769556</td>\n",
       "      <td>0.459355</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         group                  div           outcome statistic  test_stat  \\\n",
       "0     metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.000000   \n",
       "1   saliva_adh                alpha  WOMAC_P_Response       mwu  29.000000   \n",
       "2   saliva_adh                alpha  WOMAC_P_Response        tt   0.313871   \n",
       "3     metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.244898   \n",
       "4   saliva_adh                alpha  WOMAC_P_Response       mwu  17.000000   \n",
       "5   saliva_adh                alpha  WOMAC_P_Response        tt   0.682542   \n",
       "6     metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.000000   \n",
       "7    stool_adh                alpha  WOMAC_P_Response       mwu  20.000000   \n",
       "8    stool_adh                alpha  WOMAC_P_Response        tt  -0.563791   \n",
       "9     metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.244898   \n",
       "10   stool_adh                alpha  WOMAC_P_Response       mwu  21.000000   \n",
       "11   stool_adh                alpha  WOMAC_P_Response        tt   0.769556   \n",
       "\n",
       "        pval timepoint  \n",
       "0   1.000000       pre  \n",
       "1   0.955089       pre  \n",
       "2   0.758599       pre  \n",
       "3   0.620691      post  \n",
       "4   1.000000      post  \n",
       "5   0.510397      post  \n",
       "6   1.000000       pre  \n",
       "7   0.396892       pre  \n",
       "8   0.582492       pre  \n",
       "9   0.620691      post  \n",
       "10  0.638889      post  \n",
       "11  0.459355      post  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Testing for each sample type for (1) all and (2) high adh only\n",
    "# (A) Chisq of quartiles with adherence \n",
    "# (B) MWU/TT unpaired of outcomes against distance\n",
    "\n",
    "def chisq_of_df_cols(df, c1, c2):\n",
    "    groupsizes = df.groupby([c1, c2]).size()\n",
    "    ctsum = groupsizes.unstack(c1)\n",
    "    # fillna(0) is necessary to remove any NAs which will cause exceptions\n",
    "    return(scipy.stats.chi2_contingency(ctsum.fillna(0)))\n",
    "\n",
    "d_to_metric = {\n",
    "    'alpha': 'shannon_entropy',\n",
    "}\n",
    "group_var = 'Adherece_antiinflam'\n",
    "\n",
    "dfd_to_merge = {}\n",
    "g_test = ['saliva_adh', 'stool_adh']\n",
    "\n",
    "for g in g_test:\n",
    "    dfd_to_merge[g] = {}\n",
    "\n",
    "gs = []\n",
    "ds = []\n",
    "os = []\n",
    "stats = []\n",
    "ts = []\n",
    "ps = []\n",
    "times = []\n",
    "\n",
    "arr = [gs,ds,os,stats,ts,ps,times]\n",
    "\n",
    "def append_results(arr, val):\n",
    "    for a,v in zip(arr,val):\n",
    "        a.append(v)\n",
    "    return arr\n",
    "\n",
    "# for each sample type, grab relevant mapping file\n",
    "# g_test = saliva, saliva_adh, stool, etc.\n",
    "for g in g_test:\n",
    "    for d in d_to_metric:\n",
    "        # grab relevant diversity df\n",
    "        df_div = g_to_dfd[g][d]\n",
    "\n",
    "        if d == 'alpha':\n",
    "            # df_div = df_div[df_div['Timepoints'] == 'pre']\n",
    "            # df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            # df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div['shannon_entropy'] = df_div['shannon_entropy'].astype(float)\n",
    "\n",
    "        # merge with df of metadata var        \n",
    "        # grab relevant sample IDs\n",
    "        # g = saliva_adh\n",
    "        idx = 'Run_ID_' + g.split('_')[0].capitalize()\n",
    "        df_meta_sub = df_meta.dropna(subset=idx)\n",
    "        df_meta_sub = df_meta_sub.set_index(idx)\n",
    "        df_merge1 = pd.concat([df_meta_sub,df_div.drop('Timepoints',axis=1)],axis=1)\n",
    "\n",
    "        # subset on adh only\n",
    "        df_merge1 = df_merge1[df_merge1[group_var].isin(['Moderate adherence', 'High adherence'])]\n",
    "\n",
    "        # for each timepoint\n",
    "        for time in ['pre', 'post']:\n",
    "            df_merge = df_merge1[df_merge1['Timepoints'] == time]\n",
    "            # test association of div with outcomes\n",
    "            for o in ['WOMAC_P_Response']: #  outcomes:                    \n",
    "                # test association of adherence with pain outcomes\n",
    "                div_metric = d_to_metric[d]\n",
    "                x,p,dof,ef = chisq_of_df_cols(df_merge, group_var, o)\n",
    "                arr = append_results(arr, ['metadata',group_var,o,'chisq',x,p,time])\n",
    "                \n",
    "                ax = sns.boxplot(data=df_merge, x=o, y=div_metric)\n",
    "                sns.swarmplot(data=df_merge, x=o, y=div_metric, palette='dark:grey')\n",
    "                sns.despine()\n",
    "            \n",
    "                plt.tight_layout()\n",
    "                plt.savefig(path + 'outputs/jobs06/rvnr_nondiff_' + time + '_' + g + '_' + o + '_' + d + '.pdf')\n",
    "                plt.close()          \n",
    "            \n",
    "                u, p = scipy.stats.mannwhitneyu(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                                df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                                nan_policy='omit')\n",
    "    \n",
    "                arr = append_results(arr, [g,d,o,'mwu',u,p,time])\n",
    "    \n",
    "                t, p = scipy.stats.ttest_ind(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                                df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                                nan_policy='omit')\n",
    "    \n",
    "                arr = append_results(arr, [g,d,o,'tt',t,p,time])\n",
    "    \n",
    "            # save results\n",
    "            dfd_to_merge[g][d] = df_merge\n",
    "    \n",
    "            # export to Q2\n",
    "            # df_q2_type = df_merge.set_index(['Together'])\n",
    "            df_q2_type = df_merge.copy()\n",
    "            q2_row = pd.Series(data=['categorical' for i in range(len(df_merge.columns))], \n",
    "                               index=list(df_merge.columns.values), dtype=str, name='#q2:types')\n",
    "            df_q2_type = pd.concat([q2_row.to_frame().T, df_q2_type])\n",
    "            df_q2_type.index.name = '#SampleID'\n",
    "            df_q2_type.index = df_q2_type.index.map(lambda x: x.split('.guma')[0])\n",
    "            # df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + d + '_' + g + 'aggregate_outcomes.tsv', sep='\\t')\n",
    "\n",
    "df_results = pd.DataFrame.from_dict({\n",
    "    'group': gs,\n",
    "    'div': ds,\n",
    "    'outcome': os,\n",
    "    'statistic': stats,\n",
    "    'test_stat': ts,\n",
    "    'pval': ps,\n",
    "    'timepoint': times\n",
    "})\n",
    "# df_results.to_csv(path + 'outputs/df_results_aggregate.tsv', sep='\\t')\n",
    "# df_results.head()\n",
    "# df_results[df_results['pval'] < 0.05]\n",
    "df_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "22f26972-623f-459c-93fd-d37eb65d17eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>div</th>\n",
       "      <th>outcome</th>\n",
       "      <th>statistic</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.769556</td>\n",
       "      <td>0.343434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>-0.670379</td>\n",
       "      <td>0.519757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>-0.670379</td>\n",
       "      <td>0.638889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.777949</td>\n",
       "      <td>0.456353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.777949</td>\n",
       "      <td>0.431818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>1.184945</td>\n",
       "      <td>0.279964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>1.184945</td>\n",
       "      <td>0.876263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>-0.977825</td>\n",
       "      <td>0.360842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>-0.977825</td>\n",
       "      <td>0.755051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.558291</td>\n",
       "      <td>0.591262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.558291</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.731791</td>\n",
       "      <td>0.481109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.731791</td>\n",
       "      <td>0.876263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.164212</td>\n",
       "      <td>0.872849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.164212</td>\n",
       "      <td>0.267677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>1.234156</td>\n",
       "      <td>0.245859</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         group           div           outcome statistic  test_stat      pval\n",
       "0    stool_adh     pre_alpha  WOMAC_P_Response       mwu   0.769556  0.343434\n",
       "1    stool_adh     pre_alpha  WOMAC_P_Response        tt  -0.670379  0.519757\n",
       "2    stool_adh    post_alpha  WOMAC_P_Response       mwu  -0.670379  0.638889\n",
       "3    stool_adh    post_alpha  WOMAC_P_Response        tt   0.777949  0.456353\n",
       "4    stool_adh  paired_alpha  WOMAC_P_Response       mwu   0.777949  0.431818\n",
       "5    stool_adh  paired_alpha  WOMAC_P_Response        tt   1.184945  0.279964\n",
       "6    stool_adh   paired_beta  WOMAC_P_Response       mwu   1.184945  0.876263\n",
       "7    stool_adh   paired_beta  WOMAC_P_Response        tt  -0.977825  0.360842\n",
       "8   saliva_adh     pre_alpha  WOMAC_P_Response       mwu  -0.977825  0.755051\n",
       "9   saliva_adh     pre_alpha  WOMAC_P_Response        tt   0.558291  0.591262\n",
       "10  saliva_adh    post_alpha  WOMAC_P_Response       mwu   0.558291  1.000000\n",
       "11  saliva_adh    post_alpha  WOMAC_P_Response        tt   0.731791  0.481109\n",
       "12  saliva_adh  paired_alpha  WOMAC_P_Response       mwu   0.731791  0.876263\n",
       "13  saliva_adh  paired_alpha  WOMAC_P_Response        tt   0.164212  0.872849\n",
       "14  saliva_adh   paired_beta  WOMAC_P_Response       mwu   0.164212  0.267677\n",
       "15  saliva_adh   paired_beta  WOMAC_P_Response        tt   1.234156  0.245859"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# expand previous to all data, paired and unpaired diversities\n",
    "d_to_metric = {\n",
    "    'pre_alpha': 'shannon_entropy',\n",
    "    'post_alpha': 'shannon_entropy',\n",
    "    'paired_alpha': 'shannon_entropy_diff',\n",
    "    'paired_beta': 'Bray_Curtis' # 'Unweighted_Unifrac',\n",
    "}\n",
    "group_var = 'Adherece_antiinflam'\n",
    "\n",
    "dfd_to_merge = {}\n",
    "g_test = ['stool_adh','saliva_adh']\n",
    "for g in g_test:\n",
    "    dfd_to_merge[g] = {}\n",
    "\n",
    "gs = []\n",
    "ds = []\n",
    "os = []\n",
    "stats = []\n",
    "ts = []\n",
    "ps = []\n",
    "\n",
    "arr = [gs,ds,os,stats,ts,ps]\n",
    "\n",
    "def append_results(arr, val):\n",
    "    for a,v in zip(arr,val):\n",
    "        a.append(v)\n",
    "    return arr\n",
    "\n",
    "# for each sample type, grab relevant mapping file\n",
    "# g_test = saliva, saliva_adh, stool, etc.\n",
    "d_to_map = {}\n",
    "\n",
    "for g in g_test:\n",
    "    # drop duplicates so you have sample mapping to adh\n",
    "    df_map_sub = type_to_df_map[g.split('_')[0]]\n",
    "    df_map_sub.index = df_map_sub.index.map(lambda x: x.split('.')[0].replace('-',''))\n",
    "    df_map_sub = df_map_sub.dropna(how='any',subset=group_var,axis=0)           \n",
    "    \n",
    "    # figure out which samples to keep\n",
    "    # i.e. samples that have a pre and post time point\n",
    "    keep = []\n",
    "    for i in list(df_map_sub.index.values):\n",
    "        if len(df_map_sub.loc[i,:]) == 2:\n",
    "            keep.append(i)\n",
    "\n",
    "    # get unique entires in sorted order\n",
    "    # at this point we are only concerned with differences in values, \n",
    "    # as we've dropped samples with only one endpoint val\n",
    "    save = []\n",
    "    [save.append(x) for x in keep if x not in save]\n",
    "    df_map_sub = df_map_sub.loc[save,:]\n",
    "\n",
    "    # this double populates as OAD001 is an index twice, so the diff fills to both the pre and post col\n",
    "    for o in outcomes:\n",
    "        df_map_sub[o + '_diff'] = df_map_sub[df_map_sub['Timepoint'] == 'post'][o] - df_map_sub[df_map_sub['Timepoint'] == 'pre'][o] \n",
    "\n",
    "    # here we keep only the pre, but everything is identical b/w pre and post\n",
    "    df_dropdup = df_map_sub[~df_map_sub.index.duplicated(keep='first')]\n",
    "    d_to_map[g] = df_dropdup\n",
    "    \n",
    "    for d in d_to_metric:\n",
    "        # grab relevant diversity df\n",
    "        if d == 'pre_alpha' or d == 'post_alpha':\n",
    "            df_div = g_to_dfd[g]['alpha']\n",
    "        else:\n",
    "            df_div = g_to_dfd[g][d]\n",
    "        div_metric = d_to_metric[d]\n",
    "\n",
    "        # when associating alpha div vs outcomes, look at if starting adiv predicts outcome\n",
    "        if d == 'pre_alpha':\n",
    "            df_div = df_div[df_div['Timepoints'] == 'pre']\n",
    "            df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div = df_div.astype(float)\n",
    "\n",
    "        if d == 'post_alpha':\n",
    "            df_div = df_div[df_div['Timepoints'] == 'post']\n",
    "            df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div = df_div.astype(float)\n",
    "\n",
    "        # merge with df of metadata var        \n",
    "        df_merge = pd.concat([df_dropdup,df_div],axis=1)\n",
    "\n",
    "        # drop na in barcodes if samples not sequenced both pre and post\n",
    "        df_merge = df_merge.dropna(how='any',subset='BarcodeSequence')\n",
    "        \n",
    "        # test association of div with outcomes\n",
    "        for o in ['WOMAC_P_Response']:                     \n",
    "            # test association of adherence with pain outcomes\n",
    "            # flip x and y for horizontal\n",
    "            ax = sns.boxplot(data=df_merge, x=o, y=div_metric)\n",
    "            sns.swarmplot(data=df_merge, x=o, y=div_metric, palette='dark:grey')\n",
    "            sns.despine()\n",
    "        \n",
    "            plt.tight_layout()\n",
    "            plt.savefig(path + 'outputs/jobs06/WOMAC_barplot_' + g + '_' + o + '_' + d + '.pdf')\n",
    "            plt.close()          \n",
    "\n",
    "            # MWU and ttest\n",
    "            u, p = scipy.stats.mannwhitneyu(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'mwu',t,p])\n",
    "\n",
    "            t, p = scipy.stats.ttest_ind(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                         df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                         nan_policy='omit',\n",
    "                                         equal_var=False)\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'tt',t,p])\n",
    "         \n",
    "        # save results\n",
    "        dfd_to_merge[g][d] = df_merge\n",
    "\n",
    "        # export to Q2\n",
    "        df_q2_type = df_merge.set_index(['Together'])\n",
    "        q2_row = pd.Series(data=['categorical' for i in range(len(df_merge.columns))], \n",
    "                           index=list(df_merge.columns.values), dtype=str, name='#q2:types')\n",
    "        df_q2_type = pd.concat([q2_row.to_frame().T, df_q2_type])\n",
    "        df_q2_type.index.name = '#SampleID'\n",
    "        df_q2_type.index = df_q2_type.index.map(lambda x: x.split('.guma')[0])\n",
    "        # df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + d + '_' + g + '_outcomes.tsv', sep='\\t')\n",
    "\n",
    "df_results = pd.DataFrame.from_dict({\n",
    "    'group': gs,\n",
    "    'div': ds,\n",
    "    'outcome': os,\n",
    "    'statistic': stats,\n",
    "    'test_stat': ts,\n",
    "    'pval': ps\n",
    "})\n",
    "# df_results.to_csv(path + 'outputs/df_results_diff.tsv', sep='\\t')\n",
    "\n",
    "# df_results.head()\n",
    "# the div==alpha results test whether pre-alpha div state associates with pain outcome changes (differences) quartiles\n",
    "# the div==paired_alpha and paired_beta test whether the alpha and betas change in a similar way with the pain outcome\n",
    "print(len(df_results[df_results['pval'] < 0.05]))\n",
    "df_results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5db737b8-b788-4d7a-86a6-39e82ac57eb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_adh\n",
      "['Response', 'No response']\n",
      "WilcoxonResult(statistic=25.0, pvalue=0.30126953125)\n",
      "['Response']\n",
      "WilcoxonResult(statistic=4.0, pvalue=0.109375)\n",
      "['No response']\n",
      "WilcoxonResult(statistic=7.0, pvalue=1.0)\n",
      "saliva_adh\n",
      "['Response', 'No response']\n",
      "WilcoxonResult(statistic=24.0, pvalue=0.26611328125)\n",
      "['Response']\n",
      "WilcoxonResult(statistic=9.0, pvalue=0.46875)\n",
      "['No response']\n",
      "WilcoxonResult(statistic=5.0, pvalue=0.625)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>shannon_entropy_diff</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>Separate</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Together</th>\n",
       "      <th>ContactEmail</th>\n",
       "      <th>ContactName</th>\n",
       "      <th>PrimaryInvestigator</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>...</th>\n",
       "      <th>Pain_DETECT_diff</th>\n",
       "      <th>CES_D_diff</th>\n",
       "      <th>Helplesness_diff</th>\n",
       "      <th>Magnification_diff</th>\n",
       "      <th>Rumination_diff</th>\n",
       "      <th>PCS_EN_diff</th>\n",
       "      <th>Sleep_distrubance_diff</th>\n",
       "      <th>PASE_walk_diff</th>\n",
       "      <th>PASE_light_diff</th>\n",
       "      <th>BMI_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>-0.116436</td>\n",
       "      <td>AGGACAAACTAT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>Text</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-001.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>5.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.51</td>\n",
       "      <td>14.29</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>-0.327749</td>\n",
       "      <td>TCTTCCTAAAGT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-003.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-5.5</td>\n",
       "      <td>-14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>0.866474</td>\n",
       "      <td>CTTGTGCGACAA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-004.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.264032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>0.007278</td>\n",
       "      <td>CTTACACTGCTT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-001.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>5.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.081446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD012</th>\n",
       "      <td>0.535614</td>\n",
       "      <td>GCGTGGTCATTA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-012.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-1.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.72</td>\n",
       "      <td>-1.072664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TTGGACGTCCAC</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-009.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.139974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows Ã— 205 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          shannon_entropy_diff BarcodeSequence    LinkerPrimerSequence  \\\n",
       "OAD001               -0.116436    AGGACAAACTAT  CCGGACTACHVGGGTWTCTAAT   \n",
       "OAD003               -0.327749    TCTTCCTAAAGT  CCGGACTACHVGGGTWTCTAAT   \n",
       "OAD004                0.866474    CTTGTGCGACAA  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD001              0.007278    CTTACACTGCTT  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD012              0.535614    GCGTGGTCATTA  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD009                   NaN    TTGGACGTCCAC  CCGGACTACHVGGGTWTCTAAT   \n",
       "\n",
       "         Separate Timepoint              Together  \\\n",
       "OAD001       Text       pre    OAD-001.pre.saliva   \n",
       "OAD003        All       pre    OAD-003.pre.saliva   \n",
       "OAD004        All       pre    OAD-004.pre.saliva   \n",
       "VAOAD001      All       pre  VAOAD-001.pre.saliva   \n",
       "VAOAD012      All       pre  VAOAD-012.pre.saliva   \n",
       "VAOAD009      All       pre  VAOAD-009.pre.saliva   \n",
       "\n",
       "                          ContactEmail    ContactName PrimaryInvestigator  \\\n",
       "OAD001    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "OAD003    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "OAD004    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD001  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD012  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD009  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "\n",
       "         Cohort  ... Pain_DETECT_diff CES_D_diff Helplesness_diff  \\\n",
       "OAD001    NonVA  ...             -2.0       -1.0             -2.5   \n",
       "OAD003    NonVA  ...             -4.5        1.0             -1.0   \n",
       "OAD004    NonVA  ...             -1.0        3.0             -1.0   \n",
       "VAOAD001     VA  ...             -0.5        5.5              0.0   \n",
       "VAOAD012     VA  ...             -1.5        0.0              2.0   \n",
       "VAOAD009     VA  ...             -4.0        2.0              1.0   \n",
       "\n",
       "         Magnification_diff Rumination_diff PCS_EN_diff  \\\n",
       "OAD001                  5.5             1.0         4.0   \n",
       "OAD003                 -1.0            -2.0        -5.5   \n",
       "OAD004                 -1.0            -1.0        -3.0   \n",
       "VAOAD001                0.0             0.0         0.0   \n",
       "VAOAD012               -2.0            -0.5        -0.5   \n",
       "VAOAD009                0.0             0.5         1.5   \n",
       "\n",
       "         Sleep_distrubance_diff PASE_walk_diff PASE_light_diff  BMI_diff  \n",
       "OAD001                      9.0          11.51           14.29       NaN  \n",
       "OAD003                    -14.0           0.00            0.00       NaN  \n",
       "OAD004                      0.0           0.00            0.00  3.264032  \n",
       "VAOAD001                   -2.5           0.00            0.00 -0.081446  \n",
       "VAOAD012                   -1.5           0.00            6.72 -1.072664  \n",
       "VAOAD009                    2.5           0.00            0.00 -0.139974  \n",
       "\n",
       "[6 rows x 205 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test if the differences in diversity are significantly different from 0 for paired alpha\n",
    "for g in g_test:\n",
    "    print(g)\n",
    "    for d in ['paired_alpha']:\n",
    "        df_div = g_to_dfd[g][d] \n",
    "        df_map = d_to_map[g]\n",
    "\n",
    "        df_merge = pd.concat([df_div, df_map], axis=1)\n",
    "        for o in [['Response','No response'],['Response'], ['No response']]:\n",
    "            df = df_merge[df_merge['WOMAC_P_Response'].isin(o)]\n",
    "            print(o)\n",
    "            print(scipy.stats.wilcoxon(df['shannon_entropy_diff'], nan_policy='omit'))\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "872848a4-7f8e-4613-8839-f8ceab94a568",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_adh\n",
      "response\n",
      "alpha wilcoxon\n",
      "4.0 0.109375\n",
      "noresponse\n",
      "alpha wilcoxon\n",
      "7.0 1.0\n",
      "saliva_adh\n",
      "response\n",
      "alpha wilcoxon\n",
      "9.0 0.46875\n",
      "noresponse\n",
      "alpha wilcoxon\n",
      "5.0 0.625\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Separate</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Height</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>HostSubjectId</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Timepoints</th>\n",
       "      <th>Study_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>quinoa</th>\n",
       "      <th>broccoli</th>\n",
       "      <th>pork</th>\n",
       "      <th>beef</th>\n",
       "      <th>burger</th>\n",
       "      <th>Total_omega3</th>\n",
       "      <th>Adherence_omega3</th>\n",
       "      <th>Adherence_omega6</th>\n",
       "      <th>Total_o3_o6</th>\n",
       "      <th>shannon_entropy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.saliva</th>\n",
       "      <td>Text</td>\n",
       "      <td>pre</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>152.0</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001</td>\n",
       "      <td>F</td>\n",
       "      <td>OAD</td>\n",
       "      <td>pre</td>\n",
       "      <td>001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.30596529299739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.saliva</th>\n",
       "      <td>Limit 45 Characters</td>\n",
       "      <td>post</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>152.0</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001</td>\n",
       "      <td>F</td>\n",
       "      <td>OAD</td>\n",
       "      <td>post</td>\n",
       "      <td>001</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>131.0</td>\n",
       "      <td>5.22206342260189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.saliva</th>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>167.0</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003</td>\n",
       "      <td>F</td>\n",
       "      <td>OAD</td>\n",
       "      <td>pre</td>\n",
       "      <td>003</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>75.0</td>\n",
       "      <td>4.11813412588115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.saliva</th>\n",
       "      <td>All</td>\n",
       "      <td>post</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>167.0</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003</td>\n",
       "      <td>F</td>\n",
       "      <td>OAD</td>\n",
       "      <td>post</td>\n",
       "      <td>003</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>171.0</td>\n",
       "      <td>3.87363155038891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.saliva</th>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>158.0</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-004</td>\n",
       "      <td>F</td>\n",
       "      <td>OAD</td>\n",
       "      <td>pre</td>\n",
       "      <td>004</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>35.8</td>\n",
       "      <td>3.70928989877937</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 141 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Separate Timepoint               Ethnicity  \\\n",
       "id                                                                           \n",
       "OAD-001.pre.saliva                  Text       pre                Hispanic   \n",
       "OAD-001.post.saliva  Limit 45 Characters      post                Hispanic   \n",
       "OAD-003.pre.saliva                   All       pre                Hispanic   \n",
       "OAD-003.post.saliva                  All      post                Hispanic   \n",
       "OAD-004.pre.saliva                   All       pre  Not Hispanic or Latino   \n",
       "\n",
       "                    Height Cohort HostSubjectId Sex Patient_ID Timepoints  \\\n",
       "id                                                                          \n",
       "OAD-001.pre.saliva   152.0  NonVA       OAD-001   F        OAD        pre   \n",
       "OAD-001.post.saliva  152.0  NonVA       OAD-001   F        OAD       post   \n",
       "OAD-003.pre.saliva   167.0  NonVA       OAD-003   F        OAD        pre   \n",
       "OAD-003.post.saliva  167.0  NonVA       OAD-003   F        OAD       post   \n",
       "OAD-004.pre.saliva   158.0  NonVA       OAD-004   F        OAD        pre   \n",
       "\n",
       "                    Study_ID  ... quinoa broccoli pork  beef burger  \\\n",
       "id                            ...                                     \n",
       "OAD-001.pre.saliva       001  ...    0.0      0.0  0.0   0.0    0.0   \n",
       "OAD-001.post.saliva      001  ...    4.0      4.0  0.0   0.0    0.0   \n",
       "OAD-003.pre.saliva       003  ...    0.0      0.0  0.0  14.0    0.0   \n",
       "OAD-003.post.saliva      003  ...    0.0      4.0  0.0   0.0    0.0   \n",
       "OAD-004.pre.saliva       004  ...    0.0      2.8  0.0   0.0    0.0   \n",
       "\n",
       "                    Total_omega3 Adherence_omega3 Adherence_omega6  \\\n",
       "id                                                                   \n",
       "OAD-001.pre.saliva           NaN    Low adherence    Low adherence   \n",
       "OAD-001.post.saliva         48.0    Low adherence    Low adherence   \n",
       "OAD-003.pre.saliva          16.8    Low adherence   High adherence   \n",
       "OAD-003.post.saliva         48.0    Low adherence   High adherence   \n",
       "OAD-004.pre.saliva           2.8    Low adherence    Low adherence   \n",
       "\n",
       "                    Total_o3_o6   shannon_entropy  \n",
       "id                                                 \n",
       "OAD-001.pre.saliva          0.0  5.30596529299739  \n",
       "OAD-001.post.saliva       131.0  5.22206342260189  \n",
       "OAD-003.pre.saliva         75.0  4.11813412588115  \n",
       "OAD-003.post.saliva       171.0  3.87363155038891  \n",
       "OAD-004.pre.saliva         35.8  3.70928989877937  \n",
       "\n",
       "[5 rows x 141 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pre vs post for RESPONDERS only\n",
    "\n",
    "# paired beta, comparing intra-indiv difference pre_post to inter pre and inter post\n",
    "g_test = ['stool_adh','saliva_adh']\n",
    "sg = ['response', 'noresponse']\n",
    "for g in g_test: #subgroups:\n",
    "    print(g)\n",
    "    for s in sg:\n",
    "        print(s)\n",
    "        df = pd.read_csv(path + 'outputs/Qiime2_' + g + '_' + s + '/metadata.tsv', sep='\\t', index_col=0)\n",
    "        df = df.drop('#q2:types')\n",
    "\n",
    "        print('alpha wilcoxon')\n",
    "        df = df.dropna(subset=['Timepoints','shannon_entropy'])\n",
    "        df['shannon_entropy'] = df['shannon_entropy'].astype(float)\n",
    "        df_pre = df[df['Timepoints'] == 'pre']\n",
    "        df_post = df[df['Timepoints'] == 'post']\n",
    "\n",
    "        common_idx = list(set(df_pre.HostSubjectId.values).intersection(set(df_post.HostSubjectId.values)))\n",
    "        df_pre = df_pre[df_pre.HostSubjectId.isin(common_idx)]\n",
    "        df_post = df_post[df_post.HostSubjectId.isin(common_idx)]\n",
    "        \n",
    "        W, p = scipy.stats.wilcoxon(df_pre['shannon_entropy'].values, df_post['shannon_entropy'].values)\n",
    "        print(W,p)\n",
    "\n",
    "        df = pd.concat([df_pre, df_post])\n",
    "        ax = sns.swarmplot(data=df, x='Timepoint', y='shannon_entropy', hue='Timepoint', palette=['b','r'], orient='v')\n",
    "        for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "            ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "        sns.boxplot(data=df, x='Timepoint', y='shannon_entropy', orient='v', hue='Timepoint', boxprops=dict(alpha=.3))\n",
    "        sns.despine()\n",
    "    \n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/jobs18/alpha_' + g + '_' + s + '.pdf')\n",
    "        plt.close()          \n",
    "\n",
    "\n",
    "df_a.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bef8c05d-208b-4b13-bac7-01bf00e4e80e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_adh\n",
      "All\n",
      "paired alpha wilcoxon\n",
      "33.0 0.414306640625\n",
      "paired t test\n",
      "-0.7900792468039035 0.44481434457162783\n",
      "inter intra KW\n",
      "14.985539176909924 0.0005570978899780199\n",
      "Response\n",
      "paired alpha wilcoxon\n",
      "4.0 0.109375\n",
      "paired t test\n",
      "-2.1276058423185478 0.07746512839408434\n",
      "inter intra KW\n",
      "16.10169063709202 0.0003188322936852291\n",
      "No response\n",
      "paired alpha wilcoxon\n",
      "7.0 1.0\n",
      "paired t test\n",
      "0.2351990394957396 0.8256045805114028\n",
      "inter intra KW\n",
      "2.5703573471829486 0.2766011578942809\n",
      "saliva_adh\n",
      "All\n",
      "paired alpha wilcoxon\n",
      "26.0 0.19091796875\n",
      "paired t test\n",
      "-1.5438829559402965 0.14856768929110245\n",
      "inter intra KW\n",
      "17.042392642678692 0.0001992009743840896\n",
      "Response\n",
      "paired alpha wilcoxon\n",
      "9.0 0.46875\n",
      "paired t test\n",
      "-0.9798119153431403 0.365018427390056\n",
      "inter intra KW\n",
      "16.83795674134069 0.0002206399499696511\n",
      "No response\n",
      "paired alpha wilcoxon\n",
      "5.0 0.625\n",
      "paired t test\n",
      "-0.8754606396643463 0.43074573525689774\n",
      "inter intra KW\n",
      "19.225437442881685 6.687276850590436e-05\n"
     ]
    }
   ],
   "source": [
    "# Alpha div paired\n",
    "# paired beta, comparing intra-indiv difference pre_post to inter pre and inter post\n",
    "g_test = ['stool_adh','saliva_adh']\n",
    "subgroups = ['All', 'Response', 'No response']\n",
    "\n",
    "for g in g_test: #subgroups:\n",
    "    print(g)\n",
    "    df_a = pd.read_csv(path + 'outputs/Qiime2_' + g + '/metadata.tsv', sep='\\t', index_col=0)\n",
    "    df_a = df_a.drop('#q2:types')\n",
    "\n",
    "    for sg in subgroups:\n",
    "        print(sg)\n",
    "        if sg == 'All':\n",
    "            df_alpha = df_a.copy()\n",
    "        else:\n",
    "            df_alpha = df_a[df_a['WOMAC_P_Response'] == sg]\n",
    "        \n",
    "        df_alpha['SubjectID'] = df_alpha['Patient_ID'] + df_alpha['Study_ID']\n",
    "        df_alpha = df_alpha[['SubjectID', 'Timepoints', 'shannon_entropy']]\n",
    "        \n",
    "        # drop unpaired samples\n",
    "        s_remove = []\n",
    "        for s in list(df_alpha['SubjectID'].values):\n",
    "            if len(df_alpha[df_alpha['SubjectID'] == s]) != 2:\n",
    "                s_remove.append(s)\n",
    "        df_alpha = df_alpha.loc[~df_alpha['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "        \n",
    "        # set vars\n",
    "        alpha_metric = 'shannon_entropy'\n",
    "        group_var = 'Timepoints'\n",
    "        pair_var = 'SubjectID'\n",
    "        groups = ['pre','post']\n",
    "        \n",
    "        # get paired per indiv pair\n",
    "        pair_to_diff = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            df = df_alpha[df_alpha[pair_var] == p]\n",
    "            alpha_0 = float(df[df[group_var] == groups[0]][alpha_metric].values)\n",
    "            alpha_1 = float(df[df[group_var] == groups[1]][alpha_metric].values)\n",
    "            pair_to_diff[p] = alpha_0 - alpha_1\n",
    "        \n",
    "        df_paired_alpha = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[alpha_metric + '_diff'])\n",
    "        \n",
    "        # one-sided t-test, n.s.; RA-UA values \n",
    "        # t, p = scipy.stats.ttest_1samp(df_paired_alpha[alpha_metric + '_diff'],popmean=0)\n",
    "    \n",
    "        print('paired alpha wilcoxon')\n",
    "        s, p = scipy.stats.wilcoxon(df_paired_alpha[alpha_metric + '_diff'])\n",
    "        print(s, p)\n",
    "        print('paired t test')\n",
    "        t, p = scipy.stats.ttest_1samp(df_paired_alpha[alpha_metric + '_diff'], popmean=0)\n",
    "        print(t, p)\n",
    "        \n",
    "        # separate\n",
    "        df_alpha[alpha_metric] = df_alpha[alpha_metric].map(lambda x: float(x))\n",
    "        ax = sns.boxplot(data=df_alpha, x=group_var, y=alpha_metric)\n",
    "        sns.swarmplot(data=df_alpha, x=group_var, y=alpha_metric, palette='dark:grey')\n",
    "        sns.despine()\n",
    "    \n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/Qiime2_' + g + '/' + sg + '_alpha.pdf')\n",
    "        plt.close()          \n",
    "    \n",
    "        # now do beta\n",
    "        df_beta = pd.read_csv(path + 'outputs/Qiime2_' + g + '/core_metrics_results/distance-matrix.tsv',\n",
    "                              sep='\\t', index_col=0)\n",
    "        \n",
    "        # set vars\n",
    "        alpha_metric = 'shannon_entropy'\n",
    "        group_var = 'Timepoints'\n",
    "        pair_var = 'SubjectID'\n",
    "        groups = ['pre','post']\n",
    "        g0, g1 = groups[0], groups[1]\n",
    "        \n",
    "        # grab twin to pair dict\n",
    "        pair_to_ids = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            df = df_alpha[df_alpha[pair_var] == p]\n",
    "            id_0 = str(df[df[group_var] == g0].index.values[0])\n",
    "            id_1 = str(df[df[group_var] == g1].index.values[0])\n",
    "            pair_to_ids[p] = (id_0, id_1)\n",
    "        \n",
    "        # get distances for each twin pair per beta div matrix    \n",
    "        pair_to_dist = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            id_0, id_1 = pair_to_ids[p]\n",
    "            pair_to_dist[p] = df_beta.loc[id_0, id_1]\n",
    "        \n",
    "        df_paired_beta = pd.DataFrame.from_dict(pair_to_dist, orient='index', columns=['Bray_Curtis'])\n",
    "        \n",
    "        # grab inter RA distances\n",
    "        # this is from unweighted_Timepoint_significance.qzv -> download as tsv\n",
    "        df_raw = pd.read_csv(path + 'outputs/Qiime2_' + g + '/raw_data.tsv', \n",
    "                             sep='\\t', index_col=0)\n",
    "        df_0 = df_raw[df_raw['Group1'] == g0]\n",
    "        df_0 = df_0[df_0['Group2'] == g0]\n",
    "        df_1 = df_raw[df_raw['Group1'] == g1]\n",
    "        df_1 = df_1[df_1['Group2'] == g1]\n",
    "        \n",
    "        # compare distances\n",
    "        inter_twin = df_paired_beta['Bray_Curtis'].values\n",
    "        inter_0 = df_0['Distance'].values\n",
    "        inter_1 = df_1['Distance'].values\n",
    "        \n",
    "        u, p = scipy.stats.mannwhitneyu(inter_twin, inter_0)\n",
    "        #print(u, p)\n",
    "        \n",
    "        t, p = scipy.stats.ttest_ind(inter_twin, inter_1)\n",
    "        #print(t, p)\n",
    "        \n",
    "        t, p = scipy.stats.ttest_ind(inter_0, inter_1)\n",
    "        # print(t, p)\n",
    "        \n",
    "        f, p = scipy.stats.f_oneway(inter_0, inter_1, inter_twin)\n",
    "        # print('inter intra anova')\n",
    "        # print(f, p)\n",
    "\n",
    "        f, p = scipy.stats.kruskal(inter_0, inter_1, inter_twin)\n",
    "        print('inter intra KW')\n",
    "        print(f,p)\n",
    "\n",
    "        \n",
    "        category = ['intra_twin_pair']*len(inter_twin) + ['inter_' + g0 + '_only']*len(inter_0) + ['inter_' + g1 + '_only']*len(inter_1)\n",
    "        distances = list(inter_twin) + list(inter_0) + list(inter_1)\n",
    "        df_dist = pd.DataFrame(data=np.array([category,distances]).T, columns=['category','distance'])\n",
    "        df_dist['distance'] = df_dist['distance'].astype(float)\n",
    "        df_dist.to_csv(path + 'outputs/Qiime2_' + g + '/' + sg + '_inter_intra_beta_dist.tsv',sep='\\t')\n",
    "                             \n",
    "        sns.boxplot(data=df_dist, x='category', y='distance')\n",
    "        sns.swarmplot(data=df_dist, x='category', y='distance', color='black')\n",
    "        sns.despine()\n",
    "    \n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/Qiime2_' + g + '/' + sg + '_beta.pdf')\n",
    "        plt.close()          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f2ebdb4d-0d17-4c0d-b282-cd78e27f7bb7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectID</th>\n",
       "      <th>Timepoints</th>\n",
       "      <th>shannon_entropy</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>WOMAC_P_Response</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>OAD001</td>\n",
       "      <td>pre</td>\n",
       "      <td>5.097506</td>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>OAD001</td>\n",
       "      <td>post</td>\n",
       "      <td>5.838205</td>\n",
       "      <td>post</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>OAD003</td>\n",
       "      <td>pre</td>\n",
       "      <td>5.190428</td>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>OAD003</td>\n",
       "      <td>post</td>\n",
       "      <td>4.679640</td>\n",
       "      <td>post</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>OAD004</td>\n",
       "      <td>pre</td>\n",
       "      <td>5.299560</td>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SubjectID Timepoints  shannon_entropy Timepoint  \\\n",
       "OAD-001.pre.stool     OAD001        pre         5.097506       pre   \n",
       "OAD-001.post.stool    OAD001       post         5.838205      post   \n",
       "OAD-003.pre.stool     OAD003        pre         5.190428       pre   \n",
       "OAD-003.post.stool    OAD003       post         4.679640      post   \n",
       "OAD-004.pre.stool     OAD004        pre         5.299560       pre   \n",
       "\n",
       "                   WOMAC_P_Response  \n",
       "OAD-001.pre.stool       No response  \n",
       "OAD-001.post.stool      No response  \n",
       "OAD-003.pre.stool       No response  \n",
       "OAD-003.post.stool      No response  \n",
       "OAD-004.pre.stool       No response  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# alpha plots\n",
    "for g in ['saliva','stool']:\n",
    "    df_alpha = g_to_dfd[g + '_adh']['alpha']\n",
    "    df_map = type_to_df_map[g].copy()\n",
    "    df_map = df_map.set_index('Together')\n",
    "    df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.g')[0])\n",
    "    \n",
    "    df_merge = pd.concat([df_alpha, df_map], axis=1)\n",
    "    df_merge = df_merge.dropna()\n",
    "    \n",
    "    # sns.set_theme()\n",
    "    custom_params = {\"axes.spines.right\": False, \"axes.spines.top\": False}\n",
    "    sns.set_theme(style=\"ticks\", rc=custom_params)\n",
    "    plt.figure(figsize=(6,4))\n",
    "    sns.boxplot(x='Timepoint', y='shannon_entropy', data=df_merge,palette=['b','r'], boxprops=dict(alpha=0.5))#, hue='Timepoint', palette=['b','r'])\n",
    "    ax = sns.swarmplot(x='Timepoint', y='shannon_entropy', data=df_merge,color='black')\n",
    "    for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "        ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "    sns.despine()#(left=True, bottom=True)\n",
    "    \n",
    "    # ax.axes.set_title(\"\",fontsize=48)\\n\",\n",
    "    plt.tight_layout()\n",
    "    ax.set_ylabel(\"Shannon Entropy\",fontsize=16)\n",
    "    ax.set_xlabel(\"Diagnosis\",fontsize=16)\n",
    "    ax.tick_params(labelsize=16)\n",
    "    \n",
    "    # plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/Qiime2_' + g + '_adh/alpha_all.pdf')\n",
    "    plt.close()\n",
    "\n",
    "df_merge.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51757cac-d364-49d9-88b5-d27af66b38fd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Process Taxa DF ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8d8a7f96-5705-47fb-8d64-6ddcc4cd5d72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;__;__;__;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>0.014084</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.022027</td>\n",
       "      <td>0.001515</td>\n",
       "      <td>0.173283</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021230</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.152452</td>\n",
       "      <td>0.033001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.022576</td>\n",
       "      <td>0.007098</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.292202</td>\n",
       "      <td>0.001994</td>\n",
       "      <td>0.013555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000783</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>0.273829</td>\n",
       "      <td>0.067421</td>\n",
       "      <td>0.184572</td>\n",
       "      <td>0.026929</td>\n",
       "      <td>0.053673</td>\n",
       "      <td>0.037783</td>\n",
       "      <td>0.009619</td>\n",
       "      <td>0.023935</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.012060</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.181156</td>\n",
       "      <td>0.067844</td>\n",
       "      <td>0.150895</td>\n",
       "      <td>0.063229</td>\n",
       "      <td>0.051360</td>\n",
       "      <td>0.055600</td>\n",
       "      <td>0.012363</td>\n",
       "      <td>0.051513</td>\n",
       "      <td>0.006045</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.post.stool</th>\n",
       "      <td>0.174773</td>\n",
       "      <td>0.040785</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001855</td>\n",
       "      <td>0.124402</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001400</td>\n",
       "      <td>0.000371</td>\n",
       "      <td>0.014265</td>\n",
       "      <td>0.090594</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 374 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004  \\\n",
       "index                                                                                                                       \n",
       "OAD-001.post.stool                                           0.014084                                                       \n",
       "OAD-001.pre.stool                                            0.022576                                                       \n",
       "OAD-003.post.stool                                           0.273829                                                       \n",
       "OAD-003.pre.stool                                            0.181156                                                       \n",
       "OAD-004.post.stool                                           0.174773                                                       \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium  \\\n",
       "index                                                                                                                            \n",
       "OAD-001.post.stool                                           0.039046                                                            \n",
       "OAD-001.pre.stool                                            0.007098                                                            \n",
       "OAD-003.post.stool                                           0.067421                                                            \n",
       "OAD-003.pre.stool                                            0.067844                                                            \n",
       "OAD-004.post.stool                                           0.040785                                                            \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella  \\\n",
       "index                                                                                                             \n",
       "OAD-001.post.stool                                           0.022027                                             \n",
       "OAD-001.pre.stool                                            0.005128                                             \n",
       "OAD-003.post.stool                                           0.184572                                             \n",
       "OAD-003.pre.stool                                            0.150895                                             \n",
       "OAD-004.post.stool                                           0.000000                                             \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117  \\\n",
       "index                                                                                                                              \n",
       "OAD-001.post.stool                                           0.001515                                                              \n",
       "OAD-001.pre.stool                                            0.000000                                                              \n",
       "OAD-003.post.stool                                           0.026929                                                              \n",
       "OAD-003.pre.stool                                            0.063229                                                              \n",
       "OAD-004.post.stool                                           0.001855                                                              \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H  \\\n",
       "index                                                                                                                \n",
       "OAD-001.post.stool                                           0.173283                                                \n",
       "OAD-001.pre.stool                                            0.292202                                                \n",
       "OAD-003.post.stool                                           0.053673                                                \n",
       "OAD-003.pre.stool                                            0.051360                                                \n",
       "OAD-004.post.stool                                           0.124402                                                \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum  \\\n",
       "index                                                                                                                  \n",
       "OAD-001.post.stool                                           0.000000                                                  \n",
       "OAD-001.pre.stool                                            0.001994                                                  \n",
       "OAD-003.post.stool                                           0.037783                                                  \n",
       "OAD-003.pre.stool                                            0.055600                                                  \n",
       "OAD-004.post.stool                                           0.000000                                                  \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129  \\\n",
       "index                                                                                                                            \n",
       "OAD-001.post.stool                                           0.021230                                                            \n",
       "OAD-001.pre.stool                                            0.013555                                                            \n",
       "OAD-003.post.stool                                           0.009619                                                            \n",
       "OAD-003.pre.stool                                            0.012363                                                            \n",
       "OAD-004.post.stool                                           0.001400                                                            \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus  \\\n",
       "index                                                                                                                \n",
       "OAD-001.post.stool                                           0.000160                                                \n",
       "OAD-001.pre.stool                                            0.000000                                                \n",
       "OAD-003.post.stool                                           0.023935                                                \n",
       "OAD-003.pre.stool                                            0.051513                                                \n",
       "OAD-004.post.stool                                           0.000371                                                \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.post.stool                                           0.152452                                                        \n",
       "OAD-001.pre.stool                                            0.000783                                                        \n",
       "OAD-003.post.stool                                           0.005789                                                        \n",
       "OAD-003.pre.stool                                            0.006045                                                        \n",
       "OAD-004.post.stool                                           0.014265                                                        \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400  \\\n",
       "index                                                                                                                    \n",
       "OAD-001.post.stool                                           0.033001                                                    \n",
       "OAD-001.pre.stool                                            0.026849                                                    \n",
       "OAD-003.post.stool                                           0.012060                                                    \n",
       "OAD-003.pre.stool                                            0.025918                                                    \n",
       "OAD-004.post.stool                                           0.090594                                                    \n",
       "\n",
       "                    ...  d__Bacteria;p__Firmicutes_A;__;__;__;__  \\\n",
       "index               ...                                            \n",
       "OAD-001.post.stool  ...                                      0.0   \n",
       "OAD-001.pre.stool   ...                                      0.0   \n",
       "OAD-003.post.stool  ...                                      0.0   \n",
       "OAD-003.pre.stool   ...                                      0.0   \n",
       "OAD-004.post.stool  ...                                      0.0   \n",
       "\n",
       "                    d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.post.stool                                                0.0                                                        \n",
       "OAD-001.pre.stool                                                 0.0                                                        \n",
       "OAD-003.post.stool                                                0.0                                                        \n",
       "OAD-003.pre.stool                                                 0.0                                                        \n",
       "OAD-004.post.stool                                                0.0                                                        \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "index                                                                                 \n",
       "OAD-001.post.stool                                                0.0                 \n",
       "OAD-001.pre.stool                                                 0.0                 \n",
       "OAD-003.post.stool                                                0.0                 \n",
       "OAD-003.pre.stool                                                 0.0                 \n",
       "OAD-004.post.stool                                                0.0                 \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "index                                                                                                                         \n",
       "OAD-001.post.stool                                                0.0                                                         \n",
       "OAD-001.pre.stool                                                 0.0                                                         \n",
       "OAD-003.post.stool                                                0.0                                                         \n",
       "OAD-003.pre.stool                                                 0.0                                                         \n",
       "OAD-004.post.stool                                                0.0                                                         \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "index                                                                                                               \n",
       "OAD-001.post.stool                                                0.0                                               \n",
       "OAD-001.pre.stool                                                 0.0                                               \n",
       "OAD-003.post.stool                                                0.0                                               \n",
       "OAD-003.pre.stool                                                 0.0                                               \n",
       "OAD-004.post.stool                                                0.0                                               \n",
       "\n",
       "                    d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "index                                                                                                                                         \n",
       "OAD-001.post.stool                                                0.0                                                                         \n",
       "OAD-001.pre.stool                                                 0.0                                                                         \n",
       "OAD-003.post.stool                                                0.0                                                                         \n",
       "OAD-003.pre.stool                                                 0.0                                                                         \n",
       "OAD-004.post.stool                                                0.0                                                                         \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "index                                                                                                     \n",
       "OAD-001.post.stool                                                0.0                                     \n",
       "OAD-001.pre.stool                                                 0.0                                     \n",
       "OAD-003.post.stool                                                0.0                                     \n",
       "OAD-003.pre.stool                                                 0.0                                     \n",
       "OAD-004.post.stool                                                0.0                                     \n",
       "\n",
       "                    d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "index                                                                                                                                        \n",
       "OAD-001.post.stool                                                0.0                                                                        \n",
       "OAD-001.pre.stool                                                 0.0                                                                        \n",
       "OAD-003.post.stool                                                0.0                                                                        \n",
       "OAD-003.pre.stool                                                 0.0                                                                        \n",
       "OAD-004.post.stool                                                0.0                                                                        \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "index                                                                                                  \n",
       "OAD-001.post.stool                                                0.0                                  \n",
       "OAD-001.pre.stool                                                 0.0                                  \n",
       "OAD-003.post.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                 0.0                                  \n",
       "OAD-004.post.stool                                                0.0                                  \n",
       "\n",
       "                    d__Bacteria;p__Actinobacteriota;__;__;__;__  \n",
       "index                                                            \n",
       "OAD-001.post.stool                                          0.0  \n",
       "OAD-001.pre.stool                                           0.0  \n",
       "OAD-003.post.stool                                          0.0  \n",
       "OAD-003.pre.stool                                           0.0  \n",
       "OAD-004.post.stool                                          0.0  \n",
       "\n",
       "[5 rows x 374 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specific taxa profiling\n",
    "GL_to_otu = {}\n",
    "\n",
    "groups = ['saliva_adh', 'stool_adh']\n",
    "levels = ['6','7']\n",
    "for g in groups:\n",
    "    GL_to_otu[g] = {}\n",
    "    for l in levels:\n",
    "        # read in csv\n",
    "        df_otu = pd.read_csv(path + 'outputs/Qiime2_' + g + '/level-' + l + '.csv', index_col=0)\n",
    "\n",
    "        # drop specific taxa\n",
    "        # if g == 'saliva_adh':\n",
    "        #    df_otu = df_otu.drop('VAOAD-004.post.saliva')\n",
    "        \n",
    "        # filter on OTU only \n",
    "        keep = []\n",
    "        for x in df_otu.columns:\n",
    "            if 'd__' in x:\n",
    "                keep.append(x)\n",
    "        df_otu = df_otu.loc[:,keep]\n",
    "\n",
    "        # normalize by dividing by row sum\n",
    "        df_otu = df_otu.div(df_otu.sum(axis=1),axis=0)\n",
    "        \n",
    "        # save to dict\n",
    "        GL_to_otu[g]['L' + l] = df_otu\n",
    "\n",
    "                                       \n",
    "GL_to_otu['stool_adh']['L6'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "515d9fc7-6c65-439d-8733-6c92c35fc5b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;__;__;__;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.022576</td>\n",
       "      <td>0.007098</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.292202</td>\n",
       "      <td>0.001994</td>\n",
       "      <td>0.013555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000783</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.181156</td>\n",
       "      <td>0.067844</td>\n",
       "      <td>0.150895</td>\n",
       "      <td>0.063229</td>\n",
       "      <td>0.051360</td>\n",
       "      <td>0.055600</td>\n",
       "      <td>0.012363</td>\n",
       "      <td>0.051513</td>\n",
       "      <td>0.006045</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.170334</td>\n",
       "      <td>0.135468</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002088</td>\n",
       "      <td>0.173617</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003649</td>\n",
       "      <td>0.008765</td>\n",
       "      <td>0.054802</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-005.pre.stool</th>\n",
       "      <td>0.040108</td>\n",
       "      <td>0.139385</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018806</td>\n",
       "      <td>0.155290</td>\n",
       "      <td>0.037507</td>\n",
       "      <td>0.024457</td>\n",
       "      <td>0.066852</td>\n",
       "      <td>0.008969</td>\n",
       "      <td>0.018328</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-006.pre.stool</th>\n",
       "      <td>0.037219</td>\n",
       "      <td>0.167448</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007940</td>\n",
       "      <td>0.019591</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011006</td>\n",
       "      <td>0.001275</td>\n",
       "      <td>0.012769</td>\n",
       "      <td>0.162962</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 374 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004  \\\n",
       "index                                                                                                                      \n",
       "OAD-001.pre.stool                                           0.022576                                                       \n",
       "OAD-003.pre.stool                                           0.181156                                                       \n",
       "OAD-004.pre.stool                                           0.170334                                                       \n",
       "OAD-005.pre.stool                                           0.040108                                                       \n",
       "OAD-006.pre.stool                                           0.037219                                                       \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium  \\\n",
       "index                                                                                                                           \n",
       "OAD-001.pre.stool                                           0.007098                                                            \n",
       "OAD-003.pre.stool                                           0.067844                                                            \n",
       "OAD-004.pre.stool                                           0.135468                                                            \n",
       "OAD-005.pre.stool                                           0.139385                                                            \n",
       "OAD-006.pre.stool                                           0.167448                                                            \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella  \\\n",
       "index                                                                                                            \n",
       "OAD-001.pre.stool                                           0.005128                                             \n",
       "OAD-003.pre.stool                                           0.150895                                             \n",
       "OAD-004.pre.stool                                           0.000000                                             \n",
       "OAD-005.pre.stool                                           0.000000                                             \n",
       "OAD-006.pre.stool                                           0.000000                                             \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117  \\\n",
       "index                                                                                                                             \n",
       "OAD-001.pre.stool                                           0.000000                                                              \n",
       "OAD-003.pre.stool                                           0.063229                                                              \n",
       "OAD-004.pre.stool                                           0.002088                                                              \n",
       "OAD-005.pre.stool                                           0.018806                                                              \n",
       "OAD-006.pre.stool                                           0.007940                                                              \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H  \\\n",
       "index                                                                                                               \n",
       "OAD-001.pre.stool                                           0.292202                                                \n",
       "OAD-003.pre.stool                                           0.051360                                                \n",
       "OAD-004.pre.stool                                           0.173617                                                \n",
       "OAD-005.pre.stool                                           0.155290                                                \n",
       "OAD-006.pre.stool                                           0.019591                                                \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum  \\\n",
       "index                                                                                                                 \n",
       "OAD-001.pre.stool                                           0.001994                                                  \n",
       "OAD-003.pre.stool                                           0.055600                                                  \n",
       "OAD-004.pre.stool                                           0.000143                                                  \n",
       "OAD-005.pre.stool                                           0.037507                                                  \n",
       "OAD-006.pre.stool                                           0.000000                                                  \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129  \\\n",
       "index                                                                                                                           \n",
       "OAD-001.pre.stool                                           0.013555                                                            \n",
       "OAD-003.pre.stool                                           0.012363                                                            \n",
       "OAD-004.pre.stool                                           0.000000                                                            \n",
       "OAD-005.pre.stool                                           0.024457                                                            \n",
       "OAD-006.pre.stool                                           0.011006                                                            \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus  \\\n",
       "index                                                                                                               \n",
       "OAD-001.pre.stool                                           0.000000                                                \n",
       "OAD-003.pre.stool                                           0.051513                                                \n",
       "OAD-004.pre.stool                                           0.003649                                                \n",
       "OAD-005.pre.stool                                           0.066852                                                \n",
       "OAD-006.pre.stool                                           0.001275                                                \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor  \\\n",
       "index                                                                                                                       \n",
       "OAD-001.pre.stool                                           0.000783                                                        \n",
       "OAD-003.pre.stool                                           0.006045                                                        \n",
       "OAD-004.pre.stool                                           0.008765                                                        \n",
       "OAD-005.pre.stool                                           0.008969                                                        \n",
       "OAD-006.pre.stool                                           0.012769                                                        \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400  \\\n",
       "index                                                                                                                   \n",
       "OAD-001.pre.stool                                           0.026849                                                    \n",
       "OAD-003.pre.stool                                           0.025918                                                    \n",
       "OAD-004.pre.stool                                           0.054802                                                    \n",
       "OAD-005.pre.stool                                           0.018328                                                    \n",
       "OAD-006.pre.stool                                           0.162962                                                    \n",
       "\n",
       "                   ...  d__Bacteria;p__Firmicutes_A;__;__;__;__  \\\n",
       "index              ...                                            \n",
       "OAD-001.pre.stool  ...                                      0.0   \n",
       "OAD-003.pre.stool  ...                                      0.0   \n",
       "OAD-004.pre.stool  ...                                      0.0   \n",
       "OAD-005.pre.stool  ...                                      0.0   \n",
       "OAD-006.pre.stool  ...                                      0.0   \n",
       "\n",
       "                   d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "index                                                                                                                       \n",
       "OAD-001.pre.stool                                                0.0                                                        \n",
       "OAD-003.pre.stool                                                0.0                                                        \n",
       "OAD-004.pre.stool                                                0.0                                                        \n",
       "OAD-005.pre.stool                                                0.0                                                        \n",
       "OAD-006.pre.stool                                                0.0                                                        \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "index                                                                                \n",
       "OAD-001.pre.stool                                                0.0                 \n",
       "OAD-003.pre.stool                                                0.0                 \n",
       "OAD-004.pre.stool                                                0.0                 \n",
       "OAD-005.pre.stool                                                0.0                 \n",
       "OAD-006.pre.stool                                                0.0                 \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.pre.stool                                                0.0                                                         \n",
       "OAD-003.pre.stool                                                0.0                                                         \n",
       "OAD-004.pre.stool                                                0.0                                                         \n",
       "OAD-005.pre.stool                                                0.0                                                         \n",
       "OAD-006.pre.stool                                                0.0                                                         \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "index                                                                                                              \n",
       "OAD-001.pre.stool                                                0.0                                               \n",
       "OAD-003.pre.stool                                                0.0                                               \n",
       "OAD-004.pre.stool                                                0.0                                               \n",
       "OAD-005.pre.stool                                                0.0                                               \n",
       "OAD-006.pre.stool                                                0.0                                               \n",
       "\n",
       "                   d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "index                                                                                                                                        \n",
       "OAD-001.pre.stool                                                0.0                                                                         \n",
       "OAD-003.pre.stool                                                0.0                                                                         \n",
       "OAD-004.pre.stool                                                0.0                                                                         \n",
       "OAD-005.pre.stool                                                0.0                                                                         \n",
       "OAD-006.pre.stool                                                0.0                                                                         \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "index                                                                                                    \n",
       "OAD-001.pre.stool                                                0.0                                     \n",
       "OAD-003.pre.stool                                                0.0                                     \n",
       "OAD-004.pre.stool                                                0.0                                     \n",
       "OAD-005.pre.stool                                                0.0                                     \n",
       "OAD-006.pre.stool                                                0.0                                     \n",
       "\n",
       "                   d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "index                                                                                                                                       \n",
       "OAD-001.pre.stool                                                0.0                                                                        \n",
       "OAD-003.pre.stool                                                0.0                                                                        \n",
       "OAD-004.pre.stool                                                0.0                                                                        \n",
       "OAD-005.pre.stool                                                0.0                                                                        \n",
       "OAD-006.pre.stool                                                0.0                                                                        \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "index                                                                                                 \n",
       "OAD-001.pre.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                0.0                                  \n",
       "OAD-004.pre.stool                                                0.0                                  \n",
       "OAD-005.pre.stool                                                0.0                                  \n",
       "OAD-006.pre.stool                                                0.0                                  \n",
       "\n",
       "                   d__Bacteria;p__Actinobacteriota;__;__;__;__  \n",
       "index                                                           \n",
       "OAD-001.pre.stool                                          0.0  \n",
       "OAD-003.pre.stool                                          0.0  \n",
       "OAD-004.pre.stool                                          0.0  \n",
       "OAD-005.pre.stool                                          0.0  \n",
       "OAD-006.pre.stool                                          0.0  \n",
       "\n",
       "[5 rows x 374 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Microbiome testing\n",
    "gt_to_otu = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gt_to_otu[g] = {}\n",
    "    df_otu = GL_to_otu[g + '_adh']['L6'].copy()\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['post','pre']:\n",
    "        df = df_otu[df_otu.index.to_series().str.contains(t)]\n",
    "        gt_to_otu[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_otu[g]['pre'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.split('.pre')[0])\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('-',''))\n",
    "    post_df = gt_to_otu[g]['post'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.split('.post')[0])\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('-',''))\n",
    "    \n",
    "    gt_to_otu[g]['diff'] = post_df - pre_df\n",
    "    gt_to_otu[g]['all'] = pd.concat([gt_to_otu[g]['pre'],gt_to_otu[g]['post']])\n",
    "\n",
    "gt_to_otu[g][t].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "24c44852-2e85-4174-9729-3fae6a355b32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# taxa parse function\n",
    "def read_taxa(taxa, delim=';'):\n",
    "    \"\"\"\n",
    "    Converts string of OTU names (e.g. from QIIME) to shortened form.\n",
    "    ----------------------------------------------------------------------------\n",
    "    INPUTS\n",
    "    taxa  - String. Long name of OTU e.g. 'k__Archaea;p__Crenarchaeota;\n",
    "            c__Thaumarchaeota;o__Cenarchaeales;f__Cenarchaeaceae;\n",
    "            g__Nitrosopumilus' which will become 'Cenarchaeaceae Nitrosopumilus'\n",
    "    delim - String. Separates heirarchy.\n",
    "\n",
    "    OUTPUTS\n",
    "    String. Abridged taxa name.\n",
    "    \"\"\"\n",
    "    parts = taxa.split(delim) # set as param with default\n",
    "    while parts:\n",
    "        if len(parts) == 1:\n",
    "            return('u.c. ' + parts[0])\n",
    "        if not parts[-1].endswith('__'):\n",
    "            taxastr1 = parts[-2].split('__')[1]\n",
    "            taxastr2 = parts[-1].split('__')[1]\n",
    "            return taxastr1 + ' ' + taxastr2\n",
    "        else:\n",
    "            parts.pop()\n",
    "\n",
    "    # This should not be reached: \"k__;p__...\"\n",
    "    return 'Uncharacterized'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faf9a2fe-9f68-4cee-9e40-4d5b749b9f0e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Process Metabolite DF ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6bf1fdd5-b592-4a4e-a5d4-3cefc2d28c9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_N-Acetylmuramic-Acid</th>\n",
       "      <th>H_D_Fructose</th>\n",
       "      <th>H_Paraxanthine</th>\n",
       "      <th>H_L-methionine</th>\n",
       "      <th>H_Xanthuneric_acid</th>\n",
       "      <th>H_Pyridoxic_acid</th>\n",
       "      <th>H_p_coumaric_acid</th>\n",
       "      <th>H_Lactose</th>\n",
       "      <th>H_Pantothenic_acid</th>\n",
       "      <th>...</th>\n",
       "      <th>H_Glycoursodeoxycholic_acid</th>\n",
       "      <th>H_Indole_3_propionic_acid</th>\n",
       "      <th>H_Yersiniabactin_Fe_adduct</th>\n",
       "      <th>H_Glycocholic_acid</th>\n",
       "      <th>H_Glycocholate</th>\n",
       "      <th>H_glycocholic_acid</th>\n",
       "      <th>H_Glycochenodeoxycholate</th>\n",
       "      <th>H_Deoxycholate</th>\n",
       "      <th>H_PE(22:6/0:0)_[M+H]+_C27H45N1O7P1</th>\n",
       "      <th>H_PE(18:2/0:0)_[M+H]+_C23H45N1O7P1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>0.035609</td>\n",
       "      <td>0.082371</td>\n",
       "      <td>0.379626</td>\n",
       "      <td>0.201167</td>\n",
       "      <td>-0.082362</td>\n",
       "      <td>0.016484</td>\n",
       "      <td>-0.414174</td>\n",
       "      <td>-0.718066</td>\n",
       "      <td>-0.344094</td>\n",
       "      <td>0.132369</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.126092</td>\n",
       "      <td>0.446683</td>\n",
       "      <td>-0.905441</td>\n",
       "      <td>0.448360</td>\n",
       "      <td>0.483734</td>\n",
       "      <td>0.562393</td>\n",
       "      <td>0.110112</td>\n",
       "      <td>0.591415</td>\n",
       "      <td>0.282179</td>\n",
       "      <td>0.304361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.003591</td>\n",
       "      <td>1.153334</td>\n",
       "      <td>0.448950</td>\n",
       "      <td>-1.240302</td>\n",
       "      <td>0.136791</td>\n",
       "      <td>-0.077398</td>\n",
       "      <td>-0.128398</td>\n",
       "      <td>-0.483053</td>\n",
       "      <td>-0.726048</td>\n",
       "      <td>-1.082906</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.512609</td>\n",
       "      <td>0.191755</td>\n",
       "      <td>-0.670443</td>\n",
       "      <td>-0.369924</td>\n",
       "      <td>-0.382214</td>\n",
       "      <td>-1.063725</td>\n",
       "      <td>-1.327175</td>\n",
       "      <td>-0.496949</td>\n",
       "      <td>0.380516</td>\n",
       "      <td>0.421219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>0.063048</td>\n",
       "      <td>0.145153</td>\n",
       "      <td>-0.073655</td>\n",
       "      <td>0.421159</td>\n",
       "      <td>-0.232476</td>\n",
       "      <td>-0.577305</td>\n",
       "      <td>-0.208879</td>\n",
       "      <td>0.117236</td>\n",
       "      <td>-0.309148</td>\n",
       "      <td>-0.279103</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.118668</td>\n",
       "      <td>-0.014176</td>\n",
       "      <td>1.399191</td>\n",
       "      <td>-0.768953</td>\n",
       "      <td>-0.781243</td>\n",
       "      <td>-1.462742</td>\n",
       "      <td>-0.674522</td>\n",
       "      <td>-0.895976</td>\n",
       "      <td>0.380375</td>\n",
       "      <td>0.305410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>-0.142699</td>\n",
       "      <td>0.625381</td>\n",
       "      <td>0.140133</td>\n",
       "      <td>1.029497</td>\n",
       "      <td>-0.014348</td>\n",
       "      <td>-0.021789</td>\n",
       "      <td>-0.385252</td>\n",
       "      <td>0.326122</td>\n",
       "      <td>-0.530950</td>\n",
       "      <td>0.034178</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.660924</td>\n",
       "      <td>0.202382</td>\n",
       "      <td>1.602954</td>\n",
       "      <td>-0.760978</td>\n",
       "      <td>0.001860</td>\n",
       "      <td>0.136519</td>\n",
       "      <td>-0.178280</td>\n",
       "      <td>-0.888002</td>\n",
       "      <td>-0.054682</td>\n",
       "      <td>-0.804769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD007</th>\n",
       "      <td>0.340004</td>\n",
       "      <td>0.957297</td>\n",
       "      <td>0.276825</td>\n",
       "      <td>0.384845</td>\n",
       "      <td>0.148415</td>\n",
       "      <td>-0.034565</td>\n",
       "      <td>0.136779</td>\n",
       "      <td>0.969607</td>\n",
       "      <td>0.163616</td>\n",
       "      <td>-1.061064</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.891056</td>\n",
       "      <td>0.117900</td>\n",
       "      <td>1.199539</td>\n",
       "      <td>-0.348083</td>\n",
       "      <td>-0.360373</td>\n",
       "      <td>-1.041883</td>\n",
       "      <td>-0.170514</td>\n",
       "      <td>-0.475108</td>\n",
       "      <td>0.185225</td>\n",
       "      <td>-0.391874</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          H_Pyridoxamine  H_N-Acetylmuramic-Acid  H_D_Fructose  \\\n",
       "Study_ID                                                         \n",
       "OAD006          0.035609                0.082371      0.379626   \n",
       "OAD001          0.003591                1.153334      0.448950   \n",
       "OAD003          0.063048                0.145153     -0.073655   \n",
       "OAD005         -0.142699                0.625381      0.140133   \n",
       "OAD007          0.340004                0.957297      0.276825   \n",
       "\n",
       "          H_Paraxanthine  H_L-methionine  H_Xanthuneric_acid  \\\n",
       "Study_ID                                                       \n",
       "OAD006          0.201167       -0.082362            0.016484   \n",
       "OAD001         -1.240302        0.136791           -0.077398   \n",
       "OAD003          0.421159       -0.232476           -0.577305   \n",
       "OAD005          1.029497       -0.014348           -0.021789   \n",
       "OAD007          0.384845        0.148415           -0.034565   \n",
       "\n",
       "          H_Pyridoxic_acid  H_p_coumaric_acid  H_Lactose  H_Pantothenic_acid  \\\n",
       "Study_ID                                                                       \n",
       "OAD006           -0.414174          -0.718066  -0.344094            0.132369   \n",
       "OAD001           -0.128398          -0.483053  -0.726048           -1.082906   \n",
       "OAD003           -0.208879           0.117236  -0.309148           -0.279103   \n",
       "OAD005           -0.385252           0.326122  -0.530950            0.034178   \n",
       "OAD007            0.136779           0.969607   0.163616           -1.061064   \n",
       "\n",
       "          ...  H_Glycoursodeoxycholic_acid  H_Indole_3_propionic_acid  \\\n",
       "Study_ID  ...                                                           \n",
       "OAD006    ...                    -0.126092                   0.446683   \n",
       "OAD001    ...                    -0.512609                   0.191755   \n",
       "OAD003    ...                    -0.118668                  -0.014176   \n",
       "OAD005    ...                    -0.660924                   0.202382   \n",
       "OAD007    ...                    -0.891056                   0.117900   \n",
       "\n",
       "          H_Yersiniabactin_Fe_adduct  H_Glycocholic_acid  H_Glycocholate  \\\n",
       "Study_ID                                                                   \n",
       "OAD006                     -0.905441            0.448360        0.483734   \n",
       "OAD001                     -0.670443           -0.369924       -0.382214   \n",
       "OAD003                      1.399191           -0.768953       -0.781243   \n",
       "OAD005                      1.602954           -0.760978        0.001860   \n",
       "OAD007                      1.199539           -0.348083       -0.360373   \n",
       "\n",
       "          H_glycocholic_acid  H_Glycochenodeoxycholate  H_Deoxycholate  \\\n",
       "Study_ID                                                                 \n",
       "OAD006              0.562393                  0.110112        0.591415   \n",
       "OAD001             -1.063725                 -1.327175       -0.496949   \n",
       "OAD003             -1.462742                 -0.674522       -0.895976   \n",
       "OAD005              0.136519                 -0.178280       -0.888002   \n",
       "OAD007             -1.041883                 -0.170514       -0.475108   \n",
       "\n",
       "          H_PE(22:6/0:0)_[M+H]+_C27H45N1O7P1  \\\n",
       "Study_ID                                       \n",
       "OAD006                              0.282179   \n",
       "OAD001                              0.380516   \n",
       "OAD003                              0.380375   \n",
       "OAD005                             -0.054682   \n",
       "OAD007                              0.185225   \n",
       "\n",
       "          H_PE(18:2/0:0)_[M+H]+_C23H45N1O7P1  \n",
       "Study_ID                                      \n",
       "OAD006                              0.304361  \n",
       "OAD001                              0.421219  \n",
       "OAD003                              0.305410  \n",
       "OAD005                             -0.804769  \n",
       "OAD007                             -0.391874  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Metabolome testing\n",
    "gdt_to_df = {}\n",
    "for g in ['saliva','stool','plasma']:\n",
    "    gdt_to_df[g] = {}\n",
    "    for d in ['paired_beta']: # just arbitrarily \n",
    "        gdt_to_df[g][d] = {}\n",
    "        df_meta = pd.read_csv(path + 'inputs/' + g + '_normalized.csv')\n",
    "\n",
    "        # split on timepoint\n",
    "        for t in ['Baseline','After diet']:\n",
    "            df = df_meta[df_meta['Time'] == t]\n",
    "            df = df.set_index('Study_ID')\n",
    "            df.index = df.index.map(lambda x: x.split('_')[0])\n",
    "            df = df.drop('Time',axis=1)\n",
    "            gdt_to_df[g][d][t] = df\n",
    "\n",
    "        gdt_to_df[g][d]['diff'] = gdt_to_df[g][d]['After diet'] - gdt_to_df[g][d]['Baseline']\n",
    "\n",
    "gdt_to_df[g][d][t].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "dbad4aa2-7c35-490c-900b-c607c4ff549d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_N-Acetylmuramic-Acid</th>\n",
       "      <th>H_D_Fructose</th>\n",
       "      <th>H_Paraxanthine</th>\n",
       "      <th>H_L-methionine</th>\n",
       "      <th>H_Xanthuneric_acid</th>\n",
       "      <th>H_Pyridoxic_acid</th>\n",
       "      <th>H_p_coumaric_acid</th>\n",
       "      <th>H_Lactose</th>\n",
       "      <th>H_Pantothenic_acid</th>\n",
       "      <th>...</th>\n",
       "      <th>H_Glycoursodeoxycholic_acid</th>\n",
       "      <th>H_Indole_3_propionic_acid</th>\n",
       "      <th>H_Yersiniabactin_Fe_adduct</th>\n",
       "      <th>H_Glycocholic_acid</th>\n",
       "      <th>H_Glycocholate</th>\n",
       "      <th>H_glycocholic_acid</th>\n",
       "      <th>H_Glycochenodeoxycholate</th>\n",
       "      <th>H_Deoxycholate</th>\n",
       "      <th>H_PE(22:6/0:0)_[M+H]+_C27H45N1O7P1</th>\n",
       "      <th>H_PE(18:2/0:0)_[M+H]+_C23H45N1O7P1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>0.035609</td>\n",
       "      <td>0.082371</td>\n",
       "      <td>0.379626</td>\n",
       "      <td>0.201167</td>\n",
       "      <td>-0.082362</td>\n",
       "      <td>0.016484</td>\n",
       "      <td>-0.414174</td>\n",
       "      <td>-0.718066</td>\n",
       "      <td>-0.344094</td>\n",
       "      <td>0.132369</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.126092</td>\n",
       "      <td>0.446683</td>\n",
       "      <td>-0.905441</td>\n",
       "      <td>0.448360</td>\n",
       "      <td>0.483734</td>\n",
       "      <td>0.562393</td>\n",
       "      <td>0.110112</td>\n",
       "      <td>0.591415</td>\n",
       "      <td>0.282179</td>\n",
       "      <td>0.304361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.003591</td>\n",
       "      <td>1.153334</td>\n",
       "      <td>0.448950</td>\n",
       "      <td>-1.240302</td>\n",
       "      <td>0.136791</td>\n",
       "      <td>-0.077398</td>\n",
       "      <td>-0.128398</td>\n",
       "      <td>-0.483053</td>\n",
       "      <td>-0.726048</td>\n",
       "      <td>-1.082906</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.512609</td>\n",
       "      <td>0.191755</td>\n",
       "      <td>-0.670443</td>\n",
       "      <td>-0.369924</td>\n",
       "      <td>-0.382214</td>\n",
       "      <td>-1.063725</td>\n",
       "      <td>-1.327175</td>\n",
       "      <td>-0.496949</td>\n",
       "      <td>0.380516</td>\n",
       "      <td>0.421219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>0.063048</td>\n",
       "      <td>0.145153</td>\n",
       "      <td>-0.073655</td>\n",
       "      <td>0.421159</td>\n",
       "      <td>-0.232476</td>\n",
       "      <td>-0.577305</td>\n",
       "      <td>-0.208879</td>\n",
       "      <td>0.117236</td>\n",
       "      <td>-0.309148</td>\n",
       "      <td>-0.279103</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.118668</td>\n",
       "      <td>-0.014176</td>\n",
       "      <td>1.399191</td>\n",
       "      <td>-0.768953</td>\n",
       "      <td>-0.781243</td>\n",
       "      <td>-1.462742</td>\n",
       "      <td>-0.674522</td>\n",
       "      <td>-0.895976</td>\n",
       "      <td>0.380375</td>\n",
       "      <td>0.305410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>-0.142699</td>\n",
       "      <td>0.625381</td>\n",
       "      <td>0.140133</td>\n",
       "      <td>1.029497</td>\n",
       "      <td>-0.014348</td>\n",
       "      <td>-0.021789</td>\n",
       "      <td>-0.385252</td>\n",
       "      <td>0.326122</td>\n",
       "      <td>-0.530950</td>\n",
       "      <td>0.034178</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.660924</td>\n",
       "      <td>0.202382</td>\n",
       "      <td>1.602954</td>\n",
       "      <td>-0.760978</td>\n",
       "      <td>0.001860</td>\n",
       "      <td>0.136519</td>\n",
       "      <td>-0.178280</td>\n",
       "      <td>-0.888002</td>\n",
       "      <td>-0.054682</td>\n",
       "      <td>-0.804769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD007</th>\n",
       "      <td>0.340004</td>\n",
       "      <td>0.957297</td>\n",
       "      <td>0.276825</td>\n",
       "      <td>0.384845</td>\n",
       "      <td>0.148415</td>\n",
       "      <td>-0.034565</td>\n",
       "      <td>0.136779</td>\n",
       "      <td>0.969607</td>\n",
       "      <td>0.163616</td>\n",
       "      <td>-1.061064</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.891056</td>\n",
       "      <td>0.117900</td>\n",
       "      <td>1.199539</td>\n",
       "      <td>-0.348083</td>\n",
       "      <td>-0.360373</td>\n",
       "      <td>-1.041883</td>\n",
       "      <td>-0.170514</td>\n",
       "      <td>-0.475108</td>\n",
       "      <td>0.185225</td>\n",
       "      <td>-0.391874</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          H_Pyridoxamine  H_N-Acetylmuramic-Acid  H_D_Fructose  \\\n",
       "Study_ID                                                         \n",
       "OAD006          0.035609                0.082371      0.379626   \n",
       "OAD001          0.003591                1.153334      0.448950   \n",
       "OAD003          0.063048                0.145153     -0.073655   \n",
       "OAD005         -0.142699                0.625381      0.140133   \n",
       "OAD007          0.340004                0.957297      0.276825   \n",
       "\n",
       "          H_Paraxanthine  H_L-methionine  H_Xanthuneric_acid  \\\n",
       "Study_ID                                                       \n",
       "OAD006          0.201167       -0.082362            0.016484   \n",
       "OAD001         -1.240302        0.136791           -0.077398   \n",
       "OAD003          0.421159       -0.232476           -0.577305   \n",
       "OAD005          1.029497       -0.014348           -0.021789   \n",
       "OAD007          0.384845        0.148415           -0.034565   \n",
       "\n",
       "          H_Pyridoxic_acid  H_p_coumaric_acid  H_Lactose  H_Pantothenic_acid  \\\n",
       "Study_ID                                                                       \n",
       "OAD006           -0.414174          -0.718066  -0.344094            0.132369   \n",
       "OAD001           -0.128398          -0.483053  -0.726048           -1.082906   \n",
       "OAD003           -0.208879           0.117236  -0.309148           -0.279103   \n",
       "OAD005           -0.385252           0.326122  -0.530950            0.034178   \n",
       "OAD007            0.136779           0.969607   0.163616           -1.061064   \n",
       "\n",
       "          ...  H_Glycoursodeoxycholic_acid  H_Indole_3_propionic_acid  \\\n",
       "Study_ID  ...                                                           \n",
       "OAD006    ...                    -0.126092                   0.446683   \n",
       "OAD001    ...                    -0.512609                   0.191755   \n",
       "OAD003    ...                    -0.118668                  -0.014176   \n",
       "OAD005    ...                    -0.660924                   0.202382   \n",
       "OAD007    ...                    -0.891056                   0.117900   \n",
       "\n",
       "          H_Yersiniabactin_Fe_adduct  H_Glycocholic_acid  H_Glycocholate  \\\n",
       "Study_ID                                                                   \n",
       "OAD006                     -0.905441            0.448360        0.483734   \n",
       "OAD001                     -0.670443           -0.369924       -0.382214   \n",
       "OAD003                      1.399191           -0.768953       -0.781243   \n",
       "OAD005                      1.602954           -0.760978        0.001860   \n",
       "OAD007                      1.199539           -0.348083       -0.360373   \n",
       "\n",
       "          H_glycocholic_acid  H_Glycochenodeoxycholate  H_Deoxycholate  \\\n",
       "Study_ID                                                                 \n",
       "OAD006              0.562393                  0.110112        0.591415   \n",
       "OAD001             -1.063725                 -1.327175       -0.496949   \n",
       "OAD003             -1.462742                 -0.674522       -0.895976   \n",
       "OAD005              0.136519                 -0.178280       -0.888002   \n",
       "OAD007             -1.041883                 -0.170514       -0.475108   \n",
       "\n",
       "          H_PE(22:6/0:0)_[M+H]+_C27H45N1O7P1  \\\n",
       "Study_ID                                       \n",
       "OAD006                              0.282179   \n",
       "OAD001                              0.380516   \n",
       "OAD003                              0.380375   \n",
       "OAD005                             -0.054682   \n",
       "OAD007                              0.185225   \n",
       "\n",
       "          H_PE(18:2/0:0)_[M+H]+_C23H45N1O7P1  \n",
       "Study_ID                                      \n",
       "OAD006                              0.304361  \n",
       "OAD001                              0.421219  \n",
       "OAD003                              0.305410  \n",
       "OAD005                             -0.804769  \n",
       "OAD007                             -0.391874  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Metabolome testing\n",
    "gt_to_meta = {}\n",
    "for g in ['saliva','stool','plasma']:\n",
    "    gt_to_meta[g] = {}\n",
    "    df_meta = pd.read_csv(path + 'inputs/' + g + '_normalized.csv')\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['Baseline','After diet']:\n",
    "        df = df_meta[df_meta['Time'] == t]\n",
    "        df = df.set_index('Study_ID')\n",
    "        df.index = df.index.map(lambda x: x.split('_')[0])\n",
    "        df = df.drop('Time',axis=1)\n",
    "        gt_to_meta[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_meta[g]['Baseline'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('D0','D-0'))\n",
    "    pre_df.index = pre_df.index.map(lambda x: x + '.pre.' + g)\n",
    "    post_df = gt_to_meta[g]['After diet'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('D0','D-0'))\n",
    "    post_df.index = post_df.index.map(lambda x: x + '.post.' + g)\n",
    "    \n",
    "    gt_to_meta[g]['diff'] = gt_to_meta[g]['After diet'] - gt_to_meta[g]['Baseline']\n",
    "    gt_to_meta[g]['all'] = pd.concat([pre_df,post_df])\n",
    "    \n",
    "\n",
    "gt_to_meta[g][t].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "801b5fe1-2f5b-465e-b21a-943c574d1ee7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Pathway DF Processing ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8b418d85-6211-4131-9007-92d2b5f1a3ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - globo and isoglobo series</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid elongation</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of 12-, 14- and 16-membered macrolides</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of enediyne antibiotics</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.saliva</th>\n",
       "      <td>0.155479</td>\n",
       "      <td>0.025863</td>\n",
       "      <td>0.001866</td>\n",
       "      <td>0.003177</td>\n",
       "      <td>0.008013</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.026699</td>\n",
       "      <td>0.000547</td>\n",
       "      <td>0.031793</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>2.433861e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.216931e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.saliva</th>\n",
       "      <td>0.159679</td>\n",
       "      <td>0.027060</td>\n",
       "      <td>0.001949</td>\n",
       "      <td>0.003082</td>\n",
       "      <td>0.007239</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.027470</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.034357</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>1.911111e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.539150e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.saliva</th>\n",
       "      <td>0.151338</td>\n",
       "      <td>0.027415</td>\n",
       "      <td>0.002223</td>\n",
       "      <td>0.002702</td>\n",
       "      <td>0.009135</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.021911</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.030358</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.saliva</th>\n",
       "      <td>0.152707</td>\n",
       "      <td>0.026974</td>\n",
       "      <td>0.002172</td>\n",
       "      <td>0.002788</td>\n",
       "      <td>0.008887</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.022846</td>\n",
       "      <td>0.000257</td>\n",
       "      <td>0.031226</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.post.saliva</th>\n",
       "      <td>0.159885</td>\n",
       "      <td>0.027003</td>\n",
       "      <td>0.001989</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>0.007319</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.026641</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.034499</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>6.002574e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.594595e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 168 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "KEGG_Pathway         Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.post.saliva    0.155479                            0.025863   \n",
       "OAD-001.pre.saliva     0.159679                            0.027060   \n",
       "OAD-003.post.saliva    0.151338                            0.027415   \n",
       "OAD-003.pre.saliva     0.152707                            0.026974   \n",
       "OAD-004.post.saliva    0.159885                            0.027003   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.post.saliva                                           0.001866                 \n",
       "OAD-001.pre.saliva                                            0.001949                 \n",
       "OAD-003.post.saliva                                           0.002223                 \n",
       "OAD-003.pre.saliva                                            0.002172                 \n",
       "OAD-004.post.saliva                                           0.001989                 \n",
       "\n",
       "KEGG_Pathway         Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.post.saliva                                           0.003177        \n",
       "OAD-001.pre.saliva                                            0.003082        \n",
       "OAD-003.post.saliva                                           0.002702        \n",
       "OAD-003.pre.saliva                                            0.002788        \n",
       "OAD-004.post.saliva                                           0.002953        \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism  \\\n",
       "OAD-001.post.saliva                     0.008013   \n",
       "OAD-001.pre.saliva                      0.007239   \n",
       "OAD-003.post.saliva                     0.009135   \n",
       "OAD-003.pre.saliva                      0.008887   \n",
       "OAD-004.post.saliva                     0.007319   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.post.saliva                                           0.000273    \n",
       "OAD-001.pre.saliva                                            0.000166    \n",
       "OAD-003.post.saliva                                           0.000180    \n",
       "OAD-003.pre.saliva                                            0.000176    \n",
       "OAD-004.post.saliva                                           0.000132    \n",
       "\n",
       "KEGG_Pathway         Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.post.saliva                          0.026699   \n",
       "OAD-001.pre.saliva                           0.027470   \n",
       "OAD-003.post.saliva                          0.021911   \n",
       "OAD-003.pre.saliva                           0.022846   \n",
       "OAD-004.post.saliva                          0.026641   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.post.saliva                                           0.000547      \n",
       "OAD-001.pre.saliva                                            0.000309      \n",
       "OAD-003.post.saliva                                           0.000261      \n",
       "OAD-003.pre.saliva                                            0.000257      \n",
       "OAD-004.post.saliva                                           0.000200      \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.post.saliva                                         0.031793   \n",
       "OAD-001.pre.saliva                                          0.034357   \n",
       "OAD-003.post.saliva                                         0.030358   \n",
       "OAD-003.pre.saliva                                          0.031226   \n",
       "OAD-004.post.saliva                                         0.034499   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.post.saliva                                           0.000046                    \n",
       "OAD-001.pre.saliva                                            0.000045                    \n",
       "OAD-003.post.saliva                                           0.000042                    \n",
       "OAD-003.pre.saliva                                            0.000041                    \n",
       "OAD-004.post.saliva                                           0.000038                    \n",
       "\n",
       "KEGG_Pathway         ...  \\\n",
       "OAD-001.post.saliva  ...   \n",
       "OAD-001.pre.saliva   ...   \n",
       "OAD-003.post.saliva  ...   \n",
       "OAD-003.pre.saliva   ...   \n",
       "OAD-004.post.saliva  ...   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - globo and isoglobo series  \\\n",
       "OAD-001.post.saliva                                           0.000021                                                          \n",
       "OAD-001.pre.saliva                                            0.000045                                                          \n",
       "OAD-003.post.saliva                                           0.000103                                                          \n",
       "OAD-003.pre.saliva                                            0.000080                                                          \n",
       "OAD-004.post.saliva                                           0.000062                                                          \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism|Fatty acid elongation  \\\n",
       "OAD-001.post.saliva                                                0.0   \n",
       "OAD-001.pre.saliva                                                 0.0   \n",
       "OAD-003.post.saliva                                                0.0   \n",
       "OAD-003.pre.saliva                                                 0.0   \n",
       "OAD-004.post.saliva                                                0.0   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis  \\\n",
       "OAD-001.post.saliva                                           0.000189                                    \n",
       "OAD-001.pre.saliva                                            0.000095                                    \n",
       "OAD-003.post.saliva                                           0.000022                                    \n",
       "OAD-003.pre.saliva                                            0.000018                                    \n",
       "OAD-004.post.saliva                                           0.000039                                    \n",
       "\n",
       "KEGG_Pathway         Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series  \\\n",
       "OAD-001.post.saliva                                           0.000010                                               \n",
       "OAD-001.pre.saliva                                            0.000019                                               \n",
       "OAD-003.post.saliva                                           0.000060                                               \n",
       "OAD-003.pre.saliva                                            0.000043                                               \n",
       "OAD-004.post.saliva                                           0.000033                                               \n",
       "\n",
       "KEGG_Pathway         Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.post.saliva                                       2.433861e-07                                   \n",
       "OAD-001.pre.saliva                                        1.911111e-06                                   \n",
       "OAD-003.post.saliva                                       0.000000e+00                                   \n",
       "OAD-003.pre.saliva                                        0.000000e+00                                   \n",
       "OAD-004.post.saliva                                       6.002574e-07                                   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.post.saliva                                                0.0                                  \n",
       "OAD-001.pre.saliva                                                 0.0                                  \n",
       "OAD-003.post.saliva                                                0.0                                  \n",
       "OAD-003.pre.saliva                                                 0.0                                  \n",
       "OAD-004.post.saliva                                                0.0                                  \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of 12-, 14- and 16-membered macrolides  \\\n",
       "OAD-001.post.saliva                                                0.0                                                         \n",
       "OAD-001.pre.saliva                                                 0.0                                                         \n",
       "OAD-003.post.saliva                                                0.0                                                         \n",
       "OAD-003.pre.saliva                                                 0.0                                                         \n",
       "OAD-004.post.saliva                                                0.0                                                         \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of enediyne antibiotics  \\\n",
       "OAD-001.post.saliva                                                0.0                                          \n",
       "OAD-001.pre.saliva                                                 0.0                                          \n",
       "OAD-003.post.saliva                                                0.0                                          \n",
       "OAD-003.pre.saliva                                                 0.0                                          \n",
       "OAD-004.post.saliva                                                0.0                                          \n",
       "\n",
       "KEGG_Pathway         Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.post.saliva                                       0.000000e+00                           \n",
       "OAD-001.pre.saliva                                        0.000000e+00                           \n",
       "OAD-003.post.saliva                                       0.000000e+00                           \n",
       "OAD-003.pre.saliva                                        0.000000e+00                           \n",
       "OAD-004.post.saliva                                       8.594595e-07                           \n",
       "\n",
       "KEGG_Pathway         Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \n",
       "OAD-001.post.saliva                                       1.216931e-07                                         \n",
       "OAD-001.pre.saliva                                        1.539150e-07                                         \n",
       "OAD-003.post.saliva                                       0.000000e+00                                         \n",
       "OAD-003.pre.saliva                                        0.000000e+00                                         \n",
       "OAD-004.post.saliva                                       0.000000e+00                                         \n",
       "\n",
       "[5 rows x 168 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specific pathway profiling\n",
    "g_to_path = {}\n",
    "\n",
    "groups = ['saliva_adh', 'stool_adh']\n",
    "g_to_job = {\n",
    "    'saliva_adh': 'jobs07',\n",
    "    'stool_adh': 'jobs08'\n",
    "}\n",
    "for g in groups:\n",
    "    # read in csv\n",
    "    df_path = pd.read_csv(path + 'outputs/' + g_to_job[g] + '/pred_metagenome_unstrat_kegg.tsv', sep ='\\t', index_col=0)\n",
    "\n",
    "    # transpose\n",
    "    df_path = df_path.T\n",
    "    \n",
    "    # normalize by dividing by row sum\n",
    "    df_path = df_path.div(df_path.sum(axis=1),axis=0)\n",
    "\n",
    "    # filter on metabolism only \n",
    "    keep = []\n",
    "    for p in df_path.columns:\n",
    "        if 'Metabolism' in p: # or 'metabolism' in p: 168 vs 170 if you incl both for saliva\n",
    "            keep.append(p)\n",
    "    df_path = df_path[keep]\n",
    "    \n",
    "    # save to dict\n",
    "    g_to_path[g] = df_path\n",
    "                                 \n",
    "g_to_path['saliva_adh'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "34ed8f12-5226-47f7-b748-07609e554b8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Degradation of flavonoids</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis</th>\n",
       "      <th>Metabolism|Lipid metabolism|Steroid hormone biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.159285</td>\n",
       "      <td>0.035029</td>\n",
       "      <td>0.002320</td>\n",
       "      <td>0.003018</td>\n",
       "      <td>0.007937</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>0.027626</td>\n",
       "      <td>0.000229</td>\n",
       "      <td>0.026349</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000469</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>3.060096e-06</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.510913e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.161049</td>\n",
       "      <td>0.033134</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.002538</td>\n",
       "      <td>0.008606</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.029092</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.027987</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>9.470469e-06</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.511851e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.161760</td>\n",
       "      <td>0.034016</td>\n",
       "      <td>0.002254</td>\n",
       "      <td>0.002773</td>\n",
       "      <td>0.008658</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.029185</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>0.027522</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>2.489384e-06</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.100744e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-005.pre.stool</th>\n",
       "      <td>0.158015</td>\n",
       "      <td>0.033087</td>\n",
       "      <td>0.002185</td>\n",
       "      <td>0.002702</td>\n",
       "      <td>0.008593</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.028749</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>0.027538</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000405</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>2.762671e-06</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>1.103744e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.297970e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-006.pre.stool</th>\n",
       "      <td>0.157791</td>\n",
       "      <td>0.032210</td>\n",
       "      <td>0.002112</td>\n",
       "      <td>0.002819</td>\n",
       "      <td>0.008727</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.030094</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.026678</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000329</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>2.184941e-07</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.127712e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 159 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "KEGG_Pathway       Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.pre.stool    0.159285                            0.035029   \n",
       "OAD-003.pre.stool    0.161049                            0.033134   \n",
       "OAD-004.pre.stool    0.161760                            0.034016   \n",
       "OAD-005.pre.stool    0.158015                            0.033087   \n",
       "OAD-006.pre.stool    0.157791                            0.032210   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.pre.stool                                           0.002320                 \n",
       "OAD-003.pre.stool                                           0.002184                 \n",
       "OAD-004.pre.stool                                           0.002254                 \n",
       "OAD-005.pre.stool                                           0.002185                 \n",
       "OAD-006.pre.stool                                           0.002112                 \n",
       "\n",
       "KEGG_Pathway       Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.pre.stool                                           0.003018        \n",
       "OAD-003.pre.stool                                           0.002538        \n",
       "OAD-004.pre.stool                                           0.002773        \n",
       "OAD-005.pre.stool                                           0.002702        \n",
       "OAD-006.pre.stool                                           0.002819        \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism  \\\n",
       "OAD-001.pre.stool                     0.007937   \n",
       "OAD-003.pre.stool                     0.008606   \n",
       "OAD-004.pre.stool                     0.008658   \n",
       "OAD-005.pre.stool                     0.008593   \n",
       "OAD-006.pre.stool                     0.008727   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.pre.stool                                           0.000192    \n",
       "OAD-003.pre.stool                                           0.000180    \n",
       "OAD-004.pre.stool                                           0.000157    \n",
       "OAD-005.pre.stool                                           0.000186    \n",
       "OAD-006.pre.stool                                           0.000199    \n",
       "\n",
       "KEGG_Pathway       Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.pre.stool                          0.027626   \n",
       "OAD-003.pre.stool                          0.029092   \n",
       "OAD-004.pre.stool                          0.029185   \n",
       "OAD-005.pre.stool                          0.028749   \n",
       "OAD-006.pre.stool                          0.030094   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.pre.stool                                           0.000229      \n",
       "OAD-003.pre.stool                                           0.000194      \n",
       "OAD-004.pre.stool                                           0.000173      \n",
       "OAD-005.pre.stool                                           0.000216      \n",
       "OAD-006.pre.stool                                           0.000220      \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.pre.stool                                         0.026349   \n",
       "OAD-003.pre.stool                                         0.027987   \n",
       "OAD-004.pre.stool                                         0.027522   \n",
       "OAD-005.pre.stool                                         0.027538   \n",
       "OAD-006.pre.stool                                         0.026678   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.pre.stool                                           0.000012                    \n",
       "OAD-003.pre.stool                                           0.000019                    \n",
       "OAD-004.pre.stool                                           0.000016                    \n",
       "OAD-005.pre.stool                                           0.000016                    \n",
       "OAD-006.pre.stool                                           0.000013                    \n",
       "\n",
       "KEGG_Pathway       ...  \\\n",
       "OAD-001.pre.stool  ...   \n",
       "OAD-003.pre.stool  ...   \n",
       "OAD-004.pre.stool  ...   \n",
       "OAD-005.pre.stool  ...   \n",
       "OAD-006.pre.stool  ...   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Degradation of flavonoids  \\\n",
       "OAD-001.pre.stool                                           0.000469                                  \n",
       "OAD-003.pre.stool                                           0.000482                                  \n",
       "OAD-004.pre.stool                                           0.000635                                  \n",
       "OAD-005.pre.stool                                           0.000405                                  \n",
       "OAD-006.pre.stool                                           0.000329                                  \n",
       "\n",
       "KEGG_Pathway       Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis  \\\n",
       "OAD-001.pre.stool                                           0.000028                                      \n",
       "OAD-003.pre.stool                                           0.000090                                      \n",
       "OAD-004.pre.stool                                           0.000103                                      \n",
       "OAD-005.pre.stool                                           0.000055                                      \n",
       "OAD-006.pre.stool                                           0.000087                                      \n",
       "\n",
       "KEGG_Pathway       Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis  \\\n",
       "OAD-001.pre.stool                                       3.060096e-06                                    \n",
       "OAD-003.pre.stool                                       9.470469e-06                                    \n",
       "OAD-004.pre.stool                                       2.489384e-06                                    \n",
       "OAD-005.pre.stool                                       2.762671e-06                                    \n",
       "OAD-006.pre.stool                                       2.184941e-07                                    \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism|Steroid hormone biosynthesis  \\\n",
       "OAD-001.pre.stool                                           0.000026          \n",
       "OAD-003.pre.stool                                           0.000118          \n",
       "OAD-004.pre.stool                                           0.000108          \n",
       "OAD-005.pre.stool                                           0.000092          \n",
       "OAD-006.pre.stool                                           0.000088          \n",
       "\n",
       "KEGG_Pathway       Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series  \\\n",
       "OAD-001.pre.stool                                           0.000026                                               \n",
       "OAD-003.pre.stool                                           0.000090                                               \n",
       "OAD-004.pre.stool                                           0.000103                                               \n",
       "OAD-005.pre.stool                                           0.000055                                               \n",
       "OAD-006.pre.stool                                           0.000086                                               \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.pre.stool                                       0.000000e+00                                   \n",
       "OAD-003.pre.stool                                       0.000000e+00                                   \n",
       "OAD-004.pre.stool                                       0.000000e+00                                   \n",
       "OAD-005.pre.stool                                       1.103744e-07                                   \n",
       "OAD-006.pre.stool                                       0.000000e+00                                   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.pre.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                0.0                                  \n",
       "OAD-004.pre.stool                                                0.0                                  \n",
       "OAD-005.pre.stool                                                0.0                                  \n",
       "OAD-006.pre.stool                                                0.0                                  \n",
       "\n",
       "KEGG_Pathway       Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.pre.stool                                       1.510913e-06                           \n",
       "OAD-003.pre.stool                                       8.511851e-07                           \n",
       "OAD-004.pre.stool                                       6.100744e-07                           \n",
       "OAD-005.pre.stool                                       5.297970e-07                           \n",
       "OAD-006.pre.stool                                       1.127712e-07                           \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis  \\\n",
       "OAD-001.pre.stool                                                0.0                               \n",
       "OAD-003.pre.stool                                                0.0                               \n",
       "OAD-004.pre.stool                                                0.0                               \n",
       "OAD-005.pre.stool                                                0.0                               \n",
       "OAD-006.pre.stool                                                0.0                               \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \n",
       "OAD-001.pre.stool                                                0.0                                         \n",
       "OAD-003.pre.stool                                                0.0                                         \n",
       "OAD-004.pre.stool                                                0.0                                         \n",
       "OAD-005.pre.stool                                                0.0                                         \n",
       "OAD-006.pre.stool                                                0.0                                         \n",
       "\n",
       "[5 rows x 159 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pathway testing\n",
    "gt_to_path = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gt_to_path[g] = {}\n",
    "    df_path = g_to_path[g + '_adh'].copy()\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['post','pre']:\n",
    "        df = df_path[df_path.index.to_series().str.contains(t)]\n",
    "        gt_to_path[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_path[g]['pre'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.split('.pre')[0])\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('-',''))\n",
    "    post_df = gt_to_path[g]['post'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.split('.post')[0])\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('-',''))\n",
    "    \n",
    "    gt_to_path[g]['diff'] = post_df - pre_df\n",
    "    gt_to_path[g]['all'] = pd.concat([gt_to_path[g]['pre'],gt_to_path[g]['post']])\n",
    "\n",
    "gt_to_path[g][t].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81bd68ab-5702-4758-ad7a-34bf139443ea",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Diet DF Processing ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "77c18bcf-830a-458d-88c1-70f509e7c9c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dietary variables\n",
    "diet = ['animal_proteins_red',\n",
    "       'refined_grains', 'solanaceae', 'saturated', 'pastries',\n",
    "       'milk_derivates', 'beverages_prohibited', 'pro_inflammatory',\n",
    "       'chicken', 'plant_protein', 'whole_grains_pseudocereals_oats',\n",
    "       'gluten_free', 'vegetables', 'cruciferous',\n",
    "       'berries_and_enzymatic_fruit', 'other_fruit', 'fish', 'milk_subs',\n",
    "       'PUFA', 'MUFA', 'Walnuts', 'Nuts', 'probiotics', 'Tahini_avo',\n",
    "       'green_tea', 'anti_inflam_spices', 'inflammatory_spices',\n",
    "       'inflammatory', 'anti_inflammatory', 'total_antiinflammatory',\n",
    "       'Adherece_antiinflam', 'oats', 'almond_milk', 'fatty_fish',\n",
    "       'sesame_seeds', 'egg', 'chia', 'flaxseed_oil', 'walnuts',\n",
    "       'spinach', 'corn_tortilla', 'tahini', 'avocado', 'chicken.1',\n",
    "       'brown_rice', 'corn', 'quinoa', 'broccoli', 'Garbanzo_beans',\n",
    "       'pork', 'beef', 'burger', 'Total_omega3', 'Adherence_omega3',\n",
    "       'Total_omega6', 'Adherence_omega6', 'Total_o3_o6']\n",
    "\n",
    "for d in diet:\n",
    "    if 'Adh' in d:\n",
    "        diet.remove(d)\n",
    "\n",
    "type_to_df_diet = {}\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    df_map = type_to_df_map[g].reset_index().drop_duplicates(subset='#SampleID').set_index('#SampleID')\n",
    "    df_diet = df_map[diet].dropna()\n",
    "    type_to_df_diet[g] = df_diet\n",
    "\n",
    "    from sklearn.decomposition import PCA\n",
    "    npcs = 3\n",
    "    pca = PCA(n_components=3, whiten=True)\n",
    "    pca.fit(df_diet.T)\n",
    "    # Reformat and view results\n",
    "    loadings = pd.DataFrame(pca.components_,\n",
    "                            index=['PC%s' % _ for _ in range(npcs)], \n",
    "                            columns=df_diet.index).T\n",
    "    \n",
    "    df_merge = pd.concat([df_map, df_diet, loadings], axis=1)\n",
    "    plt.title(g + '_Diet_PCA')\n",
    "    sns.scatterplot(data=df_merge, x='PC1', y='PC2', hue='WOMAC_P_Response')\n",
    "    sns.despine()\n",
    "    plt.savefig(path + 'outputs/jobs06/PCA_diet_' + g + '.pdf')\n",
    "    plt.close() \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e22033b-a7a9-406a-8cb1-c1074ae78087",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Complete DA Analysis ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "856e9286-7a26-4639-a599-53c699043e5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taxa\n",
      "pvp\n",
      "saliva\n",
      "Number of sig var\n",
      "2\n",
      "Number of sig var\n",
      "5\n",
      "stool\n",
      "Number of sig var\n",
      "6\n",
      "Number of sig var\n",
      "4\n",
      "rvnr\n",
      "saliva\n",
      "Number of sig var\n",
      "9\n",
      "Number of sig var\n",
      "0\n",
      "stool\n",
      "Number of sig var\n",
      "6\n",
      "Number of sig var\n",
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "      <th>Timepoint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.022576</td>\n",
       "      <td>0.007098</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.292202</td>\n",
       "      <td>0.001994</td>\n",
       "      <td>0.013555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000783</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.181156</td>\n",
       "      <td>0.067844</td>\n",
       "      <td>0.150895</td>\n",
       "      <td>0.063229</td>\n",
       "      <td>0.051360</td>\n",
       "      <td>0.055600</td>\n",
       "      <td>0.012363</td>\n",
       "      <td>0.051513</td>\n",
       "      <td>0.006045</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.170334</td>\n",
       "      <td>0.135468</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002088</td>\n",
       "      <td>0.173617</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003649</td>\n",
       "      <td>0.008765</td>\n",
       "      <td>0.054802</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-001.pre.stool</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.064924</td>\n",
       "      <td>0.001590</td>\n",
       "      <td>0.032964</td>\n",
       "      <td>0.321408</td>\n",
       "      <td>0.010514</td>\n",
       "      <td>0.070031</td>\n",
       "      <td>0.005441</td>\n",
       "      <td>0.011820</td>\n",
       "      <td>0.020509</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-012.pre.stool</th>\n",
       "      <td>0.126033</td>\n",
       "      <td>0.139754</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015668</td>\n",
       "      <td>0.149951</td>\n",
       "      <td>0.015739</td>\n",
       "      <td>0.011277</td>\n",
       "      <td>0.000460</td>\n",
       "      <td>0.000372</td>\n",
       "      <td>0.005010</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004  \\\n",
       "OAD-001.pre.stool                                             0.022576                                                       \n",
       "OAD-003.pre.stool                                             0.181156                                                       \n",
       "OAD-004.pre.stool                                             0.170334                                                       \n",
       "VAOAD-001.pre.stool                                           0.000000                                                       \n",
       "VAOAD-012.pre.stool                                           0.126033                                                       \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium  \\\n",
       "OAD-001.pre.stool                                             0.007098                                                            \n",
       "OAD-003.pre.stool                                             0.067844                                                            \n",
       "OAD-004.pre.stool                                             0.135468                                                            \n",
       "VAOAD-001.pre.stool                                           0.064924                                                            \n",
       "VAOAD-012.pre.stool                                           0.139754                                                            \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella  \\\n",
       "OAD-001.pre.stool                                             0.005128                                             \n",
       "OAD-003.pre.stool                                             0.150895                                             \n",
       "OAD-004.pre.stool                                             0.000000                                             \n",
       "VAOAD-001.pre.stool                                           0.001590                                             \n",
       "VAOAD-012.pre.stool                                           0.000000                                             \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117  \\\n",
       "OAD-001.pre.stool                                             0.000000                                                              \n",
       "OAD-003.pre.stool                                             0.063229                                                              \n",
       "OAD-004.pre.stool                                             0.002088                                                              \n",
       "VAOAD-001.pre.stool                                           0.032964                                                              \n",
       "VAOAD-012.pre.stool                                           0.015668                                                              \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H  \\\n",
       "OAD-001.pre.stool                                             0.292202                                                \n",
       "OAD-003.pre.stool                                             0.051360                                                \n",
       "OAD-004.pre.stool                                             0.173617                                                \n",
       "VAOAD-001.pre.stool                                           0.321408                                                \n",
       "VAOAD-012.pre.stool                                           0.149951                                                \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum  \\\n",
       "OAD-001.pre.stool                                             0.001994                                                  \n",
       "OAD-003.pre.stool                                             0.055600                                                  \n",
       "OAD-004.pre.stool                                             0.000143                                                  \n",
       "VAOAD-001.pre.stool                                           0.010514                                                  \n",
       "VAOAD-012.pre.stool                                           0.015739                                                  \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129  \\\n",
       "OAD-001.pre.stool                                             0.013555                                                            \n",
       "OAD-003.pre.stool                                             0.012363                                                            \n",
       "OAD-004.pre.stool                                             0.000000                                                            \n",
       "VAOAD-001.pre.stool                                           0.070031                                                            \n",
       "VAOAD-012.pre.stool                                           0.011277                                                            \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus  \\\n",
       "OAD-001.pre.stool                                             0.000000                                                \n",
       "OAD-003.pre.stool                                             0.051513                                                \n",
       "OAD-004.pre.stool                                             0.003649                                                \n",
       "VAOAD-001.pre.stool                                           0.005441                                                \n",
       "VAOAD-012.pre.stool                                           0.000460                                                \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor  \\\n",
       "OAD-001.pre.stool                                             0.000783                                                        \n",
       "OAD-003.pre.stool                                             0.006045                                                        \n",
       "OAD-004.pre.stool                                             0.008765                                                        \n",
       "VAOAD-001.pre.stool                                           0.011820                                                        \n",
       "VAOAD-012.pre.stool                                           0.000372                                                        \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400  \\\n",
       "OAD-001.pre.stool                                             0.026849                                                    \n",
       "OAD-003.pre.stool                                             0.025918                                                    \n",
       "OAD-004.pre.stool                                             0.054802                                                    \n",
       "VAOAD-001.pre.stool                                           0.020509                                                    \n",
       "VAOAD-012.pre.stool                                           0.005010                                                    \n",
       "\n",
       "                     ...  \\\n",
       "OAD-001.pre.stool    ...   \n",
       "OAD-003.pre.stool    ...   \n",
       "OAD-004.pre.stool    ...   \n",
       "VAOAD-001.pre.stool  ...   \n",
       "VAOAD-012.pre.stool  ...   \n",
       "\n",
       "                     d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "OAD-001.pre.stool                                                  0.0                                                        \n",
       "OAD-003.pre.stool                                                  0.0                                                        \n",
       "OAD-004.pre.stool                                                  0.0                                                        \n",
       "VAOAD-001.pre.stool                                                0.0                                                        \n",
       "VAOAD-012.pre.stool                                                0.0                                                        \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "OAD-001.pre.stool                                                  0.0                 \n",
       "OAD-003.pre.stool                                                  0.0                 \n",
       "OAD-004.pre.stool                                                  0.0                 \n",
       "VAOAD-001.pre.stool                                                0.0                 \n",
       "VAOAD-012.pre.stool                                                0.0                 \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "OAD-001.pre.stool                                                  0.0                                                         \n",
       "OAD-003.pre.stool                                                  0.0                                                         \n",
       "OAD-004.pre.stool                                                  0.0                                                         \n",
       "VAOAD-001.pre.stool                                                0.0                                                         \n",
       "VAOAD-012.pre.stool                                                0.0                                                         \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "OAD-001.pre.stool                                                  0.0                                               \n",
       "OAD-003.pre.stool                                                  0.0                                               \n",
       "OAD-004.pre.stool                                                  0.0                                               \n",
       "VAOAD-001.pre.stool                                                0.0                                               \n",
       "VAOAD-012.pre.stool                                                0.0                                               \n",
       "\n",
       "                     d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "OAD-001.pre.stool                                                  0.0                                                                         \n",
       "OAD-003.pre.stool                                                  0.0                                                                         \n",
       "OAD-004.pre.stool                                                  0.0                                                                         \n",
       "VAOAD-001.pre.stool                                                0.0                                                                         \n",
       "VAOAD-012.pre.stool                                                0.0                                                                         \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "OAD-001.pre.stool                                                  0.0                                     \n",
       "OAD-003.pre.stool                                                  0.0                                     \n",
       "OAD-004.pre.stool                                                  0.0                                     \n",
       "VAOAD-001.pre.stool                                                0.0                                     \n",
       "VAOAD-012.pre.stool                                                0.0                                     \n",
       "\n",
       "                     d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "OAD-001.pre.stool                                                  0.0                                                                        \n",
       "OAD-003.pre.stool                                                  0.0                                                                        \n",
       "OAD-004.pre.stool                                                  0.0                                                                        \n",
       "VAOAD-001.pre.stool                                                0.0                                                                        \n",
       "VAOAD-012.pre.stool                                                0.0                                                                        \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "OAD-001.pre.stool                                                  0.0                                  \n",
       "OAD-003.pre.stool                                                  0.0                                  \n",
       "OAD-004.pre.stool                                                  0.0                                  \n",
       "VAOAD-001.pre.stool                                                0.0                                  \n",
       "VAOAD-012.pre.stool                                                0.0                                  \n",
       "\n",
       "                     d__Bacteria;p__Actinobacteriota;__;__;__;__  Timepoint  \n",
       "OAD-001.pre.stool                                       0.000000        pre  \n",
       "OAD-003.pre.stool                                       0.000000        pre  \n",
       "OAD-004.pre.stool                                       0.000000        pre  \n",
       "VAOAD-001.pre.stool                                     0.000000        pre  \n",
       "VAOAD-012.pre.stool                                     0.000035        pre  \n",
       "\n",
       "[5 rows x 375 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# complete block for paired testing in r v nr and non paired testing pre and post\n",
    "alpha = 0.05 # not 0.05 anymore\n",
    "\n",
    "dcgr_to_sig = {}\n",
    "\n",
    "for dt in ['taxa']:#, 'path']:\n",
    "    print(dt)\n",
    "    dcgr_to_sig[dt] = {}\n",
    "    for c in ['pvp', 'rvnr']: # pvp is pre AND post, comparing rvnr at pre and post time points\n",
    "        print(c)\n",
    "        dcgr_to_sig[dt][c] = {}        \n",
    "    \n",
    "        # define df arrays\n",
    "        xs = []\n",
    "        gs = []\n",
    "        ws = []\n",
    "        ps = []\n",
    "        os = []\n",
    "        dirs = []\n",
    "        adj_rs = []\n",
    "        \n",
    "        # define analysis specific params\n",
    "        if c == 'pvp':\n",
    "            outer_loop = ['pre', 'post']\n",
    "            comp = ['Response', 'No response']\n",
    "            var = 'WOMAC_P_Response'\n",
    "            drop = 'Timepoint'\n",
    "        elif c == 'rvnr': # pvp is pre and post, comparing rvnr at pre and posttime points\n",
    "            outer_loop = ['Response', 'No response']\n",
    "            comp = ['pre', 'post']\n",
    "            var = 'Timepoint'\n",
    "            drop = 'WOMAC_P_Response'\n",
    "\n",
    "        # for each sample type\n",
    "        for g in ['saliva','stool']:\n",
    "            print(g)\n",
    "            dcgr_to_sig[dt][c][g] = {}\n",
    "            \n",
    "            # grab relevant table table\n",
    "            if dt == 'taxa':\n",
    "                df_abun = gt_to_otu[g]['all']\n",
    "            elif dt == 'path':\n",
    "                df_abun = gt_to_path[g]['all']\n",
    "\n",
    "            df_map = type_to_df_map[g].copy()\n",
    "            df_map = df_map.set_index('Together')\n",
    "            if g == 'stool':\n",
    "                df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "            df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "\n",
    "            for r in outer_loop:\n",
    "                df_merge = pd.concat([df_abun, df_map[df_map[drop] == r]], axis=1)\n",
    "                df_merge = df_merge.dropna()\n",
    "                df_merge['SampleID'] = df_merge.index.map(lambda x: x.split('.p')[0])\n",
    "                if c == 'rvnr':\n",
    "                    df_vc = df_merge['SampleID'].value_counts()\n",
    "                    keep = df_vc[df_vc == 2].index.values\n",
    "                    df_merge = df_merge[df_merge['SampleID'].isin(keep)] \n",
    "                df_merge = df_merge.drop(['SampleID',drop],axis=1)\n",
    "                sig = []\n",
    "                for x in list(df_merge.columns.values)[:-1]:\n",
    "                    df_a = df_merge[df_merge[var] == comp[0]][x]\n",
    "                    df_b = df_merge[df_merge[var] == comp[1]][x]\n",
    "\n",
    "                    try:\n",
    "                        if c == 'rvnr':\n",
    "                            w,p = scipy.stats.wilcoxon(df_a.values, df_b.values, nan_policy='omit')\n",
    "                        elif c == 'pvp':\n",
    "                            w,p = scipy.stats.mannwhitneyu(df_a.values, df_b.values, nan_policy='omit')\n",
    "\n",
    "                        a = df_a.values\n",
    "                        b = df_b.values\n",
    "                        # if median in A is greater\n",
    "                        abr, abp = scipy.stats.mannwhitneyu(a,b,nan_policy='omit')\n",
    "                        # if median in B is greater\n",
    "                        bar, bap = scipy.stats.mannwhitneyu(b,a,nan_policy='omit')\n",
    "                        if abr > bar:\n",
    "                            dirs.append(comp[0])\n",
    "                            adj_rs.append(abr)\n",
    "                        elif abr < bar:\n",
    "                            dirs.append(comp[1])\n",
    "                            adj_rs.append(bar * -1)\n",
    "                        else:\n",
    "                            dirs.append('same')\n",
    "                            adj_rs.append(abr)\n",
    "\n",
    "                    except:\n",
    "                        w,p = 0, 1\n",
    "                        dirs.append(np.nan)\n",
    "                        adj_rs.append(np.nan)\n",
    "\n",
    "                    os.append(r)\n",
    "                    xs.append(x)\n",
    "                    gs.append(g)\n",
    "                    ws.append(w)\n",
    "                    ps.append(p)\n",
    "\n",
    "                    if p < alpha:\n",
    "                        ax = sns.boxplot(data=df_merge, y=x, x=var)\n",
    "                        sns.swarmplot(data=df_merge, y=x, x=var, palette='dark:grey')\n",
    "                        sns.despine()\n",
    "                    \n",
    "                        plt.tight_layout()\n",
    "                        plt.savefig(path + 'outputs/jobs06/' + c + '_taxa/' + c + '_' + dt + '_barplot_' + g + '_' + r + '_' + x.split('g__')[-1] + '.pdf')\n",
    "                        plt.close() \n",
    "                        sig.append(x)\n",
    "                print('Number of sig var')\n",
    "                print(len(sig))\n",
    "                dcgr_to_sig[dt][c][g][r] = sig\n",
    "                \n",
    "        df_res = pd.DataFrame({'SampleType':gs, # saliva v stool\n",
    "                      'fixed': os,\n",
    "                      'var': xs, # example pathway or taxa\n",
    "                      'direction': dirs, # direction of enrichment\n",
    "                      'Statistic': ws, \n",
    "                      'pvalue': ps,\n",
    "                      'adjusted': adj_rs                      \n",
    "                     })\n",
    "\n",
    "        # get fdr\n",
    "        # sort\n",
    "        df_res = df_res.sort_values(by='pvalue',ascending=True)\n",
    "        reject, pvals_corrected, _, _ = multitest.multipletests(df_res.pvalue, method='fdr_bh')\n",
    "        df_res['fdr'] = pvals_corrected\n",
    "        \n",
    "        df_res[df_res['pvalue'] < alpha].to_csv(path + 'outputs/jobs06/' + dt + '_' + c + '_' + str(alpha) + '.tsv', sep='\\t')\n",
    "        df_res[df_res['fdr'] < alpha].to_csv(path + 'outputs/jobs06/' + dt + '_' + c + '_' + str(alpha) + '_fdr.tsv', sep='\\t')\n",
    "\n",
    "df_merge.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "43bca09e-c560-43cf-975b-7b58541ae974",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "meta\n",
      "otu\n",
      "plasma\n",
      "meta\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>meta</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>logp</th>\n",
       "      <th>adj_test_stat</th>\n",
       "      <th>pvals</th>\n",
       "      <th>direction</th>\n",
       "      <th>fdr</th>\n",
       "      <th>hue</th>\n",
       "      <th>combhue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>direction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">post</th>\n",
       "      <th>1</th>\n",
       "      <td>H_N-Acetylmuramic-Acid</td>\n",
       "      <td>36.0</td>\n",
       "      <td>1.50515</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>post</td>\n",
       "      <td>0.4375</td>\n",
       "      <td>nofdr</td>\n",
       "      <td>post_nofdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>H_PE(22:6/0:0)_[M+H]+_C27H45N1O7P1</td>\n",
       "      <td>28.0</td>\n",
       "      <td>1.50515</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>post</td>\n",
       "      <td>0.4375</td>\n",
       "      <td>nofdr</td>\n",
       "      <td>post_nofdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H_D_Fructose</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.20412</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.06250</td>\n",
       "      <td>post</td>\n",
       "      <td>0.4375</td>\n",
       "      <td>ns</td>\n",
       "      <td>post_ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>H_Indole_3_propionic_acid</td>\n",
       "      <td>26.0</td>\n",
       "      <td>1.20412</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.06250</td>\n",
       "      <td>post</td>\n",
       "      <td>0.4375</td>\n",
       "      <td>ns</td>\n",
       "      <td>post_ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>H_Pyridoxic_acid</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.80618</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.15625</td>\n",
       "      <td>post</td>\n",
       "      <td>0.6250</td>\n",
       "      <td>ns</td>\n",
       "      <td>post_ns</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            meta  test_stat     logp  \\\n",
       "direction                                                              \n",
       "post      1               H_N-Acetylmuramic-Acid       36.0  1.50515   \n",
       "          26  H_PE(22:6/0:0)_[M+H]+_C27H45N1O7P1       28.0  1.50515   \n",
       "          2                         H_D_Fructose       27.0  1.20412   \n",
       "          19           H_Indole_3_propionic_acid       26.0  1.20412   \n",
       "          6                     H_Pyridoxic_acid       24.0  0.80618   \n",
       "\n",
       "              adj_test_stat    pvals direction     fdr    hue     combhue  \n",
       "direction                                                                  \n",
       "post      1            36.0  0.03125      post  0.4375  nofdr  post_nofdr  \n",
       "          26           28.0  0.03125      post  0.4375  nofdr  post_nofdr  \n",
       "          2            27.0  0.06250      post  0.4375     ns     post_ns  \n",
       "          19           26.0  0.06250      post  0.4375     ns     post_ns  \n",
       "          6            24.0  0.15625      post  0.6250     ns     post_ns  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# horizontal barplot for features in R, NR and both for enrichment in pre vs post\n",
    "alpha =  0.05\n",
    "\n",
    "d_to_df = {\n",
    "    'meta': gt_to_meta,\n",
    "    'otu': gt_to_otu,\n",
    "    # 'path': gt_to_path\n",
    "}\n",
    "d_to_label = {\n",
    "    'meta': 'Metabolite',\n",
    "    'otu': 'Taxa',\n",
    "    # 'path': 'Pathway'\n",
    "}\n",
    "t = 'all'\n",
    "\n",
    "gdt_to_res_pvp = {}\n",
    "for g in ['stool','plasma','saliva']:#]:\n",
    "    print(g)\n",
    "    gdt_to_res_pvp[g] = {}\n",
    "    if g == 'stool' or g == 'saliva':\n",
    "        ds = ['meta','otu']\n",
    "    elif g == 'plasma':\n",
    "        ds = ['meta']\n",
    "        \n",
    "    for d in ds:\n",
    "        print(d)\n",
    "        gdt_to_res_pvp[g][d] = {}\n",
    "        # grab original dfs\n",
    "        df_feat = d_to_df[d][g][t]\n",
    "\n",
    "        df_map = type_to_df_map[g].set_index('Together')\n",
    "        df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "        \n",
    "        df_map = df_map[['Timepoint','WOMAC_P_Response','HostSubjectId']]\n",
    "        \n",
    "        # merge df\n",
    "        df_merge = pd.concat([df_feat, df_map],axis=1)\n",
    "        df_merge = df_merge.dropna()\n",
    "        \n",
    "        # do paired test\n",
    "        # grab all sample IDs that are duplicated\n",
    "        df_vc = df_merge['HostSubjectId'].value_counts() == 2\n",
    "        df_vc = df_vc[df_vc == True]\n",
    "        keep_id = df_vc.index.values\n",
    "        \n",
    "        df_merge_complete = df_merge[df_merge['HostSubjectId'].isin(keep_id)]\n",
    "\n",
    "        for keep in [['Both','Response','No response'],['Response'],['No response']]: # adding 'Both' is a hack to make the dict hashing work\n",
    "            df_merge = df_merge_complete[df_merge_complete['WOMAC_P_Response'].isin(keep)]\n",
    "        \n",
    "            # samples are in order when subsetted on Timepoint\n",
    "            df_pre = df_merge[df_merge['Timepoint'] == 'pre'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "            df_post = df_merge[df_merge['Timepoint'] == 'post'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "            \n",
    "            # compute rank sum\n",
    "            feat = []\n",
    "            rs = []\n",
    "            adj_rs = []\n",
    "            ps = []\n",
    "            logps = []\n",
    "            dir = []\n",
    "            for x in df_pre.columns.values:\n",
    "                try:\n",
    "                    r, p = scipy.stats.wilcoxon(df_post[x].values-df_pre[x].values,nan_policy='omit')\n",
    "                except ValueError:\n",
    "                    r, p = 0, 1\n",
    "                if d == 'otu':\n",
    "                    # feat.append(x.split('g__')[-1])\n",
    "                    feat.append(read_taxa(x, ';'))\n",
    "                elif d == 'path':\n",
    "                    feat.append(x.split('|')[-1])\n",
    "                else:\n",
    "                    feat.append(x)\n",
    "                # rs.append(r)\n",
    "                ps.append(p)\n",
    "                # determine direction via MWU\n",
    "                a = df_post[x].values\n",
    "                b = df_pre[x].values\n",
    "                abr, abp = scipy.stats.mannwhitneyu(a,b,nan_policy='omit')\n",
    "                bar, bap = scipy.stats.mannwhitneyu(b,a,nan_policy='omit')\n",
    "\n",
    "                # log mean diff\n",
    "                # adj = np.log(np.abs(np.mean(a-b)))\n",
    "                # log ratio\n",
    "                # adj = np.log(np.mean(a/b))\n",
    "                # print(abr, bar, adj)\n",
    "                if abr > bar:\n",
    "                    dir.append('post')\n",
    "                    adj_rs.append(abr)\n",
    "                    # adj_rs.append(adj)\n",
    "                    rs.append(abr)\n",
    "                    logps.append(-np.log10(p))\n",
    "                elif abr < bar:\n",
    "                    dir.append('pre')\n",
    "                    adj_rs.append(bar * -1)\n",
    "                    # adj_rs.append(adj)# * -1)\n",
    "                    rs.append(bar)\n",
    "                    logps.append(np.log10(p))\n",
    "                else:\n",
    "                    dir.append('same')\n",
    "                    adj_rs.append(0)\n",
    "                    rs.append(abr)\n",
    "                    logps.append(np.nan)\n",
    "                # adj_rs as the difference post - pre\n",
    "                # adj_rs.append(np.log(np.mean(a) / np.mean(b)))\n",
    "                # adj_rs.append(np.mean(a-b))\n",
    "                        \n",
    "            # compile results df\n",
    "            df_results = pd.DataFrame({d: feat, 'test_stat': rs, 'logp': logps, 'adj_test_stat': adj_rs,'pvals': ps, 'direction': dir})\n",
    "\n",
    "            # fdr\n",
    "            df_results = df_results.sort_values(by='pvals',ascending=True)\n",
    "            reject, pvals_corrected, _, _ = multitest.multipletests(df_results.pvals, method='fdr_bh')\n",
    "            df_results['fdr'] = pvals_corrected\n",
    "\n",
    "            # grab top 10 results in both directions \n",
    "            df_results = df_results.groupby('direction').apply(lambda x: x.nlargest(10, 'test_stat'))\n",
    "\n",
    "            # create new hue\n",
    "            conditions = [ (df_results['fdr'] < alpha) & (df_results['pvals'] < alpha), \n",
    "                           (df_results['fdr'] > alpha) & (df_results['pvals'] < alpha),\n",
    "                           (df_results['pvals'] > alpha)]\n",
    "            choices     = [ \"fdr\", 'nofdr', 'ns' ]\n",
    "                \n",
    "            df_results[\"hue\"] = np.select(conditions, choices, default=np.nan)\n",
    "            df_results[\"combhue\"] = df_results['direction'] + '_' + df_results['hue']\n",
    "        \n",
    "            # df = df_results[df_results['fdr'] < alpha]\n",
    "            \n",
    "                        \n",
    "            df = df_results.dropna()#opy()#[df_results['pvals'] < alpha]\n",
    "\n",
    "            gdt_to_res_pvp[g][d][keep[0]] = df_results\n",
    "\n",
    "            df = df.sort_values(by=['logp'],ascending=False)\n",
    "            # df = df.sort_values(by=['adj_test_stat'],ascending=False)\n",
    "            # plot\n",
    "            # create plot\n",
    "            try:\n",
    "                if d == 'meta':\n",
    "                    if alpha == 0.05:\n",
    "                        plt.figure(figsize=(6,3))\n",
    "                    else:\n",
    "                        plt.figure(figsize=(6,6))\n",
    "                elif d == 'otu':\n",
    "                    if alpha == 0.05:\n",
    "                        plt.figure(figsize=(6,3))\n",
    "                    else:\n",
    "                        plt.figure(figsize=(8,8))\n",
    "                sns.set_theme(style=\"whitegrid\")\n",
    "                \n",
    "                sns.set_color_codes(\"pastel\")\n",
    "                sns.barplot(x='logp', y=d, data=df,\n",
    "                            hue=\"combhue\",orient='horizontal')#,native_scale=True)#width=1)\n",
    "                \n",
    "                ax.set_ylabel(d_to_label[d],fontsize=16)\n",
    "                # ax.set_xlabel(\"Effect Size / Adjusted Test Statistic\",fontsize=16)\n",
    "                ax.set_xlabel(\"-log10(p)\",fontsize=16)\n",
    "                ax.tick_params(labelsize=12)\n",
    "                sns.despine(left=True, bottom=True)\n",
    "                plt.tight_layout()\n",
    "                plt.title(g + '_' + d + '_' + str(keep))\n",
    "                plt.savefig(path + 'outputs/jobs18/DA_hbarplot_' + g + '_' + d + '_' + str(keep) + '_' + str(alpha) + '.pdf')\n",
    "                plt.close()\n",
    "                df.to_csv(path + 'outputs/jobs18/' + g + '_' + d + '_' + str(keep) + '_' + str(alpha) + '.tsv', sep='\\t')\n",
    "            \n",
    "            except: \n",
    "                print('no plot for :' + g + '_' + d + '_' + str(keep))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "102019da-3cc9-44a5-8631-d5ffa0152393",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "meta\n",
      "otu\n",
      "plasma\n",
      "meta\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>meta</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>logp</th>\n",
       "      <th>adj_test_stat</th>\n",
       "      <th>pvals</th>\n",
       "      <th>direction</th>\n",
       "      <th>fdr</th>\n",
       "      <th>hue</th>\n",
       "      <th>combhue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>direction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">resp</th>\n",
       "      <th>19</th>\n",
       "      <td>H_Indole_3_propionic_acid</td>\n",
       "      <td>36.0</td>\n",
       "      <td>1.456366</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0.034965</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.685315</td>\n",
       "      <td>nofdr</td>\n",
       "      <td>resp_nofdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>H_Kaempferol</td>\n",
       "      <td>34.0</td>\n",
       "      <td>1.134147</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0.073427</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.685315</td>\n",
       "      <td>ns</td>\n",
       "      <td>resp_ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>H_p_coumaric_acid</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.630291</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.234266</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.899946</td>\n",
       "      <td>ns</td>\n",
       "      <td>resp_ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>H_Hippuric_acid</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.530367</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.294872</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.899946</td>\n",
       "      <td>ns</td>\n",
       "      <td>resp_ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>H_Deoxycholate</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.530367</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.294872</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.899946</td>\n",
       "      <td>ns</td>\n",
       "      <td>resp_ns</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   meta  test_stat      logp  adj_test_stat  \\\n",
       "direction                                                                     \n",
       "resp      19  H_Indole_3_propionic_acid       36.0  1.456366           36.0   \n",
       "          17               H_Kaempferol       34.0  1.134147           34.0   \n",
       "          7           H_p_coumaric_acid       30.0  0.630291           30.0   \n",
       "          15            H_Hippuric_acid       29.0  0.530367           29.0   \n",
       "          25             H_Deoxycholate       29.0  0.530367           29.0   \n",
       "\n",
       "                 pvals direction       fdr    hue     combhue  \n",
       "direction                                                      \n",
       "resp      19  0.034965      resp  0.685315  nofdr  resp_nofdr  \n",
       "          17  0.073427      resp  0.685315     ns     resp_ns  \n",
       "          7   0.234266      resp  0.899946     ns     resp_ns  \n",
       "          15  0.294872      resp  0.899946     ns     resp_ns  \n",
       "          25  0.294872      resp  0.899946     ns     resp_ns  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# horizontal barplot for features in pre and post for enrichment in R vs NR \n",
    "alpha = 0.05\n",
    "\n",
    "d_to_df = {\n",
    "    'meta': gt_to_meta,\n",
    "    'otu': gt_to_otu,\n",
    "    # 'path': gt_to_path\n",
    "}\n",
    "d_to_label = {\n",
    "    'meta': 'Metabolite',\n",
    "    'otu': 'Taxa',\n",
    "    # 'path': 'Pathway'\n",
    "}\n",
    "t = 'all'\n",
    "\n",
    "\n",
    "gdt_to_res_rvnr = {}\n",
    "for g in ['stool','plasma','saliva']:\n",
    "    print(g)\n",
    "    gdt_to_res_rvnr[g] = {}\n",
    "    if g == 'stool' or g == 'saliva':\n",
    "        ds = ['meta','otu']\n",
    "    elif g == 'plasma':\n",
    "        ds = ['meta']\n",
    "        \n",
    "    for d in ds: #]:#,'path']:\n",
    "        print(d)\n",
    "        # grab original dfs\n",
    "        df_feat = d_to_df[d][g][t]\n",
    "        gdt_to_res_rvnr[g][d] = {}\n",
    "\n",
    "        df_map = type_to_df_map[g].set_index('Together')\n",
    "        df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "        \n",
    "        df_map = df_map[['Timepoint','WOMAC_P_Response','HostSubjectId']]\n",
    "        \n",
    "        # merge df\n",
    "        df_merge = pd.concat([df_feat, df_map],axis=1)\n",
    "        df_merge = df_merge.dropna()\n",
    "        \n",
    "        # do paired test\n",
    "        # grab all sample IDs that are duplicated\n",
    "        df_vc = df_merge['HostSubjectId'].value_counts() == 2\n",
    "        df_vc = df_vc[df_vc == True]\n",
    "        keep_id = df_vc.index.values\n",
    "        \n",
    "        df_merge_complete = df_merge[df_merge['HostSubjectId'].isin(keep_id)]\n",
    "\n",
    "        for keep in [['pre'],['post']]:\n",
    "            df_merge = df_merge_complete[df_merge_complete['Timepoint'].isin(keep)]\n",
    "        \n",
    "            # samples are in order when subsetted on Timepoint\n",
    "            df_r = df_merge[df_merge['WOMAC_P_Response'] == 'Response'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "            df_nr = df_merge[df_merge['WOMAC_P_Response'] == 'No response'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "            \n",
    "            # compute rank sum\n",
    "            feat = []\n",
    "            rs = []\n",
    "            adj_rs = []\n",
    "            ps = []\n",
    "            dir = []\n",
    "            logps = []\n",
    "\n",
    "            for x in df_r.columns.values:\n",
    "                try:\n",
    "                    r, p = scipy.stats.mannwhitneyu(df_r[x].values, df_nr[x].values,nan_policy='omit')\n",
    "                except ValueError:\n",
    "                    r, p = 0, 1\n",
    "                if d == 'otu':\n",
    "                    # feat.append(x.split('g__')[-1])\n",
    "                    feat.append(read_taxa(x, ';'))\n",
    "                elif d == 'path':\n",
    "                    feat.append(x.split('|')[-1])\n",
    "                else:\n",
    "                    feat.append(x)\n",
    "                # rs.append(r)\n",
    "                ps.append(p)\n",
    "                # determine direction via MWU\n",
    "                a = df_r[x].values\n",
    "                b = df_nr[x].values\n",
    "                abr, abp = scipy.stats.mannwhitneyu(a,b,nan_policy='omit')\n",
    "                bar, bap = scipy.stats.mannwhitneyu(b,a,nan_policy='omit')\n",
    "\n",
    "                # log mean diff\n",
    "                # adj = np.log(np.abs(np.mean(a)-np.mean(b)))\n",
    "                # print(abr, bar, adj)\n",
    "                # log ratio\n",
    "                # adj = np.log(np.mean(a)/np.mean(b))\n",
    "\n",
    "                if abr > bar:\n",
    "                    dir.append('resp')\n",
    "                    adj_rs.append(abr)\n",
    "                    # adj_rs.append(adj)\n",
    "                    rs.append(abr)\n",
    "                    logps.append(-np.log10(p))\n",
    "                elif abr < bar:\n",
    "                    dir.append('nonresp')\n",
    "                    adj_rs.append(bar * -1)\n",
    "                    # adj_rs.append(adj)# * -1)\n",
    "                    rs.append(bar)\n",
    "                    logps.append(np.log10(p))\n",
    "                else:\n",
    "                    dir.append('same')\n",
    "                    adj_rs.append(0)\n",
    "                    rs.append(abr)\n",
    "                    logps.append(np.nan)\n",
    "                # adj_rs as the difference post - pre\n",
    "                # adj_rs.append(np.log(np.mean(a) / np.mean(b)))\n",
    "                # adj_rs.append(np.mean(a-b))\n",
    "                        \n",
    "            # compile results df\n",
    "            df_results = pd.DataFrame({d: feat, 'test_stat': rs,'logp':logps, 'adj_test_stat': adj_rs,'pvals': ps, 'direction': dir})\n",
    "\n",
    "            # fdr\n",
    "            df_results = df_results.sort_values(by='pvals',ascending=True)\n",
    "            reject, pvals_corrected, _, _ = multitest.multipletests(df_results.pvals, method='fdr_bh')\n",
    "            df_results['fdr'] = pvals_corrected\n",
    "\n",
    "            # grab top 10 results in both directions \n",
    "            df_results = df_results.groupby('direction').apply(lambda x: x.nlargest(10, 'test_stat'))\n",
    "\n",
    "            # create new hue\n",
    "            conditions = [ (df_results['fdr'] < alpha) & (df_results['pvals'] < alpha), \n",
    "                           (df_results['fdr'] > alpha) & (df_results['pvals'] < alpha),\n",
    "                           (df_results['pvals'] > alpha)]\n",
    "            choices     = [ \"fdr\", 'nofdr', 'ns' ]\n",
    "                \n",
    "            df_results[\"hue\"] = np.select(conditions, choices, default=np.nan)\n",
    "            df_results[\"combhue\"] = df_results['direction'] + '_' + df_results['hue']\n",
    "        \n",
    "            # df = df_results[df_results['fdr'] < alpha]\n",
    "            \n",
    "                        \n",
    "            df = df_results.dropna()#opy()#[df_results['pvals'] < alpha]\n",
    "            \n",
    "            # filter on empty\n",
    "            # df_results = df_results[df_results[d] != '']]\n",
    "            \n",
    "            gdt_to_res_rvnr[g][d][keep[0]] = df_results\n",
    "            # df = df_results[df_results['pvals'] < alpha]\n",
    "            # df = df.sort_values(by=['adj_test_stat'],ascending=False)\n",
    "            df = df.sort_values(by=['logp'],ascending=False)\n",
    "            # plot\n",
    "            # create plot\n",
    "            try:\n",
    "                if d == 'meta':\n",
    "                    if alpha == 0.05:\n",
    "                        plt.figure(figsize=(6,3))\n",
    "                    else:\n",
    "                        plt.figure(figsize=(6,6))\n",
    "                elif d == 'otu':\n",
    "                    if alpha == 0.05:\n",
    "                        plt.figure(figsize=(6,3))\n",
    "                    else:\n",
    "                        plt.figure(figsize=(8,8))\n",
    "                sns.set_theme(style=\"whitegrid\")\n",
    "                \n",
    "                sns.set_color_codes(\"pastel\")\n",
    "                sns.barplot(x='logp', y=d, data=df,\n",
    "                            hue=\"combhue\",orient='horizontal')#,native_scale=True)#width=1)\n",
    "                \n",
    "                ax.set_ylabel(d_to_label[d],fontsize=16)\n",
    "                # ax.set_xlabel(\"Effect Size / Adjusted Test Statistic\",fontsize=16)\n",
    "                ax.set_xlabel(\"-log10(p)\",fontsize=16)\n",
    "                ax.tick_params(labelsize=12)\n",
    "                sns.despine(left=True, bottom=True)\n",
    "                plt.tight_layout()\n",
    "                plt.title(g + '_' + d + '_' + str(keep))\n",
    "                plt.savefig(path + 'outputs/jobs18/DA_hbarplot_' + g + '_' + d + '_' + str(keep) + '_' + str(alpha) + '.pdf')\n",
    "                plt.close()\n",
    "                df.to_csv(path + 'outputs/jobs18/' + g + '_' + d + '_' + str(keep) + '_' + str(alpha) + '.tsv', sep='\\t')\n",
    "            \n",
    "            except: \n",
    "                print('no plot for :' + g + '_' + d + '_' + str(keep))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "dc1aeb01-7766-4f5c-99a2-cb9498cd10ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "meta\n",
      "otu\n",
      "plasma\n",
      "meta\n",
      "saliva\n",
      "meta\n",
      "otu\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>otu</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>logp</th>\n",
       "      <th>adj_test_stat</th>\n",
       "      <th>pvals</th>\n",
       "      <th>direction</th>\n",
       "      <th>fdr</th>\n",
       "      <th>hue</th>\n",
       "      <th>combhue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>direction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">resp</th>\n",
       "      <th>36</th>\n",
       "      <td>Flavobacteriales_877923 Weeksellaceae</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1.318942</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.047980</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.683035</td>\n",
       "      <td>nofdr</td>\n",
       "      <td>resp_nofdr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Burkholderiales_597441 Neisseriaceae_563222</td>\n",
       "      <td>29.5</td>\n",
       "      <td>1.212092</td>\n",
       "      <td>29.5</td>\n",
       "      <td>0.061363</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.683035</td>\n",
       "      <td>ns</td>\n",
       "      <td>resp_ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Bacteroidia Bacteroidales</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.086133</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.082010</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.683035</td>\n",
       "      <td>ns</td>\n",
       "      <td>resp_ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>Staphylococcaceae Staphylococcus</td>\n",
       "      <td>26.0</td>\n",
       "      <td>1.054846</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.088136</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.683035</td>\n",
       "      <td>ns</td>\n",
       "      <td>resp_ns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>Pseudomonadaceae Pseudomonas_F</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.867547</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.135660</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.683035</td>\n",
       "      <td>ns</td>\n",
       "      <td>resp_ns</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       otu  test_stat  \\\n",
       "direction                                                               \n",
       "resp      36         Flavobacteriales_877923 Weeksellaceae       30.0   \n",
       "          22   Burkholderiales_597441 Neisseriaceae_563222       29.5   \n",
       "          32                     Bacteroidia Bacteroidales       27.0   \n",
       "          147             Staphylococcaceae Staphylococcus       26.0   \n",
       "          273               Pseudomonadaceae Pseudomonas_F       25.0   \n",
       "\n",
       "                   logp  adj_test_stat     pvals direction       fdr    hue  \\\n",
       "direction                                                                     \n",
       "resp      36   1.318942           30.0  0.047980      resp  0.683035  nofdr   \n",
       "          22   1.212092           29.5  0.061363      resp  0.683035     ns   \n",
       "          32   1.086133           27.0  0.082010      resp  0.683035     ns   \n",
       "          147  1.054846           26.0  0.088136      resp  0.683035     ns   \n",
       "          273  0.867547           25.0  0.135660      resp  0.683035     ns   \n",
       "\n",
       "                  combhue  \n",
       "direction                  \n",
       "resp      36   resp_nofdr  \n",
       "          22      resp_ns  \n",
       "          32      resp_ns  \n",
       "          147     resp_ns  \n",
       "          273     resp_ns  "
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# horizontal barplot for features\n",
    "# changes in responders versus changes in nonresponders\n",
    "alpha = 0.05\n",
    "\n",
    "d_to_df = {\n",
    "    'meta': gt_to_meta,\n",
    "    'otu': gt_to_otu,\n",
    "    # 'path': gt_to_path\n",
    "}\n",
    "d_to_label = {\n",
    "    'meta': 'Metabolite',\n",
    "    'otu': 'Taxa',\n",
    "    # 'path': 'Pathway'\n",
    "}\n",
    "t = 'diff'\n",
    "\n",
    "\n",
    "gdt_to_res_diff= {}\n",
    "for g in ['stool','plasma','saliva']:\n",
    "    print(g)\n",
    "    gdt_to_res_diff[g] = {}\n",
    "    if g == 'stool' or g == 'saliva':\n",
    "        ds = ['meta','otu']\n",
    "    elif g == 'plasma':\n",
    "        ds = ['meta']\n",
    "        \n",
    "    for d in ds: #]:#,'path']:\n",
    "        print(d)\n",
    "        # grab original dfs\n",
    "        df_feat = d_to_df[d][g][t]\n",
    "        gdt_to_res_diff[g][d] = {}\n",
    "\n",
    "        df_map = type_to_df_map[g].copy()\n",
    "        df_map = df_map.drop_duplicates(subset='HostSubjectId')\n",
    "        df_map = df_map.set_index('HostSubjectId')\n",
    "        df_map.index = df_map.index.map(lambda x: x.replace('-',''))\n",
    "        df_map.index.name = 'Study_ID'\n",
    "        \n",
    "        \n",
    "        df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "\n",
    "        # merge df\n",
    "        df_merge = pd.concat([df_feat, df_map],axis=1)\n",
    "        df_merge = df_merge.dropna()\n",
    "        \n",
    "        # samples are in order when subsetted on Timepoint\n",
    "        df_r = df_merge[df_merge['WOMAC_P_Response'] == 'Response'].drop(['Timepoint','WOMAC_P_Response'], axis=1)\n",
    "        df_nr = df_merge[df_merge['WOMAC_P_Response'] == 'No response'].drop(['Timepoint','WOMAC_P_Response'], axis=1)\n",
    "        \n",
    "        # compute rank sum\n",
    "        feat = []\n",
    "        rs = []\n",
    "        adj_rs = []\n",
    "        ps = []\n",
    "        dir = []\n",
    "        logps = []\n",
    "\n",
    "        for x in df_r.columns.values:\n",
    "            try:\n",
    "                r, p = scipy.stats.mannwhitneyu(df_r[x].values, df_nr[x].values,nan_policy='omit')\n",
    "            except ValueError:\n",
    "                r, p = 0, 1\n",
    "            if d == 'otu':\n",
    "                # feat.append(x.split('g__')[-1])\n",
    "                feat.append(read_taxa(x, ';'))\n",
    "            elif d == 'path':\n",
    "                feat.append(x.split('|')[-1])\n",
    "            else:\n",
    "                feat.append(x)\n",
    "            # rs.append(r)\n",
    "            ps.append(p)\n",
    "            # determine direction via MWU\n",
    "            a = df_r[x].values\n",
    "            b = df_nr[x].values\n",
    "            abr, abp = scipy.stats.mannwhitneyu(a,b,nan_policy='omit')\n",
    "            bar, bap = scipy.stats.mannwhitneyu(b,a,nan_policy='omit')\n",
    "\n",
    "            # log mean diff\n",
    "            # adj = np.log(np.abs(np.mean(a)-np.mean(b)))\n",
    "            # print(abr, bar, adj)\n",
    "            # log ratio\n",
    "            # adj = np.log(np.mean(a)/np.mean(b))\n",
    "\n",
    "            if abr > bar:\n",
    "                dir.append('resp')\n",
    "                adj_rs.append(abr)\n",
    "                # adj_rs.append(adj)\n",
    "                rs.append(abr)\n",
    "                logps.append(-np.log10(p))\n",
    "            elif abr < bar:\n",
    "                dir.append('nonresp')\n",
    "                adj_rs.append(bar * -1)\n",
    "                # adj_rs.append(adj)# * -1)\n",
    "                rs.append(bar)\n",
    "                logps.append(np.log10(p))\n",
    "            else:\n",
    "                dir.append('same')\n",
    "                adj_rs.append(0)\n",
    "                rs.append(abr)\n",
    "                logps.append(np.nan)\n",
    "            # adj_rs as the difference post - pre\n",
    "            # adj_rs.append(np.log(np.mean(a) / np.mean(b)))\n",
    "            # adj_rs.append(np.mean(a-b))\n",
    "                    \n",
    "        # compile results df\n",
    "        df_results = pd.DataFrame({d: feat, 'test_stat': rs,'logp':logps, 'adj_test_stat': adj_rs,'pvals': ps, 'direction': dir})\n",
    "\n",
    "\n",
    "        # fdr\n",
    "        df_results = df_results.sort_values(by='pvals',ascending=True)\n",
    "        reject, pvals_corrected, _, _ = multitest.multipletests(df_results.pvals, method='fdr_bh')\n",
    "        df_results['fdr'] = pvals_corrected\n",
    "\n",
    "        # grab top 10 results in both directions \n",
    "        df_results = df_results[df_results['direction'] != 'same']\n",
    "        df_results = df_results.groupby('direction').apply(lambda x: x.nlargest(10, 'test_stat'))\n",
    "\n",
    "        # create new hue\n",
    "        conditions = [ (df_results['fdr'] < alpha) & (df_results['pvals'] < alpha), \n",
    "                       (df_results['fdr'] > alpha) & (df_results['pvals'] < alpha),\n",
    "                       (df_results['pvals'] > alpha)]\n",
    "        choices     = [ \"fdr\", 'nofdr', 'ns' ]\n",
    "            \n",
    "        df_results[\"hue\"] = np.select(conditions, choices, default=np.nan)\n",
    "        df_results[\"combhue\"] = df_results['direction'] + '_' + df_results['hue']\n",
    "\n",
    "        # filter on empty\n",
    "        # df_results = df_results[df_results[d] != '']]\n",
    "        \n",
    "        gdt_to_res_diff[g][d][keep[0]] = df_results\n",
    "        # df = df_results[df_results['pvals'] < alpha]\n",
    "        # df = df.sort_values(by=['adj_test_stat'],ascending=False)\n",
    "        df = df_results.sort_values(by=['logp'],ascending=False)\n",
    "        # plot\n",
    "        # create plot\n",
    "        try:\n",
    "            if d == 'meta':\n",
    "                if alpha == 0.05:\n",
    "                    plt.figure(figsize=(6,3))\n",
    "                else:\n",
    "                    plt.figure(figsize=(6,6))\n",
    "            elif d == 'otu':\n",
    "                if alpha == 0.05:\n",
    "                    plt.figure(figsize=(6,3))\n",
    "                else:\n",
    "                    plt.figure(figsize=(8,8))\n",
    "            sns.set_theme(style=\"whitegrid\")\n",
    "            \n",
    "            sns.set_color_codes(\"pastel\")\n",
    "            sns.barplot(x='logp', y=d, data=df,\n",
    "                        hue=\"combhue\",orient='horizontal')#,native_scale=True)#width=1)\n",
    "            \n",
    "            ax.set_ylabel(d_to_label[d],fontsize=16)\n",
    "            # ax.set_xlabel(\"Effect Size / Adjusted Test Statistic\",fontsize=16)\n",
    "            ax.set_xlabel(\"-log10(p)\",fontsize=16)\n",
    "            ax.tick_params(labelsize=12)\n",
    "            sns.despine(left=True, bottom=True)\n",
    "            plt.tight_layout()\n",
    "            plt.title(g + '_' + d + '_diff')\n",
    "            plt.savefig(path + 'outputs/jobs18/DA_hbarplot_diffs_' + g + '_' + d + '_' + str(alpha) + '.pdf')\n",
    "            plt.close()\n",
    "            df.to_csv(path + 'outputs/jobs18/' + g + '_' + d + '_diff_' + str(alpha) + '.tsv', sep='\\t')\n",
    "        \n",
    "        except: \n",
    "            print('no plot for :' + g + '_' + d + '_diff')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70c6b863-4344-44ac-9ac9-949de2808948",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Mediation ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "90768c43-3b95-494b-93d3-46d7b1a4222e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mediation analyses\n",
    "# check for int_ba\n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid','Hyocholic_acid'] + ['Deoxycholic_acid','Lithocholic_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid','Hyocholic_acid']+ ['Deoxycholic_acid','Lithocholic_acid']\n",
    "         } # 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide'\n",
    "\n",
    "# filter on CUTIE results\n",
    "g_to_cutiejob = {\n",
    "    'saliva': 'jobs15',\n",
    "    'stool': 'jobs16'\n",
    "}\n",
    "\n",
    "g_to_medjob = {\n",
    "    # 'saliva': '20240917',\n",
    "    # 'stool': '20240913',\n",
    "    'stool': '20241001'\n",
    "}\n",
    "# figure out cutie plots\n",
    "#X = list(df_X.columns.values)\n",
    "#M = list(df_M.columns.values)\n",
    "#x,m = df.iloc[0,0:2].values\n",
    "#X.index(x), M.index(m)\n",
    "\n",
    "\n",
    "\n",
    "# diet as an intervention\n",
    "# mediation script looks for 'test' as a string when grouping triplets for jobs\n",
    "df_diet.to_csv(path + 'inputs/mediation_diet_test.txt', sep='\\t')\n",
    "\n",
    "# define outcome\n",
    "y = 'WOMAC_pain_diff'\n",
    "\n",
    "\n",
    "for g in ['stool', 'plasma']:\n",
    "    # import mediation DFs\n",
    "    df_M = pd.read_csv(path + 'inputs/cutie_df_meta_' + g + '.tsv', sep='\\t', index_col=0)\n",
    "    df_M.iloc[:,:-1].to_csv(path + 'inputs/mediation_meta_' + g + '_post_test.txt', sep='\\t')\n",
    "    \n",
    "    df_M = gt_to_meta[g]['diff']\n",
    "    df_M.to_csv(path + 'inputs/mediation_meta_' + g + '_diff_test.txt', sep='\\t')\n",
    "    \n",
    "    df_Y = df_meta_paired[['VAS_Pt_diff', 'WOMAC_pain_diff']]\n",
    "    df_Y.to_csv(path + 'inputs/mediation_WMdiff_test.txt', sep='\\t')\n",
    "\n",
    "    if g == 'stool':\n",
    "        df_X = pd.read_csv(path + 'inputs/cutie_df_taxa_' + g + '.tsv', sep='\\t', index_col=0)\n",
    "        df_X.index = df_X.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "        df_X.to_csv(path + 'inputs/mediation_taxa_' + g + '_post_test.txt', sep='\\t')\n",
    "    \n",
    "        df_X = gt_to_otu[g]['post'] - gt_to_otu[g]['pre']\n",
    "        df_X.to_csv(path + 'inputs/mediation_taxa_' + g + '_diff_test.txt', sep='\\t')\n",
    "\n",
    "    # split by resp and non resp\n",
    "    df = df_map.set_index('HostSubjectId')\n",
    "    df.index = df.index.map(lambda x: x.replace('-',''))\n",
    "    df = df[df['Timepoint'] == 'pre']\n",
    "    df = pd.concat([df_M, df],axis=1)\n",
    "\n",
    "    df_M_nr = df[df['WOMAC_P_Response'] == 'No response'].drop(['WOMAC_P_Response','Timepoint'], axis=1)\n",
    "    df_M_nr = df_M_nr.dropna()\n",
    "    df_M_nr.to_csv(path + 'inputs/df_' + g + '_M_nr.tsv', sep='\\t')\n",
    "    df_M_r = df[df['WOMAC_P_Response'] == 'Response'].drop(['WOMAC_P_Response','Timepoint'], axis=1)\n",
    "    df_M_r = df_M_r.dropna()\n",
    "    df_M_r.to_csv(path + 'inputs/df_' + g + '_M_r.tsv', sep='\\t')\n",
    "\n",
    "    if g == 'stool':\n",
    "        # grab OTU \n",
    "        df_post = gt_to_otu[g]['post'].copy()\n",
    "        df_pre = gt_to_otu[g]['pre'].copy()\n",
    "        \n",
    "        df_post.index = df_post.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "        df_pre.index = df_pre.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "        df_X = df_post - df_pre\n",
    "        df_X = df_X.dropna()\n",
    "        df_X.to_csv(path + 'inputs/df_X_' + g + '.tsv', sep='\\t')\n",
    "        \n",
    " \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ced57f62-7939-4437-a827-8da9c0874b66",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'saliva'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[27], line 86\u001b[0m\n\u001b[1;32m     80\u001b[0m df_X\u001b[38;5;241m.\u001b[39mto_csv(path \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124minputs/df_X_\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m g \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.tsv\u001b[39m\u001b[38;5;124m'\u001b[39m, sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     82\u001b[0m \u001b[38;5;66;03m# import mediation results\u001b[39;00m\n\u001b[1;32m     83\u001b[0m \u001b[38;5;66;03m#df_med = pd.read_csv(path + 'outputs/jobs19/' + g_to_medjob[g] + '/outputs/mediation_taxa_' + g + '_post_mediation_meta_' + g + '_post_mediation_WMdiff_test/mediation_results_all.tsv', \u001b[39;00m\n\u001b[1;32m     84\u001b[0m \u001b[38;5;66;03m#                     sep='\\t')\u001b[39;00m\n\u001b[1;32m     85\u001b[0m \u001b[38;5;66;03m# mediation_diet_mediation_taxa_stool_diff_mediation_WMdiff_test\u001b[39;00m\n\u001b[0;32m---> 86\u001b[0m df_med \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(path \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124moutputs/jobs19/\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m g_to_medjob[g] \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/outputs/mediation_diet_mediation_taxa_\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m g \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_diff_mediation_WMdiff_test/mediation_results_all.tsv\u001b[39m\u001b[38;5;124m'\u001b[39m, \n\u001b[1;32m     87\u001b[0m                      sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     89\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m t \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpartial\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcomplete\u001b[39m\u001b[38;5;124m'\u001b[39m]:\n\u001b[1;32m     90\u001b[0m     \u001b[38;5;66;03m# copy df\u001b[39;00m\n\u001b[1;32m     91\u001b[0m     df \u001b[38;5;241m=\u001b[39m df_med\u001b[38;5;241m.\u001b[39mcopy()\n",
      "\u001b[0;31mKeyError\u001b[0m: 'saliva'"
     ]
    }
   ],
   "source": [
    "# Mediation analyses\n",
    "# check for int_ba\n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid','Hyocholic_acid'] + ['Deoxycholic_acid','Lithocholic_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid','Hyocholic_acid']+ ['Deoxycholic_acid','Lithocholic_acid']\n",
    "         } # 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide'\n",
    "\n",
    "# filter on CUTIE results\n",
    "g_to_cutiejob = {\n",
    "    'saliva': 'jobs15',\n",
    "    'stool': 'jobs16'\n",
    "}\n",
    "\n",
    "g_to_medjob = {\n",
    "    # 'saliva': '20240917',\n",
    "    # 'stool': '20240913',\n",
    "    'stool': '20241001'\n",
    "}\n",
    "# figure out cutie plots\n",
    "#X = list(df_X.columns.values)\n",
    "#M = list(df_M.columns.values)\n",
    "#x,m = df.iloc[0,0:2].values\n",
    "#X.index(x), M.index(m)\n",
    "\n",
    "\n",
    "\n",
    "# diet as an intervention\n",
    "# mediation script looks for 'test' as a string when grouping triplets for jobs\n",
    "df_diet.to_csv(path + 'inputs/mediation_diet_test.txt', sep='\\t')\n",
    "\n",
    "# define outcome\n",
    "y = 'WOMAC_pain_diff'\n",
    "\n",
    "\n",
    "\n",
    "for g in ['stool', 'saliva']:\n",
    "    # import mediation DFs\n",
    "    df_M = pd.read_csv(path + 'inputs/cutie_df_meta_' + g + '.tsv', sep='\\t', index_col=0)\n",
    "    df_M.iloc[:,:-1].to_csv(path + 'inputs/mediation_meta_' + g + '_post_test.txt', sep='\\t')\n",
    "    \n",
    "    df_M = gt_to_meta[g]['diff']\n",
    "    df_M.to_csv(path + 'inputs/mediation_meta_' + g + '_diff_test.txt', sep='\\t')\n",
    "    \n",
    "    df_Y = df_meta_paired[['VAS_Pt_diff', 'WOMAC_pain_diff']]\n",
    "    df_Y.to_csv(path + 'inputs/mediation_WMdiff_test.txt', sep='\\t')\n",
    "    \n",
    "    df_X = pd.read_csv(path + 'inputs/cutie_df_taxa_' + g + '.tsv', sep='\\t', index_col=0)\n",
    "    df_X.index = df_X.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    df_X.to_csv(path + 'inputs/mediation_taxa_' + g + '_post_test.txt', sep='\\t')\n",
    "\n",
    "    df_X = gt_to_otu[g]['post'] - gt_to_otu[g]['pre']\n",
    "    df_X.to_csv(path + 'inputs/mediation_taxa_' + g + '_diff_test.txt', sep='\\t')\n",
    "\n",
    "\n",
    "    # split by resp and non resp\n",
    "    df = df_map.set_index('HostSubjectId')\n",
    "    df.index = df.index.map(lambda x: x.replace('-',''))\n",
    "    df = df[df['Timepoint'] == 'pre']\n",
    "    df = pd.concat([df_M, df],axis=1)\n",
    "\n",
    "    df_M_nr = df[df['WOMAC_P_Response'] == 'No response'].drop(['WOMAC_P_Response','Timepoint'], axis=1)\n",
    "    df_M_nr = df_M_nr.dropna()\n",
    "    df_M_nr.to_csv(path + 'inputs/df_' + g + '_M_nr.tsv', sep='\\t')\n",
    "    df_M_r = df[df['WOMAC_P_Response'] == 'Response'].drop(['WOMAC_P_Response','Timepoint'], axis=1)\n",
    "    df_M_r = df_M_r.dropna()\n",
    "    df_M_r.to_csv(path + 'inputs/df_' + g + '_M_r.tsv', sep='\\t')\n",
    "\n",
    "    # grab OTU \n",
    "    df_post = gt_to_otu[g]['post'].copy()\n",
    "    df_pre = gt_to_otu[g]['pre'].copy()\n",
    "    \n",
    "    df_post.index = df_post.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "    df_pre.index = df_pre.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "    df_X = df_post - df_pre\n",
    "    df_X = df_X.dropna()\n",
    "    df_X.to_csv(path + 'inputs/df_X_' + g + '.tsv', sep='\\t')\n",
    "        \n",
    "    # import mediation results\n",
    "    #df_med = pd.read_csv(path + 'outputs/jobs19/' + g_to_medjob[g] + '/outputs/mediation_taxa_' + g + '_post_mediation_meta_' + g + '_post_mediation_WMdiff_test/mediation_results_all.tsv', \n",
    "    #                     sep='\\t')\n",
    "    # mediation_diet_mediation_taxa_stool_diff_mediation_WMdiff_test\n",
    "    df_med = pd.read_csv(path + 'outputs/jobs19/' + g_to_medjob[g] + '/outputs/mediation_diet_mediation_taxa_' + g + '_diff_mediation_WMdiff_test/mediation_results_all.tsv', \n",
    "                         sep='\\t')\n",
    "\n",
    "    for t in ['partial','complete']:\n",
    "        # copy df\n",
    "        df = df_med.copy()\n",
    "        \n",
    "        # filter on significant total\n",
    "        df = df[df['total_effect_p'] < 0.05]\n",
    "\n",
    "        if t == 'complete':\n",
    "            # find complete mediations\n",
    "            df = df[df['direct_effect_p'] > 0.05]\n",
    "\n",
    "        elif t == 'partial':\n",
    "            # find partial mediations\n",
    "            df = df[df['direct_effect_p'] < 0.05]\n",
    "        \n",
    "        # check how many indirect; this reduces it by alot again\n",
    "        df = df[df['indirect_effect_p'] < 0.05]\n",
    "\n",
    "        # create column for metabolites\n",
    "        # df = df[df['M'].isin(['H_' + x for x in int_ba[g]])]\n",
    "        \n",
    "        # grab CUTIE results\n",
    "        # df_qt = pd.read_csv(path + 'outputs/' + g_to_cutiejob[g] + '/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "        # df_qt = df_qt[df_qt['class'] == 'TP']\n",
    "\n",
    "        for r in range(len(df)):\n",
    "            x,m,y = df.iloc[r,0:3].values\n",
    "            \n",
    "            # check if it's in CUTIE\n",
    "            # df_q = df_qt.copy()\n",
    "            # df_q = df_q[df_q['var1'] == m]\n",
    "            # df_q = df_q[df_q['var2'] == x]\n",
    "        \n",
    "            # if len(df_q) > 0: #len(df_q2) > 0: # len(df_q1) > 0 or len(df_q2) > 0:\n",
    "            # print('plotting ' + m + ' and ' + x)\n",
    "            # create plotting df\n",
    "            \n",
    "            # if X = bacteria, M = metabolite, Y = outcome\n",
    "            df_plot = pd.concat([df_X,df_M,df_Y],axis=1)\n",
    "\n",
    "            # if X = diet, M = bacteria, Y = outcome\n",
    "            df_plot = pd.concat([df_diet,df_X,df_Y],axis=1)\n",
    "            if df_X[m].nunique() > 5 and df_diet[x].nunique() > 2:\n",
    "                # path + 'outputs/jobs19/plots/'\n",
    "        \n",
    "                plt.figure(figsize=(8,6))\n",
    "                sns.set_theme(style=\"whitegrid\")\n",
    "                # sns.scatterplot(data=df_plot,x=x,y=m,hue=y)\n",
    "                sns.scatterplot(data=df_plot,x=x,y=y,hue=m,palette='coolwarm')#,size=6)\n",
    "                plt.tight_layout()\n",
    "                sns.despine()\n",
    "                plt.savefig(path + 'outputs/jobs19/' + g_to_medjob[g] + '/plots/' + g + '/' + x.split(';g__')[-1] + '_' + m.replace('/','fslash') + '_' + t + '_' + y + '_MX.pdf')\n",
    "                plt.close()\n",
    "        \n",
    "                plt.figure(figsize=(8,6))\n",
    "                sns.set_theme(style=\"whitegrid\")        \n",
    "                sns.scatterplot(data=df_plot,x=x,y=m,hue=m,palette='coolwarm')#,size=6)\n",
    "                sns.despine()\n",
    "                plt.tight_layout()\n",
    "                plt.savefig(path + 'outputs/jobs19/' + g_to_medjob[g] + '/plots/' + g + '/' + x.split(';g__')[-1] + '_' + m.replace('/','fslash') + '_' + t + '_' + y + '_YMX.pdf')\n",
    "                plt.close()\n",
    "            \n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44973fd6-4ca2-4cd0-81f7-691c681a1cdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## egg_d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens_partial_WOMAC_pain_diff_YMX.pdf\n",
    "df_plot = pd.concat([df_diet,df_X,df_Y],axis=1)\n",
    "x = 'egg'\n",
    "m = 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens'\n",
    "y = 'WOMAC_pain_diff'\n",
    "\n",
    "sns.scatterplot(df_plot, x=x, y=y)\n",
    "scipy.stats.spearmanr(df_plot[x], df_plot[y], nan_policy='omit')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01ae56d6-b961-42ad-b3bf-fb8ef3dd22b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(df_plot, x=x, y=m)\n",
    "scipy.stats.spearmanr(df_plot[x],df_plot[m],nan_policy='omit')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cbc6f27-6f8b-498a-83a9-94d4a2007c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(df_plot, x=m, y=y)\n",
    "scipy.stats.spearmanr(df_plot[m],df_plot[y],nan_policy='omit')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "001a1ab2-4876-4cf7-b308-862c073fd6b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in df_diet.columns:\n",
    "    for y in ['paired_alpha', 'paired_beta']:\n",
    "        df_merge = pd.concat([df_diet, g_to_dfd['stool_adh'][y]],axis=1)\n",
    "        ylab = g_to_dfd['stool_adh'][y].columns.values[0]\n",
    "        r, p = scipy.stats.spearmanr(df_merge[x],df_merge[ylab],nan_policy='omit')\n",
    "        if p < 0.1:\n",
    "            print(x,y)\n",
    "            print(p)\n",
    "                                     \n",
    "            plt.figure(figsize=(8,6))\n",
    "            sns.set_theme(style=\"whitegrid\")        \n",
    "            sns.scatterplot(data=df_merge,x=x,y=ylab)\n",
    "            sns.despine()\n",
    "            plt.tight_layout()\n",
    "            plt.savefig(path + 'outputs/jobs19/diet_plots/' + x +'_' +ylab + '.pdf')\n",
    "            plt.close()\n",
    "g_to_dfd['stool_adh']['paired_alpha'].head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53a8f0b2-47ab-4bb5-8f69-86d734fdd89e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mediation stability\n",
    "for p in ['100','200','1000']:\n",
    "    print('number of bootstraps: ' + p)\n",
    "    df = pd.read_csv(path + 'outputs/jobs20/df' + p + '.csv')\n",
    "    for y in ['total_effect_p','direct_effect_p','indirect_effect_p']:\n",
    "        x = df[y]\n",
    "        print('statistic: ' + y)\n",
    "        print('prop under 0.05: ')\n",
    "        print(len(x[x<0.05])/len(x))\n",
    "        print('mean and std')\n",
    "        print(np.mean(x), np.std(x))\n",
    "        print('\\n')\n",
    "    \n",
    "sns.distplot(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "754a719c-f121-424e-9fc3-a9ba86ff5ea5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Procrustes ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68ae796e-af81-4475-bb2d-316fd6e8e413",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Procrustes\n",
    "import scipy.spatial\n",
    "\n",
    "str_to_df = {\n",
    "    'diet': df_diet\n",
    "}\n",
    "\n",
    "for g in ['stool', 'saliva']:\n",
    "    for t in ['Baseline', 'After diet']:\n",
    "        str_to_df['_'.join([g, t, 'meta'])] = gt_to_meta[g][t]\n",
    "    for t in['pre', 'post']:\n",
    "        df = gt_to_otu[g][t]\n",
    "        df.index = df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "        str_to_df['_'.join([g, t, 'taxa'])] = df\n",
    "\n",
    "\n",
    "def convert_to_pcs(df, npcs):\n",
    "    pca = PCA(n_components=npcs, whiten=True)\n",
    "    pca.fit(df.T)\n",
    "    return pd.DataFrame(pca.components_, \n",
    "                        index=['PC%s' % _ for _ in range(npcs)], \n",
    "                        columns=df.index).T\n",
    "\n",
    "#df1 = convert_to_pcs(df1, 10)\n",
    "#df2 = convert_to_pcs(df2, 10)\n",
    "\n",
    "# convert to distance matrices\n",
    "from scipy.spatial.distance import squareform, pdist\n",
    "from itertools import combinations\n",
    "from skbio import DistanceMatrix\n",
    "from skbio.stats.distance import mantel\n",
    "\n",
    "# get all pairs of dfs\n",
    "# df_pairs = list(combinations(str_to_df, 2))\n",
    "\n",
    "\n",
    "# build comparisons list\n",
    "df_pairs = []\n",
    "# 'Baseline' 'After diet' # for meta\n",
    "# 'pre' 'post' # for stool\n",
    "\n",
    "# stool microbiome nad metabolome at pre and then at post\n",
    "a = ['stool_pre_taxa', 'stool_Baseline_meta']\n",
    "b = ['stool_post_taxa', 'stool_After diet_meta']\n",
    "\n",
    "c = ['saliva_pre_taxa', 'saliva_Baseline_meta']\n",
    "d = ['saliva_post_taxa', 'saliva_After diet_meta']\n",
    "\n",
    "# better fit \n",
    "\n",
    "# saliva to gut in microbiome at pre and then at post\n",
    "e = ['saliva_pre_taxa', 'stool_pre_taxa']\n",
    "f = ['saliva_post_taxa', 'stool_post_taxa']\n",
    "\n",
    "\n",
    "# repeat for metabolome \n",
    "g = ['saliva_Baseline_meta', 'stool_Baseline_meta']\n",
    "h = ['saliva_After diet_meta', 'stool_After diet_meta']\n",
    "\n",
    "# worse fit  \n",
    "\n",
    "df_pairs = [a, b, c, d, e, f, g, h]\n",
    "\n",
    "\n",
    "for d1, d2 in df_pairs:\n",
    "    print(d1, d2)\n",
    "    # retrieve df's \n",
    "    df1, df2 = str_to_df[d1], str_to_df[d2]\n",
    "    \n",
    "    # convert to distance matrix\n",
    "    dm1 = pd.DataFrame(squareform(pdist(df1.iloc[:, 1:])), columns=df1.index.unique(), index=df1.index.unique())\n",
    "    dm2 = pd.DataFrame(squareform(pdist(df2.iloc[:, 1:])), columns=df2.index.unique(), index=df2.index.unique())\n",
    "\n",
    "    # grab common samples\n",
    "    shared_idx = list(set(dm1.index).intersection(set(dm2.index)))\n",
    "    dm1 = dm1.loc[shared_idx,shared_idx]\n",
    "    dm2 = dm2.loc[shared_idx,shared_idx]\n",
    "\n",
    "    # compute mantel \n",
    "    r, p, n = mantel(x=dm1, y=dm2, method='spearman')\n",
    "    if p < 0.05:\n",
    "        print(r,p)\n",
    "\n",
    "    # compute m2\n",
    "    df1p, df2p, m2 = scipy.spatial.procrustes(dm1.values,dm2.values)\n",
    "    print(m2)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eebf72a6-8ff9-4a01-a698-8b8516a9fa3f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### PERMANOVA ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "15033391-7be6-4a47-b153-1a577639ee28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://github.com/theJohnnyBrown/permanova/blob/master/permanova.py\n",
    "\n",
    "#!/usr/bin/python\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "# Copyright (C) 2010 - 2012, University of New Orleans\n",
    "#\n",
    "# This program is free software; you can redistribute it and/or modify it under\n",
    "# the terms of the GNU General Public License as published by the Free\n",
    "# Software Foundation; either version 2 of the License, or (at your option)\n",
    "# any later version.\n",
    "#\n",
    "# Author: Johnny Brown\n",
    "\n",
    "import numpy as np\n",
    "import random as r\n",
    "from itertools import permutations, product, chain\n",
    "\n",
    "# convert to distance matrices\n",
    "from scipy.spatial.distance import squareform, pdist\n",
    "from itertools import combinations\n",
    "from skbio import DistanceMatrix\n",
    "from skbio.stats.distance import mantel\n",
    "\n",
    "from scipy import stats\n",
    "#stats.ss chokes on generators. Might be worth fixing that.\n",
    "stats.ss = lambda l: sum(a*a for a in l)\n",
    "\n",
    "def above_diagonal(n):\n",
    "    row = range(n)\n",
    "    for i in row:\n",
    "        for j in range(i+1,n):\n",
    "            yield i,j\n",
    "    \n",
    "\n",
    "def permanova_oneway(dm, levels, permutations = 200):\n",
    "    \"\"\" \n",
    "    Performs one-way permutational ANOVA on the given distance matrix.\n",
    "\n",
    "    One-way permanova tests the null hypothesis that distances between levels of\n",
    "    the variable are not significantly different from distances within levels of\n",
    "    the variable.\n",
    "\n",
    "    The test-statistic is a multivariate analogue to Fisher's F-ratio and is \n",
    "    calculated directly from any symmetric distance or dissimilarity matrix. \n",
    "    P-values are then obtained using permutations.\n",
    "\n",
    "    In the case of a Euclidean distance matrix calculated from only one \n",
    "    variable, the F-value is the same as the traditional parametric univariate \n",
    "    F-statistic.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    dm : array_like\n",
    "        The distance matrix of observations x observations. Represents a \n",
    "        symmetric n x n matrix with zeros on the diagonal.\n",
    "\n",
    "    levels : array_like\n",
    "        An array indicating the levels of the variable at each observation, such\n",
    "        that levels[i] == levels[j] means that dm[i] and dm[j] are rows \n",
    "        corresponding to observations in the same level or treatment.\n",
    "\n",
    "    permutations : int\n",
    "        The number of permutations used to compute the p-value. Default is 200.\n",
    "        If there are less than ``permutations`` unique permutations, then all of \n",
    "        them will be used\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    F-value : float\n",
    "        The computed F-value of the test.\n",
    "    p-value : float\n",
    "        The associated p-value from the F-distribution, generated by permuting \n",
    "        the levels\n",
    "\n",
    "    Notes\n",
    "    -----\n",
    "    It is assumed that all sample groups are the same size, n. \n",
    "\n",
    "    For example, if the values in levels are placebo, 5mg, and 10mg, then each \n",
    "    value must occur n times in the levels array, and \n",
    "    3n == len(levels) == len(dm) == len(dm[i]) (for all 0 <= i < len(dm))\n",
    "    \n",
    "    The algorithm is from Anderson[2] \n",
    "\n",
    "    References\n",
    "    ----------\n",
    "    .. [1] Lowry, Richard.  \"Concepts and Applications of Inferential\n",
    "           Statistics\". Chapter 14. http://faculty.vassar.edu/lowry/ch14pt1.html\n",
    "\n",
    "    .. [2] Anderson, Marti. A new method for non-parametric multivariate analysis \n",
    "           of variance. 2001. \n",
    "           http://www.entsoc.org/PDF/MUVE/6_NewMethod_MANOVA1_2.pdf\n",
    "    \n",
    "    .. [3] Jones, James. Lecture notes for Concepts of Statistics\n",
    "           http://people.richland.edu/james/lecture/m170/ch13-2wy.html\n",
    "    \"\"\"\n",
    "    bigf = f_oneway(dm,levels)\n",
    "\n",
    "    above = below = 0\n",
    "    nf = 0\n",
    "\n",
    "    #TODO make this pretty with math and functions\n",
    "    #perms = r.sample(list(perm_unique(levels)),permutations)\n",
    "    shuffledlevels = list(levels)#copy list so we can shuffle it\n",
    "    \n",
    "    for i in range(permutations):\n",
    "        r.shuffle(shuffledlevels)\n",
    "        f = f_oneway(dm,shuffledlevels)\n",
    "        \n",
    "        if f >= bigf:\n",
    "            above += 1\n",
    "\n",
    "        #debug\n",
    "        ## print shuffledlevels\n",
    "        ## print f\n",
    "\n",
    "    p = above/float(permutations)\n",
    "\n",
    "    return (bigf,p)\n",
    "\n",
    "#FIXME the right way to reuse code for one-way and n-way is to add an arg f such that\n",
    "#f(levels[i],levels[j]) returns True iff dm[i][j] should be included in the\n",
    "#desired sum of squares\n",
    "def f_oneway(dm,levels):\n",
    "    bign = len(levels)#number of observations\n",
    "    dm = np.asarray(dm)#distance matrix\n",
    "    a = len(set(levels))#number of levels\n",
    "    n = bign/a#number of observations per level\n",
    "    \n",
    "    assert dm.shape == (bign,bign) #check the dist matrix is square and the size\n",
    "                                   #corresponds to the length of levels\n",
    "\n",
    "    #total sum of squared distances                                   \n",
    "    sst = np.sum(stats.ss(r) for r in \n",
    "              (s[n+1:] for n,s in enumerate(dm[:-1])) )/float(bign)\n",
    "\n",
    "    #sum of within-group squares\n",
    "    #itertools.combinations(range(len(dm)),2)#top half of dm\n",
    "    ssw = np.sum((dm[i][j]**2 for i,j in  \n",
    "               product(range(len(dm)),range(1,len(dm)))\n",
    "               if i<j and levels[i] == levels[j]))/float(n)\n",
    "\n",
    "    ssa = sst - ssw\n",
    "\n",
    "    fstat = (ssa/float(a-1))/(ssw/float(bign-a))\n",
    "    #print (fstat,sst,ssa,ssw,a,bign,n)\n",
    "\n",
    "    return fstat\n",
    "\n",
    "def permanova_twoway(dm,levels,permutations=200):\n",
    "    \"\"\"Performs one-way permutational ANOVA on the given distance matrix.\n",
    "\n",
    "    One-way permanova tests the null hypothesis that distances between levels of\n",
    "    the variable are not significantly different from distances within levels of\n",
    "    the variable.\n",
    "\n",
    "    The test-statistic is a multivariate analogue to Fisher's F-ratio and is \n",
    "    calculated directly from any symmetric distance or dissimilarity matrix. \n",
    "    P-values are then obtained using permutations.\n",
    "\n",
    "    In the case of a Euclidean distance matrix calculated from only one \n",
    "    variable, the F-value is the same as the traditional parametric univariate \n",
    "    F-statistic.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    dm : array_like\n",
    "        The distance matrix of observations x observations. Represents a \n",
    "        symmetric n x n matrix with zeros on the diagonal.\n",
    "\n",
    "    levels : array_like\n",
    "        An array of pairs indicating the levels of the variable at each \n",
    "        observation, such that levels[i] == levels[j] means that dm[i] and dm[j] \n",
    "        are rows corresponding to observations in the same level or treatment.\n",
    "\n",
    "    permutations : int\n",
    "        The number of permutations used to compute the p-value. Default is 200.\n",
    "        If there are less than ``permutations`` unique permutations, then all of \n",
    "        them will be used\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    F-values : 3-tuple\n",
    "        The computed F-values of each test. F-values[0] corresponds to the \n",
    "        interaction of the two levels, F-values[1] corresponds to the effect\n",
    "        of variable a, and F-values[2] corresponds to the effect of variable\n",
    "        b. \n",
    "    p-value : 3-tuple\n",
    "        The associated p-values from the F-distribution, generated by permuting \n",
    "        the levels. The values correspond to (interaction, a, b) as with the \n",
    "        F-values.\n",
    "\n",
    "    Notes\n",
    "    -----\n",
    "    All sample groups must be the same size, n. \n",
    "\n",
    "    For example, if the variables are drug and dose, with levels 5mg, 10mg, 15mg\n",
    "    and drug1, drug2, then each combination of drug type and dose level must \n",
    "    be of size n, i.e. (5mg, drug1) must occur n times in the levels array and\n",
    "    6n == len(levels) == len(dm)\n",
    "\n",
    "    The algorithm is from Anderson[2] \n",
    "\n",
    "    References\n",
    "    ----------\n",
    "    .. [1] Lowry, Richard.  \"Concepts and Applications of Inferential\n",
    "           Statistics\". Chapter 14. http://faculty.vassar.edu/lowry/ch14pt1.html\n",
    "\n",
    "    .. [2] Anderson, Marti. A new method for non-parametric multivariate analysis \n",
    "           of variance. 2001. \n",
    "           http://stg-entsoc.bivings.com/PDF/MUVE/6_NewMethod_MANOVA1_2.pdf\n",
    "\n",
    "    .. [3] Jones, James. Lecture notes for Concepts of Statistics\n",
    "           http://people.richland.edu/james/lecture/m170/ch13-2wy.html\n",
    "    \"\"\"\n",
    "\n",
    "    bigf_i, bigf_a, bigf_b = f_twoway(dm,levels)\n",
    "\n",
    "    above_i = above_a = above_b = 0\n",
    "\n",
    "    #TODO make this pretty with math and functions\n",
    "    #perms = r.sample(list(perm_unique(levels)),permutations)\n",
    "    shuffledlevels = list(levels)#copy list so we can shuffle it\n",
    "\n",
    "    a_levels = list([l[0] for l in levels])\n",
    "    b_levels = list([l[1] for l in levels])\n",
    "    \n",
    "    #permutations\n",
    "    for i in range(permutations):\n",
    "        #All these are probably the wrong way to do the permutations. Wrong as in\n",
    "        #incorrect\n",
    "        ## r.shuffle(a_levels)\n",
    "        ## r.shuffle(b_levels)\n",
    "        ## shuffledlevels = zip(a_levels,b_levels)\n",
    "\n",
    "        r.shuffle(shuffledlevels)\n",
    "        \n",
    "        f_i, f_a, f_b = f_twoway(dm,shuffledlevels)\n",
    "\n",
    "        if f_i > bigf_i:\n",
    "            above_i += 1\n",
    "\n",
    "    for i in range(permutations):\n",
    "        r.shuffle(a_levels)\n",
    "\n",
    "        f_i, f_a, f_b = f_twoway(dm,zip(a_levels, [l[1] for l in levels]))\n",
    "\n",
    "        if f_a > bigf_a:\n",
    "            above_a += 1\n",
    "\n",
    "    for i in range(permutations):\n",
    "        r.shuffle(b_levels)\n",
    "\n",
    "        f_i, f_a, f_b = f_twoway(dm,zip([l[0] for l in levels], b_levels))\n",
    "            \n",
    "        if f_b > bigf_b:\n",
    "            above_b += 1\n",
    "\n",
    "\n",
    "\n",
    "    p_i,p_a,p_b = [ above/float(permutations) for above in \n",
    "                    [above_i, above_a, above_b]]\n",
    "\n",
    "    return ((bigf_i, bigf_a, bigf_b), (p_i, p_a, p_b))\n",
    "\n",
    "    \n",
    "    \n",
    "def f_twoway(dm, levels):\n",
    "    \n",
    "    bign = len(levels)#number of observations\n",
    "    dm = np.asarray(dm)#distance matrix\n",
    "    l = len(set(levels))#number of levels\n",
    "    a = len(set([l[0] for l in levels]))#number of a-levels\n",
    "    b = len(set([l[1] for l in levels]))#number of b-levels\n",
    "    n = bign/float(a*b)#number of observations per level\n",
    "\n",
    "    #sum of all distances\n",
    "    ## sst = np.sum(stats.ss(r) for r in \n",
    "    ##         (s[n+1:] for n,s in enumerate(dm[:-1])) )/float(bign)\n",
    "    sst = stats.ss(chain(*(r[i+1:] for i,r in enumerate(dm))))/float(bign)\n",
    "\n",
    "    #same level of both a and b (error, within-group)\n",
    "    ssr = select_ss(dm, levels, lambda a,b: a==b)/float(n)\n",
    "\n",
    "    #same level of a\n",
    "    sswa = select_ss(dm, levels, lambda a,b: a[0] == b[0])/float(b*n)\n",
    "\n",
    "    #same level of b\n",
    "    sswb = select_ss(dm, levels, lambda a,b: a[1] == b[1])/float(a*n)\n",
    "\n",
    "    \n",
    "    ssa = sst - sswa#effect of a\n",
    "    ssb = sst - sswb#effect of b\n",
    "    ssab = sst - ssa - ssb - ssr #interaction sum-of-squares\n",
    "\n",
    "    #these should each be separate functions?\n",
    "    f_interaction = (ssab/float((a-1)*(b-1)))/(ssr/float(bign - a*b))\n",
    "    f_a = (ssa/float((a-1)))/(ssr/float(bign - a*b))\n",
    "    f_b = (ssb/float((b-1)))/(ssr/float(bign - a*b))\n",
    "\n",
    "    return (f_interaction,f_a,f_b)\n",
    "\n",
    "def select_ss(dm, levels,  included):\n",
    "    \n",
    "    bign = len(dm)\n",
    "\n",
    "    distances = (dm[i][j] for i,j in above_diagonal(bign) \n",
    "                           if included(levels[i], levels[j]))\n",
    "\n",
    "    return stats.ss(distances)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c610c4b-54e2-43eb-a529-dec577a5187f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### PERMANOVA ANALYSES ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "61203d0e-62c8-4e39-8e85-e37ef753097c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_Baseline_meta\n",
      "(0.7309599705911948, 0.665)\n",
      "stool_After diet_meta\n",
      "(1.0518890722123193, 0.235)\n",
      "plasma_Baseline_meta\n",
      "(-0.125552000703509, 0.95)\n",
      "plasma_After diet_meta\n",
      "(0.13034044046710136, 0.77)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>SubjectID</th>\n",
       "      <th>OAD006</th>\n",
       "      <th>OAD001</th>\n",
       "      <th>OAD003</th>\n",
       "      <th>OAD007</th>\n",
       "      <th>OAD004</th>\n",
       "      <th>OAD009</th>\n",
       "      <th>OAD013</th>\n",
       "      <th>OAD012</th>\n",
       "      <th>OAD008</th>\n",
       "      <th>OAD017</th>\n",
       "      <th>VAOAD001</th>\n",
       "      <th>VAOAD002</th>\n",
       "      <th>VAOAD004</th>\n",
       "      <th>VAOAD008</th>\n",
       "      <th>VAOAD010</th>\n",
       "      <th>VAOAD009</th>\n",
       "      <th>VAOAD012</th>\n",
       "      <th>VAOAD011</th>\n",
       "      <th>VAOAD014</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SubjectID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.624672</td>\n",
       "      <td>4.238049</td>\n",
       "      <td>4.742379</td>\n",
       "      <td>3.306175</td>\n",
       "      <td>3.172851</td>\n",
       "      <td>3.745015</td>\n",
       "      <td>4.212943</td>\n",
       "      <td>3.785796</td>\n",
       "      <td>4.048351</td>\n",
       "      <td>4.478677</td>\n",
       "      <td>3.139457</td>\n",
       "      <td>3.818057</td>\n",
       "      <td>2.947746</td>\n",
       "      <td>2.912331</td>\n",
       "      <td>3.422266</td>\n",
       "      <td>5.170314</td>\n",
       "      <td>3.189969</td>\n",
       "      <td>4.123102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>4.624672</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.593756</td>\n",
       "      <td>3.507551</td>\n",
       "      <td>5.294145</td>\n",
       "      <td>4.788534</td>\n",
       "      <td>4.125055</td>\n",
       "      <td>4.283155</td>\n",
       "      <td>4.983941</td>\n",
       "      <td>3.267744</td>\n",
       "      <td>3.871621</td>\n",
       "      <td>5.003564</td>\n",
       "      <td>3.661105</td>\n",
       "      <td>6.045944</td>\n",
       "      <td>4.560048</td>\n",
       "      <td>5.599952</td>\n",
       "      <td>5.701787</td>\n",
       "      <td>4.358517</td>\n",
       "      <td>5.814690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>4.238049</td>\n",
       "      <td>4.593756</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.463564</td>\n",
       "      <td>3.916008</td>\n",
       "      <td>4.322544</td>\n",
       "      <td>4.849971</td>\n",
       "      <td>5.144340</td>\n",
       "      <td>3.786174</td>\n",
       "      <td>4.247840</td>\n",
       "      <td>4.460487</td>\n",
       "      <td>5.631385</td>\n",
       "      <td>3.144566</td>\n",
       "      <td>5.291839</td>\n",
       "      <td>3.879370</td>\n",
       "      <td>3.784153</td>\n",
       "      <td>6.289947</td>\n",
       "      <td>3.309002</td>\n",
       "      <td>6.059513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD007</th>\n",
       "      <td>4.742379</td>\n",
       "      <td>3.507551</td>\n",
       "      <td>3.463564</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.303373</td>\n",
       "      <td>4.509665</td>\n",
       "      <td>4.492377</td>\n",
       "      <td>4.218609</td>\n",
       "      <td>3.918360</td>\n",
       "      <td>3.645292</td>\n",
       "      <td>4.060945</td>\n",
       "      <td>4.983137</td>\n",
       "      <td>2.662060</td>\n",
       "      <td>5.309761</td>\n",
       "      <td>3.980563</td>\n",
       "      <td>4.519318</td>\n",
       "      <td>5.789686</td>\n",
       "      <td>4.143534</td>\n",
       "      <td>6.195704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>3.306175</td>\n",
       "      <td>5.294145</td>\n",
       "      <td>3.916008</td>\n",
       "      <td>4.303373</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.958247</td>\n",
       "      <td>3.515076</td>\n",
       "      <td>4.268502</td>\n",
       "      <td>3.410013</td>\n",
       "      <td>3.968700</td>\n",
       "      <td>3.883621</td>\n",
       "      <td>3.915517</td>\n",
       "      <td>3.647858</td>\n",
       "      <td>3.681660</td>\n",
       "      <td>3.716288</td>\n",
       "      <td>3.441643</td>\n",
       "      <td>4.310917</td>\n",
       "      <td>3.740145</td>\n",
       "      <td>3.867346</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "SubjectID    OAD006    OAD001    OAD003    OAD007    OAD004    OAD009  \\\n",
       "SubjectID                                                               \n",
       "OAD006     0.000000  4.624672  4.238049  4.742379  3.306175  3.172851   \n",
       "OAD001     4.624672  0.000000  4.593756  3.507551  5.294145  4.788534   \n",
       "OAD003     4.238049  4.593756  0.000000  3.463564  3.916008  4.322544   \n",
       "OAD007     4.742379  3.507551  3.463564  0.000000  4.303373  4.509665   \n",
       "OAD004     3.306175  5.294145  3.916008  4.303373  0.000000  2.958247   \n",
       "\n",
       "SubjectID    OAD013    OAD012    OAD008    OAD017  VAOAD001  VAOAD002  \\\n",
       "SubjectID                                                               \n",
       "OAD006     3.745015  4.212943  3.785796  4.048351  4.478677  3.139457   \n",
       "OAD001     4.125055  4.283155  4.983941  3.267744  3.871621  5.003564   \n",
       "OAD003     4.849971  5.144340  3.786174  4.247840  4.460487  5.631385   \n",
       "OAD007     4.492377  4.218609  3.918360  3.645292  4.060945  4.983137   \n",
       "OAD004     3.515076  4.268502  3.410013  3.968700  3.883621  3.915517   \n",
       "\n",
       "SubjectID  VAOAD004  VAOAD008  VAOAD010  VAOAD009  VAOAD012  VAOAD011  \\\n",
       "SubjectID                                                               \n",
       "OAD006     3.818057  2.947746  2.912331  3.422266  5.170314  3.189969   \n",
       "OAD001     3.661105  6.045944  4.560048  5.599952  5.701787  4.358517   \n",
       "OAD003     3.144566  5.291839  3.879370  3.784153  6.289947  3.309002   \n",
       "OAD007     2.662060  5.309761  3.980563  4.519318  5.789686  4.143534   \n",
       "OAD004     3.647858  3.681660  3.716288  3.441643  4.310917  3.740145   \n",
       "\n",
       "SubjectID  VAOAD014  \n",
       "SubjectID            \n",
       "OAD006     4.123102  \n",
       "OAD001     5.814690  \n",
       "OAD003     6.059513  \n",
       "OAD007     6.195704  \n",
       "OAD004     3.867346  "
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PERMANOVA metabolites\n",
    "# construct dict of data\n",
    "str_to_df = {\n",
    "    #'diet': df_diet\n",
    "}\n",
    "\n",
    "\n",
    "for g in ['stool', 'plasma','saliva']:\n",
    "    for t in ['Baseline', 'After diet']:\n",
    "        str_to_df['_'.join([g, t, 'meta'])] = gt_to_meta[g][t]\n",
    "\n",
    "# this is testing resp vs non resp at pre and post time points only, NOT the paired change in beta post-pre \n",
    "for label in str_to_df:\n",
    "    print(label)\n",
    "    df = str_to_df[label]\n",
    "\n",
    "    # get dx\n",
    "    df_dx = type_to_df_map[label.split('_')[0]].drop_duplicates(subset='HostSubjectId')\n",
    "    df_dx = df_dx.set_index('HostSubjectId')\n",
    "    df_dx.index = df_dx.index.map(lambda x: x.replace('-',''))\n",
    "    df_dx = df_dx[['Timepoint','WOMAC_P_Response']]\n",
    "\n",
    "    # get merged df and drop na\n",
    "    df.index.names = ['SubjectID']\n",
    "    df_dx.index.names = ['SubjectID']\n",
    "    \n",
    "    df_merge = pd.concat([df, df_dx], axis=1)\n",
    "    df_merge = df_merge.dropna()\n",
    "    \n",
    "    # construct distance matrix from df\n",
    "    df_data = df.loc[df_merge.index,:]\n",
    "    dx = df_dx.loc[df_merge.index,:]['WOMAC_P_Response']\n",
    "    dm = pd.DataFrame(squareform(pdist(df_data.iloc[:, 1:])), columns=df_data.index.unique(), index=df_data.index.unique())\n",
    "\n",
    "    # takes in distance matrix\n",
    "    print(permanova_oneway(dm, levels=[1 if x == 'No response' else 0 for x in dx]))\n",
    "\n",
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "f70dd3f6-3455-4b56-8be2-01dcef569ff0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasma\n",
      "stool\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_Thiamine</th>\n",
       "      <th>H_Melezitose</th>\n",
       "      <th>H_Phosphocholine</th>\n",
       "      <th>H_SN_Glycero_3_Phophocholine</th>\n",
       "      <th>H_Gamma_Valerobetaine</th>\n",
       "      <th>H_N_Acetylneuraminic_acid</th>\n",
       "      <th>H_Pyridoxine</th>\n",
       "      <th>H_N_Acetylneuraminic_Acid</th>\n",
       "      <th>H_N_Acetylmuramic_Acid</th>\n",
       "      <th>...</th>\n",
       "      <th>H_Nonhydroxylated_bile_acid</th>\n",
       "      <th>H_Hydroxydodecanoic_acid</th>\n",
       "      <th>H_Omega_Hydroxydodecanoate</th>\n",
       "      <th>H_Delta_Methyldodecenoic_Acid</th>\n",
       "      <th>H_Hydroxydecanoic_acid</th>\n",
       "      <th>H_Hydroxydecanoate</th>\n",
       "      <th>H_Methylpentanoic_acid</th>\n",
       "      <th>H_Monohydroxylated_bile_acid</th>\n",
       "      <th>H_Palmitoyl_ethanolamide</th>\n",
       "      <th>H_Linoleic_Acid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>0.001217</td>\n",
       "      <td>0.000234</td>\n",
       "      <td>0.001487</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.042887</td>\n",
       "      <td>0.000398</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000771</td>\n",
       "      <td>0.000484</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000477</td>\n",
       "      <td>0.000478</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000379</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005366</td>\n",
       "      <td>0.000742</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>0.003093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.000159</td>\n",
       "      <td>0.000747</td>\n",
       "      <td>0.002120</td>\n",
       "      <td>0.001773</td>\n",
       "      <td>0.080033</td>\n",
       "      <td>0.000577</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001202</td>\n",
       "      <td>0.000336</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000400</td>\n",
       "      <td>0.000473</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000238</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014685</td>\n",
       "      <td>0.000165</td>\n",
       "      <td>0.000340</td>\n",
       "      <td>0.001531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.000307</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>0.000252</td>\n",
       "      <td>0.000501</td>\n",
       "      <td>0.000477</td>\n",
       "      <td>0.034606</td>\n",
       "      <td>0.000342</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>0.000443</td>\n",
       "      <td>0.000427</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000263</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000233</td>\n",
       "      <td>0.000293</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.012902</td>\n",
       "      <td>0.000145</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD004</th>\n",
       "      <td>0.000819</td>\n",
       "      <td>0.005895</td>\n",
       "      <td>0.000242</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.032030</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.000271</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.002879</td>\n",
       "      <td>0.002228</td>\n",
       "      <td>0.002028</td>\n",
       "      <td>0.001658</td>\n",
       "      <td>0.000984</td>\n",
       "      <td>0.021704</td>\n",
       "      <td>0.001039</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.000490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD011</th>\n",
       "      <td>0.000266</td>\n",
       "      <td>0.005125</td>\n",
       "      <td>0.000400</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000720</td>\n",
       "      <td>0.011709</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>0.002686</td>\n",
       "      <td>0.001940</td>\n",
       "      <td>0.001686</td>\n",
       "      <td>0.001562</td>\n",
       "      <td>0.000851</td>\n",
       "      <td>0.030267</td>\n",
       "      <td>0.001059</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>0.000709</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 88 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          H_Pyridoxamine  H_Thiamine  H_Melezitose  H_Phosphocholine  \\\n",
       "Sample                                                                 \n",
       "VAOAD001        0.001217    0.000234      0.001487          0.000000   \n",
       "VAOAD009        0.000594    0.000159      0.000747          0.002120   \n",
       "OAD001          0.000307    0.000133      0.000252          0.000501   \n",
       "VAOAD004        0.000819    0.005895      0.000242          0.000000   \n",
       "VAOAD011        0.000266    0.005125      0.000400          0.000207   \n",
       "\n",
       "          H_SN_Glycero_3_Phophocholine  H_Gamma_Valerobetaine  \\\n",
       "Sample                                                          \n",
       "VAOAD001                      0.000000               0.042887   \n",
       "VAOAD009                      0.001773               0.080033   \n",
       "OAD001                        0.000477               0.034606   \n",
       "VAOAD004                      0.000513               0.032030   \n",
       "VAOAD011                      0.000720               0.011709   \n",
       "\n",
       "          H_N_Acetylneuraminic_acid  H_Pyridoxine  H_N_Acetylneuraminic_Acid  \\\n",
       "Sample                                                                         \n",
       "VAOAD001                   0.000398      0.000000                   0.000771   \n",
       "VAOAD009                   0.000577      0.000000                   0.001202   \n",
       "OAD001                     0.000342      0.000102                   0.000443   \n",
       "VAOAD004                   0.000055      0.000208                   0.000034   \n",
       "VAOAD011                   0.000072      0.000000                   0.000160   \n",
       "\n",
       "          H_N_Acetylmuramic_Acid  ...  H_Nonhydroxylated_bile_acid  \\\n",
       "Sample                            ...                                \n",
       "VAOAD001                0.000484  ...                     0.000477   \n",
       "VAOAD009                0.000336  ...                     0.000400   \n",
       "OAD001                  0.000427  ...                     0.000050   \n",
       "VAOAD004                0.000271  ...                     0.000186   \n",
       "VAOAD011                0.000102  ...                     0.000173   \n",
       "\n",
       "          H_Hydroxydodecanoic_acid  H_Omega_Hydroxydodecanoate  \\\n",
       "Sample                                                           \n",
       "VAOAD001                  0.000478                    0.000000   \n",
       "VAOAD009                  0.000473                    0.000000   \n",
       "OAD001                    0.000263                    0.000000   \n",
       "VAOAD004                  0.002879                    0.002228   \n",
       "VAOAD011                  0.002686                    0.001940   \n",
       "\n",
       "          H_Delta_Methyldodecenoic_Acid  H_Hydroxydecanoic_acid  \\\n",
       "Sample                                                            \n",
       "VAOAD001                       0.000379                0.000265   \n",
       "VAOAD009                       0.000000                0.000238   \n",
       "OAD001                         0.000233                0.000293   \n",
       "VAOAD004                       0.002028                0.001658   \n",
       "VAOAD011                       0.001686                0.001562   \n",
       "\n",
       "          H_Hydroxydecanoate  H_Methylpentanoic_acid  \\\n",
       "Sample                                                 \n",
       "VAOAD001            0.000000                0.005366   \n",
       "VAOAD009            0.000000                0.014685   \n",
       "OAD001              0.000054                0.012902   \n",
       "VAOAD004            0.000984                0.021704   \n",
       "VAOAD011            0.000851                0.030267   \n",
       "\n",
       "          H_Monohydroxylated_bile_acid  H_Palmitoyl_ethanolamide  \\\n",
       "Sample                                                             \n",
       "VAOAD001                      0.000742                  0.000441   \n",
       "VAOAD009                      0.000165                  0.000340   \n",
       "OAD001                        0.000145                  0.000065   \n",
       "VAOAD004                      0.001039                  0.000185   \n",
       "VAOAD011                      0.001059                  0.000151   \n",
       "\n",
       "          H_Linoleic_Acid  \n",
       "Sample                     \n",
       "VAOAD001         0.003093  \n",
       "VAOAD009         0.001531  \n",
       "OAD001           0.000326  \n",
       "VAOAD004         0.000490  \n",
       "VAOAD011         0.000709  \n",
       "\n",
       "[5 rows x 88 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Shannon entropy calculations\n",
    "gt_to_meta_norm = {}\n",
    "g_to_meta_norm = {}\n",
    "\n",
    "for g in ['stool','plasma','saliva']:\n",
    "    gt_to_meta_norm[g] = {}\n",
    "\n",
    "for g in ['plasma','stool','saliva']:\n",
    "    print(g)\n",
    "    df = pd.read_csv(path + 'inputs/' + g + '_removed_zeros.csv')#, index_col=0)\n",
    "\n",
    "    # split on id and timepoint and rest of data\n",
    "    df_info = df[['Sample','Time']]\n",
    "\n",
    "    # normalize\n",
    "    df_data = df.iloc[:,2:]\n",
    "    df_data = df_data.div(df_data.sum(axis=1),axis=0)\n",
    "\n",
    "    # recombine\n",
    "    df = pd.concat([df_info, df_data],axis=1)\n",
    "    g_to_meta_norm[g] = df\n",
    "\n",
    "    # split on time point\n",
    "    df_pre = df[df['Time'] == 'Baseline']\n",
    "    df_pre = df_pre.drop('Time',axis=1)\n",
    "    df_pre = df_pre.set_index('Sample')\n",
    "    df_pre.index = df_pre.index.map(lambda x: x.split('_Basal')[0].replace('_',''))\n",
    "    gt_to_meta_norm[g]['pre'] = df_pre\n",
    "    \n",
    "    df_post = df[df['Time'].isin(['Day15','After diet'])]\n",
    "    df_post = df_post.drop('Time',axis=1)\n",
    "    df_post = df_post.set_index('Sample')\n",
    "    df_post.index = df_post.index.map(lambda x: x.split('_Day15')[0].replace('_',''))\n",
    "    gt_to_meta_norm[g]['post'] = df_post\n",
    "\n",
    "\n",
    "df_post.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "ec38f5db-8aa9-4d7b-995f-6636814f47a6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plasma\n",
      "21 21\n",
      "combined pre post wsr\n",
      "69.0 0.11105823516845703\n",
      "diff\n",
      "27.0 0.17737027334550554\n",
      "Baseline\n",
      "56.0 0.3510624222698526\n",
      "After diet\n",
      "34.0 0.44214230901537405\n",
      "resp only pre post\n",
      "5.0 0.078125\n",
      "nr only pre post\n",
      "32.0 0.9658203125\n",
      "stool\n",
      "15 15\n",
      "combined pre post wsr\n",
      "29.0 0.083251953125\n",
      "diff\n",
      "23.0 0.8356643356643356\n",
      "Baseline\n",
      "19.0 0.8356643356643356\n",
      "After diet\n",
      "18.0 0.7307692307692307\n",
      "resp only pre post\n",
      "9.0 0.46875\n",
      "nr only pre post\n",
      "6.0 0.4375\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WOMAC_P_Response</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         WOMAC_P_Response\n",
       "SampleID                 \n",
       "OAD001        No response\n",
       "OAD003        No response\n",
       "OAD004        No response\n",
       "OAD005                NaN\n",
       "OAD006           Response"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pre post comparison\n",
    "# WSR on entropies\n",
    "for g in ['plasma','stool','saliva']:\n",
    "    print(g)\n",
    "    df_pre = gt_to_meta_norm[g]['pre']\n",
    "    df_pre['shannon_entropy'] = scipy.stats.entropy(df_pre, axis=1)\n",
    "\n",
    "    df_post = gt_to_meta_norm[g]['post']\n",
    "    df_post['shannon_entropy'] = scipy.stats.entropy(df_post, axis=1)\n",
    "    \n",
    "    # common samples\n",
    "    common_idx = list(set(df_pre.index.values).intersection(set(df_post.index.values)))\n",
    "    df_pre = df_pre.loc[common_idx,:]\n",
    "    df_post = df_post.loc[common_idx,:]\n",
    "    print(len(df_pre),len(df_post))\n",
    "\n",
    "    df_pre2 = df_pre.copy()\n",
    "    df_pre2.index = df_pre2.index.map(lambda x: x + '_pre')\n",
    "    df_pre2['Timepoint'] = 'pre'\n",
    "    df_pre2 = df_pre2.sort_index()\n",
    "\n",
    "    df_post2 = df_post.copy()\n",
    "    df_post2.index = df_post2.index.map(lambda x: x + '_post')\n",
    "    df_post2['Timepoint'] = 'post'\n",
    "    df_post2 = df_post2.sort_index()\n",
    "\n",
    "\n",
    "    \n",
    "    # merge\n",
    "    df_merge = pd.concat([df_pre2, df_post2])\n",
    "\n",
    "    prevals = df_merge[df_merge['Timepoint'] == 'pre']['shannon_entropy'].values\n",
    "    postvals = df_merge[df_merge['Timepoint'] == 'post']['shannon_entropy'].values\n",
    "    # get paired differences\n",
    "    # it looks like the rows are in the correct order will test\n",
    "    # for x,y in zip(df_pre.index, df_post.index):\n",
    "    #    if x != y:\n",
    "    #        print(x,y)\n",
    "    print('combined pre post wsr')\n",
    "    W,p = scipy.stats.wilcoxon(x=prevals,y=postvals, nan_policy='omit')\n",
    "    print(W,p)\n",
    "    ax = sns.swarmplot(data=df_merge, x='Timepoint', y='shannon_entropy', hue='Timepoint', palette=['b','r'], orient='v')\n",
    "    for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "        ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "    sns.boxplot(data=df_merge, x='Timepoint', y='shannon_entropy', orient='v', hue='Timepoint', boxprops=dict(alpha=.3))\n",
    "    \n",
    "    # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "    ax.set_ylabel('Shannon Entropy',fontsize=16)\n",
    "    ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "    ax.tick_params(labelsize=16)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/wsr_meta_norm_' + g  + '.pdf')\n",
    "    plt.close()            \n",
    "    \n",
    "\n",
    "    # comparing R v NR at pre and post time points\n",
    "    df_dx = type_to_df_map[g].reset_index().copy()\n",
    "    df_dx['Together'] = df_dx['Together'].map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "    df_dx = df_dx.drop_duplicates(subset=['Together'])\n",
    "    df_dx = df_dx.set_index('Together')\n",
    "    df_dx.index.name = 'SampleID'\n",
    "    df_dx = df_dx[['WOMAC_P_Response']]\n",
    "\n",
    "\n",
    "    # comparing R v NR in terms of delta pre-post\n",
    "    # df_diff = df_post.drop('WOMAC_P_Response',axis=1) - df_pre.drop('WOMAC_P_Response',axis=1)\n",
    "    df_diff = df_post - df_pre\n",
    "    df_diff = pd.concat([df_diff,df_dx],axis=1)\n",
    "    \n",
    "    rvals = df_diff[df_diff['WOMAC_P_Response'] == 'Response']['shannon_entropy'].values\n",
    "    nrvals = df_diff[df_diff['WOMAC_P_Response'] == 'No response']['shannon_entropy'].values  \n",
    "    \n",
    "    W,p = scipy.stats.mannwhitneyu(rvals,nrvals, nan_policy='omit')\n",
    "    print('diff')\n",
    "    print(W,p)\n",
    "    ax = sns.swarmplot(data=df_diff, x='WOMAC_P_Response', y='shannon_entropy', hue='WOMAC_P_Response', palette=['b','r'], orient='v')\n",
    "    sns.boxplot(data=df_diff, x='WOMAC_P_Response', y='shannon_entropy', orient='v', hue='WOMAC_P_Response', boxprops=dict(alpha=.3))\n",
    "    \n",
    "    # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "    ax.set_ylabel('Shannon Entropy',fontsize=16)\n",
    "    ax.set_xlabel('WOMAC_P_Response',fontsize=16)                \n",
    "    ax.tick_params(labelsize=16)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/wsr_meta_norm_diff_rvnr_' + g  + '.pdf')\n",
    "    plt.close()            \n",
    "\n",
    "\n",
    "    # pre\n",
    "    df_pre = pd.concat([df_pre, df_dx],axis=1)\n",
    "    df_pre = df_pre.dropna()\n",
    "\n",
    "    rvals = df_pre[df_pre['WOMAC_P_Response'] == 'Response']['shannon_entropy'].values\n",
    "    nrvals = df_pre[df_pre['WOMAC_P_Response'] == 'No response']['shannon_entropy'].values  \n",
    "    \n",
    "    W,p = scipy.stats.mannwhitneyu(rvals,nrvals, nan_policy='omit')\n",
    "    print('Baseline')\n",
    "    print(W,p)\n",
    "    ax = sns.swarmplot(data=df_pre, x='WOMAC_P_Response', y='shannon_entropy', hue='WOMAC_P_Response', palette=['b','r'], orient='v')\n",
    "    sns.boxplot(data=df_pre, x='WOMAC_P_Response', y='shannon_entropy', orient='v', hue='WOMAC_P_Response', boxprops=dict(alpha=.3))\n",
    "    \n",
    "    # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "    ax.set_ylabel('Shannon Entropy',fontsize=16)\n",
    "    ax.set_xlabel('WOMAC_P_Response',fontsize=16)                \n",
    "    ax.tick_params(labelsize=16)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/wsr_meta_norm_baseline_rvnr_' + g  + '.pdf')\n",
    "    plt.close()            \n",
    "\n",
    "\n",
    "\n",
    "    # post\n",
    "    df_post = pd.concat([df_post, df_dx],axis=1)\n",
    "    df_post = df_post.dropna()\n",
    "\n",
    "    rvals = df_post[df_post['WOMAC_P_Response'] == 'Response']['shannon_entropy'].values    \n",
    "    nrvals = df_post[df_post['WOMAC_P_Response'] == 'No response']['shannon_entropy'].values  \n",
    "    \n",
    "    W,p = scipy.stats.mannwhitneyu(rvals,nrvals, nan_policy='omit')\n",
    "    print('After diet')\n",
    "    print(W,p)\n",
    "    ax = sns.swarmplot(data=df_post, x='WOMAC_P_Response', y='shannon_entropy', hue='WOMAC_P_Response', palette=['b','r'], orient='v')\n",
    "    sns.boxplot(data=df_pre, x='WOMAC_P_Response', y='shannon_entropy', orient='v', hue='WOMAC_P_Response', boxprops=dict(alpha=.3))\n",
    "    \n",
    "    # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "    ax.set_ylabel('Shannon Entropy',fontsize=16)\n",
    "    ax.set_xlabel('WOMAC_P_Response',fontsize=16)                \n",
    "    ax.tick_params(labelsize=16)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/wsr_meta_norm_afterdiet_rvnr_' + g  + '.pdf')\n",
    "    plt.close()            \n",
    "\n",
    "\n",
    "    # responders pre vs post\n",
    "    df_dx_pre = df_dx.copy()\n",
    "    df_dx_pre.index = df_dx_pre.index.map(lambda x: x + '_pre')\n",
    "    df_dx_post = df_dx.copy()\n",
    "    df_dx_post.index = df_dx_post.index.map(lambda x: x + '_post')\n",
    "\n",
    "    df_dx2 = pd.concat([df_dx_pre, df_dx_post], axis=0)\n",
    "    \n",
    "    df_merge = pd.concat([df_pre2, df_post2])\n",
    "    df_merge2 = pd.concat([df_merge, df_dx2], axis=1)\n",
    "    df_merge = df_merge2.dropna()\n",
    "    \n",
    "    df_merge = df_merge[df_merge['WOMAC_P_Response'] == 'Response']\n",
    "\n",
    "    prevals = df_merge[df_merge['Timepoint'] == 'pre']['shannon_entropy'].values\n",
    "    postvals = df_merge[df_merge['Timepoint'] == 'post']['shannon_entropy'].values\n",
    "    # get paired differences\n",
    "    # it looks like the rows are in the correct order will test\n",
    "    # for x,y in zip(df_pre.index, df_post.index):\n",
    "    #    if x != y:\n",
    "    #        print(x,y)\n",
    "\n",
    "    W,p = scipy.stats.wilcoxon(x=prevals,y=postvals, nan_policy='omit')\n",
    "    print('resp only pre post')\n",
    "    print(W,p)\n",
    "    ax = sns.swarmplot(data=df_merge, x='Timepoint', y='shannon_entropy', hue='Timepoint', palette=['b','r'], orient='v')\n",
    "    for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "        ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "    sns.boxplot(data=df_merge, x='Timepoint', y='shannon_entropy', orient='v', hue='Timepoint', boxprops=dict(alpha=.3))\n",
    "    \n",
    "    # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "    ax.set_ylabel('Shannon Entropy',fontsize=16)\n",
    "    ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "    ax.tick_params(labelsize=16)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/wsr_meta_norm_resp_' + g  + '.pdf')\n",
    "    plt.close()            \n",
    "    \n",
    "    df_merge = df_merge2.dropna()\n",
    "    \n",
    "    df_merge = df_merge[df_merge['WOMAC_P_Response'] == 'No response']\n",
    "\n",
    "    prevals = df_merge[df_merge['Timepoint'] == 'pre']['shannon_entropy'].values\n",
    "    postvals = df_merge[df_merge['Timepoint'] == 'post']['shannon_entropy'].values\n",
    "    # get paired differences\n",
    "    # it looks like the rows are in the correct order will test\n",
    "    # for x,y in zip(df_pre.index, df_post.index):\n",
    "    #    if x != y:\n",
    "    #        print(x,y)\n",
    "\n",
    "    W,p = scipy.stats.wilcoxon(x=prevals,y=postvals, nan_policy='omit')\n",
    "    print('nr only pre post')\n",
    "    print(W,p)\n",
    "    ax = sns.swarmplot(data=df_merge, x='Timepoint', y='shannon_entropy', hue='Timepoint', palette=['b','r'], orient='v')\n",
    "    for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "        ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "    sns.boxplot(data=df_merge, x='Timepoint', y='shannon_entropy', orient='v', hue='Timepoint', boxprops=dict(alpha=.3))\n",
    "    \n",
    "    # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "    ax.set_ylabel('Shannon Entropy',fontsize=16)\n",
    "    ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "    ax.tick_params(labelsize=16)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/wsr_meta_norm_noresp_' + g  + '.pdf')\n",
    "    plt.close()            \n",
    "df_dx.head()   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "18623504-6fef-4780-85bc-80f548122cf0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "pre\n",
      "(0.9741152243294435, 0.395)\n",
      "post\n",
      "(0.6185035360941357, 0.545)\n",
      "saliva\n",
      "pre\n",
      "(0.5578591148147687, 0.91)\n",
      "post\n",
      "(0.2900106360994007, 0.77)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OAD001</th>\n",
       "      <th>OAD003</th>\n",
       "      <th>OAD004</th>\n",
       "      <th>OAD006</th>\n",
       "      <th>OAD007</th>\n",
       "      <th>OAD009</th>\n",
       "      <th>OAD012</th>\n",
       "      <th>OAD017</th>\n",
       "      <th>VAOAD001</th>\n",
       "      <th>VAOAD004</th>\n",
       "      <th>VAOAD011</th>\n",
       "      <th>VAOAD012</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.7184</td>\n",
       "      <td>0.7239</td>\n",
       "      <td>0.6918</td>\n",
       "      <td>0.8683</td>\n",
       "      <td>0.5698</td>\n",
       "      <td>0.6491</td>\n",
       "      <td>0.6825</td>\n",
       "      <td>0.6369</td>\n",
       "      <td>0.6928</td>\n",
       "      <td>0.5745</td>\n",
       "      <td>0.7305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>0.7184</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5050</td>\n",
       "      <td>0.5277</td>\n",
       "      <td>0.8907</td>\n",
       "      <td>0.7190</td>\n",
       "      <td>0.5730</td>\n",
       "      <td>0.5424</td>\n",
       "      <td>0.5941</td>\n",
       "      <td>0.7134</td>\n",
       "      <td>0.6239</td>\n",
       "      <td>0.7060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>0.7239</td>\n",
       "      <td>0.5050</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6775</td>\n",
       "      <td>0.8434</td>\n",
       "      <td>0.6633</td>\n",
       "      <td>0.6060</td>\n",
       "      <td>0.6984</td>\n",
       "      <td>0.6539</td>\n",
       "      <td>0.6216</td>\n",
       "      <td>0.7704</td>\n",
       "      <td>0.5873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>0.6918</td>\n",
       "      <td>0.5277</td>\n",
       "      <td>0.6775</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9315</td>\n",
       "      <td>0.6965</td>\n",
       "      <td>0.5613</td>\n",
       "      <td>0.4569</td>\n",
       "      <td>0.5652</td>\n",
       "      <td>0.7106</td>\n",
       "      <td>0.5498</td>\n",
       "      <td>0.7555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD007</th>\n",
       "      <td>0.8683</td>\n",
       "      <td>0.8907</td>\n",
       "      <td>0.8434</td>\n",
       "      <td>0.9315</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.8197</td>\n",
       "      <td>0.8312</td>\n",
       "      <td>0.9177</td>\n",
       "      <td>0.8834</td>\n",
       "      <td>0.8518</td>\n",
       "      <td>0.9342</td>\n",
       "      <td>0.8320</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        OAD001  OAD003  OAD004  OAD006  OAD007  OAD009  OAD012  OAD017  \\\n",
       "OAD001  0.0000  0.7184  0.7239  0.6918  0.8683  0.5698  0.6491  0.6825   \n",
       "OAD003  0.7184  0.0000  0.5050  0.5277  0.8907  0.7190  0.5730  0.5424   \n",
       "OAD004  0.7239  0.5050  0.0000  0.6775  0.8434  0.6633  0.6060  0.6984   \n",
       "OAD006  0.6918  0.5277  0.6775  0.0000  0.9315  0.6965  0.5613  0.4569   \n",
       "OAD007  0.8683  0.8907  0.8434  0.9315  0.0000  0.8197  0.8312  0.9177   \n",
       "\n",
       "        VAOAD001  VAOAD004  VAOAD011  VAOAD012  \n",
       "OAD001    0.6369    0.6928    0.5745    0.7305  \n",
       "OAD003    0.5941    0.7134    0.6239    0.7060  \n",
       "OAD004    0.6539    0.6216    0.7704    0.5873  \n",
       "OAD006    0.5652    0.7106    0.5498    0.7555  \n",
       "OAD007    0.8834    0.8518    0.9342    0.8320  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PERMANOVA OTU\n",
    "# this is testing resp vs non resp at pre and post time points only, NOT the paired change in beta post-pre \n",
    "# PERMANOVA\n",
    "\n",
    "for g in ['stool', 'saliva']:\n",
    "    print(g)\n",
    "    df_raw = pd.read_csv(path + 'outputs/Qiime2_' + g + '_adh/core_metrics_results/distance-matrix.tsv', \n",
    "                 sep='\\t', index_col=0)\n",
    "    \n",
    "    for t in ['pre', 'post']:\n",
    "        print(t)\n",
    "        # find overlapping samples\n",
    "        keep = df_dx.dropna().index.values        \n",
    "        df = df_raw[df_raw.index.str.contains(t)] # df[df.index.str.contains('b')]\n",
    "        df = df.T\n",
    "        df = df[df.index.str.contains(t)].T\n",
    "\n",
    "        # fix strings \n",
    "        df.index= df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "        df.columns = df.columns.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "\n",
    "        # filter indices\n",
    "        df = df[df.index.isin(keep)] # df[df.index.str.contains('b')]\n",
    "        df = df.T\n",
    "        df = df[df.index.isin(keep)].T\n",
    "\n",
    "        # get dx\n",
    "        df_dx = type_to_df_map[label.split('_')[0]].drop_duplicates(subset='HostSubjectId')\n",
    "        df_dx = df_dx.set_index('HostSubjectId')\n",
    "        df_dx.index = df_dx.index.map(lambda x: x.replace('-',''))\n",
    "        df_dx = df_dx[['Timepoint','WOMAC_P_Response']]\n",
    "        \n",
    "        # create dx label\n",
    "        dx = df_dx.loc[df.index.values,'WOMAC_P_Response']\n",
    "    \n",
    "        # takes in distance matrix\n",
    "        print(permanova_oneway(df, levels=[1 if x == 'No response' else 0 for x in dx]))\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "c7219e20-315c-42bc-abee-76211b89e75c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "(1.0307925253163521, 0.32)\n",
      "plasma\n",
      "(2.229966403427277, 0.0)\n"
     ]
    }
   ],
   "source": [
    "# would need to take the distance matrix of all samples\n",
    "# and then take the subset list of distances (flatten the dm) for each pre to post sample and compare those via boxplot\n",
    "# between responders and nonresponders\n",
    "\n",
    "for g in ['stool','plasma','saliva']:\n",
    "    print(g)\n",
    "    df = gt_to_meta[g]['all'].copy()\n",
    "    \n",
    "    orig_idx = df.index\n",
    "    \n",
    "    df2 = df.copy()\n",
    "    #df2.index = df2.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "\n",
    "    # df2['Timepoint'] = df2.index.map(df_dx.to_dict()['Timepoint'])\n",
    "    df_map = type_to_df_map[g].copy()\n",
    "    df = pd.concat([df2,df_map['Timepoint']],axis=1)\n",
    "    df = df.dropna()\n",
    "\n",
    "    #df2.index = orig_idx\n",
    "    #df = df2.copy()\n",
    "    #df = df.dropna()\n",
    "    dx = df['Timepoint'].values\n",
    "    df = df.drop('Timepoint',axis=1)\n",
    "    \n",
    "    # df.index = df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "    # from scipy.spatial.distance import cdist\n",
    "    #dm = pd.DataFrame(cdist(gt_to_meta['stool']['Baseline'], gt_to_meta['stool']['After diet'], 'euclidean'))\n",
    "    \n",
    "    dm = pd.DataFrame(squareform(pdist(df.iloc[:, 1:])), columns=df.index.unique(), index=df.index.unique())\n",
    "    # takes in distance matrix\n",
    "    print(permanova_oneway(dm, levels=[1 if (x == 'After diet') | (x == 'post') else 0 for x in dx]))\n",
    "\n",
    "    \n",
    "    from sklearn.manifold import MDS\n",
    "    \n",
    "    mds = MDS(n_components=2, dissimilarity='precomputed')\n",
    "    coordinates = mds.fit_transform(dm)\n",
    "    \n",
    "    df_pca = pd.DataFrame(coordinates, columns=['pc1','pc2'], index=df.index.unique())\n",
    "    df_pca['Timepoint'] = dx\n",
    "    \n",
    "    plt.figure(figsize=(6,6))\n",
    "    plt.title(g + '_metabolome')\n",
    "    sns.scatterplot(data=df_pca, x='pc1',y='pc2',hue='Timepoint')\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/pca_Timepoint_' + g + '.pdf')\n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "97ab1927-8204-4cd3-ad0d-895ae698d482",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "pre\n",
      "(0.7309599705911948, 0.595)\n",
      "post\n",
      "(1.0518890722123193, 0.315)\n",
      "plasma\n",
      "pre\n",
      "(-0.125552000703509, 0.94)\n",
      "post\n",
      "(0.13034044046710136, 0.825)\n"
     ]
    }
   ],
   "source": [
    "# would need to take the distance matrix of all samples\n",
    "# and then take the subset list of distances (flatten the dm) for each pre to post sample and compare those via boxplot\n",
    "# between responders and nonresponders\n",
    "\n",
    "for g in ['stool','plasma','saliva']:\n",
    "    print(g)\n",
    "    df = gt_to_meta[g]['all'].copy()\n",
    "    \n",
    "    orig_idx = df.index\n",
    "    \n",
    "    df2 = df.copy()\n",
    "    df2.index = df2.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "\n",
    "    df2['WOMAC_P_Response'] = df2.index.map(df_dx.to_dict()['WOMAC_P_Response'])\n",
    "    \n",
    "    df2.index = orig_idx\n",
    "\n",
    "    for x in ['pre','post']:\n",
    "        print(x)\n",
    "        df = df2.copy()\n",
    "        df = df.dropna()\n",
    "        df = df[df.index.str.contains(x)]\n",
    "        \n",
    "        dx = df['WOMAC_P_Response'].values\n",
    "        df = df.drop('WOMAC_P_Response',axis=1)\n",
    "        \n",
    "        # df.index = df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "        # from scipy.spatial.distance import cdist\n",
    "        #dm = pd.DataFrame(cdist(gt_to_meta['stool']['Baseline'], gt_to_meta['stool']['After diet'], 'euclidean'))\n",
    "        \n",
    "        dm = pd.DataFrame(squareform(pdist(df.iloc[:, 1:])), columns=df.index.unique(), index=df.index.unique())\n",
    "        # takes in distance matrix\n",
    "        print(permanova_oneway(dm, levels=[1 if x == 'No response' else 0 for x in dx]))\n",
    "    \n",
    "        \n",
    "        from sklearn.manifold import MDS\n",
    "        \n",
    "        mds = MDS(n_components=2, dissimilarity='precomputed')\n",
    "        coordinates = mds.fit_transform(dm)\n",
    "        \n",
    "        df_pca = pd.DataFrame(coordinates, columns=['pc1','pc2'], index=df.index.unique())\n",
    "        df_pca['WOMAC_P_Response'] = dx\n",
    "        \n",
    "        plt.figure(figsize=(6,6))\n",
    "        plt.title(g + '_' + x + '_metabolome')\n",
    "        sns.scatterplot(data=df_pca, x='pc1',y='pc2',hue='WOMAC_P_Response')\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/jobs18/pca_' + x + '_WOMAC_P_Response_' + g + '.pdf')\n",
    "        plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "85644667-1a3b-4756-a371-3f016098cca4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "Response\n",
      "(1.0240284216352826, 0.425)\n",
      "No response\n",
      "(0.9528270162591223, 0.43)\n",
      "plasma\n",
      "Response\n",
      "(2.128994036597561, 0.02)\n",
      "No response\n",
      "(0.3074960085333919, 0.055)\n"
     ]
    }
   ],
   "source": [
    "# would need to take the distance matrix of all samples\n",
    "# and then take the subset list of distances (flatten the dm) for each pre to post sample and compare those via boxplot\n",
    "# between responders and nonresponders\n",
    "# responders and nr only\n",
    "for g in ['stool','plasma','saliva']:\n",
    "    print(g)\n",
    "    df = gt_to_meta[g]['all'].copy()\n",
    "    \n",
    "    orig_idx = df.index\n",
    "    \n",
    "    df2 = df.copy()\n",
    "    #df2.index = df2.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "\n",
    "    # df2['Timepoint'] = df2.index.map(df_dx.to_dict()['Timepoint'])\n",
    "    df_map = type_to_df_map[g].copy()\n",
    "\n",
    "    for x in ['Response', 'No response']:\n",
    "        print(x)\n",
    "        df = pd.concat([df2,df_map[['Timepoint','WOMAC_P_Response']]],axis=1)\n",
    "        df = df.dropna()\n",
    "        df = df[df['WOMAC_P_Response'] == x]\n",
    "        #df2.index = orig_idx\n",
    "        #df = df2.copy()\n",
    "        #df = df.dropna()\n",
    "        dx = df['Timepoint'].values\n",
    "        df = df.drop(['Timepoint','WOMAC_P_Response'],axis=1)\n",
    "        \n",
    "        # df.index = df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "        # from scipy.spatial.distance import cdist\n",
    "        #dm = pd.DataFrame(cdist(gt_to_meta['stool']['Baseline'], gt_to_meta['stool']['After diet'], 'euclidean'))\n",
    "        \n",
    "        dm = pd.DataFrame(squareform(pdist(df.iloc[:, 1:])), columns=df.index.unique(), index=df.index.unique())\n",
    "        # takes in distance matrix\n",
    "        print(permanova_oneway(dm, levels=[1 if (x == 'After diet') | (x == 'post') else 0 for x in dx]))\n",
    "    \n",
    "        \n",
    "        from sklearn.manifold import MDS\n",
    "        \n",
    "        mds = MDS(n_components=2, dissimilarity='precomputed')\n",
    "        coordinates = mds.fit_transform(dm)\n",
    "        \n",
    "        df_pca = pd.DataFrame(coordinates, columns=['pc1','pc2'], index=df.index.unique())\n",
    "        df_pca['Timepoint'] = dx\n",
    "        \n",
    "        plt.figure(figsize=(6,6))\n",
    "        plt.title(g + '_' + x + '_metabolome')\n",
    "        sns.scatterplot(data=df_pca, x='pc1',y='pc2',hue='Timepoint')\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/jobs18/pca_' + x + '_Timepoint_' + g + '.pdf')\n",
    "        plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caed0a56-a43e-4873-a3db-9097292710bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PCA diet\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "df = str_to_df['diet']\n",
    "\n",
    "# get merged df and drop na\n",
    "df_merge = pd.concat([df, df_dx], axis=1)\n",
    "df_merge = df_merge.dropna()\n",
    "\n",
    "# construct distance matrix from df\n",
    "df_data = df.loc[df_merge.index,:]\n",
    "dx = df_dx.loc[df_merge.index,:]['WOMAC_P_Response']\n",
    "dm = pd.DataFrame(squareform(pdist(df_data.iloc[:, 1:])), columns=df_data.index.unique(), index=df_data.index.unique())\n",
    "\n",
    "# takes in distance matrix\n",
    "print(permanova_oneway(dm, levels=[1 if x == 'No response' else 0 for x in dx]))\n",
    "\n",
    "npcs = 3\n",
    "pca = PCA(n_components=3, whiten=True)\n",
    "pca.fit(df_data.T)\n",
    "# Reformat and view results\n",
    "loadings = pd.DataFrame(pca.components_,\n",
    "                        index=['PC%s' % _ for _ in range(npcs)], \n",
    "                        columns=df_data.index).T\n",
    "\n",
    "df_merge = pd.concat([df_dx, df_data, loadings], axis=1)\n",
    "plt.title('Diet_PCA')\n",
    "sns.scatterplot(data=df_merge, x='PC1', y='PC2', hue='WOMAC_P_Response')\n",
    "sns.despine()\n",
    "plt.savefig(path + 'outputs/jobs06/PCA_diet_Response.pdf')\n",
    "plt.close() "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8303f089-f331-4356-b6df-487184f8f8e3",
   "metadata": {},
   "source": [
    "###### Scratch ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "864433e8-4624-44d1-acac-900038e98d7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "c51924d3-36bf-4608-861f-0c40fe634333",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Akker\n",
      "1\n",
      "g__Butyribacter\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "# plotting\n",
    "df_pre = gt_to_otu['stool']['pre']\n",
    "df_post = gt_to_otu['stool']['post']\n",
    "df = gt_to_otu['stool']['all']\n",
    "# df = pd.concat([df,df_dx],axis=1)\n",
    "map = type_to_df_map['stool']\n",
    "df = pd.concat([df,map[['Timepoint','WOMAC_P_Response']]],axis=1)\n",
    "#map.index = map.index.map(lambda x: x.split('.p')[0].replace('-','')\n",
    "df = df.dropna()\n",
    "\n",
    "df['SubjectID'] = df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "df_vc = df['SubjectID'].value_counts() == 2\n",
    "df_vc = df_vc[df_vc == True]\n",
    "keep_id = df_vc.index.values\n",
    "df = df[df['SubjectID'].isin(keep_id)]\n",
    "\n",
    "queries = ['Akker','g__Butyribacter']\n",
    "for q in queries:\n",
    "    print(q)\n",
    "    taxa = []\n",
    "    \n",
    "    for x in df.columns.values:\n",
    "        if q in x:\n",
    "            taxa.append(x)\n",
    "    print(len(taxa))\n",
    "    plt.figure(figsize=(6,6))\n",
    "    sns.swarmplot(x=\"Timepoint\", y=taxa[0],\n",
    "                hue=\"WOMAC_P_Response\", palette=[\"m\", \"g\"],\n",
    "                data=df,dodge=True)\n",
    "    sns.boxplot(x=\"Timepoint\", y=taxa[0],\n",
    "                hue=\"WOMAC_P_Response\", palette=[\"m\", \"g\"], boxprops={'alpha': 0.1},\n",
    "                data=df)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    # plt.xlabel(x.split('f__')[-1])\n",
    "    plt.savefig(path + 'outputs/jobs18/' + taxa[0].split('f__')[-1] + '.pdf')\n",
    "    plt.close()\n",
    "\n",
    "# sns.despine(offset=10, trim=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "d55fac63-a786-4d41-96f9-3233c48556f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Butyrivibrio_A_180067',\n",
       " 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Butyribacter',\n",
       " 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Butyricicoccaceae;g__Agathobaculum',\n",
       " 'd__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Marinifilaceae;g__Butyricimonas',\n",
       " 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Butyricicoccaceae;g__Butyricicoccus_A_77419',\n",
       " 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Butyricicoccaceae;g__Butyricicoccus_A_77030',\n",
       " 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Butyricicoccaceae;g__Pseudobutyricicoccus']"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "id": "8d51feb2-1356-4615-81a9-ae357b3f93e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_N-Acetylmuramic-Acid</th>\n",
       "      <th>H_D_Fructose</th>\n",
       "      <th>H_Paraxanthine</th>\n",
       "      <th>H_L-methionine</th>\n",
       "      <th>H_Xanthuneric_acid</th>\n",
       "      <th>H_Pyridoxic_acid</th>\n",
       "      <th>H_p_coumaric_acid</th>\n",
       "      <th>H_Lactose</th>\n",
       "      <th>H_Pantothenic_acid</th>\n",
       "      <th>...</th>\n",
       "      <th>H_Indole_3_propionic_acid</th>\n",
       "      <th>H_Yersiniabactin_Fe_adduct</th>\n",
       "      <th>H_Glycocholic_acid</th>\n",
       "      <th>H_Glycocholate</th>\n",
       "      <th>H_glycocholic_acid</th>\n",
       "      <th>H_Glycochenodeoxycholate</th>\n",
       "      <th>H_Deoxycholate</th>\n",
       "      <th>H_PE(22:6/0:0)_[M+H]+_C27H45N1O7P1</th>\n",
       "      <th>H_PE(18:2/0:0)_[M+H]+_C23H45N1O7P1</th>\n",
       "      <th>WOMAC_P_Response</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.304465</td>\n",
       "      <td>2.126152</td>\n",
       "      <td>0.822249</td>\n",
       "      <td>0.076629</td>\n",
       "      <td>0.047375</td>\n",
       "      <td>-0.358213</td>\n",
       "      <td>0.075879</td>\n",
       "      <td>-0.994186</td>\n",
       "      <td>-1.218122</td>\n",
       "      <td>-1.071247</td>\n",
       "      <td>...</td>\n",
       "      <td>0.241408</td>\n",
       "      <td>0.076632</td>\n",
       "      <td>0.076635</td>\n",
       "      <td>0.076635</td>\n",
       "      <td>0.076634</td>\n",
       "      <td>0.076635</td>\n",
       "      <td>0.076635</td>\n",
       "      <td>0.105587</td>\n",
       "      <td>-0.173754</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>-0.045594</td>\n",
       "      <td>1.360729</td>\n",
       "      <td>0.192596</td>\n",
       "      <td>-0.190951</td>\n",
       "      <td>0.000671</td>\n",
       "      <td>-0.181145</td>\n",
       "      <td>0.307831</td>\n",
       "      <td>0.919771</td>\n",
       "      <td>0.027777</td>\n",
       "      <td>-0.311845</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.221883</td>\n",
       "      <td>1.181489</td>\n",
       "      <td>-0.079546</td>\n",
       "      <td>-0.079546</td>\n",
       "      <td>-0.079542</td>\n",
       "      <td>-0.646926</td>\n",
       "      <td>-1.303523</td>\n",
       "      <td>0.525658</td>\n",
       "      <td>1.038608</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>-0.321044</td>\n",
       "      <td>1.809679</td>\n",
       "      <td>0.044874</td>\n",
       "      <td>-0.329696</td>\n",
       "      <td>0.202123</td>\n",
       "      <td>-0.343900</td>\n",
       "      <td>0.401125</td>\n",
       "      <td>-0.523784</td>\n",
       "      <td>0.070289</td>\n",
       "      <td>0.243435</td>\n",
       "      <td>...</td>\n",
       "      <td>0.957739</td>\n",
       "      <td>0.097321</td>\n",
       "      <td>1.236516</td>\n",
       "      <td>1.251290</td>\n",
       "      <td>0.921056</td>\n",
       "      <td>0.999981</td>\n",
       "      <td>-1.234030</td>\n",
       "      <td>0.077983</td>\n",
       "      <td>0.799255</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>-0.140653</td>\n",
       "      <td>1.746173</td>\n",
       "      <td>-0.013675</td>\n",
       "      <td>0.111076</td>\n",
       "      <td>0.017588</td>\n",
       "      <td>-0.251973</td>\n",
       "      <td>-0.040747</td>\n",
       "      <td>0.332405</td>\n",
       "      <td>-0.106477</td>\n",
       "      <td>-0.168445</td>\n",
       "      <td>...</td>\n",
       "      <td>0.285837</td>\n",
       "      <td>0.986869</td>\n",
       "      <td>-0.166403</td>\n",
       "      <td>0.608725</td>\n",
       "      <td>1.424891</td>\n",
       "      <td>0.674575</td>\n",
       "      <td>-0.166402</td>\n",
       "      <td>-0.138608</td>\n",
       "      <td>-0.974873</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>0.202422</td>\n",
       "      <td>0.572669</td>\n",
       "      <td>0.739459</td>\n",
       "      <td>-0.325868</td>\n",
       "      <td>0.084594</td>\n",
       "      <td>0.416094</td>\n",
       "      <td>-0.119608</td>\n",
       "      <td>-1.302992</td>\n",
       "      <td>0.123590</td>\n",
       "      <td>0.399424</td>\n",
       "      <td>...</td>\n",
       "      <td>0.680529</td>\n",
       "      <td>0.169402</td>\n",
       "      <td>1.222726</td>\n",
       "      <td>1.270391</td>\n",
       "      <td>0.745385</td>\n",
       "      <td>0.508608</td>\n",
       "      <td>0.123904</td>\n",
       "      <td>0.200214</td>\n",
       "      <td>0.168454</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        H_Pyridoxamine  H_N-Acetylmuramic-Acid  H_D_Fructose  H_Paraxanthine  \\\n",
       "OAD001        0.304465                2.126152      0.822249        0.076629   \n",
       "OAD003       -0.045594                1.360729      0.192596       -0.190951   \n",
       "OAD004       -0.321044                1.809679      0.044874       -0.329696   \n",
       "OAD005       -0.140653                1.746173     -0.013675        0.111076   \n",
       "OAD006        0.202422                0.572669      0.739459       -0.325868   \n",
       "\n",
       "        H_L-methionine  H_Xanthuneric_acid  H_Pyridoxic_acid  \\\n",
       "OAD001        0.047375           -0.358213          0.075879   \n",
       "OAD003        0.000671           -0.181145          0.307831   \n",
       "OAD004        0.202123           -0.343900          0.401125   \n",
       "OAD005        0.017588           -0.251973         -0.040747   \n",
       "OAD006        0.084594            0.416094         -0.119608   \n",
       "\n",
       "        H_p_coumaric_acid  H_Lactose  H_Pantothenic_acid  ...  \\\n",
       "OAD001          -0.994186  -1.218122           -1.071247  ...   \n",
       "OAD003           0.919771   0.027777           -0.311845  ...   \n",
       "OAD004          -0.523784   0.070289            0.243435  ...   \n",
       "OAD005           0.332405  -0.106477           -0.168445  ...   \n",
       "OAD006          -1.302992   0.123590            0.399424  ...   \n",
       "\n",
       "        H_Indole_3_propionic_acid  H_Yersiniabactin_Fe_adduct  \\\n",
       "OAD001                   0.241408                    0.076632   \n",
       "OAD003                  -0.221883                    1.181489   \n",
       "OAD004                   0.957739                    0.097321   \n",
       "OAD005                   0.285837                    0.986869   \n",
       "OAD006                   0.680529                    0.169402   \n",
       "\n",
       "        H_Glycocholic_acid  H_Glycocholate  H_glycocholic_acid  \\\n",
       "OAD001            0.076635        0.076635            0.076634   \n",
       "OAD003           -0.079546       -0.079546           -0.079542   \n",
       "OAD004            1.236516        1.251290            0.921056   \n",
       "OAD005           -0.166403        0.608725            1.424891   \n",
       "OAD006            1.222726        1.270391            0.745385   \n",
       "\n",
       "        H_Glycochenodeoxycholate  H_Deoxycholate  \\\n",
       "OAD001                  0.076635        0.076635   \n",
       "OAD003                 -0.646926       -1.303523   \n",
       "OAD004                  0.999981       -1.234030   \n",
       "OAD005                  0.674575       -0.166402   \n",
       "OAD006                  0.508608        0.123904   \n",
       "\n",
       "        H_PE(22:6/0:0)_[M+H]+_C27H45N1O7P1  \\\n",
       "OAD001                            0.105587   \n",
       "OAD003                            0.525658   \n",
       "OAD004                            0.077983   \n",
       "OAD005                           -0.138608   \n",
       "OAD006                            0.200214   \n",
       "\n",
       "        H_PE(18:2/0:0)_[M+H]+_C23H45N1O7P1  WOMAC_P_Response  \n",
       "OAD001                           -0.173754       No response  \n",
       "OAD003                            1.038608       No response  \n",
       "OAD004                            0.799255       No response  \n",
       "OAD005                           -0.974873               NaN  \n",
       "OAD006                            0.168454          Response  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 350,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CUTIE\n",
    "# In POST samples only\n",
    "# Color by Resp and Nonresp\n",
    "\n",
    "for g in ['saliva', 'stool', 'plasma']:\n",
    "    # grab df map with diagnses etc. \n",
    "    df_map = type_to_df_map[g].set_index('Together')\n",
    "    df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "\n",
    "    if g == 'stool' or g == 'saliva':\n",
    "        df_taxa_post = gt_to_otu[g]['post']\n",
    "        df_taxa_post.to_csv(path + 'inputs/cutie_df_taxa_post_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "        df_taxa_pre = gt_to_otu[g]['pre']\n",
    "        df_taxa_pre.to_csv(path + 'inputs/cutie_df_taxa_pre_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "\n",
    "        df_taxa_pre.index = df_taxa_pre.index.map(lambda x: x.split('.')[0].replace('-',''))\n",
    "        df_taxa_post.index = df_taxa_post.index.map(lambda x: x.split('.')[0].replace('-',''))\n",
    "\n",
    "        df_taxa_diff = df_taxa_post - df_taxa_pre\n",
    "        df_taxa_diff.to_csv(path + 'inputs/cutie_df_taxa_diff_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "        df_taxa_diff = pd.concat([df_taxa_diff,df_map], axis=1)\n",
    "        # get resp only\n",
    "        df_taxa_diff_resp = df_taxa_diff[df_taxa_diff['WOMAC_P_Response'] == 'Response']\n",
    "        df_taxa_diff_nr = df_taxa_diff[df_taxa_diff['WOMAC_P_Response'] == 'No response']\n",
    "        df_taxa_diff_resp.to_csv(path + 'inputs/cutie_df_taxa_diff_resp_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "        df_taxa_diff_nr.to_csv(path + 'inputs/cutie_df_taxa_diff_nr_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "        #df_abun = gt_to_otu[g]['post'].copy() #, gt_to_path[g]['post']],axis=1)\n",
    "        #df_abun = pd.concat([gt_to_otu[g]['post'], gt_to_path[g]['post']],axis=1)\n",
    "        #df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "        #df_abun.to_csv(path + 'inputs/cutie_df_abun_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "    \n",
    "    df_meta_post = gt_to_meta[g]['After diet']\n",
    "    df_meta_pre = gt_to_meta[g]['Baseline']\n",
    "    df_meta_diff = df_meta_post - df_meta_pre\n",
    "        \n",
    "    df_meta_post = pd.concat([df_meta_post,df_map], axis=1)\n",
    "    df_meta_post.to_csv(path + 'inputs/cutie_df_meta_post_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "    df_meta_pre = pd.concat([df_meta_pre,df_map], axis=1)\n",
    "    df_meta_pre.to_csv(path + 'inputs/cutie_df_meta_pre_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "    df_meta_diff = pd.concat([df_meta_diff,df_map], axis=1)\n",
    "    df_meta_diff.to_csv(path + 'inputs/cutie_df_meta_diff_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "    # get resp only\n",
    "    df_meta_diff_resp = df_meta_diff[df_meta_diff['WOMAC_P_Response'] == 'Response']\n",
    "    df_meta_diff_nr = df_meta_diff[df_meta_diff['WOMAC_P_Response'] == 'No response']\n",
    "    df_meta_diff_resp.to_csv(path + 'inputs/cutie_df_meta_diff_resp_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "    df_meta_diff_nr.to_csv(path + 'inputs/cutie_df_meta_diff_nr_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "\n",
    "\n",
    "df_meta_diff.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "id": "b5d49a9b-36b5-4d66-919b-6183f3433d6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "diff_diff\n",
      "Total Corr: 2552\n",
      "TP: 39\n",
      "FP: 57\n",
      "Match length: 0\n",
      "pre_pre\n",
      "Total Corr: 2552\n",
      "TP: 34\n",
      "FP: 52\n",
      "Match length: 0\n",
      "post_post\n",
      "Total Corr: 2552\n",
      "TP: 72\n",
      "FP: 82\n",
      "Match length: 3\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var1</th>\n",
       "      <th>var2</th>\n",
       "      <th>pvalues</th>\n",
       "      <th>correlations</th>\n",
       "      <th>r2vals</th>\n",
       "      <th>TP_rev_indicators</th>\n",
       "      <th>FP_rev_indicators</th>\n",
       "      <th>extreme_p</th>\n",
       "      <th>extreme_r</th>\n",
       "      <th>p_ratio</th>\n",
       "      <th>r2_ratio</th>\n",
       "      <th>class</th>\n",
       "      <th>reverse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>470</th>\n",
       "      <td>H_Pyridoxic_acid</td>\n",
       "      <td>H_Pyridoxic_acid</td>\n",
       "      <td>0.022057</td>\n",
       "      <td>-0.604396</td>\n",
       "      <td>0.365294</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.060737</td>\n",
       "      <td>-0.532967</td>\n",
       "      <td>2.753633</td>\n",
       "      <td>0.777603</td>\n",
       "      <td>FP</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>796</th>\n",
       "      <td>H_Lenticin</td>\n",
       "      <td>H_Lenticin</td>\n",
       "      <td>0.023308</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.360000</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.070749</td>\n",
       "      <td>0.516484</td>\n",
       "      <td>3.035344</td>\n",
       "      <td>0.740987</td>\n",
       "      <td>FP</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1678</th>\n",
       "      <td>H_Deoxycholate</td>\n",
       "      <td>H_Deoxycholate</td>\n",
       "      <td>0.002925</td>\n",
       "      <td>0.731868</td>\n",
       "      <td>0.535631</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.010340</td>\n",
       "      <td>0.681319</td>\n",
       "      <td>3.535629</td>\n",
       "      <td>0.866632</td>\n",
       "      <td>TP</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  var1              var2   pvalues  correlations    r2vals  \\\n",
       "470   H_Pyridoxic_acid  H_Pyridoxic_acid  0.022057     -0.604396  0.365294   \n",
       "796         H_Lenticin        H_Lenticin  0.023308      0.600000  0.360000   \n",
       "1678    H_Deoxycholate    H_Deoxycholate  0.002925      0.731868  0.535631   \n",
       "\n",
       "      TP_rev_indicators  FP_rev_indicators  extreme_p  extreme_r   p_ratio  \\\n",
       "470                -1.0               -1.0   0.060737  -0.532967  2.753633   \n",
       "796                -1.0               -1.0   0.070749   0.516484  3.035344   \n",
       "1678               -1.0               -1.0   0.010340   0.681319  3.535629   \n",
       "\n",
       "      r2_ratio class reverse  \n",
       "470   0.777603    FP      No  \n",
       "796   0.740987    FP      No  \n",
       "1678  0.866632    TP      No  "
      ]
     },
     "execution_count": 351,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# correlation analyses\n",
    "# jobs25 is plasma meta diff with stool meta diff\n",
    "# var1 is stool, var2 is plasma\n",
    "\n",
    "comp_to_job = {\n",
    "    'diff_diff': 'jobs25',\n",
    "    'pre_pre': 'jobs26',\n",
    "    'post_post': 'jobs27'\n",
    "}\n",
    "\n",
    "for c in comp_to_job:\n",
    "    print(c)\n",
    "\n",
    "    df = pd.read_csv(path + 'outputs/' + comp_to_job[c] + '/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "    \n",
    "    # get number of total corr, TP and FP\n",
    "    print('Total Corr: ' + str(len(df[df['pvalues'] != np.nan])))\n",
    "    print('TP: ' + str(len(df[df['class'] == 'TP'])))\n",
    "    print('FP: ' + str(len(df[df['class'] == 'FP'])))\n",
    "\n",
    "    # strength of matching\n",
    "    df_match = df[df['var1'] == df['var2']]\n",
    "    df_match = df_match[df_match['pvalues'] < 0.05]\n",
    "\n",
    "    print('Match length: ' + str(len(df_match)))\n",
    "\n",
    "# there's 29 plasma and 88 metabolites, only 14 are in common\n",
    "# print(len(set(gt_to_meta['plasma']['Baseline'].columns.values).intersection(set(gt_to_meta['stool']['Baseline']))))\n",
    "df_match.head()    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "id": "51a50637-3f02-46d0-9671-6af857068029",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var1</th>\n",
       "      <th>var2</th>\n",
       "      <th>metabolite_log10p</th>\n",
       "      <th>taxa_log10p</th>\n",
       "      <th>correlations_resp</th>\n",
       "      <th>correlations_nr</th>\n",
       "      <th>fishersz_diff</th>\n",
       "      <th>fz_diff_p</th>\n",
       "      <th>Interest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>H_Pyridoxamine</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.841255</td>\n",
       "      <td>0.852437</td>\n",
       "      <td>-0.900000</td>\n",
       "      <td>3.160669</td>\n",
       "      <td>0.001574</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>H_Pyridoxamine</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.604038</td>\n",
       "      <td>-0.357143</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>-2.131377</td>\n",
       "      <td>0.033058</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>H_Pyridoxamine</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>0.332352</td>\n",
       "      <td>0.037062</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>H_Pyridoxamine</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.860087</td>\n",
       "      <td>0.144150</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>H_Pyridoxamine</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>0.936098</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>-0.820783</td>\n",
       "      <td>2.819416</td>\n",
       "      <td>0.004811</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32703</th>\n",
       "      <td>H_Linoleic_Acid</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...</td>\n",
       "      <td>0.329059</td>\n",
       "      <td>-0.334352</td>\n",
       "      <td>0.321429</td>\n",
       "      <td>-0.900000</td>\n",
       "      <td>2.084764</td>\n",
       "      <td>0.037091</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32712</th>\n",
       "      <td>H_Linoleic_Acid</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...</td>\n",
       "      <td>0.329059</td>\n",
       "      <td>0.145694</td>\n",
       "      <td>-0.889499</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>-1.996525</td>\n",
       "      <td>0.045877</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32843</th>\n",
       "      <td>H_Linoleic_Acid</td>\n",
       "      <td>d__Bacteria;p__Proteobacteria;c__Gammaproteoba...</td>\n",
       "      <td>0.329059</td>\n",
       "      <td>0.498516</td>\n",
       "      <td>-0.408248</td>\n",
       "      <td>0.872082</td>\n",
       "      <td>-2.049841</td>\n",
       "      <td>0.040380</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32846</th>\n",
       "      <td>H_Linoleic_Acid</td>\n",
       "      <td>d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...</td>\n",
       "      <td>0.329059</td>\n",
       "      <td>0.183944</td>\n",
       "      <td>-0.534522</td>\n",
       "      <td>0.820783</td>\n",
       "      <td>-2.027269</td>\n",
       "      <td>0.042635</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32872</th>\n",
       "      <td>H_Linoleic_Acid</td>\n",
       "      <td>d__Bacteria;p__Fusobacteriota;c__Fusobacteriia...</td>\n",
       "      <td>0.329059</td>\n",
       "      <td>-0.963333</td>\n",
       "      <td>-0.472866</td>\n",
       "      <td>0.894427</td>\n",
       "      <td>-2.260201</td>\n",
       "      <td>0.023809</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>766 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  var1                                               var2  \\\n",
       "5       H_Pyridoxamine  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   \n",
       "66      H_Pyridoxamine  d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...   \n",
       "80      H_Pyridoxamine  d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...   \n",
       "100     H_Pyridoxamine  d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...   \n",
       "120     H_Pyridoxamine  d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...   \n",
       "...                ...                                                ...   \n",
       "32703  H_Linoleic_Acid  d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...   \n",
       "32712  H_Linoleic_Acid  d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...   \n",
       "32843  H_Linoleic_Acid  d__Bacteria;p__Proteobacteria;c__Gammaproteoba...   \n",
       "32846  H_Linoleic_Acid  d__Bacteria;p__Firmicutes_A;c__Clostridia_2584...   \n",
       "32872  H_Linoleic_Acid  d__Bacteria;p__Fusobacteriota;c__Fusobacteriia...   \n",
       "\n",
       "       metabolite_log10p  taxa_log10p  correlations_resp  correlations_nr  \\\n",
       "5              -0.090177    -0.841255           0.852437        -0.900000   \n",
       "66             -0.090177    -0.604038          -0.357143         0.900000   \n",
       "80             -0.090177     0.332352           0.037062        -1.000000   \n",
       "100            -0.090177    -0.860087           0.144150         1.000000   \n",
       "120            -0.090177     0.936098           0.857143        -0.820783   \n",
       "...                  ...          ...                ...              ...   \n",
       "32703           0.329059    -0.334352           0.321429        -0.900000   \n",
       "32712           0.329059     0.145694          -0.889499         0.300000   \n",
       "32843           0.329059     0.498516          -0.408248         0.872082   \n",
       "32846           0.329059     0.183944          -0.534522         0.820783   \n",
       "32872           0.329059    -0.963333          -0.472866         0.894427   \n",
       "\n",
       "       fishersz_diff  fz_diff_p Interest  \n",
       "5           3.160669   0.001574      nan  \n",
       "66         -2.131377   0.033058      nan  \n",
       "80          5.000000   0.000000      nan  \n",
       "100        -5.000000   0.000000      nan  \n",
       "120         2.819416   0.004811      nan  \n",
       "...              ...        ...      ...  \n",
       "32703       2.084764   0.037091      nan  \n",
       "32712      -1.996525   0.045877      nan  \n",
       "32843      -2.049841   0.040380      nan  \n",
       "32846      -2.027269   0.042635      nan  \n",
       "32872      -2.260201   0.023809      nan  \n",
       "\n",
       "[766 rows x 9 columns]"
      ]
     },
     "execution_count": 383,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# need dataframe where each row is a pair of var1, var2, and the cols are\n",
    "# log10p in taxa R vs NR  - gdt_to_res\n",
    "# log10p in metabolite R vs NR - gdt_to_res\n",
    "# correlation difference - fisher's Z from chat gpt, need to implement\n",
    "# color (pos neg of original resp corr) and size also propoprtional to above DIFFERENCE\n",
    "# Special tag for bile acids of interest\n",
    "# \n",
    "\n",
    "# start with arbitrary meta taxa df\n",
    "df = pd.read_csv(path + 'outputs/jobs28/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "df = df[df['var1'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "df = df[df['var2'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "df = df[['var1','var2']]\n",
    "\n",
    "# grab pvals\n",
    "var1_ps = []\n",
    "df_res = gdt_to_res_pvp['stool']['meta']['Response']\n",
    "meta_to_logp = df_res.set_index('meta')['logp'].to_dict()\n",
    "\n",
    "for x in df.var1.values:\n",
    "    var1_ps.append(meta_to_logp[x])\n",
    "\n",
    "# repeat for taxa\n",
    "var2_ps = []\n",
    "df_res = gdt_to_res_pvp['stool']['otu']['Response']\n",
    "taxa_to_logp = df_res.set_index('otu')['logp'].to_dict()\n",
    "\n",
    "for x in df.var2.values:\n",
    "    x = read_taxa(x,';')\n",
    "    var2_ps.append(taxa_to_logp[x])\n",
    "\n",
    "# var2_ps\n",
    "df['metabolite_log10p'] = var1_ps\n",
    "df['taxa_log10p'] = var2_ps\n",
    "\n",
    "# sample sizes from observing plots\n",
    "n_nr = 5\n",
    "n_resp = 7\n",
    "\n",
    "# from chatGPT! \n",
    "def fishers_z(r1,r2,n1,n2):\n",
    "    # Calculate Fisher's z-transform for each correlation\n",
    "    z1 = np.arctanh(r1) \n",
    "    z2 = np.arctanh(r2) \n",
    "    \n",
    "    # Calculate the standard error of the difference between z1 and z2\n",
    "    se = np.sqrt(1/(n1-3) + 1/(n2-3))\n",
    "    \n",
    "    # Calculate the z-statistic\n",
    "    z = (z1 - z2) / se \n",
    "    \n",
    "    # Calculate the p-value\n",
    "    p = 2 * (1 - scipy.stats.norm.cdf(abs(z)))\n",
    "\n",
    "    return z,p\n",
    "\n",
    "\n",
    "# bring in the correlations \n",
    "df_resp = pd.read_csv(path + 'outputs/jobs28/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "df_resp = df_resp[df_resp['var1'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "df_resp = df_resp[df_resp['var2'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "df_resp = df_resp[['var1','var2','correlations']]\n",
    "df_resp = df_resp.rename(columns={'correlations': 'correlations_resp'})\n",
    "\n",
    "df_nr = pd.read_csv(path + 'outputs/jobs29/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "df_nr = df_nr[df_nr['var1'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "df_nr = df_nr[df_nr['var2'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "df_nr = df_nr[['var1','var2','correlations']]\n",
    "df_nr = df_nr.rename(columns={'correlations': 'correlations_nr'})\n",
    "\n",
    "# assuming the rows are in proper consistent order, which they seem to be\n",
    "\n",
    "df_merge = pd.concat([df,df_resp.drop(['var1','var2'],axis=1), df_nr.drop(['var1','var2'],axis=1)],axis=1)\n",
    "\n",
    "# drop na\n",
    "# interesting design choice whether to drop correlations that are NA in only one group, TBD\n",
    "df_merge = df_merge.dropna(subset=['taxa_log10p','metabolite_log10p','correlations_resp','correlations_nr'])\n",
    "\n",
    "for index, row in df_merge.iterrows():\n",
    "    z,p = fishers_z(row.loc['correlations_resp'], row.loc['correlations_nr'], n_resp, n_nr)\n",
    "    df_merge.at[index, 'fishersz_diff'] = z\n",
    "    df_merge.at[index, 'fz_diff_p'] = p\n",
    "    \n",
    "# filter on sig    \n",
    "# df_merge = df_merge[df_merge['fz_diff_p'] < 0.2]\n",
    "df_merge = df_merge[df_merge['fz_diff_p'] < 0.05]\n",
    "\n",
    "# replace inf with 5 and -inf with -5\n",
    "df_merge['fishersz_diff'] = df_merge['fishersz_diff'].replace(np.inf, 5)\n",
    "df_merge['fishersz_diff'] = df_merge['fishersz_diff'].replace(-np.inf, -5)\n",
    "\n",
    "# flag for bile acids etc.\n",
    "int_ba0 = ['Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid']\n",
    "\n",
    "# int_ba1 = ['h_' + x for x in int_ba0] + ['H_' + x for x in int_ba0]\n",
    "\n",
    "int_ba = []\n",
    "for x in df_meta.columns:\n",
    "    for t in int_ba0:\n",
    "        if t in x:\n",
    "            int_ba.append(x)\n",
    "\n",
    "\n",
    "int_taxa = []\n",
    "for x in df_otu.columns:\n",
    "    for t in ['Faecalibacterium', 'Akkermansia','Alistipes','Phascolarctobacterium','Rikenell', 'Acetati', 'Ruminococcus']:\n",
    "        if t in x:\n",
    "            int_taxa.append(x)\n",
    "\n",
    "\n",
    "df_merge.loc[(df_merge['var1'].isin(int_ba)) | (df_merge['var2'].isin(int_taxa)), 'Interest'] = 'Yes'\n",
    "\n",
    "\n",
    "df_merge\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "id": "619ebacc-4972-4069-b7dc-53f12cc0c20f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAm8AAAHPCAYAAAAFwj37AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOydd3hcxdm37+1VW6Rd9WbLTe69genE9BYCCaGlV0gneZM3CWlv8qVAKgkhQAgdQq/GGAwG995kyZbV6660vbfvj7WF17srrW3Jluy5r8sX6Jw5Z+fMab8z88zzkyQSiQQCgUAgEAgEgjGB9FRXQCAQCAQCgUCQO0K8CQQCgUAgEIwhhHgTCAQCgUAgGEMI8SYQCAQCgUAwhhDiTSAQCAQCgWAMIcSbQCAQCAQCwRhCiDeBQCAQCASCMYQQbzlQX19PfX39qa6GQCAQCAQCgRBvuRAOhwmHw6e6GoRCIbZs2UIoFDrVVRnTiHYcPkRbDh+iLYcH0Y7Dh2jL4WEk2lGItzFELBZL+a/g+BDtOHyIthw+RFsOD6Idhw/RlsPDSLSjEG8CgUAgEAgEYwgh3gQCgUAgEAjGEEK8CQQCgUAgEIwhhHgTCAQCgUAgGEMI8SYQCAQCgUAwhhDiTSAQCAQCgWAMIcSbQCAQCAQCwRhCiDeBQCAQCASCMYQQbwKBQCAQCARjCCHeBAKBQCAQCMYQ8lNdAYHgdMXlDRGJxpFJJZgN6mPe3hsIE4nGUSvlaFTiVs2EPxjBH4wilYBRr0ImG5vfox5fGH8wTp7JgkQiSVvf7woSicaQSiWYDSrkMhmxgI+ox06wZQ+xgAeFsRBV2STkhgKkymO/3gSCM4VQOIo/FEUmlWDQqU51dY4L8UYQCIaZXoefPQf7+O+q/dicAQw6JdecW8O82iJKCnQ5bV/f4uDF9xpx+0KUWHRcd94EqooNxyUCT0fc3hDN3W6ee2c/+9ucKOQyzp9XzgXzKyiz6seMiLM5/OxrcfDfd/bT7wpiNqi47jwpU8cXUGjW0tPvZ8u+Hp5d1YDdGUSllHHxwkouWViOetvT+Ha8nbpDiZS82RdiXvYJ5HkFp+agBIJRiscXpt3m5aX3G2lsd6JWyrlsaTWzJxVSYhn62TyaEOJNIBhGevr9/OLB9bR0ewaWBUJR7n9hF/mr9vPLryylojAv6/Zddh93/2sdnTbfwLLuPj/b6m0snFrEV6+fRYFRM6LHMNpxekI8+dY+Xl/bfMTSCM+9e4CX3j/Iz76wmGk1FmTS9B6skSLi7CXYUU+geTdKSznamjko8kuQSGVZt+np9/GLBzekXCtOb4g/PLGVikI9/3P7Qv7ffzbS0u0dWB8Kx3j1gybeWt/CTz51PubWPUQdXR/tNBHHs20lUY8D6+VfRq43j8jxjkUijm6CrXsJdjSgKhqPetx0FOaSjD2dI0WX3cf2hl621vdSWWzgnNlllFp1KOTZrxPB8OD2hnhmVQMvvX8wZfl9z+3EoFPyq6+cRXWJ4RTV7tgZG5+nAsEYIBiO8shre1NexkfS7w7y2/9sxukJZlzv8Yf58zPbUoTbkWzc28PKja3E4olhq/NYZPdB+1HC7SOisTg/f3AD3X2Z23AkiPR30fmfH2F78Y94t79N/9v/puNf3yXYXp91m0AoyoMv7c56rbT1ennw5d1MrsrPuD4cjfN/zzYSW/jpzPs/sJmIve3YD+Y0JWxro+Pf/4Pt1b/h2bYS+5v30/HQ9wn3NJ20OrT1ePjeX97nvud2sn53N8+83cA37lnNjv124mf4PX0y2NXYlybcDuP2hfnZv9ZhdwZOcq2OHyHeBIJhwuYI8OGOjkHLNHe5sTkyPyDszgC7G/sG3f7F1QewO/zHXcexjtMT4vE39w1aJhSJsbvRflLqE4+E6F/zDDFPf8ryRCxC74t/JOrOfD5tjgDr93QPuu+t9b3MnGjNut4fjNIW1CPTGTOud3z4ArHgyROxo5VY0Id95cPE/e6U5YmQH9ur9xE7avlIEAglP+xc3nBq3eIJ7nliC3bX2BENYxGHJ8hjb9YNWsbuDNLR6x20zGhCiDeBYJjodwfJ5QO6scOVcXl7Dg8OXzCK0xs61qqdNnj84Zzaac32TiLR+IjXJ+Zz4dv7YeZ1nj6iXkfGdX2uAIkhrpVEAvyBCNJBhn83NflRWqsyrgt11JMICVEQ97sINu3IuC7c00TU6xzxOjg9ITbuzSzWPf4Ifa7MvfGC4cHrj+T03Nja0HsSajM8CPEmEAgEI8XJC6cSCAQnyFi6XYV4EwiGiXyDmlxi5GvKMg9zlRfqh9xWp5Zj0o/Nqe3DQZ5WmVM7LZtdhkI+8o83mc6IburZmdcZLFknDFhMGoaKk5dIQKtRDBoPtWCclrCtJeM6VdlkJKoze3ILgFRrRD1uVsZ1yqLxyPWmEa+DKU/FoqnFGdcZdEoKjGIW+UiSp1Xk9NyYPanwJNRmeBDiTSAYJqxmDWfNKhu0THWJgcJ8bcZ1FpOGGTWDp3e49rwJWMyZtz8TMOWp+PQlUwYto1LImD5EOw4XUoWK/GWfQHZUWg6JXEnh1d/Imq7DYtKwZHrJoPueO6WQnfttWdfr1HIqNV5ivszD8OazrkWmHlvpD0YCmVqH5eLPINWmziSUqLRYr/gKMu3IzzDUqOTcevlUjHplat2kEr71qblYzvAZ5CONKU/NzZfWDlrGYlLnJPBGCyJViEAwTKiVcm67fCptPW6au9JnEeYb1Hz/1vkYs/Sc5WmV3HHDHH72r3V0ZJhxunBaERctrDypKTBGIzNqLFx+VjWvfdictk4hl/KTzy+iOId8esOFIr+E0lt/SaijgUDLbhSWcrTjZ6PIzy7ONCo5n71qGh02b8YZp5VFej5/5XT+36ObMm6vUsj4n0/UIP3wr2SK7NNOWoDCUnG8h3TaobRWUHb7rwm27SPYvg9VSQ3qqukozJl7w0aCiqI8fnfHMnbst7OtvpfK4jzOnpVMFTJYXKNgeJhRU8DV54zPOOPUoFPy088vwWIaOyJakkgMFTYr2LVrFwAzZsw4pfXw+/3U1dVRW1uLVnvm9r6cKCPdjjaHn71N/TyzqgG7M4BRp+Ka82qYO7kwJ1HR6/DT0OrgxdWNuH1hSq06rj1vApVFeaMuSe+puibd3hAt3R6ee/dwkl4p58+r4Px55WMySe9z7+6nzxUk36DmuvNqqB33UZLerft6Dl1LySS9H1tUxfKF5ai3PYVve4YkvXMuxnz29cjzMqcZOd0Rz8nh43Rry8NJel9e00hjmwu1SsZlS8cxa6KFEsvI9bqNRDuKnjeBYJixmrWca9YyZ5KVcDSOXCbBlJe76Co0ayk0a5kzqZBINIZaKUct7LFSMOhVzJigoqbcSCAURYIEo145ZkTbYaxmLVazltkTrfiDIZyOfirLLGg0yR6Aonwtly4dx6LpxUSicaSSpNWaXCYlZr4V04LLCLbWEfe7kZsKUZVNRG6wIFWcuXGRAkE28nRKanX5jC81EAjFkEkl5OmUQ284ChFvBIFghDCc4MQCnUYBKIanMqcpWrUCrXrst1GeTolMEqW92U6iND23W74hfThHptEh0+hQFWZOFSIQCDKjUspRKce2/Blbn6kCgUAgEAgEZzhCvAkEAoFAIBCMIYR4EwgEAoFAIBhDCPEmEAgEAoFAMIYQ4k0gEAgEAoFgDCHEm0AgEAgEAsEYQog3gUAgEAgEgjGEEG8CgUAgEAgEYwgh3gQCgUAgEAjGEEK8CQQCgUAgEIwhhHgTCAQCgUAgGEMI8SYQCAQCgUAwhhDiTSAQCAQCgWAMIcSbQCAQCAQCwRhCiDeBQCAQCASCMYT8VFdAIDidsDv9RKJxVAoZ+UbNqa7OiOLxh3F5ouQXj8ftj6PVnuoajV08vhD+YJzC4nIkEklO20QiUbr6/NicAcKRGFqVggKzmnJr3gjXdnQjk8koLi4+1dUQjGIS8RiJSBhkMqRy5amuznEhxJtAMAy09Xiob3HwxrpmXN4QBUY1l581jpoyE2WF+lNdvWHF5Q3S1uPlpfcb2bS3h1g8QWVRHtedP4Ha6nxKrafX8Y4knTYvHTYvb61voc8dJN+g5mOLqigvjA3ajl12L2+sa+bNdS3E4wl0GgVuX4g8rZIbL57EomnFWExnjpoOhiJ09fnZsd/GrsY+lHIpy2ZLqCqJUXaGi1nBR8SDfiKuXjy7VhPqaECq0GCY+zGUJeNRGK2nunrHhBBvAsEJ0tzl5lcPb6C7zz+wrKffz96mfqpLDNx1y3wqik6PF4jLG+TdLe08+PKelOWtPR7++NQ2plSZ+fZNcymxCAE3FO29Hv7yzHb2NvWnLN+wp5spVWa+8ck5lBemXzdddi9/fmY7nTYft18xFalEgsMTpNCspd8d5KGX99Bp93HdeRMoOM17fwH8gQib9/Vw75NbicYSA8s/2NGJ1aThfz+7iPFlxlNYQ8FoIBbw4N78Bo73n05ZHmjajiwvn5KbforSUn6KanfsiJg3geAE6O7zce+TW1OE25E0d7l54MVd9DkDJ7lmI0NbjzdNuB3JvhYHz68+QDAUOYm1Gns43UEef3NfmnA7zL4WB4+8tpd+V/p1s73BRkuXmy9dO4Mn36rnb//dwRMr6vnjU9tYs72DO2+cw6trDtJp9430YYwKWno8/OHxLSnC7TA2Z4BfPrSBjl7vKaiZYDQRaN6dJtwOE/P00/3kL4m67Ce5VsePEG8CwQlgcwQ42OEatMy2Bht97uBJqtHI4fGHeen9xiHLvbOpjZ7+00OsjhR2d5C1OzsHLbNhTzf97lDKsi67l+dXH+Dys8fz+Jv7cHpS1zd1utlc18O82iJe+6AJXyA87HUfTXj8YZ5ffYB4um4bwOYM0NQ1+D0qOL2Jeh30v/vY4GXcNsL29pNUoxNHiDeB4ATYvt+WU7kDbc6RrchJwOUNsWlvz5DlwtE49tOkp3GkaO/xDCo4ABIJaOpMFR1uX4TuPj/F+VpaezwZt/tgRycLpxaztb4Xp/f0Fm8uT4iNe7qHLPfe1nbi8fhJqJFgNBLzuYg6hr5OvLveOwm1GR6EeBMIToBwJJZbuWhu5UYziQTEhlIch4gncit3phKO5iYkIrHUcolD7TrYeYjG4kilEmKxePKkncbEEwniOVyTkWicWIZhVcEZQo7CPR4NDdxjox0h3gSCE2BihSmncpXFY3/CgkohozLHiRdG3dicfn+yKCnQ5VSu7KgZpxq1HJVSlpxhqs4832xihYm2Hg9lhXpUStkJ13U0o1bKc7omp47LR6E4vdtCkB2pWodEoR6ynGbczJxT9ZxqhHgTCE6A6hJD1pfoYawmDUX5ub2sRzOF+VquPX/CkOWmVJuxmE7/WY4nQoFJTfkQKWRKCnRYj2rH4nwdFy2oZMX6Fm68eHLaNnKZlGvOreHtja1cd96E0z5dSGG+lo9fMPg1KZVKmF8r8r6dycgNFgzzlg9eSCpDUz3j5FRoGBDiTSA4AYoKtNx54xykWT7W5DIpd9w4O+eeltHO1Op8plSZs65XKWV84eoZp32C4hOlOF/HHZ+YjUaVWfirlDLuuHE2xUddNyqljEsWV9HV56O+xcG3b5rLvCmFlBfqOWdOGd/99DyeX32AfIOaCTn2Co91po4rYPG0zOJMIoE7b5hNUb64Hs9kJDIZhvmXoLBUZCtB4VV3Ih9Dud6EeBMITgCVQs7MCRZ+8eWl1FbnDyyXSGDWRCu//upZTKk0I82m7sYYpVY93/rUXC5ZUoVSnvr4mFJt5v++chbjS0VOraGQSiVMqDDyf185i2Wzy5DLkteHTCrh7Fkl/PqrZzGpwpTxuqkuNfLzLyymrrmfv/13B3laJQumFhEMx/jtY5sB+MFtCzLmiDsdKS7Q8bmrpx/Ki5fszZRIYF5tIf/31bNYOK0YnUYM45/pKIyFFH/yR5jPuwmp5qNeb3X1DEpv+xXaSfPHlNuCJDFWovNOIbt27QJgxoxT26Xq9/upq6ujtrYWrfAiOm5Gqh3tDj/9nhChcAyNSk6+UU2+Yeg4i7FIMBShuz9AnzNALB7HqFdhMWnOiKSww43bG6LPHSQYjqGQQb5BRb5x6J7a9l4P7b1eVm1qwxeIYDGpuXhhFVazJq3H7kyhp89HIBwlEY9j1CtzakdBdk7Hd04iESfq7iMRDoJUhkxnRKYe2etkJNpROCwIBMOExazFYj49HnBDoVYpqC5RUGiUU1dXR8W4WrRaIdyOB4NehUGvGnjAlxbU5rRdeWEe5YV5zKstIhKJoVbKkErP7MGUogLdQDsW1ebWjoIzC4lEOuassDIhxJtAIBCMYRQyKQrZmS3aBIIzDXHHCwQCgUAgEIwhxpx4u++++7jlllsGLeNwOPjOd77DggULWLBgAT/+8Y/x+zN7TwoEAoFAIBCMJcaUePv3v//Nn//85yHL3XnnnbS1tQ2U//DDD/nZz352EmooEAgEAoFAMLKMiZi3np4efvSjH7FlyxbGjRs3aNlt27axceNGXn/9dWpqagD4+c9/zuc//3m+/e1vU1RUdDKqLBAIBAKBQDAijImetz179mA0Gnn55ZeZNWvWoGU3b96M1WodEG4ACxcuRCKRsGXLlpGuqkAgOIlEojHcvhBef3jMeBIerrPHH87Jl1MgEAiOZkz0vF1wwQVccMEFOZXt6emhpKQkZZlSqcRkMtHV1XXcdUgkEqc8bi4QCKT8V3B8jEQ7SiQSXP4o/a4QoUgMtUpOfp4Kg1Z2XKIiEAGXN0w4EketlFGQp0QmGz0veolEgicQw+kJodCXYHeFKUhIkEhOTh39oTg2Z4jX1jbT1OlCIZNy9qxSFkwtwpynQD4KP0sD4QR2V4g317XQ0OZAJpWwbHYpi6eXEInGCYYiaE1leP3hnPcZCYVx+SLJ60Qlw6RXIZWf2R6e4jmZmVhCgtMboc8VJJFIYNKrMOmV+ENRAqEYCrkUo06OUv5RYmjRlsPDke0o8rxlIRAIoFSmZ0lWqVSEQqHj3m8kEqGuru5EqjZsNDc3n+oqnBYMVzvqDGZa++DRNxtwej66xqwmDZ+9YgoF2hB+rzunfRmMZmJyE0+trGdrfS+JBCjlUs6dW86VZ48j4Oo85Q9SvdFCe3+cJ1ceoKc/+UGjUsq4eEEFF88vwdXXQTweH7HfN1nKeeTNRrY12FKWN3a4ePSNOr5y3XSqCuI5t/nJwFRQyjPvtbJ2Z/fAskmVZkoK9Pz12R3sbeofWF5bbeaWSyYhiTgIBTJ/MFryzYQSOh5d0cD2A8lt5TIJ58ws4uPn1xD09uHxeoesl1anR6bUkUgkiAbdBIPBEzzS0UOm+9uYp0ctlxCKJXB5fCe9t1Ymk2GyFJOQKJAk4vhcvcfU5hKJBF2eEaRyErEwPs/Q17hEIsFkKeOVte2s3tpBNPbRMU+vKeDKs8fzwEu7sDuDzJpo4VMXTYCwg2DAN1BOvHOGh+bmZgoKCoZlX6edeFOr1YTD6V+uoVDohBSvQqFgwoShTblHkkAgQHNzM9XV1Wg0IiHq8TKc7RiLJXh3WzcPvLQ7bZ3NGeD/PbaN79w0l/lTyrL6nx5GKpVS1+Lm5//8IOUBG47GWbmxlU17e/j5l5ZQXa04oTqfCKFIglfXtvLM2w2py8MxXv2wme377fzkswsx6kam9ycQTnDfc7vShNth4gn423O7+ennFjFtavmoGEoNRxI88kZ9inArNGu44uxx/P7xLcSOGjqta3bw439u5O4vLGbq1Oq0Y5CSYF+zg7sf3pCybTSW4J1t3Wzc18evv7yQ2opsPo5JHN4ob65vZc32PSjlMq5cNo6FU6vI04zCbstjINP9LUnEkXhsuDe+SqirEb21gtLFV4OhiIT05PRURmLQbgvwr1f3caDNicWk4foLJ1JbnVubR2MJuh0hnnv3AK3dHiZVmrj6nBoKTcpBny1uf4zfPbaVhjZn2rrdjX20dLn5+idm89tHN7Njv53djX389POLqK2tFO+cYeLIdhwuTjvxVlxczNtvv52yLBwO43Q6T2iygkQiGTX2IBqNZtTUZSwzHO3Y3uvhwVf2DFrmvud28ufvnDekZVFHr4c/PrUtRbgdidMb4tE36vjmjXMw6FXHXecTobvDlSbcjqS918t72zq44aLJI+Ln2tnvZOPeniHLPfDSbn7ztbMxG079C8fW5ebtTW0pyy4/axyPvl6XJtwOE4snuPfJbfzhG+dgMaVeoz29Dv7w9O6s23oDER56tY7v3TSbPENmf9PuPh8/+sc6+lwf9fr888XdbNzTw3c+PRdT3ti3dTvy/g527KfzP/8L8SgAEXsbvn3rKbnpJ2irT47t4Y79Nn5y/1oOnzZft4c/PL6Vm5ZP5tpzJ6BWDf463lrfy90PrOOwlm/v9fL+tk5++/WzmVhpzrrd5vr2jMLtMB5/hDXbO1g4rZh1u7qIxRP8/vGt3Putc9EfEmzinTM8DKcAHtufWBlYsGAB3d3dtLS0DCzbsGEDAHPnzj1V1RKcpuxqtA8ZdB4IRWnqHHp4o8cRoN89+BDKlroebM5TM2waiyd4e2PrkOVefP8g9hGoYyKRYNWmoX8foNPuO2XtdDTr93Sn/C2VgFGvGrJ+/e4g3X2+tOVddh9O7+AhINv296cIs6PZVNeTcf32/Ta67Om/OZaJBTz0rXxoQLgNkIhjf+OfRL3OEa+D3enn4Vf2kOlR8eyq/XT3Dx5P3e8O8vfndnB0R3I0FudfL+/OGifp9IR4amX2j63DrN/dxcKpxQN/u31hOm1DD7sLTh1jXrzFYjFsNttA3MCsWbOYO3cu3/rWt9i5cyfr16/npz/9Kddcc41IEyIYdhpaHDmVa+kaWrzl8rCMJ8BzDAHtw0koHKWhbejj9QUiBMLRIcsdK+FIjOYc2vHIeowGGo9qM51GgcOTW/xtJvHW0ZtbG3h8ma+TUDjK2p2dWberb3XmtP+xQjzkJ9SRWcBE+juJB0depPgCURo7XJnrEI3jGOKjzeMP092XWeDtberHm+VaD0djdPR6hqxfNJYgfpQybO/1IpEMf++5YHgY8+Ktq6uLs88+m9dffx1IDm/+9a9/pby8nNtuu41vfvObnHPOOdx9992ntqKC0xKdNrf4M6166AiFXMoAyE+Rj6VcJkU7xNDOYWQjYJAuk0lRKXKPT5LJRseLR6tJvUbCkTiaHNtRq06/vjIty0S260Quk1JgzD58k284NUPyI4VEKkOqzhyyIJErkchGPoZUJpMMes5VysGva6VchjzL9azXKLKeawmgUuZ2rR2t0zQq+aiIGRVkZsyJt9/85jc8+uijA3+Xl5dTX1/PddddN7CsoKCAP//5z2zbto3169dz9913o1KdXg8kwehg6YzSnMpNr7EMWaa6xDjkpAaLSU2+4dTEIykVMi5dWj1kuZkTLJjzhv9+k8ukXLywMqeyCrkUyyAC5WRyzpzylL9DkRg6jRzZECdbKoHK4vSYtZpyc9qL9mhMehX5xszXiUwm5YqzMic7V8qlTCg3Db7zMYYsLx/j4qszrsubczEyQ/6I16E4X8vyxVUZ11UV52ExDX6tFhjVXJTl2r/u/AlZnwmmPBXnzyvPuO5ISi26lGF0iQQmVpiG3E5w6hhz4k0gGE0UF+ioKskcFH6YWRMtWM1DCwmLScMF8wcXJzctn0KpVX9MdRxOaspNWEzZxaNEAjdfUotOMzK9GTXlppx6rS5cUJFTm58Mygv1mI6aYLJ+VzfnDfFSvWhBZcaXeoFRxYVzSzJs8RE3XzyewoLs12VlcR5f+fjMlN4cnUbB3V9cQtEQE2vGGhKJlLyZ55M36wKSfVFJdLVLMC2+GulJ6HmTy2VcuqSauZMLU5YXF2j5zqfnUWgefDKAUiHjhgsnsXDqR6E/EglctKCCC+dXZp0cpDj0u0N9KFy5bDxvrf8oTvz8eeVDCkrBqeW0m20qEJxM8g1qfnjbQn72r/V0Zgj0rik3cscNs8nTpucePBqDTskNF00kFImyZntqTJJUKuHTy6cwZ5J12Op+PBSatfz8i0u5+4F19DpSA+5lUgnfuWke48oMI/r7//uZhfz0gXVZZ+VWFOn5xAWTUIySZLWFZi0/+uxC/vfvawlFYgB8uLOTL14zA7cvzKYMs2eXzCjmpuVTUGcY8tLl5XHTxyYRDMX4YHdvyjqJBD51wTgWTStCIsv+eNeqFVy0oJJZE63YnQFkUglWswarSTsis4RPNfK8fAou/izGRVcTC3iQqnXIDQXIsgynjgSlVj133jAbuyuA3RXEoFVgMWkpseRWB6tZy7dumovdGcQXCJOnVWExqYccRi8v1PO/n13Erx7ekPGeuWrZeNp7vQMTaJbOLOGWS6eiVSvw+0dH3KggHUlCDGoPya5duwCYMePkTCnPht/vp66ujtraWjFt+wQYiXa0Ofw0drh48b0DuLxhLCYN155bQ2Wx4Zi/YPtcAfpcQTbt7cbhCVFRlMeMmmTvXS4i8GRgd/pp6/GyemsbwVCMWZMszJpQiDVfg3KERVM0Gqe5281/Xtubku9No5Jz+VnVXLp03JA9GSebWDxBa7ebx97cx6a93SQSSaF17bkTmDHBwtZ9PXTafRTla7hwfiVFFh1G3eBDzy6nC7srxLpdXfS5w1QXaZkzuYgCowpd3uC9wacz4jmZTjQap6vPx8Y93aze2k4snmD6+AIumF+B0xti094eCoxqFk0rptCsJU+XfM6IthweRqIdRc+bQDAMWM1arGYtMydYiETjqBSyIfM2ZaPAqKHAqGHSILmbTjUWkxaLSUttlYH29nbKykpOWhJP+aG4rO/ftgC7M4DXH0EqlWDOU2ExaU7ZhI7BkEkljCs18t1Pz8PuDOD2hZFKkjFJVrOG+bVFBAJBWltbKLeq0WqHjhk0mowYTTC+0koiHkcqGx09jYLRh1wupaIoj/JCPR9bVEUikUCrkSM/dM0snj74MLxg9CHEm0AwjOQ6E/B0IR6P43K5KC3NbeLGcKJTK9AVj6321qjkVBRl7hVLJOJ4c7C0OhqJRIJECDdBDkgkkoFeNcHYZvR9ogoEAoFAIBAIsiLEm0AgEAgEAsEYQog3gUAgEAgEgjGEEG8CgUAgEAgEYwgh3gQCgUAgEAjGEEK8CQQCgUAgEIwhhHgTCAQCwWmDRCJBrT41/r8CwclC5HkTCAQA+IMReh0BuuxeHJ4QBQY1xQU6Cs0aNGdY/rrTkVgsjs0ZoMPmxe4MYNKrqCjKw2rWpFiJ9buD9PT52NZgIxKNM6U6n+qSvFFtnRWPJ7A7A7T2eGho6UcqlSLXBynKl2LKE0JOcPohxJtAIMDm8LN+dzePvVmHPxgdWK7XKPjMldOYO9mKxSTsccYq/mCEDXu6+ftzOwmEPjq/SrmUT19Sy4ULKjDqVbT3evnVwxto701NFqzTKPjhbQuYOq4AuXx0DdhEo3H2NvXx60c24Q185MX5+IoGJpQbueuWBTn7hwoEY4XRdRcKBIKTTjAUYVNdD/98cVeKcAPwBiL85Znt7D7YTyQaO0U1FGQjGo3Rafeyv81Bb7+fbFbVO/bbuOeJrSnCDSAcjfPwq3t4d0s7Nqefn9y/Nk24AfgCEX76wDpaezwjchwnQkuPm58+sC5FuB3mQLuLn/1rHb0O/ymomUAwcgjxJhCc4fQ4Ajz+5r5By/zntb1094kX4GgiGo2xYU83d/zuXb79x/f51h/fY3+bM62c3Rng/hd2Dbqvx96sw+4IYnMGsv9eLMFTK+vTBOCpJBCK8tSKeqKxzKIVoMPm42CH6yTWSiAYeYR4EwjOcA4bpQ+GzRmg3x08STUS5ILNGeCeJ7YSjsYBcPvC3PvkVpye1PPU5wrQ5xr83IXCMTrsXuSywV8JG3Z34RhF14HDHWTj3u4hy730fuOoEp0CwYkixJtAcIaT60stGBYvv9GENxAZEG6Hae/1EgynDm+HIqllsuH2htGoBje4jydI+81TSTgaI569020AtzdMJCKG/QWnD0K8CQRnOHlaZU7l9Box43Q0YdSpMOhSz9208QXojjpPOnVu89IsJnVazOPRKORS1MrBBd7JRKWQo8hhAkWxRYdqFNVbIDhRhHgTCM5wCoxqivIHn0laXWIg36A5STUS5ILVrOGnn19M6aGZlFPH5XPHDbPTxLjFpKGyOG/QfRl0SooLdMSG6Ma6eGEVBcbRk3rDYlJz0YKKIctdfc54VEqRXEFw+iDEm0BwhlOcr+OL18zImsNLLpPy+aumDynwBCcXiUTCpEozv/n62Tzww4v48WcXUWbVp5Uz6lV8/ROzkcuy52i744bZWE1qaqvNWcvoNQquWjY+JSfcqUYhl3H1OTXkabP3Cs+ZXEhF0eDiVSAYawjxJhCc4cjlUqZUmbn784sZX2ZMWTep0szPv7iE8eXGUZug9UzHnJdMpqwfZPh7YrmJ33ztbGqOOr+lFh0//+ISZk+0YjZo+N7N87lgfnnauZ5UaeI3Xz+bssJ0cXiqKSvM49dfO5spVanCUy6TcvlZ47jzhtmYRaJewWmG6EcWCAQY9CpmT7LyA8sCHO4goXAMjUqOOU9FUYFIcDrWkculTK7K5+dfXILdFSAQiqJSyCkwqjEbPhI2VrOWr3x8Fp+4cBK9jgCxWJwCoxqLSYNBpzqFRzA4VcUGfvK5xdhdAfpdAWKxKKVWA4X5OpSK0dNTKBAMF0K8CQQCIDkMV1Kgo0SItdMWg16FQT+4CFMr5ZQX5lFeOLaGGvN0SvJ0SopMCurq6sjPyxfCTXDaIoZNBQKBQCAQCMYQQrwJBAKBQCAQjCGEeBMIBAKBQCAYQwjxJhAIBAKBQDCGEOJNIBAIBAKBYAwhxJtAIBAIBALBGEKkCjlDiccT2J0BWns87G3qA6C2Op/KYgNWk0YkZD1GYvEE/a4Abl+YaCyOTCYlkUiwu9FOIgGzJxViNWuy+ogGw1FsjgB7m/rosHmZUpWfzJYvAblMQlG+FrlMpD04TCQaw+YMcKDNycEOF0qFjFkTLRQX6CgwZrbxisbi2J0BDna4aGh1IJdJmTHBQplVj8WkIRSJ4faGSABalXzQpLfHQzQWp88ZwO4KEghFkUjAqFVSYNIQjcUJhaMUFFUilaZ/U3t8YUKRGDKpBFOeCokkeX+GIzFsDj+xeAK5XEpJgW5gXcr2/jDhSAyFTDqQKuRwG3p8EaTSpFeq1azJuH0uhCMxXIfaT62Up/mungz6XAF8gSiW0hri8fTjsDv8hCIxlAoZSqWMcDiGVCoh36DO+bi9gTDhSBylXHpc18iR5/LIHHsCwbEgxNsZSDAUZcd+G396ehsefyRlnV6j4M4b5zB7khWNSlweueD0BOm0+Xh3Sxurt7YTDMfQqeVcuKCSpTNL2bG/l2/cs5pp4wv4xo1zKLGk5lHrcwV4emUDK9Y3M67MyFeum8nOA3aeWlmPBFAoZFyyuIpJlWYqiw2n5iAHQS4/udeJ2xti1eY2HnujjnA0PrD8ybfqKTRr+P6tC6gpNyE74gPE6w+zdmcXD7y0i2A4NrD86bcbOHtWKVcuG8+b65pZu6uLWCzOpKp8brxoEjVlRoxD5EXLBY8/zIE2J3ZXgHc3t9PQ5sCkV3LXLQvYsKebVz44SJ8rSL5BzRVnjUsKUYsejy9EY7uLx1fso9PuQ69RcPlZ41g0rZhINM7ra5tYtbmNUDiGSa/iimXjWDqjdMAOyu4M0NDq4Pl39+MNRlDKZFx5Tg2Tq8ysXN/Ma2ubiRxqQ4NOyXXnTeC8eeVZBXAmItEY7b1eXllzkPe3dRCJxqguNfDJi6cwucpM/gkKlFA4ilQqGdSWq8vuo9Pm5aX3G9nf5kQul7J0RgkXLayktECH0xumodXBqx82YXcGMOepuGRJNeNKDTzy2h7mTSlmyczSAZ/YTNidAfa3OXnu3f3YHAHMBhUfP38iU6rMWM1DW8fZnQHqWx08+3YD3X0+9FolV50zngW1RZRYTp1zRTyeIJ5IIJeJgbixhCSRSAzuRCxg165dAMyYMeOU1sPv91NXV0dtbS1a7fH5TCYSCTbt7eEXD20YtNyPPrOQRdOKj/srfDQzHO14GI8/xP5WF//v0U34g9G09Ua9krtuns97W9t5a2Nr0o7oS0sHfEKdnhB/f24Ha3d1UWrR8e2b5vHXZ7dxxdnjiUTj9LmClFh0uH1hNu7p4o4b5owKn8ZYPEFPv4+mDhcuT4BJVQUUW/ToNdk9JoeDQCjCK2uaePSNuqxllHIpv71zGTVlJgDC0RjvbGrjb//dkVZ2Ro2FJTNKeOClXWR6Ei5fXMXNl9ZiOiTgorE4PX0+9rc58fjDTKkuoDhfS94gvUzhSIzdjX2s2dHB2xtbAVApZPz6a2fxh8e30mHzppS/ctl4JpQb2VzXgz8YZdZEKwq5lH+/updQJCk8DTolX//ELP767A7cvnDK9lazhrs/vxiNSsHvH9vEzIlWSq16+lxBTHoVHn+Ytze1ctPHpnDvU1sJHSFmAWZPtPLNT83JScBFo3G2NfTyy4c3Es9gar9oWjFfvX7WcQm43n4/OxvtvLOpDY1azhVnj2NciRFTXqqY7rL7eGH1Ad5Y15y2D61Kxu/uPId7ntxKY7srbX15oZ7v3zKfA+1OXljdyPdvW0BlhvvL5vDzm/9soqHVmbauqjiPH39u8aDevzZngN8/tpm9Tf1p60x6Fb/6ytKcPsz6XAG67D56HX5KLXqKCrTHbf3lC4Rp6/WyYl0zfe4gZ80sG+i9PsxwPivPZEaiHUXXyhmG3Rngz89sG7LcX57ZTs23zs3pi/JMxuYM8tf/bs8o3ABc3jAPvrKHL14zg5WbWum0+9iwp4urltUA0GHzsnZXFwBfvHYG97+wk099bAp/f34nTk9oYD81ZUYuP3scL753gJsvrT3lXo31Lf385J/rUl78118wkevOmzCokDlR7M4gj72ZXbgBhKNxHnxpNz/6zEJ0GiV2R4B/vrgrY9nlS6q45/EtGYUbwIr1LcydXMjSmaXE4wn2HOzj7gfWE4191ON32dJqblo+JWsPXZ8rQE+/b0C4Adx+RS1PrNiXJtwuXlhJKBzl3ic/uke37OulusTA56+ePiBA3b4wD7+yl2vPm8Ajr+1N2YfNEWD11nZ8gQjnzK3gg+0dPLWyYWB9eaGem5ZP4ZU1jSxfVMXLaw6mbN9h82JzBFixvgWXN8SFCyqpKjagUqb3fHX3+/i/f2cWbgAb9nQzvaYAlzfM5CozU6ry08RXJnr6/dz9wDraez9qn417urlkcTW3XFabMiTb2O7MKNwAblpey3/f2Z9RuAG093r592t7uWB+BZ++ZAr3PrGFn35+MaYj7q9wJMZTK+szCjeAlm4PD760i2/dNC/jaEU0FueVNQczCjcApzfErx/ZxP995axBh1F7+nz84qENtHR7BpZNG1/Ad26ae8zP6WAoyqpNbTzw0u6BZdvqbVjNGn715bPSRgcEow/RT3qG0W7z4vKGhyzn9oVp6/UMWe5MJhiO0nnoRTcYBztcBEJRZk6wAPDfVfuxOwMEQ1FefO8AACqlDJlUQk25iVc+OJgi3AAaO1y09Xhp6XbjcIfSfuNkYnP4+fW/N6X12Pz3nf20j/A1s3FPd1ahdSS7GvuwOZPnpa65f2Bo8EgmV5mpb3aQRXcM8PiKfbi8IWzOAL95ZFOKcAN4fW0zrT3Zj7vXEWDFhpaUZePLTGzZ15uyTCqBWROtvLWhlaNp7nLT7w5SfoQxfFefD3OeCoU8/TFeZtWzr6UfpyfI7oN9Kevae71s2N2FTqtk8lFm7gC3XFrLzx9cz5Nv1fP62ma+++f3OdiZWfxs3ddLNDZ4A76wuhGDTsmvHt7IG+ua0trvaBKJBGt3dqYIt8O8ub6Znj7fwN9ddh//fWd/1n1NrjKzZnvHoL+3tb4Xi0nDivUtVJUY0u5nmzPAO5vbBt3H+j3d2J2ZnwM2R4DXP2wadPv2Xi89Dn/W9ZFojOdXH0gRbgB7DvaxZnsHxzqAZnMGePDl3enLHQHeXNdMbIhzJDj1CPF2hnGgzZlz2WxfmoIkgVCU/Tm2UXuvl4nlyRelwxPCF4zgC0bY1+wAYEKZiaZON5MqTexu7Mu4j7U7O5kzqRCHJzgs9T9e+t1BnN7MAvKDHZ0j9rvhSIytDb1DFzyE45AA3pZlmwnlJvY0ZW7rI2nt9uANROhzBfAGIhnLDPZyD4aiaT0/Tm8oTYQWF+ho7nJn3c/63V3MmVSYWrceD1ZT+vBmLJ5g1kQra3d2ZdzXhzu7mD+lEIc7lNJblKdV4PSGUmJhEwl47p39RKKpYj0UjrJuV+b9H0m/O4hWnfyN5945QL978OvX7Qvz1lFi90g21/UM/H8wHKWxI7OwPLyvocRlIpGsY31rP1PHFdB01DlwuIM57cPuyizenN7QwHD3YAz2bHYeivPMxJvrWrLej9lo6/Fk/WhZtbn1mPcnOPkI8XaGcSwRbNJjKn1mkmtIoASIc8TTMnHECiCeiA/sK9s+pRIJ8fixncMRYZAKyEYy6FkCkmM4eslR/z2axCDrMu1rsPhP2WCzszOsynQcCQa/nqRSCfGjFJ8EyCYr4nGyzhqXSSUkEsnfO7LXJpHIvE2283qsMbFSaQ5nUJK93sm6DP8dcGStjnfv2bbLdX/SIdoy2/pkmx77eTjeeghGB0K8nWFMrsrPvWx1+pCK4CO0agWTMgw7ZaKiSM++5mTMi8WkRq9RoNcomFmTHEptbHcyvszE3qZ+Zk+yZtzHstllbN/fe8rTCxQYNFhMmetw9qzSEftdpVzGwmlFOZc/3E7za4szrm9ocTBzomXI/UwoN6LXKikwqjHqM8fzXTC/Iuv2WpUibXjSpFdy9Puzu89HVUn2oPWlM0rZsq8nZVlVsQFbhuE2iQS277exbHZZxn0tm1PGhj3dGPNUKbNvvYEIOrUC8xFxaVKphOvPn5A221OllLNs9tDn22rWDPTk3bR88pCTF4w6FZctrc66fu6Uj64BrVrO5Mrs92CeTplxWPlIpBLIN6qYPt7Cjv02xpcZU9bnG9Qoh9qHVIIlQw8ogDlPldPM/YkVpqzrzHlqLllclXHdFWePyymO8EgqCvVZPzg+trgqJeZPMDoR4u0Mo8Siy2nmlzlPRalVBK0Ohkoho7RAN2Rw7+QqM0qlbCBg+caLJlNg0qBSyrnynPEAhKMJQuEonb1eli+qTpu5Nn18AQUmNZMrzeQf44N6uLGYNPzw9oUpM0slEvjMFVNH/JqZO7kwpxyEcycXDrxMJ1WaUSnSg+0PtDupKTchH6In56blUzDolFhNGn70mYVpL+IbL5pERWH2GcAWs5rli1JfvHUt/SyekSp8EolkTN8159ak7WNypRmtWk5330dCrbIojx6HP+OQXnefj7mTC1HIpSyYmip4J1aYmDnBQiweZ+/B9GHjR17by48+s5CvfHwmNy2fwp++dS7VpZlF5cyJ1iGFzScumIhUAr/68lIumF+ZU+/swqnFGcXM9RdMpPiIe6MoX8cnLpyYdT97Gu1cuKBy0N9aNL2E3n4/580rp9fhTxuGtpo1XLKketB9nDO7LKt4s5g0XH3oPs/G+DIDVnP22b1ymZQrlo1n6rjUj+/F04tZMqNk0H1nwmrScscNs9OWlxfquXhB5eA9yYJRgUgVkgOnU6oQgO0NNn76z7VZYx6kErj7C0uYM7kwc4ExznBO2/b6wzR1uvh/j27OOBHEatLw7Zvm8vraJtZs72RcqYEff3bRwOwwty/Ew6/u5e2NrRQY1fzoMwv5yzPbuWhBJRq1HJc3TIFBTbvNS31LH1++bhblgwiFk0UikaC330+X3YfbF6CqxERhvhaNamRThQTDUVZubOWfL2SePQqgUcn53Z3LqDqUeiEajfPBzk7+8PiWtLKTKs18bFElf39uJ7EMN8THz5/AdedPHJjdGI8n6On302Hz4A9GB5Ja6wZJkRKNxtjd1MfmvT289H5yZqdcJuE3Xzub+57bycGjYrYuXFDB/Noi9hzsw+MLM218Af5QlMfeqBsQagVGNV++diZ/fGorvqNmOpcX6vnRZxaiVsn569PbKbXqmVxtJhCMolbJ6LL7WLerm+svmMAfn9qWNplj6cwSvnztzJx6eGPxBHsa7dz9r/UZJ4VcML+Cz1wx9bh6cuzOAAfanby7pQ2dWsHFi6ooL9SnJbru6fexYn0Lz65Kn7ig1cj53deX8ffndqZN3ACoKTfyrU/OZXejndfXNfOj2xdSak3PuWZ3BvjT09vY3mBLWzelysxdt8wfdMZnnyvAX5/dzua69PhLq1nDL760NJmUewicniA2RwCXL0T+oR5wg+74PuYCoShddh9rtrdjdwZZMqOECRUmCo84DpEqZHgYiXYU4i0HTjfxFo7EqGvu509PbRuYkXcYi0nNN26cw9TqApQZUgOcDgz3jeT2h+js9bFhTzerNrXi8oYoMGn42KIqZk6w8MH2Dl79sIkl00v43FXTKTyqV83hDvLah008v/oAVpOGb35yDi3dHt7e1EokEkOvVXL5WeOoKMobFTnejsTv97Nv3z6mTJly0h7uXn+Ydbu7ePDlPfiOmkBQVZzHdz49j+oSQ0o8li8QYXuDjfueS8+LdtbMEj5+wUTWbOvg/e0dxGIJaqvNXHPeBCoK84Yl9Yk/GOFAmxOHJ8QHOzppaHWgU8v5xifnYHMEBpL0mg0qrjx7PBPKTZRa9Xj8YVq63Idm8nrRaRRctWw808cXEIrE+GBHJyvWt+D2hSnK13L1OeOZMcEyIPD73UFau928+F4jPX1+DHol15w7gYoiPR/s6OTF1QcGxJ/VrOFTF09mfm3RMQ3Nx2JxOu0+3tncxjub2whHYtSUG/nEBZOoKjEc85De8dDb76PXEeD1tU00tDpRyKWcPauUJTNKKLPq6HeHaO328NraJvqcQUx5Ki5ZUkWBQcN/3qjjgvkVzJpgoagge89xvztIW4+Hl947QI8jQIFRzTXn1FBVYsgpJ57DHaS5y80Lqw8kEy5rFVx9Tg1Tx+VTlD86RzmEeBsehHg7RZxu4u0whxM+NnW6SCRgfJmREkt2e6HThZF6IDncAZyeMPFEAolEQiKR4MChbO8TK81YjGq06sw9NJFoDJsjQHOXm55+P2VWPaUWLQlAqZBhNWlHpWXZqXq4H7Z3a+vx0NbrQSmXMbHSjNWsGUioezSJRAKbM0CHzUtrlweZTMKEChNF+clEp7FYHJc3TIIEOrUC9TA7jBz+facnhNcfQSIBs0GVtLtCgj8QIRzyYs03pLVlIBQlGI4il0pTxGQikaC7z088HkehkKX0mhxJMBQlGImhlEsHrsHYoTb0BZJ1MeiUJ3TvJ9svRDwBGrUcXZZrfSRx+UJ4fCHisThWkxqNRn3U+iDBUPxQah4IheMo5NJjctEIhaMDFltq5bFfIwPnUibNapc3WhDibXgQSXoFw0qBUUOBUcP0mqGDtgVDYzZoMBtSX3415aactlXIZZRa9RmHbATpSKUSCvO1FOZrmVeb2yQGiURCoVlLoVmblnIDkjMq840jF6h95O9nQiGNUdfagTU/Pb5Mo5JnDHqXSCQ5JVRVq+RpYlQmlQzqCnCsJNvv1H74GXUqFJIYdXV1WIy1GdarMR7RXHnHcfgqpRzVcYi2w2Q7lwLBsSAmLAgEAoFAIBCMIYR4EwgEAoFAIBhDCPEmEAgEAoFAMIYQ4k0gEAjGMLFYnFA4mtUcfihC4Sjh6ND2TaeKUDiaMQ3J6U48nsz9OJQXrODYSSTGftuKqEmBgGQqiX53kEg0jlopSybRzZDYdTDikRBRTz9EI0gUSmR5BUjluc+4c/vC2J0B/MEICrmUAqMma+LPkSIcidHnChKJJmfTFRjVaZn1D+P0hHB6IxgLx+H0xTg8icoXiOD0hIjG42hUcixGzaicKTsY4XCUHkeASDSOUiGlKF+btR0G4/AM0z5nkHA0hk6twGLS5Jw+w+EJEA7H8QQiBIJRlIrkzMiifC12Z4CWbg9vrmvGG4hgNWu4dEk1JQW6IVN9+AJh7M4ga3Z0sOdgHzJpMrXGjAkWrGYNyuM41kxEY/EBb9DkNa3OyVLLG4hgdwZ4f1s7dc39yKRSls0uY0ZNAVazJuO5iMXidNi8+AIRdAXVuPxxRtsEyWA4Sk+/n0gkjlwuwaxXYzzqWnB6gvT2B1ixoZkOmw+NSsYli6upLjWcUEqRWCyOzRnA4Q4ST4BRp8Ri1hzXjNmxitsXwuYI8PamVpo73SgUUi5eWMXEChMWTZxEyAeAVKVFps3udjIaOHPOmkCQgUg0RnOnmyffqkellGE2qOmyezHqVXz8/Ik55VVLxGOE7e041z6Pb+9aSMRBJidvxnkYF12F0pLZougw0WiMxg4X97+wi/1HmFOb81TcetlUFk4rOu5EnLmSSCRo7/Xy/Lv7Wb21nWgsgVIu5aKFlVy1bDxlRyQGPpx77NlVDWxrsJFIQEmBjo9fMIFJlWYefnU32+rtAOg1Cq45t4bz51Wk5bcbjQRCYbr7kvnC3t3cTigSoyhfw62XTaWq2IDZoMKgUxGOxIjG4llTvwC4vCE+3NnJ42/uS8ktN2uihc9cMY0Si37AsD0UieEPx9HnJa2Z7M4Ard1udBoFz68+wIbd3QNJhCeUm7hp+WQ8/jD3Prkt5TdXb2mnptzID25dQPEROcvCkRiBUBSVUkYoHGPLvh7e29pOIgFzJhUSicV57cMm7n9hJ3feMIfFM4oJhmLEEwmMeiVy2bGLufZeDyvWt/DmumaC4RgmvYqPXzCBJTNKB53l6nAHeWplPa+vbU5ZvmO/DblMyrc+NYcFU4uJJxLIpRJUSjntvR7e2dxGp83H7ElWwpEYG/Z08bFF1UyvKUCllKGUy1Ae4wfZYATDUQKhKBpl+kzeTLR0u3l/WwdvrG3C448gl0k5e3YpVy+roaxQh0aloLffzz1PbGHPITeWw2yu68WgU/LTzy2isthAOBpDo5Ln/EFhdwZ4Y20Tr3zQRCCUzOsnlUo4a2YJN19Se0bMcrc7A9z/wk7W7+5OWV6dL6cq3kr3jpcJdzYAoCqdiPmcG1GVTkCmGV25NQ8j8rzlwOma5+1M5XA7TplSS327l9c+PMg5c8p5f1sHPX0+xpeZWDy9mBdW7+dr18+mutQ46P6CHQ10PfZTEtF0hwWpSkvJrb9EVZjZlxBgd6Od//3H2owZ/gE+efEkrjl3wqBZ/I+FqNdB1GnD7XSiUsrRWEpo9an54d/XDjzYj0SvUfCbr51NVYkBXyDC+t1d/OnpbWR6csyZZGXmRCuPvLY3ZXl5oZ6ffWHJKRdwLm+I9l4v63Z3olEqWDKjhFKLDn8wyv42B0a9irv/tX4g+e9FCyuZUmVmxfoWevr9VJUYuPGiSXTYPKze0sH588qZN6WQwqN6RLyBCM++3cDzqw8MLFPIpXzuqukEQlG27OtBrZRxxdnjKTCqeXbVfg52uKgsyuPa8yawp7GPqePz+dmDG9ISER/mKx+ficsb5okV+9LWVRbp+ekXlpCfp6at18OL7zXi8ga5atkEttb30tXn5cqzx7N+dzdrd3aiVMg4Z07ZgKjyBiK8sbaZUCTG2TNLueLs8ZQV5v6Cb+p08T9/+yDN/QGgzKrjZ19YkjEhrj8Y4dE36nj1g6ZB9//9W+fz3pZ2ovE4n7liGv94fie3XFrL7oN9rN3ZhUYl57Kl1RQX6Nixv5fDo2OzJ1mpLM47od6mcDhGS4+bZ95uoKnTTUVRHjdeNInqEkNWEdfS5eaXD29IsTc7jEop46efW0x1SR53P7CehiM+4I5kzmQrH1tYxQc7Omnr9TCpwsRV59RQUZiHfBCLsj5XgHuf2MqOA/aM6/MNan7ztbMosaSf39PlneP0hPjT09vYXJfqDXzZ/CIuz28kuvn5jNsZl16HafHVyDQnJm5Hoh1FzJvgjMXlj/HEm3XMmVTI7x/bwsY93bR0e3h3Sxt/eHwLV587gX++tAuPP12UHSbq6af3hXszCjeAeMiP7ZW/EfN7Mq7vcwa454mtWYUbwFMrG+jNYD5+PES9Dnoa9vJufYBfrwzwjw+DdHkl/P7xrRmFGyRf5H9+ZhtuX4iefh9/eWZ7RuEGsK3BhlRCmj9ke6+XVz88SOwUxpi4fUmh84O/fcBL7x3kqZX1fOOe1exvc/DAS7uIJRL87b87BsRSMvO9lr8+u4P9bU7cvjC7DiSFtl6jJJGA+57byU8fWE9Pf+r56enzpQg3gNsvn8rKjS088tpedjf2sbmul7sfWM8raw4il0lp7/WydlcX3//rGuZMKeShV9MdJI7kgRd3M2eSlUyjkK09Xpo63exvd/Kte99j7c5OLl5Yzf89spGX3m/k3Dnl/OHxrbz2YRMOT4iefj/PrtrPyg0tSCQS/v3qXnr6/Tg9IV79sIm7/rqGDps3p3Z2ekL89tHNGYUbQIfNxzPv7CccSY+zszsDQwo3SPqvJhNfa/hgRyefvrSWe5/axn9er+NAu5NdjXZ++9hm+txBAqEYH+7spK3HQ6/Dz879Nk6kz2J3Ux/f/dP7rN/dTU+/n811PXzvL2vY1pBufQXJoe9nVzVkFG4AoXCMe57cgsMTyircSi06ls4o5bePbebDnZ20dnt4e1Mb3/7jeym99ZnY1+LIKtwg6Rzx33cPZDwfpws9/b404SaTSlg+VZ1VuAG41j5PpK9jpKt3XAjxJjgjkclktPV4OG9eBU++VZ+2PhyN88qag1QWGbA5Ahn2kCTi7CXqyvzQHthXdyNRd+aHZ4/Dn2ZRlol3t7Qfd0D6kUT6u9ls13Hf680caHfy4R4bzfbokC/mhlYndmeQbQ22QYUmwIr1LVy0MN0M/I21zTkd60jR3edLG4qTSSX0uZOWVXkaJc1d7oF1Fy6o5Ll30v0yAR56ZQ83XjwJSArTbfUfXQOxWJy3N7WllDfnqUgAje2pPqYAb21oZe6Uj5IGS6VS/MEI+5odgx5PNBZnX0s/584pz7i+0+bhH8/vIBZPcN7ccl798CChcIzJVWb2NTtwekNp29S3OkkkEmlJZN2+MG9vbMlJfPc6/LT3Dn49vbOpDXuGa+HoIa1sdPf5sZo1XLyoiqZOFw0tDrrsvpQyi6aVsHZnJ0+/3cDBDhert7bzn9fr6O7L7Z7LRL8ryH3/3ZHRF/q+53Ziy/CR1e9KDp8Pht2Z9Cwtys8c43rJkmqefKs+7aMpGkvwjxd24M5wLgHc3hBPZuiZPZps5+N0IBKN8fra9A+CxVOtKBvfG3J718ZXiUcyt++pZEyIt3g8zp///GeWLVvGrFmz+OxnP0tLS0vW8i+88AKTJ09O+zfYNoIzC6VSyf42F0qFLM3r8jB7m/oZX2bEleXBCBDpH/yhfJiYz5lxeWeOvRl1zX0Ew5l7Mo4Fh9PDa1s+EpJalZx+dzC3bT1B6pr7hyzXafdlnGgRDMfwDtKTNNLsbkwX0FazhtZuN2qljD5X+ssrGM7cG9HnCiKXfdTl9e6WtoHzEwzHqG9JbaeJlWZ2Ddb74QoODIsb9UpcvtxeFvvbnEysMGVcp9coOdjhHvj93Y1JY/YpVfns2J9usH6Y+hYHVSXpcT4f7uzC4x/6/B3dC5mJaCye1qMdjyfY1zL09XUYbyBCPB5n5gQL63d3pa2fN6WQ97e1pyzrsvvQqOU4Pcf3Mk72Pmc+PqcnhMub/izx+MNEY0N/eLX0uFk8rTjjOp1GkfU+Pdjhxp1ldCAYjtHak7nX/0iisfgpvTdHkmAoNnAfHMmEQiXx7swfZynbdzQQDw3PyMdwMibE23333cdTTz3FL3/5S55++mkkEglf+MIXCIczX7D19fUsXLiQDz74IOVfeXnmL1TBmUcikUCtlCEbZBakSikjGosjk2UvI5Xn6E0ozRxYnGsAtUI2eF1zRSEFc95HdY7GEjnXQS6TohgktuYwg1VzOI7heMnkXxkKx9CqFURi8bR2GKquctlHbWHQKQfKS6WStHYKhqODxiyqlTKih9JhxGIJZNLcHs1KuSx7Gg0JAwIzEo0P9KaFIjG0muwxX1q1gmAoXbTq1IpB74WBOilyq/vR7SuVSo5phrdcKkEikeAPxTL6qHoDkYwTfWRSKcocruOMvymXZhymPkym+yPXmdYqhQxvIPMH2mDXolwmQZ7lepFIUq/TwRhrM8JzJdP9CBCOJZAqhn5+S+UKkIw+qTT6anQU4XCYhx56iDvuuINzzz2XKVOmcO+999LT08PKlSszbtPQ0MCUKVOwWq0p/2THMWNKcHoSDAaZMcFCS7c7a8/FeXPL2bC7C8sgfo3KwipgiIeeTI7caM24qqokt+noFy+qPCE/xcMYCou5ZVnRwAM9FIlRaFIPKrgg+VKymjScO3voD6B5U4oy9jKVWHQ5p8gYCSZXmdMe4g5PiJICHSTAbFCnvCTtzgClWXxDp48voKX7o6/5K5eNH5j5p1HJ04aN9x7sY87kzNeAWilDLpcSOhRz5PSGyDeocxIyC6cV8+6WtozrvP4w582rAOCDHcnJFQDrd3exbFb2GdCTKk0pw8eHufa8mpyM1Mus+iGFQIFRjSkvPZ3JstmDz8w+jEQCWo2CYDjKrv02li9OnxC0alMrn/zYpBSxdfHCSrpsnuP2sM03qJmfxUt36vgCCjLs16RXkT9E6haAiRVm3slyLhtaHUyvKci47ry55VmPx6BTctbMkiF/22JSY8rwcXM6oFXLOWdO+nW1tt5NfMKyIbfXz7pwVKYNGfWpQvbt24fP52Px4sUDywwGA1OnTmXTpk1cfvnladvU19ezfPnyYa1HIpHA7z+1XaeBQCDlv4Lj43D7GbVSGjtcXH/BRB56ZU/KcMiUajO11fmoFDIMWmnWcy9V56GbshjfvnVZfy9v7nLiqryM+zBqZSyoLWRTXfa4Oa1azqQK47BcfxKtiSqDi9/fMo4ud4I8jZQilYfz55WzanN71u0uW1qNQSulvFBPUb4269CRVAIXzK/gnie3pq277bJa1PJTdx+ZdTLu/vwifv/4VhyHhs0qi/OoKtbzP7ctYNcBGxctrGTF+mR4xasfNPHl62byt/9uTxkutJo0fObKafzk/rVIJXDT8smUW7Upx1VbbUatlA0Mu0ZjCTbu6eHTy6fw1Mr6gbhBnVrO1z4xm6eOiLsst+qRySRcvKhy0OD94gItRr2Slu7Mw2K11fksnl5MU6eL3Y19nD+vgpkTLOw8YMcXjLB0Zglrd3403CiRJCdVJA79/5HxVefMKaO2ypTTuTNoZVy8oJIVG7KHqdx6WS06FWn7qy7JQ6uW488y2eEwcyYVsq2+l7rmfj57xTSc3hDXXzCB5949MFDvPleQkoLkubU5AuTplHTavJwzpwy5JHbc1+HnrpxGR6+XziNi7KxmDV/7+AykRPH7U+teaFJyw0WT+MfzO7Puc8HUIvK0iqwTgVZtbuNbn5xLnyuYEts3ocLEDRdNJBoJEc0y6nnlsvG8t60j674BPnXx5Izn43R558ydbOVhqSQlbril24PHPBmtSksiy7CoRKlGM3HBCR//ke04XLNNR32qkLfeeos77riDHTt2oFZ/9HXxjW98g2AwyP33359Svr+/nyVLlnDllVeyb98+nE4ns2bN4rvf/S7jxo07rjrs2rUr6xCtYOwikUgwWsq5/6V9nDWzDL1WgdsXxmxQ0d7jpaXbxScvqMJpHzyubVxRPoF3HiLUtjdtnWbiAhSLP0GLzZm1DiZLOfc8s5uWrvSXsEYl539vn4ckZBu2a1CtUlGUp0YWC4JMgScqIaow8O83DrKtIT0WaumMYj5xXgWuvi5UKhUaUxm/+c/mtNlzcpmEL14zg7W7uth+1H4+dfFE5tVo8Lpzj2kaCVQqFSpdPv5wcihKJYvic/chV6hQ6vJRqtQ89sY+NuxJBs5bTRo++bHJhMIxnJ4QEytMlFi1ON1+/IEo+QYlsZCbgC+1p0qpUhFTFPDLh7cM9KhBMlXFBfMriMUTFBhU5KkTyKTg9MXpdQSwmjSY9HJaevyUWg08ubKeLRmEvcWk5ge3LuCfL+zKOEPx4+fXsGiSloDXSZ65EG9ISm9/gHFlJmIJaOv2UmzREo0l2H3AjlIho9Sq490t7WhUMq49dwL1LXYCoSiTK02opGG8ruwxe0djspTy1KoW1u9JneEnkcCNF05g/kQdXldf2nZKpQq/xMT/PbI16wQdc56Ku7+wmB67B4VCSlG+jvV7eonG4syaaMXmCKBSysjTKmjr9VBZZMDpDqBVSdGpwOfqJRY7/pmVUqkUvakIlz9Bd3+AQrMas0466H4LyyeyYn0LL71/MG3d1HH5fP0Ts/E6e9jdHuU/r6dPoIKk0P/CNdMpMKqxO3xYTRq0yhgeR++gs2c1ujzaHHL+9tyujALuogXlXLawELdj8IlXYxm1RkenW8mfn92VsrykQMuPrioi8dYfiAdS44+lKi3m6+6izS8hFBq+CQvz5s0blv2MevH20ksvcdddd1FXV4f0iHH9u+66i97eXv7973+nlF+/fj233XYb11xzDbfeeit+v5/77ruPffv28corr2CxWI65Drt27SKRSDBhwoQTPZwTIhAI0NzcTHV1NRrNyc28fzpxdDt6AnHaer289sFBguE4+UY1ly2tpsisRqvKMV4k7CPu6sWz9S1inj7k5iL0sy9GarASVwx9rtyBGPtbXTyzaj89/X50GjmXLqlm8fRiLIbhye82FP5QHJszxIr1LdicAUosWi5aUInVpEajTB0G6/fG6LB5WbOtg1AkzvSafGZOsKBVKWjtcfPm+haCoRiTKk0sm12GWa9ApRgbMTWeQJw+d4jVW9qxuwKUFOg4b145BQYl6mM8BpsrwvvbO3hzXQu+QIRCs5aPXzCBGTUFmHSpw6LBYJCmpiaqq6tRqjQ4vRFC0QQOd5C31rfQ1edDp1Fw4YJKJlWaqW/p46FX9qb0UuUb1Nxy2RTmTrSgVmauazwB7bYAD7y0mwPtLorytURjcfqcARbPKOHmS6akXHPH+4oIhOP0u8N8uLMLm8PPuFIj82uLMOvlKOTZ2zEWhzabnwdeTNbvMFKphKUzSvj08snk56UOGvnDCfrcYVZvaeNguwuJBJbNKWPa+AKKTMoTSg0yGBKJJOd9u/wxPP4oHx7K02bUqZLXlVFNvj55LQQjcfY2OXn41b30HjHLXaWUcdmSai4/q5o8zbFHO0VjJO/tDa2s29VFLB5nSnU+155bQ6lFg0aZeZ+n0zsnHEvQ2O7hXy/vSZkNXWrV8f1rq7HEbATq1wMJtJMXoSgaR1xjGpbfPrIdCwoyD38fK6NevK1YsYI777wzY89bOBzm73//e9o2LpcLo/GjxKp+v5/zzz+fz33uc3zxi1885jqIJL2nF9naMRyJEY7GUClkx2WFBJCIxUjEokjkciRZJikMhtsXIhyJI5VIMOWpTkkQcTyeIBqLo5BLh7QyCofD2Ox2CvLzU+7PWCxOLJ77ZIjRSjyeOOFzEIvFcXhCxBNJiyhzhlgvGPz+dniCBENR5HIp+QYNMqlkwO7I7gwMTEiwmHK3VHP7QtidQdy+EBKJhHyDGqtJk5NbwMngsF2cyxdCKpFQYFRjMQ1u5xSNxXF7gwSDAfKNupRrcrRw2FdTIZchyzKZoM8VOGSVF0Uuk2IxabCaNTlPPshGJBrD7QuTSCRnsB6dEuZoTsd3jsMTxO4M4AtEkEolWIyarJZrw8VItOPouEsHoaQkGWzZ29tLZeVHQcC9vb1MmTIl4zZHCjcArVZLeXk5PT09GcsLBJCc+XmiYkMikyE5gYkxI22DlQtSqQRljsIzGo3S3dWF2WRKWS6TSRnu+UGxQ+kM4ofykJ0MT8bhEM+yQy/f4yEYihIIR5FKJBTl61LqI5NJKS7QoVMriMbjKOWyY3LhMOhUo+J6y4ZBp8Sgy3E29yHkMilqBTQdaMaUVztoWW8gTCQaRy6T5jQRY7iQSCSoVYOfpwKjhoJBJkodLwq5bET2O5Yw56lRK+UEw1GkUgl5GuWYnGk76sXblClT0Ov1bNiwYUC8ud1u9u7dy80335xW/oknnuBPf/oT77333sBXl9frpbm5meuvv/6k1l1w5pCIx4i6+4i6eokHfUg1eciNVuQGS05G3EfjD0bodweJROOolXIspuwG8aMNlzeEwxMiGo2j0yR7g06k7oFQlJ5+H+9t7WD97i6isTjVJQauWlaDMU+FxxdCKpVi0ispzNdlTasQjydN4g/n7Usac2tPavqSw0b1Xn8EqUSCMU+ZsSeu3x2krcfDC6sP0N7rRamQcfasUhZOLcacp8JsUNPd52PXATuvrW3CH4ySb1Bx3XkTGV9mPOU2ZKOZnn4f+9ucvLi6Eac3hF6j4Ipl45k2Lp/iAt1x3a+CsUGfK0CfM4jNGSAUiSGXSdCo5JRa9FjNmjE1UjDqxZtSqeTmm2/m97//Pfn5+ZSVlfG73/2O4uJiLr74YmKxGP39/eTl5aFWqzn//PP54x//yF133cUdd9xBMBjknnvuIT8/n2uvvfZUH45glJJIJOhzHRZLSYP6XIn53Xj3fkD/6idTZi1JtQYKLrod7YR5OXvjRWNxWrs9PPpGHVv29ZBIJNN0XDi/gmvPm3BMBtJef5hgOIZaKUN/EnoW/MEo9a39PPTynoFUE4fN7a85b0IyJccx4g1EWLWplX+9tDtleXefn/W7u5k2voALF1Tw12e2U2rV88mLJ1M7Lp9Cc6p46XcFWbW5leffPTCQjFSrlnP1svEsX1xNQYaesX53gHAkjkopSxFYiUSCPneQWDSOMU+JWplbb5fdFeD9re089+6BgcTQZVY9t15Wy7TxBSTiCfR5Blz+GL97bFOa7dGTb9Xz9Mp6Pn/1DGrKjfzkn+sIHZFEuKffz6/+vZECo5qff3EplcWj01D7VNLU6eIn969LcZfoAf701Da0ajk/++ISJlWYR7wnJhqLIzuUp06QJBaLE47GUSlkI9L+vf0+Dna62bS3h/e2tQ/cO+NKDVx9Tg29TjW1VfmjJmxgKEZ9zBtALBbjnnvu4fnnnycYDLJgwQJ+8pOfUF5eTnt7OxdeeCG//vWvue666wCoq6vj97//PTt27CCRSHDWWWfxP//zPwNDsMeKiHk7vTi6HfvdQT7Y3sHTbzfg9oUpLtDymSumMXOCJRl/JpNkzYEUCwVwb3wVx/tPZf29gku/RN7M85PJHodgz8E+/vcfH2bMyG7Sq/jN186irHDwl7I/GGF/m5PH3qijtcdDVXEet1w6lQkVRjRDDNccK4fbctLkWjbstfOnp7dlLGc1a/jVV846ZgG3ZnsHv31086BlZk+yUl1i4MX3GgG4afkUli+qJP/Q8JDDHeSv/93Oxj2ZwyZmTLDwnZvmDgwnubwhNu/r4bE36rA7g5QU6Lj9iqnMnGDBG0i27Ztrm/EGIyyZUcLSGSVUFOWRiEaIBz1IFGpk6tTj7HMlPWx3ZnFZ+Pj5E0ACE8tNPL+6kYbWwa2xvvWpOTz8yt6MFleQnLjw/75+NsXHIZjHOtmek512L9/785qsjiqQzLl3zzfPpaJoZIRvd5+PTXu72VTXQ3WJgQsXVFJm1Z9wLNtIcTLeOdFYnI5eL2+ua6axwzXwQVZm1Q+buO13B2hocfLQK3vo6vNlLHPb5VOZOaGASZX5w/KbR3JGxrxB0ofye9/7Ht/73vfS1pWXl1Nfnzq1ura2lgcffPBkVU8whgmGojz7dgOvfvhRPq3uPj+/fmQTX7t+Fh/s6KDPFeSzV05jRo0l7ass5unD8f7Tg/5G/1sPoa2egTR/8I8HuzPA7x7bnNVKx+kN8eib+/jmJ+cMGu+1rb6X3/znI8FT1+zgh3//kB99ZiGLpx/fB8xQOH1R/vbf7VnX2xwBXvvgILdfMS3nF1WfK8DDr+wZstz2BhsXLahEKknOpHzyrX3Mn1I4IN6aOl1ZhRvArgN26lscLJ2pIRKNsWJ9C4++UTewvqvPx68f2cQf7lzGh7u6eP7djwznD3a4eHdzGz+8fQGaLY/hb9iEsrCKgo99FlXhR4lj9xzsyyrcAJ579wDfu3kewXBsSOEG8MLqRi5eVMmzqzLb+/S7g+w52HdGirdsbKnrGVS4QdJO6q0NLdx2+dRhF1Rddh8//PsH2J1Jm6tt9TZeWdPEb752FpOrhl8wjBWaOlzc9dcPiB7yza1r7uf1tU387o5lVBYPT3LcfleI7fttWYUbwONv1lF683yKC8LHHGt5Khidcl8gOEnYnIGMpsUAj6/Yx4LaYtp7vfz8wQ0Zzdv9jduAwTuvE7EIoa7GIevS6/DT5xrcZ3Tdzk7sjuwJI+3OAP98cXfGdf98YVdG/84TRalUUtfUP6R/45vrWrANUvej6XUEcjYQ33nATu245BT8RALW7uoiHInhC0R4Nou5/JE8tbIety+MzRng6ZXpebakUgmxOAO9e0fSafexp6mPWDRKPOAh2LKb7id/TsSVzHXn9IR4amXDkHXYe7A/4zWWieYuN+VD9MA+u2o/Tk9uvrWnO/2uIP9958DQBYE31jUP+30Sjyd4b2vbgHA7TDQW5x/P7xxSVJ6ueP1hHnh594BwO4w/GOXZd/YTGgY/53AkRq/DzzubM7tXHCYaS9De6x0z94wQb4IzmmQKh8zrnJ4QGvVHPVx7mtKTioY6c3shhGytQ5bpzcHQO54AbzC7gbQ3EM5qYG1zBkbEfFqpVKbFZ2UiFIkRCOX+MPYcwwut1+FPsSBq6XITjsYIhqO05WDM3dHrJRSO4vaGCWfwCjXqlDi9wayJY7fusyGrmjPwd8zrJOZzAhCOxujoHboOLT3uoYzWUjj6hXc0Pf3+7L6nZxiRWCzrfXE0oXAsJZZwOPCHImzI0vt7oN2F7zQ1hR8KXzBCXVPmxN3b6nuH5XkVDCdTQOXy7OmweYe8r0YLQrwJzmjUyuyzi6RSSYovYqZ4LZk2t9iYo2OgMpFrmgfFIMM5SkV2A3u5TIpyBGasxuNxjPrchhmOZSgqk5l0NtRKGeEjnAy0ajlyqRSZVDJkLqtkeUXSGD3L9RAIRQcdqi40a0j4jngJSaRIlcnYFqkktzro1AoSQ/TiHgtqpUwExB9CIpEc07Uny2L0frwo5TIKzZlTdOg1CuSyM/M8yWXSrM+OfIP6mJ4B2X9DMugz80g0KvkxfUCdSo67ZYLBIE888QTf+c53+PznP8/3v/99XnzxRaLRE+/mFAhOFhaTJqv5+JLpxWytT1rGzJpoYXyZMa2MburZOf2OpnrmkGVKrfohH+KlFt2gptoFRg0XzK/IuO7iRZUZjbNPFL/fz8KpxUOWqx2Xj9mQe16xwnxtzi/c2ZMK2XtEz+i5c8tRq+QY9SouWzq0Ld7yxVWY8pJJYKdUmdPWB8MxtBo5VRlmcEqlEpbNLiO4+dWBZQUX3YbcmBzGNeep+FgG4/SjWTitGKM+t/OjU8uz9gIe5ty55ZjyRn/szsnAqFNy9qzc4j2nji/I+WMkV5QKGdeel9mh58aLJp2xudfyDWo+efHkjOtuWj5lWPIQatUKtGoFM2qGdleqrc5Hqxkb98xxibf29nYuu+wyfv7zn7Nz5068Xi9btmzhBz/4Addffz0ul2vonQgEo4B8g5offmYhFlPqS3NypZlbLq3lqmU13PPNc/jup+dlfMAqzEUoCqsH/Q119QzkxqEfHFaThuuyPOAP8/mrp2fN0A+gUsj41McmM7+2MGX5wmnFfOLCSSOWx6jAqGTWRGvW9YdNz48lGarVlF2IHkmeVoFEwoB5/LhSw4Agl0gkLJpWPGivplop47y55cikEvK0Sr71qbmUF6amZJlYYaIgT8V3Pj2PqeM+Ci4356m46+Z5lFi0lNz6C0pu/jllX7wX/awLkSqS50kmk3LxwqpBe3nLC/UEQlEMWiWqHM7RhQsqh4zh+diiKuTDnSl5jKJSyrlyWU1OZT950aQRSa1TVWLgR59ZOPABpVbK+PQlUzh3XvmYTBI7HEgkEs6eVcpNyycPXPc6tZwvXzeTaeOGx0YKoDhfy2VLqxmsmaePL0CjlmM9zoTaJ5vjShXy1a9+lX379vG3v/2N2tqPsljv2bOHr3/96yxdupRf/epXw1rRU4lIFXJ6kakd7U4/3X1+HJ4QRWYthflaTHm5ffVF+rvoeupXRB1daeuUJTUUXfddFKbCDFum4/AEef7dA7z8fmNKLJ5SLuWr189iyYwStOqhh1c9/qS1kNcfIU+rwGLSjMgL6ci29Ibgb89sZ+tRpvQqhYxvfWou82oLj9kVobffz88eXE9rd+aYMaVcyp03zuHRN+ro6fczqdLMnTfOpuqoWWqNHU7ufmA9Tk9qag2DTslPPreISZXmlCHGfleAXmcApydEgUGN1fzR9dDT78PpSdqYmfJUFOVrhxTF8XiCA+1OfvHQhrQ6TKww8c1PziEciSIjisOf4Of/2kAsS8/alCozn7hwEr94aEPW3/va9bM471Dv45lGtuekPxjhrfUtPDjIDOarzxnPDRdNHtHZhnZnAF8wgkohw2rSZLXIGg2crHdONBbH5ggQCEXRqeXDnjw7FotT3+qgtdvDQ6/sSYt/mzXRyuVnjWNSpWlEekFHoh2PS7zNmzePX/ziF1x22WVp61566SV+/etfs379+mGp4GhAiLfTi5Fox6jLTqi3GdfGV4n5nMgNFowLr0RprUCed2xpAAKhKHZngH3N/fS7g5QV6qkpM2E1qZGPMpeFo9vS7Qtjc/jZui8ZbDyhwsSEciNWkxb5ccav2Bx+Vmxo4dU1B/EdMmGXSmDR9BKuObeGPQf78AejzJ1ciMWkoSTLMHivw09bj4fNe3uIJxLMm1JEdYnhpLoR2Bx+2m1e6psdKORSZkywUGjWYMpTD7TllCm1dDnCPPZGHVvrezn8hDbolFx+1jiWzixBo5Kzt6mfx9/cR88RE10mlhu57YppTKwwo1WfecINBr+/fYEw+1ocPPLaXpo63QPLSy06brm0lpkTLaPaMuxkczq9c6LRGG29Xrr7fNidQdptXjRKGRMqTKiVMsaVGkds+HrU5HnTaDTIsnTH6/V6xkDeX4FgWJEbLciNFjRV00lEw0gUKqSK43sJaFRyKoryRixR6Ehy2I+yptw0bPu0mrV86mNTuHhhJc5DBu96jRKrOWlSPrnSfGhyyeBf6oVmLYVmLfOmFA1b3Y4Vq1mL1axlzqTsPbESCUwoN3HXLfOxO4P4ghFkUgmmPBVWk2bgOIvydcycYCEQipJIJLdLtr8QH9nQaZTMm1LEhHIT/e4gwXAMpVxKvkF9TK4qgrGHXJ4UaEX5WvpdQabVFCAhOVHsyPtqrHBc4u3WW2/lnnvuYdq0aZSXlw8sdzqd/OMf/+DWW28dtgoKBGOFRCxKLOAhEYsiiUYIRCAcTSCXH7vxdSgcTfqDxpKG4wVG9agbXul3B3H7IhgLx+HwxRjJD3OZVIJKISNPp0Qpl6FWynB5Q2zvtNHS5UYhlzJ1fAFF+dpBYwJPFS5fiN7+AHsa7fiCEQrztRTn69Bp5JjyVCjkUvyBGHkm68BLRKtWUFmcHCJ3eUOEIzH6XUFMBjUyqYSAx43K10estY64343cWIisbCIhaQEqTfJkBMPJXtyGVge9/QGMeiVTxxVQYFTjD0Zp7fFwsMOFXCZhSnU+xfmDT4gZjEQigcsXJhFPoFXLicTi2J1B9hy04/aGKSrQMrnSjMWkQXWMw+fDjVGvwpjFNWWk8QUihMIxZHIJxgxCOxyJYXMGONDmpKvPh06tYPr4AiwmDXknMJx72ALQ64+QIIFGJcNi0uD1RwmEIshlUqzmsd27lguHJzCMdY7rDmpqasLhcHDJJZcwZ84cCgsLcTqdbNu2jWAwiFqtZsOGZDyGRCLhkUceGdZKn47E44kzNmh1rJNIJIj0deDZthL3jneQLrmZTlk5L27uo9cRwKhTce15E6itzh9yiC4ajdPW6+H51Qfo7fdj0KnocwWYNdHKxQsrj8nbdKRweIK0dLl57p0DbN+fjG8rytdy7Xk1zJlkpdQ6vD2GnTYvWxt6eXNtM/5QlIrCPC6YX4FRr6SuuT/F8aC8UM93Pz2PcaXGU3I/JRJxJJJUkd3S7ebeJ7bS2JGcyGXOUzGlOh8SCapKDKiUMiqL8nh7Yyu9jgBXnQPTaywUmrXYnH52N/bxzNsNdNl96DQKLl1azTkzi1FseozA7tXITUXI9CZCO3twBzxoZl6Iccl1+GV6nnqrnpUbW5k3pZBJFWb2NvWxblcX58wuo7xQz/0v7GR8qZFILM6Tb9Vj1Kv47qfnMaHChFwmzem5lEgk6LT52LCni5UbW5FI4AtXz+CF1QfYdlT8o0wqYfmSam64cOKIDVHJZPLjtkIcSfrdARrbXbT3eKgqMdDvCRKPw8wJlgEnjH53kJfXNPLSewfT8o1NHZfPN26cc1zPAJvTT2evjzfXN7NlXy+xWJzJVflcfc54Om0+nl99gDydkkuXVDFrYqHwxR0DHFfM2y233HJM5R999NFj/YlRxUjFvAVCUTpsXlZtaqWnz8/iGSUpN/LRnE7xB6eSo9sxEo3T3uNh874e2ro91I7LZ9r4gpytWYKdB+h6/KckwiHkl36Xv64JUNfqTitXatFx9xeXZPX3jMUT7Dpg55m367l06Th2N9rpdQQYX2qkoljPi6sb+fZNc4fFMibqthOPhEAiRa4zIVXl9iJ1eoKs3trOgy9nDvqeXGnm2zfNHTaR2dzp4n/vX4vLm56w96yZpSyZUUJ7ryfFwUCtlPG7O5ZRXZpM7eILRAbuM6c3xLJZZUwbX5AyTBYMR+np97Njvw1/IMrsyVZKC3QYjuqdcXtD2F1Beh1+IpEYVrOWfKMKU8JNoHEbgYPbUVgr0U9fhiK/hM7+MHf95X08/ghKuZTbr5hGPJFgR4MNpULG/Noiygp19LuCODwhNCo5bl+Y1Vvb+N7N8/ndY1s42JE+e1+jkvOr22qRh/3s6Y7Q3h9hSqmaao0X3r8fZVktzeOu4eWNNm6/Yhob9/bQ0OqgsjiPRdOK2LbPxrSaAvzBKKu3tqOUS5lXW0R7r4doNM7CacW8u6UdfzDC+fMqmHSoxywT9S0Ofnz/2oEg8G/cOIdnViXFZjYWTSvm65+YhWkYe0nbezzUtzn4cHsnWrWcixZWYjFqaOnx0NDqoNSiY1K5gbymVUgTcXSTF6IoKM/Jc/hEsbsCrFjXzJIZpWyu62Z3Yx/FFh0Xzq+go9fLpCozeVolD72yJ+ss4okVJr507QxsjgDtNi9VxQYqivRDOm30Of1s2NPDP17YSaa3/fnzyikwavjvISeSfIOau7+wmHGlxpP2zumye6lr7qe5y82kCjMTK00U5Z8+1m6jZsLCmcZIiLdwNMaa7R388clUI+8Co5r/+8pZGV9+QrwND0e34479Nn7+r/UpmfXNeSp+9sUljCtNz+12JFG3nY6Hv0/M60RRs4BXJRfw+sb0WaeHmV5TwA9vX5hxGLXb7uP/PbqJS5eO42//3ZGSx0unUXDHJ2bz0vsH+NFnFh33kE8iHsPd1U6rPcD6Bg+lZgUzKzWUlFhzmlixr7mf7/1lzaBlPn7+BG66ZDJK+YkNjdmdAb7/1zX0DmKpdcultZgNKp5cUZ9ipTVrooX/uW0hUqmElRtaeOClpGVYbXU+V50znj0H+7AYNSyeUYLVpGHdri7+8MSWlJfbhfMruP2KqQMCw+MP09rt4c9Pb6PzkDBRKWR844aZVO1/iuiBI2Z/SqRYbvk1D7zvYvXWdiA5A/SNdc1pYuzCBZWolDJeP+SvW1GUx+evms57W9t4Z0t7xuPON6j5wjXT+dNT2wge4QZgNWn4yXWlxF/+OfLLv49LV8XP/7U+pYxUKuHXXz2L+1/YlaEuFcyoKeCPT21PWT6xwsT/3LYgbVitp8/HN+99byAT/rhSA/OmFA0IgcH41ZeXMnOQ9DLHQluPh9/8Z1ParOTLzxpHMBxj1aakw4lOLeenn1tI3ju/JdLXTvGNP0RTPbIT0WKxOM+vPsDsSVZ+/I+1A5NuIBmn+J1Pz6PP4WfOlCLu/MPqjPuwmNT88LaF/N8jG1MstsoL9fzP7QupHCQ+ds/BPn543wdZnWQAvnjtDJ55u2FgJnRlcR4//fxi9CpG/J3T1Onif+77MMVlIt+QfA+WFZ76kYbhYCTe3ccVRNPS0sJ7773Ha6+9xuuvv86aNWvo6sr+whKkY3cE+OszO9KW97mCPPfugZRs8YKRo9Pu5U9Pb0uzRHJ4Qjz2xj48/sEtmsJ9ncS8TgAiky9i5dbsBugAuxv7sGfx66xvdXDxoioefmVPWgJWXyDCKx8cZFyZ6Zj8QY8m4razt93PD/9Tz8vrO/nHGy383/Nt2Hr6SCQGt4XxByO8ub55yN94c10z3fahrb6GotPuG1S4Hf6taCzOxxalJsHdsd+O3RnA7gzwr5eTwk2lkHHVOeP53WNbePWDJv792l7u+ssaOu1e7n1ya1qvxKrNbTR1fdSD6nAHefiVPQPCDZKWX79/Yhvh2ktSN07EsfX7WLO9A0i+ZPvdwYy9aKs2tTKlyjyQGqGtx8OHOzsGtQa6bGk1/3m9LkWUQdIC7bnNHpQTFyELeXjgxV1pZUoKdOw6YM9SlzakUmlasuj9bU627OtNK3+gw5VSz4sWVLJyY0vWeh/Js6v24xvE6i1XEokE63Z1ZUwn89qHTcyZZB1wSvEFo/ztuV0kzvkiiWgY22t/J+LoPuE6DIbNGWDauHweemVPinBL1h3+9uwO5kwuoqffnzU9xmcun8bjK/aleaO293p5+b0D+AKZn1MOT4D3trYNKtwAVqxrTrmHWrs9Odn1nShuX5g/Pb0tzR6s3x3kP2/sJTgM3qanK8ck3l599VWWL1/OJZdcwpe+9CW+853v8O1vf5svfOELXHDBBVx66aWsWLFipOp6WtHV58vqofbuljZc3lDGdYLhxeEOZhVDm+u6cQzhhxho2jnw/66oMicvyY7ezObja3d1opBLs7609xzsY2K5iZbu9CHZXPF4/DyxJvVl1W7zYfPGBkRoNpzeELsO2If8DV8wmnGY81jZsd82ZBmbM4BMKs3o3OD2h2hsdw6IslmTrKzZ3pEijN2+MPtbnVknlLy1vmWgvMMTor7VkVYmnoBGexyZPtWZwROWDORqmzu5kHW7sn/g1rc4qC79aDh89ZZ25tVmnxVbYNRkHZZcs9tGtHoJPrkxJR3GYeZOLmTtIHWpa+rP2OP8xrrmNL/ZD3d0pPytVslzPvf7WvrxD4N3pd0VGDRhcVOni7IjRjJauj14pcmeqqizh6g3/ZwOJw5PCKVSzu7GdG9kSIbPOL0hDna4sk4WsZg0A24vR/P+9k763ZnfF15/lJ0HMv/ukbR0eyg6Kh63odUx4jMwHZ4gje2Zk/qv39WVlhNR8BE5j2s8++yz/PjHP+bSSy/lW9/6FlVVVeh0OhKJBH6/n5aWFt58802++c1v8tvf/pYrr7xyJOt92iPGskcHI3Eesu0zlwCGBAkkJ+C+l0hk3jYZPTFEBRIn+bochoiOtF1k2GUiAdmadKgei9R9nMB5GeLv9PIn1jaDRctkvz6H03l1GK+loS/bU85QV8bh83E8V9CYjnwa+kIXZCHnnrcHH3yQT33qU9x7771ccskl1NbWUllZSVVVFbW1tVxyySX88Y9/5MYbb+T+++8fyTqfFhQX6LL6WJ43txzTKZrGfqaRb1BntUOZV1uEeQiXBU319IH/N8ojOXlxHm29dJilM0qIRuNZrZymjsunsd11QhMWDHkaPnl2ao9OqUWLNU+OTGcadFtTnoqZOfgDatVyDMPgDTlr0tDxUAVGNfF4IuMXukGroqbCOKCpdjTYWDa7LGX2ZJ5WwaQqM25f5t6iSxZXDZQ36VVMqkz3PZVIoMYiJebpT1mep4wPbLutwcbiGdlnQE6pMtN8RC/ZuXPL2bov+xC8wx2iuCBz7MxZ06zIWjagi7qoLkm/VrY19LJ0kLpMrc6nqTO9N+TSJdVpzgNLZ5am/B0Kx3J2J5hSZR6WlA0Wk4bz5pVnXT+u1ECn7aPe7sriPPSJ5N/Jmbrp53Q4MeepCEdiKbZqR6JWyjAb1IwrNdDnytzT3+cOMDvL/bBsdllWz2CdRs70mqFtpiqL8tKGSSdVmkdcGJoNKmoyeEZDMgm3MUeXmzORnMVbV1cXF1100ZDlLrzwQtraBvfcEyQDi792/ay05fkGNddfMHHEPCgFqZRY9Nx542wUR2X/N+lV3HJp7ZAJTxUFZQOiR9nwNhfNGdwGa+r4AixZUiRMrspn5cYWbr98apoHn04t56plNTS2O7Gajz/FgsJQwLRyLb+6ZTKXLSzhC8sr+d+PV1FYWIBEOvg1p1UrWJ6DwfryxVWUWE48KLfUoh/SZ/CypeOQSSWs2JAaZzVjggWLSYPVpOEzV0wDkvFpr3xwkO/dPI9Ll1Zz8yVT+N0d51BaoOMbN85J6zg7b145444Yysw3qPnMlVNThpcUcinf/uRs1PVHhYtIpFjNOs6elRQ3bT0eLEZ1RjF1/txyGlodA0OsZVY958wuQzOItdVrHzZx62VTUR3ll1pgVHP9gjwiDeuIqfL44jUz0rxSO21eZk60ZqzLefPKicXjRGOpL+2acmPGYdwJ5SZ0Rzg5rNrUykULK7PW+0iuv2DioJ6zuSKRSDhrZmnGpNbLF1exc799oAdVq5bz1eumI1nzTyQyBZbLvozSXHzCdRgMi0nD7kY7n7tqeprrhUQCX71+Fjvqeyku0GW1RHv4lb3cfEkt+UclEi6z6rnm3Br0WczU8w1JYTtUp/DyxVW8dcQ9VFGUR+FJyPlm0Km488Y5KdcQJAXvrZfVDnoPnOnkPNv0iiuuYNmyZXz/+98ftNzPfvYz1q5de1rFvo1cqpAIHb0+Vmxoprc/wJIZJcyeZBWpQkaYtFQhkRhtvR427OmhvcfD1PH5zKix5J4qpGM/XY/fTSIaRn7pd/nze37q29JjjYrytfziS0sosWTueYvF4uw8YOe5d/azfEk1O/bb6HX4GV9qZFyZkRfePcC3PjWXqgwv3WMl6jqUKkQqRa4zIlXldj05PQFWbW7n36/uzbh+YoWJ73563rClCmnqSKYKydQztmRGCWfNLKXT7uWJFfUDy1VKGb/9+jLGlx1OFRKmvdfLyo0tODwhzptTzrQaS8qLMBiO0tPnZ2t9L75ghHlTiii16NJm9bq8QfpcQXr6/YSjcQrNWgoMKsy48e3ffChVSAV5M85DUVBKZ3+I7/55Db5A0svy9iumEonG2dbQi0ohZ+HUIsqL9NidQfrdQbRqOf5AlHe2tPK9m+fz20c3Z4xb06jk/PLWWhTRADs7w7T1RZhapmK8LgDv34+ieAKtNR/nlc12br1sKut2d7G/1UlFkZ6lM0rZ2tDL9PEFeP0R3t3ahkohY8HUYrrsPqLRGHOmFLFqUyv+YJQLF1QwpTIfS4aPhkQiQX2rgx//Y+3AxIhvfnIOT62sp7sve8D7/Noi7rxx9rAmVG7r8VDf0s+a7Z1o1TIuXlRFoUlLU5eb+hYHpVYdtZVG9I3vIIuH0NUuRWE5SalCnAHeXNfE0pllbNjdxe6DfRQXaLloYRU9fV5qys0YdEr+9dLugdnJRzOlysznr55Ol91Hh81LdamRqqI8yodwYrE7/azb1c0DL+3KGIlw7pwyigp0PPN2Mt2OOU/F3V9Ywviyk5cqpNPuZU9jHwc73UypMjO5ypz1PTgWOaWpQl566SV+8IMfcOmll3LRRRcxbtw49PrkA9rn89HS0sKKFSt44403+NnPfsYNN9wwLBUcDZwMb9NYLD5kBn0h3oaH4W7HRCJOpK8T9+Y38ex+D+nS22ijmBc3HUrSq1dx3XkTBhwABiMajdHa4+W5d/fT7w5i1KmwO/1MG2/hY4urUgKvTxUOd5DmLjfPrtrPrsbkBAarWcO159Ywd3LRsE/v77B52VLXw5vrW/AHI1QU5XHh/ApMeSp27LenpKUoKdDxvVvmUVNmyphcdjiTYWfaVyIWRSJL7S1o7nLxu8e2DMyGLDCqqa3OJ55IUGbVo1XLqSjM453NrXT1+bly2XhmTrBSlK/F5vCz84Cdp99uoPtQtv1LllRz/pwSFBsfJbDnPRT5Jch0ZiLObmJeJ+rp52E6+wb8Uj2Pv1nHO5vbWDythJpKI902H05fmCXTS6go0vOnp7dTXWogGo2zY78dvUbBd2+ex6SKpOVYIpEY8rkUjyfotHtZt6uLlRtakUklydQTqxrSgvSlUgkXLajkpuWTRyxJbyAYwtHfj8lkzHh/x+MxpEP0Mo8Efa5kkt6efh8VhXqc3hChcJxZR3yw97sCvPheIy+vOZjWCze50sS3PjXvuO4vu9NPe6+XN9YeStIbjzOp0sw159bQ3efn2VUN5GmVXLa0mjmTCwc+XMU7Z3g45Xne3njjDf70pz/R3NycNgslkUhQUVHBV7/6Va699tphqdxoQRjTn16MVDvGYxFiHkfyBa5QEpLnEYrEUcilx+w3edgeKxKNo1TIsIxGeyxXAJcvTDQaQ6tWUGrVj+jsNIc7QCAUQyGXolLKCISiHGh30dTpQimXMWOCheJ87bB4VMZicRyHfFRlUgnmPPUJiT6XN0RPv5+d+234glGK87WUWHVoVQrMhkP2WMEwbqeDilIrGk2qsHF4gkSicaQSCeY8FTKZlIDbRdzbR6h1D3G/G5nRirJsCnKj5SN7rFAUmzNAXXM/vf1+THkqptdYkvZYgQgt3R5aul1IJVImV5kpsegGRJXXH8Z/KPGuVi3POjR3mEQigdMbIhEHrUZONBrH5giw44ANty9Mcb6O2vH5WE1JT9qRYrQ/J72BMKFwDLlMmjFfYygSw+5IWpp12LzotQpm1FiwmjUn5FubSCSwOwMDaTlUysP2WBECoShyuTRtqHS0t+VY4ZQb01966aVceumldHR00NjYiMfjIR6PYzAYGDduHJWVucU6CASH8fjD2J0B/MEoaqWMAqMG0xgNUo373MR8TuJIcYSUdPT3YXMFMenVVJcakp6OOcYyevwRXN4QoUgMnVqBWik7psS8Tk8IuzNAIBxFo5RjMamHNZt9LJ4gEovjC4QJBsMo5FICoWjWAPRwNEq33U+fK0g8kcCkV2ExaY7pmMwGDUeGlqsUMsaVGCgp0CEB9FrFoMLN6QnR0++jsd1FPJGgptxEUb42Zfg0Go3Taffy3tZ23ljXjDcQwahTceWy8SyZUUKZVX9cIu6wl2apVYfLE8IXiBCLJ9Cp5cRiCWwOH/5gBAly3P4oR2o3jz+M0x3CG0y6NADk62XIYgH6Y1qcloWEI3G0ahnGSIT8oIeEOmm0rVbJqSjKyxgPFo3GMeqVFJp1yKQStGoFKqWcfneQpk4Xz7zdQMOh1CiTqvK58aKJVJcY0+KuAFyeIP2eEH2uILFYHLNBjVopY/W2NrbX20ACEiT4ghEWTium1KIbdqEfOCRUnZ4gEm0JTl+MYDRAT38AlzeEVi2n0KzFYtLkNLFoJHB6gtidQXzBCEq5jHAkhuUoU/Skh69i4JmhkEsx6JUnJNwAItE44Wgcty9MPJGg0Kylp99Pvzv5YWDUq1Apju05MxjxcJCop49wdxOxgAeFqQiFpQy5wTJkfO1IEY3FsTsDHOxIfvRpVApmTTxxYXwqOK7Pn7KyMsrKyoa7LmckkWgMuzNIJBpDr1Eetyn0WCMWi9PU6eYfz++gvtU5sLy8UM+Xr5vJ5CrziH6dDyexUIBgyy7sb/4LapayUzmHR97ZR+iI5KhymYRrzq3hqmU1gwoMXyDCtoZeHnhxN8UFWvKNalq7PEgk8PVPzGZCuQm5PPuLJxyOUd/q4L7ndtB+RD65iqI8vvrxWUysNKJSyHE7XfiDMaRSyDfpkCtzf3A5PEFWbWojEIwwe3IharWKxg43L6w+wA0XT0kb2u3u87FqUysvrzmI/1CSUqkkOZv3tsunUnUcs2e7+3w8vmIfa7Z1DAwvmfQqbr5kCktmlKTZWnXYPGzc0000lqCuuZ8tdT3EE1Bi0fGDW+czrtRILBZn+347v3p4Q0rAvtMb4tE36nhmVQM/+8JiaqsLjkvAtfV4eOn9RorztRj0Kgw6JcFwjPtf2EXHEbMh8w1qvnTtDGZOtNLT5+PBV3ZTZs2jusSA0xti275ePn9VLU5vnIdf3ZuSNHhylZnPXVFLZaQTrbU0q0A60O7kz09vS4mnk0iStlXLF1Xx60c2EY7G0anlXDC/kupSAz19SZ/Vy88alzLk2d7r4f1tHew9aOfCaQa0Khmr9njp9SS4aEEl72xuH5gN3Njh4qmV9fziS0uZWGEaNgHXaffy+Ot1aLVKJlaYiMXivL6ujTKrjjydikdeS8ZoalRyPnXxRM6bYcFkNiKRnRwREY3GOdDu5C/PbKe156Nkwia9is9cOY0FU4vI0yqJxRM0d7q477kdNBzxXDTnHSpXW4T3kLG9Ri3HYtTkdC3anQGeWdXAyg0tRGMJPnPFVNbv7mblhpaUBOVVxXl885NzGV92Yt7AUU8/zg+fx711BRyR/Fui1GD52OfQTl6ITH1yY9r8wQgbdnfz4nsHOG9eBVazFqlEwovvNWJ3BbjzxjlZrQtHI8IeKwdGatjU5gjw1oZmtGoFOo2CTpuXqeMLmFFTgEaV3oNxOnVhH2hzctdf12RMaiuRwI8/u4gFU0dmFthwtmMiEcdXt47eF+5BNWEhvTM+jc0To8vu481DPTdHcuXZ4/n0JZPRZRiCisbirN7SzvOrD3DT8sns2G+nu8/HpAozpVYd/351D//72UVMrspuY7WtoZe7/7kuY34yqQR+/sUl5KklPPByHbubHOg1Cq49q5zzF1RitZiGPF6PL8wTK+o4b14Fb6xr5r2tHURjcSZWmLj9imm8vbGFmy+tRSqREI0lU2U8/ua+rElUrSYNd39h8TGlP+np8/O/93+YNSD+ExdO5PoLJg70Anb0etiwt4eVG1rw+MPMmVTIkhkl/OP5nTg8yR6Ze755Dr5AlLv+sibrjD9I9or88VvnDhkkfjRtPR6eemsfc6cUpQTzT6k2c+25E7j/hV30H5UQ+ts3zWVHg435tUW8traJ+hYHFpOGa86tobxQz4/+vjZrHX/xxUXUmOIo89NniDZ3ufjB3z5My2p/mOoSAxcvrGTlxlZuvGgSL77fSH2LA5VCxjlzylg2u4zacfmolXI6bV7++85+ZpcrqA7sIb7rTeIBL4rq2URnXct979i4atkEfvvo5pTf0Krl3PvNc497YksgFMXlCSGTSyAB9zyxlSvOHs+rHx5kd2PfQF1n1FjYVNeNSa/mlQ8ODmx/w7IyLh0fxDRuGvK8kU0VAlDf0s8P/vZB2izew3zp2hl8bHEV7b1e7vrzGkIZHHZuv2IqPn+EVz9sIhCKotcouGn5ZJbNGTy1VL87yL1PbGH7/mR86pVnj6fPFciaqFmllPGHO8+hqsRwXM/KWMBD31sP4d39ftYy1qvuRD99GRLJyesB/XBnJ2u2dTC9poBnVyVjiqUSmD+1mPPmlvP8uw388PZFaRZww8GosccSnDhuX4jn3t1PYb6WF99r5C/PbKexw0UgFKWhxXmqqzeiePxh/vH8jqxuBIkE/PmZ7dgcI2/PcqJEXXbsb9yPzGDBM/92/vlKPX95Zjtb63u544bZaTmvXvngIPYsuZxsjgAPvbKbmy+Zwh+f2sab65rZ3mDjmVUNPLFiH7dfMY2/P7cTty9z1vE+V4C/PL09a2LZeAL++t8dbN3vYHdTcjjMG4jw6NtN/OeNfXhcQzs39PT7mTWpkPue28mqTW0DLiH725z89J/ruPyscexr7ufbf3yPL/76bf7+3A7mTLampT4ZOGZngPe3dRCN5mYHl0gk+HBnx6AzGZ9dtX/ANcMfiPDqB008/Moe2nu9uLxhVm9t5+/P7RxIIWLSq7A5grz0fuOgwg2S8Ujvb+8YstzRbNzTzZKZpfz56W0pdd/X7OBv/93BLZfWpm3z8Ct7uGBBJb9/fAu7G/uIRON02X38/bmd7Nhvz5o3LBSJ8cjr+3D4062FItEYr6w5mFW4ATR3uUkAn75kCvc+uZX6FsfAfldubD1k05RsX4cniEERpbr9daIbniHud0MiTqRpK7z+K75wXiE79tuYcVR+QH8wypYsjgFDYXMG+Msz2/nSb97mm/e8x479Nq6/cCJ/fGrrwASJw3V95YODVBYZmFiZOoHlvx904Erk4Vz/IvHIibuBDIbbG+K+53ZkFW4AD768hz5ngIdf2Z1RuC2eXkKnzcuz7+wncCgO0RuI8M8Xd7N6S/ug12Nrt2dAuEklyd7ZwRw2QuEYj75RN/A7x0rUZRtUuAH0vfUQUdfQzg/Dhd0Z4MkV+5hfW5TyoRRPJO/Np1fWc9ascva3OU9anU6UnMXbBRdcwIUXXpjTv1zywZ3p2BwBJleaue+/OwYupO0NNtbv6mLzvp6sL+jTAbszkDJUmgmnJzSkr+VoIOroJh70wcJP8evHtg8MVR7scPH4in1cuqQ6bZtNezInX23tdjN3ShFvb2xNGXIF6HUE6On3EwzHsnqj9vT7U8zZM9Hd5884y2/19h76XENfc+9ubUWpkGX0xYzG4qxY38K7W9pweEIkErC5rpctdb2DWj29vraJ7hx9FPtcQZ5ffWDIcut3J19OPQ4/r69rTlvv9IbosHkpL9SzfHEV/e4Aa3fm5s/88pqDQ9qmHUmn3TsQR5fpHev2hXH5QhQcFTIxa6KVl94/kPHF/Mqag5w/ryLrb+5p6scZVqRZP3X3+Xk3i9n9kXTZvGzc053m+QtJKy+7M4DXH+bDHV0srVETPbAhrVwiHETTvpGePh9nzy5NW//i6gNpvY1DEYsnWLGuOWlzlki23cqNrWxvsKV5uELyo6IoX8v2BhszJ3wkIOMJ2NMTx9+4jah7aMu3E8HmCnCwY/APo2gsTr87xPaGzHVZPL2YVZsy914/9da+rM+EYDjKi+99dL9MG29hew62cxv3dtM3xLMkG54hhBtAPOglOsKeskdicwaYO6WQV4/ofT2Slm4P5jwVz68+gHcIP+vRQs7i7fbbb6enpwefz8fChQsH/bdgwYKRrPNpgdMTIhJLT4a5q9GO8VAszOmKP5jbF51jDPjaRT3Jr8egMj/tRdTa7aEwQ2qQI2NejqSn309lUR77250Z1x9oS+bpytZ+nhwfOpEsnrpO7+Av0kg0jtMVojuLpyYk/RCPzmO384CdiRWmrNt4/JGcPGEP1yEX78zmQ4byHl84xcv0SA52uCgv1KPXKonGElm9ho/GF4jkXBaS8Z0mvTpjvrbDtGbwliwr1HOgLbPvYyAUHTImyR+KkoilXiuhSCynttaoFBzIch1C8mMiGo/T3edDGc1+PUh76yk1KzJOEOh3h4gdQztCMmffut2pIrvMqmdfS3+WLaDfE8TjD6cNLXY4Isg0euLBzF7Dw0Ugx+edN4u5PCRFa7beNV8wij+YuSc1FI7Rc8SHkdmgyurlfCSJBASOwxQ+EY8RsXcMXRBG3FP2SPyBCCUFOpq6st+Dfa4gwVA0Y8/naCTniPBbb70Vk8nEXXfdJXrXhgG9VoknEEYiSfVfHF9qxOOPpGX8P504Oit8NvK0I58880SRapKxWqq4D51aju+IB7XVpMmYXDabQ4LZoGbnfhtl1v/P3nmHx1Feb/ve3ou2qHdZtmRbcjfuNqaajsE0EwiEFggESPIl5EcKKZBO6CQBEjqYXg0GDAbbuPfeJNnqdXvfne+PtdZa7a60lmVjg+/r4sLamZ19d2Z25sx5z3kebVK7p1yrli372lO6bySrk0yGJMVNX5vCnL0bqUSEUinFbEzdcJFn1SZkpcryDDS0pb7BK2SStLv/pBIRSrmk34ebbsmDvhT8cywattd2EgiGUSmkCb/FVMil4pT7MBlisRinx0+WWR13I40br0nN9tr4AKTD7iPHok6anZJKRIj7KfZXyCQJXX1yqRixqH/PVl8wRK5FmzLgNOmUSERijDoFIYk4ZRZAMObT3haiLMkyrVp22EXxKoWU4cWmmGYeRPdTvlXHjtrkwYBeI0cpl+LuFeCYNBIidi9i+dHRm+smXbecvtwEpBJxyvNTLhWnbO6KyhQd+l27vcG07cvk0sNv5hCJJUh1qWtyeyJRH17d6JGgkEvodPrJMqlpSvHwadQpkErEJ8y997BGecEFFzBv3jweeOABwuETIzo9XrFmqNhe08n8sypiN64sk5qzJxdTXmAcVOXx4w2zQdmv2KxKIe1X0PZ4QG7ORSSVI17zGndfMSp2AdapZVx77nAWLq9JeE8qX8nSPAPLNjUyZ0pxgp2NWillWFEGbm8QSwrLqGyTut+A16hVJM3cjSjJwJLCH7EbkUjEaRMKUcikSW2rRCI4b3op5QXGmG9vjlnD6RMLWb6pMeV2Z48vSPmdemPSKzl3akm/600fHe2Gt2aoOGVEYuOLQiahotjE7gM2Fq85gFopZeywvq3Nujl9YtFhydlkZqjRquXMTjHNKZeKybNqEm4qK7c0ccmp5Unfc+q4AlZuTT3NW5Stw6iIINHE+0ZaM9ScMrL/RiCzXsWMMXlJbZVyzBoyTSp0GjkzxuSxuTmCNCfJOEViAmUzUChkrN6WWCpw9qTiw5avkUklzD21nKKc6I1fLBZRlq9n+ujcpHWVWSY1Hl+IscMy2bgrfrpwTIECuTkqXXE0sRhVKX1ouxGLoud2Kp/PLXvbGV+RvPRgztRiLCkeqNRKGRfNHBL7e9OedsZV9H+eV5aYBqx8oBs1u991RFI5sozU/rqDjTVDxYrNjZwzpTjpcpNeSTAU4dxpJSeMZMhhh5h33XUXp512Gvv37z8a4/nOYNIruXBmGV0OH7ddOorbLxvNuVNL6LD74mozvo1k6JTcMre6T7+9688f0a+v5fGA1GAlY+YVhNrryGv8jJ9fPZbbLxvN1WdX8vKinQl1e+Mrs1Jm3ixGFedNK2XRijp+On8c1UMsWI0qplbn8uPLx/DsB9u4+eLqpDpbEL1A3XBh3x3RN1w4ggKTLLYNsQimjMzkznlVGE3Gfr9vrkXDpytr+fk14+OmQg1aOXdfOZa121uYNDKHR382m4fumslvbjiFT1bVpZyqUyulzJlcnLaHoUQi5oxTitD2kVGbWp1Lpim6j/UaBdeeO5wzTymKaeyVFxhjFk4QnT7NtWi5eNaQlNvsRiyKmtXLDiMrIZOKmTEmn+21nXxvTmWc1p9Rp+D2y0fzxuLEOr7zp5eyo7aTH86tjt2cFTIJF0wvZebYPFZsSV4zJBLBdedVYtGIEjJvKoWUebOH9pnpNOmVGHRyPl5Rxw/nVpPRI1AdXmLilkuqY3WTFoOKDQd82Md+H3n5KXCwe1CakY30nJ/x7NJOpo3KY9W2+LFKJSKmj849rAxmNzkWDb+/eQoP3T2Lx352KqdPLOLdr/bxo8tGx10zhpeYuO68EazZ3kK73RtXvzezyorO14Tp9O8jVhzd60xU+qW6z3XOn1GG2aDixouqkgahi9ccYOqoXCYMPxTAiUVw+sRCLp45pM/zcUi+gfyDrgzBg8LJw4pSd9iKRXD9eSPQ9ZOJT4UsIwtFfkWf6xinXoLUcOzucxajmjNPKaLT4eOimWVx539+ppabL65i4fIaqkpPnHvvSamQNDiaDgsOd1TY0hcIY9QqsGYokabQHvo2SYX4AiE27W7nkQUbsLkOTRGqFFKuP38EU0flDvji0R+DvR/DHgfOjYvx1e9EM+F8Or1imuxBFm60s353R2yqY3xlFrdeUt1nK7rN5Wfh8hre/Wofk0bmYDEoqWt2snVfBz+cW824ysw+p0dd3iBfb27kP29viesWUyul3HRxFZNGZKOSi2jvcOL0RIVCTQYFGl36UxjtNi+vLNpBZYmFzAwVwXAEsShqSn7BjCEMyTfE6Xc1tLl49v2trNjaHDftk2fVcueVY2JWTIdDbZODv7+4NlbbBtHp4LMmFXHZ6UPjmjLC4Qh1zQ721Efrxw60OPl01X5c3iBKuYRfXjeRqlILwXCYxWvqefLNTUk/UySCn84fx6SROWlPhfVk94Eu3lmyl1FDrQgCGDRyJBIxT72zJU7nTSoRceVZFZw+voB9jQ5e/WQnk6ty0KrlRCICK7Y0MW/2EHbX23np451xxzlDp+CHc6sYZgqTkZ2LSJoYFIfCEbbsbecvz6/B6YmfSsyzarn+/BE8vGA9dleA/EwtZ08qRqWUIhGLaO3ycNr4AjJNh/Sw9rc4ePWTXSglIWaUaxATpt4u4vNtDs6bVsKrn+yKq/MUi0Xce91ExgzN7FOzMF0EQWBfg51n3tvK+MosNCoZErGIvfV2Op1eJlRm88iCDbGasflnDWVqdQ5dzgBObwiNUkaWSYU1Q33UxHu9/iBrtrfy+BsbKS8wYjao8PiCbNzdzlmTirhoRlT/MRAIs2lvO/98ZV1cbadUIuLiWWWcM6UUty+I2xtEp5ZH3SrSePBpanfzj5fXsaO2E6lExO2XjeGLtQdY3ysbqVHJ+MlVYxlVbkUukwz4Whm0tdL63iP49/f2QBZhmHQBhkkXINUY45b4/CECoTAKmQTFUdD4tLv8vPPlXvbW25kxJo9wREAqEdPS6WbphkZ+9r3xFA+Cb3QyvnF7rO8qJ+2xjg7ddi0tnR5sTj9atZxssxqrUXVUraAG3ds0HCLQ0YB91Qe4tiyBcAiRTIGqajaR8lmsqheoGmLFmqFOq94kEIx2lNY1O7C7A2RlqMm1arAa1X0GOXaXn+YON+t3tVGQqSMcieALhDEblORnarEY1QPKdCSjpdNNp8PHvgY7gWCEPKuWHIsm5c2kyxE1dK9rdhAMhcmz6rAYVeRYBi6KaXf5aevy0m7zIpOJyTJpsBqVSS/8kYhAa5eHfQ12VmxpRhAEJo7IprzAiDXj0H7ptHmpaXbw9pK9bNzdhiAcEhS+YHopBVm6I/LkbOlw02bzUt/qQioRUV6YgYhot1trpwejVsrQIjNZJg1ymST2G2nu8NDa5UGrlpGfqQNB4OstzRg0cgSiAqQ6tRxfIIRMIqZqiCWhcaSbQCBMbYsdry+M0x1g5/4uJGIR1UMsiETRDPCGXW289tnuWNOQSa/k0tPKOWV4dtImnMZ2Fw5XgJomO4FAmIIsPTqNjNcX74518UrEIqaNyuOiWWUUZesOK3uZDp0OH61dHupbnIiIUJqXgVIhZdd+GwdanWTolIwaYmHF1iZe+2x3XPmAQi7hvGklnD+t9Kh5rtpdfhrbXCxauZ+GNhc6tZw5U4oozNbHajQdbj+bd7fj9AYRi0V0OXyoVTK0KhkOt5/JVbkJFlbp4HAHaOlw0273sa/RTiQSYXS5FZlUwu56G35/iOJcAwVZOqzGQ8K/R3KtDLnthOxtuLcvJ+zqQmYtRFM+HqnejFhxaFvtNi81jXY+X3sAAZCKRZw+sYiCLF3KWYaB4vEFaevysnRjAzvrulAppZwxsYiibN1R0XeLfe7xErw9+uijKZeJxWLUajVFRUVMnToVufzoZE+OJd/W4M0XCNFu81LbGNV1Ks7RY804ur6DxwODKtIbCePbv53mV/6AEE7s+BLJleTMvw9lbv9Tcjanj+YODwdanZj1SvIzdUlvlMnosHv599tb4mrLpBIxaqWUMUOtXH/+yLgalrDPTcTnQSQWIdGZDksss6bRzq/+tRy7K4BGKUUqFeP0BIlEBG67dBQzxuSjVp5Y55Dd6aO50xOVxwiGmToqj4ggUN/qQiSCjbvbWbaxkRmj87j0tHKaO90EAmGyzNGA0aA98puM1+tl586dDB06tM/zsr7FyUOvrGfHQesqsViEQibBHwjFGhHOmlTIpbOHxgzPuwlHBNbvbOX3T68gIkQzLUVZOoLhCLVNDoIHXRX+cddM5FJJVA9OBBqlLGVdYkOri3+/vYkZY/IJhMIEAhGMOgW793dRXW4lHBFindDrdrRiNiqZN7t8UO3aeuLxeNi1a1fCfrS7/Pz3va18lkIwGmDssEx+fMWYQQ8aupw+XvxoBx+vqEtYVpJn4P+unYBWLeOdL/fx8qLodL5UIkarluEPhPH6Qxi0cn5wwUi0KhltXV7ys7TkWrT91os63AHe+Hw3b36+B7EIzEYVErGIDnvUFmtqdQ43X1yd1PnlaCcMWjrd/OvNTUwbnU+nw0dLp4c8qwa1Usa6HS384MKqo1Y+EwpHvYKPxEkiXb5xb9Nu3n33XZqbmwkEAkilUoxGIzabjVAohEgkojseHDJkCM899xwmU3rdJyc5dnj9IZZuaODR1w6JuopFcOslo5gxNi/tzsVvK4Ig0NjuZs8BGwq5hJJcQ9IGipCtleYF9ycN3CCqddX86h/Ju+4vyIzWlJ/XZvPw4Mvr2bznkM6TQSvn9zdPoSQ3eRFzN8FQmHe+3JvQFBAKR30Ml6xvwGJUMf/sCmRSCZ72ZpranOxpcpOhlVFosWPNtiBR9/05EJUz+fW/vo5N6bh7NT889vpG8jO1jCw7cWpHbE4fi1bu5/mF22Ovvf3lPkaVW7hoRhn3PX1Iw2zZpkbOmlREW1fUjzcUEdi0p43TRmWi9Hfgb96HLCMHmTkPqdZ4WOMQBCGtRrD1u1pjgRtEs4q9BVU/XrGf6aPzE4K39i4Pf31hTew37/YG2dary9XtC/HQqxv49fUT+w0M2m1e/v32JmaNLeCx1zfGySxMH53H7nob+Zla/v325jjtwjyrljmTiwfd37SbZPuxrsnRZ+AGsG5nK9v2dTBt9ODZP0YiAl+tb0gauAHUNNh5eMEGbr64Kha4QfT32911rlJIufWSUfz77c109BD5zs/U8psbJiUc557sqe/izc+jNZURgQSpkGWbmsg2a/jenMqjOuPRG6cnwP/e38bs8YU8vGBD3Dls1Cr44SXVvLF4F9edN+KoTKN+U/62g8WARv/jH/8YuVzOP/7xDzZu3MjSpUvZvHkzjz76KBkZGfzzn//kvffeA+Af//jHoA74JINDc4ebhxfEq/F3K/A3tx//zgZHmwMtTu56cAl/e3Etf/zvKv7viWW0dCa2mPsadiIE+9aji3gcBNtT3zQiEYHFqw/EBW4AdleAvzy/hi5n3/prbTYv7y9N7GrtyXtf7aPd5iVob2N3g4O7n9nOox/U8ftX9/DPDw7Q0ZpaJ6sne+ttcTWKyXjp4x19alYdbzR3eOICt2427m6nudMTVwfzvXMqefyNjTzxxiae/WAbD72ynjyLlkj9Zhqe/hntHzxB0wu/pmvJy4R9qeVRBkp9q5O3luxNa933v9qHrde5s6/RkZbO4rZ9Hf0KPgM0trmYOiqPJ97clKCP9dWGBvRqOV9vbmLmmPhg6KWPUwvLHg3c3iCvL96d1rovLdqZsN+OhHa7l1c/3dXnOpv2tNNm86Zs4jpjYiFvfrEnLnADqG918fzC7fhTaLI53QFe/nhn0mU9+WBZTVrHezBpt3kpyTXwvw+2JTx82Fx+3lqyB7FIdEzPkxOJAQVvjzzyCHfeeSfnnHMOkoPF9SKRiNNPP5077riDhx56iPLycm655RaWLFkyqAM+yeCwZntylX8goTPsu4YgCHywtCbugtLS6YkzigaIhIK4Nqd3fru2f51yWZfDF+e72JP6VhedKey0uumw+foVXg2EInTYfbjcPl74oilOuHZrrZ1We4CQO7kgbDfhcITP1vTfZb55b0dSfbvjlVVJZCy6+ejrWs47KE0ilYhRKaTU9dAYC4YiLFpZh6cuvjDbueFTwo7Bt/9xeYJpiawCbNzTjqNXQ8LXfdgi9aaxD22+bpZvakQkIqWV0tKNDWiU0gQ/XrsrcExFuN3eIBvTcBaA6INbb0/iI6HL4Uvr97D7gC3BZaOb0jxDzKasN0s3NtLpSL4vnZ4AO1K8rye+QPiw3S6OlD0Hv28q7cMdtV2U5RtPCKedb4IB5SKbmpooKipKuiwvL4+GhqjCcnZ2NnZ73zeEk3wzpFLkBvr0PfwuEIkISbNLrl43QgQBIZKmQnyKaVWAiCD0KTrbn5p/KgeBhCFEoi4CriSZl1A4ApG+p+wESNsJQTg84fxvFK8/9bHxB8PIZNFnXIlElPT7Oz1BBHHipTTVVPqRcDglyuFwJEHV9bCcIdI4r0LhCKE+zglfIJyyMSHd83YwEBD6FSbuyWAOLd1thUIRJOLk+ZS+jkWkD/eFw6loP9a9i6FwBEk/+aNwRDim58mJxIAyb0OGDOG1115Luuz111+npCT6pFpbW0tmZnrClyc5tkwYnlqoc2ISUdPvEhKJmPOnl8a9JpOKGVlmjntNLJOjKhqR1jaVxambXfRaBZNTCPfq1LKkhcQ9MfYjrgtRmYsMnQKjTsn5E+Jr7wxaORa9PEHQtTdSiTilUGhPss1qVCdQw8L4PnxXJ43MYeXWaCbaHwijVkgTNOnOmFiIolcRssxaiCRNpfnDQaVM/PxUFGbrE9xMRpSmP6Z0CsWry62olNKk2mQA4yuyaLd5ae2VXemt/H+0UcgkMa2z/tBr5KjSdIFJd3vdotV9UZitS5n9crgDZKbQhxxaaESvSV6jrFRI0jqOYhHHXJy2MFtPJCKkbG7KMqlxuAMDFgv+tjOg4O3222/ns88+46KLLuKJJ55gwYIFPPHEE8ydO5dPP/2UO+64g23btvHXv/6VM888c7DHfJJBINeiZdqoRLPoKVU5aV/kvs2U5Rv44w+nMK4ikxlj8vjbHdPJTSJroamY1P/GJFKUBZUpFytkEi4/fWjCzUwkgh/NG42lH+kCi0HFmGGpmyEg2kVnMaqQaDOYOETHT+eWUVVm4ewJ2fxh/lCyMo0Jgq5Jt1OR2a/cyFVnVgx6t97RJNeioSqJMLZRp2BKdW6cYf3Li3bymxsmMXNMHqPKrdx26ShauzwoRs7GMudmlIXDMUyZS/Zlv0CqTS2EOlByTGrOOKUwrXUvnlVGlin+nB1Vbk2ruy7LpI4JHfdFeYGRFZubmJvECSLXoiHHoqGqzMInq+Kn28+YWJhSrPpoYNQpueKMYWmtO/fUIWk7fqSD1ahK6a7Rc51cizZlZvujr2u56qyKhGMnl4q5+eLqlIGX2aDi8jS+98Th2YP6ndMh26xm1dampMdFLILLTh/Krv1dx3xcJwoDejyeNWsWTz/9NI888giPPvoo4XAYmUzG2LFjefbZZxk/fjyLFy/m3HPP5c477xzkIZ9kMDDqFNx0URWzxxfELJzOnlzCkALjUWvhP5FQKWRUD7FSUWRCLCalcLJEbyVj9vfoWvx8ym1Z59zSr5p4fqaOv94+na83N7FmRwuFWTounFGGSa/s92arUcm4+eJq7n1yGe22xCd3q1HFTRdVoVZGn87NeQVM0XcydogRmUSCTK1Bok5PnDLLpOZn3xvPn59bjQjIMmmQSsW027x4/SEmjcxm1NC+A8njjRyLltsuqWbD7jY+XF6LPxBm8sgcpo7K5W8vrI1bVyIWoVRIKM0zEIkIaFRSqobko7fowHIm2pEzEUmlaQXCA0GhkHHGhEK+WFuPRCxi7uQshlrEiIQQEbGclTU+PlrbSrZZnbRL2WpUccUZQ3mpnyL2Wy8dhUnf/00z06Th7MklLN1Yz51XjGHV1mZc3iCjyq0YdQo27G4l3xqfUSrK1nHJqeWDrvPWH5UlJgoytRxoTW1Eb9IrmVKdO6hdsHKZhMtOH8q22k7qk3y2SiHlnu9PwGxUUlVmZvPexFrJlk4Pi1bW8fPvjWfdzlbqmh1UFps5fWJhv1aDYyuslBcY2X3AlnS5Ti3jmnOHp53RHSzMBhVXnlnBO1/u5e6rxvL5mgM0dbgpytYza2w+X6yr55pzKo+aWPuJzhGL9AYCAex2O2azGXGK+foTnW+rzls3oYNt9akClG8bvfejxxek3eZly94OOuxe8jN1DC3KwGpQIU9j+iTsdeHdu46Oxc8Tdh7q2pQaszCfeR2qwhFxopR9IQgCHfaocPGnqw/Q5fBRmK1j5ph8MjPU6PqYamrucLNiSxNvfr6HLqcfk17JxbPKmDQyp08pgcPF6w/GRGN31nURDEUozNaRb41qTh0NsUuXJ0C7zcuuAzYCwTDFOXpMeiV2l5/dB+VchhVmYDGq0B7Bxb6p3UVNo50v1zewbFN8cX95gZGfXT2eHIsGQYjW4gymtEKy37fXH6Td5mPdjhb2NNjJ0CmZNioXvUJEqH4LwtrXCbYd6mSWF4xAMv4SJNYiMq1GwuFITBS4tdODUacgP1PLrgNdPPLqhoR6LIVcws+/N56ibD0tnR427412QFeVWci1apIK2IbCEepbXXyyspZOhx+ZRIwvEOLU8QVIxGIeenU9Tk+AzAw1804byphh1gEJzR7JfuymqcPN315Yy679iUX8+Zla/u+6iVER5KNAa6eblVubY+LHCrmEsyYVcfakYvIztYhEItq6PDzx5qakXrBmg5JfXX8KRTk6AsEICrk0bdHtti4Pb32xl49X1MZswkSiaMnAdeeNoCAr+Xc+2jpvkYhAfZuLhctq8AZCZOiUdNi9mAxKzphY1G9geqJw3Ij0QjRoe/PNN1m5ciUOh4OMjAzGjx/PxRdfjEJxYhi7psu3PXj7rtFzP/pCYl5ZtJOPVtTGFfdKJSJ+cMFIZo3NTzsYCDo6CDs7EIJ+RHIlUp0Z6WHWPbV0uvnL82sSOlshOtV09ZzKPqckBUGg0+EjHBaQSESY9MpBzSKEQhH2tzj499tb2LovPkOg18i5/bLRVBRlDGr2trXLw5NJbmhWo4rrzh/BSx/viGU0JlRmccsl1UcUHHQ6fLR0ePhszX6c7gAzxuZj1CrwB0IYtAo0Khmddh+BUBiLUYXFkJ5FUX/0/n073H4WLq/lxY93JBSejyzJ4IfT1YQ+/CtEejegiLBeeAfikgms3dXBo69tjOsGlUrEzD9rGJNG5rBuZyub97YjEYuZUp3DsCITbm+QP/x3ZUJXa2aGinuvPyWl7mAwFMbm9CMASrkEvUZBOCLQ5fARiQjIpOJ+6zd709blwebyI5WIMRmUGPqoywqGwrTZvLR3eYiEQ2RbdWSbDwUlEb+XkKODoNtOWKZmS3OEzza0kaFXMnt8AdlmdVrZRjjkDuP1h1AqpFgMqrSmowPBEC2d3mgjjESESa9KeCBzugO02bx8taGeuiYnOrWM2RMKYy4rAyUYijq3dDr8RAQBg0aBxaiMZeSTcazuOYFgtNs1GIogl0mwGJTHVHMOIORoJ+TsBAEkOhNSvXnQrp3HTfDmcDi45ppr2LFjB7m5uVitVtra2mhsbKS8vJyXXnoJ3WF4JR7vnAzevl1078fyoZW8/OmePjXS7r5qLKf2U68yWNhcfv74zMo+W/svmT2Eq86sGJC35mCwv9nBP15ax96G5F3kMqmY3900meEl5kFRLu9y+vjTs6vZVpNch04pl3DnFWP58/OrYwHOiFIzv7hm/KAEkI3tLv707GpqGg95qI4sNXPGKUU89Op6AC4/fSjnTSs54oLvnr9vlUrFxyvqeOz1jSnXH1li5IfD2wl9/VLCMqneSuvsX3HfM2tSvv+GC0dy7tSSOJX5xjYXP334ywTP0270Gjl/vWNG0vrPwSQQCrNpdzsPvbo+JlRblmfg7vnjKEySJXJ5gyxZd4Cn390aqxsz6ZX833UTKS8wEnZ20vH5i7i3fEm0bxrUQydiPvN6ZIbDm+Z3eQJ8vaWJ/72/DYc7gFYl4+o5lUwfndvnOdDl8PHGF7t576uaWAdlVZmZH18+hqwUmfFIRDgmDgCp+C7cc4RIGN+B7bS+9SBhtw0AsVpP5kV3oiocgUgy+A9mg8GAQtu///3vNDc388ILL7B48WJeffVVFi9ezAsvvEBHRwcPPfTQoAzuJCc5mtjcQT5c1re47X/f23rMRCJbOjz9ajK9s2TfNyZaGQyFOdDiShm4RdeJ8P7SGjrsgzPGpnZ3ysANolIUa3a0MLr80A14674OmjuOXGja6Qnw0Mvr4gI3gC37Oli1tZkpVTlEIgIvL9rZ5xgHQpvNm1Q4OG4cNTZcpgpIUl8njD6ff7/T9/tfWLid9i5vLDgQhKjpfarADaJdj6u3Hn0dyIZWF79/ZmUscAPY22DnN/9eTnsS3a+aBjtPvrk5ruC/0+Hj3ieX02nzYF/zIe4tS+gO3AA8u1bRteQVIv2IbPdmy74OHn51Q0y7zeUN8uSbm1i3ozXlewRBYNmmRt5Zsi9O+mLz3g4eeW0DLk9yHbhvMnD7rhDsbKL55T/EAjeICqs3v3I/wa70dRGPNQMK3j777DPuvPNOxo8fH/f6+PHjueOOO1i0aNGgDO4kJzlaSKVS9tbb+9Vg6nL6j1mw9OX6+n7XCYUjNLanLrg+mri9QZZubOh3vRVbmgZNpPfTVf2LAn+1voFJI+OlVtIRE+6PdpuXbbXJg+mvtzQxrodsyquf7Ir5dw4GHbb0hF2/3utBkV2a8Lpbk09TR98iu75AmJauQ0Guzenng34eZgDeW7qvX9ePI+WTVfuT6nu123w09XI68flDKd0TvP4QEVcHjtUfJl3u2vIlocMQU3a4/LyQIqh+4aMddKWQ+uh0+Hj1k+QuCxt3t9PejxD3SY4e3pqNyTUZIyHcu1Yf+wGlyYCCN7fbTUFB8qmkgoICbDbbkYzpJCc56ojF4rQDjHSFaY+U/mynukkQCz5GhCORtAScu8WAj/zzhLRU+P3BMFJp/KWsy+E/YtHRVK4BEJ3OivTYfofdSyDYvy9pugRC6W3L7o0glifWaQXD6WVsvD0Em8MRIa3j6/YGCYePnnBqOByhoY+OUHuvc8IfCtPalTrTGgr4EUIpfutCpF97u54EQpGUNlJtXZ6U530oHOnz993XuXaSo0uwM3V2LdjR/8PqN8WAgrfS0lI+//zzpMs+++yzlO4LJznJ8UIoFEq7bkejSl3QO5iU5fVvDA98Y7pHUomEnDT2mV4jH5SaPIlYRHmBsd/1LMZo12lPyguMR1xsbNAqUoqralSyOGeB4aXmQT1P0t1WiVlGyJmYOVLL0gueewo8K2TpHd9cqwblIIrY9kYiETNxRGrh5N5j1CilVCfR6etGptIg0ZmTLhMrtYiV6dfvaVQyRpYl/6zKEnNKcWqVQpry9y2TijEcQ8Hik8SjLBqZcpmqdPSxG8hhMqDg7Qc/+AEvvPACv/71r1m9ejU1NTWsXr2aX//617z88stcffXVgz3OkxxFQuHIoGRKTiQikQh5mdqU6t7dDC0wHlawFAyF8fiCsf3Zl3VQb8ZXZqU0pu7GqFOQZTqygtd2m4eddZ2s29nCxt1t7G92EAz2/+Rv0CqYmkTYuTdnTCzEPEgivVOr+/+8M08p4rPVB+Jem5TCseJwsBhVXDijLOmyS04dQn2rk2yzGolYxLzTylHKU59LLk8Am9OXMsMiCAIRQRRr9LIaVRRl9930JRZBVa4UkUSKsmA40oxD31lZv4YxQ/vWFswyqdEoZextsLGvwYbHG2TeaUP7fA/AZacPOyI5lnQYOywraVf1pJHZCeK+UomEc6eWoJRLqB5i4ZwpxZx5ShG5Fg1l+QakejPmM68HQGYtQFlchTyrGADT7Pn9ajD2RKWQcvXZFch7ZXqlEhE/OH8EWlXy/aLXKPjBhSMRi6MNNdNH5zG+MguFTMK82eUJ38nm9LGn3sbmPe1sr+mktdMTm0Y+dI0ZWKY3GA4Papb4REeRUxb32+lGqreizEtP2PmbYEBtFOeccw61tbU8+eSTMZssQRCQy+XcdtttXH755YM6yJMMPuGIQEuHm637OvjqYB3T1Oo8RpaZyTapj3mb9jeBUSPljsvH8Kdnk9c1yKVifnjJqLRsfDrtPg60Ovlg2T46HX6yMtRMG52Lwx1gR10nZ0wsIteixahL3o3mD4SRSkTMnTWENz7fk/JzfjRv9IB11FyeAHsbbDz7wfY4wU6VQso5k4s4c1Ixuf3oKlmNKs6fVsp7S/clXZ5j1jBjTD567eDIBVkzVFx9dgUvfLQj6fLiHD1GrYKGtkPTbNecU5nSSigV7TYv9a1Olm1sJBCKUD3EQmaGiuohFjQqGW8v2YvDHcBsiCr1i0VQ3+LkrFOKGD0sk7wUGavmDjfbazp596t9uL1Bsi1qLp45hKIcHSa9CrvLT3OHm0Ur99Pc6cZqUHKWzEuuVcuP5o3mnseXEkoxRfnzq6qJKAS+zLqKDleYwiFyRmaJkW7/mEjtGn5w8SX8/PHluH0hpBIRWrUctzdIMBRBKhFx6yXV7NzfxdZ9HUjEIiqKTeRbtYwoNSfIwHRTNcTCkPzEDFKH3Utjm5sv19fj8YeoHmJlZJmZzAzVgMR4cywaHrh1Ku9+tY+vNjSgUki5cGYZU0bmJO3oNGjl/P7mKXy2Zj+N7dGauLMmFzN2qBW9Wo4nrwrR1Y/z9dYmGjr9ZGfLOeW0HERmLSLR4V3rCrP1/OOumXywdB91zU7yrBrOn15GQT/6cPmZWh64ddpBMVoPOrWM//e98RRl62L7yOMLsqOui6ff2cL+FmfsvRk6BXddORYBWLishi6nnyyTmnOmlpBn1fTbWS0IAg2tLva3Olm7rQUBgfKCDCqLTWSZ1Kj6kAv5tiMzWMm58l7sqz7AuXExIKCtmolx0oXIjMevvecRifQ6HA7Wr1+Pw+HAYDAwatQoDIb0pn5OJL5tUiGhcIRt+zr43dMr8fd6ApNLxfzqB5MYWWZG+i0N4HruR5cfttd08sonO+PUz4eXmLjijGGU5Ogw9qP91Nrp4c/PrWZXEgXzkWVmZo3N57HXNzKy1MxdV45NCL7c3gBfrm/gX29t5sKZZaiVUkKhCLnWqF2OQiZh1/4uRg+1UlVmGZCmmM8fZPW2Fv764tqUZtVleXp+9r0JfQpj+gIhdu/vYk+9nXe+3EvHwUJrmVTM9NF5zBiTR2GWFmvG4EhJOFx+Nh8UT373q320HPTIVMolzB5fwOzxhdz75DJ8gTBZJjXfP3c4o4ZaD0uV/UCLk/ueWhHbdjcKmYSb51axdW8H5YVGygsyaLd5yMzQ0G734g+EycxQsftAF2ajmvEVWXFeonXNDu59cnlcx2Q3VWVmfjRvNE+/t4VVWxMFWcdVZHL7ZaNp6/Ly2OsbqW061PGq18j55bXjWbhsH0s2xnd+ikVw9exCZo3ORme20G73sa/BTjAUod3mxahToFJKsRpVvLBwB5v2tMe9vyBLx91XjeXLdfV8+HXUbaJ7f587tYTzppUmZKIb26Kdob3dA6QSEXddOZY8q4aWTi/WjKiAs/EwAvtgKIzDHUAsFpGRIkCxOX0sXF5Ll9PPiFIzzR1ulAoperWcT1fVcf0FI/lyQwNvJnkoOvOUIuafffh2bp0OH80dblo7PViMKnIsyQWMu+ly+nhl0U4+XF6bsKy8wMjPrxmPQatgybp6Hn0tXh5GKhFz5xVjePOLPexL0uk9Zlgmd1w2OuUMgSAI7K23sfDrWk4bX4jHHyIUjmDUKli/s4WSPCNVZZakU/XHSiqkyxnVVnS4A2QcnF0YrAfAdBHCIcIuGwIg0RgQSwcvoD1udN6+a3zbgrf9zQ5+/I8lKadKpRIx/7x7JkXZ6VkmnWh078fS8gqefGsrO2q7OHNSEdkmNeGIgFQiZtf+Lj5fe4Crzqrg/GmlKVv2nZ4Af3l+DRt2taX8vGmjc5GKxXyxrp5TRmRz55Vj4qZXVm1t5vfPrASi9Vu/un4ST7y5kR0HOx1FIpg9voB5pw0dsOJ4TYOdux9akjKL082lpw7hqrOGIZMlDxB31nXyp+dWc/u8MbR0upFKxQhCtD5NIhaxdkcLFqOaU8fm0emMmmlnmTVpK8H35pOVdTy8YANZJjVnTSrCcPCCHokIfLm+Aa1ayvyzKgiHBYw6BaZ+fGB709rl4eePfpXUVgyi3+vPP5qG1x9m6YZ6Jlfl8pcX1uA5WOgvEsF5U0sZUWrCYlQxrCgqytzW5eGnD3+V0mh8+ug8pFIRn69J3WE8e3wBt8ytxh+Miqu6vUHkUgkalZR3v9rHxyvqUr73R/NGMbLUgoDAn55dTV3zoSyOSa/knmsn8PzCbcw/uxKXJ4hUIkYkgj89txqdWs4dl4/CrFfh8AQRAUatAmuGKiEj32H38n9PLKOhLXlnq0gEP50/jkcWbMAXCFNVZuGuK8cMqgvH0g0N1Le6qGtxsHRDY+x1qUTEz64eT32rq0/ZlUtnl0fP+TQzhE0dbu7/70pqmw7t0xyLht/eMClp5jocEXj7iz3874NtKbc5otTEbZeO5ta/LAZg9rh8zp5cjM0V/Q099e4WtiSxzepm9vgCfji3OumD3f5mB+9/tY+po/P4x0vrYuekUi7hBxeMRECgNNcQO3d7ciyCt2RailVDzNx5+Vgyj7BE5HjhG9V5q6iooLKyMq3/hg8fPiiDO8ngEwpHWLSyrs8at1A4wsdf1w24puJEweYMsnxTE52O6FPxP19ZzyMLNvDgy+v4YFkNHl+IlxftpM2WupOtrcvbZ+AGsHxjI+MrowXYK7c2xwUKXU4fz7y3Nfb3TRdV8fyH22KBG4AgwGerD7BsYyNOz+FpUnWzeW9bv4EbwAfLa2nqSN5N5w+GeX9pDXNPLefJtzZS1+yMZfFsLj+vfrqLscOy+HhFLTsP2PjVv5Zzx9+/YF+9bUBjbuvy8NzBm25Lp4fnPtzOIws28MiCDTzx5ibUSimnjMjB4QkSEYR+ZV8gql7f1O6mqd2F0xNgb70tZeAGB2+8S/YSiUS4YEYpDzy3Oha4QfTYvLd0H+GIwCcr6/D4ot2atU2OlIEbwMQRWSxZ13cn2xdrD0SzZVoFQ/KNjCq3kmPR4PGHEozee/Pyop0EQmGe/WBbXOAG0azRwws2cOulo6ltdGBz+Wm3e2lsd/ObGybR0unhQIsbARGVxSYqik1kWzRJSyn2tzhTBm4Q3T8ffV3HrLH5AGze285bS/YO2rWl3ebhg2U1KOSSuMBNLhVz5ZkVWAwq3vw8uYxIN+98uTdtXUCPL8hT72yJC9wgqkeYSq+t3eZhwWfJZUK62bqvk9YuD2IRDMk3MH1MHr98YhkPPLuK+lZXn4EbRM+VZF2wgiCwo66L2RMK+dOzq+POSV8gzONvbCTPquXTVXU43AO7thwJbl+Qf7+1OUFLcfOeDt5YvJtgml3X30XSnn+57bbbBtVm53CIRCI8+uijvPbaazgcDsaNG8dvfvOblF2tXV1d/OEPf+DLL78E4Oyzz+aee+751ipEHw42pz+huDsZi9fsZ+6pQ76xzsajjUwmSzrN2ZP8TC1VZZa4m3Vvtu5rT7msm4gAdpcflUKK1x9iR10nxTnRrGaXwx9Xr2UyKFm7M7nY54fLa5g8Mhud+vCmE7ocPr7oJ1DoxusPYXNG/VR7Y3f6WbqxkVlj87nyzAo++ro2pgtmNig5a1IRCrmEEaWHCsD9wTBPvbuFX/3glJTF3KnodPiSTjnqNXJ+eEk1X29u4tHXNsYeREx6JfPPqmDiiKyEGiCXJ8DuAzZe+nhHTAj5ktlD2FefWnC4mxVbmhhbkXnQIiv5zWTRyjquml1M2NlJKKxi+ebU8gNKuQSnJ5hUx6wnEQHqmhxxvpNOj5+mNne/7+2w+3B7g6xKIah7oMVJS6eHlz/ZGdvHBVk6LjutnMtOL2f1tmZK8/RA35nedHT4Nu9t59Tx+Xx0MFO4aEUdF04vTekqcDh0OvyU5Rv4fO2ha5pCLuHuK8cC0OHw4e7j9wtRKaA9B2z4/CGKc/V9ZuA6HT5Wb4vu01yLhhyLhpZOTyzA6rD7Epo5Omy+Pq8h3WzY3UZBlo7LzxjG0+9uJRQWyLVo4qbMUxERose0t0ep3eVnW00HKoUUVxIZGEGApRsayTJrsDkDR+wScrh02L2sTSFu/Mnq/Vx86pBB9WX+NpF28Hb77bcf9sYFQeCxxx7j8ssvx2o9PAuSnjz++OO88sorPPDAA2RlZfHXv/6VG2+8kffffx+5PPGGcMcdd+D3+/nf//6Hw+Hg//7v/7jvvvv485//POAxfFsQBCGhzi0ZvkA4Tsfq24ZYLI7TuOqJQibhhgtHYnP6WbWtmSff3MSFM8qoLDYleDN6fek9GfoCYeQyMV5/b22tQxnQUeUWgqFIypo0uyvAQHqCwxHhsLTqwikCA0EQqCoz09wZ9RrtSYfdx0sf72R4iYmLZpbFqfS32bwEgxE4zOeAZJlCsQh+OLeap9/dkpAx63T4eOS1DZzTUMz8sytiNyK3N8CHy2sTps7srgC+FMFY73EIQur9MrbMyE0zdIjWPkNH0x6keiuXjb0QizKXV75sTFhfJhWnDAJ7k9idKsIXSE8TLBSO9JmNbLd544LjAy1O/v7SOs6fVkJlsYlQqO/ffzgipBWUAHHntD8YTisL3I3LG0AqFiedEgyHBVRKWZz24bXnDEcsFvHMu1uZd1p5Wp/h8Ye454ll/N91Exldbk2ZqAgEI2TolHz/3OE0dbipa3YwpTqXgiwtz3+4Pak+X7qd/D5/CLlMgkYpi9WSymWStM+VZOdFMCwgk4hTOjhANGtekmdAEI694oA/kPozg6HvngrC4XBUK9IjkQiPPfYYra2pbUP6IxAI8Mwzz3D77bczc+ZMKioqePDBB2lpaeGTTz5JWH/9+vWsWrWKBx54gBEjRjB58mR+97vf8c4779DSklgY/F1DIZfESRAo5BJK8wyU5hnitJuKc/R9Sh+c6ASDQQpSSDF8/7zhfLCshhc/3sHuAza21XTywLOreea9rQlTC8W58XWBQ/KNTK7KoThHj0ohJcukRquSYTYoYzeYnvtfo5LFGkMuP2MY9S1OMg52pOZYNAzJN2I2RAPGUeUWNuxsw+Y6PDV2jUpKfmb6tXIaVfLjLpdLOGtyMf96a1PS5QDbajqpb3VRmK1j5pg88jO1nDWpCL1WQbvNS22TndpGO43trn6nRHRqWYJ0yrjKrISp5958uLyWxh5TeY3tnqQ1T83tbopy+q/rjOqKCXHNCN2Y9Ep+MEWN/+3f49u3gYjXRaClBu/CfzJDvYdJFYn6Ym5fKHaM+yM3U4PXH6K+1cmeAza8/lBadY9iEejUcizG5IX4cqkYmTT55f+9pTWU5Bn6ldGRiKPTqv1h0MrRa+RoDm6vOEePVt1/MbjDHWDphgZ+/a+veeDZ1eze35VwM1erpNQ22hlRakYsgllj8xhWlIHXG0QuEyc8bKUiy6TGHwjztxfW0pbEfqsbvUbOrZdU8/R7W3h50U6Wb2piwae7ePLNzdxwYVXSzFVv4/lUFOfoaevysnlvO9NH5wHRADvHml7mKVmGSquUIpGIyO+jE3bSyGzUCskx07PsiVErx6BNvn8Ks3Rov4ExnSgc9bvzkfZD7NixA7fbzaRJk2Kv6fV6hg8fzurVqzn33HPj1l+zZg1Wq5WyskP6TBMnTkQkErF27VrOOeecIxrPiY5eo+DyM4bx+OsbmXvqENRKWayD6dypJXh9Id78Yg+XnzE0LYmME5VIJEKuRY1eI49zWjDplQRDkaRTFV+sq+eCGaVxF+iiHB1qpZR5p5ZSXWZm94Eu2m0+xpdnMLTQRLvDR3NXtK1/9FAru/Z3kd9jasNqVHH6hAJWbGnC5w/zzpf7uOuqsTjdAeqanDg80YJlqzHaqffb/3zNyDIzRm363XEqhYw5k4tYujExC9SbiqIMzAYVgVC0SL620UG7zYs1Q82wQiP7mx0pM4NyWbQrzqRXsnF3Gxl6JZdVZFGSq2dvvY22Lg8NbW5C4QhZJjV1CimFOXpyLZqkmQ6LUcWEyixWbTv00DW5KofHX08dPHbz7lf7KMnVIxaL+Gh5csunrTUdnDO1hI++ru1zWxdML2V4iRkhIjBheBare4znksmZiFb8F5JkLYJr3uaCs37Nih3x9UqRg84R1gxVn4HClOocJGIxf31+DWt2tMT2+62XVJOZoaK1j/dOGJ6NXCbmhgtG8qfnEs3pLzt9aKwb9Iu1B1jfq25z6YZGrj23MuX2IdoJOiTfgEhEynMC4IyJRRg0cubPqUQll1CYrYs1nvTF9poO/vz8obFv2tPOQz+ZFWdMn2vRoNfIOWNiEedMKabT4WPdzlZkEjE3XliFUa/gjAkFfNJHqUhBli4WqDrcAepbnSkL5XUaOUs3NmJ3xWey3N4gn66u4yfl4xLeYzGqqCwxsb0P/1uFXEJ5QcbB2tGdPHDrNNRKKSu3NGMxKGMlF6mwGlVJ9R9VShkzx+Rjc/k5dVw+n6891CBTmmfg4lllyKUSGtvdbNjdTmmuHotBdcw6PS1GFTddVMVfX1gb97pYBDddXNWvBMp3meM+tdLcHK0vyMmJF9HLzMykqSmxrqSlpSVhXblcjtFoTLp+ugiCgMdz5GbXR4LX6437/0Apz9dzx+VjePz1jQlFrmaDktsvG01xjvYb/75Hi+79p5SE+On8sdHi7UtGIZdJ8AVC6NRy7r5yDA8v2JAwvbOzrot8izL2UKJTivnbbZP4eMV+fvrY13E3MbFYxJWnlZJr0fCrfy3n/OmlXHnmMDRy4vbtxbPK8AXDNLQ5ufWSUSxaWZtQoyaViLjqrAp+cMFIHO7AYR8ba4aakaVmtqTQ74LoBfPacysREebDZTU89+H2uOnWmWPycBycfinM0nHdeSNAFJ0W0qlkqJRSnv1gG+t2xgcCKoWUWy+t5kCLkwWfHioez8xQcf35IwgEQ2RnKJI+6H3/3OHs3N8Vu1kKQnrTUHsO2HC4fYhFIrbXJfcnFQRYtrGRuacOSSojAVGpF51azp3/+IL5cyq54YIRXHLqENptPoKhMKXZaoSuAkhmoxMJoRGS2zwtXF7D7fNG87unVyb9PsNLMpg+Ko+fP/JVwtTnq5/u4vrzR/DQK+sJJJkON+mVnDquAH8gjNcf4q4rx/LBshpqG+1kmTXMmVKMWBSdNn763a1cOKOUwmwd73x5SLtPp4lORe5vdiISQYZegVkno2fTsDcQ4evNzdxwwUj+886WpN+zJFfP1OocXN4gZXkGtu7rQCSCTKMCuTR1DXVYEPHuV/FagqFwhD37uzBrJdjcIWzOADKZmEtOHUJTh4fHXt+YEAyPGWbl2nOG0+HwJZyXEK0/vOrMYTS1uxCLRUQiAvsa7FQW6ZOejw5PmFXbktcRrtneit3lg0h8bZkEuOXiKn7x2LKkAZhIBHdePgaDVkZhto79zU5+/uhS5kwp4rrzh+Pzh/j+ecN58s1NSYNkiVjEnVeMQd3rutKNWa9gzfYWzphYRHW5lU9X7ac0V09ZgZGtezuoKDah18hxeQL87cW1TByRzZkTizBqxIN2z+mLUWUZ/POumbR2eXD7QuhUMrJMasx62bfmHtRzPw5W7f1xH7x1f+netW0KhQK7PbHY2Ov1Jq2DUygU+P0D76YJBoNs35663fxYUltbe0Tvz8gq5e8vrk1awNph9/G3F9dy/y1TqGvZe0Sfc7yzb98+1Fo9v7lhMv98eR17e2goTRqRzQO3TuMXj31Fz3urSi5i586dhA92y5nNFpZstvHOssTC7UhE4MVP9nLjhSO47LRyXlq0C51KRsQDHs+hm7pEIuGy2SW4fQIfLE8M3CBad/Xch9u59dJqjFr5gM7FWy+p4ok3N7M5SeeaVCLmZ1ePRSsPsmyTjaffTZQ18AXCyKUSyguMXHNOJX97cW0sqPrV9RN5cdEONu9J3LbXH+IfL63jdzdNpizPgMcXxukJ8Oqnu3jo1Q3cfdVYQn4PjiQegzKZjN9eP46FKxvisgb9IZOJ6ezoQCyKZgRTsWxTI2eeUsTtl43mva/2xTKuRp2CM08pIkOn4MGX1x3sJt1PYZaOf7y0LpatFYtFXDZ9NtPG5xFe80bC9vV6DUatIs7XsihbxxVnDsPjD/LT+eP4cHlNnN7ayDIz155TyS8eW5a0Zq3D7mPBp7v4yfxxLN3QyPLNjYQjAkq5hFlj86kut7J47QEqijJ47sPt6NQyZo0rYNa4fLocPt5YvJsOu4+p1bmcOi6fNz7fw+WnD42J8140s4xhhRn84rGlsWuEViXj9stGkakJYrdFj7HWYKYs30Bzh5sfXz6G95ftY+/BBhCtSsZpEwopydXT1uXjby+uISLA+dNKKMk1sGVvBwapnUAgeS2WRqvHYkjMuqiVEpZsaOSZ93fgD4QpyzNw40VV/Pm5NUkDo/U72/B4N3HbpdUYdXv5akMjwVAEiVjE5Kocpo/O4/mF27l41pBYE4hEAnv27Ek6Nl1GJlqVLGmtn1Ylw+V00liXeB7L5XL+cNME3viihhVbmmPHtbzAyPfOHooCJ411rdx9+Ugeem0rNY0OFi6vY+HyaJPHuIpM7rpyLG99voeaHrMCw4sz+N6coYj87Wzblrp5ZPqoImpbvGyv6eCWi6to7nTjC4TpcPh46NX1CEI06L/yzGEs39TIZ+IDTB+ZQUdr9Fp0pPecVIjFYsw5pfz77c1s65GZPGVEFteeU0lrQ2rB8hOR2tpazObkVm2Hy3EfvCmV0R9wIBCI/RvA7/ejUiVWQCuVyqQ/Or/ff0QRr0wmY8iQIQN+/2Dg9Xqpra2luLg46XdPB5FIxFtL6pIGbt24vUGWb25i7swRMKAS+eObnvsxhIK/vBAfuAGs2NqMyaDkktlDWfBptM1fKhEztMiMWXfId7HT5uPtr+JFNXvz0sc7+fUNk1i3q41XP93NlLtnUFSU+NPzBQMsWdd3gPLSxzv5wy1TqKzse0orFbdfNooOm5+FK2qob3Ujl4mZNSaP4aUWLDoZTl+YFz9an/S922o6mH92JUXZOu57akWs2F8hk6CQS5MGbt0IAjz34XaGl5h558u95Fg0/OjSUfzlhTWs3tbC6RMK+vxO1507jLmzygiFhX6nDAFOG19AYV4mXn+EKVX+WFCRjEUr68i1qLnjstHUNTuICNGmkk9W1cXJbFw6u5w/Pbs6rskhEhF4ZUk9xZdUk6/5lLD70OdIdCb2O6XMO60cjUpGOBItHq9vcxEIRvj3W5sIRwRmjc1n9vgCwhEBiVhEp93L8k1NKRskAOqanfz1hTXMmVzMb2+cTOtBU/SlGxr5aEUdd14xhn+9FdWndHqCvPdVoiPGsk2N/PTqcXyyaj/vfLmXGy+qYntNB5Orcrjn8WVxHa0ub5C/PL+Wv94xncrKqOq8SCTC5rPx2Osb0WvknDahkHOnliAIEAiGWbzmAO98uZdxFZlMG53H4jUHeGvJXkwGJc3tbuaeWoZWmTqwvlifycqtzbFu0SEFRiwZav74v0NTqZedXs7iNfv7nFLcub+Lpg4P2SYNN19cFZuiX7u9hb88vwaxWBQnSF5VZiXXnHzaUCQScensch5/YxNyqQidShKry7x4ZjEWoxqLsTTlWG48fyiXn1aKPxBCKpWgVUlRykSAEog29f18fjUddj+frztAm82HUiZh+ugsSnLU3PO9ahyeUFS8Wy5Br5YezGBmp/zMbkaX6ijNVuMP+mlqtbO9ppOaehtmXfe1KMSri7bygwtG8vzCbUweYSE3N5fGxkZyc3Pj7r+DhS8o8MJH22jtcGDRH7om7j3Qwccr9nHx9DLE4hOjcU4ikSCRSJKWgPS85wwWx33w1j0F2traSmFhYez11tZWKioqEtbPzs7m008/jXstEAhgs9nIykptdtwfIpHouJEaUalUAx5LU7uLT1anFvfs5pNV+5k9oYBcy8BEYU8EVCoVtS1edu1PPq322eoD3HfTZBZ8ugulXML/XTeRHIs27kK/cVd7v51zbl+IpnY3Z0ws4pEFG2jt8pGXmWi7sn1NU5+1QxCVerE5/AMWUFarIceiY0iB4aBtkjiu/qiutT2ltILTEy0Cb+pwxwUwp08s5Os+pDG62X3AxrlTS4CoLta6na1UDbHw5fp6xldmUlFs6lOOSHuwHvvqOZX846V1KddTyCRMHJGNSqXiQGsXFqMqmhVJ8cCiUkgZXmqmpsnBIws2JF3HqFNgc/pSdqe+trKdu4efBmveBECsUGO66Gf89tmapM0Vd14xJrafP1pRF5PRALjijKF9So10EwoLvLe0htI8Y8K41UpZnwFNNw2trlgwLBLB2VOK+XpzU1IpknBEYOnGxuh0OdFSkm5rPYc7wFtfJM+SrNvZyu2XjWbxmmjd2Vtf7OXuK8didwfJNGWkHFuxGh68axaNbS4UcgnWDBVPvBFf72jUKVmyvn8ZnOWbGpk5Np/fPb0yYdmZpxSxdEN0G9VDzGSa1Kj7kOM5ZWQ2Qa8djSyI5KC4sVIuRasSaG5OvzQnFIBUEpIy4IxROgRBh0gEIpEft/1Q458ECIWgcwCziv5gmGKTgEmpZGplYtAnDXXx/TNycHY1E5BJkEqldHZ2IhYPfn9jMBRh4hA5E4ckCz4jNDY1nFBOPxKJhMzMTAwGQ9Jr2UCTLsk47oO3iooKtFotK1eujAVvDoeDbdu2cfXVVyesP2HCBP72t79RV1cX04FbuTL6gx07duyxG/hxSjgi4Pb2f1F3e/vXofo24OvjBucPhpGIRfz+5ilkm9VYjYkK8840hS27p2ogum+T0bsIOhXJ5AgOF6VChlKR2MnVn4zMZ6v3M6o8XvZHp5FT3+pM8Y54emaT9rc4KcrWsWl3e1SKIywg7aMOqptxwzK5eFYZb32ROK2vkEu478ZJ5Bx86HD7grzw0XZ+NG80T765KW76EqJTXT+9ehz2PkzjAfRqOZ2O1Me6w+5DMfJU9EY9MmMm8sxiahwy2m2J09siEX1KtyjkfRen96a3pM8ZEwvTllhwe4OoDkpweH0hCjJ1bNydWruwud1NOBxBIhETDIZxpHHOCkJ8Q0Onw4dcJiYQ7H+MOQe11CD64NKYRBA4HSkN50EXid5Mqcohx6Jh4de1lOYZuOPyMf1qnfndNobkyDFb8pArlEjEYqQScUoXluMNtzdAJBLVwEuGQiZBJhWjkEXdPHw+HwqFAonk8D1q+8PrDyLvTJ1FzzGrUZwAqgeCIBAKhXA4HDQ1NeH1ehNq7web436vyOVyrr76av72t79hMpnIy8vjr3/9K9nZ2ZxxxhmEw2E6OzvR6XQolUpGjRrF2LFjueuuu/jtb3+Lx+PhN7/5DRdddNERZd6+LchlErJM6n6FH7PM6j5rhb4tGLQKxCKS1haZ9Eq0ahkVfcghZJrSa+NXK6UxdXNTkloegLw05TyOZkt/f56ge+vtzD01XjvrQIuDomwD0Hc3q0QsQtzjaXTssEyWbmzEYlQiEomQppCu6I1eq2DeaUOZPiqPD5bXsLOuC6lEzBkTCxlXkUmORRu7kWbolbR1efnPO5u58qxhCEK0izEiwJihVtQHNcK8/hD6Pr57m83bp9zK0AIjerMVVd75h76v25Z0XUGI3iBTYXP6MRuUcV3QqRCJoqblY4ZZycpQM3KIBbvLj1SSXiCRmaGmq4dIb2GOnkhEYMWW5BmkqiGW2AOMXB6VoOndqdobpVxCuEcwWZZnwObyxwLsdFErpYwsM9PUcSiAE4Rok1W3Lloqcq0aLEYV804rp6ndTWaGmhFlZmoa7Gzc08bvbppMYbauT39SgHA4HJvFGazapWNNBAmhsIDUEyHZ47lWI8cXCKNWK1EqJAiCgFKpPCrBm0gsRSoLp2zEUKlUyPv4rRxv6HQ6FAoF7e3tZGZmHpV91s1xH7xBVHQ3FApx77334vP5mDBhAk8//TRyuZz6+npOO+00HnjgAebOnYtIJOLRRx/lvvvu49prr0WhUMQcFk4SvVhfNLOMf76SvK6pm4tmlpE5SObixzOZJjVnnFKU1CfyqrOG9auplWNOlBvpTbZZTZ5VwzPvbcVsUJKZwtexJFePQibpM/tVnKNPWyNsIFiNKmaMziMQiuBw+9lR2xkX2PoCYQwaOfmZ2pgR+bKNTVw4o4yXF/UtGTG1OpcddZ0o5BJOHVeASiHlQIuTq84aFsuupItOLUdXKOe2XD0eXwixSJRUT8tiVDJ7XD6L19bzxBub0GvkBzX4JOi1cv7wzCruvGIMuw90Mb4yK2Ug4PWH8AfD5Fm1cY4YcLBp4fShqJTxQbXFqKIoW4dUKmbSiBzUSik6tQyFXIpYHJXpiEQElm9qIBCKoFPLcXgCfL25kfOmlVLTuJX+GFVuxRcIM74ii5YuD0+8sZE7LhuDxxfqV4pELhWjUclwuANYjSoKsnSY9UpGlVsPThPHZxozdAqqhljiXps5toD3liaXYulm1tj8mEyNSBSd9l66oYEbLxrZ7/eLG69MwoUzyli2sYFRQzPJNmvwB0LMmVLMCwt39PneySNz8PlDzB5fQI45avUVCIYZXhwVlU43uxMMRrPmx0sJzUCQSsR4/QEMWjm2XplTiViEUiHF6Qkik4qPWOqrP2RSMUadgq4kWW2zQZlSi/B4RqPR0NbWRjAYPKrB2xHtmUgkwo4dO/jyyy9xuVzYbLa45RKJhOeee46SkpIj+RgkEgk/+9nP+Prrr1m/fj3//ve/yc+PeuXl5+ezc+dO5s6dG1vfbDbz8MMPs379elasWMFvf/tbFIpja/txPFNRbKK8wJhyeVm+geFpiG9+G9Cp5Vxy6hCuPGtYTBDSmqHix1eMYdywzH4t4TKtBn50ycgEMdluxGIRt11SzVPvbqXL6efOK8aktBzLMqv5wQUjUn6WVCLi5ourDjtj0ZNQKEJju4sl6+r511ubeeGj7eze34Xd5cfu9tNu86JTywmHIxRm6bjrqrGxOrXoGMTIZCLuuXYCM8fmxaaC9zc5uOy0oSk/V6eWcdGsMnIsaq47bwStnR6eeW8rhdk6CrMGHpDKpBIMWkVKIVStSs615w7nlrlVWDNUBEMRtGo5s8cX8vyH27lk1hCqysxcdvpQMnRKvjenMuX01yuf7ORnV49j1tj82PcuztHz+5snU5pnSFhfr5Hz48vHUD3EyqKVdbz48Q7e+HwPLZ0eOh0+/vXWJhranJw1qZgcs4ZwOILVqOLqOZWU5Oox9SMwKxbBZacNZc22FjbvbcfrD/HLayfS0OZiaKGRH80bTV8zeZefMYxPVtUhFou44cKRNHe6ONDiRKuR8usfnMKkkdmIxSLEYhFTqnL49Q2TUPbKgmRmqDhvWurru0mvZFxlFttqOhlWlMFvb5jErv2dnD+9FM1h2qUBGLUK7v3BJMQiWLW1mTc+38PYYZlxwte9OW9aCaFIhLsf+pJ7HlvGwq9r6XL4kMskaNXyAU3LDcQqMhSO4AuE6LB7aWp309rpwesLHhP/TkEQCATD2Jx+nJ4AakX0AcJqVKKQiZFKROjUMjIzVHTYvGRmqI5JrZlYLMagkWM1KmO/KZlERJZJjVop/cYsOY+EYzVmkTDA0Pqdd97h73//O62trYjFYl577TUeeeQRZDIZf//735PKdZyobN4c7dqqqqr6Rsfh8XjYvn07lZWVR/zk19jm4o3Pd/P52vpY/Y1MKmbW2HwumV2elor7iUqy/Wh3RX00gyEBuVyMUROVd9jf4kImjSqUZ2aokvoeutpb2NXo4T8L98WyUQCluXpuPKeMbIOEv79Vw0Wzysi3asnrQ+28sc3J1n2dvPLJzriOyrJ8A9+bU0lxlgZzxsCOjT8QZu2OFv7+4toEfbCqMjNzTy3nj/9dmdCAMaUqh4piE6u3NXPTxVUIAjz22kYumllGhl5BKCQgk4rwBsK0dnl4ZdGu2BSxSASjh1q5Zk4lCz7dxddbohpZCrmEmWPymTA8k9JcY0pR1MEgEAjR2OHG6Q4iCAJymQSXJ0Bju5uvtzRx3rRSJlRkIRBhX6Od1k4fby3ZE9ehOrzExOWnD8WsVxAhapMUCkcDQZNOga5XnZQgCOzc38WvnlyetMkhz6rl7qvG8vriXQwtNGHSK7G7/Og1cmwuP42tLs6dVsJ9T61ManAvFou4+8qx0elVT4AOu49cc9SNYc32Fq46axib9nYgFolY8OmueP9cvZJLZ5ez+0AXjW1uLphRyvtLa9he20l+ppb/973xOJw+VCoZErEYkUhEW5eHB19Zh1gk4oHbpsUaZiIRgZpGO2t2tPDeV/tidZtiUVQo+KKZZagUEnzBCF12H6u2NTNpZA4eX5Ap1bmolemXANicfl78aHtccwfAVWcNZXxlNh8tr2XJ+oZY5tpqVHHRrDKqh1j4x0trqWk8VJc5pSqHW+ZWp+3A0I3P56OmpoaSkpLD6r4MhsLYXYHYFHVPVAoJmRnquOnBUDhCMBQhGAwjkYiRy6J1dd2WUaFwBKkk+lp/04qCIOD1h2hsd8cy41ajEoVcenCKPdp04Q+ECUcEzAYlcpkEqURMOBzG5/MdtWnTngSCYQRBQCQSnVBTpb1Jdo4M5r27mwEFbx9++CF33303F1xwAaeeeip33XUXb7zxBjt37uS+++7juuuu48477xyUAR4PHM3gzeEORMUJvcGoSrZZE3sC6c1gngCCILC/2UldsyN2c1HKJRTlRDMhJ+ITT7r03o+NbS7+8N+VHGiJ3uDOnVqCTCLmna/2xi52UomIn8wfx8TKbOQ9bJJCzk7sq97HtfUrRNXn4dKXEAiDUgrqzp1ENn2IYdKFdFjG8atnNzN9TB6XnzYUc5Lsm9Pl4dEFG9nb5OLsycVk6BSEIwJSiZjaJjufrtrPVaeXcc70oQMqjt5R28n/e/SrlFOblcUmqsosLPhsV8Kyi2aWMe+0cmxOP/94eR2XnTaUVz7ZSU1jtHZSq5Jx9ZxKHC4fk6tzcbiDhMPRqUCjTo7XH+JAiwtfIIQgRM81s0GJNUPdb53RkRCOCGzZ284f/7sqrgng1HH5ZJs1vLxoJ2IRPHT3LFq7vHyx9gCnTSiIyoX4QwSCYRRyKWqlFJNBSafdlyAXcs7UYuadOhRLxqHv0dju4s5/LEnZeDBtVC4qpZSxQzN5adFODrQcCizyrFqunlPBtn2dnDu1mN31Nt5YvId2mxeVUsrpEwqpLM5ApZDxx/+uStCQu/XSUaza2kxtk4PaJgdnTSoiy6SJHY8ss5qmNhfeQJh9DXY+W3Mgrokmy6TmrivH8IvHlgHRAPy8qVH5i/eW7qM0T8/vbpqCQaugvtXJnQ8uIces4cxTig7WSEUztGt3tLB8UxPXnT8ci0GFLxBmW00HX6yrxx8I88+7ZlKWb+zz+LV0umnt9KJVS9myt5N/v705brlGKeXmi6t57bNdXD2nAoNWgdMTRCwSoVXL2LXfxvtL93HFGcN46NX4UpGfzh/LzLEFfX5+bwYSvAmCgN3lp60PSzeNUkpmhhqpVEwwFKalw4O3xzkmFovItWjosHvx+g+9rpCJyTZr+gx2/MEwB1qcCb97lVwSmwWICAJikQiJRBT3gHosg7dvC8cqeBtQzduTTz7JFVdcwW9/+9uYWCnA3Llz6ejoYMGCBd+q4O1o4fQEeGHhdhYetOdRyCTcf+tUhhambp8fLFo6Pfz80a8SZCHUSin/vGvWYdcgnag4PQH++cq6WOCmVkopzTMkSC+EwgJ/fX4ND//01DiZjpCzE8eahQihACx7DiVRxSaA7j1rX/EO1ktG4guE+GRlHWdMKEwavHXYvCzfGpUDePaDRJFcgJc+3cv4yiyyM42H9T3d3iDPL9zeZ03a9tpOzplaHFOa78n7S2uYM7mYA60uplbn8vzC7XFZRpc3yL/e2sRP54/j+YXbuemiKrJ6NXNkmzU43AEEISrPcSy8FJs73PzhmZUJ2a/P19Zz/fkjsBiVKGRSttV08sSbURmKrzY2UpSt4+KZQxhZZkEpl2LUKahrdnD/f1clZC0/XFbLiBIzMzLyY69t2dvRZ8foKSOz2bq3g49X1sUFbgANbS4+WFpDYbYOXyDMrLEFjB2WSSAY7Vh2ePwcaHbx2OuJ3bN1zc6Yjt700Xms3dHKSx/vjC2/cGYp63dF+HB5bcqxtXR6aLf5yDGraerwIAjRoO22S0ehUUrZ1+CgrcuLXiNn1dZm/IEwtU2OhMCqm9c+282ls8t5qpcLw8Kva7npoqqUgUd9q5NfPLYUuyvAzRdX8conOxPWmT2+kA+/ruFAq4sHnl2DSiHFbFASCkdo7fLGzmObK2pR19JDW+Olj3dSPcR62Nm3wyUYivTZqQxROaFQOIJYIsLm9McFbhDNcDa1uzHplXj9hzLy/mCEDrsPa4ppTkEQcB78zfXm/j/8hpbmJv70jydQyMTkWrV9TpV+/vnnFBQUfON6p8fLOL5pBjSpXVNTwxlnnJF02ahRo04awKdJW5cnFrhB9AnpX29twuVJTzLiSKhrdiTV8/L4Qv12on6b6LB72V57SOdt4vDsmOZTbyJCVNizJyFbazRw64OIx4HY20lxjoFQWOBAa3LLpO01qUVuu3F6gv121iXD5vTHqfinYmddFyVJzNpD4QhNHW6+XF+P2aCKC9y6EQTYuLsdjzeYVA5FJpVgNqiwGFXHzAS7rsmRUptt8ZoDTBuVx+zxBby+eHf8+5qd/PPV9fgDYYwH6/H2NdiT2lEBvL1kL46DsjFuXzBpA0w3CrkElydIRbGJDSk6Nbfs62BoYQY1jdGpW71GgcWoIkOvZMXmZjQqWVwg0pMVW5oZMzQzqcjvhMpsvkxDF+3ztQeYMzm+lm3pxkZOGRmVP9i6rwOHO8CilalV/buxOf1JTe6XbmxM2egTDkd458tD07AKuSTpuiV5enb0+P16/SHqW100d3jiHkC+3tzEmKHxEjeN7W7srvSkfo6EUDjSp+ByN15/iHA4gj3VPokIJJNac3uDKaVhwmEhZmnXF/5ghHAfepUNDQ3ccsstdHT0f406mhwv4zgeGFDwZjab2bs3uXXS3r17T9gW6mONJ8mTeVuXt1+trcHAnqT2opveT/PfZnprRKmV0j4vdu09AichEu43cIsRDsWkV7ov0r1JZrmTdFORw3e9CKX5Hl8gjCyFRIw/ECZ4sNYrFXaXH5VS1q9w8bGiry5ghzuARilDp5YlrSsD4gK/3t2XvbcVCET3SzgcwR9IfSzlUjG+g/VFfRGOCEmvES5vkEAf14hIRIjqToUjCSUYUokYXx9j6ybZeWB3HQrCvIEQ4YiQ1rYgeRdyd41TMkLhCA09HhBSZYzTLfrxBUJJM3zho9xNCSSV40hGt15fX0MSBOhdMCH08R4BASFtvc7j4zfbF0e7+/VEYkDB2znnnMPDDz/MRx99FLOiEolEbNmyhccff5yzzz57UAf5bcViUKHp9UR6+sRCjNqj3xlbmmdMuWxIks65bysGrSImUgpRHbMRJakfPsYOO+SMIBJLkGjS21cilT4m25BlUieI/QIUJcl4JUPXh/p7KlSKqExFfxRkamlNIZppNigpzNYhk4pT1mWOKDXTaff1qWN2LCnJTb1PR5aZ2VNvY1tNZ9xx7UanlpGhP7Sv++rQHjPUikEXbdJSKaSU5KY+L9y+EBk6BYFgGH3KLlkZoXAkqcNJeYERjUqWsu7RalTh8YVQKaQJAWKnw5eWO0dproG6Xhn4qjILe+ptQFQTTqWQUpzGOSsSgSzJ+Z7TR62WQi5lzuTi2N+pJCOCoUjCNTQZhVk6mntlKqUSEapjIACb6rfSG7lMcrBYP/VtWSwWJYRYUoko5WeI0yz+v37+RTzzzNPcfvvtjBkzhlNOOYX777+fUChEY2NjbKbtmmuu4ZFHHgGiiZobb7yRMWPGMG3aNH7yk5/Q1nYok/y9732PX/7yl8ybN4/x48fz9ttvA/DGG28wZ84cqqurmTNnDs8++yyRHg+Xb7/9Nueeey5VVVVMnz6dP/7xjwQCgZgsWO9xfFcZUPB25513Mnr0aO68807GjRsHRA/UvHnzKC4u5sc//vGgDvLbSpZJzQO3TWPsMCtZJjVXnTmMc6eWJL2xDzaZJhWnjU8s1p01Np8s84mrYXS4WI0qrjv/kDzHzv1dDCvKSHpTLcnVJ9yspAYr8syiPj9DWVyFLaKmzeYlx6whK0VnZX6mtt8Aq6rMlFJqpC8sBhUXzey7RkQujVplJctC5Vm1ZJrUTK7KZcm6ei6elbitomwdcpmEaaNzyTlOzqEci4YJwxPFuZVyCVOrc1m9vYWlGxu57PShccdFIZPwi2smYOnRTJFtVjOql84ZRIO1c6aWxAq9ZVJJnLxKbyKRaAH7qq3NXDq7POk6l8wuZ8XmJrKT7MfyggyWb2rkgunJPTSvPXc4n63eT2N7ohvBy4t2cOHMspRj62bqqNy4rk6dWkZliYkdtV0oZBJKcw2oFFLmntp/3dGociub9iZO2V92+tA4a7beVJdb+NG8UeRZtbR1eZPu+y/WHWD2+MIk745n4ojshJKHWWPzB/RbOlykEjFqRd9BolgcPeekEjHWFGPSa+R4fInlCBaDKmVwK5GI+5Wc6f78xx97lAkTJvDWW29x++2389xzz/Hhhx+SlZXFq6++CsAjjzzC9ddfT0tLC1dddRUFBQW8/vrrPPnkk7hcLq644go8nkNB8ptvvsk111zDyy+/zMyZM3n11Vf585//zG233cYHH3zAnXfeyX/+8x/+9re/AbBjxw7uvfdebr/9dj7++GPuv/9+3nnnHZ566ilycnJ47bXX4sbxXWZAjx1yuZynnnqKZcuWsWLFCmw2GzqdjokTJzJz5sxvdafiYCISiSjJNfDzaybgD4TRaxVpP6UdKXqNgmvPHc6U6lw+XF6DIAicM7WUoYXGfu1hvk1IJGKmj84j26TmpY930NzhYd2OVn5/8xTeW7qPZRsbkUnFzJlczJmnFCVc7KU6M6bTv0/L639BCCRmrMRqPYZpl/OPRY0oZBJunluFJYXDQqbFwE+uGMXv/rc2qTWZRinlpgtGoNMdfmAkFos4dVw+Szc2xDpEeyISwXXnj+CDZYmCq3KpmLuuHEOGTok/EGJ4iYm2Li93XD6arzc14fYFGTssE4NWweLV+/nRZaORHyeWNnqNghsvHMnQQiOLVu7H5QkyqtzKqePyeX7hdiIRgaI8PRajkj/dNo02m5dgKIzVqCYzQxWX3bIY1dw2bxTLNjXywbIaPL4Q4yuzuPTU8oSgPseqYcaYvJT1Ze8vq+GWudUs3dDAHZeP5oNlNdS3uMjL1HLOlGK210a10HLMiY1DVqOKYcUmAoEwP7p0FK8t3k1rl4eSXAPzz65g8+42Lpldzr/e2pTw3gMtLnRqOeMqMlm7ozVhOcDVZ1fgD4QwG5RRAeDKLKZU5fCfd6INCVeeOQxrRvQcLszSMbkqJ6W3rUoRzaD19qLNzFBR2Y+WpEGr4KxJxZwyIgepRMTEEdn87OEv46ayd9R2cc6UEnIsGpqSBKsAs8cXsL2mMy4LqZBJuGjWkGMiSSGViLEYlTS0uZNOlYuArIOdphD1Sc3P1NJu8+ILhJFKRJj0SlQKKf5gOFq+EIogl0kw6aM6bX3dcxUyCWqFNOkUfHR8UeeT6dOnc8011wBQXFzM66+/ztq1aznzzDMxmaLHymAwoNFo+M9//kNmZia//vWvY9v55z//yaRJk/joo49iuquVlZWcf/4h15HHH3+cm2++mfPOOw+AgoICXC4X9913Hz/+8Y+pr69HJBKRn59Pbm4uubm5PP3002i1WiQSScI4BkJ3SUH3DKxELDomCZPBZkBX2LfffpuZM2cydepUpk6dGresra2Nt99+mxtvvHFQBvhdQK2UHZbe0WCRoVcycUQ2Y4ZlIiAgT6JhdjSJRATabV6aOtx0OnzoNXJyrVqsxmMjENmNViVjzLBMyguM+INh1EoZKoWUH86t5qozhyESicjQK5MG1mKFCnlWCdlX3ItjzULcO1ZAJIRIKkc7YjrqUafxzgY3Yqmc3908mTyrNkGJH6I1Pu02LzK5jN/ecAoLPtvNlr3RolyJWMTUUTlcftoQCnMG3olszVBz7/Wn8MXaet5esgenJ/oUXz3EwtVzKlDJZazocRMWiWDssCyuPbeSwqyoNl22Wcu0UXnUNNl56/O9WDNU6LUKVm5pYvSwTO6ePxaPL8TiNftx+0LkWjTkmDXkHgXdwKZ2F61dXhpaXchlYopzDWQao+PpSY5Fy1mnFFOcbaDV5mF7bSd/fWENYpGI86aWcPGpQ7AYowGxxagiGArT2O5mzY4WbE4/Jr2SbLMm+l0sWi45tZzpo/MIRwQydMq4afduDBoFP7hgJNYMFe9+uS/Oy9SglXPhjDJ21HQycUQ2n6+tp6rMwmnjC2mzeVi1tZmLZw2hJFdPIBihtdVFTaOd+lYXeq2c4cVmRhQbkdn2492xjBHnjyWsKkXqaES8/yOGT5iJ0qrnp1eP57f//jpOLxDgj8+s4Hc3T2VCZRavf76b9oMSFqV5Bq46axi5Zg3/fGU950wpQSYVs2VfB39+bjURAS6dPYTTJxYiPSgbYdQpueXiaoqydbzz5b64DtvhJSYumF7G/97fFlfHm2PW8JsbTklb26+7YUStlPGX26fzxJub2F7TGVv+3/e28rPvjeerDQ18tmp/rKnEpFdy3rQS/IEwr356SP5GpZDy2xsnkd+H3uJgo5BLybNq6HL6cXmCsalP1UHJHIVcErOOE4tFqBRSci0aIoKAiEPWcXKZBKVcQkSIaukl053sjVQqJtOkjgpxu/xxjikSiYg8qxaRSERZWXxGVqfTxRwlerNt2zb27t3LmDFj4l73+/1x9fDd/uIAnZ2dNDc389BDD/Hoo4/GXo9EIvj9furr65k+fTpjxozhkksuobi4mClTpnDaaacxcuThuXGkIhAM4/QEsLn8dM/UqhQSMvTKWObzRGFAwds999zDq6++SkZG4o1k+/btPPzwwyeDtxOIb8KCxOMLsm5HK4+9vhFXj85EhVzC1WdXMnt8/jHNALq9AdrtPvyBECpFiBxzVDTTmsLKqidisYRIwIu2aib68WdDOAQSKRGvi3AwwMzxRZyhVGE2qJLWKbm9QVZubeKJNzbhC4RRKaScOr6A0ycUolRI0WvkLN3QgMMTJhQKIz2CIDszQ82ls8uZWZ2J1x9EIhGToVMhkcvYuLud0jwDp44vIBwRkIhF1DTa6XL4yLFoYk+nuVYt2WYNpbkGPL4Q4bCASilFJhHz8qKdfLH2QNwNIkOn4I7Lx1BVZh4Uk+lQOMLOuk4efHl9QsflKSOyue78EQki02qllIJsHZkmNWPKrcydNSQ6RZWhQntQ6T+qtxhg894OXli4Iy4Q0ahk/OD8EUwYnkUoHMHtDSEIAoLgIytDlfSYmPRK5p9VwZmnFNHU7sbpCWDQRGvdPluzn7YuHxq1nOvOHY43EMLrC1FeYMCkV5Bp0tJh97JyawPPvLc1rrFGJILpo/K4YkYuwfWfwLqPAOhew7thIdmX/5L8opH8+bbptNnddNr9BMMRLEYVcqkEnz9EZoaamy+qxumNSkk0tbtZu72VwllD+PEVY2jp9BAIRbi0yMQF00tQyKIyHL2nOk0GJVecMYwzJhbh8AQIhyOolVIUMgm7D9jI0CuQy8RYjCounFFGQZZuQNOVYnF0puJX102kwxEV1ZbLJBi1CnyBEEq5hN/eOJlAKIxYLEKrkuH1hWi1efjp/HEIgoBOLYudvwOZIfIHQjGnBLEkjEQiQpKsBbQX4Ui00Uchl2DQymONB8GwgD8YRioVJ3SSSiRikv3S0wnYEt8jxmxQotfIo9k/QYhao0kPifwejrB+JBJh0qRJ/OY3v0lYptMdCop7auF117Xdc889TJkyJeF9OTk5yOVynnvuObZt28bSpUtZunQpr7zyChdddBEPPPBA2uNLhi8QoqHNnTCr4fWH8ba5MWjlmPTKEyaAS/tKevPNN7Nnzx4gmna87bbbkh7sjo4OCgv7r0E4yXebTXva+fPzaxJe9wfCPP3uFgRB4LxpJQO6UB0uje0uFq2o48PltXj9IYw6BVeeOYzxlVkpfUi7EYQInj1raX3nn7HXRBIZQvhQQJp9xa9Q54xOuY0Nu9t48OVDAqJef4gPl9Xw4cG/TxmRjTVDxb1PLufPP5rGsKKBW5eF/V58NRvxffwUEVcXEZEY77g57M4+kweeW5v0Pe9+tY/7bpzMmB5F/W02Ly8v2smX6xsIRwSKc/TMP7uCumYHvWeGupx+fv/MSh64dSrD+2gGSZfaJge/+feKpF3ZK7c2Y3f7+clV48g+OOXY2unhhY+2s7fexpVnVvDG53vYU29DLOoO9kbi8Qd5/6saskxqXvw40SfT7Q3y8IIN/GT+OHz+IAs+243bG2RcRRYXzSzFYlRh0icGJDKphFyLNq7xIBQOM7zUjEgEasWhxoNuIc/CzEq8/hAtnR7kUglnTyrmg2U1sS5fQYAvNzQQCIa59oL/I/DO7+I/NByi5dUHyLvhb0gVVrbu62LBp7v43jmVrNneylfr62PHaOywTGaNzefhBesJhQWyTGrGV2byzHvbYq4MOrWMK8+sYOPuNpra3fz06nEJDRkdDh+vfrqTxWsOEApHz4dbLxnFKSNzGDMsMxq4yCRHHLzbnX6+WFfP8x9tjwW0+Zlafnz5GMpy9fzlhTWxruDzp5UwsszCgk9209ThRqeWcdnpwyjNNw4ocOuwe3nj050MywG1zQciaczhQCbte+rSHwjT1O5J2c8ZjggYtYp+p+8CwTAeXwivP4RGKY0+NKV5jeztXCARiwZc4lReXs6HH34YC7gAbDYbP//5z7nuuuuYNGlSwnvMZjNms5n9+/dz5ZVXxl7/8MMP+eSTT/jzn//MkiVL2Lx5Mz/60Y8YPnw4N910E0888QRPPvkkDzzwwIDHGwyFaUwSuPXE7gqglEtOmLKhtEPMm2++mYkTJzJx4kQAhg8fHvu7+79JkyZx5ZVX8s9//vNojfck3wLabV6efDOxHqcnzy/cTpsttan2YOF0+3n/qxre+HxPLNNic/p54o1NbN3XkVTSoychRzvti56Ke61n4AbQ/tG/CTk7SUa7zct/UoibdrNyazOVxSYigsB/3t6M8wh0AAMttbS88VfCroPaWEIEt6GEJ97akvI9ggCPv7Ex1shgc/n5y/Or+XxtfayGp7bJwZ+fW8Ols8uT+rxGIgL/e38bNufha9T1xO0N8Ppnu/uU09lR2xULPOwuP399YQ2fr63n8oMq+90dkxEBvt7SzO+eWsGO2i6Gl5p44/PdKbcL8PS7WwgEI7R1efH4Qny1oYE//nd1UmPtVEglEnRqOVqVPGkm1huI8Obnu/nl48t46NX17K238aN5oxLWW7G1GacqB8SJAZEQDuLdv53lmxt59oNtjCw1U9voYMm6+rjget3OVj5fe4AzJxUD0Xq2f7y0Ls5Oy+kJ8u+3o+4gLZ0e7n1yOS2dh+rLHG4/D72yjkUr98ckYmqbHPzyiWXUt7pQK2XoNYpBybpure3gqXe3xGUi61td/P6ZlRh0yljgplFKKS/M4IFnV9PU4Y59j6ff3cKiFXX9/q574w+GeeWTqL2bRCTCqJPT5fTT1OGhvtWF1x9KKWMRDIVpt3n7FOLocvr7lOCBaODW0OaizebF5Q3S0uWlsd19TLxRu+vLdu3ahdPp5KqrrsLpdHL33Xezfft2duzYwU9+8hM2bdpEeXnyJhyRSMQNN9zA888/z/PPP8/+/fv59NNPue+++5DL5cjlcqRSKY899hj/+9//OHDgAJs3b+bzzz+PTc92uxN0jyNd/GlI8wB02n2EUmg5Hm+k/WsaO3YsY8eOjf196623UlBweNYiJzkJRAOW/oRmg6EI+5ucSaUSBnUsdh8fr6hNuuzVT3YxrCijzzGE7O1EvMlFd2Pr2FoIOTuQ6hIzZunsC4gGJGV5Rnbut9Fx0Dz+cIn4PXR9tSDhdYc8E5tzT5/vbe7w0G7zYtIraev0sGu/LWGdUDjChl1tjCgxs2Vfoojm9tpO2m0+jLqBK9q32Xx8vSV5cXxPPl5RR1WZhbYuLzvquijO0bO3wZ5UsLe+zYVCLiEcjqQU9O3G5vQnCAx3Onxsr+3EYlBiOILv1k1Ll49XPjlUo7VlXwdl+QbKC4zsPmCLW3fZlhYuHH0aznUfJ441rOSlg1nESVU5KR+Y1u9q49TxBazY3ITN5U8q3g3w+ZoDTKrKYcm6emoaHTEHjTabl017Eo93KBzh4xV13HDBiEEpCLe5fLz8caLLAkR19uwuP1qVDJc3yIwx+XzUQwC9J68v3s3MsfmxzGw6tNu8fLqqDqNGgkopxeb0ExFFz4NwRKC5w0NBljZpFiwUFvAH+w4IBCEaIPYV4Hp8wQT9xEAwes4e7RkKo9HIJZdcwl/+8hfq6uq49957eeGFF/j73//OVVddhUQiYfTo0Tz77LN96rxef/31KBQKnn/+ef785z9jNpuZO3cud911FwBTp07lj3/8I8888wwPPvggSqWSmTNn8otf/AKAjIyMhHH0RyQipK1dGgxHmxmk30Ap0eEyoEehI517Psl3m97CuKlIJZw6mHh8wZSq+Y3trrhC82T0F7h1IwSSXzzS3RddLh/agzIi/QUYqYgE/AQ76nu9KiJNbeDYWPsSvm1odWHNSF3PlK6oayoCwXCfUx/ddDqi9YvdWUprhorGttTHyub0p137GQpHEInihVH3NzvwJ5EkOVykUmlS94rdB2wUZusSgrcOuw9xSfImlqBYid11KOPb17kcDEUwG5Q0p+jYhKgjwZSqaFdtfYsTDrotOPs4H/bW2wiEIqgGIXjz+cOxLFrS8bW5MeoUuLxBMk1qlm1qTL6dQLhP67JkeP2hWOAkEokIhgUkPe6e4YiQ8rxMV1i2P2HrVN2i/kCYATSg86c//Sn278WLFycsf/7552PepgD3338/999/f2z58OHDefrpp1Nu//nnn0/6+vz585k/f37K982dOzfWrZqM3uPoD0FIfWxSrX8ikPYvqrKykk2bok9uFRUVVFZWpvxv+PDhR23AJznxSdadl4xjocGkUcpQypM/tRZm6ZD3c0MXq9PrWBMrkn8XZZr7wqxXxYKmdPdf4hiUyDOLe70qoEpzc0pFdD/1pc1VmKOnuSO5bRMMfOzdpNsRZs1QxZo9IJo5LMxKLShr0iuSWjglQyaVJCjal+QaYvvnSAiFQklFdIeXmNnXYE94PcukJuJIbrElD3tiGl+CIKQUThaJogXtbTZvn13BhVk6Wrqix7anoHRf50NFsWnQ5DjUCmms6zkZBVk6ug4+8DW1u1OKEauV0n5115K9pzu4jwhCguiwVCJCnKJxQZxmnVZ/57UmhSLBYJx332ZEosOTAjlRpM7S/ka33XYbWVlZsX/39d+tt9561AZ8khMfi1GVUqi2G4VcQkEfF+pBG0uGigumJxctveqsigRz9d7I9NZ+XRZk5jyk+uRTCRajksw+MlXdDC00sq/BzshSMxbjwKbmxHIVGdPn0dtgR+9rxJxCe66b/Ewt1oNyGtYMFdVJBFMVcglVpWa21yav76seYhnw2LvJNKmYNjq33/XOnlSMTBrtFh4z1MqBFicF2dqknqqluQbc3hCRiNBvAGc2KBP8MLNMaoYVmdBrBsfgPNOo4IYLRqCQSxAdbKrIMqmT6vNNHm7FufHzpNsxKsJcc04lAMs2NnLGKckbySYOz2bL3vbYlHAygWqxKCpqu2JzM2aDMi54sxhUTByRnfAehUzCGRMLB027Uq9VcNVZFUmXmfRK1EppbMr3qw0NzJlSTLKPvuqsCixpdJH3xGpUcd60qCiy1xfCqJPHgjmZVEyOWdOHUK4o5QNiNyIR/bqS9Awgu1HKJShkx4em4vGKWCxK27VIJhWfEFOmcBjTpj/60Y9i/7799tuPymBO8t3AZFBy26Wj+M1/vk7pyfeDC0amVBofTLQqOadPLESllPD2kr3YXQFyLBrmn13B0MKMlBZE3UgNFixzbqHl9T8nX0EkxjLnZiQaY9LFZoOKH14yivueWpHyM2aOyWPj7jZkUjE/uHAkGtXh17t1I88sIvuqX9Hx0X8IdjaBRIo20MaPLj2V3z2zKunxEItF3DZvdExvy6BVcOcVY3nnqz18tLwOfzBMVZmZS2aX8/S7W5N+rlQi5ppzKjFojyzAUSlkzJ01hFVbm1N6wY4ut8acCfQaObdfNpp3vtzLgk92cfu80Xz0dS0bdrchlYg5bUIB804rx+cPsXB5LZefPpT/vr8t5efffHEV/kCI4hw9bm+QCcOzOGNiEWbD4HWoKWQizplawrAiEwdanWyr6eSJJPVqs8blo3HWERASp0NFMiXK/GFMUFq5/bLRPPfhNqrLrZwztYRPVtYRDEUQi0VMG5XLqHIrj722AYAXP9rB3VeN5aWPd8TqGq1GFVecOYyPVtQxtNDIjy4bHdeFrdPI+eHcaoqzdby3tAavP0RVmZkfXDBy0PX9hhVl8OPLR/Pf97fFMtFDCzO4+eIqWjrc5Jg1NHW48fpD7Kjr4jc3TOa5hdvYW2/HalRx9ZwKxlVkHXZAKZNKuGhGKUppBEEI0OX0YzLqUBzsNO0ruyiTSrAYoiK9qSbkzHolUmnfY5JJJeRZNfj8YXyBqP2ZQi75RqSeTjTkMjFSiajfqWmz4cSRChEJaU7wrl69+rA2PGHChAEN6Hhk8+ZoN2BVVdU3Oo5uKYHKyspY182Jii8QYuveDh59fUNMJBSisgTXXzCSSSOy0Q6gKD8dku1HfzBEa4eXYDhaNJydwn80GeGAD//+bdGuUvuhKSyZKRfLOTejyBuKWJr6u3j9ITbsauPR1zbE1ZNJJSJOn1hElknN0o0N3HbpKEpzDYNS/B1y2Yh4XYikUiQ6M4GIiO21nTzx+qa4uqL8TC23XTqKoUUZCSLOoXCYtk4PTpcbq1lHMATvflXDh8tr4uqrciwabp83morijEEprBYEgd0HbDyyYAO1Pfw3JWIRs8dHg7GcXk0moXCEDruPSCSCQi7FHwgjFkOGThm78XY6fDjdfvbU23nmva1xx8JsUHLTRVVUD7EcFPoMEo5EfTWzzOkFKIIg0NblpaHNRW2TA6lERHlBBllmNRk6ZcJ5aXP62Li7jX+/vSXhvDhrUjHnjjYQfPkn9DYUF8kUZF/5K5T5wxCJxAhCVAzb4w+hUUgJhCLYnH4cngBL1tezfGMjESHqpHHWpGIumFGKWimly+EnHImglEtjx9NkUKZslgmHI3Q6fIQjAlq1LKafN9gIgkBThzum82Y1KGnp9PC/D7YxqtxKhl6JWBQ9n0UiuHhWGVKJBJlUlFTO5XBwutzU1dZSUFiEShUVE+/vAQ+iRfO+QIi2Lm9cja1ELMJsUB7Mqh1/05/dNW9KpRKJ5Pgb3+HgP9itG04RwJn0irTkWvrD5/NRU1NDSUlJTOfuaNy70w7eKioqYnPBqd4iEokQBAGRSMT27dsHZYDHAyeDt6NHh91LW5cXlzeASiHDmqHCkkLMdrA4Wvsx5Owk5OhACPoQy1VI9BakWmNa73W4/bTbvNS3umjr8qJVyyjNM2DQyGnt8qCUS7FkqDAeYeaqPzodPjrsXnz+MCqlFItBFcu4JaP3vvQFw7S0u2jp9OILhDAZVFiNyn6nnw8HpyfA+p2tyKViNCo5nQ4fMqmYDL2SmkY7FcUmSnL0A65dEQSBhjYX7TbvQT0tGZYM1RF1Pnt8QdZsb+Ffb21OaPjItWj4yfxx5JkV7NgRf146PD46bH5aOjy0dHnQqeUUZeswaqQoPS3Y136Eb9uXEA4hVqhRjZmDbsQ0VFn5iESpb0Iddi/7m524vcFYIbxGKcViVJFr1Q6om/mbxOn2Y3cHsDn9NHe6kUkk5GVq0WtkGHXKQXOPSXZjTpdgKEwoLBAOR2LOCRKJKKYVdzzybQre4JBOXpfTRygsIAI0KunBc0Q8KA/Gxyp4S3va9LnnnhuUDzzJSXpiNkSdB74NSHWmpHIg/WFz+nj2w218uuoAUomYWWNyOXtKMRt2tVHf6mJ4sYmCbB0vLNzOlWdWHNX9ZdIr0zKyToVSJqEox0BRTt91gAPFFwjxxdp6/t1DG08hExMMRWL6ZSqFlL/ePj2uLutwEIlE5GfqktonhcIRmjs87K234fQEqCgykWVSo0tSJxZ7TyjMii1NcULMPWlsd3PPY0v5021T426QXU4fz76/jc/WHMCglZNj1mB3+2lq95Br1nDtecNZbhvHtFkzkUnAHRB4f10n7V/v4A+3mFIGm+12L/94aR2b9ySaxQNcdlo5F80c0ud3Op6wOX089+F2Plm1H7E46iLi9YewuwKIxSLuuXYC4yoyv9HMViAYprXLg9ef2CkulYjItWgGRQfvJH0jl0mQyyRoVbJoEkoEEnF62dPjjbTPlm5x3pOc5CSDy/LNTXy66gAAZfl6po8p4JePL49NryxecwCTXsm9101k0Yo6Ljt96AlppDwYtNu8PPVuvKBwbw0trz/E0+9t5efXjE/ZoTcQwhGBzXva+d3TK+MEVc+aVMTVZ1emzFC22rw89trGPrcdCEV4/I1N/PDC0thrX29u4rM10fPC7gpgdx3K2DV2uHl98W5GlVv542uJwsKPLNjA/31/YkLpgSAILFlbnzJwA1jw2W5GD8ukqiyxKeV4ZMOuNj5ZtR+ASIS4budIROBPz67m4Z/MojBF9+nRpltnLFngBlGJkOYOD7lWzXE5dfpt5ERpSuiLAX+Dmpoa7r77bqZOnUpVVRUzZszg7rvvjjOlPclJTtI3bTYPryw6JDw6/+xKHl6wPkF7rvOg/ZBGJTsmzhPHK+t2tKal2bR+Zyvtg7yf2ro8PPDs6gQl/I9X1FHTmCjj0c3O2q6UWoI92VNvJyTIY5/18qLkgrTd7D5gozhHn7SjcsveDtrtid+/zeblzS/6FmQGeP2z3Xh8yU3Jjyc6Hb6kdmY9CUcElm9qPCytr8EkGIr0qY0I0eC9P03Jk5ykJwMK3vbs2cO8efNYvnw5M2bM4Pvf/z5Tpkxh2bJlzJs372QAd5KTpInDFe1ciyGQ0nFh9bYWyvIN8et/hwiEwqza1pL2+l2DLPLcXQOXjIVf16a031m9vTntz2jtio7ZcbB+qz+aOtwpddaakgjuOlyBfgMJgI2723B5j//gzekJ9Kkr2M3yzU24v6FgNByJpOyq78nhCgef5LvNgCbZ//a3v5Gfn8/zzz+PTneoLsTpdHLttdfy4IMP8uijjw7aIE9yku8KffUPCZw4ApJHBQGEPh0iE1Yf3I/vY4NCRDi4QuLxGYhge7pv6W4QS77s8D+3JyfEmZbmdxSE9Nc9yUlOBAaUeVu9ejW33HJLXOAGoNPpANTiMAAAkeVJREFUuOmmmw5bVuQkJ/muotPI40RRJRJxytqpscMyqW1ypC04+W1DLpMwdmhm2utn9NElOxAsRhWKFGKrZ00uTlmHOLYi/TF3ixjre50XqcixaFL6NuYk8e7UqWVJhYp7M7zEhCpNx4lvEq1alpbI9cQR2ajT+N5HA4lYRLL4WiSKyrOEwxEcLj9dTj8O93czq36Sw2dAwZtUKkUuT35hkcvlBAL9p+VPcpKTRAVQLzttaOzvBZ/u4rZLRyWIiOo1cq48cxgddm+f3qHfdiaOyE56I+xNVZl50O3VrBkqfnb1uITOtOmjcynLT91dO6LEnJbwZ3GOHoX4oBerUcWls8v7Xb++xZW0lquiKCPp97dkqLlwRmnC67257PShR02nbTAxG6ICwn0hFsGM0XmD5vRwuMik4gTpFZEIZBIxj7+xkZ898hW/fWoFdz24hPv/t5qWPvxbT3KSbgYUvFVVVfHiiy8mTPEIgsALL7zAyJEjB2VwJznJtx2RSMT0MXlMqY5aPm3a086KzU08cNs0Lp5VxsQR2Vx//nB+df0pfLxyP3Mml5wwCuBHA4tRxdVnV/a5jkIm4YYLqwY9+JBKxIwZlskjP5nFD+dWc9VZw/jL7dO56eJqMnSp5VUsRhU3XtT3NVEqEXHrJdV4nF1A9LyYOTafyVWJtlPRbSq56qxhvLd0X8KyDJ2CH18xJqnUh0QcFX4eWmBMOZZzpxZTmnt0pF6OBuMrsphandwyTSSCn8wflzQLeawQi8Vk6BRx9lfdgdu2mngrua37OnjwlfUnM3An6ZcB5cV//OMfc+WVV3LeeecxZ84crFYrbW1tLFy4kLq6Ov773/8O9jhPcpJvLSa9kh/OreKcycW88fluNu1pZ2+DnWvOqeDUcQUsWXcAm8vP/LOGfWs08QaKSiHlrElFKGQSnl+4HX8wXn4hM0PFz6+ZMGCNt/6QSyUUZusPS3ZCLpPEMj9PvbMFXyB+zGaDkp9ePZ4Cq4qdnYeK1k16JT+8ZBTnTCnh9cW7aWhzoVXJuWBGKSNLLYTCEaaNzmPx6gOEwlG3hwtmlHHquPwEh4meWI0qfnHtRL7e0shrn+2ONUYUZumYf3YFI0rN6E+gqfkMvZKb51Yxa1w+ryzayb5GOxKxiMlVucydNYT8LC3yfrxFjzZymYRssxpfIESX04/XF0oI3LrZuq+DdpsXveboHoNhw4bx+9//noULF7J27VoMBgNXX301N998MxBNxjz99NO88cYbNDQ0oFAoGD9+PPfeey8FBQVpbeMkR4+0HRZ6s2LFCv7+97+zZcuWWNHsyJEjufvuu5k8efJgj/Mb5aTDwreL43k/enxB/IEwUokYAYFgKIJcJjluFe+/qX0ZCodp6/Kxp97GvgY7cpmYUeVWsk0aTIaj60IxUELhCO02L/sa7ew5YEMqETOyzEyuRYvFqOpzX3p8QXyBMFKJKO6mHgiG6XL6CYUiyGViTAbVYU0Pttu8uL1BRKLo1LyxjwziiYDD7cftCeByu8iyGNBrB/ecPBKHhW6CoTDbajq598nlKdf54w+nUD3EOtBhpsWwYcPQ6XT86le/orq6mnfeeYcnnniCF198kfHjx/O///2Pxx57jPvuu4+RI0fS1NTEr371K8rLy3nsscfS2sZ3kePOYaE3kyZN4rXXXsPr9eJwONDr9ahU3+2swOESCIZps3lpao8aKVuMSjIz1N/57MrxgMcXpN3mpcvpRySK1tZYjaq0bWwcbj8ddh9efwi5VILZqOxzaq0btVKGWimj3ealw+7FH4xmVCIRIaUkxHcRqURCjkVDjkXD9NF5R7y9cDhCm81Lp8OHIAjoNQqsGSqUg6h6L5WIyTZryDZrmFJ1aJrP6QlQ2+TA4w2gMubhCwr0vrx3nxe9kcskZJkGdjMIhsIEQmG8/hAiESgV0j67V08E9BoFUlGYxv2N5GXGT/2GvU5Cjk6EgBeRTIFEb0aqPrbCveFwhEhE6LdpRHuMmisuvvhiLrzwQgDuvPNOXnrpJdauXcv48eMpLCzk/vvvZ8qUKSiVSgoLC5kzZw4ffPBB2ts4URAEASEcgsjBrLhYgkgiPa5/CwO+MoVCId566y2WL1+Ow+HAZDIxadIkzj///JTNDCc5RKfdxxuf7+bD5TWEehjlWo0qfnz5GCpLTYPmx3eSw6OxzcV/3tnC2h0tMbkFqUTMmacUMu+0oX0WwofCEWoa7Dz51iZ27bfFXs/P1HLzxdVUFGf0GRC4vQHW72rjP29vobOHTllRto7b5o2mPN+A9OR5MWC8/hA2px8BAaNWgVopo9PhY9HKOt78fE9Ma0ssFjGlOodr5gwnx3J06qVC4Qg1jXb+/fYWdtQemkIbVpTBTRdVUZprOGpK8G1dHt78Yi+LVtTGBIS7m2KmjcrDqFPg8QVjrg5GnQKVou/bRTJj+u5soz8YRq2IeqcezRuiwx3A7gyiM+XRraUshEP4W2rp/Ox5wsY8whkFSBzNiNv3YphxJX5dAQaj9qjWkgqCgD8Ypq3Liy8QRi4VM7zUzLZ9HQnrjigd/GabVJSVlcX9rdVqCQajenizZ89m3bp1PPHEEzQ2NrJv3z52795NVlZW2ts4ERDCIcIeO2G3HYSDJ41IjERtQKIxIJIcn13XAxrVgQMHuP7666mvryc/Px+z2UxtbS3vvfcezz33HP/73//IyMgY7LF+a3C4/Dz17ma+2tCYsKzN5uXX/17OH2+dxshS8zcwuu82zR1u7n1yeYKLQSgc4cPltRxodfHT+WMx6ZNfXPfW2/nFY0sTVPjrW138+t/L+b/vTzzYMZl4AwuFIizf3MTDr25IWFbX7OSex5bywG3TqCw+fP/Uk0B9q5P/vb+NVduiorljhmZyw4UjePWTXSxZ3xC3biQisHRDI7v32/jjLVPIOgoF73vqbdzz2LKEc2VnXRc/f/Qr7v/hNCpLBv9Yt9k8/PF/q9hbH+8K4XAH+Ndbm+mwe5k1toDnPtzG6u1RUeSJw7O59rzhFCTxe41u08uHy/bxwbJavP4QVWVmvn/eCLbt6+CFj3bgD4YxaOV8b04lk6tyBr2eKxSOsLfexpNvbmJPfbTmbfroPOafXYHR30jL4pdxjbqKV5a1sm+1i4LMMq6cOpnQincRV8zii40ZnDqu4KhNuQeCYepbXbGHwWA4wq1zq3n8zU1xAdyIUjN3XTnmqNe7dZMs0dJdSfWf//yHRx55hAsuuIAJEyZwzTXX8NlnnyVk3vraxvGOEA4RtLch+Ht1+AoRwu4uIkE/MmPmcRnADehR409/+hORSIS33nqLTz75hFdeeYXPPvuMBQsW0NXVxQMPPDDY4/xW0dLpSRq4dRMR4Ik3NmJPod90kqNDJCLw1YaGPu2nNu9pp67JmXSZw+Xn8Tc2JNyMuxEEeHjBhpTbb7N5+debm5Mug6jNz+Ovb0xLef8k8TR3uLnn8WWs3NqMcFBPd9OeNvbWOxICt560dHpYtqlx0G9GdpefRxakPldCYYGHF6w/Ksd6y56OhMCtJ/mZOv7vyWWs2tYS21crtzbzy8eW0ZzUtcHPw6+u5/XFhzKXm/d28IvHlmLNUBMMRaei7K4Aj762kRWbmwZ9f9a3OPnFY0vZc/B7hSMCX6yr576nVtDe1oWj6jJ++dwONuzpxOEOsLWmi3tf2EnrkPMRr3sTvcjDy5/swBcYfJeD7in5nl9ZEA4FcH+9fTq/vXES/7xrJr/8/gSyTN9cZ2xPnnjiCW677TbuueceLrvsMkaPHk1tbe0JE5ilgxAKJAZuPZcHPESCx+f1dkDB2/Lly/nJT35CRUVF3OvV1dXcfffdLF68eFAG920kEhH4eGVdv+vtb3YOujfjSfqm0+HjrTR8H9/4fE9S38c2m5d9DY4+3+twB2hJYedT02RP6J7sTW2Tg44knpUn6ZttNR0JgdCYoZms3NbU73vf/GJPSsuygdJu87K/OflDQDf1rS7abP1bPx0ONpefBZ/tSrk826ym3eaNTZf2fu/mJNN8rTYvG3a1JbweDEVYubWJMcPiRYqfW7h9UK9twVCY977aF1d+0k19qwu3OpfnlrSQzL3smc8aCZVOpiwjzKIVdbR1Df5vKxQWkprSC0LU01QiEaPXKMgya45Zxi0dcnJyWLZsGfv27aOmpoYHH3yQRYsWfWt0XIVIhJDL1u96YbcNIdL3dfmbYEDBm1qtRiZLXlBpMpmQSE7W5KQiGArT0OpKa12P76TX3bEkEArj9PRfq9Hc4cYfSPwxp3u8khmGAzS1pSfOefK8OHw272lPeM1kUNLa2f/N2u4KDLppeLo+m+5B9hcNBsN9eoHmWrTsa0ydldu0OzFIa+8j4NnbYCfPGi9bYncF8Ayij6fHF2JbbXLZDYCAII1l5HpT3+oirMtGHHQjEolwegY/MImkmakKhY6vAOEvf/kLPp+Pq6++mmuuuYZdu3Zx33330dHRQX19/Tc9vCNHiCCE+/99CaHgkXvNHQUGNJE7f/58HnzwQaqqqsjOPiQi6XK5+Ne//sUVV1wxaAP8tvH/27vv8Kiq9IHj3+klmWTSeyC00DuhgwIiuiq21UVEVxF1Vextf7J2sZcFxYZdLGtX7GADpBNAKaElkN4nyfT6+yMmZJhJZSbJwPk8j88u996599w7JzPvnPbKZVIiw9s2oUOpOHkXY+0KcqkUpVzaOIC7ORFhShR+3pvmUif5e70/URFt+9Ut6kX7+Vv3zWRxtCkFlUopQyYL7CB7VRtnLasCONsVQCaTEBGm9JoM01R1nZXhSc0vUZGR7PscW5oFnRClpfqYFk+VUtbm+28LlVJGcmw4Bc38KJbjIjEmzO/+KJ0Kqa0Wj0aPy+1B28qkjI5o6/yM5tKrBUtOTo7Ptqa9ZoMGDeL999/HarWiVqsbG2Wafr+3do5uTSJBIpW1GsBJpLK2v4mdqM215bLLLmv8b8OGDRw+fJiZM2cyb948br31VubPn8/UqVPZvXs3+fn5wSxzSJPJpJw5MaPV42Ii1WLJkE4WHanitLE9Wj3u3Km9/a7eHxOpbjXPokopIznO/wKq/dKjaG2Jrli9qBcdMXpAAopjZm5uyylj7CD/GQyamjWuJ9ERgR3IHhOpIaaVwfFROlVjrtNAidKpW0yPlVtUS9+0KL8zLxVyKWMG+j6v+GhNszNyp45MZeOf3l3TZ0/qFdDZlGqlnAum9fG7T6dVEO6o5JLJ/t/nOVOSUez7iVKbhqF9YoMyy1Muq/9R2BIJbQ/ohcCQSGXIwvStHicL09cHcN1Mm4M3j8fj9d/IkSMZNmwYAOXl5djtdgYOHMjgwYOpqPDtohCOSo0LJyOl5fQzV549qNOmiwv15DIZf5uYgbqFFrSEaC39e/ifARgTqeGa84a2eI1LZw0grpn3NSZSw+ypvf3ua7Bg9hBRLzogKSaMB6+Z4BUMqRQy+qTq6dNCXlKNSs6s8T0CvoxErL71unL1uUOI0wd2kVmJpD7zgF7XfGtZ9t4y7l8wziu4jNWrefDq8ST7CdJiIjXcf9U4BvY8usJARJiSOy4dxcFCQ2NLtkQC08ekcdakwKd465kUwS1zRqJVH205S4oN4+FrJxCjdpFeuZ4bz+mF/q9WQp1WwYJZPRno3IW7zxQ2Hbaz8KLhhAdhMWy5TNpqPmJ9hOqkTnvXVSQKFRJZC++5TIFE2T0Xru5whoWTSTAyLJRUmnjy3S1ea4FBfe7Bq2YP5pSRqT4fJN05M0Aoaek5ut0eDhQYeOSNTT5dS71SIrhz3hifMTxe57Y62LKnlGUf78DUZGyaUi7l0jMGMG10WovdTNW1Vj7/7SCf/3rQK+G4Winj2vOHMm5wUqsLfHamUKuTlTWWxskHUREq4vRaSqvMvPDxDrJzyryOTYjWcvflY+idEhmUtclMFgdb95by4ic7MTYZ2xamUXDNeUMYOyjR78K8gZBfWsfTK7ZysPDoWDCJBCYMTWb+2YOIi6qfuFBVa0VC/fjA1lp860x2KmutOJwuIsNVxOk11JnrF7s2WR2N21pbL66jGhdarrHgctpJjI0gLjocl82M5cA2arZ8g6PfdByKCBQuM/KcVWgHTMKVNhKVLrJNXegdzbDgdnswWx2UVlu8/q4lkvoW1sjw7hm8uVwun27TE43b6cBZU4bH7j12U6LQINfHIZW3L6DvrAwLAQ/ezGYzW7ZsYcqUKYE8bZcKVnosQ52V0ioz6/8oxmhx0DdVz+De9U33/sZPhdoXZXfVludYYbBQVGFk3+Fq5HIpAzNiiI/StCl9kMvtoaLaTGGFiQqDBX24irSE+hRIijYssGu1OamosXAg34DBaCMxJoyeSRHE6jXd7gP+RKmTRrOdCoOFA4U1OJ1ueiTqSIjWEh3kLmq320O5wUJplQlDnRWNUkpKvI7EGB3SdqS56ogao41yg4WSChMymYSUeB2xkeqgBYydxV+d9LhdOGsrcVSX4DJWI9NGoIhORh4R0641vI4nPZbH48HpcuNw1v8nk0pQKmTIZdKgv9cddTIEb1BfPzwuZ/3kBDxI5Mr6DAsd6C7t1umxCgsLuffee9m8eXOzKynv2bPnuAp2MtDr1Oh1ajKb6YYTuk6sXkOsXtOh/IIyqYSEmLAOL+yqVslJjdeR2syCqELghWuVhGuV9ExueThDoEmlEhKitSREaxs/4PVhUZ3yZR4ZXt/i0ydVH/RrdTWJVIZCH49CH9/6wcEqg0SCQi5r0w84oXNJpLL6QE3RfZZqaU2HgrdHH32U7OxsLrroIrZt24ZGo2H48OGsW7eOffv2sXTp0kCXUziBVddacbrqf4lGRai7dT454Si7w0Wt2UVkjP98nMcy1FmpMdpxezyEqRXE6jVtDlLMVgeGOhsutweVUkZsZNtf293ZHC6qaqxYbd5pndrC6XJRVWvD5fKgkEuJiQz830+d2V6f/5T6gK+t+X2r66yUVppZs70Qg9FGSlw444ckERupQdeGLkoh+NxuDy53w5hESbdr2Rea16HgbfPmzdx8881cdtllrFixglWrVnHHHXdw6623cuWVV7J69WqmT58e6LIKJ5iyKjPb95fz8er9lBvMRISpOGdKLyYMSSIptvlxZScLu8OFzeFCo5J3qw9Vu9NFQamRL347yNrthThdbjJSIpkzsz+Z6XqfruVak41dhyp5+5s9jcs1hGkUnDulF9PGpBMf1XzYZ7O7yCuu4b3vc8jeV4bHU5+0e/bU3pw6MjUoaas6i9vtoaCsjk9+PsBv2QU4XZ76tE4jUrhwWl9S43XImglQPR4PBWVGvtuQx/frD2NzuIgIU3L+qX2YODSZxAA8F0OdlX1HDHzwYw4HC2tQyKVMGZHC2ZN6kRYf3mKO3fzSOha/uclneY73f8hhzIAErr1gaIvveyB5PB7sDhdyuazZ59mZPB4PbrcHiVSCtAOBttvtwe3xIO3g6xvKYHe4qK6zYbQ48HjqZxNHR6jRquRBy6krBE6HgjeTycSAAQOA+qS0DS1tMpmMuXPn8thjjwWuhMIJqbTKzEOvbeBwk1Xmq2qtvLlyN5//epDF/5pIWsLJ2W1osTk4XFzH578eoKjCxKCMGGZN6ElqXHinrwV1LKfTzY595Tz8xiavgdcHC2p4+PWNTB2RwvzZg4n6K4CrM9v57JeDfPzTfq/zmCwOVnyfw9odRdw7fxzx0b5f5A6niy17Snn8nc1ea2QaLQ5WfLeX1ZuP8NA1E1oNVGx2F0UVRn7fWUx1nZUJQ5PplRJBhFbVpa13BwoM/N+L67wWfHa5PfyytYD1O4t5+F8Tmp3ZfKDAwD0v/t6Yjgrqs3e8uXI336zL5eFrJza7fEdbVNdZef3LP/ll29HUYTa7ix83HuGnzfksunIsw/vF+f1RUVxh5P9eXNdsaq/Ne0oxvruFuy4bE9RlbyQSCYXldazdUcz2fWWkxoVzxoSepMTp2rwmYyA1BExGiwOrzYVcLiUiTIlSLm3T37XL7cbhcFNjsuNwuFAqZX+9XtaueuzxeDBbnRRXmGg64N3hdFNaZUajkpMQrfVZWudE4XC6sNpdGM2O+vdAq0DRzmfYHXTo3YmPj6e8vH6l7R49elBTU0NZWf1MrcjISCorfVOoCEIDm93Je9/v9QrcmjLU2Xj6va3UmrpnTrlgcrrcbPijhDuWrmHdzmJyi2pZuS6XW5791WtmYFcpqTKx+E3vwK2pX7MLvbIZFJYZfQK3pg6X1PHt+lxcfvoKy6rMPPnulmYXNy+pNLPiu73YWshH6XS62bynhJue+YUPfsyhus5GjdHGCx/t5IWPd3C4uLbZewmmyhoLj7292W+mDqjvSn30zc1+00hVGCwsfnOTV+DWVFm1hde/+rPZ/W3xx4EKr8CtKZfbw+I3N1Fa5ZupweX2sHpzfqs5WffkVZNb1HIqueMhkUgoqbZx23O/8e63e/jzYCXfbTjMLc/++lcLbue/5zaHi8JyE1W1Nsw2J7UmO4VlRsw2Z5vKY7E5KSgzUmuyY7G7qDHaKSgztlj//XE43RRXegdux17HaLGfUDlMGzicLoorTJRUmjFa6odiHCk1YrI6Qu5+OxS8TZ06lf/+979s27aNpKQkEhMTef311zEajXzyySckJCQEupzCCaTCYOHXbS2nVzlYUBOUPIPdXYXBwouf7vTZ7nC6efGTHdSZujav4I795X5zSDb13vc5GOps2OxOvvjtYKvn/HpdHuV+gpTsfeW4Wgms1mwv9PvaBuU1Fp77IBuPp37NtknDknnmvW1s+LOYHzYe5rYlv1FU0bZ0dYFUUmlutX5X1Vr9lq2k0kSFoeVcq5t2lXT476e6zsqK7/a2eIzD6ebPg77reVYYLHy19lCbrvPhjzlBSUcFEKbT8863e72W6wFwe2DJh9md/tnidLmpqrH61GcPUF5taTWri93horza4hNweTz1OZUd7UitZbU7eerR+zj/zCkUFhzx2W+oruK0aZO55ZZb23zOUODxeDBaHNgc/n8oOtsz2LQb6FDwduONNxIREcGSJUsAuOWWW3j77bcZM2YMX331FVdccUVACymcWAxGe6tfygDFLeRgPFFV1libbTE5UFCDoQtbI50uN5t2lbZ6XGG5EZPFgdHiYJefRObHsticPjllXS43m3e3fi2X2+OTfqmpSoOlsXVreL841u4o8tpvs7vYk1fd6nUCLbeF/KFNHfSTk3PfkdbL6/aAwdhygNccs8VBUUXreXY37S7x+Ts2Wx1tzr2bW1SLNYA5TpuSKrRs2eO//tSZHVTVdezZdJTL7fF5LhIJ6JUO4jzlOAt2YyvJxWXx3xvhcrub/dFkc7jb9HnawGRxcs31txEWHs7zzz7q0+L00tKnUChU3HPPojafMxS43Z5mf/y6PfWt9KGkQ2PeoqKi+Oijjxq7Ss855xySk5PZvn07Q4cOJSsrK6CFFE4sbc71d2IOuWhRSwOqJRKQdeFMXAktl8/rWGl911Vbc4L6HNWO17ZUJJn0aCWyO1yE+1ngWKPq/PFPTcvV8nG+N9fWySsS36fatte1sY7JpFKfZ9+eK0qlkva9oF08yGVSHM18IXfJxAUJNDSdSSSQoDRT+90y6vKPLqulShtI/Dk3oNAf23vVcnnbczcSCYTrIrjh5rt56N47+OHbLzn9zNkA/L72F9b+tpoHHv0vEZGdu2xO0ElaqdshtspBh74en3/+eUpLS4mPP7pmzujRo7nqqqtITk7mwQcfDFgBhRNPdIS61VXWJRJazGRwooqJVDebumj0gIQW0xoFm0wmZdrotFaPG9grhgitkogwJROGJrd6fJROhV7nvXSETCrh1FGtX0ujkrc46D1Wr25MibRjfzkThiZ75ZmM02vom6Zv9TqB1je9bdcckOE7YWFAz9bXhVTKpcR0MC9qRJjS73WPNW10ms+XYbhW2WrO1gbD+sb5zREcCC6bkenN1NWEaC0xAc5V2xqFTIKuScacSIWD2u+WYcv3Xg/Vlr+bsi+f92mBk8skKBX+v661anm7JjI1lGPshClMnXY6r7+yFEN1FWaTkZeWPsnfzrmAceMnkJeXyzXXXMPEiROZMmUKt912W+NYd4C8vDzmz5/PqFGjGDFiBPPnz/ebqL67kEmljZ8FvvskyNv4Y7G76FDw9sILL1Ba6r9JeseOHXz00UfHVSjhxBan13BeKzk8xw1OOilzeMZFabn7sjE+iayjI9RcefagLl/9vm+avtU0QnNPzyRcq0Quk3LamPQWW8YALj6tH7F+cnj2TdO3mgrs7Mm9ms0VC/XPc9GVWYSp5ThdHlZ8v5eb54zkyrMHcfvcUSy+biIJ0Z2/3Ei8XkO/VgK4XikRfpfTiI/W0reFfKwAp4/v2eJzaUm4Vsnc0/u3eExkuJIMPwsax+o1XDyjX5uuc94pvYOWKstkrOW8U3qTGu/9A1CrlnPXZaODnjnjWFKplCidqjEAUztqfAK3Brb83ThrvMcTKuQyEqK0Pi2GcpmE2Eh1u5YSUiqkja3a195wG0qlkjdefZ5333oFtVrDFVffiNtWy2XzLiUtLY133nmHF198EaPRyD/+8Q/M5vrhLLfeeivx8fF88sknfPTRR0ilUm644YY2l6MraNRydFrvzxSppD4Pbqgtntzmv5x//OMf7NixA6gf+HfxxRc3e2yg00gJJxaZTMrp43typKyOtduLfPZn9tBz1ezBXR6odJX+PaP5722nsvNAOfmlRgZlRNM3Xd8lQcaxEmLCeODq8fznpd+98nE2WDB7ML1T9I3/To4N57a5o3hqxVa/s0YnD09mwhD/rXOJMWE8ePV4Fr30u99xgKP6x3PmhJ6ttjr0S4/iv7eeQlGFCZPFQVqCjlH949Cqu26h2IhwFbddMop7X1nvd9ZmnF7DnZeO8ZsHNzJcxe2Xjub+5esprvB97YjMeM4/pc9xfRn1SdWzYPZgXv3iT9+yhyl58OoJJPhZ3gVg9MAE+m7Ssz/f0Oz5Tx/XwyewCjR9mIyHrplAfmkd+48YSIoLo09qZEDWwOsIpUJGUkwYdqcLilsez+u2+k5UUavkpMSHY7M7sTncqBUyVEpZmxdNbqCQy0iJDaOw3IQuIpLrb7qLR+6/C5lczmNPv0hSXCTvvPES8fHxLFq0qDE91nPPPce4ceP47rvvOP/88zly5AgTJ04kNTUVuVzO4sWLOXToEG63G2kbhwV0NrlMSpxeg16nwmZ3IZNJUMplIbksSptzmx48eJBvv/0Wj8fDCy+8wIUXXkhiYqLXMVKplIiICGbOnOnVpRrqgpXbtL1OlDySDWqMNoorTHy9LpeiChPRESr+NqkX6fE6otvY9dIRJ9pz7ArFFSZ25Vby/YY8HE43gzKimZHVg8QYLRqVd9Btt7soKDfy9bpc1u4oxOl00ydNz8Uz+tErJbLFfLEej4fCciMb/yzh2/V5WO1O0hJ0/H1aX3omRxLdyd1fgVZaZeKPAxV88vMBqmutREWoOXdqb4b3i2s1WC+rNrP/iIHPfj1AnclOYkwY55/Sh7REXUCei8XmoKTSzE9b8tl1qBKFXMqscT3pnxFNUisBUHm1mfd+yOHnLfleg+k1KjkXTe/L9Kz0xrUAAy2Yf9/Hk9u0KVtJLoWv3d7s/pT5T6FKzOjw+dvC7nBhtTupMdq5/urLSE5O4bnnnkMhl/Kvf13L2rVrUSqP/sCRSCRYLBbmz5/PHXfcwYoVK1i8eDEajYZx48YxefJkzjjjDCIiIoJa7u6uWyemf/755/n73//eaUuC2Gw2HnvsMb777jusViuTJ0/mvvvuIyYmpsUy+kvTtWvXLuTy9jXVi+AtuJwuNzaHC6Vc2ilN1yfqc+wKNXVmSkrLSEtOQKttuSvK7nBRa6pfP0qjkhOubXvLl8fjwVBnw+3xoFbKW+1ODTWGOhtWu51aQzVpyXFoNG3v1jNZHTgcbtRKGeogdEO6XG6sdhdSqaRd3Zw2u5OKGiuFZUbqLHaidWqSY8OI1WuCuth0KARvLksdJR89gS1/t88+VdpAEv9+JzJN5yxS7nK7ufyyy0hJSeHxxx8H4KqrrsLtdrNo0SJsNhsqlaoxMb1OpyM6un5MZGVlJb/++ivr16/nt99+Q6FQ8PnnnxMbG9spZe+OunVi+oZ+7YMHD7Ju3TrKysqYN28e+fn59O/fn/DwwDaH33///WzdupWlS5eiVCq57777uOmmm3j33XebfU1OTg6zZ8/mjjvu8Nre3sBNCD65TNqt0j8JbaeQgbGmEpJbb2lXKmQdHscokdTnvbXZnbg99UsvdIdUR4Gi16kwm10U5pXjSTr6xedye7DZncikElRK/59dYWoFBLEBUiaTEqZp/9+nSiknJS78pJx41BqZRkf8OTdQ9uXzXgFc/WzThZ0WuEH9QH6JROI1+aRv37588803JCUl4Xa7UavV1NXVcdddd3HFFVfQp08fli1bxtVXX83555/P+eefT2lpKVOmTGHTpk2ceeaZnVb+k1WHIhm32829997LJ598gsfjQSKRcMYZZ/DCCy+Qn5/Pu+++69Ol2lGlpaV8/vnnvPzyy4wePRqAZ555hlmzZrF9+3aGDx/u93X79u1jzpw5xMXFBaQcgiB0DYfTRVm1hZ37y/ktuxCn203ftChmjEknLkrjNYvvRFFrtFFabeaHDYdxut0M6xtPbKQanVZBXJT2pB0PeiJR6BNI/PudOGsqcFuNSNXhyCNjOzVwa84ll1zChx9+yO233878+fNRq9U8/fTT7N69m759+xIZGckvv/zCkSNHuO222wgPD+fjjz9GoVAwePDgri7+SaFDwduyZcv46quvePjhhznllFOYOHEiAHfddRfXXXcdzz77bGPz6/HaunUrAGPHjm3clpGRQUJCAps3b/YbvFksFo4cOUKfPn0CUgb4Kx+cuWsXjbVYLF7/K3TMyfIcHa76BTxVcimKIDU4B/tZOpwedhys5tn3t3ktUro3r5qv1hzilJGpzDsjk3B1aLfc2pweLFYHOn0MNSYnr36xg925VVxz3lAOFhp46dOdmCwOpFIJE4Yk8ffpfUnQK0NtaapOEcw6abPZcLvduFwuXK62ZzVollKLPC7da1NAzttOHo8Hj8fTeO3k5GTeeustnnnmGa688krkcjnDhg3jjTfeQK/XA/DSSy/x5JNPcvnll2O1Wunfvz8vvvgiKSkpXXIP3YXL5cLtdmOxWHC769cZbFonu7Tb9JNPPuHGG2/kggsu8HqT+vfvz4033shTTz0VkMJBfctbVFQUKpX3rKv4+HiKi4v9vmb//v243W6+++47HnzwQex2O1lZWdx+++0dnkjhcDjYs8f/1O7OlpeX19VFOCGcqM9RExaBnXA++fUQ+aVG0hN1XHBKLxTuWiwm/yu4H69gPEuFQoFFEs2T725t9phfthUAbv6WFYuptirgZQg2TVgkZreG/60+SH6pkdT4cM6eXD+m77oLh/HOt3soaZJpxO32sHZHERv+LOHeK0cjd1Rgt3dtyrTuKlh/33K5HJvtxMq7/PLLLwP147Ua9OrVi+eff97n2IZjUlLqJzg0t/9kZbPZcDqdHDrkmyYuLy+vxbH67dGh4K2iooIBAwb43ZeQkEBtbdsTDhcUFDB9+vRm9990001eM14aqFSqZv+A9u+vT4St0+lYsmQJFRUVPPPMM1x22WV89tln7RoM3EChUAS0Ja8jLBYLeXl59OzZs0P3INQ7kZ+jRCIhe38Vj761qXFbZY2V7fvKueefWQwdmBbQBMzBfJYmq5tFL29o9bhfthVx3tS+pKeEXk7lzXsreWrFlsZ/7z1czd7DW7lr3mg27y71CtyacrrcPPP+Dp5YOJFIbWitTxVswayTNpuNoqIiVCrVcU1YCBUej6dxwkJbM28I9QF+enp6Y6NT0zoZsGt05EU9evTg119/ZcKECT77Nm3aRI8ePdp8roSEBL755ptm9//6669+f1nabLZm/zAvuOACZsyYQWST9B59+/Zl6tSp/Pzzzx0aTCmRSLrNzESNRtNtyhLKTsTnWGGw8NKnf/hs93jgpc928uTCKX4XxD1ewXiWBRXVbcqxCbBmRxGX/21gQK8fbGXVZl753HcdNQC708WaHYUtvt5gtFFWZSUpVozr9ScYdVIqlSKVSpHJZI2zL09kDT1r9anuTvz7DQSZTIZUKkWj0fgE+IH8MdGh4O3yyy/n3nvvxeFwcOqppyKRSDh8+DAbN27k9ddf5+67727zuRQKBb17N7/afk5ODgaDAbvd7tUCV1ZW1uKkiMhj8rIlJCSg1+spKSlpc9kEIdTUmGzNJmqvMFipM9tDJnNFbTNJpP05UGDA6XQjD6HFNmuN9mbv0enyNJuXs6m84hqG9RPBmyCcbDoUvP3973+nqqqKl156iffeew+oT5WhUCi46qqrmDNnTsAKOGrUKNxuN1u3bmX8+PEAHDp0iNLS0sbZp8d6+umnWb16NV9//XVjU29BQQHV1dVd3vUpCMGkaGHJFYmk7UnNu4P25BpUKqRIQmzpkJZWdW9rD1Uw1nUTBKH76/Bf/jXXXMPcuXPJzs7GYDAQERHB8OHDfVq8jldCQgJ/+9vfWLRoUeNqzvfddx9ZWVmNM03tdjs1NTVERkaiVCqZNWsWb775Jg899BDz5s2joqKCxYsXM3LkSCZPnhzQ8glCdxITqWFEvziy95X77BvVP6HNScO7g1i9BqlUgtvd+hi9U0elhdy6bzGRagb1imHXoUqffW63h4Rord/UWU1l9mg9gbwgCCeeDgVvNTU1LFmyhG3btvmdnCCRSFi1atVxF67BQw89xOLFixsXB54yZQqLFi1q3J+dnc1ll13G22+/zdixYxk0aBDLly/nueee4/zzz0epVDJ9+nTuuusuMehSOKGFaRRce8FQHnh1g9d4sdT4cK4+N7TyxcbpNUwalsxv2S2P/dKo5PRJ1XdOoQIoXKvkhr8P475X1lNWfXRZi1i9mli9hnOn9Oblz33HLzYYNziRWH3oBOOCIAROh4K3//znP6xevZrJkyfTv3//QJfJh1ar5eGHH+bhhx/2u3/s2LHk5OT4bHv//feDXjZB6G6SY8NZfN1EiitMlFWbSYgOIzFGS0xkaIx1a6BSyrl01gD25FVRXu1/zS6pVMLdl48hLio0J56kxut47PpJFJQbKSwzEqdXERmu4eE3NjJ+SDIXnNqHT385wLEThIf1jeXqc4cQrjnxFigWBKF1HQrefv/9d+68804uv/zyQJdHEIQAiInUhFyw5k9SbBiP/Gsi//sxh1+2FXgt1JuZrueqc4fQOyUy5LpMm4qL0hIXpaV/mo69e/eSFJPJXfPG8OoXfxAfpeX2uaPILaqlqMJIZLiKU0emkhwXTmS4qvWTC4JwQupQ8BYWFkZGRkagyyIIguAjKSaMf10wjAun96OqxorH40GnVRJ7gqXG8ng8uN1uFDIY0ieWh66ZQIXBgtHiYMLQJGIi1ESEKZHLxZINgnCy61DwNnfuXN544w1GjRpFWFhYoMskCILgRamQnXRJziPDVaJ1TRAEvzq0bsCll15KSUkJU6dO5aKLLuKyyy7z+k90pwqCIAhC97Vr1y7OOussBg8ezE033cS8efPa9LqlS5cybdq0IJeucxx7L5mZmXz66adA/SoWt912G8OHD2f06NFUVFTwyCOPMGrUKEaMGMGff/pfYLuzdKjl7d577yU3N5eMjAxUKpVPup1Apt8RBEEQBCGwli1bhkQiYeXKlSgUCtGLBqxduxadTgfAb7/9xsqVK1m2bBmZmZlUVFTw9ttvc++993LKKad0OE96oHQoePvpp5+49dZbufrqqwNdHkEQBEEQgqy2tpaBAwcGNN9mqIuLO5qtpK6uDoBp06YhkUjYtKk+X/SUKVNISUnpkvI11aHgTalUMmTIkECXRegCVpsTk9UBgFatQCNWbBcEPB4PBqMNl8uDXCZBr+ue66m5XC4MBiMutweFXEpUVERXFylkGOpsOF1uZFIJet3JlXh92rRpFBbWr5/4+eefk5KSQkpKCu+88w4Ar732Gu+//z4lJSXExcVxwQUXcP3113s9o1dffZV3332X6upqhg0bxkMPPdQYCNbV1fHEE0/w448/4nA4GDRoEHfccUdj3LB06VLWrVtHcnIyv/zyC7Nnz2bRokU888wzrFy5ksrKSlJTU7n88suZM2cOBQUFTJ8+3e+9PPbYY5x33nltuu8PP/yQ5cuXU1payqRJk0hOTvban5mZyaOPPkphYSHPP/88AP379yclJaXxec2YMYOsrKzGZ9VVOvRNfe655/L+++8zduxYpNLQSbcjHGUw2qg0WCivtmC01OdXDNcqideriY7Uotd1zkBpk9VBpcGCxeZEpZARHakhIqzzZxDWmmxU1lixO1xoVApi9ep2LWjrcTlx1lbgqC7BZapFFq5HEZWIPCIGibT9swOraq2N5dGq5cTqW59Z6XK5KTdYKKk0YaizodepSIoJIzZK26GlNAx19WWw2p2oVXJiIzWNA+idLjfF5UYqaqx4NCkUVdqI98iICAvtAfYut4fiCiNb9pSycm0uRrOdyHAV507tzbC+cSR34oQJs9WBoc6GxeakrNpCucGMPlxFRnIk0TollQYT6/8s5ftN+ZgtDqIi1Jw3JYPBGVEkJ0a1en6bw0VFtYUakw2JREK0TtXhumKzOyk3WCitNGN3uoiN1BAfpUUihapa219/V/X1OFg/EO0OF+UGCzV1VmThSdSYXWg0Hp+grKjCyM79FXz+60EMRhvhGgV/m5jBmIEJJMWGh/SyM2318ccfc91115GYmMg999zDY489RllZGVDfs/bSSy/x3HPPkZ6ezubNm7n33ntJS0tj9uzZABQWFrJ161Zefvll7HY7d955J/fccw8rVqzA4/GwYMECFAoFL7/8MuHh4XzxxRfMmTOH//3vfwwcOBCoX1x/yJAhfPHFF7hcLt577z2+++47nn32WRISEvj555+5//776du3LyNGjGDt2rWN5Xe73dx6663U1tZy2mmntemev/76ax588EH+fffdjB+XxapVq3luyVISExPxeLzryZVXXklERASLFy9m7dq1qNVq1q9fz8KFC/noo49IT08P1FvRYR36KwoPD+f3339n2rRpDB061KevXCKRsHjx4oAUUAi8qlorB/Kr2bq3jJ+25GO1uwBQKWVMG5XG6AHx9EmNIjqIqZTcbg+HS2p5/as/2b6vonF7n9RIFpw7hD6pepSK4C+JYLe72F9g4NXP/+BgYQ1Qn1dyRL94rjh7EOkJOqStfJg7TTUY//iF6t8+xOM4mhReqtISPf0ywvqPR6Zp25e+2epg16FKXvn8D0oqj6ZGykzXc835Q+mVHInMT37SWpONNdsLeevrPVhszsbtGpWcf541kEnDUtocFNvsLvblV/PSJzs5UlrXuL1nUgT/umAoMRFqvl2fR3m1hVNGpaKQS9maU8bugxX88+zBZCQHNkVeZ3E4nOzKreLh1zdhc7gat5usTpZ9spMwtZz7F4wns0dUUFtpJBIJBWV1bPizGIlEwoc/7vN6Twf2jObMiT157oNsr3XvTFYjSz/+g4gwJQ9elUXv9JjGfXaHC7PVgVopR62SU1pl4v0fcvhlawGuv9KPadVyzj+1D6dl9SA6Qo3J4qDGWF+fI3Uqwpr5MVNhsLDi+738tCW/MZXZ4N4xnDUxg427S+iTokellFFjtFNebebsyb1IS9AF9BmWVpr4cNU+ft6a3/hMIsOVXHJ6fyYMSUKvU+PxeDiQb+A/L/+OyXr0eZosDl7/ahfvfruH+xaMY1BGjN+/sRNJdHQ0CoUCtVpNXFwcCsXR9/bIkSOoVCpSU1NJTk7m9NNPJzU1ldTU1MZj5HI5Tz75ZOP4sH/84x88++yzAGzYsIHs7GzWr19PdHR9+rZbb72Vbdu28fbbb/PYY481nufGG29sPMe7776LRqMhOTmFxMQELr30Unr16kVGRgYymcyrS/ORRx4hNzeXjz76iPDwtn22vv3225wxaxYXnTkNj93CP8+dSfbmjeQcPITLXItMfTSOCQsLayxXw3UbUn9GR0ej1+vb9qCDqEPB26effkpERH3zvL8ZFydT83OosTlc7DtSzWe/HGB3bpX3PruLb9fnkVdcy/mn9GFEZhwqZXB+JecV13Ln82uw2V1e2w8U1PDvF9Zy/4LxjMgM/oDQnQcrePC1DV4r2Hs8sC2nDKkE/nn2IMLVCqIj1X7rtctmoXbLtxjWfuSzz20zU/HNS3jcLnTDpiOVt9yS53S52bS7lKdXbPXZl3PEwF3Pr+Xx6yfRN927VcVmd7FqUz5vrNzl8zqLzcmLn+zEbndx5sSMNgXEfx6s4IFjngnUv2f/fmEtt186mqyBiezJq+KlT3disjoZ1T+euWcM5PWvdnHV7MH0SAyt7ruyajMlFSYefG0jDqfb7zEmq5P7Xl3Pw9dOIDJMhUQCep0KxV/rrrltFlxWI1KFCpm24/dfUevgpU//YGT/eN5cudtrn1wm4ewpvXjq3a2NQdexak127l2+iSevG0dCXCT5ZUY++Xk/e/OqSI3X8ffpfflx42FWb873ep3Z6uTdb/dSVWtl1tievL7y6A+rEf3imO/nfa2qtfDMe1v54+DR/KzJsWFMG51GcYWJ2EgNH/yYQ53ZQUK0lnOn9uarNYc4a3KvgNWR0ioT9766nqJyk9f2GqOdFz/ZSUGZkTkzM6k12ln08u+YmwRuTdmdbu5/dQPP3jI15OpvIJ1zzjl88sknzJw5k8zMTLKyspg1a5ZXF2NsbGxjcAMQERGB1WoF6mexAj7dnHa7HZvt6I/bmJgYdDodLrcbq83FtFnn8d33P3LqqacwcOAgJk2axJlnnkFMTIzXeVasWMEHH3zAW2+91a6xZ/tycjh98jg89qPZWoYNHkDOgUO4asvB5WjzubqDDk9YEEJTVY2FqhqrT+DW1J68KiprrVTVWEkKQjdRndnOi5/u9AncGrg98NwH2Tx90xRi9cHLElBjdvHcB9t8ghSAOTMzMVkc3LFkDSqljPlnDyJrUKJPV6qrrgLDuk9avE7VqrfQZgxHGp3Y4nEVBgsvfLS92f0Op5sXPtnBg1eP9+qerDBYeOfb3c2+DuCtb3aTNSix1W6/CoOF/36Y7feZQP17k1tUw+87ili7s6hx+2/ZhWzfV86988fxxa8H+dcFQxuDmu6uutbK578ewGx1Nhu4NTBbnazZXkRmeiQrvstheGY8F5zahwh3LZU/voH5wFaUcenEnXUdysRe7f4hGxYewee/HGTy8BS/wfi4IUn8uq2w2cCtQa3Jzs5DVWS6pdz2398a76uk0szWvaUsvGg42/eXU2Gw+rx2YEYM/35xHSbL0S+z7H3l3P38Wp6+aYpXHTpcXOcVuAH8bWIGe/76fPlx05HG7aVVZl7+7A8WzB7Ml78dZP45x59r1+PxsG5HkU/g1tRXaw5xxrgebNpd2mzg1sDhdPPNulyumj04ZOpvoEVHR/PFF1+QnZ3NmjVrWLNmDe+88w4LFy5szC8ukzX/bNxuN+Hh4Y1LbjSlVB5t/Ver63t2zBYnJVVm4hJSePXtT/hjx1ayt25i1apVvPzySzz66KONY9p+++03Fi9ezOLFixk5cmSb78njdoMEPB7vv++m77HLZGjz+bqDE7ttWPBRYbDw09b8Vo/7aUs+5TX+80kGogx785oPHqG+a7e0qvkP5OMlkUgoq7JQY7T77Iv7a1zOl2sOYbE5MdTZePq9bRSUGX2OteTuBE/LX/gepx1b6aFWy5RXXNvYhd2cgwU1lBu835ecI1Ve3Wf+OF0ecg5Xt1qG0ioz1XW2Fo/pkRjhFbg1qDXZ2by7BIfTRXFl8N67QCuuNJGeEMGa7MI2Hb968xFqTQ5OHZ3GV2sOsWZ7IdUbvsS8fzN43NjL8ih+/yGcNeXtLotUqWPtzmIAv4HGiH7xbNpd0qZzFZSbWfHdXp+A1OOBj1fv57SsHj6v6ZkUQV5RrVfg1sBocbB5T2njvy02J5/+csC7/FIJYRoFQ3rHsnqL/8+Zz349QHJcOBWG4/98qayx+pTBn6JKE5//erBN5/xx0xEqa3yD2pPFF198wfvvv8+oUaNYuHAhb7/9NhdeeCHffPNNm17fr18/jEYjdrudHj16NP736quvsnr1aq9jHU534/fMl599yLo1PzFi1FiuvHohL772HuPGjWu8bk5ODrfccgtXXXVV49i7tvK4nWT27kX2H94/iP7cu79d5+lORPB2krE53JQ2GUvVnJJKE/ZWAomOMlvb1jwdzA9QmUxGRTPBad90PTv2+37x7vHTWmkvbz0QBnBU+gY7xyoq9w0O/TFbvL/Uj5TUNXOkt6bj15pTVdvyM5dKJS2+f7sOVZLZIxqrNTh1JxgOF9cilUqwt9Lq1qDWZEcmkzZO3vh1WyFWj3cLkttSh8tc0+6y2BxuNCoZ1XXNvw/uVlrdGkRHqNnTzI+kogqT31bt1IRwDhQYmj3ntr1luFz1z8lqd1Jc4R2kh6nlGIw2HC53s+WsMFjRaZWttoK1hd3h8vsD7FhOlxuDseUfJQ0cTjd2R+jU30Cz2Ww8/vjjfP755xQWFrJt2zY2bdrEiBEj2vT6yZMnM2DAAG6++WbWr1/P4cOHefzxx/nkk0/o3bu317FutwfXXz88DdVVvLz0KTb+/htlpcX8vm4de/bsYcSIEZSXl3Pttdcybtw4Lr/8csrLyxv/a1jSo0VuN/MvuZDVa9bzxgefcLigkPc+/ZJVv63zOTRU1qkV60KcZGRSCeFaRasfZGEaRasD9Tuqrd0R4ZrgzTp1u93Nnr+0ysyQ3rFs3VvmtT0twbfLUaZt2+D8toyBamsqJKXC+zdXW1+n17X+PJsblN7A7fagbmEcZEK0lsoaC4qMmGaP6W5i9Zo2f7EDyGVSPJ6jXzp90yJRuCx4hSJSOVJV+xc9VcilWG3OZrsT29MLa7I4idNrqDX5BjfhGoXfYQu1Rjvx0dpmz5kSH9Y4mF8ukxKuUVDaZL/V7iJMrUDRwoB/tVKGy+VGoTj+tgOZTIJcJsXpajnwlkokKOXSNgfo8hN8wkJLLrroImpqali2bBnFxcVEREQwc+ZM7rzzzja9XiaT8frrr/Pkk09yyy23YLFY6N27N0uXLmX8+PFex0ok9f95PHDJZQtwuVy8tPQpqg1VREfF8I9/zOGaa67hiy++oKioiKKiIlatWuV1jvPOO89rEoRfEglTxmfx2KI7ePHNFbzw2jsMG9Sfyy46n29X/XLMoaExZl8EbyeZuCgNk4en8P4POS0eN3VkKvFRzX+IH4+YSDVxURrKq5vvNlEpZCTFBm/Fb7fbTXJcmN8P/oMFNcye3Lu+C6m4FoBJw5LomeQbqGkzszCs+7jV66nTBrR6TL/0KKSS+nFlzYmOUBMT6d1iMrxfXDNHexvWt/XjkmLDWv2SqzHZSYkLp/CYlkKpBE4bm86HP+7l/FP7tKlM3UF6go7t+8oZ2jeWnfsrWj1+/JAk3B4Pa7YXkhij5dxT+hBlC8OW8ztumxmkMuLOvAZ5RGy7yyJ1W+mVEklkuBKpVOLTenWkpI6+aXr25xtaP5fEw9xZ/XnwtY0++86cmMHPfoZP/HmwgjvGj2b15iM+4x4lEpgx5mhXq06r5OzJvXjug+zGbQ6nG4VcypGSWjJ7RPntqp82Jp2cw9WMH5LU6j20JjpCzbQxafyw4XCLx8VGajhlVBo/bGz5OIDh/WKJ7KSlkrpS03XKjg1+FixYwIIF9cGU1WpFrVY3jnNbuHAhCxcu9Dr+/PPP5/zzz2/8d3R0NI8++miz1244h9vjQR+uorrOhlwu54oFN3DFgvpxdXFRGiLDlEgkEp/zt5dEKkeiUDNr2hRmTZvite+mBUfTee75YztSpcbvPY0dO5acnJa/NzvTyfvz4iQV+dc6US2t46YPV9ErOSJoH2AxkRoWzG55kec5MzOJiwruwqhR4XIuPq2f333LPtnB5OEpPPKvCSy57RSuu2C436VTFJFxqHu2fC9hgyYjj2i9JSpWr2HW+IwWj5l/ziCf7q5YvYaxg1qeDDF+SBKxka1P/oiLUnPxaZktHuNwurjj0lH073l01muUTsXtl45mw58l/GPmAHTa0PnyS4gJY0ZWOjOz2rZ20/TRaQzuFcsVZw/isesnkRwbjiq5NylXPU3y5YtJvfpZwgZOQqpof8uxqa6aeWcOYOOfJZwyMtVn/+rNR5g51nes2rGkEpgwOIEBGTHcPGcEkeH1ZdGo5MydlYlcJvEbALo9sHlPCbfPHYVaebSFXKOSc9dlY0g5ZsLLoF4xPp8lP28tQCqVMHtKb/qm6Ru3SyQwYWgSafHhTB2VGpCFjxVyGbMn90LVwizqPqmRREeqOXNCzza1XF40I7PVFmghMKQSCfpwVX2Q9tc2iQSiI1SEaxQBawWTyGTIdTFA8+eTKNRIZKHzvouWt5OMTqskIymCGy4cxhsrd/u0niTHhnHl2YPokRTR6qKwx2NY31hunTOSFz/d6bWGlVwmYc7MTGZkpSNvYUZTIEglMGtcT9xuDx+t3uc16F8ikZAQraVvmh6Nqvk/aJk2grizrqPs8yXYCvb47Nf0GU30tHlIVa23YmpUDcGkh2/X53m1fCjkUq6aPZiRfpZP0WmVXHPeEGx2F9v9jNUbmRnPgnOHEN6G91MukzFzbA/sDhcf/bTfq+VHJq1/b8YNTuSFj3Zw1sQMLjtzIE5n/Sr1X609xKmj0uidEnrrvGUkRxKuUTD39P6s+H5vs8ddd8FQ+vbQE3FMcCqRSFHo41Hoj295G7fbTWqshmmj0yg3WKg12dnSZJJAndnB4ZJazhjfg2/X+29Fkkjg1n8MJzkuHJVawbRRaQztHYvZ6kStlBETqeZgYQ3f/J6H4ZjJKb1TIrh4Ribx0Vr6pUU1jguN1WuI1Wt8uhMTY8J4+NoJ3PfK+sYxqjv2l9MrJZLcohpGZsZz9uReOJxulAope/OqiYpQ079H9HE9p6ZS43U8/K8JPPrmZp8xm0P7xHDjxSOIidQQplGw8O/DWfK/7c2e69IzBoTsOoWhSi6XEqvXoNepcHs8SCUS5HIp0jYGbtnZ2Vx55ZUtHjNjxgyeePxx5PoEnDVlPpPMJEot8sg4JLLQCYkknlAZndeF/vjjD4AuTwlmNpvZs2cPAwYMQKs9vi7N0ioT+aV19V8Gf3UN9kiMIFyrID1RR0J08JMUO11uKgwWjpTWUVZlJkqnpmdyBHF6TVAX6D32OdocLioMFnKLajDU2UiI1pKWoCNOr2nzYp1OowGnoYTa7FW46iqRRyWhGzYNeWQ88rD2rRlltjqoqLGy+1AlVbVWUuPD6ZOqJzZKg7KF8YIGo43SShM/bjpCWZWZ+GgtM7N6EB+jRd/GcXENrDYnFTUW9uRVUV5tISFaS/8eUcTqNaiUcsqqzZRWmtm0uwS7w0X/ntH0To0kJTY8pBc4NVoc7D9Szbvf7WHfEUPj9iG9Y5g7awC9UiJaDOaPR9N6qVJrKK82Y6izYbQ4+C27gKJyE2EaOWdN6kVafBiHi2t474f9HCqq//utX1w6jn/M6ENGUgRqTcstW+XVZgrLTew7XI1cLmFQrxgSorUdahErrzaTX1rH6i35WO1OhvSOJWtgIm6Ph4JSIwaTjYQoLclxYcTptUEZT1thsFBUYeRggQEJHgb3jiM+Wuu1rI7F5iS3qIYV3+1l54GjXeT90vXMnTWAfmn6Fn/kWK1WcnNzycjIaFzq4kTmr9u0u7HZbJSUtDwDW6vVEhcXh8fjweNy4nHY6hdUl0qRqrT13aoBuj9/dSSQ390NRPDWBidi8Ab1qYAqDBZs9vqWL5VC1uH0OKEkGH9IDRo+HFpbkDfYnE43cnnwgyir1UpBQQHJyckBf5Zdqc5sp6rWisPhRqmUEh2hDuoEGmi5XrpcbmwOF3K51CuANxjqqK6z4XC6/2pV0xAW3rXvg9vtCdpkp7awWCwcOHCA3r17N1snjZb699dur584EaVTtSm1mwjehNZ0VvAWOm2EQsDJpPVdg0LgSCQSJF0cuAGdErhBfTdfTU2NT4LnUKfTKoM6bKC9ZDIpWj8tmnq9Dr1e5+cVXacrAzeo/wFlt7e8fEi4Rhn0YFwQgil0+zcEQRAEQRBOQiJ4EwRBEARBCCEieBMEQRAEQQghIngTBEEQBEEIISJ4EwRBEARBCCFitqnQ5RxOFzaHC4VMiqqFvJndndtuwWWsxu10IJErkeuikSo6P9OAze7E7qhfFDXYz9PpBl1kNC2tXB5q3H8toVNRY8HucKFSyomNVBMXpHRxDeRyOTExbc8JW7/Uj5kKgxWH04VaJSc2UuM34XxLHE43Eknw8nnanS4cTjcqhazb5AwtrzZTWWPFaneikMuI1WuI02u6fKZsd7Vlyxb++c9/8uabbzJ27NiuLo6ACN6ELmSos1JYbmLlmkMUV5nQh6s4e3IveiZF+OTv7M7cDjv28iMY1n2Med8WwANSGWEDxqMffx7K2NROWbm7qsbC4ZI6vvjtIJU1VmL1as6Z3Jv0RF3An2dxhYk9eZX8sPEIDqebQb1MnDoqjYRobbNJ1UNBda2V37IL+HDVPurMjsbtep2K80/pQ3JcGAnRWuL0GsICsNREZY2FCoOFNdsLKa0yExelZcrwCOL0EmJaCMIqayz8sPEwn/96ELP1aIaSmEg1l585gFEDEokIa758DqeL0iozW/eWsWVPKTKphOlj0umbpicxJjALdFcYLBwpqeX7jYdxON1EaJXMGt+TpNgwItu5aHSg1BhtbPyzhHe/20N1k+wS4RoFf5/el1NHpREV0Xnrt1VUVHDw4EFycnLIzMykd+/exMa2PyduMNXV1XH33Xfjdjef71jofCJ4E7pEhcHCCx9vZ8ueMq/tW/eW0TMpgnuuyArYl0gwuZ0OzAe3Ufbp094pV9wuTLvWYtqzgcR//B+aHoORSNu+yGVFtYWiSiNVNVaSYsNIiA5rMR9tWbWZx9/azL4m+SrzimvZsqeMzB567pw3hvgAtRwdyDew6OXfMVmOBjf7jlTz2S8HufLsQZw2tgfhmu4ZwNWa7JRWmdidW4VKKWNgz2jio7Qo5FKKKkzsLzAglUq55vyh/L6ziN93FgNgqLPx+le7mD4mDY1KjlIu5fxT+x5XEFJSaeK97/fy89YCr+1frTnElOEpzDtzgN+/gapaK0s/3M7WnDKffZU1Vp55P5uLZ/Tj3FP6+H0f7HYXW/aW8sQ7W3A1SX+2dW8ZOq2CB6+ZQJ9UfYfvq+Henv9oO5OGpTCsbxzVtVbio7Xszq3k699zueJvg/zmCg6mOpOdj1bv44vfDvnsM1ocvLFyN/vzDVx93hCiApB3tTUVFRXcf//9bNiwoXHbuHHjuP/++7tVAPfoo4+SlpZGUVFRVxdFaEIEb0Knszlc/G/1Pp/ArUFecS3Pvr+Ne67IatOq58FSUmGisNyIQi4lOS6MWL1v8OM0lPoGbk25nZT+7zFSFjyDMjqpTdc9UlrHvS//3pgrEmBAzyjuuHS03647o8XOS5/u9Arcmso5bOCVz/7gljkjCTvOoKq4wugTuDX1+le7SE/QMWpAwnFdJxiq66y8tXI3q7fkN26TSODOeaPRqhUsfmMTNoercfs5k3tx5oSefPN7XuPxv24r5OFrx1NSaWbH/nJ6p+hJjgvD47TjNJTiqCxGFhGNQp+ITNv84rl2u4vvNxz2Cdwa/La9kOhINXNn9Ud9TNf39n3lfgO3pj5ctY+sQYn0S4/y2Xe4tI7H3t6Mv9w6dWYHi176nWdvnkpSbMd+PNUYbbz86U7OGJ/Bsk92UGs6umBueqKOi2f046t1B7lkZn8ULaR7C7QjpbV+A7em1u4oYurIVMYNbtvf6vE4ePCgV+AGsGHDBg4dOtRtgrcvv/ySnTt3smzZMs4999yuLo7QhAjehE5XYbDwwwb/SbUb7M6tz6nZVcFbQVkddy5d09h11iNRx71XjfNqvfJ43Jj2rG8+cGs4zmnHemR3m4I3g9HGk+9s8QrcAPbkVfP5bwe54qyByI9JU1NpsLJ5dykt2birhIoay3EHb38eqmw2cGvw1je76ZOm77Kusebsya3yCtwAPB5wONw8/N5GHE631/YvfjvEjRcPR62UYbXXB3WXntGfN77aTc6RagDUShlPLJxMvGkfpf97HKiPiCLHnkPUpL8jVftv7SyuMvHV2pYDiW/W5TJ9TDo9k47mxq2utfLBDzltut+v1hzihouGo2qSJ9hqd/LxT/v8Bm4NTBYHfxys6HDwVlZtZkBGDG99vdsrcAM4UlLH7zuLkUrqPweSYsM7dI32sticfPrzgTYd+/4PexnYM5qIINffnBz/7+O+ffvIysoK6rXboqCggMWLF/PMM88QFtb9e0FONt1j9KhwUimtMnt11zTnz4MVrR4TDB6Ph29+z/Ma83S4pI4Dx7Rsucx1GHetadM5jTt/wW23tnpcVY2VvOJav/u+X3+Yqhqbz/b9BQbfg/042MbjmmO2OvhhY8tBN0BuUS0Go285u5LJ4uDTX3y/vHVaBbUmu1fg1tSGP0sY2T8eALlMQkSYsjFwA7DaXfzvx33U5GTTELgB1Gz8CmddZbPlqc8p7GqxzHanmwqDxWtbjclOcaWpxdc12LS7hLpjgqcao40NfxS3+trvNxxuNUhvzs79FcREqpst54Y/ixmeGU+FofW/h0CpM9vZsrfl1soGhwprqTV37N7bIzMz0+/2fv36Bf3aBQUFZGZmNvtfaWkpd955JxdddBEjR44MenmE9hMtb0Knczhb/tJqYG3lyy1Y3G4PZVW+XzxNBzgD4PHgcbbtQ97jcuBpw4Dflp6NzeHyG/Ta2/ic7M0EKG3ldntwONp2Dk8bgvPO5HS5vQb2N1AqZJhtzb+HJoujMf+vXCb1WyfLayy44o5t0fTgcTafX9PdxufjOqbOeFpqMjuG0+n2Od7thrZc2u5wtbmMx7LanbjczU+WaKjDx95bMHncnnbdj7sdz7mjevfuzbhx43zGvPXq1Svo105ISOCbb75pdv9HH32E2WzmhhtuEBMVuikRvAmdLiaibTMfex/noOmOksmk/G1iLzbuOtoVKZNKGNzLexkHqVKNMiEDZ03rv+hVyX2QKlvvhtGHq9Cq5X4DjQEZ0YRrfbs90xLalpg8Ne74uqg0KjkDekZzsLCmxePC1PJuN+M0XKtkwpAkPiyt89peXWsluYWuu+F94/hlW/24NKvdhU6rQKWQNY6Ng/qxcVpDPk2fijwqCVm473izBm3tUj72OI1Kjkopa7XVDiA5LtxnqRi1SkZybBhFFS233g3MiEaj7tjXQ5+0KGpNNp/n1CA9UUdplZm+afoOnb8jlEoZiTFaSirNrR4bplGgVgZ/LF5sbCz3338/hw4dYt++ffTr149evXp1yng3hUJB7969m93/6aefUlZWxvjx4722L1iwgKysLJYvXx7sIgqtEN2mQqeLi9LQI7HlgCNco2hzUBIM/dL1LLoyi8z0KIb3i+OxGyaREuc97kOqVBM59qw2nU83bFqbZpvGRmm55ryhPtvlMilXnTMYnda3RSMpNozoVpY3iIlUH/fsXZlMymlje7R63Owpvdu91liwyaQSpo1O81k+w+2Bihozp4xK9XlNUkwY0ZFqCsuNjds+Wr2fe+ePZVT/ePqm6bn1kpEM7xdH5KhZRJ1yCcr4HuiGzyDpH/cg10U3W55YvZqBGc3vB+iXHkXcMUu8xOo1zMxq/T0AuPi0fj73G6VTM2em/+66pmaO7dHhNdl6JupYt72QC6b18dknlUq48NS+5BYaiO3E5YCidGouPq31+waYPaVXp5UtNjaWrKwsLr30UrKysrrNRIV33nmHr7/+mk8//ZT333+fl156CYCHH36YRx55pItLJ4BoeRO6QGS4ilvmjOTuF9b67YaSSiXcOW80cV0YAIRplIwdlMTgXrHIpBLUKv9/KoqYZLT9xmDet7nZc+mGz0AeGd+m68qkEsYNTuSx6yfx8U/7KCo3MTAjmtlTepPaTDAbq9dwx6Wj+M/Lv+N0+Xb3yGVS7rh0dEACqoRoLXNPz2TF9/4HW/dI1DF9THq3XOw0OS6cJxZOZuXaQ6zZXohSIeOsiRlMHp7CtFFSxgxI4Ku1h7DaXIwekEBqfDivfP6H1zmsNiflBgsJ0VquPm9Ik1Y7FfoJ5xExciYSpRqprOWWx+gIDdeeN5T/vPI7NUbf7tWIMCX/umCoz1pvcpmUMyf25KctRzD5aZ1tkBofTqafmaYAQ/vEMqp/PFubGQN26az+xxXox0ZpOWdKH9btLGThRcP5cdNhyqos9EqJZPqYNH7ceJgrzh6EroV16IJhSO8YEqK1lFY13/oWEaZk6ojUbll/O1NKSgoALpcLq9VKZWX9+M2EhAQSErrfTPKTkcTTnkEUJ6k//qj/AB8yZEiXlsNsNrNnzx4GDBiAVhvc1d6Dze32kF9Wx4c/5rBuZ3HjeJTh/WK57IyB9EyOCNoyAoF+js66KqrXfkxd9o/eM09lciKzziYy62/IW+hCa47V5sTqcKFVyVEqWn4WTqebvJJa3vlmD9n7yvB46pe7GJkZz7wzBtAjKSJgq9vXme3sOlTJW1/vpqCsvlVKq5Zz9qRezBzXI2DryQWL0+WixmhHIpEQpVMhkRz9oi4sM/Lz1nw27S4ht8h74khKXDj//NtACsrrmDQsJSDrEB4uruX7DXms2pyPxeZErZQxfUw6s8b3oGdSZLOvO1hg4P7lGzAcOw4T6JUcyd2Xj2lxtmhljZVNu0v48MecxpnNvVIiuHTWAAb0jCbcTwtve9idLg4X1fLRT/vQh6uJilBRXGFCIZdy7tQ+QW1Vb+nvu7DcyMOvb2yst03F6tXcd9V4r9m9x7JareTm5pKRkYFa3bnr1HWFpsHbaaedxttvvy0yLLTCXx0Jxne3CN7aQARvwWO1O6k01KepUSpkREeoj3s5i9YE4zm6HTactRXYSg7hMhqQR8SijO+BPDIOqbzzxn+ZrQ4qa6zY7C5UShkxkeqgjT8z1NmoqjVjMluI0YeTEKNDdgK0WFQYzOSXGvnm91xqTXaiI9TMGteTGL0atUJGVKQmoPdpd7ooqzJjd7iQ4CFeryYsrPVW0rJqM0dKavnm9zxMFgfxURrOmJBBYkzr3egNKgwWzFYHEkn9TNpAL+9isjioqv0rfZdSToxe47V0STC09vddVWOhqMLEN7/nUVljISJMyZkTMkiND281BdrJGryp1Wpkss5bky+UdVbwJrpNhS6lVspJie+ctZ6CSapQoYxJQRmT0qXl0KoVnTZZQK9ToZS52FNyCH3KgBMicAOI1WuJ1WsZ2icWh9ONQi5FFsScnEq5jNR4XeMHfELUgDa9Lj5KS3yUluH94nE63SgUsna/B/Vd6cEbnhCmUQT9x1h7RUdqiI7UMKBndKe8v4IQDCJ4EwRB8EMmC40vdblM2m0SvoeSUHl/BcEfUXMFQRAEQRBCiAjeBEEQBEEQQogI3gRBEARBEEKICN4EQRAEQRBCiAjeBEEQBEEQQogI3gRBEARBEEKIWCpEEISQVVhuxGpzIpVK0IcriYroXjlVj1dZtRm73YVEAtGRajSq7rVmmiAIXUMEb4LwF4vNgcXmJFyjbDUdldC1isqNbNhVzBe/HqKqtj6904Ce0fxjZia9UiLRBzhTQGcrrjCxJ6+KT3/ez5HSOpQKGZOHpfC3iRmkJYajUoiP7q7i9nhwOF0oXO4Tep244uJinnjiCTZu3IjD4WDo0KHcfffd9O3bt6uLJiCCN0Ggzmwjv9TIV2sOUVJpJjM9itPGpZOeEIFCfuJ+OIeqonIjT7+3jX1Hqr2278mr4r5X1rNg9mBOGZVKRNjxBXButxvzX8F8ZzKYXLzy+Z9eieNtdherNh/hl20FLLoyi2F94pC3sW4eKanlj4OV7D5USY8kHWMGJtIjUYdUKup2e5gtDnKLDNSZ7JRVWdBoPETpVCgVsg4HcVarlZKSEoxGIzqdjoSEhG6Rdstut3P11VcTFRXFf//7X3Q6HS+99BKXX345K1euJDo6uquLeNITwZtwUnO73WzbW87T722lIcvvgQIDqzYf4eFrJ9C/p/iQ6k7cbjdrthf6BG5NvfrFnwzIiD6u4K2grI6fthSwJ7eSEf3jmTQ0meS44Kdx0+l0bMsp9wrcmnK63Dz+9haeunEy6YnNJ1BvcLi4lntfWd/YOsl2+PinAzx49XhRt9vpj0MVvPzxNi4/LRGHy43b6sRkdZIYrSVcq0AiaV9qsrKyMl5++WVWrlyJy+VCJpNx1llncc011xAfHx+ku2ibLVu2sG/fPn755RciIiJQq9U88cQTZGVl8dNPP3HhhRd2afkEMWFBOMkVlZt4+bOdjYFbA5vDxetf7aKqxtI1BRP8Kio38dXaQ60e91t2IU6nu0PXqDCYeeGjHXy0eh9/HqrknW/28NY3e6g12Tp0vvZQhsXw6S8HWjzGYnNyoKCm1XOZrQ4++fnA0cCtyetf+fwPKkXdbrOiciMvf/YHHj/7yg0WHO2sa1arlZdffpkvvvgCl8sF1CeB/+KLL3j55ZexWq2tnCG4+vbtyyuvvOITRHo8HmpqWq97QvCJ4E04qVXVWakzO/zu25NXRa3J3sklElpisjqoMbb+nuzJrcJo6dh7V11n489DlV7b1v9R5BMEBYPdCeXVrQdVuw5VtHpMdZ2NTbuK/e7bn29o03MU6tWY7M2+Ly63B5erfcFbSUkJK1eu9Ltv5cqVlJSUtLuMgRQXF8fUqVO9tr399tvYbDYmTpzYRaUSmgq5btN77rkHl8vFY4891uJxBQUFPPTQQ2zevBm1Ws15553HrbfeikwmBqILR7WU0FsukyCVtq8rRAiutr4fMpmkw+OQ5DIpUqkEt/toO4tSIUPazm6xjmjrNRTy1j/HJBJQKeWYrE7f60glyETdbjOFrJVn1c66YTQaG1vcjuVyuTAaje06X3sVFBQwffr0ZvevXbuWuLi4xn//+OOPPPvss8ybN4/+/fsHtWxC24RM8OZyuXjqqaf4+OOPOe+881o81uFwMH/+fDIyMvjggw84cuQI99xzDyqVihtvvLGTSiyEgphIDanx4RSU+X5YThyWTKz+xFp6ItRFhilJT9BxpLSuxeMmD09Bp+3YRIM4vZpzJvfi818PNm6bc1omCTHaDp2vPeQyF/3So1oc0weQNSix1XMlRmuZNb4n732/12ffhCFJxER2/cD4UBEVoSazRxSV1b71TqmQIm8tuDuGTqdDJpP5DeBkMhnh4cEdX5mQkMA333zT7P6mExI+/vhjHn/8cc4880z+/e9/B7VcQtuFRPB28OBB/v3vf5Ofn09ycnKrx3///fcUFRXx0UcfERERQb9+/aisrOSJJ57g2muvRans3NljQveVEK3lljkjeWD5Bq8u0tT4cC6ekYlWLdbV6k7io8P4x8xMnnhnS7PHaFRyBveO6fA1wrUqzpnci9H9EzhcWkuv5EiSYsM6ZXmOuqoS5szM5IHlG5o9JjU+nKQ2BJJSqZRTRqaw70g1W/aUNm7vnRLJ3NP7E97B4PZkFBOp4V/nD+W/H2z22i6TSkiI0rapJbSphIQEzjrrLL744guffWeffTaJia0H58dDoVDQu3fvVo975plnWL58OZdeeimLFi1q96QMIXhCInjbtGkTAwYM4MUXX+Tmm29u9fgtW7YwaNAgIiKOzsYaN24cRqORvXv3MnTo0CCWVgg1/dKjePS6ieQV11JcYaJXSiQp8eEkxwZ/dqHQfgN6RnP+qX349Gffgf0alZz/zB9LeoLuuK4RF6UlLkrLsH5xrR8cQHa7nZ6pYcw/ZxCvfbnLZ39ijJa7LhtDUhvrZlJsONdfOJTKGitVtVZ0WhWxejWJMWGBLvoJr3eqnpsvHklNZRERYUrCwzSolHJUHVgTUq1Wc8011wB4zTY9++yzufrqq7vFciFPPvkky5cv5+abb+bqq68WgVs3ExLB25w5c9p1fElJic8vl4ZZM0VFRR0K3jweD2azud2vCySLxeL1v0LH+HuOsRFyYiOikQ6Ixe2uH3zc1e93KOiKOqlVwlkTezB2UCLfb8jjQH4NCoWUqSNSGd4vjgS9ostn63VEwzOUuG1MGZ7E4F6xrNtZxP58AyqljOlj0kiPDydaJ29X3dQqQRunJi3uaEBwItftYNbJqHApllop+nAlKlX912dzY9daExMTw2233ca8efOoq6tDp9MRHx+PSqXq8DkDZdOmTY0tbmeeeSZlZWWNwZtWqyUsTAT/zXG5XLjdbiwWS+N3SdM6qdUGZvhFlwdv7R042RZWq9Wr1Q1Apapf88lm69h0f4fDwZ49ezr02kDLy8vr6iKcEMRzDJyuepbnTUjAI01DIgGHpYaq4gNU+Z9gGTKaPsspA6OZNiwKqcRDVVUlpQXFlDb/UqGJYNVJuVyO1WrFc+z6Qh0UFxfX+B3n8Xi6xQ+Phu7cd999l3fffddr39VXX821117bFcUKCTabDafTyaFDvksa5eXlERPT8SEdTXV58NaegZNtpVarsdu9p8E3BG0djXoVCgV9+vTp0GsDxWKxkJeXR8+ePdFoxED6jhLPMXC617MMJyUlpYvL0HGtPcuoKH3nFyoEBbNO2mw2ioqKUKlU3aJrM1geeeQRHnnkETweDzabDZVKJbpN20Eul5Oent7YaNS0TgbsGgE7Uwe1deBkeyQmJrJv3z6vbWVl9SuWJyQkdOicEokkYM2dx0uj0XSbsoQy8RwDRzzLwBHPMjCC8RylUilSqRSZTHZSLDvV0H0rkUhOivsNBJlMhlQqRaPR+AT4gfwxcUIu0jtmzBh2797ttVbO+vXrCQsLE2vUCIIgCIIQ0k6I4M1ut1NeXt7YVTpjxgzi4uK4+eab2bt3L6tWreLZZ5/lyiuvFMuECIIgCIIQ0k6I4C07O5tJkyaRnZ0N1E9OWL58OW63m4suuogHHniASy65hOuuu66LSyoIgiAIgnB8unzMW3u98847PtvGjh1LTk6O17YePXrw+uuvd1axBEEQBEEQOsUJ0fImCIIgCIJwsgi5ljdBONG5XG7KDRbKqs0YzQ6idCriorQBy7NaWWOhrNpCSaUJjUpOanw4cfr61eKF7sXpdFFusHCwoIYjpXVEhqsY3CuGWL2GMM2JnbqtxmijvNpCabUZuUxCcmw4sXoNGpWop4Ig/goEoRupM9n5/Y8iyqotpMaF43C5sTlc/JZdSNagRAb1ikHZgXQ8UB8U7s838OS7WyirPrr6vEwq4YwJGVw4re8Jn6zcYnNiqLPi8YBep+oWuWvtjvoAzWiyER6dgqt+UXZMFjvrdhbz8qc7sTvdXq+ZOjKVf/5t4HEF9A0/EqprrchkUmIi1cREdvVaffUOl9Ty9Iqt5BbVNm6TSurve96ZA4lrx30b6mxY7U4UcikWmwuTxU6YRkFspAa1CARPSh6XE4/bBRIpEqkMiTT0OiFFzRWEbsLpcrN2RyE6rZLt+47wv1X1axVKJDBxaDKF5UZUShkDMzq2QveR0jr+78V1OI4JBFxuDyvXHsJsdbBg9uATNmF5UYWRN77axaZdJXiAkZnxXDV7MKnxx5cH9XhU1Vj49JcDfL0uF6fLg1wm4fRxPfn79L7sya1i6f+2+33dr9sKcDpdXH/hcHRh7X+/LDYHG/8sYdknO7HYnADE6tXcNW8M/dKjkEq7bkHWonIj/7dsHbUm74XW3R74eWsB1bVWbps7Cr2u5R8aTjf8caCClz7byYwx6eSX1vHTlnxcbg9SCZwyKo15ZwwIWIu20P15PG7cdivOmnJwOQAJUo0OWXgUUnnX/5Brj9ALNwXhBFVhsFBVZ+OTnw+w70h143aPB9buKOJISR2/bMun1tT+FG9Wm5P3f8jxCdya+mlLPuWGEzNvboXBzL0vr2fDnyW4PfXPdOveMu558XdKq0xdUiany83Xv+fyxW+HcLo8f23z8PW6XD7/9SAb/mw5z9e6ncUdfr8OF9fx9HvbGgM3gAqDlUUv/05xZdc8D6j/IfFrdoFP4NbU9v0VFFW0XEaJREJxpZVFL60jSqemtMrMj5uO4HLXP2e3p76+v//DXmx2Z4vnEk4cHocDZ1XxX4EbgAe3pRZXbUV9S1wIEcGbIHQTReUmkmPDOFBg8Lv/py359ErWU1nT/tyHVbVWNrYSDABs3FXS7nOHgsIyE6VVvsnYq2qt5BXXdUGJ6oP1L37zzX8I8M26XIb2bT2nc2sBnj82u5NPft7fzD4Xfx6oaPc5A6W61spXa/w/k6a+W3+4xdyiYbpIPv5pP24PTBqWzKrNR/wet3pzPhWGrs0larVaqa6u7vScpm63myVLljB58mSGDRvGlVdeyeHDhzu1DJ3J43HjMhkA33rjtpnwuEIriBfdpoLQTdSabC22jNkcLqRSCTZ7+38h2p0u3G3Io11efWK2vJVV+wZuDYrKjc3uCyab3dXse3nsGLfmVHSg5c3qcFFY3nzL1eGSrglmob7lrc7saPW4yloLDqe7+fGfEjn5pfXva0t/My63x6v1sTMZDAYOHDjAihUryM/PJy0tjblz59KnTx/0en3Qr79s2TI++OADHn30URISEnjyySdZsGABK1euPDEXs/d48Dibb9HF3ba/ue5CtLwJQjcRHalGqWj+TzJco8DucKFRt/83l0ohQyFv/c89PbHrxn8FU0p8eLP7MpIjOrEkR2nUcnRa/+NswtRynG0I4NISmr+v5mhVcjLT9c3u798zqt3nDBTFXxMnWpMer2t54o7LzoC/7sPmcDX7nFUKWZfM2jUYDLz22mtce+21rFmzhry8PNasWcO1117La6+9hsFgCOr17XY7r7/+OgsXLmTq1Kn079+fZ599ltLSUn788cegXrvLSKRIlM3VLQlIQyt3qwjeBKGbSIwO41BBDaP6x/vdf9akDPJLa4ntwIzAWL2G6aPTWjxGIqkfxH8iSooJa/wyb6pnko60hK4JWGMjNVx6xgC/+/4xM5M/Dpa3+HqJBEb1T2j3dRVyGedM6Y3Mz6QEvU5Fv/SuC96iI9X8fXrfVo+bnpXe4n6jsZbZU3qjUsj4aXM+50zu7fe4C07tQ1xU58+wPnDgAO+//77ffe+//z4HDx4M6vX37t2LyWRi3LhxjdsiIiIYOHAgmzdvDuq1u4pEIkGmjcRf2CPVRiCRhVZHpAjeBKGbiIvSMCAjhglDkzllZCpyWf2Xq1Yt56Lp/XB7PMwan9GhlgKFXMa5U/sQ0cLMxDmnZRKrPzGXComKUHPHpaO55PRMonQq9OEq/j69L/+5clyXLY8hlUqYNCyZGy8eTpROBYA+XMX1Fw7j1FFpnDmhFy1N+jzvlD4dnimZFh/Ow9dOICXuaMvdkN4xLP7XRBJjwjp0zkAZMyCRlLjmy3D6uB4kRGtbPU+8XsFTN00hOS4Mh8vNFWcNRB9e/5zDNQrmnzOYMyb0RC7r3BYXq9XKihUrWjxmxYoVQR0DV1JSP7Y1KSnJa3t8fDzFxe0fRxkqJHIFitgUpKow6lvb5Mgi4pCHR4XcciGhFWoKwglMIpEwol8cew5XoVHLue7CYXg89bMS84pqmDU+gx6JHe/iS4kP59HrJrL8iz/J3ne0VSciTMm8MwYwYUgSGlVoTZdvj7goLRfNyGTm2B5AfaAkk3XtB3ZEmIrTsnowol88Zqsdm8VEamIUGo2KzHQ5D1w9gSX/y/Yai6hWyrj4tExmjEnr8Dp1crmMwb1jeez6SdQYbUilEqIiVIRrun6sU3y0lvsXjOe97/fyW3Zh4wzRMI2Ci2f05dRRaejauJxNz6QIbrxoOCarE7VCxuThKVjtLlRKGbGRmi5ZEsVisZCfn9/iMfn5+VgsFtTq4PyYsljq69OxY9tUKhU1NTVBuWZ3IJFIkChUSPTxeNzu+n+HWItbg9AstSCcoNQqOSP6xdMzKYKqGis2hwutWsGU4SkBWX8tPTGCuy4fQ0W1hTqzHblMSnSkmji9Bomk69b26iwyqaTbLETbVKxeg9nsYU9hMZ4EPQBKpYzh/eJ4cuFkyqstGIw2NCo5CdFa4vSagASeep0K/V+tft1JYkwY1184jItm9KPGaEcmlRAdoSZW3/6AS6WUN2YP0XaDrBQajYa0tDTy8vKaPSYtLQ2NJnj1tCEotNvtXgGizWYL6nW7i/qFeUNrjNuxRPAmCN1QlE5NVCuLkHZUmFpBWFLXf4kJbRMTqemWAWewqZRyUuN1pJ5gwzDVajVz585lzZo1zR4zd+7coLW6wdHu0rKyMtLTj44fLCsro3///kG7rhA4odXJKwiCIAghrk+fPsyZM8fvvjlz5tC7t/8JFoHSv39/wsPD2bhxY+O22tpadu/ezejRo4N6bSEwRMubIAiCIHQivV7P/PnzOeWUU3zWeevdu3fQ13lTKpVceumlPPXUU0RHR5OSksKTTz5JYmIip512WlCvLQSGCN4EQRAEoZPp9XpGjRrFoEGDsFgsaDSaoHaVHuvGG2/E6XSyaNEirFYrY8aM4bXXXjsxF+g9AYngTRAEQRC6iFqt7tSgrYFMJuOOO+7gjjvu6PRrC8dPjHkTBEEQBEEIISJ4EwRBEARBCCEieBMEQRAEQQghIngTBEEQBEEIISJ4EwRBEARBCCEieBMEQRAEQQghIngTBEEQBEEIISJ4EwRBEARBCCEieBMEQRAEQQghIsOCIAiCIHSB8vJySktLMZvNaLVaEhISiIuL6+piCSFABG+CIAiC0ImsVivZ2dksXryY4uLixu1JSUncc889DB8+POgpswwGA8888wy//PILRqORzMxMbrvtNkaPHh3U6wqBIbpNBUEQBKETZWdnc+ONN3oFbgDFxcUsXLiQ7OzsoJfh1ltvZceOHTzzzDN8/PHHDBo0iPnz53Pw4MGgX1s4fiJ4EwRBOE5OUw22kkNY8v7EWpCDs7aiq4skdFPl5eUsXrwYj8fjd7/H4+HRRx+lvLw8aGU4fPgw69at47777mP06NH06tWLe+65h4SEBFauXBm06wqBI7pNBSGAXFYTHocNqVKDVKXp6uIElcPporLOSVRCBha7B602uNcz1Fmx2Fwo5FJiItVIJJLgXrANPC4ntpJDVHzzEvayw43bZeFRxMz4J5peI5BpwrxeY7TYqTPZkUgkREeoUSpkrV6nutaK2epEoZASE6lBJj1671W1VhxONxqVjIgwVeBurgPMVgcOpxudVolU2vXvT3dUWlrq0+J2rKKiIkpLS4M2/i0qKopXXnmFwYMHN26TSCR4PB5qamqCck0hsETwJggB4LZbsZUcovrX93FUFqJM7EXUlH+gjE9HKld2dfECrrDcyBe/HeSnzfnYHC4GZkRzxdmD6JsW5RVYBILF5mB3bhXLv/iTgjIjEWFKLj6tH5OHpxClC+64oNbYSg5R9PZ/wO302u4yVlP2+bPEnHENuqGnIpUrcLs95BXXsPyLP/njYCVymYRTRqZy8WmZJMaE+T2/2epg16FKXv38T4orTWjVcs47pQ+nZfVALpewdU8p7363l/JqC71SIph/9mD6pUehVnl/tBvNdkxWByqlHH144AO8OpOdnCPVfLR6H0azg8kjUjhlZGqz93UyM5vNAT2uIyIiIpg6darXtm+//ZYjR44wadKkoF1XCBwRvAlCAFjz91DywcON/7YczMZyaAfJlz+COqVfF5Ys8MqqzDzz3lb2HTE0btudW8U9y9bxxMLJ9E7VB/R6e3Kruf/VDY3/rjXZefXzP6mssXLJ6f1RtaHlKhicploqvn3FJ3BrquqH19H2HII0OonCciN3Pb8Wq91V/3qXh1Wb89l7uJqHrpmA1k+Mvzevmgdf29j4b7PVyYrv9iKVSJBJJbz59e7GfYcKa7nnpd95+JoJDOt3tMWmsMzIko+2s/tQJemJOm6ZM5I+AXyPHE4XqzYf4fWvdjVuW/HdXn7anM/D104gPjrITbIhRtvGJuq2HhcIW7du5f/+7/+YPn0606ZN67TrCh0nxrwJwnFymQxU/vC67w6Pm8rVb+OymDq/UEFUWmX2Ctwa2J1uvvk9F4fTFbBrGeqsLP/yT7/7vvztIBXVloBdq71cdZXYS3NbPMbjcmArPojL7WHV5iONgVtTBWVG8kvrfLbXGG281sy9R4Yr+eDHHL/7XvvqT2qNNgDqzHae+2Abuw9VAnCkpI4Hlm+gvDpwrTrlBgvvfrvHZ3txpYk9eVUBu86JIiEhgaSkpBaPSU5OJiEhoVPKs2rVKubPn8/QoUN55plnOuWawvETwZsgHCeXzYKjqsjvPlvRftz24HV/dIXCcmOz+3IOV2M0OwJ2Lavd5TewgfqWq1qzPWDXai+3rW3vq6OqCKvdyZ8HK5s95mCBwWcMn8Xm5Egz9y6RSPwGggC5RbVY7PWtgUazg72Hq732G+ps1JgC99zqTHbsTrfffdv3lwXsOieKuLg47rnnnmbHbEokEv797393ynpv7777LgsXLmTKlCm8+uqrQV+eRAgcEbwJwnGSyhVItRF+9yn0CUhOsDFvsfrmJ2IkxmjRqAI3GkMhlxKla36MllbddSM/JHJFm46ThUWhUkhJiWt+/FdSbLjP7EOFrPl7l0klzU4IiIlUo5DVdyWrlDISjum2VCllhKnbVva20KoVNDd3JCM5MmDXOZEMHz6cJUuWkJyc7LU9OTmZJUuWMGLEiKCX4b333uOhhx5i7ty5PPfccyiVJ9bn1IlOBG+CcJxkuhiipvzD777oUy5BHnZifYElx4YRHeH/F/rsqX18Bssfj5hIDXNO7+9335iBCcRGdt2MXrkuBpkuppWjJKjT+iOXyTh7ci+/QU6YRkGvVN86EqPXMGdmpt+zFlYYOWVUqt99l84aQHRk/fsTHaHmtrkjCdPUB2tKuZRb54wkLipwzy1Wr+HUUWk+2zUqOaMyO6frL9So1WrGjx/Pa6+9xptvvsmyZct48803ee211xg/fnzQW8Byc3NZvHgxp512Gtdccw2VlZWUl5dTXl5OXZ3/1l6hexETFgThOEkkEsL6j8PjtGNY+xFuqwmpNoKYafNQ9xjc+glCTHJcOPddNY5n399GXnEtUB+ALJg9mJ5JuoBfb/zgRCprLHzy0wGcrvruuayBCVxz3tDGoKQryCNiiJl5JWWfPNnsMbqRM5FHxAKQlqDjrsvG8Pz/tmO01HctJ8WEcedlo0mKCfM7u3D8kCTKDRY+++UATld9y9zg3jHMzOqBXCZFAvy8tQC324NKKePiGf0YPdA7YMpMj+a/t06lxmgnXKsgXq9BLgvc73aNSs5lZw4gOkLNyrWHsNpdDO4dw4LZQ0huobVRqO9C7Yp0WN9//z0Oh4Mff/yRH3/80Wvfeeedx2OPPdbpZRLaR+JpbqVAodEff/wBwJAhQ7q0HGazmT179jBgwIBOnYl0ognWc/R43DhrKv5a502NLCK2W6xFFiwVBgsVBgs2u5PoSDUpcbqgre1ld7ioMFioM9tRq+TE6jUB7frrKJfVhGnvRiq/fxWPs+k4Mgm60bOImnABcl1U41a320OFwUJVnRW5VEJ0pKaxFbO5eml3uCivtlBrsqFRyYmJ1KALq+/istqdVBgsWGxOwjVK4qICG5i1h8vlprLGisvtISJMQZima7rhgvk5abVayc3NJSMj46QYH+ZyubBarajVamSyrpnVHWr81ZFg1EnR8iYIASKRSFHo47u6GJ0mVq9Bq/SwZ88eYlIHBHVRVqVCRnJceNDO31EydRi6wVPQpA/EVpaHo7IIeXgkqpR+yHWxPgs1S6US4qO17Vo+Q6mQkRIfTgq+969WykmND3xrZ0fIZFKxLIggdBIRvAmCIBwHiVyOIjoRRXRiVxdFEISThJiwIAiCIAiCEEJE8CYIgiAIghBCRPAmCIIgCIIQQkTwJgiCIAiCEEJE8CYIgiAIghBCRPAmCIIgCIIQQsRSIYIgCILQRUwmEzabDbVaLRZfF9pMtLwJgiAIQicrLi7ml19+4eabb2bBggXcdNNN/PLLLxQXF3d6WXJzcxkxYgSffvppp19b6BjR8iYIgiAInSg3N5cbbriB0tLSxm2HDx8mOzubhIQEnn/+eTIyMjqlLA6Hg9tvv91vbl2h+xItb4IgCILQSYqLi30Ct6ZKS0tZuHBhp7XALV26lLCwsE65lhA4IngTBCFkVdVaKCo3UlJpwmx1dHVxAs5QZ6W0ykR5tRmbw9XVxRECICcnp9nArUFJSQk5OTlBL8vmzZv58MMPefzxx4N+LSGwQq7b9J577sHlcvHYY4+1eNzzzz/P0qVLfbbv2rULuTzkblvoBC63B6fThVIhQyIJXpL1zuRwulHIO/4bze324OiGz6TCYCF7Xxkf/JBDWbUFgBGZ8cw5rR89kyPQqBRdXMLjU1lj4Y8DFXzw4z4Ky43IZVKmjEhh9pTepCWEo5DLurqIQgeYTCZWrFjRpmPfe+89srKygjaJoba2ljvvvJNFixaRlJQUlGsIwRMyUYzL5eKpp57i448/5rzzzmv1+JycHGbPns0dd9zhtV0EbsKxbHYXBeV1fPd7HvllRob0jmHqiFSS48KRSrtPwNJWLpebwnIjv24rZF9+NZk9ourvJzYMmaxtgZzTWX+O1VvyOZBvoF+6nmmj00iOC0fexnMES4XBwpPvbmF3bpXX9uycMrJzyrjh78OYOiIVtSo0/9brLG6e/WA7uw5VNm5zutz8tCWfX7YVsOiKLEb0i0d+HEG50DVsNhtVVVWtHwhUVlZis9mCFrzdf//9DB8+nLPPPjso5xeCKyQ+3Q4ePMi///1v8vPzSU5ObtNr9u3bx5w5c4iLiwty6YRQ5nJ72JZTyqNvbcbjqd+261Aln/96kEevm0SfNH2Xlq+9qmut5BbV8vAbG3E43QBs31fOZz8f4OF/TWRAz+g2nWd3XiX3vbIep6v+ofxxsIIvfjvEI/+awMCMmKCVvzVut4dVm4/4BG5NPf/RDvqlR5GRHNmJJQsMjUbLT1sKvAK3ptxuD4vf3MTzt59KSryuk0snHC+VSkV0dDSHDx9u9diYmBhUKlVQyvH555+zZcsWvvrqq6CcXwi+kAjeNm3axIABA3jxxRe5+eabWz3eYrFw5MgR+vTpE7AyeDyeLp+NY7FYvP5X6Jimz9FgcvHfD7IbA7cGVruLZZ/uYNE/R6NWhEbrm8PlYdchA299s7sxcGtgd7pZ8mE2D109Fq2q5RYbo8XNcx9kNwZuDZyu+u2PXDOOMHX9OTq7TtaYXHz+68FWj1u16QiXnNYHicTT6rHdhcViQa7R88Vv21o8zunysC2njGidHM+xFVcIap202Wy43W5cLhcuV/vHIKrVaubOnUt2dnarx15yySWoVKoOXac1n3zyCZWVlUydOtVr+3333cfy5csbg7qG+uXxeIJSjhORy+XC7XZjsVhwu+s/h5vWyUC1pIZE8DZnzpx2Hb9//37cbjffffcdDz74IHa7naysLG6//Xbi4+M7VAaHw8GePXs69NpAy8vL6+oinBDy8vKQaBMxWZ1+9+8/YqC0shZzdVEnl6xjwqNTsNidlFT6/5FRUGaktLIOa03L96PQJVFe7f+Lr7jCRGllHbZa73N0Vp1UhCdhsrQ+MWFbThlTBkdgrKnohFIFjjw8qdn62NSGP0vonwS1NYbgFypEBatOyuVybDZbh1/fp08fEhISWpy0kJiYSJ8+fbBarR2+TksefPBBn3Ofe+65XHvttcycOdNn3/Hc78nGZrPhdDo5dOiQz768vDxiYgLTc9HlwVtBQQHTp09vdv/atWvb3fW5f/9+AHQ6HUuWLKGiooJnnnmGyy67jM8++wyNRtPucioUioC25HWExWIhLy+Pnj17dugehHpNn2NJTcstF1qNlh6JAzqpZMenss6BRGJv8RiNWk1Gcsv3U1jR8heGRqOmV0r9OTq7ThaUt/3LLDY2lrTk0Bk2YbFYKKxo25ekRCIhKSmZlGQx0PxYwayTNpuNoqIiVCoVarW6Q+dISkri+eefZ+HChZSUlPjsT0xMZOnSpUGdRJCenu53e0JCAr169Wr8t8fjwWazoVKputWEpe5OLpeTnp7e2O3dtE4G7BoBO1MHJSQk8M033zS7Pzq6bWN0mrrggguYMWMGkZFHx7z07duXqVOn8vPPP3PmmWe2+5wSiaTbpC7RaDTdpiyhTKPRECcBnVZBndm3NWdgrxhiIjVotcouKF37SeVOjpSaSI0Pp6DM6LO/Z1IEMXotWm3L42hi9FISY7R+W/DSEnR/PRPvL67OqpP6CIgIU1JrajlIHTc4Cb1O0+YJGt2FWmlv0/1NHp5CeJj4AdeSYNRJqVSKVCpFJpMhk3V8xm9GRgavvvoqOTk5vPfee1RWVhITE8PcuXPJzMwkMTExgKVuu4Z7a9DQVSqRSI7rfk8mMpkMqVSKRqPxCfAD+WOiy4M3hUJB7969A37epoEb1AeJer3e7y8d4eQVF6Xh9ktH8+DyDbjcR1vhwjQKrj1vCOEhErgBqJVyBvWK5roLh/Hg8g1Y7UfHqGhUcm66eASR4a0PgI6OUHPbJaNY9NLvXmuLqZUybv7HCPS6jrU4BEKcXsOF0/ry+le7mj1GKoFTRqaGXOAG4LRUc/6pfXhz5e5mj1EpZAzu3XWTRoTASEpKIikpiaysrMbWra78Ud4Z68oJgdPlwVswPP3006xevZqvv/66sam3oKCA6urqLu/6FLoXiUTC0N4xLLntFNZsLyS3qJZh/eIYlRlPclx4Vxev3fThaiK0Kp695RS27i1l3+Fq+veMZmT/eJJj276Ket/0KJ67dSob/ywh53AVAzJiGDMwgeTYrn0mEomEKSNS2LG/nK17y/zsh9vmjiKpHffanVgsFqYMS2d3biWbdvmOiVLIpdw7fyyJMaF5f4IvrVYrelKEdjshgje73U5NTQ2RkZEolUpmzZrFm2++yUMPPcS8efOoqKhg8eLFjBw5ksmTJ3d1cYVuRi6XkZ4YwdxZEV1dlICQSiWkxoeTGt/xQEsmlZAaryN1WvdbjiImUsONFw1n7+Fq3v8hh8MltcikUiYNT+bcKb1JjdehVIRuF0+YWsr1Fw5n1jgD/1u1j9ziWlQKGdPHpDFjTDopceHIQnD9QUEQAueECN6ys7O57LLLePvttxk7diyDBg1i+fLlPPfcc5x//vkolUqmT5/OXXfdJQZdCsIJIDpSw4ShGgb3isFqdyGR1I+FUylPiI80oiPURA9MpH/PaKw2JxKJhCidKiS7ggVBCLyQ+6R75513fLaNHTvWp79+7NixvP/++51VLEEQukBEuIoTo73UP51WiS6Exl0KgtA5xM84QRAEQWgHsTiy0JzOqhsieBMEQRCENlAoFABdnm1H6L5MJhMSiaSxrgRLyHWbCoIgCEJXkMlk6PV6ysrqZzprtdoTehy1y+VqzK4g1nlrnsfjwel0UltbS21tLXq9PujPSwRvgiAIgtBGDQvoNgRwJzK3243T6UQulyOVio661shkMpKSknzWmQ0GEbwJgiAIQhvVpyZLIj4+Hoej9Ty7ocxisXDo0CHS09NFSsZWyOVyZDJZp7XEiuBNEARBENrpeFNkhQK32w1wXLlcheAQ7aCCIAiCIAghRARvgiAIgiAIIUQEb4IgCIIgCCFEBG+CIAiCIAghROIRS0W3atu2bXg8HpTKrk1T4/F4cDgcKBSKE3ptoWATzzFwxLMMHPEsA0M8x8ARzzIwmj5HlUpFZmbmcZ9TzDZtg+5SaSUSSZcHkCcC8RwDRzzLwBHPMjDEcwwc8SwDIxjPUbS8CYIgCIIghBAx5k0QBEEQBCGEiOBNEARBEAQhhIjgTRAEQRAEIYSI4E0QBEEQBCGEiOBNEARBEAQhhIjgTRAEQRAEIYSI4E0QBEEQBCGEiOBNEARBEAQhhIjgTRAEQRAEIYSI4E0QBEEQBCGEiOBNEARBEAQhhIjgTRAEQRAEIYSI4C1E3HPPPdx9992tHvf888+TmZnp85/T6eyEUnZ/bX2OBQUFXHPNNYwcOZIJEybw5JNP4nK5OqGE3ZvNZuOBBx5g/PjxjBgxghtvvJHKysoWXyPqZD23282SJUuYPHkyw4YN48orr+Tw4cPNHl9dXc1tt93GmDFjGDNmDP/5z38wm82dWOLuqb3P8bPPPvNb/1p6zclo2bJlzJs3r8VjRJ1sXVueYyDqpAjeujmXy8Xjjz/Oxx9/3Kbjc3JymD17NmvXrvX6Ty6XB7mk3Vt7nqPD4WD+/PlIJBI++OADHnzwQT7++GNeeOGFTihp93b//fezbt06li5dyltvvUV+fj433XRTi68RdbLesmXL+OCDD3j44Yf58MMPkUgkLFiwALvd7vf4G2+8kfz8fN58802WLFnCunXreOCBBzq51N1Pe59jTk4OWVlZPvUvNTW1k0vefTXUsdaIOtmytj7HgNRJj9BtHThwwPP3v//dM27cOM8pp5ziueuuu1p9zcyZMz1vvPFG8AsXQtr7HL/66ivP4MGDPTU1NY3bPvjgA8/IkSM9Npst2MXttkpKSjz9+/f3/Prrr43bDh065OnXr58nOzu72deJOunx2Gw2z4gRIzzvvfde47aamhrP0KFDPStXrvQ5ftu2bZ5+/fp5Dhw40LhtzZo1nszMTE9JSUmnlLk7au9z9Hg8niuuuMLz8MMPd1YRQ0pJSYln/vz5nuHDh3tmzZrlufTSS5s9VtTJ5rXnOXo8gamTouWtG9u0aRMDBgxg5cqVbYrILRYLR44coU+fPp1QutDR3ue4ZcsWBg0aREREROO2cePGYTQa2bt3bzCL2q1t3boVgLFjxzZuy8jIICEhgc2bN/t9jaiT9fbu3YvJZGLcuHGN2yIiIhg4cKDfZ7dlyxbi4uLo3bt347asrCwkEknj+3Ayau9zhPpWjpO9/jVn165dREZG8uWXXzJs2LAWjxV1snnteY4QmDp5cvVbhJg5c+a06/j9+/fjdrv57rvvePDBB7Hb7WRlZXH77bcTHx8fpFJ2f+19jiUlJSQmJnpta3h+RUVFDB06NGBlCyWlpaVERUWhUqm8tsfHx1NcXOz3NaJO1ispKQEgKSnJa3tzz660tNTnWKVSiV6vb/ZZnwza+xyrqqqoqKhg8+bNvPPOOxgMBoYNG8btt99ORkZGp5S5O5s2bRrTpk1r07GiTjavPc8xUHVSBG9dpKCggOnTpze7f+3atcTFxbXrnPv37wdAp9OxZMkSKioqeOaZZ7jsssv47LPP0Gg0x1Xm7igYz9FqtXq1ugGNAYvNZmt/IUNEa8/ypptuQqlU+mxXqVTNPpeTsU76Y7FYAHyen0qloqamxu/x7X3WJ4P2Psd9+/YBIJPJePzxxzGbzSxbtoxLLrmEr776itjY2OAX+gQh6mRgBKpOiuCtiyQkJPDNN980uz86Orrd57zggguYMWMGkZGRjdv69u3L1KlT+fnnnznzzDM7VNbuLBjPUa1W+wx+bvhw0mq17T5fqGjtWf76669+B4XbbLZmg7CTsU76o1arAbDb7Y3/H5p/dv7qYMPxJ3IdbE17n+O4cePYtGmTV/174YUXOPXUU/n000+5+uqrg1/oE4Sok4ERqDopgrcuolAovMYOBErTCgH1X8h6vb6xu+FEE4znmJiY2PjrqEFZWRlQ/zxPVK09y5ycHAwGA3a73esXeFlZmU83c1MnW530p6G7qaysjPT09MbtZWVl9O/f3+f4xMREVq1a5bXNbrdjMBhO6DrYmvY+R/Ctf1qtltTUVEpLS4NX0BOQqJOBE4g6KSYsnECefvppzjzzTDweT+O2goICqqurxYDddhgzZgy7d+/GaDQ2blu/fj1hYWHNfkGcDEaNGoXb7fYanHzo0CFKS0sZPXq039eIOlmvf//+hIeHs3HjxsZttbW17N692++zGzNmDCUlJV7rPjW8duTIkcEvcDfV3uf43nvvMXbsWKxWa+M2o9FIXl7eSVX/AkHUycAIVJ0UwVsIs9vtlJeXNzZlz5o1i/z8fB566CFyc3PZvHkzCxcuZOTIkUyePLmLS9t9HfscZ8yYQVxcHDfffDN79+5l1apVPPvss1x55ZV+x3ycLBISEvjb3/7GokWL2LhxIzt37uS2224jKyuL4cOHA6JONkepVHLppZfy1FNPsXr1avbu3cstt9xCYmIip512Gi6Xi/Ly8sYP9GHDhjFy5EhuueUWdu7cyYYNG7jvvvs499xzT+pWjvY+x1NPPRWPx8Odd97J/v37+eOPP1i4cCHR0dGcd955XXw33Zuok4ERrDopgrcQlp2dzaRJk8jOzgZg0KBBLF++nD179nD++edzww03MGDAAF566SUkEkkXl7b7OvY5qlQqli9fjtvt5qKLLuKBBx7gkksu4brrruvikna9hx56iPHjx3PDDTcwf/58evXq5bUopaiTzbvxxhu58MILWbRoEXPmzEEmk/Haa6+hVCopLi5m0qRJjWMOJRIJzz//PKmpqVx++eXcfPPNTJkyhfvvv79rb6IbaM9zTEpK4q233sJkMjFnzhz++c9/otPpePvtt73GzAm+RJ0MjGDVSYmnaX+GIAiCIAiC0K2JljdBEARBEIQQIoI3QRAEQRCEECKCN0EQBEEQhBAigjdBEARBEIQQIoI3QRAEQRCEECKCN0EQBEEQhBAigjdBEARBEIQQIoI3QRC6LbEMpSAIgi8RvAmC0C2tXr2au+66q6uL0azMzEyWLl0a1GsYjUamTZvGp59+6rOvrKyMW2+9lbFjxzJy5EhuvPFGkWxdEE4S8q4ugCAIgj9vvvlmVxehSxkMBv71r39RWFjos8/pdLJgwQLMZjP3338/TqeTp59+miuvvJLPP/8chULRBSUWBKGziOBNEAShm1m1ahWPPPIIZrPZ7/7vvvuOvXv3snLlSvr27QvAgAEDOOuss/jmm2+YPXt2ZxZXEIROJrpNBUHodubNm8emTZvYtGkTmZmZbNy4kb1793LDDTcwbtw4Bg0axOTJk3n44YexWq0AvP3222RmZnp1MW7evJkBAwawZMmSxm0fffQR559/PsOHD2fo0KHMnj27MWn08SgrK+Pf//43U6dOZejQoVx44YWsXr3a6xij0ci9997L+PHjGTFiBLfccgtvvvkmmZmZjcfU1taycOFCsrKyWL58ud9rrV27loyMjMbADaBPnz707t2b3377DYCNGzeSmZnJ2rVrmTt3LkOHDuW0007j3XffPe57FQSha4ngTRCEbue+++5j4MCBDBw4kA8//JDU1FTmzp2LxWLhscce49VXX+WMM87gnXfeaexenTdvHllZWTz++ONUVVVhMpm4++67GTx4MNdddx0AK1as4N5772X69Om8/PLLPPnkkygUCu644w6Kioo6XN6KigouvPBCNm3axC233MLSpUtJSUnh+uuv58svv2w87vrrr+fbb79l4cKFPPvss5hMJp5++mmvc6nVar7++msef/xxoqKi/F7v4MGD9OzZ02d7eno6ubm5XttuueUWBg4cyAsvvMDEiRN56KGHeOeddzp8r4IgdD3RbSoIQrfTp08fwsPDARg+fDhr165lwIAB/Pe//23cPmHCBNavX8/mzZu59tprkUgkLF68mHPOOYcnn3wSpVJJVVUVr7/+OnJ5/Uddfn4+V155Jddff33jtVJTUzn//PPZtm0bycnJHSrvG2+8QVVVFd9++y1paWkATJ06lX/+85888cQTnHXWWWzcuJENGzawdOlSZs6cCcCUKVM4++yzOXDgQOO5lEolvXr1avF6tbW19OjRw2d7WFgYJpPJa9uMGTO45557AJg8eTJlZWW8+OKLzJ07F6lU/H4XhFAkgjdBELq9SZMmMWnSJBwOB7m5ueTl5ZGTk0NVVRV6vb7xuLS0NO666y7uv/9+AB566CGvIOfuu+8GoK6ujry8PPLy8li/fj0ADoejw+XbtGkTI0aMaAzcGpxzzjn8+9//5tChQ2zYsAGFQsGMGTMa90ulUs4444x2z1r1eDxIJJI2bT92/NvMmTNZvXo1ubm59O7du13XFQShexDBmyAI3Z7b7eaZZ55hxYoVmM1mkpKSGDp0KCqVyufYM844g0cffRSXy8WkSZO89h05coR7772XDRs2IJfL6dWrV+N4s+NZU66mpobU1FSf7bGxsUB9S1l1dTV6vd6ntavhmPbQ6XQYjUaf7WazGZ1O57UtPj7e698xMTGNZRIEITSJ4E0QhG7vlVde4c033+T+++/n9NNPbwxQLrzwQp9jH374YdRqNRqNhkWLFvHaa68B9QHg1VdfjUKh4H//+x8DBw5ELpdz4MABr3FpHREZGUlFRYXP9vLycgCioqJISEiguroat9vtFcBVVla2+3oZGRns2bPHZ/uRI0cYOnSo1zaDweD174brNQRxgiCEHjHgQRCEbqlpgLN161b69OnDhRde2Bi4lZaWsm/fPtxud+Nxq1at4ssvv+Tuu+/mvvvuY+3atXzwwQcAVFdXk5uby4UXXsjQoUMbx8E1zM5sep72GjNmDNnZ2eTn53tt//LLL4mLi6NHjx5kZWXhdDr56aefvI5ZtWpVu683adIkDh486DVW7sCBAxw8eJCJEyd6HXvs9b777jtSUlJIT09v93UFQegeRMubIAjdUkREBNnZ2axfv54ePXqwdu1aXnnlFYYPH87hw4d5+eWXsdvtWCwWAKqqqrjvvvuYOHEi5513HgCnn346jz/+OBMnTiQtLY2UlBRWrFhBYmIiERERrF27lrfeegug8TwdccUVV/Dll19yxRVXcMMNNxAVFcXnn3/Ohg0bWLx4MVKplDFjxjBx4kTuueceKioqSE5O5uOPP2bv3r1+x6+15Mwzz+Sll15iwYIF3HbbbQA8/fTT9OvXj1mzZnkd++abb6JWqxk+fDg//PADP//8s88MV0EQQotoeRMEoVuaO3cuCoWCBQsWMGjQIObMmcPbb7/NggULeO2115g9ezY33HAD+/fvp6amhgceeACTycQDDzzQeI7//Oc/yOVy/u///g+Px8OyZctISEjg7rvv5uabb2b79u28+OKL9OrViy1btnS4rHFxcbz//vsMHjyYRx55hJtuuoni4mKWLVvGBRdc0Hjcs88+y7Rp03j66ae56aabUCqVzJkzB61W267rKZVK3njjDQYNGsR//vMfHnzwQYYPH87y5csbWxQb/N///R+//vor//rXv9ixYwdLlizhrLPO6vC9CoLQ9SQekflZEAQh6AoLC9m+fTvTp09HrVY3br/xxhvJz8/ns88+C+j1Nm7cyGWXXcbbb7/N2LFjA3puQRC6lug2FQRB+IvL5WrTrNNjW7faQiqVcvfddzN9+nQuvPBCZDIZv/32Gz/88AOPPvpoR4orCMJJSgRvgiAIf/nnP//Jpk2bWj0uJyen3edOSkri1Vdf5YUXXuDmm2/G6XTSu3dvnnrqKdGNKQhCu4huU0EQhL8cOnTIJ0OBP0OGDOmE0giCIPgngjdBEARBEIQQImabCoIgCIIghBARvAmCIAiCIIQQEbwJgiAIgiCEEBG8CYIgCIIghBARvAmCIAiCIIQQEbwJgiAIgiCEEBG8CYIgCIIghJD/B82VivVlsVm5AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# scatterplot test\n",
    "sns.scatterplot(data=df_merge, x='taxa_log10p', y='metabolite_log10p', size='fishersz_diff', hue='Interest')\n",
    "sns.despine()\n",
    "plt.tight_layout()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 376,
   "id": "11f23fbf-15f4-414c-bcb4-68f19c73a13f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Interest\n",
       "nan    711\n",
       "Yes     55\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 376,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merge['Interest'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "id": "5e9341dc-b7b0-4f3f-8f09-36c9080357eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var1</th>\n",
       "      <th>var2</th>\n",
       "      <th>metabolite_log10p</th>\n",
       "      <th>taxa_log10p</th>\n",
       "      <th>correlations_resp</th>\n",
       "      <th>correlations_nr</th>\n",
       "      <th>fishersz_diff</th>\n",
       "      <th>fz_diff_p</th>\n",
       "      <th>Interest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [var1, var2, metabolite_log10p, taxa_log10p, correlations_resp, correlations_nr, fishersz_diff, fz_diff_p, Interest]\n",
       "Index: []"
      ]
     },
     "execution_count": 377,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v1 = 'H_Deoxycholic_acid'\n",
    "\n",
    "df = df_merge[df_merge['var1'] == v1]\n",
    "\n",
    "vg = 'Rumino'\n",
    "v2 = []\n",
    "for t in df.var2.values:\n",
    "    if vg in t:\n",
    "        v2.append(t)\n",
    "        \n",
    "print(v2)\n",
    "df = df[df['var2'].isin(v2)]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 378,
   "id": "177028f6-2876-4da8-9b8e-543f105ffc65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var1</th>\n",
       "      <th>var2</th>\n",
       "      <th>metabolite_log10p</th>\n",
       "      <th>taxa_log10p</th>\n",
       "      <th>correlations_resp</th>\n",
       "      <th>correlations_nr</th>\n",
       "      <th>fishersz_diff</th>\n",
       "      <th>fz_diff_p</th>\n",
       "      <th>Interest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>766</th>\n",
       "      <td>H_Melezitose</td>\n",
       "      <td>Acutalibacteraceae Ruminococcus_E</td>\n",
       "      <td>0.028029</td>\n",
       "      <td>-0.604038</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.974679</td>\n",
       "      <td>-2.391057</td>\n",
       "      <td>0.016800</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1141</th>\n",
       "      <td>H_Phosphocholine</td>\n",
       "      <td>Acutalibacteraceae Ruminococcus_E</td>\n",
       "      <td>0.162727</td>\n",
       "      <td>-0.604038</td>\n",
       "      <td>-0.750000</td>\n",
       "      <td>0.974679</td>\n",
       "      <td>-3.638724</td>\n",
       "      <td>0.000274</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1891</th>\n",
       "      <td>H_Gamma_Valerobetaine</td>\n",
       "      <td>Acutalibacteraceae Ruminococcus_E</td>\n",
       "      <td>0.527426</td>\n",
       "      <td>-0.604038</td>\n",
       "      <td>-0.392857</td>\n",
       "      <td>0.974679</td>\n",
       "      <td>-2.994654</td>\n",
       "      <td>0.002748</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3001</th>\n",
       "      <td>H_N_Acetylneuraminic_Acid</td>\n",
       "      <td>Ruminococcaceae Faecalibacterium</td>\n",
       "      <td>0.425969</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.214286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3033</th>\n",
       "      <td>H_N_Acetylneuraminic_Acid</td>\n",
       "      <td>Acidaminococcaceae Phascolarctobacterium_A</td>\n",
       "      <td>0.425969</td>\n",
       "      <td>-0.860087</td>\n",
       "      <td>0.450469</td>\n",
       "      <td>-0.974679</td>\n",
       "      <td>3.075615</td>\n",
       "      <td>0.002101</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3384</th>\n",
       "      <td>H_N_Acetylmuramic_Acid</td>\n",
       "      <td>Rikenellaceae Alistipes_A_871400</td>\n",
       "      <td>0.961082</td>\n",
       "      <td>0.090177</td>\n",
       "      <td>0.607143</td>\n",
       "      <td>-0.900000</td>\n",
       "      <td>2.513325</td>\n",
       "      <td>0.011960</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3490</th>\n",
       "      <td>H_N_Acetylmuramic_Acid</td>\n",
       "      <td>Lachnospiraceae Ruminococcus_B</td>\n",
       "      <td>0.961082</td>\n",
       "      <td>0.545080</td>\n",
       "      <td>-0.630488</td>\n",
       "      <td>0.894427</td>\n",
       "      <td>-2.524016</td>\n",
       "      <td>0.011602</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3758</th>\n",
       "      <td>H_D_Fructose</td>\n",
       "      <td>Lachnospiraceae Acetatifactor</td>\n",
       "      <td>0.237978</td>\n",
       "      <td>1.107210</td>\n",
       "      <td>0.928571</td>\n",
       "      <td>-0.700000</td>\n",
       "      <td>2.904325</td>\n",
       "      <td>0.003680</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3865</th>\n",
       "      <td>H_D_Fructose</td>\n",
       "      <td>Lachnospiraceae Ruminococcus_B</td>\n",
       "      <td>0.237978</td>\n",
       "      <td>0.545080</td>\n",
       "      <td>-0.866921</td>\n",
       "      <td>0.447214</td>\n",
       "      <td>-2.080499</td>\n",
       "      <td>0.037480</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4258</th>\n",
       "      <td>H_Hyocholic_acid</td>\n",
       "      <td>Rikenellaceae Tidjanibacter</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.745422</td>\n",
       "      <td>0.757240</td>\n",
       "      <td>-0.670820</td>\n",
       "      <td>2.080709</td>\n",
       "      <td>0.037461</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4516</th>\n",
       "      <td>H_N_Acetyl_L_Glutamic_acid</td>\n",
       "      <td>Acutalibacteraceae Ruminococcus_E</td>\n",
       "      <td>0.329059</td>\n",
       "      <td>-0.604038</td>\n",
       "      <td>-0.714286</td>\n",
       "      <td>0.974679</td>\n",
       "      <td>-3.549725</td>\n",
       "      <td>0.000386</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4884</th>\n",
       "      <td>H_Paraxanthine</td>\n",
       "      <td>Rikenellaceae Alistipes_A_871400</td>\n",
       "      <td>-0.028029</td>\n",
       "      <td>0.090177</td>\n",
       "      <td>-0.928571</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>-2.537136</td>\n",
       "      <td>0.011176</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5259</th>\n",
       "      <td>H_L_methionine</td>\n",
       "      <td>Rikenellaceae Alistipes_A_871400</td>\n",
       "      <td>0.237978</td>\n",
       "      <td>0.090177</td>\n",
       "      <td>-0.607143</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5266</th>\n",
       "      <td>H_L_methionine</td>\n",
       "      <td>Acutalibacteraceae Ruminococcus_E</td>\n",
       "      <td>0.237978</td>\n",
       "      <td>-0.604038</td>\n",
       "      <td>-0.821429</td>\n",
       "      <td>0.564288</td>\n",
       "      <td>-2.078803</td>\n",
       "      <td>0.037635</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5283</th>\n",
       "      <td>H_L_methionine</td>\n",
       "      <td>Acidaminococcaceae Phascolarctobacterium_A</td>\n",
       "      <td>0.237978</td>\n",
       "      <td>-0.860087</td>\n",
       "      <td>0.558581</td>\n",
       "      <td>-0.872082</td>\n",
       "      <td>2.277618</td>\n",
       "      <td>0.022749</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5752</th>\n",
       "      <td>H_Xanthurenic_acid</td>\n",
       "      <td>Rikenellaceae Alistipes_A_871404</td>\n",
       "      <td>0.090177</td>\n",
       "      <td>1.365377</td>\n",
       "      <td>-0.738769</td>\n",
       "      <td>0.782624</td>\n",
       "      <td>-2.309250</td>\n",
       "      <td>0.020930</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6001</th>\n",
       "      <td>H_Pyridoxic_acid</td>\n",
       "      <td>Ruminococcaceae Faecalibacterium</td>\n",
       "      <td>0.329059</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.321429</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>-2.084764</td>\n",
       "      <td>0.037091</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6033</th>\n",
       "      <td>H_Pyridoxic_acid</td>\n",
       "      <td>Acidaminococcaceae Phascolarctobacterium_A</td>\n",
       "      <td>0.329059</td>\n",
       "      <td>-0.860087</td>\n",
       "      <td>0.252262</td>\n",
       "      <td>-0.974679</td>\n",
       "      <td>2.812966</td>\n",
       "      <td>0.004909</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6865</th>\n",
       "      <td>H_Lactose</td>\n",
       "      <td>Lachnospiraceae Ruminococcus_B</td>\n",
       "      <td>0.329059</td>\n",
       "      <td>0.545080</td>\n",
       "      <td>-0.394055</td>\n",
       "      <td>0.894427</td>\n",
       "      <td>-2.148005</td>\n",
       "      <td>0.031713</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7141</th>\n",
       "      <td>H_Pentanoic_acid</td>\n",
       "      <td>Acutalibacteraceae Ruminococcus_E</td>\n",
       "      <td>0.028029</td>\n",
       "      <td>-0.604038</td>\n",
       "      <td>0.464286</td>\n",
       "      <td>-0.974679</td>\n",
       "      <td>3.095790</td>\n",
       "      <td>0.001963</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7516</th>\n",
       "      <td>H_Glutamyl_leucine</td>\n",
       "      <td>Acutalibacteraceae Ruminococcus_E</td>\n",
       "      <td>-0.162727</td>\n",
       "      <td>-0.604038</td>\n",
       "      <td>0.178571</td>\n",
       "      <td>0.974679</td>\n",
       "      <td>-2.306821</td>\n",
       "      <td>0.021065</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7533</th>\n",
       "      <td>H_Glutamyl_leucine</td>\n",
       "      <td>Acidaminococcaceae Phascolarctobacterium_A</td>\n",
       "      <td>-0.162727</td>\n",
       "      <td>-0.860087</td>\n",
       "      <td>0.558581</td>\n",
       "      <td>-0.820783</td>\n",
       "      <td>2.066891</td>\n",
       "      <td>0.038744</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8365</th>\n",
       "      <td>H_D_pantothenic_acid</td>\n",
       "      <td>Lachnospiraceae Ruminococcus_B</td>\n",
       "      <td>0.660052</td>\n",
       "      <td>0.545080</td>\n",
       "      <td>-0.334947</td>\n",
       "      <td>0.894427</td>\n",
       "      <td>-2.069253</td>\n",
       "      <td>0.038522</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8658</th>\n",
       "      <td>H_Tryptophan</td>\n",
       "      <td>Acidaminococcaceae Phascolarctobacterium_A</td>\n",
       "      <td>-0.028029</td>\n",
       "      <td>-0.860087</td>\n",
       "      <td>0.612637</td>\n",
       "      <td>-0.820783</td>\n",
       "      <td>2.161997</td>\n",
       "      <td>0.030618</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9783</th>\n",
       "      <td>H_Epigallocatechin</td>\n",
       "      <td>Acidaminococcaceae Phascolarctobacterium_A</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.860087</td>\n",
       "      <td>0.558581</td>\n",
       "      <td>-0.974679</td>\n",
       "      <td>3.243601</td>\n",
       "      <td>0.001180</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10240</th>\n",
       "      <td>H_Lenticin</td>\n",
       "      <td>Lachnospiraceae Ruminococcus_B</td>\n",
       "      <td>-0.329059</td>\n",
       "      <td>0.545080</td>\n",
       "      <td>0.315244</td>\n",
       "      <td>-0.894427</td>\n",
       "      <td>2.043812</td>\n",
       "      <td>0.040972</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10252</th>\n",
       "      <td>H_Lenticin</td>\n",
       "      <td>Rikenellaceae Alistipes_A_871404</td>\n",
       "      <td>-0.329059</td>\n",
       "      <td>1.365377</td>\n",
       "      <td>0.306319</td>\n",
       "      <td>-0.894427</td>\n",
       "      <td>2.032404</td>\n",
       "      <td>0.042113</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10908</th>\n",
       "      <td>H_Chlorogenic_acid.1</td>\n",
       "      <td>Acidaminococcaceae Phascolarctobacterium_A</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.860087</td>\n",
       "      <td>0.504525</td>\n",
       "      <td>-0.820783</td>\n",
       "      <td>1.979814</td>\n",
       "      <td>0.047724</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11320</th>\n",
       "      <td>H_Sinapine</td>\n",
       "      <td>Ruminococcaceae Ruminococcus_C_58660</td>\n",
       "      <td>0.660052</td>\n",
       "      <td>0.545080</td>\n",
       "      <td>-0.768408</td>\n",
       "      <td>0.707107</td>\n",
       "      <td>-2.191393</td>\n",
       "      <td>0.028423</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11626</th>\n",
       "      <td>H_Biotin</td>\n",
       "      <td>Ruminococcaceae Faecalibacterium</td>\n",
       "      <td>1.329059</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.714286</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>-2.303041</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11634</th>\n",
       "      <td>H_Biotin</td>\n",
       "      <td>Rikenellaceae Alistipes_A_871400</td>\n",
       "      <td>1.329059</td>\n",
       "      <td>0.090177</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11740</th>\n",
       "      <td>H_Biotin</td>\n",
       "      <td>Lachnospiraceae Ruminococcus_B</td>\n",
       "      <td>1.329059</td>\n",
       "      <td>0.545080</td>\n",
       "      <td>-0.512272</td>\n",
       "      <td>0.894427</td>\n",
       "      <td>-2.320302</td>\n",
       "      <td>0.020325</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12391</th>\n",
       "      <td>H_Catechin</td>\n",
       "      <td>Acutalibacteraceae Ruminococcus_E</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.604038</td>\n",
       "      <td>-0.285714</td>\n",
       "      <td>0.974679</td>\n",
       "      <td>-2.854611</td>\n",
       "      <td>0.004309</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13533</th>\n",
       "      <td>H_Isofraxidin</td>\n",
       "      <td>Acidaminococcaceae Phascolarctobacterium_A</td>\n",
       "      <td>0.527426</td>\n",
       "      <td>-0.860087</td>\n",
       "      <td>0.918956</td>\n",
       "      <td>-0.410391</td>\n",
       "      <td>2.330596</td>\n",
       "      <td>0.019775</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14634</th>\n",
       "      <td>H_Kaempferol</td>\n",
       "      <td>Rikenellaceae Alistipes_A_871400</td>\n",
       "      <td>-0.162727</td>\n",
       "      <td>0.090177</td>\n",
       "      <td>-0.071429</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15391</th>\n",
       "      <td>H_Stercobilin</td>\n",
       "      <td>Acutalibacteraceae Ruminococcus_E</td>\n",
       "      <td>0.028029</td>\n",
       "      <td>-0.604038</td>\n",
       "      <td>-0.214286</td>\n",
       "      <td>-0.974679</td>\n",
       "      <td>2.263921</td>\n",
       "      <td>0.023579</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16134</th>\n",
       "      <td>H_indole_3_propionic_acid</td>\n",
       "      <td>Rikenellaceae Alistipes_A_871400</td>\n",
       "      <td>-0.660052</td>\n",
       "      <td>0.090177</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16501</th>\n",
       "      <td>H_Korseveriline</td>\n",
       "      <td>Ruminococcaceae Faecalibacterium</td>\n",
       "      <td>-0.660052</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.464286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16510</th>\n",
       "      <td>H_Korseveriline</td>\n",
       "      <td>Akkermansiaceae Akkermansia</td>\n",
       "      <td>-0.660052</td>\n",
       "      <td>1.333487</td>\n",
       "      <td>0.821429</td>\n",
       "      <td>-0.718185</td>\n",
       "      <td>2.384549</td>\n",
       "      <td>0.017100</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16533</th>\n",
       "      <td>H_Korseveriline</td>\n",
       "      <td>Acidaminococcaceae Phascolarctobacterium_A</td>\n",
       "      <td>-0.660052</td>\n",
       "      <td>-0.860087</td>\n",
       "      <td>0.360375</td>\n",
       "      <td>-0.974679</td>\n",
       "      <td>2.950940</td>\n",
       "      <td>0.003168</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16876</th>\n",
       "      <td>H_Dihydroxylated_bile_acid</td>\n",
       "      <td>Ruminococcaceae Faecalibacterium</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.714286</td>\n",
       "      <td>0.700000</td>\n",
       "      <td>-2.035945</td>\n",
       "      <td>0.041756</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18751</th>\n",
       "      <td>H_Tomatidine</td>\n",
       "      <td>Ruminococcaceae Faecalibacterium</td>\n",
       "      <td>-0.806180</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.142857</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18783</th>\n",
       "      <td>H_Tomatidine</td>\n",
       "      <td>Acidaminococcaceae Phascolarctobacterium_A</td>\n",
       "      <td>-0.806180</td>\n",
       "      <td>-0.860087</td>\n",
       "      <td>0.846881</td>\n",
       "      <td>-0.974679</td>\n",
       "      <td>3.952877</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19126</th>\n",
       "      <td>H_Tetrahydroxylated_bile_acid</td>\n",
       "      <td>Ruminococcaceae Faecalibacterium</td>\n",
       "      <td>-1.329059</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>-0.900000</td>\n",
       "      <td>2.334257</td>\n",
       "      <td>0.019582</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19158</th>\n",
       "      <td>H_Tetrahydroxylated_bile_acid</td>\n",
       "      <td>Acidaminococcaceae Phascolarctobacterium_A</td>\n",
       "      <td>-1.329059</td>\n",
       "      <td>-0.860087</td>\n",
       "      <td>-0.144150</td>\n",
       "      <td>0.974679</td>\n",
       "      <td>-2.682870</td>\n",
       "      <td>0.007299</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22509</th>\n",
       "      <td>H_Lys-C16:0</td>\n",
       "      <td>Rikenellaceae Alistipes_A_871400</td>\n",
       "      <td>0.329059</td>\n",
       "      <td>0.090177</td>\n",
       "      <td>0.642857</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22615</th>\n",
       "      <td>H_Lys-C16:0</td>\n",
       "      <td>Lachnospiraceae Ruminococcus_B</td>\n",
       "      <td>0.329059</td>\n",
       "      <td>0.545080</td>\n",
       "      <td>-0.394055</td>\n",
       "      <td>0.894427</td>\n",
       "      <td>-2.148005</td>\n",
       "      <td>0.031713</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23251</th>\n",
       "      <td>H_PE(16:1/0:0)_[M+H]+_C21H43N1O7P1</td>\n",
       "      <td>Ruminococcaceae Faecalibacterium</td>\n",
       "      <td>0.162727</td>\n",
       "      <td>-0.090177</td>\n",
       "      <td>-0.321429</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23283</th>\n",
       "      <td>H_PE(16:1/0:0)_[M+H]+_C21H43N1O7P1</td>\n",
       "      <td>Acidaminococcaceae Phascolarctobacterium_A</td>\n",
       "      <td>0.162727</td>\n",
       "      <td>-0.860087</td>\n",
       "      <td>0.666694</td>\n",
       "      <td>-0.974679</td>\n",
       "      <td>3.444518</td>\n",
       "      <td>0.000572</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23634</th>\n",
       "      <td>H_Arg-C16:0</td>\n",
       "      <td>Rikenellaceae Alistipes_A_871400</td>\n",
       "      <td>0.090177</td>\n",
       "      <td>0.090177</td>\n",
       "      <td>0.678571</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23740</th>\n",
       "      <td>H_Arg-C16:0</td>\n",
       "      <td>Lachnospiraceae Ruminococcus_B</td>\n",
       "      <td>0.090177</td>\n",
       "      <td>0.545080</td>\n",
       "      <td>-0.492569</td>\n",
       "      <td>0.894427</td>\n",
       "      <td>-2.289866</td>\n",
       "      <td>0.022029</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24010</th>\n",
       "      <td>H_PC(16:1/0:0)_[M+H]+_C24H49N1O7P1</td>\n",
       "      <td>Akkermansiaceae Akkermansia</td>\n",
       "      <td>-1.107210</td>\n",
       "      <td>1.333487</td>\n",
       "      <td>-0.035714</td>\n",
       "      <td>0.974679</td>\n",
       "      <td>-2.556509</td>\n",
       "      <td>0.010573</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29268</th>\n",
       "      <td>H_Nonhydroxylated_bile_acid</td>\n",
       "      <td>Ruminococcaceae Ruminococcus_D</td>\n",
       "      <td>0.237978</td>\n",
       "      <td>0.300870</td>\n",
       "      <td>-0.684712</td>\n",
       "      <td>0.872082</td>\n",
       "      <td>-2.516831</td>\n",
       "      <td>0.011842</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30127</th>\n",
       "      <td>H_Omega_Hydroxydodecanoate</td>\n",
       "      <td>Rikenellaceae Alistipes_A_871404</td>\n",
       "      <td>0.527426</td>\n",
       "      <td>1.365377</td>\n",
       "      <td>0.612637</td>\n",
       "      <td>-0.782624</td>\n",
       "      <td>2.038322</td>\n",
       "      <td>0.041518</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30760</th>\n",
       "      <td>H_Hydroxydecanoic_acid</td>\n",
       "      <td>Akkermansiaceae Akkermansia</td>\n",
       "      <td>1.107210</td>\n",
       "      <td>1.333487</td>\n",
       "      <td>0.928571</td>\n",
       "      <td>-0.359092</td>\n",
       "      <td>2.336839</td>\n",
       "      <td>0.019448</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     var1  \\\n",
       "766                          H_Melezitose   \n",
       "1141                     H_Phosphocholine   \n",
       "1891                H_Gamma_Valerobetaine   \n",
       "3001            H_N_Acetylneuraminic_Acid   \n",
       "3033            H_N_Acetylneuraminic_Acid   \n",
       "3384               H_N_Acetylmuramic_Acid   \n",
       "3490               H_N_Acetylmuramic_Acid   \n",
       "3758                         H_D_Fructose   \n",
       "3865                         H_D_Fructose   \n",
       "4258                     H_Hyocholic_acid   \n",
       "4516           H_N_Acetyl_L_Glutamic_acid   \n",
       "4884                       H_Paraxanthine   \n",
       "5259                       H_L_methionine   \n",
       "5266                       H_L_methionine   \n",
       "5283                       H_L_methionine   \n",
       "5752                   H_Xanthurenic_acid   \n",
       "6001                     H_Pyridoxic_acid   \n",
       "6033                     H_Pyridoxic_acid   \n",
       "6865                            H_Lactose   \n",
       "7141                     H_Pentanoic_acid   \n",
       "7516                   H_Glutamyl_leucine   \n",
       "7533                   H_Glutamyl_leucine   \n",
       "8365                 H_D_pantothenic_acid   \n",
       "8658                         H_Tryptophan   \n",
       "9783                   H_Epigallocatechin   \n",
       "10240                          H_Lenticin   \n",
       "10252                          H_Lenticin   \n",
       "10908                H_Chlorogenic_acid.1   \n",
       "11320                          H_Sinapine   \n",
       "11626                            H_Biotin   \n",
       "11634                            H_Biotin   \n",
       "11740                            H_Biotin   \n",
       "12391                          H_Catechin   \n",
       "13533                       H_Isofraxidin   \n",
       "14634                        H_Kaempferol   \n",
       "15391                       H_Stercobilin   \n",
       "16134           H_indole_3_propionic_acid   \n",
       "16501                     H_Korseveriline   \n",
       "16510                     H_Korseveriline   \n",
       "16533                     H_Korseveriline   \n",
       "16876          H_Dihydroxylated_bile_acid   \n",
       "18751                        H_Tomatidine   \n",
       "18783                        H_Tomatidine   \n",
       "19126       H_Tetrahydroxylated_bile_acid   \n",
       "19158       H_Tetrahydroxylated_bile_acid   \n",
       "22509                         H_Lys-C16:0   \n",
       "22615                         H_Lys-C16:0   \n",
       "23251  H_PE(16:1/0:0)_[M+H]+_C21H43N1O7P1   \n",
       "23283  H_PE(16:1/0:0)_[M+H]+_C21H43N1O7P1   \n",
       "23634                         H_Arg-C16:0   \n",
       "23740                         H_Arg-C16:0   \n",
       "24010  H_PC(16:1/0:0)_[M+H]+_C24H49N1O7P1   \n",
       "29268         H_Nonhydroxylated_bile_acid   \n",
       "30127          H_Omega_Hydroxydodecanoate   \n",
       "30760              H_Hydroxydecanoic_acid   \n",
       "\n",
       "                                             var2  metabolite_log10p  \\\n",
       "766             Acutalibacteraceae Ruminococcus_E           0.028029   \n",
       "1141            Acutalibacteraceae Ruminococcus_E           0.162727   \n",
       "1891            Acutalibacteraceae Ruminococcus_E           0.527426   \n",
       "3001             Ruminococcaceae Faecalibacterium           0.425969   \n",
       "3033   Acidaminococcaceae Phascolarctobacterium_A           0.425969   \n",
       "3384             Rikenellaceae Alistipes_A_871400           0.961082   \n",
       "3490               Lachnospiraceae Ruminococcus_B           0.961082   \n",
       "3758                Lachnospiraceae Acetatifactor           0.237978   \n",
       "3865               Lachnospiraceae Ruminococcus_B           0.237978   \n",
       "4258                  Rikenellaceae Tidjanibacter          -0.090177   \n",
       "4516            Acutalibacteraceae Ruminococcus_E           0.329059   \n",
       "4884             Rikenellaceae Alistipes_A_871400          -0.028029   \n",
       "5259             Rikenellaceae Alistipes_A_871400           0.237978   \n",
       "5266            Acutalibacteraceae Ruminococcus_E           0.237978   \n",
       "5283   Acidaminococcaceae Phascolarctobacterium_A           0.237978   \n",
       "5752             Rikenellaceae Alistipes_A_871404           0.090177   \n",
       "6001             Ruminococcaceae Faecalibacterium           0.329059   \n",
       "6033   Acidaminococcaceae Phascolarctobacterium_A           0.329059   \n",
       "6865               Lachnospiraceae Ruminococcus_B           0.329059   \n",
       "7141            Acutalibacteraceae Ruminococcus_E           0.028029   \n",
       "7516            Acutalibacteraceae Ruminococcus_E          -0.162727   \n",
       "7533   Acidaminococcaceae Phascolarctobacterium_A          -0.162727   \n",
       "8365               Lachnospiraceae Ruminococcus_B           0.660052   \n",
       "8658   Acidaminococcaceae Phascolarctobacterium_A          -0.028029   \n",
       "9783   Acidaminococcaceae Phascolarctobacterium_A          -0.090177   \n",
       "10240              Lachnospiraceae Ruminococcus_B          -0.329059   \n",
       "10252            Rikenellaceae Alistipes_A_871404          -0.329059   \n",
       "10908  Acidaminococcaceae Phascolarctobacterium_A          -0.090177   \n",
       "11320        Ruminococcaceae Ruminococcus_C_58660           0.660052   \n",
       "11626            Ruminococcaceae Faecalibacterium           1.329059   \n",
       "11634            Rikenellaceae Alistipes_A_871400           1.329059   \n",
       "11740              Lachnospiraceae Ruminococcus_B           1.329059   \n",
       "12391           Acutalibacteraceae Ruminococcus_E          -0.090177   \n",
       "13533  Acidaminococcaceae Phascolarctobacterium_A           0.527426   \n",
       "14634            Rikenellaceae Alistipes_A_871400          -0.162727   \n",
       "15391           Acutalibacteraceae Ruminococcus_E           0.028029   \n",
       "16134            Rikenellaceae Alistipes_A_871400          -0.660052   \n",
       "16501            Ruminococcaceae Faecalibacterium          -0.660052   \n",
       "16510                 Akkermansiaceae Akkermansia          -0.660052   \n",
       "16533  Acidaminococcaceae Phascolarctobacterium_A          -0.660052   \n",
       "16876            Ruminococcaceae Faecalibacterium          -0.000000   \n",
       "18751            Ruminococcaceae Faecalibacterium          -0.806180   \n",
       "18783  Acidaminococcaceae Phascolarctobacterium_A          -0.806180   \n",
       "19126            Ruminococcaceae Faecalibacterium          -1.329059   \n",
       "19158  Acidaminococcaceae Phascolarctobacterium_A          -1.329059   \n",
       "22509            Rikenellaceae Alistipes_A_871400           0.329059   \n",
       "22615              Lachnospiraceae Ruminococcus_B           0.329059   \n",
       "23251            Ruminococcaceae Faecalibacterium           0.162727   \n",
       "23283  Acidaminococcaceae Phascolarctobacterium_A           0.162727   \n",
       "23634            Rikenellaceae Alistipes_A_871400           0.090177   \n",
       "23740              Lachnospiraceae Ruminococcus_B           0.090177   \n",
       "24010                 Akkermansiaceae Akkermansia          -1.107210   \n",
       "29268              Ruminococcaceae Ruminococcus_D           0.237978   \n",
       "30127            Rikenellaceae Alistipes_A_871404           0.527426   \n",
       "30760                 Akkermansiaceae Akkermansia           1.107210   \n",
       "\n",
       "       taxa_log10p  correlations_resp  correlations_nr  fishersz_diff  \\\n",
       "766      -0.604038           0.107143         0.974679      -2.391057   \n",
       "1141     -0.604038          -0.750000         0.974679      -3.638724   \n",
       "1891     -0.604038          -0.392857         0.974679      -2.994654   \n",
       "3001     -0.090177          -0.214286         1.000000      -5.000000   \n",
       "3033     -0.860087           0.450469        -0.974679       3.075615   \n",
       "3384      0.090177           0.607143        -0.900000       2.513325   \n",
       "3490      0.545080          -0.630488         0.894427      -2.524016   \n",
       "3758      1.107210           0.928571        -0.700000       2.904325   \n",
       "3865      0.545080          -0.866921         0.447214      -2.080499   \n",
       "4258     -0.745422           0.757240        -0.670820       2.080709   \n",
       "4516     -0.604038          -0.714286         0.974679      -3.549725   \n",
       "4884      0.090177          -0.928571         0.500000      -2.537136   \n",
       "5259      0.090177          -0.607143        -1.000000       5.000000   \n",
       "5266     -0.604038          -0.821429         0.564288      -2.078803   \n",
       "5283     -0.860087           0.558581        -0.872082       2.277618   \n",
       "5752      1.365377          -0.738769         0.782624      -2.309250   \n",
       "6001     -0.090177          -0.321429         0.900000      -2.084764   \n",
       "6033     -0.860087           0.252262        -0.974679       2.812966   \n",
       "6865      0.545080          -0.394055         0.894427      -2.148005   \n",
       "7141     -0.604038           0.464286        -0.974679       3.095790   \n",
       "7516     -0.604038           0.178571         0.974679      -2.306821   \n",
       "7533     -0.860087           0.558581        -0.820783       2.066891   \n",
       "8365      0.545080          -0.334947         0.894427      -2.069253   \n",
       "8658     -0.860087           0.612637        -0.820783       2.161997   \n",
       "9783     -0.860087           0.558581        -0.974679       3.243601   \n",
       "10240     0.545080           0.315244        -0.894427       2.043812   \n",
       "10252     1.365377           0.306319        -0.894427       2.032404   \n",
       "10908    -0.860087           0.504525        -0.820783       1.979814   \n",
       "11320     0.545080          -0.768408         0.707107      -2.191393   \n",
       "11626    -0.090177          -0.714286         0.800000      -2.303041   \n",
       "11634     0.090177           0.500000        -1.000000       5.000000   \n",
       "11740     0.545080          -0.512272         0.894427      -2.320302   \n",
       "12391    -0.604038          -0.285714         0.974679      -2.854611   \n",
       "13533    -0.860087           0.918956        -0.410391       2.330596   \n",
       "14634     0.090177          -0.071429        -1.000000       5.000000   \n",
       "15391    -0.604038          -0.214286        -0.974679       2.263921   \n",
       "16134     0.090177           0.250000        -1.000000       5.000000   \n",
       "16501    -0.090177          -0.464286         1.000000      -5.000000   \n",
       "16510     1.333487           0.821429        -0.718185       2.384549   \n",
       "16533    -0.860087           0.360375        -0.974679       2.950940   \n",
       "16876    -0.090177          -0.714286         0.700000      -2.035945   \n",
       "18751    -0.090177          -0.142857         1.000000      -5.000000   \n",
       "18783    -0.860087           0.846881        -0.974679       3.952877   \n",
       "19126    -0.090177           0.500000        -0.900000       2.334257   \n",
       "19158    -0.860087          -0.144150         0.974679      -2.682870   \n",
       "22509     0.090177           0.642857        -1.000000       5.000000   \n",
       "22615     0.545080          -0.394055         0.894427      -2.148005   \n",
       "23251    -0.090177          -0.321429         1.000000      -5.000000   \n",
       "23283    -0.860087           0.666694        -0.974679       3.444518   \n",
       "23634     0.090177           0.678571        -1.000000       5.000000   \n",
       "23740     0.545080          -0.492569         0.894427      -2.289866   \n",
       "24010     1.333487          -0.035714         0.974679      -2.556509   \n",
       "29268     0.300870          -0.684712         0.872082      -2.516831   \n",
       "30127     1.365377           0.612637        -0.782624       2.038322   \n",
       "30760     1.333487           0.928571        -0.359092       2.336839   \n",
       "\n",
       "       fz_diff_p Interest  \n",
       "766     0.016800      Yes  \n",
       "1141    0.000274      Yes  \n",
       "1891    0.002748      Yes  \n",
       "3001    0.000000      Yes  \n",
       "3033    0.002101      Yes  \n",
       "3384    0.011960      Yes  \n",
       "3490    0.011602      Yes  \n",
       "3758    0.003680      Yes  \n",
       "3865    0.037480      Yes  \n",
       "4258    0.037461      Yes  \n",
       "4516    0.000386      Yes  \n",
       "4884    0.011176      Yes  \n",
       "5259    0.000000      Yes  \n",
       "5266    0.037635      Yes  \n",
       "5283    0.022749      Yes  \n",
       "5752    0.020930      Yes  \n",
       "6001    0.037091      Yes  \n",
       "6033    0.004909      Yes  \n",
       "6865    0.031713      Yes  \n",
       "7141    0.001963      Yes  \n",
       "7516    0.021065      Yes  \n",
       "7533    0.038744      Yes  \n",
       "8365    0.038522      Yes  \n",
       "8658    0.030618      Yes  \n",
       "9783    0.001180      Yes  \n",
       "10240   0.040972      Yes  \n",
       "10252   0.042113      Yes  \n",
       "10908   0.047724      Yes  \n",
       "11320   0.028423      Yes  \n",
       "11626   0.021277      Yes  \n",
       "11634   0.000000      Yes  \n",
       "11740   0.020325      Yes  \n",
       "12391   0.004309      Yes  \n",
       "13533   0.019775      Yes  \n",
       "14634   0.000000      Yes  \n",
       "15391   0.023579      Yes  \n",
       "16134   0.000000      Yes  \n",
       "16501   0.000000      Yes  \n",
       "16510   0.017100      Yes  \n",
       "16533   0.003168      Yes  \n",
       "16876   0.041756      Yes  \n",
       "18751   0.000000      Yes  \n",
       "18783   0.000077      Yes  \n",
       "19126   0.019582      Yes  \n",
       "19158   0.007299      Yes  \n",
       "22509   0.000000      Yes  \n",
       "22615   0.031713      Yes  \n",
       "23251   0.000000      Yes  \n",
       "23283   0.000572      Yes  \n",
       "23634   0.000000      Yes  \n",
       "23740   0.022029      Yes  \n",
       "24010   0.010573      Yes  \n",
       "29268   0.011842      Yes  \n",
       "30127   0.041518      Yes  \n",
       "30760   0.019448      Yes  "
      ]
     },
     "execution_count": 378,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df_merge.copy()#df_merge[df_merge['metabolite_log10p'] < -1]\n",
    "df['var2'] = df['var2'].map(lambda x: read_taxa(x))\n",
    "df = df[df['Interest'] == 'Yes']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "id": "a514e789-3bd4-4e20-8d36-e3da4b83d733",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium',\n",
       "       'd__Bacteria;p__Firmicutes_C;c__Negativicutes;o__Acidaminococcales;f__Acidaminococcaceae;g__Phascolarctobacterium_A',\n",
       "       'd__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 367,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['var2'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "643b1518-8e68-46f5-8c35-9d78cb472f91",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_Thiamine</th>\n",
       "      <th>H_Melezitose</th>\n",
       "      <th>H_Phosphocholine</th>\n",
       "      <th>H_SN_Glycero_3_Phophocholine</th>\n",
       "      <th>H_Gamma_Valerobetaine</th>\n",
       "      <th>H_N_Acetylneuraminic_acid</th>\n",
       "      <th>H_Pyridoxine</th>\n",
       "      <th>H_N_Acetylneuraminic_Acid</th>\n",
       "      <th>H_N_Acetylmuramic_Acid</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;__;__;__;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.210066</td>\n",
       "      <td>-0.777345</td>\n",
       "      <td>0.224452</td>\n",
       "      <td>0.324467</td>\n",
       "      <td>-0.103097</td>\n",
       "      <td>-0.045994</td>\n",
       "      <td>0.161577</td>\n",
       "      <td>0.811554</td>\n",
       "      <td>0.403052</td>\n",
       "      <td>0.409099</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>-0.057791</td>\n",
       "      <td>-0.128497</td>\n",
       "      <td>0.136301</td>\n",
       "      <td>-0.462694</td>\n",
       "      <td>0.115211</td>\n",
       "      <td>-0.225795</td>\n",
       "      <td>0.385639</td>\n",
       "      <td>0.298280</td>\n",
       "      <td>0.298811</td>\n",
       "      <td>0.226231</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>-0.173576</td>\n",
       "      <td>-0.201427</td>\n",
       "      <td>-1.095070</td>\n",
       "      <td>-1.438525</td>\n",
       "      <td>-1.701619</td>\n",
       "      <td>-0.491732</td>\n",
       "      <td>-1.365669</td>\n",
       "      <td>-0.199607</td>\n",
       "      <td>-0.352783</td>\n",
       "      <td>-0.319787</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>1.017082</td>\n",
       "      <td>2.961867</td>\n",
       "      <td>2.046746</td>\n",
       "      <td>1.047499</td>\n",
       "      <td>2.068942</td>\n",
       "      <td>0.817433</td>\n",
       "      <td>1.913159</td>\n",
       "      <td>0.418271</td>\n",
       "      <td>1.459702</td>\n",
       "      <td>2.211136</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00003</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD007</th>\n",
       "      <td>-0.012481</td>\n",
       "      <td>-1.179298</td>\n",
       "      <td>-0.239259</td>\n",
       "      <td>1.407852</td>\n",
       "      <td>-0.744968</td>\n",
       "      <td>0.504381</td>\n",
       "      <td>-0.191796</td>\n",
       "      <td>-0.651325</td>\n",
       "      <td>0.846103</td>\n",
       "      <td>-0.102527</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD009</th>\n",
       "      <td>0.063274</td>\n",
       "      <td>0.143755</td>\n",
       "      <td>-0.790354</td>\n",
       "      <td>0.736583</td>\n",
       "      <td>0.519140</td>\n",
       "      <td>-0.018743</td>\n",
       "      <td>0.894693</td>\n",
       "      <td>-0.134895</td>\n",
       "      <td>1.285396</td>\n",
       "      <td>0.421072</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>-0.000054</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000036</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD012</th>\n",
       "      <td>-0.400539</td>\n",
       "      <td>0.739819</td>\n",
       "      <td>1.373349</td>\n",
       "      <td>-0.345965</td>\n",
       "      <td>-0.137104</td>\n",
       "      <td>0.132952</td>\n",
       "      <td>0.217476</td>\n",
       "      <td>-0.147902</td>\n",
       "      <td>0.684251</td>\n",
       "      <td>0.441743</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000074</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD017</th>\n",
       "      <td>-0.312795</td>\n",
       "      <td>-0.705903</td>\n",
       "      <td>-0.303385</td>\n",
       "      <td>-0.007038</td>\n",
       "      <td>-0.419213</td>\n",
       "      <td>0.180189</td>\n",
       "      <td>-0.246269</td>\n",
       "      <td>-1.325523</td>\n",
       "      <td>-0.483948</td>\n",
       "      <td>0.191478</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>-0.116967</td>\n",
       "      <td>0.377655</td>\n",
       "      <td>-0.180781</td>\n",
       "      <td>-0.544268</td>\n",
       "      <td>-0.544268</td>\n",
       "      <td>-0.299064</td>\n",
       "      <td>-0.259821</td>\n",
       "      <td>-0.544268</td>\n",
       "      <td>1.130724</td>\n",
       "      <td>0.432946</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD004</th>\n",
       "      <td>-0.084239</td>\n",
       "      <td>1.562695</td>\n",
       "      <td>-1.457749</td>\n",
       "      <td>-1.715716</td>\n",
       "      <td>-0.746018</td>\n",
       "      <td>-0.058798</td>\n",
       "      <td>-0.959669</td>\n",
       "      <td>-0.606696</td>\n",
       "      <td>-0.976076</td>\n",
       "      <td>0.301023</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD011</th>\n",
       "      <td>-0.274656</td>\n",
       "      <td>1.855500</td>\n",
       "      <td>-0.017258</td>\n",
       "      <td>0.535871</td>\n",
       "      <td>-0.219883</td>\n",
       "      <td>0.141015</td>\n",
       "      <td>-0.206757</td>\n",
       "      <td>-0.566620</td>\n",
       "      <td>-0.350967</td>\n",
       "      <td>-0.434815</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>-0.000016</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD012</th>\n",
       "      <td>-0.522389</td>\n",
       "      <td>-2.148315</td>\n",
       "      <td>-0.678315</td>\n",
       "      <td>-0.737065</td>\n",
       "      <td>-1.004191</td>\n",
       "      <td>-0.432072</td>\n",
       "      <td>-1.102651</td>\n",
       "      <td>0.219508</td>\n",
       "      <td>0.110740</td>\n",
       "      <td>-0.651285</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.000035</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows Ã— 463 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          H_Pyridoxamine  H_Thiamine  H_Melezitose  H_Phosphocholine  \\\n",
       "SampleID                                                               \n",
       "OAD001          0.210066   -0.777345      0.224452          0.324467   \n",
       "OAD003         -0.057791   -0.128497      0.136301         -0.462694   \n",
       "OAD004         -0.173576   -0.201427     -1.095070         -1.438525   \n",
       "OAD006          1.017082    2.961867      2.046746          1.047499   \n",
       "OAD007         -0.012481   -1.179298     -0.239259          1.407852   \n",
       "OAD009          0.063274    0.143755     -0.790354          0.736583   \n",
       "OAD012         -0.400539    0.739819      1.373349         -0.345965   \n",
       "OAD017         -0.312795   -0.705903     -0.303385         -0.007038   \n",
       "VAOAD001       -0.116967    0.377655     -0.180781         -0.544268   \n",
       "VAOAD004       -0.084239    1.562695     -1.457749         -1.715716   \n",
       "VAOAD011       -0.274656    1.855500     -0.017258          0.535871   \n",
       "VAOAD012       -0.522389   -2.148315     -0.678315         -0.737065   \n",
       "\n",
       "          H_SN_Glycero_3_Phophocholine  H_Gamma_Valerobetaine  \\\n",
       "SampleID                                                        \n",
       "OAD001                       -0.103097              -0.045994   \n",
       "OAD003                        0.115211              -0.225795   \n",
       "OAD004                       -1.701619              -0.491732   \n",
       "OAD006                        2.068942               0.817433   \n",
       "OAD007                       -0.744968               0.504381   \n",
       "OAD009                        0.519140              -0.018743   \n",
       "OAD012                       -0.137104               0.132952   \n",
       "OAD017                       -0.419213               0.180189   \n",
       "VAOAD001                     -0.544268              -0.299064   \n",
       "VAOAD004                     -0.746018              -0.058798   \n",
       "VAOAD011                     -0.219883               0.141015   \n",
       "VAOAD012                     -1.004191              -0.432072   \n",
       "\n",
       "          H_N_Acetylneuraminic_acid  H_Pyridoxine  H_N_Acetylneuraminic_Acid  \\\n",
       "SampleID                                                                       \n",
       "OAD001                     0.161577      0.811554                   0.403052   \n",
       "OAD003                     0.385639      0.298280                   0.298811   \n",
       "OAD004                    -1.365669     -0.199607                  -0.352783   \n",
       "OAD006                     1.913159      0.418271                   1.459702   \n",
       "OAD007                    -0.191796     -0.651325                   0.846103   \n",
       "OAD009                     0.894693     -0.134895                   1.285396   \n",
       "OAD012                     0.217476     -0.147902                   0.684251   \n",
       "OAD017                    -0.246269     -1.325523                  -0.483948   \n",
       "VAOAD001                  -0.259821     -0.544268                   1.130724   \n",
       "VAOAD004                  -0.959669     -0.606696                  -0.976076   \n",
       "VAOAD011                  -0.206757     -0.566620                  -0.350967   \n",
       "VAOAD012                  -1.102651      0.219508                   0.110740   \n",
       "\n",
       "          H_N_Acetylmuramic_Acid  ...  \\\n",
       "SampleID                          ...   \n",
       "OAD001                  0.409099  ...   \n",
       "OAD003                  0.226231  ...   \n",
       "OAD004                 -0.319787  ...   \n",
       "OAD006                  2.211136  ...   \n",
       "OAD007                 -0.102527  ...   \n",
       "OAD009                  0.421072  ...   \n",
       "OAD012                  0.441743  ...   \n",
       "OAD017                  0.191478  ...   \n",
       "VAOAD001                0.432946  ...   \n",
       "VAOAD004                0.301023  ...   \n",
       "VAOAD011               -0.434815  ...   \n",
       "VAOAD012               -0.651285  ...   \n",
       "\n",
       "          d__Bacteria;p__Firmicutes_A;__;__;__;__  \\\n",
       "SampleID                                            \n",
       "OAD001                                   0.000000   \n",
       "OAD003                                   0.000000   \n",
       "OAD004                                   0.000000   \n",
       "OAD006                                   0.000000   \n",
       "OAD007                                   0.000000   \n",
       "OAD009                                   0.000029   \n",
       "OAD012                                   0.000000   \n",
       "OAD017                                   0.000000   \n",
       "VAOAD001                                 0.000000   \n",
       "VAOAD004                                 0.000000   \n",
       "VAOAD011                                 0.000000   \n",
       "VAOAD012                                 0.000000   \n",
       "\n",
       "          d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "SampleID                                                                                                           \n",
       "OAD001                                             0.000000                                                        \n",
       "OAD003                                             0.000000                                                        \n",
       "OAD004                                             0.000000                                                        \n",
       "OAD006                                             0.000000                                                        \n",
       "OAD007                                             0.000000                                                        \n",
       "OAD009                                            -0.000054                                                        \n",
       "OAD012                                             0.000000                                                        \n",
       "OAD017                                             0.000000                                                        \n",
       "VAOAD001                                           0.000000                                                        \n",
       "VAOAD004                                           0.000000                                                        \n",
       "VAOAD011                                           0.000000                                                        \n",
       "VAOAD012                                           0.000000                                                        \n",
       "\n",
       "          d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "SampleID                                                                    \n",
       "OAD001                                             0.000000                 \n",
       "OAD003                                             0.000000                 \n",
       "OAD004                                             0.000000                 \n",
       "OAD006                                             0.000000                 \n",
       "OAD007                                             0.000000                 \n",
       "OAD009                                             0.000000                 \n",
       "OAD012                                            -0.000074                 \n",
       "OAD017                                             0.000000                 \n",
       "VAOAD001                                           0.000000                 \n",
       "VAOAD004                                           0.000000                 \n",
       "VAOAD011                                           0.000000                 \n",
       "VAOAD012                                           0.000000                 \n",
       "\n",
       "          d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "SampleID                                                                                                            \n",
       "OAD001                                             0.000000                                                         \n",
       "OAD003                                             0.000000                                                         \n",
       "OAD004                                             0.000000                                                         \n",
       "OAD006                                             0.000000                                                         \n",
       "OAD007                                             0.000000                                                         \n",
       "OAD009                                             0.000000                                                         \n",
       "OAD012                                             0.000000                                                         \n",
       "OAD017                                             0.000000                                                         \n",
       "VAOAD001                                           0.000000                                                         \n",
       "VAOAD004                                           0.000000                                                         \n",
       "VAOAD011                                           0.000036                                                         \n",
       "VAOAD012                                           0.000000                                                         \n",
       "\n",
       "          d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "SampleID                                                                                                  \n",
       "OAD001                                             0.000000                                               \n",
       "OAD003                                             0.000000                                               \n",
       "OAD004                                             0.000000                                               \n",
       "OAD006                                             0.000000                                               \n",
       "OAD007                                             0.000000                                               \n",
       "OAD009                                             0.000000                                               \n",
       "OAD012                                             0.000000                                               \n",
       "OAD017                                             0.000000                                               \n",
       "VAOAD001                                           0.000000                                               \n",
       "VAOAD004                                           0.000000                                               \n",
       "VAOAD011                                          -0.000016                                               \n",
       "VAOAD012                                           0.000000                                               \n",
       "\n",
       "          d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "SampleID                                                                                                                            \n",
       "OAD001                                              0.00000                                                                         \n",
       "OAD003                                              0.00000                                                                         \n",
       "OAD004                                              0.00000                                                                         \n",
       "OAD006                                              0.00003                                                                         \n",
       "OAD007                                              0.00000                                                                         \n",
       "OAD009                                              0.00000                                                                         \n",
       "OAD012                                              0.00000                                                                         \n",
       "OAD017                                              0.00000                                                                         \n",
       "VAOAD001                                            0.00000                                                                         \n",
       "VAOAD004                                            0.00000                                                                         \n",
       "VAOAD011                                            0.00000                                                                         \n",
       "VAOAD012                                            0.00000                                                                         \n",
       "\n",
       "          d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "SampleID                                                                                        \n",
       "OAD001                                             0.000000                                     \n",
       "OAD003                                             0.000000                                     \n",
       "OAD004                                             0.000000                                     \n",
       "OAD006                                             0.000000                                     \n",
       "OAD007                                             0.000026                                     \n",
       "OAD009                                             0.000000                                     \n",
       "OAD012                                             0.000000                                     \n",
       "OAD017                                             0.000000                                     \n",
       "VAOAD001                                           0.000000                                     \n",
       "VAOAD004                                           0.000000                                     \n",
       "VAOAD011                                           0.000000                                     \n",
       "VAOAD012                                           0.000000                                     \n",
       "\n",
       "          d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "SampleID                                                                                                                           \n",
       "OAD001                                             0.000000                                                                        \n",
       "OAD003                                             0.000000                                                                        \n",
       "OAD004                                             0.000000                                                                        \n",
       "OAD006                                             0.000000                                                                        \n",
       "OAD007                                             0.000000                                                                        \n",
       "OAD009                                            -0.000036                                                                        \n",
       "OAD012                                             0.000000                                                                        \n",
       "OAD017                                             0.000000                                                                        \n",
       "VAOAD001                                           0.000000                                                                        \n",
       "VAOAD004                                           0.000000                                                                        \n",
       "VAOAD011                                           0.000000                                                                        \n",
       "VAOAD012                                           0.000000                                                                        \n",
       "\n",
       "          d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "SampleID                                                                                     \n",
       "OAD001                                                  0.0                                  \n",
       "OAD003                                                  0.0                                  \n",
       "OAD004                                                  0.0                                  \n",
       "OAD006                                                  0.0                                  \n",
       "OAD007                                                  0.0                                  \n",
       "OAD009                                                  0.0                                  \n",
       "OAD012                                                  0.0                                  \n",
       "OAD017                                                  0.0                                  \n",
       "VAOAD001                                                0.0                                  \n",
       "VAOAD004                                                0.0                                  \n",
       "VAOAD011                                                0.0                                  \n",
       "VAOAD012                                                0.0                                  \n",
       "\n",
       "          d__Bacteria;p__Actinobacteriota;__;__;__;__  \n",
       "SampleID                                               \n",
       "OAD001                                       0.000000  \n",
       "OAD003                                       0.000000  \n",
       "OAD004                                       0.000000  \n",
       "OAD006                                       0.000000  \n",
       "OAD007                                       0.000000  \n",
       "OAD009                                       0.000000  \n",
       "OAD012                                       0.000000  \n",
       "OAD017                                       0.000000  \n",
       "VAOAD001                                     0.000000  \n",
       "VAOAD004                                     0.000000  \n",
       "VAOAD011                                     0.000000  \n",
       "VAOAD012                                    -0.000035  \n",
       "\n",
       "[12 rows x 463 columns]"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 300x300 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 300x300 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 300x300 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 300x300 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 300x300 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 300x300 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 300x300 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 300x300 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 300x300 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plot \n",
    "df_meta_diff = pd.read_csv(path + 'inputs/cutie_df_meta_diff_stool.tsv', sep='\\t', index_col=0)#'SampleID')\n",
    "df_taxa_diff = pd.read_csv(path + 'inputs/cutie_df_taxa_diff_stool.tsv', sep='\\t', index_col=0)#, index_label='SampleID')\n",
    "df_taxa_diff.index = df_taxa_diff.index.map(lambda x: x.split('.')[0].replace('-',''))\n",
    "\n",
    "g='stool'\n",
    "df_map = type_to_df_map[g].set_index('Together')\n",
    "df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "df_map\n",
    "\n",
    "# df_merge_diff = pd.concat([df_meta_diff, df_taxa_diff, df_map],axis=1)#,axis=1)\n",
    "df_merge_diff = pd.concat([df_meta_diff, df_taxa_diff],axis=1)#,axis=1)\n",
    "df_merge_diff = df_merge_diff.dropna()#(subset='WOMAC_P_Response')\n",
    "\n",
    "df = df_merge.copy()\n",
    "df = df[df['metabolite_log10p'] < -1]\n",
    "df = df[df['Interest'] == 'Yes']\n",
    "for i in range(len(df)):\n",
    "    row = df.iloc[i,:]\n",
    "    var1 = row['var1']\n",
    "    var2 = row['var2']\n",
    "\n",
    "    plt.figure(figsize=(3,3))\n",
    "    sns.lmplot(data=df_merge_diff, x=var1, y=var2, hue='WOMAC_P_Response')\n",
    "    plt.ylabel(read_taxa(var2))               \n",
    "    plt.tight_layout()\n",
    "    sns.despine()    \n",
    "    plt.savefig(path + 'outputs/jobs18/corrs_stool_diff/' + 'corr_meta10p_' + var1.replace('/','_') + '_' + read_taxa(var2) + '.pdf')\n",
    "\n",
    "    plt.close()\n",
    "\n",
    "df_merge_diff\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 381,
   "id": "8cced40a-b941-4147-8d78-454412c9ecd2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var1</th>\n",
       "      <th>var2</th>\n",
       "      <th>pvalues</th>\n",
       "      <th>correlations</th>\n",
       "      <th>r2vals</th>\n",
       "      <th>TP_rev_indicators</th>\n",
       "      <th>FP_rev_indicators</th>\n",
       "      <th>extreme_p</th>\n",
       "      <th>extreme_r</th>\n",
       "      <th>p_ratio</th>\n",
       "      <th>r2_ratio</th>\n",
       "      <th>class</th>\n",
       "      <th>reverse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [var1, var2, pvalues, correlations, r2vals, TP_rev_indicators, FP_rev_indicators, extreme_p, extreme_r, p_ratio, r2_ratio, class, reverse]\n",
       "Index: []"
      ]
     },
     "execution_count": 381,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(path + 'outputs/jobs30/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "df = df[df['var1'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "df = df[df['var2'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "# df = df[['var1','var2','correlations']]\n",
    "df = df[df['var1'].isin(int_ba)]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 384,
   "id": "bb8357fd-526f-4a65-85dd-740254449432",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 384,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "int_ba"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "d2c459a7-763e-45b0-be0c-be4fdb5e06f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clostridium\n",
      "d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Clostridium_Q_135853\n",
      "d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Clostridium_Q_134516\n",
      "d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Clostridium_Q_135822\n",
      "d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Acutalibacteraceae;g__Clostridium_A\n",
      "d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;g__Clostridium_T\n",
      "h_Glycocholic_acid\n",
      "H_Glycocholic_acid\n",
      "H_Glycocholate\n",
      "H_Dihydroxylated_bile_acid\n",
      "Pseudomonas\n",
      "d__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Pseudomonadales_650611;f__Pseudomonadaceae;g__Pseudomonas_F\n",
      "H_Glycocholate\n",
      "H_Monohydroxylated_bile_acid\n",
      "Paraprevotella\n",
      "d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Paraprevotella\n",
      "H_Deoxycholate\n",
      "H_Deoxycholic_acid\n",
      "Bacteroides\n",
      "d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H\n",
      "H_Tetrahydroxylated_bile_acid\n",
      "Collinsella\n",
      "d__Bacteria;p__Actinobacteriota;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Collinsella\n",
      "H_Dihydroxylated_bile_acid\n",
      "Parabaceteroides\n",
      "H_Dihydroxylated_bile_acid\n",
      "Akkermansia\n",
      "d__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia\n",
      "H_Deoxycholate\n",
      "Bifidobacterium\n",
      "d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Bifidobacteriaceae;g__Bifidobacterium_388775\n",
      "H_Deoxycholic_acid\n",
      "h_Glycocholic_acid\n",
      "H_Glycocholic_acid\n",
      "H_Dihydroxylated_bile_acid\n",
      "Sutterella\n",
      "d__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Burkholderiales_595427;f__Burkholderiaceae_A_595427;g__Sutterella\n",
      "H_Deoxycholic_acid\n",
      "Veilonella\n",
      "H_Dihydroxylated_bile_acid\n",
      "H_Deoxycholate\n",
      "H_Deoxycholic_acid\n",
      "Faecalibacterium\n",
      "d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium\n",
      "H_Deoxycholate\n",
      "H_Deoxycholic_acid\n",
      "Prevotella\n",
      "d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella\n",
      "H_Deoxycholic_acid\n",
      "H_Glycocholate\n",
      "Christensenella\n",
      "d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Christensenellales;f__;g__\n",
      "d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Christensenellales;f__CAG-74;g__\n",
      "d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Christensenellales;f__Christensenellaceae;g__\n",
      "d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Christensenellales;f__Borkfalkiaceae;g__Borkfalkia\n",
      "d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Christensenellales;f__CAG-74;g__Onthenecus\n",
      "d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Christensenellales;f__CAG-74;g__SFMI01\n",
      "H_Deoxycholic_acid\n",
      "H_Dihydroxylated_bile_acid\n",
      "Oribacterium\n",
      "H_Dihydroxylated_bile_acid\n",
      "Peptostreptococcus\n",
      "H_Dihydroxylated_bile_acid\n",
      "H_Deoxycholate\n",
      "H_Deoxycholic_acid\n",
      "Neisseria\n",
      "d__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Burkholderiales_597441;f__Neisseriaceae_563222;g__Neisseria_563205\n",
      "H_Dihydroxylated_bile_acid\n",
      "Coprobacillus\n",
      "H_Deoxycholic_acid\n"
     ]
    }
   ],
   "source": [
    "# manual mapping for microbemasst of taxa to correlations with bile acids\n",
    "\n",
    "taxa_to_meta_to_match = {\n",
    "    'Clostridium': \n",
    "    {\n",
    "        'Glycocholic': 3,\n",
    "        'Glycocholate': 5,\n",
    "        'Dihydroxylated': -30,\n",
    "    },\n",
    "    'Pseudomonas':\n",
    "    {\n",
    "        'Glycocholate': 33, # or -33?\n",
    "        'Monohydroxylated': 5,\n",
    "    },\n",
    "    'Paraprevotella': \n",
    "    {\n",
    "        'Deoxycholate': 29,\n",
    "        'Deoxycholic': 2,\n",
    "    },\n",
    "    'Bacteroides':\n",
    "    {\n",
    "        'Tetrahydroxylated': -6,\n",
    "    },\n",
    "    'Collinsella':\n",
    "    {\n",
    "        'Dihydroxylated': -8\n",
    "    },\n",
    "    'Parabaceteroides':\n",
    "    {\n",
    "        'Dihydroxylated': 32\n",
    "    },\n",
    "    'Akkermansia':\n",
    "    {\n",
    "        'Glycholic': -4,\n",
    "        'Deoxycholate': 29,\n",
    "    },\n",
    "    'Bifidobacterium':\n",
    "    {\n",
    "        'Deoxycholic': -81,\n",
    "        'Glycocholic': -2,\n",
    "        'Dihydroxylated': -93,\n",
    "    },\n",
    "    'Sutterella':\n",
    "    {\n",
    "        'Deoxycholic': 4\n",
    "    },\n",
    "    'Veilonella':\n",
    "    {\n",
    "        'Dihydroxylated': 9, # or -9?\n",
    "        'Deoxycholate': -9,\n",
    "        'Deoxycholic': -9,\n",
    "    },\n",
    "    'Faecalibacterium':\n",
    "    {\n",
    "        'Deoxycholate': -2,\n",
    "        'Deoxycholic': -2\n",
    "        \n",
    "    },\n",
    "    'Prevotella':\n",
    "    {\n",
    "        'Deoxycholic': 65,\n",
    "        'Glycocholate': 4\n",
    "    },\n",
    "    'Christensenella':\n",
    "    {\n",
    "        'Deoxycholic': 3,\n",
    "        'Dihydroxylated': 4\n",
    "    },\n",
    "    'Oribacterium':\n",
    "    {\n",
    "        'Dihydroxylated': -4,\n",
    "    },\n",
    "    'Peptostreptococcus':\n",
    "    {\n",
    "        'Dihydroxylated': 5,\n",
    "        'Deoxycholate': 5,\n",
    "        'Deoxycholic': 5\n",
    "    },\n",
    "    'Neisseria':\n",
    "    {\n",
    "        'Dihydroxylated': 2\n",
    "    },\n",
    "    'Coprobacillus':\n",
    "    {\n",
    "        'Deoxycholic': 9\n",
    "    }\n",
    "}\n",
    "metas = list(set(df_merge['var1']))\n",
    "taxas =list(set(df_merge['var2']))\n",
    "\n",
    "for t in taxa_to_meta_to_match:\n",
    "    print(t)\n",
    "    meta_to_match = taxa_to_meta_to_match[t]\n",
    "    for y in taxas:\n",
    "        if t in y:\n",
    "            print(y)\n",
    "    for m in meta_to_match:\n",
    "        for x in metas:\n",
    "            if m in x:\n",
    "                print(x)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "082111cb-07ca-416c-94b2-adda8b22a9e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for row in corr\n",
    "# for row in df_merge:\n",
    "df_merge_new = df_merge.copy()\n",
    "df_merge_new['MM'] = 0\n",
    "int_rows = []\n",
    "\n",
    "for i in range(len(df_merge)):\n",
    "    row = df_merge.iloc[i,:]\n",
    "    \n",
    "    for t in taxa_to_meta_to_match:\n",
    "        meta_to_match = taxa_to_meta_to_match[t]\n",
    "        if t in row['var2']:    \n",
    "            for m in meta_to_match:\n",
    "                if m in row['var1']:\n",
    "                    matches = meta_to_match[m]\n",
    "                    row['MM'] = matches\n",
    "                    int_rows.append(row)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "00cf15b4-ece2-49d2-9b16-4fff87be5a02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['H_Glycocholic_acid',\n",
       " 'H_Glycocholic_acid',\n",
       " 'H_Glycocholate',\n",
       " 'H_Glycochenodeoxycholate',\n",
       " 'H_Deoxycholate']"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# int rows\n",
    "int_ba"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "74801029-460c-4a60-9c85-3f7cd790b79d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H_Glycocholic_acid d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;g__Clostridium_T\n"
     ]
    }
   ],
   "source": [
    "row = int_rows[0]\n",
    "print(row['var1'], row['var2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6f3aa88-2e19-42b3-bd79-3e0b66119a1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlations of taxa of interest with metabolites\n",
    "gtr_to_otucorr = {}\n",
    "for g in ['saliva','stool']:\n",
    "    print(g)\n",
    "    gtr_to_otucorr[g] = {}\n",
    "    for time in ['diff','all']:\n",
    "        print(time)\n",
    "        gtr_to_otucorr[g][time] = {}\n",
    "        # grab otu table of differences or of all samples\n",
    "        df_otu = gt_to_otu[g][time]\n",
    "        for comp in ['horizontal', 'vertical']:\n",
    "            if comp == 'vertical':\n",
    "                gr_to_sigotu = to_sigotu[comp]\n",
    "                rvs = ['pre', 'post']\n",
    "            elif comp == 'horizontal':\n",
    "                gr_to_sigotu = to_sigotu[comp]\n",
    "                rvs = ['Response', 'No response']\n",
    "            for rv in rvs: # ['pre', 'post']: # Response No response\n",
    "                print(rv)\n",
    "                sigcorr = []\n",
    "                # grab taxa of interest\n",
    "                taxa_int = gr_to_sigotu[g][rv]\n",
    "    \n",
    "                # merge otu table with metabolites            \n",
    "                df_merge = pd.concat([df_otu.loc[:,taxa_int], gt_to_meta[g][time]], axis=1)\n",
    "                meta = gt_to_meta[g][time].columns\n",
    "                \n",
    "                # look at taxa int and correlate with metabolites\n",
    "                # df_merge \n",
    "                for t in taxa_int:\n",
    "                    for m in meta:\n",
    "                        df_sub = df_merge.loc[:,[t,m]]\n",
    "                        df = type_to_df_map[g].copy()\n",
    "                        if time is 'all':\n",
    "                            df = df.set_index('Together')\n",
    "                            if g == 'stool':\n",
    "                                df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "                        elif time is 'diff':\n",
    "                            df = df[df['Timepoint'] == 'pre']\n",
    "                        df_sub = pd.concat([df_sub, df['WOMAC_P_Response']],axis=1)\n",
    "                        df_sub = df_sub.dropna(subset=[t,m])\n",
    "                        r, p = scipy.stats.spearmanr(df_sub[t].values, df_sub[m].values)\n",
    "                        if p < 0.05:\n",
    "                            sigcorr.append([t,m])\n",
    "                            sns.scatterplot(data=df_sub, x=t, y=m, hue='WOMAC_P_Response')\n",
    "                            sns.despine()\n",
    "                        \n",
    "                            plt.tight_layout()\n",
    "                            plt.savefig(path + 'outputs/' + g_to_job[g + '_adh'] + '/corr_taxa_' + comp + '_' + time + '_' + t.split(';g__')[-1] + '_' + m.replace('/','fslash') + '.pdf')\n",
    "                            plt.close()   \n",
    "                gtr_to_otucorr[g][time][rv] = sigcorr\n",
    "                print(len(sigcorr))\n",
    "           \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f6c73d6-71f4-44ee-ac66-5fe0711fe3f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlations of pathways of interest with metabolites\n",
    "gtr_to_pathcorr = {}\n",
    "for g in ['saliva','stool']:\n",
    "    print(g)\n",
    "    gtr_to_pathcorr[g] = {}\n",
    "    for time in ['diff','all']:\n",
    "        print(time)\n",
    "        gtr_to_pathcorr[g][time] = {}\n",
    "        # grab otu table of differences or of all samples\n",
    "        df_path = gt_to_path[g][time]\n",
    "        for comp in ['horizontal', 'vertical']:\n",
    "            if comp == 'vertical':\n",
    "                gr_to_sigpath = to_sigpath[comp]\n",
    "                rvs = ['pre', 'post']\n",
    "            elif comp == 'horizontal':\n",
    "                gr_to_sigpath = to_sigpath[comp]\n",
    "                rvs = ['Response', 'No response']\n",
    "            for rv in rvs: # ['pre', 'post']: # Response No response\n",
    "                print(rv)\n",
    "                sigcorr = []\n",
    "                # grab taxa of interest\n",
    "                taxa_int = gr_to_sigpath[g][rv]\n",
    "    \n",
    "                # merge otu table with metabolites            \n",
    "                df_merge = pd.concat([df_path.loc[:,taxa_int], gt_to_meta[g][time]], axis=1)\n",
    "                meta = gt_to_meta[g][time].columns\n",
    "                \n",
    "                # look at taxa int and correlate with metabolites\n",
    "                # df_merge \n",
    "                for t in taxa_int:\n",
    "                    for m in meta:\n",
    "                        df_sub = df_merge.loc[:,[t,m]]\n",
    "                        df = type_to_df_map[g].copy()\n",
    "                        if time is 'all':\n",
    "                            df = df.set_index('Together')\n",
    "                            if g == 'stool':\n",
    "                                df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "                        elif time is 'diff':\n",
    "                            df = df[df['Timepoint'] == 'pre']\n",
    "                        df_sub = pd.concat([df_sub, df['WOMAC_P_Response']],axis=1)\n",
    "                        df_sub = df_sub.dropna(subset=[t,m])\n",
    "                        r, p = scipy.stats.spearmanr(df_sub[t].values, df_sub[m].values)\n",
    "                        if p < 0.05:\n",
    "                            sigcorr.append([t,m])\n",
    "                            sns.scatterplot(data=df_sub, x=t, y=m, hue='WOMAC_P_Response')\n",
    "                            sns.despine()\n",
    "                        \n",
    "                            plt.tight_layout()\n",
    "                            plt.savefig(path + 'outputs/' + g_to_job[g + '_adh'] + '/corr_path_' + comp + '_' + time + '_' + t.split(';g__')[-1] + '_' + m.replace('/','fslash') + '.pdf')\n",
    "                            plt.close()   \n",
    "                gtr_to_pathcorr[g][time][rv] = sigcorr\n",
    "                print(len(sigcorr))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5abd2d7-1c78-4e99-a89e-bb864ac5547f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# L Methionine\n",
    "# Kaempferol\n",
    "# Tryptophan\n",
    "# Tyrosine Stool\n",
    "# Usnic acid\n",
    "# Xanthurenic acid\n",
    "# Tryhydroxycholestenoic acid Stool\n",
    "# Isofraxadine\n",
    "int = ['Isofrax', 'Xanth', 'ydroxychol', 'Usnic', 'Tyrosine', 'Tryptophan', 'Kaempferol', 'Methion']\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    for time in ['diff','all']:\n",
    "        for r in ['Response', 'No response']:\n",
    "            test = gtr_to_pathcorr[g][time][r]\n",
    "            for t in test:\n",
    "                p, m = t # unpack pathway, metabolite, from tuple representing sig corr\n",
    "                for i in int:\n",
    "                    if i in m:\n",
    "                        print(t)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f30bf9df-2904-4ca5-b179-4a4eaa5e71f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "int = ['Isofrax', 'Xanth', 'ydroxychol', 'Usnic', 'Tyrosine', 'Tryptophan', 'Kaempferol', 'Methion']\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    for time in ['diff','all']:\n",
    "        for r in ['Response', 'No response']:\n",
    "            test = gtr_to_otucorr[g][time][r]\n",
    "            for t in test:\n",
    "                p, m = t # unpack pathway, metabolite, from tuple representing sig corr\n",
    "                for i in int:\n",
    "                    if i in m:\n",
    "                        print(t)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fdbda6d-641c-47ab-9c83-d392cf0299a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# anaerobic vs aerobic in response\n",
    "# bile acids -> list? saliva and stool of interest \n",
    "#\n",
    "# Bile acids found in Stool: \n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid', 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide']\n",
    "         }\n",
    "\n",
    "mod_ba = {\n",
    "    'stool': [],\n",
    "    'saliva': []\n",
    "}\n",
    "full = list(gt_to_meta[g]['After diet'].columns.values)\n",
    "for g in ['stool','saliva']:\n",
    "    ba = int_ba[g]\n",
    "    for b in ba:\n",
    "        for f in full:\n",
    "            if b in f:\n",
    "                mod_ba[g].append(f)\n",
    "\n",
    "g_to_cutiejob = {\n",
    "    'saliva': 'jobs15',\n",
    "    'stool': 'jobs16'\n",
    "}\n",
    "\n",
    "\n",
    "# from difflib import SequenceMatcher\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    print(g)\n",
    "    n_corr = 0\n",
    "    # import CUITIE results\n",
    "    df_qt = pd.read_csv(path + 'outputs/' + g_to_cutiejob[g] + '/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "    df_qt = df_qt[df_qt['class'] == 'TP']\n",
    "    # iterate through rows\n",
    "    for r in range(len(df_qt)):\n",
    "        meta = df_qt.iloc[r,0]\n",
    "        abun = df_qt.iloc[r,1]\n",
    "        \n",
    "        #for i in int_ba:\n",
    "        #    s = SequenceMatcher(None, meta, i)\n",
    "        #    if s.ratio() > 0.9:\n",
    "        #        print(i)\n",
    "        \n",
    "        #if 'bile' in meta:\n",
    "        #    print(meta)\n",
    "        output = [r]\n",
    "        \n",
    "        for i in int_ba[g]:\n",
    "            if i in meta:\n",
    "                output.append(meta)\n",
    "\n",
    "        for c in ['horizontal','vertical']:\n",
    "            if c == 'vertical':\n",
    "                queries = ['pre', 'post']\n",
    "            if c == 'horizontal':\n",
    "                queries = ['Response','No response']\n",
    "            output.append(c)\n",
    "            for r in queries:\n",
    "                int_taxa = to_sigotu[c][g][r]\n",
    "                for i in int_taxa:\n",
    "                    if i in abun:\n",
    "                        output.append('taxa')\n",
    "                        output.append(i)\n",
    "                int_path = to_sigpath[c][g][r]\n",
    "                for i in int_path:\n",
    "                    if i in abun:\n",
    "                        output.append('path')\n",
    "                        output.append(i)\n",
    "        #output.append()\n",
    "        #output.append()\n",
    "        if len(output) > 5:\n",
    "            # print(output)\n",
    "            n_corr += 1\n",
    "\n",
    "    print(n_corr)\n",
    "\n",
    "# pathway prediction of response vs non response @ baseline\n",
    "# specific taxa at baseline\n",
    "\n",
    "\n",
    "# pathways of interest\n",
    "# ex inflammation in non responders\n",
    "# heatmap of differential pathways in responders/nonresponders \n",
    "# and correlation of those pathways with bacteria\n",
    "# send list of upregulated pathways \n",
    "\n",
    "int_taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76722aff-f034-4fff-a4da-c63dd1fd75ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotations for pathways\n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid', 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide']\n",
    "         }\n",
    "\n",
    "for g in ['saliva','stool']:\n",
    "    # int_ba[g] for metabolite (row) annotations\n",
    "    #df_map = type_to_df_map[g].set_index('Together')\n",
    "    #df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    #df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    #df_meta = pd.concat([gt_to_meta[g]['After diet'],df_map], axis=1)\n",
    "    df_meta = gt_to_meta[g]['After diet']\n",
    "    df_meta_R = df_meta.T\n",
    "    df_meta_R['var_type'] = df_meta_R.index.map(lambda x: 'BA' if x in mod_ba[g] else 'NonBA')\n",
    "    df_meta_R = df_meta_R['var_type'].reset_index()\n",
    "    # df_meta_R['index'] = df_meta_R['index'].str.lower()\n",
    "    df_meta_R = df_meta_R.sort_values(by='index',ascending=True)\n",
    "    df_meta_R.to_csv(path + 'inputs/row_' + g + '.tsv', sep='\\t')\n",
    "    \n",
    "    # dcgr_to_sig for column (taxa) annotations\n",
    "    df_abun = gt_to_otu[g]['post']\n",
    "    df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_abun_R = df_abun.T\n",
    "    df_abun_R['var_type'] = df_abun_R.index.map(lambda x: 'Sig' if x in dcgr_to_sig['taxa']['rvnr'][g]['Response'] else 'Nonsig')\n",
    "    df_abun_R = df_abun_R['var_type'].reset_index()\n",
    "    #df_abun_R['index'] = df_abun_R['index'].str.lower()\n",
    "    df_abun_R = df_abun_R.sort_values(by='index',ascending=True)\n",
    "    df_abun_R.to_csv(path + 'inputs/col_' + g + '.tsv', sep='\\t')\n",
    "\n",
    "df_abun_R.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fbf3e56-2d5f-41fa-8ed7-718d92ed6400",
   "metadata": {},
   "outputs": [],
   "source": [
    "# play testing mediation\n",
    "g = 'stool'\n",
    "df_M = pd.read_csv(path + 'inputs/cutie_df_meta_stool.tsv', sep='\\t', index_col=0)\n",
    "df_M.iloc[:,:-1].to_csv(path + 'inputs/mediation_meta_stool_post_test.txt', sep='\\t')\n",
    "\n",
    "df_Y = df_meta_paired[['VAS_Pt_diff', 'WOMAC_pain_diff']]\n",
    "df_Y.to_csv(path + 'inputs/mediation_WMdiff_test.txt', sep='\\t')\n",
    "\n",
    "df_X = pd.read_csv(path + 'inputs/cutie_df_taxa_stool.tsv', sep='\\t', index_col=0)\n",
    "df_X.index = df_X.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "df_X.to_csv(path + 'inputs/mediation_taxa_stool_post_test.txt', sep='\\t')\n",
    "df_X.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11360b5e-dd8b-4857-b205-4dd8c5c685b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for corr heatmaps \n",
    "for g in ['saliva','stool']:\n",
    "    df_abun = pd.concat([gt_to_otu[g]['post'], gt_to_path[g]['post']],axis=1)\n",
    "    df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_map = type_to_df_map[g].set_index('Together')\n",
    "    df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_meta = pd.concat([gt_to_meta[g]['After diet'],df_map], axis=1)\n",
    "\n",
    "    df_quant_R = df_abun.copy()\n",
    "    #df_quant_R = pd.concat([df_abun, df_meta],axis=1)# .copy() #([df_quant, df_asv100, df_path_top100], axis=1)\n",
    "    # df_quant_R = df_quant_R.drop(['WOMAC_P_Response'],axis=1)\n",
    "    # df_quant_R.to_csv(path + 'inputs/df_quant_' + g + '.tsv', sep='\\t')\n",
    "    df_quant_R = df_quant_R.T\n",
    "    # this is in the order they were entered i.e. \n",
    "    # [df_olink, df_fa, df_acpa_fecal, df_acpa_plasma, df_plasma, df_rbfa, df_mb]\n",
    "     # ['rbfa']*len(df_rbfa.columns) + \\  \n",
    "    df_quant_R['var_type'] = ['Taxa']*len(gt_to_otu[g]['post'].columns) + \\\n",
    "                             ['Pathways']*len(gt_to_path[g]['post'].columns) # + \\\n",
    "                             # ['Metabolite']*len(gt_to_meta[g]['After diet'].columns)\n",
    "    # df_quant_R = df_quant_R[~df_quant_R['var_type'].isin(['Metagenomic_ASVs','Metagenomic_Pathways'])]\n",
    "    df_quant_R = df_quant_R['var_type'].reset_index()\n",
    "    df_quant_R['index'] = df_quant_R['index'].str.lower()\n",
    "    df_quant_R = df_quant_R.sort_values(by='index',ascending=True)\n",
    "    df_quant_R.to_csv(path + 'inputs/df_quant_R_corr_labels_' + g + '.tsv', sep='\\t')\n",
    "    print(len(df_quant_R))\n",
    "df_quant_R.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2707d857-ee9c-466f-983c-3d94600478d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4 way paired boxplot trial\n",
    "g = 'saliva'\n",
    "df_alpha = g_to_dfd[g + '_adh']['alpha']\n",
    "\n",
    "df_map = type_to_df_map[g].copy()\n",
    "df_map = df_map.set_index('Together')\n",
    "df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "\n",
    "df_merge = pd.concat([df_alpha, df_map], axis=1)\n",
    "df_merge = df_merge.dropna()\n",
    "\n",
    "sns.boxplot(x='Timepoint', y='shannon_entropy',\n",
    "            hue='WOMAC_P_Response', palette=[\"r\", \"g\"],\n",
    "            data=df_merge)\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a341d51b-1db7-4c98-b4e8-65a312f999e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test if each sub correlation is sig\n",
    "m = 'H_Nonhydroxylated_bile_acid'\n",
    "y = 'WOMAC_pain_diff'\n",
    "y = 'VAS_Pt_diff'\n",
    "x = 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;__'\n",
    "\n",
    "df = df_plot[[x,y,m]]\n",
    "\n",
    "df_corr = df.dropna()\n",
    "print('Y on X')\n",
    "print(scipy.stats.pearsonr(df_corr[y].values, df_corr[x].values))\n",
    "print('M on X')\n",
    "print(scipy.stats.pearsonr(df_corr[m].values, df_corr[x].values))\n",
    "\n",
    "df['meta'] = df[m].apply(lambda x: 'high' if -1.25 <= x <= 0.25 else 'low')\n",
    "\n",
    "# do each linear model separately\n",
    "df_corr = df[df['meta'] == 'low']\n",
    "df_corr = df_corr.dropna()\n",
    "print(scipy.stats.pearsonr(df_corr[y].values, df_corr[x].values))\n",
    "\n",
    "df_corr = df[df['meta'] == 'high']\n",
    "df_corr = df_corr.dropna()\n",
    "print(scipy.stats.pearsonr(df_corr[y].values, df_corr[x].values))\n",
    "# df_corr\n",
    "df.columns = ['Lachno','Pain','H_Nonhydroxylated_BA', 'meta']\n",
    "df.to_csv(path + 'inputs/df_med_R.tsv', sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "101b41e6-d1b0-4d18-a49c-7c338f54078d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "061973e1-2fae-4d91-8847-ea6ce3450a9e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ad3a4bc-37a9-4ae4-9075-376b97b35337",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2faeb2e-f734-4d0c-b0a4-8d9b5421a6fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniq_to_taxa['OAD005']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6219d1d-abce-4661-8592-bca518dab9ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(path + 'outputs/Qiime2_saliva_adh_withOAD007/level-6.csv', index_col=0)\n",
    "# df = pd.read_csv(path + 'outputs/Qiime2_stool_adh/level-6.csv', index_col=0)\n",
    "df = pd.read_csv(path + 'outputs/Qiime2_saliva_adh/level-6.csv', index_col=0)\n",
    "keep = []\n",
    "for x in df.columns:\n",
    "    if 'd__' in x:\n",
    "        keep.append(x)\n",
    "df = df[keep]\n",
    "\n",
    "\n",
    "# df = gt_to_otu['saliva']['post']\n",
    "\n",
    "uniques = []\n",
    "from collections import defaultdict\n",
    "uniq_to_taxa = defaultdict(list)\n",
    "for t in df.columns:\n",
    "    if df[t].nunique() == 2:\n",
    "        x = df[t]\n",
    "        x = x[x != 0]\n",
    "        uniques.append(x.index[0])\n",
    "        uniq_to_taxa[x.index[0]].append(t)\n",
    "pd.Series(uniques).value_counts()\n",
    "# df = df.astype(bool).sum(axis=1)\n",
    "\n",
    "#df = df.loc['OAD-007.post.saliva',:]\n",
    "#df = df[df > 0]\n",
    "\n",
    "#len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edd2c113-64a7-43e1-826d-ee86582ec9db",
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f23883a-75bd-40da-8c0b-135e04f1f4b4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
