{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3f62dc73-df18-4548-bc0d-aea2a125fef0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "import pandas as pd\n",
    "import scipy.stats\n",
    "import statsmodels.stats.multitest\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "\n",
    "# disable warnings, use w caution\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "# project specific libs\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import pathlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a8b42930-bcde-48f3-9753-8779e1aa11fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# project specific path\n",
    "path = '/Users/KevinBu/Desktop/clemente_lab/Projects/oa/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a3013747-2e0e-412a-b719-b2a6a28afbf2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>Separate</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Together</th>\n",
       "      <th>ContactEmail</th>\n",
       "      <th>ContactName</th>\n",
       "      <th>PrimaryInvestigator</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>RawDataNotes</th>\n",
       "      <th>...</th>\n",
       "      <th>broccoli</th>\n",
       "      <th>Garbanzo_beans</th>\n",
       "      <th>pork</th>\n",
       "      <th>beef</th>\n",
       "      <th>burger</th>\n",
       "      <th>Total_omega3</th>\n",
       "      <th>Adherence_omega3</th>\n",
       "      <th>Total_omega6</th>\n",
       "      <th>Adherence_omega6</th>\n",
       "      <th>Total_o3_o6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>TTCAGTTCGTTA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-001.pre.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>CGGCCAGAAGCA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>post</td>\n",
       "      <td>OAD-001.post.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001.post.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>72.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>131.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>GACGTTAAGAAT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-003.pre.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>53.2</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>TCGCTACAGATG</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>post</td>\n",
       "      <td>OAD-003.post.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003.post.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>108.0</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>171.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>ATGGGACCTTCA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-004.pre.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-004.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>2.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>28.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>35.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 188 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   BarcodeSequence    LinkerPrimerSequence Separate Timepoint  \\\n",
       "#SampleID                                                                       \n",
       "OAD-001.pre.stool     TTCAGTTCGTTA  CCGGACTACHVGGGTWTCTAAT      All       pre   \n",
       "OAD-001.post.stool    CGGCCAGAAGCA  CCGGACTACHVGGGTWTCTAAT      All      post   \n",
       "OAD-003.pre.stool     GACGTTAAGAAT  CCGGACTACHVGGGTWTCTAAT      All       pre   \n",
       "OAD-003.post.stool    TCGCTACAGATG  CCGGACTACHVGGGTWTCTAAT      All      post   \n",
       "OAD-004.pre.stool     ATGGGACCTTCA  CCGGACTACHVGGGTWTCTAAT      All       pre   \n",
       "\n",
       "                                            Together  \\\n",
       "#SampleID                                              \n",
       "OAD-001.pre.stool    OAD-001.pre.stool.guma.plate313   \n",
       "OAD-001.post.stool  OAD-001.post.stool.guma.plate313   \n",
       "OAD-003.pre.stool    OAD-003.pre.stool.guma.plate313   \n",
       "OAD-003.post.stool  OAD-003.post.stool.guma.plate313   \n",
       "OAD-004.pre.stool    OAD-004.pre.stool.guma.plate313   \n",
       "\n",
       "                                    ContactEmail    ContactName  \\\n",
       "#SampleID                                                         \n",
       "OAD-001.pre.stool   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-001.post.stool  rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-003.pre.stool   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-003.post.stool  rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-004.pre.stool   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "\n",
       "                   PrimaryInvestigator Cohort  \\\n",
       "#SampleID                                       \n",
       "OAD-001.pre.stool           Jose Scher  NonVA   \n",
       "OAD-001.post.stool          Jose Scher  NonVA   \n",
       "OAD-003.pre.stool           Jose Scher  NonVA   \n",
       "OAD-003.post.stool          Jose Scher  NonVA   \n",
       "OAD-004.pre.stool           Jose Scher  NonVA   \n",
       "\n",
       "                                        RawDataNotes  ... broccoli  \\\n",
       "#SampleID                                             ...            \n",
       "OAD-001.pre.stool    OAD-001.pre.stool.guma.plate313  ...      0.0   \n",
       "OAD-001.post.stool  OAD-001.post.stool.guma.plate313  ...      4.0   \n",
       "OAD-003.pre.stool    OAD-003.pre.stool.guma.plate313  ...      0.0   \n",
       "OAD-003.post.stool  OAD-003.post.stool.guma.plate313  ...      4.0   \n",
       "OAD-004.pre.stool    OAD-004.pre.stool.guma.plate313  ...      2.8   \n",
       "\n",
       "                   Garbanzo_beans pork  beef burger Total_omega3  \\\n",
       "#SampleID                                                          \n",
       "OAD-001.pre.stool             0.0  0.0   0.0    0.0          NaN   \n",
       "OAD-001.post.stool            0.0  0.0   0.0    0.0         48.0   \n",
       "OAD-003.pre.stool             0.0  0.0  14.0    0.0         16.8   \n",
       "OAD-003.post.stool            0.0  0.0   0.0    0.0         48.0   \n",
       "OAD-004.pre.stool             0.0  0.0   0.0    0.0          2.8   \n",
       "\n",
       "                   Adherence_omega3 Total_omega6 Adherence_omega6 Total_o3_o6  \n",
       "#SampleID                                                                      \n",
       "OAD-001.pre.stool     Low adherence          NaN    Low adherence         0.0  \n",
       "OAD-001.post.stool    Low adherence         72.0    Low adherence       131.0  \n",
       "OAD-003.pre.stool     Low adherence         53.2   High adherence        75.0  \n",
       "OAD-003.post.stool    Low adherence        108.0   High adherence       171.0  \n",
       "OAD-004.pre.stool     Low adherence         28.0    Low adherence        35.8  \n",
       "\n",
       "[5 rows x 188 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# from AC Q2 run of merged saliva stool\n",
    "df_map = pd.read_csv(path + 'inputs/Qiime2_0/qiime_mapping_file.tsv', sep='\\t', index_col=0)\n",
    "q2_row = df_map.loc['#q2:types',:]\n",
    "df_map = df_map.drop('#q2:types')\n",
    "\n",
    "# change index so it matches metadata file\n",
    "df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "\n",
    "# drop MOC and elution buffer\n",
    "df_map = df_map.drop(['MOC.320','elutionbuffer.plate313'])\n",
    "\n",
    "# grab metadata\n",
    "df_meta = pd.read_csv(path + 'inputs/Metadata_OA.csv')\n",
    "\n",
    "# rename 'Run_ID_Saliva' to be correct\n",
    "df_meta['Timepoints'] = df_meta['Timepoints'].apply(lambda x: 'pre' if x == '0' else 'post')\n",
    "df_meta['Patient_ID'] = df_meta['Patient_ID'].apply(lambda x: x[:-3])  \n",
    "df_meta['Study_ID'] = df_meta['Study_ID'].apply(lambda x: x.split('_')[0][-3:]) \n",
    "\n",
    "# create per sample type mapping files\n",
    "type_to_ST = {'saliva':'Saliva','stool':'fecal'}\n",
    "type_to_df_map = {}\n",
    "\n",
    "# split into specimen type\n",
    "for t in type_to_ST:\n",
    "    # subset on specimen type\n",
    "    df_map_type = df_map[df_map['SpecimenType'] == type_to_ST[t]]\n",
    "\n",
    "    # as to not overwrite df meta\n",
    "    df_meta_type = df_meta.copy()\n",
    "\n",
    "    # create new sample ID for specimen type and set as index\n",
    "    df_meta_type['#SampleID'] = df_meta['Patient_ID'] + '-' + df_meta['Study_ID'] + '.' + df_meta['Timepoints'] + '.' + t\n",
    "    df_meta_type = df_meta_type.set_index('#SampleID')\n",
    "\n",
    "    # create full mapping file\n",
    "    df_map_type = pd.concat([df_map_type, df_meta_type],axis=1)\n",
    "\n",
    "    # use only sequenced samples\n",
    "    df_map_type = df_map_type.dropna(how='any',subset='BarcodeSequence')\n",
    "\n",
    "    # drop all na columns\n",
    "    df_map_type = df_map_type.dropna(how='all',axis=1)\n",
    "\n",
    "    # drop VAD OA 015 because misdx with PsA not OA\n",
    "    if t == 'saliva':\n",
    "        df_map_type = df_map_type.drop(['VAOAD-015.pre.saliva','VAOAD-015.post.saliva'])\n",
    "    if t == 'stool':\n",
    "        df_map_type = df_map_type.drop(['VAOAD-015.pre.stool','VAOAD-015.post.stool'])\n",
    "\n",
    "    # populate dict of mapping files\n",
    "    type_to_df_map[t] = df_map_type\n",
    "\n",
    "    # export for q2\n",
    "    df_q2_type = pd.concat([q2_row.to_frame().T, df_map_type])\n",
    "    df_q2_type.index.name = '#SampleID'\n",
    "    df_q2_type.iloc[0,:] = 'categorical'\n",
    "    df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + t + '.tsv', sep='\\t')\n",
    "    df_q2_type = df_q2_type[df_q2_type['Adherece_antiinflam'].isin(['Moderate adherence', 'High adherence','categorical'])]\n",
    "    df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh.tsv', sep='\\t')\n",
    "    df_q2_type = df_q2_type[df_q2_type['Adherece_antiinflam'].isin(['Moderate adherence', 'High adherence','categorical'])]\n",
    "    df_q2_resp = df_q2_type[df_q2_type['WOMAC_P_Response'].isin(['categorical', 'Response'])]\n",
    "    df_q2_resp.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_response.tsv', sep='\\t')\n",
    "    df_q2_nonresp = df_q2_type[df_q2_type['WOMAC_P_Response'].isin(['categorical', 'No response'])]\n",
    "    df_q2_nonresp.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_noresponse.tsv', sep='\\t')\n",
    "\n",
    "type_to_df_map['stool'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "95e4ef90-0416-4803-a5b2-d730031630c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n",
      "['VAS_Pt', 'VAS_overall', 'WOMAC_pain', 'WOMAC_stiffness', 'WOMAC_activity', 'WOMAC_total', 'Pain_DETECT', 'CES_D', 'Helplesness', 'Magnification', 'Rumination', 'PCS_EN', 'Sleep_distrubance', 'PASE_walk', 'PASE_light', 'BMI']\n",
      "32\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var</th>\n",
       "      <th>effect</th>\n",
       "      <th>pval</th>\n",
       "      <th>stat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VAS_Pt</td>\n",
       "      <td>198.5</td>\n",
       "      <td>0.008316</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VAS_overall</td>\n",
       "      <td>168.5</td>\n",
       "      <td>0.131314</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_pain</td>\n",
       "      <td>170.0</td>\n",
       "      <td>0.017900</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_stiffness</td>\n",
       "      <td>172.5</td>\n",
       "      <td>0.013052</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_activity</td>\n",
       "      <td>164.5</td>\n",
       "      <td>0.032592</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               var  effect      pval stat\n",
       "0           VAS_Pt   198.5  0.008316  MWU\n",
       "0      VAS_overall   168.5  0.131314  MWU\n",
       "0       WOMAC_pain   170.0  0.017900  MWU\n",
       "0  WOMAC_stiffness   172.5  0.013052  MWU\n",
       "0   WOMAC_activity   164.5  0.032592  MWU"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###\n",
    "# Hypothesis 1: There will be a measurable difference in WOMAC pain response scores and \n",
    "# other outcomes from baseline to after the dietary intervention.\n",
    "###\n",
    "\n",
    "# outcome variables\n",
    "outcomes = ['VAS_Pt', 'VAS_overall', 'WOMAC_pain', 'WOMAC_stiffness', 'WOMAC_activity', 'WOMAC_total', 'Pain_DETECT', \n",
    "            'CES_D', 'Helplesness', 'Magnification', 'Rumination', 'PCS_EN', 'Sleep_distrubance', 'PASE_walk', 'PASE_light', \n",
    "            #'PASE_gardening', # Where did this go? gardening_improve is binary\n",
    "            'BMI']\n",
    "\n",
    "\n",
    "# hypothesis 1\n",
    "# create a new df_meta\n",
    "df_meta = pd.read_csv(path + 'inputs/Metadata_OA.csv')\n",
    "\n",
    "# rename 'Run_ID_Saliva' to be correct\n",
    "df_meta['Timepoints'] = df_meta['Timepoints'].apply(lambda x: 'pre' if x == '0' else 'post')\n",
    "df_meta['Patient_ID'] = df_meta['Patient_ID'].apply(lambda x: x[:-3])  \n",
    "df_meta['Study_ID'] = df_meta['Study_ID'].apply(lambda x: x.split('_')[0][-3:]) \n",
    "df_meta['#SampleID'] = df_meta['Patient_ID'] + '-' + df_meta['Study_ID'] + '.' + df_meta['Timepoints'] + '.stool'\n",
    "df_meta = df_meta.set_index('#SampleID')\n",
    "\n",
    "# convert % to floats for calculations down the road\n",
    "bin = []\n",
    "cont = []\n",
    "df_paired_os = []\n",
    "for w in outcomes:\n",
    "    df_w = df_meta[w]\n",
    "    if df_w.nunique() > 2: # do spearman\n",
    "        df_meta[w] = df_meta[w].astype(str).str.replace('%','').astype(float).values\n",
    "        cont.append(w)\n",
    "    else:\n",
    "        bin.append(w)\n",
    "\n",
    "    # compute difference and store it\n",
    "    df_md = df_meta.copy()\n",
    "    df_md['SubjectID'] = df_md['Patient_ID'] + df_md['Study_ID']\n",
    "    \n",
    "    # first drop unpaired samples\n",
    "    s_remove = []\n",
    "    for s in list(df_md['SubjectID'].values):\n",
    "        if len(df_md[df_md['SubjectID'] == s]) != 2:\n",
    "            s_remove.append(s)\n",
    "    df_md = df_md.loc[~df_md['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "    \n",
    "    # set vars\n",
    "    group_var = 'Timepoints'\n",
    "    pair_var = 'SubjectID'\n",
    "    groups = ['pre','post']\n",
    "    \n",
    "    # get paired per indiv pair\n",
    "    pair_to_diff = {}\n",
    "    for p in list(df_md[pair_var].values):\n",
    "        df = df_md[df_md[pair_var] == p]\n",
    "        t0 = float(df[df[group_var] == groups[0]][w].values)\n",
    "        tf = float(df[df[group_var] == groups[1]][w].values)\n",
    "        pair_to_diff[p] = tf - t0\n",
    "    \n",
    "    df_paired_o = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[w + '_diff'])\n",
    "    df_paired_os.append(df_paired_o)\n",
    "\n",
    "df_meta_paired = pd.concat([*df_paired_os], axis=1)    \n",
    "\n",
    "print(bin)\n",
    "print(cont)\n",
    "\n",
    "# split into all and mod high only\n",
    "a = 'modhigh'\n",
    "job = 'jobs02'\n",
    "df_meta = df_meta[df_meta['Adherece_antiinflam'].isin(['Moderate adherence', 'High adherence'])]\n",
    "        \n",
    "print(len(df_meta))\n",
    "\n",
    "df_results = pd.DataFrame(columns=['var','effect','pval','stat'])\n",
    "# do post treatment vals of binary vars differ from pre treatment 'unpaired'\n",
    "for b in bin:\n",
    "    ct_table_ind=pd.crosstab(df_meta[\"Timepoints\"],df_meta[b])\n",
    "    chi2_stat, p, dof, expected = scipy.stats.chi2_contingency(ct_table_ind)\n",
    "    row=pd.DataFrame.from_dict({'var': [b],'effect':[chi2_stat],'pval':[p],'stat':['chi2']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "# fishers exact\n",
    "for b in bin:\n",
    "    ct_table_ind=pd.crosstab(df_meta[\"Timepoints\"],df_meta[b])\n",
    "    fisher, p = scipy.stats.fisher_exact(ct_table_ind)\n",
    "    row=pd.DataFrame.from_dict({'var': [b],'effect':[t],'pval':[p],'stat':['fisher']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "    \n",
    "# do post treatment vals of continuous vars differ from pre treatment unpaired\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    try:\n",
    "        W,p = scipy.stats.mannwhitneyu(x=df_pre[c].values,y=df_post[c].values, nan_policy='omit')\n",
    "    except:\n",
    "        W,p = 0, 1\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['mwu']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v')\n",
    "        sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette='dark:grey', hue=None, orient='v')\n",
    "    \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/mwu_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['MWU']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "    \n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    t,p = scipy.stats.ttest_ind(a=df_pre[c].values,b=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['ttest']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v')\n",
    "        sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette='dark:grey', hue=None, orient='v')\n",
    "    \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/tt_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['ttest']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "# unpaired and then paired\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    W,p = scipy.stats.wilcoxon(x=df_pre[c].values,y=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['WSR']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v')\n",
    "        sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette='dark:grey', hue=None, orient='v')\n",
    "    \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/wsr_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['WSR']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    t,p = scipy.stats.ttest_rel(a=df_pre[c].values,b=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['pairedt']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v')\n",
    "        sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette='dark:grey', hue=None, orient='v')\n",
    "    \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/pairedt_' + c  + '.pdf')\n",
    "        plt.close()          \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['pairedt']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "df_results.to_csv(path + 'outputs/' + job + '/outcome_testing.tsv', sep='\\t')\n",
    "df_results.head()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "298b4824-3c24-485e-8705-5daf4a85d6d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unweighted_Unifrac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.6672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>0.2837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>0.4237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>0.3257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>0.1022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unweighted_Unifrac\n",
       "OAD001              0.6672\n",
       "OAD003              0.2837\n",
       "OAD004              0.4237\n",
       "OAD005              0.3257\n",
       "OAD006              0.1022"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###\n",
    "# Hypothesis 2: There will be an association between oral and gut microbiome and pain outcomes\n",
    "###\n",
    "# construct alpha, beta and paired alpha dataframes\n",
    "g_to_dfd = {}\n",
    "# g_test = ['stool','saliva_adh', 'stool_adh', 'saliva']\n",
    "g_test = ['saliva_adh', 'stool_adh']\n",
    "\n",
    "for g in g_test:\n",
    "    # maps diversity type to dataframe\n",
    "    g_to_dfd[g] = {}\n",
    "\n",
    "    # get alpha diversities\n",
    "    df_alpha = pd.read_csv(path + 'outputs/Qiime2_' + g + '/metadata.tsv', sep='\\t', index_col=0)\n",
    "    df_alpha = df_alpha.drop('#q2:types')\n",
    "    df_alpha['SubjectID'] = df_alpha['Patient_ID'] + df_alpha['Study_ID']\n",
    "    df_alpha = df_alpha[['SubjectID', 'Timepoints', 'shannon_entropy']]\n",
    "    g_to_dfd[g]['alpha'] = df_alpha\n",
    "\n",
    "    # get paired alpha div, first drop unpaired samples\n",
    "    s_remove = []\n",
    "    for s in list(df_alpha['SubjectID'].values):\n",
    "        if len(df_alpha[df_alpha['SubjectID'] == s]) != 2:\n",
    "            s_remove.append(s)\n",
    "    df_alpha = df_alpha.loc[~df_alpha['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "    \n",
    "    # set vars\n",
    "    alpha_metric = 'shannon_entropy'\n",
    "    group_var = 'Timepoints'\n",
    "    pair_var = 'SubjectID'\n",
    "    groups = ['pre','post']\n",
    "    \n",
    "    # get paired per indiv pair\n",
    "    pair_to_diff = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        df = df_alpha[df_alpha[pair_var] == p]\n",
    "        alpha_0 = float(df[df[group_var] == groups[0]][alpha_metric].values)\n",
    "        alpha_1 = float(df[df[group_var] == groups[1]][alpha_metric].values)\n",
    "        pair_to_diff[p] = alpha_1 - alpha_0\n",
    "    \n",
    "    df_paired_alpha = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[alpha_metric + '_diff'])\n",
    "    g_to_dfd[g]['paired_alpha'] = df_paired_alpha\n",
    "\n",
    "    # get beta div\n",
    "    df_beta = pd.read_csv(path + 'outputs/Qiime2_' + g + '/core_metrics_results/distance-matrix.tsv',\n",
    "                              sep='\\t', index_col=0)\n",
    "        \n",
    "    # grab twin to pair dict\n",
    "    pair_to_ids = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        df = df_alpha[df_alpha[pair_var] == p]\n",
    "        id_0 = str(df[df[group_var] == groups[0]].index.values[0])\n",
    "        id_1 = str(df[df[group_var] == groups[1]].index.values[0])\n",
    "        pair_to_ids[p] = (id_0, id_1)\n",
    "    \n",
    "    # get distances for each twin pair per beta div matrix    \n",
    "    pair_to_dist = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        id_0, id_1 = pair_to_ids[p]\n",
    "        pair_to_dist[p] = df_beta.loc[id_0, id_1]\n",
    "    \n",
    "    df_paired_beta = pd.DataFrame.from_dict(pair_to_dist, orient='index', columns=['Unweighted_Unifrac'])\n",
    "    g_to_dfd[g]['paired_beta'] = df_paired_beta\n",
    "\n",
    "g_to_dfd['stool_adh']['paired_beta'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9d4f38ff-41d6-4f31-94ab-69879de56109",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>div</th>\n",
       "      <th>outcome</th>\n",
       "      <th>statistic</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.232031</td>\n",
       "      <td>0.630022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>1.153857</td>\n",
       "      <td>0.788413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.756449</td>\n",
       "      <td>0.456450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.232031</td>\n",
       "      <td>0.630022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.756449</td>\n",
       "      <td>0.980536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.182238</td>\n",
       "      <td>0.856866</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        group                  div           outcome statistic  test_stat  \\\n",
       "0    metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.232031   \n",
       "1  saliva_adh                alpha  WOMAC_P_Response       mwu   1.153857   \n",
       "2  saliva_adh                alpha  WOMAC_P_Response        tt   0.756449   \n",
       "3    metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.232031   \n",
       "4   stool_adh                alpha  WOMAC_P_Response       mwu   0.756449   \n",
       "5   stool_adh                alpha  WOMAC_P_Response        tt   0.182238   \n",
       "\n",
       "       pval  \n",
       "0  0.630022  \n",
       "1  0.788413  \n",
       "2  0.456450  \n",
       "3  0.630022  \n",
       "4  0.980536  \n",
       "5  0.856866  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Testing for each sample type for (1) all and (2) high adh only\n",
    "# (A) Chisq of quartiles with adherence \n",
    "# (B) MWU/TT unpaired of outcomes against distance\n",
    "\n",
    "def chisq_of_df_cols(df, c1, c2):\n",
    "    groupsizes = df.groupby([c1, c2]).size()\n",
    "    ctsum = groupsizes.unstack(c1)\n",
    "    # fillna(0) is necessary to remove any NAs which will cause exceptions\n",
    "    return(scipy.stats.chi2_contingency(ctsum.fillna(0)))\n",
    "\n",
    "d_to_metric = {\n",
    "    'alpha': 'shannon_entropy',\n",
    "}\n",
    "group_var = 'Adherece_antiinflam'\n",
    "\n",
    "dfd_to_merge = {}\n",
    "g_test = ['saliva_adh', 'stool_adh']\n",
    "\n",
    "for g in g_test:\n",
    "    dfd_to_merge[g] = {}\n",
    "\n",
    "gs = []\n",
    "ds = []\n",
    "os = []\n",
    "stats = []\n",
    "ts = []\n",
    "ps = []\n",
    "\n",
    "arr = [gs,ds,os,stats,ts,ps]\n",
    "\n",
    "def append_results(arr, val):\n",
    "    for a,v in zip(arr,val):\n",
    "        a.append(v)\n",
    "    return arr\n",
    "\n",
    "# for each sample type, grab relevant mapping file\n",
    "# g_test = saliva, saliva_adh, stool, etc.\n",
    "for g in g_test:\n",
    "    for d in d_to_metric:\n",
    "        # grab relevant diversity df\n",
    "        df_div = g_to_dfd[g][d]\n",
    "\n",
    "        if d == 'alpha':\n",
    "            # df_div = df_div[df_div['Timepoints'] == 'pre']\n",
    "            # df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            # df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div['shannon_entropy'] = df_div['shannon_entropy'].astype(float)\n",
    "\n",
    "        # merge with df of metadata var        \n",
    "        # grab relevant sample IDs\n",
    "        # g = saliva_adh\n",
    "        idx = 'Run_ID_' + g.split('_')[0].capitalize()\n",
    "        df_meta_sub = df_meta.dropna(subset=idx)\n",
    "        df_meta_sub = df_meta_sub.set_index(idx)\n",
    "        df_merge = pd.concat([df_meta_sub,df_div],axis=1)\n",
    "\n",
    "        # subset on adh only\n",
    "        df_merge = df_merge[df_merge[group_var].isin(['Moderate adherence', 'High adherence'])]\n",
    "\n",
    "        # test association of div with outcomes\n",
    "        for o in ['WOMAC_P_Response']: #  outcomes:                    \n",
    "            # test association of adherence with pain outcomes\n",
    "            div_metric = d_to_metric[d]\n",
    "            x,p,dof,ef = chisq_of_df_cols(df_merge, group_var, o)\n",
    "            arr = append_results(arr, ['metadata',group_var,o,'chisq',x,p])\n",
    "            \n",
    "            ax = sns.boxplot(data=df_merge, x=o, y=div_metric)\n",
    "            sns.swarmplot(data=df_merge, x=o, y=div_metric, palette='dark:grey')\n",
    "            sns.despine()\n",
    "        \n",
    "            plt.tight_layout()\n",
    "            plt.savefig(path + 'outputs/jobs06/rvnr_nondiff_' + g + '_' + o + '_' + d + '.pdf')\n",
    "            plt.close()          \n",
    "        \n",
    "            u, p = scipy.stats.mannwhitneyu(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'mwu',t,p])\n",
    "\n",
    "            t, p = scipy.stats.ttest_ind(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'tt',t,p])\n",
    "\n",
    "        # save results\n",
    "        dfd_to_merge[g][d] = df_merge\n",
    "\n",
    "        # export to Q2\n",
    "        # df_q2_type = df_merge.set_index(['Together'])\n",
    "        df_q2_type = df_merge.copy()\n",
    "        q2_row = pd.Series(data=['categorical' for i in range(len(df_merge.columns))], \n",
    "                           index=list(df_merge.columns.values), dtype=str, name='#q2:types')\n",
    "        df_q2_type = pd.concat([q2_row.to_frame().T, df_q2_type])\n",
    "        df_q2_type.index.name = '#SampleID'\n",
    "        df_q2_type.index = df_q2_type.index.map(lambda x: x.split('.guma')[0])\n",
    "        # df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + d + '_' + g + 'aggregate_outcomes.tsv', sep='\\t')\n",
    "\n",
    "df_results = pd.DataFrame.from_dict({\n",
    "    'group': gs,\n",
    "    'div': ds,\n",
    "    'outcome': os,\n",
    "    'statistic': stats,\n",
    "    'test_stat': ts,\n",
    "    'pval': ps\n",
    "})\n",
    "# df_results.to_csv(path + 'outputs/df_results_aggregate.tsv', sep='\\t')\n",
    "# df_results.head()\n",
    "# df_results[df_results['pval'] < 0.05]\n",
    "df_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "22f26972-623f-459c-93fd-d37eb65d17eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>div</th>\n",
       "      <th>outcome</th>\n",
       "      <th>statistic</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.182238</td>\n",
       "      <td>0.343434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>-0.594256</td>\n",
       "      <td>0.565542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>-0.594256</td>\n",
       "      <td>0.638889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.769556</td>\n",
       "      <td>0.459355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.769556</td>\n",
       "      <td>0.431818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>1.289488</td>\n",
       "      <td>0.226254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>1.289488</td>\n",
       "      <td>0.876263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>-1.033580</td>\n",
       "      <td>0.325681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>-1.033580</td>\n",
       "      <td>0.755051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.565329</td>\n",
       "      <td>0.584310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.565329</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.682542</td>\n",
       "      <td>0.510397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.682542</td>\n",
       "      <td>0.876263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.155526</td>\n",
       "      <td>0.879502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.155526</td>\n",
       "      <td>0.267677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>1.182922</td>\n",
       "      <td>0.264200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         group           div           outcome statistic  test_stat      pval\n",
       "0    stool_adh     pre_alpha  WOMAC_P_Response       mwu   0.182238  0.343434\n",
       "1    stool_adh     pre_alpha  WOMAC_P_Response        tt  -0.594256  0.565542\n",
       "2    stool_adh    post_alpha  WOMAC_P_Response       mwu  -0.594256  0.638889\n",
       "3    stool_adh    post_alpha  WOMAC_P_Response        tt   0.769556  0.459355\n",
       "4    stool_adh  paired_alpha  WOMAC_P_Response       mwu   0.769556  0.431818\n",
       "5    stool_adh  paired_alpha  WOMAC_P_Response        tt   1.289488  0.226254\n",
       "6    stool_adh   paired_beta  WOMAC_P_Response       mwu   1.289488  0.876263\n",
       "7    stool_adh   paired_beta  WOMAC_P_Response        tt  -1.033580  0.325681\n",
       "8   saliva_adh     pre_alpha  WOMAC_P_Response       mwu  -1.033580  0.755051\n",
       "9   saliva_adh     pre_alpha  WOMAC_P_Response        tt   0.565329  0.584310\n",
       "10  saliva_adh    post_alpha  WOMAC_P_Response       mwu   0.565329  1.000000\n",
       "11  saliva_adh    post_alpha  WOMAC_P_Response        tt   0.682542  0.510397\n",
       "12  saliva_adh  paired_alpha  WOMAC_P_Response       mwu   0.682542  0.876263\n",
       "13  saliva_adh  paired_alpha  WOMAC_P_Response        tt   0.155526  0.879502\n",
       "14  saliva_adh   paired_beta  WOMAC_P_Response       mwu   0.155526  0.267677\n",
       "15  saliva_adh   paired_beta  WOMAC_P_Response        tt   1.182922  0.264200"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# expand previous to all data, paired and unpaired diversities\n",
    "d_to_metric = {\n",
    "    'pre_alpha': 'shannon_entropy',\n",
    "    'post_alpha': 'shannon_entropy',\n",
    "    'paired_alpha': 'shannon_entropy_diff',\n",
    "    'paired_beta': 'Unweighted_Unifrac',\n",
    "}\n",
    "group_var = 'Adherece_antiinflam'\n",
    "\n",
    "dfd_to_merge = {}\n",
    "g_test = ['stool_adh','saliva_adh']\n",
    "for g in g_test:\n",
    "    dfd_to_merge[g] = {}\n",
    "\n",
    "gs = []\n",
    "ds = []\n",
    "os = []\n",
    "stats = []\n",
    "ts = []\n",
    "ps = []\n",
    "\n",
    "arr = [gs,ds,os,stats,ts,ps]\n",
    "\n",
    "def append_results(arr, val):\n",
    "    for a,v in zip(arr,val):\n",
    "        a.append(v)\n",
    "    return arr\n",
    "\n",
    "# for each sample type, grab relevant mapping file\n",
    "# g_test = saliva, saliva_adh, stool, etc.\n",
    "d_to_map = {}\n",
    "\n",
    "for g in g_test:\n",
    "    # drop duplicates so you have sample mapping to adh\n",
    "    df_map_sub = type_to_df_map[g.split('_')[0]]\n",
    "    df_map_sub.index = df_map_sub.index.map(lambda x: x.split('.')[0].replace('-',''))\n",
    "    df_map_sub = df_map_sub.dropna(how='any',subset=group_var,axis=0)           \n",
    "    \n",
    "    # figure out which samples to keep\n",
    "    # i.e. samples that have a pre and post time point\n",
    "    keep = []\n",
    "    for i in list(df_map_sub.index.values):\n",
    "        if len(df_map_sub.loc[i,:]) == 2:\n",
    "            keep.append(i)\n",
    "\n",
    "    # get unique entires in sorted order\n",
    "    # at this point we are only concerned with differences in values, \n",
    "    # as we've dropped samples with only one endpoint val\n",
    "    save = []\n",
    "    [save.append(x) for x in keep if x not in save]\n",
    "    df_map_sub = df_map_sub.loc[save,:]\n",
    "\n",
    "    # this double populates as OAD001 is an index twice, so the diff fills to both the pre and post col\n",
    "    for o in outcomes:\n",
    "        df_map_sub[o + '_diff'] = df_map_sub[df_map_sub['Timepoint'] == 'post'][o] - df_map_sub[df_map_sub['Timepoint'] == 'pre'][o] \n",
    "\n",
    "    # here we keep only the pre, but everything is identical b/w pre and post\n",
    "    df_dropdup = df_map_sub[~df_map_sub.index.duplicated(keep='first')]\n",
    "    d_to_map[g] = df_dropdup\n",
    "    \n",
    "    for d in d_to_metric:\n",
    "        # grab relevant diversity df\n",
    "        if d == 'pre_alpha' or d == 'post_alpha':\n",
    "            df_div = g_to_dfd[g]['alpha']\n",
    "        else:\n",
    "            df_div = g_to_dfd[g][d]\n",
    "        div_metric = d_to_metric[d]\n",
    "\n",
    "        # when associating alpha div vs outcomes, look at if starting adiv predicts outcome\n",
    "        if d == 'pre_alpha':\n",
    "            df_div = df_div[df_div['Timepoints'] == 'pre']\n",
    "            df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div = df_div.astype(float)\n",
    "\n",
    "        if d == 'post_alpha':\n",
    "            df_div = df_div[df_div['Timepoints'] == 'post']\n",
    "            df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div = df_div.astype(float)\n",
    "\n",
    "        # merge with df of metadata var        \n",
    "        df_merge = pd.concat([df_dropdup,df_div],axis=1)\n",
    "\n",
    "        # drop na in barcodes if samples not sequenced both pre and post\n",
    "        df_merge = df_merge.dropna(how='any',subset='BarcodeSequence')\n",
    "        \n",
    "        # test association of div with outcomes\n",
    "        for o in ['WOMAC_P_Response']:                     \n",
    "            # test association of adherence with pain outcomes            \n",
    "            ax = sns.boxplot(data=df_merge, y=o, x=div_metric)\n",
    "            sns.swarmplot(data=df_merge, y=o, x=div_metric, palette='dark:grey')\n",
    "            sns.despine()\n",
    "        \n",
    "            plt.tight_layout()\n",
    "            plt.savefig(path + 'outputs/jobs06/WOMAC_barplot_' + g + '_' + o + '_' + d + '.pdf')\n",
    "            plt.close()          \n",
    "\n",
    "            # MWU and ttest\n",
    "            u, p = scipy.stats.mannwhitneyu(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'mwu',t,p])\n",
    "\n",
    "            t, p = scipy.stats.ttest_ind(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'tt',t,p])\n",
    "         \n",
    "        # save results\n",
    "        dfd_to_merge[g][d] = df_merge\n",
    "\n",
    "        # export to Q2\n",
    "        df_q2_type = df_merge.set_index(['Together'])\n",
    "        q2_row = pd.Series(data=['categorical' for i in range(len(df_merge.columns))], \n",
    "                           index=list(df_merge.columns.values), dtype=str, name='#q2:types')\n",
    "        df_q2_type = pd.concat([q2_row.to_frame().T, df_q2_type])\n",
    "        df_q2_type.index.name = '#SampleID'\n",
    "        df_q2_type.index = df_q2_type.index.map(lambda x: x.split('.guma')[0])\n",
    "        # df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + d + '_' + g + '_outcomes.tsv', sep='\\t')\n",
    "\n",
    "df_results = pd.DataFrame.from_dict({\n",
    "    'group': gs,\n",
    "    'div': ds,\n",
    "    'outcome': os,\n",
    "    'statistic': stats,\n",
    "    'test_stat': ts,\n",
    "    'pval': ps\n",
    "})\n",
    "# df_results.to_csv(path + 'outputs/df_results_diff.tsv', sep='\\t')\n",
    "\n",
    "# df_results.head()\n",
    "# the div==alpha results test whether pre-alpha div state associates with pain outcome changes (differences) quartiles\n",
    "# the div==paired_alpha and paired_beta test whether the alpha and betas change in a similar way with the pain outcome\n",
    "print(len(df_results[df_results['pval'] < 0.05]))\n",
    "df_results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5db737b8-b788-4d7a-86a6-39e82ac57eb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_adh\n",
      "Response\n",
      "WilcoxonResult(statistic=4.0, pvalue=0.109375)\n",
      "No response\n",
      "WilcoxonResult(statistic=7.0, pvalue=1.0)\n",
      "saliva_adh\n",
      "Response\n",
      "WilcoxonResult(statistic=9.0, pvalue=0.46875)\n",
      "No response\n",
      "WilcoxonResult(statistic=5.0, pvalue=0.625)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>shannon_entropy_diff</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>Separate</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Together</th>\n",
       "      <th>ContactEmail</th>\n",
       "      <th>ContactName</th>\n",
       "      <th>PrimaryInvestigator</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>...</th>\n",
       "      <th>Pain_DETECT_diff</th>\n",
       "      <th>CES_D_diff</th>\n",
       "      <th>Helplesness_diff</th>\n",
       "      <th>Magnification_diff</th>\n",
       "      <th>Rumination_diff</th>\n",
       "      <th>PCS_EN_diff</th>\n",
       "      <th>Sleep_distrubance_diff</th>\n",
       "      <th>PASE_walk_diff</th>\n",
       "      <th>PASE_light_diff</th>\n",
       "      <th>BMI_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>-0.116436</td>\n",
       "      <td>AGGACAAACTAT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>Text</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-001.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>5.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.51</td>\n",
       "      <td>14.29</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>-0.327749</td>\n",
       "      <td>TCTTCCTAAAGT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-003.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-5.5</td>\n",
       "      <td>-14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>0.866474</td>\n",
       "      <td>CTTGTGCGACAA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-004.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.264032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>0.007278</td>\n",
       "      <td>CTTACACTGCTT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-001.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>5.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.081446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD012</th>\n",
       "      <td>0.535614</td>\n",
       "      <td>GCGTGGTCATTA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-012.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-1.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.72</td>\n",
       "      <td>-1.072664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TTGGACGTCCAC</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-009.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.139974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows Ã— 205 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          shannon_entropy_diff BarcodeSequence    LinkerPrimerSequence  \\\n",
       "OAD001               -0.116436    AGGACAAACTAT  CCGGACTACHVGGGTWTCTAAT   \n",
       "OAD003               -0.327749    TCTTCCTAAAGT  CCGGACTACHVGGGTWTCTAAT   \n",
       "OAD004                0.866474    CTTGTGCGACAA  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD001              0.007278    CTTACACTGCTT  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD012              0.535614    GCGTGGTCATTA  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD009                   NaN    TTGGACGTCCAC  CCGGACTACHVGGGTWTCTAAT   \n",
       "\n",
       "         Separate Timepoint              Together  \\\n",
       "OAD001       Text       pre    OAD-001.pre.saliva   \n",
       "OAD003        All       pre    OAD-003.pre.saliva   \n",
       "OAD004        All       pre    OAD-004.pre.saliva   \n",
       "VAOAD001      All       pre  VAOAD-001.pre.saliva   \n",
       "VAOAD012      All       pre  VAOAD-012.pre.saliva   \n",
       "VAOAD009      All       pre  VAOAD-009.pre.saliva   \n",
       "\n",
       "                          ContactEmail    ContactName PrimaryInvestigator  \\\n",
       "OAD001    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "OAD003    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "OAD004    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD001  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD012  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD009  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "\n",
       "         Cohort  ... Pain_DETECT_diff CES_D_diff Helplesness_diff  \\\n",
       "OAD001    NonVA  ...             -2.0       -1.0             -2.5   \n",
       "OAD003    NonVA  ...             -4.5        1.0             -1.0   \n",
       "OAD004    NonVA  ...             -1.0        3.0             -1.0   \n",
       "VAOAD001     VA  ...             -0.5        5.5              0.0   \n",
       "VAOAD012     VA  ...             -1.5        0.0              2.0   \n",
       "VAOAD009     VA  ...             -4.0        2.0              1.0   \n",
       "\n",
       "         Magnification_diff Rumination_diff PCS_EN_diff  \\\n",
       "OAD001                  5.5             1.0         4.0   \n",
       "OAD003                 -1.0            -2.0        -5.5   \n",
       "OAD004                 -1.0            -1.0        -3.0   \n",
       "VAOAD001                0.0             0.0         0.0   \n",
       "VAOAD012               -2.0            -0.5        -0.5   \n",
       "VAOAD009                0.0             0.5         1.5   \n",
       "\n",
       "         Sleep_distrubance_diff PASE_walk_diff PASE_light_diff  BMI_diff  \n",
       "OAD001                      9.0          11.51           14.29       NaN  \n",
       "OAD003                    -14.0           0.00            0.00       NaN  \n",
       "OAD004                      0.0           0.00            0.00  3.264032  \n",
       "VAOAD001                   -2.5           0.00            0.00 -0.081446  \n",
       "VAOAD012                   -1.5           0.00            6.72 -1.072664  \n",
       "VAOAD009                    2.5           0.00            0.00 -0.139974  \n",
       "\n",
       "[6 rows x 205 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test if the differences in diversity are significantly different from 0 for paired alpha\n",
    "for g in g_test:\n",
    "    print(g)\n",
    "    for d in ['paired_alpha']:\n",
    "        df_div = g_to_dfd[g][d] \n",
    "        df_map = d_to_map[g]\n",
    "\n",
    "        df_merge = pd.concat([df_div, df_map], axis=1)\n",
    "        for o in ['Response', 'No response']:\n",
    "            df = df_merge[df_merge['WOMAC_P_Response'] == o]\n",
    "            print(o)\n",
    "            print(scipy.stats.wilcoxon(df['shannon_entropy_diff'], nan_policy='omit'))\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bef8c05d-208b-4b13-bac7-01bf00e4e80e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_adh\n",
      "All\n",
      "paired alpha wilcoxon\n",
      "33.0 0.414306640625\n",
      "Response\n",
      "paired alpha wilcoxon\n",
      "4.0 0.109375\n",
      "No response\n",
      "paired alpha wilcoxon\n",
      "7.0 1.0\n",
      "saliva_adh\n",
      "All\n",
      "paired alpha wilcoxon\n",
      "26.0 0.19091796875\n",
      "Response\n",
      "paired alpha wilcoxon\n",
      "9.0 0.46875\n",
      "No response\n",
      "paired alpha wilcoxon\n",
      "5.0 0.625\n"
     ]
    }
   ],
   "source": [
    "# Alpha div paired\n",
    "# paired beta, comparing intra-indiv difference pre_post to inter pre and inter post\n",
    "g_test = ['stool_adh','saliva_adh']\n",
    "subgroups = ['All', 'Response', 'No response']\n",
    "\n",
    "for g in g_test: #subgroups:\n",
    "    print(g)\n",
    "    df_a = pd.read_csv(path + 'outputs/Qiime2_' + g + '/metadata.tsv', sep='\\t', index_col=0)\n",
    "    df_a = df_a.drop('#q2:types')\n",
    "\n",
    "    for sg in subgroups:\n",
    "        print(sg)\n",
    "        if sg == 'All':\n",
    "            df_alpha = df_a.copy()\n",
    "        else:\n",
    "            df_alpha = df_a[df_a['WOMAC_P_Response'] == sg]\n",
    "        \n",
    "        df_alpha['SubjectID'] = df_alpha['Patient_ID'] + df_alpha['Study_ID']\n",
    "        df_alpha = df_alpha[['SubjectID', 'Timepoints', 'shannon_entropy']]\n",
    "        \n",
    "        # drop unpaired samples\n",
    "        s_remove = []\n",
    "        for s in list(df_alpha['SubjectID'].values):\n",
    "            if len(df_alpha[df_alpha['SubjectID'] == s]) != 2:\n",
    "                s_remove.append(s)\n",
    "        df_alpha = df_alpha.loc[~df_alpha['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "        \n",
    "        # set vars\n",
    "        alpha_metric = 'shannon_entropy'\n",
    "        group_var = 'Timepoints'\n",
    "        pair_var = 'SubjectID'\n",
    "        groups = ['pre','post']\n",
    "        \n",
    "        # get paired per indiv pair\n",
    "        pair_to_diff = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            df = df_alpha[df_alpha[pair_var] == p]\n",
    "            alpha_0 = float(df[df[group_var] == groups[0]][alpha_metric].values)\n",
    "            alpha_1 = float(df[df[group_var] == groups[1]][alpha_metric].values)\n",
    "            pair_to_diff[p] = alpha_0 - alpha_1\n",
    "        \n",
    "        df_paired_alpha = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[alpha_metric + '_diff'])\n",
    "        \n",
    "        # one-sided t-test, n.s.; RA-UA values \n",
    "        # t, p = scipy.stats.ttest_1samp(df_paired_alpha[alpha_metric + '_diff'],popmean=0)\n",
    "    \n",
    "        print('paired alpha wilcoxon')\n",
    "        s, p = scipy.stats.wilcoxon(df_paired_alpha[alpha_metric + '_diff'])\n",
    "        print(s, p)\n",
    "        \n",
    "        # separate\n",
    "        df_alpha[alpha_metric] = df_alpha[alpha_metric].map(lambda x: float(x))\n",
    "        ax = sns.boxplot(data=df_alpha, x=group_var, y=alpha_metric)\n",
    "        sns.swarmplot(data=df_alpha, x=group_var, y=alpha_metric, palette='dark:grey')\n",
    "        sns.despine()\n",
    "    \n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/Qiime2_' + g + '/' + sg + '_alpha.pdf')\n",
    "        plt.close()          \n",
    "    \n",
    "        # now do beta\n",
    "        df_beta = pd.read_csv(path + 'outputs/Qiime2_' + g + '/core_metrics_results/distance-matrix.tsv',\n",
    "                              sep='\\t', index_col=0)\n",
    "        \n",
    "        # set vars\n",
    "        alpha_metric = 'shannon_entropy'\n",
    "        group_var = 'Timepoints'\n",
    "        pair_var = 'SubjectID'\n",
    "        groups = ['pre','post']\n",
    "        g0, g1 = groups[0], groups[1]\n",
    "        \n",
    "        # grab twin to pair dict\n",
    "        pair_to_ids = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            df = df_alpha[df_alpha[pair_var] == p]\n",
    "            id_0 = str(df[df[group_var] == g0].index.values[0])\n",
    "            id_1 = str(df[df[group_var] == g1].index.values[0])\n",
    "            pair_to_ids[p] = (id_0, id_1)\n",
    "        \n",
    "        # get distances for each twin pair per beta div matrix    \n",
    "        pair_to_dist = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            id_0, id_1 = pair_to_ids[p]\n",
    "            pair_to_dist[p] = df_beta.loc[id_0, id_1]\n",
    "        \n",
    "        df_paired_beta = pd.DataFrame.from_dict(pair_to_dist, orient='index', columns=['Unweighted_Unifrac'])\n",
    "        \n",
    "        # grab inter RA distances\n",
    "        # this is from unweighted_Timepoint_significance.qzv -> download as tsv\n",
    "        df_raw = pd.read_csv(path + 'outputs/Qiime2_' + g + '/raw_data.tsv', \n",
    "                             sep='\\t', index_col=0)\n",
    "        df_0 = df_raw[df_raw['Group1'] == g0]\n",
    "        df_0 = df_0[df_0['Group2'] == g0]\n",
    "        df_1 = df_raw[df_raw['Group1'] == g1]\n",
    "        df_1 = df_1[df_1['Group2'] == g1]\n",
    "        \n",
    "        # compare distances\n",
    "        inter_twin = df_paired_beta['Unweighted_Unifrac'].values\n",
    "        inter_0 = df_0['Distance'].values\n",
    "        inter_1 = df_1['Distance'].values\n",
    "        \n",
    "        u, p = scipy.stats.mannwhitneyu(inter_twin, inter_0)\n",
    "        #print(u, p)\n",
    "        \n",
    "        t, p = scipy.stats.ttest_ind(inter_twin, inter_1)\n",
    "        #print(t, p)\n",
    "        \n",
    "        t, p = scipy.stats.ttest_ind(inter_0, inter_1)\n",
    "        # print(t, p)\n",
    "        \n",
    "        f, p = scipy.stats.f_oneway(inter_0, inter_1, inter_twin)\n",
    "        # print('inter intra anova')\n",
    "        p# rint(f, p)\n",
    "        \n",
    "        category = ['intra_twin_pair']*len(inter_twin) + ['inter_' + g0 + '_only']*len(inter_0) + ['inter_' + g1 + '_only']*len(inter_1)\n",
    "        distances = list(inter_twin) + list(inter_0) + list(inter_1)\n",
    "        df_dist = pd.DataFrame(data=np.array([category,distances]).T, columns=['category','distance'])\n",
    "        df_dist['distance'] = df_dist['distance'].astype(float)\n",
    "        df_dist.to_csv(path + 'outputs/Qiime2_' + g + '/' + sg + '_inter_intra_beta_dist.tsv',sep='\\t')\n",
    "                             \n",
    "        sns.boxplot(data=df_dist, x='category', y='distance')\n",
    "        sns.swarmplot(data=df_dist, x='category', y='distance', color='black')\n",
    "        sns.despine()\n",
    "    \n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/Qiime2_' + g + '/' + sg + '_beta.pdf')\n",
    "        plt.close()          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8d8a7f96-5705-47fb-8d64-6ddcc4cd5d72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;__;__;__;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>0.014084</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.022027</td>\n",
       "      <td>0.001515</td>\n",
       "      <td>0.173283</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021230</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.152452</td>\n",
       "      <td>0.033001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.022576</td>\n",
       "      <td>0.007098</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.292202</td>\n",
       "      <td>0.001994</td>\n",
       "      <td>0.013555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000783</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>0.273829</td>\n",
       "      <td>0.067421</td>\n",
       "      <td>0.184572</td>\n",
       "      <td>0.026929</td>\n",
       "      <td>0.053673</td>\n",
       "      <td>0.037783</td>\n",
       "      <td>0.009619</td>\n",
       "      <td>0.023935</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.012060</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.181156</td>\n",
       "      <td>0.067844</td>\n",
       "      <td>0.150895</td>\n",
       "      <td>0.063229</td>\n",
       "      <td>0.051360</td>\n",
       "      <td>0.055600</td>\n",
       "      <td>0.012363</td>\n",
       "      <td>0.051513</td>\n",
       "      <td>0.006045</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.post.stool</th>\n",
       "      <td>0.174773</td>\n",
       "      <td>0.040785</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001855</td>\n",
       "      <td>0.124402</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001400</td>\n",
       "      <td>0.000371</td>\n",
       "      <td>0.014265</td>\n",
       "      <td>0.090594</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 374 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004  \\\n",
       "index                                                                                                                       \n",
       "OAD-001.post.stool                                           0.014084                                                       \n",
       "OAD-001.pre.stool                                            0.022576                                                       \n",
       "OAD-003.post.stool                                           0.273829                                                       \n",
       "OAD-003.pre.stool                                            0.181156                                                       \n",
       "OAD-004.post.stool                                           0.174773                                                       \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium  \\\n",
       "index                                                                                                                            \n",
       "OAD-001.post.stool                                           0.039046                                                            \n",
       "OAD-001.pre.stool                                            0.007098                                                            \n",
       "OAD-003.post.stool                                           0.067421                                                            \n",
       "OAD-003.pre.stool                                            0.067844                                                            \n",
       "OAD-004.post.stool                                           0.040785                                                            \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella  \\\n",
       "index                                                                                                             \n",
       "OAD-001.post.stool                                           0.022027                                             \n",
       "OAD-001.pre.stool                                            0.005128                                             \n",
       "OAD-003.post.stool                                           0.184572                                             \n",
       "OAD-003.pre.stool                                            0.150895                                             \n",
       "OAD-004.post.stool                                           0.000000                                             \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117  \\\n",
       "index                                                                                                                              \n",
       "OAD-001.post.stool                                           0.001515                                                              \n",
       "OAD-001.pre.stool                                            0.000000                                                              \n",
       "OAD-003.post.stool                                           0.026929                                                              \n",
       "OAD-003.pre.stool                                            0.063229                                                              \n",
       "OAD-004.post.stool                                           0.001855                                                              \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H  \\\n",
       "index                                                                                                                \n",
       "OAD-001.post.stool                                           0.173283                                                \n",
       "OAD-001.pre.stool                                            0.292202                                                \n",
       "OAD-003.post.stool                                           0.053673                                                \n",
       "OAD-003.pre.stool                                            0.051360                                                \n",
       "OAD-004.post.stool                                           0.124402                                                \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum  \\\n",
       "index                                                                                                                  \n",
       "OAD-001.post.stool                                           0.000000                                                  \n",
       "OAD-001.pre.stool                                            0.001994                                                  \n",
       "OAD-003.post.stool                                           0.037783                                                  \n",
       "OAD-003.pre.stool                                            0.055600                                                  \n",
       "OAD-004.post.stool                                           0.000000                                                  \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129  \\\n",
       "index                                                                                                                            \n",
       "OAD-001.post.stool                                           0.021230                                                            \n",
       "OAD-001.pre.stool                                            0.013555                                                            \n",
       "OAD-003.post.stool                                           0.009619                                                            \n",
       "OAD-003.pre.stool                                            0.012363                                                            \n",
       "OAD-004.post.stool                                           0.001400                                                            \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus  \\\n",
       "index                                                                                                                \n",
       "OAD-001.post.stool                                           0.000160                                                \n",
       "OAD-001.pre.stool                                            0.000000                                                \n",
       "OAD-003.post.stool                                           0.023935                                                \n",
       "OAD-003.pre.stool                                            0.051513                                                \n",
       "OAD-004.post.stool                                           0.000371                                                \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.post.stool                                           0.152452                                                        \n",
       "OAD-001.pre.stool                                            0.000783                                                        \n",
       "OAD-003.post.stool                                           0.005789                                                        \n",
       "OAD-003.pre.stool                                            0.006045                                                        \n",
       "OAD-004.post.stool                                           0.014265                                                        \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400  \\\n",
       "index                                                                                                                    \n",
       "OAD-001.post.stool                                           0.033001                                                    \n",
       "OAD-001.pre.stool                                            0.026849                                                    \n",
       "OAD-003.post.stool                                           0.012060                                                    \n",
       "OAD-003.pre.stool                                            0.025918                                                    \n",
       "OAD-004.post.stool                                           0.090594                                                    \n",
       "\n",
       "                    ...  d__Bacteria;p__Firmicutes_A;__;__;__;__  \\\n",
       "index               ...                                            \n",
       "OAD-001.post.stool  ...                                      0.0   \n",
       "OAD-001.pre.stool   ...                                      0.0   \n",
       "OAD-003.post.stool  ...                                      0.0   \n",
       "OAD-003.pre.stool   ...                                      0.0   \n",
       "OAD-004.post.stool  ...                                      0.0   \n",
       "\n",
       "                    d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.post.stool                                                0.0                                                        \n",
       "OAD-001.pre.stool                                                 0.0                                                        \n",
       "OAD-003.post.stool                                                0.0                                                        \n",
       "OAD-003.pre.stool                                                 0.0                                                        \n",
       "OAD-004.post.stool                                                0.0                                                        \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "index                                                                                 \n",
       "OAD-001.post.stool                                                0.0                 \n",
       "OAD-001.pre.stool                                                 0.0                 \n",
       "OAD-003.post.stool                                                0.0                 \n",
       "OAD-003.pre.stool                                                 0.0                 \n",
       "OAD-004.post.stool                                                0.0                 \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "index                                                                                                                         \n",
       "OAD-001.post.stool                                                0.0                                                         \n",
       "OAD-001.pre.stool                                                 0.0                                                         \n",
       "OAD-003.post.stool                                                0.0                                                         \n",
       "OAD-003.pre.stool                                                 0.0                                                         \n",
       "OAD-004.post.stool                                                0.0                                                         \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "index                                                                                                               \n",
       "OAD-001.post.stool                                                0.0                                               \n",
       "OAD-001.pre.stool                                                 0.0                                               \n",
       "OAD-003.post.stool                                                0.0                                               \n",
       "OAD-003.pre.stool                                                 0.0                                               \n",
       "OAD-004.post.stool                                                0.0                                               \n",
       "\n",
       "                    d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "index                                                                                                                                         \n",
       "OAD-001.post.stool                                                0.0                                                                         \n",
       "OAD-001.pre.stool                                                 0.0                                                                         \n",
       "OAD-003.post.stool                                                0.0                                                                         \n",
       "OAD-003.pre.stool                                                 0.0                                                                         \n",
       "OAD-004.post.stool                                                0.0                                                                         \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "index                                                                                                     \n",
       "OAD-001.post.stool                                                0.0                                     \n",
       "OAD-001.pre.stool                                                 0.0                                     \n",
       "OAD-003.post.stool                                                0.0                                     \n",
       "OAD-003.pre.stool                                                 0.0                                     \n",
       "OAD-004.post.stool                                                0.0                                     \n",
       "\n",
       "                    d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "index                                                                                                                                        \n",
       "OAD-001.post.stool                                                0.0                                                                        \n",
       "OAD-001.pre.stool                                                 0.0                                                                        \n",
       "OAD-003.post.stool                                                0.0                                                                        \n",
       "OAD-003.pre.stool                                                 0.0                                                                        \n",
       "OAD-004.post.stool                                                0.0                                                                        \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "index                                                                                                  \n",
       "OAD-001.post.stool                                                0.0                                  \n",
       "OAD-001.pre.stool                                                 0.0                                  \n",
       "OAD-003.post.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                 0.0                                  \n",
       "OAD-004.post.stool                                                0.0                                  \n",
       "\n",
       "                    d__Bacteria;p__Actinobacteriota;__;__;__;__  \n",
       "index                                                            \n",
       "OAD-001.post.stool                                          0.0  \n",
       "OAD-001.pre.stool                                           0.0  \n",
       "OAD-003.post.stool                                          0.0  \n",
       "OAD-003.pre.stool                                           0.0  \n",
       "OAD-004.post.stool                                          0.0  \n",
       "\n",
       "[5 rows x 374 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specific taxa profiling\n",
    "GL_to_otu = {}\n",
    "\n",
    "groups = ['saliva_adh', 'stool_adh']\n",
    "levels = ['6','7']\n",
    "for g in groups:\n",
    "    GL_to_otu[g] = {}\n",
    "    for l in levels:\n",
    "        # read in csv\n",
    "        df_otu = pd.read_csv(path + 'outputs/Qiime2_' + g + '/level-' + l + '.csv', index_col=0)\n",
    "\n",
    "        # filter on OTU only \n",
    "        keep = []\n",
    "        for x in df_otu.columns:\n",
    "            if 'd__' in x:\n",
    "                keep.append(x)\n",
    "        df_otu = df_otu.loc[:,keep]\n",
    "\n",
    "        # normalize by dividing by row sum\n",
    "        df_otu = df_otu.div(df_otu.sum(axis=1),axis=0)\n",
    "\n",
    "        # save to dict\n",
    "        GL_to_otu[g]['L' + l] = df_otu\n",
    "\n",
    "                                       \n",
    "GL_to_otu['stool_adh']['L6'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6bf1fdd5-b592-4a4e-a5d4-3cefc2d28c9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_Thiamine</th>\n",
       "      <th>H_Melezitose</th>\n",
       "      <th>H_Phosphocholine</th>\n",
       "      <th>H_SN_Glycero_3_Phophocholine</th>\n",
       "      <th>H_Gamma_Valerobetaine</th>\n",
       "      <th>H_N_Acetylneuraminic_acid</th>\n",
       "      <th>H_Pyridoxine</th>\n",
       "      <th>H_N_Acetylneuraminic_Acid</th>\n",
       "      <th>H_N_Acetylmuramic_Acid</th>\n",
       "      <th>...</th>\n",
       "      <th>H_Nonhydroxylated_bile_acid</th>\n",
       "      <th>H_Hydroxydodecanoic_acid</th>\n",
       "      <th>H_Omega_Hydroxydodecanoate</th>\n",
       "      <th>H_Delta_Methyldodecenoic_Acid</th>\n",
       "      <th>H_Hydroxydecanoic_acid</th>\n",
       "      <th>H_Hydroxydecanoate</th>\n",
       "      <th>H_Methylpentanoic_acid</th>\n",
       "      <th>H_Monohydroxylated_bile_acid</th>\n",
       "      <th>H_Palmitoyl_ethanolamide</th>\n",
       "      <th>H_Linoleic_Acid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>0.312342</td>\n",
       "      <td>-0.312477</td>\n",
       "      <td>0.119853</td>\n",
       "      <td>-0.492876</td>\n",
       "      <td>-0.671376</td>\n",
       "      <td>-0.202952</td>\n",
       "      <td>0.203945</td>\n",
       "      <td>-0.455869</td>\n",
       "      <td>0.352628</td>\n",
       "      <td>0.151464</td>\n",
       "      <td>...</td>\n",
       "      <td>0.054336</td>\n",
       "      <td>-0.330037</td>\n",
       "      <td>-0.886045</td>\n",
       "      <td>-0.314427</td>\n",
       "      <td>-0.322677</td>\n",
       "      <td>-0.650357</td>\n",
       "      <td>-0.361695</td>\n",
       "      <td>0.132794</td>\n",
       "      <td>-0.067464</td>\n",
       "      <td>-0.112641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>0.000428</td>\n",
       "      <td>-0.481251</td>\n",
       "      <td>-0.179337</td>\n",
       "      <td>0.983217</td>\n",
       "      <td>0.167836</td>\n",
       "      <td>0.067823</td>\n",
       "      <td>0.364897</td>\n",
       "      <td>-0.371942</td>\n",
       "      <td>0.545435</td>\n",
       "      <td>-0.007664</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.022614</td>\n",
       "      <td>-0.334936</td>\n",
       "      <td>-0.802118</td>\n",
       "      <td>-0.929470</td>\n",
       "      <td>-0.369578</td>\n",
       "      <td>-0.566431</td>\n",
       "      <td>0.075375</td>\n",
       "      <td>-0.519189</td>\n",
       "      <td>-0.180631</td>\n",
       "      <td>-0.418351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>-0.285212</td>\n",
       "      <td>-0.556445</td>\n",
       "      <td>-0.651325</td>\n",
       "      <td>0.357509</td>\n",
       "      <td>-0.401875</td>\n",
       "      <td>-0.295787</td>\n",
       "      <td>0.137736</td>\n",
       "      <td>-0.165902</td>\n",
       "      <td>0.112170</td>\n",
       "      <td>0.096791</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.923477</td>\n",
       "      <td>-0.589028</td>\n",
       "      <td>-1.535141</td>\n",
       "      <td>-0.525277</td>\n",
       "      <td>-0.278312</td>\n",
       "      <td>-0.600484</td>\n",
       "      <td>0.019657</td>\n",
       "      <td>-0.576831</td>\n",
       "      <td>-0.901479</td>\n",
       "      <td>-1.089906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD004</th>\n",
       "      <td>0.140904</td>\n",
       "      <td>1.089055</td>\n",
       "      <td>-0.668077</td>\n",
       "      <td>-1.061336</td>\n",
       "      <td>-0.370602</td>\n",
       "      <td>-0.329379</td>\n",
       "      <td>-0.654931</td>\n",
       "      <td>0.142971</td>\n",
       "      <td>-1.006689</td>\n",
       "      <td>-0.100373</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.354297</td>\n",
       "      <td>0.449897</td>\n",
       "      <td>0.489189</td>\n",
       "      <td>0.413723</td>\n",
       "      <td>0.474250</td>\n",
       "      <td>0.661963</td>\n",
       "      <td>0.245553</td>\n",
       "      <td>0.279290</td>\n",
       "      <td>-0.444319</td>\n",
       "      <td>-0.912164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD011</th>\n",
       "      <td>-0.346855</td>\n",
       "      <td>1.028262</td>\n",
       "      <td>-0.450323</td>\n",
       "      <td>-0.025939</td>\n",
       "      <td>-0.223029</td>\n",
       "      <td>-0.766374</td>\n",
       "      <td>-0.538410</td>\n",
       "      <td>-1.091423</td>\n",
       "      <td>-0.328522</td>\n",
       "      <td>-0.524686</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.385428</td>\n",
       "      <td>0.419748</td>\n",
       "      <td>0.429119</td>\n",
       "      <td>0.333756</td>\n",
       "      <td>0.448461</td>\n",
       "      <td>0.598846</td>\n",
       "      <td>0.390035</td>\n",
       "      <td>0.287682</td>\n",
       "      <td>-0.532141</td>\n",
       "      <td>-0.752280</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 88 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          H_Pyridoxamine  H_Thiamine  H_Melezitose  H_Phosphocholine  \\\n",
       "Study_ID                                                               \n",
       "VAOAD001        0.312342   -0.312477      0.119853         -0.492876   \n",
       "VAOAD009        0.000428   -0.481251     -0.179337          0.983217   \n",
       "OAD001         -0.285212   -0.556445     -0.651325          0.357509   \n",
       "VAOAD004        0.140904    1.089055     -0.668077         -1.061336   \n",
       "VAOAD011       -0.346855    1.028262     -0.450323         -0.025939   \n",
       "\n",
       "          H_SN_Glycero_3_Phophocholine  H_Gamma_Valerobetaine  \\\n",
       "Study_ID                                                        \n",
       "VAOAD001                     -0.671376              -0.202952   \n",
       "VAOAD009                      0.167836               0.067823   \n",
       "OAD001                       -0.401875              -0.295787   \n",
       "VAOAD004                     -0.370602              -0.329379   \n",
       "VAOAD011                     -0.223029              -0.766374   \n",
       "\n",
       "          H_N_Acetylneuraminic_acid  H_Pyridoxine  H_N_Acetylneuraminic_Acid  \\\n",
       "Study_ID                                                                       \n",
       "VAOAD001                   0.203945     -0.455869                   0.352628   \n",
       "VAOAD009                   0.364897     -0.371942                   0.545435   \n",
       "OAD001                     0.137736     -0.165902                   0.112170   \n",
       "VAOAD004                  -0.654931      0.142971                  -1.006689   \n",
       "VAOAD011                  -0.538410     -1.091423                  -0.328522   \n",
       "\n",
       "          H_N_Acetylmuramic_Acid  ...  H_Nonhydroxylated_bile_acid  \\\n",
       "Study_ID                          ...                                \n",
       "VAOAD001                0.151464  ...                     0.054336   \n",
       "VAOAD009               -0.007664  ...                    -0.022614   \n",
       "OAD001                  0.096791  ...                    -0.923477   \n",
       "VAOAD004               -0.100373  ...                    -0.354297   \n",
       "VAOAD011               -0.524686  ...                    -0.385428   \n",
       "\n",
       "          H_Hydroxydodecanoic_acid  H_Omega_Hydroxydodecanoate  \\\n",
       "Study_ID                                                         \n",
       "VAOAD001                 -0.330037                   -0.886045   \n",
       "VAOAD009                 -0.334936                   -0.802118   \n",
       "OAD001                   -0.589028                   -1.535141   \n",
       "VAOAD004                  0.449897                    0.489189   \n",
       "VAOAD011                  0.419748                    0.429119   \n",
       "\n",
       "          H_Delta_Methyldodecenoic_Acid  H_Hydroxydecanoic_acid  \\\n",
       "Study_ID                                                          \n",
       "VAOAD001                      -0.314427               -0.322677   \n",
       "VAOAD009                      -0.929470               -0.369578   \n",
       "OAD001                        -0.525277               -0.278312   \n",
       "VAOAD004                       0.413723                0.474250   \n",
       "VAOAD011                       0.333756                0.448461   \n",
       "\n",
       "          H_Hydroxydecanoate  H_Methylpentanoic_acid  \\\n",
       "Study_ID                                               \n",
       "VAOAD001           -0.650357               -0.361695   \n",
       "VAOAD009           -0.566431                0.075375   \n",
       "OAD001             -0.600484                0.019657   \n",
       "VAOAD004            0.661963                0.245553   \n",
       "VAOAD011            0.598846                0.390035   \n",
       "\n",
       "          H_Monohydroxylated_bile_acid  H_Palmitoyl_ethanolamide  \\\n",
       "Study_ID                                                           \n",
       "VAOAD001                      0.132794                 -0.067464   \n",
       "VAOAD009                     -0.519189                 -0.180631   \n",
       "OAD001                       -0.576831                 -0.901479   \n",
       "VAOAD004                      0.279290                 -0.444319   \n",
       "VAOAD011                      0.287682                 -0.532141   \n",
       "\n",
       "          H_Linoleic_Acid  \n",
       "Study_ID                   \n",
       "VAOAD001        -0.112641  \n",
       "VAOAD009        -0.418351  \n",
       "OAD001          -1.089906  \n",
       "VAOAD004        -0.912164  \n",
       "VAOAD011        -0.752280  \n",
       "\n",
       "[5 rows x 88 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Metabolome testing\n",
    "gdt_to_df = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gdt_to_df[g] = {}\n",
    "    for d in ['paired_beta']: # just arbitrarily \n",
    "        gdt_to_df[g][d] = {}\n",
    "        df_merge = dfd_to_merge[g + '_adh'][d]\n",
    "        df_meta = pd.read_csv(path + 'inputs/' + g + '_normalized.csv')\n",
    "\n",
    "        # split on timepoint\n",
    "        for t in ['Baseline','After diet']:\n",
    "            df = df_meta[df_meta['Time'] == t]\n",
    "            df = df.set_index('Study_ID')\n",
    "            df.index = df.index.map(lambda x: x.split('_')[0])\n",
    "            df = df.drop('Time',axis=1)\n",
    "            gdt_to_df[g][d][t] = df\n",
    "\n",
    "        gdt_to_df[g][d]['diff'] = gdt_to_df[g][d]['After diet'] - gdt_to_df[g][d]['Baseline']\n",
    "\n",
    "gdt_to_df[g][d][t].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "dbad4aa2-7c35-490c-900b-c607c4ff549d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_Thiamine</th>\n",
       "      <th>H_Melezitose</th>\n",
       "      <th>H_Phosphocholine</th>\n",
       "      <th>H_SN_Glycero_3_Phophocholine</th>\n",
       "      <th>H_Gamma_Valerobetaine</th>\n",
       "      <th>H_N_Acetylneuraminic_acid</th>\n",
       "      <th>H_Pyridoxine</th>\n",
       "      <th>H_N_Acetylneuraminic_Acid</th>\n",
       "      <th>H_N_Acetylmuramic_Acid</th>\n",
       "      <th>...</th>\n",
       "      <th>H_Nonhydroxylated_bile_acid</th>\n",
       "      <th>H_Hydroxydodecanoic_acid</th>\n",
       "      <th>H_Omega_Hydroxydodecanoate</th>\n",
       "      <th>H_Delta_Methyldodecenoic_Acid</th>\n",
       "      <th>H_Hydroxydecanoic_acid</th>\n",
       "      <th>H_Hydroxydecanoate</th>\n",
       "      <th>H_Methylpentanoic_acid</th>\n",
       "      <th>H_Monohydroxylated_bile_acid</th>\n",
       "      <th>H_Palmitoyl_ethanolamide</th>\n",
       "      <th>H_Linoleic_Acid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>0.312342</td>\n",
       "      <td>-0.312477</td>\n",
       "      <td>0.119853</td>\n",
       "      <td>-0.492876</td>\n",
       "      <td>-0.671376</td>\n",
       "      <td>-0.202952</td>\n",
       "      <td>0.203945</td>\n",
       "      <td>-0.455869</td>\n",
       "      <td>0.352628</td>\n",
       "      <td>0.151464</td>\n",
       "      <td>...</td>\n",
       "      <td>0.054336</td>\n",
       "      <td>-0.330037</td>\n",
       "      <td>-0.886045</td>\n",
       "      <td>-0.314427</td>\n",
       "      <td>-0.322677</td>\n",
       "      <td>-0.650357</td>\n",
       "      <td>-0.361695</td>\n",
       "      <td>0.132794</td>\n",
       "      <td>-0.067464</td>\n",
       "      <td>-0.112641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>0.000428</td>\n",
       "      <td>-0.481251</td>\n",
       "      <td>-0.179337</td>\n",
       "      <td>0.983217</td>\n",
       "      <td>0.167836</td>\n",
       "      <td>0.067823</td>\n",
       "      <td>0.364897</td>\n",
       "      <td>-0.371942</td>\n",
       "      <td>0.545435</td>\n",
       "      <td>-0.007664</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.022614</td>\n",
       "      <td>-0.334936</td>\n",
       "      <td>-0.802118</td>\n",
       "      <td>-0.929470</td>\n",
       "      <td>-0.369578</td>\n",
       "      <td>-0.566431</td>\n",
       "      <td>0.075375</td>\n",
       "      <td>-0.519189</td>\n",
       "      <td>-0.180631</td>\n",
       "      <td>-0.418351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>-0.285212</td>\n",
       "      <td>-0.556445</td>\n",
       "      <td>-0.651325</td>\n",
       "      <td>0.357509</td>\n",
       "      <td>-0.401875</td>\n",
       "      <td>-0.295787</td>\n",
       "      <td>0.137736</td>\n",
       "      <td>-0.165902</td>\n",
       "      <td>0.112170</td>\n",
       "      <td>0.096791</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.923477</td>\n",
       "      <td>-0.589028</td>\n",
       "      <td>-1.535141</td>\n",
       "      <td>-0.525277</td>\n",
       "      <td>-0.278312</td>\n",
       "      <td>-0.600484</td>\n",
       "      <td>0.019657</td>\n",
       "      <td>-0.576831</td>\n",
       "      <td>-0.901479</td>\n",
       "      <td>-1.089906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD004</th>\n",
       "      <td>0.140904</td>\n",
       "      <td>1.089055</td>\n",
       "      <td>-0.668077</td>\n",
       "      <td>-1.061336</td>\n",
       "      <td>-0.370602</td>\n",
       "      <td>-0.329379</td>\n",
       "      <td>-0.654931</td>\n",
       "      <td>0.142971</td>\n",
       "      <td>-1.006689</td>\n",
       "      <td>-0.100373</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.354297</td>\n",
       "      <td>0.449897</td>\n",
       "      <td>0.489189</td>\n",
       "      <td>0.413723</td>\n",
       "      <td>0.474250</td>\n",
       "      <td>0.661963</td>\n",
       "      <td>0.245553</td>\n",
       "      <td>0.279290</td>\n",
       "      <td>-0.444319</td>\n",
       "      <td>-0.912164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD011</th>\n",
       "      <td>-0.346855</td>\n",
       "      <td>1.028262</td>\n",
       "      <td>-0.450323</td>\n",
       "      <td>-0.025939</td>\n",
       "      <td>-0.223029</td>\n",
       "      <td>-0.766374</td>\n",
       "      <td>-0.538410</td>\n",
       "      <td>-1.091423</td>\n",
       "      <td>-0.328522</td>\n",
       "      <td>-0.524686</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.385428</td>\n",
       "      <td>0.419748</td>\n",
       "      <td>0.429119</td>\n",
       "      <td>0.333756</td>\n",
       "      <td>0.448461</td>\n",
       "      <td>0.598846</td>\n",
       "      <td>0.390035</td>\n",
       "      <td>0.287682</td>\n",
       "      <td>-0.532141</td>\n",
       "      <td>-0.752280</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 88 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          H_Pyridoxamine  H_Thiamine  H_Melezitose  H_Phosphocholine  \\\n",
       "Study_ID                                                               \n",
       "VAOAD001        0.312342   -0.312477      0.119853         -0.492876   \n",
       "VAOAD009        0.000428   -0.481251     -0.179337          0.983217   \n",
       "OAD001         -0.285212   -0.556445     -0.651325          0.357509   \n",
       "VAOAD004        0.140904    1.089055     -0.668077         -1.061336   \n",
       "VAOAD011       -0.346855    1.028262     -0.450323         -0.025939   \n",
       "\n",
       "          H_SN_Glycero_3_Phophocholine  H_Gamma_Valerobetaine  \\\n",
       "Study_ID                                                        \n",
       "VAOAD001                     -0.671376              -0.202952   \n",
       "VAOAD009                      0.167836               0.067823   \n",
       "OAD001                       -0.401875              -0.295787   \n",
       "VAOAD004                     -0.370602              -0.329379   \n",
       "VAOAD011                     -0.223029              -0.766374   \n",
       "\n",
       "          H_N_Acetylneuraminic_acid  H_Pyridoxine  H_N_Acetylneuraminic_Acid  \\\n",
       "Study_ID                                                                       \n",
       "VAOAD001                   0.203945     -0.455869                   0.352628   \n",
       "VAOAD009                   0.364897     -0.371942                   0.545435   \n",
       "OAD001                     0.137736     -0.165902                   0.112170   \n",
       "VAOAD004                  -0.654931      0.142971                  -1.006689   \n",
       "VAOAD011                  -0.538410     -1.091423                  -0.328522   \n",
       "\n",
       "          H_N_Acetylmuramic_Acid  ...  H_Nonhydroxylated_bile_acid  \\\n",
       "Study_ID                          ...                                \n",
       "VAOAD001                0.151464  ...                     0.054336   \n",
       "VAOAD009               -0.007664  ...                    -0.022614   \n",
       "OAD001                  0.096791  ...                    -0.923477   \n",
       "VAOAD004               -0.100373  ...                    -0.354297   \n",
       "VAOAD011               -0.524686  ...                    -0.385428   \n",
       "\n",
       "          H_Hydroxydodecanoic_acid  H_Omega_Hydroxydodecanoate  \\\n",
       "Study_ID                                                         \n",
       "VAOAD001                 -0.330037                   -0.886045   \n",
       "VAOAD009                 -0.334936                   -0.802118   \n",
       "OAD001                   -0.589028                   -1.535141   \n",
       "VAOAD004                  0.449897                    0.489189   \n",
       "VAOAD011                  0.419748                    0.429119   \n",
       "\n",
       "          H_Delta_Methyldodecenoic_Acid  H_Hydroxydecanoic_acid  \\\n",
       "Study_ID                                                          \n",
       "VAOAD001                      -0.314427               -0.322677   \n",
       "VAOAD009                      -0.929470               -0.369578   \n",
       "OAD001                        -0.525277               -0.278312   \n",
       "VAOAD004                       0.413723                0.474250   \n",
       "VAOAD011                       0.333756                0.448461   \n",
       "\n",
       "          H_Hydroxydecanoate  H_Methylpentanoic_acid  \\\n",
       "Study_ID                                               \n",
       "VAOAD001           -0.650357               -0.361695   \n",
       "VAOAD009           -0.566431                0.075375   \n",
       "OAD001             -0.600484                0.019657   \n",
       "VAOAD004            0.661963                0.245553   \n",
       "VAOAD011            0.598846                0.390035   \n",
       "\n",
       "          H_Monohydroxylated_bile_acid  H_Palmitoyl_ethanolamide  \\\n",
       "Study_ID                                                           \n",
       "VAOAD001                      0.132794                 -0.067464   \n",
       "VAOAD009                     -0.519189                 -0.180631   \n",
       "OAD001                       -0.576831                 -0.901479   \n",
       "VAOAD004                      0.279290                 -0.444319   \n",
       "VAOAD011                      0.287682                 -0.532141   \n",
       "\n",
       "          H_Linoleic_Acid  \n",
       "Study_ID                   \n",
       "VAOAD001        -0.112641  \n",
       "VAOAD009        -0.418351  \n",
       "OAD001          -1.089906  \n",
       "VAOAD004        -0.912164  \n",
       "VAOAD011        -0.752280  \n",
       "\n",
       "[5 rows x 88 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Metabolome testing\n",
    "gt_to_meta = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gt_to_meta[g] = {}\n",
    "    df_meta = pd.read_csv(path + 'inputs/' + g + '_normalized.csv')\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['Baseline','After diet']:\n",
    "        df = df_meta[df_meta['Time'] == t]\n",
    "        df = df.set_index('Study_ID')\n",
    "        df.index = df.index.map(lambda x: x.split('_')[0])\n",
    "        df = df.drop('Time',axis=1)\n",
    "        gt_to_meta[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_meta[g]['Baseline'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('D0','D-0'))\n",
    "    pre_df.index = pre_df.index.map(lambda x: x + '.pre.' + g)\n",
    "    post_df = gt_to_meta[g]['After diet'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('D0','D-0'))\n",
    "    post_df.index = post_df.index.map(lambda x: x + '.post.' + g)\n",
    "    \n",
    "    gt_to_meta[g]['diff'] = gt_to_meta[g]['After diet'] - gt_to_meta[g]['Baseline']\n",
    "    gt_to_meta[g]['all'] = pd.concat([pre_df,post_df])\n",
    "    \n",
    "\n",
    "gt_to_meta[g][t].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d5c7924c-8c76-4970-b7fc-9d3c68132504",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;__;__;__;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.022576</td>\n",
       "      <td>0.007098</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.292202</td>\n",
       "      <td>0.001994</td>\n",
       "      <td>0.013555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000783</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.181156</td>\n",
       "      <td>0.067844</td>\n",
       "      <td>0.150895</td>\n",
       "      <td>0.063229</td>\n",
       "      <td>0.051360</td>\n",
       "      <td>0.055600</td>\n",
       "      <td>0.012363</td>\n",
       "      <td>0.051513</td>\n",
       "      <td>0.006045</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.170334</td>\n",
       "      <td>0.135468</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002088</td>\n",
       "      <td>0.173617</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003649</td>\n",
       "      <td>0.008765</td>\n",
       "      <td>0.054802</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-005.pre.stool</th>\n",
       "      <td>0.040108</td>\n",
       "      <td>0.139385</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018806</td>\n",
       "      <td>0.155290</td>\n",
       "      <td>0.037507</td>\n",
       "      <td>0.024457</td>\n",
       "      <td>0.066852</td>\n",
       "      <td>0.008969</td>\n",
       "      <td>0.018328</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-006.pre.stool</th>\n",
       "      <td>0.037219</td>\n",
       "      <td>0.167448</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007940</td>\n",
       "      <td>0.019591</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011006</td>\n",
       "      <td>0.001275</td>\n",
       "      <td>0.012769</td>\n",
       "      <td>0.162962</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 374 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004  \\\n",
       "index                                                                                                                      \n",
       "OAD-001.pre.stool                                           0.022576                                                       \n",
       "OAD-003.pre.stool                                           0.181156                                                       \n",
       "OAD-004.pre.stool                                           0.170334                                                       \n",
       "OAD-005.pre.stool                                           0.040108                                                       \n",
       "OAD-006.pre.stool                                           0.037219                                                       \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium  \\\n",
       "index                                                                                                                           \n",
       "OAD-001.pre.stool                                           0.007098                                                            \n",
       "OAD-003.pre.stool                                           0.067844                                                            \n",
       "OAD-004.pre.stool                                           0.135468                                                            \n",
       "OAD-005.pre.stool                                           0.139385                                                            \n",
       "OAD-006.pre.stool                                           0.167448                                                            \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella  \\\n",
       "index                                                                                                            \n",
       "OAD-001.pre.stool                                           0.005128                                             \n",
       "OAD-003.pre.stool                                           0.150895                                             \n",
       "OAD-004.pre.stool                                           0.000000                                             \n",
       "OAD-005.pre.stool                                           0.000000                                             \n",
       "OAD-006.pre.stool                                           0.000000                                             \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117  \\\n",
       "index                                                                                                                             \n",
       "OAD-001.pre.stool                                           0.000000                                                              \n",
       "OAD-003.pre.stool                                           0.063229                                                              \n",
       "OAD-004.pre.stool                                           0.002088                                                              \n",
       "OAD-005.pre.stool                                           0.018806                                                              \n",
       "OAD-006.pre.stool                                           0.007940                                                              \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H  \\\n",
       "index                                                                                                               \n",
       "OAD-001.pre.stool                                           0.292202                                                \n",
       "OAD-003.pre.stool                                           0.051360                                                \n",
       "OAD-004.pre.stool                                           0.173617                                                \n",
       "OAD-005.pre.stool                                           0.155290                                                \n",
       "OAD-006.pre.stool                                           0.019591                                                \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum  \\\n",
       "index                                                                                                                 \n",
       "OAD-001.pre.stool                                           0.001994                                                  \n",
       "OAD-003.pre.stool                                           0.055600                                                  \n",
       "OAD-004.pre.stool                                           0.000143                                                  \n",
       "OAD-005.pre.stool                                           0.037507                                                  \n",
       "OAD-006.pre.stool                                           0.000000                                                  \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129  \\\n",
       "index                                                                                                                           \n",
       "OAD-001.pre.stool                                           0.013555                                                            \n",
       "OAD-003.pre.stool                                           0.012363                                                            \n",
       "OAD-004.pre.stool                                           0.000000                                                            \n",
       "OAD-005.pre.stool                                           0.024457                                                            \n",
       "OAD-006.pre.stool                                           0.011006                                                            \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus  \\\n",
       "index                                                                                                               \n",
       "OAD-001.pre.stool                                           0.000000                                                \n",
       "OAD-003.pre.stool                                           0.051513                                                \n",
       "OAD-004.pre.stool                                           0.003649                                                \n",
       "OAD-005.pre.stool                                           0.066852                                                \n",
       "OAD-006.pre.stool                                           0.001275                                                \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor  \\\n",
       "index                                                                                                                       \n",
       "OAD-001.pre.stool                                           0.000783                                                        \n",
       "OAD-003.pre.stool                                           0.006045                                                        \n",
       "OAD-004.pre.stool                                           0.008765                                                        \n",
       "OAD-005.pre.stool                                           0.008969                                                        \n",
       "OAD-006.pre.stool                                           0.012769                                                        \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400  \\\n",
       "index                                                                                                                   \n",
       "OAD-001.pre.stool                                           0.026849                                                    \n",
       "OAD-003.pre.stool                                           0.025918                                                    \n",
       "OAD-004.pre.stool                                           0.054802                                                    \n",
       "OAD-005.pre.stool                                           0.018328                                                    \n",
       "OAD-006.pre.stool                                           0.162962                                                    \n",
       "\n",
       "                   ...  d__Bacteria;p__Firmicutes_A;__;__;__;__  \\\n",
       "index              ...                                            \n",
       "OAD-001.pre.stool  ...                                      0.0   \n",
       "OAD-003.pre.stool  ...                                      0.0   \n",
       "OAD-004.pre.stool  ...                                      0.0   \n",
       "OAD-005.pre.stool  ...                                      0.0   \n",
       "OAD-006.pre.stool  ...                                      0.0   \n",
       "\n",
       "                   d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "index                                                                                                                       \n",
       "OAD-001.pre.stool                                                0.0                                                        \n",
       "OAD-003.pre.stool                                                0.0                                                        \n",
       "OAD-004.pre.stool                                                0.0                                                        \n",
       "OAD-005.pre.stool                                                0.0                                                        \n",
       "OAD-006.pre.stool                                                0.0                                                        \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "index                                                                                \n",
       "OAD-001.pre.stool                                                0.0                 \n",
       "OAD-003.pre.stool                                                0.0                 \n",
       "OAD-004.pre.stool                                                0.0                 \n",
       "OAD-005.pre.stool                                                0.0                 \n",
       "OAD-006.pre.stool                                                0.0                 \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.pre.stool                                                0.0                                                         \n",
       "OAD-003.pre.stool                                                0.0                                                         \n",
       "OAD-004.pre.stool                                                0.0                                                         \n",
       "OAD-005.pre.stool                                                0.0                                                         \n",
       "OAD-006.pre.stool                                                0.0                                                         \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "index                                                                                                              \n",
       "OAD-001.pre.stool                                                0.0                                               \n",
       "OAD-003.pre.stool                                                0.0                                               \n",
       "OAD-004.pre.stool                                                0.0                                               \n",
       "OAD-005.pre.stool                                                0.0                                               \n",
       "OAD-006.pre.stool                                                0.0                                               \n",
       "\n",
       "                   d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "index                                                                                                                                        \n",
       "OAD-001.pre.stool                                                0.0                                                                         \n",
       "OAD-003.pre.stool                                                0.0                                                                         \n",
       "OAD-004.pre.stool                                                0.0                                                                         \n",
       "OAD-005.pre.stool                                                0.0                                                                         \n",
       "OAD-006.pre.stool                                                0.0                                                                         \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "index                                                                                                    \n",
       "OAD-001.pre.stool                                                0.0                                     \n",
       "OAD-003.pre.stool                                                0.0                                     \n",
       "OAD-004.pre.stool                                                0.0                                     \n",
       "OAD-005.pre.stool                                                0.0                                     \n",
       "OAD-006.pre.stool                                                0.0                                     \n",
       "\n",
       "                   d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "index                                                                                                                                       \n",
       "OAD-001.pre.stool                                                0.0                                                                        \n",
       "OAD-003.pre.stool                                                0.0                                                                        \n",
       "OAD-004.pre.stool                                                0.0                                                                        \n",
       "OAD-005.pre.stool                                                0.0                                                                        \n",
       "OAD-006.pre.stool                                                0.0                                                                        \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "index                                                                                                 \n",
       "OAD-001.pre.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                0.0                                  \n",
       "OAD-004.pre.stool                                                0.0                                  \n",
       "OAD-005.pre.stool                                                0.0                                  \n",
       "OAD-006.pre.stool                                                0.0                                  \n",
       "\n",
       "                   d__Bacteria;p__Actinobacteriota;__;__;__;__  \n",
       "index                                                           \n",
       "OAD-001.pre.stool                                          0.0  \n",
       "OAD-003.pre.stool                                          0.0  \n",
       "OAD-004.pre.stool                                          0.0  \n",
       "OAD-005.pre.stool                                          0.0  \n",
       "OAD-006.pre.stool                                          0.0  \n",
       "\n",
       "[5 rows x 374 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Microbiome testing\n",
    "gt_to_otu = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gt_to_otu[g] = {}\n",
    "    df_otu = GL_to_otu[g + '_adh']['L6'].copy()\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['post','pre']:\n",
    "        df = df_otu[df_otu.index.to_series().str.contains(t)]\n",
    "        gt_to_otu[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_otu[g]['pre'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.split('.pre')[0])\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('-',''))\n",
    "    post_df = gt_to_otu[g]['post'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.split('.post')[0])\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('-',''))\n",
    "    \n",
    "    gt_to_otu[g]['diff'] = post_df - pre_df\n",
    "    gt_to_otu[g]['all'] = pd.concat([gt_to_otu[g]['pre'],gt_to_otu[g]['post']])\n",
    "\n",
    "gt_to_otu[g][t].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2dc087d8-1169-468f-b476-1fdd9c28baec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saliva\n",
      "Response\n",
      "Number of sigtaxa\n",
      "9\n",
      "No response\n",
      "Number of sigtaxa\n",
      "0\n",
      "stool\n",
      "Response\n",
      "Number of sigtaxa\n",
      "6\n",
      "No response\n",
      "Number of sigtaxa\n",
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "      <th>Timepoint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.022576</td>\n",
       "      <td>0.007098</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.292202</td>\n",
       "      <td>0.001994</td>\n",
       "      <td>0.013555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000783</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.181156</td>\n",
       "      <td>0.067844</td>\n",
       "      <td>0.150895</td>\n",
       "      <td>0.063229</td>\n",
       "      <td>0.051360</td>\n",
       "      <td>0.055600</td>\n",
       "      <td>0.012363</td>\n",
       "      <td>0.051513</td>\n",
       "      <td>0.006045</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.170334</td>\n",
       "      <td>0.135468</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002088</td>\n",
       "      <td>0.173617</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003649</td>\n",
       "      <td>0.008765</td>\n",
       "      <td>0.054802</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-001.pre.stool</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.064924</td>\n",
       "      <td>0.001590</td>\n",
       "      <td>0.032964</td>\n",
       "      <td>0.321408</td>\n",
       "      <td>0.010514</td>\n",
       "      <td>0.070031</td>\n",
       "      <td>0.005441</td>\n",
       "      <td>0.011820</td>\n",
       "      <td>0.020509</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-012.pre.stool</th>\n",
       "      <td>0.126033</td>\n",
       "      <td>0.139754</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015668</td>\n",
       "      <td>0.149951</td>\n",
       "      <td>0.015739</td>\n",
       "      <td>0.011277</td>\n",
       "      <td>0.000460</td>\n",
       "      <td>0.000372</td>\n",
       "      <td>0.005010</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004  \\\n",
       "OAD-001.pre.stool                                             0.022576                                                       \n",
       "OAD-003.pre.stool                                             0.181156                                                       \n",
       "OAD-004.pre.stool                                             0.170334                                                       \n",
       "VAOAD-001.pre.stool                                           0.000000                                                       \n",
       "VAOAD-012.pre.stool                                           0.126033                                                       \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium  \\\n",
       "OAD-001.pre.stool                                             0.007098                                                            \n",
       "OAD-003.pre.stool                                             0.067844                                                            \n",
       "OAD-004.pre.stool                                             0.135468                                                            \n",
       "VAOAD-001.pre.stool                                           0.064924                                                            \n",
       "VAOAD-012.pre.stool                                           0.139754                                                            \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella  \\\n",
       "OAD-001.pre.stool                                             0.005128                                             \n",
       "OAD-003.pre.stool                                             0.150895                                             \n",
       "OAD-004.pre.stool                                             0.000000                                             \n",
       "VAOAD-001.pre.stool                                           0.001590                                             \n",
       "VAOAD-012.pre.stool                                           0.000000                                             \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117  \\\n",
       "OAD-001.pre.stool                                             0.000000                                                              \n",
       "OAD-003.pre.stool                                             0.063229                                                              \n",
       "OAD-004.pre.stool                                             0.002088                                                              \n",
       "VAOAD-001.pre.stool                                           0.032964                                                              \n",
       "VAOAD-012.pre.stool                                           0.015668                                                              \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H  \\\n",
       "OAD-001.pre.stool                                             0.292202                                                \n",
       "OAD-003.pre.stool                                             0.051360                                                \n",
       "OAD-004.pre.stool                                             0.173617                                                \n",
       "VAOAD-001.pre.stool                                           0.321408                                                \n",
       "VAOAD-012.pre.stool                                           0.149951                                                \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum  \\\n",
       "OAD-001.pre.stool                                             0.001994                                                  \n",
       "OAD-003.pre.stool                                             0.055600                                                  \n",
       "OAD-004.pre.stool                                             0.000143                                                  \n",
       "VAOAD-001.pre.stool                                           0.010514                                                  \n",
       "VAOAD-012.pre.stool                                           0.015739                                                  \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129  \\\n",
       "OAD-001.pre.stool                                             0.013555                                                            \n",
       "OAD-003.pre.stool                                             0.012363                                                            \n",
       "OAD-004.pre.stool                                             0.000000                                                            \n",
       "VAOAD-001.pre.stool                                           0.070031                                                            \n",
       "VAOAD-012.pre.stool                                           0.011277                                                            \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus  \\\n",
       "OAD-001.pre.stool                                             0.000000                                                \n",
       "OAD-003.pre.stool                                             0.051513                                                \n",
       "OAD-004.pre.stool                                             0.003649                                                \n",
       "VAOAD-001.pre.stool                                           0.005441                                                \n",
       "VAOAD-012.pre.stool                                           0.000460                                                \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor  \\\n",
       "OAD-001.pre.stool                                             0.000783                                                        \n",
       "OAD-003.pre.stool                                             0.006045                                                        \n",
       "OAD-004.pre.stool                                             0.008765                                                        \n",
       "VAOAD-001.pre.stool                                           0.011820                                                        \n",
       "VAOAD-012.pre.stool                                           0.000372                                                        \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400  \\\n",
       "OAD-001.pre.stool                                             0.026849                                                    \n",
       "OAD-003.pre.stool                                             0.025918                                                    \n",
       "OAD-004.pre.stool                                             0.054802                                                    \n",
       "VAOAD-001.pre.stool                                           0.020509                                                    \n",
       "VAOAD-012.pre.stool                                           0.005010                                                    \n",
       "\n",
       "                     ...  \\\n",
       "OAD-001.pre.stool    ...   \n",
       "OAD-003.pre.stool    ...   \n",
       "OAD-004.pre.stool    ...   \n",
       "VAOAD-001.pre.stool  ...   \n",
       "VAOAD-012.pre.stool  ...   \n",
       "\n",
       "                     d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "OAD-001.pre.stool                                                  0.0                                                        \n",
       "OAD-003.pre.stool                                                  0.0                                                        \n",
       "OAD-004.pre.stool                                                  0.0                                                        \n",
       "VAOAD-001.pre.stool                                                0.0                                                        \n",
       "VAOAD-012.pre.stool                                                0.0                                                        \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "OAD-001.pre.stool                                                  0.0                 \n",
       "OAD-003.pre.stool                                                  0.0                 \n",
       "OAD-004.pre.stool                                                  0.0                 \n",
       "VAOAD-001.pre.stool                                                0.0                 \n",
       "VAOAD-012.pre.stool                                                0.0                 \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "OAD-001.pre.stool                                                  0.0                                                         \n",
       "OAD-003.pre.stool                                                  0.0                                                         \n",
       "OAD-004.pre.stool                                                  0.0                                                         \n",
       "VAOAD-001.pre.stool                                                0.0                                                         \n",
       "VAOAD-012.pre.stool                                                0.0                                                         \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "OAD-001.pre.stool                                                  0.0                                               \n",
       "OAD-003.pre.stool                                                  0.0                                               \n",
       "OAD-004.pre.stool                                                  0.0                                               \n",
       "VAOAD-001.pre.stool                                                0.0                                               \n",
       "VAOAD-012.pre.stool                                                0.0                                               \n",
       "\n",
       "                     d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "OAD-001.pre.stool                                                  0.0                                                                         \n",
       "OAD-003.pre.stool                                                  0.0                                                                         \n",
       "OAD-004.pre.stool                                                  0.0                                                                         \n",
       "VAOAD-001.pre.stool                                                0.0                                                                         \n",
       "VAOAD-012.pre.stool                                                0.0                                                                         \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "OAD-001.pre.stool                                                  0.0                                     \n",
       "OAD-003.pre.stool                                                  0.0                                     \n",
       "OAD-004.pre.stool                                                  0.0                                     \n",
       "VAOAD-001.pre.stool                                                0.0                                     \n",
       "VAOAD-012.pre.stool                                                0.0                                     \n",
       "\n",
       "                     d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "OAD-001.pre.stool                                                  0.0                                                                        \n",
       "OAD-003.pre.stool                                                  0.0                                                                        \n",
       "OAD-004.pre.stool                                                  0.0                                                                        \n",
       "VAOAD-001.pre.stool                                                0.0                                                                        \n",
       "VAOAD-012.pre.stool                                                0.0                                                                        \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "OAD-001.pre.stool                                                  0.0                                  \n",
       "OAD-003.pre.stool                                                  0.0                                  \n",
       "OAD-004.pre.stool                                                  0.0                                  \n",
       "VAOAD-001.pre.stool                                                0.0                                  \n",
       "VAOAD-012.pre.stool                                                0.0                                  \n",
       "\n",
       "                     d__Bacteria;p__Actinobacteriota;__;__;__;__  Timepoint  \n",
       "OAD-001.pre.stool                                       0.000000        pre  \n",
       "OAD-003.pre.stool                                       0.000000        pre  \n",
       "OAD-004.pre.stool                                       0.000000        pre  \n",
       "VAOAD-001.pre.stool                                     0.000000        pre  \n",
       "VAOAD-012.pre.stool                                     0.000035        pre  \n",
       "\n",
       "[5 rows x 375 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# paired testing in resp vs non resp \n",
    "# does pre vs post testing in each category alone\n",
    "\n",
    "to_sigotu = {}\n",
    "\n",
    "gr_to_sigotu = {}\n",
    "\n",
    "gs = []\n",
    "rs = []\n",
    "ts = []\n",
    "ws = []\n",
    "ps = []\n",
    "\n",
    "for g in ['saliva','stool']:\n",
    "    print(g)\n",
    "    r_to_sigotu = {}\n",
    "    # grab taxa table\n",
    "    df_otu = gt_to_otu[g]['all']\n",
    "\n",
    "    df = type_to_df_map[g].copy()\n",
    "    df = df.set_index('Together')\n",
    "    if g == 'stool':\n",
    "        df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "    df = df[['Timepoint','WOMAC_P_Response']]\n",
    "    for r in ['Response', 'No response']:\n",
    "        print(r)\n",
    "        df_merge = pd.concat([df_otu, df[df['WOMAC_P_Response'] == r]], axis=1)\n",
    "        df_merge = df_merge.dropna()\n",
    "        df_merge['SampleID'] = df_merge.index.map(lambda x: x.split('.p')[0])\n",
    "        df_vc = df_merge['SampleID'].value_counts()\n",
    "        keep = df_vc[df_vc == 2].index.values\n",
    "        df_merge = df_merge[df_merge['SampleID'].isin(keep)] \n",
    "        df_merge = df_merge.drop(['SampleID','WOMAC_P_Response'],axis=1)\n",
    "        sig_otu = []\n",
    "        for t in list(df_merge.columns.values)[:-1]:\n",
    "            df_pre = df_merge[df_merge['Timepoint'] == 'pre'][t]\n",
    "            df_post = df_merge[df_merge['Timepoint'] == 'post'][t]\n",
    "\n",
    "            try:\n",
    "                w,p = scipy.stats.wilcoxon(df_pre.values, df_post.values, nan_policy='omit')\n",
    "            except:\n",
    "                w,p = 0, 1\n",
    "            if p < 0.05:\n",
    "                gs.append(g)\n",
    "                rs.append(r)\n",
    "                ts.append(t)\n",
    "                ws.append(w)\n",
    "                ps.append(p)\n",
    "\n",
    "                ax = sns.boxplot(data=df_merge, y=t, x='Timepoint')\n",
    "                sns.swarmplot(data=df_merge, y=t, x='Timepoint', palette='dark:grey')\n",
    "                sns.despine()\n",
    "            \n",
    "                plt.tight_layout()\n",
    "                plt.savefig(path + 'outputs/jobs06/rvnr_taxa_barplot_' + g + '_' + r + '_' + t.split('g__')[-1] + '.pdf')\n",
    "                plt.close() \n",
    "                sig_otu.append(t)\n",
    "        print('Number of sigtaxa')\n",
    "        print(len(sig_otu))\n",
    "        r_to_sigotu[r] = sig_otu\n",
    "    gr_to_sigotu[g] = r_to_sigotu\n",
    "\n",
    "to_sigotu['horizontal'] = gr_to_sigotu\n",
    "\n",
    "pd.DataFrame({'SampleType':gs, 'WOMAC_P_Response': rs, 'Timepoint': ts, 'WilcoxonRankSum': ws, 'pvalue': ps}).to_csv(path + 'outputs/jobs06/taxa_rvnr.tsv', sep='\\t')\n",
    "df_merge.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "88b10eda-ea18-4792-8a75-22696138d035",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saliva\n",
      "pre\n",
      "Number of sigtaxa\n",
      "2\n",
      "post\n",
      "Number of sigtaxa\n",
      "5\n",
      "stool\n",
      "pre\n",
      "Number of sigtaxa\n",
      "6\n",
      "post\n",
      "Number of sigtaxa\n",
      "4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "      <th>WOMAC_P_Response</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>0.014084</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.022027</td>\n",
       "      <td>0.001515</td>\n",
       "      <td>0.173283</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021230</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.152452</td>\n",
       "      <td>0.033001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>0.273829</td>\n",
       "      <td>0.067421</td>\n",
       "      <td>0.184572</td>\n",
       "      <td>0.026929</td>\n",
       "      <td>0.053673</td>\n",
       "      <td>0.037783</td>\n",
       "      <td>0.009619</td>\n",
       "      <td>0.023935</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.012060</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.post.stool</th>\n",
       "      <td>0.174773</td>\n",
       "      <td>0.040785</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001855</td>\n",
       "      <td>0.124402</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001400</td>\n",
       "      <td>0.000371</td>\n",
       "      <td>0.014265</td>\n",
       "      <td>0.090594</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-006.post.stool</th>\n",
       "      <td>0.038928</td>\n",
       "      <td>0.185602</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006453</td>\n",
       "      <td>0.020567</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>0.012512</td>\n",
       "      <td>0.000831</td>\n",
       "      <td>0.010729</td>\n",
       "      <td>0.174328</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00003</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-007.post.stool</th>\n",
       "      <td>0.168776</td>\n",
       "      <td>0.099334</td>\n",
       "      <td>0.000753</td>\n",
       "      <td>0.045066</td>\n",
       "      <td>0.228742</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010929</td>\n",
       "      <td>0.018782</td>\n",
       "      <td>0.039342</td>\n",
       "      <td>0.010033</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004  \\\n",
       "OAD-001.post.stool                                           0.014084                                                       \n",
       "OAD-003.post.stool                                           0.273829                                                       \n",
       "OAD-004.post.stool                                           0.174773                                                       \n",
       "OAD-006.post.stool                                           0.038928                                                       \n",
       "OAD-007.post.stool                                           0.168776                                                       \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium  \\\n",
       "OAD-001.post.stool                                           0.039046                                                            \n",
       "OAD-003.post.stool                                           0.067421                                                            \n",
       "OAD-004.post.stool                                           0.040785                                                            \n",
       "OAD-006.post.stool                                           0.185602                                                            \n",
       "OAD-007.post.stool                                           0.099334                                                            \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella  \\\n",
       "OAD-001.post.stool                                           0.022027                                             \n",
       "OAD-003.post.stool                                           0.184572                                             \n",
       "OAD-004.post.stool                                           0.000000                                             \n",
       "OAD-006.post.stool                                           0.000000                                             \n",
       "OAD-007.post.stool                                           0.000753                                             \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117  \\\n",
       "OAD-001.post.stool                                           0.001515                                                              \n",
       "OAD-003.post.stool                                           0.026929                                                              \n",
       "OAD-004.post.stool                                           0.001855                                                              \n",
       "OAD-006.post.stool                                           0.006453                                                              \n",
       "OAD-007.post.stool                                           0.045066                                                              \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H  \\\n",
       "OAD-001.post.stool                                           0.173283                                                \n",
       "OAD-003.post.stool                                           0.053673                                                \n",
       "OAD-004.post.stool                                           0.124402                                                \n",
       "OAD-006.post.stool                                           0.020567                                                \n",
       "OAD-007.post.stool                                           0.228742                                                \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum  \\\n",
       "OAD-001.post.stool                                           0.000000                                                  \n",
       "OAD-003.post.stool                                           0.037783                                                  \n",
       "OAD-004.post.stool                                           0.000000                                                  \n",
       "OAD-006.post.stool                                           0.000106                                                  \n",
       "OAD-007.post.stool                                           0.000000                                                  \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129  \\\n",
       "OAD-001.post.stool                                           0.021230                                                            \n",
       "OAD-003.post.stool                                           0.009619                                                            \n",
       "OAD-004.post.stool                                           0.001400                                                            \n",
       "OAD-006.post.stool                                           0.012512                                                            \n",
       "OAD-007.post.stool                                           0.010929                                                            \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus  \\\n",
       "OAD-001.post.stool                                           0.000160                                                \n",
       "OAD-003.post.stool                                           0.023935                                                \n",
       "OAD-004.post.stool                                           0.000371                                                \n",
       "OAD-006.post.stool                                           0.000831                                                \n",
       "OAD-007.post.stool                                           0.018782                                                \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor  \\\n",
       "OAD-001.post.stool                                           0.152452                                                        \n",
       "OAD-003.post.stool                                           0.005789                                                        \n",
       "OAD-004.post.stool                                           0.014265                                                        \n",
       "OAD-006.post.stool                                           0.010729                                                        \n",
       "OAD-007.post.stool                                           0.039342                                                        \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400  \\\n",
       "OAD-001.post.stool                                           0.033001                                                    \n",
       "OAD-003.post.stool                                           0.012060                                                    \n",
       "OAD-004.post.stool                                           0.090594                                                    \n",
       "OAD-006.post.stool                                           0.174328                                                    \n",
       "OAD-007.post.stool                                           0.010033                                                    \n",
       "\n",
       "                    ...  \\\n",
       "OAD-001.post.stool  ...   \n",
       "OAD-003.post.stool  ...   \n",
       "OAD-004.post.stool  ...   \n",
       "OAD-006.post.stool  ...   \n",
       "OAD-007.post.stool  ...   \n",
       "\n",
       "                    d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "OAD-001.post.stool                                                0.0                                                        \n",
       "OAD-003.post.stool                                                0.0                                                        \n",
       "OAD-004.post.stool                                                0.0                                                        \n",
       "OAD-006.post.stool                                                0.0                                                        \n",
       "OAD-007.post.stool                                                0.0                                                        \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "OAD-001.post.stool                                                0.0                 \n",
       "OAD-003.post.stool                                                0.0                 \n",
       "OAD-004.post.stool                                                0.0                 \n",
       "OAD-006.post.stool                                                0.0                 \n",
       "OAD-007.post.stool                                                0.0                 \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "OAD-001.post.stool                                                0.0                                                         \n",
       "OAD-003.post.stool                                                0.0                                                         \n",
       "OAD-004.post.stool                                                0.0                                                         \n",
       "OAD-006.post.stool                                                0.0                                                         \n",
       "OAD-007.post.stool                                                0.0                                                         \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "OAD-001.post.stool                                                0.0                                               \n",
       "OAD-003.post.stool                                                0.0                                               \n",
       "OAD-004.post.stool                                                0.0                                               \n",
       "OAD-006.post.stool                                                0.0                                               \n",
       "OAD-007.post.stool                                                0.0                                               \n",
       "\n",
       "                    d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "OAD-001.post.stool                                            0.00000                                                                         \n",
       "OAD-003.post.stool                                            0.00000                                                                         \n",
       "OAD-004.post.stool                                            0.00000                                                                         \n",
       "OAD-006.post.stool                                            0.00003                                                                         \n",
       "OAD-007.post.stool                                            0.00000                                                                         \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "OAD-001.post.stool                                           0.000000                                     \n",
       "OAD-003.post.stool                                           0.000000                                     \n",
       "OAD-004.post.stool                                           0.000000                                     \n",
       "OAD-006.post.stool                                           0.000000                                     \n",
       "OAD-007.post.stool                                           0.000026                                     \n",
       "\n",
       "                    d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "OAD-001.post.stool                                                0.0                                                                        \n",
       "OAD-003.post.stool                                                0.0                                                                        \n",
       "OAD-004.post.stool                                                0.0                                                                        \n",
       "OAD-006.post.stool                                                0.0                                                                        \n",
       "OAD-007.post.stool                                                0.0                                                                        \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "OAD-001.post.stool                                                0.0                                  \n",
       "OAD-003.post.stool                                                0.0                                  \n",
       "OAD-004.post.stool                                                0.0                                  \n",
       "OAD-006.post.stool                                                0.0                                  \n",
       "OAD-007.post.stool                                                0.0                                  \n",
       "\n",
       "                    d__Bacteria;p__Actinobacteriota;__;__;__;__  \\\n",
       "OAD-001.post.stool                                          0.0   \n",
       "OAD-003.post.stool                                          0.0   \n",
       "OAD-004.post.stool                                          0.0   \n",
       "OAD-006.post.stool                                          0.0   \n",
       "OAD-007.post.stool                                          0.0   \n",
       "\n",
       "                    WOMAC_P_Response  \n",
       "OAD-001.post.stool       No response  \n",
       "OAD-003.post.stool       No response  \n",
       "OAD-004.post.stool       No response  \n",
       "OAD-006.post.stool          Response  \n",
       "OAD-007.post.stool          Response  \n",
       "\n",
       "[5 rows x 375 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# paired testing in resp vs non resp \n",
    "# does pre vs post testing in each category alone\n",
    "gr_to_sigotu = {}\n",
    "\n",
    "gs = []\n",
    "ts = []\n",
    "xs = []\n",
    "ws = []\n",
    "ps = []\n",
    "\n",
    "for g in ['saliva','stool']:\n",
    "    print(g)\n",
    "    r_to_sigotu = {}\n",
    "    # grab taxa table\n",
    "    df_otu = gt_to_otu[g]['all']\n",
    "\n",
    "    df = type_to_df_map[g].copy()\n",
    "    df = df.set_index('Together')\n",
    "    if g == 'stool':\n",
    "        df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "    df = df[['Timepoint','WOMAC_P_Response']]\n",
    "    for t in ['pre', 'post']:\n",
    "        print(t)\n",
    "        df_merge = pd.concat([df_otu, df[df['Timepoint'] == t]], axis=1)\n",
    "        df_merge = df_merge.dropna()\n",
    "        df_merge['SampleID'] = df_merge.index.map(lambda x: x.split('.p')[0])\n",
    "        #df_vc = df_merge['SampleID'].value_counts()\n",
    "        #keep = df_vc[df_vc == 2].index.values\n",
    "        #df_merge = df_merge[df_merge['SampleID'].isin(keep)] \n",
    "        df_merge = df_merge.drop(['SampleID','Timepoint'],axis=1)\n",
    "        sig_otu = []\n",
    "        for x in list(df_merge.columns.values)[:-1]:\n",
    "            df_r = df_merge[df_merge['WOMAC_P_Response'] == 'Response'][x]\n",
    "            df_nr = df_merge[df_merge['WOMAC_P_Response'] == 'No response'][x]\n",
    "\n",
    "            try:\n",
    "                w,p = scipy.stats.mannwhitneyu(df_r.values, df_nr.values, nan_policy='omit')\n",
    "            except:\n",
    "                w,p = 0, 1\n",
    "            if p < 0.05:\n",
    "\n",
    "                gs.append(g)\n",
    "                ts.append(t)\n",
    "                xs.append(x)\n",
    "                ws.append(w)\n",
    "                ps.append(p)\n",
    "\n",
    "                ax = sns.boxplot(data=df_merge, y=x, x='WOMAC_P_Response')\n",
    "                sns.swarmplot(data=df_merge, y=x, x='WOMAC_P_Response', palette='dark:grey')\n",
    "                sns.despine()\n",
    "            \n",
    "                plt.tight_layout()\n",
    "                plt.savefig(path + 'outputs/jobs06/pvp_taxa_barplot_' + g + '_' + t + '_' + x.split('g__')[-1] + '.pdf')\n",
    "                plt.close() \n",
    "                sig_otu.append(x)\n",
    "        print('Number of sigtaxa')\n",
    "        print(len(sig_otu))\n",
    "        r_to_sigotu[t] = sig_otu\n",
    "    gr_to_sigotu[g] = r_to_sigotu\n",
    "\n",
    "to_sigotu['vertical'] = gr_to_sigotu\n",
    "\n",
    "pd.DataFrame({'SampleType':gs, 'WOMAC_P_Response': xs, 'Timepoint': ts, 'WilcoxonRankSum': ws, 'pvalue': ps}).to_csv(path + 'outputs/jobs06/taxa_pvp.tsv', sep='\\t')\n",
    "df_merge.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8b418d85-6211-4131-9007-92d2b5f1a3ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of 12-, 14- and 16-membered macrolides</th>\n",
       "      <th>Cellular Processes|Cellular community - eukaryotes|Signaling pathways regulating pluripotency of stem cells</th>\n",
       "      <th>Environmental Information Processing|Signal transduction|Plant hormone signal transduction</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of enediyne antibiotics</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "      <th>Organismal Systems|Endocrine system|Relaxin signaling pathway</th>\n",
       "      <th>Human Diseases|Endocrine and metabolic disease|AGE-RAGE signaling pathway in diabetic complications</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.saliva</th>\n",
       "      <td>0.155479</td>\n",
       "      <td>0.025863</td>\n",
       "      <td>0.001866</td>\n",
       "      <td>0.003177</td>\n",
       "      <td>0.008013</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.026699</td>\n",
       "      <td>0.000547</td>\n",
       "      <td>0.031793</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>...</td>\n",
       "      <td>2.433861e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.216931e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.saliva</th>\n",
       "      <td>0.159679</td>\n",
       "      <td>0.027060</td>\n",
       "      <td>0.001949</td>\n",
       "      <td>0.003082</td>\n",
       "      <td>0.007239</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.027470</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.034357</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>...</td>\n",
       "      <td>1.911111e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.539150e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.saliva</th>\n",
       "      <td>0.151338</td>\n",
       "      <td>0.027415</td>\n",
       "      <td>0.002223</td>\n",
       "      <td>0.002702</td>\n",
       "      <td>0.009135</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.021911</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.030358</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.saliva</th>\n",
       "      <td>0.152707</td>\n",
       "      <td>0.026974</td>\n",
       "      <td>0.002172</td>\n",
       "      <td>0.002788</td>\n",
       "      <td>0.008887</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.022846</td>\n",
       "      <td>0.000257</td>\n",
       "      <td>0.031226</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.post.saliva</th>\n",
       "      <td>0.159885</td>\n",
       "      <td>0.027003</td>\n",
       "      <td>0.001989</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>0.007319</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.026641</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.034499</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>...</td>\n",
       "      <td>6.002574e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.594595e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 443 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "KEGG_Pathway         Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.post.saliva    0.155479                            0.025863   \n",
       "OAD-001.pre.saliva     0.159679                            0.027060   \n",
       "OAD-003.post.saliva    0.151338                            0.027415   \n",
       "OAD-003.pre.saliva     0.152707                            0.026974   \n",
       "OAD-004.post.saliva    0.159885                            0.027003   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.post.saliva                                           0.001866                 \n",
       "OAD-001.pre.saliva                                            0.001949                 \n",
       "OAD-003.post.saliva                                           0.002223                 \n",
       "OAD-003.pre.saliva                                            0.002172                 \n",
       "OAD-004.post.saliva                                           0.001989                 \n",
       "\n",
       "KEGG_Pathway         Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.post.saliva                                           0.003177        \n",
       "OAD-001.pre.saliva                                            0.003082        \n",
       "OAD-003.post.saliva                                           0.002702        \n",
       "OAD-003.pre.saliva                                            0.002788        \n",
       "OAD-004.post.saliva                                           0.002953        \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism  \\\n",
       "OAD-001.post.saliva                     0.008013   \n",
       "OAD-001.pre.saliva                      0.007239   \n",
       "OAD-003.post.saliva                     0.009135   \n",
       "OAD-003.pre.saliva                      0.008887   \n",
       "OAD-004.post.saliva                     0.007319   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.post.saliva                                           0.000273    \n",
       "OAD-001.pre.saliva                                            0.000166    \n",
       "OAD-003.post.saliva                                           0.000180    \n",
       "OAD-003.pre.saliva                                            0.000176    \n",
       "OAD-004.post.saliva                                           0.000132    \n",
       "\n",
       "KEGG_Pathway         Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.post.saliva                          0.026699   \n",
       "OAD-001.pre.saliva                           0.027470   \n",
       "OAD-003.post.saliva                          0.021911   \n",
       "OAD-003.pre.saliva                           0.022846   \n",
       "OAD-004.post.saliva                          0.026641   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.post.saliva                                           0.000547      \n",
       "OAD-001.pre.saliva                                            0.000309      \n",
       "OAD-003.post.saliva                                           0.000261      \n",
       "OAD-003.pre.saliva                                            0.000257      \n",
       "OAD-004.post.saliva                                           0.000200      \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.post.saliva                                         0.031793   \n",
       "OAD-001.pre.saliva                                          0.034357   \n",
       "OAD-003.post.saliva                                         0.030358   \n",
       "OAD-003.pre.saliva                                          0.031226   \n",
       "OAD-004.post.saliva                                         0.034499   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.post.saliva                                           0.000046                    \n",
       "OAD-001.pre.saliva                                            0.000045                    \n",
       "OAD-003.post.saliva                                           0.000042                    \n",
       "OAD-003.pre.saliva                                            0.000041                    \n",
       "OAD-004.post.saliva                                           0.000038                    \n",
       "\n",
       "KEGG_Pathway         ...  \\\n",
       "OAD-001.post.saliva  ...   \n",
       "OAD-001.pre.saliva   ...   \n",
       "OAD-003.post.saliva  ...   \n",
       "OAD-003.pre.saliva   ...   \n",
       "OAD-004.post.saliva  ...   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.post.saliva                                       2.433861e-07                                   \n",
       "OAD-001.pre.saliva                                        1.911111e-06                                   \n",
       "OAD-003.post.saliva                                       0.000000e+00                                   \n",
       "OAD-003.pre.saliva                                        0.000000e+00                                   \n",
       "OAD-004.post.saliva                                       6.002574e-07                                   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.post.saliva                                                0.0                                  \n",
       "OAD-001.pre.saliva                                                 0.0                                  \n",
       "OAD-003.post.saliva                                                0.0                                  \n",
       "OAD-003.pre.saliva                                                 0.0                                  \n",
       "OAD-004.post.saliva                                                0.0                                  \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of 12-, 14- and 16-membered macrolides  \\\n",
       "OAD-001.post.saliva                                                0.0                                                         \n",
       "OAD-001.pre.saliva                                                 0.0                                                         \n",
       "OAD-003.post.saliva                                                0.0                                                         \n",
       "OAD-003.pre.saliva                                                 0.0                                                         \n",
       "OAD-004.post.saliva                                                0.0                                                         \n",
       "\n",
       "KEGG_Pathway         Cellular Processes|Cellular community - eukaryotes|Signaling pathways regulating pluripotency of stem cells  \\\n",
       "OAD-001.post.saliva                                                0.0                                                             \n",
       "OAD-001.pre.saliva                                                 0.0                                                             \n",
       "OAD-003.post.saliva                                                0.0                                                             \n",
       "OAD-003.pre.saliva                                                 0.0                                                             \n",
       "OAD-004.post.saliva                                                0.0                                                             \n",
       "\n",
       "KEGG_Pathway         Environmental Information Processing|Signal transduction|Plant hormone signal transduction  \\\n",
       "OAD-001.post.saliva                                                0.0                                            \n",
       "OAD-001.pre.saliva                                                 0.0                                            \n",
       "OAD-003.post.saliva                                                0.0                                            \n",
       "OAD-003.pre.saliva                                                 0.0                                            \n",
       "OAD-004.post.saliva                                                0.0                                            \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of enediyne antibiotics  \\\n",
       "OAD-001.post.saliva                                                0.0                                          \n",
       "OAD-001.pre.saliva                                                 0.0                                          \n",
       "OAD-003.post.saliva                                                0.0                                          \n",
       "OAD-003.pre.saliva                                                 0.0                                          \n",
       "OAD-004.post.saliva                                                0.0                                          \n",
       "\n",
       "KEGG_Pathway         Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.post.saliva                                       0.000000e+00                           \n",
       "OAD-001.pre.saliva                                        0.000000e+00                           \n",
       "OAD-003.post.saliva                                       0.000000e+00                           \n",
       "OAD-003.pre.saliva                                        0.000000e+00                           \n",
       "OAD-004.post.saliva                                       8.594595e-07                           \n",
       "\n",
       "KEGG_Pathway         Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \\\n",
       "OAD-001.post.saliva                                       1.216931e-07                                          \n",
       "OAD-001.pre.saliva                                        1.539150e-07                                          \n",
       "OAD-003.post.saliva                                       0.000000e+00                                          \n",
       "OAD-003.pre.saliva                                        0.000000e+00                                          \n",
       "OAD-004.post.saliva                                       0.000000e+00                                          \n",
       "\n",
       "KEGG_Pathway         Organismal Systems|Endocrine system|Relaxin signaling pathway  \\\n",
       "OAD-001.post.saliva                                                0.0               \n",
       "OAD-001.pre.saliva                                                 0.0               \n",
       "OAD-003.post.saliva                                                0.0               \n",
       "OAD-003.pre.saliva                                                 0.0               \n",
       "OAD-004.post.saliva                                                0.0               \n",
       "\n",
       "KEGG_Pathway         Human Diseases|Endocrine and metabolic disease|AGE-RAGE signaling pathway in diabetic complications  \n",
       "OAD-001.post.saliva                                                0.0                                                    \n",
       "OAD-001.pre.saliva                                                 0.0                                                    \n",
       "OAD-003.post.saliva                                                0.0                                                    \n",
       "OAD-003.pre.saliva                                                 0.0                                                    \n",
       "OAD-004.post.saliva                                                0.0                                                    \n",
       "\n",
       "[5 rows x 443 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specific pathway profiling\n",
    "g_to_path = {}\n",
    "\n",
    "groups = ['saliva_adh', 'stool_adh']\n",
    "g_to_job = {\n",
    "    'saliva_adh': 'jobs07',\n",
    "    'stool_adh': 'jobs08'\n",
    "}\n",
    "for g in groups:\n",
    "    # read in csv\n",
    "    df_path = pd.read_csv(path + 'outputs/' + g_to_job[g] + '/pred_metagenome_unstrat_kegg.tsv', sep ='\\t', index_col=0)\n",
    "\n",
    "    # transpose\n",
    "    df_path = df_path.T\n",
    "    \n",
    "    # normalize by dividing by row sum\n",
    "    df_path = df_path.div(df_path.sum(axis=1),axis=0)\n",
    "\n",
    "    # save to dict\n",
    "    g_to_path[g] = df_path\n",
    "                                 \n",
    "g_to_path['saliva_adh'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "34ed8f12-5226-47f7-b748-07609e554b8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Human Diseases|Infectious disease: bacterial|Bacterial invasion of epithelial cells</th>\n",
       "      <th>Human Diseases|Infectious disease: bacterial|Yersinia infection</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Cellular Processes|Cellular community - eukaryotes|Signaling pathways regulating pluripotency of stem cells</th>\n",
       "      <th>Environmental Information Processing|Signal transduction|Plant hormone signal transduction</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Environmental Information Processing|Signal transduction|Rap1 signaling pathway</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.159285</td>\n",
       "      <td>0.035029</td>\n",
       "      <td>0.002320</td>\n",
       "      <td>0.003018</td>\n",
       "      <td>0.007937</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>0.027626</td>\n",
       "      <td>0.000229</td>\n",
       "      <td>0.026349</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>6.262338e-06</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.063603e-06</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.510913e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.161049</td>\n",
       "      <td>0.033134</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.002538</td>\n",
       "      <td>0.008606</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.029092</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.027987</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>2.013513e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.511851e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.161760</td>\n",
       "      <td>0.034016</td>\n",
       "      <td>0.002254</td>\n",
       "      <td>0.002773</td>\n",
       "      <td>0.008658</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.029185</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>0.027522</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>7.082122e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.183607e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.100744e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-005.pre.stool</th>\n",
       "      <td>0.158015</td>\n",
       "      <td>0.033087</td>\n",
       "      <td>0.002185</td>\n",
       "      <td>0.002702</td>\n",
       "      <td>0.008593</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.028749</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>0.027538</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>1.593033e-06</td>\n",
       "      <td>1.103744e-07</td>\n",
       "      <td>3.679146e-08</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.297970e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-006.pre.stool</th>\n",
       "      <td>0.157791</td>\n",
       "      <td>0.032210</td>\n",
       "      <td>0.002112</td>\n",
       "      <td>0.002819</td>\n",
       "      <td>0.008727</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.030094</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.026678</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.127712e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 406 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "KEGG_Pathway       Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.pre.stool    0.159285                            0.035029   \n",
       "OAD-003.pre.stool    0.161049                            0.033134   \n",
       "OAD-004.pre.stool    0.161760                            0.034016   \n",
       "OAD-005.pre.stool    0.158015                            0.033087   \n",
       "OAD-006.pre.stool    0.157791                            0.032210   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.pre.stool                                           0.002320                 \n",
       "OAD-003.pre.stool                                           0.002184                 \n",
       "OAD-004.pre.stool                                           0.002254                 \n",
       "OAD-005.pre.stool                                           0.002185                 \n",
       "OAD-006.pre.stool                                           0.002112                 \n",
       "\n",
       "KEGG_Pathway       Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.pre.stool                                           0.003018        \n",
       "OAD-003.pre.stool                                           0.002538        \n",
       "OAD-004.pre.stool                                           0.002773        \n",
       "OAD-005.pre.stool                                           0.002702        \n",
       "OAD-006.pre.stool                                           0.002819        \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism  \\\n",
       "OAD-001.pre.stool                     0.007937   \n",
       "OAD-003.pre.stool                     0.008606   \n",
       "OAD-004.pre.stool                     0.008658   \n",
       "OAD-005.pre.stool                     0.008593   \n",
       "OAD-006.pre.stool                     0.008727   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.pre.stool                                           0.000192    \n",
       "OAD-003.pre.stool                                           0.000180    \n",
       "OAD-004.pre.stool                                           0.000157    \n",
       "OAD-005.pre.stool                                           0.000186    \n",
       "OAD-006.pre.stool                                           0.000199    \n",
       "\n",
       "KEGG_Pathway       Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.pre.stool                          0.027626   \n",
       "OAD-003.pre.stool                          0.029092   \n",
       "OAD-004.pre.stool                          0.029185   \n",
       "OAD-005.pre.stool                          0.028749   \n",
       "OAD-006.pre.stool                          0.030094   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.pre.stool                                           0.000229      \n",
       "OAD-003.pre.stool                                           0.000194      \n",
       "OAD-004.pre.stool                                           0.000173      \n",
       "OAD-005.pre.stool                                           0.000216      \n",
       "OAD-006.pre.stool                                           0.000220      \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.pre.stool                                         0.026349   \n",
       "OAD-003.pre.stool                                         0.027987   \n",
       "OAD-004.pre.stool                                         0.027522   \n",
       "OAD-005.pre.stool                                         0.027538   \n",
       "OAD-006.pre.stool                                         0.026678   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.pre.stool                                           0.000012                    \n",
       "OAD-003.pre.stool                                           0.000019                    \n",
       "OAD-004.pre.stool                                           0.000016                    \n",
       "OAD-005.pre.stool                                           0.000016                    \n",
       "OAD-006.pre.stool                                           0.000013                    \n",
       "\n",
       "KEGG_Pathway       ...  \\\n",
       "OAD-001.pre.stool  ...   \n",
       "OAD-003.pre.stool  ...   \n",
       "OAD-004.pre.stool  ...   \n",
       "OAD-005.pre.stool  ...   \n",
       "OAD-006.pre.stool  ...   \n",
       "\n",
       "KEGG_Pathway       Human Diseases|Infectious disease: bacterial|Bacterial invasion of epithelial cells  \\\n",
       "OAD-001.pre.stool                                           0.000008                                     \n",
       "OAD-003.pre.stool                                           0.000011                                     \n",
       "OAD-004.pre.stool                                           0.000001                                     \n",
       "OAD-005.pre.stool                                           0.000014                                     \n",
       "OAD-006.pre.stool                                           0.000002                                     \n",
       "\n",
       "KEGG_Pathway       Human Diseases|Infectious disease: bacterial|Yersinia infection  \\\n",
       "OAD-001.pre.stool                                       6.262338e-06                 \n",
       "OAD-003.pre.stool                                       2.013513e-07                 \n",
       "OAD-004.pre.stool                                       7.082122e-07                 \n",
       "OAD-005.pre.stool                                       1.593033e-06                 \n",
       "OAD-006.pre.stool                                       0.000000e+00                 \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.pre.stool                                       0.000000e+00                                   \n",
       "OAD-003.pre.stool                                       0.000000e+00                                   \n",
       "OAD-004.pre.stool                                       0.000000e+00                                   \n",
       "OAD-005.pre.stool                                       1.103744e-07                                   \n",
       "OAD-006.pre.stool                                       0.000000e+00                                   \n",
       "\n",
       "KEGG_Pathway       Cellular Processes|Cellular community - eukaryotes|Signaling pathways regulating pluripotency of stem cells  \\\n",
       "OAD-001.pre.stool                                       1.063603e-06                                                             \n",
       "OAD-003.pre.stool                                       0.000000e+00                                                             \n",
       "OAD-004.pre.stool                                       0.000000e+00                                                             \n",
       "OAD-005.pre.stool                                       3.679146e-08                                                             \n",
       "OAD-006.pre.stool                                       0.000000e+00                                                             \n",
       "\n",
       "KEGG_Pathway       Environmental Information Processing|Signal transduction|Plant hormone signal transduction  \\\n",
       "OAD-001.pre.stool                                       0.000000e+00                                            \n",
       "OAD-003.pre.stool                                       0.000000e+00                                            \n",
       "OAD-004.pre.stool                                       3.183607e-07                                            \n",
       "OAD-005.pre.stool                                       0.000000e+00                                            \n",
       "OAD-006.pre.stool                                       0.000000e+00                                            \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.pre.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                0.0                                  \n",
       "OAD-004.pre.stool                                                0.0                                  \n",
       "OAD-005.pre.stool                                                0.0                                  \n",
       "OAD-006.pre.stool                                                0.0                                  \n",
       "\n",
       "KEGG_Pathway       Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.pre.stool                                       1.510913e-06                           \n",
       "OAD-003.pre.stool                                       8.511851e-07                           \n",
       "OAD-004.pre.stool                                       6.100744e-07                           \n",
       "OAD-005.pre.stool                                       5.297970e-07                           \n",
       "OAD-006.pre.stool                                       1.127712e-07                           \n",
       "\n",
       "KEGG_Pathway       Environmental Information Processing|Signal transduction|Rap1 signaling pathway  \\\n",
       "OAD-001.pre.stool                                                0.0                                 \n",
       "OAD-003.pre.stool                                                0.0                                 \n",
       "OAD-004.pre.stool                                                0.0                                 \n",
       "OAD-005.pre.stool                                                0.0                                 \n",
       "OAD-006.pre.stool                                                0.0                                 \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis  \\\n",
       "OAD-001.pre.stool                                                0.0                               \n",
       "OAD-003.pre.stool                                                0.0                               \n",
       "OAD-004.pre.stool                                                0.0                               \n",
       "OAD-005.pre.stool                                                0.0                               \n",
       "OAD-006.pre.stool                                                0.0                               \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \n",
       "OAD-001.pre.stool                                                0.0                                         \n",
       "OAD-003.pre.stool                                                0.0                                         \n",
       "OAD-004.pre.stool                                                0.0                                         \n",
       "OAD-005.pre.stool                                                0.0                                         \n",
       "OAD-006.pre.stool                                                0.0                                         \n",
       "\n",
       "[5 rows x 406 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pathway testing\n",
    "gt_to_path = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gt_to_path[g] = {}\n",
    "    df_path = g_to_path[g + '_adh'].copy()\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['post','pre']:\n",
    "        df = df_path[df_path.index.to_series().str.contains(t)]\n",
    "        gt_to_path[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_path[g]['pre'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.split('.pre')[0])\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('-',''))\n",
    "    post_df = gt_to_path[g]['post'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.split('.post')[0])\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('-',''))\n",
    "    \n",
    "    gt_to_path[g]['diff'] = post_df - pre_df\n",
    "    gt_to_path[g]['all'] = pd.concat([gt_to_path[g]['pre'],gt_to_path[g]['post']])\n",
    "\n",
    "gt_to_path[g][t].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "50230423-0a1a-48e8-a2b9-5990f87f2bca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saliva\n",
      "Response\n",
      "Number of sigpath\n",
      "32\n",
      "No response\n",
      "Number of sigpath\n",
      "0\n",
      "stool\n",
      "Response\n",
      "Number of sigpath\n",
      "22\n",
      "No response\n",
      "Number of sigpath\n",
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Human Diseases|Infectious disease: bacterial|Yersinia infection</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Cellular Processes|Cellular community - eukaryotes|Signaling pathways regulating pluripotency of stem cells</th>\n",
       "      <th>Environmental Information Processing|Signal transduction|Plant hormone signal transduction</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Environmental Information Processing|Signal transduction|Rap1 signaling pathway</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "      <th>Timepoint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.159285</td>\n",
       "      <td>0.035029</td>\n",
       "      <td>0.002320</td>\n",
       "      <td>0.003018</td>\n",
       "      <td>0.007937</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>0.027626</td>\n",
       "      <td>0.000229</td>\n",
       "      <td>0.026349</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>...</td>\n",
       "      <td>6.262338e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.510913e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.161049</td>\n",
       "      <td>0.033134</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.002538</td>\n",
       "      <td>0.008606</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.029092</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.027987</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>...</td>\n",
       "      <td>2.013513e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.511851e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.161760</td>\n",
       "      <td>0.034016</td>\n",
       "      <td>0.002254</td>\n",
       "      <td>0.002773</td>\n",
       "      <td>0.008658</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.029185</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>0.027522</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>7.082122e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.183607e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.100744e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-001.pre.stool</th>\n",
       "      <td>0.163414</td>\n",
       "      <td>0.036621</td>\n",
       "      <td>0.002214</td>\n",
       "      <td>0.002779</td>\n",
       "      <td>0.008978</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>0.028735</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>...</td>\n",
       "      <td>1.303797e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.026500e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.425694e-07</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-012.pre.stool</th>\n",
       "      <td>0.164662</td>\n",
       "      <td>0.036528</td>\n",
       "      <td>0.002363</td>\n",
       "      <td>0.002908</td>\n",
       "      <td>0.009032</td>\n",
       "      <td>0.000170</td>\n",
       "      <td>0.028994</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.027414</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.141516e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.499616e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 407 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.pre.stool      0.159285                            0.035029   \n",
       "OAD-003.pre.stool      0.161049                            0.033134   \n",
       "OAD-004.pre.stool      0.161760                            0.034016   \n",
       "VAOAD-001.pre.stool    0.163414                            0.036621   \n",
       "VAOAD-012.pre.stool    0.164662                            0.036528   \n",
       "\n",
       "                     Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.pre.stool                                             0.002320                 \n",
       "OAD-003.pre.stool                                             0.002184                 \n",
       "OAD-004.pre.stool                                             0.002254                 \n",
       "VAOAD-001.pre.stool                                           0.002214                 \n",
       "VAOAD-012.pre.stool                                           0.002363                 \n",
       "\n",
       "                     Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.pre.stool                                             0.003018        \n",
       "OAD-003.pre.stool                                             0.002538        \n",
       "OAD-004.pre.stool                                             0.002773        \n",
       "VAOAD-001.pre.stool                                           0.002779        \n",
       "VAOAD-012.pre.stool                                           0.002908        \n",
       "\n",
       "                     Metabolism|Lipid metabolism  \\\n",
       "OAD-001.pre.stool                       0.007937   \n",
       "OAD-003.pre.stool                       0.008606   \n",
       "OAD-004.pre.stool                       0.008658   \n",
       "VAOAD-001.pre.stool                     0.008978   \n",
       "VAOAD-012.pre.stool                     0.009032   \n",
       "\n",
       "                     Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.pre.stool                                             0.000192    \n",
       "OAD-003.pre.stool                                             0.000180    \n",
       "OAD-004.pre.stool                                             0.000157    \n",
       "VAOAD-001.pre.stool                                           0.000195    \n",
       "VAOAD-012.pre.stool                                           0.000170    \n",
       "\n",
       "                     Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.pre.stool                            0.027626   \n",
       "OAD-003.pre.stool                            0.029092   \n",
       "OAD-004.pre.stool                            0.029185   \n",
       "VAOAD-001.pre.stool                          0.028735   \n",
       "VAOAD-012.pre.stool                          0.028994   \n",
       "\n",
       "                     Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.pre.stool                                             0.000229      \n",
       "OAD-003.pre.stool                                             0.000194      \n",
       "OAD-004.pre.stool                                             0.000173      \n",
       "VAOAD-001.pre.stool                                           0.000203      \n",
       "VAOAD-012.pre.stool                                           0.000194      \n",
       "\n",
       "                     Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.pre.stool                                           0.026349   \n",
       "OAD-003.pre.stool                                           0.027987   \n",
       "OAD-004.pre.stool                                           0.027522   \n",
       "VAOAD-001.pre.stool                                         0.026849   \n",
       "VAOAD-012.pre.stool                                         0.027414   \n",
       "\n",
       "                     Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.pre.stool                                             0.000012                    \n",
       "OAD-003.pre.stool                                             0.000019                    \n",
       "OAD-004.pre.stool                                             0.000016                    \n",
       "VAOAD-001.pre.stool                                           0.000027                    \n",
       "VAOAD-012.pre.stool                                           0.000013                    \n",
       "\n",
       "                     ...  \\\n",
       "OAD-001.pre.stool    ...   \n",
       "OAD-003.pre.stool    ...   \n",
       "OAD-004.pre.stool    ...   \n",
       "VAOAD-001.pre.stool  ...   \n",
       "VAOAD-012.pre.stool  ...   \n",
       "\n",
       "                     Human Diseases|Infectious disease: bacterial|Yersinia infection  \\\n",
       "OAD-001.pre.stool                                         6.262338e-06                 \n",
       "OAD-003.pre.stool                                         2.013513e-07                 \n",
       "OAD-004.pre.stool                                         7.082122e-07                 \n",
       "VAOAD-001.pre.stool                                       1.303797e-07                 \n",
       "VAOAD-012.pre.stool                                       0.000000e+00                 \n",
       "\n",
       "                     Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.pre.stool                                                  0.0                                   \n",
       "OAD-003.pre.stool                                                  0.0                                   \n",
       "OAD-004.pre.stool                                                  0.0                                   \n",
       "VAOAD-001.pre.stool                                                0.0                                   \n",
       "VAOAD-012.pre.stool                                                0.0                                   \n",
       "\n",
       "                     Cellular Processes|Cellular community - eukaryotes|Signaling pathways regulating pluripotency of stem cells  \\\n",
       "OAD-001.pre.stool                                             0.000001                                                             \n",
       "OAD-003.pre.stool                                             0.000000                                                             \n",
       "OAD-004.pre.stool                                             0.000000                                                             \n",
       "VAOAD-001.pre.stool                                           0.000000                                                             \n",
       "VAOAD-012.pre.stool                                           0.000000                                                             \n",
       "\n",
       "                     Environmental Information Processing|Signal transduction|Plant hormone signal transduction  \\\n",
       "OAD-001.pre.stool                                         0.000000e+00                                            \n",
       "OAD-003.pre.stool                                         0.000000e+00                                            \n",
       "OAD-004.pre.stool                                         3.183607e-07                                            \n",
       "VAOAD-001.pre.stool                                       0.000000e+00                                            \n",
       "VAOAD-012.pre.stool                                       3.141516e-07                                            \n",
       "\n",
       "                     Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.pre.stool                                                  0.0                                  \n",
       "OAD-003.pre.stool                                                  0.0                                  \n",
       "OAD-004.pre.stool                                                  0.0                                  \n",
       "VAOAD-001.pre.stool                                                0.0                                  \n",
       "VAOAD-012.pre.stool                                                0.0                                  \n",
       "\n",
       "                     Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.pre.stool                                         1.510913e-06                           \n",
       "OAD-003.pre.stool                                         8.511851e-07                           \n",
       "OAD-004.pre.stool                                         6.100744e-07                           \n",
       "VAOAD-001.pre.stool                                       1.026500e-06                           \n",
       "VAOAD-012.pre.stool                                       3.499616e-07                           \n",
       "\n",
       "                     Environmental Information Processing|Signal transduction|Rap1 signaling pathway  \\\n",
       "OAD-001.pre.stool                                                  0.0                                 \n",
       "OAD-003.pre.stool                                                  0.0                                 \n",
       "OAD-004.pre.stool                                                  0.0                                 \n",
       "VAOAD-001.pre.stool                                                0.0                                 \n",
       "VAOAD-012.pre.stool                                                0.0                                 \n",
       "\n",
       "                     Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis  \\\n",
       "OAD-001.pre.stool                                                  0.0                               \n",
       "OAD-003.pre.stool                                                  0.0                               \n",
       "OAD-004.pre.stool                                                  0.0                               \n",
       "VAOAD-001.pre.stool                                                0.0                               \n",
       "VAOAD-012.pre.stool                                                0.0                               \n",
       "\n",
       "                     Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \\\n",
       "OAD-001.pre.stool                                         0.000000e+00                                          \n",
       "OAD-003.pre.stool                                         0.000000e+00                                          \n",
       "OAD-004.pre.stool                                         0.000000e+00                                          \n",
       "VAOAD-001.pre.stool                                       1.425694e-07                                          \n",
       "VAOAD-012.pre.stool                                       0.000000e+00                                          \n",
       "\n",
       "                     Timepoint  \n",
       "OAD-001.pre.stool          pre  \n",
       "OAD-003.pre.stool          pre  \n",
       "OAD-004.pre.stool          pre  \n",
       "VAOAD-001.pre.stool        pre  \n",
       "VAOAD-012.pre.stool        pre  \n",
       "\n",
       "[5 rows x 407 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# paired testing in resp vs non resp \n",
    "to_sigpath = {}\n",
    "\n",
    "gr_to_sigpath = {} \n",
    "\n",
    "gs = []\n",
    "rs = []\n",
    "ts = []\n",
    "ws = []\n",
    "ps = []\n",
    "\n",
    "for g in ['saliva','stool']:#'stool_adh']:\n",
    "    print(g)\n",
    "    r_to_sigpath = {}\n",
    "    # grab taxa table\n",
    "    df_path = gt_to_path[g]['all']\n",
    "\n",
    "    df = type_to_df_map[g].copy()\n",
    "    df = df.set_index('Together')\n",
    "    if g == 'stool':\n",
    "        df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "    df = df[['Timepoint','WOMAC_P_Response']]\n",
    "    for r in ['Response', 'No response']:\n",
    "        print(r)\n",
    "        sig_path = []\n",
    "        df_merge = pd.concat([df_path, df[df['WOMAC_P_Response'] == r]], axis=1)\n",
    "        df_merge = df_merge.dropna()\n",
    "        df_merge['SampleID'] = df_merge.index.map(lambda x: x.split('.p')[0])\n",
    "        df_vc = df_merge['SampleID'].value_counts()\n",
    "        keep = df_vc[df_vc == 2].index.values\n",
    "        df_merge = df_merge[df_merge['SampleID'].isin(keep)] \n",
    "        df_merge = df_merge.drop(['SampleID','WOMAC_P_Response'],axis=1)\n",
    "\n",
    "        for t in list(df_merge.columns.values)[:-1]:\n",
    "            df_pre = df_merge[df_merge['Timepoint'] == 'pre'][t]\n",
    "            df_post = df_merge[df_merge['Timepoint'] == 'post'][t]\n",
    "\n",
    "            try:\n",
    "                w,p = scipy.stats.wilcoxon(df_pre.values, df_post.values, nan_policy='omit')\n",
    "            except:\n",
    "                w,p = 0, 1\n",
    "            if p < 0.05:\n",
    "\n",
    "                gs.append(g)\n",
    "                rs.append(r)\n",
    "                ts.append(t)\n",
    "                ws.append(w)\n",
    "                ps.append(p)\n",
    "   \n",
    "                ax = sns.boxplot(data=df_merge, y=t, x='Timepoint')\n",
    "                sns.swarmplot(data=df_merge, y=t, x='Timepoint', palette='dark:grey')\n",
    "                sns.despine()\n",
    "            \n",
    "                plt.tight_layout()\n",
    "                plt.savefig(path + 'outputs/jobs06/rvnr_path_barplot_' + g + '_' + r + '_' + t.split('g__')[-1] + '.pdf')\n",
    "                plt.close()    \n",
    "                sig_path.append(t)\n",
    "        print('Number of sigpath')\n",
    "        print(len(sig_path))\n",
    "        r_to_sigpath[r] = sig_path\n",
    "    gr_to_sigpath[g] = r_to_sigpath\n",
    "\n",
    "pd.DataFrame({'SampleType':gs, 'WOMAC_P_Response': rs, 'Timepoint': ts, 'WilcoxonRankSum': ws, 'pvalue': ps}).to_csv(path + 'outputs/jobs06/path_rvnr.tsv', sep='\\t')\n",
    "\n",
    "\n",
    "to_sigpath['horizontal'] = gr_to_sigpath\n",
    "\n",
    "df_merge.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "aea74477-f627-46f2-80d5-63477a813826",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saliva\n",
      "pre\n",
      "Number of sigpath\n",
      "3\n",
      "post\n",
      "Number of sigpath\n",
      "6\n",
      "stool\n",
      "pre\n",
      "Number of sigpath\n",
      "6\n",
      "post\n",
      "Number of sigpath\n",
      "11\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Human Diseases|Infectious disease: bacterial|Yersinia infection</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Cellular Processes|Cellular community - eukaryotes|Signaling pathways regulating pluripotency of stem cells</th>\n",
       "      <th>Environmental Information Processing|Signal transduction|Plant hormone signal transduction</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Environmental Information Processing|Signal transduction|Rap1 signaling pathway</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "      <th>WOMAC_P_Response</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>0.158589</td>\n",
       "      <td>0.033424</td>\n",
       "      <td>0.002252</td>\n",
       "      <td>0.002844</td>\n",
       "      <td>0.008684</td>\n",
       "      <td>0.000159</td>\n",
       "      <td>0.029109</td>\n",
       "      <td>0.000198</td>\n",
       "      <td>0.026486</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>...</td>\n",
       "      <td>2.050906e-06</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>9.696548e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.119511e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>0.164939</td>\n",
       "      <td>0.033380</td>\n",
       "      <td>0.002203</td>\n",
       "      <td>0.002569</td>\n",
       "      <td>0.008234</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.029464</td>\n",
       "      <td>0.000243</td>\n",
       "      <td>0.028377</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>...</td>\n",
       "      <td>2.408971e-06</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.566727e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.post.stool</th>\n",
       "      <td>0.165587</td>\n",
       "      <td>0.035285</td>\n",
       "      <td>0.002349</td>\n",
       "      <td>0.002926</td>\n",
       "      <td>0.009046</td>\n",
       "      <td>0.000168</td>\n",
       "      <td>0.029548</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>0.027896</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>8.859027e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.261718e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.110141e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-006.post.stool</th>\n",
       "      <td>0.158672</td>\n",
       "      <td>0.031957</td>\n",
       "      <td>0.002104</td>\n",
       "      <td>0.002818</td>\n",
       "      <td>0.008811</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>0.030133</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>0.026968</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>8.318388e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.472433e-09</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-007.post.stool</th>\n",
       "      <td>0.159561</td>\n",
       "      <td>0.035646</td>\n",
       "      <td>0.002210</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.008615</td>\n",
       "      <td>0.000188</td>\n",
       "      <td>0.028568</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.026092</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>...</td>\n",
       "      <td>1.779090e-06</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.456909e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 407 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.post.stool    0.158589                            0.033424   \n",
       "OAD-003.post.stool    0.164939                            0.033380   \n",
       "OAD-004.post.stool    0.165587                            0.035285   \n",
       "OAD-006.post.stool    0.158672                            0.031957   \n",
       "OAD-007.post.stool    0.159561                            0.035646   \n",
       "\n",
       "                    Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.post.stool                                           0.002252                 \n",
       "OAD-003.post.stool                                           0.002203                 \n",
       "OAD-004.post.stool                                           0.002349                 \n",
       "OAD-006.post.stool                                           0.002104                 \n",
       "OAD-007.post.stool                                           0.002210                 \n",
       "\n",
       "                    Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.post.stool                                           0.002844        \n",
       "OAD-003.post.stool                                           0.002569        \n",
       "OAD-004.post.stool                                           0.002926        \n",
       "OAD-006.post.stool                                           0.002818        \n",
       "OAD-007.post.stool                                           0.002707        \n",
       "\n",
       "                    Metabolism|Lipid metabolism  \\\n",
       "OAD-001.post.stool                     0.008684   \n",
       "OAD-003.post.stool                     0.008234   \n",
       "OAD-004.post.stool                     0.009046   \n",
       "OAD-006.post.stool                     0.008811   \n",
       "OAD-007.post.stool                     0.008615   \n",
       "\n",
       "                    Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.post.stool                                           0.000159    \n",
       "OAD-003.post.stool                                           0.000194    \n",
       "OAD-004.post.stool                                           0.000168    \n",
       "OAD-006.post.stool                                           0.000203    \n",
       "OAD-007.post.stool                                           0.000188    \n",
       "\n",
       "                    Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.post.stool                          0.029109   \n",
       "OAD-003.post.stool                          0.029464   \n",
       "OAD-004.post.stool                          0.029548   \n",
       "OAD-006.post.stool                          0.030133   \n",
       "OAD-007.post.stool                          0.028568   \n",
       "\n",
       "                    Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.post.stool                                           0.000198      \n",
       "OAD-003.post.stool                                           0.000243      \n",
       "OAD-004.post.stool                                           0.000183      \n",
       "OAD-006.post.stool                                           0.000216      \n",
       "OAD-007.post.stool                                           0.000199      \n",
       "\n",
       "                    Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.post.stool                                         0.026486   \n",
       "OAD-003.post.stool                                         0.028377   \n",
       "OAD-004.post.stool                                         0.027896   \n",
       "OAD-006.post.stool                                         0.026968   \n",
       "OAD-007.post.stool                                         0.026092   \n",
       "\n",
       "                    Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.post.stool                                           0.000029                    \n",
       "OAD-003.post.stool                                           0.000023                    \n",
       "OAD-004.post.stool                                           0.000016                    \n",
       "OAD-006.post.stool                                           0.000013                    \n",
       "OAD-007.post.stool                                           0.000018                    \n",
       "\n",
       "                    ...  \\\n",
       "OAD-001.post.stool  ...   \n",
       "OAD-003.post.stool  ...   \n",
       "OAD-004.post.stool  ...   \n",
       "OAD-006.post.stool  ...   \n",
       "OAD-007.post.stool  ...   \n",
       "\n",
       "                    Human Diseases|Infectious disease: bacterial|Yersinia infection  \\\n",
       "OAD-001.post.stool                                       2.050906e-06                 \n",
       "OAD-003.post.stool                                       2.408971e-06                 \n",
       "OAD-004.post.stool                                       8.859027e-07                 \n",
       "OAD-006.post.stool                                       0.000000e+00                 \n",
       "OAD-007.post.stool                                       1.779090e-06                 \n",
       "\n",
       "                    Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.post.stool                                       0.000000e+00                                   \n",
       "OAD-003.post.stool                                       0.000000e+00                                   \n",
       "OAD-004.post.stool                                       0.000000e+00                                   \n",
       "OAD-006.post.stool                                       8.318388e-07                                   \n",
       "OAD-007.post.stool                                       0.000000e+00                                   \n",
       "\n",
       "                    Cellular Processes|Cellular community - eukaryotes|Signaling pathways regulating pluripotency of stem cells  \\\n",
       "OAD-001.post.stool                                       9.696548e-07                                                             \n",
       "OAD-003.post.stool                                       0.000000e+00                                                             \n",
       "OAD-004.post.stool                                       0.000000e+00                                                             \n",
       "OAD-006.post.stool                                       0.000000e+00                                                             \n",
       "OAD-007.post.stool                                       0.000000e+00                                                             \n",
       "\n",
       "                    Environmental Information Processing|Signal transduction|Plant hormone signal transduction  \\\n",
       "OAD-001.post.stool                                       0.000000e+00                                            \n",
       "OAD-003.post.stool                                       0.000000e+00                                            \n",
       "OAD-004.post.stool                                       1.261718e-07                                            \n",
       "OAD-006.post.stool                                       0.000000e+00                                            \n",
       "OAD-007.post.stool                                       0.000000e+00                                            \n",
       "\n",
       "                    Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.post.stool                                                0.0                                  \n",
       "OAD-003.post.stool                                                0.0                                  \n",
       "OAD-004.post.stool                                                0.0                                  \n",
       "OAD-006.post.stool                                                0.0                                  \n",
       "OAD-007.post.stool                                                0.0                                  \n",
       "\n",
       "                    Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.post.stool                                       2.119511e-06                           \n",
       "OAD-003.post.stool                                       1.566727e-06                           \n",
       "OAD-004.post.stool                                       4.110141e-07                           \n",
       "OAD-006.post.stool                                       8.472433e-09                           \n",
       "OAD-007.post.stool                                       9.456909e-07                           \n",
       "\n",
       "                    Environmental Information Processing|Signal transduction|Rap1 signaling pathway  \\\n",
       "OAD-001.post.stool                                                0.0                                 \n",
       "OAD-003.post.stool                                                0.0                                 \n",
       "OAD-004.post.stool                                                0.0                                 \n",
       "OAD-006.post.stool                                                0.0                                 \n",
       "OAD-007.post.stool                                                0.0                                 \n",
       "\n",
       "                    Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis  \\\n",
       "OAD-001.post.stool                                                0.0                               \n",
       "OAD-003.post.stool                                                0.0                               \n",
       "OAD-004.post.stool                                                0.0                               \n",
       "OAD-006.post.stool                                                0.0                               \n",
       "OAD-007.post.stool                                                0.0                               \n",
       "\n",
       "                    Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \\\n",
       "OAD-001.post.stool                                                0.0                                          \n",
       "OAD-003.post.stool                                                0.0                                          \n",
       "OAD-004.post.stool                                                0.0                                          \n",
       "OAD-006.post.stool                                                0.0                                          \n",
       "OAD-007.post.stool                                                0.0                                          \n",
       "\n",
       "                    WOMAC_P_Response  \n",
       "OAD-001.post.stool       No response  \n",
       "OAD-003.post.stool       No response  \n",
       "OAD-004.post.stool       No response  \n",
       "OAD-006.post.stool          Response  \n",
       "OAD-007.post.stool          Response  \n",
       "\n",
       "[5 rows x 407 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# paired testing in resp vs non resp \n",
    "gr_to_sigpath = {} \n",
    "\n",
    "gs = []\n",
    "ts = []\n",
    "xs = []\n",
    "ws = []\n",
    "ps = []\n",
    "\n",
    "for g in ['saliva','stool']:#'stool_adh']:\n",
    "    print(g)\n",
    "    r_to_sigpath = {}\n",
    "    # grab taxa table\n",
    "    df_path = gt_to_path[g]['all']\n",
    "\n",
    "    df = type_to_df_map[g].copy()\n",
    "    df = df.set_index('Together')\n",
    "    if g == 'stool':\n",
    "        df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "    df = df[['Timepoint','WOMAC_P_Response']]\n",
    "    for t in ['pre', 'post']:\n",
    "        print(t)\n",
    "        sig_path = []\n",
    "        df_merge = pd.concat([df_path, df[df['Timepoint'] == t]], axis=1)\n",
    "        df_merge = df_merge.dropna()\n",
    "        df_merge['SampleID'] = df_merge.index.map(lambda x: x.split('.p')[0])\n",
    "        df_merge = df_merge.drop(['SampleID','Timepoint'],axis=1)\n",
    "\n",
    "        for x in list(df_merge.columns.values)[:-1]:\n",
    "            df_r = df_merge[df_merge['WOMAC_P_Response'] == 'Response'][x]\n",
    "            df_nr = df_merge[df_merge['WOMAC_P_Response'] == 'No response'][x]\n",
    "\n",
    "            try:\n",
    "                w,p = scipy.stats.mannwhitneyu(df_r.values, df_nr.values, nan_policy='omit')\n",
    "            except:\n",
    "                w,p = 0, 1\n",
    "            if p < 0.05:\n",
    "\n",
    "                gs.append(g)\n",
    "                ts.append(t)\n",
    "                xs.append(x)\n",
    "                ws.append(w)\n",
    "                ps.append(p)\n",
    "                \n",
    "                ax = sns.boxplot(data=df_merge, y=x, x='WOMAC_P_Response')\n",
    "                sns.swarmplot(data=df_merge, y=x, x='WOMAC_P_Response', palette='dark:grey')\n",
    "                sns.despine()\n",
    "            \n",
    "                plt.tight_layout()\n",
    "                plt.savefig(path + 'outputs/jobs06/pvp_path_barplot_' + g + '_' + t + '_' + x.split('g__')[-1] + '.pdf')\n",
    "                plt.close()    \n",
    "                sig_path.append(x)\n",
    "        print('Number of sigpath')\n",
    "        print(len(sig_path))\n",
    "        r_to_sigpath[t] = sig_path\n",
    "    gr_to_sigpath[g] = r_to_sigpath\n",
    "\n",
    "\n",
    "to_sigpath['vertical'] = gr_to_sigpath\n",
    "\n",
    "pd.DataFrame({'SampleType':gs, 'WOMAC_P_Response': xs, 'Timepoint': ts, 'WilcoxonRankSum': ws, 'pvalue': ps}).to_csv(path + 'outputs/jobs06/path_pvp.tsv', sep='\\t')\n",
    "\n",
    "df_merge.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b6f3aa88-2e19-42b3-bd79-3e0b66119a1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saliva\n",
      "diff\n",
      "Response\n",
      "29\n",
      "No response\n",
      "0\n",
      "pre\n",
      "12\n",
      "post\n",
      "16\n",
      "all\n",
      "Response\n",
      "30\n",
      "No response\n",
      "0\n",
      "pre\n",
      "6\n",
      "post\n",
      "19\n",
      "stool\n",
      "diff\n",
      "Response\n",
      "25\n",
      "No response\n",
      "0\n",
      "pre\n",
      "45\n",
      "post\n",
      "15\n",
      "all\n",
      "Response\n",
      "42\n",
      "No response\n",
      "0\n",
      "pre\n",
      "13\n",
      "post\n",
      "9\n"
     ]
    }
   ],
   "source": [
    "# correlations of taxa of interest with metabolites\n",
    "gtr_to_otucorr = {}\n",
    "for g in ['saliva','stool']:\n",
    "    print(g)\n",
    "    gtr_to_otucorr[g] = {}\n",
    "    for time in ['diff','all']:\n",
    "        print(time)\n",
    "        gtr_to_otucorr[g][time] = {}\n",
    "        # grab otu table of differences or of all samples\n",
    "        df_otu = gt_to_otu[g][time]\n",
    "        for comp in ['horizontal', 'vertical']:\n",
    "            if comp == 'vertical':\n",
    "                gr_to_sigotu = to_sigotu[comp]\n",
    "                rvs = ['pre', 'post']\n",
    "            elif comp == 'horizontal':\n",
    "                gr_to_sigotu = to_sigotu[comp]\n",
    "                rvs = ['Response', 'No response']\n",
    "            for rv in rvs: # ['pre', 'post']: # Response No response\n",
    "                print(rv)\n",
    "                sigcorr = []\n",
    "                # grab taxa of interest\n",
    "                taxa_int = gr_to_sigotu[g][rv]\n",
    "    \n",
    "                # merge otu table with metabolites            \n",
    "                df_merge = pd.concat([df_otu.loc[:,taxa_int], gt_to_meta[g][time]], axis=1)\n",
    "                meta = gt_to_meta[g][time].columns\n",
    "                \n",
    "                # look at taxa int and correlate with metabolites\n",
    "                # df_merge \n",
    "                for t in taxa_int:\n",
    "                    for m in meta:\n",
    "                        df_sub = df_merge.loc[:,[t,m]]\n",
    "                        df = type_to_df_map[g].copy()\n",
    "                        if time is 'all':\n",
    "                            df = df.set_index('Together')\n",
    "                            if g == 'stool':\n",
    "                                df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "                        elif time is 'diff':\n",
    "                            df = df[df['Timepoint'] == 'pre']\n",
    "                        df_sub = pd.concat([df_sub, df['WOMAC_P_Response']],axis=1)\n",
    "                        df_sub = df_sub.dropna(subset=[t,m])\n",
    "                        r, p = scipy.stats.spearmanr(df_sub[t].values, df_sub[m].values)\n",
    "                        if p < 0.05:\n",
    "                            sigcorr.append([t,m])\n",
    "                            sns.scatterplot(data=df_sub, x=t, y=m, hue='WOMAC_P_Response')\n",
    "                            sns.despine()\n",
    "                        \n",
    "                            plt.tight_layout()\n",
    "                            plt.savefig(path + 'outputs/' + g_to_job[g + '_adh'] + '/corr_taxa_' + comp + '_' + time + '_' + t.split(';g__')[-1] + '_' + m.replace('/','fslash') + '.pdf')\n",
    "                            plt.close()   \n",
    "                gtr_to_otucorr[g][time][rv] = sigcorr\n",
    "                print(len(sigcorr))\n",
    "           \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7f6c73d6-71f4-44ee-ac66-5fe0711fe3f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saliva\n",
      "diff\n",
      "Response\n",
      "38\n",
      "No response\n",
      "0\n",
      "pre\n",
      "6\n",
      "post\n",
      "7\n",
      "all\n",
      "Response\n",
      "86\n",
      "No response\n",
      "0\n",
      "pre\n",
      "6\n",
      "post\n",
      "22\n",
      "stool\n",
      "diff\n",
      "Response\n",
      "88\n",
      "No response\n",
      "0\n",
      "pre\n",
      "16\n",
      "post\n",
      "45\n",
      "all\n",
      "Response\n",
      "178\n",
      "No response\n",
      "0\n",
      "pre\n",
      "28\n",
      "post\n",
      "48\n"
     ]
    }
   ],
   "source": [
    "# correlations of pathways of interest with metabolites\n",
    "gtr_to_pathcorr = {}\n",
    "for g in ['saliva','stool']:\n",
    "    print(g)\n",
    "    gtr_to_pathcorr[g] = {}\n",
    "    for time in ['diff','all']:\n",
    "        print(time)\n",
    "        gtr_to_pathcorr[g][time] = {}\n",
    "        # grab otu table of differences or of all samples\n",
    "        df_path = gt_to_path[g][time]\n",
    "        for comp in ['horizontal', 'vertical']:\n",
    "            if comp == 'vertical':\n",
    "                gr_to_sigpath = to_sigpath[comp]\n",
    "                rvs = ['pre', 'post']\n",
    "            elif comp == 'horizontal':\n",
    "                gr_to_sigpath = to_sigpath[comp]\n",
    "                rvs = ['Response', 'No response']\n",
    "            for rv in rvs: # ['pre', 'post']: # Response No response\n",
    "                print(rv)\n",
    "                sigcorr = []\n",
    "                # grab taxa of interest\n",
    "                taxa_int = gr_to_sigpath[g][rv]\n",
    "    \n",
    "                # merge otu table with metabolites            \n",
    "                df_merge = pd.concat([df_path.loc[:,taxa_int], gt_to_meta[g][time]], axis=1)\n",
    "                meta = gt_to_meta[g][time].columns\n",
    "                \n",
    "                # look at taxa int and correlate with metabolites\n",
    "                # df_merge \n",
    "                for t in taxa_int:\n",
    "                    for m in meta:\n",
    "                        df_sub = df_merge.loc[:,[t,m]]\n",
    "                        df = type_to_df_map[g].copy()\n",
    "                        if time is 'all':\n",
    "                            df = df.set_index('Together')\n",
    "                            if g == 'stool':\n",
    "                                df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "                        elif time is 'diff':\n",
    "                            df = df[df['Timepoint'] == 'pre']\n",
    "                        df_sub = pd.concat([df_sub, df['WOMAC_P_Response']],axis=1)\n",
    "                        df_sub = df_sub.dropna(subset=[t,m])\n",
    "                        r, p = scipy.stats.spearmanr(df_sub[t].values, df_sub[m].values)\n",
    "                        if p < 0.05:\n",
    "                            sigcorr.append([t,m])\n",
    "                            sns.scatterplot(data=df_sub, x=t, y=m, hue='WOMAC_P_Response')\n",
    "                            sns.despine()\n",
    "                        \n",
    "                            plt.tight_layout()\n",
    "                            plt.savefig(path + 'outputs/' + g_to_job[g + '_adh'] + '/corr_path_' + comp + '_' + time + '_' + t.split(';g__')[-1] + '_' + m.replace('/','fslash') + '.pdf')\n",
    "                            plt.close()   \n",
    "                gtr_to_pathcorr[g][time][rv] = sigcorr\n",
    "                print(len(sigcorr))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f5abd2d7-1c78-4e99-a89e-bb864ac5547f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Metabolism|Xenobiotics biodegradation and metabolism|Caprolactam degradation', 'H_Tyrosine']\n",
      "['Metabolism|Nucleotide metabolism', 'H_Tyrosine']\n",
      "['Metabolism|Lipid metabolism|Arachidonic acid metabolism', 'H_Tyrosine']\n",
      "['Human Diseases|Cancer: specific types|Renal cell carcinoma', 'H_Isofraxidin']\n",
      "['Human Diseases|Cancer: specific types|Renal cell carcinoma', 'H_Tyrosine']\n",
      "['Human Diseases|Cancer: specific types|Renal cell carcinoma', 'H_Trihydroxycholestanoic_acid']\n",
      "['Human Diseases|Endocrine and metabolic disease|Cushing syndrome', 'H_Isofraxidin']\n",
      "['Human Diseases|Endocrine and metabolic disease|Cushing syndrome', 'H_Tyrosine']\n",
      "['Human Diseases|Endocrine and metabolic disease|Cushing syndrome', 'H_Trihydroxycholestanoic_acid']\n",
      "['Metabolism|Xenobiotics biodegradation and metabolism|Caprolactam degradation', 'H_Tyrosine']\n",
      "['Metabolism|Xenobiotics biodegradation and metabolism|Aminobenzoate degradation', 'H_Tyrosine']\n",
      "['Human Diseases|Infectious disease: parasitic|African trypanosomiasis', 'H_Tryptophan']\n",
      "['Metabolism|Lipid metabolism|Arachidonic acid metabolism', 'H_Tyrosine']\n",
      "['Metabolism|Lipid metabolism|Arachidonic acid metabolism', 'H_Trihydroxycholestanoic_acid']\n",
      "['Metabolism|Biosynthesis of other secondary metabolites|Flavone and flavonol biosynthesis', 'H_Tryptophan']\n",
      "['Human Diseases|Infectious disease: parasitic|Chagas disease', 'H_Tryptophan']\n",
      "['Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis', 'H_Xanthurenic_acid']\n",
      "['Metabolism|Amino acid metabolism|Tyrosine metabolism', 'H_Tryptophan']\n",
      "['Metabolism|Glycan biosynthesis and metabolism', 'H_Tryptophan']\n",
      "['Metabolism|Metabolism of cofactors and vitamins|Lipoic acid metabolism', 'H_Tryptophan']\n"
     ]
    }
   ],
   "source": [
    "# L Methionine\n",
    "# Kaempferol\n",
    "# Tryptophan\n",
    "# Tyrosine Stool\n",
    "# Usnic acid\n",
    "# Xanthurenic acid\n",
    "# Tryhydroxycholestenoic acid Stool\n",
    "# Isofraxadine\n",
    "int = ['Isofrax', 'Xanth', 'ydroxychol', 'Usnic', 'Tyrosine', 'Tryptophan', 'Kaempferol', 'Methion']\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    for time in ['diff','all']:\n",
    "        for r in ['Response', 'No response']:\n",
    "            test = gtr_to_pathcorr[g][time][r]\n",
    "            for t in test:\n",
    "                p, m = t # unpack pathway, metabolite, from tuple representing sig corr\n",
    "                for i in int:\n",
    "                    if i in m:\n",
    "                        print(t)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f30bf9df-2904-4ca5-b179-4a4eaa5e71f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['d__Bacteria;p__Firmicutes_B_370539;c__Peptococcia;o__Peptococcales;f__Peptococcaceae_369362;g__Peptococcus', 'H_Tyrosine']\n",
      "['d__Bacteria;p__Firmicutes_B_370539;c__Peptococcia;o__Peptococcales;f__Peptococcaceae_369362;g__Peptococcus', 'H_Trihydroxycholestanoic_acid']\n",
      "['d__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia', 'H_Isofraxidin']\n",
      "['d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871404', 'H_Kaempferol']\n",
      "['d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871404', 'H_Tyrosine']\n",
      "['d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Phocea', 'H_Xanthurenic_acid']\n",
      "['d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Phocea', 'H_Kaempferol']\n"
     ]
    }
   ],
   "source": [
    "int = ['Isofrax', 'Xanth', 'ydroxychol', 'Usnic', 'Tyrosine', 'Tryptophan', 'Kaempferol', 'Methion']\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    for time in ['diff','all']:\n",
    "        for r in ['Response', 'No response']:\n",
    "            test = gtr_to_otucorr[g][time][r]\n",
    "            for t in test:\n",
    "                p, m = t # unpack pathway, metabolite, from tuple representing sig corr\n",
    "                for i in int:\n",
    "                    if i in m:\n",
    "                        print(t)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "8d51feb2-1356-4615-81a9-ae357b3f93e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_Thiamine</th>\n",
       "      <th>H_Melezitose</th>\n",
       "      <th>H_Phosphocholine</th>\n",
       "      <th>H_SN_Glycero_3_Phophocholine</th>\n",
       "      <th>H_Gamma_Valerobetaine</th>\n",
       "      <th>H_N_Acetylneuraminic_acid</th>\n",
       "      <th>H_Pyridoxine</th>\n",
       "      <th>H_N_Acetylneuraminic_Acid</th>\n",
       "      <th>H_N_Acetylmuramic_Acid</th>\n",
       "      <th>...</th>\n",
       "      <th>H_Hydroxydodecanoic_acid</th>\n",
       "      <th>H_Omega_Hydroxydodecanoate</th>\n",
       "      <th>H_Delta_Methyldodecenoic_Acid</th>\n",
       "      <th>H_Hydroxydecanoic_acid</th>\n",
       "      <th>H_Hydroxydecanoate</th>\n",
       "      <th>H_Methylpentanoic_acid</th>\n",
       "      <th>H_Monohydroxylated_bile_acid</th>\n",
       "      <th>H_Palmitoyl_ethanolamide</th>\n",
       "      <th>H_Linoleic_Acid</th>\n",
       "      <th>WOMAC_P_Response</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>0.312342</td>\n",
       "      <td>-0.312477</td>\n",
       "      <td>0.119853</td>\n",
       "      <td>-0.492876</td>\n",
       "      <td>-0.671376</td>\n",
       "      <td>-0.202952</td>\n",
       "      <td>0.203945</td>\n",
       "      <td>-0.455869</td>\n",
       "      <td>0.352628</td>\n",
       "      <td>0.151464</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.330037</td>\n",
       "      <td>-0.886045</td>\n",
       "      <td>-0.314427</td>\n",
       "      <td>-0.322677</td>\n",
       "      <td>-0.650357</td>\n",
       "      <td>-0.361695</td>\n",
       "      <td>0.132794</td>\n",
       "      <td>-0.067464</td>\n",
       "      <td>-0.112641</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>0.000428</td>\n",
       "      <td>-0.481251</td>\n",
       "      <td>-0.179337</td>\n",
       "      <td>0.983217</td>\n",
       "      <td>0.167836</td>\n",
       "      <td>0.067823</td>\n",
       "      <td>0.364897</td>\n",
       "      <td>-0.371942</td>\n",
       "      <td>0.545435</td>\n",
       "      <td>-0.007664</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.334936</td>\n",
       "      <td>-0.802118</td>\n",
       "      <td>-0.929470</td>\n",
       "      <td>-0.369578</td>\n",
       "      <td>-0.566431</td>\n",
       "      <td>0.075375</td>\n",
       "      <td>-0.519189</td>\n",
       "      <td>-0.180631</td>\n",
       "      <td>-0.418351</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>-0.285212</td>\n",
       "      <td>-0.556445</td>\n",
       "      <td>-0.651325</td>\n",
       "      <td>0.357509</td>\n",
       "      <td>-0.401875</td>\n",
       "      <td>-0.295787</td>\n",
       "      <td>0.137736</td>\n",
       "      <td>-0.165902</td>\n",
       "      <td>0.112170</td>\n",
       "      <td>0.096791</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.589028</td>\n",
       "      <td>-1.535141</td>\n",
       "      <td>-0.525277</td>\n",
       "      <td>-0.278312</td>\n",
       "      <td>-0.600484</td>\n",
       "      <td>0.019657</td>\n",
       "      <td>-0.576831</td>\n",
       "      <td>-0.901479</td>\n",
       "      <td>-1.089906</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD004</th>\n",
       "      <td>0.140904</td>\n",
       "      <td>1.089055</td>\n",
       "      <td>-0.668077</td>\n",
       "      <td>-1.061336</td>\n",
       "      <td>-0.370602</td>\n",
       "      <td>-0.329379</td>\n",
       "      <td>-0.654931</td>\n",
       "      <td>0.142971</td>\n",
       "      <td>-1.006689</td>\n",
       "      <td>-0.100373</td>\n",
       "      <td>...</td>\n",
       "      <td>0.449897</td>\n",
       "      <td>0.489189</td>\n",
       "      <td>0.413723</td>\n",
       "      <td>0.474250</td>\n",
       "      <td>0.661963</td>\n",
       "      <td>0.245553</td>\n",
       "      <td>0.279290</td>\n",
       "      <td>-0.444319</td>\n",
       "      <td>-0.912164</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD011</th>\n",
       "      <td>-0.346855</td>\n",
       "      <td>1.028262</td>\n",
       "      <td>-0.450323</td>\n",
       "      <td>-0.025939</td>\n",
       "      <td>-0.223029</td>\n",
       "      <td>-0.766374</td>\n",
       "      <td>-0.538410</td>\n",
       "      <td>-1.091423</td>\n",
       "      <td>-0.328522</td>\n",
       "      <td>-0.524686</td>\n",
       "      <td>...</td>\n",
       "      <td>0.419748</td>\n",
       "      <td>0.429119</td>\n",
       "      <td>0.333756</td>\n",
       "      <td>0.448461</td>\n",
       "      <td>0.598846</td>\n",
       "      <td>0.390035</td>\n",
       "      <td>0.287682</td>\n",
       "      <td>-0.532141</td>\n",
       "      <td>-0.752280</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 89 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          H_Pyridoxamine  H_Thiamine  H_Melezitose  H_Phosphocholine  \\\n",
       "VAOAD001        0.312342   -0.312477      0.119853         -0.492876   \n",
       "VAOAD009        0.000428   -0.481251     -0.179337          0.983217   \n",
       "OAD001         -0.285212   -0.556445     -0.651325          0.357509   \n",
       "VAOAD004        0.140904    1.089055     -0.668077         -1.061336   \n",
       "VAOAD011       -0.346855    1.028262     -0.450323         -0.025939   \n",
       "\n",
       "          H_SN_Glycero_3_Phophocholine  H_Gamma_Valerobetaine  \\\n",
       "VAOAD001                     -0.671376              -0.202952   \n",
       "VAOAD009                      0.167836               0.067823   \n",
       "OAD001                       -0.401875              -0.295787   \n",
       "VAOAD004                     -0.370602              -0.329379   \n",
       "VAOAD011                     -0.223029              -0.766374   \n",
       "\n",
       "          H_N_Acetylneuraminic_acid  H_Pyridoxine  H_N_Acetylneuraminic_Acid  \\\n",
       "VAOAD001                   0.203945     -0.455869                   0.352628   \n",
       "VAOAD009                   0.364897     -0.371942                   0.545435   \n",
       "OAD001                     0.137736     -0.165902                   0.112170   \n",
       "VAOAD004                  -0.654931      0.142971                  -1.006689   \n",
       "VAOAD011                  -0.538410     -1.091423                  -0.328522   \n",
       "\n",
       "          H_N_Acetylmuramic_Acid  ...  H_Hydroxydodecanoic_acid  \\\n",
       "VAOAD001                0.151464  ...                 -0.330037   \n",
       "VAOAD009               -0.007664  ...                 -0.334936   \n",
       "OAD001                  0.096791  ...                 -0.589028   \n",
       "VAOAD004               -0.100373  ...                  0.449897   \n",
       "VAOAD011               -0.524686  ...                  0.419748   \n",
       "\n",
       "          H_Omega_Hydroxydodecanoate  H_Delta_Methyldodecenoic_Acid  \\\n",
       "VAOAD001                   -0.886045                      -0.314427   \n",
       "VAOAD009                   -0.802118                      -0.929470   \n",
       "OAD001                     -1.535141                      -0.525277   \n",
       "VAOAD004                    0.489189                       0.413723   \n",
       "VAOAD011                    0.429119                       0.333756   \n",
       "\n",
       "          H_Hydroxydecanoic_acid  H_Hydroxydecanoate  H_Methylpentanoic_acid  \\\n",
       "VAOAD001               -0.322677           -0.650357               -0.361695   \n",
       "VAOAD009               -0.369578           -0.566431                0.075375   \n",
       "OAD001                 -0.278312           -0.600484                0.019657   \n",
       "VAOAD004                0.474250            0.661963                0.245553   \n",
       "VAOAD011                0.448461            0.598846                0.390035   \n",
       "\n",
       "          H_Monohydroxylated_bile_acid  H_Palmitoyl_ethanolamide  \\\n",
       "VAOAD001                      0.132794                 -0.067464   \n",
       "VAOAD009                     -0.519189                 -0.180631   \n",
       "OAD001                       -0.576831                 -0.901479   \n",
       "VAOAD004                      0.279290                 -0.444319   \n",
       "VAOAD011                      0.287682                 -0.532141   \n",
       "\n",
       "          H_Linoleic_Acid  WOMAC_P_Response  \n",
       "VAOAD001        -0.112641       No response  \n",
       "VAOAD009        -0.418351       No response  \n",
       "OAD001          -1.089906       No response  \n",
       "VAOAD004        -0.912164          Response  \n",
       "VAOAD011        -0.752280          Response  \n",
       "\n",
       "[5 rows x 89 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CUTIE\n",
    "# In POST samples only\n",
    "# Color by Resp and Nonresp\n",
    "\n",
    "for g in ['saliva', 'stool']:\n",
    "\n",
    "    df_abun = pd.concat([gt_to_otu[g]['post'], gt_to_path[g]['post']],axis=1)\n",
    "    df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_map = type_to_df_map[g].set_index('Together')\n",
    "    df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_meta = pd.concat([gt_to_meta[g]['After diet'],df_map], axis=1)\n",
    "\n",
    "    df_abun.to_csv(path + 'inputs/cutie_df_abun_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "    df_meta.to_csv(path + 'inputs/cutie_df_meta_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1fdbda6d-641c-47ab-9c83-d392cf0299a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "28\n",
      "saliva\n",
      "97\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Tannerellaceae;g__Tannerella',\n",
       " 'd__Bacteria;p__Spirochaetota;c__Spirochaetia;o__Treponematales;f__Treponemataceae;g__Treponema_B_986141',\n",
       " 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Johnsonella',\n",
       " 'd__Bacteria;p__Synergistota;c__Synergistia;o__Synergistales;f__Aminobacteriaceae;g__CAJPSE01',\n",
       " 'd__Bacteria;p__Fusobacteriota;c__Fusobacteriia;o__Fusobacteriales_993777;f__Leptotrichiaceae;g__Pseudoleptotrichia']"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# anaerobic vs aerobic in response\n",
    "# bile acids -> list? saliva and stool of interest \n",
    "#\n",
    "# Bile acids found in Stool: \n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid', 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide']\n",
    "         }\n",
    "\n",
    "g_to_cutiejob = {\n",
    "    'saliva': 'jobs15',\n",
    "    'stool': 'jobs16'\n",
    "}\n",
    "\n",
    "\n",
    "# from difflib import SequenceMatcher\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    print(g)\n",
    "    n_corr = 0\n",
    "    # import CUITIE results\n",
    "    df_qt = pd.read_csv(path + 'outputs/' + g_to_cutiejob[g] + '/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "    df_qt = df_qt[df_qt['class'] == 'TP']\n",
    "    # iterate through rows\n",
    "    for r in range(len(df_qt)):\n",
    "        meta = df_qt.iloc[r,0]\n",
    "        abun = df_qt.iloc[r,1]\n",
    "        \n",
    "        #for i in int_ba:\n",
    "        #    s = SequenceMatcher(None, meta, i)\n",
    "        #    if s.ratio() > 0.9:\n",
    "        #        print(i)\n",
    "        \n",
    "        #if 'bile' in meta:\n",
    "        #    print(meta)\n",
    "        output = [r]\n",
    "        \n",
    "        for i in int_ba[g]:\n",
    "            if i in meta:\n",
    "                output.append(meta)\n",
    "\n",
    "        for c in ['horizontal','vertical']:\n",
    "            if c == 'vertical':\n",
    "                queries = ['pre', 'post']\n",
    "            if c == 'horizontal':\n",
    "                queries = ['Response','No response']\n",
    "            output.append(c)\n",
    "            for r in queries:\n",
    "                int_taxa = to_sigotu[c][g][r]\n",
    "                for i in int_taxa:\n",
    "                    if i in abun:\n",
    "                        output.append('taxa')\n",
    "                        output.append(i)\n",
    "                int_path = to_sigpath[c][g][r]\n",
    "                for i in int_path:\n",
    "                    if i in abun:\n",
    "                        output.append('path')\n",
    "                        output.append(i)\n",
    "        #output.append()\n",
    "        #output.append()\n",
    "        if len(output) > 5:\n",
    "            # print(output)\n",
    "            n_corr += 1\n",
    "\n",
    "    print(n_corr)\n",
    "\n",
    "# pathway prediction of response vs non response @ baseline\n",
    "# specific taxa at baseline\n",
    "\n",
    "\n",
    "# pathways of interest\n",
    "# ex inflammation in non responders\n",
    "# heatmap of differential pathways in responders/nonresponders \n",
    "# and correlation of those pathways with bacteria\n",
    "# send list of upregulated pathways \n",
    "\n",
    "int_taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "76722aff-f034-4fff-a4da-c63dd1fd75ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "meta\n",
      "otu\n",
      "path\n",
      "saliva\n",
      "meta\n",
      "otu\n",
      "path\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>path</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>adj_test_stat</th>\n",
       "      <th>pvals</th>\n",
       "      <th>direction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>Metabolism|Biosynthesis of other secondary met...</td>\n",
       "      <td>32.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>0.012207</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Metabolism|Biosynthesis of other secondary met...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>0.009277</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>Human Diseases|Immune disease|Primary immunode...</td>\n",
       "      <td>36.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.020996</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>Organismal Systems|Aging</td>\n",
       "      <td>37.0</td>\n",
       "      <td>107.0</td>\n",
       "      <td>0.002441</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>Metabolism|Metabolism of terpenoids and polyke...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>107.0</td>\n",
       "      <td>0.002441</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  path  test_stat  \\\n",
       "224  Metabolism|Biosynthesis of other secondary met...       32.0   \n",
       "30   Metabolism|Biosynthesis of other secondary met...       35.0   \n",
       "294  Human Diseases|Immune disease|Primary immunode...       36.0   \n",
       "176                           Organismal Systems|Aging       37.0   \n",
       "304  Metabolism|Metabolism of terpenoids and polyke...       37.0   \n",
       "\n",
       "     adj_test_stat     pvals direction  \n",
       "224          112.0  0.012207      post  \n",
       "30           109.0  0.009277      post  \n",
       "294          108.0  0.020996      post  \n",
       "176          107.0  0.002441      post  \n",
       "304          107.0  0.002441      post  "
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# horizontal barplot for features\n",
    "\n",
    "d_to_df = {\n",
    "    'meta': gt_to_meta,\n",
    "    'otu': gt_to_otu,\n",
    "    'path': gt_to_path\n",
    "}\n",
    "t = 'all'\n",
    "\n",
    "gd_to_res = {}\n",
    "for g in ['stool','saliva']:\n",
    "    print(g)\n",
    "    gd_to_res[g] = {}\n",
    "    for d in ['meta','otu','path']:\n",
    "        print(d)\n",
    "        # grab original dfs\n",
    "        df_feat = d_to_df[d][g][t]\n",
    "\n",
    "        df_map = type_to_df_map[g].set_index('Together')\n",
    "        df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "        \n",
    "        df_map = df_map[['Timepoint','WOMAC_P_Response','HostSubjectId']]\n",
    "        \n",
    "        # merge df\n",
    "        df_merge = pd.concat([df_feat, df_map],axis=1)\n",
    "        df_merge = df_merge.dropna()\n",
    "        \n",
    "        # do paired test\n",
    "        # grab all sample IDs that are duplicated\n",
    "        df_vc = df_merge['HostSubjectId'].value_counts() == 2\n",
    "        df_vc = df_vc[df_vc == True]\n",
    "        keep_id = df_vc.index.values\n",
    "        \n",
    "        df_merge = df_merge[df_merge['HostSubjectId'].isin(keep_id)]\n",
    "        \n",
    "        # samples are in order when subsetted on Timepoint\n",
    "        df_pre = df_merge[df_merge['Timepoint'] == 'pre'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "        df_post = df_merge[df_merge['Timepoint'] == 'post'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "        \n",
    "        # compute rank sum\n",
    "        feat = []\n",
    "        rs = []\n",
    "        adj_rs = []\n",
    "        ps = []\n",
    "        dir = []\n",
    "        for x in df_pre.columns.values:\n",
    "            try:\n",
    "                r, p = scipy.stats.wilcoxon(df_post[x].values-df_pre[x].values,nan_policy='omit')\n",
    "            except ValueError:\n",
    "                r, p = 0, 1\n",
    "            if d == 'otu':\n",
    "                feat.append(x.split('__')[-1])\n",
    "            else:\n",
    "                feat.append(x)\n",
    "            # rs.append(r)\n",
    "            ps.append(p)\n",
    "            # determine direction via MWU\n",
    "            a = df_post[x].values\n",
    "            b = df_pre[x].values\n",
    "            abr, abp = scipy.stats.mannwhitneyu(a,b,nan_policy='omit')\n",
    "            bar, bap = scipy.stats.mannwhitneyu(b,a,nan_policy='omit')\n",
    "            if abr > bar:\n",
    "                dir.append('post')\n",
    "                adj_rs.append(abr)\n",
    "                rs.append(bar)\n",
    "            elif abr < bar:\n",
    "                dir.append('pre')\n",
    "                adj_rs.append(bar * -1)\n",
    "                rs.append(bar)\n",
    "            else:\n",
    "                dir.append('same')\n",
    "                adj_rs.append(abr)\n",
    "                rs.append(abr)\n",
    "        # compile results df\n",
    "        df_results = pd.DataFrame({d: feat, 'test_stat': rs, 'adj_test_stat': adj_rs,'pvals': ps, 'direction': dir})\n",
    "        gd_to_res[g][d] = df_results\n",
    "        df = df_results[df_results['pvals'] < 0.05]\n",
    "        df = df.sort_values(by=['adj_test_stat'],ascending=False)\n",
    "        # plot\n",
    "        # create plot\n",
    "        sns.set_theme(style=\"whitegrid\")\n",
    "        \n",
    "        # Plot the total crashes\n",
    "        sns.set_color_codes(\"pastel\")\n",
    "        sns.barplot(x='adj_test_stat', y=d, data=df,\n",
    "                    hue=\"direction\",orient='horizontal')\n",
    "        \n",
    "        sns.despine(left=True, bottom=True)\n",
    "        sns.tight_layou\n",
    "        plt.savefig(path + 'outputs/jobs18/DA_hbarplot_' + g + '_' + d + '.pdf')\n",
    "        plt.close()\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "11360b5e-dd8b-4857-b205-4dd8c5c685b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "881\n",
      "780\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>var_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>cellular processes</td>\n",
       "      <td>Pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>cellular processes|cell growth and death</td>\n",
       "      <td>Pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>620</th>\n",
       "      <td>cellular processes|cell growth and death|apopt...</td>\n",
       "      <td>Pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>698</th>\n",
       "      <td>cellular processes|cell growth and death|apopt...</td>\n",
       "      <td>Pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>755</th>\n",
       "      <td>cellular processes|cell growth and death|apopt...</td>\n",
       "      <td>Pathways</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 index  var_type\n",
       "434                                 cellular processes  Pathways\n",
       "495           cellular processes|cell growth and death  Pathways\n",
       "620  cellular processes|cell growth and death|apopt...  Pathways\n",
       "698  cellular processes|cell growth and death|apopt...  Pathways\n",
       "755  cellular processes|cell growth and death|apopt...  Pathways"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for corr heatmaps \n",
    "for g in ['saliva','stool']:\n",
    "    df_abun = pd.concat([gt_to_otu[g]['post'], gt_to_path[g]['post']],axis=1)\n",
    "    df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_map = type_to_df_map[g].set_index('Together')\n",
    "    df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_meta = pd.concat([gt_to_meta[g]['After diet'],df_map], axis=1)\n",
    "\n",
    "    df_quant_R = df_abun.copy()\n",
    "    #df_quant_R = pd.concat([df_abun, df_meta],axis=1)# .copy() #([df_quant, df_asv100, df_path_top100], axis=1)\n",
    "    # df_quant_R = df_quant_R.drop(['WOMAC_P_Response'],axis=1)\n",
    "    # df_quant_R.to_csv(path + 'inputs/df_quant_' + g + '.tsv', sep='\\t')\n",
    "    df_quant_R = df_quant_R.T\n",
    "    # this is in the order they were entered i.e. \n",
    "    # [df_olink, df_fa, df_acpa_fecal, df_acpa_plasma, df_plasma, df_rbfa, df_mb]\n",
    "     # ['rbfa']*len(df_rbfa.columns) + \\  \n",
    "    df_quant_R['var_type'] = ['Taxa']*len(gt_to_otu[g]['post'].columns) + \\\n",
    "                             ['Pathways']*len(gt_to_path[g]['post'].columns) # + \\\n",
    "                             # ['Metabolite']*len(gt_to_meta[g]['After diet'].columns)\n",
    "    # df_quant_R = df_quant_R[~df_quant_R['var_type'].isin(['Metagenomic_ASVs','Metagenomic_Pathways'])]\n",
    "    df_quant_R = df_quant_R['var_type'].reset_index()\n",
    "    df_quant_R['index'] = df_quant_R['index'].str.lower()\n",
    "    df_quant_R = df_quant_R.sort_values(by='index',ascending=True)\n",
    "    df_quant_R.to_csv(path + 'inputs/df_quant_R_corr_labels_' + g + '.tsv', sep='\\t')\n",
    "    print(len(df_quant_R))\n",
    "df_quant_R.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "2707d857-ee9c-466f-983c-3d94600478d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjcAAAGwCAYAAABVdURTAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABLB0lEQVR4nO3deVgVZf8G8HvYDzuibIrghiyioJTijphrrr1qRrmj5q65ZKWgmFi5pm8qSuKWWqHmmposLmgJapkg4goSiiaKyiYwvz98nZ9HROFw4MBwf65rrsvzzMwz3zmB3c4884wgiqIIIiIiIpnQ0nQBREREROrEcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLJSLcONKIrIzMwEp/ghIiKSn2oZbh49egQzMzM8evRI06UQERGRmlXLcENERETyxXBDREREssJwQ0RERLLCcENERESywnBDREREsqKj6QIqq4KCAjx9+lTTZZBM6erqQltbW9NlEBHJEsPNS0RRxO3bt/HgwQNNl0IyZ25uDhsbGwiCoOlSiIhkheHmJc+DjZWVFQwNDfk/HlI7URSRlZWF9PR0AICtra2GKyIikheGmxcUFBRIwcbS0lLT5ZCMKRQKAEB6ejqsrKx4i4qISI04oPgFz8fYGBoaargSqg6e/5xxbBcRkXox3LwCb0VRReDPGRFR+WC4ISIi0qCIiAj4+PggIiJC06XIBsMNERGRhmRnZyMwMBD//PMPAgMDkZ2dremSZIHhhoiISEPWrl0rPTmZnp6OkJAQDVckDww3pbBmzRqYmJggPz9fanv8+DF0dXXRrl07pW2PHz8OQRBw+fJlAEBMTAx69OgBCwsLGBgYwN3dHUuWLEFBQYHSfoIgQBAEnD59Wqk9NzcXlpaWEAQBUVFRRWobPXo0tLW1sX379lfWfuXKFQwfPhx16tSBvr4+6tWrh8GDByM2NrZE5/68LkEQYGJiAi8vL+zcubNE+wYGBkr7amlpwc7ODn5+fkhJSSnR/kREcnTz5k2sW7cOoigCeDZNREhICG7evKnhyqo+hptS8PHxwePHj5UCwfHjx2FjY4MzZ84gKytLao+KioKdnR2cnJywa9cudOjQAXXq1EFkZCQuXbqEyZMn48svv8T7778v/WA/Z29vjw0bNii17dq1C8bGxq+sKysrCzt27MCMGTMQGhpaZH1sbCxatGiBy5cvY+3atYiPj8euXbvg7OyMTz75pMTnv2HDBqSlpeHMmTNo1qwZBgwYgFOnTpVoXzc3N6SlpeHWrVvYsWMHLly4gIEDB5b42EREciKKIubPn1/k7//i2ql0GG5KoXHjxrCzs1O6chIVFYU+ffqgQYMGiImJUWr38fHBkydP4O/vj969eyMkJAQeHh5wdHTEqFGjsHHjRvz888/48ccflY4zdOhQbN++Xene6/fff4+hQ4e+sq6ffvoJrq6umD17Nk6ePIkbN25I60RRxLBhw9CoUSMcP34cPXv2RIMGDeDh4YGAgAD88ssvJT7/5zPqOjs7Y82aNTAwMMCePXtKtK+Ojg5sbGxgZ2eHdu3awd/fH6dPn0ZmZqa0zd69e9GiRQsYGBigfv36mDdvntJVssDAQNStWxf6+vqws7PDpEmTpHWOjo4ICgrCBx98AGNjY9jZ2WHlypVKNSQnJ6NPnz4wNjaGqakpBg4ciDt37ij17+Hhgc2bN8PR0RFmZmZ4//338ejRI2mbn3/+Ge7u7lAoFLC0tETnzp3x5MkTaf2GDRvg4uICAwMDODs747vvvivx90tE1ce1a9dw4sSJIlfvCwoKcOLECVy7dk1DlckDw00pdezYEZGRkdLnyMhIdOzYER06dJDa8/LycOrUKfj4+ODw4cP4999/MX369CJ99erVC05OTti2bZtSe4sWLVCvXj2Eh4cDAFJSUnDs2DF89NFHr6wpNDQUH374IczMzNCjRw+lqz7nz5/HxYsX8cknn0BLq+h/bnNz81J/B8CzdyPp6OioNEfL7du3sXPnTmhra0uT1x06dAgffvghJk2ahPj4eKxduxZhYWH48ssvATwLFcuWLcPatWuRlJSE3bt3w93dXanfb775Bk2bNsXZs2cxe/ZsTJ06FUeOHAHwLOT17dsX9+/fR3R0NI4cOYKrV69i0KBBSn1cvXoVu3fvxr59+7Bv3z5ER0dj0aJFAIC0tDQMHjwYI0aMQEJCAqKiotC/f3/pX1jr1q3D559/ji+//BIJCQlYuHAh5syZg40bN5b6OyIieatfvz7atm1bZAJPbW1ttGvXDvXr19dQZTIhVkMPHz4UAYgPHz5Uas/Ozhbj4+PF7OzsYvcNCQkRjYyMxKdPn4qZmZmijo6OeOfOHXH79u1i69atRVEUxejoaBGAePXqVXHRokUiADEjI+OV/fXu3Vt0cXGRPgMQd+3aJS5fvlz08fERRVEU582bJ/br10/MyMgQAYiRkZHS9pcvXxZ1dXXFu3fviqIoirt27RLt7e3FgoICURRFcceOHSIA8ezZs6X+nl70vC5RFMWcnBwxKChIBCAeOHDgjfsGBASIWlpaopGRkahQKEQAIgBx0qRJ0jbt2rUTFy5cqLTf5s2bRVtbW1EURXHJkiWik5OTmJeX98pjODg4iN26dVNqGzRokNi9e3dRFEXx8OHDora2tpicnCytv3jxoghA/OOPP6Q6DQ0NxczMTGmbGTNmiC1bthRFURTj4uJEAOKNGzdeWYO9vb34ww8/KLUFBQWJ3t7er9y+JD9vRCRfN27cEF1dXUUnJydpcXV1FW/evKnp0qo8jV+5SU1NxYcffghLS0sYGhrCw8MDcXFxxW4fFRWlNLj1+XLp0qUKqff5raYzZ87g+PHjcHJygpWVFTp06IAzZ87gyZMniIqKQt26dZWSt1jM/VNRFF85mduHH36IU6dO4dq1awgLC8OIESNeuX9oaCi6du2KmjVrAgB69OiBJ0+e4LffflM6rjomjBs8eDCMjY1haGiIpUuXYvHixejevXuJ9m3cuDHOnz+PM2fO4Msvv4SHh4d0VQYA4uLiMH/+fBgbG0uLv78/0tLSkJWVhQEDBiA7Oxv169eHv78/du3apXTLCgC8vb2LfE5ISAAAJCQkwN7eHvb29tJ6V1dXmJubS9sAz25vmZiYSJ9tbW2lJxmaNWsGX19fuLu7Y8CAAVi3bh0yMjIAAHfv3kVKSgpGjhypdA4LFizA1atXS/QdEVH14uDgAH9/f+nvZ0EQMHr0aNStW1fDlVV9Gn23VEZGBtq0aQMfHx8cPHgQVlZWuHr1aolulSQmJsLU1FT6XKtWrXKs9P81bNhQGhickZGBDh06AABsbGxQr149nDx5EpGRkejUqRMAwMnJCcCz/7m2bt26SH+XLl2Cq6trkXZLS0u8++67GDlyJHJyctC9e3elsR/As3uzmzZtwu3bt6Gjo6PUHhoaii5duigd38PDo0znvmzZMnTu3BmmpqawsrIq1b56enpo2LAhgGeDi5OSkvDxxx9j8+bNAIDCwkLMmzcP/fv3L7KvgYEB7O3tkZiYiCNHjuC3337DuHHj8M033yA6Ohq6urrFHvf5XxrFhciX21/uSxAEFBYWAnh2ufjIkSOIiYnB4cOHsXLlSnz++ef4/fffpVcprFu3Di1btlTqg++NIqLijBkzBjt37sSdO3dgbW2N0aNHa7okWdBouPnqq6+KPBnk6OhYon2trKxUHi9SVj4+PoiKikJGRgZmzJghtXfo0AGHDh3C6dOnMXz4cABAly5dUKNGDSxZsqRIuNmzZw+SkpIQFBT0yuOMGDECPXr0wKxZs175P8gDBw7g0aNHOHfunNL6S5cuwc/PD//++y88PDzg6uqKJUuWYNCgQUXG3Tx48KDE36ONjY0UUMpqzpw5cHJywtSpU9G8eXM0b94ciYmJr+1foVCgd+/e6N27N8aPHw9nZ2dcuHABzZs3B4Aij8+fPn0azs7OAJ5dpUlOTkZKSop09SY+Ph4PHz6Ei4tLiesWBAFt2rRBmzZtMHfuXDg4OGDXrl2YNm0aateujWvXrsHPz6+0XwcRVVMKhQKBgYEICgrCnDlzpJfqUtloNNzs2bMHXbt2xYABAxAdHY3atWtj3Lhx8Pf3f+O+np6eyMnJgaurK7744gv4+PgUu21ubi5yc3Olzy8+oaMKHx8fjB8/Hk+fPpWu3ADPws3HH3+MnJwcqR4jIyOsXbsW77//PkaPHo0JEybA1NQUR48exYwZM/Cf//yn2Eeiu3Xrhrt37ypdoXpRaGgoevbsiWbNmim1u7m5YcqUKdiyZQsmT56MDRs2oHPnzmjfvj0+++wzODs74/Hjx9i7dy8OHz6M6OjoMn0fqqhfvz769OmDuXPnYt++fZg7dy7effdd2NvbY8CAAdDS0sJff/2FCxcuYMGCBQgLC0NBQQFatmwJQ0NDbN68GQqFAg4ODlKfJ0+exNdff42+ffviyJEj+Omnn7B//34AQOfOndG0aVP4+flh+fLlyM/Px7hx49ChQwd4eXmVqObff/8dR48eRZcuXWBlZYXff/8dd+/elcJRYGAgJk2aBFNTU3Tv3h25ubmIjY1FRkYGpk2bpv4vkYhkoVOnTtLVflITTQ740dfXF/X19cXZs2eLZ8+eFdesWSMaGBiIGzduLHafS5cuiSEhIWJcXJwYExMjfvzxx6IgCGJ0dHSx+wQEBEiDWF9cVBlQLIqieP36dRGA6OzsrNSekpIiAhAbNGhQZJ9jx46J3bp1E83MzEQ9PT3R1dVVXLx4sZifn6+0HV4YuPuyFwcU3759W9TR0RF//PHHV247ceJE0d3dXfqcmJgoDhkyRLSzsxP19PREBwcHcfDgwSUeaPy6ut4kICBAbNasWZH2kydPigDE06dPi6Ioir/++qvYunVrUaFQiKampuLbb78thoSEiKL4bKB0y5YtRVNTU9HIyEhs1aqV+Ntvv0l9OTg4iPPmzRMHDhwoGhoaitbW1uLy5cuVjnfz5k2xd+/eopGRkWhiYiIOGDBAvH379mvrXLZsmejg4CCKoijGx8eLXbt2FWvVqiXq6+uLTk5O4sqVK5W237p1q+jh4SHq6emJFhYWYvv27cWdO3e+8nvhgGIiovIhiKLmZgrS09ODl5eX0vwwkyZNwpkzZ0o8ORzw7JFqQRCKnXPlVVdu7O3t8fDhQ6WrIjk5Obh+/Trq1asHAwMDFc6INMXR0RFTpkzBlClTNF1KifHnjYiofGj0aSlbW9sig2ldXFyQnJxcqn5atWqFpKSkYtfr6+vD1NRUaSEiIiJ50mi4adOmDRITE5XaLl++rDSOoiTOnTsHW1tbdZZWrSxcuFDp8eUXl5I86l3cvsbGxjh+/HgFnAEREdH/0+iA4qlTp6J169ZYuHAhBg4ciD/++AMhISFKb0WdPXs2UlNTsWnTJgDA8uXL4ejoCDc3N+Tl5WHLli0IDw+XZvOl0hs7dmyxg5pLMnL//Pnzxa6rXbu2qmWVyouvnCAioupNo+Hmrbfewq5duzB79mzMnz8f9erVw/Lly5UepU1LS1O6TZWXl4fp06cjNTUVCoUCbm5u2L9/P3r06KGJU5CFGjVqoEaNGirvr67Hw4mIiNRBowOKNSUzMxNmZmYcUEwaxZ83IqLyofHXLxARERGpE8MNERERyQrDDREREcmKRgcUy1FBQYH0osWKoKWlxRczEhERvYDhRo0KCgrg07Yt7ty/X2HHtK5RA5EnTjDgEBER/Q/DjRoVFhbizv37+CUvr0K+2HwAfe7fR2FhYYnDzbBhw7Bx40YEBwfj008/ldp3796Nfv36oRo+PEdERDLDMTflQKcCF1UYGBjgq6++QkZGhoo9lMzTp0/LtX8iIqJXYbiphjp37gwbGxsEBwe/drvw8HC4ublBX18fjo6OWLJkyWu3DwwMhIeHB77//nvUr18f+vr6EEURDx8+xOjRo2FlZQVTU1N06tQJf/75p7Tfn3/+CR8fH5iYmMDU1BQtWrRAbGwsACAsLAzm5ubYvXs3nJycYGBggHfeeQcpKSlKx169ejUaNGgAPT09NG7cGJs3b1ZaLwgC1q9fj379+sHQ0BCNGjVSetFqRkYG/Pz8UKtWLSgUCjRq1AgbNmyQ1qempmLQoEGwsLCApaUl+vTpw1mRiYgqKYabakhbWxsLFy7EypUrcevWrVduExcXh4EDB+L999/HhQsXEBgYiDlz5iAsLOy1fV+5cgU//vgjwsPDpdcy9OzZE7dv38aBAwcQFxeH5s2bw9fXF/f/NzbJz88PderUwZkzZxAXF4dPP/0Uurq6Up9ZWVn48ssvsXHjRpw8eRKZmZl4//33pfW7du3C5MmT8cknn+Dvv//GmDFjMHz4cERGRirVNm/ePAwcOBB//fUXevToAT8/P6mGOXPmID4+HgcPHkRCQgJWr16NmjVrSsf38fGBsbExjh07hhMnTsDY2BjdunVDXl5eqb57IiIqfxxzU03169cPHh4eCAgIQGhoaJH1S5cuha+vL+bMmQMAcHJyQnx8PL755hsMGzas2H7z8vKwefNm1KpVCwAQERGBCxcuID09Hfr6+gCAxYsXY/fu3fj5558xevRoJCcnY8aMGXB2dgYANGrUSKnPp0+fYtWqVWjZsiUAYOPGjXBxccEff/yBt99+G4sXL8awYcMwbtw4AMC0adNw+vRpLF68GD4+PlI/w4YNw+DBgwFACnd//PEHunXrhuTkZHh6esLLywsA4OjoKO23fft2aGlpYf369RAEAQCwYcMGmJubIyoqCl26dCnZl05ERBWCV26qsa+++gobN25EfHx8kXUJCQlo06aNUlubNm2QlJSEgoKCYvt0cHCQgg3w7ArQ48ePYWlpqfS28OvXr+Pq1asAnoWRUaNGoXPnzli0aJHU/pyOjo4UOgDA2dkZ5ubmSEhIeG2tz9c/17RpU+nPRkZGMDExQXp6OgDg448/xvbt2+Hh4YGZM2ciJiZG6RyuXLkCExMTqf4aNWogJyenSK1ERKR5vHJTjbVv3x5du3bFZ599VuRqjCiK0lWKF9vexMjISOlzYWEhbG1tERUVVWRbc3NzAM/G6nzwwQfYv38/Dh48iICAAGzfvh39+vWTtn25lpfbXlXry20v3up6vs/zOYm6d++OmzdvYv/+/fjtt9/g6+uL8ePHY/HixSgsLESLFi2wdevWIjW8GOSIiKhy4JWbam7RokXYu3ev0pUKAHB1dcWJEyeU2mJiYuDk5FSqOXWaN2+O27dvQ0dHBw0bNlRano9pAZ7d9po6dSoOHz6M/v37Kw3mzc/PlwYYA0BiYiIePHgg3cZycXF5Za0uLi4lrhN4FlSGDRuGLVu2YPny5QgJCZHOISkpCVZWVkXOwczMrFTHICKi8sdwUw7yK3ApK3d3d/j5+WHlypVK7Z988gmOHj2KoKAgXL58GRs3bsSqVaswffr0UvXfuXNneHt7o2/fvjh06BBu3LiBmJgYfPHFF4iNjUV2djYmTJiAqKgo3Lx5EydPnsSZM2eUgomuri4mTpyI33//HWfPnsXw4cPRqlUrvP322wCAGTNmICwsDGvWrEFSUhKWLl2KnTt3lqrWuXPn4pdffsGVK1dw8eJF7Nu3T6rBz88PNWvWRJ8+fXD8+HFcv34d0dHRmDx5crEDsomISHN4W0qNtLS0YF2jBvpU8AzFWlply6hBQUH48ccfldqaN2+OH3/8EXPnzkVQUBBsbW0xf/781w4mfhVBEHDgwAF8/vnnGDFiBO7evQsbGxu0b98e1tbW0NbWxr///oshQ4bgzp07qFmzJvr374958+ZJfRgaGmLWrFn44IMPcOvWLbRt2xbff/+9tL5v375YsWIFvvnmG0yaNAn16tXDhg0b0LFjxxLXqaenh9mzZ+PGjRtQKBRo164dtm/fLh3/2LFjmDVrFvr3749Hjx6hdu3a8PX1hampaam+DyIiKn+CWA2npM3MzISZmRkePnyo9D+nnJwcXL9+HfXq1YOBgYFKffPdUuoVFhaGKVOm4MGDB5ouRe3U8fNGRERF8cqNmmlra8s6bBAREVV2HHNDREREssJwQ5XasGHDZHlLioiIyg/DDREREckKww0RERHJCsMNERERyQrDDREREckKww0RERHJCue5UTNO4kdERKRZDDdqVFBQgDbt2yDjXkaFHdOipgVOHjvJgENERPQ/DDdqVFhYiIx7GUjvmg4IFXBAEcChZ8ctabgZNmwYNm7cCODZbMp2dnbo2bMnFi5cCAsLi3IsloiIqGIw3JQHARUzmknFu1/dunXDhg0bkJ+fj/j4eIwYMQIPHjzAtm3b1FsfERGRBnBAcTWkr68PGxsb1KlTB126dMGgQYNw+PBhaf2GDRvg4uICAwMDODs747vvvpPW5eXlYcKECbC1tYWBgQEcHR0RHBwsrRcEAatXr0b37t2hUChQr149/PTTT0rHv3DhAjp16gSFQgFLS0uMHj0ajx8/ltYPGzYMffv2xeLFi2FrawtLS0uMHz8eT58+lbb57rvv0KhRIxgYGMDa2hr/+c9/pHWiKOLrr79G/fr1oVAo0KxZM/z8889q/Q6JiKjy4pWbau7atWv49ddfoaurCwBYt24dAgICsGrVKnh6euLcuXPw9/eHkZERhg4dim+//RZ79uzBjz/+iLp16yIlJQUpKSlKfc6ZMweLFi3CihUrsHnzZgwePBhNmjSBi4sLsrKy0K1bN7Rq1QpnzpxBeno6Ro0ahQkTJiAsLEzqIzIyEra2toiMjMSVK1cwaNAgeHh4wN/fH7GxsZg0aRI2b96M1q1b4/79+zh+/Li07xdffIGdO3di9erVaNSoEY4dO4YPP/wQtWrVQocOHSrkeyUiIs1huKmG9u3bB2NjYxQUFCAnJwcAsHTpUgBAUFAQlixZgv79+wMA6tWrh/j4eKxduxZDhw5FcnIyGjVqhLZt20IQBDg4OBTpf8CAARg1apTU35EjR7By5Up899132Lp1K7Kzs7Fp0yYYGRkBAFatWoVevXrhq6++grW1NQDAwsICq1atgra2NpydndGzZ08cPXoU/v7+SE5OhpGREd59912YmJjAwcEBnp6eAIAnT55g6dKliIiIgLe3NwCgfv36OHHiBNauXctwQ0RUDTDcVEM+Pj5YvXo1srKysH79ely+fBkTJ07E3bt3kZKSgpEjR8Lf31/aPj8/H2ZmZgCe3TJ655130LhxY3Tr1g3vvvsuunTpotT/81Dx4ufz588DABISEtCsWTMp2ABAmzZtUFhYiMTERCncuLm5KQ2StrW1xYULFwAA77zzDhwcHFC/fn1069YN3bp1Q79+/WBoaIj4+Hjk5OTgnXfeUaohLy9PCkBERCRvDDfVkJGRERo2bAgA+Pbbb+Hj44N58+ZhwoQJAJ7dmmrZsqXSPs+DRvPmzXH9+nUcPHgQv/32GwYOHIjOnTu/cUyLIDx7fEwURenPxW0DQLpN9uK65/MHmZiY4OzZs4iKisLhw4cxd+5cBAYG4syZM9I2+/fvR+3atZX60NfXf22NREQkDxxQTAgICMDixYtRUFCA2rVr49q1a2jYsKHSUq9ePWl7U1NTDBo0COvWrcOOHTsQHh6O+/fvS+tPnz6t1P/p06fh7OwMAHB1dcX58+fx5MkTaf3JkyehpaUFJyenEteso6ODzp074+uvv8Zff/2FGzduICIiAq6urtDX10dycnKRc7C3t1f1KyIioiqEV27KgwiVH9Mu9XHUoGPHjnBzc8PChQsRGBiISZMmwdTUFN27d0dubi5iY2ORkZGBadOmYdmyZbC1tYWHhwe0tLTw008/wcbGBubm5lJ/P/30E7y8vNC2bVts3boVf/zxB0JDQwEAfn5+CAgIwNChQxEYGIi7d+9i4sSJ+Oijj6RbUm+yb98+XLt2De3bt4eFhQUOHDiAwsJCNG7cGCYmJpg+fTqmTp2KwsJCtG3bFpmZmYiJiYGxsTGGDh2qni+NiIgqLYYbNdLS0oJFTQvgUMUd06KmBbS0yn4Bbtq0aRg+fDiuXLmC9evX45tvvsHMmTNhZGQEd3d3TJkyBQBgbGyMr776CklJSdDW1sZbb72FAwcOKNUwb948bN++HePGjYONjQ22bt0KV1dXAIChoSEOHTqEyZMn46233oKhoSHee+89aUBzSZibm2Pnzp0IDAxETk4OGjVqhG3btsHNzQ3As0HMVlZWCA4OxrVr12Bubo7mzZvjs88+K/P3RERElZ8giqKa/v1fdWRmZsLMzAwPHz6Eqamp1J6Tk4Pr16+jXr16MDAwUKnv6v5uKUEQsGvXLvTt21fTpVR66vh5IyKionjlRs20tbUrVdggIiKqbjigmIiIiGSFV25IrarhXU4iIqpkeOXmFfg/aKoI/DkjIiofDDcveD5xXFZWloYroerg+c/ZyxMWEhFR2fC21Au0tbVhbm6O9PR0AM8eWy5uNl0iVYmiiKysLKSnp8Pc3JwD0ImI1Izh5iU2NjYAIAUcovJibm4u/bwREZH6aDzcpKamYtasWTh48CCys7Ph5OSE0NBQtGjRoth9oqOjMW3aNFy8eBF2dnaYOXMmxo4dq5Z6BEGAra0trKys8PTpU7X0SfQyXV1dXrEhIionGg03GRkZaNOmDXx8fHDw4EFYWVnh6tWrSlP5v+z69evo0aMH/P39sWXLFpw8eRLjxo1DrVq18N5776mtNs5XQ0REVDVpdIbiTz/9FCdPnsTx48dLvM+sWbOwZ88eJCQkSG1jx47Fn3/+iVOnTpWoj+JmKCYiIqKqT6NPS+3ZswdeXl4YMGAArKys4OnpiXXr1r12n1OnTqFLly5KbV27dkVsbGyxt5Fyc3ORmZmptBAREZE8aTTcXLt2DatXr0ajRo1w6NAhjB07FpMmTcKmTZuK3ef27dtF3h5tbW2N/Px83Lt375X7BAcHw8zMTFrs7e3Veh5ERERUeWg03BQWFqJ58+ZYuHAhPD09MWbMGPj7+2P16tWv3e/lx7Of31kr7rHt2bNn4+HDh9KSkpKinhMgIiKiSkej4cbW1haurq5KbS4uLkhOTi52HxsbG9y+fVupLT09HTo6OrC0tHzlPvr6+jA1NVVaiIiISJ40Gm7atGmDxMREpbbLly/DwcGh2H28vb1x5MgRpbbDhw/Dy8uLM70SERGRZsPN1KlTcfr0aSxcuBBXrlzBDz/8gJCQEIwfP17aZvbs2RgyZIj0eezYsbh58yamTZuGhIQEfP/99wgNDcX06dM1cQpERERUyWg03Lz11lvYtWsXtm3bhiZNmiAoKAjLly+Hn5+ftE1aWprSbap69erhwIEDiIqKgoeHB4KCgvDtt9+qdY4bIiIiqro0Os+NpnCeGyIiIvniW8GJiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWVAo3gYGBuHnzprprISIiIiozlcLN3r170aBBA/j6+uKHH35ATk6OuusiIiIiUolK4SYuLg5nz55F06ZNMXXqVNja2uLjjz/GmTNnStVPYGAgBEFQWmxsbIrdPioqqsj2giDg0qVLqpwGERERyZDKY26aNm2KZcuWITU1Fd9//z1SU1PRpk0buLu7Y8WKFXj48GGJ+nFzc0NaWpq0XLhw4Y37JCYmKu3TqFEjVU+DiIiIZKbMA4oLCwuRl5eH3NxciKKIGjVqYPXq1bC3t8eOHTveuL+Ojg5sbGykpVatWm/cx8rKSmkfbW3t126fm5uLzMxMpYWIiIjkSeVwExcXhwkTJsDW1hZTp06Fp6cnEhISEB0djUuXLiEgIACTJk16Yz9JSUmws7NDvXr18P777+PatWtv3MfT0xO2trbw9fVFZGTkG7cPDg6GmZmZtNjb25foHImIiKjqEURRFEu7U9OmTZGQkIAuXbrA398fvXr1KnL15O7du7C2tkZhYWGx/Rw8eBBZWVlwcnLCnTt3sGDBAly6dAkXL16EpaVlke0TExNx7NgxtGjRArm5udi8eTPWrFmDqKgotG/fvtjj5ObmIjc3V/qcmZkJe3t7PHz4EKampqU9fSIiIqrEVAo3QUFBGDFiBGrXrq3WYp48eYIGDRpg5syZmDZtWon26dWrFwRBwJ49e0p8nMzMTJiZmTHcEBERyZBKt6XmzJkjBRtRFKFCPnolIyMjuLu7IykpqcT7tGrVqlTbExERkbypPOYmNDQUTZo0gYGBAQwMDNCkSROsX7++TMXk5uYiISEBtra2Jd7n3LlzpdqeiIiI5E1HlZ3mzJmDZcuWYeLEifD29gYAnDp1ClOnTsWNGzewYMGCEvUzffp09OrVC3Xr1kV6ejoWLFiAzMxMDB06FAAwe/ZspKamYtOmTQCA5cuXw9HREW5ubsjLy8OWLVsQHh6O8PBwVU6DiIiIZEilcLN69WqsW7cOgwcPltp69+6Npk2bYuLEiSUON7du3cLgwYNx79491KpVC61atcLp06fh4OAAAEhLS0NycrK0fV5eHqZPn47U1FQoFAq4ublh//796NGjhyqnQURERDKk0oBiCwsL/PHHH0Umz7t8+TLefvttPHjwQF31lQsOKCYiIpIvlcbcfPjhh1i9enWR9pCQEPj5+ZW5KCIiIiJVqXRbCng2oPjw4cNo1aoVAOD06dNISUnBkCFDlB7jXrp0admrJCIiIiohlW5L+fj4lKxzQUBERESpiypvvC1FREQkXyqFm6qO4YaIiEi+yvzizFu3biE1NVUdtRARERGVmUrhprCwEPPnz4eZmRkcHBxQt25dmJubIygo6LXvkiIiIiIqbyoNKP78888RGhqKRYsWoU2bNhBFESdPnkRgYCBycnLw5ZdfqrtOIiIiohJRacyNnZ0d1qxZg969eyu1//LLLxg3blylv03FMTdERETypdJtqfv378PZ2blIu7OzM+7fv1/mooiIiIhUpVK4adasGVatWlWkfdWqVWjWrFmZiyIiIiJSlUpjbr7++mv07NkTv/32G7y9vSEIAmJiYpCSkoIDBw6ou0Z6BVEUkZ2dXa795+TkAAAMDAwgCEK5HUuhUJRr/0REVL2oPM/NP//8g//+97+4dOkSRFGEq6srxo0bBzs7O3XXqHZyGHOTlZUFT09PTZehFufOnYOhoaGmyyAiIpko9ZWbp0+fokuXLli7di2fiiIiIqJKp9ThRldXF3///TdvI2iYQqHAuXPnyq3/7OxstG7dGgAQExMDhUJRbscqz76JiKj6UWnMzZAhQ6R5bkgzBEGosFs5CoWCt41eg+OfiIgqF5XCTV5eHtavX48jR47Ay8sLRkZGSuv5JnCqTrKzszn+iYioElEp3Pz9999o3rw5AODy5ctqLYiIiIioLFQKN5GRkequg6jK4vgnIqLKRaVwM2LECKxYsQImJiZK7U+ePMHEiRPx/fffq6U4oqqA45+IiCoXlWYo3rhx4ysHUGZnZ2PTpk1lLoqIiIhIVaW6cpOZmQlRFCGKIh49egQDAwNpXUFBAQ4cOAArKyu1F0lERERUUqUKN+bm5hAEAYIgwMnJqch6QRAwb948tRVHREREVFqlCjeRkZEQRRGdOnVCeHg4atSoIa3T09ODg4NDlXj9AhEREclXqcJNhw4dAADXr1+Hvb09tLRUGrJDRESvwYkhicpGpaelHBwc8ODBA/zxxx9IT09HYWGh0vohQ4aopTgiouqIE0MSlY1K4Wbv3r3w8/PDkydPYGJiopTKBUFguCEiIiKNUSncfPLJJxgxYgQWLlzIRE5EpGacGJKobFQKN6mpqZg0aRKDDRFROeDEkERlo9KI4K5duyI2NlbdtRARERGVmUpXbnr27IkZM2YgPj4e7u7u0NXVVVrfu3dvtRRHREREVFoqhRt/f38AwPz584usEwQBBQUFZauKiIiISEUqhZuXH/0mIiIiqizKPAvf84mgiIiIiCoDlcJNQUEBgoKCULt2bRgbG+PatWsAgDlz5iA0NFStBRIRERGVhkrh5ssvv0RYWBi+/vpr6OnpSe3u7u5Yv3692oojIiIiKi2Vws2mTZsQEhICPz8/aGtrS+1NmzbFpUuX1FYcERERUWmpPIlfw4YNi7QXFhbi6dOnZS6qqivvl95VhBfrr+rnAvDlfURE1YlK4cbNzQ3Hjx+Hg4ODUvtPP/0km5e9lYWcXnoHQJqmvSrjy/uIiKoPlcJNQEAAPvroI6SmpqKwsBA7d+5EYmIiNm3ahH379qm7RiIiIqISUync9OrVCzt27MDChQshCALmzp2L5s2bY+/evXjnnXfUXWOVtiMvDwaaLkIFIoDc//1ZH0BVvKGTA2DQCwPeiYioelAp3ADP3i/VtWvX126zbds29O7dG0ZGRqoepsoz+N9SFcnpXb5VedwQxz8REZWOyuGmJMaMGYOWLVuifv365XkYojeSw7ghQB7nwfFPRFTeyjxD8euIolie3RMREREVUa5Xbt4kMDAQ8+bNU2qztrbG7du3i90nOjoa06ZNw8WLF2FnZ4eZM2di7Nix5V0qycRd37sQtatY6BYBPH8XrTaq5AAooUBAraO1NF0GEVUTGg03wLPHyn/77Tfp84uTAr7s+vXr6NGjB/z9/bFlyxacPHkS48aNQ61atfDee+9VRLlUxYnaYiX4qVeBrqYLKBsRVSxQvgbnsap8OI6LXqbxv+Z1dHRgY2NTom3XrFmDunXrYvny5QAAFxcXxMbGYvHixQw3RFQhOI9V5VOe47jKO8yKoii9gNrAwKBcQ1p1CoEaDzdJSUmws7ODvr4+WrZsiYULFxY7APnUqVPo0qWLUlvXrl0RGhqKp0+fQlf31f+8zc3NRW5urvQ5MzNTfSdARESyJacwW50G85druHFwcCg2cABAy5YtsWnTJjg5OeHOnTtYsGABWrdujYsXL8LS0rLI9rdv34a1tbVSm7W1NfLz83Hv3j3Y2tq+8jjBwcFFxvYQEZUV57HSHM5jRa9TpnCTl5eH9PR0FBYWKrXXrVsXAPD333+/dv/u3btLf3Z3d4e3tzcaNGiAjRs3Ytq0aa/c5+VLas+fyHrdpbbZs2cr9ZeZmQl7e/vX1kZE9Cacx0r+FAoFzp07V279Z2dnS7cGY2JioFCU33+Z8uy7slEp3CQlJWHEiBGIiYlRahdFEYIgoKCgoJg9X8/IyAju7u5ISkp65XobG5siT1Klp6dDR0fnlVd6ntPX14e+vr5KNRERUfUlCEKF3cpRKBTV5rZReVMp3AwbNgw6OjrYt28fbG1t1TZAKTc3FwkJCWjXrt0r13t7e2Pv3r1KbYcPH4aXl9drb38RERFR9aFSuDl//jzi4uLg7OxcpoNPnz4dvXr1Qt26dZGeno4FCxYgMzMTQ4cOBfDsdlJqaio2bdoEABg7dixWrVqFadOmwd/fH6dOnUJoaCi2bdtWpjqIiIhIPlQKN66urrh3716ZD37r1i0MHjwY9+7dQ61atdCqVSucPn0aDg4OAIC0tDQkJydL29erVw8HDhzA1KlT8d///hd2dnb49ttv+Rg4EVE1VpXn6uGcQ+VDEFV4R0JERAS++OILLFy4EO7u7kVuCZmamqqtwPKQmZkJMzMzPHz4sFxqzcrKkh4d/KWKPk0hBzkA+rz0NEV6l/RKMAFCNZQPWB22AlD1H0fl73fl8Krfb9K8yvL7rdJf8507dwYA+Pr6KrWXdUAxERERUVmpFG4iIyPVXQcREVGZ8N1xmlEZ3x2nUrjp0KGDuusgIiIqE747TjMq47vjVP4xePDgAUJDQ5GQkABBEODq6ooRI0bAzMxMnfURERERlYqWKjvFxsaiQYMGWLZsGe7fv4979+5h6dKlaNCgAc6ePavuGomIiIhKTKUrN1OnTkXv3r2xbt066Og86yI/Px+jRo3ClClTcOzYMbUWSURERFRSKoWb2NhYpWADADo6Opg5cya8vLzUVhwRERFRaal0W8rU1FRpcr3nUlJSYGJiUuaiiIiIiFSl0pWbQYMGYeTIkVi8eDFat24NQRBw4sQJzJgxA4MHD1Z3jUTqk6/pAqopfu9EVIFUCjeLFy+GIAgYMmQI8vOf/a2lq6uLjz/+GIsWLVJrgVVdjqYLqMZe9d1bRVhVeB1ERFSxVAo3enp6WLFiBYKDg3H16lWIooiGDRtWiimXK5tBnB6ciIioQpVpuiNDQ0O4u7urqxaicpfeie+W0oh8XjUjooqj0l/zT548waJFi3D06FGkp6ejsLBQaf21a9fUUpwc7OCL9TQmB6+4cqYDhhsiIplT6a/5UaNGITo6Gh999BFsbW0rxevNKyuD/y1ERERUMVQKNwcPHsT+/fvRpk0bdddDREREVCYqzXNjYWGBGjVqqLsWIiIiojJTKdwEBQVh7ty5yMrKUnc9RERERGWi0m2pJUuW4OrVq7C2toajoyN0dZXf186XZxIREZGmqBRu+vbtq+YyiIiIyogzYWtGJfzeVQo3AQEB6q6DiIioTDiXEj1Xphk/8vLyXjnPTd26dctUFBEREZGqVAo3ly9fxsiRIxETE6PULooiBEFAQUGBWoojIiIqKc5AriGVcAZylX4Mhg8fDh0dHezbt4+T+BERUeXAGcjpf1T6MTh//jzi4uLg7Oys7nqIiIiIykSleW5cXV1x7949dddCREREVGYqhZuvvvoKM2fORFRUFP79919kZmYqLURERESaotJtqc6dOwMAfH19ldo5oJiIiIg0TaVwExkZqe46iIiIiNRCpXDToUMHdddBREREpBZlemguKysLycnJyMvLU2pv2rRpmYoiIiIiUpVK4ebu3bsYPnw4Dh48+Mr1HHNDREREmqJSuJkyZQoyMjJw+vRp+Pj4YNeuXbhz5w4WLFiAJUuWqLtGIrURCgSIEDVdRumIAJ7/e0EbQBWcM1MoqIJFl0COpguoxvjd0+uoFG4iIiLwyy+/4K233oKWlhYcHBzwzjvvwNTUFMHBwejZs6e66yRSi1pHa2m6BJKRQXp6mi6BiF5BpXlunjx5AiurZ++RqFGjBu7evQsAcHd3x9mzZ9VXHREREVEpqXTlpnHjxkhMTISjoyM8PDywdu1aODo6Ys2aNbC1tVV3jURlFhMTA4VCoekyVJKdnY3WrVsDqNrn8VxVr/9FO/LyYKDpIqqpHPDKGRVP5TE3aWlpAICAgAB07doVW7duhZ6eHsLCwtRZH5FaKBQKGBoaarqMMpPLeciFwf8WIqpcVAo3fn5+0p89PT1x48YNXLp0CXXr1kXNmjXVVhwRERFRaanl5fCGhoZo3ry5OroiIqoyyuuJHRFAbjn1XdH0UT4P+PFpKXodlcJNQUEBwsLCcPToUaSnp6OwsFBpfUREhFqKIyKqzDjmg6hyUincTJ48GWFhYejZsyeaNGkCQZDnHBZERERU9agUbrZv344ff/wRPXr0UHc9RESVmkKhwLlz58r1GKIoIidHHjdeDAwMyvUfwC8+TUj0nErhRk9PDw0bNlR3LURElZ4gCBXyxJqRkVG5H4NIrlSaxO+TTz7BihUrIIpVbBp7IiIikr0SX7np37+/0ueIiAgcPHgQbm5u0NXVVVq3c+dO9VRHREREVEolvnJjZmamtPTr1w8dOnRAzZo1i6xTRXBwMARBwJQpU4rdJioqCoIgFFkuXbqk0jGJiIhIfkp85WbDhg3Sn7Ozs1FYWCjdE75x4wZ2794NFxcXdO3atdRFnDlzBiEhIWjatGmJtk9MTISpqan0uVYtvgyRiIiInlFpzE2fPn2wefNmAMCDBw/QqlUrLFmyBH379sXq1atL1dfjx4/h5+eHdevWwcLCokT7WFlZwcbGRlq0tbVLfQ5EREQkTyqFm7Nnz6Jdu3YAgJ9//hnW1ta4efMmNm3ahG+//bZUfY0fPx49e/ZE586dS7yPp6cnbG1t4evri8jIyDdun5ubi8zMTKWFiIiI5EmlR8GzsrJgYmICADh8+DD69+8PLS0ttGrVCjdv3ixxP9u3b8fZs2dx5syZEm1va2uLkJAQtGjRArm5udi8eTN8fX0RFRWF9u3bF7tfcHAw5s2bV+K6iIiIqOpSKdw0bNgQu3fvRr9+/XDo0CFMnToVAJCenq40FuZ1UlJSMHnyZBw+fBgGBiV7r27jxo3RuHFj6bO3tzdSUlKwePHi14ab2bNnY9q0adLnzMxM2Nvbl+iYREREVLWoFG7mzp2LDz74AFOnToWvry+8vb0BPLuK4+npWaI+4uLikJ6ejhYtWkhtBQUFOHbsGFatWoXc3NwSjaVp1aoVtmzZ8tpt9PX1oa+vX6K6iIioahIKBIioYvOviQAK/vdnbZTPW0bLmVBQ+YpWKdz85z//Qdu2bZGWloZmzZpJ7b6+vujXr1+J+vD19cWFCxeU2oYPHw5nZ2fMmjWrxIOEz507B1tb25IXT0REslTrKJ+cpWdUCjcApCeVXvT222+XeH8TExM0adJEqc3IyAiWlpZS++zZs5GamopNmzYBAJYvXw5HR0e4ubkhLy8PW7ZsQXh4OMLDw1U9DSIiIpIZlcNNRUhLS0NycrL0OS8vD9OnT0dqaioUCgXc3Nywf/9+vsCTiKiaqogXmZanF1/8GRMTA4VCoeGKyqay1F+pwk1UVJTS57CwMKXPM2fOxMyZMyuuICIiqtQq6kWmFUGhUMjmXDRNpXluiIiIiCorhhsiIiKSFYYbIiIikpVKNeaGiIioMhFFEdnZ2eXW/4t9l+dxgGdjegSh8s1JUx4YboiIiIqRnZ1d4slpy+r5U1Pl5dy5c9VmwDJvSxEREZGs8MoNERFRMcp7Hh1RFJGTkwMAMDAwKNfbRpVlDpqKwHBDRERUjIqYR8fIyKhc+6+OeFuKiIiIZIXhhoiIiGSF4YaIiIhkheGGiIiIZIXhhoiIiGSF4YaIiIhkheGGiIiIZIXz3BCVEd89Q0RUuTDcEJUR3z1DRFS58LYUERERyQqv3BCVEd89Q0RUuTDcEJUR3z1DRFS58LYUERERyQrDDREREckKww0RERHJCsMNERERyQrDDREREckKww0RERHJCsMNERERyQrDDREREckKww0RERHJCmcoLmc5mi5ARSKA3P/9WR9AVXxPdFX97omIqGwYbsrZID09TZdARERUrfC2FBEREckKr9yUg/J+S3RFyM7ORuvWrQEAMTExVf5t0VW9fiIiKjmGm3JQEW+JrkgKhUJW50NERPLG21JEREQkKww3REREJCsMN0RERCQrDDdEREQkKww3REREJCsMN0RERCQrDDdEREQkKww3REREJCsMN0RERCQrDDdEREQkK5Um3AQHB0MQBEyZMuW120VHR6NFixYwMDBA/fr1sWbNmoopkIiIiKqEShFuzpw5g5CQEDRt2vS1212/fh09evRAu3btcO7cOXz22WeYNGkSwsPDK6hSIiIiquw0Hm4eP34MPz8/rFu3DhYWFq/dds2aNahbty6WL18OFxcXjBo1CiNGjMDixYsrqFoiIiKq7DQebsaPH4+ePXuic+fOb9z21KlT6NKli1Jb165dERsbi6dPnxa7X25uLjIzM5UWIiIikieNhpvt27fj7NmzCA4OLtH2t2/fhrW1tVKbtbU18vPzce/evWL3Cw4OhpmZmbTY29uXqW4iIiKqvDQWblJSUjB58mRs2bIFBgYGJd5PEASlz6IovrL9RbNnz8bDhw+lJSUlRbWiiYiIqNLT0dSB4+LikJ6ejhYtWkhtBQUFOHbsGFatWoXc3Fxoa2sr7WNjY4Pbt28rtaWnp0NHRweWlpbFHktfXx/6+vrqPQEiIiKqlDQWbnx9fXHhwgWltuHDh8PZ2RmzZs0qEmwAwNvbG3v37lVqO3z4MLy8vKCrq1uu9RIREVHVoLFwY2JigiZNmii1GRkZwdLSUmqfPXs2UlNTsWnTJgDA2LFjsWrVKkybNg3+/v44deoUQkNDsW3btgqvn4iIiConjT8t9TppaWlITk6WPterVw8HDhxAVFQUPDw8EBQUhG+//RbvvfeeBqskIiKiykQQn4/IrUYyMzNhZmaGhw8fwtTUVNPlVEpZWVnw9PQEAJw7dw6GhoYaroiIiKhkKvWVGyIiIqLSYrghIqqGIiIi4OPjg4iICE2XQqR2DDdERNVMdnY2AgMD8c8//yAwMBDZ2dmaLolIrRhuiIiqmbVr1yI9PR3As7nCQkJCNFwRkXox3BARVSM3b97EunXrpNndRVFESEgIbt68qeHKiNSH4YaIqJoQRRHz58/Hyw/JFtdOVFUx3BARVRPXrl3DiRMnUFBQoNReUFCAEydO4Nq1axqqjEi9GG6IiKqJ+vXro23btkVeb6OtrY127dqhfv36GqqMSL0YboiIqglBEDB37lwIglCidqKqiuGGiKgacXBwgL+/vxRkBEHA6NGjUbduXQ1XRqQ+DDdERNXMmDFjYGVlBQCwtrbG6NGjNVwRkXox3BARVTMKhQKBgYGws7NDQEAAFAqFpksiUisdTRdAREQVr1OnTujUqZOmyyAqF7xyQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssIZiqsoURSRnZ1dbv2/2Hd5Hgd4NhU830ZMRETqwnBTRWVnZ8PT07NCjtW6dety7f/cuXMwNDQs12MQEVH1wdtSREREJCu8clNFKRQKnDt3rtz6F0UROTk5AAADA4NyvW3ENxITEZE6MdxUUYIglPutHCMjo3Ltn4iIqDzwthQRERHJCsMNERERyQrDDREREckKww0RERHJCsMNERERyQrDDREREckKww0RERHJCsMNERERyQrDDREREckKww0RERHJCsMNERERyQrDDREREckKww0RERHJCsMNERERyQrDDREREckKww0RERHJCsMNURUQEREBHx8fREREaLoUIqJKT6PhZvXq1WjatClMTU1hamoKb29vHDx4sNjto6KiIAhCkeXSpUsVWDVRxcrOzkZgYCD++ecfBAYGIjs7W9MlERFVahoNN3Xq1MGiRYsQGxuL2NhYdOrUCX369MHFixdfu19iYiLS0tKkpVGjRhVUMVHFW7t2LdLT0wEA6enpCAkJ0XBFRESVmyCKoqjpIl5Uo0YNfPPNNxg5cmSRdVFRUfDx8UFGRgbMzc1L3Gdubi5yc3Olz5mZmbC3t8fDhw9hamqqjrKJysXNmzfRo0cP5OfnS206Ojo4cOAAHBwcNFgZEVHlVWnG3BQUFGD79u148uQJvL29X7utp6cnbG1t4evri8jIyDf2HRwcDDMzM2mxt7dXV9lE5UYURcyfPx8v//ujuHYiInpG41duLly4AG9vb+Tk5MDY2Bg//PADevTo8cptExMTcezYMbRo0QK5ubnYvHkz1qxZg6ioKLRv377YY/DKDVVFV69eLfZ3AQAOHDiABg0aVGBFRERVg8bDTV5eHpKTk/HgwQOEh4dj/fr1iI6Ohqura4n279WrFwRBwJ49e0p8zMzMTJiZmTHcUKUmiiJGjRqFU6dOoaCgQGrX1tZG69atsW7dOgiCoMEKiYgqJ43fltLT00PDhg3h5eWF4OBgNGvWDCtWrCjx/q1atUJSUlI5VkikGYIgYO7cuUUCTHHtRET0jMbDzctEUVS6hfQm586dg62tbTlWRKQ5Dg4O8Pf3l4KMIAgYPXo06tatq+HKiIgqLx1NHvyzzz5D9+7dYW9vj0ePHmH79u2IiorCr7/+CgCYPXs2UlNTsWnTJgDA8uXL4ejoCDc3N+Tl5WHLli0IDw9HeHi4Jk+DqFyNGTMGO3fuxJ07d2BtbY3Ro0druiQiokpNo+Hmzp07+Oijj5CWlgYzMzM0bdoUv/76K9555x0AQFpaGpKTk6Xt8/LyMH36dKSmpkKhUMDNzQ379+9/7aBLoqpOoVAgMDAQQUFBmDNnDhQKhaZLIiKq1DQ+oFgTOKCYiIhIvirdmBsiIiKismC4ISIiIllhuCEiIiJZYbghIiIiWWG4ISIiIllhuCEiIiJZYbghIiIiWWG4ISIiIllhuCEiIiJZ0ejrFzTl+aTMmZmZGq6EiIiISsvExER6ofCrVMtw8+jRIwCAvb29hishIiKi0nrT65Oq5bulCgsL8c8//7wx+ZE8ZGZmwt7eHikpKXyXGJHM8Pe7euKVm1fQ0tJCnTp1NF0GVTBTU1P+5UckU/z9phdxQDERERHJCsMNERERyQrDDcmevr4+AgICoK+vr+lSiEjN+PtNr1ItBxQTERGRfPHKDREREckKww0RERHJCsMNERERyQrDDREREckKww0REcleVFQUBEHAgwcPNF0KVQCGG6o28vLyNF0CERFVAIYbqrI6duyICRMmYMKECTA3N4elpSW++OIL6a3vjo6OWLBgAYYNGwYzMzP4+/sDAGJiYtC+fXsoFArY29tj0qRJePLkiSZPhYjw5t/pjIwMDBkyBBYWFjA0NET37t2RlJQk7X/z5k306tULFhYWMDIygpubGw4cOIAbN27Ax8cHAGBhYQFBEDBs2DBNnCJVEIYbqtI2btwIHR0d/P777/j222+xbNkyrF+/Xlr/zTffoEmTJoiLi8OcOXNw4cIFdO3aFf3798dff/2FHTt24MSJE5gwYYIGz4KInnvd7/SwYcMQGxuLPXv24NSpUxBFET169MDTp08BAOPHj0dubi6OHTuGCxcu4KuvvoKxsTHs7e0RHh4OAEhMTERaWhpWrFihsXOk8sdJ/KjK6tixI9LT03Hx4kXp7bCffvop9uzZg/j4eDg6OsLT0xO7du2S9hkyZAgUCgXWrl0rtZ04cQIdOnTAkydPYGBgUOHnQUTPvO53+pdffoGTkxNOnjyJ1q1bAwD+/fdf2NvbY+PGjRgwYACaNm2K9957DwEBAUX6joqKgo+PDzIyMmBubl6Rp0UawCs3VKW1atVK6bX33t7eSEpKQkFBAQDAy8tLafu4uDiEhYXB2NhYWrp27YrCwkJcv369QmsnoqKK+52Oj4+Hjo4OWrZsKa2ztLRE48aNkZCQAACYNGkSFixYgDZt2iAgIAB//fVXhddPlQPDDcmakZGR0ufCwkKMGTMG58+fl5Y///wTSUlJaNCggYaqJCJViaIohaFRo0bh2rVr+Oijj3DhwgV4eXlh5cqVGq6QNIHhhqq006dPF/ncqFEjaGtrv3L75s2b4+LFi2jYsGGRRU9PryJKJqLXKO532tXVFfn5+fj999+ldf/++y8uX74MFxcXqc3e3h5jx47Fzp078cknn2DdunUAIP1+P7+qS/LGcENVWkpKCqZNm4bExERs27YNK1euxOTJk4vdftasWTh16hTGjx+P8+fPIykpCXv27MHEiRMrsGoiKk5xv9ONGjVCnz594O/vjxMnTuDPP//Ehx9+iNq1a6NPnz4AgClTpuDQoUO4fv06zp49i4iICCn4ODg4QBAE7Nu3D3fv3sXjx481eZpUznQ0XQBRWQwZMgTZ2dl4++23oa2tjYkTJ2L06NHFbt+0aVNER0fj888/R7t27SCKIho0aIBBgwZVYNVEVJzX/U5v2LABkydPxrvvvou8vDy0b98eBw4cgK6uLoBnV2XGjx+PW7duwdTUFN26dcOyZcsAALVr18a8efPw6aefYvjw4RgyZAjCwsI0dZpUzvi0FFVZHTt2hIeHB5YvX67pUohIDfg7TerC21JEREQkKww3REREJCu8LUVERESywis3REREJCsMN0RERCQrDDdEREQkKww3REREJCsMN0RERCQrDDdEVGECAwPh4eGh6TJKJCoqCoIg4MGDB5ouhYhKiY+CE5FaPH8zc3GGDh2KVatWITc3F5aWlhVUlery8vJw//59WFtbv/Hcnhs2bBgePHiA3bt3l29xRPRafLcUEalFWlqa9OcdO3Zg7ty5SExMlNoUCgWMjY1hbGysifJKTU9PDzY2Npoug4hUwNtSRKQWNjY20mJmZgZBEIq0vXxbatiwYejbty8WLlwIa2trmJubY968ecjPz8eMGTNQo0YN1KlTB99//73SsVJTUzFo0CBYWFjA0tISffr0wY0bN4r0O2/ePFhZWcHU1BRjxoxBXl6etE1ubi4mTZoEKysrGBgYoG3btjhz5oy0/uXbUmFhYTA3N8ehQ4fg4uICY2NjdOvWTQp1gYGB2LhxI3755RcIggBBEBAVFaX275mI3ozhhog0KiIiAv/88w+OHTuGpUuXIjAwEO+++y4sLCzw+++/Y+zYsRg7dixSUlIAAFlZWfDx8YGxsTGOHTuGEydOSEHjxfBy9OhRJCQkIDIyEtu2bcOuXbswb948af3MmTMRHh6OjRs34uzZs2jYsCG6du2K+/fvF1trVlYWFi9ejM2bN+PYsWNITk7G9OnTAQDTp0/HwIEDpcCTlpaG1q1bl9O3RkSvJRIRqdmGDRtEMzOzIu0BAQFis2bNpM9Dhw4VHRwcxIKCAqmtcePGYrt27aTP+fn5opGRkbht2zZRFEUxNDRUbNy4sVhYWChtk5ubKyoUCvHQoUNSvzVq1BCfPHkibbN69WrR2NhYLCgoEB8/fizq6uqKW7duldbn5eWJdnZ24tdffy2KoihGRkaKAMSMjAzpnACIV65ckfb573//K1pbWyudT58+fUrxTRFReeCYGyLSKDc3N2hp/f9FZGtrazRp0kT6rK2tDUtLS6SnpwMA4uLicOXKFZiYmCj1k5OTg6tXr0qfmzVrBkNDQ+mzt7c3Hj9+jJSUFDx8+BBPnz5FmzZtpPW6urp4++23kZCQUGythoaGaNCggfTZ1tZWqouIKg+GGyLSKF1dXaXPgiC8sq2wsBAAUFhYiBYtWmDr1q1F+qpVq9YbjycIAsT/PST68lNQoii+9smoV9Ul8oFTokqHY26IqEpp3rw5kpKSYGVlhYYNGyotZmZm0nZ//vknsrOzpc+nT5+GsbEx6tSpg4YNG0JPTw8nTpyQ1j99+hSxsbFwcXFRuTY9PT0UFBSovD8RqQfDDRFVKX5+fqhZsyb69OmD48eP4/r164iOjsbkyZNx69Ytabu8vDyMHDkS8fHxOHjwIAICAjBhwgRoaWnByMgIH3/8MWbMmIFff/0V8fHx8Pf3R1ZWFkaOHKlybY6Ojvjrr7+QmJiIe/fu4enTp+o4ZSIqJd6WIqIqxdDQEMeOHcOsWbPQv39/PHr0CLVr14avry9MTU2l7Xx9fdGoUSO0b98eubm5eP/99xEYGCitX7RoEQoLC/HRRx/h0aNH8PLywqFDh2BhYaFybf7+/oiKioKXlxceP36MyMhIdOzYsQxnS0Sq4AzFRCQ7nCmYqHrjbSkiIiKSFYYbIiIikhXeliIiIiJZ4ZUbIiIikhWGGyIiIpIVhhsiIiKSFYYbIiIikhWGGyIiIpIVhhsiIiKSFYYbIiIikhWGGyIiIpKV/wP4rPq490qWGwAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# 4 way paired boxplot trial\n",
    "g = 'saliva'\n",
    "df_alpha = g_to_dfd[g + '_adh']['alpha']\n",
    "\n",
    "df_map = type_to_df_map[g].copy()\n",
    "df_map = df_map.set_index('Together')\n",
    "df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "\n",
    "df_merge = pd.concat([df_alpha, df_map], axis=1)\n",
    "df_merge = df_merge.dropna()\n",
    "\n",
    "sns.boxplot(x='Timepoint', y='shannon_entropy',\n",
    "            hue='WOMAC_P_Response', palette=[\"r\", \"g\"],\n",
    "            data=df_merge)\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "021fdc95-b226-4456-a324-b0768af749ef",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
