{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3f62dc73-df18-4548-bc0d-aea2a125fef0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "import pandas as pd\n",
    "import scipy.stats\n",
    "import statsmodels.stats.multitest\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "\n",
    "# disable warnings, use w caution\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "# project specific libs\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import pathlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a8b42930-bcde-48f3-9753-8779e1aa11fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# project specific path\n",
    "path = '/Users/KevinBu/Desktop/clemente_lab/Projects/oa/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0343d6a9-1c06-4ed7-90a1-ed193002a1dd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Process Mapping Files ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a3013747-2e0e-412a-b719-b2a6a28afbf2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>Separate</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Together</th>\n",
       "      <th>ContactEmail</th>\n",
       "      <th>ContactName</th>\n",
       "      <th>PrimaryInvestigator</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>RawDataNotes</th>\n",
       "      <th>...</th>\n",
       "      <th>broccoli</th>\n",
       "      <th>Garbanzo_beans</th>\n",
       "      <th>pork</th>\n",
       "      <th>beef</th>\n",
       "      <th>burger</th>\n",
       "      <th>Total_omega3</th>\n",
       "      <th>Adherence_omega3</th>\n",
       "      <th>Total_omega6</th>\n",
       "      <th>Adherence_omega6</th>\n",
       "      <th>Total_o3_o6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>TTCAGTTCGTTA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-001.pre.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>CGGCCAGAAGCA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>post</td>\n",
       "      <td>OAD-001.post.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001.post.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>72.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>131.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>GACGTTAAGAAT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-003.pre.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>53.2</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>TCGCTACAGATG</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>post</td>\n",
       "      <td>OAD-003.post.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003.post.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>108.0</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>171.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>ATGGGACCTTCA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-004.pre.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-004.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>2.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>28.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>35.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 188 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   BarcodeSequence    LinkerPrimerSequence Separate Timepoint  \\\n",
       "#SampleID                                                                       \n",
       "OAD-001.pre.stool     TTCAGTTCGTTA  CCGGACTACHVGGGTWTCTAAT      All       pre   \n",
       "OAD-001.post.stool    CGGCCAGAAGCA  CCGGACTACHVGGGTWTCTAAT      All      post   \n",
       "OAD-003.pre.stool     GACGTTAAGAAT  CCGGACTACHVGGGTWTCTAAT      All       pre   \n",
       "OAD-003.post.stool    TCGCTACAGATG  CCGGACTACHVGGGTWTCTAAT      All      post   \n",
       "OAD-004.pre.stool     ATGGGACCTTCA  CCGGACTACHVGGGTWTCTAAT      All       pre   \n",
       "\n",
       "                                            Together  \\\n",
       "#SampleID                                              \n",
       "OAD-001.pre.stool    OAD-001.pre.stool.guma.plate313   \n",
       "OAD-001.post.stool  OAD-001.post.stool.guma.plate313   \n",
       "OAD-003.pre.stool    OAD-003.pre.stool.guma.plate313   \n",
       "OAD-003.post.stool  OAD-003.post.stool.guma.plate313   \n",
       "OAD-004.pre.stool    OAD-004.pre.stool.guma.plate313   \n",
       "\n",
       "                                    ContactEmail    ContactName  \\\n",
       "#SampleID                                                         \n",
       "OAD-001.pre.stool   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-001.post.stool  rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-003.pre.stool   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-003.post.stool  rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-004.pre.stool   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "\n",
       "                   PrimaryInvestigator Cohort  \\\n",
       "#SampleID                                       \n",
       "OAD-001.pre.stool           Jose Scher  NonVA   \n",
       "OAD-001.post.stool          Jose Scher  NonVA   \n",
       "OAD-003.pre.stool           Jose Scher  NonVA   \n",
       "OAD-003.post.stool          Jose Scher  NonVA   \n",
       "OAD-004.pre.stool           Jose Scher  NonVA   \n",
       "\n",
       "                                        RawDataNotes  ... broccoli  \\\n",
       "#SampleID                                             ...            \n",
       "OAD-001.pre.stool    OAD-001.pre.stool.guma.plate313  ...      0.0   \n",
       "OAD-001.post.stool  OAD-001.post.stool.guma.plate313  ...      4.0   \n",
       "OAD-003.pre.stool    OAD-003.pre.stool.guma.plate313  ...      0.0   \n",
       "OAD-003.post.stool  OAD-003.post.stool.guma.plate313  ...      4.0   \n",
       "OAD-004.pre.stool    OAD-004.pre.stool.guma.plate313  ...      2.8   \n",
       "\n",
       "                   Garbanzo_beans pork  beef burger Total_omega3  \\\n",
       "#SampleID                                                          \n",
       "OAD-001.pre.stool             0.0  0.0   0.0    0.0          NaN   \n",
       "OAD-001.post.stool            0.0  0.0   0.0    0.0         48.0   \n",
       "OAD-003.pre.stool             0.0  0.0  14.0    0.0         16.8   \n",
       "OAD-003.post.stool            0.0  0.0   0.0    0.0         48.0   \n",
       "OAD-004.pre.stool             0.0  0.0   0.0    0.0          2.8   \n",
       "\n",
       "                   Adherence_omega3 Total_omega6 Adherence_omega6 Total_o3_o6  \n",
       "#SampleID                                                                      \n",
       "OAD-001.pre.stool     Low adherence          NaN    Low adherence         0.0  \n",
       "OAD-001.post.stool    Low adherence         72.0    Low adherence       131.0  \n",
       "OAD-003.pre.stool     Low adherence         53.2   High adherence        75.0  \n",
       "OAD-003.post.stool    Low adherence        108.0   High adherence       171.0  \n",
       "OAD-004.pre.stool     Low adherence         28.0    Low adherence        35.8  \n",
       "\n",
       "[5 rows x 188 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# from AC Q2 run of merged saliva stool\n",
    "df_map = pd.read_csv(path + 'inputs/Qiime2_0/qiime_mapping_file.tsv', sep='\\t', index_col=0)\n",
    "q2_row = df_map.loc['#q2:types',:]\n",
    "df_map = df_map.drop('#q2:types')\n",
    "\n",
    "# change index so it matches metadata file\n",
    "df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "\n",
    "# drop MOC and elution buffer\n",
    "df_map = df_map.drop(['MOC.320','elutionbuffer.plate313'])\n",
    "\n",
    "# grab metadata\n",
    "df_meta = pd.read_csv(path + 'inputs/Metadata_OA.csv')\n",
    "\n",
    "# rename 'Run_ID_Saliva' to be correct\n",
    "df_meta['Timepoints'] = df_meta['Timepoints'].apply(lambda x: 'pre' if x == '0' else 'post')\n",
    "df_meta['Patient_ID'] = df_meta['Patient_ID'].apply(lambda x: x[:-3])  \n",
    "df_meta['Study_ID'] = df_meta['Study_ID'].apply(lambda x: x.split('_')[0][-3:]) \n",
    "df_outcomes = df_meta.copy()\n",
    "\n",
    "# create per sample type mapping files\n",
    "type_to_ST = {'saliva':'Saliva','stool':'fecal'}\n",
    "type_to_df_map = {}\n",
    "\n",
    "# split into specimen type\n",
    "for t in type_to_ST:\n",
    "    # subset on specimen type\n",
    "    df_map_type = df_map[df_map['SpecimenType'] == type_to_ST[t]]\n",
    "\n",
    "    # as to not overwrite df meta\n",
    "    df_meta_type = df_meta.copy()\n",
    "\n",
    "    # create new sample ID for specimen type and set as index\n",
    "    df_meta_type['#SampleID'] = df_meta['Patient_ID'] + '-' + df_meta['Study_ID'] + '.' + df_meta['Timepoints'] + '.' + t\n",
    "    df_meta_type = df_meta_type.set_index('#SampleID')\n",
    "\n",
    "    # create full mapping file\n",
    "    df_map_type = pd.concat([df_map_type, df_meta_type],axis=1)\n",
    "\n",
    "    # use only sequenced samples\n",
    "    df_map_type = df_map_type.dropna(how='any',subset='BarcodeSequence')\n",
    "\n",
    "    # drop all na columns\n",
    "    df_map_type = df_map_type.dropna(how='all',axis=1)\n",
    "\n",
    "    # drop VAD OA 015 because misdx with PsA not OA\n",
    "    if t == 'saliva':\n",
    "        df_map_type = df_map_type.drop(['VAOAD-015.pre.saliva','VAOAD-015.post.saliva'])\n",
    "    if t == 'stool':\n",
    "        df_map_type = df_map_type.drop(['VAOAD-015.pre.stool','VAOAD-015.post.stool'])\n",
    "\n",
    "    # populate dict of mapping files\n",
    "    type_to_df_map[t] = df_map_type\n",
    "\n",
    "    # export for q2\n",
    "    df_q2_type = pd.concat([q2_row.to_frame().T, df_map_type])\n",
    "    df_q2_type.index.name = '#SampleID'\n",
    "    df_q2_type.iloc[0,:] = 'categorical'\n",
    "    df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + t + '.tsv', sep='\\t')\n",
    "    df_q2_type = df_q2_type[df_q2_type['Adherece_antiinflam'].isin(['Moderate adherence', 'High adherence','categorical'])]\n",
    "    df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh.tsv', sep='\\t')\n",
    "    df_q2_type = df_q2_type[df_q2_type['Adherece_antiinflam'].isin(['Moderate adherence', 'High adherence','categorical'])]\n",
    "    df_q2_resp = df_q2_type[df_q2_type['WOMAC_P_Response'].isin(['categorical', 'Response'])]\n",
    "    df_q2_resp.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_response.tsv', sep='\\t')\n",
    "    df_q2_nonresp = df_q2_type[df_q2_type['WOMAC_P_Response'].isin(['categorical', 'No response'])]\n",
    "    df_q2_nonresp.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_noresponse.tsv', sep='\\t')\n",
    "\n",
    "type_to_df_map['stool'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "248b05fc-430d-4a56-9d93-bfc3334b840a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Assess Outcomes ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "95e4ef90-0416-4803-a5b2-d730031630c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n",
      "['VAS_Pt', 'VAS_overall', 'WOMAC_pain', 'WOMAC_stiffness', 'WOMAC_activity', 'WOMAC_total', 'Pain_DETECT', 'CES_D', 'Helplesness', 'Magnification', 'Rumination', 'PCS_EN', 'Sleep_distrubance', 'PASE_walk', 'PASE_light', 'BMI']\n",
      "32\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var</th>\n",
       "      <th>effect</th>\n",
       "      <th>pval</th>\n",
       "      <th>stat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VAS_Pt</td>\n",
       "      <td>198.5</td>\n",
       "      <td>0.008316</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VAS_overall</td>\n",
       "      <td>168.5</td>\n",
       "      <td>0.131314</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_pain</td>\n",
       "      <td>170.0</td>\n",
       "      <td>0.017900</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_stiffness</td>\n",
       "      <td>172.5</td>\n",
       "      <td>0.013052</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_activity</td>\n",
       "      <td>164.5</td>\n",
       "      <td>0.032592</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               var  effect      pval stat\n",
       "0           VAS_Pt   198.5  0.008316  MWU\n",
       "0      VAS_overall   168.5  0.131314  MWU\n",
       "0       WOMAC_pain   170.0  0.017900  MWU\n",
       "0  WOMAC_stiffness   172.5  0.013052  MWU\n",
       "0   WOMAC_activity   164.5  0.032592  MWU"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###\n",
    "# Hypothesis 1: There will be a measurable difference in WOMAC pain response scores and \n",
    "# other outcomes from baseline to after the dietary intervention.\n",
    "###\n",
    "\n",
    "# outcome variables\n",
    "outcomes = ['VAS_Pt', 'VAS_overall', 'WOMAC_pain', 'WOMAC_stiffness', 'WOMAC_activity', 'WOMAC_total', 'Pain_DETECT', \n",
    "            'CES_D', 'Helplesness', 'Magnification', 'Rumination', 'PCS_EN', 'Sleep_distrubance', 'PASE_walk', 'PASE_light', \n",
    "            #'PASE_gardening', # Where did this go? gardening_improve is binary\n",
    "            'BMI']\n",
    "\n",
    "\n",
    "# hypothesis 1\n",
    "# create a new df_meta\n",
    "df_meta = pd.read_csv(path + 'inputs/Metadata_OA.csv')\n",
    "\n",
    "# rename 'Run_ID_Saliva' to be correct\n",
    "df_meta['Timepoints'] = df_meta['Timepoints'].apply(lambda x: 'pre' if x == '0' else 'post')\n",
    "df_meta['Patient_ID'] = df_meta['Patient_ID'].apply(lambda x: x[:-3])  \n",
    "df_meta['Study_ID'] = df_meta['Study_ID'].apply(lambda x: x.split('_')[0][-3:]) \n",
    "df_meta['#SampleID'] = df_meta['Patient_ID'] + '-' + df_meta['Study_ID'] + '.' + df_meta['Timepoints'] + '.stool'\n",
    "df_meta = df_meta.set_index('#SampleID')\n",
    "\n",
    "# convert % to floats for calculations down the road\n",
    "bin = []\n",
    "cont = []\n",
    "df_paired_os = []\n",
    "for w in outcomes:\n",
    "    df_w = df_meta[w]\n",
    "    if df_w.nunique() > 2: # do spearman\n",
    "        df_meta[w] = df_meta[w].astype(str).str.replace('%','').astype(float).values\n",
    "        cont.append(w)\n",
    "    else:\n",
    "        bin.append(w)\n",
    "\n",
    "    # compute difference and store it\n",
    "    df_md = df_meta.copy()\n",
    "    df_md['SubjectID'] = df_md['Patient_ID'] + df_md['Study_ID']\n",
    "    \n",
    "    # first drop unpaired samples\n",
    "    s_remove = []\n",
    "    for s in list(df_md['SubjectID'].values):\n",
    "        if len(df_md[df_md['SubjectID'] == s]) != 2:\n",
    "            s_remove.append(s)\n",
    "    df_md = df_md.loc[~df_md['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "    \n",
    "    # set vars\n",
    "    group_var = 'Timepoints'\n",
    "    pair_var = 'SubjectID'\n",
    "    groups = ['pre','post']\n",
    "    \n",
    "    # get paired per indiv pair\n",
    "    pair_to_diff = {}\n",
    "    for p in list(df_md[pair_var].values):\n",
    "        df = df_md[df_md[pair_var] == p]\n",
    "        t0 = float(df[df[group_var] == groups[0]][w].values)\n",
    "        tf = float(df[df[group_var] == groups[1]][w].values)\n",
    "        pair_to_diff[p] = tf - t0\n",
    "    \n",
    "    df_paired_o = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[w + '_diff'])\n",
    "    df_paired_os.append(df_paired_o)\n",
    "\n",
    "df_meta_paired = pd.concat([*df_paired_os], axis=1)    \n",
    "\n",
    "print(bin)\n",
    "print(cont)\n",
    "\n",
    "# split into all and mod high only\n",
    "a = 'modhigh'\n",
    "job = 'jobs02'\n",
    "df_meta = df_meta[df_meta['Adherece_antiinflam'].isin(['Moderate adherence', 'High adherence'])]\n",
    "        \n",
    "print(len(df_meta))\n",
    "\n",
    "df_results = pd.DataFrame(columns=['var','effect','pval','stat'])\n",
    "# do post treatment vals of binary vars differ from pre treatment 'unpaired'\n",
    "for b in bin:\n",
    "    ct_table_ind=pd.crosstab(df_meta[\"Timepoints\"],df_meta[b])\n",
    "    chi2_stat, p, dof, expected = scipy.stats.chi2_contingency(ct_table_ind)\n",
    "    row=pd.DataFrame.from_dict({'var': [b],'effect':[chi2_stat],'pval':[p],'stat':['chi2']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "# fishers exact\n",
    "for b in bin:\n",
    "    ct_table_ind=pd.crosstab(df_meta[\"Timepoints\"],df_meta[b])\n",
    "    fisher, p = scipy.stats.fisher_exact(ct_table_ind)\n",
    "    row=pd.DataFrame.from_dict({'var': [b],'effect':[t],'pval':[p],'stat':['fisher']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "    \n",
    "# do post treatment vals of continuous vars differ from pre treatment unpaired\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    try:\n",
    "        W,p = scipy.stats.mannwhitneyu(x=df_pre[c].values,y=df_post[c].values, nan_policy='omit')\n",
    "    except:\n",
    "        W,p = 0, 1\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['mwu']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v')\n",
    "        sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette='dark:grey', hue=None, orient='v')\n",
    "    \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/mwu_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['MWU']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "    \n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    t,p = scipy.stats.ttest_ind(a=df_pre[c].values,b=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['ttest']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v')\n",
    "        sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette='dark:grey', hue=None, orient='v')\n",
    "    \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/tt_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['ttest']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "# unpaired and then paired\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    W,p = scipy.stats.wilcoxon(x=df_pre[c].values,y=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['WSR']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.swarmplot(data=df_meta, x='Timepoints', y=c, hue='WOMAC_P_Response', palette=['b','r'], orient='v')\n",
    "        for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "            ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "        sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v', hue=None, boxprops=dict(alpha=.3))\n",
    "        \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/wsr_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['WSR']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    t,p = scipy.stats.ttest_rel(a=df_pre[c].values,b=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['pairedt']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette=['b','r'], hue='WOMAC_P_Response',orient='v')\n",
    "        for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "            ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "        sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v', hue=None, boxprops=dict(alpha=0.3))\n",
    "        \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/pairedt_' + c  + '.pdf')\n",
    "        plt.close()          \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['pairedt']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "df_results.to_csv(path + 'outputs/' + job + '/outcome_testing.tsv', sep='\\t')\n",
    "df_results.head()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0379f4d-6b07-489a-bfae-75b6085b9a15",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Grab Paired Alpha ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "298b4824-3c24-485e-8705-5daf4a85d6d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unweighted_Unifrac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.6672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>0.2837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>0.4237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>0.3257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>0.1022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unweighted_Unifrac\n",
       "OAD001              0.6672\n",
       "OAD003              0.2837\n",
       "OAD004              0.4237\n",
       "OAD005              0.3257\n",
       "OAD006              0.1022"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###\n",
    "# Hypothesis 2: There will be an association between oral and gut microbiome and pain outcomes\n",
    "###\n",
    "# construct alpha, beta and paired alpha dataframes\n",
    "g_to_dfd = {}\n",
    "# g_test = ['stool','saliva_adh', 'stool_adh', 'saliva']\n",
    "g_test = ['saliva_adh', 'stool_adh']\n",
    "\n",
    "for g in g_test:\n",
    "    # maps diversity type to dataframe\n",
    "    g_to_dfd[g] = {}\n",
    "\n",
    "    # get alpha diversities\n",
    "    df_alpha = pd.read_csv(path + 'outputs/Qiime2_' + g + '/metadata.tsv', sep='\\t', index_col=0)\n",
    "    df_alpha = df_alpha.drop('#q2:types')\n",
    "    df_alpha['SubjectID'] = df_alpha['Patient_ID'] + df_alpha['Study_ID']\n",
    "    df_alpha = df_alpha[['SubjectID', 'Timepoints', 'shannon_entropy']]\n",
    "    g_to_dfd[g]['alpha'] = df_alpha\n",
    "\n",
    "    # get paired alpha div, first drop unpaired samples\n",
    "    s_remove = []\n",
    "    for s in list(df_alpha['SubjectID'].values):\n",
    "        if len(df_alpha[df_alpha['SubjectID'] == s]) != 2:\n",
    "            s_remove.append(s)\n",
    "    df_alpha = df_alpha.loc[~df_alpha['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "    \n",
    "    # set vars\n",
    "    alpha_metric = 'shannon_entropy'\n",
    "    group_var = 'Timepoints'\n",
    "    pair_var = 'SubjectID'\n",
    "    groups = ['pre','post']\n",
    "    \n",
    "    # get paired per indiv pair\n",
    "    pair_to_diff = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        df = df_alpha[df_alpha[pair_var] == p]\n",
    "        alpha_0 = float(df[df[group_var] == groups[0]][alpha_metric].values)\n",
    "        alpha_1 = float(df[df[group_var] == groups[1]][alpha_metric].values)\n",
    "        pair_to_diff[p] = alpha_1 - alpha_0\n",
    "    \n",
    "    df_paired_alpha = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[alpha_metric + '_diff'])\n",
    "    g_to_dfd[g]['paired_alpha'] = df_paired_alpha\n",
    "\n",
    "    # get beta div\n",
    "    df_beta = pd.read_csv(path + 'outputs/Qiime2_' + g + '/core_metrics_results/distance-matrix.tsv',\n",
    "                              sep='\\t', index_col=0)\n",
    "        \n",
    "    # grab twin to pair dict\n",
    "    pair_to_ids = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        df = df_alpha[df_alpha[pair_var] == p]\n",
    "        id_0 = str(df[df[group_var] == groups[0]].index.values[0])\n",
    "        id_1 = str(df[df[group_var] == groups[1]].index.values[0])\n",
    "        pair_to_ids[p] = (id_0, id_1)\n",
    "    \n",
    "    # get distances for each twin pair per beta div matrix    \n",
    "    pair_to_dist = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        id_0, id_1 = pair_to_ids[p]\n",
    "        pair_to_dist[p] = df_beta.loc[id_0, id_1]\n",
    "    \n",
    "    df_paired_beta = pd.DataFrame.from_dict(pair_to_dist, orient='index', columns=['Unweighted_Unifrac'])\n",
    "    g_to_dfd[g]['paired_beta'] = df_paired_beta\n",
    "\n",
    "g_to_dfd['stool_adh']['paired_beta'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e018facb-88c3-4c10-be22-dbe70ee41351",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Chi2 Association of Outcome with Alpha ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9d4f38ff-41d6-4f31-94ab-69879de56109",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>div</th>\n",
       "      <th>outcome</th>\n",
       "      <th>statistic</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.232031</td>\n",
       "      <td>0.630022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>1.153857</td>\n",
       "      <td>0.788413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.756449</td>\n",
       "      <td>0.456450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.232031</td>\n",
       "      <td>0.630022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.756449</td>\n",
       "      <td>0.980536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.182238</td>\n",
       "      <td>0.856866</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        group                  div           outcome statistic  test_stat  \\\n",
       "0    metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.232031   \n",
       "1  saliva_adh                alpha  WOMAC_P_Response       mwu   1.153857   \n",
       "2  saliva_adh                alpha  WOMAC_P_Response        tt   0.756449   \n",
       "3    metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.232031   \n",
       "4   stool_adh                alpha  WOMAC_P_Response       mwu   0.756449   \n",
       "5   stool_adh                alpha  WOMAC_P_Response        tt   0.182238   \n",
       "\n",
       "       pval  \n",
       "0  0.630022  \n",
       "1  0.788413  \n",
       "2  0.456450  \n",
       "3  0.630022  \n",
       "4  0.980536  \n",
       "5  0.856866  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Testing for each sample type for (1) all and (2) high adh only\n",
    "# (A) Chisq of quartiles with adherence \n",
    "# (B) MWU/TT unpaired of outcomes against distance\n",
    "\n",
    "def chisq_of_df_cols(df, c1, c2):\n",
    "    groupsizes = df.groupby([c1, c2]).size()\n",
    "    ctsum = groupsizes.unstack(c1)\n",
    "    # fillna(0) is necessary to remove any NAs which will cause exceptions\n",
    "    return(scipy.stats.chi2_contingency(ctsum.fillna(0)))\n",
    "\n",
    "d_to_metric = {\n",
    "    'alpha': 'shannon_entropy',\n",
    "}\n",
    "group_var = 'Adherece_antiinflam'\n",
    "\n",
    "dfd_to_merge = {}\n",
    "g_test = ['saliva_adh', 'stool_adh']\n",
    "\n",
    "for g in g_test:\n",
    "    dfd_to_merge[g] = {}\n",
    "\n",
    "gs = []\n",
    "ds = []\n",
    "os = []\n",
    "stats = []\n",
    "ts = []\n",
    "ps = []\n",
    "\n",
    "arr = [gs,ds,os,stats,ts,ps]\n",
    "\n",
    "def append_results(arr, val):\n",
    "    for a,v in zip(arr,val):\n",
    "        a.append(v)\n",
    "    return arr\n",
    "\n",
    "# for each sample type, grab relevant mapping file\n",
    "# g_test = saliva, saliva_adh, stool, etc.\n",
    "for g in g_test:\n",
    "    for d in d_to_metric:\n",
    "        # grab relevant diversity df\n",
    "        df_div = g_to_dfd[g][d]\n",
    "\n",
    "        if d == 'alpha':\n",
    "            # df_div = df_div[df_div['Timepoints'] == 'pre']\n",
    "            # df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            # df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div['shannon_entropy'] = df_div['shannon_entropy'].astype(float)\n",
    "\n",
    "        # merge with df of metadata var        \n",
    "        # grab relevant sample IDs\n",
    "        # g = saliva_adh\n",
    "        idx = 'Run_ID_' + g.split('_')[0].capitalize()\n",
    "        df_meta_sub = df_meta.dropna(subset=idx)\n",
    "        df_meta_sub = df_meta_sub.set_index(idx)\n",
    "        df_merge = pd.concat([df_meta_sub,df_div],axis=1)\n",
    "\n",
    "        # subset on adh only\n",
    "        df_merge = df_merge[df_merge[group_var].isin(['Moderate adherence', 'High adherence'])]\n",
    "\n",
    "        # test association of div with outcomes\n",
    "        for o in ['WOMAC_P_Response']: #  outcomes:                    \n",
    "            # test association of adherence with pain outcomes\n",
    "            div_metric = d_to_metric[d]\n",
    "            x,p,dof,ef = chisq_of_df_cols(df_merge, group_var, o)\n",
    "            arr = append_results(arr, ['metadata',group_var,o,'chisq',x,p])\n",
    "            \n",
    "            ax = sns.boxplot(data=df_merge, x=o, y=div_metric)\n",
    "            sns.swarmplot(data=df_merge, x=o, y=div_metric, palette='dark:grey')\n",
    "            sns.despine()\n",
    "        \n",
    "            plt.tight_layout()\n",
    "            plt.savefig(path + 'outputs/jobs06/rvnr_nondiff_' + g + '_' + o + '_' + d + '.pdf')\n",
    "            plt.close()          \n",
    "        \n",
    "            u, p = scipy.stats.mannwhitneyu(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'mwu',t,p])\n",
    "\n",
    "            t, p = scipy.stats.ttest_ind(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'tt',t,p])\n",
    "\n",
    "        # save results\n",
    "        dfd_to_merge[g][d] = df_merge\n",
    "\n",
    "        # export to Q2\n",
    "        # df_q2_type = df_merge.set_index(['Together'])\n",
    "        df_q2_type = df_merge.copy()\n",
    "        q2_row = pd.Series(data=['categorical' for i in range(len(df_merge.columns))], \n",
    "                           index=list(df_merge.columns.values), dtype=str, name='#q2:types')\n",
    "        df_q2_type = pd.concat([q2_row.to_frame().T, df_q2_type])\n",
    "        df_q2_type.index.name = '#SampleID'\n",
    "        df_q2_type.index = df_q2_type.index.map(lambda x: x.split('.guma')[0])\n",
    "        # df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + d + '_' + g + 'aggregate_outcomes.tsv', sep='\\t')\n",
    "\n",
    "df_results = pd.DataFrame.from_dict({\n",
    "    'group': gs,\n",
    "    'div': ds,\n",
    "    'outcome': os,\n",
    "    'statistic': stats,\n",
    "    'test_stat': ts,\n",
    "    'pval': ps\n",
    "})\n",
    "# df_results.to_csv(path + 'outputs/df_results_aggregate.tsv', sep='\\t')\n",
    "# df_results.head()\n",
    "# df_results[df_results['pval'] < 0.05]\n",
    "df_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "22f26972-623f-459c-93fd-d37eb65d17eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>div</th>\n",
       "      <th>outcome</th>\n",
       "      <th>statistic</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.182238</td>\n",
       "      <td>0.343434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>-0.594256</td>\n",
       "      <td>0.565542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>-0.594256</td>\n",
       "      <td>0.638889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.769556</td>\n",
       "      <td>0.459355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.769556</td>\n",
       "      <td>0.431818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>1.289488</td>\n",
       "      <td>0.226254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>1.289488</td>\n",
       "      <td>0.876263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>-1.033580</td>\n",
       "      <td>0.325681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>-1.033580</td>\n",
       "      <td>0.755051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.565329</td>\n",
       "      <td>0.584310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.565329</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.682542</td>\n",
       "      <td>0.510397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.682542</td>\n",
       "      <td>0.876263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.155526</td>\n",
       "      <td>0.879502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.155526</td>\n",
       "      <td>0.267677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>1.182922</td>\n",
       "      <td>0.264200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         group           div           outcome statistic  test_stat      pval\n",
       "0    stool_adh     pre_alpha  WOMAC_P_Response       mwu   0.182238  0.343434\n",
       "1    stool_adh     pre_alpha  WOMAC_P_Response        tt  -0.594256  0.565542\n",
       "2    stool_adh    post_alpha  WOMAC_P_Response       mwu  -0.594256  0.638889\n",
       "3    stool_adh    post_alpha  WOMAC_P_Response        tt   0.769556  0.459355\n",
       "4    stool_adh  paired_alpha  WOMAC_P_Response       mwu   0.769556  0.431818\n",
       "5    stool_adh  paired_alpha  WOMAC_P_Response        tt   1.289488  0.226254\n",
       "6    stool_adh   paired_beta  WOMAC_P_Response       mwu   1.289488  0.876263\n",
       "7    stool_adh   paired_beta  WOMAC_P_Response        tt  -1.033580  0.325681\n",
       "8   saliva_adh     pre_alpha  WOMAC_P_Response       mwu  -1.033580  0.755051\n",
       "9   saliva_adh     pre_alpha  WOMAC_P_Response        tt   0.565329  0.584310\n",
       "10  saliva_adh    post_alpha  WOMAC_P_Response       mwu   0.565329  1.000000\n",
       "11  saliva_adh    post_alpha  WOMAC_P_Response        tt   0.682542  0.510397\n",
       "12  saliva_adh  paired_alpha  WOMAC_P_Response       mwu   0.682542  0.876263\n",
       "13  saliva_adh  paired_alpha  WOMAC_P_Response        tt   0.155526  0.879502\n",
       "14  saliva_adh   paired_beta  WOMAC_P_Response       mwu   0.155526  0.267677\n",
       "15  saliva_adh   paired_beta  WOMAC_P_Response        tt   1.182922  0.264200"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# expand previous to all data, paired and unpaired diversities\n",
    "d_to_metric = {\n",
    "    'pre_alpha': 'shannon_entropy',\n",
    "    'post_alpha': 'shannon_entropy',\n",
    "    'paired_alpha': 'shannon_entropy_diff',\n",
    "    'paired_beta': 'Unweighted_Unifrac',\n",
    "}\n",
    "group_var = 'Adherece_antiinflam'\n",
    "\n",
    "dfd_to_merge = {}\n",
    "g_test = ['stool_adh','saliva_adh']\n",
    "for g in g_test:\n",
    "    dfd_to_merge[g] = {}\n",
    "\n",
    "gs = []\n",
    "ds = []\n",
    "os = []\n",
    "stats = []\n",
    "ts = []\n",
    "ps = []\n",
    "\n",
    "arr = [gs,ds,os,stats,ts,ps]\n",
    "\n",
    "def append_results(arr, val):\n",
    "    for a,v in zip(arr,val):\n",
    "        a.append(v)\n",
    "    return arr\n",
    "\n",
    "# for each sample type, grab relevant mapping file\n",
    "# g_test = saliva, saliva_adh, stool, etc.\n",
    "d_to_map = {}\n",
    "\n",
    "for g in g_test:\n",
    "    # drop duplicates so you have sample mapping to adh\n",
    "    df_map_sub = type_to_df_map[g.split('_')[0]]\n",
    "    df_map_sub.index = df_map_sub.index.map(lambda x: x.split('.')[0].replace('-',''))\n",
    "    df_map_sub = df_map_sub.dropna(how='any',subset=group_var,axis=0)           \n",
    "    \n",
    "    # figure out which samples to keep\n",
    "    # i.e. samples that have a pre and post time point\n",
    "    keep = []\n",
    "    for i in list(df_map_sub.index.values):\n",
    "        if len(df_map_sub.loc[i,:]) == 2:\n",
    "            keep.append(i)\n",
    "\n",
    "    # get unique entires in sorted order\n",
    "    # at this point we are only concerned with differences in values, \n",
    "    # as we've dropped samples with only one endpoint val\n",
    "    save = []\n",
    "    [save.append(x) for x in keep if x not in save]\n",
    "    df_map_sub = df_map_sub.loc[save,:]\n",
    "\n",
    "    # this double populates as OAD001 is an index twice, so the diff fills to both the pre and post col\n",
    "    for o in outcomes:\n",
    "        df_map_sub[o + '_diff'] = df_map_sub[df_map_sub['Timepoint'] == 'post'][o] - df_map_sub[df_map_sub['Timepoint'] == 'pre'][o] \n",
    "\n",
    "    # here we keep only the pre, but everything is identical b/w pre and post\n",
    "    df_dropdup = df_map_sub[~df_map_sub.index.duplicated(keep='first')]\n",
    "    d_to_map[g] = df_dropdup\n",
    "    \n",
    "    for d in d_to_metric:\n",
    "        # grab relevant diversity df\n",
    "        if d == 'pre_alpha' or d == 'post_alpha':\n",
    "            df_div = g_to_dfd[g]['alpha']\n",
    "        else:\n",
    "            df_div = g_to_dfd[g][d]\n",
    "        div_metric = d_to_metric[d]\n",
    "\n",
    "        # when associating alpha div vs outcomes, look at if starting adiv predicts outcome\n",
    "        if d == 'pre_alpha':\n",
    "            df_div = df_div[df_div['Timepoints'] == 'pre']\n",
    "            df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div = df_div.astype(float)\n",
    "\n",
    "        if d == 'post_alpha':\n",
    "            df_div = df_div[df_div['Timepoints'] == 'post']\n",
    "            df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div = df_div.astype(float)\n",
    "\n",
    "        # merge with df of metadata var        \n",
    "        df_merge = pd.concat([df_dropdup,df_div],axis=1)\n",
    "\n",
    "        # drop na in barcodes if samples not sequenced both pre and post\n",
    "        df_merge = df_merge.dropna(how='any',subset='BarcodeSequence')\n",
    "        \n",
    "        # test association of div with outcomes\n",
    "        for o in ['WOMAC_P_Response']:                     \n",
    "            # test association of adherence with pain outcomes\n",
    "            # flip x and y for horizontal\n",
    "            ax = sns.boxplot(data=df_merge, x=o, y=div_metric)\n",
    "            sns.swarmplot(data=df_merge, x=o, y=div_metric, palette='dark:grey')\n",
    "            sns.despine()\n",
    "        \n",
    "            plt.tight_layout()\n",
    "            plt.savefig(path + 'outputs/jobs06/WOMAC_barplot_' + g + '_' + o + '_' + d + '.pdf')\n",
    "            plt.close()          \n",
    "\n",
    "            # MWU and ttest\n",
    "            u, p = scipy.stats.mannwhitneyu(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'mwu',t,p])\n",
    "\n",
    "            t, p = scipy.stats.ttest_ind(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'tt',t,p])\n",
    "         \n",
    "        # save results\n",
    "        dfd_to_merge[g][d] = df_merge\n",
    "\n",
    "        # export to Q2\n",
    "        df_q2_type = df_merge.set_index(['Together'])\n",
    "        q2_row = pd.Series(data=['categorical' for i in range(len(df_merge.columns))], \n",
    "                           index=list(df_merge.columns.values), dtype=str, name='#q2:types')\n",
    "        df_q2_type = pd.concat([q2_row.to_frame().T, df_q2_type])\n",
    "        df_q2_type.index.name = '#SampleID'\n",
    "        df_q2_type.index = df_q2_type.index.map(lambda x: x.split('.guma')[0])\n",
    "        # df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + d + '_' + g + '_outcomes.tsv', sep='\\t')\n",
    "\n",
    "df_results = pd.DataFrame.from_dict({\n",
    "    'group': gs,\n",
    "    'div': ds,\n",
    "    'outcome': os,\n",
    "    'statistic': stats,\n",
    "    'test_stat': ts,\n",
    "    'pval': ps\n",
    "})\n",
    "# df_results.to_csv(path + 'outputs/df_results_diff.tsv', sep='\\t')\n",
    "\n",
    "# df_results.head()\n",
    "# the div==alpha results test whether pre-alpha div state associates with pain outcome changes (differences) quartiles\n",
    "# the div==paired_alpha and paired_beta test whether the alpha and betas change in a similar way with the pain outcome\n",
    "print(len(df_results[df_results['pval'] < 0.05]))\n",
    "df_results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5db737b8-b788-4d7a-86a6-39e82ac57eb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_adh\n",
      "['Response', 'No response']\n",
      "WilcoxonResult(statistic=25.0, pvalue=0.30126953125)\n",
      "['Response']\n",
      "WilcoxonResult(statistic=4.0, pvalue=0.109375)\n",
      "['No response']\n",
      "WilcoxonResult(statistic=7.0, pvalue=1.0)\n",
      "saliva_adh\n",
      "['Response', 'No response']\n",
      "WilcoxonResult(statistic=24.0, pvalue=0.26611328125)\n",
      "['Response']\n",
      "WilcoxonResult(statistic=9.0, pvalue=0.46875)\n",
      "['No response']\n",
      "WilcoxonResult(statistic=5.0, pvalue=0.625)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>shannon_entropy_diff</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>Separate</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Together</th>\n",
       "      <th>ContactEmail</th>\n",
       "      <th>ContactName</th>\n",
       "      <th>PrimaryInvestigator</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>...</th>\n",
       "      <th>Pain_DETECT_diff</th>\n",
       "      <th>CES_D_diff</th>\n",
       "      <th>Helplesness_diff</th>\n",
       "      <th>Magnification_diff</th>\n",
       "      <th>Rumination_diff</th>\n",
       "      <th>PCS_EN_diff</th>\n",
       "      <th>Sleep_distrubance_diff</th>\n",
       "      <th>PASE_walk_diff</th>\n",
       "      <th>PASE_light_diff</th>\n",
       "      <th>BMI_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>-0.116436</td>\n",
       "      <td>AGGACAAACTAT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>Text</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-001.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>5.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.51</td>\n",
       "      <td>14.29</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>-0.327749</td>\n",
       "      <td>TCTTCCTAAAGT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-003.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-5.5</td>\n",
       "      <td>-14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>0.866474</td>\n",
       "      <td>CTTGTGCGACAA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-004.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.264032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>0.007278</td>\n",
       "      <td>CTTACACTGCTT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-001.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>5.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.081446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD012</th>\n",
       "      <td>0.535614</td>\n",
       "      <td>GCGTGGTCATTA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-012.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-1.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.72</td>\n",
       "      <td>-1.072664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TTGGACGTCCAC</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-009.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.139974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows Ã— 205 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          shannon_entropy_diff BarcodeSequence    LinkerPrimerSequence  \\\n",
       "OAD001               -0.116436    AGGACAAACTAT  CCGGACTACHVGGGTWTCTAAT   \n",
       "OAD003               -0.327749    TCTTCCTAAAGT  CCGGACTACHVGGGTWTCTAAT   \n",
       "OAD004                0.866474    CTTGTGCGACAA  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD001              0.007278    CTTACACTGCTT  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD012              0.535614    GCGTGGTCATTA  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD009                   NaN    TTGGACGTCCAC  CCGGACTACHVGGGTWTCTAAT   \n",
       "\n",
       "         Separate Timepoint              Together  \\\n",
       "OAD001       Text       pre    OAD-001.pre.saliva   \n",
       "OAD003        All       pre    OAD-003.pre.saliva   \n",
       "OAD004        All       pre    OAD-004.pre.saliva   \n",
       "VAOAD001      All       pre  VAOAD-001.pre.saliva   \n",
       "VAOAD012      All       pre  VAOAD-012.pre.saliva   \n",
       "VAOAD009      All       pre  VAOAD-009.pre.saliva   \n",
       "\n",
       "                          ContactEmail    ContactName PrimaryInvestigator  \\\n",
       "OAD001    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "OAD003    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "OAD004    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD001  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD012  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD009  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "\n",
       "         Cohort  ... Pain_DETECT_diff CES_D_diff Helplesness_diff  \\\n",
       "OAD001    NonVA  ...             -2.0       -1.0             -2.5   \n",
       "OAD003    NonVA  ...             -4.5        1.0             -1.0   \n",
       "OAD004    NonVA  ...             -1.0        3.0             -1.0   \n",
       "VAOAD001     VA  ...             -0.5        5.5              0.0   \n",
       "VAOAD012     VA  ...             -1.5        0.0              2.0   \n",
       "VAOAD009     VA  ...             -4.0        2.0              1.0   \n",
       "\n",
       "         Magnification_diff Rumination_diff PCS_EN_diff  \\\n",
       "OAD001                  5.5             1.0         4.0   \n",
       "OAD003                 -1.0            -2.0        -5.5   \n",
       "OAD004                 -1.0            -1.0        -3.0   \n",
       "VAOAD001                0.0             0.0         0.0   \n",
       "VAOAD012               -2.0            -0.5        -0.5   \n",
       "VAOAD009                0.0             0.5         1.5   \n",
       "\n",
       "         Sleep_distrubance_diff PASE_walk_diff PASE_light_diff  BMI_diff  \n",
       "OAD001                      9.0          11.51           14.29       NaN  \n",
       "OAD003                    -14.0           0.00            0.00       NaN  \n",
       "OAD004                      0.0           0.00            0.00  3.264032  \n",
       "VAOAD001                   -2.5           0.00            0.00 -0.081446  \n",
       "VAOAD012                   -1.5           0.00            6.72 -1.072664  \n",
       "VAOAD009                    2.5           0.00            0.00 -0.139974  \n",
       "\n",
       "[6 rows x 205 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test if the differences in diversity are significantly different from 0 for paired alpha\n",
    "for g in g_test:\n",
    "    print(g)\n",
    "    for d in ['paired_alpha']:\n",
    "        df_div = g_to_dfd[g][d] \n",
    "        df_map = d_to_map[g]\n",
    "\n",
    "        df_merge = pd.concat([df_div, df_map], axis=1)\n",
    "        for o in [['Response','No response'],['Response'], ['No response']]:\n",
    "            df = df_merge[df_merge['WOMAC_P_Response'].isin(o)]\n",
    "            print(o)\n",
    "            print(scipy.stats.wilcoxon(df['shannon_entropy_diff'], nan_policy='omit'))\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bef8c05d-208b-4b13-bac7-01bf00e4e80e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_adh\n",
      "All\n",
      "paired alpha wilcoxon\n",
      "33.0 0.414306640625\n",
      "Response\n",
      "paired alpha wilcoxon\n",
      "4.0 0.109375\n",
      "No response\n",
      "paired alpha wilcoxon\n",
      "7.0 1.0\n",
      "saliva_adh\n",
      "All\n",
      "paired alpha wilcoxon\n",
      "26.0 0.19091796875\n",
      "Response\n",
      "paired alpha wilcoxon\n",
      "9.0 0.46875\n",
      "No response\n",
      "paired alpha wilcoxon\n",
      "5.0 0.625\n"
     ]
    }
   ],
   "source": [
    "# Alpha div paired\n",
    "# paired beta, comparing intra-indiv difference pre_post to inter pre and inter post\n",
    "g_test = ['stool_adh','saliva_adh']\n",
    "subgroups = ['All', 'Response', 'No response']\n",
    "\n",
    "for g in g_test: #subgroups:\n",
    "    print(g)\n",
    "    df_a = pd.read_csv(path + 'outputs/Qiime2_' + g + '/metadata.tsv', sep='\\t', index_col=0)\n",
    "    df_a = df_a.drop('#q2:types')\n",
    "\n",
    "    for sg in subgroups:\n",
    "        print(sg)\n",
    "        if sg == 'All':\n",
    "            df_alpha = df_a.copy()\n",
    "        else:\n",
    "            df_alpha = df_a[df_a['WOMAC_P_Response'] == sg]\n",
    "        \n",
    "        df_alpha['SubjectID'] = df_alpha['Patient_ID'] + df_alpha['Study_ID']\n",
    "        df_alpha = df_alpha[['SubjectID', 'Timepoints', 'shannon_entropy']]\n",
    "        \n",
    "        # drop unpaired samples\n",
    "        s_remove = []\n",
    "        for s in list(df_alpha['SubjectID'].values):\n",
    "            if len(df_alpha[df_alpha['SubjectID'] == s]) != 2:\n",
    "                s_remove.append(s)\n",
    "        df_alpha = df_alpha.loc[~df_alpha['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "        \n",
    "        # set vars\n",
    "        alpha_metric = 'shannon_entropy'\n",
    "        group_var = 'Timepoints'\n",
    "        pair_var = 'SubjectID'\n",
    "        groups = ['pre','post']\n",
    "        \n",
    "        # get paired per indiv pair\n",
    "        pair_to_diff = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            df = df_alpha[df_alpha[pair_var] == p]\n",
    "            alpha_0 = float(df[df[group_var] == groups[0]][alpha_metric].values)\n",
    "            alpha_1 = float(df[df[group_var] == groups[1]][alpha_metric].values)\n",
    "            pair_to_diff[p] = alpha_0 - alpha_1\n",
    "        \n",
    "        df_paired_alpha = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[alpha_metric + '_diff'])\n",
    "        \n",
    "        # one-sided t-test, n.s.; RA-UA values \n",
    "        # t, p = scipy.stats.ttest_1samp(df_paired_alpha[alpha_metric + '_diff'],popmean=0)\n",
    "    \n",
    "        print('paired alpha wilcoxon')\n",
    "        s, p = scipy.stats.wilcoxon(df_paired_alpha[alpha_metric + '_diff'])\n",
    "        print(s, p)\n",
    "        \n",
    "        # separate\n",
    "        df_alpha[alpha_metric] = df_alpha[alpha_metric].map(lambda x: float(x))\n",
    "        ax = sns.boxplot(data=df_alpha, x=group_var, y=alpha_metric)\n",
    "        sns.swarmplot(data=df_alpha, x=group_var, y=alpha_metric, palette='dark:grey')\n",
    "        sns.despine()\n",
    "    \n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/Qiime2_' + g + '/' + sg + '_alpha.pdf')\n",
    "        plt.close()          \n",
    "    \n",
    "        # now do beta\n",
    "        df_beta = pd.read_csv(path + 'outputs/Qiime2_' + g + '/core_metrics_results/distance-matrix.tsv',\n",
    "                              sep='\\t', index_col=0)\n",
    "        \n",
    "        # set vars\n",
    "        alpha_metric = 'shannon_entropy'\n",
    "        group_var = 'Timepoints'\n",
    "        pair_var = 'SubjectID'\n",
    "        groups = ['pre','post']\n",
    "        g0, g1 = groups[0], groups[1]\n",
    "        \n",
    "        # grab twin to pair dict\n",
    "        pair_to_ids = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            df = df_alpha[df_alpha[pair_var] == p]\n",
    "            id_0 = str(df[df[group_var] == g0].index.values[0])\n",
    "            id_1 = str(df[df[group_var] == g1].index.values[0])\n",
    "            pair_to_ids[p] = (id_0, id_1)\n",
    "        \n",
    "        # get distances for each twin pair per beta div matrix    \n",
    "        pair_to_dist = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            id_0, id_1 = pair_to_ids[p]\n",
    "            pair_to_dist[p] = df_beta.loc[id_0, id_1]\n",
    "        \n",
    "        df_paired_beta = pd.DataFrame.from_dict(pair_to_dist, orient='index', columns=['Unweighted_Unifrac'])\n",
    "        \n",
    "        # grab inter RA distances\n",
    "        # this is from unweighted_Timepoint_significance.qzv -> download as tsv\n",
    "        df_raw = pd.read_csv(path + 'outputs/Qiime2_' + g + '/raw_data.tsv', \n",
    "                             sep='\\t', index_col=0)\n",
    "        df_0 = df_raw[df_raw['Group1'] == g0]\n",
    "        df_0 = df_0[df_0['Group2'] == g0]\n",
    "        df_1 = df_raw[df_raw['Group1'] == g1]\n",
    "        df_1 = df_1[df_1['Group2'] == g1]\n",
    "        \n",
    "        # compare distances\n",
    "        inter_twin = df_paired_beta['Unweighted_Unifrac'].values\n",
    "        inter_0 = df_0['Distance'].values\n",
    "        inter_1 = df_1['Distance'].values\n",
    "        \n",
    "        u, p = scipy.stats.mannwhitneyu(inter_twin, inter_0)\n",
    "        #print(u, p)\n",
    "        \n",
    "        t, p = scipy.stats.ttest_ind(inter_twin, inter_1)\n",
    "        #print(t, p)\n",
    "        \n",
    "        t, p = scipy.stats.ttest_ind(inter_0, inter_1)\n",
    "        # print(t, p)\n",
    "        \n",
    "        f, p = scipy.stats.f_oneway(inter_0, inter_1, inter_twin)\n",
    "        # print('inter intra anova')\n",
    "        p# rint(f, p)\n",
    "        \n",
    "        category = ['intra_twin_pair']*len(inter_twin) + ['inter_' + g0 + '_only']*len(inter_0) + ['inter_' + g1 + '_only']*len(inter_1)\n",
    "        distances = list(inter_twin) + list(inter_0) + list(inter_1)\n",
    "        df_dist = pd.DataFrame(data=np.array([category,distances]).T, columns=['category','distance'])\n",
    "        df_dist['distance'] = df_dist['distance'].astype(float)\n",
    "        df_dist.to_csv(path + 'outputs/Qiime2_' + g + '/' + sg + '_inter_intra_beta_dist.tsv',sep='\\t')\n",
    "                             \n",
    "        sns.boxplot(data=df_dist, x='category', y='distance')\n",
    "        sns.swarmplot(data=df_dist, x='category', y='distance', color='black')\n",
    "        sns.despine()\n",
    "    \n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/Qiime2_' + g + '/' + sg + '_beta.pdf')\n",
    "        plt.close()          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f2ebdb4d-0d17-4c0d-b282-cd78e27f7bb7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectID</th>\n",
       "      <th>Timepoints</th>\n",
       "      <th>shannon_entropy</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>WOMAC_P_Response</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>OAD001</td>\n",
       "      <td>pre</td>\n",
       "      <td>5.097506</td>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>OAD001</td>\n",
       "      <td>post</td>\n",
       "      <td>5.838205</td>\n",
       "      <td>post</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>OAD003</td>\n",
       "      <td>pre</td>\n",
       "      <td>5.190428</td>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>OAD003</td>\n",
       "      <td>post</td>\n",
       "      <td>4.679640</td>\n",
       "      <td>post</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>OAD004</td>\n",
       "      <td>pre</td>\n",
       "      <td>5.299560</td>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SubjectID Timepoints  shannon_entropy Timepoint  \\\n",
       "OAD-001.pre.stool     OAD001        pre         5.097506       pre   \n",
       "OAD-001.post.stool    OAD001       post         5.838205      post   \n",
       "OAD-003.pre.stool     OAD003        pre         5.190428       pre   \n",
       "OAD-003.post.stool    OAD003       post         4.679640      post   \n",
       "OAD-004.pre.stool     OAD004        pre         5.299560       pre   \n",
       "\n",
       "                   WOMAC_P_Response  \n",
       "OAD-001.pre.stool       No response  \n",
       "OAD-001.post.stool      No response  \n",
       "OAD-003.pre.stool       No response  \n",
       "OAD-003.post.stool      No response  \n",
       "OAD-004.pre.stool       No response  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# alpha plots\n",
    "for g in ['saliva','stool']:\n",
    "    df_alpha = g_to_dfd[g + '_adh']['alpha']\n",
    "    df_map = type_to_df_map[g].copy()\n",
    "    df_map = df_map.set_index('Together')\n",
    "    df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.g')[0])\n",
    "    \n",
    "    df_merge = pd.concat([df_alpha, df_map], axis=1)\n",
    "    df_merge = df_merge.dropna()\n",
    "    \n",
    "    # sns.set_theme()\n",
    "    custom_params = {\"axes.spines.right\": False, \"axes.spines.top\": False}\n",
    "    sns.set_theme(style=\"ticks\", rc=custom_params)\n",
    "    plt.figure(figsize=(6,4))\n",
    "    sns.boxplot(x='Timepoint', y='shannon_entropy', data=df_merge,palette=['b','r'], boxprops=dict(alpha=0.5))#, hue='Timepoint', palette=['b','r'])\n",
    "    ax = sns.swarmplot(x='Timepoint', y='shannon_entropy', data=df_merge,color='black')\n",
    "    for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "        ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "    sns.despine()#(left=True, bottom=True)\n",
    "    \n",
    "    # ax.axes.set_title(\"\",fontsize=48)\\n\",\n",
    "    plt.tight_layout()\n",
    "    ax.set_ylabel(\"Shannon Entropy\",fontsize=16)\n",
    "    ax.set_xlabel(\"Diagnosis\",fontsize=16)\n",
    "    ax.tick_params(labelsize=16)\n",
    "    \n",
    "    # plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/Qiime2_' + g + '_adh/alpha_all.pdf')\n",
    "    plt.close()\n",
    "\n",
    "df_merge.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51757cac-d364-49d9-88b5-d27af66b38fd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Process Taxa DF ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8d8a7f96-5705-47fb-8d64-6ddcc4cd5d72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;__;__;__;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>0.014084</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.022027</td>\n",
       "      <td>0.001515</td>\n",
       "      <td>0.173283</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021230</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.152452</td>\n",
       "      <td>0.033001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.022576</td>\n",
       "      <td>0.007098</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.292202</td>\n",
       "      <td>0.001994</td>\n",
       "      <td>0.013555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000783</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>0.273829</td>\n",
       "      <td>0.067421</td>\n",
       "      <td>0.184572</td>\n",
       "      <td>0.026929</td>\n",
       "      <td>0.053673</td>\n",
       "      <td>0.037783</td>\n",
       "      <td>0.009619</td>\n",
       "      <td>0.023935</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.012060</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.181156</td>\n",
       "      <td>0.067844</td>\n",
       "      <td>0.150895</td>\n",
       "      <td>0.063229</td>\n",
       "      <td>0.051360</td>\n",
       "      <td>0.055600</td>\n",
       "      <td>0.012363</td>\n",
       "      <td>0.051513</td>\n",
       "      <td>0.006045</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.post.stool</th>\n",
       "      <td>0.174773</td>\n",
       "      <td>0.040785</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001855</td>\n",
       "      <td>0.124402</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001400</td>\n",
       "      <td>0.000371</td>\n",
       "      <td>0.014265</td>\n",
       "      <td>0.090594</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 374 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004  \\\n",
       "index                                                                                                                       \n",
       "OAD-001.post.stool                                           0.014084                                                       \n",
       "OAD-001.pre.stool                                            0.022576                                                       \n",
       "OAD-003.post.stool                                           0.273829                                                       \n",
       "OAD-003.pre.stool                                            0.181156                                                       \n",
       "OAD-004.post.stool                                           0.174773                                                       \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium  \\\n",
       "index                                                                                                                            \n",
       "OAD-001.post.stool                                           0.039046                                                            \n",
       "OAD-001.pre.stool                                            0.007098                                                            \n",
       "OAD-003.post.stool                                           0.067421                                                            \n",
       "OAD-003.pre.stool                                            0.067844                                                            \n",
       "OAD-004.post.stool                                           0.040785                                                            \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella  \\\n",
       "index                                                                                                             \n",
       "OAD-001.post.stool                                           0.022027                                             \n",
       "OAD-001.pre.stool                                            0.005128                                             \n",
       "OAD-003.post.stool                                           0.184572                                             \n",
       "OAD-003.pre.stool                                            0.150895                                             \n",
       "OAD-004.post.stool                                           0.000000                                             \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117  \\\n",
       "index                                                                                                                              \n",
       "OAD-001.post.stool                                           0.001515                                                              \n",
       "OAD-001.pre.stool                                            0.000000                                                              \n",
       "OAD-003.post.stool                                           0.026929                                                              \n",
       "OAD-003.pre.stool                                            0.063229                                                              \n",
       "OAD-004.post.stool                                           0.001855                                                              \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H  \\\n",
       "index                                                                                                                \n",
       "OAD-001.post.stool                                           0.173283                                                \n",
       "OAD-001.pre.stool                                            0.292202                                                \n",
       "OAD-003.post.stool                                           0.053673                                                \n",
       "OAD-003.pre.stool                                            0.051360                                                \n",
       "OAD-004.post.stool                                           0.124402                                                \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum  \\\n",
       "index                                                                                                                  \n",
       "OAD-001.post.stool                                           0.000000                                                  \n",
       "OAD-001.pre.stool                                            0.001994                                                  \n",
       "OAD-003.post.stool                                           0.037783                                                  \n",
       "OAD-003.pre.stool                                            0.055600                                                  \n",
       "OAD-004.post.stool                                           0.000000                                                  \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129  \\\n",
       "index                                                                                                                            \n",
       "OAD-001.post.stool                                           0.021230                                                            \n",
       "OAD-001.pre.stool                                            0.013555                                                            \n",
       "OAD-003.post.stool                                           0.009619                                                            \n",
       "OAD-003.pre.stool                                            0.012363                                                            \n",
       "OAD-004.post.stool                                           0.001400                                                            \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus  \\\n",
       "index                                                                                                                \n",
       "OAD-001.post.stool                                           0.000160                                                \n",
       "OAD-001.pre.stool                                            0.000000                                                \n",
       "OAD-003.post.stool                                           0.023935                                                \n",
       "OAD-003.pre.stool                                            0.051513                                                \n",
       "OAD-004.post.stool                                           0.000371                                                \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.post.stool                                           0.152452                                                        \n",
       "OAD-001.pre.stool                                            0.000783                                                        \n",
       "OAD-003.post.stool                                           0.005789                                                        \n",
       "OAD-003.pre.stool                                            0.006045                                                        \n",
       "OAD-004.post.stool                                           0.014265                                                        \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400  \\\n",
       "index                                                                                                                    \n",
       "OAD-001.post.stool                                           0.033001                                                    \n",
       "OAD-001.pre.stool                                            0.026849                                                    \n",
       "OAD-003.post.stool                                           0.012060                                                    \n",
       "OAD-003.pre.stool                                            0.025918                                                    \n",
       "OAD-004.post.stool                                           0.090594                                                    \n",
       "\n",
       "                    ...  d__Bacteria;p__Firmicutes_A;__;__;__;__  \\\n",
       "index               ...                                            \n",
       "OAD-001.post.stool  ...                                      0.0   \n",
       "OAD-001.pre.stool   ...                                      0.0   \n",
       "OAD-003.post.stool  ...                                      0.0   \n",
       "OAD-003.pre.stool   ...                                      0.0   \n",
       "OAD-004.post.stool  ...                                      0.0   \n",
       "\n",
       "                    d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.post.stool                                                0.0                                                        \n",
       "OAD-001.pre.stool                                                 0.0                                                        \n",
       "OAD-003.post.stool                                                0.0                                                        \n",
       "OAD-003.pre.stool                                                 0.0                                                        \n",
       "OAD-004.post.stool                                                0.0                                                        \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "index                                                                                 \n",
       "OAD-001.post.stool                                                0.0                 \n",
       "OAD-001.pre.stool                                                 0.0                 \n",
       "OAD-003.post.stool                                                0.0                 \n",
       "OAD-003.pre.stool                                                 0.0                 \n",
       "OAD-004.post.stool                                                0.0                 \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "index                                                                                                                         \n",
       "OAD-001.post.stool                                                0.0                                                         \n",
       "OAD-001.pre.stool                                                 0.0                                                         \n",
       "OAD-003.post.stool                                                0.0                                                         \n",
       "OAD-003.pre.stool                                                 0.0                                                         \n",
       "OAD-004.post.stool                                                0.0                                                         \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "index                                                                                                               \n",
       "OAD-001.post.stool                                                0.0                                               \n",
       "OAD-001.pre.stool                                                 0.0                                               \n",
       "OAD-003.post.stool                                                0.0                                               \n",
       "OAD-003.pre.stool                                                 0.0                                               \n",
       "OAD-004.post.stool                                                0.0                                               \n",
       "\n",
       "                    d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "index                                                                                                                                         \n",
       "OAD-001.post.stool                                                0.0                                                                         \n",
       "OAD-001.pre.stool                                                 0.0                                                                         \n",
       "OAD-003.post.stool                                                0.0                                                                         \n",
       "OAD-003.pre.stool                                                 0.0                                                                         \n",
       "OAD-004.post.stool                                                0.0                                                                         \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "index                                                                                                     \n",
       "OAD-001.post.stool                                                0.0                                     \n",
       "OAD-001.pre.stool                                                 0.0                                     \n",
       "OAD-003.post.stool                                                0.0                                     \n",
       "OAD-003.pre.stool                                                 0.0                                     \n",
       "OAD-004.post.stool                                                0.0                                     \n",
       "\n",
       "                    d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "index                                                                                                                                        \n",
       "OAD-001.post.stool                                                0.0                                                                        \n",
       "OAD-001.pre.stool                                                 0.0                                                                        \n",
       "OAD-003.post.stool                                                0.0                                                                        \n",
       "OAD-003.pre.stool                                                 0.0                                                                        \n",
       "OAD-004.post.stool                                                0.0                                                                        \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "index                                                                                                  \n",
       "OAD-001.post.stool                                                0.0                                  \n",
       "OAD-001.pre.stool                                                 0.0                                  \n",
       "OAD-003.post.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                 0.0                                  \n",
       "OAD-004.post.stool                                                0.0                                  \n",
       "\n",
       "                    d__Bacteria;p__Actinobacteriota;__;__;__;__  \n",
       "index                                                            \n",
       "OAD-001.post.stool                                          0.0  \n",
       "OAD-001.pre.stool                                           0.0  \n",
       "OAD-003.post.stool                                          0.0  \n",
       "OAD-003.pre.stool                                           0.0  \n",
       "OAD-004.post.stool                                          0.0  \n",
       "\n",
       "[5 rows x 374 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specific taxa profiling\n",
    "GL_to_otu = {}\n",
    "\n",
    "groups = ['saliva_adh', 'stool_adh']\n",
    "levels = ['6','7']\n",
    "for g in groups:\n",
    "    GL_to_otu[g] = {}\n",
    "    for l in levels:\n",
    "        # read in csv\n",
    "        df_otu = pd.read_csv(path + 'outputs/Qiime2_' + g + '/level-' + l + '.csv', index_col=0)\n",
    "\n",
    "        # filter on OTU only \n",
    "        keep = []\n",
    "        for x in df_otu.columns:\n",
    "            if 'd__' in x:\n",
    "                keep.append(x)\n",
    "        df_otu = df_otu.loc[:,keep]\n",
    "\n",
    "        # normalize by dividing by row sum\n",
    "        df_otu = df_otu.div(df_otu.sum(axis=1),axis=0)\n",
    "\n",
    "        # save to dict\n",
    "        GL_to_otu[g]['L' + l] = df_otu\n",
    "\n",
    "                                       \n",
    "GL_to_otu['stool_adh']['L6'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "515d9fc7-6c65-439d-8733-6c92c35fc5b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;__;__;__;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.022576</td>\n",
       "      <td>0.007098</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.292202</td>\n",
       "      <td>0.001994</td>\n",
       "      <td>0.013555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000783</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.181156</td>\n",
       "      <td>0.067844</td>\n",
       "      <td>0.150895</td>\n",
       "      <td>0.063229</td>\n",
       "      <td>0.051360</td>\n",
       "      <td>0.055600</td>\n",
       "      <td>0.012363</td>\n",
       "      <td>0.051513</td>\n",
       "      <td>0.006045</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.170334</td>\n",
       "      <td>0.135468</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002088</td>\n",
       "      <td>0.173617</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003649</td>\n",
       "      <td>0.008765</td>\n",
       "      <td>0.054802</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-005.pre.stool</th>\n",
       "      <td>0.040108</td>\n",
       "      <td>0.139385</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018806</td>\n",
       "      <td>0.155290</td>\n",
       "      <td>0.037507</td>\n",
       "      <td>0.024457</td>\n",
       "      <td>0.066852</td>\n",
       "      <td>0.008969</td>\n",
       "      <td>0.018328</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-006.pre.stool</th>\n",
       "      <td>0.037219</td>\n",
       "      <td>0.167448</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007940</td>\n",
       "      <td>0.019591</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011006</td>\n",
       "      <td>0.001275</td>\n",
       "      <td>0.012769</td>\n",
       "      <td>0.162962</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 374 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004  \\\n",
       "index                                                                                                                      \n",
       "OAD-001.pre.stool                                           0.022576                                                       \n",
       "OAD-003.pre.stool                                           0.181156                                                       \n",
       "OAD-004.pre.stool                                           0.170334                                                       \n",
       "OAD-005.pre.stool                                           0.040108                                                       \n",
       "OAD-006.pre.stool                                           0.037219                                                       \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium  \\\n",
       "index                                                                                                                           \n",
       "OAD-001.pre.stool                                           0.007098                                                            \n",
       "OAD-003.pre.stool                                           0.067844                                                            \n",
       "OAD-004.pre.stool                                           0.135468                                                            \n",
       "OAD-005.pre.stool                                           0.139385                                                            \n",
       "OAD-006.pre.stool                                           0.167448                                                            \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella  \\\n",
       "index                                                                                                            \n",
       "OAD-001.pre.stool                                           0.005128                                             \n",
       "OAD-003.pre.stool                                           0.150895                                             \n",
       "OAD-004.pre.stool                                           0.000000                                             \n",
       "OAD-005.pre.stool                                           0.000000                                             \n",
       "OAD-006.pre.stool                                           0.000000                                             \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117  \\\n",
       "index                                                                                                                             \n",
       "OAD-001.pre.stool                                           0.000000                                                              \n",
       "OAD-003.pre.stool                                           0.063229                                                              \n",
       "OAD-004.pre.stool                                           0.002088                                                              \n",
       "OAD-005.pre.stool                                           0.018806                                                              \n",
       "OAD-006.pre.stool                                           0.007940                                                              \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H  \\\n",
       "index                                                                                                               \n",
       "OAD-001.pre.stool                                           0.292202                                                \n",
       "OAD-003.pre.stool                                           0.051360                                                \n",
       "OAD-004.pre.stool                                           0.173617                                                \n",
       "OAD-005.pre.stool                                           0.155290                                                \n",
       "OAD-006.pre.stool                                           0.019591                                                \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum  \\\n",
       "index                                                                                                                 \n",
       "OAD-001.pre.stool                                           0.001994                                                  \n",
       "OAD-003.pre.stool                                           0.055600                                                  \n",
       "OAD-004.pre.stool                                           0.000143                                                  \n",
       "OAD-005.pre.stool                                           0.037507                                                  \n",
       "OAD-006.pre.stool                                           0.000000                                                  \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129  \\\n",
       "index                                                                                                                           \n",
       "OAD-001.pre.stool                                           0.013555                                                            \n",
       "OAD-003.pre.stool                                           0.012363                                                            \n",
       "OAD-004.pre.stool                                           0.000000                                                            \n",
       "OAD-005.pre.stool                                           0.024457                                                            \n",
       "OAD-006.pre.stool                                           0.011006                                                            \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus  \\\n",
       "index                                                                                                               \n",
       "OAD-001.pre.stool                                           0.000000                                                \n",
       "OAD-003.pre.stool                                           0.051513                                                \n",
       "OAD-004.pre.stool                                           0.003649                                                \n",
       "OAD-005.pre.stool                                           0.066852                                                \n",
       "OAD-006.pre.stool                                           0.001275                                                \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor  \\\n",
       "index                                                                                                                       \n",
       "OAD-001.pre.stool                                           0.000783                                                        \n",
       "OAD-003.pre.stool                                           0.006045                                                        \n",
       "OAD-004.pre.stool                                           0.008765                                                        \n",
       "OAD-005.pre.stool                                           0.008969                                                        \n",
       "OAD-006.pre.stool                                           0.012769                                                        \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400  \\\n",
       "index                                                                                                                   \n",
       "OAD-001.pre.stool                                           0.026849                                                    \n",
       "OAD-003.pre.stool                                           0.025918                                                    \n",
       "OAD-004.pre.stool                                           0.054802                                                    \n",
       "OAD-005.pre.stool                                           0.018328                                                    \n",
       "OAD-006.pre.stool                                           0.162962                                                    \n",
       "\n",
       "                   ...  d__Bacteria;p__Firmicutes_A;__;__;__;__  \\\n",
       "index              ...                                            \n",
       "OAD-001.pre.stool  ...                                      0.0   \n",
       "OAD-003.pre.stool  ...                                      0.0   \n",
       "OAD-004.pre.stool  ...                                      0.0   \n",
       "OAD-005.pre.stool  ...                                      0.0   \n",
       "OAD-006.pre.stool  ...                                      0.0   \n",
       "\n",
       "                   d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "index                                                                                                                       \n",
       "OAD-001.pre.stool                                                0.0                                                        \n",
       "OAD-003.pre.stool                                                0.0                                                        \n",
       "OAD-004.pre.stool                                                0.0                                                        \n",
       "OAD-005.pre.stool                                                0.0                                                        \n",
       "OAD-006.pre.stool                                                0.0                                                        \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "index                                                                                \n",
       "OAD-001.pre.stool                                                0.0                 \n",
       "OAD-003.pre.stool                                                0.0                 \n",
       "OAD-004.pre.stool                                                0.0                 \n",
       "OAD-005.pre.stool                                                0.0                 \n",
       "OAD-006.pre.stool                                                0.0                 \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.pre.stool                                                0.0                                                         \n",
       "OAD-003.pre.stool                                                0.0                                                         \n",
       "OAD-004.pre.stool                                                0.0                                                         \n",
       "OAD-005.pre.stool                                                0.0                                                         \n",
       "OAD-006.pre.stool                                                0.0                                                         \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "index                                                                                                              \n",
       "OAD-001.pre.stool                                                0.0                                               \n",
       "OAD-003.pre.stool                                                0.0                                               \n",
       "OAD-004.pre.stool                                                0.0                                               \n",
       "OAD-005.pre.stool                                                0.0                                               \n",
       "OAD-006.pre.stool                                                0.0                                               \n",
       "\n",
       "                   d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "index                                                                                                                                        \n",
       "OAD-001.pre.stool                                                0.0                                                                         \n",
       "OAD-003.pre.stool                                                0.0                                                                         \n",
       "OAD-004.pre.stool                                                0.0                                                                         \n",
       "OAD-005.pre.stool                                                0.0                                                                         \n",
       "OAD-006.pre.stool                                                0.0                                                                         \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "index                                                                                                    \n",
       "OAD-001.pre.stool                                                0.0                                     \n",
       "OAD-003.pre.stool                                                0.0                                     \n",
       "OAD-004.pre.stool                                                0.0                                     \n",
       "OAD-005.pre.stool                                                0.0                                     \n",
       "OAD-006.pre.stool                                                0.0                                     \n",
       "\n",
       "                   d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "index                                                                                                                                       \n",
       "OAD-001.pre.stool                                                0.0                                                                        \n",
       "OAD-003.pre.stool                                                0.0                                                                        \n",
       "OAD-004.pre.stool                                                0.0                                                                        \n",
       "OAD-005.pre.stool                                                0.0                                                                        \n",
       "OAD-006.pre.stool                                                0.0                                                                        \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "index                                                                                                 \n",
       "OAD-001.pre.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                0.0                                  \n",
       "OAD-004.pre.stool                                                0.0                                  \n",
       "OAD-005.pre.stool                                                0.0                                  \n",
       "OAD-006.pre.stool                                                0.0                                  \n",
       "\n",
       "                   d__Bacteria;p__Actinobacteriota;__;__;__;__  \n",
       "index                                                           \n",
       "OAD-001.pre.stool                                          0.0  \n",
       "OAD-003.pre.stool                                          0.0  \n",
       "OAD-004.pre.stool                                          0.0  \n",
       "OAD-005.pre.stool                                          0.0  \n",
       "OAD-006.pre.stool                                          0.0  \n",
       "\n",
       "[5 rows x 374 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Microbiome testing\n",
    "gt_to_otu = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gt_to_otu[g] = {}\n",
    "    df_otu = GL_to_otu[g + '_adh']['L6'].copy()\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['post','pre']:\n",
    "        df = df_otu[df_otu.index.to_series().str.contains(t)]\n",
    "        gt_to_otu[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_otu[g]['pre'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.split('.pre')[0])\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('-',''))\n",
    "    post_df = gt_to_otu[g]['post'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.split('.post')[0])\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('-',''))\n",
    "    \n",
    "    gt_to_otu[g]['diff'] = post_df - pre_df\n",
    "    gt_to_otu[g]['all'] = pd.concat([gt_to_otu[g]['pre'],gt_to_otu[g]['post']])\n",
    "\n",
    "gt_to_otu[g][t].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faf9a2fe-9f68-4cee-9e40-4d5b749b9f0e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Process Metabolite DF ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6bf1fdd5-b592-4a4e-a5d4-3cefc2d28c9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_Thiamine</th>\n",
       "      <th>H_Melezitose</th>\n",
       "      <th>H_Phosphocholine</th>\n",
       "      <th>H_SN_Glycero_3_Phophocholine</th>\n",
       "      <th>H_Gamma_Valerobetaine</th>\n",
       "      <th>H_N_Acetylneuraminic_acid</th>\n",
       "      <th>H_Pyridoxine</th>\n",
       "      <th>H_N_Acetylneuraminic_Acid</th>\n",
       "      <th>H_N_Acetylmuramic_Acid</th>\n",
       "      <th>...</th>\n",
       "      <th>H_Nonhydroxylated_bile_acid</th>\n",
       "      <th>H_Hydroxydodecanoic_acid</th>\n",
       "      <th>H_Omega_Hydroxydodecanoate</th>\n",
       "      <th>H_Delta_Methyldodecenoic_Acid</th>\n",
       "      <th>H_Hydroxydecanoic_acid</th>\n",
       "      <th>H_Hydroxydecanoate</th>\n",
       "      <th>H_Methylpentanoic_acid</th>\n",
       "      <th>H_Monohydroxylated_bile_acid</th>\n",
       "      <th>H_Palmitoyl_ethanolamide</th>\n",
       "      <th>H_Linoleic_Acid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>0.312342</td>\n",
       "      <td>-0.312477</td>\n",
       "      <td>0.119853</td>\n",
       "      <td>-0.492876</td>\n",
       "      <td>-0.671376</td>\n",
       "      <td>-0.202952</td>\n",
       "      <td>0.203945</td>\n",
       "      <td>-0.455869</td>\n",
       "      <td>0.352628</td>\n",
       "      <td>0.151464</td>\n",
       "      <td>...</td>\n",
       "      <td>0.054336</td>\n",
       "      <td>-0.330037</td>\n",
       "      <td>-0.886045</td>\n",
       "      <td>-0.314427</td>\n",
       "      <td>-0.322677</td>\n",
       "      <td>-0.650357</td>\n",
       "      <td>-0.361695</td>\n",
       "      <td>0.132794</td>\n",
       "      <td>-0.067464</td>\n",
       "      <td>-0.112641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>0.000428</td>\n",
       "      <td>-0.481251</td>\n",
       "      <td>-0.179337</td>\n",
       "      <td>0.983217</td>\n",
       "      <td>0.167836</td>\n",
       "      <td>0.067823</td>\n",
       "      <td>0.364897</td>\n",
       "      <td>-0.371942</td>\n",
       "      <td>0.545435</td>\n",
       "      <td>-0.007664</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.022614</td>\n",
       "      <td>-0.334936</td>\n",
       "      <td>-0.802118</td>\n",
       "      <td>-0.929470</td>\n",
       "      <td>-0.369578</td>\n",
       "      <td>-0.566431</td>\n",
       "      <td>0.075375</td>\n",
       "      <td>-0.519189</td>\n",
       "      <td>-0.180631</td>\n",
       "      <td>-0.418351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>-0.285212</td>\n",
       "      <td>-0.556445</td>\n",
       "      <td>-0.651325</td>\n",
       "      <td>0.357509</td>\n",
       "      <td>-0.401875</td>\n",
       "      <td>-0.295787</td>\n",
       "      <td>0.137736</td>\n",
       "      <td>-0.165902</td>\n",
       "      <td>0.112170</td>\n",
       "      <td>0.096791</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.923477</td>\n",
       "      <td>-0.589028</td>\n",
       "      <td>-1.535141</td>\n",
       "      <td>-0.525277</td>\n",
       "      <td>-0.278312</td>\n",
       "      <td>-0.600484</td>\n",
       "      <td>0.019657</td>\n",
       "      <td>-0.576831</td>\n",
       "      <td>-0.901479</td>\n",
       "      <td>-1.089906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD004</th>\n",
       "      <td>0.140904</td>\n",
       "      <td>1.089055</td>\n",
       "      <td>-0.668077</td>\n",
       "      <td>-1.061336</td>\n",
       "      <td>-0.370602</td>\n",
       "      <td>-0.329379</td>\n",
       "      <td>-0.654931</td>\n",
       "      <td>0.142971</td>\n",
       "      <td>-1.006689</td>\n",
       "      <td>-0.100373</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.354297</td>\n",
       "      <td>0.449897</td>\n",
       "      <td>0.489189</td>\n",
       "      <td>0.413723</td>\n",
       "      <td>0.474250</td>\n",
       "      <td>0.661963</td>\n",
       "      <td>0.245553</td>\n",
       "      <td>0.279290</td>\n",
       "      <td>-0.444319</td>\n",
       "      <td>-0.912164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD011</th>\n",
       "      <td>-0.346855</td>\n",
       "      <td>1.028262</td>\n",
       "      <td>-0.450323</td>\n",
       "      <td>-0.025939</td>\n",
       "      <td>-0.223029</td>\n",
       "      <td>-0.766374</td>\n",
       "      <td>-0.538410</td>\n",
       "      <td>-1.091423</td>\n",
       "      <td>-0.328522</td>\n",
       "      <td>-0.524686</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.385428</td>\n",
       "      <td>0.419748</td>\n",
       "      <td>0.429119</td>\n",
       "      <td>0.333756</td>\n",
       "      <td>0.448461</td>\n",
       "      <td>0.598846</td>\n",
       "      <td>0.390035</td>\n",
       "      <td>0.287682</td>\n",
       "      <td>-0.532141</td>\n",
       "      <td>-0.752280</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 88 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          H_Pyridoxamine  H_Thiamine  H_Melezitose  H_Phosphocholine  \\\n",
       "Study_ID                                                               \n",
       "VAOAD001        0.312342   -0.312477      0.119853         -0.492876   \n",
       "VAOAD009        0.000428   -0.481251     -0.179337          0.983217   \n",
       "OAD001         -0.285212   -0.556445     -0.651325          0.357509   \n",
       "VAOAD004        0.140904    1.089055     -0.668077         -1.061336   \n",
       "VAOAD011       -0.346855    1.028262     -0.450323         -0.025939   \n",
       "\n",
       "          H_SN_Glycero_3_Phophocholine  H_Gamma_Valerobetaine  \\\n",
       "Study_ID                                                        \n",
       "VAOAD001                     -0.671376              -0.202952   \n",
       "VAOAD009                      0.167836               0.067823   \n",
       "OAD001                       -0.401875              -0.295787   \n",
       "VAOAD004                     -0.370602              -0.329379   \n",
       "VAOAD011                     -0.223029              -0.766374   \n",
       "\n",
       "          H_N_Acetylneuraminic_acid  H_Pyridoxine  H_N_Acetylneuraminic_Acid  \\\n",
       "Study_ID                                                                       \n",
       "VAOAD001                   0.203945     -0.455869                   0.352628   \n",
       "VAOAD009                   0.364897     -0.371942                   0.545435   \n",
       "OAD001                     0.137736     -0.165902                   0.112170   \n",
       "VAOAD004                  -0.654931      0.142971                  -1.006689   \n",
       "VAOAD011                  -0.538410     -1.091423                  -0.328522   \n",
       "\n",
       "          H_N_Acetylmuramic_Acid  ...  H_Nonhydroxylated_bile_acid  \\\n",
       "Study_ID                          ...                                \n",
       "VAOAD001                0.151464  ...                     0.054336   \n",
       "VAOAD009               -0.007664  ...                    -0.022614   \n",
       "OAD001                  0.096791  ...                    -0.923477   \n",
       "VAOAD004               -0.100373  ...                    -0.354297   \n",
       "VAOAD011               -0.524686  ...                    -0.385428   \n",
       "\n",
       "          H_Hydroxydodecanoic_acid  H_Omega_Hydroxydodecanoate  \\\n",
       "Study_ID                                                         \n",
       "VAOAD001                 -0.330037                   -0.886045   \n",
       "VAOAD009                 -0.334936                   -0.802118   \n",
       "OAD001                   -0.589028                   -1.535141   \n",
       "VAOAD004                  0.449897                    0.489189   \n",
       "VAOAD011                  0.419748                    0.429119   \n",
       "\n",
       "          H_Delta_Methyldodecenoic_Acid  H_Hydroxydecanoic_acid  \\\n",
       "Study_ID                                                          \n",
       "VAOAD001                      -0.314427               -0.322677   \n",
       "VAOAD009                      -0.929470               -0.369578   \n",
       "OAD001                        -0.525277               -0.278312   \n",
       "VAOAD004                       0.413723                0.474250   \n",
       "VAOAD011                       0.333756                0.448461   \n",
       "\n",
       "          H_Hydroxydecanoate  H_Methylpentanoic_acid  \\\n",
       "Study_ID                                               \n",
       "VAOAD001           -0.650357               -0.361695   \n",
       "VAOAD009           -0.566431                0.075375   \n",
       "OAD001             -0.600484                0.019657   \n",
       "VAOAD004            0.661963                0.245553   \n",
       "VAOAD011            0.598846                0.390035   \n",
       "\n",
       "          H_Monohydroxylated_bile_acid  H_Palmitoyl_ethanolamide  \\\n",
       "Study_ID                                                           \n",
       "VAOAD001                      0.132794                 -0.067464   \n",
       "VAOAD009                     -0.519189                 -0.180631   \n",
       "OAD001                       -0.576831                 -0.901479   \n",
       "VAOAD004                      0.279290                 -0.444319   \n",
       "VAOAD011                      0.287682                 -0.532141   \n",
       "\n",
       "          H_Linoleic_Acid  \n",
       "Study_ID                   \n",
       "VAOAD001        -0.112641  \n",
       "VAOAD009        -0.418351  \n",
       "OAD001          -1.089906  \n",
       "VAOAD004        -0.912164  \n",
       "VAOAD011        -0.752280  \n",
       "\n",
       "[5 rows x 88 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Metabolome testing\n",
    "gdt_to_df = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gdt_to_df[g] = {}\n",
    "    for d in ['paired_beta']: # just arbitrarily \n",
    "        gdt_to_df[g][d] = {}\n",
    "        df_merge = dfd_to_merge[g + '_adh'][d]\n",
    "        df_meta = pd.read_csv(path + 'inputs/' + g + '_normalized.csv')\n",
    "\n",
    "        # split on timepoint\n",
    "        for t in ['Baseline','After diet']:\n",
    "            df = df_meta[df_meta['Time'] == t]\n",
    "            df = df.set_index('Study_ID')\n",
    "            df.index = df.index.map(lambda x: x.split('_')[0])\n",
    "            df = df.drop('Time',axis=1)\n",
    "            gdt_to_df[g][d][t] = df\n",
    "\n",
    "        gdt_to_df[g][d]['diff'] = gdt_to_df[g][d]['After diet'] - gdt_to_df[g][d]['Baseline']\n",
    "\n",
    "gdt_to_df[g][d][t].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dbad4aa2-7c35-490c-900b-c607c4ff549d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_Thiamine</th>\n",
       "      <th>H_Melezitose</th>\n",
       "      <th>H_Phosphocholine</th>\n",
       "      <th>H_SN_Glycero_3_Phophocholine</th>\n",
       "      <th>H_Gamma_Valerobetaine</th>\n",
       "      <th>H_N_Acetylneuraminic_acid</th>\n",
       "      <th>H_Pyridoxine</th>\n",
       "      <th>H_N_Acetylneuraminic_Acid</th>\n",
       "      <th>H_N_Acetylmuramic_Acid</th>\n",
       "      <th>...</th>\n",
       "      <th>H_Nonhydroxylated_bile_acid</th>\n",
       "      <th>H_Hydroxydodecanoic_acid</th>\n",
       "      <th>H_Omega_Hydroxydodecanoate</th>\n",
       "      <th>H_Delta_Methyldodecenoic_Acid</th>\n",
       "      <th>H_Hydroxydecanoic_acid</th>\n",
       "      <th>H_Hydroxydecanoate</th>\n",
       "      <th>H_Methylpentanoic_acid</th>\n",
       "      <th>H_Monohydroxylated_bile_acid</th>\n",
       "      <th>H_Palmitoyl_ethanolamide</th>\n",
       "      <th>H_Linoleic_Acid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>0.312342</td>\n",
       "      <td>-0.312477</td>\n",
       "      <td>0.119853</td>\n",
       "      <td>-0.492876</td>\n",
       "      <td>-0.671376</td>\n",
       "      <td>-0.202952</td>\n",
       "      <td>0.203945</td>\n",
       "      <td>-0.455869</td>\n",
       "      <td>0.352628</td>\n",
       "      <td>0.151464</td>\n",
       "      <td>...</td>\n",
       "      <td>0.054336</td>\n",
       "      <td>-0.330037</td>\n",
       "      <td>-0.886045</td>\n",
       "      <td>-0.314427</td>\n",
       "      <td>-0.322677</td>\n",
       "      <td>-0.650357</td>\n",
       "      <td>-0.361695</td>\n",
       "      <td>0.132794</td>\n",
       "      <td>-0.067464</td>\n",
       "      <td>-0.112641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>0.000428</td>\n",
       "      <td>-0.481251</td>\n",
       "      <td>-0.179337</td>\n",
       "      <td>0.983217</td>\n",
       "      <td>0.167836</td>\n",
       "      <td>0.067823</td>\n",
       "      <td>0.364897</td>\n",
       "      <td>-0.371942</td>\n",
       "      <td>0.545435</td>\n",
       "      <td>-0.007664</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.022614</td>\n",
       "      <td>-0.334936</td>\n",
       "      <td>-0.802118</td>\n",
       "      <td>-0.929470</td>\n",
       "      <td>-0.369578</td>\n",
       "      <td>-0.566431</td>\n",
       "      <td>0.075375</td>\n",
       "      <td>-0.519189</td>\n",
       "      <td>-0.180631</td>\n",
       "      <td>-0.418351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>-0.285212</td>\n",
       "      <td>-0.556445</td>\n",
       "      <td>-0.651325</td>\n",
       "      <td>0.357509</td>\n",
       "      <td>-0.401875</td>\n",
       "      <td>-0.295787</td>\n",
       "      <td>0.137736</td>\n",
       "      <td>-0.165902</td>\n",
       "      <td>0.112170</td>\n",
       "      <td>0.096791</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.923477</td>\n",
       "      <td>-0.589028</td>\n",
       "      <td>-1.535141</td>\n",
       "      <td>-0.525277</td>\n",
       "      <td>-0.278312</td>\n",
       "      <td>-0.600484</td>\n",
       "      <td>0.019657</td>\n",
       "      <td>-0.576831</td>\n",
       "      <td>-0.901479</td>\n",
       "      <td>-1.089906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD004</th>\n",
       "      <td>0.140904</td>\n",
       "      <td>1.089055</td>\n",
       "      <td>-0.668077</td>\n",
       "      <td>-1.061336</td>\n",
       "      <td>-0.370602</td>\n",
       "      <td>-0.329379</td>\n",
       "      <td>-0.654931</td>\n",
       "      <td>0.142971</td>\n",
       "      <td>-1.006689</td>\n",
       "      <td>-0.100373</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.354297</td>\n",
       "      <td>0.449897</td>\n",
       "      <td>0.489189</td>\n",
       "      <td>0.413723</td>\n",
       "      <td>0.474250</td>\n",
       "      <td>0.661963</td>\n",
       "      <td>0.245553</td>\n",
       "      <td>0.279290</td>\n",
       "      <td>-0.444319</td>\n",
       "      <td>-0.912164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD011</th>\n",
       "      <td>-0.346855</td>\n",
       "      <td>1.028262</td>\n",
       "      <td>-0.450323</td>\n",
       "      <td>-0.025939</td>\n",
       "      <td>-0.223029</td>\n",
       "      <td>-0.766374</td>\n",
       "      <td>-0.538410</td>\n",
       "      <td>-1.091423</td>\n",
       "      <td>-0.328522</td>\n",
       "      <td>-0.524686</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.385428</td>\n",
       "      <td>0.419748</td>\n",
       "      <td>0.429119</td>\n",
       "      <td>0.333756</td>\n",
       "      <td>0.448461</td>\n",
       "      <td>0.598846</td>\n",
       "      <td>0.390035</td>\n",
       "      <td>0.287682</td>\n",
       "      <td>-0.532141</td>\n",
       "      <td>-0.752280</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 88 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          H_Pyridoxamine  H_Thiamine  H_Melezitose  H_Phosphocholine  \\\n",
       "Study_ID                                                               \n",
       "VAOAD001        0.312342   -0.312477      0.119853         -0.492876   \n",
       "VAOAD009        0.000428   -0.481251     -0.179337          0.983217   \n",
       "OAD001         -0.285212   -0.556445     -0.651325          0.357509   \n",
       "VAOAD004        0.140904    1.089055     -0.668077         -1.061336   \n",
       "VAOAD011       -0.346855    1.028262     -0.450323         -0.025939   \n",
       "\n",
       "          H_SN_Glycero_3_Phophocholine  H_Gamma_Valerobetaine  \\\n",
       "Study_ID                                                        \n",
       "VAOAD001                     -0.671376              -0.202952   \n",
       "VAOAD009                      0.167836               0.067823   \n",
       "OAD001                       -0.401875              -0.295787   \n",
       "VAOAD004                     -0.370602              -0.329379   \n",
       "VAOAD011                     -0.223029              -0.766374   \n",
       "\n",
       "          H_N_Acetylneuraminic_acid  H_Pyridoxine  H_N_Acetylneuraminic_Acid  \\\n",
       "Study_ID                                                                       \n",
       "VAOAD001                   0.203945     -0.455869                   0.352628   \n",
       "VAOAD009                   0.364897     -0.371942                   0.545435   \n",
       "OAD001                     0.137736     -0.165902                   0.112170   \n",
       "VAOAD004                  -0.654931      0.142971                  -1.006689   \n",
       "VAOAD011                  -0.538410     -1.091423                  -0.328522   \n",
       "\n",
       "          H_N_Acetylmuramic_Acid  ...  H_Nonhydroxylated_bile_acid  \\\n",
       "Study_ID                          ...                                \n",
       "VAOAD001                0.151464  ...                     0.054336   \n",
       "VAOAD009               -0.007664  ...                    -0.022614   \n",
       "OAD001                  0.096791  ...                    -0.923477   \n",
       "VAOAD004               -0.100373  ...                    -0.354297   \n",
       "VAOAD011               -0.524686  ...                    -0.385428   \n",
       "\n",
       "          H_Hydroxydodecanoic_acid  H_Omega_Hydroxydodecanoate  \\\n",
       "Study_ID                                                         \n",
       "VAOAD001                 -0.330037                   -0.886045   \n",
       "VAOAD009                 -0.334936                   -0.802118   \n",
       "OAD001                   -0.589028                   -1.535141   \n",
       "VAOAD004                  0.449897                    0.489189   \n",
       "VAOAD011                  0.419748                    0.429119   \n",
       "\n",
       "          H_Delta_Methyldodecenoic_Acid  H_Hydroxydecanoic_acid  \\\n",
       "Study_ID                                                          \n",
       "VAOAD001                      -0.314427               -0.322677   \n",
       "VAOAD009                      -0.929470               -0.369578   \n",
       "OAD001                        -0.525277               -0.278312   \n",
       "VAOAD004                       0.413723                0.474250   \n",
       "VAOAD011                       0.333756                0.448461   \n",
       "\n",
       "          H_Hydroxydecanoate  H_Methylpentanoic_acid  \\\n",
       "Study_ID                                               \n",
       "VAOAD001           -0.650357               -0.361695   \n",
       "VAOAD009           -0.566431                0.075375   \n",
       "OAD001             -0.600484                0.019657   \n",
       "VAOAD004            0.661963                0.245553   \n",
       "VAOAD011            0.598846                0.390035   \n",
       "\n",
       "          H_Monohydroxylated_bile_acid  H_Palmitoyl_ethanolamide  \\\n",
       "Study_ID                                                           \n",
       "VAOAD001                      0.132794                 -0.067464   \n",
       "VAOAD009                     -0.519189                 -0.180631   \n",
       "OAD001                       -0.576831                 -0.901479   \n",
       "VAOAD004                      0.279290                 -0.444319   \n",
       "VAOAD011                      0.287682                 -0.532141   \n",
       "\n",
       "          H_Linoleic_Acid  \n",
       "Study_ID                   \n",
       "VAOAD001        -0.112641  \n",
       "VAOAD009        -0.418351  \n",
       "OAD001          -1.089906  \n",
       "VAOAD004        -0.912164  \n",
       "VAOAD011        -0.752280  \n",
       "\n",
       "[5 rows x 88 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Metabolome testing\n",
    "gt_to_meta = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gt_to_meta[g] = {}\n",
    "    df_meta = pd.read_csv(path + 'inputs/' + g + '_normalized.csv')\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['Baseline','After diet']:\n",
    "        df = df_meta[df_meta['Time'] == t]\n",
    "        df = df.set_index('Study_ID')\n",
    "        df.index = df.index.map(lambda x: x.split('_')[0])\n",
    "        df = df.drop('Time',axis=1)\n",
    "        gt_to_meta[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_meta[g]['Baseline'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('D0','D-0'))\n",
    "    pre_df.index = pre_df.index.map(lambda x: x + '.pre.' + g)\n",
    "    post_df = gt_to_meta[g]['After diet'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('D0','D-0'))\n",
    "    post_df.index = post_df.index.map(lambda x: x + '.post.' + g)\n",
    "    \n",
    "    gt_to_meta[g]['diff'] = gt_to_meta[g]['After diet'] - gt_to_meta[g]['Baseline']\n",
    "    gt_to_meta[g]['all'] = pd.concat([pre_df,post_df])\n",
    "    \n",
    "\n",
    "gt_to_meta[g][t].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "801b5fe1-2f5b-465e-b21a-943c574d1ee7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Pathway DF Processing ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8b418d85-6211-4131-9007-92d2b5f1a3ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - globo and isoglobo series</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid elongation</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of 12-, 14- and 16-membered macrolides</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of enediyne antibiotics</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.saliva</th>\n",
       "      <td>0.155479</td>\n",
       "      <td>0.025863</td>\n",
       "      <td>0.001866</td>\n",
       "      <td>0.003177</td>\n",
       "      <td>0.008013</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.026699</td>\n",
       "      <td>0.000547</td>\n",
       "      <td>0.031793</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>2.433861e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.216931e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.saliva</th>\n",
       "      <td>0.159679</td>\n",
       "      <td>0.027060</td>\n",
       "      <td>0.001949</td>\n",
       "      <td>0.003082</td>\n",
       "      <td>0.007239</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.027470</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.034357</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>1.911111e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.539150e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.saliva</th>\n",
       "      <td>0.151338</td>\n",
       "      <td>0.027415</td>\n",
       "      <td>0.002223</td>\n",
       "      <td>0.002702</td>\n",
       "      <td>0.009135</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.021911</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.030358</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.saliva</th>\n",
       "      <td>0.152707</td>\n",
       "      <td>0.026974</td>\n",
       "      <td>0.002172</td>\n",
       "      <td>0.002788</td>\n",
       "      <td>0.008887</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.022846</td>\n",
       "      <td>0.000257</td>\n",
       "      <td>0.031226</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.post.saliva</th>\n",
       "      <td>0.159885</td>\n",
       "      <td>0.027003</td>\n",
       "      <td>0.001989</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>0.007319</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.026641</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.034499</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>6.002574e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.594595e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 168 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "KEGG_Pathway         Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.post.saliva    0.155479                            0.025863   \n",
       "OAD-001.pre.saliva     0.159679                            0.027060   \n",
       "OAD-003.post.saliva    0.151338                            0.027415   \n",
       "OAD-003.pre.saliva     0.152707                            0.026974   \n",
       "OAD-004.post.saliva    0.159885                            0.027003   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.post.saliva                                           0.001866                 \n",
       "OAD-001.pre.saliva                                            0.001949                 \n",
       "OAD-003.post.saliva                                           0.002223                 \n",
       "OAD-003.pre.saliva                                            0.002172                 \n",
       "OAD-004.post.saliva                                           0.001989                 \n",
       "\n",
       "KEGG_Pathway         Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.post.saliva                                           0.003177        \n",
       "OAD-001.pre.saliva                                            0.003082        \n",
       "OAD-003.post.saliva                                           0.002702        \n",
       "OAD-003.pre.saliva                                            0.002788        \n",
       "OAD-004.post.saliva                                           0.002953        \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism  \\\n",
       "OAD-001.post.saliva                     0.008013   \n",
       "OAD-001.pre.saliva                      0.007239   \n",
       "OAD-003.post.saliva                     0.009135   \n",
       "OAD-003.pre.saliva                      0.008887   \n",
       "OAD-004.post.saliva                     0.007319   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.post.saliva                                           0.000273    \n",
       "OAD-001.pre.saliva                                            0.000166    \n",
       "OAD-003.post.saliva                                           0.000180    \n",
       "OAD-003.pre.saliva                                            0.000176    \n",
       "OAD-004.post.saliva                                           0.000132    \n",
       "\n",
       "KEGG_Pathway         Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.post.saliva                          0.026699   \n",
       "OAD-001.pre.saliva                           0.027470   \n",
       "OAD-003.post.saliva                          0.021911   \n",
       "OAD-003.pre.saliva                           0.022846   \n",
       "OAD-004.post.saliva                          0.026641   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.post.saliva                                           0.000547      \n",
       "OAD-001.pre.saliva                                            0.000309      \n",
       "OAD-003.post.saliva                                           0.000261      \n",
       "OAD-003.pre.saliva                                            0.000257      \n",
       "OAD-004.post.saliva                                           0.000200      \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.post.saliva                                         0.031793   \n",
       "OAD-001.pre.saliva                                          0.034357   \n",
       "OAD-003.post.saliva                                         0.030358   \n",
       "OAD-003.pre.saliva                                          0.031226   \n",
       "OAD-004.post.saliva                                         0.034499   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.post.saliva                                           0.000046                    \n",
       "OAD-001.pre.saliva                                            0.000045                    \n",
       "OAD-003.post.saliva                                           0.000042                    \n",
       "OAD-003.pre.saliva                                            0.000041                    \n",
       "OAD-004.post.saliva                                           0.000038                    \n",
       "\n",
       "KEGG_Pathway         ...  \\\n",
       "OAD-001.post.saliva  ...   \n",
       "OAD-001.pre.saliva   ...   \n",
       "OAD-003.post.saliva  ...   \n",
       "OAD-003.pre.saliva   ...   \n",
       "OAD-004.post.saliva  ...   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - globo and isoglobo series  \\\n",
       "OAD-001.post.saliva                                           0.000021                                                          \n",
       "OAD-001.pre.saliva                                            0.000045                                                          \n",
       "OAD-003.post.saliva                                           0.000103                                                          \n",
       "OAD-003.pre.saliva                                            0.000080                                                          \n",
       "OAD-004.post.saliva                                           0.000062                                                          \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism|Fatty acid elongation  \\\n",
       "OAD-001.post.saliva                                                0.0   \n",
       "OAD-001.pre.saliva                                                 0.0   \n",
       "OAD-003.post.saliva                                                0.0   \n",
       "OAD-003.pre.saliva                                                 0.0   \n",
       "OAD-004.post.saliva                                                0.0   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis  \\\n",
       "OAD-001.post.saliva                                           0.000189                                    \n",
       "OAD-001.pre.saliva                                            0.000095                                    \n",
       "OAD-003.post.saliva                                           0.000022                                    \n",
       "OAD-003.pre.saliva                                            0.000018                                    \n",
       "OAD-004.post.saliva                                           0.000039                                    \n",
       "\n",
       "KEGG_Pathway         Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series  \\\n",
       "OAD-001.post.saliva                                           0.000010                                               \n",
       "OAD-001.pre.saliva                                            0.000019                                               \n",
       "OAD-003.post.saliva                                           0.000060                                               \n",
       "OAD-003.pre.saliva                                            0.000043                                               \n",
       "OAD-004.post.saliva                                           0.000033                                               \n",
       "\n",
       "KEGG_Pathway         Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.post.saliva                                       2.433861e-07                                   \n",
       "OAD-001.pre.saliva                                        1.911111e-06                                   \n",
       "OAD-003.post.saliva                                       0.000000e+00                                   \n",
       "OAD-003.pre.saliva                                        0.000000e+00                                   \n",
       "OAD-004.post.saliva                                       6.002574e-07                                   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.post.saliva                                                0.0                                  \n",
       "OAD-001.pre.saliva                                                 0.0                                  \n",
       "OAD-003.post.saliva                                                0.0                                  \n",
       "OAD-003.pre.saliva                                                 0.0                                  \n",
       "OAD-004.post.saliva                                                0.0                                  \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of 12-, 14- and 16-membered macrolides  \\\n",
       "OAD-001.post.saliva                                                0.0                                                         \n",
       "OAD-001.pre.saliva                                                 0.0                                                         \n",
       "OAD-003.post.saliva                                                0.0                                                         \n",
       "OAD-003.pre.saliva                                                 0.0                                                         \n",
       "OAD-004.post.saliva                                                0.0                                                         \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of enediyne antibiotics  \\\n",
       "OAD-001.post.saliva                                                0.0                                          \n",
       "OAD-001.pre.saliva                                                 0.0                                          \n",
       "OAD-003.post.saliva                                                0.0                                          \n",
       "OAD-003.pre.saliva                                                 0.0                                          \n",
       "OAD-004.post.saliva                                                0.0                                          \n",
       "\n",
       "KEGG_Pathway         Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.post.saliva                                       0.000000e+00                           \n",
       "OAD-001.pre.saliva                                        0.000000e+00                           \n",
       "OAD-003.post.saliva                                       0.000000e+00                           \n",
       "OAD-003.pre.saliva                                        0.000000e+00                           \n",
       "OAD-004.post.saliva                                       8.594595e-07                           \n",
       "\n",
       "KEGG_Pathway         Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \n",
       "OAD-001.post.saliva                                       1.216931e-07                                         \n",
       "OAD-001.pre.saliva                                        1.539150e-07                                         \n",
       "OAD-003.post.saliva                                       0.000000e+00                                         \n",
       "OAD-003.pre.saliva                                        0.000000e+00                                         \n",
       "OAD-004.post.saliva                                       0.000000e+00                                         \n",
       "\n",
       "[5 rows x 168 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specific pathway profiling\n",
    "g_to_path = {}\n",
    "\n",
    "groups = ['saliva_adh', 'stool_adh']\n",
    "g_to_job = {\n",
    "    'saliva_adh': 'jobs07',\n",
    "    'stool_adh': 'jobs08'\n",
    "}\n",
    "for g in groups:\n",
    "    # read in csv\n",
    "    df_path = pd.read_csv(path + 'outputs/' + g_to_job[g] + '/pred_metagenome_unstrat_kegg.tsv', sep ='\\t', index_col=0)\n",
    "\n",
    "    # transpose\n",
    "    df_path = df_path.T\n",
    "    \n",
    "    # normalize by dividing by row sum\n",
    "    df_path = df_path.div(df_path.sum(axis=1),axis=0)\n",
    "\n",
    "    # filter on metabolism only \n",
    "    keep = []\n",
    "    for p in df_path.columns:\n",
    "        if 'Metabolism' in p: # or 'metabolism' in p: 168 vs 170 if you incl both for saliva\n",
    "            keep.append(p)\n",
    "    df_path = df_path[keep]\n",
    "    \n",
    "    # save to dict\n",
    "    g_to_path[g] = df_path\n",
    "                                 \n",
    "g_to_path['saliva_adh'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "34ed8f12-5226-47f7-b748-07609e554b8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Degradation of flavonoids</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis</th>\n",
       "      <th>Metabolism|Lipid metabolism|Steroid hormone biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.159285</td>\n",
       "      <td>0.035029</td>\n",
       "      <td>0.002320</td>\n",
       "      <td>0.003018</td>\n",
       "      <td>0.007937</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>0.027626</td>\n",
       "      <td>0.000229</td>\n",
       "      <td>0.026349</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000469</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>3.060096e-06</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.510913e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.161049</td>\n",
       "      <td>0.033134</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.002538</td>\n",
       "      <td>0.008606</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.029092</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.027987</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>9.470469e-06</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.511851e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.161760</td>\n",
       "      <td>0.034016</td>\n",
       "      <td>0.002254</td>\n",
       "      <td>0.002773</td>\n",
       "      <td>0.008658</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.029185</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>0.027522</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>2.489384e-06</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.100744e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-005.pre.stool</th>\n",
       "      <td>0.158015</td>\n",
       "      <td>0.033087</td>\n",
       "      <td>0.002185</td>\n",
       "      <td>0.002702</td>\n",
       "      <td>0.008593</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.028749</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>0.027538</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000405</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>2.762671e-06</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>1.103744e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.297970e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-006.pre.stool</th>\n",
       "      <td>0.157791</td>\n",
       "      <td>0.032210</td>\n",
       "      <td>0.002112</td>\n",
       "      <td>0.002819</td>\n",
       "      <td>0.008727</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.030094</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.026678</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000329</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>2.184941e-07</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.127712e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 159 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "KEGG_Pathway       Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.pre.stool    0.159285                            0.035029   \n",
       "OAD-003.pre.stool    0.161049                            0.033134   \n",
       "OAD-004.pre.stool    0.161760                            0.034016   \n",
       "OAD-005.pre.stool    0.158015                            0.033087   \n",
       "OAD-006.pre.stool    0.157791                            0.032210   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.pre.stool                                           0.002320                 \n",
       "OAD-003.pre.stool                                           0.002184                 \n",
       "OAD-004.pre.stool                                           0.002254                 \n",
       "OAD-005.pre.stool                                           0.002185                 \n",
       "OAD-006.pre.stool                                           0.002112                 \n",
       "\n",
       "KEGG_Pathway       Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.pre.stool                                           0.003018        \n",
       "OAD-003.pre.stool                                           0.002538        \n",
       "OAD-004.pre.stool                                           0.002773        \n",
       "OAD-005.pre.stool                                           0.002702        \n",
       "OAD-006.pre.stool                                           0.002819        \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism  \\\n",
       "OAD-001.pre.stool                     0.007937   \n",
       "OAD-003.pre.stool                     0.008606   \n",
       "OAD-004.pre.stool                     0.008658   \n",
       "OAD-005.pre.stool                     0.008593   \n",
       "OAD-006.pre.stool                     0.008727   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.pre.stool                                           0.000192    \n",
       "OAD-003.pre.stool                                           0.000180    \n",
       "OAD-004.pre.stool                                           0.000157    \n",
       "OAD-005.pre.stool                                           0.000186    \n",
       "OAD-006.pre.stool                                           0.000199    \n",
       "\n",
       "KEGG_Pathway       Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.pre.stool                          0.027626   \n",
       "OAD-003.pre.stool                          0.029092   \n",
       "OAD-004.pre.stool                          0.029185   \n",
       "OAD-005.pre.stool                          0.028749   \n",
       "OAD-006.pre.stool                          0.030094   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.pre.stool                                           0.000229      \n",
       "OAD-003.pre.stool                                           0.000194      \n",
       "OAD-004.pre.stool                                           0.000173      \n",
       "OAD-005.pre.stool                                           0.000216      \n",
       "OAD-006.pre.stool                                           0.000220      \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.pre.stool                                         0.026349   \n",
       "OAD-003.pre.stool                                         0.027987   \n",
       "OAD-004.pre.stool                                         0.027522   \n",
       "OAD-005.pre.stool                                         0.027538   \n",
       "OAD-006.pre.stool                                         0.026678   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.pre.stool                                           0.000012                    \n",
       "OAD-003.pre.stool                                           0.000019                    \n",
       "OAD-004.pre.stool                                           0.000016                    \n",
       "OAD-005.pre.stool                                           0.000016                    \n",
       "OAD-006.pre.stool                                           0.000013                    \n",
       "\n",
       "KEGG_Pathway       ...  \\\n",
       "OAD-001.pre.stool  ...   \n",
       "OAD-003.pre.stool  ...   \n",
       "OAD-004.pre.stool  ...   \n",
       "OAD-005.pre.stool  ...   \n",
       "OAD-006.pre.stool  ...   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Degradation of flavonoids  \\\n",
       "OAD-001.pre.stool                                           0.000469                                  \n",
       "OAD-003.pre.stool                                           0.000482                                  \n",
       "OAD-004.pre.stool                                           0.000635                                  \n",
       "OAD-005.pre.stool                                           0.000405                                  \n",
       "OAD-006.pre.stool                                           0.000329                                  \n",
       "\n",
       "KEGG_Pathway       Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis  \\\n",
       "OAD-001.pre.stool                                           0.000028                                      \n",
       "OAD-003.pre.stool                                           0.000090                                      \n",
       "OAD-004.pre.stool                                           0.000103                                      \n",
       "OAD-005.pre.stool                                           0.000055                                      \n",
       "OAD-006.pre.stool                                           0.000087                                      \n",
       "\n",
       "KEGG_Pathway       Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis  \\\n",
       "OAD-001.pre.stool                                       3.060096e-06                                    \n",
       "OAD-003.pre.stool                                       9.470469e-06                                    \n",
       "OAD-004.pre.stool                                       2.489384e-06                                    \n",
       "OAD-005.pre.stool                                       2.762671e-06                                    \n",
       "OAD-006.pre.stool                                       2.184941e-07                                    \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism|Steroid hormone biosynthesis  \\\n",
       "OAD-001.pre.stool                                           0.000026          \n",
       "OAD-003.pre.stool                                           0.000118          \n",
       "OAD-004.pre.stool                                           0.000108          \n",
       "OAD-005.pre.stool                                           0.000092          \n",
       "OAD-006.pre.stool                                           0.000088          \n",
       "\n",
       "KEGG_Pathway       Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series  \\\n",
       "OAD-001.pre.stool                                           0.000026                                               \n",
       "OAD-003.pre.stool                                           0.000090                                               \n",
       "OAD-004.pre.stool                                           0.000103                                               \n",
       "OAD-005.pre.stool                                           0.000055                                               \n",
       "OAD-006.pre.stool                                           0.000086                                               \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.pre.stool                                       0.000000e+00                                   \n",
       "OAD-003.pre.stool                                       0.000000e+00                                   \n",
       "OAD-004.pre.stool                                       0.000000e+00                                   \n",
       "OAD-005.pre.stool                                       1.103744e-07                                   \n",
       "OAD-006.pre.stool                                       0.000000e+00                                   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.pre.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                0.0                                  \n",
       "OAD-004.pre.stool                                                0.0                                  \n",
       "OAD-005.pre.stool                                                0.0                                  \n",
       "OAD-006.pre.stool                                                0.0                                  \n",
       "\n",
       "KEGG_Pathway       Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.pre.stool                                       1.510913e-06                           \n",
       "OAD-003.pre.stool                                       8.511851e-07                           \n",
       "OAD-004.pre.stool                                       6.100744e-07                           \n",
       "OAD-005.pre.stool                                       5.297970e-07                           \n",
       "OAD-006.pre.stool                                       1.127712e-07                           \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis  \\\n",
       "OAD-001.pre.stool                                                0.0                               \n",
       "OAD-003.pre.stool                                                0.0                               \n",
       "OAD-004.pre.stool                                                0.0                               \n",
       "OAD-005.pre.stool                                                0.0                               \n",
       "OAD-006.pre.stool                                                0.0                               \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \n",
       "OAD-001.pre.stool                                                0.0                                         \n",
       "OAD-003.pre.stool                                                0.0                                         \n",
       "OAD-004.pre.stool                                                0.0                                         \n",
       "OAD-005.pre.stool                                                0.0                                         \n",
       "OAD-006.pre.stool                                                0.0                                         \n",
       "\n",
       "[5 rows x 159 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pathway testing\n",
    "gt_to_path = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gt_to_path[g] = {}\n",
    "    df_path = g_to_path[g + '_adh'].copy()\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['post','pre']:\n",
    "        df = df_path[df_path.index.to_series().str.contains(t)]\n",
    "        gt_to_path[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_path[g]['pre'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.split('.pre')[0])\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('-',''))\n",
    "    post_df = gt_to_path[g]['post'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.split('.post')[0])\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('-',''))\n",
    "    \n",
    "    gt_to_path[g]['diff'] = post_df - pre_df\n",
    "    gt_to_path[g]['all'] = pd.concat([gt_to_path[g]['pre'],gt_to_path[g]['post']])\n",
    "\n",
    "gt_to_path[g][t].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81bd68ab-5702-4758-ad7a-34bf139443ea",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Diet DF Processing ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "77c18bcf-830a-458d-88c1-70f509e7c9c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dietary variables\n",
    "diet = ['animal_proteins_red',\n",
    "       'refined_grains', 'solanaceae', 'saturated', 'pastries',\n",
    "       'milk_derivates', 'beverages_prohibited', 'pro_inflammatory',\n",
    "       'chicken', 'plant_protein', 'whole_grains_pseudocereals_oats',\n",
    "       'gluten_free', 'vegetables', 'cruciferous',\n",
    "       'berries_and_enzymatic_fruit', 'other_fruit', 'fish', 'milk_subs',\n",
    "       'PUFA', 'MUFA', 'Walnuts', 'Nuts', 'probiotics', 'Tahini_avo',\n",
    "       'green_tea', 'anti_inflam_spices', 'inflammatory_spices',\n",
    "       'inflammatory', 'anti_inflammatory', 'total_antiinflammatory',\n",
    "       'Adherece_antiinflam', 'oats', 'almond_milk', 'fatty_fish',\n",
    "       'sesame_seeds', 'egg', 'chia', 'flaxseed_oil', 'walnuts',\n",
    "       'spinach', 'corn_tortilla', 'tahini', 'avocado', 'chicken.1',\n",
    "       'brown_rice', 'corn', 'quinoa', 'broccoli', 'Garbanzo_beans',\n",
    "       'pork', 'beef', 'burger', 'Total_omega3', 'Adherence_omega3',\n",
    "       'Total_omega6', 'Adherence_omega6', 'Total_o3_o6']\n",
    "\n",
    "for d in diet:\n",
    "    if 'Adh' in d:\n",
    "        diet.remove(d)\n",
    "\n",
    "type_to_df_diet = {}\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    df_map = type_to_df_map[g].reset_index().drop_duplicates(subset='#SampleID').set_index('#SampleID')\n",
    "    df_diet = df_map[diet].dropna()\n",
    "    type_to_df_diet[g] = df_diet\n",
    "\n",
    "    from sklearn.decomposition import PCA\n",
    "    npcs = 3\n",
    "    pca = PCA(n_components=3, whiten=True)\n",
    "    pca.fit(df_diet.T)\n",
    "    # Reformat and view results\n",
    "    loadings = pd.DataFrame(pca.components_,\n",
    "                            index=['PC%s' % _ for _ in range(npcs)], \n",
    "                            columns=df_diet.index).T\n",
    "    \n",
    "    df_merge = pd.concat([df_map, df_diet, loadings], axis=1)\n",
    "    plt.title(g + '_Diet_PCA')\n",
    "    sns.scatterplot(data=df_merge, x='PC1', y='PC2', hue='WOMAC_P_Response')\n",
    "    sns.despine()\n",
    "    plt.savefig(path + 'outputs/jobs06/PCA_diet_' + g + '.pdf')\n",
    "    plt.close() \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e22033b-a7a9-406a-8cb1-c1074ae78087",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Complete DA Analysis ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "856e9286-7a26-4639-a599-53c699043e5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taxa\n",
      "pvp\n",
      "saliva\n",
      "Number of sig var\n",
      "2\n",
      "Number of sig var\n",
      "5\n",
      "stool\n",
      "Number of sig var\n",
      "6\n",
      "Number of sig var\n",
      "4\n",
      "rvnr\n",
      "saliva\n",
      "Number of sig var\n",
      "9\n",
      "Number of sig var\n",
      "0\n",
      "stool\n",
      "Number of sig var\n",
      "6\n",
      "Number of sig var\n",
      "0\n",
      "path\n",
      "pvp\n",
      "saliva\n",
      "Number of sig var\n",
      "3\n",
      "Number of sig var\n",
      "5\n",
      "stool\n",
      "Number of sig var\n",
      "4\n",
      "Number of sig var\n",
      "6\n",
      "rvnr\n",
      "saliva\n",
      "Number of sig var\n",
      "22\n",
      "Number of sig var\n",
      "0\n",
      "stool\n",
      "Number of sig var\n",
      "11\n",
      "Number of sig var\n",
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis</th>\n",
       "      <th>Metabolism|Lipid metabolism|Steroid hormone biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "      <th>Timepoint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.159285</td>\n",
       "      <td>0.035029</td>\n",
       "      <td>0.002320</td>\n",
       "      <td>0.003018</td>\n",
       "      <td>0.007937</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>0.027626</td>\n",
       "      <td>0.000229</td>\n",
       "      <td>0.026349</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>3.060096e-06</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.510913e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.161049</td>\n",
       "      <td>0.033134</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.002538</td>\n",
       "      <td>0.008606</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.029092</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.027987</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>9.470469e-06</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.511851e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.161760</td>\n",
       "      <td>0.034016</td>\n",
       "      <td>0.002254</td>\n",
       "      <td>0.002773</td>\n",
       "      <td>0.008658</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.029185</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>0.027522</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>2.489384e-06</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.100744e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-001.pre.stool</th>\n",
       "      <td>0.163414</td>\n",
       "      <td>0.036621</td>\n",
       "      <td>0.002214</td>\n",
       "      <td>0.002779</td>\n",
       "      <td>0.008978</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>0.028735</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>8.806406e-06</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.026500e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.425694e-07</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-012.pre.stool</th>\n",
       "      <td>0.164662</td>\n",
       "      <td>0.036528</td>\n",
       "      <td>0.002363</td>\n",
       "      <td>0.002908</td>\n",
       "      <td>0.009032</td>\n",
       "      <td>0.000170</td>\n",
       "      <td>0.028994</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.027414</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>1.871888e-07</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.499616e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 160 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.pre.stool      0.159285                            0.035029   \n",
       "OAD-003.pre.stool      0.161049                            0.033134   \n",
       "OAD-004.pre.stool      0.161760                            0.034016   \n",
       "VAOAD-001.pre.stool    0.163414                            0.036621   \n",
       "VAOAD-012.pre.stool    0.164662                            0.036528   \n",
       "\n",
       "                     Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.pre.stool                                             0.002320                 \n",
       "OAD-003.pre.stool                                             0.002184                 \n",
       "OAD-004.pre.stool                                             0.002254                 \n",
       "VAOAD-001.pre.stool                                           0.002214                 \n",
       "VAOAD-012.pre.stool                                           0.002363                 \n",
       "\n",
       "                     Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.pre.stool                                             0.003018        \n",
       "OAD-003.pre.stool                                             0.002538        \n",
       "OAD-004.pre.stool                                             0.002773        \n",
       "VAOAD-001.pre.stool                                           0.002779        \n",
       "VAOAD-012.pre.stool                                           0.002908        \n",
       "\n",
       "                     Metabolism|Lipid metabolism  \\\n",
       "OAD-001.pre.stool                       0.007937   \n",
       "OAD-003.pre.stool                       0.008606   \n",
       "OAD-004.pre.stool                       0.008658   \n",
       "VAOAD-001.pre.stool                     0.008978   \n",
       "VAOAD-012.pre.stool                     0.009032   \n",
       "\n",
       "                     Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.pre.stool                                             0.000192    \n",
       "OAD-003.pre.stool                                             0.000180    \n",
       "OAD-004.pre.stool                                             0.000157    \n",
       "VAOAD-001.pre.stool                                           0.000195    \n",
       "VAOAD-012.pre.stool                                           0.000170    \n",
       "\n",
       "                     Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.pre.stool                            0.027626   \n",
       "OAD-003.pre.stool                            0.029092   \n",
       "OAD-004.pre.stool                            0.029185   \n",
       "VAOAD-001.pre.stool                          0.028735   \n",
       "VAOAD-012.pre.stool                          0.028994   \n",
       "\n",
       "                     Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.pre.stool                                             0.000229      \n",
       "OAD-003.pre.stool                                             0.000194      \n",
       "OAD-004.pre.stool                                             0.000173      \n",
       "VAOAD-001.pre.stool                                           0.000203      \n",
       "VAOAD-012.pre.stool                                           0.000194      \n",
       "\n",
       "                     Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.pre.stool                                           0.026349   \n",
       "OAD-003.pre.stool                                           0.027987   \n",
       "OAD-004.pre.stool                                           0.027522   \n",
       "VAOAD-001.pre.stool                                         0.026849   \n",
       "VAOAD-012.pre.stool                                         0.027414   \n",
       "\n",
       "                     Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.pre.stool                                             0.000012                    \n",
       "OAD-003.pre.stool                                             0.000019                    \n",
       "OAD-004.pre.stool                                             0.000016                    \n",
       "VAOAD-001.pre.stool                                           0.000027                    \n",
       "VAOAD-012.pre.stool                                           0.000013                    \n",
       "\n",
       "                     ...  \\\n",
       "OAD-001.pre.stool    ...   \n",
       "OAD-003.pre.stool    ...   \n",
       "OAD-004.pre.stool    ...   \n",
       "VAOAD-001.pre.stool  ...   \n",
       "VAOAD-012.pre.stool  ...   \n",
       "\n",
       "                     Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis  \\\n",
       "OAD-001.pre.stool                                             0.000028                                      \n",
       "OAD-003.pre.stool                                             0.000090                                      \n",
       "OAD-004.pre.stool                                             0.000103                                      \n",
       "VAOAD-001.pre.stool                                           0.000102                                      \n",
       "VAOAD-012.pre.stool                                           0.000106                                      \n",
       "\n",
       "                     Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis  \\\n",
       "OAD-001.pre.stool                                         3.060096e-06                                    \n",
       "OAD-003.pre.stool                                         9.470469e-06                                    \n",
       "OAD-004.pre.stool                                         2.489384e-06                                    \n",
       "VAOAD-001.pre.stool                                       8.806406e-06                                    \n",
       "VAOAD-012.pre.stool                                       1.871888e-07                                    \n",
       "\n",
       "                     Metabolism|Lipid metabolism|Steroid hormone biosynthesis  \\\n",
       "OAD-001.pre.stool                                             0.000026          \n",
       "OAD-003.pre.stool                                             0.000118          \n",
       "OAD-004.pre.stool                                             0.000108          \n",
       "VAOAD-001.pre.stool                                           0.000105          \n",
       "VAOAD-012.pre.stool                                           0.000071          \n",
       "\n",
       "                     Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series  \\\n",
       "OAD-001.pre.stool                                             0.000026                                               \n",
       "OAD-003.pre.stool                                             0.000090                                               \n",
       "OAD-004.pre.stool                                             0.000103                                               \n",
       "VAOAD-001.pre.stool                                           0.000102                                               \n",
       "VAOAD-012.pre.stool                                           0.000106                                               \n",
       "\n",
       "                     Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.pre.stool                                                  0.0                                   \n",
       "OAD-003.pre.stool                                                  0.0                                   \n",
       "OAD-004.pre.stool                                                  0.0                                   \n",
       "VAOAD-001.pre.stool                                                0.0                                   \n",
       "VAOAD-012.pre.stool                                                0.0                                   \n",
       "\n",
       "                     Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.pre.stool                                                  0.0                                  \n",
       "OAD-003.pre.stool                                                  0.0                                  \n",
       "OAD-004.pre.stool                                                  0.0                                  \n",
       "VAOAD-001.pre.stool                                                0.0                                  \n",
       "VAOAD-012.pre.stool                                                0.0                                  \n",
       "\n",
       "                     Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.pre.stool                                         1.510913e-06                           \n",
       "OAD-003.pre.stool                                         8.511851e-07                           \n",
       "OAD-004.pre.stool                                         6.100744e-07                           \n",
       "VAOAD-001.pre.stool                                       1.026500e-06                           \n",
       "VAOAD-012.pre.stool                                       3.499616e-07                           \n",
       "\n",
       "                     Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis  \\\n",
       "OAD-001.pre.stool                                                  0.0                               \n",
       "OAD-003.pre.stool                                                  0.0                               \n",
       "OAD-004.pre.stool                                                  0.0                               \n",
       "VAOAD-001.pre.stool                                                0.0                               \n",
       "VAOAD-012.pre.stool                                                0.0                               \n",
       "\n",
       "                     Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \\\n",
       "OAD-001.pre.stool                                         0.000000e+00                                          \n",
       "OAD-003.pre.stool                                         0.000000e+00                                          \n",
       "OAD-004.pre.stool                                         0.000000e+00                                          \n",
       "VAOAD-001.pre.stool                                       1.425694e-07                                          \n",
       "VAOAD-012.pre.stool                                       0.000000e+00                                          \n",
       "\n",
       "                     Timepoint  \n",
       "OAD-001.pre.stool          pre  \n",
       "OAD-003.pre.stool          pre  \n",
       "OAD-004.pre.stool          pre  \n",
       "VAOAD-001.pre.stool        pre  \n",
       "VAOAD-012.pre.stool        pre  \n",
       "\n",
       "[5 rows x 160 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# complete block for paired testing in r v nr and non paired testing pre and post\n",
    "\n",
    "dcgr_to_sig = {}\n",
    "\n",
    "for dt in ['taxa', 'path']:\n",
    "    print(dt)\n",
    "    dcgr_to_sig[dt] = {}\n",
    "    for c in ['pvp', 'rvnr']: # pvp is pre AND post, comparing rvnr at pre and post time points\n",
    "        print(c)\n",
    "        dcgr_to_sig[dt][c] = {}        \n",
    "    \n",
    "        # define df arrays\n",
    "        xs = []\n",
    "        gs = []\n",
    "        ws = []\n",
    "        ps = []\n",
    "        os = []\n",
    "        dirs = []\n",
    "        adj_rs = []\n",
    "        \n",
    "        # define analysis specific params\n",
    "        if c == 'pvp':\n",
    "            outer_loop = ['pre', 'post']\n",
    "            comp = ['Response', 'No response']\n",
    "            var = 'WOMAC_P_Response'\n",
    "            drop = 'Timepoint'\n",
    "        elif c == 'rvnr': # pvp is pre and post, comparing rvnr at pre and posttime points\n",
    "            outer_loop = ['Response', 'No response']\n",
    "            comp = ['pre', 'post']\n",
    "            var = 'Timepoint'\n",
    "            drop = 'WOMAC_P_Response'\n",
    "\n",
    "        # for each sample type\n",
    "        for g in ['saliva','stool']:\n",
    "            print(g)\n",
    "            dcgr_to_sig[dt][c][g] = {}\n",
    "            \n",
    "            # grab relevant table table\n",
    "            if dt == 'taxa':\n",
    "                df_abun = gt_to_otu[g]['all']\n",
    "            elif dt == 'path':\n",
    "                df_abun = gt_to_path[g]['all']\n",
    "\n",
    "            df_map = type_to_df_map[g].copy()\n",
    "            df_map = df_map.set_index('Together')\n",
    "            if g == 'stool':\n",
    "                df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "            df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "\n",
    "            for r in outer_loop:\n",
    "                df_merge = pd.concat([df_abun, df_map[df_map[drop] == r]], axis=1)\n",
    "                df_merge = df_merge.dropna()\n",
    "                df_merge['SampleID'] = df_merge.index.map(lambda x: x.split('.p')[0])\n",
    "                if c == 'rvnr':\n",
    "                    df_vc = df_merge['SampleID'].value_counts()\n",
    "                    keep = df_vc[df_vc == 2].index.values\n",
    "                    df_merge = df_merge[df_merge['SampleID'].isin(keep)] \n",
    "                df_merge = df_merge.drop(['SampleID',drop],axis=1)\n",
    "                sig = []\n",
    "                for x in list(df_merge.columns.values)[:-1]:\n",
    "                    df_a = df_merge[df_merge[var] == comp[0]][x]\n",
    "                    df_b = df_merge[df_merge[var] == comp[1]][x]\n",
    "\n",
    "                    try:\n",
    "                        if c == 'rvnr':\n",
    "                            w,p = scipy.stats.wilcoxon(df_a.values, df_b.values, nan_policy='omit')\n",
    "                        elif c == 'pvp':\n",
    "                            w,p = scipy.stats.mannwhitneyu(df_a.values, df_b.values, nan_policy='omit')\n",
    "\n",
    "                        a = df_a.values\n",
    "                        b = df_b.values\n",
    "                        # if median in A is greater\n",
    "                        abr, abp = scipy.stats.mannwhitneyu(a,b,nan_policy='omit')\n",
    "                        # if median in B is greater\n",
    "                        bar, bap = scipy.stats.mannwhitneyu(b,a,nan_policy='omit')\n",
    "                        if abr > bar:\n",
    "                            dirs.append(comp[0])\n",
    "                            adj_rs.append(abr)\n",
    "                        elif abr < bar:\n",
    "                            dirs.append(comp[1])\n",
    "                            adj_rs.append(bar * -1)\n",
    "                        else:\n",
    "                            dirs.append('same')\n",
    "                            adj_rs.append(abr)\n",
    "\n",
    "                    except:\n",
    "                        w,p = 0, 1\n",
    "                        dirs.append(np.nan)\n",
    "                        adj_rs.append(np.nan)\n",
    "\n",
    "                    os.append(r)\n",
    "                    xs.append(x)\n",
    "                    gs.append(g)\n",
    "                    ws.append(w)\n",
    "                    ps.append(p)\n",
    "\n",
    "                    if p < 0.05:\n",
    "                        ax = sns.boxplot(data=df_merge, y=x, x=var)\n",
    "                        sns.swarmplot(data=df_merge, y=x, x=var, palette='dark:grey')\n",
    "                        sns.despine()\n",
    "                    \n",
    "                        plt.tight_layout()\n",
    "                        plt.savefig(path + 'outputs/jobs06/' + c + '_' + dt + '_barplot_' + g + '_' + r + '_' + x.split('g__')[-1] + '.pdf')\n",
    "                        plt.close() \n",
    "                        sig.append(x)\n",
    "                print('Number of sig var')\n",
    "                print(len(sig))\n",
    "                dcgr_to_sig[dt][c][g][r] = sig\n",
    "                \n",
    "        df_res = pd.DataFrame({'SampleType':gs, # saliva v stool\n",
    "                      'fixed': os,\n",
    "                      'var': xs, # example pathway or taxa\n",
    "                      'direction': dirs, # direction of enrichment\n",
    "                      'Statistic': ws, \n",
    "                      'pvalue': ps,\n",
    "                      'adjusted': adj_rs                      \n",
    "                     })\n",
    "        df_res[df_res['pvalue'] < 0.05].to_csv(path + 'outputs/jobs06/' + dt + '_' + c + '.tsv', sep='\\t')\n",
    "df_merge.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "43bca09e-c560-43cf-975b-7b58541ae974",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "meta\n",
      "otu\n",
      "no plot for :stool_otu_['No response']\n",
      "path\n",
      "no plot for :stool_path_['No response']\n",
      "saliva\n",
      "meta\n",
      "no plot for :saliva_meta_['No response']\n",
      "otu\n",
      "no plot for :saliva_otu_['No response']\n",
      "path\n",
      "no plot for :saliva_path_['No response']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>path</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>adj_test_stat</th>\n",
       "      <th>pvals</th>\n",
       "      <th>direction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [path, test_stat, adj_test_stat, pvals, direction]\n",
       "Index: []"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# horizontal barplot for features\n",
    "\n",
    "d_to_df = {\n",
    "    'meta': gt_to_meta,\n",
    "    'otu': gt_to_otu,\n",
    "    'path': gt_to_path\n",
    "}\n",
    "d_to_label = {\n",
    "    'meta': 'Metabolite',\n",
    "    'otu': 'Taxa',\n",
    "    'path': 'Pathway'\n",
    "}\n",
    "t = 'all'\n",
    "\n",
    "gd_to_res = {}\n",
    "for g in ['stool','saliva']:\n",
    "    print(g)\n",
    "    gd_to_res[g] = {}\n",
    "    for d in ['meta','otu','path']:\n",
    "        print(d)\n",
    "        # grab original dfs\n",
    "        df_feat = d_to_df[d][g][t]\n",
    "\n",
    "        df_map = type_to_df_map[g].set_index('Together')\n",
    "        df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "        \n",
    "        df_map = df_map[['Timepoint','WOMAC_P_Response','HostSubjectId']]\n",
    "        \n",
    "        # merge df\n",
    "        df_merge = pd.concat([df_feat, df_map],axis=1)\n",
    "        df_merge = df_merge.dropna()\n",
    "        \n",
    "        # do paired test\n",
    "        # grab all sample IDs that are duplicated\n",
    "        df_vc = df_merge['HostSubjectId'].value_counts() == 2\n",
    "        df_vc = df_vc[df_vc == True]\n",
    "        keep_id = df_vc.index.values\n",
    "        \n",
    "        df_merge_complete = df_merge[df_merge['HostSubjectId'].isin(keep_id)]\n",
    "\n",
    "        for keep in [['Response','No response'],['Response'],['No response']]:\n",
    "            df_merge = df_merge_complete[df_merge_complete['WOMAC_P_Response'].isin(keep)]\n",
    "        \n",
    "            # samples are in order when subsetted on Timepoint\n",
    "            df_pre = df_merge[df_merge['Timepoint'] == 'pre'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "            df_post = df_merge[df_merge['Timepoint'] == 'post'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "            \n",
    "            # compute rank sum\n",
    "            feat = []\n",
    "            rs = []\n",
    "            adj_rs = []\n",
    "            ps = []\n",
    "            dir = []\n",
    "            for x in df_pre.columns.values:\n",
    "                try:\n",
    "                    r, p = scipy.stats.wilcoxon(df_post[x].values-df_pre[x].values,nan_policy='omit')\n",
    "                except ValueError:\n",
    "                    r, p = 0, 1\n",
    "                if d == 'otu':\n",
    "                    feat.append(x.split('__')[-1])\n",
    "                elif d == 'path':\n",
    "                    feat.append(x.split('|')[-1])\n",
    "                else:\n",
    "                    feat.append(x)\n",
    "                # rs.append(r)\n",
    "                ps.append(p)\n",
    "                # determine direction via MWU\n",
    "                a = df_post[x].values\n",
    "                b = df_pre[x].values\n",
    "                abr, abp = scipy.stats.mannwhitneyu(a,b,nan_policy='omit')\n",
    "                bar, bap = scipy.stats.mannwhitneyu(b,a,nan_policy='omit')\n",
    "                if abr > bar:\n",
    "                    dir.append('post')\n",
    "                    adj_rs.append(abr)\n",
    "                    rs.append(bar)\n",
    "                elif abr < bar:\n",
    "                    dir.append('pre')\n",
    "                    adj_rs.append(bar * -1)\n",
    "                    rs.append(bar)\n",
    "                else:\n",
    "                    dir.append('same')\n",
    "                    adj_rs.append(abr)\n",
    "                    rs.append(abr)\n",
    "            # compile results df\n",
    "            df_results = pd.DataFrame({d: feat, 'test_stat': rs, 'adj_test_stat': adj_rs,'pvals': ps, 'direction': dir})\n",
    "            gd_to_res[g][d] = df_results\n",
    "            df = df_results[df_results['pvals'] < 0.05]\n",
    "            df = df.sort_values(by=['adj_test_stat'],ascending=False)\n",
    "            # plot\n",
    "            # create plot\n",
    "            try:\n",
    "                plt.figure(figsize=(8,6))\n",
    "                sns.set_theme(style=\"whitegrid\")\n",
    "                \n",
    "                sns.set_color_codes(\"pastel\")\n",
    "                sns.barplot(x='adj_test_stat', y=d, data=df,\n",
    "                            hue=\"direction\",orient='horizontal')#,native_scale=True)#width=1)\n",
    "                \n",
    "                ax.set_ylabel(d_to_label[d],fontsize=16)\n",
    "                ax.set_xlabel(\"Effect Size / Adjusted Test Statistic\",fontsize=16)\n",
    "                ax.tick_params(labelsize=16)\n",
    "                sns.despine(left=True, bottom=True)\n",
    "                plt.tight_layout()\n",
    "                plt.title(g + '_' + d + '_' + str(keep))\n",
    "                plt.savefig(path + 'outputs/jobs18/DA_hbarplot_' + g + '_' + d + '_' + str(keep) + '.pdf')\n",
    "                plt.close()\n",
    "                df.to_csv(path + 'outputs/jobs18/' + g + '_' + d + '_' + str(keep) + '.tsv', sep='\\t')\n",
    "            \n",
    "            except: \n",
    "                print('no plot for :' + g + '_' + d + '_' + str(keep))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70c6b863-4344-44ac-9ac9-949de2808948",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Mediation ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "90768c43-3b95-494b-93d3-46d7b1a4222e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X</th>\n",
       "      <th>M</th>\n",
       "      <th>Y</th>\n",
       "      <th>split_value</th>\n",
       "      <th>total_effect</th>\n",
       "      <th>total_effect_p</th>\n",
       "      <th>total_effect_ci_lower</th>\n",
       "      <th>total_effect_ci_upper</th>\n",
       "      <th>direct_effect</th>\n",
       "      <th>direct_effect_p</th>\n",
       "      <th>...</th>\n",
       "      <th>b2_effect_ci_lower</th>\n",
       "      <th>b2_effect_ci_upper</th>\n",
       "      <th>b3_effect</th>\n",
       "      <th>b3_effect_p</th>\n",
       "      <th>b3_effect_ci_lower</th>\n",
       "      <th>b3_effect_ci_upper</th>\n",
       "      <th>X_sparsity</th>\n",
       "      <th>M_sparsity</th>\n",
       "      <th>Y_sparsity</th>\n",
       "      <th>n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [X, M, Y, split_value, total_effect, total_effect_p, total_effect_ci_lower, total_effect_ci_upper, direct_effect, direct_effect_p, direct_effect_ci_lower, direct_effect_ci_upper, indirect_effect, indirect_effect_p, indirect_effect_ci_lower, indirect_effect_ci_upper, prop_mediated, prop_mediated_p, prop_mediated_ci_lower, prop_mediated_ci_upper, b2_effect, b2_effect_p, b2_effect_ci_lower, b2_effect_ci_upper, b3_effect, b3_effect_p, b3_effect_ci_lower, b3_effect_ci_upper, X_sparsity, M_sparsity, Y_sparsity, n]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 32 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Mediation analyses\n",
    "# check for int_ba\n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid'] + ['Deoxycholic_acid','Lithocholic_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid', 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide']+ ['Deoxycholic_acid','Lithocholic_acid']\n",
    "         }\n",
    "\n",
    "# filter on CUTIE results\n",
    "g_to_cutiejob = {\n",
    "    'saliva': 'jobs15',\n",
    "    'stool': 'jobs16'\n",
    "}\n",
    "\n",
    "g_to_medjob = {\n",
    "    'saliva': '20240917',\n",
    "    'stool': '20240913'\n",
    "}\n",
    "# figure out cutie plots\n",
    "#X = list(df_X.columns.values)\n",
    "#M = list(df_M.columns.values)\n",
    "#x,m = df.iloc[0,0:2].values\n",
    "#X.index(x), M.index(m)\n",
    "\n",
    "# define outcome\n",
    "y = 'WOMAC_pain_diff'\n",
    "\n",
    "\n",
    "\n",
    "for g in ['stool', 'saliva']:\n",
    "    # import mediation DFs\n",
    "    df_M = pd.read_csv(path + 'inputs/cutie_df_meta_' + g + '.tsv', sep='\\t', index_col=0)\n",
    "    df_M.iloc[:,:-1].to_csv(path + 'inputs/mediation_meta_' + g + '_post_test.txt', sep='\\t')\n",
    "    \n",
    "    df_Y = df_meta_paired[['VAS_Pt_diff', 'WOMAC_pain_diff']]\n",
    "    df_Y.to_csv(path + 'inputs/mediation_WMdiff_test.txt', sep='\\t')\n",
    "    \n",
    "    df_X = pd.read_csv(path + 'inputs/cutie_df_taxa_' + g + '.tsv', sep='\\t', index_col=0)\n",
    "    df_X.index = df_X.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    df_X.to_csv(path + 'inputs/mediation_taxa_' + g + '_post_test.txt', sep='\\t')\n",
    "\n",
    "    # import mediation results\n",
    "    df_med = pd.read_csv(path + 'outputs/jobs19/' + g_to_medjob[g] + '/outputs/mediation_taxa_' + g + '_post_mediation_meta_' + g + '_post_mediation_WMdiff_test/mediation_results_all.tsv', \n",
    "                         sep='\\t')\n",
    "\n",
    "    for t in ['partial','complete']:\n",
    "        # copy df\n",
    "        df = df_med.copy()\n",
    "        \n",
    "        # filter on significant total\n",
    "        df = df[df['total_effect_p'] < 0.05]\n",
    "\n",
    "        if t == 'complete':\n",
    "            # find complete mediations\n",
    "            df = df[df['direct_effect_p'] > 0.05]\n",
    "\n",
    "        elif t == 'partial':\n",
    "            # find partial mediations\n",
    "            df = df[df['direct_effect_p'] < 0.05]\n",
    "        \n",
    "        # check how many indirect; this reduces it by alot again\n",
    "        df = df[df['indirect_effect_p'] < 0.05]\n",
    "\n",
    "        # create column for metabolites\n",
    "        df = df[df['M'].isin(['H_' + x for x in int_ba[g]])]\n",
    "        \n",
    "        # grab CUTIE results\n",
    "        df_qt = pd.read_csv(path + 'outputs/' + g_to_cutiejob[g] + '/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "        df_qt = df_qt[df_qt['class'] == 'TP']\n",
    "\n",
    "        for r in range(len(df)):\n",
    "            x,m,y = df.iloc[r,0:3].values\n",
    "            \n",
    "            # check if it's in CUTIE\n",
    "            df_q = df_qt.copy()\n",
    "            # df_q = df_q[df_q['var1'] == m]\n",
    "            df_q = df_q[df_q['var2'] == x]\n",
    "        \n",
    "            if len(df_q) > 0: #len(df_q2) > 0: # len(df_q1) > 0 or len(df_q2) > 0:\n",
    "                # print('plotting ' + m + ' and ' + x)\n",
    "                # create plotting df\n",
    "                df_plot = pd.concat([df_X,df_M,df_Y],axis=1)\n",
    "                # path + 'outputs/jobs19/plots/'\n",
    "        \n",
    "                plt.figure(figsize=(8,6))\n",
    "                sns.set_theme(style=\"whitegrid\")\n",
    "                # sns.scatterplot(data=df_plot,x=x,y=m,hue=y)\n",
    "                sns.scatterplot(data=df_plot,x=x,y=y,hue=m,palette='coolwarm')#,size=6)\n",
    "                plt.tight_layout()\n",
    "                sns.despine()\n",
    "                plt.savefig(path + 'outputs/jobs19/plots/' + g + '/' + x.split(';g__')[-1] + '_' + m.replace('/','fslash') + '_' + t + '_' + y + '_MX.pdf')\n",
    "                plt.close()\n",
    "        \n",
    "                plt.figure(figsize=(8,6))\n",
    "                sns.set_theme(style=\"whitegrid\")        \n",
    "                sns.scatterplot(data=df_plot,x=x,y=m,hue=m,palette='coolwarm')#,size=6)\n",
    "                sns.despine()\n",
    "                plt.tight_layout()\n",
    "                plt.savefig(path + 'outputs/jobs19/plots/' + g + '/' + x.split(';g__')[-1] + '_' + m.replace('/','fslash') + '_' + t + '_' + y + '_YMX.pdf')\n",
    "                plt.close()\n",
    "        \n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "53a8f0b2-47ab-4bb5-8f69-86d734fdd89e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of bootstraps: 100\n",
      "statistic: total_effect_p\n",
      "prop under 0.05: \n",
      "0.55\n",
      "mean and std\n",
      "0.051199999999999996 0.03326499661806687\n",
      "\n",
      "\n",
      "statistic: direct_effect_p\n",
      "prop under 0.05: \n",
      "0.0\n",
      "mean and std\n",
      "0.6058 0.09083149233608354\n",
      "\n",
      "\n",
      "statistic: indirect_effect_p\n",
      "prop under 0.05: \n",
      "0.05\n",
      "mean and std\n",
      "0.10980000000000001 0.04004946940971876\n",
      "\n",
      "\n",
      "number of bootstraps: 200\n",
      "statistic: total_effect_p\n",
      "prop under 0.05: \n",
      "0.38\n",
      "mean and std\n",
      "0.0524 0.02209615351141461\n",
      "\n",
      "\n",
      "statistic: direct_effect_p\n",
      "prop under 0.05: \n",
      "0.0\n",
      "mean and std\n",
      "0.6138999999999999 0.05740897142433402\n",
      "\n",
      "\n",
      "statistic: indirect_effect_p\n",
      "prop under 0.05: \n",
      "0.01\n",
      "mean and std\n",
      "0.113 0.029410882339705485\n",
      "\n",
      "\n",
      "number of bootstraps: 1000\n",
      "statistic: total_effect_p\n",
      "prop under 0.05: \n",
      "0.55\n",
      "mean and std\n",
      "0.050300000000000004 0.01037834283496166\n",
      "\n",
      "\n",
      "statistic: direct_effect_p\n",
      "prop under 0.05: \n",
      "0.0\n",
      "mean and std\n",
      "0.6142 0.032411726273063594\n",
      "\n",
      "\n",
      "statistic: indirect_effect_p\n",
      "prop under 0.05: \n",
      "0.0\n",
      "mean and std\n",
      "0.11530000000000001 0.014550257729676132\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Axes: xlabel='indirect_effect_p', ylabel='Density'>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjYAAAG1CAYAAADqer7eAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABiIElEQVR4nO3dd3hUZd4+8PtMy8yk94SSThJCCQGC9EAQUQGVRV9XRQVdC7iia0X5Wfb1tezaFUVFdLFgAwRlRRcVCWIoCR1CSQ/pvc0k087vj0DWSEkymcmZcn+uiytm5syZ7zwmM3ee8xRBFEURRERERC5AJnUBRERERLbCYENEREQug8GGiIiIXAaDDREREbkMBhsiIiJyGQw2RERE5DIYbIiIiMhlKKQuoL/t378foihCqVRKXQoRERH1kNFohCAISElJuehxbtdjI4oizq5JKIoiDAYDuEah7bFt7Yvtaz9sW/ti+9qXK7fv7z+/L8btemzO9tSMGDECOp0OOTk5iIuLg1arlbgy18K2tS+2r/2wbe2L7Wtfrty+hw8f7tFxbtdjQ0RERK6LwYaIiIhcBoMNERERuQwGGyIiInIZkgeb2tpaPPzwwxg/fjxSUlJw5513Ijc3t/P+nJwcLFiwAKNGjcK0adOwevVqCaslIiIiRyZ5sFm8eDFKSkqwatUqrFu3Dmq1GgsXLoRer0d9fT0WLVqEqKgorF+/Hvfeey9ef/11rF+/XuqyiYiIyAFJOt27vr4egwYNwuLFizFkyBAAwJIlS3D11Vfj1KlTyMzMhEqlwtNPPw2FQoHY2FgUFRVh1apVmD9/vpSlExERkQOStMfG398fr7zySmeoqampwerVqxEWFoa4uDhkZWUhNTUVCsV/89f48eNRUFCA2tpaqcomIiIiB+UwC/Q98cQT+PLLL6FSqbBy5UpotVpUVFQgPj6+y3EhISEAgLKyMgQGBlr1XKIoQqfTQa/XA0DnV7Idtq19sX3th21rX2xf+3Ll9hVFEYIgdHucwwSbW2+9Fddffz0+++wz3HPPPVi7di3a2tqgUqm6HOfh4QEAaG9vt/q5jEYjcnJyOr8vLCy0+lx0cWxb+2L72g/b1r7Yvvblqu37x0xwPg4TbOLi4gAAzzzzDA4cOIBPPvkEarUaBoOhy3FnA01flopWKpWIi4uDXq9HYWEhoqKioNForC+ezsG2tS+2r/2wbe2L7Wtfrty+v58xfTGSBpva2lpkZmbiiiuugFwuBwDIZDLExsaiqqoKYWFhqKqq6vKYs9+HhoZa/byCIHQJRhqNxuX21HAUbFv7YvvaD9vWvti+9uWK7duTy1CAxIOHq6qq8OCDD2LPnj2dtxmNRhw7dgyxsbFITU1FdnY2zGZz5/2ZmZmIjo62enwNERERuS5Jg01iYiImT56Mv//978jKysLJkyfx6KOPoqmpCQsXLsT8+fPR0tKC5cuXIzc3Fxs2bMCaNWtw1113SVk2EREROShJg40gCHjttdcwfvx43H///bjuuuvQ2NiITz/9FAMGDEBgYCDef/99FBQUYN68eVixYgUeeeQRzJs3T8qyiYiIyEFJPnjY29sbTz/9NJ5++unz3j9y5Eh88cUX/VsUEREROSXJgw0ROa8WnQG6NpPUZXShVSvgpe1+SigRuSYGGyKymq7NhONFdTCaLFKXAgBQKmRIjAxgsCFyYww2RNQnRpPFYYINEZHku3sTERER2QqDDREREbkMBhsiIiJyGQw2RERE5DIYbIiIiMhlMNgQERGRy2CwISIiIpfBYENEREQug8GGiIiIXAaDDREREbkMBhsiIiJyGQw2RERE5DIYbIiIiMhlMNgQERGRy2CwISIiIpfBYENEREQug8GGiIiIXAaDDREREbkMBhsiIiJyGQw2RERE5DIYbIiIiMhlMNgQERGRy2CwISIiIpfBYENEREQug8GGiIiIXAaDDREREbkMBhsiIiJyGQw2RERE5DIYbIiIiMhlMNgQERGRy2CwISIiIpfBYENEREQug8GGiIiIXAaDDREREbkMBhsiIiJyGQw2RERE5DIYbIiIiMhlMNgQERGRy2CwISIiIpfBYENEREQug8GGiIiIXAaDDREREbkMBhsiIiJyGQw2RERE5DIYbIiIiMhlKKQuoKGhAa+88gp++eUXtLS0ICEhAQ8++CDGjh0LAHjsscewYcOGLo8JDQ1FRkaGFOUSERGRA5M82DzwwAOora3FK6+8goCAAKxduxa33347NmzYgNjYWJw4cQJ33303FixY0PkYuVwuYcVERETkqCS9FFVUVISdO3fiqaeewtixYxETE4Ply5cjNDQUmzdvhtlsRm5uLkaMGIHg4ODOfwEBAVKWTURERA5K0mDj7++P9957D8OHD++8TRAEiKKIxsZGFBYWor29HbGxsRJWSURERM5C0ktRPj4+SEtL63Lbli1bUFxcjMmTJ+PkyZMQBAFr1qxBRkYGZDIZ0tLScP/998Pb29vq5xVFETqdDnq9HgA6v5LtsG3ty1Ha12g0wmgwwGCy2OycoiiiVW9CTWMbmnUGKBUyeKjkCPXXQqvu5i3LIoPRaIROp7P6+R2lbV0V29e+XLl9RVGEIAjdHif5GJvfy87OxuOPP44ZM2YgPT0db7zxBmQyGQYOHIh33nkHRUVF+Mc//oGTJ09izZo1kMms63AyGo3Iycnp/L6wsNBGr4D+iG1rX1K2r1wuh6DyRmVVJdoMpj6fz2wRUVprQH55O1razh+U/L3kGBykwqBgFWTneYNTqxQI9xdQXd4Ms9ncp3r4s2tfbF/7ctX2ValU3R4jiKIo9kMt3frxxx/x0EMPITk5Ge+++y7UajVEUURzczN8fHw6jztw4ACuv/56fPnll0hOTu718xw+fBiiKCIuLg56vR6FhYWIioqCRqOx5ctxe2xb+3KU9q1vNuJwbnWfe2xKq1ux/UAZWvUdAUkQAH9vD/h6qWAyi2jVG1HX1N55vJ+3CuOTQjE41KvLeVQKGUbEBcPfW2l1LY7Stq6K7Wtfrty+ubm5EAQBI0aMuOhxDtFj88knn+DZZ5/FzJkz8dJLL3UmMkEQuoQaAIiPjwcAVFRUWBVszp5Xq9V2fq/RaLp8T7bDtrUvqdu3pU0HpUoFyKwLNhaLiD3HKpB9vAoA4KlRInlIEIZFB0Kl7Dr7sUVnwMmSBuw/UYWGZgO+312CEXFBmDQyHPIzvbdKhQxKpdImbSJ127o6tq99uWL79uQyFOAAwWbt2rV45plncPPNN+Pxxx/vcnnpwQcfRENDA1avXt152+HDhwEAcXFx/V4rEdmO2WLBf3YVI7+sEQCQFB2AyckDoFScfzkHL60KoxNCMCw6EHuPVeBgbg0O59agqk6HKyZEwVNjfS8NEbkOSWdFFRQU4LnnnsPMmTNx1113oba2FtXV1aiurkZzczPmzJmDnTt3YuXKlSguLsb27dvx+OOPY86cOZwpReTEzGYLvs8sQn5ZI2QyAZddEoHpYwZfMNT8nodKjsmjBuLKiVFQKWWorNPh6+25aNYZ+qFyInJ0kvbY/PDDDzAajdi6dSu2bt3a5b558+bhhRdewOuvv4533nkH77zzDry9vTF37lzcf//90hRMRH0miiK27i1GYXkT5DIBV06MQkSYT/cP/IPoAb64bkY8vsnIR2OLAV//kotr04fYoWIiciaSBpu7774bd99990WPmTVrFmbNmtVPFRGRvWXlVCHvdCNkgoDZk6IxONT6pRv8vDwwb1osNmXkobHFgPXbcjEiNgghAa41toCIeo6bYBJRv8kvbcSeYxUAgLTRA/sUas7y1qowLy0OPp4qNLUa8PoX+6Fv7/v0cyJyTgw2RNQvmloN+HFvMQBgRFwQkqIDbXZuT40ScyfHQOOhQFFFM/75cRbMFodYyYKI+hmDDRHZnSiK+GlvMYwmC8IDtZg0coDNn8PP2wNXTY2BSiFDVk4lPv0+p/sHEZHLYbAhIrs7eKoGZTWtUMhlmJEaAbmsZ+tR9FZ4oCcWzkkCAHz10ylkHi6zy/MQkeNisCEiu6pvbsOuI+UAgEnJA+Dr5WHX57tkWDiuntqxHMSrn+1DaXWLXZ+PiBwLgw0R2Y0oitixvxRmi4jBoV4YFh3QL8+7aE4ShscGQt9uxkufZMFow006icixMdgQkd3klTaipKoFcpmAtJRBPV4Sva/kchkevHEMvDRK5J5uxNofjvfL8xKR9BhsiMguDCYzdh7sGOMyOiHE7peg/ijIT4N7/2cUAGD9tlM4lFvdr89PRNJgsCEiu9h3vAoteiN8PFUYnRgiSQ0TRw7AZZdEQhSBV9bu47YLRG6AwYaIbK5Fb8TBUx09JJNGDoBC3n9vNX+82HXH1cMxMNgTtY1tWPHVAYhi9+vbyOXd71lFRI5J8t29icj17D1WAZNZRHigFtEDer8PlLXkMgEmiwVVdbout982dzie+9ce/HaoHBu25WLKqIEXPIfRaISg8kZ9sxEtbboLHtcbWrUCXlqVTc5FRBfHYENENlXX1IacgjoAwIQRA/ptwDAAyGQC2gxmlFQ2nDMTasKIcPx6sAyf/nAcgoALjvkxGgyorKpEaIgIparvYUSpkCExMoDBhqifMNgQkU3tOlIOEUDMAF+EB3lKUoPRZDkn2IyMC0J+aSPKalrxw64iXJMWe97QZTBZ0GYwwWCyADJOEydyNhxjQ0Q2U12vQ0FZEwQA44eHSV1OF4IgYEbqYCjkMpTVtOLgqRqpSyIiO2CwISKb2XOsEgAwJMIf/j5qias5l4+nByYnd+xTtetIOeqa2iSuiIhsjcGGiGyiql6HwvKO3prUoaFSl3NBSdEBiAj1htnSsTGnhbuAE7kUBhsisom9Z3pr4iP84efdv4vx9YYgCJg+djA8lHJU1eux70SV1CURkQ0x2BBRn9U06Dt7a8Y6cG/NWV4aJaakdEz53nusAtUNeokrIiJbYbAhoj7bf7JjMb64wX4O3Vvze/GD/RA70BcWEfhxTzHMZs6AInIFDDZE1CdNrQacKqkHAKTEB0tcTc8JgoC00YOg8VCgrqmtc+AzETk3Bhsi6pP9J6ogisCgEC8E+2ulLqdXNB4KTBs9CEDH6yivaZW4IiLqKwYbIrJai96II/m1AICUeGk2uuyrmIG+SIj0hwjgp6zicxb2IyLnwmBDRFbbll0Co8mCQF81Bod6SV2O1aYkD4SXRonGFgP25nCWFJEzY7AhIqu0G834aW8xAGB0Qki/7gllax4qOaaPGQwAOFpQj5pGo8QVEZG1GGyIyCo/7y1Gs84Ib60KsYP8pC6nzyLCvDE8JhAAcKBAhzaDSeKKiMgaDDZE1Gtmi4ivf8kDAIxODIZc5ry9Nb83YWQ4fD1VaDOIyDhQDlHkqsREzobBhoh6bdfhcpTXtsJTo+zs5XAFKoUcM8YOhEwAiipacCiXG2USORsGGyLqtU0ZHb0100cPglIhl7ga2wr0VSMpQgMA+O1QOarqdRJXRES9wWBDRL2SW9KAnMI6yGVC54BbVxMZokJUmDcsoogfdhXBYDRLXRIR9RCDDRH1yre/5gMAJicPdJrtE3pLEARMHRUOb60STa0G/LLvNMfbEDkJBhsi6rGG5nZk7C8FAMydEi1xNfbloZLjsksiIQjAqZIGHD2zECEROTYGGyLqsR92FcJktmDIYD8kRAZIXY7dhQV6YvzwcADAjgOlKK1qkbgiIuoOgw0R9YjJbMF3vxUCAOZOiZG2mH6UEh+MIYP9YBGB73cVoqm1XeqSiOgiGGyIqEcyD5WjrqkNft4emJw8QOpy+o0gCEgfOxgh/hq0Gcz4985CDiYmcmAMNkTUI2cHDV8+Psrlpnh3RyGX4YqJ0dCqFahrasOPe4s5mJjIQTHYEFG3fj/F+4qJUVKXIwkvjRJXTIiCXCagoKwJu49WSF0SEZ0Hgw0Rdev3U7wDfNQSVyOdsEBPTBszCACQfbwKR/K4MjGRo2GwIaKLcqcp3j2RGBmAsUNDAQDb95ci93SDtAURURcMNkR0UT/sdq8p3j0xLikUw87skbV1TzFKKpslroiIzmKwIaILMpkt+G5nIQD3muLdHUEQMDVlIGIH+sJiEbElsxBVddxTisgRMNgQ0QW56xTvnpAJAmaOi8DAYC8YTRZ8+2s+6pvapC6LyO0x2BDRBbnzFO+ekMtluHJiFILPrHGzKSMPDS1cwI9ISgw2RHRenOLdMyqlHHMnxyDAR43WNhM2bc/j6sREEmKwIaLz4hTvntN4KHD11Bj4e3ugRW/Exu35aNYZpC6LyC0x2BDROTjFu/e0aiWunhoLXy8VmnUGbNyehxa9UeqyiNwOgw0RnYNTvK3jqVHimqmx8PFUoanVgE3b89DKcEPUrxhsiKgLTvHuGy+tCldPjYWXVomGlnas35aLplZeliLqLww2RNQFp3j3nY+nCtdMjYWnRom6pja8vDab4Yaon0gebBoaGvDkk09i6tSpGD16NG644QZkZWV13p+Tk4MFCxZg1KhRmDZtGlavXi1htUSuj1O8bcPXy6Mj3KgVOF3Vgife/Q0tHFBMZHeSB5sHHngABw8exCuvvIJ169Zh2LBhuP3225GXl4f6+nosWrQIUVFRWL9+Pe699168/vrrWL9+vdRlE7mkUyX1yCmsg0Iu4EpO8e4zP28PzJ8+BD6eKuSXNuLJ9zI55obIzhRSPnlRURF27tyJzz77DKNHjwYALF++HBkZGdi8eTPUajVUKhWefvppKBQKxMbGoqioCKtWrcL8+fOlLJ3IJX2z48wU71ED4c8p3jYR4KvGgzeOwUufZuNUSQOeXpWJv985AVq1UurSiFySpD02/v7+eO+99zB8+PDO2wRBgCiKaGxsRFZWFlJTU6FQ/Dd/jR8/HgUFBaitrZWiZCKXVdfUhl8PdEzxvoqDhm1qUIgX/u/uifDSKHG8qB7/u3o32tpNUpdF5JIk7bHx8fFBWlpal9u2bNmC4uJiTJ48Ga+++iri4+O73B8SEgIAKCsrQ2BgoFXPK4oidDod9Ho9AHR+Jdth29qXPdr3m+15MJlFxEf4YmCgB3S67jd1NBqNMBoMMJgsNqujL5QyCyxmS59qMhqNXb72mUUGk8mIMH8lHr81Bc98uA9H82vxzOpMPLogBQqF5CMC+hXfG+zLldtXFEUIgtDtcZIGmz/Kzs7G448/jhkzZiA9PR3PP/88VCpVl2M8PDwAAO3t1i9ZbjQakZOT0/l9YWGh1eeii2Pb2pet2tdkFvH9rnIAwMjB8i6/Hxcil8shqLxRWVWJNoNj9D74eqkR4u+Bqupq6Nv7NlC3prbGJjV5aVRoGeCJmlo9RFHELZdF4MMthTiUV4cXP83G/KkDe/RmbWsqhYC21kZYLNKEUr432Jertu8fM8H5OEyw+fHHH/HQQw8hOTkZr7zyCgBArVbDYOj65nQ20Gi1WqufS6lUIi4uDnq9HoWFhYiKioJGo7G+eDoH29a+bN2+2/eXobWtFIE+Hpg3MwUKec96EeqbjQgNER2mx8ZTrYBWo0VIcHCfemxqamsQFBgEpbLv42A81QrIFWpUN7XBaLJAodTgsnGD8O/MYhzIbYBcrsD4YaF9fp7eUCrkSIwKQOSgsH59XoDvDfbmyu2bm5vbo+McIth88sknePbZZzFz5ky89NJLnYksLCwMVVVVXY49+31oqPVvBIIgdAlGGo2mT0GJLoxta1+2aF9RFPH97tMAgNmTY+Dj7dXjx7a06aBUqQCZYwQbpUoBmVxmk5qUSmWP/jrsaU2QyQFZR89M5AB/TBttwbbs08g+UQN/Hy0SIv37/Fw9JpNBqVRK+rvJ9wb7csX27WnPpuQXd9euXYtnnnkGN910E1577bUubySpqanIzs6G2WzuvC0zMxPR0dFWj68hoq6OFdQhv7QRKqUcs8ZHSV2O20iKDsTohI4xg9uyS1BZ1/2YJiLqnqTBpqCgAM899xxmzpyJu+66C7W1taiurkZ1dTWam5sxf/58tLS0YPny5cjNzcWGDRuwZs0a3HXXXVKWTeRSvj0zxXv6mEHw8ex7DwX13PjhYYgK94HZIuK73wq4xg2RDUgabH744QcYjUZs3boVkydP7vLv2WefRWBgIN5//30UFBRg3rx5WLFiBR555BHMmzdPyrKJXEZVvQ6ZRzoGDc+dzCne/U0QBMwcF4EAHzV0bSZs3VMEi0WUuiwipybpGJu7774bd99990WPGTlyJL744ot+qojIvXy3swAWi4jkIUGIDPeRuhy3pFLKcfmESHz10ymUVrdi77EKXDI8XOqyiJyW5GNsiEgabe0m/LCrCABw1ZRYiatxb/7eakwbPQgAkHW8CiWVzRJXROS8GGyI3NS2fafRojciPNATY4f273RjOld8hD+GxXRMivhxb7HDrA1E5GwYbIjckCiKnYOGZ0+OhkzW/wvE0bkmjRwAP28P6NpMyNhfKnU5RE6JwYbIDR04WY2SymZoPOS4NDVC6nLoDKVChktTIyAIwKmSBpwqqZe6JCKnw2BD5IY2ZuQBAGakRsBTw12mHUlogBZjEzsuDWbsL4Wem2US9QqDDZGbKSpvwr7jVZAJwNVTOWjYEY0ZGopAXzXaDGbsPFgmdTlEToXBhsjNfL29Y7+VCSMHICzQU+Jq6HzkMgHTxwwGAJworucsKaJeYLAhciO1jXps39exL9S8NPbWOLLQAC1GxHbMkvpl32mYzI6xHxeRo2OwIXIjm38tgMksYlhMIBIiA6Quh7oxfng4PDVKNLUacOBktdTlEDkFBhsiN6FvN2FLZiEA9tY4C5VSjokjOlYhzj5ehRbuJUXULQYbIjexdU8RWvVGDAz2RGpSmNTlUA8NGeyHsEAtTGYLMg+XS10OkcNjsCFyA2azBZsyOhbkuzotjgvyORFBEDAleSAA4GRxPSpqWyWuiMixMdgQuYHfDpejqk4HH08V0scOlroc6qWQAC0SI/0BAJmHyyGK3AGc6EIYbIhcnCiK+PqXjinecyZFw0Mpl7gissa4YWGQywSU1bSimNO/iS6IwYbIxR3KrcGpkgaoFDJcOSla6nLISt5aFYbHBgEAdh1hrw3RhTDYELm4r346CQC4bHwkfL08JK6G+mJsYghUChlqGtpwqqRB6nKIHBKDDZELO1FUh4OnaiCXCZg3LU7qcqiP1B4KpCSEAAD25lTCwl4bonMw2BC5sK9+OgUAmD5mMEL8tRJXQ7YwMi4IHio5GprbkcteG6JzMNgQuajC8ibsPloBQQDmp7O3xlWolHKMGhIMAMjKqeRYG6I/YLAhclHrzvTWTBw5AINCvCWuhmxpRFwQPJRy1De3I/d0o9TlEDkUBhsiF1Re04odBzo2u7wufYjE1ZCteSjlGDmkY4YUe22IumKwIXJB67edgkUExiSGIHaQn9TlkB0kxwVDqZChrqkNRRVc14boLAYbIhdT26jHT3tLAADXzYiXuBqyFw+VHMNiAgEA+09USVwNkeOwKtiUlZXZug4ispGN2/NgMlswLCaw84OPXFNyXBBkQsdqxNxDiqiDVcFmxowZWLRoEb799lu0t7fbuiYislJTqwFbMgsBANfN4NgaV+elVSE+wg8AsP9ktbTFEDkIq4LNSy+9BIVCgWXLlmHSpEl48sknceDAARuXRkS99e2OfLQbzIgZ6IvRZxZyI9d2dsG+/NJGNDTzD00iq4LN7NmzsWrVKvzyyy+46667sG/fPvz5z3/G5Zdfjvfeew+VlZW2rpOIuqFrM+LbX/MBAP8zIx6CIEhcEfWHAB81IsM6pvMfyauRuBoi6fVp8HBwcDDuuOMObN68GV9//TVCQkLw6quvIj09HYsXL0Z2drat6iSibnyfWYhWvREDg70wfkS41OVQPxoR1zH1O6ewDgaTWeJqiKTV51lRWVlZeOKJJ7Bw4UJkZWVh0qRJePzxx2EymbBgwQJ8+OGHtqiTiC7CYDTj6+15AIBr04dALmNvjTuJCPWGr5cKBpMFJ4vqpS6HSFIKax5UVFSETZs24ZtvvkFpaSkGDhyIW265BfPnz0dYWBgA4KabbsJDDz2ElStXYtGiRTYtmoi6+nFvMRqa2xHsr8G0MYOkLof6mSAIGBEbhF8PluFQXg2GxQTyUiS5LauCzaxZs+Dh4YFLL70UzzzzDCZMmHDe42JiYlBYWNiX+oioGyazBeu35QIA/jQtDgo5l6dyR4lRAdh1pAL1Te0orW7hNhrktqwKNk888QSuuuoqeHtf/BdnyZIlWLJkiVWFEVHPZOwvRVWdDn5eHph5SaTU5ZBEPJRyJEb640h+LQ7n1jLYkNuy6k+7H374AVVV51/p8vjx45g7d26fiiKinrFYRKz7+SQA4KqpMfBQyiWuiKR0dhBxQVkjmloNEldDJI0e99hkZWV1brS2Z88e7N27F3V1decct23bNpSUlNiuQiK6oKzj1SipbIGnWoErJ0ZLXQ5JLMBHjYHBXiitbsHR/FpM4Ow4ckM9Djbr1q3Dxo0bIQgCBEHA3//+93OOORt85syZY7sKiei8RFHE1xkFAIDZk2PgqVFKXBE5gpFxQSitbsGxglqkJoVyzBW5nR4Hm+XLl+NPf/oTRFHErbfeiieffBJxcXFdjpHJZPDx8cGQIVzKncje8ivakV/aBJVSjqumxEhdDjmIqAE+8NIq0aIz4lRJA4ZGBUhdElG/6nGw8fb2xrhx4wAAH330EYYNGwZPT0+7FUZEF7fjaDMA4PLxkfD18pC4GnIUMkHAiJggZB4px5G8GgYbcjs9DjYbN25EWloa/P39UVZW1u0O39dcc01fayOiCzhR3IDCqnbI5QKuSYvr/gHkVhKj/LH7aAWq6vWoadAjyE8jdUlE/abHwWbZsmX48ssv4e/vj2XLll30WEEQGGyI7GjjmbE1U0eFI9ifH1rUlVatRPQAH+SVNiKnsA5TRg2UuiSiftPjYPPTTz8hODi487+JSBoFZY3Yd6IGggBcPSVK6nLIQSVFByCvtBEniuoxYUQ4BxGT2+hxsBk4cOB5//ssk8mElpYW+Pn52aQwIjq/dT+dAgAkDdYgPJDj3Oj8BoV6dw4izi9tRHyEv9QlEfULqyK8yWTCihUr8M033wAAMjMzMXHiREyYMAG33norGhsbbVokEXUoq27BrwdLAQBThnFlWbowmSB0Dhw+VnDummNErsqqYPPmm29i5cqVaG7umJXx3HPPwd/fH4899hiKi4vx8ssv27RIIuqwflsuLCIwOj4IYf4qqcshB3c22JRWt6CxpV3iaoj6h1XBZvPmzXjggQdw0003IT8/H6dOncLixYtxyy234G9/+xt+/vlnW9dJ5PZqG/X4OatjVe+rp3KVYeqet1aFiNCOnr2cQvbakHuwKthUVVUhOTkZAJCRkQGZTIapU6cCAMLCwjp7cojIdr7JyIfJbEFSdAASI/2kLoecxNDojl6b44V1sFhEiashsj+rgk1ISAhOnz4NANi6dSuGDh2KgICOX579+/cjLCzMdhUSEVr0RmzJLAQAzE/nyt7Uc9EDfKBWydHaZkJxBf/oJNdnVbC56qqr8Pzzz+P2229HdnY25s+fDwB49tln8eabb3J3byIb2/JbAfTtJkSGeWNsYqjU5ZATkctkSIw8M4i4sFbiaojsz6pgs3TpUtx2220QBAEPPvggbrzxRgDA4cOHcdttt2Hx4sU2LZLInbUbzfgmIx9AR2+NTCZIXBE5m7OXowrLm9CqN0pcDZF99Xgdm98TBAF33XUX7rrrri63f/75530q5u2330ZmZiY+/vjjztsee+wxbNiwoctxoaGhyMjI6NNzETmLn/cWo6GlHcH+Gq4gS1YJ8FEjLFCLilodThbXY9wwDhcg12VVsAGA5uZm7Nq1CzqdDqJ47oC03m6p8K9//QtvvPEGUlNTu9x+4sQJ3H333ViwYEHnbXK53KqaiZyN2WzBhl9yAQDz0uK4eixZLTEyABW1OhwvqkdqEi9nkuuyKths374d999/P/R6/Xnv781eUZWVlVi+fDmys7MRHd11CqvZbEZubi6WLFnSuZ0DkTvZeagMFbU6eGtVmDkuQupyyInFDfLDjgOlqGtqQ3XD+d+7iVyBVcHmlVdeQUxMDB577DGEhoZCJrP+r8ijR4/C19cX33zzDd566y2UlpZ23ldYWIj29nbExsZaff7zEUUROp2uM5hdKKCR9di2fSeKIr768SQA4PLxg2AxG6DTGQA4TvsajUYYDQYYTBZJ6zhLKbPAYrb0qSaj0djlqyPUZAsCgMgwL+SXNeNIbjWmJodDp9P1ex2O8rPrqly5fUVRhCB0P8bQqmCTn5+Pt99+G2PHjrXm4V2kp6cjPT39vPedPHkSgiBgzZo1nevlpKWl4f7774e3t/XLyRuNRuTk5HR+X1hYaPW56OLYttbLLWtDYUUzlHIB0f76Lj+zZ0nZvnK5HILKG5VVlWgzmCSr4/d8vdQI8fdAVXU19O2GPp2rprbG4Wrqq0AvC/IBnCiuR2V1NarLW2A2myWphe8N9uWq7atSdb/iulXBZsCAAWhpabHmob1y6tQpyGQyDBw4EO+88w6Kiorwj3/8AydPnsSaNWus7ilSKpWIi4uDXq9HYWEhoqKioNFobFy9e2Pb9t2XmVkAgJnjBmPMqIQu9zlK+9Y3GxEaIjpMj42nWgGtRouQ4OA+9djU1NYgKDAISqXSIWqyldAQEUeKTkHfbkZlo4C0lPh+r8FRfnZdlSu3b25ubo+OsyrY3HXXXXjrrbcwYsQIDBo0yJpT9Mi9996LhQsXwsfHBwAQHx+P4OBgXH/99Th8+HDn6se9JQgCtFpt5/cajabL92Q7bFvrnCiqw7GCeshlAq6dkQit9vxvUFK3b0ubDkqVCpA5RrBRqhSQyWU2qUmpVPbor8P+rMkW4iMCcPBUNXYfq8IVk2x7mb83pP7ZdXWu2L49uQwFWBlsvv32W1RWVmLmzJkICAiAWq0+58l//PFHa059znnOhpqz4uM7/sKoqKiwOtgQObr12zr+MkkbPQjB/q71VxdJKzHSHwdPVePgqWo06wzw1nIzVXItVgWbsLCwftk24cEHH0RDQwNWr17dedvhw4cBAHFxcXZ/fiIpnK5qxq4j5QCA+dP5c062FeSnQZCfBjUNeuw4UIorJ3JDVXItVgWb559/3tZ1nNecOXOwePFirFy5ErNnz0ZBQQH+93//F3PmzLH5TCkiR7Fxex5EEbhkWBgiwny6fwBRLyVFByBjfyl+ziphsCGXY/UCfQCQl5eHnTt3oqqqCjfffDNKSkqQmJgILy8vmxQ3ffp0vP7663jnnXfwzjvvwNvbG3PnzsX9999vk/MTOZrGlnZsyyoBAFyTxvBO9pEQ6Y9fD5ThRFE9Tlc1Y1CI9bNMiRyNVcHGbDbjqaeewvr16zvnlV9xxRV46623UFJSgk8++cSqS1UvvPDCObfNmjULs2bNsqZMIqfzfWYhDCYL4gb5YlhMoNTlkIvyVCsxPDYQh3JrsC37NG6+YqjUJRHZjFXzpVeuXIlvv/0W//d//4edO3d2bqnw6KOPwmKx4NVXX7VpkUTuwGgy4987CwAAV0+N7fEMACJrTBwRDgDYll0Ci+XcbXGInJVVwWb9+vVYunQp5s+fDz8/v87bExMTsXTpUuzcudNW9RG5jYz9pahvbkegrxqTkrnZJdnXqPhgeGqUqK7X40i+bRYjJHIEVgWbmpoaDB16/q7L0NBQNDU19akoIncjiiI2ZeQBAGZPioZS0f2vJjeDpb5QKuSdu8X/tLdE4mqIbMeqMTaRkZHYvn07Jk6ceM59e/bsQWRkZJ8LI5JKi84AXVv/bhGQU1iHgrImqJQyjEkMRVXdf/fwEQDI5ALM5v9eLjAajRBU3qhvNqKlrf/3+wEAQQAMRmmW4yfbSB8zGN9nFuK3Q2W4+08jofHo03wSIodg1U/xrbfeiieffBJGoxHTp0+HIAgoKirC7t278cEHH2DZsmW2rpOo3+jaTDheVAdjPy5/f7a3ZmhUAArKGrvcp/FQIDzIEyWVzZ01GQ0GVFZVIjRE7FjRVgJn6yLnlRjlj/AgT5TXtCLzcDnSxw6WuiSiPrMq2Fx33XWoq6vDO++8g7Vr1wIAHnjgASiVSvzlL3/BDTfcYNMiifqb0WTpt2BT39yGgrKOy7fDY4LOeV6lwnJOTQaTBW0GU8feQxIt03+2LnJegiAgfexgfPr9cfycVcxgQy7B6n7HO+64A3PnzsWePXugUCjg7e2N5OTkLoOJiah7h051DNyMCveBn7eHxNWQu5k+piPYHMqtQXW9nlt4kNPrdbDZvHkzPv/8cxw8eBAmU8c4BLVajdGjR+OGG27ApZdeavMiiVxVm8GE40X1AIBRQ4IlrobcUWiAFsNjA3Ekrxa/7CvBdTP6f8dvIlvqcbCxWCx46KGH8N133yEkJARXXnklgoKCAACVlZXYs2cP7r33Xlx99dXnXWiPiM51vLAeJrMFgb5qDAjmeBWSRvqYwTiSV4tt2SW4Nn0I11Aip9bjYLN27Vp8//33WLZsGW655RbIZF2no1osFnz22Wd47rnnMGXKFMyePdvmxRK5ElEUcTiv4zLUiNggfpiQZCYlD8A7Gw6hpLIFuacbMGSwv9QlEVmtx+vYbNiwAddffz0WLlx4TqgBAJlMhptuugn/8z//gy+//NKmRRK5oqKKZjS1GuChlCM+wk/qcsiNadVKjD+zEvHPWVzThpxbj4NNYWEh0tLSuj1uypQpyM/P71NRRO7gbG9NYlQAlAoutkfSmjE2AkDHCtj9udQBka31ONjo9Xr4+vp2e5y/vz/q6ur6VBSRq2toaUdxRTMAYEQsN7sk6SUPCUKAjweaWg3Yd7xS6nKIrNbjYCOKYo+WcJfJZLBYmPaJLuZIbkdvTWSYN3y9OMWbpCeXy5A2umMdm5+zeTmKnJdVe0URkfUMJjNyijp6NUfEBUlcDdF/nV2gb8/RSjTrDBJXQ2SdXq1j8/TTT8PLy+uix7S0tPSpICJXd7K4AQajBb5eKkSEektdDlGnqHAfxAzwRX5ZI3YcKMWVE6OlLomo13rcY5OamgpPT0+IonjRf56enhg7dqw9ayZyWqIo4nAup3iT45p+pteGs6PIWfW4x+bjjz+2Zx1EbqGsphV1TW1QyGVIjAyQuhyic6SlDMSHm4/iRFE9SqtbMDD44r30RI6GY2yI+tHR/FoAQEKkPzxUnOJNjsffR43RCSEAgG3stSEnxGBD1E/07SbklTYCAIZFs7eGHFf6mI7LUduyS2CxiBJXQ9Q7DDZE/eREUT0sFhEh/hoE+2ulLofogsYND4NWrUBVvR5HC2qlLoeoVxhsiPqBKIqdHxBJ0VyQjxybh1KOyckDAfByFDkfBhuiflBe04qG5nYo5DIMGewndTlE3Tq7ps2vB8vQZjBJXA1RzzHYEPWDowUdC/LFR/hBpeSgYXJ8Q6MCEBqghb7dhN1HKqQuh6jHGGyI7KzNYELe6QYAvAxFzkMmEzB9DNe0IefDYENkZyeK6mG2iAj0VSPEXyN1OUQ9dvZy1IGTVahp0EtcDVHPMNgQ2ZEoijh25jLUsJhArjRMTiU8yBPDYgJhEdlrQ86DwYbIjirrdGdWGhYQH+EvdTlEvXZpagQA4Me9xRBFrmlDjo/BhsiOzq40HDfIDx4cNExOaFLyAGg85Civae3sfSRyZAw2RHbSbjQjl4OGyclpPBSda9r8uKdY4mqIusdgQ2QnJ4vrYTKLCPBRIyyQKw2T87p0XMflqF8PlkLfzjVtyLEx2BDZgSiKnZehkqIDOGiYnNrQqAAMDPZEm8GMnQdLpS6H6KIYbIjsoKpej9rGNshlAhIiOWiYnJsgCJjROYiYs6PIsTHYENnBsTP7QsUO8oNapZC4GqK+Sx87GDKhY0B8WXWL1OUQXRCDDZGNGYxmnCxuAAAMiw6QthgiGwn01SAlIQRAx9RvIkfFYENkY6dKGmAyW+Dn7YHwIE+pyyGymbODiH/OKoHZwjVtyDEx2BDZ2NnLUBw0TK7mkmFh8NYqUdvYhv0nqqQuh+i8GGyIbKi6Xoeqej1kMgGJkbwMRa5FqZBj2pmNMf+zu0jiaojOj8GGyIbOrswaM8AXGg8OGibXM2t8JABg99EK1DW1SVwN0bkYbIhsxGgy40RxPQBgWAx7a8g1RYb5YGhUACwWkSsRk0NisCGykdzTjTCaLPDxVGFgsJfU5RDZzeUTOnptfthdBAsHEZODYbAhspGzKw0Piw7koGFyaZOSB8JTo0RVnQ4HTlVLXQ5RFww2RDZQ26hHZZ0OMgFIjOJKw+TaPJRyTB8zCADww65CaYsh+gMGGyIbOHpm0HD0AF9o1UqJqyGyv1njowAAu49UoJ6DiMmBMNgQ9ZHJbMHJoo5Bw0lcaZjcRFS4DxIj/WG2iFyJmBwKgw1RH+WdbkC70QxvrRKDQ72lLoeo35zttfkPBxGTA2GwIeqjs5ehkjhomNzM5FED4KlWoKJWh0O5HERMjsGhgs3bb7+Nm2++ucttOTk5WLBgAUaNGoVp06Zh9erVElVHdK66pjaU17RCEIChUbwMRe5FrVJ0rkT8fSZXIibH4DDB5l//+hfeeOONLrfV19dj0aJFiIqKwvr163Hvvffi9ddfx/r16yWqkqirsysNR4X7wFPDQcPkfs6uRLzrSDlqG/USV0MESL7me2VlJZYvX47s7GxER0d3ue/LL7+ESqXC008/DYVCgdjYWBQVFWHVqlWYP3++RBUTdTCZLThR9N/LUETuKHqAL5KiA3CsoA4/7CrCjbMSpS6J3JzkPTZHjx6Fr68vvvnmGyQnJ3e5LysrC6mpqVAo/pu/xo8fj4KCAtTW1vZ3qURd5Jc2os1ghpdGiYgwDhom9zVnUgwA4PvMQhhNFomrIXcneY9Neno60tPTz3tfRUUF4uPju9wWEhICACgrK0NgoHV/JYuiCJ1OB72+o9v07FeyHWduW6PRCKPBAEM3b9BH8msAAPGDfWEyGu1Wj1JmgcVs6VKT8czzGe34vNbUJTVb1GTrtnXEdoJFBqPRCJ1OZ5PTJcf6wt9bhfrmdmzPLsTEEWEXPNaZ3xucgSu3ryiKPZqgIXmwuZi2tjaoVKout3l4eAAA2tvbrT6v0WhETk5O5/eFhYVWn4suztnaVi6XQ1B5o7KqEm0G0wWPa2kzo7ym40PBX2NAeUW53Wry9VIjxN8DVdXV0LcbutxXU1tjt+ftzsXqkoota7JV2zpiO6lVCoT7C6gub4bZbLbJOUdGemD7EQO+3nYC/or6bo93tvcGZ+Oq7fvHTHA+Dh1s1Go1DIaubwRnA41Wq7X6vEqlEnFxcdDr9SgsLERUVBQ0Gk2faqWunLlt65uNCA0RL/rX9e5jVQCAwSGeiIkcaNd6PNUKaDVahAQHd+mxqamtQVBgEJRKaQYtn68uqdmiJlu3rSO2k0ohQ1BgEPy9w212ztCBbfj12K8orjZA4zcIUeHnvzzrzO8NzsCV2zc3N7dHxzl0sAkLC0NVVVWX285+HxoaavV5BUHoEow0Gk2fghJdmDO2bUubDkqVCpCd/0PIbLHgVEkjAGB4XHCP/oLoC6VKAZlcdt6alEql3Z/fmrqkYsuabNW2DtlOChmUSqVNfze1Wi0mjhyAHQdK8VN2Oe79n4u/Rzvje4MzccX27ek6YZIPHr6Y1NRUZGdnd+kqzczMRHR0tNXja4j6qqCsCfp2E7RqBaLCfKQuh8hhzJ7UMbP1l32n0aJzjMtu5H4cOtjMnz8fLS0tWL58OXJzc7FhwwasWbMGd911l9SlkRs7lt8xI29oVABkMq40THRWUnQAogf4wGA0c/8okoxDB5vAwEC8//77KCgowLx587BixQo88sgjmDdvntSlkZtqam1HSVULAG54SfRHgiBg9pmp3//eWcD9o0gSDjXG5oUXXjjntpEjR+KLL76QoBqicx3N71iQb3CoF3w8PSSuhsjxpI0eiA83H0VFrQ7ZxyuRmnThqd9E9uDQPTZEjsRssSCnsCPYDIvhGC+i81GrFJg5LgIA8E1GvsTVkDtisCHqoS6DhsN9pS6HyGHNnRwDmUzAgVPVKChrlLoccjMMNkQ9dCSvY9BwUlQA5Bw0THRBIQFaTBo5AACwKSNP4mrI3TDYEPVAQ3M7Sqs7Bg0P5YaXRN26Ji0WALB932nUNbVJXA25EwYboh44WtDRWxMZ5g0fT2kWxCNyJvER/hgaFQCTWcR3OwukLofcCIMNUTdMZguOc9AwUa+d7bX57rfCi+69RmRLDDZE3cgvbUSbwQwvjRKR4VxpmKinLhkejtAALZp1BmzLKpG6HHITDDZE3ThyZqXhpOgAyHq4VwkRAXKZgKumdizYtykjjwv2Ub9gsCG6iLqmNpTXtEIQOGiYyBqXpkbAU61AaXUrso5XSl0OuQEGG6KLOHqmtyYq3AdeGqXE1RA5H61aiVnjowAAG3/h1G+yPwYbogswmMw4XsRBw0R9NWdyDOQyAYfzapB7mgv2kX0x2BBdwMmiehiMFvh6qRAR6i11OUROK9hfg7TRgwAAm3YUSlsMuTwGG6LzEEURh8+sNDwiNggCBw0T9cn86XEAgL05VahuNEpcDbkyBhui8zhd1YK6pjYo5DIkRgVIXQ6RzfV3VI8I88Elw8IgikDm8ZZ+fnZyJwqpCyByRAdOVQMAEiP94aGUS1wNkW3JZQJMFguq6nT9+rwzUgdj99EKHCzQIb+sGSHn+ZtBq1bAS8vVvcl6DDZEf1DTqEd+accAxxFxQRJXQ2R7MpmANoMZJZUNMJos/frcA4O0KK3R4fOtpzB9bESX+5QKGRIjAxhsqE8YbIj+4Jfs0xBFYFCIFwJ81FKXQ2Q3RpOl34PNiLhAlNbocLSgDmOHhkHtwY8hsi2OsSH6nXajGRkHSgF0DBomItsaFOwJH60cJrOIw3k1UpdDLojBhuh3MvadRqveCG+tClEDuC8Uka0JgoC4cA8AwKHcGhhNZokrIlfDYEN0hiiK2PxrAQAgeUgQ94UispPwACV8tEq0Gcw4ml8ndTnkYhhsiM44VlCH/LJGKBUyrjRMZEeCICB5SMfv2P6TVTCZ+3ecD7k2BhuiM77+JRcAMH54ODQc0EhkV0MG+8FLq4SuzYRjBey1IdthsCECcLqqGXuOVQAALrskopujiaiv5DIBYxJCAAD7TlTBzF4bshEGGyIAG7fnQRSBS4aFYUCQl9TlELmFoVEB8NQo0ao3IqeQvTZkGww25Pbqm9rwc1YJAGDetDiJqyFyH3K5DKPP9Npks9eGbITBhtzet7/mw2iyICHSH0nR3BeKqD8lRQdAq1agRcdeG7INBhtya/p2E7b8VggA+NO0OO7iTdTPFL/rtdlzrJIzpKjPGGzIrW3dXYQWvREDgjxxyfBwqcshcktJ0YHQeCjQ1GrAriPlUpdDTo7BhtyWyWzBxow8AMA10+Igl7G3hkgKSoUMKfHBAIB/7yzgWBvqEwYbclu/HixDdb0evl4qpI8dLHU5RG5teGxHr01Vvb5zMD+RNRhsyC2JoogN204BAOZOjoGHUi5xRUTuTamQI3VoKADgs60nuIcUWY3BhtzSriPlKChrgsZDjismRktdDhEBGBkXBF8vFarr9fjP7mKpyyEnxWBDbsdiEbH2hxMAgDmTY+DjqZK4IiICAIVChjmTYgAAX/54Eu1G9tpQ7zHYkNvJPFKOwvImaDwUXJCPyMFMGTUQwf4a1DW1dS7FQNQbDDbkViwWEZ/9cBwAcNWUGHhr2VtD5EiUChn+PDMBALDu55PQt5skroicDYMNuZXfDpehqKIZWrUC16TFSl0OEZ1H+tjBCA/yRGOLAZt/zZe6HHIyDDbkNiwWEZ/9p2NszVVTYuHF3hoih6SQy3DDZR29Nhu25aJVb5S4InImDDbkNnYeLENxRTM81Qpczd4aIoc2NWUQBod6oUVvxNfbc6Uuh5wIgw25BbNFxGdbO8bWXD01Fl4apcQVEdHFyGUCbrp8KABg0/Y81De3SVwROQsGG3ILOw6UoqSyBZ4aJa6ayt4aImcwcUQ44iP80GYw44utJ6Uuh5wEgw25PKPJjE+25AAA5qXFwpO9NUROQRAELJw9DADwfWYhympaJK6InAGDDbm8734rRGWdDgE+HriavTVETmVEXBDGJIbAbBHx6ZbjUpdDToDBhlxai96IL7Z2zIS6cdZQqD0UEldERL116+wkCAKQcaAUuSUNUpdDDo7Bhlzaup9OollnxOBQb1yayh28iZxR9ABfpI0eBABY8+9jEldDjo7BhlxWWU0LNmV0LO61cE4S5HL+uBM5qwWXD4VCLsOBU9XYf6JK6nLIgfGdnlzWB98chclsQUp8MFKHhkpdDhH1QWiAFldOjAIArPnuGCwWUdqCyGEx2JBL2n+iCruPVkAmE3DHNSMgCILUJRFRH/3PpfHQeCiQd7oR2/eflrocclAMNuRyjCYLVm06AgCYMykag0O9Ja6IiGzB18sD180YAgD46N/H0GbgBpl0LqcINqWlpUhISDjn31dffSV1aeSANmXkoaSyGT6eqs79ZojINVw9NRYh/hrUNLZh4/Y8qcshB+QUc19PnDgBDw8P/Pjjj10uKXh78y9x6qqqTofPz0zvvm3uMG50SeRiVEo5Fs4ehn9+koV1P5/CzHERCPTVSF0WORCn6LE5efIkoqOjERISguDg4M5/arVa6tLIwby38TDaDWYMiwlE+lhO7yZyRZNHDUBCpD/aDWZ8+j0X7aOunCLYnDhxAnFxcVKXQQ5u56Ey7D5aAblMwOL5IzlgmMhFCYKAv1w9HADw495i5Jc2SlwRORKnuBR18uRJBAcH48Ybb0RhYSEiIyOxZMkSTJkyxarziaIInU4HvV4PAJ1fyXb6u21bdEasXH8QAHDV5CgE+yig0+msOpfRaITRYIDBZLFliVZTyiywmC1dajIajV2+OkpdUrNFTbZuW1dtJ2tdtH0tMhiNxh797kYEqzFxRCh+O1yJ974+iCcWjeEfM+j/997+JIpij/4fO3ywMRgMKCwshEajwSOPPAKtVotvvvkGd9xxBz788ENMmDCh1+c0Go3Iycnp/L6wsNCGFdPv9Vfbfp1Zh8YWA4J8FEgKN3T5/9sbcrkcgsoblVWVDjPjwtdLjRB/D1RVV0PfbuhyX01tjURVXbwuqdiyJlu1rau3k7XO175qlQLh/gKqy5thNpu7Pce4GAG7jwJHC+qx8cf9SBzEsTZnuernmkrV/bhJhw82KpUKe/fuhUKh6HxBw4cPR15eHlavXm1VsFEqlYiLi4Ner0dhYSGioqKg0fAXwpb6s233n6zBwYLTEATgvutTEB/h16fz1TcbERoiOsxf155qBbQaLUKCg7v02NTU1iAoMAhKpTS7lZ+vLqnZoiZbt62rtpO1Lta+KoUMQYFB8PcO7/H5CuvV2JhRiG2HdZg9LRkqpdzWJTsVV/5cy83N7dFxDh9sAECr1Z5zW3x8PH799VerzicIQpdzajSa8z4H9Z2927ap1YB3N3bsHTN3SgxGJQ7o8zlb2nRQqlSAzDE+hJQqBWRy2XlrUiqVPfoLpr/rkoota7JV27p6O1ldw3naV6mQQalU9uo944ZZScg4UIHKej2+31OGP8/kEg+Aa36u9fRSo8MPHj5+/DhSUlKQlZXV5fYjR45wQLGbE0URb687iPrmdgwK8cItVyZJXRIR9TOtWonbrxoGAPjqx5OorLNubB25DofvsYmPj8eQIUPw97//HU899RT8/f3x5Zdf4sCBA1i3bp3U5VEftegM0LVZN5Yl83A5dh4qg1wmYNHsJDQ2t/e5HkEADMbur+0TkeOYMmogfthVhEO5NXh/02EsX3SJ1CWRhBw+2MhkMrzzzjt46aWXcP/996OpqQlJSUn48MMPkZDALkdnp2sz4XhRHYy9vM5f19SGz/7TsRDfuGFhaNYbcTiv74M9NR4KhAd59vk8RNR/BEHAnfNG4L6Xf8GuIxXIPl6JMYnc+NZdOXywAYCAgAA899xzUpdBdmI0WXoVbExmC/69swBGkwUDgjwxakhwr4PRhSgVjjEGgoh6JzLMB3OnxGDj9jy8+/VhvPVwEJQK9x5I7K4cfowN0R9l7C9FbWMbNB4KXHZJJGQyrl1BRMANlyXA39sD5TWt3EfKjTHYkFM5XlSHnMI6AMDMcRHw1Egz1ZmIHI9WrcSiuR0Dib/gQGK3xWBDTqOuqQ3b95UCAFKTQjE4lJugElFX00YPwvDYQLQbzHh73UGIoih1SdTPGGzIKRhMZnyfWQiT2YJBIV4YO5QDA4noXIIg4J5rk6FUyLDvRBW27zstdUnUzxhsyOGJoogf9xSjvrkdWrUCM8dFQMY9YYjoAgaFeOP6mfEAgFWbjqCxpe9LQZDzYLAhh7frSDkKypoglwm4YkIUtGqOqyGii/vTtCGICvdBU6sBq785InU51I8YbMih5RTWYd+JagDA9LGDERbINWaIqHtKhQx/vS4ZggBsyz6NfcerpC6J+gmDDTmssuoW/JLdcX18bGIIEiL8Ja6IiJxJQmQA5k6OAQC8tf4g2tqtW+WcnAuDDTmkxpZ2bMkshEUUETvQF+OGhUldEhE5oQVXDEWwvwZVdTp8vCVH6nKoHzDYkMPRt5uweWcB2gxmBPtrMCM1ose7uhIR/Z7GQ4F7rk0GAHyzIx8HT1ZLXBHZG4MNORSD0YzNv+ajobkdXholrpwYDaWCP6ZE7sIef8KMSQzF5ROiAACvfb4PLTqDHZ6FHIVT7BVF7sFstmBLZiGq6vVQq+S4akoMvLiyMJHbkMsEmCwWVNlhxeCrJsdg3/FKVNXr8drn+3HnNSN6/FitWgEvrcrmNZF9MNiQQ7CIIn7cW4zTVS1QyGWYMzkG/j5qqcsion4kkwloM5hRUtlgs41tf2/62MH48seT2H20AgE+aiREdj8hQamQITEygMHGiTDYkOREUcSOA6XIPd0ImSDgyolRCA3QSl0WEUnEaLLYJdgE+WowJjEUWTmV+DmrBCEBWvYKuyAOXiBJiaKInYfKcCSvFgBw6bjB3AOKiOxm7NBQhPhr0G4046e9xdxLygUx2JBkRFHE19vzkJXTsXBWWspADBnMtWqIyH7kMgGXjouAQi7gdFUL9p3gwn2uhsGGJPP5f07g3zsLAACTkwdgeGyQxBURkTvw91ZjyqiBAIDdRypwuqpZ4orIlhhsSBJf/HgCa/9zAgAwddRAJA8JlrgiInInSdGBSIzyhwjgP7uL0aI3Sl0S2QiDDfW7dT+fwidbjgMA5k+Pw+jEEIkrIiJ3lJYyCIG+aujbTfjPrkKYLRxv4woYbKjfiKKIz344jjX/PgYAWHBFIq6cGC1xVUTkrhRyGS6fEAWVQobyWh0yD5dJXRLZAIMN9QtRFLHm38c6Lz/dfMVQXH9pgsRVEZG78/PywIzUCADAwVM1OFXSIG1B1GcMNmR3FouI974+jPXbcgEAf7l6OP7n0niJqyIi6hAz0Bcp8R3j/H7OKkalHVY+pv7DYEN2ZbGIWPHVAWzeWQBBAO65NhlXT42Vuiwioi7GDw9HRJg3TGYR/95ZgKbWdqlLIisx2JDdmMwiXv/yMLbuKYZMAO7/8+jOjeiIiByJTCZg1iWRnYOJN/9agDaDSeqyyAoMNmQXujYTPv2lBruOVkIhl+GRm1ORPnaw1GUREV2QSinHnMkx8NQoUd/cju8zi2A2235rB7IvBhuyufrmNvz9gywUVLZDrZLj6TvGY1LyAKnLIiLqlpdGiTmToqFUyFBa3YIf95Zw2wUnw2BDNlVR24pH3/wVheXN8FTL8NTtY7n4HhE5lSA/DWaNj4QgADmFdfh860mGGyfCYEM2c7K4Hg+/sQPlta0I9dfgtpkhiBngI3VZRES9Fhnmg+ljOi6f/7i3GB98e5ThxkkopC6AXMPOQ2V45dNsGEwWxAzwxSMLklFxOl/qsoiIrDY0KgACgJ+ySrBxex7kMgG3zk6CIAhSl0YXwWBDfSKKIr7+JRcfbu5YTXjs0FA8cvNYiGYDKiSujYior0bEBSEs0BOf/nAc67flQiYTcPMVQxluHBiDDVnNZLbgnQ2H8MOuIgDAnMnR+MvVIyCXCdDpDBJXR0RkG+ljB8NTo8R7Gw/jq59OQRSBW65kuHFUDDZklYbmdvzj4704klcLmQDcfvVwXDWFC+8RkWuaOyUGFlHE+5uOYN3Pp1DX1Ia/XjcKSgWHqjoaBhvqtVMl9XjuX3tR06CHxkOBhxaMwbikMKnLIiKyq6unxkKtUuDt9Qfxc1YJ6hrb8NjCVGjVSqlLo99h1KRe+WlvMR5d8StqGvQYGOyFl++bylBDRG5j1vhIPHHbJVCr5DhwqhqPrvgVtY16qcui32GwoR4xmsx4d8MhvPb5fhhNFoxLCsPL903F4FBvqUsjIupXY4eG4vklk+Hn7YHC8iY89HoGTpXUS10WncFgQ90qrW7BQ2/swOadBQCAGy5LwPJF4+CpYfcrEbmnuMF+eGnpVAwK8UJNYxseeXMHvt2Rz7VuHACDDV2QKIr4aW8x7n/lF+SXNsLHU4Unb78EN85KhEzG2QBE5N5CA7R4celUTBgRDpNZxHsbD+MfH2WhVW+UujS3xsHDdF66NiNWrj+EX/adBgCMjAvCAzeORqCvRuLKiIgch5dGicduTcW3O/Lx4eaj2HmoDPmljXjk5rGIG+wndXluicGGzpGVU4m3vjqAmsY2yGQCbpyVgGvT4yFnLw0R0TkEQcBVU2ORGBWAf3y0F+W1rXjwjQzMS4vFDbMS4aGUS12iW2GwoU6NLe14f9ORzl6a8EBP3H9DCpKiAyWujIjI8cVH+OO1B6bhrXUHsfNgGdZvy8Vvh8vx1+uSMTKOmwH3FwYbgiiK2HGgFO9tPIzGFgNkAnDV1FjcdHki1Cr+iBAR9ZS3VoVlt6Ri15FyrFx/COU1rVi+8jfMHBeBW2cnwdfLQ+oSXR4/tdzcyeJ6rP7mCI4V1AEAIsK8cd/1KYiP8Je4MiIi5zV+eDhGxAZhzXfHsOW3QmzdU4xfD5Zi3rQhuCYtFhoPfvzaC1vWTdU06PHRd8ewLbvjspNKKce16UNwbfoQLhFORGQDnhollsxPRlrKILy/6TByTzdi7Q/H8d3OAvx5ZjwuGx/F91s7YLBxM3VNbdi0PQ+bdxbAYDQDAKaPGYRbrkxCkB9nPBER2dqwmEC8fF8adh4qw8ff5aC8thXvfH0Y63/JxdzJMbjskkiuC2ZDDDZuoqK2FRu25eLHvcUwmiwAgKToANx+1XBediIisjOZTMCUUQMxYUQ4fthVhM+3nkB1vR4ffHsUn/3nBGaNj8TcyTEICdBKXarTY7BxYaIo4kheLb7fVYhfD5bBYulYEXNoVACunTEEqUNDIQicwk1E1F8UchlmT4rGzHER2JZ9GpsyclFS2YKN2/PwTUYeUhJCkD52MC4ZHs5p4lZisHFBtY16/JxVgq17ilFe09p5++iEEFw3YwiGxQQy0BARSUillGPW+EjMHBeB/SersPGXPBw4VY3s41XIPl4FrVqBSSMHYGrKQAyLCeJYnF5gsHERlXU67D5ajt1HKnAkv7azd0bjocDUlIG4fEIU4gb5SVskERF1IZMJGJMYijGJoSirbsG27NP4ObsEVXU6bN1TjK17iqHxUGB0QghSkzqO8/PmlPGLcYpgY7FYsGLFCnz11VdoamrCmDFj8NRTTyEyMlLq0iSjazPieGE9juTXYO+xShSWN3W5Pyk6ADPHRWJy8gCoOa2QiMjhDQj2wk2XJ+KGyxKQU1iHbdkl2H20Ag3N7dh5qAw7D5UB6FiWIyk6EEnRAUiKDkSIv4a98L/jFJ94b7/9Nj7//HM8//zzCA0NxYsvvog77rgDmzdvhkqlkro8u2szmFBS2Yyi8ibknW7EsYI6FJY3wvK7TWRlApAUE4hLhoXjkmFhCA/ylK5gIiKymkwmYFhMIIbFBGLJfBG5pxuw91glsnIqkHu6EcUVzSiuaMb3mYUAAD9vD0SH+yBqgC8GBqph0hkQazTDXYchO3ywMRgM+OCDD/Dwww8jLS0NAPDqq69iypQp2Lp1K2bPni1xhX1nMlvQ3GpAXVMbqur1qKrXoapOh8o6HYorm1FR2wpRPPdxYYFaJEUHInlIEMYODYOPp+uHPCIidyKTCYiP8Ed8hD9uujwRjS3tOFZQh2MFtThWUIu8041oaG7H/uZq7D9Z3fm4d7b8jEBfNcKDPBEe6ImQAC38vdXw9/GAv7cH/L3V8PXycMmxOw4fbI4fP47W1laMHz++8zYfHx8kJSVh7969DhNszGYLjhXUoUVvRLvRDIPRjHaDGe1nvhqMHf+tbzehWWdAU6sBTS0GNLW2o7XN1O35fb1UiAzzQVS4D4ZGB2BoVAB32iYicjO+Xh6YMCIcE0aEA+jo0S8sb0JhWRMKyhqRX9qAgrJGtBtF1Da2obaxDUfyai94Pm+tCj6eSmjUSniqFdCqldD+7quHUg6lQgalXAbl2f9WyKBSyKE4898KmQxyuQCZTIBcJsDfWy3pOCBBFM/XF+A4/vOf/+Dee+/FwYMHoVarO2+/77770NbWhnfffbdX59u3bx9EUYRSqYQoijCZTFAoFH2+PtmiN0LXg4ByMTLhvz8YcvmZrzIBCrkMMifbWbunbWuxiDCYLDhvl5QEBJkAhUwGo9mxaxIBWCxmyGRySPWT4Sxt1Vu2bltXbSdrXax9HbGtIAhQKZznPVgURRiNRsjlClhEwGwRYTZbYLaIsIgiLJaO912Lnds3wMcDCrlte4OMRiMEQcDo0aMvepzD99jo9XoAOGcsjYeHBxobG3t9vrMfsoIgQBAEm43R8daq4K3lpaCzetq2crkAjY1/+G1B4YDds+fW5BgrlTpHW/WW7dvWNdvJWhdvX0dsK2chCAI8PFxz1tTZz+3uOHywOdtLYzAYuvTYtLe3Q6Pp/aWYlJQUm9VGREREjsXhY3F4eMd1xKqqqi63V1VVISwsTIqSiIiIyEE5fLBJTEyEl5cXdu/e3XlbU1MTjh07hrFjx0pYGRERETkah78UpVKpsGDBArz00ksICAjAwIED8eKLLyIsLAwzZ86UujwiIiJyIA4fbABg6dKlMJlM+H//7/+hra0NqampWL16tVsszkdEREQ95/DTvYmIiIh6yuHH2BARERH1FIMNERERuQwGGyIiInIZDDZERETkMhhsiIiIyGUw2BAREZHLYLAhIiIil+EywcZiseCNN97AlClTkJycjNtuuw1FRUUXPL6+vh4PPvggUlNTkZqaiieeeAI6na7LMYcOHcJNN92EkSNHIi0tDW+88QYsFou9X4pDskf7fvvtt5g9ezaSk5Nx5ZVXYv369fZ+GQ6pt237+8fdfvvtePPNN8+5b8uWLbjyyisxYsQIzJ07FxkZGfYo3SnYun0tFgvef/99zJo1C6NGjcLs2bPx1Vdf2at8h2ePn9+zDAYD5s6di2XLltmyZKdhj7Z1i8810UW8+eab4oQJE8RffvlFzMnJEW+77TZx5syZYnt7+3mPX7BggXjdddeJR44cEX/77Tdx+vTp4iOPPNJ5f35+vpicnCwuW7ZMzM/PF7/77jtx1KhR4nvvvddfL8mh2Lp9f/vtNzEpKUn87LPPxOLiYvGTTz4RExMTxZ9//rm/XpLD6G3biqIo6vV68YEHHhDj4+PFN954o8t9mZmZ4rBhw8SPP/5YzM3NFV944QVx+PDhYm5urr1fikOydfu+/fbbYmpqqvjdd9+JRUVF4hdffCEOGzZM3LBhg71fikOydfv+3jPPPCPGx8eLjz76qD1Kd3i2blt3+VxziWDT3t4upqSkiGvXru28rbGxURw5cqS4efPmc47ft2+fGB8f3+WNfseOHWJCQoJYUVEhiqIoPvroo+L8+fNFi8XSeczrr78u3n333XZ8JY7JHu37f//3f+K8efO6PO6aa64R//d//9dOr8Ix9bZtRVEUs7Ozxcsvv1ycMWOGOHbs2HPevG677Tbx/vvv73Lb9ddfLz7xxBO2fwEOzh7tO3XqVHHlypVdbnv88cfFG2+80fYvwMHZo33PysjIECdOnCjOnj3bLYONPdrWXT7XXOJS1PHjx9Ha2orx48d33ubj44OkpCTs3bv3nOOzsrIQHByM2NjYztvGjRsHQRCQnZ0NANixYwfmzJkDQRA6j1m6dClWrlxpx1fimOzRvn5+fsjNzcWuXbsgiiJ2796NvLw8JCcn2/8FOZDeti3Q8bM5c+ZMbNy4Ed7e3l3us1gs2LdvX5fzAcAll1yCrKws278AB2eP9n3hhRdwzTXXnPO4xsZGm9buDGzdvmfV1dXhsccewzPPPAN/f3+71O7o7NG27vK55hSbYHanoqICABAeHt7l9pCQEJSXl59zfGVl5TnHqlQq+Pn5oby8HC0tLaipqYG3tzcef/xxZGRkwMfHB9dccw1uv/12yOVy+70YB2Tr9gWAW265BYcPH8att94KuVwOs9mMO+64A1dddZWdXoVj6m3bAsB99913wfM1NTVBp9MhLCysx+dzZbZuX5lMhgkTJnS57fTp0/j3v/+NP//5z32s1vnYun3PWr58OaZPn4709HR8+OGHfS/UCdm6bd3pc80lemz0ej0AnLPbt4eHB9rb2897/Pl2Bj97fEtLCwDgH//4BwYMGIBVq1bhL3/5C959912sWLHCDq/Asdm6fQGgvLwcDQ0NePLJJ7F+/XosW7YMH330ETZs2GCHV+C4etu23Wlra7Pp+Zydrdv3j6qrq3HnnXciMDAQixcv7vP5nI092vfzzz9HXl4eHnvssT7X58xs3bbu9LnmEj02arUaQMcI+rP/DQDt7e3QaDTnPd5gMJxze3t7O7RaLZRKJQBg4sSJ+Otf/woAGDp0KOrq6vDWW29h6dKlXbryXJ2t2xfo6P6cO3cubrrpJgAd7dvY2Ih//OMfuOaaayCTuUTm7lZv27Y7Hh4enef7PWvP5+xs3b6/l5+fjzvvvBNGoxEff/wxfH19+3Q+Z2Tr9s3Pz8eLL76I1atXd75XuCtbt607fa65xKfH2a66qqqqLrdXVVWd0yUPAGFhYeccazAY0NDQgNDQUPj5+cHDwwPx8fFdjhkyZAh0Oh3q6ups/Aocm63bt66uDgUFBRgxYkSXY0aNGoWGhgY0NDTY9gU4sN62bXf8/Pyg1Wptdj5nZ+v2PSs7Oxt//vOf4eHhgc8//xwRERF9qtNZ2bp9v/vuO7S2tmLRokVISUlBSkoKsrKy8O233yIlJQVlZWU2qdsZ2OO9wV0+11wi2CQmJsLLywu7d+/uvK2pqQnHjh3D2LFjzzk+NTUVFRUVXdYDOPvY0aNHQy6XY/To0Th48GCXx504cQI+Pj7w8/OzzwtxULZuXz8/P2g0Gpw4caLL406ePAkfHx8EBATY6ZU4nt62bXcEQcDo0aOxZ8+eLrfv3r0bY8aM6XO9zsbW7Qt0rAPyl7/8BUOGDMHatWvPGQPhTmzdvgsWLMAPP/yAjRs3dv4bPnw40tPTsXHjRoSEhNiyfIdm67Z1p881l7gUpVKpsGDBArz00ksICAjAwIED8eKLLyIsLAwzZ86E2WxGXV0dvL29oVarkZycjNGjR+Nvf/sbnn76aeh0Ojz11FO45pprEBoaCgBYvHgxFi1ahDfffBNXX301jh49ivfeew8LFy50qUFWPWGP9r311luxcuVKBAcHY8yYMcjOzsY777yDJUuWSPxq+1dv27YnFi1ahDvvvBNJSUmYOnUq1q9fj5ycHDz77LN2fjWOx9btazKZ8NBDDyEwMBAvvPACDAYDqqurAXR8cLhTKAds375+fn7nfMCq1Wp4enoiMjLSTq/CMdnjvcFtPteknm9uKyaTSfznP/8pjh8/Xhw1apR4xx13iCUlJaIoimJJSYkYHx8vrl+/vvP4mpoa8d577xVHjRolXnLJJeJTTz0ltrW1dTlnRkaGOG/ePHHYsGHitGnTxHfffVc0m839+rocha3b12QyiR988IF4+eWXi8nJyeLs2bPFtWvXdllfwV30tm1/b/r06eddB+Trr78WZ86cKY4YMUKcN2+e+Ntvv9n1NTgyW7Zvdna2GB8ff95/06dP75fX42js8fP7ewsWLHDLdWxE0T5t6w6fa4IoiqLU4YqIiIjIFlxijA0RERERwGBDRERELoTBhoiIiFwGgw0RERG5DAYbIiIichkMNkREROQyGGyIiIjIZTDYEJHDc5XltlzldRA5MgYbIheXnp6OZcuW9ekcGzZsQEJCAk6fPg0AePPNN5GQkGCL8rqVm5uLG264wW7nX7NmDSZPnoyRI0fi7bffxsmTJzFv3jwMHz4cV155pc2e56effsKjjz5qs/MR0fm5xF5RRHRhK1asgJeXl03Ped1112HKlCk2PeeFbNmyBfv377fLuVtaWvDCCy8gLS0Nt99+OwYNGoTnn38epaWlWLFiBQIDA232XP/6179sdi4iujAGGyIXl5SUZPNzhoWFISwszObn7W+NjY2wWCyYOXMmUlNTAQD19fWIj4/HtGnTpC2OiKzCS1FELu7spajTp08jISEBW7ZswdKlS5GSkoLU1FQsX74cra2tncdbLBa8/fbbmDZtGpKTk7FkyRI0NjZ2OecfL0XdfPPNeOihh7B06VKMHj0ad955JwCgvb0d//znP5GWlobhw4dj7ty5+O6777qcSxRFfPrpp5g9ezZGjhyJmTNnYtWqVRBFEW+++SZWrFgBAEhISMCbb77Zq9eelZWFBQsWIDk5GePGjcOjjz6Kuro6AB2X19LT0wEAjz/+OBISEpCQkIA9e/Zg7969SEhIwIYNGwAAZWVleOCBBzBu3DgkJyfj1ltvxbFjx7o8V2trK55//nlMnToVo0aNwp/+9Cf8/PPPne2zZ88e7NmzBwkJCdi9e3ePX0N6ejpeffVVPP/88xg3bhzGjRuHhx9+GPX19b1qCyJ3wR4bIjfz1FNPYf78+Xj77bdx6NAhvPrqqwgICMCDDz4IAHjxxRfx0Ucf4e6778aoUaPw/fff4+WXX+72vFu2bMHll1+Ot956C2azGaIo4p577sG+ffuwdOlSxMbGYuvWrfjb3/4Gg8GAa665BgDwyiuvYPXq1Vi4cCEmTZqEo0eP4tVXX4XBYMB1112HiooKrFu3Dl988UWveon27t2LRYsWYfz48XjttdfQ2NiI119/HbfccgvWrVuHadOmYcWKFfjrX/+KxYsXd/bQ/P3vf+9sp4iICNTV1eHPf/4zNBoNnnjiCWg0GqxZswY33XQT1q1bh9jYWFgsFvzlL39BXl5e52vdtGkT/vrXv+LDDz/EU089hYcffrjzvHFxcb34PwasXbsWkZGReO6551BXV4eXX34Z+fn5+OqrryCT8e9Tot9jsCFyM2lpaZ2DWCdMmICdO3fil19+wYMPPoimpiZ8/PHHuOWWW3DvvfcCAKZMmYLKykrs2LHjoueVyWR45plnoNVqAQA7d+7Ejh078Oqrr3YOwp0yZQr0ej1eeuklzJkzBzqdDh9++CFuvvlmPPLIIwCASZMmoa6uDtnZ2bjnnns6w8yoUaN69TpffvllREdH491334VcLgcAJCcnY/bs2Vi/fj1uuukmDB06FAAQERHRef6z45HOfv/qq6+ioaEBn332GQYOHAgAmDp1Kq688kq8/vrreOONN5CRkYF9+/bh7bffxowZMwAA48ePR1FREXbt2oX77rvvnPP2hiAI+PDDD+Ht7Q0ACAgIwD333IOMjAxeMiP6A0Z9Ijfzxw/WsLAw6HQ6AMCBAwdgNBo7P5zPuuKKK7o976BBgzpDDQBkZmZCEASkpaXBZDJ1/ktPT0d1dTVOnTrV+XwzZ87scq5ly5bhgw8+sPIVAnq9HgcPHkRaWhpEUex87sGDByM2NhY7d+7s8bkyMzMxdOhQhIaGdp5HJpNh6tSp+O233wB0XPJSKpWYPn165+MEQcBnn32G++67z+rXcdb06dM7Qw3QcXlKqVQiKyurz+cmcjXssSFyMxqNpsv3Mpmsc32Vs2NpAgICuhwTHBzc7XmDgoK6fN/Q0ABRFDF69OjzHl9VVXXB5+urpqYmWCwWrFq1CqtWrTrnfg8Pjx6fq6GhAUVFRRg2bNh579fr9WhoaICfn5/dLguFhIR0+V4mk8HPzw9NTU12eT4iZ8ZgQ0Sd/P39AQC1tbWIiYnpvL2hoaHX5/L29oZWq8VHH3103vsjIyOxb98+AEBdXV2X5ysvL0dRURHGjBnT6+cFAE9PTwiCgIULF2L27Nnn3P/HcHcx3t7eGDduXOelsj9SqVTw9vZGQ0MDLBZLl3CTk5MDk8mEESNG9P5F/M4f299sNqO+vt7mgZDIFfBSFBF1SklJgVqtxvfff9/l9m3btvX6XOPGjYNOp4MoihgxYkTnv1OnTuGtt96CyWTCyJEjoVQq8dNPP3V57Jo1a3DfffdBEASrekG8vLyQlJSE/Pz8Ls89ZMgQrFixolezksaNG4eCggJER0d3Odc333yDr776CnK5HGPHjoXRaMT27ds7HyeKIpYvX46VK1cCQJ96c3bs2AGDwdD5/U8//QSTyYQJEyZYfU4iV8UeGyLq5OnpiSVLluC1116DRqPB+PHjsX37dquCTVpaGlJTU7FkyRIsWbIEsbGxOHToEN58801Mnjy5s7fhlltuwZo1a6BSqTB+/HgcPnwYn3zyCR544AEoFAr4+PgAADZv3ozk5GQMHjy4R8//wAMP4M4778SDDz6Iq666CmazGR988AEOHjyIxYsX9/h1LFy4EJs2bcLChQtx2223wd/fH9999x2+/PJLPPbYYwCAadOmISUlBY899hjuu+8+REZG4ttvv8XJkyfxxBNPAAB8fHywf/9+ZGZmIikpCb6+vj2uoaKiAosXL8Ytt9yC8vJyvPLKK5g8eTIuueSSHp+DyF0w2BBRF3fddRe0Wi3WrFmDNWvWICUlBY8++iiefvrpXp1HJpPhvffew+uvv453330XtbW1CA0NxcKFC3HPPfd0Hvfwww8jKCgIn332GT744AMMGjQIjz/+OG688UYAwGWXXYZNmzZh2bJluPbaa3tcx+TJk7F69WqsWLECS5cuhVKpxLBhw/Dhhx/2amZSaGgoPv/8c7z88st4+umn0d7ejqioKDz77LO49tprAQByuRyrVq3Cyy+/jDfffBM6nQ6JiYl4//33kZKSAgC46aabcOTIEdxxxx14/vnnMXfu3B7XMHv2bPj4+OD++++HVqvFvHnz8Le//a3HjydyJ4LIXdmIiBxWeno6xo0bhxdeeEHqUoicAntsiMhpiKIIs9nc7XFyuRyCIPRDRdaxWCywWCzdHqdQ8C2aqLf4W0NETuPrr7/uHNdyMc8//zz+9Kc/9UNF1nn88cfx9ddfd3vcHwdVE1H3eCmKiJxGfX09Tp8+3e1xgwYN6py67ohOnz7do72eEhISoFKp+qEiItfBYENEREQug+vYEBERkctgsCEiIiKXwWBDRERELoPBhoiIiFwGgw0RERG5DAYbIiIichkMNkREROQy/j+YboyMlJk9EwAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# mediation stability\n",
    "for p in ['100','200','1000']:\n",
    "    print('number of bootstraps: ' + p)\n",
    "    df = pd.read_csv(path + 'outputs/jobs20/df' + p + '.csv')\n",
    "    for y in ['total_effect_p','direct_effect_p','indirect_effect_p']:\n",
    "        x = df[y]\n",
    "        print('statistic: ' + y)\n",
    "        print('prop under 0.05: ')\n",
    "        print(len(x[x<0.05])/len(x))\n",
    "        print('mean and std')\n",
    "        print(np.mean(x), np.std(x))\n",
    "        print('\\n')\n",
    "    \n",
    "sns.distplot(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "754a719c-f121-424e-9fc3-a9ba86ff5ea5",
   "metadata": {},
   "source": [
    "###### Procrustes ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "68ae796e-af81-4475-bb2d-316fd6e8e413",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_pre_taxa stool_Baseline_meta\n",
      "0.46850198561860146\n",
      "stool_post_taxa stool_After diet_meta\n",
      "0.31613285036266187\n",
      "saliva_pre_taxa saliva_Baseline_meta\n",
      "0.4658971051150502\n",
      "saliva_post_taxa saliva_After diet_meta\n",
      "0.44393321940218644\n",
      "saliva_pre_taxa stool_pre_taxa\n",
      "0.6735031124724702\n",
      "saliva_post_taxa stool_post_taxa\n",
      "0.4247307345866329\n",
      "saliva_Baseline_meta stool_Baseline_meta\n",
      "0.32144702449016405\n",
      "saliva_After diet_meta stool_After diet_meta\n",
      "0.3159891086236145\n"
     ]
    }
   ],
   "source": [
    "# Procrustes\n",
    "import scipy.spatial\n",
    "\n",
    "str_to_df = {\n",
    "    'diet': df_diet\n",
    "}\n",
    "\n",
    "for g in ['stool', 'saliva']:\n",
    "    for t in ['Baseline', 'After diet']:\n",
    "        str_to_df['_'.join([g, t, 'meta'])] = gt_to_meta[g][t]\n",
    "    for t in['pre', 'post']:\n",
    "        df = gt_to_otu[g][t]\n",
    "        df.index = df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "        str_to_df['_'.join([g, t, 'taxa'])] = df\n",
    "\n",
    "\n",
    "def convert_to_pcs(df, npcs):\n",
    "    pca = PCA(n_components=npcs, whiten=True)\n",
    "    pca.fit(df.T)\n",
    "    return pd.DataFrame(pca.components_, \n",
    "                        index=['PC%s' % _ for _ in range(npcs)], \n",
    "                        columns=df.index).T\n",
    "\n",
    "#df1 = convert_to_pcs(df1, 10)\n",
    "#df2 = convert_to_pcs(df2, 10)\n",
    "\n",
    "# convert to distance matrices\n",
    "from scipy.spatial.distance import squareform, pdist\n",
    "from itertools import combinations\n",
    "from skbio import DistanceMatrix\n",
    "from skbio.stats.distance import mantel\n",
    "\n",
    "# get all pairs of dfs\n",
    "# df_pairs = list(combinations(str_to_df, 2))\n",
    "\n",
    "\n",
    "# build comparisons list\n",
    "df_pairs = []\n",
    "# 'Baseline' 'After diet' # for meta\n",
    "# 'pre' 'post' # for stool\n",
    "\n",
    "# stool microbiome nad metabolome at pre and then at post\n",
    "a = ['stool_pre_taxa', 'stool_Baseline_meta']\n",
    "b = ['stool_post_taxa', 'stool_After diet_meta']\n",
    "\n",
    "c = ['saliva_pre_taxa', 'saliva_Baseline_meta']\n",
    "d = ['saliva_post_taxa', 'saliva_After diet_meta']\n",
    "\n",
    "# better fit \n",
    "\n",
    "# saliva to gut in microbiome at pre and then at post\n",
    "e = ['saliva_pre_taxa', 'stool_pre_taxa']\n",
    "f = ['saliva_post_taxa', 'stool_post_taxa']\n",
    "\n",
    "\n",
    "# repeat for metabolome \n",
    "g = ['saliva_Baseline_meta', 'stool_Baseline_meta']\n",
    "h = ['saliva_After diet_meta', 'stool_After diet_meta']\n",
    "\n",
    "# worse fit  \n",
    "\n",
    "df_pairs = [a, b, c, d, e, f, g, h]\n",
    "\n",
    "\n",
    "for d1, d2 in df_pairs:\n",
    "    print(d1, d2)\n",
    "    # retrieve df's \n",
    "    df1, df2 = str_to_df[d1], str_to_df[d2]\n",
    "    \n",
    "    # convert to distance matrix\n",
    "    dm1 = pd.DataFrame(squareform(pdist(df1.iloc[:, 1:])), columns=df1.index.unique(), index=df1.index.unique())\n",
    "    dm2 = pd.DataFrame(squareform(pdist(df2.iloc[:, 1:])), columns=df2.index.unique(), index=df2.index.unique())\n",
    "\n",
    "    # grab common samples\n",
    "    shared_idx = list(set(dm1.index).intersection(set(dm2.index)))\n",
    "    dm1 = dm1.loc[shared_idx,shared_idx]\n",
    "    dm2 = dm2.loc[shared_idx,shared_idx]\n",
    "\n",
    "    # compute mantel \n",
    "    r, p, n = mantel(x=dm1, y=dm2, method='spearman')\n",
    "    if p < 0.05:\n",
    "        print(r,p)\n",
    "\n",
    "    # compute m2\n",
    "    df1p, df2p, m2 = scipy.spatial.procrustes(dm1.values,dm2.values)\n",
    "    print(m2)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eebf72a6-8ff9-4a01-a698-8b8516a9fa3f",
   "metadata": {},
   "source": [
    "###### PERMANOVA ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "15033391-7be6-4a47-b153-1a577639ee28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://github.com/theJohnnyBrown/permanova/blob/master/permanova.py\n",
    "\n",
    "#!/usr/bin/python\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "# Copyright (C) 2010 - 2012, University of New Orleans\n",
    "#\n",
    "# This program is free software; you can redistribute it and/or modify it under\n",
    "# the terms of the GNU General Public License as published by the Free\n",
    "# Software Foundation; either version 2 of the License, or (at your option)\n",
    "# any later version.\n",
    "#\n",
    "# Author: Johnny Brown\n",
    "\n",
    "import numpy as np\n",
    "import random as r\n",
    "from itertools import permutations, product, chain\n",
    "\n",
    "from scipy import stats\n",
    "#stats.ss chokes on generators. Might be worth fixing that.\n",
    "stats.ss = lambda l: sum(a*a for a in l)\n",
    "\n",
    "def above_diagonal(n):\n",
    "    row = range(n)\n",
    "    for i in row:\n",
    "        for j in range(i+1,n):\n",
    "            yield i,j\n",
    "    \n",
    "\n",
    "def permanova_oneway(dm, levels, permutations = 200):\n",
    "    \"\"\" \n",
    "    Performs one-way permutational ANOVA on the given distance matrix.\n",
    "\n",
    "    One-way permanova tests the null hypothesis that distances between levels of\n",
    "    the variable are not significantly different from distances within levels of\n",
    "    the variable.\n",
    "\n",
    "    The test-statistic is a multivariate analogue to Fisher's F-ratio and is \n",
    "    calculated directly from any symmetric distance or dissimilarity matrix. \n",
    "    P-values are then obtained using permutations.\n",
    "\n",
    "    In the case of a Euclidean distance matrix calculated from only one \n",
    "    variable, the F-value is the same as the traditional parametric univariate \n",
    "    F-statistic.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    dm : array_like\n",
    "        The distance matrix of observations x observations. Represents a \n",
    "        symmetric n x n matrix with zeros on the diagonal.\n",
    "\n",
    "    levels : array_like\n",
    "        An array indicating the levels of the variable at each observation, such\n",
    "        that levels[i] == levels[j] means that dm[i] and dm[j] are rows \n",
    "        corresponding to observations in the same level or treatment.\n",
    "\n",
    "    permutations : int\n",
    "        The number of permutations used to compute the p-value. Default is 200.\n",
    "        If there are less than ``permutations`` unique permutations, then all of \n",
    "        them will be used\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    F-value : float\n",
    "        The computed F-value of the test.\n",
    "    p-value : float\n",
    "        The associated p-value from the F-distribution, generated by permuting \n",
    "        the levels\n",
    "\n",
    "    Notes\n",
    "    -----\n",
    "    It is assumed that all sample groups are the same size, n. \n",
    "\n",
    "    For example, if the values in levels are placebo, 5mg, and 10mg, then each \n",
    "    value must occur n times in the levels array, and \n",
    "    3n == len(levels) == len(dm) == len(dm[i]) (for all 0 <= i < len(dm))\n",
    "    \n",
    "    The algorithm is from Anderson[2] \n",
    "\n",
    "    References\n",
    "    ----------\n",
    "    .. [1] Lowry, Richard.  \"Concepts and Applications of Inferential\n",
    "           Statistics\". Chapter 14. http://faculty.vassar.edu/lowry/ch14pt1.html\n",
    "\n",
    "    .. [2] Anderson, Marti. A new method for non-parametric multivariate analysis \n",
    "           of variance. 2001. \n",
    "           http://www.entsoc.org/PDF/MUVE/6_NewMethod_MANOVA1_2.pdf\n",
    "    \n",
    "    .. [3] Jones, James. Lecture notes for Concepts of Statistics\n",
    "           http://people.richland.edu/james/lecture/m170/ch13-2wy.html\n",
    "    \"\"\"\n",
    "    bigf = f_oneway(dm,levels)\n",
    "\n",
    "    above = below = 0\n",
    "    nf = 0\n",
    "\n",
    "    #TODO make this pretty with math and functions\n",
    "    #perms = r.sample(list(perm_unique(levels)),permutations)\n",
    "    shuffledlevels = list(levels)#copy list so we can shuffle it\n",
    "    \n",
    "    for i in range(permutations):\n",
    "        r.shuffle(shuffledlevels)\n",
    "        f = f_oneway(dm,shuffledlevels)\n",
    "        \n",
    "        if f >= bigf:\n",
    "            above += 1\n",
    "\n",
    "        #debug\n",
    "        ## print shuffledlevels\n",
    "        ## print f\n",
    "\n",
    "    p = above/float(permutations)\n",
    "\n",
    "    return (bigf,p)\n",
    "\n",
    "#FIXME the right way to reuse code for one-way and n-way is to add an arg f such that\n",
    "#f(levels[i],levels[j]) returns True iff dm[i][j] should be included in the\n",
    "#desired sum of squares\n",
    "def f_oneway(dm,levels):\n",
    "    bign = len(levels)#number of observations\n",
    "    dm = np.asarray(dm)#distance matrix\n",
    "    a = len(set(levels))#number of levels\n",
    "    n = bign/a#number of observations per level\n",
    "    \n",
    "    assert dm.shape == (bign,bign) #check the dist matrix is square and the size\n",
    "                                   #corresponds to the length of levels\n",
    "\n",
    "    #total sum of squared distances                                   \n",
    "    sst = np.sum(stats.ss(r) for r in \n",
    "              (s[n+1:] for n,s in enumerate(dm[:-1])) )/float(bign)\n",
    "\n",
    "    #sum of within-group squares\n",
    "    #itertools.combinations(range(len(dm)),2)#top half of dm\n",
    "    ssw = np.sum((dm[i][j]**2 for i,j in  \n",
    "               product(range(len(dm)),range(1,len(dm)))\n",
    "               if i<j and levels[i] == levels[j]))/float(n)\n",
    "\n",
    "    ssa = sst - ssw\n",
    "\n",
    "    fstat = (ssa/float(a-1))/(ssw/float(bign-a))\n",
    "    #print (fstat,sst,ssa,ssw,a,bign,n)\n",
    "\n",
    "    return fstat\n",
    "\n",
    "def permanova_twoway(dm,levels,permutations=200):\n",
    "    \"\"\"Performs one-way permutational ANOVA on the given distance matrix.\n",
    "\n",
    "    One-way permanova tests the null hypothesis that distances between levels of\n",
    "    the variable are not significantly different from distances within levels of\n",
    "    the variable.\n",
    "\n",
    "    The test-statistic is a multivariate analogue to Fisher's F-ratio and is \n",
    "    calculated directly from any symmetric distance or dissimilarity matrix. \n",
    "    P-values are then obtained using permutations.\n",
    "\n",
    "    In the case of a Euclidean distance matrix calculated from only one \n",
    "    variable, the F-value is the same as the traditional parametric univariate \n",
    "    F-statistic.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    dm : array_like\n",
    "        The distance matrix of observations x observations. Represents a \n",
    "        symmetric n x n matrix with zeros on the diagonal.\n",
    "\n",
    "    levels : array_like\n",
    "        An array of pairs indicating the levels of the variable at each \n",
    "        observation, such that levels[i] == levels[j] means that dm[i] and dm[j] \n",
    "        are rows corresponding to observations in the same level or treatment.\n",
    "\n",
    "    permutations : int\n",
    "        The number of permutations used to compute the p-value. Default is 200.\n",
    "        If there are less than ``permutations`` unique permutations, then all of \n",
    "        them will be used\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    F-values : 3-tuple\n",
    "        The computed F-values of each test. F-values[0] corresponds to the \n",
    "        interaction of the two levels, F-values[1] corresponds to the effect\n",
    "        of variable a, and F-values[2] corresponds to the effect of variable\n",
    "        b. \n",
    "    p-value : 3-tuple\n",
    "        The associated p-values from the F-distribution, generated by permuting \n",
    "        the levels. The values correspond to (interaction, a, b) as with the \n",
    "        F-values.\n",
    "\n",
    "    Notes\n",
    "    -----\n",
    "    All sample groups must be the same size, n. \n",
    "\n",
    "    For example, if the variables are drug and dose, with levels 5mg, 10mg, 15mg\n",
    "    and drug1, drug2, then each combination of drug type and dose level must \n",
    "    be of size n, i.e. (5mg, drug1) must occur n times in the levels array and\n",
    "    6n == len(levels) == len(dm)\n",
    "\n",
    "    The algorithm is from Anderson[2] \n",
    "\n",
    "    References\n",
    "    ----------\n",
    "    .. [1] Lowry, Richard.  \"Concepts and Applications of Inferential\n",
    "           Statistics\". Chapter 14. http://faculty.vassar.edu/lowry/ch14pt1.html\n",
    "\n",
    "    .. [2] Anderson, Marti. A new method for non-parametric multivariate analysis \n",
    "           of variance. 2001. \n",
    "           http://stg-entsoc.bivings.com/PDF/MUVE/6_NewMethod_MANOVA1_2.pdf\n",
    "\n",
    "    .. [3] Jones, James. Lecture notes for Concepts of Statistics\n",
    "           http://people.richland.edu/james/lecture/m170/ch13-2wy.html\n",
    "    \"\"\"\n",
    "\n",
    "    bigf_i, bigf_a, bigf_b = f_twoway(dm,levels)\n",
    "\n",
    "    above_i = above_a = above_b = 0\n",
    "\n",
    "    #TODO make this pretty with math and functions\n",
    "    #perms = r.sample(list(perm_unique(levels)),permutations)\n",
    "    shuffledlevels = list(levels)#copy list so we can shuffle it\n",
    "\n",
    "    a_levels = list([l[0] for l in levels])\n",
    "    b_levels = list([l[1] for l in levels])\n",
    "    \n",
    "    #permutations\n",
    "    for i in range(permutations):\n",
    "        #All these are probably the wrong way to do the permutations. Wrong as in\n",
    "        #incorrect\n",
    "        ## r.shuffle(a_levels)\n",
    "        ## r.shuffle(b_levels)\n",
    "        ## shuffledlevels = zip(a_levels,b_levels)\n",
    "\n",
    "        r.shuffle(shuffledlevels)\n",
    "        \n",
    "        f_i, f_a, f_b = f_twoway(dm,shuffledlevels)\n",
    "\n",
    "        if f_i > bigf_i:\n",
    "            above_i += 1\n",
    "\n",
    "    for i in range(permutations):\n",
    "        r.shuffle(a_levels)\n",
    "\n",
    "        f_i, f_a, f_b = f_twoway(dm,zip(a_levels, [l[1] for l in levels]))\n",
    "\n",
    "        if f_a > bigf_a:\n",
    "            above_a += 1\n",
    "\n",
    "    for i in range(permutations):\n",
    "        r.shuffle(b_levels)\n",
    "\n",
    "        f_i, f_a, f_b = f_twoway(dm,zip([l[0] for l in levels], b_levels))\n",
    "            \n",
    "        if f_b > bigf_b:\n",
    "            above_b += 1\n",
    "\n",
    "\n",
    "\n",
    "    p_i,p_a,p_b = [ above/float(permutations) for above in \n",
    "                    [above_i, above_a, above_b]]\n",
    "\n",
    "    return ((bigf_i, bigf_a, bigf_b), (p_i, p_a, p_b))\n",
    "\n",
    "    \n",
    "    \n",
    "def f_twoway(dm, levels):\n",
    "    \n",
    "    bign = len(levels)#number of observations\n",
    "    dm = np.asarray(dm)#distance matrix\n",
    "    l = len(set(levels))#number of levels\n",
    "    a = len(set([l[0] for l in levels]))#number of a-levels\n",
    "    b = len(set([l[1] for l in levels]))#number of b-levels\n",
    "    n = bign/float(a*b)#number of observations per level\n",
    "\n",
    "    #sum of all distances\n",
    "    ## sst = np.sum(stats.ss(r) for r in \n",
    "    ##         (s[n+1:] for n,s in enumerate(dm[:-1])) )/float(bign)\n",
    "    sst = stats.ss(chain(*(r[i+1:] for i,r in enumerate(dm))))/float(bign)\n",
    "\n",
    "    #same level of both a and b (error, within-group)\n",
    "    ssr = select_ss(dm, levels, lambda a,b: a==b)/float(n)\n",
    "\n",
    "    #same level of a\n",
    "    sswa = select_ss(dm, levels, lambda a,b: a[0] == b[0])/float(b*n)\n",
    "\n",
    "    #same level of b\n",
    "    sswb = select_ss(dm, levels, lambda a,b: a[1] == b[1])/float(a*n)\n",
    "\n",
    "    \n",
    "    ssa = sst - sswa#effect of a\n",
    "    ssb = sst - sswb#effect of b\n",
    "    ssab = sst - ssa - ssb - ssr #interaction sum-of-squares\n",
    "\n",
    "    #these should each be separate functions?\n",
    "    f_interaction = (ssab/float((a-1)*(b-1)))/(ssr/float(bign - a*b))\n",
    "    f_a = (ssa/float((a-1)))/(ssr/float(bign - a*b))\n",
    "    f_b = (ssb/float((b-1)))/(ssr/float(bign - a*b))\n",
    "\n",
    "    return (f_interaction,f_a,f_b)\n",
    "\n",
    "def select_ss(dm, levels,  included):\n",
    "    \n",
    "    bign = len(dm)\n",
    "\n",
    "    distances = (dm[i][j] for i,j in above_diagonal(bign) \n",
    "                           if included(levels[i], levels[j]))\n",
    "\n",
    "    return stats.ss(distances)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "61203d0e-62c8-4e39-8e85-e37ef753097c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "diet\n",
      "(2.1883126287024086, 0.1)\n",
      "stool_Baseline_meta\n",
      "(0.7309599705911948, 0.64)\n",
      "stool_After diet_meta\n",
      "(1.0518890722123193, 0.235)\n",
      "saliva_Baseline_meta\n",
      "(0.9239761536787098, 0.45)\n",
      "saliva_After diet_meta\n",
      "(0.08004394231598018, 0.68)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VAOAD008</th>\n",
       "      <th>VAOAD011</th>\n",
       "      <th>VAOAD001</th>\n",
       "      <th>OAD004</th>\n",
       "      <th>VAOAD004</th>\n",
       "      <th>OAD001</th>\n",
       "      <th>VAOAD012</th>\n",
       "      <th>OAD006</th>\n",
       "      <th>OAD003</th>\n",
       "      <th>OAD007</th>\n",
       "      <th>OAD012</th>\n",
       "      <th>OAD017</th>\n",
       "      <th>OAD009</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VAOAD008</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.141737</td>\n",
       "      <td>5.291336</td>\n",
       "      <td>7.374733</td>\n",
       "      <td>3.469745</td>\n",
       "      <td>7.804589</td>\n",
       "      <td>5.751012</td>\n",
       "      <td>5.031966</td>\n",
       "      <td>5.131356</td>\n",
       "      <td>6.885489</td>\n",
       "      <td>6.147435</td>\n",
       "      <td>5.693628</td>\n",
       "      <td>5.464290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD011</th>\n",
       "      <td>7.141737</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.766258</td>\n",
       "      <td>6.376212</td>\n",
       "      <td>8.409915</td>\n",
       "      <td>6.783297</td>\n",
       "      <td>6.827353</td>\n",
       "      <td>5.531482</td>\n",
       "      <td>5.249290</td>\n",
       "      <td>7.679534</td>\n",
       "      <td>8.122889</td>\n",
       "      <td>7.303300</td>\n",
       "      <td>6.146555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>5.291336</td>\n",
       "      <td>7.766258</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.348406</td>\n",
       "      <td>5.754649</td>\n",
       "      <td>6.805472</td>\n",
       "      <td>6.221437</td>\n",
       "      <td>5.005707</td>\n",
       "      <td>6.221551</td>\n",
       "      <td>7.627605</td>\n",
       "      <td>4.483618</td>\n",
       "      <td>7.241581</td>\n",
       "      <td>5.189725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>7.374733</td>\n",
       "      <td>6.376212</td>\n",
       "      <td>7.348406</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.787354</td>\n",
       "      <td>3.799295</td>\n",
       "      <td>7.329036</td>\n",
       "      <td>6.306350</td>\n",
       "      <td>6.303166</td>\n",
       "      <td>6.501106</td>\n",
       "      <td>6.413057</td>\n",
       "      <td>6.192060</td>\n",
       "      <td>6.309455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD004</th>\n",
       "      <td>3.469745</td>\n",
       "      <td>8.409915</td>\n",
       "      <td>5.754649</td>\n",
       "      <td>8.787354</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.780640</td>\n",
       "      <td>6.150334</td>\n",
       "      <td>6.608156</td>\n",
       "      <td>6.724492</td>\n",
       "      <td>8.345605</td>\n",
       "      <td>6.708331</td>\n",
       "      <td>7.895603</td>\n",
       "      <td>6.857276</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          VAOAD008  VAOAD011  VAOAD001    OAD004  VAOAD004    OAD001  \\\n",
       "VAOAD008  0.000000  7.141737  5.291336  7.374733  3.469745  7.804589   \n",
       "VAOAD011  7.141737  0.000000  7.766258  6.376212  8.409915  6.783297   \n",
       "VAOAD001  5.291336  7.766258  0.000000  7.348406  5.754649  6.805472   \n",
       "OAD004    7.374733  6.376212  7.348406  0.000000  8.787354  3.799295   \n",
       "VAOAD004  3.469745  8.409915  5.754649  8.787354  0.000000  8.780640   \n",
       "\n",
       "          VAOAD012    OAD006    OAD003    OAD007    OAD012    OAD017    OAD009  \n",
       "VAOAD008  5.751012  5.031966  5.131356  6.885489  6.147435  5.693628  5.464290  \n",
       "VAOAD011  6.827353  5.531482  5.249290  7.679534  8.122889  7.303300  6.146555  \n",
       "VAOAD001  6.221437  5.005707  6.221551  7.627605  4.483618  7.241581  5.189725  \n",
       "OAD004    7.329036  6.306350  6.303166  6.501106  6.413057  6.192060  6.309455  \n",
       "VAOAD004  6.150334  6.608156  6.724492  8.345605  6.708331  7.895603  6.857276  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PERMANOVA\n",
    "df_dx = df_map.drop_duplicates(subset='HostSubjectId')\n",
    "df_dx = df_dx.set_index('HostSubjectId')\n",
    "df_dx.index = df_dx.index.map(lambda x: x.replace('-',''))\n",
    "\n",
    "# construct dict of data\n",
    "str_to_df = {\n",
    "    'diet': df_diet\n",
    "}\n",
    "\n",
    "for g in ['stool', 'saliva']:\n",
    "    for t in ['Baseline', 'After diet']:\n",
    "        str_to_df['_'.join([g, t, 'meta'])] = gt_to_meta[g][t]\n",
    "\n",
    "# this is testing resp vs non resp at pre and post time points only, NOT the paired change in beta post-pre \n",
    "for label in str_to_df:\n",
    "    print(label)\n",
    "    df = str_to_df[label]\n",
    "\n",
    "    # get merged df and drop na\n",
    "    df_merge = pd.concat([df, df_dx], axis=1)\n",
    "    df_merge = df_merge.dropna()\n",
    "    \n",
    "    # construct distance matrix from df\n",
    "    df_data = df.loc[df_merge.index,:]\n",
    "    dx = df_dx.loc[df_merge.index,:]['WOMAC_P_Response']\n",
    "    dm = pd.DataFrame(squareform(pdist(df_data.iloc[:, 1:])), columns=df_data.index.unique(), index=df_data.index.unique())\n",
    "\n",
    "    # takes in distance matrix\n",
    "    print(permanova_oneway(dm, levels=[1 if x == 'No response' else 0 for x in dx]))\n",
    "\n",
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "caed0a56-a43e-4873-a3db-9097292710bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2.1883126287024086, 0.085)\n"
     ]
    }
   ],
   "source": [
    "# PCA diet\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "df = str_to_df['diet']\n",
    "\n",
    "# get merged df and drop na\n",
    "df_merge = pd.concat([df, df_dx], axis=1)\n",
    "df_merge = df_merge.dropna()\n",
    "\n",
    "# construct distance matrix from df\n",
    "df_data = df.loc[df_merge.index,:]\n",
    "dx = df_dx.loc[df_merge.index,:]['WOMAC_P_Response']\n",
    "dm = pd.DataFrame(squareform(pdist(df_data.iloc[:, 1:])), columns=df_data.index.unique(), index=df_data.index.unique())\n",
    "\n",
    "# takes in distance matrix\n",
    "print(permanova_oneway(dm, levels=[1 if x == 'No response' else 0 for x in dx]))\n",
    "\n",
    "npcs = 3\n",
    "pca = PCA(n_components=3, whiten=True)\n",
    "pca.fit(df_data.T)\n",
    "# Reformat and view results\n",
    "loadings = pd.DataFrame(pca.components_,\n",
    "                        index=['PC%s' % _ for _ in range(npcs)], \n",
    "                        columns=df_data.index).T\n",
    "\n",
    "df_merge = pd.concat([df_dx, df_data, loadings], axis=1)\n",
    "plt.title('Diet_PCA')\n",
    "sns.scatterplot(data=df_merge, x='PC1', y='PC2', hue='WOMAC_P_Response')\n",
    "sns.despine()\n",
    "plt.savefig(path + 'outputs/jobs06/PCA_diet_Response.pdf')\n",
    "plt.close() "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8303f089-f331-4356-b6df-487184f8f8e3",
   "metadata": {},
   "source": [
    "###### Scratch ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "864433e8-4624-44d1-acac-900038e98d7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b6f3aa88-2e19-42b3-bd79-3e0b66119a1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saliva\n",
      "diff\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'to_sigotu' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[27], line 16\u001b[0m\n\u001b[1;32m     14\u001b[0m     rvs \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpre\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpost\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m     15\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m comp \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhorizontal\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[0;32m---> 16\u001b[0m     gr_to_sigotu \u001b[38;5;241m=\u001b[39m to_sigotu[comp]\n\u001b[1;32m     17\u001b[0m     rvs \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mResponse\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNo response\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m     18\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m rv \u001b[38;5;129;01min\u001b[39;00m rvs: \u001b[38;5;66;03m# ['pre', 'post']: # Response No response\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'to_sigotu' is not defined"
     ]
    }
   ],
   "source": [
    "# correlations of taxa of interest with metabolites\n",
    "gtr_to_otucorr = {}\n",
    "for g in ['saliva','stool']:\n",
    "    print(g)\n",
    "    gtr_to_otucorr[g] = {}\n",
    "    for time in ['diff','all']:\n",
    "        print(time)\n",
    "        gtr_to_otucorr[g][time] = {}\n",
    "        # grab otu table of differences or of all samples\n",
    "        df_otu = gt_to_otu[g][time]\n",
    "        for comp in ['horizontal', 'vertical']:\n",
    "            if comp == 'vertical':\n",
    "                gr_to_sigotu = to_sigotu[comp]\n",
    "                rvs = ['pre', 'post']\n",
    "            elif comp == 'horizontal':\n",
    "                gr_to_sigotu = to_sigotu[comp]\n",
    "                rvs = ['Response', 'No response']\n",
    "            for rv in rvs: # ['pre', 'post']: # Response No response\n",
    "                print(rv)\n",
    "                sigcorr = []\n",
    "                # grab taxa of interest\n",
    "                taxa_int = gr_to_sigotu[g][rv]\n",
    "    \n",
    "                # merge otu table with metabolites            \n",
    "                df_merge = pd.concat([df_otu.loc[:,taxa_int], gt_to_meta[g][time]], axis=1)\n",
    "                meta = gt_to_meta[g][time].columns\n",
    "                \n",
    "                # look at taxa int and correlate with metabolites\n",
    "                # df_merge \n",
    "                for t in taxa_int:\n",
    "                    for m in meta:\n",
    "                        df_sub = df_merge.loc[:,[t,m]]\n",
    "                        df = type_to_df_map[g].copy()\n",
    "                        if time is 'all':\n",
    "                            df = df.set_index('Together')\n",
    "                            if g == 'stool':\n",
    "                                df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "                        elif time is 'diff':\n",
    "                            df = df[df['Timepoint'] == 'pre']\n",
    "                        df_sub = pd.concat([df_sub, df['WOMAC_P_Response']],axis=1)\n",
    "                        df_sub = df_sub.dropna(subset=[t,m])\n",
    "                        r, p = scipy.stats.spearmanr(df_sub[t].values, df_sub[m].values)\n",
    "                        if p < 0.05:\n",
    "                            sigcorr.append([t,m])\n",
    "                            sns.scatterplot(data=df_sub, x=t, y=m, hue='WOMAC_P_Response')\n",
    "                            sns.despine()\n",
    "                        \n",
    "                            plt.tight_layout()\n",
    "                            plt.savefig(path + 'outputs/' + g_to_job[g + '_adh'] + '/corr_taxa_' + comp + '_' + time + '_' + t.split(';g__')[-1] + '_' + m.replace('/','fslash') + '.pdf')\n",
    "                            plt.close()   \n",
    "                gtr_to_otucorr[g][time][rv] = sigcorr\n",
    "                print(len(sigcorr))\n",
    "           \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f6c73d6-71f4-44ee-ac66-5fe0711fe3f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlations of pathways of interest with metabolites\n",
    "gtr_to_pathcorr = {}\n",
    "for g in ['saliva','stool']:\n",
    "    print(g)\n",
    "    gtr_to_pathcorr[g] = {}\n",
    "    for time in ['diff','all']:\n",
    "        print(time)\n",
    "        gtr_to_pathcorr[g][time] = {}\n",
    "        # grab otu table of differences or of all samples\n",
    "        df_path = gt_to_path[g][time]\n",
    "        for comp in ['horizontal', 'vertical']:\n",
    "            if comp == 'vertical':\n",
    "                gr_to_sigpath = to_sigpath[comp]\n",
    "                rvs = ['pre', 'post']\n",
    "            elif comp == 'horizontal':\n",
    "                gr_to_sigpath = to_sigpath[comp]\n",
    "                rvs = ['Response', 'No response']\n",
    "            for rv in rvs: # ['pre', 'post']: # Response No response\n",
    "                print(rv)\n",
    "                sigcorr = []\n",
    "                # grab taxa of interest\n",
    "                taxa_int = gr_to_sigpath[g][rv]\n",
    "    \n",
    "                # merge otu table with metabolites            \n",
    "                df_merge = pd.concat([df_path.loc[:,taxa_int], gt_to_meta[g][time]], axis=1)\n",
    "                meta = gt_to_meta[g][time].columns\n",
    "                \n",
    "                # look at taxa int and correlate with metabolites\n",
    "                # df_merge \n",
    "                for t in taxa_int:\n",
    "                    for m in meta:\n",
    "                        df_sub = df_merge.loc[:,[t,m]]\n",
    "                        df = type_to_df_map[g].copy()\n",
    "                        if time is 'all':\n",
    "                            df = df.set_index('Together')\n",
    "                            if g == 'stool':\n",
    "                                df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "                        elif time is 'diff':\n",
    "                            df = df[df['Timepoint'] == 'pre']\n",
    "                        df_sub = pd.concat([df_sub, df['WOMAC_P_Response']],axis=1)\n",
    "                        df_sub = df_sub.dropna(subset=[t,m])\n",
    "                        r, p = scipy.stats.spearmanr(df_sub[t].values, df_sub[m].values)\n",
    "                        if p < 0.05:\n",
    "                            sigcorr.append([t,m])\n",
    "                            sns.scatterplot(data=df_sub, x=t, y=m, hue='WOMAC_P_Response')\n",
    "                            sns.despine()\n",
    "                        \n",
    "                            plt.tight_layout()\n",
    "                            plt.savefig(path + 'outputs/' + g_to_job[g + '_adh'] + '/corr_path_' + comp + '_' + time + '_' + t.split(';g__')[-1] + '_' + m.replace('/','fslash') + '.pdf')\n",
    "                            plt.close()   \n",
    "                gtr_to_pathcorr[g][time][rv] = sigcorr\n",
    "                print(len(sigcorr))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5abd2d7-1c78-4e99-a89e-bb864ac5547f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# L Methionine\n",
    "# Kaempferol\n",
    "# Tryptophan\n",
    "# Tyrosine Stool\n",
    "# Usnic acid\n",
    "# Xanthurenic acid\n",
    "# Tryhydroxycholestenoic acid Stool\n",
    "# Isofraxadine\n",
    "int = ['Isofrax', 'Xanth', 'ydroxychol', 'Usnic', 'Tyrosine', 'Tryptophan', 'Kaempferol', 'Methion']\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    for time in ['diff','all']:\n",
    "        for r in ['Response', 'No response']:\n",
    "            test = gtr_to_pathcorr[g][time][r]\n",
    "            for t in test:\n",
    "                p, m = t # unpack pathway, metabolite, from tuple representing sig corr\n",
    "                for i in int:\n",
    "                    if i in m:\n",
    "                        print(t)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f30bf9df-2904-4ca5-b179-4a4eaa5e71f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "int = ['Isofrax', 'Xanth', 'ydroxychol', 'Usnic', 'Tyrosine', 'Tryptophan', 'Kaempferol', 'Methion']\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    for time in ['diff','all']:\n",
    "        for r in ['Response', 'No response']:\n",
    "            test = gtr_to_otucorr[g][time][r]\n",
    "            for t in test:\n",
    "                p, m = t # unpack pathway, metabolite, from tuple representing sig corr\n",
    "                for i in int:\n",
    "                    if i in m:\n",
    "                        print(t)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d51feb2-1356-4615-81a9-ae357b3f93e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CUTIE\n",
    "# In POST samples only\n",
    "# Color by Resp and Nonresp\n",
    "\n",
    "for g in ['saliva', 'stool']:\n",
    "\n",
    "    df_taxa = gt_to_otu[g]['post']\n",
    "    df_taxa.to_csv(path + 'inputs/cutie_df_taxa_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "    \n",
    "    df_abun = pd.concat([gt_to_otu[g]['post'], gt_to_path[g]['post']],axis=1)\n",
    "    df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_map = type_to_df_map[g].set_index('Together')\n",
    "    df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_meta = pd.concat([gt_to_meta[g]['After diet'],df_map], axis=1)\n",
    "\n",
    "    df_abun.to_csv(path + 'inputs/cutie_df_abun_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "    df_meta.to_csv(path + 'inputs/cutie_df_meta_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fdbda6d-641c-47ab-9c83-d392cf0299a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# anaerobic vs aerobic in response\n",
    "# bile acids -> list? saliva and stool of interest \n",
    "#\n",
    "# Bile acids found in Stool: \n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid', 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide']\n",
    "         }\n",
    "\n",
    "mod_ba = {\n",
    "    'stool': [],\n",
    "    'saliva': []\n",
    "}\n",
    "full = list(gt_to_meta[g]['After diet'].columns.values)\n",
    "for g in ['stool','saliva']:\n",
    "    ba = int_ba[g]\n",
    "    for b in ba:\n",
    "        for f in full:\n",
    "            if b in f:\n",
    "                mod_ba[g].append(f)\n",
    "\n",
    "g_to_cutiejob = {\n",
    "    'saliva': 'jobs15',\n",
    "    'stool': 'jobs16'\n",
    "}\n",
    "\n",
    "\n",
    "# from difflib import SequenceMatcher\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    print(g)\n",
    "    n_corr = 0\n",
    "    # import CUITIE results\n",
    "    df_qt = pd.read_csv(path + 'outputs/' + g_to_cutiejob[g] + '/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "    df_qt = df_qt[df_qt['class'] == 'TP']\n",
    "    # iterate through rows\n",
    "    for r in range(len(df_qt)):\n",
    "        meta = df_qt.iloc[r,0]\n",
    "        abun = df_qt.iloc[r,1]\n",
    "        \n",
    "        #for i in int_ba:\n",
    "        #    s = SequenceMatcher(None, meta, i)\n",
    "        #    if s.ratio() > 0.9:\n",
    "        #        print(i)\n",
    "        \n",
    "        #if 'bile' in meta:\n",
    "        #    print(meta)\n",
    "        output = [r]\n",
    "        \n",
    "        for i in int_ba[g]:\n",
    "            if i in meta:\n",
    "                output.append(meta)\n",
    "\n",
    "        for c in ['horizontal','vertical']:\n",
    "            if c == 'vertical':\n",
    "                queries = ['pre', 'post']\n",
    "            if c == 'horizontal':\n",
    "                queries = ['Response','No response']\n",
    "            output.append(c)\n",
    "            for r in queries:\n",
    "                int_taxa = to_sigotu[c][g][r]\n",
    "                for i in int_taxa:\n",
    "                    if i in abun:\n",
    "                        output.append('taxa')\n",
    "                        output.append(i)\n",
    "                int_path = to_sigpath[c][g][r]\n",
    "                for i in int_path:\n",
    "                    if i in abun:\n",
    "                        output.append('path')\n",
    "                        output.append(i)\n",
    "        #output.append()\n",
    "        #output.append()\n",
    "        if len(output) > 5:\n",
    "            # print(output)\n",
    "            n_corr += 1\n",
    "\n",
    "    print(n_corr)\n",
    "\n",
    "# pathway prediction of response vs non response @ baseline\n",
    "# specific taxa at baseline\n",
    "\n",
    "\n",
    "# pathways of interest\n",
    "# ex inflammation in non responders\n",
    "# heatmap of differential pathways in responders/nonresponders \n",
    "# and correlation of those pathways with bacteria\n",
    "# send list of upregulated pathways \n",
    "\n",
    "int_taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76722aff-f034-4fff-a4da-c63dd1fd75ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotations for pathways\n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid', 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide']\n",
    "         }\n",
    "\n",
    "for g in ['saliva','stool']:\n",
    "    # int_ba[g] for metabolite (row) annotations\n",
    "    #df_map = type_to_df_map[g].set_index('Together')\n",
    "    #df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    #df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    #df_meta = pd.concat([gt_to_meta[g]['After diet'],df_map], axis=1)\n",
    "    df_meta = gt_to_meta[g]['After diet']\n",
    "    df_meta_R = df_meta.T\n",
    "    df_meta_R['var_type'] = df_meta_R.index.map(lambda x: 'BA' if x in mod_ba[g] else 'NonBA')\n",
    "    df_meta_R = df_meta_R['var_type'].reset_index()\n",
    "    # df_meta_R['index'] = df_meta_R['index'].str.lower()\n",
    "    df_meta_R = df_meta_R.sort_values(by='index',ascending=True)\n",
    "    df_meta_R.to_csv(path + 'inputs/row_' + g + '.tsv', sep='\\t')\n",
    "    \n",
    "    # dcgr_to_sig for column (taxa) annotations\n",
    "    df_abun = gt_to_otu[g]['post']\n",
    "    df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_abun_R = df_abun.T\n",
    "    df_abun_R['var_type'] = df_abun_R.index.map(lambda x: 'Sig' if x in dcgr_to_sig['taxa']['rvnr'][g]['Response'] else 'Nonsig')\n",
    "    df_abun_R = df_abun_R['var_type'].reset_index()\n",
    "    #df_abun_R['index'] = df_abun_R['index'].str.lower()\n",
    "    df_abun_R = df_abun_R.sort_values(by='index',ascending=True)\n",
    "    df_abun_R.to_csv(path + 'inputs/col_' + g + '.tsv', sep='\\t')\n",
    "\n",
    "df_abun_R.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fbf3e56-2d5f-41fa-8ed7-718d92ed6400",
   "metadata": {},
   "outputs": [],
   "source": [
    "# play testing mediation\n",
    "g = 'stool'\n",
    "df_M = pd.read_csv(path + 'inputs/cutie_df_meta_stool.tsv', sep='\\t', index_col=0)\n",
    "df_M.iloc[:,:-1].to_csv(path + 'inputs/mediation_meta_stool_post_test.txt', sep='\\t')\n",
    "\n",
    "df_Y = df_meta_paired[['VAS_Pt_diff', 'WOMAC_pain_diff']]\n",
    "df_Y.to_csv(path + 'inputs/mediation_WMdiff_test.txt', sep='\\t')\n",
    "\n",
    "df_X = pd.read_csv(path + 'inputs/cutie_df_taxa_stool.tsv', sep='\\t', index_col=0)\n",
    "df_X.index = df_X.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "df_X.to_csv(path + 'inputs/mediation_taxa_stool_post_test.txt', sep='\\t')\n",
    "df_X.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11360b5e-dd8b-4857-b205-4dd8c5c685b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for corr heatmaps \n",
    "for g in ['saliva','stool']:\n",
    "    df_abun = pd.concat([gt_to_otu[g]['post'], gt_to_path[g]['post']],axis=1)\n",
    "    df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_map = type_to_df_map[g].set_index('Together')\n",
    "    df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_meta = pd.concat([gt_to_meta[g]['After diet'],df_map], axis=1)\n",
    "\n",
    "    df_quant_R = df_abun.copy()\n",
    "    #df_quant_R = pd.concat([df_abun, df_meta],axis=1)# .copy() #([df_quant, df_asv100, df_path_top100], axis=1)\n",
    "    # df_quant_R = df_quant_R.drop(['WOMAC_P_Response'],axis=1)\n",
    "    # df_quant_R.to_csv(path + 'inputs/df_quant_' + g + '.tsv', sep='\\t')\n",
    "    df_quant_R = df_quant_R.T\n",
    "    # this is in the order they were entered i.e. \n",
    "    # [df_olink, df_fa, df_acpa_fecal, df_acpa_plasma, df_plasma, df_rbfa, df_mb]\n",
    "     # ['rbfa']*len(df_rbfa.columns) + \\  \n",
    "    df_quant_R['var_type'] = ['Taxa']*len(gt_to_otu[g]['post'].columns) + \\\n",
    "                             ['Pathways']*len(gt_to_path[g]['post'].columns) # + \\\n",
    "                             # ['Metabolite']*len(gt_to_meta[g]['After diet'].columns)\n",
    "    # df_quant_R = df_quant_R[~df_quant_R['var_type'].isin(['Metagenomic_ASVs','Metagenomic_Pathways'])]\n",
    "    df_quant_R = df_quant_R['var_type'].reset_index()\n",
    "    df_quant_R['index'] = df_quant_R['index'].str.lower()\n",
    "    df_quant_R = df_quant_R.sort_values(by='index',ascending=True)\n",
    "    df_quant_R.to_csv(path + 'inputs/df_quant_R_corr_labels_' + g + '.tsv', sep='\\t')\n",
    "    print(len(df_quant_R))\n",
    "df_quant_R.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2707d857-ee9c-466f-983c-3d94600478d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4 way paired boxplot trial\n",
    "g = 'saliva'\n",
    "df_alpha = g_to_dfd[g + '_adh']['alpha']\n",
    "\n",
    "df_map = type_to_df_map[g].copy()\n",
    "df_map = df_map.set_index('Together')\n",
    "df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "\n",
    "df_merge = pd.concat([df_alpha, df_map], axis=1)\n",
    "df_merge = df_merge.dropna()\n",
    "\n",
    "sns.boxplot(x='Timepoint', y='shannon_entropy',\n",
    "            hue='WOMAC_P_Response', palette=[\"r\", \"g\"],\n",
    "            data=df_merge)\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a341d51b-1db7-4c98-b4e8-65a312f999e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test if each sub correlation is sig\n",
    "m = 'H_Nonhydroxylated_bile_acid'\n",
    "y = 'WOMAC_pain_diff'\n",
    "y = 'VAS_Pt_diff'\n",
    "x = 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;__'\n",
    "\n",
    "df = df_plot[[x,y,m]]\n",
    "\n",
    "df_corr = df.dropna()\n",
    "print('Y on X')\n",
    "print(scipy.stats.pearsonr(df_corr[y].values, df_corr[x].values))\n",
    "print('M on X')\n",
    "print(scipy.stats.pearsonr(df_corr[m].values, df_corr[x].values))\n",
    "\n",
    "df['meta'] = df[m].apply(lambda x: 'high' if -1.25 <= x <= 0.25 else 'low')\n",
    "\n",
    "# do each linear model separately\n",
    "df_corr = df[df['meta'] == 'low']\n",
    "df_corr = df_corr.dropna()\n",
    "print(scipy.stats.pearsonr(df_corr[y].values, df_corr[x].values))\n",
    "\n",
    "df_corr = df[df['meta'] == 'high']\n",
    "df_corr = df_corr.dropna()\n",
    "print(scipy.stats.pearsonr(df_corr[y].values, df_corr[x].values))\n",
    "# df_corr\n",
    "df.columns = ['Lachno','Pain','H_Nonhydroxylated_BA', 'meta']\n",
    "df.to_csv(path + 'inputs/df_med_R.tsv', sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "101b41e6-d1b0-4d18-a49c-7c338f54078d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "061973e1-2fae-4d91-8847-ea6ce3450a9e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8ad3a4bc-37a9-4ae4-9075-376b97b35337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OAD007      207\n",
       "OAD005        9\n",
       "OAD001        8\n",
       "OAD004        8\n",
       "OAD012        5\n",
       "OAD003        3\n",
       "OAD006        2\n",
       "VAOAD004      2\n",
       "VAOAD012      2\n",
       "OAD017        2\n",
       "VAOAD011      1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = gt_to_otu['saliva']['post']\n",
    "\n",
    "uniques = []\n",
    "from collections import defaultdict\n",
    "uniq_to_taxa = defaultdict(list)\n",
    "for t in df.columns:\n",
    "    if df[t].nunique() == 2:\n",
    "        x = df[t]\n",
    "        x = x[x != 0]\n",
    "        uniques.append(x.index[0])\n",
    "        uniq_to_taxa[x.index[0]].append(t)\n",
    "pd.Series(uniques).value_counts()\n",
    "# df = df.astype(bool).sum(axis=1)\n",
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c2faeb2e-f734-4d0c-b0a4-8d9b5421a6fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Caldalkalibacillales;__;__',\n",
       " 'd__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Nevskiales;f__Nevskiaceae;__',\n",
       " 'd__Bacteria;p__Firmicutes_D;c__Bacilli;o__Bacillales_B_310392;f__Bacillaceae_G_310392;g__Bacillus_A',\n",
       " 'd__Bacteria;p__Actinobacteriota;c__Coriobacteriia;o__Coriobacteriales;f__Eggerthellaceae;g__Cryptobacterium',\n",
       " 'd__Bacteria;p__Firmicutes_C;c__Negativicutes;o__Selenomonadales;f__Selenomonadaceae_42771;__',\n",
       " 'd__Bacteria;p__Firmicutes_D;c__Bacilli;o__Bacillales_B_302584;f__Bacillaceae_C;g__Weizmannia',\n",
       " 'd__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Glutamicibacter',\n",
       " 'd__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400',\n",
       " 'd__Bacteria;p__Firmicutes_D;c__Bacilli;o__Tumebacillales;f__Tumebacillaceae;g__Tumebacillus_A']"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniq_to_taxa['OAD005']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6219d1d-abce-4661-8592-bca518dab9ee",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
