{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3f62dc73-df18-4548-bc0d-aea2a125fef0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "import pandas as pd\n",
    "import scipy.stats\n",
    "import statsmodels.stats.multitest\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "\n",
    "# disable warnings, use w caution\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "# project specific libs\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import pathlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a8b42930-bcde-48f3-9753-8779e1aa11fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# project specific path\n",
    "path = '/Users/KevinBu/Desktop/clemente_lab/Projects/oa/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0343d6a9-1c06-4ed7-90a1-ed193002a1dd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Process Mapping Files ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a3013747-2e0e-412a-b719-b2a6a28afbf2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>Separate</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Together</th>\n",
       "      <th>ContactEmail</th>\n",
       "      <th>ContactName</th>\n",
       "      <th>PrimaryInvestigator</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>RawDataNotes</th>\n",
       "      <th>...</th>\n",
       "      <th>broccoli</th>\n",
       "      <th>Garbanzo_beans</th>\n",
       "      <th>pork</th>\n",
       "      <th>beef</th>\n",
       "      <th>burger</th>\n",
       "      <th>Total_omega3</th>\n",
       "      <th>Adherence_omega3</th>\n",
       "      <th>Total_omega6</th>\n",
       "      <th>Adherence_omega6</th>\n",
       "      <th>Total_o3_o6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.plasma</th>\n",
       "      <td>TTCAGTTCGTTA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-001.pre.plasma.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.plasma</th>\n",
       "      <td>CGGCCAGAAGCA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>post</td>\n",
       "      <td>OAD-001.post.plasma.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001.post.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>72.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>131.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.plasma</th>\n",
       "      <td>GACGTTAAGAAT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-003.pre.plasma.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>53.2</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.plasma</th>\n",
       "      <td>TCGCTACAGATG</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>post</td>\n",
       "      <td>OAD-003.post.plasma.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003.post.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>108.0</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>171.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.plasma</th>\n",
       "      <td>ATGGGACCTTCA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-004.pre.plasma.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-004.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>2.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>28.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>35.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 188 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    BarcodeSequence    LinkerPrimerSequence Separate  \\\n",
       "#SampleID                                                              \n",
       "OAD-001.pre.plasma     TTCAGTTCGTTA  CCGGACTACHVGGGTWTCTAAT      All   \n",
       "OAD-001.post.plasma    CGGCCAGAAGCA  CCGGACTACHVGGGTWTCTAAT      All   \n",
       "OAD-003.pre.plasma     GACGTTAAGAAT  CCGGACTACHVGGGTWTCTAAT      All   \n",
       "OAD-003.post.plasma    TCGCTACAGATG  CCGGACTACHVGGGTWTCTAAT      All   \n",
       "OAD-004.pre.plasma     ATGGGACCTTCA  CCGGACTACHVGGGTWTCTAAT      All   \n",
       "\n",
       "                    Timepoint                           Together  \\\n",
       "#SampleID                                                          \n",
       "OAD-001.pre.plasma        pre   OAD-001.pre.plasma.guma.plate313   \n",
       "OAD-001.post.plasma      post  OAD-001.post.plasma.guma.plate313   \n",
       "OAD-003.pre.plasma        pre   OAD-003.pre.plasma.guma.plate313   \n",
       "OAD-003.post.plasma      post  OAD-003.post.plasma.guma.plate313   \n",
       "OAD-004.pre.plasma        pre   OAD-004.pre.plasma.guma.plate313   \n",
       "\n",
       "                                     ContactEmail    ContactName  \\\n",
       "#SampleID                                                          \n",
       "OAD-001.pre.plasma   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-001.post.plasma  rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-003.pre.plasma   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-003.post.plasma  rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-004.pre.plasma   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "\n",
       "                    PrimaryInvestigator Cohort  \\\n",
       "#SampleID                                        \n",
       "OAD-001.pre.plasma           Jose Scher  NonVA   \n",
       "OAD-001.post.plasma          Jose Scher  NonVA   \n",
       "OAD-003.pre.plasma           Jose Scher  NonVA   \n",
       "OAD-003.post.plasma          Jose Scher  NonVA   \n",
       "OAD-004.pre.plasma           Jose Scher  NonVA   \n",
       "\n",
       "                                         RawDataNotes  ... broccoli  \\\n",
       "#SampleID                                              ...            \n",
       "OAD-001.pre.plasma    OAD-001.pre.stool.guma.plate313  ...      0.0   \n",
       "OAD-001.post.plasma  OAD-001.post.stool.guma.plate313  ...      4.0   \n",
       "OAD-003.pre.plasma    OAD-003.pre.stool.guma.plate313  ...      0.0   \n",
       "OAD-003.post.plasma  OAD-003.post.stool.guma.plate313  ...      4.0   \n",
       "OAD-004.pre.plasma    OAD-004.pre.stool.guma.plate313  ...      2.8   \n",
       "\n",
       "                    Garbanzo_beans pork  beef burger Total_omega3  \\\n",
       "#SampleID                                                           \n",
       "OAD-001.pre.plasma             0.0  0.0   0.0    0.0          NaN   \n",
       "OAD-001.post.plasma            0.0  0.0   0.0    0.0         48.0   \n",
       "OAD-003.pre.plasma             0.0  0.0  14.0    0.0         16.8   \n",
       "OAD-003.post.plasma            0.0  0.0   0.0    0.0         48.0   \n",
       "OAD-004.pre.plasma             0.0  0.0   0.0    0.0          2.8   \n",
       "\n",
       "                    Adherence_omega3 Total_omega6 Adherence_omega6 Total_o3_o6  \n",
       "#SampleID                                                                       \n",
       "OAD-001.pre.plasma     Low adherence          NaN    Low adherence         0.0  \n",
       "OAD-001.post.plasma    Low adherence         72.0    Low adherence       131.0  \n",
       "OAD-003.pre.plasma     Low adherence         53.2   High adherence        75.0  \n",
       "OAD-003.post.plasma    Low adherence        108.0   High adherence       171.0  \n",
       "OAD-004.pre.plasma     Low adherence         28.0    Low adherence        35.8  \n",
       "\n",
       "[5 rows x 188 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# from AC Q2 run of merged saliva stool\n",
    "df_map = pd.read_csv(path + 'inputs/Qiime2_0/qiime_mapping_file.tsv', sep='\\t', index_col=0)\n",
    "q2_row = df_map.loc['#q2:types',:]\n",
    "df_map = df_map.drop('#q2:types')\n",
    "\n",
    "# change index so it matches metadata file\n",
    "df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "\n",
    "# drop MOC and elution buffer\n",
    "df_map = df_map.drop(['MOC.320','elutionbuffer.plate313'])\n",
    "\n",
    "# grab metadata\n",
    "df_meta = pd.read_csv(path + 'inputs/Metadata_OA.csv')\n",
    "\n",
    "# rename 'Run_ID_Saliva' to be correct\n",
    "df_meta['Timepoints'] = df_meta['Timepoints'].apply(lambda x: 'pre' if x == '0' else 'post')\n",
    "df_meta['Patient_ID'] = df_meta['Patient_ID'].apply(lambda x: x[:-3])  \n",
    "df_meta['Study_ID'] = df_meta['Study_ID'].apply(lambda x: x.split('_')[0][-3:]) \n",
    "df_outcomes = df_meta.copy()\n",
    "\n",
    "# create per sample type mapping files\n",
    "type_to_ST = {'saliva':'Saliva','stool':'fecal'}\n",
    "type_to_df_map = {}\n",
    "\n",
    "# split into specimen type\n",
    "for t in type_to_ST:\n",
    "    # subset on specimen type\n",
    "    df_map_type = df_map[df_map['SpecimenType'] == type_to_ST[t]]\n",
    "\n",
    "    # as to not overwrite df meta\n",
    "    df_meta_type = df_meta.copy()\n",
    "\n",
    "    # create new sample ID for specimen type and set as index\n",
    "    df_meta_type['#SampleID'] = df_meta['Patient_ID'] + '-' + df_meta['Study_ID'] + '.' + df_meta['Timepoints'] + '.' + t\n",
    "    df_meta_type = df_meta_type.set_index('#SampleID')\n",
    "\n",
    "    # create full mapping file\n",
    "    df_map_type = pd.concat([df_map_type, df_meta_type],axis=1)\n",
    "\n",
    "    # use only sequenced samples\n",
    "    df_map_type = df_map_type.dropna(how='any',subset='BarcodeSequence')\n",
    "\n",
    "    # drop all na columns\n",
    "    df_map_type = df_map_type.dropna(how='all',axis=1)\n",
    "\n",
    "    # drop VAD OA 015 because misdx with PsA not OA\n",
    "    # drop OAD007 in saliva bc possible contaminants\n",
    "    if t == 'saliva':\n",
    "        df_map_type = df_map_type.drop(['VAOAD-015.pre.saliva','VAOAD-015.post.saliva'])\n",
    "        # df_map_type = df_map_type.drop(['OAD-007.pre.saliva','OAD-007.post.saliva'])\n",
    "\n",
    "    if t == 'stool':\n",
    "        df_map_type = df_map_type.drop(['VAOAD-015.pre.stool','VAOAD-015.post.stool'])\n",
    "\n",
    "    # populate dict of mapping files\n",
    "    type_to_df_map[t] = df_map_type\n",
    "\n",
    "    # export for q2\n",
    "    df_q2_type = pd.concat([q2_row.to_frame().T, df_map_type])\n",
    "    df_q2_type.index.name = '#SampleID'\n",
    "    df_q2_type.iloc[0,:] = 'categorical'\n",
    "    df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + t + '.tsv', sep='\\t')\n",
    "    df_q2_type = df_q2_type[df_q2_type['Adherece_antiinflam'].isin(['Moderate adherence', 'High adherence','categorical'])]\n",
    "    df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh.tsv', sep='\\t')\n",
    "\n",
    "    # split by response category\n",
    "    df_q2_resp = df_q2_type[df_q2_type['WOMAC_P_Response'].isin(['categorical', 'Response'])]\n",
    "    df_q2_resp.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_response.tsv', sep='\\t')\n",
    "    df_q2_nonresp = df_q2_type[df_q2_type['WOMAC_P_Response'].isin(['categorical', 'No response'])]\n",
    "    df_q2_nonresp.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_noresponse.tsv', sep='\\t')\n",
    "\n",
    "    # split by timepoint\n",
    "    df_q2_pre = df_q2_type[df_q2_type['Timepoint'].isin(['pre','categorical'])]\n",
    "    df_q2_pre = df_q2_pre.dropna(subset='WOMAC_P_Response')\n",
    "    df_q2_pre.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_pre.tsv', sep='\\t')\n",
    "    df_q2_post = df_q2_type[df_q2_type['Timepoint'].isin(['post', 'categorical'])]\n",
    "    df_q2_post = df_q2_post.dropna(subset='WOMAC_P_Response')\n",
    "    df_q2_post.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_post.tsv', sep='\\t')\n",
    "\n",
    "# create new fake df map for plasma \n",
    "df_plasma = type_to_df_map['stool'].copy()\n",
    "df_plasma.index = df_plasma.index.map(lambda x: x.replace('.stool','.plasma'))\n",
    "df_plasma['Together'] = df_plasma['Together'].map(lambda x: x.replace('.stool','.plasma'))\n",
    "type_to_df_map['plasma'] = df_plasma\n",
    "\n",
    "type_to_df_map['plasma'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0f862250-1b36-4345-bd4a-ec9600c20d36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>Separate</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Together</th>\n",
       "      <th>ContactEmail</th>\n",
       "      <th>ContactName</th>\n",
       "      <th>PrimaryInvestigator</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>RawDataNotes</th>\n",
       "      <th>...</th>\n",
       "      <th>broccoli</th>\n",
       "      <th>Garbanzo_beans</th>\n",
       "      <th>pork</th>\n",
       "      <th>beef</th>\n",
       "      <th>burger</th>\n",
       "      <th>Total_omega3</th>\n",
       "      <th>Adherence_omega3</th>\n",
       "      <th>Total_omega6</th>\n",
       "      <th>Adherence_omega6</th>\n",
       "      <th>Total_o3_o6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>TTCAGTTCGTTA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-001.pre.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>CGGCCAGAAGCA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>post</td>\n",
       "      <td>OAD-001.post.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001.post.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>72.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>131.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>GACGTTAAGAAT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-003.pre.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>53.2</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>TCGCTACAGATG</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>post</td>\n",
       "      <td>OAD-003.post.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003.post.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>108.0</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>171.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>ATGGGACCTTCA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-004.pre.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-004.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>2.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>28.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>35.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 188 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   BarcodeSequence    LinkerPrimerSequence Separate Timepoint  \\\n",
       "#SampleID                                                                       \n",
       "OAD-001.pre.stool     TTCAGTTCGTTA  CCGGACTACHVGGGTWTCTAAT      All       pre   \n",
       "OAD-001.post.stool    CGGCCAGAAGCA  CCGGACTACHVGGGTWTCTAAT      All      post   \n",
       "OAD-003.pre.stool     GACGTTAAGAAT  CCGGACTACHVGGGTWTCTAAT      All       pre   \n",
       "OAD-003.post.stool    TCGCTACAGATG  CCGGACTACHVGGGTWTCTAAT      All      post   \n",
       "OAD-004.pre.stool     ATGGGACCTTCA  CCGGACTACHVGGGTWTCTAAT      All       pre   \n",
       "\n",
       "                                            Together  \\\n",
       "#SampleID                                              \n",
       "OAD-001.pre.stool    OAD-001.pre.stool.guma.plate313   \n",
       "OAD-001.post.stool  OAD-001.post.stool.guma.plate313   \n",
       "OAD-003.pre.stool    OAD-003.pre.stool.guma.plate313   \n",
       "OAD-003.post.stool  OAD-003.post.stool.guma.plate313   \n",
       "OAD-004.pre.stool    OAD-004.pre.stool.guma.plate313   \n",
       "\n",
       "                                    ContactEmail    ContactName  \\\n",
       "#SampleID                                                         \n",
       "OAD-001.pre.stool   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-001.post.stool  rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-003.pre.stool   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-003.post.stool  rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-004.pre.stool   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "\n",
       "                   PrimaryInvestigator Cohort  \\\n",
       "#SampleID                                       \n",
       "OAD-001.pre.stool           Jose Scher  NonVA   \n",
       "OAD-001.post.stool          Jose Scher  NonVA   \n",
       "OAD-003.pre.stool           Jose Scher  NonVA   \n",
       "OAD-003.post.stool          Jose Scher  NonVA   \n",
       "OAD-004.pre.stool           Jose Scher  NonVA   \n",
       "\n",
       "                                        RawDataNotes  ... broccoli  \\\n",
       "#SampleID                                             ...            \n",
       "OAD-001.pre.stool    OAD-001.pre.stool.guma.plate313  ...      0.0   \n",
       "OAD-001.post.stool  OAD-001.post.stool.guma.plate313  ...      4.0   \n",
       "OAD-003.pre.stool    OAD-003.pre.stool.guma.plate313  ...      0.0   \n",
       "OAD-003.post.stool  OAD-003.post.stool.guma.plate313  ...      4.0   \n",
       "OAD-004.pre.stool    OAD-004.pre.stool.guma.plate313  ...      2.8   \n",
       "\n",
       "                   Garbanzo_beans pork  beef burger Total_omega3  \\\n",
       "#SampleID                                                          \n",
       "OAD-001.pre.stool             0.0  0.0   0.0    0.0          NaN   \n",
       "OAD-001.post.stool            0.0  0.0   0.0    0.0         48.0   \n",
       "OAD-003.pre.stool             0.0  0.0  14.0    0.0         16.8   \n",
       "OAD-003.post.stool            0.0  0.0   0.0    0.0         48.0   \n",
       "OAD-004.pre.stool             0.0  0.0   0.0    0.0          2.8   \n",
       "\n",
       "                   Adherence_omega3 Total_omega6 Adherence_omega6 Total_o3_o6  \n",
       "#SampleID                                                                      \n",
       "OAD-001.pre.stool     Low adherence          NaN    Low adherence         0.0  \n",
       "OAD-001.post.stool    Low adherence         72.0    Low adherence       131.0  \n",
       "OAD-003.pre.stool     Low adherence         53.2   High adherence        75.0  \n",
       "OAD-003.post.stool    Low adherence        108.0   High adherence       171.0  \n",
       "OAD-004.pre.stool     Low adherence         28.0    Low adherence        35.8  \n",
       "\n",
       "[5 rows x 188 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type_to_df_map['stool'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "248b05fc-430d-4a56-9d93-bfc3334b840a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Assess Outcomes ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "95e4ef90-0416-4803-a5b2-d730031630c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n",
      "['VAS_Pt', 'VAS_overall', 'WOMAC_pain', 'WOMAC_stiffness', 'WOMAC_activity', 'WOMAC_total', 'Pain_DETECT', 'CES_D', 'Helplesness', 'Magnification', 'Rumination', 'PCS_EN', 'Sleep_distrubance', 'PASE_walk', 'PASE_light', 'BMI']\n",
      "32\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var</th>\n",
       "      <th>effect</th>\n",
       "      <th>pval</th>\n",
       "      <th>stat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VAS_Pt</td>\n",
       "      <td>198.5</td>\n",
       "      <td>0.008316</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VAS_overall</td>\n",
       "      <td>168.5</td>\n",
       "      <td>0.131314</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_pain</td>\n",
       "      <td>170.0</td>\n",
       "      <td>0.017900</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_stiffness</td>\n",
       "      <td>172.5</td>\n",
       "      <td>0.013052</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_activity</td>\n",
       "      <td>164.5</td>\n",
       "      <td>0.032592</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               var  effect      pval stat\n",
       "0           VAS_Pt   198.5  0.008316  MWU\n",
       "0      VAS_overall   168.5  0.131314  MWU\n",
       "0       WOMAC_pain   170.0  0.017900  MWU\n",
       "0  WOMAC_stiffness   172.5  0.013052  MWU\n",
       "0   WOMAC_activity   164.5  0.032592  MWU"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###\n",
    "# Hypothesis 1: There will be a measurable difference in WOMAC pain response scores and \n",
    "# other outcomes from baseline to after the dietary intervention.\n",
    "###\n",
    "\n",
    "# outcome variables\n",
    "outcomes = ['VAS_Pt', 'VAS_overall', 'WOMAC_pain', 'WOMAC_stiffness', 'WOMAC_activity', 'WOMAC_total', 'Pain_DETECT', \n",
    "            'CES_D', 'Helplesness', 'Magnification', 'Rumination', 'PCS_EN', 'Sleep_distrubance', 'PASE_walk', 'PASE_light', \n",
    "            #'PASE_gardening', # Where did this go? gardening_improve is binary\n",
    "            'BMI']\n",
    "\n",
    "\n",
    "# hypothesis 1\n",
    "# create a new df_meta\n",
    "df_meta = pd.read_csv(path + 'inputs/Metadata_OA.csv')\n",
    "\n",
    "# rename 'Run_ID_Saliva' to be correct\n",
    "df_meta['Timepoints'] = df_meta['Timepoints'].apply(lambda x: 'pre' if x == '0' else 'post')\n",
    "df_meta['Patient_ID'] = df_meta['Patient_ID'].apply(lambda x: x[:-3])  \n",
    "df_meta['Study_ID'] = df_meta['Study_ID'].apply(lambda x: x.split('_')[0][-3:]) \n",
    "df_meta['#SampleID'] = df_meta['Patient_ID'] + '-' + df_meta['Study_ID'] + '.' + df_meta['Timepoints'] + '.stool'\n",
    "df_meta = df_meta.set_index('#SampleID')\n",
    "\n",
    "# convert % to floats for calculations down the road\n",
    "bin = []\n",
    "cont = []\n",
    "df_paired_os = []\n",
    "for w in outcomes:\n",
    "    df_w = df_meta[w]\n",
    "    if df_w.nunique() > 2: # do spearman\n",
    "        df_meta[w] = df_meta[w].astype(str).str.replace('%','').astype(float).values\n",
    "        cont.append(w)\n",
    "    else:\n",
    "        bin.append(w)\n",
    "\n",
    "    # compute difference and store it\n",
    "    df_md = df_meta.copy()\n",
    "    df_md['SubjectID'] = df_md['Patient_ID'] + df_md['Study_ID']\n",
    "    \n",
    "    # first drop unpaired samples\n",
    "    s_remove = []\n",
    "    for s in list(df_md['SubjectID'].values):\n",
    "        if len(df_md[df_md['SubjectID'] == s]) != 2:\n",
    "            s_remove.append(s)\n",
    "    df_md = df_md.loc[~df_md['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "    \n",
    "    # set vars\n",
    "    group_var = 'Timepoints'\n",
    "    pair_var = 'SubjectID'\n",
    "    groups = ['pre','post']\n",
    "    \n",
    "    # get paired per indiv pair\n",
    "    pair_to_diff = {}\n",
    "    for p in list(df_md[pair_var].values):\n",
    "        df = df_md[df_md[pair_var] == p]\n",
    "        t0 = float(df[df[group_var] == groups[0]][w].values)\n",
    "        tf = float(df[df[group_var] == groups[1]][w].values)\n",
    "        pair_to_diff[p] = tf - t0\n",
    "    \n",
    "    df_paired_o = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[w + '_diff'])\n",
    "    df_paired_os.append(df_paired_o)\n",
    "\n",
    "df_meta_paired = pd.concat([*df_paired_os], axis=1)    \n",
    "\n",
    "print(bin)\n",
    "print(cont)\n",
    "\n",
    "# split into all and mod high only\n",
    "a = 'modhigh'\n",
    "job = 'jobs02'\n",
    "df_meta = df_meta[df_meta['Adherece_antiinflam'].isin(['Moderate adherence', 'High adherence'])]\n",
    "        \n",
    "print(len(df_meta))\n",
    "\n",
    "df_results = pd.DataFrame(columns=['var','effect','pval','stat'])\n",
    "# do post treatment vals of binary vars differ from pre treatment 'unpaired'\n",
    "for b in bin:\n",
    "    ct_table_ind=pd.crosstab(df_meta[\"Timepoints\"],df_meta[b])\n",
    "    chi2_stat, p, dof, expected = scipy.stats.chi2_contingency(ct_table_ind)\n",
    "    row=pd.DataFrame.from_dict({'var': [b],'effect':[chi2_stat],'pval':[p],'stat':['chi2']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "# fishers exact\n",
    "for b in bin:\n",
    "    ct_table_ind=pd.crosstab(df_meta[\"Timepoints\"],df_meta[b])\n",
    "    fisher, p = scipy.stats.fisher_exact(ct_table_ind)\n",
    "    row=pd.DataFrame.from_dict({'var': [b],'effect':[t],'pval':[p],'stat':['fisher']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "    \n",
    "# do post treatment vals of continuous vars differ from pre treatment unpaired\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    try:\n",
    "        W,p = scipy.stats.mannwhitneyu(x=df_pre[c].values,y=df_post[c].values, nan_policy='omit')\n",
    "    except:\n",
    "        W,p = 0, 1\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['mwu']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v')\n",
    "        sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette='dark:grey', hue=None, orient='v')\n",
    "    \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/mwu_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['MWU']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "    \n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    t,p = scipy.stats.ttest_ind(a=df_pre[c].values,b=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['ttest']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v')\n",
    "        sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette='dark:grey', hue=None, orient='v')\n",
    "    \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/tt_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['ttest']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "# unpaired and then paired\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    W,p = scipy.stats.wilcoxon(x=df_pre[c].values,y=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['WSR']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.swarmplot(data=df_meta, x='Timepoints', y=c, hue='WOMAC_P_Response', palette=['b','r'], orient='v')\n",
    "        for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "            ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "        sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v', hue=None, boxprops=dict(alpha=.3))\n",
    "        \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/wsr_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['WSR']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    t,p = scipy.stats.ttest_rel(a=df_pre[c].values,b=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['pairedt']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette=['b','r'], hue='WOMAC_P_Response',orient='v')\n",
    "        for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "            ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "        sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v', hue=None, boxprops=dict(alpha=0.3))\n",
    "        \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/pairedt_' + c  + '.pdf')\n",
    "        plt.close()          \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['pairedt']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "df_results.to_csv(path + 'outputs/' + job + '/outcome_testing.tsv', sep='\\t')\n",
    "df_results.head()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0379f4d-6b07-489a-bfae-75b6085b9a15",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Grab Paired Alpha ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "298b4824-3c24-485e-8705-5daf4a85d6d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bray_Curtis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.6672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>0.2837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>0.4237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>0.3257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>0.1022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Bray_Curtis\n",
       "OAD001       0.6672\n",
       "OAD003       0.2837\n",
       "OAD004       0.4237\n",
       "OAD005       0.3257\n",
       "OAD006       0.1022"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###\n",
    "# Hypothesis 2: There will be an association between oral and gut microbiome and pain outcomes\n",
    "###\n",
    "# construct alpha, beta and paired alpha dataframes\n",
    "g_to_dfd = {}\n",
    "# g_test = ['stool','saliva_adh', 'stool_adh', 'saliva']\n",
    "g_test = ['saliva_adh', 'stool_adh']\n",
    "\n",
    "for g in g_test:\n",
    "    # maps diversity type to dataframe\n",
    "    g_to_dfd[g] = {}\n",
    "\n",
    "    # get alpha diversities\n",
    "    df_alpha = pd.read_csv(path + 'outputs/Qiime2_' + g + '/metadata.tsv', sep='\\t', index_col=0)\n",
    "    df_alpha = df_alpha.drop('#q2:types')\n",
    "    df_alpha['SubjectID'] = df_alpha['Patient_ID'] + df_alpha['Study_ID']\n",
    "    df_alpha = df_alpha[['SubjectID', 'Timepoints', 'shannon_entropy']]\n",
    "    g_to_dfd[g]['alpha'] = df_alpha\n",
    "\n",
    "    # get paired alpha div, first drop unpaired samples\n",
    "    s_remove = []\n",
    "    for s in list(df_alpha['SubjectID'].values):\n",
    "        if len(df_alpha[df_alpha['SubjectID'] == s]) != 2:\n",
    "            s_remove.append(s)\n",
    "    df_alpha = df_alpha.loc[~df_alpha['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "    \n",
    "    # set vars\n",
    "    alpha_metric = 'shannon_entropy'\n",
    "    group_var = 'Timepoints'\n",
    "    pair_var = 'SubjectID'\n",
    "    groups = ['pre','post']\n",
    "    \n",
    "    # get paired per indiv pair\n",
    "    pair_to_diff = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        df = df_alpha[df_alpha[pair_var] == p]\n",
    "        alpha_0 = float(df[df[group_var] == groups[0]][alpha_metric].values)\n",
    "        alpha_1 = float(df[df[group_var] == groups[1]][alpha_metric].values)\n",
    "        pair_to_diff[p] = alpha_1 - alpha_0\n",
    "    \n",
    "    df_paired_alpha = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[alpha_metric + '_diff'])\n",
    "    g_to_dfd[g]['paired_alpha'] = df_paired_alpha\n",
    "\n",
    "    # get beta div\n",
    "    df_beta = pd.read_csv(path + 'outputs/Qiime2_' + g + '/core_metrics_results/distance-matrix.tsv',\n",
    "                              sep='\\t', index_col=0)\n",
    "        \n",
    "    # grab twin to pair dict\n",
    "    pair_to_ids = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        df = df_alpha[df_alpha[pair_var] == p]\n",
    "        id_0 = str(df[df[group_var] == groups[0]].index.values[0])\n",
    "        id_1 = str(df[df[group_var] == groups[1]].index.values[0])\n",
    "        pair_to_ids[p] = (id_0, id_1)\n",
    "    \n",
    "    # get distances for each twin pair per beta div matrix    \n",
    "    pair_to_dist = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        id_0, id_1 = pair_to_ids[p]\n",
    "        pair_to_dist[p] = df_beta.loc[id_0, id_1]\n",
    "    \n",
    "    df_paired_beta = pd.DataFrame.from_dict(pair_to_dist, orient='index', columns=['Bray_Curtis'])\n",
    "    g_to_dfd[g]['paired_beta'] = df_paired_beta\n",
    "\n",
    "g_to_dfd['stool_adh']['paired_beta'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e018facb-88c3-4c10-be22-dbe70ee41351",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Chi2 Association of Outcome with Alpha ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9d4f38ff-41d6-4f31-94ab-69879de56109",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>div</th>\n",
       "      <th>outcome</th>\n",
       "      <th>statistic</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>pval</th>\n",
       "      <th>timepoint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>0.955089</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.313871</td>\n",
       "      <td>0.758599</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.244898</td>\n",
       "      <td>0.620691</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>17.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.682542</td>\n",
       "      <td>0.510397</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>0.396892</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>-0.563791</td>\n",
       "      <td>0.582492</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.244898</td>\n",
       "      <td>0.620691</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>21.000000</td>\n",
       "      <td>0.638889</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.769556</td>\n",
       "      <td>0.459355</td>\n",
       "      <td>post</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         group                  div           outcome statistic  test_stat  \\\n",
       "0     metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.000000   \n",
       "1   saliva_adh                alpha  WOMAC_P_Response       mwu  29.000000   \n",
       "2   saliva_adh                alpha  WOMAC_P_Response        tt   0.313871   \n",
       "3     metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.244898   \n",
       "4   saliva_adh                alpha  WOMAC_P_Response       mwu  17.000000   \n",
       "5   saliva_adh                alpha  WOMAC_P_Response        tt   0.682542   \n",
       "6     metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.000000   \n",
       "7    stool_adh                alpha  WOMAC_P_Response       mwu  20.000000   \n",
       "8    stool_adh                alpha  WOMAC_P_Response        tt  -0.563791   \n",
       "9     metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.244898   \n",
       "10   stool_adh                alpha  WOMAC_P_Response       mwu  21.000000   \n",
       "11   stool_adh                alpha  WOMAC_P_Response        tt   0.769556   \n",
       "\n",
       "        pval timepoint  \n",
       "0   1.000000       pre  \n",
       "1   0.955089       pre  \n",
       "2   0.758599       pre  \n",
       "3   0.620691      post  \n",
       "4   1.000000      post  \n",
       "5   0.510397      post  \n",
       "6   1.000000       pre  \n",
       "7   0.396892       pre  \n",
       "8   0.582492       pre  \n",
       "9   0.620691      post  \n",
       "10  0.638889      post  \n",
       "11  0.459355      post  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Testing for each sample type for (1) all and (2) high adh only\n",
    "# (A) Chisq of quartiles with adherence \n",
    "# (B) MWU/TT unpaired of outcomes against distance\n",
    "\n",
    "def chisq_of_df_cols(df, c1, c2):\n",
    "    groupsizes = df.groupby([c1, c2]).size()\n",
    "    ctsum = groupsizes.unstack(c1)\n",
    "    # fillna(0) is necessary to remove any NAs which will cause exceptions\n",
    "    return(scipy.stats.chi2_contingency(ctsum.fillna(0)))\n",
    "\n",
    "d_to_metric = {\n",
    "    'alpha': 'shannon_entropy',\n",
    "}\n",
    "group_var = 'Adherece_antiinflam'\n",
    "\n",
    "dfd_to_merge = {}\n",
    "g_test = ['saliva_adh', 'stool_adh']\n",
    "\n",
    "for g in g_test:\n",
    "    dfd_to_merge[g] = {}\n",
    "\n",
    "gs = []\n",
    "ds = []\n",
    "os = []\n",
    "stats = []\n",
    "ts = []\n",
    "ps = []\n",
    "times = []\n",
    "\n",
    "arr = [gs,ds,os,stats,ts,ps,times]\n",
    "\n",
    "def append_results(arr, val):\n",
    "    for a,v in zip(arr,val):\n",
    "        a.append(v)\n",
    "    return arr\n",
    "\n",
    "# for each sample type, grab relevant mapping file\n",
    "# g_test = saliva, saliva_adh, stool, etc.\n",
    "for g in g_test:\n",
    "    for d in d_to_metric:\n",
    "        # grab relevant diversity df\n",
    "        df_div = g_to_dfd[g][d]\n",
    "\n",
    "        if d == 'alpha':\n",
    "            # df_div = df_div[df_div['Timepoints'] == 'pre']\n",
    "            # df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            # df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div['shannon_entropy'] = df_div['shannon_entropy'].astype(float)\n",
    "\n",
    "        # merge with df of metadata var        \n",
    "        # grab relevant sample IDs\n",
    "        # g = saliva_adh\n",
    "        idx = 'Run_ID_' + g.split('_')[0].capitalize()\n",
    "        df_meta_sub = df_meta.dropna(subset=idx)\n",
    "        df_meta_sub = df_meta_sub.set_index(idx)\n",
    "        df_merge1 = pd.concat([df_meta_sub,df_div.drop('Timepoints',axis=1)],axis=1)\n",
    "\n",
    "        # subset on adh only\n",
    "        df_merge1 = df_merge1[df_merge1[group_var].isin(['Moderate adherence', 'High adherence'])]\n",
    "\n",
    "        # for each timepoint\n",
    "        for time in ['pre', 'post']:\n",
    "            df_merge = df_merge1[df_merge1['Timepoints'] == time]\n",
    "            # test association of div with outcomes\n",
    "            for o in ['WOMAC_P_Response']: #  outcomes:                    \n",
    "                # test association of adherence with pain outcomes\n",
    "                div_metric = d_to_metric[d]\n",
    "                x,p,dof,ef = chisq_of_df_cols(df_merge, group_var, o)\n",
    "                arr = append_results(arr, ['metadata',group_var,o,'chisq',x,p,time])\n",
    "                \n",
    "                ax = sns.boxplot(data=df_merge, x=o, y=div_metric)\n",
    "                sns.swarmplot(data=df_merge, x=o, y=div_metric, palette='dark:grey')\n",
    "                sns.despine()\n",
    "            \n",
    "                plt.tight_layout()\n",
    "                plt.savefig(path + 'outputs/jobs06/rvnr_nondiff_' + time + '_' + g + '_' + o + '_' + d + '.pdf')\n",
    "                plt.close()          \n",
    "            \n",
    "                u, p = scipy.stats.mannwhitneyu(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                                df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                                nan_policy='omit')\n",
    "    \n",
    "                arr = append_results(arr, [g,d,o,'mwu',u,p,time])\n",
    "    \n",
    "                t, p = scipy.stats.ttest_ind(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                                df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                                nan_policy='omit')\n",
    "    \n",
    "                arr = append_results(arr, [g,d,o,'tt',t,p,time])\n",
    "    \n",
    "            # save results\n",
    "            dfd_to_merge[g][d] = df_merge\n",
    "    \n",
    "            # export to Q2\n",
    "            # df_q2_type = df_merge.set_index(['Together'])\n",
    "            df_q2_type = df_merge.copy()\n",
    "            q2_row = pd.Series(data=['categorical' for i in range(len(df_merge.columns))], \n",
    "                               index=list(df_merge.columns.values), dtype=str, name='#q2:types')\n",
    "            df_q2_type = pd.concat([q2_row.to_frame().T, df_q2_type])\n",
    "            df_q2_type.index.name = '#SampleID'\n",
    "            df_q2_type.index = df_q2_type.index.map(lambda x: x.split('.guma')[0])\n",
    "            # df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + d + '_' + g + 'aggregate_outcomes.tsv', sep='\\t')\n",
    "\n",
    "df_results = pd.DataFrame.from_dict({\n",
    "    'group': gs,\n",
    "    'div': ds,\n",
    "    'outcome': os,\n",
    "    'statistic': stats,\n",
    "    'test_stat': ts,\n",
    "    'pval': ps,\n",
    "    'timepoint': times\n",
    "})\n",
    "# df_results.to_csv(path + 'outputs/df_results_aggregate.tsv', sep='\\t')\n",
    "# df_results.head()\n",
    "# df_results[df_results['pval'] < 0.05]\n",
    "df_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "22f26972-623f-459c-93fd-d37eb65d17eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>div</th>\n",
       "      <th>outcome</th>\n",
       "      <th>statistic</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.769556</td>\n",
       "      <td>0.343434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>-0.670379</td>\n",
       "      <td>0.519757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>-0.670379</td>\n",
       "      <td>0.638889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.777949</td>\n",
       "      <td>0.456353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.777949</td>\n",
       "      <td>0.431818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>1.184945</td>\n",
       "      <td>0.279964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>1.184945</td>\n",
       "      <td>0.876263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>-0.977825</td>\n",
       "      <td>0.360842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>-0.977825</td>\n",
       "      <td>0.755051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.558291</td>\n",
       "      <td>0.591262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.558291</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.731791</td>\n",
       "      <td>0.481109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.731791</td>\n",
       "      <td>0.876263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.164212</td>\n",
       "      <td>0.872849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.164212</td>\n",
       "      <td>0.267677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>1.234156</td>\n",
       "      <td>0.245859</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         group           div           outcome statistic  test_stat      pval\n",
       "0    stool_adh     pre_alpha  WOMAC_P_Response       mwu   0.769556  0.343434\n",
       "1    stool_adh     pre_alpha  WOMAC_P_Response        tt  -0.670379  0.519757\n",
       "2    stool_adh    post_alpha  WOMAC_P_Response       mwu  -0.670379  0.638889\n",
       "3    stool_adh    post_alpha  WOMAC_P_Response        tt   0.777949  0.456353\n",
       "4    stool_adh  paired_alpha  WOMAC_P_Response       mwu   0.777949  0.431818\n",
       "5    stool_adh  paired_alpha  WOMAC_P_Response        tt   1.184945  0.279964\n",
       "6    stool_adh   paired_beta  WOMAC_P_Response       mwu   1.184945  0.876263\n",
       "7    stool_adh   paired_beta  WOMAC_P_Response        tt  -0.977825  0.360842\n",
       "8   saliva_adh     pre_alpha  WOMAC_P_Response       mwu  -0.977825  0.755051\n",
       "9   saliva_adh     pre_alpha  WOMAC_P_Response        tt   0.558291  0.591262\n",
       "10  saliva_adh    post_alpha  WOMAC_P_Response       mwu   0.558291  1.000000\n",
       "11  saliva_adh    post_alpha  WOMAC_P_Response        tt   0.731791  0.481109\n",
       "12  saliva_adh  paired_alpha  WOMAC_P_Response       mwu   0.731791  0.876263\n",
       "13  saliva_adh  paired_alpha  WOMAC_P_Response        tt   0.164212  0.872849\n",
       "14  saliva_adh   paired_beta  WOMAC_P_Response       mwu   0.164212  0.267677\n",
       "15  saliva_adh   paired_beta  WOMAC_P_Response        tt   1.234156  0.245859"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# expand previous to all data, paired and unpaired diversities\n",
    "d_to_metric = {\n",
    "    'pre_alpha': 'shannon_entropy',\n",
    "    'post_alpha': 'shannon_entropy',\n",
    "    'paired_alpha': 'shannon_entropy_diff',\n",
    "    'paired_beta': 'Bray_Curtis' # 'Unweighted_Unifrac',\n",
    "}\n",
    "group_var = 'Adherece_antiinflam'\n",
    "\n",
    "dfd_to_merge = {}\n",
    "g_test = ['stool_adh','saliva_adh']\n",
    "for g in g_test:\n",
    "    dfd_to_merge[g] = {}\n",
    "\n",
    "gs = []\n",
    "ds = []\n",
    "os = []\n",
    "stats = []\n",
    "ts = []\n",
    "ps = []\n",
    "\n",
    "arr = [gs,ds,os,stats,ts,ps]\n",
    "\n",
    "def append_results(arr, val):\n",
    "    for a,v in zip(arr,val):\n",
    "        a.append(v)\n",
    "    return arr\n",
    "\n",
    "# for each sample type, grab relevant mapping file\n",
    "# g_test = saliva, saliva_adh, stool, etc.\n",
    "d_to_map = {}\n",
    "\n",
    "for g in g_test:\n",
    "    # drop duplicates so you have sample mapping to adh\n",
    "    df_map_sub = type_to_df_map[g.split('_')[0]]\n",
    "    df_map_sub.index = df_map_sub.index.map(lambda x: x.split('.')[0].replace('-',''))\n",
    "    df_map_sub = df_map_sub.dropna(how='any',subset=group_var,axis=0)           \n",
    "    \n",
    "    # figure out which samples to keep\n",
    "    # i.e. samples that have a pre and post time point\n",
    "    keep = []\n",
    "    for i in list(df_map_sub.index.values):\n",
    "        if len(df_map_sub.loc[i,:]) == 2:\n",
    "            keep.append(i)\n",
    "\n",
    "    # get unique entires in sorted order\n",
    "    # at this point we are only concerned with differences in values, \n",
    "    # as we've dropped samples with only one endpoint val\n",
    "    save = []\n",
    "    [save.append(x) for x in keep if x not in save]\n",
    "    df_map_sub = df_map_sub.loc[save,:]\n",
    "\n",
    "    # this double populates as OAD001 is an index twice, so the diff fills to both the pre and post col\n",
    "    for o in outcomes:\n",
    "        df_map_sub[o + '_diff'] = df_map_sub[df_map_sub['Timepoint'] == 'post'][o] - df_map_sub[df_map_sub['Timepoint'] == 'pre'][o] \n",
    "\n",
    "    # here we keep only the pre, but everything is identical b/w pre and post\n",
    "    df_dropdup = df_map_sub[~df_map_sub.index.duplicated(keep='first')]\n",
    "    d_to_map[g] = df_dropdup\n",
    "    \n",
    "    for d in d_to_metric:\n",
    "        # grab relevant diversity df\n",
    "        if d == 'pre_alpha' or d == 'post_alpha':\n",
    "            df_div = g_to_dfd[g]['alpha']\n",
    "        else:\n",
    "            df_div = g_to_dfd[g][d]\n",
    "        div_metric = d_to_metric[d]\n",
    "\n",
    "        # when associating alpha div vs outcomes, look at if starting adiv predicts outcome\n",
    "        if d == 'pre_alpha':\n",
    "            df_div = df_div[df_div['Timepoints'] == 'pre']\n",
    "            df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div = df_div.astype(float)\n",
    "\n",
    "        if d == 'post_alpha':\n",
    "            df_div = df_div[df_div['Timepoints'] == 'post']\n",
    "            df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div = df_div.astype(float)\n",
    "\n",
    "        # merge with df of metadata var        \n",
    "        df_merge = pd.concat([df_dropdup,df_div],axis=1)\n",
    "\n",
    "        # drop na in barcodes if samples not sequenced both pre and post\n",
    "        df_merge = df_merge.dropna(how='any',subset='BarcodeSequence')\n",
    "        \n",
    "        # test association of div with outcomes\n",
    "        for o in ['WOMAC_P_Response']:                     \n",
    "            # test association of adherence with pain outcomes\n",
    "            # flip x and y for horizontal\n",
    "            ax = sns.boxplot(data=df_merge, x=o, y=div_metric)\n",
    "            sns.swarmplot(data=df_merge, x=o, y=div_metric, palette='dark:grey')\n",
    "            sns.despine()\n",
    "        \n",
    "            plt.tight_layout()\n",
    "            plt.savefig(path + 'outputs/jobs06/WOMAC_barplot_' + g + '_' + o + '_' + d + '.pdf')\n",
    "            plt.close()          \n",
    "\n",
    "            # MWU and ttest\n",
    "            u, p = scipy.stats.mannwhitneyu(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'mwu',t,p])\n",
    "\n",
    "            t, p = scipy.stats.ttest_ind(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                         df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                         nan_policy='omit',\n",
    "                                         equal_var=False)\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'tt',t,p])\n",
    "         \n",
    "        # save results\n",
    "        dfd_to_merge[g][d] = df_merge\n",
    "\n",
    "        # export to Q2\n",
    "        df_q2_type = df_merge.set_index(['Together'])\n",
    "        q2_row = pd.Series(data=['categorical' for i in range(len(df_merge.columns))], \n",
    "                           index=list(df_merge.columns.values), dtype=str, name='#q2:types')\n",
    "        df_q2_type = pd.concat([q2_row.to_frame().T, df_q2_type])\n",
    "        df_q2_type.index.name = '#SampleID'\n",
    "        df_q2_type.index = df_q2_type.index.map(lambda x: x.split('.guma')[0])\n",
    "        # df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + d + '_' + g + '_outcomes.tsv', sep='\\t')\n",
    "\n",
    "df_results = pd.DataFrame.from_dict({\n",
    "    'group': gs,\n",
    "    'div': ds,\n",
    "    'outcome': os,\n",
    "    'statistic': stats,\n",
    "    'test_stat': ts,\n",
    "    'pval': ps\n",
    "})\n",
    "# df_results.to_csv(path + 'outputs/df_results_diff.tsv', sep='\\t')\n",
    "\n",
    "# df_results.head()\n",
    "# the div==alpha results test whether pre-alpha div state associates with pain outcome changes (differences) quartiles\n",
    "# the div==paired_alpha and paired_beta test whether the alpha and betas change in a similar way with the pain outcome\n",
    "print(len(df_results[df_results['pval'] < 0.05]))\n",
    "df_results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5db737b8-b788-4d7a-86a6-39e82ac57eb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_adh\n",
      "['Response', 'No response']\n",
      "WilcoxonResult(statistic=25.0, pvalue=0.30126953125)\n",
      "['Response']\n",
      "WilcoxonResult(statistic=4.0, pvalue=0.109375)\n",
      "['No response']\n",
      "WilcoxonResult(statistic=7.0, pvalue=1.0)\n",
      "saliva_adh\n",
      "['Response', 'No response']\n",
      "WilcoxonResult(statistic=24.0, pvalue=0.26611328125)\n",
      "['Response']\n",
      "WilcoxonResult(statistic=9.0, pvalue=0.46875)\n",
      "['No response']\n",
      "WilcoxonResult(statistic=5.0, pvalue=0.625)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>shannon_entropy_diff</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>Separate</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Together</th>\n",
       "      <th>ContactEmail</th>\n",
       "      <th>ContactName</th>\n",
       "      <th>PrimaryInvestigator</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>...</th>\n",
       "      <th>Pain_DETECT_diff</th>\n",
       "      <th>CES_D_diff</th>\n",
       "      <th>Helplesness_diff</th>\n",
       "      <th>Magnification_diff</th>\n",
       "      <th>Rumination_diff</th>\n",
       "      <th>PCS_EN_diff</th>\n",
       "      <th>Sleep_distrubance_diff</th>\n",
       "      <th>PASE_walk_diff</th>\n",
       "      <th>PASE_light_diff</th>\n",
       "      <th>BMI_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>-0.116436</td>\n",
       "      <td>AGGACAAACTAT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>Text</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-001.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>5.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.51</td>\n",
       "      <td>14.29</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>-0.327749</td>\n",
       "      <td>TCTTCCTAAAGT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-003.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-5.5</td>\n",
       "      <td>-14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>0.866474</td>\n",
       "      <td>CTTGTGCGACAA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-004.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.264032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>0.007278</td>\n",
       "      <td>CTTACACTGCTT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-001.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>5.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.081446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD012</th>\n",
       "      <td>0.535614</td>\n",
       "      <td>GCGTGGTCATTA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-012.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-1.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.72</td>\n",
       "      <td>-1.072664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TTGGACGTCCAC</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-009.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.139974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 205 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          shannon_entropy_diff BarcodeSequence    LinkerPrimerSequence  \\\n",
       "OAD001               -0.116436    AGGACAAACTAT  CCGGACTACHVGGGTWTCTAAT   \n",
       "OAD003               -0.327749    TCTTCCTAAAGT  CCGGACTACHVGGGTWTCTAAT   \n",
       "OAD004                0.866474    CTTGTGCGACAA  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD001              0.007278    CTTACACTGCTT  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD012              0.535614    GCGTGGTCATTA  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD009                   NaN    TTGGACGTCCAC  CCGGACTACHVGGGTWTCTAAT   \n",
       "\n",
       "         Separate Timepoint              Together  \\\n",
       "OAD001       Text       pre    OAD-001.pre.saliva   \n",
       "OAD003        All       pre    OAD-003.pre.saliva   \n",
       "OAD004        All       pre    OAD-004.pre.saliva   \n",
       "VAOAD001      All       pre  VAOAD-001.pre.saliva   \n",
       "VAOAD012      All       pre  VAOAD-012.pre.saliva   \n",
       "VAOAD009      All       pre  VAOAD-009.pre.saliva   \n",
       "\n",
       "                          ContactEmail    ContactName PrimaryInvestigator  \\\n",
       "OAD001    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "OAD003    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "OAD004    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD001  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD012  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD009  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "\n",
       "         Cohort  ... Pain_DETECT_diff CES_D_diff Helplesness_diff  \\\n",
       "OAD001    NonVA  ...             -2.0       -1.0             -2.5   \n",
       "OAD003    NonVA  ...             -4.5        1.0             -1.0   \n",
       "OAD004    NonVA  ...             -1.0        3.0             -1.0   \n",
       "VAOAD001     VA  ...             -0.5        5.5              0.0   \n",
       "VAOAD012     VA  ...             -1.5        0.0              2.0   \n",
       "VAOAD009     VA  ...             -4.0        2.0              1.0   \n",
       "\n",
       "         Magnification_diff Rumination_diff PCS_EN_diff  \\\n",
       "OAD001                  5.5             1.0         4.0   \n",
       "OAD003                 -1.0            -2.0        -5.5   \n",
       "OAD004                 -1.0            -1.0        -3.0   \n",
       "VAOAD001                0.0             0.0         0.0   \n",
       "VAOAD012               -2.0            -0.5        -0.5   \n",
       "VAOAD009                0.0             0.5         1.5   \n",
       "\n",
       "         Sleep_distrubance_diff PASE_walk_diff PASE_light_diff  BMI_diff  \n",
       "OAD001                      9.0          11.51           14.29       NaN  \n",
       "OAD003                    -14.0           0.00            0.00       NaN  \n",
       "OAD004                      0.0           0.00            0.00  3.264032  \n",
       "VAOAD001                   -2.5           0.00            0.00 -0.081446  \n",
       "VAOAD012                   -1.5           0.00            6.72 -1.072664  \n",
       "VAOAD009                    2.5           0.00            0.00 -0.139974  \n",
       "\n",
       "[6 rows x 205 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test if the differences in diversity are significantly different from 0 for paired alpha\n",
    "for g in g_test:\n",
    "    print(g)\n",
    "    for d in ['paired_alpha']:\n",
    "        df_div = g_to_dfd[g][d] \n",
    "        df_map = d_to_map[g]\n",
    "\n",
    "        df_merge = pd.concat([df_div, df_map], axis=1)\n",
    "        for o in [['Response','No response'],['Response'], ['No response']]:\n",
    "            df = df_merge[df_merge['WOMAC_P_Response'].isin(o)]\n",
    "            print(o)\n",
    "            print(scipy.stats.wilcoxon(df['shannon_entropy_diff'], nan_policy='omit'))\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "872848a4-7f8e-4613-8839-f8ceab94a568",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_adh\n",
      "response\n",
      "alpha wilcoxon\n",
      "4.0 0.109375\n",
      "noresponse\n",
      "alpha wilcoxon\n",
      "7.0 1.0\n",
      "saliva_adh\n",
      "response\n",
      "alpha wilcoxon\n",
      "9.0 0.46875\n",
      "noresponse\n",
      "alpha wilcoxon\n",
      "5.0 0.625\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Separate</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Height</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>HostSubjectId</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Timepoints</th>\n",
       "      <th>Study_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>quinoa</th>\n",
       "      <th>broccoli</th>\n",
       "      <th>pork</th>\n",
       "      <th>beef</th>\n",
       "      <th>burger</th>\n",
       "      <th>Total_omega3</th>\n",
       "      <th>Adherence_omega3</th>\n",
       "      <th>Adherence_omega6</th>\n",
       "      <th>Total_o3_o6</th>\n",
       "      <th>shannon_entropy</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.saliva</th>\n",
       "      <td>Text</td>\n",
       "      <td>pre</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>152.0</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001</td>\n",
       "      <td>F</td>\n",
       "      <td>OAD</td>\n",
       "      <td>pre</td>\n",
       "      <td>001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.305965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.saliva</th>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>167.0</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003</td>\n",
       "      <td>F</td>\n",
       "      <td>OAD</td>\n",
       "      <td>pre</td>\n",
       "      <td>003</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>75.0</td>\n",
       "      <td>4.118134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.saliva</th>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>158.0</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-004</td>\n",
       "      <td>F</td>\n",
       "      <td>OAD</td>\n",
       "      <td>pre</td>\n",
       "      <td>004</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>35.8</td>\n",
       "      <td>3.709290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-001.pre.saliva</th>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>175.2</td>\n",
       "      <td>VA</td>\n",
       "      <td>VAOAD-001</td>\n",
       "      <td>M</td>\n",
       "      <td>VAOAD</td>\n",
       "      <td>pre</td>\n",
       "      <td>001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>57.0</td>\n",
       "      <td>4.985869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-012.pre.saliva</th>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>170.0</td>\n",
       "      <td>VA</td>\n",
       "      <td>VAOAD-012</td>\n",
       "      <td>M</td>\n",
       "      <td>VAOAD</td>\n",
       "      <td>pre</td>\n",
       "      <td>012</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>39.2</td>\n",
       "      <td>11.2</td>\n",
       "      <td>5.6</td>\n",
       "      <td>8.4</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>77.8</td>\n",
       "      <td>4.070449</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 141 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Separate Timepoint               Ethnicity Height Cohort  \\\n",
       "id                                                                              \n",
       "OAD-001.pre.saliva       Text       pre                Hispanic  152.0  NonVA   \n",
       "OAD-003.pre.saliva        All       pre                Hispanic  167.0  NonVA   \n",
       "OAD-004.pre.saliva        All       pre  Not Hispanic or Latino  158.0  NonVA   \n",
       "VAOAD-001.pre.saliva      All       pre  Not Hispanic or Latino  175.2     VA   \n",
       "VAOAD-012.pre.saliva      All       pre  Not Hispanic or Latino  170.0     VA   \n",
       "\n",
       "                     HostSubjectId Sex Patient_ID Timepoints Study_ID  ...  \\\n",
       "id                                                                     ...   \n",
       "OAD-001.pre.saliva         OAD-001   F        OAD        pre      001  ...   \n",
       "OAD-003.pre.saliva         OAD-003   F        OAD        pre      003  ...   \n",
       "OAD-004.pre.saliva         OAD-004   F        OAD        pre      004  ...   \n",
       "VAOAD-001.pre.saliva     VAOAD-001   M      VAOAD        pre      001  ...   \n",
       "VAOAD-012.pre.saliva     VAOAD-012   M      VAOAD        pre      012  ...   \n",
       "\n",
       "                     quinoa broccoli  pork  beef burger Total_omega3  \\\n",
       "id                                                                     \n",
       "OAD-001.pre.saliva      0.0      0.0   0.0   0.0    0.0          NaN   \n",
       "OAD-003.pre.saliva      0.0      0.0   0.0  14.0    0.0         16.8   \n",
       "OAD-004.pre.saliva      0.0      2.8   0.0   0.0    0.0          2.8   \n",
       "VAOAD-001.pre.saliva    0.0      0.0  35.0   1.0    2.0          6.0   \n",
       "VAOAD-012.pre.saliva    0.0      0.0  39.2  11.2    5.6          8.4   \n",
       "\n",
       "                     Adherence_omega3 Adherence_omega6 Total_o3_o6  \\\n",
       "id                                                                   \n",
       "OAD-001.pre.saliva      Low adherence    Low adherence         0.0   \n",
       "OAD-003.pre.saliva      Low adherence   High adherence        75.0   \n",
       "OAD-004.pre.saliva      Low adherence    Low adherence        35.8   \n",
       "VAOAD-001.pre.saliva    Low adherence    Low adherence        57.0   \n",
       "VAOAD-012.pre.saliva   High adherence   High adherence        77.8   \n",
       "\n",
       "                     shannon_entropy  \n",
       "id                                    \n",
       "OAD-001.pre.saliva          5.305965  \n",
       "OAD-003.pre.saliva          4.118134  \n",
       "OAD-004.pre.saliva          3.709290  \n",
       "VAOAD-001.pre.saliva        4.985869  \n",
       "VAOAD-012.pre.saliva        4.070449  \n",
       "\n",
       "[5 rows x 141 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pre vs post for RESPONDERS only\n",
    "\n",
    "# paired beta, comparing intra-indiv difference pre_post to inter pre and inter post\n",
    "g_test = ['stool_adh','saliva_adh']\n",
    "sg = ['response', 'noresponse']\n",
    "for g in g_test: #subgroups:\n",
    "    print(g)\n",
    "    for s in sg:\n",
    "        print(s)\n",
    "        df = pd.read_csv(path + 'outputs/Qiime2_' + g + '_' + s + '/metadata.tsv', sep='\\t', index_col=0)\n",
    "        df = df.drop('#q2:types')\n",
    "\n",
    "        print('alpha wilcoxon')\n",
    "        df = df.dropna(subset=['Timepoints','shannon_entropy'])\n",
    "        df['shannon_entropy'] = df['shannon_entropy'].astype(float)\n",
    "        df_pre = df[df['Timepoints'] == 'pre']\n",
    "        df_post = df[df['Timepoints'] == 'post']\n",
    "\n",
    "        common_idx = list(set(df_pre.HostSubjectId.values).intersection(set(df_post.HostSubjectId.values)))\n",
    "        df_pre = df_pre[df_pre.HostSubjectId.isin(common_idx)]\n",
    "        df_post = df_post[df_post.HostSubjectId.isin(common_idx)]\n",
    "        \n",
    "        W, p = scipy.stats.wilcoxon(df_pre['shannon_entropy'].values, df_post['shannon_entropy'].values)\n",
    "        print(W,p)\n",
    "\n",
    "        df = pd.concat([df_pre, df_post])\n",
    "        ax = sns.swarmplot(data=df, x='Timepoint', y='shannon_entropy', hue='Timepoint', palette=['b','r'], orient='v')\n",
    "        for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "            ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "        sns.boxplot(data=df, x='Timepoint', y='shannon_entropy', orient='v', hue='Timepoint', boxprops=dict(alpha=.3))\n",
    "        sns.despine()\n",
    "    \n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/jobs18/alpha_' + g + '_' + s + '.pdf')\n",
    "        plt.close()          \n",
    "\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bef8c05d-208b-4b13-bac7-01bf00e4e80e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_adh\n",
      "All\n",
      "paired alpha wilcoxon\n",
      "33.0 0.414306640625\n",
      "paired t test\n",
      "-0.7900792468039035 0.44481434457162783\n",
      "inter intra KW\n",
      "14.985539176909924 0.0005570978899780199\n",
      "Response\n",
      "paired alpha wilcoxon\n",
      "4.0 0.109375\n",
      "paired t test\n",
      "-2.1276058423185478 0.07746512839408434\n",
      "inter intra KW\n",
      "16.10169063709202 0.0003188322936852291\n",
      "No response\n",
      "paired alpha wilcoxon\n",
      "7.0 1.0\n",
      "paired t test\n",
      "0.2351990394957396 0.8256045805114028\n",
      "inter intra KW\n",
      "2.5703573471829486 0.2766011578942809\n",
      "saliva_adh\n",
      "All\n",
      "paired alpha wilcoxon\n",
      "26.0 0.19091796875\n",
      "paired t test\n",
      "-1.5438829559402965 0.14856768929110245\n",
      "inter intra KW\n",
      "17.042392642678692 0.0001992009743840896\n",
      "Response\n",
      "paired alpha wilcoxon\n",
      "9.0 0.46875\n",
      "paired t test\n",
      "-0.9798119153431403 0.365018427390056\n",
      "inter intra KW\n",
      "16.83795674134069 0.0002206399499696511\n",
      "No response\n",
      "paired alpha wilcoxon\n",
      "5.0 0.625\n",
      "paired t test\n",
      "-0.8754606396643463 0.43074573525689774\n",
      "inter intra KW\n",
      "19.225437442881685 6.687276850590436e-05\n"
     ]
    }
   ],
   "source": [
    "# Alpha div paired\n",
    "# paired beta, comparing intra-indiv difference pre_post to inter pre and inter post\n",
    "g_test = ['stool_adh','saliva_adh']\n",
    "subgroups = ['All', 'Response', 'No response']\n",
    "\n",
    "for g in g_test: #subgroups:\n",
    "    print(g)\n",
    "    df_a = pd.read_csv(path + 'outputs/Qiime2_' + g + '/metadata.tsv', sep='\\t', index_col=0)\n",
    "    df_a = df_a.drop('#q2:types')\n",
    "\n",
    "    for sg in subgroups:\n",
    "        print(sg)\n",
    "        if sg == 'All':\n",
    "            df_alpha = df_a.copy()\n",
    "        else:\n",
    "            df_alpha = df_a[df_a['WOMAC_P_Response'] == sg]\n",
    "        \n",
    "        df_alpha['SubjectID'] = df_alpha['Patient_ID'] + df_alpha['Study_ID']\n",
    "        df_alpha = df_alpha[['SubjectID', 'Timepoints', 'shannon_entropy']]\n",
    "        \n",
    "        # drop unpaired samples\n",
    "        s_remove = []\n",
    "        for s in list(df_alpha['SubjectID'].values):\n",
    "            if len(df_alpha[df_alpha['SubjectID'] == s]) != 2:\n",
    "                s_remove.append(s)\n",
    "        df_alpha = df_alpha.loc[~df_alpha['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "        \n",
    "        # set vars\n",
    "        alpha_metric = 'shannon_entropy'\n",
    "        group_var = 'Timepoints'\n",
    "        pair_var = 'SubjectID'\n",
    "        groups = ['pre','post']\n",
    "        \n",
    "        # get paired per indiv pair\n",
    "        pair_to_diff = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            df = df_alpha[df_alpha[pair_var] == p]\n",
    "            alpha_0 = float(df[df[group_var] == groups[0]][alpha_metric].values)\n",
    "            alpha_1 = float(df[df[group_var] == groups[1]][alpha_metric].values)\n",
    "            pair_to_diff[p] = alpha_0 - alpha_1\n",
    "        \n",
    "        df_paired_alpha = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[alpha_metric + '_diff'])\n",
    "        \n",
    "        # one-sided t-test, n.s.; RA-UA values \n",
    "        # t, p = scipy.stats.ttest_1samp(df_paired_alpha[alpha_metric + '_diff'],popmean=0)\n",
    "    \n",
    "        print('paired alpha wilcoxon')\n",
    "        s, p = scipy.stats.wilcoxon(df_paired_alpha[alpha_metric + '_diff'])\n",
    "        print(s, p)\n",
    "        print('paired t test')\n",
    "        t, p = scipy.stats.ttest_1samp(df_paired_alpha[alpha_metric + '_diff'], popmean=0)\n",
    "        print(t, p)\n",
    "        \n",
    "        # separate\n",
    "        df_alpha[alpha_metric] = df_alpha[alpha_metric].map(lambda x: float(x))\n",
    "        ax = sns.boxplot(data=df_alpha, x=group_var, y=alpha_metric)\n",
    "        sns.swarmplot(data=df_alpha, x=group_var, y=alpha_metric, palette='dark:grey')\n",
    "        sns.despine()\n",
    "    \n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/Qiime2_' + g + '/' + sg + '_alpha.pdf')\n",
    "        plt.close()          \n",
    "    \n",
    "        # now do beta\n",
    "        df_beta = pd.read_csv(path + 'outputs/Qiime2_' + g + '/core_metrics_results/distance-matrix.tsv',\n",
    "                              sep='\\t', index_col=0)\n",
    "        \n",
    "        # set vars\n",
    "        alpha_metric = 'shannon_entropy'\n",
    "        group_var = 'Timepoints'\n",
    "        pair_var = 'SubjectID'\n",
    "        groups = ['pre','post']\n",
    "        g0, g1 = groups[0], groups[1]\n",
    "        \n",
    "        # grab twin to pair dict\n",
    "        pair_to_ids = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            df = df_alpha[df_alpha[pair_var] == p]\n",
    "            id_0 = str(df[df[group_var] == g0].index.values[0])\n",
    "            id_1 = str(df[df[group_var] == g1].index.values[0])\n",
    "            pair_to_ids[p] = (id_0, id_1)\n",
    "        \n",
    "        # get distances for each twin pair per beta div matrix    \n",
    "        pair_to_dist = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            id_0, id_1 = pair_to_ids[p]\n",
    "            pair_to_dist[p] = df_beta.loc[id_0, id_1]\n",
    "        \n",
    "        df_paired_beta = pd.DataFrame.from_dict(pair_to_dist, orient='index', columns=['Bray_Curtis'])\n",
    "        \n",
    "        # grab inter RA distances\n",
    "        # this is from unweighted_Timepoint_significance.qzv -> download as tsv\n",
    "        df_raw = pd.read_csv(path + 'outputs/Qiime2_' + g + '/raw_data.tsv', \n",
    "                             sep='\\t', index_col=0)\n",
    "        df_0 = df_raw[df_raw['Group1'] == g0]\n",
    "        df_0 = df_0[df_0['Group2'] == g0]\n",
    "        df_1 = df_raw[df_raw['Group1'] == g1]\n",
    "        df_1 = df_1[df_1['Group2'] == g1]\n",
    "        \n",
    "        # compare distances\n",
    "        inter_twin = df_paired_beta['Bray_Curtis'].values\n",
    "        inter_0 = df_0['Distance'].values\n",
    "        inter_1 = df_1['Distance'].values\n",
    "        \n",
    "        u, p = scipy.stats.mannwhitneyu(inter_twin, inter_0)\n",
    "        #print(u, p)\n",
    "        \n",
    "        t, p = scipy.stats.ttest_ind(inter_twin, inter_1)\n",
    "        #print(t, p)\n",
    "        \n",
    "        t, p = scipy.stats.ttest_ind(inter_0, inter_1)\n",
    "        # print(t, p)\n",
    "        \n",
    "        f, p = scipy.stats.f_oneway(inter_0, inter_1, inter_twin)\n",
    "        # print('inter intra anova')\n",
    "        # print(f, p)\n",
    "\n",
    "        f, p = scipy.stats.kruskal(inter_0, inter_1, inter_twin)\n",
    "        print('inter intra KW')\n",
    "        print(f,p)\n",
    "\n",
    "        \n",
    "        category = ['intra_twin_pair']*len(inter_twin) + ['inter_' + g0 + '_only']*len(inter_0) + ['inter_' + g1 + '_only']*len(inter_1)\n",
    "        distances = list(inter_twin) + list(inter_0) + list(inter_1)\n",
    "        df_dist = pd.DataFrame(data=np.array([category,distances]).T, columns=['category','distance'])\n",
    "        df_dist['distance'] = df_dist['distance'].astype(float)\n",
    "        df_dist.to_csv(path + 'outputs/Qiime2_' + g + '/' + sg + '_inter_intra_beta_dist.tsv',sep='\\t')\n",
    "                             \n",
    "        sns.boxplot(data=df_dist, x='category', y='distance')\n",
    "        sns.swarmplot(data=df_dist, x='category', y='distance', color='black')\n",
    "        sns.despine()\n",
    "    \n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/Qiime2_' + g + '/' + sg + '_beta.pdf')\n",
    "        plt.close()          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f2ebdb4d-0d17-4c0d-b282-cd78e27f7bb7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectID</th>\n",
       "      <th>Timepoints</th>\n",
       "      <th>shannon_entropy</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>WOMAC_P_Response</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>OAD001</td>\n",
       "      <td>pre</td>\n",
       "      <td>5.097506</td>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>OAD001</td>\n",
       "      <td>post</td>\n",
       "      <td>5.838205</td>\n",
       "      <td>post</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>OAD003</td>\n",
       "      <td>pre</td>\n",
       "      <td>5.190428</td>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>OAD003</td>\n",
       "      <td>post</td>\n",
       "      <td>4.679640</td>\n",
       "      <td>post</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>OAD004</td>\n",
       "      <td>pre</td>\n",
       "      <td>5.299560</td>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SubjectID Timepoints  shannon_entropy Timepoint  \\\n",
       "OAD-001.pre.stool     OAD001        pre         5.097506       pre   \n",
       "OAD-001.post.stool    OAD001       post         5.838205      post   \n",
       "OAD-003.pre.stool     OAD003        pre         5.190428       pre   \n",
       "OAD-003.post.stool    OAD003       post         4.679640      post   \n",
       "OAD-004.pre.stool     OAD004        pre         5.299560       pre   \n",
       "\n",
       "                   WOMAC_P_Response  \n",
       "OAD-001.pre.stool       No response  \n",
       "OAD-001.post.stool      No response  \n",
       "OAD-003.pre.stool       No response  \n",
       "OAD-003.post.stool      No response  \n",
       "OAD-004.pre.stool       No response  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# alpha plots\n",
    "for g in ['saliva','stool']:\n",
    "    df_alpha = g_to_dfd[g + '_adh']['alpha']\n",
    "    df_map = type_to_df_map[g].copy()\n",
    "    df_map = df_map.set_index('Together')\n",
    "    df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.g')[0])\n",
    "    \n",
    "    df_merge = pd.concat([df_alpha, df_map], axis=1)\n",
    "    df_merge = df_merge.dropna()\n",
    "    \n",
    "    # sns.set_theme()\n",
    "    custom_params = {\"axes.spines.right\": False, \"axes.spines.top\": False}\n",
    "    sns.set_theme(style=\"ticks\", rc=custom_params)\n",
    "    plt.figure(figsize=(6,4))\n",
    "    sns.boxplot(x='Timepoint', y='shannon_entropy', data=df_merge,palette=['b','r'], boxprops=dict(alpha=0.5))#, hue='Timepoint', palette=['b','r'])\n",
    "    ax = sns.swarmplot(x='Timepoint', y='shannon_entropy', data=df_merge,color='black')\n",
    "    for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "        ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "    sns.despine()#(left=True, bottom=True)\n",
    "    \n",
    "    # ax.axes.set_title(\"\",fontsize=48)\\n\",\n",
    "    plt.tight_layout()\n",
    "    ax.set_ylabel(\"Shannon Entropy\",fontsize=16)\n",
    "    ax.set_xlabel(\"Diagnosis\",fontsize=16)\n",
    "    ax.tick_params(labelsize=16)\n",
    "    \n",
    "    # plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/Qiime2_' + g + '_adh/alpha_all.pdf')\n",
    "    plt.close()\n",
    "\n",
    "df_merge.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51757cac-d364-49d9-88b5-d27af66b38fd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Process Taxa DF ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8d8a7f96-5705-47fb-8d64-6ddcc4cd5d72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;__;__;__;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>0.014084</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.022027</td>\n",
       "      <td>0.001515</td>\n",
       "      <td>0.173283</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021230</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.152452</td>\n",
       "      <td>0.033001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.022576</td>\n",
       "      <td>0.007098</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.292202</td>\n",
       "      <td>0.001994</td>\n",
       "      <td>0.013555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000783</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>0.273829</td>\n",
       "      <td>0.067421</td>\n",
       "      <td>0.184572</td>\n",
       "      <td>0.026929</td>\n",
       "      <td>0.053673</td>\n",
       "      <td>0.037783</td>\n",
       "      <td>0.009619</td>\n",
       "      <td>0.023935</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.012060</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.181156</td>\n",
       "      <td>0.067844</td>\n",
       "      <td>0.150895</td>\n",
       "      <td>0.063229</td>\n",
       "      <td>0.051360</td>\n",
       "      <td>0.055600</td>\n",
       "      <td>0.012363</td>\n",
       "      <td>0.051513</td>\n",
       "      <td>0.006045</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.post.stool</th>\n",
       "      <td>0.174773</td>\n",
       "      <td>0.040785</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001855</td>\n",
       "      <td>0.124402</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001400</td>\n",
       "      <td>0.000371</td>\n",
       "      <td>0.014265</td>\n",
       "      <td>0.090594</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 374 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004  \\\n",
       "index                                                                                                                       \n",
       "OAD-001.post.stool                                           0.014084                                                       \n",
       "OAD-001.pre.stool                                            0.022576                                                       \n",
       "OAD-003.post.stool                                           0.273829                                                       \n",
       "OAD-003.pre.stool                                            0.181156                                                       \n",
       "OAD-004.post.stool                                           0.174773                                                       \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium  \\\n",
       "index                                                                                                                            \n",
       "OAD-001.post.stool                                           0.039046                                                            \n",
       "OAD-001.pre.stool                                            0.007098                                                            \n",
       "OAD-003.post.stool                                           0.067421                                                            \n",
       "OAD-003.pre.stool                                            0.067844                                                            \n",
       "OAD-004.post.stool                                           0.040785                                                            \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella  \\\n",
       "index                                                                                                             \n",
       "OAD-001.post.stool                                           0.022027                                             \n",
       "OAD-001.pre.stool                                            0.005128                                             \n",
       "OAD-003.post.stool                                           0.184572                                             \n",
       "OAD-003.pre.stool                                            0.150895                                             \n",
       "OAD-004.post.stool                                           0.000000                                             \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117  \\\n",
       "index                                                                                                                              \n",
       "OAD-001.post.stool                                           0.001515                                                              \n",
       "OAD-001.pre.stool                                            0.000000                                                              \n",
       "OAD-003.post.stool                                           0.026929                                                              \n",
       "OAD-003.pre.stool                                            0.063229                                                              \n",
       "OAD-004.post.stool                                           0.001855                                                              \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H  \\\n",
       "index                                                                                                                \n",
       "OAD-001.post.stool                                           0.173283                                                \n",
       "OAD-001.pre.stool                                            0.292202                                                \n",
       "OAD-003.post.stool                                           0.053673                                                \n",
       "OAD-003.pre.stool                                            0.051360                                                \n",
       "OAD-004.post.stool                                           0.124402                                                \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum  \\\n",
       "index                                                                                                                  \n",
       "OAD-001.post.stool                                           0.000000                                                  \n",
       "OAD-001.pre.stool                                            0.001994                                                  \n",
       "OAD-003.post.stool                                           0.037783                                                  \n",
       "OAD-003.pre.stool                                            0.055600                                                  \n",
       "OAD-004.post.stool                                           0.000000                                                  \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129  \\\n",
       "index                                                                                                                            \n",
       "OAD-001.post.stool                                           0.021230                                                            \n",
       "OAD-001.pre.stool                                            0.013555                                                            \n",
       "OAD-003.post.stool                                           0.009619                                                            \n",
       "OAD-003.pre.stool                                            0.012363                                                            \n",
       "OAD-004.post.stool                                           0.001400                                                            \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus  \\\n",
       "index                                                                                                                \n",
       "OAD-001.post.stool                                           0.000160                                                \n",
       "OAD-001.pre.stool                                            0.000000                                                \n",
       "OAD-003.post.stool                                           0.023935                                                \n",
       "OAD-003.pre.stool                                            0.051513                                                \n",
       "OAD-004.post.stool                                           0.000371                                                \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.post.stool                                           0.152452                                                        \n",
       "OAD-001.pre.stool                                            0.000783                                                        \n",
       "OAD-003.post.stool                                           0.005789                                                        \n",
       "OAD-003.pre.stool                                            0.006045                                                        \n",
       "OAD-004.post.stool                                           0.014265                                                        \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400  \\\n",
       "index                                                                                                                    \n",
       "OAD-001.post.stool                                           0.033001                                                    \n",
       "OAD-001.pre.stool                                            0.026849                                                    \n",
       "OAD-003.post.stool                                           0.012060                                                    \n",
       "OAD-003.pre.stool                                            0.025918                                                    \n",
       "OAD-004.post.stool                                           0.090594                                                    \n",
       "\n",
       "                    ...  d__Bacteria;p__Firmicutes_A;__;__;__;__  \\\n",
       "index               ...                                            \n",
       "OAD-001.post.stool  ...                                      0.0   \n",
       "OAD-001.pre.stool   ...                                      0.0   \n",
       "OAD-003.post.stool  ...                                      0.0   \n",
       "OAD-003.pre.stool   ...                                      0.0   \n",
       "OAD-004.post.stool  ...                                      0.0   \n",
       "\n",
       "                    d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.post.stool                                                0.0                                                        \n",
       "OAD-001.pre.stool                                                 0.0                                                        \n",
       "OAD-003.post.stool                                                0.0                                                        \n",
       "OAD-003.pre.stool                                                 0.0                                                        \n",
       "OAD-004.post.stool                                                0.0                                                        \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "index                                                                                 \n",
       "OAD-001.post.stool                                                0.0                 \n",
       "OAD-001.pre.stool                                                 0.0                 \n",
       "OAD-003.post.stool                                                0.0                 \n",
       "OAD-003.pre.stool                                                 0.0                 \n",
       "OAD-004.post.stool                                                0.0                 \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "index                                                                                                                         \n",
       "OAD-001.post.stool                                                0.0                                                         \n",
       "OAD-001.pre.stool                                                 0.0                                                         \n",
       "OAD-003.post.stool                                                0.0                                                         \n",
       "OAD-003.pre.stool                                                 0.0                                                         \n",
       "OAD-004.post.stool                                                0.0                                                         \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "index                                                                                                               \n",
       "OAD-001.post.stool                                                0.0                                               \n",
       "OAD-001.pre.stool                                                 0.0                                               \n",
       "OAD-003.post.stool                                                0.0                                               \n",
       "OAD-003.pre.stool                                                 0.0                                               \n",
       "OAD-004.post.stool                                                0.0                                               \n",
       "\n",
       "                    d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "index                                                                                                                                         \n",
       "OAD-001.post.stool                                                0.0                                                                         \n",
       "OAD-001.pre.stool                                                 0.0                                                                         \n",
       "OAD-003.post.stool                                                0.0                                                                         \n",
       "OAD-003.pre.stool                                                 0.0                                                                         \n",
       "OAD-004.post.stool                                                0.0                                                                         \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "index                                                                                                     \n",
       "OAD-001.post.stool                                                0.0                                     \n",
       "OAD-001.pre.stool                                                 0.0                                     \n",
       "OAD-003.post.stool                                                0.0                                     \n",
       "OAD-003.pre.stool                                                 0.0                                     \n",
       "OAD-004.post.stool                                                0.0                                     \n",
       "\n",
       "                    d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "index                                                                                                                                        \n",
       "OAD-001.post.stool                                                0.0                                                                        \n",
       "OAD-001.pre.stool                                                 0.0                                                                        \n",
       "OAD-003.post.stool                                                0.0                                                                        \n",
       "OAD-003.pre.stool                                                 0.0                                                                        \n",
       "OAD-004.post.stool                                                0.0                                                                        \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "index                                                                                                  \n",
       "OAD-001.post.stool                                                0.0                                  \n",
       "OAD-001.pre.stool                                                 0.0                                  \n",
       "OAD-003.post.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                 0.0                                  \n",
       "OAD-004.post.stool                                                0.0                                  \n",
       "\n",
       "                    d__Bacteria;p__Actinobacteriota;__;__;__;__  \n",
       "index                                                            \n",
       "OAD-001.post.stool                                          0.0  \n",
       "OAD-001.pre.stool                                           0.0  \n",
       "OAD-003.post.stool                                          0.0  \n",
       "OAD-003.pre.stool                                           0.0  \n",
       "OAD-004.post.stool                                          0.0  \n",
       "\n",
       "[5 rows x 374 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specific taxa profiling\n",
    "GL_to_otu = {}\n",
    "\n",
    "groups = ['saliva_adh', 'stool_adh']\n",
    "levels = ['6','7']\n",
    "for g in groups:\n",
    "    GL_to_otu[g] = {}\n",
    "    for l in levels:\n",
    "        # read in csv\n",
    "        df_otu = pd.read_csv(path + 'outputs/Qiime2_' + g + '/level-' + l + '.csv', index_col=0)\n",
    "\n",
    "        # drop specific taxa\n",
    "        # if g == 'saliva_adh':\n",
    "        #    df_otu = df_otu.drop('VAOAD-004.post.saliva')\n",
    "        \n",
    "        # filter on OTU only \n",
    "        keep = []\n",
    "        for x in df_otu.columns:\n",
    "            if 'd__' in x:\n",
    "                keep.append(x)\n",
    "        df_otu = df_otu.loc[:,keep]\n",
    "\n",
    "        # normalize by dividing by row sum\n",
    "        df_otu = df_otu.div(df_otu.sum(axis=1),axis=0)\n",
    "        \n",
    "        # save to dict\n",
    "        GL_to_otu[g]['L' + l] = df_otu\n",
    "\n",
    "                                       \n",
    "GL_to_otu['stool_adh']['L6'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "515d9fc7-6c65-439d-8733-6c92c35fc5b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;__;__;__;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.022576</td>\n",
       "      <td>0.007098</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.292202</td>\n",
       "      <td>0.001994</td>\n",
       "      <td>0.013555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000783</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.181156</td>\n",
       "      <td>0.067844</td>\n",
       "      <td>0.150895</td>\n",
       "      <td>0.063229</td>\n",
       "      <td>0.051360</td>\n",
       "      <td>0.055600</td>\n",
       "      <td>0.012363</td>\n",
       "      <td>0.051513</td>\n",
       "      <td>0.006045</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.170334</td>\n",
       "      <td>0.135468</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002088</td>\n",
       "      <td>0.173617</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003649</td>\n",
       "      <td>0.008765</td>\n",
       "      <td>0.054802</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-005.pre.stool</th>\n",
       "      <td>0.040108</td>\n",
       "      <td>0.139385</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.018806</td>\n",
       "      <td>0.155290</td>\n",
       "      <td>0.037507</td>\n",
       "      <td>0.024457</td>\n",
       "      <td>0.066852</td>\n",
       "      <td>0.008969</td>\n",
       "      <td>0.018328</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-006.pre.stool</th>\n",
       "      <td>0.037219</td>\n",
       "      <td>0.167448</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007940</td>\n",
       "      <td>0.019591</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011006</td>\n",
       "      <td>0.001275</td>\n",
       "      <td>0.012769</td>\n",
       "      <td>0.162962</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 374 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004  \\\n",
       "index                                                                                                                      \n",
       "OAD-001.pre.stool                                           0.022576                                                       \n",
       "OAD-003.pre.stool                                           0.181156                                                       \n",
       "OAD-004.pre.stool                                           0.170334                                                       \n",
       "OAD-005.pre.stool                                           0.040108                                                       \n",
       "OAD-006.pre.stool                                           0.037219                                                       \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium  \\\n",
       "index                                                                                                                           \n",
       "OAD-001.pre.stool                                           0.007098                                                            \n",
       "OAD-003.pre.stool                                           0.067844                                                            \n",
       "OAD-004.pre.stool                                           0.135468                                                            \n",
       "OAD-005.pre.stool                                           0.139385                                                            \n",
       "OAD-006.pre.stool                                           0.167448                                                            \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella  \\\n",
       "index                                                                                                            \n",
       "OAD-001.pre.stool                                           0.005128                                             \n",
       "OAD-003.pre.stool                                           0.150895                                             \n",
       "OAD-004.pre.stool                                           0.000000                                             \n",
       "OAD-005.pre.stool                                           0.000000                                             \n",
       "OAD-006.pre.stool                                           0.000000                                             \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117  \\\n",
       "index                                                                                                                             \n",
       "OAD-001.pre.stool                                           0.000000                                                              \n",
       "OAD-003.pre.stool                                           0.063229                                                              \n",
       "OAD-004.pre.stool                                           0.002088                                                              \n",
       "OAD-005.pre.stool                                           0.018806                                                              \n",
       "OAD-006.pre.stool                                           0.007940                                                              \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H  \\\n",
       "index                                                                                                               \n",
       "OAD-001.pre.stool                                           0.292202                                                \n",
       "OAD-003.pre.stool                                           0.051360                                                \n",
       "OAD-004.pre.stool                                           0.173617                                                \n",
       "OAD-005.pre.stool                                           0.155290                                                \n",
       "OAD-006.pre.stool                                           0.019591                                                \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum  \\\n",
       "index                                                                                                                 \n",
       "OAD-001.pre.stool                                           0.001994                                                  \n",
       "OAD-003.pre.stool                                           0.055600                                                  \n",
       "OAD-004.pre.stool                                           0.000143                                                  \n",
       "OAD-005.pre.stool                                           0.037507                                                  \n",
       "OAD-006.pre.stool                                           0.000000                                                  \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129  \\\n",
       "index                                                                                                                           \n",
       "OAD-001.pre.stool                                           0.013555                                                            \n",
       "OAD-003.pre.stool                                           0.012363                                                            \n",
       "OAD-004.pre.stool                                           0.000000                                                            \n",
       "OAD-005.pre.stool                                           0.024457                                                            \n",
       "OAD-006.pre.stool                                           0.011006                                                            \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus  \\\n",
       "index                                                                                                               \n",
       "OAD-001.pre.stool                                           0.000000                                                \n",
       "OAD-003.pre.stool                                           0.051513                                                \n",
       "OAD-004.pre.stool                                           0.003649                                                \n",
       "OAD-005.pre.stool                                           0.066852                                                \n",
       "OAD-006.pre.stool                                           0.001275                                                \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor  \\\n",
       "index                                                                                                                       \n",
       "OAD-001.pre.stool                                           0.000783                                                        \n",
       "OAD-003.pre.stool                                           0.006045                                                        \n",
       "OAD-004.pre.stool                                           0.008765                                                        \n",
       "OAD-005.pre.stool                                           0.008969                                                        \n",
       "OAD-006.pre.stool                                           0.012769                                                        \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400  \\\n",
       "index                                                                                                                   \n",
       "OAD-001.pre.stool                                           0.026849                                                    \n",
       "OAD-003.pre.stool                                           0.025918                                                    \n",
       "OAD-004.pre.stool                                           0.054802                                                    \n",
       "OAD-005.pre.stool                                           0.018328                                                    \n",
       "OAD-006.pre.stool                                           0.162962                                                    \n",
       "\n",
       "                   ...  d__Bacteria;p__Firmicutes_A;__;__;__;__  \\\n",
       "index              ...                                            \n",
       "OAD-001.pre.stool  ...                                      0.0   \n",
       "OAD-003.pre.stool  ...                                      0.0   \n",
       "OAD-004.pre.stool  ...                                      0.0   \n",
       "OAD-005.pre.stool  ...                                      0.0   \n",
       "OAD-006.pre.stool  ...                                      0.0   \n",
       "\n",
       "                   d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "index                                                                                                                       \n",
       "OAD-001.pre.stool                                                0.0                                                        \n",
       "OAD-003.pre.stool                                                0.0                                                        \n",
       "OAD-004.pre.stool                                                0.0                                                        \n",
       "OAD-005.pre.stool                                                0.0                                                        \n",
       "OAD-006.pre.stool                                                0.0                                                        \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "index                                                                                \n",
       "OAD-001.pre.stool                                                0.0                 \n",
       "OAD-003.pre.stool                                                0.0                 \n",
       "OAD-004.pre.stool                                                0.0                 \n",
       "OAD-005.pre.stool                                                0.0                 \n",
       "OAD-006.pre.stool                                                0.0                 \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.pre.stool                                                0.0                                                         \n",
       "OAD-003.pre.stool                                                0.0                                                         \n",
       "OAD-004.pre.stool                                                0.0                                                         \n",
       "OAD-005.pre.stool                                                0.0                                                         \n",
       "OAD-006.pre.stool                                                0.0                                                         \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "index                                                                                                              \n",
       "OAD-001.pre.stool                                                0.0                                               \n",
       "OAD-003.pre.stool                                                0.0                                               \n",
       "OAD-004.pre.stool                                                0.0                                               \n",
       "OAD-005.pre.stool                                                0.0                                               \n",
       "OAD-006.pre.stool                                                0.0                                               \n",
       "\n",
       "                   d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "index                                                                                                                                        \n",
       "OAD-001.pre.stool                                                0.0                                                                         \n",
       "OAD-003.pre.stool                                                0.0                                                                         \n",
       "OAD-004.pre.stool                                                0.0                                                                         \n",
       "OAD-005.pre.stool                                                0.0                                                                         \n",
       "OAD-006.pre.stool                                                0.0                                                                         \n",
       "\n",
       "                   d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "index                                                                                                    \n",
       "OAD-001.pre.stool                                                0.0                                     \n",
       "OAD-003.pre.stool                                                0.0                                     \n",
       "OAD-004.pre.stool                                                0.0                                     \n",
       "OAD-005.pre.stool                                                0.0                                     \n",
       "OAD-006.pre.stool                                                0.0                                     \n",
       "\n",
       "                   d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "index                                                                                                                                       \n",
       "OAD-001.pre.stool                                                0.0                                                                        \n",
       "OAD-003.pre.stool                                                0.0                                                                        \n",
       "OAD-004.pre.stool                                                0.0                                                                        \n",
       "OAD-005.pre.stool                                                0.0                                                                        \n",
       "OAD-006.pre.stool                                                0.0                                                                        \n",
       "\n",
       "                   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "index                                                                                                 \n",
       "OAD-001.pre.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                0.0                                  \n",
       "OAD-004.pre.stool                                                0.0                                  \n",
       "OAD-005.pre.stool                                                0.0                                  \n",
       "OAD-006.pre.stool                                                0.0                                  \n",
       "\n",
       "                   d__Bacteria;p__Actinobacteriota;__;__;__;__  \n",
       "index                                                           \n",
       "OAD-001.pre.stool                                          0.0  \n",
       "OAD-003.pre.stool                                          0.0  \n",
       "OAD-004.pre.stool                                          0.0  \n",
       "OAD-005.pre.stool                                          0.0  \n",
       "OAD-006.pre.stool                                          0.0  \n",
       "\n",
       "[5 rows x 374 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Microbiome testing\n",
    "gt_to_otu = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gt_to_otu[g] = {}\n",
    "    df_otu = GL_to_otu[g + '_adh']['L6'].copy()\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['post','pre']:\n",
    "        df = df_otu[df_otu.index.to_series().str.contains(t)]\n",
    "        gt_to_otu[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_otu[g]['pre'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.split('.pre')[0])\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('-',''))\n",
    "    post_df = gt_to_otu[g]['post'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.split('.post')[0])\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('-',''))\n",
    "    \n",
    "    gt_to_otu[g]['diff'] = post_df - pre_df\n",
    "    gt_to_otu[g]['all'] = pd.concat([gt_to_otu[g]['pre'],gt_to_otu[g]['post']])\n",
    "\n",
    "gt_to_otu[g][t].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "24c44852-2e85-4174-9729-3fae6a355b32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# taxa parse function\n",
    "def read_taxa(taxa, delim=';'):\n",
    "    \"\"\"\n",
    "    Converts string of OTU names (e.g. from QIIME) to shortened form.\n",
    "    ----------------------------------------------------------------------------\n",
    "    INPUTS\n",
    "    taxa  - String. Long name of OTU e.g. 'k__Archaea;p__Crenarchaeota;\n",
    "            c__Thaumarchaeota;o__Cenarchaeales;f__Cenarchaeaceae;\n",
    "            g__Nitrosopumilus' which will become 'Cenarchaeaceae Nitrosopumilus'\n",
    "    delim - String. Separates heirarchy.\n",
    "\n",
    "    OUTPUTS\n",
    "    String. Abridged taxa name.\n",
    "    \"\"\"\n",
    "    parts = taxa.split(delim) # set as param with default\n",
    "    while parts:\n",
    "        if len(parts) == 1:\n",
    "            return('u.c. ' + parts[0])\n",
    "        if not parts[-1].endswith('__'):\n",
    "            taxastr1 = parts[-2].split('__')[1]\n",
    "            taxastr2 = parts[-1].split('__')[1]\n",
    "            return taxastr1 + ' ' + taxastr2\n",
    "        else:\n",
    "            parts.pop()\n",
    "\n",
    "    # This should not be reached: \"k__;p__...\"\n",
    "    return 'Uncharacterized'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faf9a2fe-9f68-4cee-9e40-4d5b749b9f0e",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Process Metabolite DF ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6bf1fdd5-b592-4a4e-a5d4-3cefc2d28c9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_N-Acetylmuramic-Acid</th>\n",
       "      <th>H_D_Fructose</th>\n",
       "      <th>H_Paraxanthine</th>\n",
       "      <th>H_L-methionine</th>\n",
       "      <th>H_Xanthuneric_acid</th>\n",
       "      <th>H_Pyridoxic_acid</th>\n",
       "      <th>H_p_coumaric_acid</th>\n",
       "      <th>H_Lactose</th>\n",
       "      <th>H_Pantothenic_acid</th>\n",
       "      <th>...</th>\n",
       "      <th>H_Glycoursodeoxycholic_acid</th>\n",
       "      <th>H_Indole_3_propionic_acid</th>\n",
       "      <th>H_Yersiniabactin_Fe_adduct</th>\n",
       "      <th>H_Glycocholic_acid</th>\n",
       "      <th>H_Glycocholate</th>\n",
       "      <th>H_glycocholic_acid</th>\n",
       "      <th>H_Glycochenodeoxycholate</th>\n",
       "      <th>H_Deoxycholate</th>\n",
       "      <th>H_PE(22:6/0:0)_[M+H]+_C27H45N1O7P1</th>\n",
       "      <th>H_PE(18:2/0:0)_[M+H]+_C23H45N1O7P1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>0.035609</td>\n",
       "      <td>0.082371</td>\n",
       "      <td>0.379626</td>\n",
       "      <td>0.201167</td>\n",
       "      <td>-0.082362</td>\n",
       "      <td>0.016484</td>\n",
       "      <td>-0.414174</td>\n",
       "      <td>-0.718066</td>\n",
       "      <td>-0.344094</td>\n",
       "      <td>0.132369</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.126092</td>\n",
       "      <td>0.446683</td>\n",
       "      <td>-0.905441</td>\n",
       "      <td>0.448360</td>\n",
       "      <td>0.483734</td>\n",
       "      <td>0.562393</td>\n",
       "      <td>0.110112</td>\n",
       "      <td>0.591415</td>\n",
       "      <td>0.282179</td>\n",
       "      <td>0.304361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.003591</td>\n",
       "      <td>1.153334</td>\n",
       "      <td>0.448950</td>\n",
       "      <td>-1.240302</td>\n",
       "      <td>0.136791</td>\n",
       "      <td>-0.077398</td>\n",
       "      <td>-0.128398</td>\n",
       "      <td>-0.483053</td>\n",
       "      <td>-0.726048</td>\n",
       "      <td>-1.082906</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.512609</td>\n",
       "      <td>0.191755</td>\n",
       "      <td>-0.670443</td>\n",
       "      <td>-0.369924</td>\n",
       "      <td>-0.382214</td>\n",
       "      <td>-1.063725</td>\n",
       "      <td>-1.327175</td>\n",
       "      <td>-0.496949</td>\n",
       "      <td>0.380516</td>\n",
       "      <td>0.421219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>0.063048</td>\n",
       "      <td>0.145153</td>\n",
       "      <td>-0.073655</td>\n",
       "      <td>0.421159</td>\n",
       "      <td>-0.232476</td>\n",
       "      <td>-0.577305</td>\n",
       "      <td>-0.208879</td>\n",
       "      <td>0.117236</td>\n",
       "      <td>-0.309148</td>\n",
       "      <td>-0.279103</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.118668</td>\n",
       "      <td>-0.014176</td>\n",
       "      <td>1.399191</td>\n",
       "      <td>-0.768953</td>\n",
       "      <td>-0.781243</td>\n",
       "      <td>-1.462742</td>\n",
       "      <td>-0.674522</td>\n",
       "      <td>-0.895976</td>\n",
       "      <td>0.380375</td>\n",
       "      <td>0.305410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>-0.142699</td>\n",
       "      <td>0.625381</td>\n",
       "      <td>0.140133</td>\n",
       "      <td>1.029497</td>\n",
       "      <td>-0.014348</td>\n",
       "      <td>-0.021789</td>\n",
       "      <td>-0.385252</td>\n",
       "      <td>0.326122</td>\n",
       "      <td>-0.530950</td>\n",
       "      <td>0.034178</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.660924</td>\n",
       "      <td>0.202382</td>\n",
       "      <td>1.602954</td>\n",
       "      <td>-0.760978</td>\n",
       "      <td>0.001860</td>\n",
       "      <td>0.136519</td>\n",
       "      <td>-0.178280</td>\n",
       "      <td>-0.888002</td>\n",
       "      <td>-0.054682</td>\n",
       "      <td>-0.804769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD007</th>\n",
       "      <td>0.340004</td>\n",
       "      <td>0.957297</td>\n",
       "      <td>0.276825</td>\n",
       "      <td>0.384845</td>\n",
       "      <td>0.148415</td>\n",
       "      <td>-0.034565</td>\n",
       "      <td>0.136779</td>\n",
       "      <td>0.969607</td>\n",
       "      <td>0.163616</td>\n",
       "      <td>-1.061064</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.891056</td>\n",
       "      <td>0.117900</td>\n",
       "      <td>1.199539</td>\n",
       "      <td>-0.348083</td>\n",
       "      <td>-0.360373</td>\n",
       "      <td>-1.041883</td>\n",
       "      <td>-0.170514</td>\n",
       "      <td>-0.475108</td>\n",
       "      <td>0.185225</td>\n",
       "      <td>-0.391874</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          H_Pyridoxamine  H_N-Acetylmuramic-Acid  H_D_Fructose  \\\n",
       "Study_ID                                                         \n",
       "OAD006          0.035609                0.082371      0.379626   \n",
       "OAD001          0.003591                1.153334      0.448950   \n",
       "OAD003          0.063048                0.145153     -0.073655   \n",
       "OAD005         -0.142699                0.625381      0.140133   \n",
       "OAD007          0.340004                0.957297      0.276825   \n",
       "\n",
       "          H_Paraxanthine  H_L-methionine  H_Xanthuneric_acid  \\\n",
       "Study_ID                                                       \n",
       "OAD006          0.201167       -0.082362            0.016484   \n",
       "OAD001         -1.240302        0.136791           -0.077398   \n",
       "OAD003          0.421159       -0.232476           -0.577305   \n",
       "OAD005          1.029497       -0.014348           -0.021789   \n",
       "OAD007          0.384845        0.148415           -0.034565   \n",
       "\n",
       "          H_Pyridoxic_acid  H_p_coumaric_acid  H_Lactose  H_Pantothenic_acid  \\\n",
       "Study_ID                                                                       \n",
       "OAD006           -0.414174          -0.718066  -0.344094            0.132369   \n",
       "OAD001           -0.128398          -0.483053  -0.726048           -1.082906   \n",
       "OAD003           -0.208879           0.117236  -0.309148           -0.279103   \n",
       "OAD005           -0.385252           0.326122  -0.530950            0.034178   \n",
       "OAD007            0.136779           0.969607   0.163616           -1.061064   \n",
       "\n",
       "          ...  H_Glycoursodeoxycholic_acid  H_Indole_3_propionic_acid  \\\n",
       "Study_ID  ...                                                           \n",
       "OAD006    ...                    -0.126092                   0.446683   \n",
       "OAD001    ...                    -0.512609                   0.191755   \n",
       "OAD003    ...                    -0.118668                  -0.014176   \n",
       "OAD005    ...                    -0.660924                   0.202382   \n",
       "OAD007    ...                    -0.891056                   0.117900   \n",
       "\n",
       "          H_Yersiniabactin_Fe_adduct  H_Glycocholic_acid  H_Glycocholate  \\\n",
       "Study_ID                                                                   \n",
       "OAD006                     -0.905441            0.448360        0.483734   \n",
       "OAD001                     -0.670443           -0.369924       -0.382214   \n",
       "OAD003                      1.399191           -0.768953       -0.781243   \n",
       "OAD005                      1.602954           -0.760978        0.001860   \n",
       "OAD007                      1.199539           -0.348083       -0.360373   \n",
       "\n",
       "          H_glycocholic_acid  H_Glycochenodeoxycholate  H_Deoxycholate  \\\n",
       "Study_ID                                                                 \n",
       "OAD006              0.562393                  0.110112        0.591415   \n",
       "OAD001             -1.063725                 -1.327175       -0.496949   \n",
       "OAD003             -1.462742                 -0.674522       -0.895976   \n",
       "OAD005              0.136519                 -0.178280       -0.888002   \n",
       "OAD007             -1.041883                 -0.170514       -0.475108   \n",
       "\n",
       "          H_PE(22:6/0:0)_[M+H]+_C27H45N1O7P1  \\\n",
       "Study_ID                                       \n",
       "OAD006                              0.282179   \n",
       "OAD001                              0.380516   \n",
       "OAD003                              0.380375   \n",
       "OAD005                             -0.054682   \n",
       "OAD007                              0.185225   \n",
       "\n",
       "          H_PE(18:2/0:0)_[M+H]+_C23H45N1O7P1  \n",
       "Study_ID                                      \n",
       "OAD006                              0.304361  \n",
       "OAD001                              0.421219  \n",
       "OAD003                              0.305410  \n",
       "OAD005                             -0.804769  \n",
       "OAD007                             -0.391874  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Metabolome testing\n",
    "gdt_to_df = {}\n",
    "for g in ['saliva','stool','plasma']:\n",
    "    gdt_to_df[g] = {}\n",
    "    for d in ['paired_beta']: # just arbitrarily \n",
    "        gdt_to_df[g][d] = {}\n",
    "        df_meta = pd.read_csv(path + 'inputs/' + g + '_normalized.csv')\n",
    "\n",
    "        # split on timepoint\n",
    "        for t in ['Baseline','After diet']:\n",
    "            df = df_meta[df_meta['Time'] == t]\n",
    "            df = df.set_index('Study_ID')\n",
    "            df.index = df.index.map(lambda x: x.split('_')[0])\n",
    "            df = df.drop('Time',axis=1)\n",
    "            gdt_to_df[g][d][t] = df\n",
    "\n",
    "        gdt_to_df[g][d]['diff'] = gdt_to_df[g][d]['After diet'] - gdt_to_df[g][d]['Baseline']\n",
    "\n",
    "gdt_to_df[g][d][t].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "dbad4aa2-7c35-490c-900b-c607c4ff549d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_N-Acetylmuramic-Acid</th>\n",
       "      <th>H_D_Fructose</th>\n",
       "      <th>H_Paraxanthine</th>\n",
       "      <th>H_L-methionine</th>\n",
       "      <th>H_Xanthuneric_acid</th>\n",
       "      <th>H_Pyridoxic_acid</th>\n",
       "      <th>H_p_coumaric_acid</th>\n",
       "      <th>H_Lactose</th>\n",
       "      <th>H_Pantothenic_acid</th>\n",
       "      <th>...</th>\n",
       "      <th>H_Glycoursodeoxycholic_acid</th>\n",
       "      <th>H_Indole_3_propionic_acid</th>\n",
       "      <th>H_Yersiniabactin_Fe_adduct</th>\n",
       "      <th>H_Glycocholic_acid</th>\n",
       "      <th>H_Glycocholate</th>\n",
       "      <th>H_glycocholic_acid</th>\n",
       "      <th>H_Glycochenodeoxycholate</th>\n",
       "      <th>H_Deoxycholate</th>\n",
       "      <th>H_PE(22:6/0:0)_[M+H]+_C27H45N1O7P1</th>\n",
       "      <th>H_PE(18:2/0:0)_[M+H]+_C23H45N1O7P1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>0.035609</td>\n",
       "      <td>0.082371</td>\n",
       "      <td>0.379626</td>\n",
       "      <td>0.201167</td>\n",
       "      <td>-0.082362</td>\n",
       "      <td>0.016484</td>\n",
       "      <td>-0.414174</td>\n",
       "      <td>-0.718066</td>\n",
       "      <td>-0.344094</td>\n",
       "      <td>0.132369</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.126092</td>\n",
       "      <td>0.446683</td>\n",
       "      <td>-0.905441</td>\n",
       "      <td>0.448360</td>\n",
       "      <td>0.483734</td>\n",
       "      <td>0.562393</td>\n",
       "      <td>0.110112</td>\n",
       "      <td>0.591415</td>\n",
       "      <td>0.282179</td>\n",
       "      <td>0.304361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.003591</td>\n",
       "      <td>1.153334</td>\n",
       "      <td>0.448950</td>\n",
       "      <td>-1.240302</td>\n",
       "      <td>0.136791</td>\n",
       "      <td>-0.077398</td>\n",
       "      <td>-0.128398</td>\n",
       "      <td>-0.483053</td>\n",
       "      <td>-0.726048</td>\n",
       "      <td>-1.082906</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.512609</td>\n",
       "      <td>0.191755</td>\n",
       "      <td>-0.670443</td>\n",
       "      <td>-0.369924</td>\n",
       "      <td>-0.382214</td>\n",
       "      <td>-1.063725</td>\n",
       "      <td>-1.327175</td>\n",
       "      <td>-0.496949</td>\n",
       "      <td>0.380516</td>\n",
       "      <td>0.421219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>0.063048</td>\n",
       "      <td>0.145153</td>\n",
       "      <td>-0.073655</td>\n",
       "      <td>0.421159</td>\n",
       "      <td>-0.232476</td>\n",
       "      <td>-0.577305</td>\n",
       "      <td>-0.208879</td>\n",
       "      <td>0.117236</td>\n",
       "      <td>-0.309148</td>\n",
       "      <td>-0.279103</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.118668</td>\n",
       "      <td>-0.014176</td>\n",
       "      <td>1.399191</td>\n",
       "      <td>-0.768953</td>\n",
       "      <td>-0.781243</td>\n",
       "      <td>-1.462742</td>\n",
       "      <td>-0.674522</td>\n",
       "      <td>-0.895976</td>\n",
       "      <td>0.380375</td>\n",
       "      <td>0.305410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>-0.142699</td>\n",
       "      <td>0.625381</td>\n",
       "      <td>0.140133</td>\n",
       "      <td>1.029497</td>\n",
       "      <td>-0.014348</td>\n",
       "      <td>-0.021789</td>\n",
       "      <td>-0.385252</td>\n",
       "      <td>0.326122</td>\n",
       "      <td>-0.530950</td>\n",
       "      <td>0.034178</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.660924</td>\n",
       "      <td>0.202382</td>\n",
       "      <td>1.602954</td>\n",
       "      <td>-0.760978</td>\n",
       "      <td>0.001860</td>\n",
       "      <td>0.136519</td>\n",
       "      <td>-0.178280</td>\n",
       "      <td>-0.888002</td>\n",
       "      <td>-0.054682</td>\n",
       "      <td>-0.804769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD007</th>\n",
       "      <td>0.340004</td>\n",
       "      <td>0.957297</td>\n",
       "      <td>0.276825</td>\n",
       "      <td>0.384845</td>\n",
       "      <td>0.148415</td>\n",
       "      <td>-0.034565</td>\n",
       "      <td>0.136779</td>\n",
       "      <td>0.969607</td>\n",
       "      <td>0.163616</td>\n",
       "      <td>-1.061064</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.891056</td>\n",
       "      <td>0.117900</td>\n",
       "      <td>1.199539</td>\n",
       "      <td>-0.348083</td>\n",
       "      <td>-0.360373</td>\n",
       "      <td>-1.041883</td>\n",
       "      <td>-0.170514</td>\n",
       "      <td>-0.475108</td>\n",
       "      <td>0.185225</td>\n",
       "      <td>-0.391874</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          H_Pyridoxamine  H_N-Acetylmuramic-Acid  H_D_Fructose  \\\n",
       "Study_ID                                                         \n",
       "OAD006          0.035609                0.082371      0.379626   \n",
       "OAD001          0.003591                1.153334      0.448950   \n",
       "OAD003          0.063048                0.145153     -0.073655   \n",
       "OAD005         -0.142699                0.625381      0.140133   \n",
       "OAD007          0.340004                0.957297      0.276825   \n",
       "\n",
       "          H_Paraxanthine  H_L-methionine  H_Xanthuneric_acid  \\\n",
       "Study_ID                                                       \n",
       "OAD006          0.201167       -0.082362            0.016484   \n",
       "OAD001         -1.240302        0.136791           -0.077398   \n",
       "OAD003          0.421159       -0.232476           -0.577305   \n",
       "OAD005          1.029497       -0.014348           -0.021789   \n",
       "OAD007          0.384845        0.148415           -0.034565   \n",
       "\n",
       "          H_Pyridoxic_acid  H_p_coumaric_acid  H_Lactose  H_Pantothenic_acid  \\\n",
       "Study_ID                                                                       \n",
       "OAD006           -0.414174          -0.718066  -0.344094            0.132369   \n",
       "OAD001           -0.128398          -0.483053  -0.726048           -1.082906   \n",
       "OAD003           -0.208879           0.117236  -0.309148           -0.279103   \n",
       "OAD005           -0.385252           0.326122  -0.530950            0.034178   \n",
       "OAD007            0.136779           0.969607   0.163616           -1.061064   \n",
       "\n",
       "          ...  H_Glycoursodeoxycholic_acid  H_Indole_3_propionic_acid  \\\n",
       "Study_ID  ...                                                           \n",
       "OAD006    ...                    -0.126092                   0.446683   \n",
       "OAD001    ...                    -0.512609                   0.191755   \n",
       "OAD003    ...                    -0.118668                  -0.014176   \n",
       "OAD005    ...                    -0.660924                   0.202382   \n",
       "OAD007    ...                    -0.891056                   0.117900   \n",
       "\n",
       "          H_Yersiniabactin_Fe_adduct  H_Glycocholic_acid  H_Glycocholate  \\\n",
       "Study_ID                                                                   \n",
       "OAD006                     -0.905441            0.448360        0.483734   \n",
       "OAD001                     -0.670443           -0.369924       -0.382214   \n",
       "OAD003                      1.399191           -0.768953       -0.781243   \n",
       "OAD005                      1.602954           -0.760978        0.001860   \n",
       "OAD007                      1.199539           -0.348083       -0.360373   \n",
       "\n",
       "          H_glycocholic_acid  H_Glycochenodeoxycholate  H_Deoxycholate  \\\n",
       "Study_ID                                                                 \n",
       "OAD006              0.562393                  0.110112        0.591415   \n",
       "OAD001             -1.063725                 -1.327175       -0.496949   \n",
       "OAD003             -1.462742                 -0.674522       -0.895976   \n",
       "OAD005              0.136519                 -0.178280       -0.888002   \n",
       "OAD007             -1.041883                 -0.170514       -0.475108   \n",
       "\n",
       "          H_PE(22:6/0:0)_[M+H]+_C27H45N1O7P1  \\\n",
       "Study_ID                                       \n",
       "OAD006                              0.282179   \n",
       "OAD001                              0.380516   \n",
       "OAD003                              0.380375   \n",
       "OAD005                             -0.054682   \n",
       "OAD007                              0.185225   \n",
       "\n",
       "          H_PE(18:2/0:0)_[M+H]+_C23H45N1O7P1  \n",
       "Study_ID                                      \n",
       "OAD006                              0.304361  \n",
       "OAD001                              0.421219  \n",
       "OAD003                              0.305410  \n",
       "OAD005                             -0.804769  \n",
       "OAD007                             -0.391874  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Metabolome testing\n",
    "gt_to_meta = {}\n",
    "for g in ['saliva','stool','plasma']:\n",
    "    gt_to_meta[g] = {}\n",
    "    df_meta = pd.read_csv(path + 'inputs/' + g + '_normalized.csv')\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['Baseline','After diet']:\n",
    "        df = df_meta[df_meta['Time'] == t]\n",
    "        df = df.set_index('Study_ID')\n",
    "        df.index = df.index.map(lambda x: x.split('_')[0])\n",
    "        df = df.drop('Time',axis=1)\n",
    "        gt_to_meta[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_meta[g]['Baseline'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('D0','D-0'))\n",
    "    pre_df.index = pre_df.index.map(lambda x: x + '.pre.' + g)\n",
    "    post_df = gt_to_meta[g]['After diet'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('D0','D-0'))\n",
    "    post_df.index = post_df.index.map(lambda x: x + '.post.' + g)\n",
    "    \n",
    "    gt_to_meta[g]['diff'] = gt_to_meta[g]['After diet'] - gt_to_meta[g]['Baseline']\n",
    "    gt_to_meta[g]['all'] = pd.concat([pre_df,post_df])\n",
    "    \n",
    "\n",
    "gt_to_meta[g][t].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "801b5fe1-2f5b-465e-b21a-943c574d1ee7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Pathway DF Processing ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8b418d85-6211-4131-9007-92d2b5f1a3ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - globo and isoglobo series</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid elongation</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of 12-, 14- and 16-membered macrolides</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of enediyne antibiotics</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.saliva</th>\n",
       "      <td>0.155479</td>\n",
       "      <td>0.025863</td>\n",
       "      <td>0.001866</td>\n",
       "      <td>0.003177</td>\n",
       "      <td>0.008013</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.026699</td>\n",
       "      <td>0.000547</td>\n",
       "      <td>0.031793</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>2.433861e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.216931e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.saliva</th>\n",
       "      <td>0.159679</td>\n",
       "      <td>0.027060</td>\n",
       "      <td>0.001949</td>\n",
       "      <td>0.003082</td>\n",
       "      <td>0.007239</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.027470</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.034357</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>1.911111e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.539150e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.saliva</th>\n",
       "      <td>0.151338</td>\n",
       "      <td>0.027415</td>\n",
       "      <td>0.002223</td>\n",
       "      <td>0.002702</td>\n",
       "      <td>0.009135</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.021911</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.030358</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.saliva</th>\n",
       "      <td>0.152707</td>\n",
       "      <td>0.026974</td>\n",
       "      <td>0.002172</td>\n",
       "      <td>0.002788</td>\n",
       "      <td>0.008887</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.022846</td>\n",
       "      <td>0.000257</td>\n",
       "      <td>0.031226</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.post.saliva</th>\n",
       "      <td>0.159885</td>\n",
       "      <td>0.027003</td>\n",
       "      <td>0.001989</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>0.007319</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.026641</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.034499</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>6.002574e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.594595e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 168 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "KEGG_Pathway         Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.post.saliva    0.155479                            0.025863   \n",
       "OAD-001.pre.saliva     0.159679                            0.027060   \n",
       "OAD-003.post.saliva    0.151338                            0.027415   \n",
       "OAD-003.pre.saliva     0.152707                            0.026974   \n",
       "OAD-004.post.saliva    0.159885                            0.027003   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.post.saliva                                           0.001866                 \n",
       "OAD-001.pre.saliva                                            0.001949                 \n",
       "OAD-003.post.saliva                                           0.002223                 \n",
       "OAD-003.pre.saliva                                            0.002172                 \n",
       "OAD-004.post.saliva                                           0.001989                 \n",
       "\n",
       "KEGG_Pathway         Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.post.saliva                                           0.003177        \n",
       "OAD-001.pre.saliva                                            0.003082        \n",
       "OAD-003.post.saliva                                           0.002702        \n",
       "OAD-003.pre.saliva                                            0.002788        \n",
       "OAD-004.post.saliva                                           0.002953        \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism  \\\n",
       "OAD-001.post.saliva                     0.008013   \n",
       "OAD-001.pre.saliva                      0.007239   \n",
       "OAD-003.post.saliva                     0.009135   \n",
       "OAD-003.pre.saliva                      0.008887   \n",
       "OAD-004.post.saliva                     0.007319   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.post.saliva                                           0.000273    \n",
       "OAD-001.pre.saliva                                            0.000166    \n",
       "OAD-003.post.saliva                                           0.000180    \n",
       "OAD-003.pre.saliva                                            0.000176    \n",
       "OAD-004.post.saliva                                           0.000132    \n",
       "\n",
       "KEGG_Pathway         Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.post.saliva                          0.026699   \n",
       "OAD-001.pre.saliva                           0.027470   \n",
       "OAD-003.post.saliva                          0.021911   \n",
       "OAD-003.pre.saliva                           0.022846   \n",
       "OAD-004.post.saliva                          0.026641   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.post.saliva                                           0.000547      \n",
       "OAD-001.pre.saliva                                            0.000309      \n",
       "OAD-003.post.saliva                                           0.000261      \n",
       "OAD-003.pre.saliva                                            0.000257      \n",
       "OAD-004.post.saliva                                           0.000200      \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.post.saliva                                         0.031793   \n",
       "OAD-001.pre.saliva                                          0.034357   \n",
       "OAD-003.post.saliva                                         0.030358   \n",
       "OAD-003.pre.saliva                                          0.031226   \n",
       "OAD-004.post.saliva                                         0.034499   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.post.saliva                                           0.000046                    \n",
       "OAD-001.pre.saliva                                            0.000045                    \n",
       "OAD-003.post.saliva                                           0.000042                    \n",
       "OAD-003.pre.saliva                                            0.000041                    \n",
       "OAD-004.post.saliva                                           0.000038                    \n",
       "\n",
       "KEGG_Pathway         ...  \\\n",
       "OAD-001.post.saliva  ...   \n",
       "OAD-001.pre.saliva   ...   \n",
       "OAD-003.post.saliva  ...   \n",
       "OAD-003.pre.saliva   ...   \n",
       "OAD-004.post.saliva  ...   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - globo and isoglobo series  \\\n",
       "OAD-001.post.saliva                                           0.000021                                                          \n",
       "OAD-001.pre.saliva                                            0.000045                                                          \n",
       "OAD-003.post.saliva                                           0.000103                                                          \n",
       "OAD-003.pre.saliva                                            0.000080                                                          \n",
       "OAD-004.post.saliva                                           0.000062                                                          \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism|Fatty acid elongation  \\\n",
       "OAD-001.post.saliva                                                0.0   \n",
       "OAD-001.pre.saliva                                                 0.0   \n",
       "OAD-003.post.saliva                                                0.0   \n",
       "OAD-003.pre.saliva                                                 0.0   \n",
       "OAD-004.post.saliva                                                0.0   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis  \\\n",
       "OAD-001.post.saliva                                           0.000189                                    \n",
       "OAD-001.pre.saliva                                            0.000095                                    \n",
       "OAD-003.post.saliva                                           0.000022                                    \n",
       "OAD-003.pre.saliva                                            0.000018                                    \n",
       "OAD-004.post.saliva                                           0.000039                                    \n",
       "\n",
       "KEGG_Pathway         Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series  \\\n",
       "OAD-001.post.saliva                                           0.000010                                               \n",
       "OAD-001.pre.saliva                                            0.000019                                               \n",
       "OAD-003.post.saliva                                           0.000060                                               \n",
       "OAD-003.pre.saliva                                            0.000043                                               \n",
       "OAD-004.post.saliva                                           0.000033                                               \n",
       "\n",
       "KEGG_Pathway         Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.post.saliva                                       2.433861e-07                                   \n",
       "OAD-001.pre.saliva                                        1.911111e-06                                   \n",
       "OAD-003.post.saliva                                       0.000000e+00                                   \n",
       "OAD-003.pre.saliva                                        0.000000e+00                                   \n",
       "OAD-004.post.saliva                                       6.002574e-07                                   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.post.saliva                                                0.0                                  \n",
       "OAD-001.pre.saliva                                                 0.0                                  \n",
       "OAD-003.post.saliva                                                0.0                                  \n",
       "OAD-003.pre.saliva                                                 0.0                                  \n",
       "OAD-004.post.saliva                                                0.0                                  \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of 12-, 14- and 16-membered macrolides  \\\n",
       "OAD-001.post.saliva                                                0.0                                                         \n",
       "OAD-001.pre.saliva                                                 0.0                                                         \n",
       "OAD-003.post.saliva                                                0.0                                                         \n",
       "OAD-003.pre.saliva                                                 0.0                                                         \n",
       "OAD-004.post.saliva                                                0.0                                                         \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of enediyne antibiotics  \\\n",
       "OAD-001.post.saliva                                                0.0                                          \n",
       "OAD-001.pre.saliva                                                 0.0                                          \n",
       "OAD-003.post.saliva                                                0.0                                          \n",
       "OAD-003.pre.saliva                                                 0.0                                          \n",
       "OAD-004.post.saliva                                                0.0                                          \n",
       "\n",
       "KEGG_Pathway         Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.post.saliva                                       0.000000e+00                           \n",
       "OAD-001.pre.saliva                                        0.000000e+00                           \n",
       "OAD-003.post.saliva                                       0.000000e+00                           \n",
       "OAD-003.pre.saliva                                        0.000000e+00                           \n",
       "OAD-004.post.saliva                                       8.594595e-07                           \n",
       "\n",
       "KEGG_Pathway         Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \n",
       "OAD-001.post.saliva                                       1.216931e-07                                         \n",
       "OAD-001.pre.saliva                                        1.539150e-07                                         \n",
       "OAD-003.post.saliva                                       0.000000e+00                                         \n",
       "OAD-003.pre.saliva                                        0.000000e+00                                         \n",
       "OAD-004.post.saliva                                       0.000000e+00                                         \n",
       "\n",
       "[5 rows x 168 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specific pathway profiling\n",
    "g_to_path = {}\n",
    "\n",
    "groups = ['saliva_adh', 'stool_adh']\n",
    "g_to_job = {\n",
    "    'saliva_adh': 'jobs07',\n",
    "    'stool_adh': 'jobs08'\n",
    "}\n",
    "for g in groups:\n",
    "    # read in csv\n",
    "    df_path = pd.read_csv(path + 'outputs/' + g_to_job[g] + '/pred_metagenome_unstrat_kegg.tsv', sep ='\\t', index_col=0)\n",
    "\n",
    "    # transpose\n",
    "    df_path = df_path.T\n",
    "    \n",
    "    # normalize by dividing by row sum\n",
    "    df_path = df_path.div(df_path.sum(axis=1),axis=0)\n",
    "\n",
    "    # filter on metabolism only \n",
    "    keep = []\n",
    "    for p in df_path.columns:\n",
    "        if 'Metabolism' in p: # or 'metabolism' in p: 168 vs 170 if you incl both for saliva\n",
    "            keep.append(p)\n",
    "    df_path = df_path[keep]\n",
    "    \n",
    "    # save to dict\n",
    "    g_to_path[g] = df_path\n",
    "                                 \n",
    "g_to_path['saliva_adh'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "34ed8f12-5226-47f7-b748-07609e554b8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Degradation of flavonoids</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis</th>\n",
       "      <th>Metabolism|Lipid metabolism|Steroid hormone biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.159285</td>\n",
       "      <td>0.035029</td>\n",
       "      <td>0.002320</td>\n",
       "      <td>0.003018</td>\n",
       "      <td>0.007937</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>0.027626</td>\n",
       "      <td>0.000229</td>\n",
       "      <td>0.026349</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000469</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>3.060096e-06</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.510913e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.161049</td>\n",
       "      <td>0.033134</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.002538</td>\n",
       "      <td>0.008606</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.029092</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.027987</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>9.470469e-06</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.511851e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.161760</td>\n",
       "      <td>0.034016</td>\n",
       "      <td>0.002254</td>\n",
       "      <td>0.002773</td>\n",
       "      <td>0.008658</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.029185</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>0.027522</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>2.489384e-06</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.100744e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-005.pre.stool</th>\n",
       "      <td>0.158015</td>\n",
       "      <td>0.033087</td>\n",
       "      <td>0.002185</td>\n",
       "      <td>0.002702</td>\n",
       "      <td>0.008593</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.028749</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>0.027538</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000405</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>2.762671e-06</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>1.103744e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.297970e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-006.pre.stool</th>\n",
       "      <td>0.157791</td>\n",
       "      <td>0.032210</td>\n",
       "      <td>0.002112</td>\n",
       "      <td>0.002819</td>\n",
       "      <td>0.008727</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.030094</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.026678</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000329</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>2.184941e-07</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.127712e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 159 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "KEGG_Pathway       Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.pre.stool    0.159285                            0.035029   \n",
       "OAD-003.pre.stool    0.161049                            0.033134   \n",
       "OAD-004.pre.stool    0.161760                            0.034016   \n",
       "OAD-005.pre.stool    0.158015                            0.033087   \n",
       "OAD-006.pre.stool    0.157791                            0.032210   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.pre.stool                                           0.002320                 \n",
       "OAD-003.pre.stool                                           0.002184                 \n",
       "OAD-004.pre.stool                                           0.002254                 \n",
       "OAD-005.pre.stool                                           0.002185                 \n",
       "OAD-006.pre.stool                                           0.002112                 \n",
       "\n",
       "KEGG_Pathway       Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.pre.stool                                           0.003018        \n",
       "OAD-003.pre.stool                                           0.002538        \n",
       "OAD-004.pre.stool                                           0.002773        \n",
       "OAD-005.pre.stool                                           0.002702        \n",
       "OAD-006.pre.stool                                           0.002819        \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism  \\\n",
       "OAD-001.pre.stool                     0.007937   \n",
       "OAD-003.pre.stool                     0.008606   \n",
       "OAD-004.pre.stool                     0.008658   \n",
       "OAD-005.pre.stool                     0.008593   \n",
       "OAD-006.pre.stool                     0.008727   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.pre.stool                                           0.000192    \n",
       "OAD-003.pre.stool                                           0.000180    \n",
       "OAD-004.pre.stool                                           0.000157    \n",
       "OAD-005.pre.stool                                           0.000186    \n",
       "OAD-006.pre.stool                                           0.000199    \n",
       "\n",
       "KEGG_Pathway       Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.pre.stool                          0.027626   \n",
       "OAD-003.pre.stool                          0.029092   \n",
       "OAD-004.pre.stool                          0.029185   \n",
       "OAD-005.pre.stool                          0.028749   \n",
       "OAD-006.pre.stool                          0.030094   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.pre.stool                                           0.000229      \n",
       "OAD-003.pre.stool                                           0.000194      \n",
       "OAD-004.pre.stool                                           0.000173      \n",
       "OAD-005.pre.stool                                           0.000216      \n",
       "OAD-006.pre.stool                                           0.000220      \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.pre.stool                                         0.026349   \n",
       "OAD-003.pre.stool                                         0.027987   \n",
       "OAD-004.pre.stool                                         0.027522   \n",
       "OAD-005.pre.stool                                         0.027538   \n",
       "OAD-006.pre.stool                                         0.026678   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.pre.stool                                           0.000012                    \n",
       "OAD-003.pre.stool                                           0.000019                    \n",
       "OAD-004.pre.stool                                           0.000016                    \n",
       "OAD-005.pre.stool                                           0.000016                    \n",
       "OAD-006.pre.stool                                           0.000013                    \n",
       "\n",
       "KEGG_Pathway       ...  \\\n",
       "OAD-001.pre.stool  ...   \n",
       "OAD-003.pre.stool  ...   \n",
       "OAD-004.pre.stool  ...   \n",
       "OAD-005.pre.stool  ...   \n",
       "OAD-006.pre.stool  ...   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Degradation of flavonoids  \\\n",
       "OAD-001.pre.stool                                           0.000469                                  \n",
       "OAD-003.pre.stool                                           0.000482                                  \n",
       "OAD-004.pre.stool                                           0.000635                                  \n",
       "OAD-005.pre.stool                                           0.000405                                  \n",
       "OAD-006.pre.stool                                           0.000329                                  \n",
       "\n",
       "KEGG_Pathway       Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis  \\\n",
       "OAD-001.pre.stool                                           0.000028                                      \n",
       "OAD-003.pre.stool                                           0.000090                                      \n",
       "OAD-004.pre.stool                                           0.000103                                      \n",
       "OAD-005.pre.stool                                           0.000055                                      \n",
       "OAD-006.pre.stool                                           0.000087                                      \n",
       "\n",
       "KEGG_Pathway       Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis  \\\n",
       "OAD-001.pre.stool                                       3.060096e-06                                    \n",
       "OAD-003.pre.stool                                       9.470469e-06                                    \n",
       "OAD-004.pre.stool                                       2.489384e-06                                    \n",
       "OAD-005.pre.stool                                       2.762671e-06                                    \n",
       "OAD-006.pre.stool                                       2.184941e-07                                    \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism|Steroid hormone biosynthesis  \\\n",
       "OAD-001.pre.stool                                           0.000026          \n",
       "OAD-003.pre.stool                                           0.000118          \n",
       "OAD-004.pre.stool                                           0.000108          \n",
       "OAD-005.pre.stool                                           0.000092          \n",
       "OAD-006.pre.stool                                           0.000088          \n",
       "\n",
       "KEGG_Pathway       Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series  \\\n",
       "OAD-001.pre.stool                                           0.000026                                               \n",
       "OAD-003.pre.stool                                           0.000090                                               \n",
       "OAD-004.pre.stool                                           0.000103                                               \n",
       "OAD-005.pre.stool                                           0.000055                                               \n",
       "OAD-006.pre.stool                                           0.000086                                               \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.pre.stool                                       0.000000e+00                                   \n",
       "OAD-003.pre.stool                                       0.000000e+00                                   \n",
       "OAD-004.pre.stool                                       0.000000e+00                                   \n",
       "OAD-005.pre.stool                                       1.103744e-07                                   \n",
       "OAD-006.pre.stool                                       0.000000e+00                                   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.pre.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                0.0                                  \n",
       "OAD-004.pre.stool                                                0.0                                  \n",
       "OAD-005.pre.stool                                                0.0                                  \n",
       "OAD-006.pre.stool                                                0.0                                  \n",
       "\n",
       "KEGG_Pathway       Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.pre.stool                                       1.510913e-06                           \n",
       "OAD-003.pre.stool                                       8.511851e-07                           \n",
       "OAD-004.pre.stool                                       6.100744e-07                           \n",
       "OAD-005.pre.stool                                       5.297970e-07                           \n",
       "OAD-006.pre.stool                                       1.127712e-07                           \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis  \\\n",
       "OAD-001.pre.stool                                                0.0                               \n",
       "OAD-003.pre.stool                                                0.0                               \n",
       "OAD-004.pre.stool                                                0.0                               \n",
       "OAD-005.pre.stool                                                0.0                               \n",
       "OAD-006.pre.stool                                                0.0                               \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \n",
       "OAD-001.pre.stool                                                0.0                                         \n",
       "OAD-003.pre.stool                                                0.0                                         \n",
       "OAD-004.pre.stool                                                0.0                                         \n",
       "OAD-005.pre.stool                                                0.0                                         \n",
       "OAD-006.pre.stool                                                0.0                                         \n",
       "\n",
       "[5 rows x 159 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pathway testing\n",
    "gt_to_path = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gt_to_path[g] = {}\n",
    "    df_path = g_to_path[g + '_adh'].copy()\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['post','pre']:\n",
    "        df = df_path[df_path.index.to_series().str.contains(t)]\n",
    "        gt_to_path[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_path[g]['pre'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.split('.pre')[0])\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('-',''))\n",
    "    post_df = gt_to_path[g]['post'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.split('.post')[0])\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('-',''))\n",
    "    \n",
    "    gt_to_path[g]['diff'] = post_df - pre_df\n",
    "    gt_to_path[g]['all'] = pd.concat([gt_to_path[g]['pre'],gt_to_path[g]['post']])\n",
    "\n",
    "gt_to_path[g][t].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81bd68ab-5702-4758-ad7a-34bf139443ea",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Diet DF Processing ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "77c18bcf-830a-458d-88c1-70f509e7c9c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>animal_proteins_red</th>\n",
       "      <th>refined_grains</th>\n",
       "      <th>solanaceae</th>\n",
       "      <th>saturated</th>\n",
       "      <th>pastries</th>\n",
       "      <th>milk_derivates</th>\n",
       "      <th>beverages_prohibited</th>\n",
       "      <th>pro_inflammatory</th>\n",
       "      <th>chicken</th>\n",
       "      <th>plant_protein</th>\n",
       "      <th>...</th>\n",
       "      <th>corn</th>\n",
       "      <th>quinoa</th>\n",
       "      <th>broccoli</th>\n",
       "      <th>Garbanzo_beans</th>\n",
       "      <th>pork</th>\n",
       "      <th>beef</th>\n",
       "      <th>burger</th>\n",
       "      <th>Total_omega3</th>\n",
       "      <th>Total_omega6</th>\n",
       "      <th>Total_o3_o6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>131.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>14.0</td>\n",
       "      <td>18.4</td>\n",
       "      <td>5.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.6</td>\n",
       "      <td>36.4</td>\n",
       "      <td>33.6</td>\n",
       "      <td>113.6</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.2</td>\n",
       "      <td>54.8</td>\n",
       "      <td>96.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>-2.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.8</td>\n",
       "      <td>10.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.4</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.2</td>\n",
       "      <td>56.0</td>\n",
       "      <td>83.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>9.2</td>\n",
       "      <td>22.4</td>\n",
       "      <td>-5.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>3.2</td>\n",
       "      <td>3.2</td>\n",
       "      <td>23.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-78.4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.8</td>\n",
       "      <td>88.8</td>\n",
       "      <td>147.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>5.6</td>\n",
       "      <td>3.6</td>\n",
       "      <td>8.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>8.4</td>\n",
       "      <td>5.6</td>\n",
       "      <td>42.8</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-39.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.4</td>\n",
       "      <td>68.8</td>\n",
       "      <td>101.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           animal_proteins_red  refined_grains  solanaceae  saturated  \\\n",
       "#SampleID                                                               \n",
       "OAD001                     0.0             0.0         0.0        0.0   \n",
       "OAD003                    14.0            18.4         5.6        0.0   \n",
       "OAD004                    -2.0            12.0         2.8        0.0   \n",
       "OAD005                     9.2            22.4        -5.2        0.0   \n",
       "OAD006                     5.6             3.6         8.4        0.0   \n",
       "\n",
       "           pastries  milk_derivates  beverages_prohibited  pro_inflammatory  \\\n",
       "#SampleID                                                                     \n",
       "OAD001          0.0             0.0                   0.0               NaN   \n",
       "OAD003          5.6            36.4                  33.6             113.6   \n",
       "OAD004          0.0             0.8                  10.0              32.0   \n",
       "OAD005          2.8             3.2                   3.2              23.6   \n",
       "OAD006          2.8             8.4                   5.6              42.8   \n",
       "\n",
       "           chicken  plant_protein  ...  corn  quinoa  broccoli  \\\n",
       "#SampleID                          ...                           \n",
       "OAD001         0.0            2.0  ...   0.0     4.0       4.0   \n",
       "OAD003         6.0            0.0  ...   8.0     0.0       4.0   \n",
       "OAD004        -2.0            0.4  ...   0.0     0.0      11.2   \n",
       "OAD005         0.0            4.0  ...   4.0     0.0       2.0   \n",
       "OAD006        -2.0            6.0  ...   0.0     0.0       0.0   \n",
       "\n",
       "           Garbanzo_beans  pork  beef  burger  Total_omega3  Total_omega6  \\\n",
       "#SampleID                                                                   \n",
       "OAD001                0.0   0.0   0.0     0.0           NaN           NaN   \n",
       "OAD003                0.0   0.0 -14.0     0.0          31.2          54.8   \n",
       "OAD004                0.0   0.0   2.0     0.0          25.2          56.0   \n",
       "OAD005                0.0 -78.4   2.0     0.0          48.8          88.8   \n",
       "OAD006                0.0 -39.2   0.0     0.0          32.4          68.8   \n",
       "\n",
       "           Total_o3_o6  \n",
       "#SampleID               \n",
       "OAD001           131.0  \n",
       "OAD003            96.0  \n",
       "OAD004            83.2  \n",
       "OAD005           147.6  \n",
       "OAD006           101.2  \n",
       "\n",
       "[5 rows x 54 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dietary variables\n",
    "diet = ['animal_proteins_red',\n",
    "       'refined_grains', 'solanaceae', 'saturated', 'pastries',\n",
    "       'milk_derivates', 'beverages_prohibited', 'pro_inflammatory',\n",
    "       'chicken', 'plant_protein', 'whole_grains_pseudocereals_oats',\n",
    "       'gluten_free', 'vegetables', 'cruciferous',\n",
    "       'berries_and_enzymatic_fruit', 'other_fruit', 'fish', 'milk_subs',\n",
    "       'PUFA', 'MUFA', 'Walnuts', 'Nuts', 'probiotics', 'Tahini_avo',\n",
    "       'green_tea', 'anti_inflam_spices', 'inflammatory_spices',\n",
    "       'inflammatory', 'anti_inflammatory', 'total_antiinflammatory',\n",
    "       'Adherece_antiinflam', 'oats', 'almond_milk', 'fatty_fish',\n",
    "       'sesame_seeds', 'egg', 'chia', 'flaxseed_oil', 'walnuts',\n",
    "       'spinach', 'corn_tortilla', 'tahini', 'avocado', 'chicken.1',\n",
    "       'brown_rice', 'corn', 'quinoa', 'broccoli', 'Garbanzo_beans',\n",
    "       'pork', 'beef', 'burger', 'Total_omega3', 'Adherence_omega3',\n",
    "       'Total_omega6', 'Adherence_omega6', 'Total_o3_o6']\n",
    "\n",
    "for d in diet:\n",
    "    if 'Adh' in d:\n",
    "        diet.remove(d)\n",
    "\n",
    "# type_to_df_diet = {}\n",
    "gt_to_df_diet = {}\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    gt_to_df_diet[g] = {}\n",
    "    for t in ['pre','post']:\n",
    "        df_map = type_to_df_map[g].copy()\n",
    "        df_map = df_map[df_map['Timepoint'] == t]\n",
    "        df_diet = df_map.loc[:,diet]\n",
    "        \n",
    "        # type_to_df_diet[g] = df_diet\n",
    "        gt_to_df_diet[g][t] = df_diet.copy()\n",
    "\n",
    "        df_diet = df_diet.dropna(axis=1)\n",
    "        \n",
    "        from sklearn.decomposition import PCA\n",
    "        npcs = 3\n",
    "        pca = PCA(n_components=3, whiten=True)\n",
    "        pca.fit(df_diet.T)\n",
    "        # Reformat and view results\n",
    "        loadings = pd.DataFrame(pca.components_,\n",
    "                                index=['PC%s' % _ for _ in range(npcs)], \n",
    "                                columns=df_diet.index).T\n",
    "        \n",
    "        df_merge = pd.concat([df_map, df_diet, loadings], axis=1)\n",
    "        plt.title(g + '_Diet_PCA')\n",
    "        sns.scatterplot(data=df_merge, x='PC1', y='PC2', hue='WOMAC_P_Response')\n",
    "        sns.despine()\n",
    "        plt.savefig(path + 'outputs/jobs06/PCA_diet_' + g + '_' + t + '.pdf')\n",
    "        plt.close() \n",
    "\n",
    "    gt_to_df_diet[g]['diff'] = gt_to_df_diet[g]['post'] - gt_to_df_diet[g]['pre']\n",
    "\n",
    "gt_to_df_diet[g]['diff'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a78534-daba-47ff-87b8-0a1eb95002d1",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Complete DA Analysis ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "856e9286-7a26-4639-a599-53c699043e5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taxa\n",
      "pvp\n",
      "saliva\n",
      "Number of sig var\n",
      "2\n",
      "Number of sig var\n",
      "5\n",
      "stool\n",
      "Number of sig var\n",
      "6\n",
      "Number of sig var\n",
      "4\n",
      "rvnr\n",
      "saliva\n",
      "Number of sig var\n",
      "9\n",
      "Number of sig var\n",
      "0\n",
      "stool\n",
      "Number of sig var\n",
      "6\n",
      "Number of sig var\n",
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "      <th>Timepoint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.022576</td>\n",
       "      <td>0.007098</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.292202</td>\n",
       "      <td>0.001994</td>\n",
       "      <td>0.013555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000783</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.181156</td>\n",
       "      <td>0.067844</td>\n",
       "      <td>0.150895</td>\n",
       "      <td>0.063229</td>\n",
       "      <td>0.051360</td>\n",
       "      <td>0.055600</td>\n",
       "      <td>0.012363</td>\n",
       "      <td>0.051513</td>\n",
       "      <td>0.006045</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.170334</td>\n",
       "      <td>0.135468</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002088</td>\n",
       "      <td>0.173617</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003649</td>\n",
       "      <td>0.008765</td>\n",
       "      <td>0.054802</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-001.pre.stool</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.064924</td>\n",
       "      <td>0.001590</td>\n",
       "      <td>0.032964</td>\n",
       "      <td>0.321408</td>\n",
       "      <td>0.010514</td>\n",
       "      <td>0.070031</td>\n",
       "      <td>0.005441</td>\n",
       "      <td>0.011820</td>\n",
       "      <td>0.020509</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-012.pre.stool</th>\n",
       "      <td>0.126033</td>\n",
       "      <td>0.139754</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015668</td>\n",
       "      <td>0.149951</td>\n",
       "      <td>0.015739</td>\n",
       "      <td>0.011277</td>\n",
       "      <td>0.000460</td>\n",
       "      <td>0.000372</td>\n",
       "      <td>0.005010</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004  \\\n",
       "OAD-001.pre.stool                                             0.022576                                                       \n",
       "OAD-003.pre.stool                                             0.181156                                                       \n",
       "OAD-004.pre.stool                                             0.170334                                                       \n",
       "VAOAD-001.pre.stool                                           0.000000                                                       \n",
       "VAOAD-012.pre.stool                                           0.126033                                                       \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium  \\\n",
       "OAD-001.pre.stool                                             0.007098                                                            \n",
       "OAD-003.pre.stool                                             0.067844                                                            \n",
       "OAD-004.pre.stool                                             0.135468                                                            \n",
       "VAOAD-001.pre.stool                                           0.064924                                                            \n",
       "VAOAD-012.pre.stool                                           0.139754                                                            \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella  \\\n",
       "OAD-001.pre.stool                                             0.005128                                             \n",
       "OAD-003.pre.stool                                             0.150895                                             \n",
       "OAD-004.pre.stool                                             0.000000                                             \n",
       "VAOAD-001.pre.stool                                           0.001590                                             \n",
       "VAOAD-012.pre.stool                                           0.000000                                             \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117  \\\n",
       "OAD-001.pre.stool                                             0.000000                                                              \n",
       "OAD-003.pre.stool                                             0.063229                                                              \n",
       "OAD-004.pre.stool                                             0.002088                                                              \n",
       "VAOAD-001.pre.stool                                           0.032964                                                              \n",
       "VAOAD-012.pre.stool                                           0.015668                                                              \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H  \\\n",
       "OAD-001.pre.stool                                             0.292202                                                \n",
       "OAD-003.pre.stool                                             0.051360                                                \n",
       "OAD-004.pre.stool                                             0.173617                                                \n",
       "VAOAD-001.pre.stool                                           0.321408                                                \n",
       "VAOAD-012.pre.stool                                           0.149951                                                \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum  \\\n",
       "OAD-001.pre.stool                                             0.001994                                                  \n",
       "OAD-003.pre.stool                                             0.055600                                                  \n",
       "OAD-004.pre.stool                                             0.000143                                                  \n",
       "VAOAD-001.pre.stool                                           0.010514                                                  \n",
       "VAOAD-012.pre.stool                                           0.015739                                                  \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129  \\\n",
       "OAD-001.pre.stool                                             0.013555                                                            \n",
       "OAD-003.pre.stool                                             0.012363                                                            \n",
       "OAD-004.pre.stool                                             0.000000                                                            \n",
       "VAOAD-001.pre.stool                                           0.070031                                                            \n",
       "VAOAD-012.pre.stool                                           0.011277                                                            \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus  \\\n",
       "OAD-001.pre.stool                                             0.000000                                                \n",
       "OAD-003.pre.stool                                             0.051513                                                \n",
       "OAD-004.pre.stool                                             0.003649                                                \n",
       "VAOAD-001.pre.stool                                           0.005441                                                \n",
       "VAOAD-012.pre.stool                                           0.000460                                                \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor  \\\n",
       "OAD-001.pre.stool                                             0.000783                                                        \n",
       "OAD-003.pre.stool                                             0.006045                                                        \n",
       "OAD-004.pre.stool                                             0.008765                                                        \n",
       "VAOAD-001.pre.stool                                           0.011820                                                        \n",
       "VAOAD-012.pre.stool                                           0.000372                                                        \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400  \\\n",
       "OAD-001.pre.stool                                             0.026849                                                    \n",
       "OAD-003.pre.stool                                             0.025918                                                    \n",
       "OAD-004.pre.stool                                             0.054802                                                    \n",
       "VAOAD-001.pre.stool                                           0.020509                                                    \n",
       "VAOAD-012.pre.stool                                           0.005010                                                    \n",
       "\n",
       "                     ...  \\\n",
       "OAD-001.pre.stool    ...   \n",
       "OAD-003.pre.stool    ...   \n",
       "OAD-004.pre.stool    ...   \n",
       "VAOAD-001.pre.stool  ...   \n",
       "VAOAD-012.pre.stool  ...   \n",
       "\n",
       "                     d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "OAD-001.pre.stool                                                  0.0                                                        \n",
       "OAD-003.pre.stool                                                  0.0                                                        \n",
       "OAD-004.pre.stool                                                  0.0                                                        \n",
       "VAOAD-001.pre.stool                                                0.0                                                        \n",
       "VAOAD-012.pre.stool                                                0.0                                                        \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "OAD-001.pre.stool                                                  0.0                 \n",
       "OAD-003.pre.stool                                                  0.0                 \n",
       "OAD-004.pre.stool                                                  0.0                 \n",
       "VAOAD-001.pre.stool                                                0.0                 \n",
       "VAOAD-012.pre.stool                                                0.0                 \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "OAD-001.pre.stool                                                  0.0                                                         \n",
       "OAD-003.pre.stool                                                  0.0                                                         \n",
       "OAD-004.pre.stool                                                  0.0                                                         \n",
       "VAOAD-001.pre.stool                                                0.0                                                         \n",
       "VAOAD-012.pre.stool                                                0.0                                                         \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "OAD-001.pre.stool                                                  0.0                                               \n",
       "OAD-003.pre.stool                                                  0.0                                               \n",
       "OAD-004.pre.stool                                                  0.0                                               \n",
       "VAOAD-001.pre.stool                                                0.0                                               \n",
       "VAOAD-012.pre.stool                                                0.0                                               \n",
       "\n",
       "                     d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "OAD-001.pre.stool                                                  0.0                                                                         \n",
       "OAD-003.pre.stool                                                  0.0                                                                         \n",
       "OAD-004.pre.stool                                                  0.0                                                                         \n",
       "VAOAD-001.pre.stool                                                0.0                                                                         \n",
       "VAOAD-012.pre.stool                                                0.0                                                                         \n",
       "\n",
       "                     d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "OAD-001.pre.stool                                                  0.0                                     \n",
       "OAD-003.pre.stool                                                  0.0                                     \n",
       "OAD-004.pre.stool                                                  0.0                                     \n",
       "VAOAD-001.pre.stool                                                0.0                                     \n",
       "VAOAD-012.pre.stool                                                0.0                                     \n",
       "\n",
       "                     d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "OAD-001.pre.stool                                                  0.0                                                                        \n",
       "OAD-003.pre.stool                                                  0.0                                                                        \n",
       "OAD-004.pre.stool                                                  0.0                                                                        \n",
       "VAOAD-001.pre.stool                                                0.0                                                                        \n",
       "VAOAD-012.pre.stool                                                0.0                                                                        \n",
       "\n",
       "                     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "OAD-001.pre.stool                                                  0.0                                  \n",
       "OAD-003.pre.stool                                                  0.0                                  \n",
       "OAD-004.pre.stool                                                  0.0                                  \n",
       "VAOAD-001.pre.stool                                                0.0                                  \n",
       "VAOAD-012.pre.stool                                                0.0                                  \n",
       "\n",
       "                     d__Bacteria;p__Actinobacteriota;__;__;__;__  Timepoint  \n",
       "OAD-001.pre.stool                                       0.000000        pre  \n",
       "OAD-003.pre.stool                                       0.000000        pre  \n",
       "OAD-004.pre.stool                                       0.000000        pre  \n",
       "VAOAD-001.pre.stool                                     0.000000        pre  \n",
       "VAOAD-012.pre.stool                                     0.000035        pre  \n",
       "\n",
       "[5 rows x 375 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# complete block for paired testing in r v nr and non paired testing pre and post\n",
    "import statsmodels\n",
    "alpha = 0.05 # not 0.05 anymore\n",
    "\n",
    "dcgr_to_sig = {}\n",
    "\n",
    "for dt in ['taxa']:#, 'path']:\n",
    "    print(dt)\n",
    "    dcgr_to_sig[dt] = {}\n",
    "    for c in ['pvp', 'rvnr']: # pvp is pre AND post, comparing rvnr at pre and post time points\n",
    "        print(c)\n",
    "        dcgr_to_sig[dt][c] = {}        \n",
    "    \n",
    "        # define df arrays\n",
    "        xs = []\n",
    "        gs = []\n",
    "        ws = []\n",
    "        ps = []\n",
    "        os = []\n",
    "        dirs = []\n",
    "        adj_rs = []\n",
    "        \n",
    "        # define analysis specific params\n",
    "        if c == 'pvp':\n",
    "            outer_loop = ['pre', 'post']\n",
    "            comp = ['Response', 'No response']\n",
    "            var = 'WOMAC_P_Response'\n",
    "            drop = 'Timepoint'\n",
    "        elif c == 'rvnr': # pvp is pre and post, comparing rvnr at pre and posttime points\n",
    "            outer_loop = ['Response', 'No response']\n",
    "            comp = ['pre', 'post']\n",
    "            var = 'Timepoint'\n",
    "            drop = 'WOMAC_P_Response'\n",
    "\n",
    "        # for each sample type\n",
    "        for g in ['saliva','stool']:\n",
    "            print(g)\n",
    "            dcgr_to_sig[dt][c][g] = {}\n",
    "            \n",
    "            # grab relevant table table\n",
    "            if dt == 'taxa':\n",
    "                df_abun = gt_to_otu[g]['all']\n",
    "            elif dt == 'path':\n",
    "                df_abun = gt_to_path[g]['all']\n",
    "\n",
    "            df_map = type_to_df_map[g].copy()\n",
    "            df_map = df_map.set_index('Together')\n",
    "            if g == 'stool':\n",
    "                df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "            df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "\n",
    "            for r in outer_loop:\n",
    "                df_merge = pd.concat([df_abun, df_map[df_map[drop] == r]], axis=1)\n",
    "                df_merge = df_merge.dropna()\n",
    "                df_merge['SampleID'] = df_merge.index.map(lambda x: x.split('.p')[0])\n",
    "                if c == 'rvnr':\n",
    "                    df_vc = df_merge['SampleID'].value_counts()\n",
    "                    keep = df_vc[df_vc == 2].index.values\n",
    "                    df_merge = df_merge[df_merge['SampleID'].isin(keep)] \n",
    "                df_merge = df_merge.drop(['SampleID',drop],axis=1)\n",
    "                sig = []\n",
    "                for x in list(df_merge.columns.values)[:-1]:\n",
    "                    df_a = df_merge[df_merge[var] == comp[0]][x]\n",
    "                    df_b = df_merge[df_merge[var] == comp[1]][x]\n",
    "\n",
    "                    try:\n",
    "                        if c == 'rvnr':\n",
    "                            w,p = scipy.stats.wilcoxon(df_a.values, df_b.values, nan_policy='omit')\n",
    "                        elif c == 'pvp':\n",
    "                            w,p = scipy.stats.mannwhitneyu(df_a.values, df_b.values, nan_policy='omit')\n",
    "\n",
    "                        a = df_a.values\n",
    "                        b = df_b.values\n",
    "                        # if median in A is greater\n",
    "                        abr, abp = scipy.stats.mannwhitneyu(a,b,nan_policy='omit')\n",
    "                        # if median in B is greater\n",
    "                        bar, bap = scipy.stats.mannwhitneyu(b,a,nan_policy='omit')\n",
    "                        if abr > bar:\n",
    "                            dirs.append(comp[0])\n",
    "                            adj_rs.append(abr)\n",
    "                        elif abr < bar:\n",
    "                            dirs.append(comp[1])\n",
    "                            adj_rs.append(bar * -1)\n",
    "                        else:\n",
    "                            dirs.append('same')\n",
    "                            adj_rs.append(abr)\n",
    "\n",
    "                    except:\n",
    "                        w,p = 0, 1\n",
    "                        dirs.append(np.nan)\n",
    "                        adj_rs.append(np.nan)\n",
    "\n",
    "                    os.append(r)\n",
    "                    xs.append(x)\n",
    "                    gs.append(g)\n",
    "                    ws.append(w)\n",
    "                    ps.append(p)\n",
    "\n",
    "                    if p < alpha:\n",
    "                        ax = sns.boxplot(data=df_merge, y=x, x=var)\n",
    "                        sns.swarmplot(data=df_merge, y=x, x=var, palette='dark:grey')\n",
    "                        sns.despine()\n",
    "                    \n",
    "                        plt.tight_layout()\n",
    "                        plt.savefig(path + 'outputs/jobs06/' + c + '_taxa/' + c + '_' + dt + '_barplot_' + g + '_' + r + '_' + x.split('g__')[-1] + '.pdf')\n",
    "                        plt.close() \n",
    "                        sig.append(x)\n",
    "                print('Number of sig var')\n",
    "                print(len(sig))\n",
    "                dcgr_to_sig[dt][c][g][r] = sig\n",
    "                \n",
    "        df_res = pd.DataFrame({'SampleType':gs, # saliva v stool\n",
    "                      'fixed': os,\n",
    "                      'var': xs, # example pathway or taxa\n",
    "                      'direction': dirs, # direction of enrichment\n",
    "                      'Statistic': ws, \n",
    "                      'pvalue': ps,\n",
    "                      'adjusted': adj_rs                      \n",
    "                     })\n",
    "\n",
    "        # get fdr\n",
    "        # sort\n",
    "        df_res = df_res.sort_values(by='pvalue',ascending=True)\n",
    "        reject, pvals_corrected, _, _ = statsmodels.stats.multitest.multipletests(df_res.pvalue, method='fdr_bh')\n",
    "        df_res['fdr'] = pvals_corrected\n",
    "        \n",
    "        df_res[df_res['pvalue'] < alpha].to_csv(path + 'outputs/jobs06/' + dt + '_' + c + '_' + str(alpha) + '.tsv', sep='\\t')\n",
    "        df_res[df_res['fdr'] < alpha].to_csv(path + 'outputs/jobs06/' + dt + '_' + c + '_' + str(alpha) + '_fdr.tsv', sep='\\t')\n",
    "\n",
    "df_merge.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "43bca09e-c560-43cf-975b-7b58541ae974",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "meta\n",
      "otu\n",
      "plasma\n",
      "meta\n",
      "saliva\n",
      "meta\n",
      "otu\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>logp</th>\n",
       "      <th>adj_test_stat</th>\n",
       "      <th>pvals</th>\n",
       "      <th>direction</th>\n",
       "      <th>fdr</th>\n",
       "      <th>hue</th>\n",
       "      <th>Enrichment_Significance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>direction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">post</th>\n",
       "      <th>51</th>\n",
       "      <td>Porphyromonadaceae Porphyromonas_A_859424</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.963333</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.108809</td>\n",
       "      <td>post</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NS</td>\n",
       "      <td>Post_NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Lachnospiraceae Lachnoanaerobaculum</td>\n",
       "      <td>16.5</td>\n",
       "      <td>0.841255</td>\n",
       "      <td>16.5</td>\n",
       "      <td>0.144127</td>\n",
       "      <td>post</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NS</td>\n",
       "      <td>Post_NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Enterobacterales_A_737866 Enterobacteriaceae_A</td>\n",
       "      <td>17.5</td>\n",
       "      <td>0.745422</td>\n",
       "      <td>17.5</td>\n",
       "      <td>0.179712</td>\n",
       "      <td>post</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NS</td>\n",
       "      <td>Post_NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>Bacteroidaceae Phocaeicola_A_858004</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.745422</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.179712</td>\n",
       "      <td>post</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NS</td>\n",
       "      <td>Post_NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Amphibacillaceae Ornithinibacillus_287146</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.745422</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.179712</td>\n",
       "      <td>post</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NS</td>\n",
       "      <td>Post_NS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      feature  test_stat  \\\n",
       "direction                                                                  \n",
       "post      51        Porphyromonadaceae Porphyromonas_A_859424       16.0   \n",
       "          44              Lachnospiraceae Lachnoanaerobaculum       16.5   \n",
       "          87   Enterobacterales_A_737866 Enterobacteriaceae_A       17.5   \n",
       "          239             Bacteroidaceae Phocaeicola_A_858004       16.0   \n",
       "          53        Amphibacillaceae Ornithinibacillus_287146       16.0   \n",
       "\n",
       "                   logp  adj_test_stat     pvals direction  fdr hue  \\\n",
       "direction                                                             \n",
       "post      51   0.963333           16.0  0.108809      post  1.0  NS   \n",
       "          44   0.841255           16.5  0.144127      post  1.0  NS   \n",
       "          87   0.745422           17.5  0.179712      post  1.0  NS   \n",
       "          239  0.745422           16.0  0.179712      post  1.0  NS   \n",
       "          53   0.745422           16.0  0.179712      post  1.0  NS   \n",
       "\n",
       "              Enrichment_Significance  \n",
       "direction                              \n",
       "post      51                  Post_NS  \n",
       "          44                  Post_NS  \n",
       "          87                  Post_NS  \n",
       "          239                 Post_NS  \n",
       "          53                  Post_NS  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# horizontal barplot for features in R, NR and both for enrichment in pre vs post\n",
    "alpha =  0.05\n",
    "\n",
    "d_to_df = {\n",
    "    'meta': gt_to_meta,\n",
    "    'otu': gt_to_otu,\n",
    "    # 'path': gt_to_path\n",
    "}\n",
    "d_to_label = {\n",
    "    'meta': 'Metabolite',\n",
    "    'otu': 'Taxa',\n",
    "    # 'path': 'Pathway'\n",
    "}\n",
    "t = 'all'\n",
    "\n",
    "gdt_to_res_pvp = {}\n",
    "for g in ['stool','plasma','saliva']:#]:\n",
    "    print(g)\n",
    "    gdt_to_res_pvp[g] = {}\n",
    "    if g == 'stool' or g == 'saliva':\n",
    "        ds = ['meta','otu']\n",
    "    elif g == 'plasma':\n",
    "        ds = ['meta']\n",
    "        \n",
    "    for d in ds:\n",
    "        print(d)\n",
    "        gdt_to_res_pvp[g][d] = {}\n",
    "        # grab original dfs\n",
    "        df_feat = d_to_df[d][g][t]\n",
    "\n",
    "        df_map = type_to_df_map[g].set_index('Together')\n",
    "        df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "        \n",
    "        df_map = df_map[['Timepoint','WOMAC_P_Response','HostSubjectId']]\n",
    "        \n",
    "        # merge df\n",
    "        df_merge = pd.concat([df_feat, df_map],axis=1)\n",
    "        df_merge = df_merge.dropna()\n",
    "        \n",
    "        # do paired test\n",
    "        # grab all sample IDs that are duplicated\n",
    "        df_vc = df_merge['HostSubjectId'].value_counts() == 2\n",
    "        df_vc = df_vc[df_vc == True]\n",
    "        keep_id = df_vc.index.values\n",
    "        \n",
    "        df_merge_complete = df_merge[df_merge['HostSubjectId'].isin(keep_id)]\n",
    "\n",
    "        for keep in [['Both','Response','No response'],['Response'],['No response']]: # adding 'Both' is a hack to make the dict hashing work\n",
    "            df_merge = df_merge_complete[df_merge_complete['WOMAC_P_Response'].isin(keep)]\n",
    "        \n",
    "            # samples are in order when subsetted on Timepoint\n",
    "            df_pre = df_merge[df_merge['Timepoint'] == 'pre'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "            df_post = df_merge[df_merge['Timepoint'] == 'post'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "            \n",
    "            # compute rank sum\n",
    "            feat = []\n",
    "            rs = []\n",
    "            adj_rs = []\n",
    "            ps = []\n",
    "            logps = []\n",
    "            dir = []\n",
    "            for x in df_pre.columns.values:\n",
    "                try:\n",
    "                    r, p = scipy.stats.wilcoxon(df_post[x].values-df_pre[x].values,nan_policy='omit')\n",
    "                except ValueError:\n",
    "                    r, p = 0, 1\n",
    "                if d == 'otu':\n",
    "                    # feat.append(x.split('g__')[-1])\n",
    "                    feat.append(read_taxa(x, ';'))\n",
    "                elif d == 'path':\n",
    "                    feat.append(x.split('|')[-1])\n",
    "                else:\n",
    "                    feat.append(x)\n",
    "                # rs.append(r)\n",
    "                ps.append(p)\n",
    "                # determine direction via MWU\n",
    "                a = df_post[x].values\n",
    "                b = df_pre[x].values\n",
    "                abr, abp = scipy.stats.mannwhitneyu(a,b,nan_policy='omit')\n",
    "                bar, bap = scipy.stats.mannwhitneyu(b,a,nan_policy='omit')\n",
    "\n",
    "                # log mean diff\n",
    "                # adj = np.log(np.abs(np.mean(a-b)))\n",
    "                # log ratio\n",
    "                # adj = np.log(np.mean(a/b))\n",
    "                # print(abr, bar, adj)\n",
    "                if abr > bar:\n",
    "                    dir.append('post')\n",
    "                    adj_rs.append(abr)\n",
    "                    # adj_rs.append(adj)\n",
    "                    rs.append(abr)\n",
    "                    logps.append(-np.log10(p))\n",
    "                elif abr < bar:\n",
    "                    dir.append('pre')\n",
    "                    adj_rs.append(bar * -1)\n",
    "                    # adj_rs.append(adj)# * -1)\n",
    "                    rs.append(bar)\n",
    "                    logps.append(np.log10(p))\n",
    "                else:\n",
    "                    dir.append('same')\n",
    "                    adj_rs.append(0)\n",
    "                    rs.append(abr)\n",
    "                    logps.append(np.nan)\n",
    "                # adj_rs as the difference post - pre\n",
    "                # adj_rs.append(np.log(np.mean(a) / np.mean(b)))\n",
    "                # adj_rs.append(np.mean(a-b))\n",
    "                        \n",
    "            # compile results df\n",
    "            df_results = pd.DataFrame({'feature': feat, 'test_stat': rs, 'logp': logps, 'adj_test_stat': adj_rs,'pvals': ps, 'direction': dir})\n",
    "\n",
    "            # fdr\n",
    "            df_results = df_results.sort_values(by='pvals',ascending=True)\n",
    "            reject, pvals_corrected, _, _ = statsmodels.stats.multitest.multipletests(df_results.pvals, method='fdr_bh')\n",
    "            df_results['fdr'] = pvals_corrected\n",
    "\n",
    "            # create new hue\n",
    "            conditions = [ (df_results['fdr'] < alpha) & (df_results['pvals'] < alpha), \n",
    "                           (df_results['fdr'] > alpha) & (df_results['pvals'] < alpha),\n",
    "                           (df_results['pvals'] > alpha)]\n",
    "            choices     = [ \"FDR\", 'NonFDR', 'NS' ]\n",
    "                \n",
    "            df_results[\"hue\"] = np.select(conditions, choices, default=np.nan)\n",
    "            df_results[\"Enrichment_Significance\"] = df_results['direction'] + '_' + df_results['hue']\n",
    "            df_results[\"Enrichment_Significance\"] = df_results['direction'].map(lambda x: x.capitalize()) + '_' + df_results['hue']\n",
    "            df_results[\"Enrichment_Significance\"] = df_results[\"Enrichment_Significance\"].map(lambda x: x.replace('Resp','R'))\n",
    "            df_results[\"Enrichment_Significance\"] = df_results[\"Enrichment_Significance\"].map(lambda x: x.replace('Nonresp','NR'))\n",
    "        \n",
    "            # df = df_results[df_results['fdr'] < alpha]\n",
    "            gdt_to_res_pvp[g][d][keep[0]] = df_results\n",
    "                        \n",
    "\n",
    "            # grab top 10 results in both directions \n",
    "            df_results = df_results.groupby('direction').apply(lambda x: x.nlargest(10, 'test_stat'))\n",
    "\n",
    "            df = df_results.dropna()#opy()#[df_results['pvals'] < alpha]\n",
    "\n",
    "            df = df.sort_values(by=['logp'],ascending=False)\n",
    "            # df = df.sort_values(by=['adj_test_stat'],ascending=False)\n",
    "            # plot\n",
    "            # create plot\n",
    "            #try:\n",
    "            if d == 'meta':\n",
    "                if alpha == 0.05:\n",
    "                    plt.figure(figsize=(6,3))\n",
    "                else:\n",
    "                    plt.figure(figsize=(6,6))\n",
    "            elif d == 'otu':\n",
    "                if alpha == 0.05:\n",
    "                    plt.figure(figsize=(6,3))\n",
    "                else:\n",
    "                    plt.figure(figsize=(8,8))\n",
    "            sns.set_theme(style=\"whitegrid\")\n",
    "            \n",
    "            sns.set_color_codes(\"pastel\")\n",
    "            sns.barplot(x='logp', y='feature', data=df,\n",
    "                        hue=\"Enrichment_Significance\",orient='horizontal')#,native_scale=True)#width=1)\n",
    "            \n",
    "            ax.set_ylabel(d_to_label[d],fontsize=16)\n",
    "            # ax.set_xlabel(\"Effect Size / Adjusted Test Statistic\",fontsize=16)\n",
    "            ax.set_xlabel(\"-log10(p)\",fontsize=16)\n",
    "            ax.tick_params(labelsize=12)\n",
    "            sns.despine(left=True, bottom=True)\n",
    "            plt.tight_layout()\n",
    "            plt.title(g + '_' + d + '_' + str(keep))\n",
    "            plt.savefig(path + 'outputs/jobs18/DA_hbarplot_' + g + '_' + d + '_' + str(keep) + '_' + str(alpha) + '.pdf')\n",
    "            plt.close()\n",
    "            df.to_csv(path + 'outputs/jobs18/' + g + '_' + d + '_' + str(keep) + '_' + str(alpha) + '.tsv', sep='\\t')\n",
    "\n",
    "            # save for R plotting\n",
    "            # d = 'otu'\n",
    "            # g = 'saliva'\n",
    "            # keep = ['Response']\n",
    "            if len(keep) == 3:\n",
    "                suffix = ''\n",
    "            else:\n",
    "                suffix = '_' + keep[0].lower().replace(' ','')\n",
    "            \n",
    "            if d == 'otu':\n",
    "                prefix = 'Qiime2_'\n",
    "            elif d == 'meta':\n",
    "                prefix = 'Meta_'\n",
    "            \n",
    "            df.to_csv(path + 'outputs/' + prefix + g + '_adh' + suffix + '/barplot_data.tsv', sep='\\t')\n",
    "\n",
    "            #except: \n",
    "            #    print('no plot for :' + g + '_' + d + '_' + str(keep))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a0bc8415-1b9b-48ea-8971-4468821e41ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "meta\n",
      "otu\n",
      "plasma\n",
      "meta\n",
      "saliva\n",
      "meta\n",
      "otu\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>logp</th>\n",
       "      <th>adj_test_stat</th>\n",
       "      <th>pvals</th>\n",
       "      <th>direction</th>\n",
       "      <th>fdr</th>\n",
       "      <th>hue</th>\n",
       "      <th>Enrichment_Significance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>direction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">resp</th>\n",
       "      <th>273</th>\n",
       "      <td>Pseudomonadaceae Pseudomonas_F</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.867547</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.135660</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.723611</td>\n",
       "      <td>NS</td>\n",
       "      <td>R_NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Bacteroidia Bacteroidales</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.867547</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.135660</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.723611</td>\n",
       "      <td>NS</td>\n",
       "      <td>R_NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Pseudonocardiaceae Pseudonocardia</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.867547</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.135660</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.723611</td>\n",
       "      <td>NS</td>\n",
       "      <td>R_NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>Bacilli Lactobacillales</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.867547</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.135660</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.723611</td>\n",
       "      <td>NS</td>\n",
       "      <td>R_NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>CSP1-4 GWC2-73-18</td>\n",
       "      <td>22.5</td>\n",
       "      <td>0.583012</td>\n",
       "      <td>22.5</td>\n",
       "      <td>0.261209</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.723611</td>\n",
       "      <td>NS</td>\n",
       "      <td>R_NS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         feature  test_stat      logp  \\\n",
       "direction                                                               \n",
       "resp      273     Pseudomonadaceae Pseudomonas_F       25.0  0.867547   \n",
       "          32           Bacteroidia Bacteroidales       25.0  0.867547   \n",
       "          95   Pseudonocardiaceae Pseudonocardia       25.0  0.867547   \n",
       "          361            Bacilli Lactobacillales       25.0  0.867547   \n",
       "          82                   CSP1-4 GWC2-73-18       22.5  0.583012   \n",
       "\n",
       "               adj_test_stat     pvals direction       fdr hue  \\\n",
       "direction                                                        \n",
       "resp      273           25.0  0.135660      resp  0.723611  NS   \n",
       "          32            25.0  0.135660      resp  0.723611  NS   \n",
       "          95            25.0  0.135660      resp  0.723611  NS   \n",
       "          361           25.0  0.135660      resp  0.723611  NS   \n",
       "          82            22.5  0.261209      resp  0.723611  NS   \n",
       "\n",
       "              Enrichment_Significance  \n",
       "direction                              \n",
       "resp      273                    R_NS  \n",
       "          32                     R_NS  \n",
       "          95                     R_NS  \n",
       "          361                    R_NS  \n",
       "          82                     R_NS  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# horizontal barplot for features in pre and post for enrichment in R vs NR \n",
    "alpha = 0.05\n",
    "\n",
    "d_to_df = {\n",
    "    'meta': gt_to_meta,\n",
    "    'otu': gt_to_otu,\n",
    "    # 'path': gt_to_path\n",
    "}\n",
    "d_to_label = {\n",
    "    'meta': 'Metabolite',\n",
    "    'otu': 'Taxa',\n",
    "    # 'path': 'Pathway'\n",
    "}\n",
    "t = 'all'\n",
    "\n",
    "\n",
    "gdt_to_res_rvnr = {}\n",
    "for g in ['stool','plasma','saliva']:\n",
    "    print(g)\n",
    "    gdt_to_res_rvnr[g] = {}\n",
    "    if g == 'stool' or g == 'saliva':\n",
    "        ds = ['meta','otu']\n",
    "    elif g == 'plasma':\n",
    "        ds = ['meta']\n",
    "        \n",
    "    for d in ds: #]:#,'path']:\n",
    "        print(d)\n",
    "        # grab original dfs\n",
    "        df_feat = d_to_df[d][g][t]\n",
    "        gdt_to_res_rvnr[g][d] = {}\n",
    "\n",
    "        df_map = type_to_df_map[g].set_index('Together')\n",
    "        df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "        \n",
    "        df_map = df_map[['Timepoint','WOMAC_P_Response','HostSubjectId']]\n",
    "        \n",
    "        # merge df\n",
    "        df_merge = pd.concat([df_feat, df_map],axis=1)\n",
    "        df_merge = df_merge.dropna()\n",
    "        \n",
    "        # do paired test\n",
    "        # grab all sample IDs that are duplicated\n",
    "        df_vc = df_merge['HostSubjectId'].value_counts() == 2\n",
    "        df_vc = df_vc[df_vc == True]\n",
    "        keep_id = df_vc.index.values\n",
    "        \n",
    "        df_merge_complete = df_merge[df_merge['HostSubjectId'].isin(keep_id)]\n",
    "\n",
    "        for keep in [['pre'],['post']]:\n",
    "            df_merge = df_merge_complete[df_merge_complete['Timepoint'].isin(keep)]\n",
    "        \n",
    "            # samples are in order when subsetted on Timepoint\n",
    "            df_r = df_merge[df_merge['WOMAC_P_Response'] == 'Response'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "            df_nr = df_merge[df_merge['WOMAC_P_Response'] == 'No response'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "            \n",
    "            # compute rank sum\n",
    "            feat = []\n",
    "            rs = []\n",
    "            adj_rs = []\n",
    "            ps = []\n",
    "            dir = []\n",
    "            logps = []\n",
    "\n",
    "            for x in df_r.columns.values:\n",
    "                try:\n",
    "                    r, p = scipy.stats.mannwhitneyu(df_r[x].values, df_nr[x].values,nan_policy='omit')\n",
    "                except ValueError:\n",
    "                    r, p = 0, 1\n",
    "                if d == 'otu':\n",
    "                    # feat.append(x.split('g__')[-1])\n",
    "                    feat.append(read_taxa(x, ';'))\n",
    "                elif d == 'path':\n",
    "                    feat.append(x.split('|')[-1])\n",
    "                else:\n",
    "                    feat.append(x)\n",
    "                # rs.append(r)\n",
    "                ps.append(p)\n",
    "                # determine direction via MWU\n",
    "                a = df_r[x].values\n",
    "                b = df_nr[x].values\n",
    "                abr, abp = scipy.stats.mannwhitneyu(a,b,nan_policy='omit')\n",
    "                bar, bap = scipy.stats.mannwhitneyu(b,a,nan_policy='omit')\n",
    "\n",
    "                # log mean diff\n",
    "                # adj = np.log(np.abs(np.mean(a)-np.mean(b)))\n",
    "                # print(abr, bar, adj)\n",
    "                # log ratio\n",
    "                # adj = np.log(np.mean(a)/np.mean(b))\n",
    "\n",
    "                if abr > bar:\n",
    "                    dir.append('resp')\n",
    "                    adj_rs.append(abr)\n",
    "                    # adj_rs.append(adj)\n",
    "                    rs.append(abr)\n",
    "                    logps.append(-np.log10(p))\n",
    "                elif abr < bar:\n",
    "                    dir.append('nonresp')\n",
    "                    adj_rs.append(bar * -1)\n",
    "                    # adj_rs.append(adj)# * -1)\n",
    "                    rs.append(bar)\n",
    "                    logps.append(np.log10(p))\n",
    "                else:\n",
    "                    dir.append('same')\n",
    "                    adj_rs.append(0)\n",
    "                    rs.append(abr)\n",
    "                    logps.append(np.nan)\n",
    "                # adj_rs as the difference post - pre\n",
    "                # adj_rs.append(np.log(np.mean(a) / np.mean(b)))\n",
    "                # adj_rs.append(np.mean(a-b))\n",
    "                        \n",
    "            # compile results df\n",
    "            df_results = pd.DataFrame({'feature': feat, 'test_stat': rs,'logp':logps, 'adj_test_stat': adj_rs,'pvals': ps, 'direction': dir})\n",
    "\n",
    "            # fdr\n",
    "            df_results = df_results.sort_values(by='pvals',ascending=True)\n",
    "            reject, pvals_corrected, _, _ = statsmodels.stats.multitest.multipletests(df_results.pvals, method='fdr_bh')\n",
    "            df_results['fdr'] = pvals_corrected\n",
    "\n",
    "            # create new hue\n",
    "            conditions = [ (df_results['fdr'] < alpha) & (df_results['pvals'] < alpha), \n",
    "                           (df_results['fdr'] > alpha) & (df_results['pvals'] < alpha),\n",
    "                           (df_results['pvals'] > alpha)]\n",
    "            choices     = [ \"FDR\", 'NonFDR', 'NS' ]\n",
    "                \n",
    "            df_results[\"hue\"] = np.select(conditions, choices, default=np.nan)\n",
    "            df_results[\"Enrichment_Significance\"] = df_results['direction'].map(lambda x: x.capitalize()) + '_' + df_results['hue']\n",
    "            df_results[\"Enrichment_Significance\"] = df_results[\"Enrichment_Significance\"].map(lambda x: x.replace('Resp','R'))\n",
    "            df_results[\"Enrichment_Significance\"] = df_results[\"Enrichment_Significance\"].map(lambda x: x.replace('Nonresp','NR'))\n",
    "            # df = df_results[df_results['fdr'] < alpha]\n",
    "            \n",
    "            \n",
    "            # filter on empty\n",
    "            # df_results = df_results[df_results[d] != '']]\n",
    "            \n",
    "            gdt_to_res_rvnr[g][d][keep[0]] = df_results\n",
    "\n",
    "            # grab top 10 results in both directions \n",
    "            df_results = df_results.groupby('direction').apply(lambda x: x.nlargest(10, 'test_stat'))\n",
    "                        \n",
    "            df = df_results.dropna()#opy()#[df_results['pvals'] < alpha]\n",
    "\n",
    "            # df = df_results[df_results['pvals'] < alpha]\n",
    "            # df = df.sort_values(by=['adj_test_stat'],ascending=False)\n",
    "            df = df.sort_values(by=['logp'],ascending=False)\n",
    "            # plot\n",
    "            # create plot\n",
    "            try:\n",
    "                if d == 'meta':\n",
    "                    if alpha == 0.05:\n",
    "                        plt.figure(figsize=(6,3))\n",
    "                    else:\n",
    "                        plt.figure(figsize=(6,6))\n",
    "                elif d == 'otu':\n",
    "                    if alpha == 0.05:\n",
    "                        plt.figure(figsize=(6,3))\n",
    "                    else:\n",
    "                        plt.figure(figsize=(8,8))\n",
    "                sns.set_theme(style=\"whitegrid\")\n",
    "                \n",
    "                sns.set_color_codes(\"pastel\")\n",
    "                sns.barplot(x='logp', y='feature', data=df,\n",
    "                            hue=\"Enrichment_Significance\",orient='horizontal')#,native_scale=True)#width=1)\n",
    "                \n",
    "                ax.set_ylabel(d_to_label[d],fontsize=16)\n",
    "                # ax.set_xlabel(\"Effect Size / Adjusted Test Statistic\",fontsize=16)\n",
    "                ax.set_xlabel(\"-log10(p)\",fontsize=16)\n",
    "                ax.tick_params(labelsize=12)\n",
    "                sns.despine(left=True, bottom=True)\n",
    "                plt.tight_layout()\n",
    "                plt.title(g + '_' + d + '_' + str(keep))\n",
    "                plt.savefig(path + 'outputs/jobs18/DA_hbarplot_' + g + '_' + d + '_' + str(keep) + '_' + str(alpha) + '.pdf')\n",
    "                plt.close()\n",
    "                df.to_csv(path + 'outputs/jobs18/' + g + '_' + d + '_' + str(keep) + '_' + str(alpha) + '.tsv', sep='\\t')\n",
    "    \n",
    "                # save for R plotting\n",
    "                # d = 'otu'\n",
    "                # g = 'saliva'\n",
    "                # keep = ['pre']                \n",
    "                if d == 'otu':\n",
    "                    prefix = 'Qiime2_'\n",
    "                elif d == 'meta':\n",
    "                    prefix = 'Meta_'\n",
    "                \n",
    "                df.to_csv(path + 'outputs/' + prefix + g + '_adh_' + keep[0] + '/barplot_data.tsv', sep='\\t')\n",
    "\n",
    "            except: \n",
    "                print('no plot for :' + g + '_' + d + '_' + str(keep))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "dc1aeb01-7766-4f5c-99a2-cb9498cd10ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "meta\n",
      "otu\n",
      "plasma\n",
      "meta\n",
      "saliva\n",
      "meta\n",
      "otu\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>logp</th>\n",
       "      <th>adj_test_stat</th>\n",
       "      <th>pvals</th>\n",
       "      <th>direction</th>\n",
       "      <th>fdr</th>\n",
       "      <th>hue</th>\n",
       "      <th>Enrichment_Significance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>direction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">resp</th>\n",
       "      <th>36</th>\n",
       "      <td>Flavobacteriales_877923 Weeksellaceae</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1.318942</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.047980</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.683035</td>\n",
       "      <td>NonFDR</td>\n",
       "      <td>R_NonFDR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Burkholderiales_597441 Neisseriaceae_563222</td>\n",
       "      <td>29.5</td>\n",
       "      <td>1.212092</td>\n",
       "      <td>29.5</td>\n",
       "      <td>0.061363</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.683035</td>\n",
       "      <td>NS</td>\n",
       "      <td>R_NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Bacteroidia Bacteroidales</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.086133</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.082010</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.683035</td>\n",
       "      <td>NS</td>\n",
       "      <td>R_NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>Staphylococcaceae Staphylococcus</td>\n",
       "      <td>26.0</td>\n",
       "      <td>1.054846</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.088136</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.683035</td>\n",
       "      <td>NS</td>\n",
       "      <td>R_NS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>Pseudomonadaceae Pseudomonas_F</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.867547</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.135660</td>\n",
       "      <td>resp</td>\n",
       "      <td>0.683035</td>\n",
       "      <td>NS</td>\n",
       "      <td>R_NS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   feature  test_stat  \\\n",
       "direction                                                               \n",
       "resp      36         Flavobacteriales_877923 Weeksellaceae       30.0   \n",
       "          22   Burkholderiales_597441 Neisseriaceae_563222       29.5   \n",
       "          32                     Bacteroidia Bacteroidales       27.0   \n",
       "          147             Staphylococcaceae Staphylococcus       26.0   \n",
       "          273               Pseudomonadaceae Pseudomonas_F       25.0   \n",
       "\n",
       "                   logp  adj_test_stat     pvals direction       fdr     hue  \\\n",
       "direction                                                                      \n",
       "resp      36   1.318942           30.0  0.047980      resp  0.683035  NonFDR   \n",
       "          22   1.212092           29.5  0.061363      resp  0.683035      NS   \n",
       "          32   1.086133           27.0  0.082010      resp  0.683035      NS   \n",
       "          147  1.054846           26.0  0.088136      resp  0.683035      NS   \n",
       "          273  0.867547           25.0  0.135660      resp  0.683035      NS   \n",
       "\n",
       "              Enrichment_Significance  \n",
       "direction                              \n",
       "resp      36                 R_NonFDR  \n",
       "          22                     R_NS  \n",
       "          32                     R_NS  \n",
       "          147                    R_NS  \n",
       "          273                    R_NS  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# horizontal barplot for features\n",
    "# changes in responders versus changes in nonresponders\n",
    "alpha = 0.05\n",
    "\n",
    "d_to_df = {\n",
    "    'meta': gt_to_meta,\n",
    "    'otu': gt_to_otu,\n",
    "    # 'path': gt_to_path\n",
    "}\n",
    "d_to_label = {\n",
    "    'meta': 'Metabolite',\n",
    "    'otu': 'Taxa',\n",
    "    # 'path': 'Pathway'\n",
    "}\n",
    "t = 'diff'\n",
    "\n",
    "\n",
    "gdt_to_res_diff= {}\n",
    "for g in ['stool','plasma','saliva']:\n",
    "    print(g)\n",
    "    gdt_to_res_diff[g] = {}\n",
    "    if g == 'stool' or g == 'saliva':\n",
    "        ds = ['meta','otu']\n",
    "    elif g == 'plasma':\n",
    "        ds = ['meta']\n",
    "        \n",
    "    for d in ds: #]:#,'path']:\n",
    "        print(d)\n",
    "        # grab original dfs\n",
    "        df_feat = d_to_df[d][g][t]\n",
    "        gdt_to_res_diff[g][d] = {}\n",
    "\n",
    "        df_map = type_to_df_map[g].copy()\n",
    "        df_map = df_map.drop_duplicates(subset='HostSubjectId')\n",
    "        df_map = df_map.set_index('HostSubjectId')\n",
    "        df_map.index = df_map.index.map(lambda x: x.replace('-',''))\n",
    "        df_map.index.name = 'Study_ID'\n",
    "        \n",
    "        \n",
    "        df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "\n",
    "        # merge df\n",
    "        df_merge = pd.concat([df_feat, df_map],axis=1)\n",
    "        df_merge = df_merge.dropna()\n",
    "        \n",
    "        # samples are in order when subsetted on Timepoint\n",
    "        df_r = df_merge[df_merge['WOMAC_P_Response'] == 'Response'].drop(['Timepoint','WOMAC_P_Response'], axis=1)\n",
    "        df_nr = df_merge[df_merge['WOMAC_P_Response'] == 'No response'].drop(['Timepoint','WOMAC_P_Response'], axis=1)\n",
    "        \n",
    "        # compute rank sum\n",
    "        feat = []\n",
    "        rs = []\n",
    "        adj_rs = []\n",
    "        ps = []\n",
    "        dir = []\n",
    "        logps = []\n",
    "\n",
    "        for x in df_r.columns.values:\n",
    "            try:\n",
    "                r, p = scipy.stats.mannwhitneyu(df_r[x].values, df_nr[x].values,nan_policy='omit')\n",
    "            except ValueError:\n",
    "                r, p = 0, 1\n",
    "            if d == 'otu':\n",
    "                # feat.append(x.split('g__')[-1])\n",
    "                feat.append(read_taxa(x, ';'))\n",
    "            elif d == 'path':\n",
    "                feat.append(x.split('|')[-1])\n",
    "            else:\n",
    "                feat.append(x)\n",
    "            # rs.append(r)\n",
    "            ps.append(p)\n",
    "            # determine direction via MWU\n",
    "            a = df_r[x].values\n",
    "            b = df_nr[x].values\n",
    "            abr, abp = scipy.stats.mannwhitneyu(a,b,nan_policy='omit')\n",
    "            bar, bap = scipy.stats.mannwhitneyu(b,a,nan_policy='omit')\n",
    "\n",
    "            # log mean diff\n",
    "            # adj = np.log(np.abs(np.mean(a)-np.mean(b)))\n",
    "            # print(abr, bar, adj)\n",
    "            # log ratio\n",
    "            # adj = np.log(np.mean(a)/np.mean(b))\n",
    "\n",
    "            if abr > bar:\n",
    "                dir.append('resp')\n",
    "                adj_rs.append(abr)\n",
    "                # adj_rs.append(adj)\n",
    "                rs.append(abr)\n",
    "                logps.append(-np.log10(p))\n",
    "            elif abr < bar:\n",
    "                dir.append('nonresp')\n",
    "                adj_rs.append(bar * -1)\n",
    "                # adj_rs.append(adj)# * -1)\n",
    "                rs.append(bar)\n",
    "                logps.append(np.log10(p))\n",
    "            else:\n",
    "                dir.append('same')\n",
    "                adj_rs.append(0)\n",
    "                rs.append(abr)\n",
    "                logps.append(np.nan)\n",
    "            # adj_rs as the difference post - pre\n",
    "            # adj_rs.append(np.log(np.mean(a) / np.mean(b)))\n",
    "            # adj_rs.append(np.mean(a-b))\n",
    "                    \n",
    "        # compile results df\n",
    "        df_results = pd.DataFrame({'feature': feat, 'test_stat': rs,'logp':logps, 'adj_test_stat': adj_rs,'pvals': ps, 'direction': dir})\n",
    "\n",
    "\n",
    "        # fdr\n",
    "        df_results = df_results.sort_values(by='pvals',ascending=True)\n",
    "        reject, pvals_corrected, _, _ = statsmodels.stats.multitest.multipletests(df_results.pvals, method='fdr_bh')\n",
    "        df_results['fdr'] = pvals_corrected\n",
    "\n",
    "        # grab top 10 results in both directions \n",
    "        df_results = df_results[df_results['direction'] != 'same']\n",
    "        df_results = df_results.groupby('direction').apply(lambda x: x.nlargest(10, 'test_stat'))\n",
    "\n",
    "        # create new hue\n",
    "        conditions = [ (df_results['fdr'] < alpha) & (df_results['pvals'] < alpha), \n",
    "                       (df_results['fdr'] > alpha) & (df_results['pvals'] < alpha),\n",
    "                       (df_results['pvals'] > alpha)]\n",
    "        choices     = [ \"FDR\", 'NonFDR', 'NS' ]\n",
    "            \n",
    "        df_results[\"hue\"] = np.select(conditions, choices, default=np.nan)\n",
    "        df_results[\"Enrichment_Significance\"] = df_results['direction'].map(lambda x: x.capitalize()) + '_' + df_results['hue']\n",
    "        df_results[\"Enrichment_Significance\"] = df_results[\"Enrichment_Significance\"].map(lambda x: x.replace('Resp','R'))\n",
    "        df_results[\"Enrichment_Significance\"] = df_results[\"Enrichment_Significance\"].map(lambda x: x.replace('Nonresp','NR'))\n",
    "\n",
    "        # filter on empty\n",
    "        # df_results = df_results[df_results[d] != '']]\n",
    "        \n",
    "        gdt_to_res_diff[g][d][keep[0]] = df_results\n",
    "        # df = df_results[df_results['pvals'] < alpha]\n",
    "        # df = df.sort_values(by=['adj_test_stat'],ascending=False)\n",
    "        df = df_results.sort_values(by=['logp'],ascending=False)\n",
    "        # plot\n",
    "        # create plot\n",
    "        try:\n",
    "            if d == 'meta':\n",
    "                if alpha == 0.05:\n",
    "                    plt.figure(figsize=(6,3))\n",
    "                else:\n",
    "                    plt.figure(figsize=(6,6))\n",
    "            elif d == 'otu':\n",
    "                if alpha == 0.05:\n",
    "                    plt.figure(figsize=(6,3))\n",
    "                else:\n",
    "                    plt.figure(figsize=(8,8))\n",
    "            sns.set_theme(style=\"whitegrid\")\n",
    "            \n",
    "            sns.set_color_codes(\"pastel\")\n",
    "            sns.barplot(x='logp', y='feature', data=df,\n",
    "                        hue=\"Enrichment_Significance\",orient='horizontal')#,native_scale=True)#width=1)\n",
    "            \n",
    "            ax.set_ylabel(d_to_label[d],fontsize=16)\n",
    "            # ax.set_xlabel(\"Effect Size / Adjusted Test Statistic\",fontsize=16)\n",
    "            ax.set_xlabel(\"-log10(p)\",fontsize=16)\n",
    "            ax.tick_params(labelsize=12)\n",
    "            sns.despine(left=True, bottom=True)\n",
    "            plt.tight_layout()\n",
    "            plt.title(g + '_' + d + '_diff')\n",
    "            plt.savefig(path + 'outputs/jobs18/DA_hbarplot_diffs_' + g + '_' + d + '_' + str(alpha) + '.pdf')\n",
    "            plt.close()\n",
    "            df.to_csv(path + 'outputs/jobs18/' + g + '_' + d + '_diff_' + str(alpha) + '.tsv', sep='\\t')\n",
    "\n",
    "            # save for R plotting\n",
    "            # d = 'otu'\n",
    "            # g = 'saliva'\n",
    "            \n",
    "            if d == 'otu':\n",
    "                prefix = 'Qiime2_'\n",
    "            elif d == 'meta':\n",
    "                prefix = 'Meta_'\n",
    "            \n",
    "            df.to_csv(path + 'outputs/' + prefix + g + '_adh_diff/barplot_data.tsv', sep='\\t')\n",
    "        \n",
    "        except: \n",
    "            print('no plot for :' + g + '_' + d + '_diff')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "90768c43-3b95-494b-93d3-46d7b1a4222e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nfor g in ['stool', 'plasma']:\\n    # import mediation DFs\\n    df_M = pd.read_csv(path + 'inputs/cutie_df_meta_' + g + '.tsv', sep='\\t', index_col=0)\\n    df_M.iloc[:,:-1].to_csv(path + 'inputs/mediation_meta_' + g + '_post_test.txt', sep='\\t')\\n    \\n    df_M = gt_to_meta[g]['diff']\\n    df_M.to_csv(path + 'inputs/mediation_meta_' + g + '_diff_test.txt', sep='\\t')\\n    \\n    df_Y = df_meta_paired[['VAS_Pt_diff', 'WOMAC_pain_diff']]\\n    df_Y.to_csv(path + 'inputs/mediation_WMdiff_test.txt', sep='\\t')\\n\\n    if g == 'stool':\\n        df_X = pd.read_csv(path + 'inputs/cutie_df_taxa_' + g + '.tsv', sep='\\t', index_col=0)\\n        df_X.index = df_X.index.map(lambda x: x.split('.post')[0].replace('-',''))\\n        df_X.to_csv(path + 'inputs/mediation_taxa_' + g + '_post_test.txt', sep='\\t')\\n    \\n        df_X = gt_to_otu[g]['post'] - gt_to_otu[g]['pre']\\n        df_X.to_csv(path + 'inputs/mediation_taxa_' + g + '_diff_test.txt', sep='\\t')\\n\\n    # split by resp and non resp\\n    df = df_map.set_index('HostSubjectId')\\n    df.index = df.index.map(lambda x: x.replace('-',''))\\n    df = df[df['Timepoint'] == 'pre']\\n    df = pd.concat([df_M, df],axis=1)\\n\\n    df_M_nr = df[df['WOMAC_P_Response'] == 'No response'].drop(['WOMAC_P_Response','Timepoint'], axis=1)\\n    df_M_nr = df_M_nr.dropna()\\n    df_M_nr.to_csv(path + 'inputs/df_' + g + '_M_nr.tsv', sep='\\t')\\n    df_M_r = df[df['WOMAC_P_Response'] == 'Response'].drop(['WOMAC_P_Response','Timepoint'], axis=1)\\n    df_M_r = df_M_r.dropna()\\n    df_M_r.to_csv(path + 'inputs/df_' + g + '_M_r.tsv', sep='\\t')\\n\\n    if g == 'stool':\\n        # grab OTU \\n        df_post = gt_to_otu[g]['post'].copy()\\n        df_pre = gt_to_otu[g]['pre'].copy()\\n        \\n        df_post.index = df_post.index.map(lambda x: x.split('.p')[0].replace('-',''))\\n        df_pre.index = df_pre.index.map(lambda x: x.split('.p')[0].replace('-',''))\\n        df_X = df_post - df_pre\\n        df_X = df_X.dropna()\\n        df_X.to_csv(path + 'inputs/df_X_' + g + '.tsv', sep='\\t')\\n        \\n \\n\""
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Mediation analyses\n",
    "# check for int_ba\n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid','Hyocholic_acid'] + ['Deoxycholic_acid','Lithocholic_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid','Hyocholic_acid']+ ['Deoxycholic_acid','Lithocholic_acid']\n",
    "         } # 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide'\n",
    "\n",
    "# filter on CUTIE results\n",
    "g_to_cutiejob = {\n",
    "    'saliva': 'jobs15',\n",
    "    'stool': 'jobs16'\n",
    "}\n",
    "\n",
    "g_to_medjob = {\n",
    "    # 'saliva': '20240917',\n",
    "    # 'stool': '20240913',\n",
    "    'stool': '20241001'\n",
    "}\n",
    "# figure out cutie plots\n",
    "#X = list(df_X.columns.values)\n",
    "#M = list(df_M.columns.values)\n",
    "#x,m = df.iloc[0,0:2].values\n",
    "#X.index(x), M.index(m)\n",
    "\n",
    "\n",
    "\n",
    "# diet as an intervention\n",
    "# mediation script looks for 'test' as a string when grouping triplets for jobs\n",
    "df_diet.to_csv(path + 'inputs/mediation_diet_test.txt', sep='\\t')\n",
    "\n",
    "# define outcome\n",
    "y = 'WOMAC_pain_diff'\n",
    "\n",
    "'''\n",
    "for g in ['stool', 'plasma']:\n",
    "    # import mediation DFs\n",
    "    df_M = pd.read_csv(path + 'inputs/cutie_df_meta_' + g + '.tsv', sep='\\t', index_col=0)\n",
    "    df_M.iloc[:,:-1].to_csv(path + 'inputs/mediation_meta_' + g + '_post_test.txt', sep='\\t')\n",
    "    \n",
    "    df_M = gt_to_meta[g]['diff']\n",
    "    df_M.to_csv(path + 'inputs/mediation_meta_' + g + '_diff_test.txt', sep='\\t')\n",
    "    \n",
    "    df_Y = df_meta_paired[['VAS_Pt_diff', 'WOMAC_pain_diff']]\n",
    "    df_Y.to_csv(path + 'inputs/mediation_WMdiff_test.txt', sep='\\t')\n",
    "\n",
    "    if g == 'stool':\n",
    "        df_X = pd.read_csv(path + 'inputs/cutie_df_taxa_' + g + '.tsv', sep='\\t', index_col=0)\n",
    "        df_X.index = df_X.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "        df_X.to_csv(path + 'inputs/mediation_taxa_' + g + '_post_test.txt', sep='\\t')\n",
    "    \n",
    "        df_X = gt_to_otu[g]['post'] - gt_to_otu[g]['pre']\n",
    "        df_X.to_csv(path + 'inputs/mediation_taxa_' + g + '_diff_test.txt', sep='\\t')\n",
    "\n",
    "    # split by resp and non resp\n",
    "    df = df_map.set_index('HostSubjectId')\n",
    "    df.index = df.index.map(lambda x: x.replace('-',''))\n",
    "    df = df[df['Timepoint'] == 'pre']\n",
    "    df = pd.concat([df_M, df],axis=1)\n",
    "\n",
    "    df_M_nr = df[df['WOMAC_P_Response'] == 'No response'].drop(['WOMAC_P_Response','Timepoint'], axis=1)\n",
    "    df_M_nr = df_M_nr.dropna()\n",
    "    df_M_nr.to_csv(path + 'inputs/df_' + g + '_M_nr.tsv', sep='\\t')\n",
    "    df_M_r = df[df['WOMAC_P_Response'] == 'Response'].drop(['WOMAC_P_Response','Timepoint'], axis=1)\n",
    "    df_M_r = df_M_r.dropna()\n",
    "    df_M_r.to_csv(path + 'inputs/df_' + g + '_M_r.tsv', sep='\\t')\n",
    "\n",
    "    if g == 'stool':\n",
    "        # grab OTU \n",
    "        df_post = gt_to_otu[g]['post'].copy()\n",
    "        df_pre = gt_to_otu[g]['pre'].copy()\n",
    "        \n",
    "        df_post.index = df_post.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "        df_pre.index = df_pre.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "        df_X = df_post - df_pre\n",
    "        df_X = df_X.dropna()\n",
    "        df_X.to_csv(path + 'inputs/df_X_' + g + '.tsv', sep='\\t')\n",
    "        \n",
    " \n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ced57f62-7939-4437-a827-8da9c0874b66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n\\nfor g in [\\'stool\\', \\'saliva\\']:\\n    # import mediation DFs\\n    df_M = pd.read_csv(path + \\'inputs/cutie_df_meta_\\' + g + \\'.tsv\\', sep=\\'\\t\\', index_col=0)\\n    df_M.iloc[:,:-1].to_csv(path + \\'inputs/mediation_meta_\\' + g + \\'_post_test.txt\\', sep=\\'\\t\\')\\n    \\n    df_M = gt_to_meta[g][\\'diff\\']\\n    df_M.to_csv(path + \\'inputs/mediation_meta_\\' + g + \\'_diff_test.txt\\', sep=\\'\\t\\')\\n    \\n    df_Y = df_meta_paired[[\\'VAS_Pt_diff\\', \\'WOMAC_pain_diff\\']]\\n    df_Y.to_csv(path + \\'inputs/mediation_WMdiff_test.txt\\', sep=\\'\\t\\')\\n    \\n    df_X = pd.read_csv(path + \\'inputs/cutie_df_taxa_\\' + g + \\'.tsv\\', sep=\\'\\t\\', index_col=0)\\n    df_X.index = df_X.index.map(lambda x: x.split(\\'.post\\')[0].replace(\\'-\\',\\'\\'))\\n    df_X.to_csv(path + \\'inputs/mediation_taxa_\\' + g + \\'_post_test.txt\\', sep=\\'\\t\\')\\n\\n    df_X = gt_to_otu[g][\\'post\\'] - gt_to_otu[g][\\'pre\\']\\n    df_X.to_csv(path + \\'inputs/mediation_taxa_\\' + g + \\'_diff_test.txt\\', sep=\\'\\t\\')\\n\\n\\n    # split by resp and non resp\\n    df = df_map.set_index(\\'HostSubjectId\\')\\n    df.index = df.index.map(lambda x: x.replace(\\'-\\',\\'\\'))\\n    df = df[df[\\'Timepoint\\'] == \\'pre\\']\\n    df = pd.concat([df_M, df],axis=1)\\n\\n    df_M_nr = df[df[\\'WOMAC_P_Response\\'] == \\'No response\\'].drop([\\'WOMAC_P_Response\\',\\'Timepoint\\'], axis=1)\\n    df_M_nr = df_M_nr.dropna()\\n    df_M_nr.to_csv(path + \\'inputs/df_\\' + g + \\'_M_nr.tsv\\', sep=\\'\\t\\')\\n    df_M_r = df[df[\\'WOMAC_P_Response\\'] == \\'Response\\'].drop([\\'WOMAC_P_Response\\',\\'Timepoint\\'], axis=1)\\n    df_M_r = df_M_r.dropna()\\n    df_M_r.to_csv(path + \\'inputs/df_\\' + g + \\'_M_r.tsv\\', sep=\\'\\t\\')\\n\\n    # grab OTU \\n    df_post = gt_to_otu[g][\\'post\\'].copy()\\n    df_pre = gt_to_otu[g][\\'pre\\'].copy()\\n    \\n    df_post.index = df_post.index.map(lambda x: x.split(\\'.p\\')[0].replace(\\'-\\',\\'\\'))\\n    df_pre.index = df_pre.index.map(lambda x: x.split(\\'.p\\')[0].replace(\\'-\\',\\'\\'))\\n    df_X = df_post - df_pre\\n    df_X = df_X.dropna()\\n    df_X.to_csv(path + \\'inputs/df_X_\\' + g + \\'.tsv\\', sep=\\'\\t\\')\\n        \\n    # import mediation results\\n    #df_med = pd.read_csv(path + \\'outputs/jobs19/\\' + g_to_medjob[g] + \\'/outputs/mediation_taxa_\\' + g + \\'_post_mediation_meta_\\' + g + \\'_post_mediation_WMdiff_test/mediation_results_all.tsv\\', \\n    #                     sep=\\'\\t\\')\\n    # mediation_diet_mediation_taxa_stool_diff_mediation_WMdiff_test\\n    df_med = pd.read_csv(path + \\'outputs/jobs19/\\' + g_to_medjob[g] + \\'/outputs/mediation_diet_mediation_taxa_\\' + g + \\'_diff_mediation_WMdiff_test/mediation_results_all.tsv\\', \\n                         sep=\\'\\t\\')\\n\\n    for t in [\\'partial\\',\\'complete\\']:\\n        # copy df\\n        df = df_med.copy()\\n        \\n        # filter on significant total\\n        df = df[df[\\'total_effect_p\\'] < 0.05]\\n\\n        if t == \\'complete\\':\\n            # find complete mediations\\n            df = df[df[\\'direct_effect_p\\'] > 0.05]\\n\\n        elif t == \\'partial\\':\\n            # find partial mediations\\n            df = df[df[\\'direct_effect_p\\'] < 0.05]\\n        \\n        # check how many indirect; this reduces it by alot again\\n        df = df[df[\\'indirect_effect_p\\'] < 0.05]\\n\\n        # create column for metabolites\\n        # df = df[df[\\'M\\'].isin([\\'H_\\' + x for x in int_ba[g]])]\\n        \\n        # grab CUTIE results\\n        # df_qt = pd.read_csv(path + \\'outputs/\\' + g_to_cutiejob[g] + \\'/data_processing/summary_df_resample_1.txt\\', sep=\\'\\t\\')\\n        # df_qt = df_qt[df_qt[\\'class\\'] == \\'TP\\']\\n\\n        for r in range(len(df)):\\n            x,m,y = df.iloc[r,0:3].values\\n            \\n            # check if it\\'s in CUTIE\\n            # df_q = df_qt.copy()\\n            # df_q = df_q[df_q[\\'var1\\'] == m]\\n            # df_q = df_q[df_q[\\'var2\\'] == x]\\n        \\n            # if len(df_q) > 0: #len(df_q2) > 0: # len(df_q1) > 0 or len(df_q2) > 0:\\n            # print(\\'plotting \\' + m + \\' and \\' + x)\\n            # create plotting df\\n            \\n            # if X = bacteria, M = metabolite, Y = outcome\\n            df_plot = pd.concat([df_X,df_M,df_Y],axis=1)\\n\\n            # if X = diet, M = bacteria, Y = outcome\\n            df_plot = pd.concat([df_diet,df_X,df_Y],axis=1)\\n            if df_X[m].nunique() > 5 and df_diet[x].nunique() > 2:\\n                # path + \\'outputs/jobs19/plots/\\'\\n        \\n                plt.figure(figsize=(8,6))\\n                sns.set_theme(style=\"whitegrid\")\\n                # sns.scatterplot(data=df_plot,x=x,y=m,hue=y)\\n                sns.scatterplot(data=df_plot,x=x,y=y,hue=m,palette=\\'coolwarm\\')#,size=6)\\n                plt.tight_layout()\\n                sns.despine()\\n                plt.savefig(path + \\'outputs/jobs19/\\' + g_to_medjob[g] + \\'/plots/\\' + g + \\'/\\' + x.split(\\';g__\\')[-1] + \\'_\\' + m.replace(\\'/\\',\\'fslash\\') + \\'_\\' + t + \\'_\\' + y + \\'_MX.pdf\\')\\n                plt.close()\\n        \\n                plt.figure(figsize=(8,6))\\n                sns.set_theme(style=\"whitegrid\")        \\n                sns.scatterplot(data=df_plot,x=x,y=m,hue=m,palette=\\'coolwarm\\')#,size=6)\\n                sns.despine()\\n                plt.tight_layout()\\n                plt.savefig(path + \\'outputs/jobs19/\\' + g_to_medjob[g] + \\'/plots/\\' + g + \\'/\\' + x.split(\\';g__\\')[-1] + \\'_\\' + m.replace(\\'/\\',\\'fslash\\') + \\'_\\' + t + \\'_\\' + y + \\'_YMX.pdf\\')\\n                plt.close()\\n            \\ndf.head()\\n'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Mediation analyses\n",
    "# check for int_ba\n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid','Hyocholic_acid'] + ['Deoxycholic_acid','Lithocholic_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid','Hyocholic_acid']+ ['Deoxycholic_acid','Lithocholic_acid']\n",
    "         } # 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide'\n",
    "\n",
    "# filter on CUTIE results\n",
    "g_to_cutiejob = {\n",
    "    'saliva': 'jobs15',\n",
    "    'stool': 'jobs16'\n",
    "}\n",
    "\n",
    "g_to_medjob = {\n",
    "    # 'saliva': '20240917',\n",
    "    # 'stool': '20240913',\n",
    "    'stool': '20241001'\n",
    "}\n",
    "# figure out cutie plots\n",
    "#X = list(df_X.columns.values)\n",
    "#M = list(df_M.columns.values)\n",
    "#x,m = df.iloc[0,0:2].values\n",
    "#X.index(x), M.index(m)\n",
    "\n",
    "\n",
    "\n",
    "# diet as an intervention\n",
    "# mediation script looks for 'test' as a string when grouping triplets for jobs\n",
    "df_diet.to_csv(path + 'inputs/mediation_diet_test.txt', sep='\\t')\n",
    "\n",
    "# define outcome\n",
    "y = 'WOMAC_pain_diff'\n",
    "\n",
    "'''\n",
    "\n",
    "for g in ['stool', 'saliva']:\n",
    "    # import mediation DFs\n",
    "    df_M = pd.read_csv(path + 'inputs/cutie_df_meta_' + g + '.tsv', sep='\\t', index_col=0)\n",
    "    df_M.iloc[:,:-1].to_csv(path + 'inputs/mediation_meta_' + g + '_post_test.txt', sep='\\t')\n",
    "    \n",
    "    df_M = gt_to_meta[g]['diff']\n",
    "    df_M.to_csv(path + 'inputs/mediation_meta_' + g + '_diff_test.txt', sep='\\t')\n",
    "    \n",
    "    df_Y = df_meta_paired[['VAS_Pt_diff', 'WOMAC_pain_diff']]\n",
    "    df_Y.to_csv(path + 'inputs/mediation_WMdiff_test.txt', sep='\\t')\n",
    "    \n",
    "    df_X = pd.read_csv(path + 'inputs/cutie_df_taxa_' + g + '.tsv', sep='\\t', index_col=0)\n",
    "    df_X.index = df_X.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    df_X.to_csv(path + 'inputs/mediation_taxa_' + g + '_post_test.txt', sep='\\t')\n",
    "\n",
    "    df_X = gt_to_otu[g]['post'] - gt_to_otu[g]['pre']\n",
    "    df_X.to_csv(path + 'inputs/mediation_taxa_' + g + '_diff_test.txt', sep='\\t')\n",
    "\n",
    "\n",
    "    # split by resp and non resp\n",
    "    df = df_map.set_index('HostSubjectId')\n",
    "    df.index = df.index.map(lambda x: x.replace('-',''))\n",
    "    df = df[df['Timepoint'] == 'pre']\n",
    "    df = pd.concat([df_M, df],axis=1)\n",
    "\n",
    "    df_M_nr = df[df['WOMAC_P_Response'] == 'No response'].drop(['WOMAC_P_Response','Timepoint'], axis=1)\n",
    "    df_M_nr = df_M_nr.dropna()\n",
    "    df_M_nr.to_csv(path + 'inputs/df_' + g + '_M_nr.tsv', sep='\\t')\n",
    "    df_M_r = df[df['WOMAC_P_Response'] == 'Response'].drop(['WOMAC_P_Response','Timepoint'], axis=1)\n",
    "    df_M_r = df_M_r.dropna()\n",
    "    df_M_r.to_csv(path + 'inputs/df_' + g + '_M_r.tsv', sep='\\t')\n",
    "\n",
    "    # grab OTU \n",
    "    df_post = gt_to_otu[g]['post'].copy()\n",
    "    df_pre = gt_to_otu[g]['pre'].copy()\n",
    "    \n",
    "    df_post.index = df_post.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "    df_pre.index = df_pre.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "    df_X = df_post - df_pre\n",
    "    df_X = df_X.dropna()\n",
    "    df_X.to_csv(path + 'inputs/df_X_' + g + '.tsv', sep='\\t')\n",
    "        \n",
    "    # import mediation results\n",
    "    #df_med = pd.read_csv(path + 'outputs/jobs19/' + g_to_medjob[g] + '/outputs/mediation_taxa_' + g + '_post_mediation_meta_' + g + '_post_mediation_WMdiff_test/mediation_results_all.tsv', \n",
    "    #                     sep='\\t')\n",
    "    # mediation_diet_mediation_taxa_stool_diff_mediation_WMdiff_test\n",
    "    df_med = pd.read_csv(path + 'outputs/jobs19/' + g_to_medjob[g] + '/outputs/mediation_diet_mediation_taxa_' + g + '_diff_mediation_WMdiff_test/mediation_results_all.tsv', \n",
    "                         sep='\\t')\n",
    "\n",
    "    for t in ['partial','complete']:\n",
    "        # copy df\n",
    "        df = df_med.copy()\n",
    "        \n",
    "        # filter on significant total\n",
    "        df = df[df['total_effect_p'] < 0.05]\n",
    "\n",
    "        if t == 'complete':\n",
    "            # find complete mediations\n",
    "            df = df[df['direct_effect_p'] > 0.05]\n",
    "\n",
    "        elif t == 'partial':\n",
    "            # find partial mediations\n",
    "            df = df[df['direct_effect_p'] < 0.05]\n",
    "        \n",
    "        # check how many indirect; this reduces it by alot again\n",
    "        df = df[df['indirect_effect_p'] < 0.05]\n",
    "\n",
    "        # create column for metabolites\n",
    "        # df = df[df['M'].isin(['H_' + x for x in int_ba[g]])]\n",
    "        \n",
    "        # grab CUTIE results\n",
    "        # df_qt = pd.read_csv(path + 'outputs/' + g_to_cutiejob[g] + '/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "        # df_qt = df_qt[df_qt['class'] == 'TP']\n",
    "\n",
    "        for r in range(len(df)):\n",
    "            x,m,y = df.iloc[r,0:3].values\n",
    "            \n",
    "            # check if it's in CUTIE\n",
    "            # df_q = df_qt.copy()\n",
    "            # df_q = df_q[df_q['var1'] == m]\n",
    "            # df_q = df_q[df_q['var2'] == x]\n",
    "        \n",
    "            # if len(df_q) > 0: #len(df_q2) > 0: # len(df_q1) > 0 or len(df_q2) > 0:\n",
    "            # print('plotting ' + m + ' and ' + x)\n",
    "            # create plotting df\n",
    "            \n",
    "            # if X = bacteria, M = metabolite, Y = outcome\n",
    "            df_plot = pd.concat([df_X,df_M,df_Y],axis=1)\n",
    "\n",
    "            # if X = diet, M = bacteria, Y = outcome\n",
    "            df_plot = pd.concat([df_diet,df_X,df_Y],axis=1)\n",
    "            if df_X[m].nunique() > 5 and df_diet[x].nunique() > 2:\n",
    "                # path + 'outputs/jobs19/plots/'\n",
    "        \n",
    "                plt.figure(figsize=(8,6))\n",
    "                sns.set_theme(style=\"whitegrid\")\n",
    "                # sns.scatterplot(data=df_plot,x=x,y=m,hue=y)\n",
    "                sns.scatterplot(data=df_plot,x=x,y=y,hue=m,palette='coolwarm')#,size=6)\n",
    "                plt.tight_layout()\n",
    "                sns.despine()\n",
    "                plt.savefig(path + 'outputs/jobs19/' + g_to_medjob[g] + '/plots/' + g + '/' + x.split(';g__')[-1] + '_' + m.replace('/','fslash') + '_' + t + '_' + y + '_MX.pdf')\n",
    "                plt.close()\n",
    "        \n",
    "                plt.figure(figsize=(8,6))\n",
    "                sns.set_theme(style=\"whitegrid\")        \n",
    "                sns.scatterplot(data=df_plot,x=x,y=m,hue=m,palette='coolwarm')#,size=6)\n",
    "                sns.despine()\n",
    "                plt.tight_layout()\n",
    "                plt.savefig(path + 'outputs/jobs19/' + g_to_medjob[g] + '/plots/' + g + '/' + x.split(';g__')[-1] + '_' + m.replace('/','fslash') + '_' + t + '_' + y + '_YMX.pdf')\n",
    "                plt.close()\n",
    "            \n",
    "df.head()\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "44973fd6-4ca2-4cd0-81f7-691c681a1cdc",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df_X' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[27], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m## egg_d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens_partial_WOMAC_pain_diff_YMX.pdf\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m df_plot \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([df_diet,df_X,df_Y],axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m      3\u001b[0m x \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124megg\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      4\u001b[0m m \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124md__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens\u001b[39m\u001b[38;5;124m'\u001b[39m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df_X' is not defined"
     ]
    }
   ],
   "source": [
    "## egg_d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens_partial_WOMAC_pain_diff_YMX.pdf\n",
    "df_plot = pd.concat([df_diet,df_X,df_Y],axis=1)\n",
    "x = 'egg'\n",
    "m = 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens'\n",
    "y = 'WOMAC_pain_diff'\n",
    "\n",
    "sns.scatterplot(df_plot, x=x, y=y)\n",
    "scipy.stats.spearmanr(df_plot[x], df_plot[y], nan_policy='omit')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01ae56d6-b961-42ad-b3bf-fb8ef3dd22b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(df_plot, x=x, y=m)\n",
    "scipy.stats.spearmanr(df_plot[x],df_plot[m],nan_policy='omit')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cbc6f27-6f8b-498a-83a9-94d4a2007c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(df_plot, x=m, y=y)\n",
    "scipy.stats.spearmanr(df_plot[m],df_plot[y],nan_policy='omit')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "001a1ab2-4876-4cf7-b308-862c073fd6b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in df_diet.columns:\n",
    "    for y in ['paired_alpha', 'paired_beta']:\n",
    "        df_merge = pd.concat([df_diet, g_to_dfd['stool_adh'][y]],axis=1)\n",
    "        ylab = g_to_dfd['stool_adh'][y].columns.values[0]\n",
    "        r, p = scipy.stats.spearmanr(df_merge[x],df_merge[ylab],nan_policy='omit')\n",
    "        if p < 0.1:\n",
    "            print(x,y)\n",
    "            print(p)\n",
    "                                     \n",
    "            plt.figure(figsize=(8,6))\n",
    "            sns.set_theme(style=\"whitegrid\")        \n",
    "            sns.scatterplot(data=df_merge,x=x,y=ylab)\n",
    "            sns.despine()\n",
    "            plt.tight_layout()\n",
    "            plt.savefig(path + 'outputs/jobs19/diet_plots/' + x +'_' +ylab + '.pdf')\n",
    "            plt.close()\n",
    "g_to_dfd['stool_adh']['paired_alpha'].head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53a8f0b2-47ab-4bb5-8f69-86d734fdd89e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mediation stability\n",
    "for p in ['100','200','1000']:\n",
    "    print('number of bootstraps: ' + p)\n",
    "    df = pd.read_csv(path + 'outputs/jobs20/df' + p + '.csv')\n",
    "    for y in ['total_effect_p','direct_effect_p','indirect_effect_p']:\n",
    "        x = df[y]\n",
    "        print('statistic: ' + y)\n",
    "        print('prop under 0.05: ')\n",
    "        print(len(x[x<0.05])/len(x))\n",
    "        print('mean and std')\n",
    "        print(np.mean(x), np.std(x))\n",
    "        print('\\n')\n",
    "    \n",
    "sns.distplot(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "754a719c-f121-424e-9fc3-a9ba86ff5ea5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Procrustes ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68ae796e-af81-4475-bb2d-316fd6e8e413",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Procrustes\n",
    "import scipy.spatial\n",
    "\n",
    "str_to_df = {\n",
    "    'diet': df_diet\n",
    "}\n",
    "\n",
    "for g in ['stool', 'saliva']:\n",
    "    for t in ['Baseline', 'After diet']:\n",
    "        str_to_df['_'.join([g, t, 'meta'])] = gt_to_meta[g][t]\n",
    "    for t in['pre', 'post']:\n",
    "        df = gt_to_otu[g][t]\n",
    "        df.index = df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "        str_to_df['_'.join([g, t, 'taxa'])] = df\n",
    "\n",
    "\n",
    "def convert_to_pcs(df, npcs):\n",
    "    pca = PCA(n_components=npcs, whiten=True)\n",
    "    pca.fit(df.T)\n",
    "    return pd.DataFrame(pca.components_, \n",
    "                        index=['PC%s' % _ for _ in range(npcs)], \n",
    "                        columns=df.index).T\n",
    "\n",
    "#df1 = convert_to_pcs(df1, 10)\n",
    "#df2 = convert_to_pcs(df2, 10)\n",
    "\n",
    "# convert to distance matrices\n",
    "from scipy.spatial.distance import squareform, pdist\n",
    "from itertools import combinations\n",
    "from skbio import DistanceMatrix\n",
    "from skbio.stats.distance import mantel\n",
    "\n",
    "# get all pairs of dfs\n",
    "# df_pairs = list(combinations(str_to_df, 2))\n",
    "\n",
    "\n",
    "# build comparisons list\n",
    "df_pairs = []\n",
    "# 'Baseline' 'After diet' # for meta\n",
    "# 'pre' 'post' # for stool\n",
    "\n",
    "# stool microbiome nad metabolome at pre and then at post\n",
    "a = ['stool_pre_taxa', 'stool_Baseline_meta']\n",
    "b = ['stool_post_taxa', 'stool_After diet_meta']\n",
    "\n",
    "c = ['saliva_pre_taxa', 'saliva_Baseline_meta']\n",
    "d = ['saliva_post_taxa', 'saliva_After diet_meta']\n",
    "\n",
    "# better fit \n",
    "\n",
    "# saliva to gut in microbiome at pre and then at post\n",
    "e = ['saliva_pre_taxa', 'stool_pre_taxa']\n",
    "f = ['saliva_post_taxa', 'stool_post_taxa']\n",
    "\n",
    "\n",
    "# repeat for metabolome \n",
    "g = ['saliva_Baseline_meta', 'stool_Baseline_meta']\n",
    "h = ['saliva_After diet_meta', 'stool_After diet_meta']\n",
    "\n",
    "# worse fit  \n",
    "\n",
    "df_pairs = [a, b, c, d, e, f, g, h]\n",
    "\n",
    "\n",
    "for d1, d2 in df_pairs:\n",
    "    print(d1, d2)\n",
    "    # retrieve df's \n",
    "    df1, df2 = str_to_df[d1], str_to_df[d2]\n",
    "    \n",
    "    # convert to distance matrix\n",
    "    dm1 = pd.DataFrame(squareform(pdist(df1.iloc[:, 1:])), columns=df1.index.unique(), index=df1.index.unique())\n",
    "    dm2 = pd.DataFrame(squareform(pdist(df2.iloc[:, 1:])), columns=df2.index.unique(), index=df2.index.unique())\n",
    "\n",
    "    # grab common samples\n",
    "    shared_idx = list(set(dm1.index).intersection(set(dm2.index)))\n",
    "    dm1 = dm1.loc[shared_idx,shared_idx]\n",
    "    dm2 = dm2.loc[shared_idx,shared_idx]\n",
    "\n",
    "    # compute mantel \n",
    "    r, p, n = mantel(x=dm1, y=dm2, method='spearman')\n",
    "    if p < 0.05:\n",
    "        print(r,p)\n",
    "\n",
    "    # compute m2\n",
    "    df1p, df2p, m2 = scipy.spatial.procrustes(dm1.values,dm2.values)\n",
    "    print(m2)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c610c4b-54e2-43eb-a529-dec577a5187f",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### PERMANOVA ANALYSES ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61203d0e-62c8-4e39-8e85-e37ef753097c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PERMANOVA metabolites\n",
    "# construct dict of data\n",
    "str_to_df = {\n",
    "    #'diet': df_diet\n",
    "}\n",
    "\n",
    "import labtools.statistics\n",
    "\n",
    "for g in ['stool', 'plasma','saliva']:\n",
    "    for t in ['Baseline', 'After diet']:\n",
    "        str_to_df['_'.join([g, t, 'meta'])] = gt_to_meta[g][t]\n",
    "\n",
    "# this is testing resp vs non resp at pre and post time points only, NOT the paired change in beta post-pre \n",
    "for label in str_to_df:\n",
    "    print(label)\n",
    "    df = str_to_df[label]\n",
    "\n",
    "    # get dx\n",
    "    df_dx = type_to_df_map[label.split('_')[0]].drop_duplicates(subset='HostSubjectId')\n",
    "    df_dx = df_dx.set_index('HostSubjectId')\n",
    "    df_dx.index = df_dx.index.map(lambda x: x.replace('-',''))\n",
    "    df_dx = df_dx[['Timepoint','WOMAC_P_Response']]\n",
    "\n",
    "    # get merged df and drop na\n",
    "    df.index.names = ['SubjectID']\n",
    "    df_dx.index.names = ['SubjectID']\n",
    "    \n",
    "    df_merge = pd.concat([df, df_dx], axis=1)\n",
    "    df_merge = df_merge.dropna()\n",
    "    \n",
    "    # construct distance matrix from df\n",
    "    df_data = df.loc[df_merge.index,:]\n",
    "    dx = df_dx.loc[df_merge.index,:]['WOMAC_P_Response']\n",
    "    dm = pd.DataFrame(squareform(pdist(df_data)), columns=df_data.index.unique(), index=df_data.index.unique())\n",
    "\n",
    "    # takes in distance matrix\n",
    "    print(labtools.statistics.permanova_oneway(dm, levels=[1 if x == 'No response' else 0 for x in dx]))\n",
    "\n",
    "dm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f70dd3f6-3455-4b56-8be2-01dcef569ff0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Shannon entropy calculations\n",
    "gt_to_meta_norm = {}\n",
    "g_to_meta_norm = {}\n",
    "\n",
    "for g in ['stool','plasma','saliva']:\n",
    "    gt_to_meta_norm[g] = {}\n",
    "\n",
    "for g in ['plasma','stool','saliva']:\n",
    "    print(g)\n",
    "    df = pd.read_csv(path + 'inputs/' + g + '_removed_zeros.csv')#, index_col=0)\n",
    "\n",
    "    # split on id and timepoint and rest of data\n",
    "    df_info = df[['Sample','Time']]\n",
    "\n",
    "    # normalize\n",
    "    df_data = df.iloc[:,2:]\n",
    "    df_data = df_data.div(df_data.sum(axis=1),axis=0)\n",
    "\n",
    "    # recombine\n",
    "    df = pd.concat([df_info, df_data],axis=1)\n",
    "    g_to_meta_norm[g] = df\n",
    "\n",
    "    # split on time point\n",
    "    df_pre = df[df['Time'] == 'Baseline']\n",
    "    df_pre = df_pre.drop('Time',axis=1)\n",
    "    df_pre = df_pre.set_index('Sample')\n",
    "    df_pre.index = df_pre.index.map(lambda x: x.split('_Basal')[0].replace('_',''))\n",
    "    gt_to_meta_norm[g]['pre'] = df_pre\n",
    "    \n",
    "    df_post = df[df['Time'].isin(['Day15','After diet'])]\n",
    "    df_post = df_post.drop('Time',axis=1)\n",
    "    df_post = df_post.set_index('Sample')\n",
    "    df_post.index = df_post.index.map(lambda x: x.split('_Day15')[0].replace('_',''))\n",
    "    gt_to_meta_norm[g]['post'] = df_post\n",
    "\n",
    "\n",
    "df_post.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec38f5db-8aa9-4d7b-995f-6636814f47a6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# pre post comparison\n",
    "# WSR on entropies\n",
    "for g in ['plasma','stool','saliva']:\n",
    "    print(g)\n",
    "    df_pre = gt_to_meta_norm[g]['pre']\n",
    "    df_pre['shannon_entropy'] = scipy.stats.entropy(df_pre, axis=1)\n",
    "\n",
    "    df_post = gt_to_meta_norm[g]['post']\n",
    "    df_post['shannon_entropy'] = scipy.stats.entropy(df_post, axis=1)\n",
    "    \n",
    "    # common samples\n",
    "    common_idx = list(set(df_pre.index.values).intersection(set(df_post.index.values)))\n",
    "    df_pre = df_pre.loc[common_idx,:]\n",
    "    df_post = df_post.loc[common_idx,:]\n",
    "    print(len(df_pre),len(df_post))\n",
    "\n",
    "    df_pre2 = df_pre.copy()\n",
    "    df_pre2.index = df_pre2.index.map(lambda x: x + '_pre')\n",
    "    df_pre2['Timepoint'] = 'pre'\n",
    "    df_pre2 = df_pre2.sort_index()\n",
    "\n",
    "    df_post2 = df_post.copy()\n",
    "    df_post2.index = df_post2.index.map(lambda x: x + '_post')\n",
    "    df_post2['Timepoint'] = 'post'\n",
    "    df_post2 = df_post2.sort_index()\n",
    "\n",
    "\n",
    "    \n",
    "    # merge\n",
    "    df_merge = pd.concat([df_pre2, df_post2])\n",
    "\n",
    "    prevals = df_merge[df_merge['Timepoint'] == 'pre']['shannon_entropy'].values\n",
    "    postvals = df_merge[df_merge['Timepoint'] == 'post']['shannon_entropy'].values\n",
    "    # get paired differences\n",
    "    # it looks like the rows are in the correct order will test\n",
    "    # for x,y in zip(df_pre.index, df_post.index):\n",
    "    #    if x != y:\n",
    "    #        print(x,y)\n",
    "    print('combined pre post wsr')\n",
    "    W,p = scipy.stats.wilcoxon(x=prevals,y=postvals, nan_policy='omit')\n",
    "    print(W,p)\n",
    "    ax = sns.swarmplot(data=df_merge, x='Timepoint', y='shannon_entropy', hue='Timepoint', palette=['b','r'], orient='v')\n",
    "    for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "        ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "    sns.boxplot(data=df_merge, x='Timepoint', y='shannon_entropy', orient='v', hue='Timepoint', boxprops=dict(alpha=.3))\n",
    "    \n",
    "    # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "    ax.set_ylabel('Shannon Entropy',fontsize=16)\n",
    "    ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "    ax.tick_params(labelsize=16)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/wsr_meta_norm_' + g  + '.pdf')\n",
    "    plt.close()            \n",
    "    \n",
    "\n",
    "    # comparing R v NR at pre and post time points\n",
    "    df_dx = type_to_df_map[g].reset_index().copy()\n",
    "    df_dx['Together'] = df_dx['Together'].map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "    df_dx = df_dx.drop_duplicates(subset=['Together'])\n",
    "    df_dx = df_dx.set_index('Together')\n",
    "    df_dx.index.name = 'SampleID'\n",
    "    df_dx = df_dx[['WOMAC_P_Response']]\n",
    "\n",
    "\n",
    "    # comparing R v NR in terms of delta pre-post\n",
    "    # df_diff = df_post.drop('WOMAC_P_Response',axis=1) - df_pre.drop('WOMAC_P_Response',axis=1)\n",
    "    df_diff = df_post - df_pre\n",
    "    df_diff = pd.concat([df_diff,df_dx],axis=1)\n",
    "    \n",
    "    rvals = df_diff[df_diff['WOMAC_P_Response'] == 'Response']['shannon_entropy'].values\n",
    "    nrvals = df_diff[df_diff['WOMAC_P_Response'] == 'No response']['shannon_entropy'].values  \n",
    "    \n",
    "    W,p = scipy.stats.mannwhitneyu(rvals,nrvals, nan_policy='omit')\n",
    "    print('diff')\n",
    "    print(W,p)\n",
    "    ax = sns.swarmplot(data=df_diff, x='WOMAC_P_Response', y='shannon_entropy', hue='WOMAC_P_Response', palette=['b','r'], orient='v')\n",
    "    sns.boxplot(data=df_diff, x='WOMAC_P_Response', y='shannon_entropy', orient='v', hue='WOMAC_P_Response', boxprops=dict(alpha=.3))\n",
    "    \n",
    "    # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "    ax.set_ylabel('Shannon Entropy',fontsize=16)\n",
    "    ax.set_xlabel('WOMAC_P_Response',fontsize=16)                \n",
    "    ax.tick_params(labelsize=16)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/wsr_meta_norm_diff_rvnr_' + g  + '.pdf')\n",
    "    plt.close()            \n",
    "\n",
    "\n",
    "    # pre\n",
    "    df_pre = pd.concat([df_pre, df_dx],axis=1)\n",
    "    df_pre = df_pre.dropna()\n",
    "\n",
    "    rvals = df_pre[df_pre['WOMAC_P_Response'] == 'Response']['shannon_entropy'].values\n",
    "    nrvals = df_pre[df_pre['WOMAC_P_Response'] == 'No response']['shannon_entropy'].values  \n",
    "    \n",
    "    W,p = scipy.stats.mannwhitneyu(rvals,nrvals, nan_policy='omit')\n",
    "    print('Baseline')\n",
    "    print(W,p)\n",
    "    ax = sns.swarmplot(data=df_pre, x='WOMAC_P_Response', y='shannon_entropy', hue='WOMAC_P_Response', palette=['b','r'], orient='v')\n",
    "    sns.boxplot(data=df_pre, x='WOMAC_P_Response', y='shannon_entropy', orient='v', hue='WOMAC_P_Response', boxprops=dict(alpha=.3))\n",
    "    \n",
    "    # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "    ax.set_ylabel('Shannon Entropy',fontsize=16)\n",
    "    ax.set_xlabel('WOMAC_P_Response',fontsize=16)                \n",
    "    ax.tick_params(labelsize=16)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/wsr_meta_norm_baseline_rvnr_' + g  + '.pdf')\n",
    "    plt.close()            \n",
    "\n",
    "\n",
    "\n",
    "    # post\n",
    "    df_post = pd.concat([df_post, df_dx],axis=1)\n",
    "    df_post = df_post.dropna()\n",
    "\n",
    "    rvals = df_post[df_post['WOMAC_P_Response'] == 'Response']['shannon_entropy'].values    \n",
    "    nrvals = df_post[df_post['WOMAC_P_Response'] == 'No response']['shannon_entropy'].values  \n",
    "    \n",
    "    W,p = scipy.stats.mannwhitneyu(rvals,nrvals, nan_policy='omit')\n",
    "    print('After diet')\n",
    "    print(W,p)\n",
    "    ax = sns.swarmplot(data=df_post, x='WOMAC_P_Response', y='shannon_entropy', hue='WOMAC_P_Response', palette=['b','r'], orient='v')\n",
    "    sns.boxplot(data=df_pre, x='WOMAC_P_Response', y='shannon_entropy', orient='v', hue='WOMAC_P_Response', boxprops=dict(alpha=.3))\n",
    "    \n",
    "    # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "    ax.set_ylabel('Shannon Entropy',fontsize=16)\n",
    "    ax.set_xlabel('WOMAC_P_Response',fontsize=16)                \n",
    "    ax.tick_params(labelsize=16)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/wsr_meta_norm_afterdiet_rvnr_' + g  + '.pdf')\n",
    "    plt.close()            \n",
    "\n",
    "\n",
    "    # responders pre vs post\n",
    "    df_dx_pre = df_dx.copy()\n",
    "    df_dx_pre.index = df_dx_pre.index.map(lambda x: x + '_pre')\n",
    "    df_dx_post = df_dx.copy()\n",
    "    df_dx_post.index = df_dx_post.index.map(lambda x: x + '_post')\n",
    "\n",
    "    df_dx2 = pd.concat([df_dx_pre, df_dx_post], axis=0)\n",
    "    \n",
    "    df_merge = pd.concat([df_pre2, df_post2])\n",
    "    df_merge2 = pd.concat([df_merge, df_dx2], axis=1)\n",
    "    df_merge = df_merge2.dropna()\n",
    "    \n",
    "    df_merge = df_merge[df_merge['WOMAC_P_Response'] == 'Response']\n",
    "\n",
    "    prevals = df_merge[df_merge['Timepoint'] == 'pre']['shannon_entropy'].values\n",
    "    postvals = df_merge[df_merge['Timepoint'] == 'post']['shannon_entropy'].values\n",
    "    # get paired differences\n",
    "    # it looks like the rows are in the correct order will test\n",
    "    # for x,y in zip(df_pre.index, df_post.index):\n",
    "    #    if x != y:\n",
    "    #        print(x,y)\n",
    "\n",
    "    W,p = scipy.stats.wilcoxon(x=prevals,y=postvals, nan_policy='omit')\n",
    "    print('resp only pre post')\n",
    "    print(W,p)\n",
    "    ax = sns.swarmplot(data=df_merge, x='Timepoint', y='shannon_entropy', hue='Timepoint', palette=['b','r'], orient='v')\n",
    "    for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "        ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "    sns.boxplot(data=df_merge, x='Timepoint', y='shannon_entropy', orient='v', hue='Timepoint', boxprops=dict(alpha=.3))\n",
    "    \n",
    "    # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "    ax.set_ylabel('Shannon Entropy',fontsize=16)\n",
    "    ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "    ax.tick_params(labelsize=16)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/wsr_meta_norm_resp_' + g  + '.pdf')\n",
    "    plt.close()            \n",
    "    \n",
    "    df_merge = df_merge2.dropna()\n",
    "    \n",
    "    df_merge = df_merge[df_merge['WOMAC_P_Response'] == 'No response']\n",
    "\n",
    "    prevals = df_merge[df_merge['Timepoint'] == 'pre']['shannon_entropy'].values\n",
    "    postvals = df_merge[df_merge['Timepoint'] == 'post']['shannon_entropy'].values\n",
    "    # get paired differences\n",
    "    # it looks like the rows are in the correct order will test\n",
    "    # for x,y in zip(df_pre.index, df_post.index):\n",
    "    #    if x != y:\n",
    "    #        print(x,y)\n",
    "\n",
    "    W,p = scipy.stats.wilcoxon(x=prevals,y=postvals, nan_policy='omit')\n",
    "    print('nr only pre post')\n",
    "    print(W,p)\n",
    "    ax = sns.swarmplot(data=df_merge, x='Timepoint', y='shannon_entropy', hue='Timepoint', palette=['b','r'], orient='v')\n",
    "    for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "        ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "    sns.boxplot(data=df_merge, x='Timepoint', y='shannon_entropy', orient='v', hue='Timepoint', boxprops=dict(alpha=.3))\n",
    "    \n",
    "    # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "    ax.set_ylabel('Shannon Entropy',fontsize=16)\n",
    "    ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "    ax.tick_params(labelsize=16)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/wsr_meta_norm_noresp_' + g  + '.pdf')\n",
    "    plt.close()            \n",
    "df_dx.head()   \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6af0b7f-09d0-47c6-bc23-4530c76eae11",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Entropies for R ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e123a37b-4118-48cc-8b78-e87920b88906",
   "metadata": {},
   "outputs": [],
   "source": [
    "paired_dirs = [\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_stool_adh\",\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_saliva_adh\",\n",
    "  \n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_stool_adh_response\",\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_stool_adh_noresponse\",\n",
    "  \n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_saliva_adh_response\",\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_saliva_adh_noresponse\"\n",
    "]\n",
    "\n",
    "unpaired_dirs = [\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_saliva_adh_pre\",\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_saliva_adh_post\",\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_stool_adh_pre\",\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_stool_adh_post\"\n",
    "]\n",
    "import os\n",
    "\n",
    "# create merge\n",
    "for g in ['stool','plasma','saliva']:\n",
    "    pre_df = gt_to_meta_norm[g]['pre'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x + '.pre')\n",
    "    post_df = gt_to_meta_norm[g]['post'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x + '.post')\n",
    "    gt_to_meta_norm[g]['all'] = pd.concat([pre_df, post_df])\n",
    "\n",
    "    comparisons = ['_pre','_post','_response','_noresponse','']\n",
    "    for c in comparisons:\n",
    "        # make new folder for analysis\n",
    "        outdir = \"/Users/KevinBu/Desktop/clemente_lab/Projects/oa/outputs/Meta_\" + g + '_adh' + c\n",
    "        # os.mkdir(outdir)\n",
    "\n",
    "        # grab mapping file\n",
    "        df_map = type_to_df_map[g].copy()\n",
    "        df_map.index = df_map['HostSubjectId'] + '.' + df_map['Timepoint'] #replace('_'\n",
    "        df_map.index = df_map.index.map(lambda x: x.replace('-',''))\n",
    "        \n",
    "        # merge with metabolite data\n",
    "        df_meta = gt_to_meta_norm[g]['all'].copy()\n",
    "        df_merge = pd.concat([df_map, df_meta],axis=1)\n",
    "\n",
    "        # subset on cols of interest\n",
    "        if c == '_pre':\n",
    "            df_merge = df_merge[df_merge['Timepoint'] == 'pre']\n",
    "        if c == '_post':\n",
    "            df_merge = df_merge[df_merge['Timepoint'] == 'post']\n",
    "        if c == '_response':\n",
    "            df_merge = df_merge[df_merge['WOMAC_P_Response'] == 'Response']\n",
    "        if c == '_noresponse':\n",
    "            df_merge = df_merge[df_merge['WOMAC_P_Response'] == 'No response']\n",
    "\n",
    "        # dropna\n",
    "        df_merge = df_merge.dropna(subset='shannon_entropy')\n",
    "\n",
    "        # need to make a mapping file with columns Study_ID, Timepoint, WOMAC_P_Response, and shannon_entropy \n",
    "        df_merge = df_merge.loc[:,['Study_ID','Timepoint','HostSubjectId','WOMAC_pain','WOMAC_P_Response','shannon_entropy']]\n",
    "\n",
    "        # export to folder\n",
    "        df_merge.to_csv(outdir + '/metadata.tsv', sep='\\t')\n",
    "        \n",
    "df_merge.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "563c0ded-57c4-4b1e-babf-b7376b28fdbc",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### PERMANOVA OTU ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18623504-6fef-4780-85bc-80f548122cf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PERMANOVA OTU\n",
    "# this is testing resp vs non resp at pre and post time points only, NOT the paired change in beta post-pre \n",
    "# this takes care of\n",
    "'''\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_stool_adh/\",\n",
    "  \n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_saliva_adh/\",\n",
    " \n",
    "  \n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_saliva_adh_pre/\",\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_saliva_adh_post/\",\n",
    "\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_stool_adh_pre/\",\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_stool_adh_post/\",\n",
    "\n",
    "'''\n",
    "np.random.seed(0) # this doesn't actually fix PERMANOVA results :(\n",
    "\n",
    "for g in ['stool', 'saliva']:\n",
    "    print(g)\n",
    "    df_raw = pd.read_csv(path + 'outputs/Qiime2_' + g + '_adh/core_metrics_results/distance-matrix.tsv', \n",
    "                 sep='\\t', index_col=0)\n",
    "\n",
    "    # only compute for paired samples\n",
    "    df_map = type_to_df_map[g].copy()\n",
    "    df_map = df_map[df_map.duplicated(subset=['HostSubjectId'], keep=False)]\n",
    "    \n",
    "    for t in [['pre'], ['post'],['pre','post']]:\n",
    "        print(t)\n",
    "        df_map_sub = df_map[df_map['Timepoint'].isin(t)]\n",
    "\n",
    "        # drop NA in pain response\n",
    "        df_map_sub = df_map_sub.dropna(subset='WOMAC_pain')\n",
    "        \n",
    "        # find overlapping samples between df_raw and df_map_sub\n",
    "        keep = list(set(df_map_sub.index.values).intersection(set(df_raw.index.values)))\n",
    "        \n",
    "        # get dx\n",
    "        df_map_sub = df_map_sub.loc[keep,:]\n",
    "        dx = df_map_sub['WOMAC_P_Response'].values\n",
    "\n",
    "        # subest dm\n",
    "        df = df_raw.loc[keep,keep]\n",
    "        \n",
    "        # takes in distance matrix\n",
    "        print(permanova_oneway(df, levels=[0 if x == 'No response' else 1 for x in dx]))\n",
    "    \n",
    "        from sklearn.manifold import MDS\n",
    "        \n",
    "        mds = MDS(n_components=2, dissimilarity='precomputed')\n",
    "        coordinates = mds.fit_transform(df)\n",
    "        \n",
    "        df_pca = pd.DataFrame(coordinates, columns=['PC1','PC2'], index=df.index.unique())\n",
    "        df_pca['Timepoint'] = dx\n",
    "\n",
    "        df_merge = pd.concat([df_pca.drop('Timepoint',axis=1), df_map_sub],axis=1)\n",
    "        df_merge = df_merge[['PC1','PC2','Timepoint','HostSubjectId','WOMAC_P_Response']]\n",
    "\n",
    "        # save\n",
    "        if len(t) == 1:\n",
    "            df_merge.to_csv(path + 'outputs/Qiime2_' + g + '_adh_' + t[0] + '/pcoa.tsv', sep='\\t') \n",
    "        else:\n",
    "            df_merge.to_csv(path + 'outputs/Qiime2_' + g + '_adh' + '/pcoa.tsv', sep='\\t') \n",
    "\n",
    "\n",
    "df_merge.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f78e9337-4333-446c-a9f5-c3af16a566a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PERMANOVA OTU\n",
    "# this is testing resp vs non resp at pre and post time points only, NOT the paired change in beta post-pre \n",
    "# this takes care of\n",
    "'''\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_stool_adh_response/\",\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_stool_adh_noresponse/\",\n",
    "  \n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_saliva_adh_response/\",\n",
    "  \"~/Desktop/clemente_lab/Projects/oa/outputs/Qiime2_saliva_adh_noresponse/\",\n",
    "\n",
    "'''\n",
    "np.random.seed(0) # this doesn't actually fix PERMANOVA results :(\n",
    "\n",
    "for g in ['stool', 'saliva']:\n",
    "    print(g)\n",
    "    df_raw = pd.read_csv(path + 'outputs/Qiime2_' + g + '_adh/core_metrics_results/distance-matrix.tsv', \n",
    "                 sep='\\t', index_col=0)\n",
    "\n",
    "    # only compute for paired samples\n",
    "    df_map = type_to_df_map[g].copy()\n",
    "    df_map = df_map[df_map.duplicated(subset=['HostSubjectId'], keep=False)]\n",
    "    \n",
    "    for t in [['Response'], ['No response']]:\n",
    "        print(t)\n",
    "        df_map_sub = df_map[df_map['WOMAC_P_Response'].isin(t)]\n",
    "\n",
    "        # drop NA in pain response\n",
    "        df_map_sub = df_map_sub.dropna(subset='WOMAC_pain')\n",
    "        \n",
    "        # find overlapping samples between df_raw and df_map_sub\n",
    "        keep = list(set(df_map_sub.index.values).intersection(set(df_raw.index.values)))\n",
    "        \n",
    "        # get dx\n",
    "        df_map_sub = df_map_sub.loc[keep,:]\n",
    "        dx = df_map_sub['Timepoint'].values\n",
    "\n",
    "        # subest dm\n",
    "        df = df_raw.loc[keep,keep]\n",
    "        \n",
    "        # takes in distance matrix\n",
    "        print(permanova_oneway(df, levels=[0 if x == 'pre' else 1 for x in dx]))\n",
    "    \n",
    "        from sklearn.manifold import MDS\n",
    "        \n",
    "        mds = MDS(n_components=2, dissimilarity='precomputed')\n",
    "        coordinates = mds.fit_transform(df)\n",
    "        \n",
    "        df_pca = pd.DataFrame(coordinates, columns=['PC1','PC2'], index=df.index.unique())\n",
    "        df_pca['Timepoint'] = dx\n",
    "\n",
    "        df_merge = pd.concat([df_pca.drop('Timepoint',axis=1), df_map_sub],axis=1)\n",
    "        df_merge = df_merge[['PC1','PC2','Timepoint','HostSubjectId','WOMAC_P_Response']]\n",
    "\n",
    "        # save\n",
    "        if t[0] == 'Response':\n",
    "            df_merge.to_csv(path + 'outputs/Qiime2_' + g + '_adh_response/pcoa.tsv', sep='\\t') \n",
    "        if t[0] == 'No response':\n",
    "            df_merge.to_csv(path + 'outputs/Qiime2_' + g + '_adh_noresponse/pcoa.tsv', sep='\\t') \n",
    "        \n",
    "\n",
    "df_merge.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7219e20-315c-42bc-abee-76211b89e75c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for metabolites, compare pre and post timepoints INDEP of respones status\n",
    "'''\n",
    "for g in ['stool','plasma','saliva']:\n",
    "    print(g)\n",
    "    df = gt_to_meta[g]['all'].copy()\n",
    "    \n",
    "    orig_idx = df.index\n",
    "    \n",
    "    df2 = df.copy()\n",
    "    #df2.index = df2.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "\n",
    "    # df2['Timepoint'] = df2.index.map(df_dx.to_dict()['Timepoint'])\n",
    "    df_map = type_to_df_map[g].copy().set_index('Together')\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('guma')[0])\n",
    "    df = pd.concat([df2,df_map['Timepoint']],axis=1)\n",
    "    df = df.dropna()\n",
    "\n",
    "    #df2.index = orig_idx\n",
    "    #df = df2.copy()\n",
    "    #df = df.dropna()\n",
    "    dx = df['Timepoint'].values\n",
    "    df = df.drop('Timepoint',axis=1)\n",
    "    \n",
    "    # df.index = df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "    # from scipy.spatial.distance import cdist\n",
    "    #dm = pd.DataFrame(cdist(gt_to_meta['stool']['Baseline'], gt_to_meta['stool']['After diet'], 'euclidean'))\n",
    "    \n",
    "    dm = pd.DataFrame(squareform(pdist(df)), columns=df.index.unique(), index=df.index.unique())\n",
    "    # takes in distance matrix\n",
    "    print(permanova_oneway(dm, levels=[1 if (x == 'After diet') | (x == 'post') else 0 for x in dx]))\n",
    "\n",
    "    \n",
    "    from sklearn.manifold import MDS\n",
    "    \n",
    "    mds = MDS(n_components=2, dissimilarity='precomputed')\n",
    "    coordinates = mds.fit_transform(dm)\n",
    "    \n",
    "    df_pca = pd.DataFrame(coordinates, columns=['PC1','PC2'], index=df.index.unique())\n",
    "    df_pca['Timepoint'] = dx\n",
    "\n",
    "    df_merge = pd.concat([df_pca.drop('Timepoint',axis=1), df_map],axis=1)\n",
    "    df_merge = df_merge[['PC1','PC2','Timepoint','HostSubjectId','WOMAC_P_Response']]\n",
    "    df_merge = df_merge.dropna(subset='PC1')\n",
    "\n",
    "    # save\n",
    "    df_merge.to_csv(path + 'outputs/Meta_' + g + '_adh/pcoa.tsv', sep='\\t') \n",
    "    \n",
    "    \n",
    "    plt.figure(figsize=(6,6))\n",
    "    plt.title(g + '_metabolome')\n",
    "    sns.scatterplot(data=df_pca, x='PC1',y='PC2',hue='Timepoint')\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/jobs18/pca_Timepoint_' + g + '.pdf')\n",
    "    plt.close()\n",
    "    \n",
    "df_merge.head()\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97ab1927-8204-4cd3-ad0d-895ae698d482",
   "metadata": {},
   "outputs": [],
   "source": [
    "# would need to take the distance matrix of all samples\n",
    "# and then take the subset list of distances (flatten the dm) for each pre to post sample and compare those via boxplot\n",
    "# between responders and nonresponders\n",
    "'''\n",
    "for g in ['stool','plasma','saliva']:\n",
    "    print(g)\n",
    "    df_map = type_to_df_map[g].copy()\n",
    "    df = gt_to_meta[g]['all'].copy()\n",
    "    \n",
    "    orig_idx = df.index\n",
    "    \n",
    "    df2 = df.copy()\n",
    "    df2.index = df2.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "\n",
    "    df2['WOMAC_P_Response'] = df2.index.map(df_dx.to_dict()['WOMAC_P_Response'])\n",
    "    \n",
    "    df2.index = orig_idx\n",
    "\n",
    "    for x in ['pre','post']:\n",
    "        print(x)\n",
    "        df = df2.copy()\n",
    "        df = df.dropna()\n",
    "        df = df[df.index.str.contains(x)]\n",
    "        \n",
    "        dx = df['WOMAC_P_Response'].values\n",
    "        df = df.drop('WOMAC_P_Response',axis=1)\n",
    "        \n",
    "        # df.index = df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "        # from scipy.spatial.distance import cdist\n",
    "        #dm = pd.DataFrame(cdist(gt_to_meta['stool']['Baseline'], gt_to_meta['stool']['After diet'], 'euclidean'))\n",
    "        \n",
    "        dm = pd.DataFrame(squareform(pdist(df)), columns=df.index.unique(), index=df.index.unique())\n",
    "        # takes in distance matrix\n",
    "        print(labtools.statistics.permanova_oneway(dm, levels=[1 if x == 'No response' else 0 for x in dx]))\n",
    "    \n",
    "        \n",
    "        from sklearn.manifold import MDS\n",
    "        \n",
    "        mds = MDS(n_components=2, dissimilarity='precomputed')\n",
    "        coordinates = mds.fit_transform(dm)\n",
    "        \n",
    "        df_pca = pd.DataFrame(coordinates, columns=['PC1','PC2'], index=df.index.unique())\n",
    "        df_pca['WOMAC_P_Response'] = dx\n",
    "    \n",
    "        df_merge = pd.concat([df_pca.drop('WOMAC_P_Response',axis=1), df_map],axis=1)\n",
    "        df_merge = df_merge[['PC1','PC2','Timepoint','HostSubjectId','WOMAC_P_Response']]\n",
    "        df_merge = df_merge.dropna(subset=['PC1','WOMAC_P_Response'])\n",
    "    \n",
    "        # save\n",
    "        df_merge.to_csv(path + 'outputs/Meta_' + g + '_adh_' + x + '/pcoa.tsv', sep='\\t') \n",
    "        \n",
    "        plt.figure(figsize=(6,6))\n",
    "        plt.title(g + '_' + x + '_metabolome')\n",
    "        sns.scatterplot(data=df_pca, x='PC1',y='PC2',hue='WOMAC_P_Response')\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/jobs18/pca_' + x + '_WOMAC_P_Response_' + g + '.pdf')\n",
    "        plt.close()\n",
    "\n",
    "df_merge.head()\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85644667-1a3b-4756-a371-3f016098cca4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# would need to take the distance matrix of all samples\n",
    "# and then take the subset list of distances (flatten the dm) for each pre to post sample and compare those via boxplot\n",
    "# between responders and nonresponders\n",
    "# responders and nr only\n",
    "'''\n",
    "for g in ['stool','plasma','saliva']:\n",
    "    print(g)\n",
    "    df = gt_to_meta[g]['all'].copy()\n",
    "    \n",
    "    orig_idx = df.index\n",
    "    \n",
    "    df2 = df.copy()\n",
    "    #df2.index = df2.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "\n",
    "    # df2['Timepoint'] = df2.index.map(df_dx.to_dict()['Timepoint'])\n",
    "    df_map = type_to_df_map[g].copy()\n",
    "\n",
    "    for x in ['Response', 'No response']:\n",
    "        print(x)\n",
    "        df = pd.concat([df2,df_map[['Timepoint','WOMAC_P_Response']]],axis=1)\n",
    "        df = df.dropna()\n",
    "        df = df[df['WOMAC_P_Response'] == x]\n",
    "        #df2.index = orig_idx\n",
    "        #df = df2.copy()\n",
    "        #df = df.dropna()\n",
    "        dx = df['Timepoint'].values\n",
    "        df = df.drop(['Timepoint','WOMAC_P_Response'],axis=1)\n",
    "        \n",
    "        # df.index = df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "        # from scipy.spatial.distance import cdist\n",
    "        #dm = pd.DataFrame(cdist(gt_to_meta['stool']['Baseline'], gt_to_meta['stool']['After diet'], 'euclidean'))\n",
    "        \n",
    "        dm = pd.DataFrame(squareform(pdist(df)), columns=df.index.unique(), index=df.index.unique())\n",
    "        # takes in distance matrix\n",
    "        print(permanova_oneway(dm, levels=[1 if (x == 'After diet') | (x == 'post') else 0 for x in dx]))\n",
    "    \n",
    "        \n",
    "        from sklearn.manifold import MDS\n",
    "        \n",
    "        mds = MDS(n_components=2, dissimilarity='precomputed')\n",
    "        coordinates = mds.fit_transform(dm)\n",
    "        \n",
    "        df_pca = pd.DataFrame(coordinates, columns=['PC1','PC2'], index=df.index.unique())\n",
    "        df_pca['Timepoint'] = dx\n",
    "    \n",
    "        df_merge = pd.concat([df_pca.drop('Timepoint',axis=1), df_map],axis=1)\n",
    "        df_merge = df_merge[['PC1','PC2','Timepoint','HostSubjectId','WOMAC_P_Response']]\n",
    "        df_merge = df_merge.dropna(subset='PC1')\n",
    "    \n",
    "        # save\n",
    "        if x == 'Response':\n",
    "            df_merge.to_csv(path + 'outputs/Meta_' + g + '_adh_response/pcoa.tsv', sep='\\t') \n",
    "        if x == 'No response':\n",
    "            df_merge.to_csv(path + 'outputs/Meta_' + g + '_adh_noresponse/pcoa.tsv', sep='\\t') \n",
    "        \n",
    "        \n",
    "        plt.figure(figsize=(6,6))\n",
    "        plt.title(g + '_' + x + '_metabolome')\n",
    "        sns.scatterplot(data=df_pca, x='PC1',y='PC2',hue='Timepoint')\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/jobs18/pca_' + x + '_Timepoint_' + g + '.pdf')\n",
    "        plt.close()\n",
    "\n",
    "df_merge.head()\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caed0a56-a43e-4873-a3db-9097292710bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PCA diet\n",
    "'''\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "# df = str_to_df['diet']\n",
    "df = df_diet\n",
    "\n",
    "# get merged df and drop na\n",
    "df_merge = pd.concat([df, df_dx], axis=1)\n",
    "df_merge = df_merge.dropna()\n",
    "\n",
    "# construct distance matrix from df\n",
    "df_data = df.loc[df_merge.index,:]\n",
    "dx = df_dx.loc[df_merge.index,:]['WOMAC_P_Response']\n",
    "dm = pd.DataFrame(squareform(pdist(df_data.iloc[:, 1:])), columns=df_data.index.unique(), index=df_data.index.unique())\n",
    "\n",
    "# takes in distance matrix\n",
    "print(labtools.statistics.permanova_oneway(dm, levels=[1 if x == 'No response' else 0 for x in dx]))\n",
    "\n",
    "npcs = 3\n",
    "pca = PCA(n_components=3, whiten=True)\n",
    "pca.fit(df_data.T)\n",
    "# Reformat and view results\n",
    "loadings = pd.DataFrame(pca.components_,\n",
    "                        index=['PC%s' % _ for _ in range(npcs)], \n",
    "                        columns=df_data.index).T\n",
    "\n",
    "df_merge = pd.concat([df_dx, df_data, loadings], axis=1)\n",
    "plt.title('Diet_PCA')\n",
    "sns.scatterplot(data=df_merge, x='PC1', y='PC2', hue='WOMAC_P_Response')\n",
    "sns.despine()\n",
    "plt.savefig(path + 'outputs/jobs06/PCA_diet_Response.pdf')\n",
    "plt.close() \n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8303f089-f331-4356-b6df-487184f8f8e3",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Paired Boxplots ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c51924d3-36bf-4608-861f-0c40fe634333",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotting\n",
    "'''\n",
    "df_pre = gt_to_otu['stool']['pre']\n",
    "df_post = gt_to_otu['stool']['post']\n",
    "df = gt_to_otu['stool']['all']\n",
    "# df = pd.concat([df,df_dx],axis=1)\n",
    "map = type_to_df_map['stool'].copy()\n",
    "df = pd.concat([df,map[['Timepoint','WOMAC_P_Response']]],axis=1)\n",
    "#map.index = map.index.map(lambda x: x.split('.p')[0].replace('-','')\n",
    "df = df.dropna()\n",
    "\n",
    "df['SubjectID'] = df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "df_vc = df['SubjectID'].value_counts() == 2\n",
    "df_vc = df_vc[df_vc == True]\n",
    "keep_id = df_vc.index.values\n",
    "df = df[df['SubjectID'].isin(keep_id)]\n",
    "\n",
    "taxa_int = []\n",
    "queries = ['Akker','g__Butyribacter','Anaerostipes','Limivivens']\n",
    "for q in queries:\n",
    "    print(q)\n",
    "    taxa = []\n",
    "    \n",
    "    for x in df.columns.values:\n",
    "        if q in x:\n",
    "            taxa.append(x)\n",
    "            taxa_int.append(x)\n",
    "    print(len(taxa))\n",
    "    print(taxa)\n",
    "    plt.figure(figsize=(6,6))\n",
    "    sns.swarmplot(x=\"Timepoint\", y=taxa[0],\n",
    "                hue=\"WOMAC_P_Response\", palette=[\"m\", \"g\"],\n",
    "                data=df,dodge=True)\n",
    "    sns.boxplot(x=\"Timepoint\", y=taxa[0],\n",
    "                hue=\"WOMAC_P_Response\", palette=[\"m\", \"g\"], boxprops={'alpha': 0.1},\n",
    "                data=df)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    # plt.xlabel(x.split('f__')[-1])\n",
    "    plt.savefig(path + 'outputs/jobs18/' + taxa[0].split('f__')[-1] + '.pdf')\n",
    "    plt.close()\n",
    "\n",
    "# sns.despine(offset=10, trim=True)\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d55fac63-a786-4d41-96f9-3233c48556f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotting\n",
    "'''\n",
    "df = gt_to_meta['stool']['all']\n",
    "\n",
    "map = type_to_df_map['stool'].copy()\n",
    "df = pd.concat([df,map[['Timepoint','WOMAC_P_Response']]],axis=1)\n",
    "\n",
    "df = df.dropna()\n",
    "\n",
    "df['SubjectID'] = df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "df_vc = df['SubjectID'].value_counts() == 2\n",
    "df_vc = df_vc[df_vc == True]\n",
    "keep_id = df_vc.index.values\n",
    "df = df[df['SubjectID'].isin(keep_id)]\n",
    "\n",
    "meta_int = []\n",
    "queries = ['H_Hydroxydodecanoic','H_Hydroxydodecanoic','H_Omega_Hydroxydodecanoate', 'H_Biotin','H_Hydroxydecanoate']\n",
    "for q in queries:\n",
    "    print(q)\n",
    "    var = []\n",
    "    \n",
    "    for x in df.columns.values:\n",
    "        if q in x:\n",
    "            var.append(x)\n",
    "            meta_int.append(x)\n",
    "    print(len(var))\n",
    "    print(var)\n",
    "    plt.figure(figsize=(6,6))\n",
    "    sns.swarmplot(x=\"Timepoint\", y=var[0],\n",
    "                hue=\"WOMAC_P_Response\", palette=[\"m\", \"g\"],\n",
    "                data=df,dodge=True)\n",
    "    sns.boxplot(x=\"Timepoint\", y=var[0],\n",
    "                hue=\"WOMAC_P_Response\", palette=[\"m\", \"g\"], boxprops={'alpha': 0.1},\n",
    "                data=df)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    # plt.xlabel(x.split('f__')[-1])\n",
    "    plt.savefig(path + 'outputs/jobs18/stool_pairedbox_' + var[0] + '.pdf')\n",
    "    plt.close()\n",
    "\n",
    "# sns.despine(offset=10, trim=True)\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ba60f29-54e3-40eb-a0aa-34deafb77957",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotting\n",
    "'''\n",
    "df = gt_to_meta['plasma']['all']\n",
    "\n",
    "map = type_to_df_map['plasma'].copy()\n",
    "df = pd.concat([df,map[['Timepoint','WOMAC_P_Response']]],axis=1)\n",
    "\n",
    "df = df.dropna()\n",
    "\n",
    "df['SubjectID'] = df.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "df_vc = df['SubjectID'].value_counts() == 2\n",
    "df_vc = df_vc[df_vc == True]\n",
    "keep_id = df_vc.index.values\n",
    "df = df[df['SubjectID'].isin(keep_id)]\n",
    "\n",
    "meta_int = []\n",
    "queries = ['H_Theobromine', 'H_Paraxanthine', 'H_Lenticin','H_N-Acetylmuramic-Acid']\n",
    "for q in queries:\n",
    "    print(q)\n",
    "    var = []\n",
    "    \n",
    "    for x in df.columns.values:\n",
    "        if q in x:\n",
    "            var.append(x)\n",
    "            meta_int.append(x)\n",
    "    print(len(var))\n",
    "    print(var)\n",
    "    plt.figure(figsize=(6,6))\n",
    "    sns.swarmplot(x=\"Timepoint\", y=var[0],\n",
    "                hue=\"WOMAC_P_Response\", palette=[\"m\", \"g\"],\n",
    "                data=df,dodge=True)\n",
    "    sns.boxplot(x=\"Timepoint\", y=var[0],\n",
    "                hue=\"WOMAC_P_Response\", palette=[\"m\", \"g\"], boxprops={'alpha': 0.1},\n",
    "                data=df)\n",
    "    sns.despine()\n",
    "    plt.tight_layout()\n",
    "    # plt.xlabel(x.split('f__')[-1])\n",
    "    plt.savefig(path + 'outputs/jobs18/plasma_pairedbox_' + var[0] + '.pdf')\n",
    "    plt.close()\n",
    "\n",
    "# sns.despine(offset=10, trim=True)\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cbf78bc-4ccf-4fc5-8231-b7506b740b13",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Scatterplot ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cb25185-ddec-489f-bc37-b9ccada37da7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlation matrices\n",
    "# differences in taxa versus all points?\n",
    "'''\n",
    "df_otu  = gt_to_otu['stool']['diff']\n",
    "df_meta = gt_to_meta['stool']['diff']\n",
    "\n",
    "df_merge = pd.concat([df_otu, df_meta, df_dx], axis=1)\n",
    "df_merge = df_merge.dropna()\n",
    "\n",
    "\n",
    "for x in taxa_int:\n",
    "    for y in meta_int:\n",
    "        r,p = scipy.stats.spearmanr(df_merge[x].values, df_merge[y].values, nan_policy='omit')\n",
    "        if p < 0.05:\n",
    "            print(x,y)\n",
    "            print(p)\n",
    "        plt.figure(figsize=(6,6))\n",
    "        sns.lmplot(data=df_merge, x=x,y=y,hue='WOMAC_P_Response')\n",
    "        plt.savefig(path + 'outputs/jobs18/scatter_' + x.split('_')[-1] + '_' + y + '.pdf') \n",
    "        plt.close()\n",
    "        \n",
    "\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05ed278e-e0a7-4ea4-a2dd-f95350da1e8c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### CUTIE ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d51feb2-1356-4615-81a9-ae357b3f93e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CUTIE\n",
    "# In POST samples only\n",
    "# Color by Resp and Nonresp\n",
    "'''\n",
    "for g in ['saliva', 'stool', 'plasma']:\n",
    "    # grab df map with diagnses etc. \n",
    "    df_map = type_to_df_map[g].set_index('Together')\n",
    "    df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "\n",
    "    if g == 'stool' or g == 'saliva':\n",
    "        df_taxa_post = gt_to_otu[g]['post']\n",
    "        df_taxa_post.to_csv(path + 'inputs/cutie_df_taxa_post_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "        df_taxa_pre = gt_to_otu[g]['pre']\n",
    "        df_taxa_pre.to_csv(path + 'inputs/cutie_df_taxa_pre_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "\n",
    "        df_taxa_pre.index = df_taxa_pre.index.map(lambda x: x.split('.')[0].replace('-',''))\n",
    "        df_taxa_post.index = df_taxa_post.index.map(lambda x: x.split('.')[0].replace('-',''))\n",
    "\n",
    "        df_taxa_diff = df_taxa_post - df_taxa_pre\n",
    "        df_taxa_diff.to_csv(path + 'inputs/cutie_df_taxa_diff_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "        df_taxa_diff = pd.concat([df_taxa_diff,df_map], axis=1)\n",
    "        # get resp only\n",
    "        df_taxa_diff_resp = df_taxa_diff[df_taxa_diff['WOMAC_P_Response'] == 'Response']\n",
    "        df_taxa_diff_nr = df_taxa_diff[df_taxa_diff['WOMAC_P_Response'] == 'No response']\n",
    "        df_taxa_diff_resp.to_csv(path + 'inputs/cutie_df_taxa_diff_resp_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "        df_taxa_diff_nr.to_csv(path + 'inputs/cutie_df_taxa_diff_nr_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "        #df_abun = gt_to_otu[g]['post'].copy() #, gt_to_path[g]['post']],axis=1)\n",
    "        #df_abun = pd.concat([gt_to_otu[g]['post'], gt_to_path[g]['post']],axis=1)\n",
    "        #df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "        #df_abun.to_csv(path + 'inputs/cutie_df_abun_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "    \n",
    "    df_meta_post = gt_to_meta[g]['After diet']\n",
    "    df_meta_pre = gt_to_meta[g]['Baseline']\n",
    "    df_meta_diff = df_meta_post - df_meta_pre\n",
    "        \n",
    "    df_meta_post = pd.concat([df_meta_post,df_map], axis=1)\n",
    "    df_meta_post.to_csv(path + 'inputs/cutie_df_meta_post_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "    df_meta_pre = pd.concat([df_meta_pre,df_map], axis=1)\n",
    "    df_meta_pre.to_csv(path + 'inputs/cutie_df_meta_pre_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "    df_meta_diff = pd.concat([df_meta_diff,df_map], axis=1)\n",
    "    df_meta_diff.to_csv(path + 'inputs/cutie_df_meta_diff_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "    # get resp only\n",
    "    df_meta_diff_resp = df_meta_diff[df_meta_diff['WOMAC_P_Response'] == 'Response']\n",
    "    df_meta_diff_nr = df_meta_diff[df_meta_diff['WOMAC_P_Response'] == 'No response']\n",
    "    df_meta_diff_resp.to_csv(path + 'inputs/cutie_df_meta_diff_resp_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "    df_meta_diff_nr.to_csv(path + 'inputs/cutie_df_meta_diff_nr_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "\n",
    "\n",
    "df_meta_diff.head()\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5d49a9b-36b5-4d66-919b-6183f3433d6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlation analyses\n",
    "# jobs25 is plasma meta diff with stool meta diff\n",
    "# var1 is stool, var2 is plasma\n",
    "\n",
    "comp_to_job = {\n",
    "    'diff_diff': 'jobs25',\n",
    "    'pre_pre': 'jobs26',\n",
    "    'post_post': 'jobs27'\n",
    "}\n",
    "'''\n",
    "for c in comp_to_job:\n",
    "    print(c)\n",
    "\n",
    "    df = pd.read_csv(path + 'outputs/' + comp_to_job[c] + '/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "    \n",
    "    # get number of total corr, TP and FP\n",
    "    print('Total Corr: ' + str(len(df[df['pvalues'] != np.nan])))\n",
    "    print('TP: ' + str(len(df[df['class'] == 'TP'])))\n",
    "    print('FP: ' + str(len(df[df['class'] == 'FP'])))\n",
    "\n",
    "    # strength of matching\n",
    "    df_match = df[df['var1'] == df['var2']]\n",
    "    df_match = df_match[df_match['pvalues'] < 0.05]\n",
    "\n",
    "    print('Match length: ' + str(len(df_match)))\n",
    "\n",
    "# there's 29 plasma and 88 metabolites, only 14 are in common\n",
    "# print(len(set(gt_to_meta['plasma']['Baseline'].columns.values).intersection(set(gt_to_meta['stool']['Baseline']))))\n",
    "df_match.head()    \n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "807adee0-f229-42fb-9679-d627fb4a0fe5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Integrative plot ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51a50637-3f02-46d0-9671-6af857068029",
   "metadata": {},
   "outputs": [],
   "source": [
    "# need dataframe where each row is a pair of var1, var2, and the cols are\n",
    "# log10p in taxa R vs NR  - gdt_to_res\n",
    "# log10p in metabolite R vs NR - gdt_to_res\n",
    "# correlation difference - fisher's Z from chat gpt, need to implement\n",
    "# color (pos neg of original resp corr) and size also propoprtional to above DIFFERENCE\n",
    "# Special tag for bile acids of interest\n",
    "# \n",
    "'''\n",
    "# start with arbitrary meta taxa df\n",
    "df = pd.read_csv(path + 'outputs/jobs28/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "df = df[df['var1'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "df = df[df['var2'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "df = df[['var1','var2']]\n",
    "\n",
    "# grab pvals\n",
    "var1_ps = []\n",
    "df_res = gdt_to_res_pvp['stool']['meta']['Response']\n",
    "meta_to_logp = df_res.set_index('meta')['logp'].to_dict()\n",
    "\n",
    "for x in df.var1.values:\n",
    "    var1_ps.append(meta_to_logp[x])\n",
    "\n",
    "# repeat for taxa\n",
    "var2_ps = []\n",
    "df_res = gdt_to_res_pvp['stool']['otu']['Response']\n",
    "taxa_to_logp = df_res.set_index('otu')['logp'].to_dict()\n",
    "\n",
    "for x in df.var2.values:\n",
    "    x = read_taxa(x,';')\n",
    "    var2_ps.append(taxa_to_logp[x])\n",
    "\n",
    "# var2_ps\n",
    "df['metabolite_log10p'] = var1_ps\n",
    "df['taxa_log10p'] = var2_ps\n",
    "\n",
    "# sample sizes from observing plots\n",
    "n_nr = 5\n",
    "n_resp = 7\n",
    "\n",
    "# from chatGPT! \n",
    "def fishers_z(r1,r2,n1,n2):\n",
    "    # Calculate Fisher's z-transform for each correlation\n",
    "    z1 = np.arctanh(r1) \n",
    "    z2 = np.arctanh(r2) \n",
    "    \n",
    "    # Calculate the standard error of the difference between z1 and z2\n",
    "    se = np.sqrt(1/(n1-3) + 1/(n2-3))\n",
    "    \n",
    "    # Calculate the z-statistic\n",
    "    z = (z1 - z2) / se \n",
    "    \n",
    "    # Calculate the p-value\n",
    "    p = 2 * (1 - scipy.stats.norm.cdf(abs(z)))\n",
    "\n",
    "    return z,p\n",
    "\n",
    "\n",
    "# bring in the correlations \n",
    "df_resp = pd.read_csv(path + 'outputs/jobs28/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "df_resp = df_resp[df_resp['var1'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "df_resp = df_resp[df_resp['var2'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "df_resp = df_resp[['var1','var2','correlations']]\n",
    "df_resp = df_resp.rename(columns={'correlations': 'correlations_resp'})\n",
    "\n",
    "df_nr = pd.read_csv(path + 'outputs/jobs29/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "df_nr = df_nr[df_nr['var1'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "df_nr = df_nr[df_nr['var2'] != 'WOMAC_P_Response'] # gets rid of the WOMAC_P_Response mirror'd col\n",
    "df_nr = df_nr[['var1','var2','correlations']]\n",
    "df_nr = df_nr.rename(columns={'correlations': 'correlations_nr'})\n",
    "\n",
    "# assuming the rows are in proper consistent order, which they seem to be\n",
    "\n",
    "df_merge = pd.concat([df,df_resp.drop(['var1','var2'],axis=1), df_nr.drop(['var1','var2'],axis=1)],axis=1)\n",
    "\n",
    "# drop na\n",
    "# interesting design choice whether to drop correlations that are NA in only one group, TBD\n",
    "df_merge = df_merge.dropna(subset=['taxa_log10p','metabolite_log10p','correlations_resp','correlations_nr'])\n",
    "\n",
    "for index, row in df_merge.iterrows():\n",
    "    z,p = fishers_z(row.loc['correlations_resp'], row.loc['correlations_nr'], n_resp, n_nr)\n",
    "    df_merge.at[index, 'fishersz_diff'] = z\n",
    "    df_merge.at[index, 'fz_diff_p'] = p\n",
    "    \n",
    "# filter on sig    \n",
    "# df_merge = df_merge[df_merge['fz_diff_p'] < 0.2]\n",
    "df_merge = df_merge.sort_values(by='fz_diff_p',ascending=True)\n",
    "reject, pvals_corrected, _, _ = multitest.multipletests(df_merge.fz_diff_p, method='fdr_bh')\n",
    "df_merge['fdr'] = pvals_corrected\n",
    "\n",
    "df_merge = df_merge[df_merge['fdr'] < 0.05]\n",
    "\n",
    "# replace inf with 5 and -inf with -5\n",
    "df_merge['fishersz_diff'] = df_merge['fishersz_diff'].replace(np.inf, 5)\n",
    "df_merge['fishersz_diff'] = df_merge['fishersz_diff'].replace(-np.inf, -5)\n",
    "\n",
    "# flag for bile acids etc.\n",
    "query_ba = ['Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid']\n",
    "\n",
    "# int_ba1 = ['h_' + x for x in int_ba0] + ['H_' + x for x in int_ba0]\n",
    "\n",
    "int_ba = []\n",
    "\n",
    "for x in df_meta.columns:\n",
    "    for t in query_ba:\n",
    "        if t in x:\n",
    "            int_ba.append(x)\n",
    "\n",
    "\n",
    "int_taxa = []\n",
    "query_taxa = ['Faecalibacterium', 'Akkermansia','Alistipes',\n",
    "              'Anaerostipes',\n",
    "              'Limivivens',\n",
    "              'Phascolarctobacterium','Rikenell', 'Acetati', 'Ruminococcus']\n",
    "for x in df_otu.columns:\n",
    "    for t in query_taxa:\n",
    "        if t in x:\n",
    "            int_taxa.append(x)\n",
    "\n",
    "\n",
    "df_merge.loc[(df_merge['var1'].isin(int_ba)) | (df_merge['var2'].isin(int_taxa)), 'Interest'] = 'Yes'\n",
    "\n",
    "\n",
    "df_merge\n",
    "'''\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "643b1518-8e68-46f5-8c35-9d78cb472f91",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot \n",
    "'''\n",
    "df_meta_diff = pd.read_csv(path + 'inputs/cutie_df_meta_diff_stool.tsv', sep='\\t', index_col=0)#'SampleID')\n",
    "df_taxa_diff = pd.read_csv(path + 'inputs/cutie_df_taxa_diff_stool.tsv', sep='\\t', index_col=0)#, index_label='SampleID')\n",
    "df_taxa_diff.index = df_taxa_diff.index.map(lambda x: x.split('.')[0].replace('-',''))\n",
    "\n",
    "g='stool'\n",
    "df_map = type_to_df_map[g].set_index('Together')\n",
    "df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "df_map\n",
    "\n",
    "# df_merge_diff = pd.concat([df_meta_diff, df_taxa_diff, df_map],axis=1)#,axis=1)\n",
    "df_merge_diff = pd.concat([df_meta_diff, df_taxa_diff],axis=1)#,axis=1)\n",
    "df_merge_diff = df_merge_diff.dropna()#(subset='WOMAC_P_Response')\n",
    "\n",
    "df = df_merge.copy()\n",
    "# df = df[df['metabolite_log10p'] < -1]\n",
    "df = df[df['Interest'] == 'Yes']\n",
    "for i in range(len(df)):\n",
    "    row = df.iloc[i,:]\n",
    "    var1 = row['var1']\n",
    "    var2 = row['var2']\n",
    "\n",
    "    plt.figure(figsize=(3,3))\n",
    "    sns.lmplot(data=df_merge_diff, x=var1, y=var2, hue='WOMAC_P_Response')\n",
    "    plt.ylabel(read_taxa(var2))               \n",
    "    plt.tight_layout()\n",
    "    sns.despine()    \n",
    "    # plt.savefig(path + 'outputs/jobs18/corrs_stool_diff/' + 'corr_meta10p_' + var1.replace('/','_') + '_' + read_taxa(var2) + '.pdf')\n",
    "    plt.savefig(path + 'outputs/jobs18/corrs_stool_diff_int/' + 'corr_' + var1.replace('/','_') + '_' + read_taxa(var2) + '.pdf')\n",
    "\n",
    "    plt.close()\n",
    "\n",
    "df_merge_diff\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a5f61f1-0309-49e5-8d25-7ba72d4c0da2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Mediation ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a0d3480-2510-4f3b-aa38-068db1ffe1cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# play testing mediation\n",
    "g = 'stool'\n",
    "df_M = pd.read_csv(path + 'inputs/cutie_df_meta_stool.tsv', sep='\\t', index_col=0)\n",
    "df_M.iloc[:,:-1].to_csv(path + 'inputs/mediation_meta_stool_post_test.txt', sep='\\t')\n",
    "\n",
    "df_Y = df_meta_paired[['VAS_Pt_diff', 'WOMAC_pain_diff']]\n",
    "df_Y.to_csv(path + 'inputs/mediation_WMdiff_test.txt', sep='\\t')\n",
    "\n",
    "df_X = pd.read_csv(path + 'inputs/cutie_df_taxa_stool.tsv', sep='\\t', index_col=0)\n",
    "df_X.index = df_X.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "df_X.to_csv(path + 'inputs/mediation_taxa_stool_post_test.txt', sep='\\t')\n",
    "df_X.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b691f31c-b12b-4bca-b4ab-90189c2984a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for corr heatmaps \n",
    "for g in ['saliva','stool']:\n",
    "    df_abun = pd.concat([gt_to_otu[g]['post'], gt_to_path[g]['post']],axis=1)\n",
    "    df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_map = type_to_df_map[g].set_index('Together')\n",
    "    df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_meta = pd.concat([gt_to_meta[g]['After diet'],df_map], axis=1)\n",
    "\n",
    "    df_quant_R = df_abun.copy()\n",
    "    #df_quant_R = pd.concat([df_abun, df_meta],axis=1)# .copy() #([df_quant, df_asv100, df_path_top100], axis=1)\n",
    "    # df_quant_R = df_quant_R.drop(['WOMAC_P_Response'],axis=1)\n",
    "    # df_quant_R.to_csv(path + 'inputs/df_quant_' + g + '.tsv', sep='\\t')\n",
    "    df_quant_R = df_quant_R.T\n",
    "    # this is in the order they were entered i.e. \n",
    "    # [df_olink, df_fa, df_acpa_fecal, df_acpa_plasma, df_plasma, df_rbfa, df_mb]\n",
    "     # ['rbfa']*len(df_rbfa.columns) + \\  \n",
    "    df_quant_R['var_type'] = ['Taxa']*len(gt_to_otu[g]['post'].columns) + \\\n",
    "                             ['Pathways']*len(gt_to_path[g]['post'].columns) # + \\\n",
    "                             # ['Metabolite']*len(gt_to_meta[g]['After diet'].columns)\n",
    "    # df_quant_R = df_quant_R[~df_quant_R['var_type'].isin(['Metagenomic_ASVs','Metagenomic_Pathways'])]\n",
    "    df_quant_R = df_quant_R['var_type'].reset_index()\n",
    "    df_quant_R['index'] = df_quant_R['index'].str.lower()\n",
    "    df_quant_R = df_quant_R.sort_values(by='index',ascending=True)\n",
    "    df_quant_R.to_csv(path + 'inputs/df_quant_R_corr_labels_' + g + '.tsv', sep='\\t')\n",
    "    print(len(df_quant_R))\n",
    "df_quant_R.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1bfb6c5-3ef9-4a9f-aa21-03a8f3d4194f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4 way paired boxplot trial\n",
    "'''\n",
    "g = 'saliva'\n",
    "df_alpha = g_to_dfd[g + '_adh']['alpha']\n",
    "\n",
    "df_map = type_to_df_map[g].copy()\n",
    "df_map = df_map.set_index('Together')\n",
    "df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "\n",
    "df_merge = pd.concat([df_alpha, df_map], axis=1)\n",
    "df_merge = df_merge.dropna()\n",
    "\n",
    "sns.boxplot(x='Timepoint', y='shannon_entropy',\n",
    "            hue='WOMAC_P_Response', palette=[\"r\", \"g\"],\n",
    "            data=df_merge)\n",
    "sns.despine()\n",
    "\n",
    "# test if each sub correlation is sig\n",
    "m = 'H_Nonhydroxylated_bile_acid'\n",
    "y = 'WOMAC_pain_diff'\n",
    "y = 'VAS_Pt_diff'\n",
    "x = 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;__'\n",
    "\n",
    "df = df_plot[[x,y,m]]\n",
    "\n",
    "df_corr = df.dropna()\n",
    "print('Y on X')\n",
    "print(scipy.stats.pearsonr(df_corr[y].values, df_corr[x].values))\n",
    "print('M on X')\n",
    "print(scipy.stats.pearsonr(df_corr[m].values, df_corr[x].values))\n",
    "\n",
    "df['meta'] = df[m].apply(lambda x: 'high' if -1.25 <= x <= 0.25 else 'low')\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63265521-846f-4759-b807-c6e822612aa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "# do each linear model separately\n",
    "df_corr = df[df['meta'] == 'low']\n",
    "df_corr = df_corr.dropna()\n",
    "print(scipy.stats.pearsonr(df_corr[y].values, df_corr[x].values))\n",
    "\n",
    "df_corr = df[df['meta'] == 'high']\n",
    "df_corr = df_corr.dropna()\n",
    "print(scipy.stats.pearsonr(df_corr[y].values, df_corr[x].values))\n",
    "# df_corr\n",
    "df.columns = ['Lachno','Pain','H_Nonhydroxylated_BA', 'meta']\n",
    "df.to_csv(path + 'inputs/df_med_R.tsv', sep='\\t')\n",
    "df\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a93620d-aa39-47e6-8c20-b44c82d19abb",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### DFs for R ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec28d742-b98f-4edf-b166-9aca33830cd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlation of diet pro or anti inflam with diversity\n",
    "int_var = ['total_antiinflammatory','anti_inflammatory','inflammatory','pro_inflammatory']\n",
    "df_d = df_diet[int_var]\n",
    "\n",
    "df_y = type_to_df_map['stool']\n",
    "# df_y = df_y[['VAS_Pt','VAS_response','WOMAC_pain','WOMAC_P_Response']]\n",
    "df_y = df_y[['WOMAC_pain','Timepoint','anti_inflammatory']]#,'WOMAC_P_Response']]\n",
    "df_y = df_y.dropna()\n",
    "df_pre  = df_y[df_y['Timepoint'] == 'pre'].drop('Timepoint',axis=1)\n",
    "df_pre.index = df_pre.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "df_post  = df_y[df_y['Timepoint'] == 'post'].drop('Timepoint',axis=1)\n",
    "df_post.index = df_post.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "\n",
    "# df_diff\n",
    "df_diff = df_post - df_pre\n",
    "df_diff.columns = df_diff.columns.map(lambda x: x + '_diff')\n",
    "df_diff = df_diff.dropna()\n",
    "df_diff.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccce3c1c-b1ad-4c30-b54d-e8ca9b87767d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# need particular DF for taxa\n",
    "# for anaerostipes,\n",
    "g = 'stool'\n",
    "y='Lachnospiraceae_Anaerostipes'\n",
    "#x='Timepoint'\n",
    "#hue='WOMAC_P_Response'\n",
    "#order=['pre','post']\n",
    "#hue_order=['No response','Response']\n",
    "\n",
    "'''\n",
    "hue='Timepoint'\n",
    "x='WOMAC_P_Response'\n",
    "hue_order=['pre','post']\n",
    "order=['No response','Response']\n",
    "\n",
    "# copy dfs and format\n",
    "df_map = type_to_df_map[g].copy()\n",
    "df_map = df_map[df_map.duplicated(subset='HostSubjectId', keep=False)]\n",
    "\n",
    "# format taxa table\n",
    "df_taxa = gt_to_otu[g]['all'].copy()\n",
    "df_taxa.columns = df_taxa.columns.map(lambda x: x.split('f__')[-1].replace(';g_',''))\n",
    "\n",
    "# merge\n",
    "df_merge = pd.concat([df_map, df_taxa],axis=1)\n",
    "\n",
    "# subset on interest\n",
    "keep = ['WOMAC_P_Response','Timepoint','WOMAC_pain','HostSubjectId']\n",
    "taxa = ['Lachnospiraceae_Limivivens','Lachnospiraceae_Anaerostipes','Lachnospiraceae_Butyribacter','Akkermansiaceae_Akkermansia']\n",
    "\n",
    "v_to_v ={\n",
    "'WOMAC_pain': 'WOMAC_Pain',#_Improvement',\n",
    "'anti_inflammatory': 'Anti_Inflammatory_Diet_Score',#_Change',\n",
    "'plasma_meta_alpha_diff': 'Plasma_Metabolite_Alpha_Diversity',#_Difference',\n",
    "#'stool_meta_alpha_diff': 'Gut_Metabolite_Alpha_Diversity',#Difference',\n",
    "'stool_taxa_alpha_diff': 'Gut_Microbiome_Alpha_Diversity',#Difference',\n",
    "'H_Tetrahydroxylated_bile_acid_diff_stool': 'Plasma_Tetrahydroxylated_Bile_Acid',\n",
    "'H_Glycocholic_acid_diff_plasma': 'Plasma_Glycocholic_Acid', \n",
    "'H_Glycochenodeoxycholate_diff_plasma':'Plasma_Glycochenodeoxycholate',\n",
    "'H_Deoxycholate_diff_plasma':'Plasma_Deoxycholate',\n",
    "'d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens': 'Gut_Lachnospiraceae_Limivivens',\n",
    "'d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Anaerostipes': 'Gut_Lachnospiraceae_Anaerostipes',\n",
    "'d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Butyribacter': 'Gut_Lachnospiraceae_Butyribacter',\n",
    "'d__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia': 'Gut_Akkermansia'\n",
    "\n",
    "}\n",
    "\n",
    "df_merge = df_merge.rename(v_to_v)\n",
    "df_merge = df_merge.loc[:,keep+taxa]\n",
    "\n",
    "\n",
    "# dropna\n",
    "df_merge = df_merge.dropna()\n",
    "\n",
    "# test plot\n",
    "plt.figure(figsize=(4,3))\n",
    "sns.boxplot(data=df_merge,x=x,y=y,order=order,hue=hue,hue_order=hue_order,fill=False)\n",
    "sns.stripplot(data=df_merge,x=x,y=y,hue=hue,hue_order=hue_order,dodge=True)\n",
    "sns.despine()\n",
    "# plt.tight_layout()\n",
    "\n",
    "print(len(df_merge))\n",
    "df_merge.Timepoint = df_merge.Timepoint.map(lambda x: x.capitalize())\n",
    "df_merge.to_csv(path + 'inputs/df_R_boxplots.tsv', sep='\\t')\n",
    "df_merge.head()\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "142c0448-b438-4f20-bf1c-862fdabc6904",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the diff versoin\n",
    "'''\n",
    "df_taxa = gt_to_otu[g]['diff'].copy()\n",
    "df_taxa.columns = df_taxa.columns.map(lambda x: x.split('f__')[-1].replace(';g_',''))\n",
    "\n",
    "df_map = type_to_df_map[g].copy()\n",
    "# keep duplicated samples only and then drop\n",
    "df_map = df_map[df_map.duplicated(subset='HostSubjectId', keep=False)]\n",
    "df_map = df_map.drop_duplicates(subset='HostSubjectId')\n",
    "df_map.index = df_map.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "\n",
    "# merge\n",
    "df_merge = pd.concat([df_map, df_taxa],axis=1)\n",
    "\n",
    "# subset on interest\n",
    "keep = ['WOMAC_P_Response','Timepoint']\n",
    "taxa = [y]\n",
    "df_merge = df_merge.loc[:,keep+taxa]\n",
    "\n",
    "# dropna\n",
    "df_merge = df_merge.dropna()\n",
    "\n",
    "# merge again\n",
    "df_merge = pd.concat([df_merge,df_diff],axis=1)\n",
    "\n",
    "# test plot\n",
    "plt.figure(figsize=(4,3))\n",
    "sns.boxplot(data=df_merge,x=x,y=y,order=order,hue=x,fill=False)\n",
    "sns.stripplot(data=df_merge,x=x,y=y,hue=x,dodge=False)\n",
    "sns.despine()\n",
    "# plt.tight_layout()\n",
    "\n",
    "print(len(df_merge))\n",
    "df_merge.head()\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4e2c2dd-4c6f-490d-83ff-a2d8823595a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for taxa barplots\n",
    "# gt_to_otu[\n",
    "\n",
    "# taxa barplot\n",
    "L_to_start_str = {\n",
    "    'L2': '|c  ', # for phylum you want to split on class\n",
    "    'L6': ';g__'\n",
    "}\n",
    "L_to_end_str = {\n",
    "    'L2': '|p  ', \n",
    "    'L6': ''\n",
    "}\n",
    "\n",
    "g = 'stool'\n",
    "t = 'pre'\n",
    "for g in ['stool','saliva']:\n",
    "    for t in ['pre','post']:\n",
    "        df_taxa = gt_to_otu[g][t].copy()\n",
    "        df_merge = pd.concat([df_taxa],axis=1)\n",
    "        #df_merge = df_merge[df_merge['Visit'] == v]\n",
    "        #df_merge = df_merge.dropna(subset=df_taxa.columns[0]) \n",
    "        keep = df_merge.index\n",
    "        \n",
    "        # get string to split on\n",
    "        # ss,es = L_to_start_str[l], L_to_end_str[l]\n",
    "        \n",
    "        # get taxa data\n",
    "        df = df_taxa.copy().loc[keep,:]\n",
    "        # df.columns = df.columns.map(lambda x: x.split(ss)[0])\n",
    "        \n",
    "        # normally this should happen AFTER the groupby but there's silly inconsistences\n",
    "        # like bacillota === firmicutes but it's not consistent in the mapping\n",
    "        \n",
    "        # split string on ending '|'\n",
    "        df.columns = df.columns.map(lambda x: x.split(';g__')[-1])\n",
    "        df = df.groupby(df.columns, axis=1).sum()\n",
    "        \n",
    "        # set name of index\n",
    "        df.index.name = 'Taxon'\n",
    "        \n",
    "        # transpose\n",
    "        df = df.T\n",
    "        df = df.rename({'':'Unclassified'})\n",
    "        \n",
    "        # check that no dup rows\n",
    "        if np.max(df.index.value_counts()) > 1:\n",
    "            print('dup row')\n",
    "\n",
    "        # annoying specific renames\n",
    "        df = df.rename({'d__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Enterobacterales_A_737866;f__Pasteurellaceae;__': 'Uncl Pasteurellaceae'})\n",
    "        \n",
    "        # save output\n",
    "        df.to_csv(path + 'outputs/jobs35/df_q2R_' + g + '_' + t + '.tsv', sep='\\t', index_label='Taxon')\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "565df936-19b8-4ece-b152-1afcf37098a1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "384908e5-ad03-4e90-b9d2-95a1a7b5cb15",
   "metadata": {},
   "source": [
    "### New section ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f58fe742-ef3b-494b-8068-8bf3eefe1f21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>WOMAC_P_Response</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>pre</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>pre</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD007</th>\n",
       "      <td>pre</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD008</th>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD009</th>\n",
       "      <td>pre</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD012</th>\n",
       "      <td>pre</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD013</th>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD017</th>\n",
       "      <td>pre</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD002</th>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD004</th>\n",
       "      <td>pre</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD008</th>\n",
       "      <td>pre</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD010</th>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD011</th>\n",
       "      <td>pre</td>\n",
       "      <td>Response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD012</th>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD014</th>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Timepoint WOMAC_P_Response\n",
       "Study_ID                           \n",
       "OAD001         pre      No response\n",
       "OAD003         pre      No response\n",
       "OAD004         pre      No response\n",
       "OAD005         pre              NaN\n",
       "OAD006         pre         Response\n",
       "OAD007         pre         Response\n",
       "OAD008         pre      No response\n",
       "OAD009         pre         Response\n",
       "OAD012         pre         Response\n",
       "OAD013         pre      No response\n",
       "OAD017         pre         Response\n",
       "VAOAD001       pre      No response\n",
       "VAOAD002       pre      No response\n",
       "VAOAD004       pre         Response\n",
       "VAOAD008       pre         Response\n",
       "VAOAD009       pre      No response\n",
       "VAOAD010       pre      No response\n",
       "VAOAD011       pre         Response\n",
       "VAOAD012       pre      No response\n",
       "VAOAD014       pre      No response"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_otu = gt_to_otu['stool']['all']\n",
    "df_map\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1e5deae-a915-4348-827c-e4b351bef8b8",
   "metadata": {},
   "source": [
    "###### Triple Correlations ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "79bbbea3-cb4b-4eb7-940a-c102001314bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WOMAC_pain</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>-1.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>-1.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>-2.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD007</th>\n",
       "      <td>-6.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           WOMAC_pain\n",
       "#SampleID            \n",
       "OAD001           -1.5\n",
       "OAD003           -1.5\n",
       "OAD004            1.0\n",
       "OAD006           -2.5\n",
       "OAD007           -6.0"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# correlation of diet pro or anti inflam with diversity\n",
    "# \n",
    "df_dx = type_to_df_map['stool'].drop_duplicates(subset='HostSubjectId')['WOMAC_P_Response']\n",
    "\n",
    "int_var = ['total_antiinflammatory','anti_inflammatory','inflammatory','pro_inflammatory']\n",
    "df_d = df_diet[int_var]\n",
    "\n",
    "df_y = type_to_df_map['stool']\n",
    "# df_y = df_y[['VAS_Pt','VAS_response','WOMAC_pain','WOMAC_P_Response']]\n",
    "df_y = df_y[['WOMAC_pain','Timepoint']]#,'WOMAC_P_Response']]\n",
    "df_y = df_y.dropna()\n",
    "df_pre  = df_y[df_y['Timepoint'] == 'pre'].drop('Timepoint',axis=1)\n",
    "df_pre.index = df_pre.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "df_post  = df_y[df_y['Timepoint'] == 'post'].drop('Timepoint',axis=1)\n",
    "df_post.index = df_post.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "\n",
    "# df_diff\n",
    "df_diff = df_post - df_pre\n",
    "df_diff = df_diff.dropna()\n",
    "df_diff.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "f60310d5-5455-4ff0-82da-4e463a7450bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>animal_proteins_red</th>\n",
       "      <th>refined_grains</th>\n",
       "      <th>solanaceae</th>\n",
       "      <th>saturated</th>\n",
       "      <th>pastries</th>\n",
       "      <th>milk_derivates</th>\n",
       "      <th>beverages_prohibited</th>\n",
       "      <th>pro_inflammatory</th>\n",
       "      <th>chicken</th>\n",
       "      <th>plant_protein</th>\n",
       "      <th>...</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis</th>\n",
       "      <th>Metabolism|Lipid metabolism|Steroid hormone biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "      <th>WOMAC_pain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>1.282808e-05</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.085977e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>14.0</td>\n",
       "      <td>18.4</td>\n",
       "      <td>5.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.6</td>\n",
       "      <td>36.4</td>\n",
       "      <td>33.6</td>\n",
       "      <td>113.6</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>4.251192e-06</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.155420e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>-2.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.8</td>\n",
       "      <td>10.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.4</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>-7.975351e-07</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.990602e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>9.2</td>\n",
       "      <td>22.4</td>\n",
       "      <td>-5.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>3.2</td>\n",
       "      <td>3.2</td>\n",
       "      <td>23.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>8.057778e-07</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>-1.103744e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.230988e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>5.6</td>\n",
       "      <td>3.6</td>\n",
       "      <td>8.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>8.4</td>\n",
       "      <td>5.6</td>\n",
       "      <td>42.8</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>-3.364107e-08</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>8.318388e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.042987e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 704 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        animal_proteins_red  refined_grains  solanaceae  saturated  pastries  \\\n",
       "OAD001                  0.0             0.0         0.0        0.0       0.0   \n",
       "OAD003                 14.0            18.4         5.6        0.0       5.6   \n",
       "OAD004                 -2.0            12.0         2.8        0.0       0.0   \n",
       "OAD005                  9.2            22.4        -5.2        0.0       2.8   \n",
       "OAD006                  5.6             3.6         8.4        0.0       2.8   \n",
       "\n",
       "        milk_derivates  beverages_prohibited  pro_inflammatory  chicken  \\\n",
       "OAD001             0.0                   0.0               NaN      0.0   \n",
       "OAD003            36.4                  33.6             113.6      6.0   \n",
       "OAD004             0.8                  10.0              32.0     -2.0   \n",
       "OAD005             3.2                   3.2              23.6      0.0   \n",
       "OAD006             8.4                   5.6              42.8     -2.0   \n",
       "\n",
       "        plant_protein  ...  \\\n",
       "OAD001            2.0  ...   \n",
       "OAD003            0.0  ...   \n",
       "OAD004            0.4  ...   \n",
       "OAD005            4.0  ...   \n",
       "OAD006            6.0  ...   \n",
       "\n",
       "        Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis  \\\n",
       "OAD001                                           0.000034                                      \n",
       "OAD003                                           0.000023                                      \n",
       "OAD004                                           0.000033                                      \n",
       "OAD005                                           0.000016                                      \n",
       "OAD006                                           0.000011                                      \n",
       "\n",
       "        Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis  \\\n",
       "OAD001                                       1.282808e-05                                    \n",
       "OAD003                                       4.251192e-06                                    \n",
       "OAD004                                      -7.975351e-07                                    \n",
       "OAD005                                       8.057778e-07                                    \n",
       "OAD006                                      -3.364107e-08                                    \n",
       "\n",
       "        Metabolism|Lipid metabolism|Steroid hormone biosynthesis  \\\n",
       "OAD001                                           0.000041          \n",
       "OAD003                                           0.000020          \n",
       "OAD004                                           0.000012          \n",
       "OAD005                                           0.000019          \n",
       "OAD006                                           0.000009          \n",
       "\n",
       "        Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series  \\\n",
       "OAD001                                           0.000033                                               \n",
       "OAD003                                           0.000023                                               \n",
       "OAD004                                           0.000033                                               \n",
       "OAD005                                           0.000016                                               \n",
       "OAD006                                           0.000011                                               \n",
       "\n",
       "        Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD001                                       0.000000e+00                                   \n",
       "OAD003                                       0.000000e+00                                   \n",
       "OAD004                                       0.000000e+00                                   \n",
       "OAD005                                      -1.103744e-07                                   \n",
       "OAD006                                       8.318388e-07                                   \n",
       "\n",
       "        Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD001                                                0.0                                  \n",
       "OAD003                                                0.0                                  \n",
       "OAD004                                                0.0                                  \n",
       "OAD005                                                0.0                                  \n",
       "OAD006                                                0.0                                  \n",
       "\n",
       "        Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD001                                       6.085977e-07                           \n",
       "OAD003                                       7.155420e-07                           \n",
       "OAD004                                      -1.990602e-07                           \n",
       "OAD005                                      -1.230988e-07                           \n",
       "OAD006                                      -1.042987e-07                           \n",
       "\n",
       "        Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis  \\\n",
       "OAD001                                                0.0                               \n",
       "OAD003                                                0.0                               \n",
       "OAD004                                                0.0                               \n",
       "OAD005                                                0.0                               \n",
       "OAD006                                                0.0                               \n",
       "\n",
       "        Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \\\n",
       "OAD001                                                0.0                                          \n",
       "OAD003                                                0.0                                          \n",
       "OAD004                                                0.0                                          \n",
       "OAD005                                                0.0                                          \n",
       "OAD006                                                0.0                                          \n",
       "\n",
       "        WOMAC_pain  \n",
       "OAD001        -1.5  \n",
       "OAD003        -1.5  \n",
       "OAD004         1.0  \n",
       "OAD005         NaN  \n",
       "OAD006        -2.5  \n",
       "\n",
       "[5 rows x 704 columns]"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get dfs of diffs\n",
    "g = 'stool'\n",
    "t = 'diff'\n",
    "\n",
    "df_diet = gt_to_df_diet[g][t].copy()\n",
    "df_meta_stool = gt_to_meta['stool'][t].copy()\n",
    "df_meta_stool.columns = df_meta_stool.columns.map(lambda x: x + '_diff_stool')\n",
    "df_meta_plasma = gt_to_meta['plasma'][t].copy()\n",
    "df_meta_plasma.columns = df_meta_plasma.columns.map(lambda x: x + '_diff_plasma')\n",
    "df_otu_stool = gt_to_otu['stool'][t].copy()\n",
    "df_path_stool = gt_to_path[g][t].copy()\n",
    "df_outcomes = df_diff['WOMAC_pain']\n",
    "\n",
    "# merge all dfs\n",
    "df_merge = pd.concat([df_diet, df_meta_stool, df_meta_plasma, df_otu_stool, df_path_stool, df_outcomes],axis=1)\n",
    "df_merge.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "f04b9fa6-8373-4f91-9f1d-01e74ff5233c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>animal_proteins_red</th>\n",
       "      <th>refined_grains</th>\n",
       "      <th>solanaceae</th>\n",
       "      <th>saturated</th>\n",
       "      <th>pastries</th>\n",
       "      <th>milk_derivates</th>\n",
       "      <th>beverages_prohibited</th>\n",
       "      <th>pro_inflammatory</th>\n",
       "      <th>chicken</th>\n",
       "      <th>plant_protein</th>\n",
       "      <th>...</th>\n",
       "      <th>Saliva__Metabolism|Lipid metabolism|Fatty acid elongation</th>\n",
       "      <th>Saliva__Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis</th>\n",
       "      <th>Saliva__Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series</th>\n",
       "      <th>Saliva__Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Saliva__Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Saliva__Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of 12-, 14- and 16-membered macrolides</th>\n",
       "      <th>Saliva__Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of enediyne antibiotics</th>\n",
       "      <th>Saliva__Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Saliva__Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "      <th>WOMAC_pain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001.pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>1.911111e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.539150e-07</td>\n",
       "      <td>14.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003.pre</th>\n",
       "      <td>-14.0</td>\n",
       "      <td>-22.4</td>\n",
       "      <td>-5.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-5.6</td>\n",
       "      <td>-36.4</td>\n",
       "      <td>-33.6</td>\n",
       "      <td>-117.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>13.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004.pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-14.0</td>\n",
       "      <td>-2.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.8</td>\n",
       "      <td>-14.0</td>\n",
       "      <td>-42.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.6</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>4.396613e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.211511e-08</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005.pre</th>\n",
       "      <td>-11.2</td>\n",
       "      <td>-22.4</td>\n",
       "      <td>-2.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.8</td>\n",
       "      <td>-11.2</td>\n",
       "      <td>-11.2</td>\n",
       "      <td>-61.6</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006.pre</th>\n",
       "      <td>-5.6</td>\n",
       "      <td>-5.6</td>\n",
       "      <td>-8.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.8</td>\n",
       "      <td>-8.4</td>\n",
       "      <td>-5.6</td>\n",
       "      <td>-44.8</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>1.145692e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.091966e-08</td>\n",
       "      <td>14.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1358 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            animal_proteins_red  refined_grains  solanaceae  saturated  \\\n",
       "OAD001.pre                  0.0             0.0         0.0        0.0   \n",
       "OAD003.pre                -14.0           -22.4        -5.6        0.0   \n",
       "OAD004.pre                  0.0           -14.0        -2.8        0.0   \n",
       "OAD005.pre                -11.2           -22.4        -2.8        0.0   \n",
       "OAD006.pre                 -5.6            -5.6        -8.4        0.0   \n",
       "\n",
       "            pastries  milk_derivates  beverages_prohibited  pro_inflammatory  \\\n",
       "OAD001.pre       0.0             0.0                   0.0               NaN   \n",
       "OAD003.pre      -5.6           -36.4                 -33.6            -117.6   \n",
       "OAD004.pre       0.0            -2.8                 -14.0             -42.0   \n",
       "OAD005.pre      -2.8           -11.2                 -11.2             -61.6   \n",
       "OAD006.pre      -2.8            -8.4                  -5.6             -44.8   \n",
       "\n",
       "            chicken  plant_protein  ...  \\\n",
       "OAD001.pre      0.0            0.0  ...   \n",
       "OAD003.pre      0.0            6.0  ...   \n",
       "OAD004.pre      6.0            5.6  ...   \n",
       "OAD005.pre      6.0            0.0  ...   \n",
       "OAD006.pre      6.0            0.0  ...   \n",
       "\n",
       "            Saliva__Metabolism|Lipid metabolism|Fatty acid elongation  \\\n",
       "OAD001.pre                                                0.0           \n",
       "OAD003.pre                                                0.0           \n",
       "OAD004.pre                                                0.0           \n",
       "OAD005.pre                                                0.0           \n",
       "OAD006.pre                                                0.0           \n",
       "\n",
       "            Saliva__Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis  \\\n",
       "OAD001.pre                                           0.000095                                            \n",
       "OAD003.pre                                           0.000018                                            \n",
       "OAD004.pre                                           0.000067                                            \n",
       "OAD005.pre                                           0.000127                                            \n",
       "OAD006.pre                                           0.000019                                            \n",
       "\n",
       "            Saliva__Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series  \\\n",
       "OAD001.pre                                           0.000019                                                       \n",
       "OAD003.pre                                           0.000043                                                       \n",
       "OAD004.pre                                           0.000028                                                       \n",
       "OAD005.pre                                           0.000071                                                       \n",
       "OAD006.pre                                           0.000053                                                       \n",
       "\n",
       "            Saliva__Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD001.pre                                       1.911111e-06                                           \n",
       "OAD003.pre                                       0.000000e+00                                           \n",
       "OAD004.pre                                       4.396613e-07                                           \n",
       "OAD005.pre                                       0.000000e+00                                           \n",
       "OAD006.pre                                       1.145692e-06                                           \n",
       "\n",
       "            Saliva__Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD001.pre                                                0.0                                          \n",
       "OAD003.pre                                                0.0                                          \n",
       "OAD004.pre                                                0.0                                          \n",
       "OAD005.pre                                                0.0                                          \n",
       "OAD006.pre                                                0.0                                          \n",
       "\n",
       "            Saliva__Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of 12-, 14- and 16-membered macrolides  \\\n",
       "OAD001.pre                                                0.0                                                                 \n",
       "OAD003.pre                                                0.0                                                                 \n",
       "OAD004.pre                                                0.0                                                                 \n",
       "OAD005.pre                                                0.0                                                                 \n",
       "OAD006.pre                                                0.0                                                                 \n",
       "\n",
       "            Saliva__Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of enediyne antibiotics  \\\n",
       "OAD001.pre                                                0.0                                                  \n",
       "OAD003.pre                                                0.0                                                  \n",
       "OAD004.pre                                                0.0                                                  \n",
       "OAD005.pre                                                0.0                                                  \n",
       "OAD006.pre                                                0.0                                                  \n",
       "\n",
       "            Saliva__Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD001.pre                                       0.000000e+00                                   \n",
       "OAD003.pre                                       0.000000e+00                                   \n",
       "OAD004.pre                                       1.211511e-08                                   \n",
       "OAD005.pre                                       0.000000e+00                                   \n",
       "OAD006.pre                                       0.000000e+00                                   \n",
       "\n",
       "            Saliva__Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \\\n",
       "OAD001.pre                                       1.539150e-07                                                  \n",
       "OAD003.pre                                       0.000000e+00                                                  \n",
       "OAD004.pre                                       0.000000e+00                                                  \n",
       "OAD005.pre                                       0.000000e+00                                                  \n",
       "OAD006.pre                                       5.091966e-08                                                  \n",
       "\n",
       "            WOMAC_pain  \n",
       "OAD001.pre        14.5  \n",
       "OAD003.pre        13.5  \n",
       "OAD004.pre        10.0  \n",
       "OAD005.pre         NaN  \n",
       "OAD006.pre        14.5  \n",
       "\n",
       "[5 rows x 1358 columns]"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dfs of absolutes\n",
    "df_diet_pre = gt_to_df_diet['stool']['pre'].copy()\n",
    "df_diet_pre.index = df_diet_pre.index.map(lambda x: x + '.pre')\n",
    "df_diet_post = gt_to_df_diet['stool']['pre'].copy()\n",
    "df_diet_post.index = df_diet_post.index.map(lambda x: x + '.post')\n",
    "df_diet = pd.concat([df_diet_pre, df_diet_post])\n",
    "\n",
    "\n",
    "df_meta_stool = gt_to_meta['stool']['all'].copy()\n",
    "df_meta_stool.index = df_meta_stool.index.map(lambda x: x.replace('.stool','').replace('-',''))\n",
    "df_meta_stool.columns = df_meta_stool.columns.map(lambda x: 'Stool__' + x)\n",
    "\n",
    "df_meta_plasma = gt_to_meta['plasma']['all'].copy()\n",
    "df_meta_plasma.index = df_meta_plasma.index.map(lambda x: x.replace('.plasma','').replace('-',''))\n",
    "df_meta_plasma.columns = df_meta_plasma.columns.map(lambda x: 'Plasma__' + x)\n",
    "\n",
    "df_otu_stool = gt_to_otu['stool']['all'].copy()\n",
    "df_otu_stool.index = df_otu_stool.index.map(lambda x: x.replace('.stool','').replace('-',''))\n",
    "df_otu_stool.columns = df_otu_stool.columns.map(lambda x: 'Stool__' + x)\n",
    "\n",
    "df_path_stool = gt_to_path['stool']['all'].copy()\n",
    "df_path_stool.index = df_path_stool.index.map(lambda x: x.replace('.stool','').replace('-',''))\n",
    "df_path_stool.columns = df_path_stool.columns.map(lambda x: 'Stool__' + x)\n",
    "\n",
    "\n",
    "df_meta_saliva = gt_to_meta['saliva']['all'].copy()\n",
    "df_meta_saliva.index = df_meta_saliva.index.map(lambda x: x.replace('.saliva','').replace('-',''))\n",
    "df_meta_saliva.columns = df_meta_saliva.columns.map(lambda x: 'Saliva__' + x)\n",
    "\n",
    "df_otu_saliva = gt_to_otu['saliva']['all'].copy()\n",
    "df_otu_saliva.index = df_otu_saliva.index.map(lambda x: x.replace('.saliva','').replace('-',''))\n",
    "df_otu_saliva.columns = df_otu_saliva.columns.map(lambda x: 'Saliva__' + x)\n",
    "\n",
    "df_path_saliva = gt_to_path['saliva']['all'].copy()\n",
    "df_path_saliva.index = df_path_saliva.index.map(lambda x: x.replace('.saliva','').replace('-',''))\n",
    "df_path_saliva.columns = df_path_saliva.columns.map(lambda x: 'Saliva__' + x)\n",
    "\n",
    "df_outcomes = type_to_df_map['stool'].copy()\n",
    "df_outcomes = df_outcomes[['WOMAC_pain','Timepoint']]#,'WOMAC_P_Response']]\n",
    "df_outcomes = df_outcomes.dropna()\n",
    "df_pre  = df_outcomes[df_outcomes['Timepoint'] == 'pre'].drop('Timepoint',axis=1)\n",
    "df_pre.index = df_pre.index.map(lambda x: x.split('.p')[0].replace('-','') + '.pre')\n",
    "df_post  = df_outcomes[df_outcomes['Timepoint'] == 'post'].drop('Timepoint',axis=1)\n",
    "df_post.index = df_post.index.map(lambda x: x.split('.p')[0].replace('-','')  + '.post')\n",
    "\n",
    "\n",
    "df_outcomes = pd.concat([df_pre,df_post])\n",
    "\n",
    "# merge all dfs\n",
    "df_merge = pd.concat([df_diet, \n",
    "                      df_meta_plasma,\n",
    "                      df_meta_stool, df_otu_stool, df_path_stool, \n",
    "                      df_meta_saliva, df_otu_saliva, df_path_saliva, \n",
    "                      df_outcomes],axis=1)\n",
    "df_merge.head() # n=42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "91481c64-6d37-4a45-b416-985003aade74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find everything associated with pain\n",
    "\n",
    "d_to_df = {\n",
    "    'diet': df_diet,\n",
    "    'meta_stool': df_meta_stool,\n",
    "    'meta_plasma': df_meta_plasma,\n",
    "    'taxa_stool': df_otu_stool,\n",
    "    'path_stool': df_path_stool,\n",
    "    'meta_saliva': df_meta_saliva,\n",
    "    'taxa_saliva': df_otu_saliva,\n",
    "    'path_saliva': df_path_saliva,\n",
    "}\n",
    "\n",
    "d_to_sig = {}\n",
    "\n",
    "for d in d_to_df:\n",
    "    # get df of datatype of interest\n",
    "    df = d_to_df[d].copy()\n",
    "\n",
    "    # merge with outcome data\n",
    "    df_m = pd.concat([df, df_outcomes],axis=1)\n",
    "    \n",
    "    # identify significant dietary correlates with WOMAC pain \n",
    "    sig_feat = []\n",
    "    for x in df.columns:\n",
    "        df_sub = df_m[[x,'WOMAC_pain']].dropna()\n",
    "        r, p = scipy.stats.pearsonr(df_sub[x],df_sub['WOMAC_pain'])\n",
    "        #if p < 0.05:\n",
    "        if r > 0.5 or r < -0.5:\n",
    "            sig_feat.append(x)\n",
    "\n",
    "    d_to_sig[d] = sig_feat\n",
    "\n",
    "sig_feat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "0838c3f6-1e02-4f3f-891e-84ae57fa0e4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture \n",
    "import os\n",
    "# diet -> bacteria -> outcome\n",
    "for duet in [['diet','taxa_stool'],\n",
    "             ['diet','meta_stool'],\n",
    "             ['meta_stool','taxa_stool'],\n",
    "             ['diet','path_stool'],\n",
    "             ['meta_stool','path_stool'],\n",
    "             \n",
    "             ['diet','taxa_saliva'],\n",
    "             ['diet','meta_saliva'],\n",
    "             ['meta_saliva','taxa_saliva'],\n",
    "             ['diet','path_saliva'],\n",
    "             ['meta_saliva','path_saliva']]:\n",
    "    v1,v2=duet\n",
    "    try:\n",
    "        os.mkdir(path + 'outputs/jobs36/' + v1 + '_' + v2 + '/')\n",
    "    except:\n",
    "        pass\n",
    "    for x in d_to_sig[v1]:\n",
    "        for y in d_to_sig[v2]:\n",
    "            df_sub = df_merge[[x,y,'WOMAC_pain']].dropna()\n",
    "            r, p = scipy.stats.pearsonr(df_sub[x],df_sub[y])\n",
    "            if r > 0.5 or r < -0.5:\n",
    "            #if p < 0.05:\n",
    "\n",
    "                # x,y\n",
    "                plt.figure(figsize=(6,4))\n",
    "                sns.lmplot(data=df_sub,x=x,y=y)\n",
    "                sns.despine()\n",
    "                plt.tight_layout()\n",
    "                plt.ylabel(y.split(';f__')[-1])\n",
    "                plt.savefig(path + 'outputs/jobs36/' + v1 + '_' + v2 + '/' + x + '_' + y.split(';f__')[-1] + '.pdf')\n",
    "                plt.close()\n",
    "    \n",
    "                # x,pain\n",
    "                for i,v in enumerate([x,y]):\n",
    "                    plt.figure(figsize=(6,4))\n",
    "                    sns.lmplot(data=df_sub,x=v,y='WOMAC_pain')\n",
    "                    sns.despine()\n",
    "                    plt.tight_layout()\n",
    "                    plt.xlabel(v.split(';f__')[-1])\n",
    "                    plt.savefig(path + 'outputs/jobs36/' + v1 + '_' + v2 + '/' + x + '_' + y.split(';f__')[-1] + '_pain' + str(i) + '.pdf')\n",
    "                    plt.close()\n",
    "\n",
    "            # y,pain\n",
    "\n",
    "\n",
    "# diet -> metabolite -> outcome\n",
    "\n",
    "# diet -> bacteria -> metabolite -> outcome\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "7da7ea24-096e-43d7-8981-4d3495cb4ff0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PearsonRResult(statistic=-0.08633427047751045, pvalue=0.6384774545048038)\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeQAAAHjCAYAAADyq2xBAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABhg0lEQVR4nO3deXxU1d0/8M/dZk0mgbAawBAIayEgi6BoaZBq1WrV9KnPU9GfS21V7GOtVas+pbZa7U+rFYWq1Za6/UAD4ooKok9ti2yVoLKFJSIBEoyQYTIzd//9MZmQkKAzySRzZ/J5v14YcufmzvdA5JNz7j3nCLZt2yAiIqK0EtNdABERETGQiYiIHIGBTERE5AAMZCIiIgdgIBMRETkAA5mIiMgBGMhEREQO0OMCefv27di+fXu6yyAiImqlxwWypmnQNC0l11JVFRs3boSqqim5nhNkY5sAtiuTZGObALYrk6SrTT0ukFPJNM1WH7NBNrYJYLsySTa2CWC7Mkm62uSoQF64cCHmzJnT6lhdXR1uvvlmTJ48Gaeeeip+/vOf48svv0xThURERF3DMYG8aNEizJ8/v9UxTdNw1VVX4fPPP8df//pXPPHEE9iyZQtuu+22NFVJRETUNeR0F1BbW4s777wTGzduxNChQ1u99vrrr6OmpgYrV65Enz59AAB33HEH7r77boRCIeTk5KSjZCIiopRLew/5008/RV5eHl599VWUlpa2eu2DDz7AtGnTmsMYAM444wysWrWKYUxERFkl7T3ksrIylJWVtftadXU1Jk+ejAULFmD58uUwDAMzZszAL37xCwQCgQ6/p23bCIfDHf76uEgk0upjNsjGNgFsVybJxjYBbFcmSXWbfD5fQucJTtoP+fbbb0dNTQ2effZZAMDs2bNx5MgRTJ8+Hddccw0aGhpw3333oXfv3nj22WchCELS7/Hxxx+nbNoTERHR15k0aVJC56W9h/xVFEWBz+fDH/7wByiKAgDIy8vD97//fXz88ccYP358h687fPjwTtcXiURQXV2NoqIieL3eTl/PCbKxTQDblUmysU0A25VJ0tUmRwfygAEDYFlWcxgDQElJCQBg3759HQ5kQRASHkJIhNfrTen1nCAb2wSwXZkkG9sEsF2ZpLvblPaHur7K5MmTsW3bNkSj0eZjO3bsAACcfPLJ6SqLiIgo5RwdyJdeeikkScLPf/5z7NixAxs3bsRdd92FU089FWPHjk13eURERCnj6EDu3bs3nn/+eRiGgf/4j//Addddh3HjxmHBggXpLo2IiCilHHUP+f77729zrKioCE888UQaqiEiIuo+jgpkop7OsmzsrmlAsFFDwO9CcWEeRDH56X1ElHkYyEQOUVl1CBWrq1BTF4JhWpAlEYX9clBeVoLSkr7pLo+Iupij7yET9RSVVYewoKIS1fuD8Lgl9Mp1w+OWUH0giAUVlaisOpTuEomoizGQidLMsmxUrK5CJGqgIM8NtyJBFAW4FQkFATciqoGK1VWwLMcsqkdEXYCBTJRmu2saUFMXQq5fabMcrCAIyPUqqKkLYXdNQ5oqJKLuwEAmSrNgowbDtKBI7f/vqMgiDNNCsJFrsBNlMwYyUZoF/C7IkgjdtNp9XTdiD3gF/K5uroyIuhMDmSjNigvzUNgvB0fDOo7ffM22bRyN6Cjsl4Piwrw0VUhE3YGBTJRmoiigvKwEXreM+qAKVTNhWTZUzUR9UIXPLaO8rITzkYmyHAOZyAFKS/rihvJSFA0MIKqZOHxURVQzUTQwgOvLSzkPmagH4MIgRA5RWtIX44b14UpdRD0UA5nIQURRwPDB+ekug4jSgEPWRERETWLryQexc38Uu2uC3bogD3vIREREOLae/Oe1R6GqGlZ8FMLg/rndtp48e8hERNTjtVxP3usSkeMV4XWJ3bqePAOZiIh6tOPXk3cpEgRBgKub15NnIBMRUY/mlPXkGchERNSjOWU9eQYyERH1aE5ZT56BTEREPZpT1pNnIBMRUY/WZj153YRt21D17l1PnoFMREQ9Xuv15C2EIhaimtWt68lzYRAiIiIcW09+y65afLptF8aOGoYxw/p323ryDGQiIqImoiiguDAANehBcWGgWzd34ZB1BxmGhQ82HcDfPwnig00HYBjtP51HRESUCPaQO2DZ+ztR8e4ONEZ02Dbw/sef4G8rtqN81ghcPHN4ussjIqIMxEBO0rL3d+KZN7bAsmxIkgDbtiEIAkJhHc+8sQUAGMpERJQ0DlknwTAsVLy7A5ZlQ5EFiKIAUYh9VGQhth7quzs4fE1EREljICfhg001aIwYkCQBgtD6j04QREiSgMaIgQ821aSpQiIiylQM5CQcOhyBDRvCCR66EwTAho1DhyPdWxgREWU8BnIS+vbyQoAA+wQ7cNk2IEBA317e7i2MiIgyHgM5CWdMKITfK8M0bdh26/vEtm3BNG34vTLOmFCYpgqJiChTMZCTIMsiymeNgCgK0A0blmXDsmMfdcOOrYc6awRkmX+sRESUHE57SlJ8SlPLeciCYCPHp3AeMhERdRgDuQMunjkcF8woxrvr9mBL1V6MKRmCWVOHsmdMREQdxkDuIFkWccaEgejjPoLRowcyjImIqFOYIkRERA7gqEBeuHAh5syZc8LX77rrLpSVlXVjRURE1BPJcvcPIDsmkBctWoT58+ef8PVVq1bhpZde6saKiIioJ7FtG6pm4GhER1Tvvm0X49J+D7m2thZ33nknNm7ciKFDh7Z7Tl1dHf7nf/4HU6dORU0Nl6UkIqLUsG0bmm4iqsV+mZaFcFiDZXf/ngRp7yF/+umnyMvLw6uvvorS0tI2r9u2jdtvvx0XXnghpk6dmoYKiYgom9i2jahmoCGkovZwGPXBKEIRHYZpnXAlxu6Q9h5yWVnZV94XXrRoEQ4dOoTHH38cTzzxREre07ZthMPhTl8nEom0+pgNsrFNANuVSbKxTQDb5QSabkFt6g2bpgXrBOmrRlUAQDQaTcn7+ny+hM5LeyB/lW3btuGxxx7D888/D5fLlbLr6rqOrVu3pux61dXVKbuWU2RjmwC2K5NkY5sAtqu7iKIIURRh2gJ0A4jqJgzDgGmaCfWCAzke7N27F6ZpdrqWSZMmJXSeYwNZVVXccsstuO666zBq1KiUXltRFAwf3vkVtSKRCKqrq1FUVASvNzs2lMjGNgFsVybJxjYBbFd3sW1AMyxouomIZsAy7RP2hE9EjaoINnyBIUOGwOPxdFGlbTk2kCsrK1FVVYXHHnsMCxYsABDr2RqGgYkTJ+Luu+/GBRdc0KFrC4KQ8BBCIrxeb0qv5wTZ2CaA7cok2dgmgO3qCpZlQ9UNRDUTqmbCsgHblqAoEqB08KINgMfj6dY2OTaQx48fj3feeafVsWeffRbvvPMOnn32WRQUFKSpMiIiSre2IWyn9YGsVHBsIHs8Hpx88smtjuXl5UGW5TbHiYgo+2VjCLfk2EAmIiLK9hBuyVGBfP/993/l6zfeeCNuvPHGbqqGiIjSwbRiK2apevaHcEuOCuRMYhgWPth0AFuqgvhCPcDtF4mIOqGnhnBLDOQOWPb+TlS8uwONER22Dbz/8Sf424rtKJ81AhfP7Px0KiKiniAewlHNhKb3zBBuiYGcpGXv78Qzb2yBZdmQJAG2bUMQBITCOp55YwsAMJSJiE6gZQirugm7h4dwSxxjTYJhWKh4dwcsy4YiCxBFAaIQ+6jIAizLRsW7O2AY3b8oORGRU5mmhcaoji+DUdQdDuNISEVENWBZDOOWGMhJ+GBTDRojBiRJgCC0/qMTBBGSJKAxYuCDTdyRioh6tngI1zdEUXckgoajDOGvwyHrJBw6HIGN2BB1ewQBsGHj0GHnL7JORJQq8X8TTdNCVDMQ1WJLV9q2DWZv4hjISejbywsBAmw7FrwW7Ng9ZDs21GDbgAABfXulfz1XIqLuYFk2LEHGkaMqhLDNEO4EDlkn4YwJhfB7ZRimDd2wYBg2TAswjKbPTRt+r4wzJhSmu1Qioi5h2zZ0w0RjREN9QwSHjkRwJBhFWDViT0mnu8AMxh5yEmRZxGmlJ+HtNZ+1+aaLf35a6Umcj0xEWcG0bBiGCcO0jnVETKv5yWi76RzT4oOsqcBAToJl2aitD8PjlqBqrffUFATA7ZJQWx+GZdkQxfbvMxMROVEsbE0Yhg3DsqAbFkzTho1Y8rLn2/UYyEnYXdOAmroQ+uZ74ZIlBBtVRFUNHrcLAb8bmm6ipi6E3TUNGD44P93lEhG1YVk2dDPW0zWaery6YTXf++UT0OnDQE5CsFGDYVpQJAWCAOT6FMiiCa8n9rkiiwhFdAQbtXSXSkTUPOSsN4WuYVgwrNiDV+z1Og8DOQkBvwuyJEI3LSgQcDSsIxI1YVg6cn0KdMOGLIkI+F3pLpWIeijdsKDqBlTVhNbifi85HwM5CcWFeSjsl4Nt1V82rbsaOx6KRlHfEIVLkTCqqDeKC/PSWygR9Sha04YMUd2EYRz7t4kyCwM5CaIooH+BD5t2HGrzmmUDUc1E/wIfH+gioi5l2zY03WxeD9o0LYZwFuD8nCQYhoV/Ve7/ynP+Vbmfa1kTUcrZto2oZqAhpKLucBj1wShCER26wTDOFuwhJyG+lrUsCZBEEZZtw7IsiKIIURBgWlbzWtbfmjw43eUSUYazLBuq3rQzUg/dI7gnYSAn4fi1rAUBEAUB8aWtuZY1EXUW9wjuuRjISWi5ljXauU3MtayJqCMM02oKYQuawT2CeyreQ05CfC1r07Rh263vE9t2bFUbrmVNRInQDQuhiIb6IxEcOhxBQ0hDVOP2hD0ZAzkJsiyifNYIiKIA3bBju5zYsY+6EVsus3zWCK5lTUTtEmUXGqMGvjgSwRdHmkI4Piyd7uIo7ThknaSLZw4HAFS8uwONET02TC3YyPEpKJ81ovl1IiIAUHUTqmqi4WgUh4+qCIRUeLz8oZ3aYiB3wMUzh+OCGcV4d90ebKnaizElQzBr6lD2jImo1RzhqGbCtCzYdiyYdd1gT5hOiIHcQbIs4owJA9HHfQSjRw9kGBP1YKZpxXrCTb94H5g6goFMWc+ybOyuCWLn/ijcgSDGDPNyNTXqsNgzI8dt2GDazbslEXUUA5myWmXVIVSsrsLntUehqhpWfBTC4P65KC8rQWlJ33SXRw5n2zY0w4JhmDDMpt9zwwbqIgxkylqVVYewoKISkaiBHK8ERRQhyyKqDwSxoKISN5SXMpSpFdOyoekGND3W+9WbwpdbFVJ3YCBTVrIsGxWrqxCJGijIc8O2bZimAJciocAloz6oomJ1FcYN68Ph6x5O1U1oTZs06IYFG+z9UnowkCkr7a5pQE1dCLl+BYIgxHo5TQRBQK5XQU1dCLtrGjB8cH76CqVuZ9t27OGrpvWhTYubM5AzMJApKwUbNRimBUVS2n1dkUWEIjqCjVo3V0bpEB+K5iYN5GQMZMpKAb8LsiRCNy24RanN67phQZZEBPyuNFRH3YHrQ1OmYSBTViouzENhvxxUHwjCFWg9R9y2bRyN6CgaGEBxYV6aKqSuoMWHouP3gzkViTIIV7OgrCSKAsrLSuB1xx7gUnWz+d5hfVCFzy2jvKyED3RlONuObVXYEFJRdziM+oYIgmEttjgHw5gyDHvIlLVKS/rihvLSFvOQLbjdFooGBjgPOYO1uh/MVbEoizCQKauVlvTFuGF9sGVXLT7dtgtjRw3DmGH92TPOMPH7wapuQdP5UBZlJwYyZT1RFFBcGIAa9KC4MMAwzgCyLMMwLYTCGqIa7wdTz8B7yB2kaSaWvrcbL6/5Ekvf2w1NM9NdElFGM63Y/eBQxEAwYuGLhigaGnk/mHoOR/WQFy5ciDVr1uDZZ59tPrZ69WosWLAAu3fvRq9evXD22Wfjv//7v+HxeNJW52MVm7Bq7V6YTasJVO7ZhaXv78ZZpw7B3PIJaauLKBOYZmw9aNOyYZo2DCu2TKVp2rBhIxxWEY6osLhaB/UwjgnkRYsWYf78+ZgyZUrzsQ0bNmDu3Lm46aabcPbZZ+Ozzz7Dr371Kxw5cgT33XdfWup8rGIT3l7zWZvjpmU3H2coE8X+n9ANE7puwbBigRvbmAGwwfWhiY6X9iHr2tpaXHPNNXjkkUcwdOjQVq8tXrwY06ZNw7XXXouTTz4ZZ555Jn72s5/h1VdfhaZ1/wpLmmZi1dq9zZ8LAiA0fYxbtXYvh6+pRzJNCxFVR0NIxRdHIqj7Mowvg1EEwxrCUQOqbsK07OYHshjGRK2lvYf86aefIi8vD6+++ioWLFiAmpqa5teuuuoqiGLbnxkMw0AoFELv3r079J62bSMcDif9dUvf2908TC0IOPYvih373LZjvYIlK7fikm8Vd6i2dItEIq0+Zgu2K3UEQYgNOZuth5x1w4Rlo9W64R2hRtVWH7MF25U54m2JRqMpuZ7P50voPMHu7P89KXT77bejpqam1T3kljRNw3/8x39AFEUsW7asQ+/x8ccfd7h3/fKaL1G5Jxbk7T2nG/+DLB3qw0XTO/bDApFTiKIIURRhQ4BhAZaJWPiaNiwzNv/Xsi3e66WsFMjxwKsAptn5Ec9JkyYldF7ae8iJMgwDt956K3bu3Innn3++U9dSFAXDhw9P+uu2HNyNyj27Yp+07CEf9/nIoQMxenTm9pCrq6tRVFQEr9eb7nJShu1Kjm5YzXN+u3sdaDWqYl/NPgwqHAS3x909b9oN2K7MoUZVBBu+wJAhQ7r1AeKMCORQKISbbroJa9euxfz581FaWtqp6wmCkPAQQks/mD0aS9+PDVvbdoth66bhagCQRAE/mD0aLlfbDQ0yidfr7dCfkdOxXW1Zlg2t6eErzYjN+Y0NPYsQJBHuE+yY1dXcHndW/fAUx3ZliAbA4/F0678Xjg/kuro6/OhHP8K+ffvw5z//GdOmTUtbLS6XhLNOHdL8NLXd4h5y3FmnDsn4MKbsZtt20/KTZvO+wFz5iij9HB3IDQ0NuOKKKxAKhfDCCy9g5MiR6S6peUpTy3nIQKxnzHnI5BRW05Qjy449eGVaTb+a5v/aTQ9fMYOJnMPRgXzffffh888/x1NPPYXevXvj0KFDza/17t0bkpSenujc8gm49oJxWLJyK7bvOYCRQwdmxTA1Zab4TARVN6E39Xq1pqUmOdeXKHM4NpAty8Kbb74JXddxxRVXtHn93XffxaBBg9JQWYzLJeGSbxVj6wAVo0cXM4yp2xmmFdvxyJRw6EgUisvisDNRBnNUIN9///3NvxdFEZs3b05jNUTOEh+GVjULqmHCMCyEI1E0HA2jj2FCTs+zV0SUIo4K5ExiWTZ21wSxc38U7kAQY4Z5uYsQpYRt2zBMG6YVW/PZMG3ohtW07GTrh6/YIybKHgzkDqisOtRi03sNKz4KYXD/XG56T0k59pCVBcOIrXhlmLFj8QeuGLhEPQcDOUmVVYewoKISkaiBHK8ERRQhyyKqDwSxoKISN5SXMpQptmazFVvFKr5+s2UCpm01LzdpW0DTNgsMXiJiICfDsmxUrK5CJGqgIM8dm89pCnApEgpcMuqDKipWV2HcsD4cvu5BbDs2p1fTLai62byjUTxs+aQzESWCgZyE3TUNqKkLIdevQBCEVovoC4KAXK+CmroQdtc0YPjg/PQVSl0uvqSkqlnN833ZyyWizmAgJyHYqMEwLSgnWEpQkUWEIjqCjd2/NSR1LcO0mle20nQrNhzNACaiFGIgJyHgd0GWROimBbfYdt6xbliQJREBvysN1VEqxZaWNKA1bbBgmDZXtiKiLsVATkJxYR4K++Wg+kAQrkDrfZpt28bRiI6igQEUF+alqULqKNu2oRkWNC02FB1f6YrD0ETUXcSvP4XiRFFAeVkJvO7YA1yaYcKlxHrM9UEVPreM8rISPtDlUIIQ+3sxLRuqZqAxoqEhpKK+IYraw2HUN0QQDGuIarG9fhnGRNSd2ENOUmlJX9xQXoqK1VWob4hAkUQoLgXFJ+XjO6cVYWxxn3SXSMexLBuGacG0ZRw5qkKI2Oz9EpHjMJA7oLSkL8YN64Pte+rwyfbP0K9/Pwwf3AeyLOLw0ShkUYQsC5AlCYrMQYjuYpoWjPhiG01zfXUjtrtRJBLFkVAEvVQDXpFrTBKR8zCQO0gUBZw8MBeH613o19cPCGha5tACAAgCIAAQRAGKJMElx0JaFASIoghRFCBxaLtZfNUqQYgNLQuCAFE4NswMtJ7vqxkmdNNqPcfXbpr52868X8uyYfGxaCJyMAZyB1mWjc8OHEXNFxpUIYThQzyt7h3b8VAwbZimgagWC2kgFtRALLFjAR0LH0kSocgiXHJ29KxjazLHfkiJ9Vhje/MCx/4sLKv1qlXHCPE/qNgHIfYy5/sSUbZiIHdAfC3rLxuikEQLh4/WY0AfP86ZVoRRRb1P+HXxILHj/7UBCzZgxs8wY71qIbbblSQKTcsq2s1fb7fo/sV7kaIYC3VFii3jqUgiJKn7Aj0+VGxZseDVmzZEMEwL6PCazFzeioh6FgZyklquZd2vlxuyJCIcFVFzKITn396GH5496itD+evEw8syLRhmIme3FhsqF5p7oHaL46IgHOuRiwIkQYAoxYbOhaaTDNOGJCkwDAuqZgCI9UqbNz9ouifb6t3jPyiAazITEXUUAzkJx69l7VIk2JYJRZGQr0g40qjhrQ+rMWJIr7RNfTq+F92S2V6AN/8nJhqN4suQhkAwCk98wTF2VomIulzm36jsRsevZd2SIAjwuxXU1ofxee3RNFWYvHiPPP7LsmzoutE8D9dmGBMRdQsGchKOrWXd/h+bLAswLQuhsN7NlRERUaZjICeh5VrW7TEMG5IoIsfHea5ERJQcBnIS4mtZHw3rrbZeBGJPQjeqOvoX+DC4f26aKiQiokzFQE7C8WtZq7oJG7HFKo40avC6ZJwzrYhrWRMRUdIYyEmKr2VdNDAAVbMRUS1ouoXCvjn4r05OeSIiop6L05464Pi1rPv07YvhQ/qwZ0xERB3GQO6gVmtZ989hGBMRUacwkIkcxLJsfF57FKGwjhyfgsH9c/nDHnUav68yAwOZyCG2VX+Jtz6sRm19GKZlQRJF9C/wfe0a6URfhd9XmYMPdRE5wLbqL/H829tQUxeC2yUh4HfB7ZKa10jfVv1lukukDMTvq8zCQCZKM8uy8daH1YiqBvJzXHDJIkRBgEsWke93IaoZeOvDau7nTEnh91XmYSATpdnntUdRWx+G35s9a6RT+vH7KvMwkInSLBTWYVoWZKn9h2y4Rjp1BL+vMg8DmSjNcnwKJFGEYbY/dMg10qkj+H2VeRjIRGk2uH8u+hf40BjlGumUOvy+yjwMZKI0E0UB50wrgscl40ijBk23YNk2NN3iGunUYfy+yjwMZCIHGFXUGz88exQK++ZA1U0EGzWousk10qlT+H2VWbgwCJFDjCrqjRFDenFFJUopfl9lDgYykYPE1kgPpLsMyjL8vsoMHLLuIMOw8OEntajcHcaGrYdgGFa6SyIiogzmqB7ywoULsWbNGjz77LPNx7Zu3Yp7770Xn3zyCfLz8zFnzhxcffXVaawSWPb+Try4ajsMw4Isi2iMHMJLq3dh4qh+OH38SfB5ZPg9CnweBS5FbDMpn4iI6HiOCeRFixZh/vz5mDJlSvOxw4cP48orr8RZZ52Fu+++G5s2bcLdd9+N/Px8XHLJJWmpc9n7O/HMG1tgNi03p+qxnrFmWFj7yUGs/eRgq/NlSYDPo8DnkeHzKE1BLcdC26s0HWt6zdvyPBmSxAEMIqKeIu2BXFtbizvvvBMbN27E0KFDW7324osvwuVy4de//jVkWcawYcPw2Wef4c9//nNaAtkwLFS8uwOWZUMUgESWgDVMG8FGDcFGLen3c7ukFr3t1oHeMrybz2k65nXLENkrJyLKKGkP5E8//RR5eXl49dVXsWDBAtTU1DS/tmHDBkyZMgWyfKzMadOm4YknnkB9fT0KCgo69J62bSMcDif9dR9sOoDGiA5JEiAIgFsUIYhAJGrGrtt0niQKzT3ozlA1E6pm4nBQTerrBABetwyvR4LPLcPrkeFvCmqfW24O7XhPPX6OzyPDNmLL6KnR5N7T6eLtYbucLxvbBLBdmSTelmg0mpLr+Xy+hM5LeyCXlZWhrKys3dcOHjyIESNGtDrWr18/AMD+/fs7HMi6rmPr1q1Jf92WqiBsOxboAgR4XBL8XhkHtVi4W3bs15QSHyYW+xHVLai6DVW3EG36qGp20/HYa/Fzotqxczub5TaAsGogrBqoR3L/k4gC4HGJcCtfwK2IcCsCPE0f3YrY9NqxY/HP3YoIKQOmUeyr2ZfuErpENrYrG9sEsF2ZIpDjwd69e2GaZqevNWnSpITOS3sgf5VoNAqXy9XqmNvtBgCoasd/GlMUBcOHD0/6675QD+D9jz+BIAgQRKH5YS1BEAABECwbgg0MO/kkjBjRt0O12U0r6YSjsUCNqEbs99Fjv29s8fuweuy1qGqis/1yywbCqoWwCgDJfSO6FLG5B+5LpEfedMzjlrp8iF2NqthXsw+DCgfB7XF36Xt1p2xsVza2CWC7MokaVRFs+AJDhgyBx+Pptvd1dCB7PB5oWut7r/EgTnQIoD2CIHTo62dNHYq/rdiOUFiHKCA2Noymj7YN0wL8HhnTxw+CLHf8gSwfgPy85L/OsuymoNabgrvpY0RvCnUdjVEdjZHjz9Gh6Z2ftqXpFjRdw5FQcvfL40Ps8Xvg7d0nb30f/di5bkVK6il2t8cNr9ebZMucLxvblY1tAtiujNEQy6DOZE2yHB3IAwYMQF1dXatj8c/79+/f7fXIsojyWSPwzBtboBs2LMuODVNbsTAWBeDs6UWdCuPOEEUBfm/sae1kGaaFcFTHl0dCqNr1GXr16Q/DEluFecuQD7cI+87eL285xJ4sSRTafWK91QNvXgWyYOHwlxpyeoXRO1+Ez6NA5lPsROQgjg7kKVOmYPHixTBNE5IkAQDWrFmDoUOHdvj+cWddPDM21B1/2joWyLGe8dnTi/DtU09OS12dJUsiAn43FNFC6LALw4b2Tuin3fgQe2NURziio/GE4a0j1KJnHo7qiESNTg+xm5aNo2EdRxPd0/Vfh5t/61akNsHdcg5522OxnjmfYieirtChQN6zZw/+93//F+FwGJbVeqhTEATccMMNKSnukksuwVNPPYU777wT11xzDTZv3oy//e1vuPvuu1Ny/Y66eOZwXDCjGH//92fYvvsAek0q7PQwdaYSBAFulwS3S0LvQHL3Wiy7aYj9uCAPtzOs3txTV3WEIwZUvfMPWqi6CVU3cfhox55i9x03bzw2l/y4qWgtA96b/BA7EfUcSQfy8uXL8ctf/rLN/ppxqQzkgoICPPXUU7j33ntx0UUXoW/fvrj11ltx0UUXpeT6nSHLIqZ9oz88aEC/AX17ZBh3ligI8DfdH072Ebj4EHs8uBuPG0ZvFe5RA6GwhlBYharbaR1iF0XhhAvBtDenvOW9c4XfY0RZLelA/tOf/oTTTjsN99xzDwYMGJDSn/bvv//+NsfGjx+PJUuWpOw9KDvEh9gD/sSe6oxEIti1axeKi4shye7mB9zaPvR2bIg9HDUQiqR2iN1Kdoi9BZcithpO93kUeBQBajSE3fX7kJfrbfd+utctc2cfogyQdCDv378fv/71rzFw4MCuqIeoS7UcYu/V0SH2aHyYve0Dbi2fZm95H13VOj/EHnuKXW13iH3T7r0n/LrmIfZ4L/wET7O3tyKc28UhdqLuknQgDx06FAcOHOiKWogcreUQO/KTm94RH2KPzylvOax+fE+9Mdp6yN0wUzjE3pDcykMth9jj4d0yuI8fYm95z5xD7ETJSTqQf/7zn+O3v/0tCgsLMWHChOaFOojoxJIdYo+zbRu6YaGxVY/8WKA3hCI4WPclFLcPqm7HhtybznXCEPsJH3BremL9+Pvpfo+S2CLxRFko6UC+9957UV9fj//zf/5Pu68LgoAtW7Z0ti4iQuz/J5ciwaW0P8QeuzduYNiwYW2mqSUzxN6qpx7VUzrEfqQDT7G7FAF+7xHkeF3tPuDm98rwuds+1c4hdspkSQfyBRdc0BV1EFGKdWaI3TSt5gfcThjcXTjEHlvXXcWXSW6scvwQe7z33eYpds/xK8NxiJ3SL+lAnjt3blfUQUQOIkkiAn4XAn7X15/cQssh9nZXdmsK7+bpaseFfTqH2BVZPDYVrUXvu709zFv9nk+xU4okFMjr16/HmDFj4Pf7sX79+q89f8qUKZ0ujIgyT+sh9uS+1rJtRFUDXx4JYcfOavTuMwCGLbZd+a2dsI+mYIhdNywcOZr8EDsQe4q9Vc+8Va88dkwRbRz5UoMn0IheeYhNW+MQO7WQUCDPmTMHL774IsaPH485c+ZAEIQ2C4PEjwmC0KGtDYmoZxMFAT6PAiHPg375CoadnJ/wZgUth9hP9BR7c89cbd0z7+wQOwBEmnZmS+gp9g+PLd8qigJ87tYbqcSH0dtdOMZ7bMhdkaVO103OklAgP/PMMxg2bFjz74mInCRVQ+xtVn1rcX+8Mb7EazzsVQMnWLAwYZZlIxTREYp0bIi95QNuJ5xTflywez0yJJH3y50ooUCeOnVqu78nIspkqRhib7sOe/tD7KGIhmAoAs0UUvIUu25YaAipaAh1bIi93SfXW60Ed+yBt/hrHGLvWh3aXGLTpk1Yt24ddF1vHrq2bRvhcBgbN27Eiy++mNIiiYicJj7E7vMoAL5+aD2+fOuwYcPgcrkRVttZtrWp99143BB7y2H4VA6x1ye7UIwgtHnAzS0L0NVG7Kjbi0COt00vPf45h9i/XtKB/Pzzz+Oee+5pd3MJURQxY8aMlBRGRJStJElErs+FXF9yQ+wAoOlm03anbYfY21ujveU99E4PsdsthtgPR1q99vFn+77ya5uH2NtMSzt+x7SeO8SedCA/99xzmDFjBh588EE8+eSTOHr0KO644w787//+L26//XbOUyYi6kLNQ+y5yX2dZdtQVfPY1LN2Folpvmd+XNin6in2jg6xe9zSseHzdrY6bT3H/FjPvCND7JZl4/PaEOq/0JB34ChGDfV227S2pAN53759uP3225GXl4dx48bh0Ucfhcfjwdlnn409e/bgmWeewfnnn98VtRIRUQeJggBvU48zkSH2lkzLQiRqtFkspjGqI3g0gv219VA8fmi63eapdsO0Ol17VDURVc1ODbF/5ZanTWu0137ZiH99fABfHA6jIM+N1Zs/QUG+F+VlJSgtSXaT2OQlHciKosDjiS3hV1RUhM8++wy6rkNRFJxyyin4y1/+kvIiiYgofSRRRI7PhZx2hthj98a1dpdvBWJD7O0tCnP8EHt709VSOsSOyNee31Jj1EC/Xl5UHwhiQUUlbigv7fJQTjqQR48ejffeew+nnnoqTj75ZFiWhU2bNmHKlCk4ePBgV9RIREQZKj7Enp+b3MYqxw+xtzenvPWyrU2BruqIqp0fYjdMG8FGDf17+1AfVFGxugrjhvXp0uHrpAP5yiuvxNy5c9HQ0ID77rsPs2bNwq233oqzzz4br732GiZNmtQVdRIRUQ+SyiH2lr3v9tZoPxLSUN8QadMjD/hdEAQBuV4FNXUh7K5pwPDB+Slr4/GSDuSzzjoLjz/+OHbt2gUA+M1vfoOf//znWLx4McaNG4df/epXKS+SiIgoUV81xN6eT3fV429vftq8sIxpWsjPlSEKMlTdgiKLCEV0BBu1riy7Y/OQZ86ciZkzZwIAevXqxfvGRESUsXJ8CiRRhGHacMkiIImQRKG5x6wbVtOe5slPU0tGhwI5Go1i2bJl2LhxIxoaGlBQUIDp06fj/PPPhyx36JJERERpMbh/LvoX+FBzKATluNC1bRtHIzqKBgZQXJjXpXUkPdt63759OPfcc/Gb3/wGmzdvRigUwsaNG3H77bejvLwcDQ0NXVFn1hEACELslyjE7peIghA7lu7iiIh6EFEUcM60InhcMo40atB0E7YNqLqJ+qAKn1tGeVlJl89HTro7+7vf/Q4A8PLLL2P06NHNxz/99FPMnTsX//f//l/ce++9qavQ4RQptvqMS5Fg2zZsALaN5t+LggBRFCA1/RIFAYKI5t9DEJoD2Ubs3oVp2TBNO3adFjvEWrbd/JoVe5PYqzY6vY8sEVFPNqqoN3549ii89WE1Dn7RiIhqQTeAooEB585DXrt2LX7729+2CmMAGDt2LG666Sbcd999PSyQLeTnuODzJfcU4InIUmKDFnZzOB8LcLNp5xrTtI49LRj/ga4p3Ds7r4+IKFuNKuqNEUN6YefeL1D/xSGMGXEyRg3t59yVurxeLySp/UXCc3Jy2l3jOpuZZufnu3WEIAiQJeGEAW5adnMWC0Lsc90woekWdMOCblqxvyv2romImomigMH9c+BBA4oG5nZbGAMdCOTLL78cDz30EMaOHYtBgwY1Hz9y5Agef/xxXH755Skt0KkMw8IHmw5gS1UQX6gHMGvqUMiycxZAl477JoqHt7dpbr5l2TDMWDAbhtXc29ZlES5F5hZrRETdLOlA3rNnDw4fPoxzzjkHEydORL9+/XDkyBF89NFHiEaj8Hg8WLt2LYBYL+5vf/tbyotOt2Xv70TFuzvQGIkt7fb+x5/gbyu2o3zWCFw8c3i6y0uIKApwibEVdFrK8Qg4kutCr1wPZNkFzTChm1bsfjV700REXaZDm0uMHDmy+fO6ujoAsXvIcS33SM42y97fiWfe2ALLsiFJAmzbhiAICIV1PPPGFgDImFBuj23bMA0dHpcIX9Okervp3nS8N60bFgzTanqwjCFNRJQKSQfys88+m/C569evRzgchs/nS/ZtHMkwLFS8uwOWZUORY09I2xYgiLEnpXXDRsW7O3DBjGJHDV93liAIzevRxtm2Dc2wYBgmjKanvg3Tbnqg7NjT5kRElJguSw3TNHH55Zdjz549XfUW3e6DTTVojBiQJAGC0PqPThBESJKAxoiBDzbVpKnC7iMIAtyKBL/XhbwcN3rletA334sBBX70yfehV64bAZ8LXrcMWRKb51gTEVH7unRZrWwbsj50OAIb9gkfeBKE2NSiQ4eT2+Yr2yiyCEUWgRabu5hND5CpmomIasC0sut7g4ios7jOZRL69vJCQNP6pu1ksm0DAgT07ZWaOcnZRJJESJIIj0tGjleJ7bKi6jAtzo0mIgK6cMg6G50xoRB+r9y0ipbV6jXbtmCaNvxeGWdMKExThZlBkkTk+l3om+9DXo4bbkVCN071IyJyJAZyEmRZRPmsERBFAbphw7JiDzNZlg3dsCGKAspnjciqB7q6kigK8HsU9Mn3onfAA79Xabo/n+7KiIi6H4eskxSf0tRyHrIg2MjxKRk1D9lp3C4ZbpcM23ZB02Mrimm6CS3+1DaHtYkoyzGQO+DimcNxwYxivLtuD7ZU7cWYkiGOW6krUwmC0BTOsc9Ny4bRtOSnqpvQDJPhTERZiYHcQbIs4owJA9HHfQSjRw9kGHcRSRQgNQV0LoCIqqMxYjCYiSjrMEU6yLJs7K4JYuf+KHbXBGFxGk+38Lpj95x75brhcUmxLSyJiLJAl/WQRVHE3Llz0a9fv05fS9d1PPbYY3jllVfQ0NCA0aNH45ZbbsEpp5ySgkqTV1l1CBWrq/B57VGoqoYVH4UwuH9ut+2ZSbFg9roV6IYFVTMQUY2mHazSXRkRUcck3EPWdR2LFi3C22+/3eq4aZo477zz8OSTT8Kyjk0FEgQBc+fORd++nQ+oP/3pT1i6dCnuueceLF++HMXFxfjRj36E2traTl87WZVVh7CgohLV+4PwukTkeEV4XSKqDwSxoKISlVWHur2mnkyRReT4XOjbK7Y6mEsW25siTkTkeAkFsqZp+MlPfoLf//73+Pjjj1u9dvjwYbjdbjz00EO4/vrrYRhGyot89913cf7552PGjBk4+eSTcfvttyMUCmHTpk0pf6+vYlk2KlZXIRI1UJDnhkuRmtd5Lgi4EVENVKyu4vB1msSHswN+V2z6VLoLIiJKQkJD1kuWLMGGDRvw8MMP45xzzmn1Wp8+fbBs2TK8/vrruP3221FRUYFLL700pUXm5+fjvffew2WXXYaBAwdiyZIlcLlcGD16dIeuZ9s2wuFw0l+3uyaIz2uPIscrwW6afwzEgloUAb9Hwue1R7FlVy2KCwMdqi3dIpFIq4+ZSASQ6xYQ1SxENSM2fUpVAQBqVE1vcSkWb082tSsb2wSwXZkk3pZoNJqS6yW6wZJgJ7Dg9MUXX4wzzjgDP/vZz77yvN/85jfYvHkzKioqEqsyQTt27MDPfvYz7Ny5E5IkQRRFPPLII5g1a1bS1/r444+haVqH6ti5P4rlH36JHK/Y7nrWtm0jFLHwvWm9MfwkT4feg1JHEARIkgQbAnQDiOomdN2AYfIJbSL6aoEcD7xK7LZsZ02aNCmh8xLqIX/22We45ZZbvva8M888E6+99lpCb5yMXbt2IRAIYMGCBejfvz9eeukl3HbbbXjuuecwatSopK+nKAqGD09+AQ93IIgVH4UgyyJcigTLsqFqKtwuN0RRgKqbcLstjB01LKN7yNXV1SgqKoLXmz1rckciEezbtw8DhhUDgoyIZjQtgZrZyaxGVeyr2YdBhYPg9ri//gsyQDa2CWC7MokaVRFs+AJDhgyBx9N9nauEAlmWZei6ntB5J9oJqaNqamrwi1/8AosWLcLkyZMBAOPGjcPOnTvx6KOPYsGCBUlfUxCEDu3RPGaYF4P756L6QBAFLhli0x14URQgCAIaoyaKBgYwZlh/iBm+OLPX682afazjDMNArj/WLsuyoeoGIqoJVTdhW7E9nDOV2+POqh+ggOxsE8B2ZYwGwOPxdOu/gwk91FVSUoK1a9d+7Xlr167FoEGDOl1US5s3b4au6xg3blyr46Wlpaiurk7pe30dURRQXlYCr1tGfVCN/UNu21B1E/VBFT63jPKykowP455AFAV43Qp6Bzzol+9FIMcFRRK5jjYRpU1CgXzhhRfi//2//4fNmzef8JyPP/4Yzz//PL7zne+krDgAGDhwIABg+/btrY7v2LEDJ598ckrfKxGlJX1xQ3kpigYGENUshCIWopqFooEBXF9eynnIGUiSROR4Xejbq2nBEe4+RURpkNCQdXl5OV5//XXMmTMH5eXlmDlzJgYNGgTLslBTU4O///3vePHFFzFy5EjMmTMnpQWOHz8ekydPxm233YZ58+ZhwIABWL58OdasWYMXXnghpe+VqNKSvhg3rA+27KrFp9t2YeyoYVkxTN3TCYLQvOCIqpuIqgYimgGLezYTUTdIKJAFQcATTzyB3/3ud1iyZEmrILRtG7Is4/vf/z5uvvnmlN8AF0URCxcuxB//+Ef88pe/RENDA0aMGIFFixZhwoQJKX2v5OoSUFwYgBr0oLgwwDDOMm5FgluRELBdiGoGopoJVTNhcecpIuoiCS+d6fF48Jvf/AY33XQTPvzwQxw8eBCiKKKwsBDTpk1Dbm5ulxWZl5eHefPmYd68eV32HkTtadlrjj8IFtVMRNVYOBMRpUrSa1n37t0b5557bruvhUIhvPzyyykftiZygviDYF63At1rIRzVEVYNrsxGRCmRks0lNm/ejMWLF2PFihWIRqMMZMp6iiwiL8cNv1dBY0RHRDVgMpiJqBM6HMiRSASvvfYaFi9ejK1bt0IURZx++un43ve+l8LyiJxNlmLBnONVEI4aaFR1WGZmz2kmovRIOpC3b9+OxYsX47XXXkMoFIIgCLjiiitw9dVXp2RnJ6JMJEkicv0u+LwKIlEdjVEDpmkxmIkoYQkFsqZpePPNN7F48WJUVlYiNzcX3/3ud/Htb38bV155JWbNmsUwJgIgiQJyfC74PAoiqo5wlPs0E1FiEgrkM844A+FwGKeffjoefPBBzJ49Gy6XC0ePHu3q+ogykigK8HtjwRzVDISjsV2neJuZiE4koUBuaGhAcXExxowZg8GDB8PlcnV1XURZoeW0KU03EeFiI0R0AgkF8sqVK7F06VK8/PLL+NOf/oShQ4fikksuwcyZM7u4PKLs4VIkuJoWG9F0E6pmQdUNGKbFnjMRJbaW9eDBg3HTTTdh9erVeOKJJ1BSUoJHHnkEF1xwAQRBwLp16xAKhbq6VqKsIAgC3C4ZgRwX+vbyoXeeF7leFxSZm1sQ9WRJPWUtCALOPPNMnHnmmWhoaMArr7yCl19+GY8++iiefPJJfPOb38R3v/tdzJ49u6vqJco6zct0IrZMZyisQzNMDmkT9TAdnoecl5eHyy+/HJdffjm2bNmCiooKvPHGG1i5ciW2bt2ayhqJegyPS4ZbkRBWDTSGdRicOkXUYyQ0ZP11xowZg1/96lf44IMP8Ic//CEVl3Q8TTOx9L3deHnNl1j63m5ompnukihLCIIAv0dBn3wvcn0uSNy4hKhHSKiHvH79+oQv2BPmIz9WsQmr1u5tXiqxcs8uLH1/N846dQjmlk9Ib3GUNURRiC024pERisTmNHNDC6LslVAgz5kzB0KLp03sFv8oHH9cEISsHrJ+rGIT3l7zWZvjpmU3H2coUypJTctz+jwKQhENUdXgU9lEWSihQFYUBbquY8yYMTjvvPMwbty4rq7LkTTNxKq1e5s/FwQANgABzQ/grFq7F9deMA4ul5SWGil7KbKIXrkeqB4TobAGVedtEqJsklAgr1mzBitXrsQbb7yBhx9+GCeddBLOO+88nHfeeRg+fHhX1+gYy97f2TxMffz0FKEplE3LxrL3d+LSb49MQ4XUE7gVCe48L6KaAdE24HEprUaqiCgzJRTIOTk5uOiii3DRRRfhyJEjeOutt/Dmm2/iiSeewPDhw3H++efj3HPPxaBBg7q63rQ6+EVjSs8j6gyPS0bA70J+joJeuW7YghybLmVxtymiTJT0U9b5+fm49NJL8cwzz+D9999HeXk53nvvPXz729/GpZdeimeffbYr6nSEAX38KT2PKBUMQ4fHJaEgz4N++V7k5brhcckQ2WsmyiidmvbUr18/XH755Xjsscdw7bXX4pNPPsHvfve7VNXmOBfPHN48BeX4h13jn0uigItn9pxhfHIWSRLh9ygoyPOgT74Hfq/CYCbKEB1eGOTLL7/EO++8gxUrVmDDhg1wuVw466yzcO6556ayPkdxuSScdeqQ5qepm0O5RTifdeoQPtBFjqDIEvJzJPg9ChqjOiKcNkXkaEkF8uHDh1uFsCiKOPPMM/HAAw/gW9/6Frxeb1fV6RjxKU0t5yEDsZ4x5yGTEymyiPwcN/zN06ZMBjORAyUUyC+99BJWrFiBdevWQRAEnH766bj33nsxa9Ys5OTkdHWNjjO3fAKuvWAclqzciu17DmDk0IH4wezR7BmTo8WnTWkeE6GIDlXjfGYiJ0kokP/nf/4HkiThlFNOwVlnnYW8vDzYto1Vq1a1e/73vve9VNboSC6XhEu+VYytA1SMHl3MMKaM4VIk9FYkqJoRC2adG1kQOUHCQ9amaWL9+vVfu4ymIAg9IpCJMp3bJcPtkhHVYhtZqNxhiiitEgrkd999t6vryDiWZWN3TRA790fhDgQxZpgXIjcBoAzkccnwuGREVB2hsA7dtBjMRGmQUCAXFhZ2dR0ZpbLqECpWV+Hz2qNQVQ0rPgphcP9clJeVoLQk+zfXoOzkdStNwWygMcJgJupuST1lrWkaXn/9dbz77ruoqamBbds46aSTcNZZZ+G8886Dx+Ppqjodo7LqEBZUVCISNZDjlaCIImRZRPWBIBZUVOKG8lKGMmUsQRDg8yjweRSomoGIZsY2s+DqX0RdLuGFQaqqqnDhhRfijjvuwIYNGyBJEjweDyorK3HnnXfiwgsvxPbt27uy1rSzLBsVq6sQiRooyHPDpUgQBAEuRUJBwI2IaqBidRUsPrpKWcDtkpGf40bfXj7k5bjgUSTwrgxR10moh3zkyBH8+Mc/hsvlwp///GfMmDGj1WL269atw913343rrrsOL7/8MvLy8rqs4HTaXdOAmroQcv2xxfyP34Yy16ugpi6E3TUNGD44P32FEqWQJArwe13wewHdMBFVTURUA4bFIW2iVEqoh/zcc89B13U8//zzOOOMM9rsLDN16lQ899xzsG07q9eyDjZqMEwLitT+H5siizBMC8FGrZsrI+oeiiwh1+9C315e9A544PPIkEQB7DgTdV5Cgfz222/jiiuuQEFBwQnP6dWrF6644gq8/fbbKSvOaQJ+F2RJhG5a7b6uGxZkSUTA7+rmyoi6lyAI8Lhk9Mr1oG++F3k5Lrg5pE3UKQkF8r59+/CNb3zja88bO3Ys9u3b1+minKq4MA+F/XJwNKy3Gq4GANu2cTSio7BfDooLs3PInqg9kiTC73WhT74XBXle5HpjP7hyTwui5CQUyLIsQ1XVrz0vHA5n9XrWoiigvKwEXreM+qDatMKRDVU3UR9U4XPLKC8r4Xxk6rFcioRAjgv9ennRK9cNn0eGyCFtooQkFMijRo3C6tWrv/a89957DyNHjux0UU5WWtIXN5SXomhgAFHNQihiIapZKBoYwPWc8kQEIDak7XUr6JUb26M5kOOCSxY5pE30FRJ6yvqSSy7BXXfdhbPPPhunnXZau+d88MEHqKiowAMPPJDSAp2otKQvxg3rgy27avHptl0YO2oYxgzrz54xUTskSUSO14UcrwuqZiCsGtxxiqgdCQXyhRdeiHfeeQfXXnstLrnkEpSVlWHQoEFQFAU1NTV46623sHTpUsyePRvf+c53urpmRxBFAcWFAahBD4oLAwxjogTE18/WvbEdpxjMRMckFMiCIGD+/Pn4wx/+gBdeeAEvvvhiq9cVRcHVV1+NG2+8sUuKBIDly5fjySefxOeff44hQ4Zg7ty5PSb8qXO47rjzKLKEXrkSdK+JcNRARGtaDYzZTD1YQoG8ZMkSzJgxA7fddht+/OMfY926ddi3bx9s20ZhYSFmzJjRpfsiv/LKK7jjjjtw2223YebMmXj99ddx8803Y8CAAZg4cWKXvS9lPq477myKLCEvR0Ku5UJUMxBRDaiq0GatA6KeIKFAnjdvHgRBwJAhQzBjxgycdtpp+MEPfgC/39/V9cG2bTzyyCO44oorcMUVVwAAbrjhBvz73//GunXrGMh0Qlx3PHOI4rE1tF2ihS9yvVBkCYIA9pqpx0gokF977TWsW7cOGzduxMqVK/H8889DlmWUlpbi9NNPx4wZMzBu3Lgu+al29+7dqKmpwXe/+91Wx59++umUvxdlj+PXHbdtG6bZtO64KzZtrWJ1FcYN68Pha4eRJAGKaKJvvgeCpCAcNaHpvNdM2S+hQC4pKUFJSQl++MMfAgD27t2LdevWYcOGDVi2bBnmz5+PvLw8TJs2DTNmzMD3v//9lBVYXV0NIDbH+eqrr8aWLVswaNAgXHfddSgrK+vQNW3bRjgc7nRtkUik1cdskC1t2l0TxOe1R5HjlWDbdvOGH5ZlQxQBv0fC57VHsWVXLYoLA2mutuOy5e+rpUgkAsuyEIlE4PUCXgVwiUBUt2JraBtWRoazGlVbfcwW2diueFui0WhKrufz+RI6T7CPX3KqAz788EMsXrwYq1atgmma2Lp1a2cv2eyVV17BrbfeikGDBmHu3LkYNWoU3n77bTz++OP461//iunTpyd1vY8//hiaxrWms93O/VEs//BL5HjFdkdubNtGKGLhe9N6Y/hJ2b9taDYQRRGiKMK0BWi6jYhmwDAM7q5GXSKQ44FXAUzT7PS1Jk2alNB5Se2HHHfw4EF88MEH+Ne//oV169bhyy+/RCAQwFlnnYUZM2Z05JInpCgKAODqq6/GRRddBAAYPXo0tmzZ0qFAjl9z+PDhna4tEomguroaRUVFWbNCWba0yR0IYsVHIciyCJciwbJsqJoKt8sNURSg6ibcbgtjRw3L+B5yNvx9tZRom0zLhqrFdp7STcvxwaxGVeyr2YdBhYPg9rjTXU7KZGO71KiKYMMXGDJkCDye7vuBPaFA1jQN69evxwcffIB//OMf2LVrFyRJwvjx4/HDH/6wS+8hDxgwAAAwYsSIVseHDx+O999/v0PXFAQh4SGERHi93pRezwkyvU1jhnkxuH8uqg8EUeCSITatSSeKsSd4G6MmigYGsmZBl0z/+2pPIm3KbfoYX3BE1czY9KmuL6/D3B531vzw1FLWtasB8Hg83fr/VUKBfOqppyIajWLw4ME47bTTcNNNN2HatGldOtUpbsyYMfD7/aisrMTkyZObj+/YsQNDhgzp8venzBRfd3xBRSXqgyr8Hql53fHGqMl1x7NMfMER07IRUWMLjuiGCYd3molaSSiQI5EI8vPzMW3aNEybNg1Tp07tljAGYj+hXHPNNViwYAH69++P8ePH44033sA///lPLFq0qFtqoMwUX3f82DxkC253bN1xzkPOTpIoNC3TCai6iagam9ts2Vx0hJwvoUBevnw5/v73v+Mf//gHXn75ZViWhXHjxmHGjBmYMWMGSktLu3Qi//XXXw+v14uHH34YtbW1GDZsGB599FGceuqpXfaelB247njP5VYkuBUJuT4XIpqBaNSAxl4zOVhCgTxq1CiMGjUK1157LcLhMNasWYN//OMfeOWVV/DYY48hEAhg+vTpzQE9cODAlBd65ZVX4sorr0z5dSn7cd3xnk0UBfg9CvweBbphIhI1EdF0mFyqkxwm6aesfT4fZs2ahVmzZgEAPvvsM/zzn//Ehx9+iN/+9rcwDANbtmxJeaFERJ2lyBKUHAm5ttK0VKcZexCMyUwO0KFpTwBQX1+Pjz76CBs3bsTmzZuxZcsWWJaFU045JZX1ERGlXHy/Zq9bgWFaiERj95oNy2KvmdIm4UDetWsX/v3vfzf/2rt3L4DYdKTTTjsN1157LaZOnZpdj70TUdaTJRG5fhdyfAoiqoEI7zVTmiQUyFOnTsXRo0ebd3eaPn06/vu//xvTpk1D7969u7pGIqIuJwjHNrhQNQONUQOqZjCYqdskFMjTpk3D6aefjtNOOw2DBw/u6pqIiNIqPq9Z0000RmPzmnmfmbpaQoE8f/78rq6DiMhxXIoElyJB95pobLrP7PQlOilzdfihLiKinkKRJeTnSMjxKmiM6IioBkwGM6UYA5mIKEGyJCIvx40cr4Jw1ECjqsMynb12NmUOBjIRUZKkpiezfV4F4YiOsKrDZDBTJzGQiYg6SBIF5Ppd8HsVhFUd4QjnMlPHMZCJiDpJbNrUwu9REFYNhCM6dJPBTMlhIBMRpYggxNbN9rllRDUDjZHYIiMMZkoEA5nIQSzLxu6aBgQbNQT8LhQX5nEzjAzUcmnOSNNQtqry75G+GgOZyCEqqw6hYnUVaupCMEwLsiSisF8O927OcPFglkQTAb+HP2DRCYnpLoCIYmG8oKIS1fuD8Lgl9Mp1w+OWUH0giAUVlaisOpTuEqmTXLIIrwvok+eBzyND7MI95CkzMZCJ0syybFSsrkIkaqAgzw23IkEUBbgVCQUBNyKqgYrVVVwhKguYpglZEtEr14M++Qxmao2BTJRmu2saUFMXQq5fgXDcP86CICDXq6CmLoTdNQ1pqpC6giJLTcHsRY5XgcSh7B6PgUyUZsFGDYZpQZHa/99RkUUYpoVgo9bNlVF3UOTY6l998r3I9boYzD0YA5kozQJ+F2RJhG5a7b6uG7EHvAJ+VzdXRt1JlkQEclzom+9FwOeCJAlgNPcsDGSiNCsuzENhvxwcDeuwj5uwats2jkZ0FPbLQXFhXpoqpO4UX5azb74PAb8LiiSCt5l7BgYyUZqJooDyshJ43TLqgypUzYRl2VA1E/VBFT63jPKyEk6X6WEkUUCOz4W+vbzIy3HDJTOYsx0DmcgBSkv64obyUhQNDCCqmTh8VEVUM1E0MIDry0s5D7kHi6/+1beXLzYdzsUns7MVFwYhcojSkr4YN6wPV+qiE4ovMqIbFiJRA2GN2z9mEwYykYOIooDhg/PTXQY5nCKLUHJcyLEUhKM6QtHY9o+U2RjIREQZSmy6z+z1xPZlbozqMLmATMZiIBMRZbj4vsw+j4xQREdENRjMGYgPdRERZQlJOrbIiN+r8OGvDMMeMhFRlpElEfk5bvg9sXvMYdXgWugZgIFMRJSl4sty+r0KGpuGsi2LT2U7FQOZiCjLyU1D2Tk+FyJRHeGoAcOyYDOZHYWBTETUQ8RX//J7FURUAxHVgKab4Gi2MzCQiYh6GEEQ4PMo8HkUaLoZC2etaTib4Zw2DGQioh7MpUhwKRJyLRcimoFo1IBmsNecDgxkIiKCKMbWzPa37DXzIbBuxUAmIqJWmnvNPhciqo6IakIzTA5ndzEGMhERtUsUBfi9Lvi9gKoZCIV1qAzmLsNAJiKir+V2yXC7ZERUHaGwDt3ktKlUy7ilM/fs2YOJEydi2bJlaa3Dsmzsrgli5/4odtcEuQoOEfUIXreCPvle5OW44ZJFcHXO1MmoHrKu67jlllsQDofTWkdl1SFUrK7C57VHoaoaVnwUwuD+uSgvK+FG8kSU9QQh9gCYzy1DtE14PW4GcwpkVA/50Ucfhd/vT2sNlVWHsKCiEtX7g/C6ROR4RXhdIqoPBLGgohKVVYfSWh8RUXcRBAFej4Q8v4SA3w1JZCp3RsYE8vr167FkyRL8/ve/T1sNlmWjYnUVIlEDBXluuBQJgiDApUgoCLgRUQ1UrK7i8DUR9SiGrsPvkdE334tcr4vB3EEZMWQdDAZx66234q677sLAgQM7fT3btjs07L27JojPa48ixyvBtu3m4LUsG6II+D0SPq89ii27alFcGOh0nekQiURafcwWbFfmyMY2AT2nXbII+N1AOGoiHNUzcl9mNaoCAKLRaEqu5/P5EjovIwL517/+NSZMmIDvfve7KbmeruvYunVr0l+3c38UqqpBEUWY5rGfAFUt9pdn2zZU1cKn23ZBDXpSUmu6VFdXp7uELsF2ZY5sbBPQc9olSRIgSIhqFsKqDl030lNYBwVyPNi7dy9M0+z0tSZNmpTQeY4P5OXLl2PDhg147bXXUnZNRVEwfPjwpL/OHQhixUchyLIIlyLBsmyomgq3yw1RFKDqJtxuC2NHDcvoHnJ1dTWKiorg9XrTXU7KsF2ZIxvbBPTsdhmmhVDEQFTLjH2Z1aiKYMMXGDJkCDye7utcOT6Qly5divr6esycObPV8Xnz5uHpp5/GG2+8kfQ1BUFIeAihpTHDvBjcPxfVB4IocMkQm+7Ai6IAQRDQGDVRNDCAMcP6Q8zweyher7dDf0ZOx3ZljmxsE9Bz2xXIzbDFRRoAj8fTrX9Xjg/kBx98sM04/re//W389Kc/xbnnntuttYiigPKyEiyoqER9UIXfE7uXrOomGqMmfG4Z5WUlGR/GRERdIb64SFQz0JgpwdyNHB/I/fv3b/d4QUEBCgsLu7kaoLSkL24oL20xD9mC222haGCA85CJiBLgccnwMJjbcHwgO1FpSV+MG9YHW3bV4tNtuzB21LCsGKYmIupO8WBWNQOhiAFNN3r0to8ZGcjbt29PdwkQRQHFhQGoQQ+KCwMMYyKiDooPZauagcaoAVXrmcGckYFMRETZJx7Mmm6iMaojqpqwetBYNgOZiIgcJb4fs+410Rg1EIkaPSKYGchERORIiiwhP0eC36MgHNURUY2MXPkrUQxkIiJyNEUWkZfjRo5XQWPEQFjNzCU5vw4DmYiIMoIkiQjkuOD3KQhHdDSqOizTRrZEMwOZiIgyiiQKyPW74PMqiER1NEYNmKaV8cHMQCYioowkiQJyfC74PAoiqo7GiAHDsjJ2kREGMhERZTRRFOD3xoI5rBoIR3ToZuYFMwOZiIiygiAI8HsU+NxNy3JGDGgZtCwnA5mIiLKKIAjwuhV43bGh7HDEyIj1shnIRESUteLBHFsvW4eqOzeYGchERJT1Wm79GArrjhzKFtNdQKayLBu7a4LYuT+K3TVBWFk4SZ2IKNt4XDL65HvRK9cNlyxCcNC+QOwhd0Bl1aEW+yFrWPFRCIP753I/ZCKiDOF1K/C4ZISjTdOlHDCPmT3kJFVWHcKCikpU7w/C6xKR4xXhdYmoPhDEgopKVFYdSneJRESUAEGITZfqk+9Frs8FKc3b6DKQk2BZNipWVyESNVCQ54ZLkSAIAlyKhIKAGxHVQMXqKg5fExFlELFp5a+++V7keJW0BTMDOQm7axpQUxdCrl+BcNyNB0EQkOtVUFMXwu6ahjRVSEREHSVJsU0s+uR5ocjdH48M5CQEGzUYpgVFav+PTZFFGKaFYKPWzZUREVGqSJIACWa3vy8DOQkBvwuyJEI3rXZf1w0LsiQi4Hd1c2VERJRKltX+v/NdiYGchOLCPBT2y8HRsA77uAlstm3jaERHYb8cFBfmpalCIiLKVAzkJIiigPKyEnjdMuqDatOKLzZU3UR9UIXPLaO8rARimp/UIyKizMNATlJpSV/cUF6KooEBRDULoYiFqGahaGAA15eXch4yERF1CBcG6YDSkr4YN6wPtuyqxafbdmHsqGEYM6w/e8ZERNRhDOQOEkUBxYUBqEEPigsDDGMiIuoUDlkTERE5AAOZiIjIARjIREREDsBAJiIicgAGMhERkQMwkImIiByAgUxEROQADGQiIiIHYCATERE5AAOZiIjIARjIREREDsBAJiIicgAGMhERkQNkRCAfOXIEv/rVr3DmmWfilFNOwX/+539iw4YN6S6LiIgoZTIikG+++WZUVlbioYceQkVFBcaOHYurr74au3btSndpREREKeH4QP7ss8/wz3/+E/PmzcPkyZNRXFyMO++8E/3798frr7+e7vKIiIhSwvGB3KtXLzz55JP4xje+0XxMEATYto2GhoY0VkZERJQ6croL+DqBQADf/OY3Wx1bsWIF9u7dixkzZnTomrZtIxwOd7q2SCTS6mM2yMY2AWxXJsnGNgFsVyZJdZt8Pl9C5wm2bdspecdusnHjRlxzzTWYPn06Fi5cmPTXf/zxx9A0rQsqIyIiamvSpEkJnef4HnJLq1atwi233ILS0lI89NBDHb6OoigYPnx4p+uJRCKorq5GUVERvF5vp6/nBNnYJoDtyiTZ2CaA7cok6WpTxgTyc889h3vvvRezZ8/Ggw8+CJfL1eFrCYKQ8BBCIrxeb0qv5wTZ2CaA7cok2dgmgO3KJN3dJsc/1AUAL7zwAn7729/ihz/8If74xz92KoyJiIicyPE95D179uB3v/sdZs+ejR//+Meor69vfs3j8SA3NzeN1REREaWG4wP57bffhq7rWLlyJVauXNnqtYsuugj3339/miojIiJKHccH8k9+8hP85Cc/SXcZREREXSoj7iETERFlOwYyERGRAzCQiYiIHICBTERE5AAMZCIiIgdgIBMRETkAA5mIiMgBGMhEREQOwEAmIiJyAAYyERGRAzCQiYiIHICBTERE5AAMZCIiIgdgIBMRETkAA5mIiMgBGMhEREQOwEAmIiJyAAYyERGRAzCQiYiIHICBTERE5AAMZCIiIgdgIBMRETkAA5mIiMgBGMhEREQOwEAmIiJyAAYyERGRAzCQiYiIHICBTERE5AAMZCIiIgdgIBMRETkAA5mIiMgBGMhEREQOwEAmIiJyAAYyERGRAzCQiYiIHICBTERE5AByugsgIqKuZVk2dtc0INioIeB3obgwD6IopLssOk5GBLJlWXjsscfw0ksvIRgMYtKkSZg3bx5OPvnkdJdGRORolVWHULG6CjV1IRimBVkSUdgvB+VlJSgt6Zvu8qiFjBiyXrhwIRYvXox77rkHS5YsgSAI+NGPfgRN09JdGhGRY1VWHcKCikpU7w/C45bQK9cNj1tC9YEgFlRUorLqULpLpBYcH8iapuEvf/kLbrzxRnzzm9/EqFGj8PDDD6O2thYrV65Md3lERI5kWTYqVlchEjVQkOeGW5EgigLcioSCgBsR1UDF6ipYlp3uUqmJ44est23bhsbGRkybNq35WCAQwJgxY7B+/Xqcd955SV/Ttm2Ew+FO1xaJRFp9zAbZ2CaA7cok2dgmoPvbtbsmiM9rjyLHK8G2bdh26+D1eyR8XnsUW3bVorgw0OH3yca/r1S3yefzJXSe4wP54MGDAICBAwe2Ot6vXz8cOHCgQ9fUdR1bt27tdG1x1dXVKbuWU2RjmwC2K5NkY5uA7mvXzv1RqKoGRRRhmm0f4LJtG6pq4dNtu6AGPZ1+v2z8+0pVmyZNmpTQeY4P5PhPKC6Xq9Vxt9uNhoaGDl1TURQMHz48JbVVV1ejqKgIXq+309dzgmxsE8B2ZZJsbBPQ/e1yB4JY8VEIsizCpUhtXld1E263hbGjhnW6h5xtf1/papPjA9njif3kpmla8+8BQFXVDv9BCYKQ8BBCIrxeb0qv5wTZ2CaA7cok2dgmoPvaNWaYF4P756L6QBAFLhmCcKyXbNs2GqMmigYGMGZY/5RMgcrGv6/ubpPjH+qKD1XX1dW1Ol5XV4cBAwakoyQiIscTRQHlZSXwumXUB1WomgnLsqFqJuqDKnxuGeVlJZyP7CCOD+RRo0YhJycHa9eubT4WDAaxZcsWTJ48OY2VERE5W2lJX9xQXoqigQFENROHj6qIarGe8fXlpZyH7DCOH7J2uVy47LLL8OCDD6J3794oLCzEAw88gAEDBmD27NnpLo+IyNFKS/pi3LA+XKkrAzg+kAHgpz/9KQzDwF133YVoNIopU6bg6aefbvOgFxERtSWKAoYPzk93GfQ1MiKQJUnCL37xC/ziF79IdylERERdwvH3kImIiHoCBjIREZEDMJCJiIgcgIFMRETkAAxkIiIiB2AgExEROQADmYiIyAEYyERERA7AQCYiInIABjIREZEDCLZt2+kuojv9+9//hm3bKVkH27Zt6LoORVFa7TWaybKxTQDblUmysU0A25VJUt0ml8uFkSNHfu15GbGWdSql8htGEISs2+AiG9sEsF2ZJBvbBLBdmSRdbepxPWQiIiIn4j1kIiIiB2AgExEROQADmYiIyAEYyERERA7AQCYiInIABjIREZEDMJCJiIgcgIFMRETkAAxkIiIiB2AgExEROQADmYiIyAEYyERERA7AQO4Ay7Iwf/58nHHGGSgtLcVVV12Fzz77LN1lddqRI0fwq1/9CmeeeSZOOeUU/Od//ic2bNiQ7rJSZs+ePZg4cSKWLVuW7lJSYvny5Tj33HMxbtw4nHfeeVixYkW6S+o0Xdfx8MMPY+bMmZg4cSL+67/+C//+97/TXVaHLVy4EHPmzGl1bOvWrbjsssswYcIEzJw5E08//XSaquu49tq1evVqXHLJJZg4cSLKysrw+9//HtFoNE0VJq+9NrV01113oaysrEtrYCB3wMKFC7F48WLcc889WLJkCQRBwI9+9CNompbu0jrl5ptvRmVlJR566CFUVFRg7NixuPrqq7Fr1650l9Zpuq7jlltuQTgcTncpKfHKK6/gjjvuwA9+8AO8/vrrOPfcc3HzzTfjo48+SndpnfKnP/0JS5cuxT333IPly5ejuLgYP/rRj1BbW5vu0pK2aNEizJ8/v9Wxw4cP48orr0RRURGWLl2KG2+8EY888giWLl2apiqT1167NmzYgLlz5+Lss8/G8uXL8etf/xorVqzA3XffnaYqk9Nem1patWoVXnrppa4vxKakqKpqT5w40X7hhReajzU0NNjjx4+3X3/99TRW1jnV1dX2iBEj7I0bNzYfsyzLnj17tv3HP/4xjZWlxh/+8Ad7zpw59ogRI+ylS5emu5xOsSzL/ta3vmXff//9rY5fddVV9uOPP56mqlLjggsusO+7777mz48ePWqPGDHCfuutt9JYVXIOHjxoX3311faECRPsc845x77sssuaX3v88cftM844w9Z1vfnYH/7wB/vss89OR6lJ+ap2/fznP7evvPLKVucvX77cHjNmjK2qaneXmrCvalNcbW2tPW3aNPuyyy6zv/Wtb3VpPewhJ2nbtm1obGzEtGnTmo8FAgGMGTMG69evT2NlndOrVy88+eST+MY3vtF8TBAE2LaNhoaGNFbWeevXr8eSJUvw+9//Pt2lpMTu3btRU1OD7373u62OP/300/jxj3+cpqpSIz8/H++99x727dsH0zSxZMkSuFwujB49Ot2lJezTTz9FXl4eXn31VZSWlrZ6bcOGDZgyZQpkWW4+Nm3aNOzZswf19fXdXWpSvqpdV111FW699dY2X2MYBkKhUHeVmLSvahMA2LaN22+/HRdeeCGmTp3a5fXIX38KtXTw4EEAwMCBA1sd79evHw4cOJCOklIiEAjgm9/8ZqtjK1aswN69ezFjxow0VdV5wWAQt956K+666642f2eZqrq6GgAQDodx9dVXY8uWLRg0aBCuu+66Lr/H1dXuvPNO/OxnP8OsWbMgSRJEUcQjjzyCIUOGpLu0hJWVlZ3w7+HgwYMYMWJEq2P9+vUDAOzfvx8FBQVdXl9HfVW7xowZ0+pzTdPw17/+FWPHjkXv3r27o7wO+ao2AbGh7EOHDuHxxx/HE0880eX1sIecpEgkAgBwuVytjrvdbqiqmo6SusTGjRtxxx13YNasWRn9j/yvf/1rTJgwoU1vMpPFexy33XYbzj//fPzlL3/B6aefjuuvvx5r1qxJc3Wds2vXLgQCASxYsABLlizBxRdfjNtuuw3btm1Ld2kpEY1G2/23A0DW/PthGAZuvfVW7Ny5E/PmzUt3OR22bds2PPbYY3jggQfa/J11FfaQk+TxeADEfgKM/x6I/c/k9XrTVVZKrVq1CrfccgtKS0vx0EMPpbucDlu+fDk2bNiA1157Ld2lpJSiKACAq6++GhdddBEAYPTo0diyZQv++te/Yvr06eksr8Nqamrwi1/8AosWLcLkyZMBAOPGjcPOnTvx6KOPYsGCBWmusPM8Hk+bhz/jQezz+dJRUkqFQiHcdNNNWLt2LebPn9/uMHAmUFUVt9xyC6677jqMGjWq296XPeQkxYc96+rqWh2vq6vDgAED0lFSSj333HO48cYbceaZZ+LPf/5zqx86Ms3SpUtRX1/fPIVm4sSJAIB58+bhvPPOS3N1HRf/Pjt+6HP48OHYt29fOkpKic2bN0PXdYwbN67V8dLS0uZh+kw3YMCAdv/tAID+/funo6SUqaurww9/+EN89NFH+POf/5zRI2uVlZWoqqrCY4891vxvxxNPPIH9+/dj4sSJePXVV7vkfdlDTtKoUaOQk5ODtWvXNt/XCgaD2LJlCy677LI0V9c5L7zwAn77299izpw5uOOOOyCKmf3z2oMPPthmHuS3v/1t/PSnP8W5556bpqo6b8yYMfD7/aisrGzuSQLAjh07Mupe6/HiP+xu374d48ePbz6+Y8cOnHzyyekqK6WmTJmCxYsXwzRNSJIEAFizZg2GDh3q6PvHX6ehoQFXXHEFQqEQXnjhBYwcOTLdJXXK+PHj8c4777Q69uyzz+Kdd97Bs88+22V/VwzkJLlcLlx22WV48MEH0bt3bxQWFuKBBx7AgAEDMHv27HSX12F79uzB7373O8yePRs//vGPWz3x6fF4kJubm8bqOuZEPY6CggIUFhZ2czWp4/F4cM0112DBggXo378/xo8fjzfeeAP//Oc/sWjRonSX12Hjx4/H5MmTcdttt2HevHkYMGAAli9fjjVr1uCFF15Id3kpcckll+Cpp57CnXfeiWuuuQabN2/G3/72t4yZr3si9913Hz7//HM89dRT6N27Nw4dOtT8Wu/evZt/+MgUHo+nzQ+BeXl5kGW5S384ZCB3wE9/+lMYhoG77roL0WgUU6ZMwdNPP91tN/67wttvvw1d17Fy5UqsXLmy1WsXXXQR7r///jRVRu25/vrr4fV68fDDD6O2thbDhg3Do48+ilNPPTXdpXWYKIpYuHAh/vjHP+KXv/wlGhoaMGLECCxatAgTJkxId3kpUVBQgKeeegr33nsvLrroIvTt2xe33npr87MAmciyLLz55pvQdR1XXHFFm9ffffddDBo0KA2VZR7Btm073UUQERH1dJl9k5CIiChLMJCJiIgcgIFMRETkAAxkIiIiB2AgExEROQADmYiIyAEYyERERA7AQCYiInIABjJRFpozZw5GjhyJSy+99ITn/OxnP8PIkSNx++23A4jtDRv/fXtGjhyJRx99tM17nOjXhg0bWn39I488gpEjR2b0lnxEXYlLZxJlKVEUsWnTJhw4cKB544a4SCSC999/v9PvMWbMmBMG7PDhw5t/b1kWli9fjhEjRuC1117DrbfeCr/f3+n3J8om7CETZakxY8bA7XbjrbfeavPa6tWr4Xa7O73lX05ODiZMmNDur5ycnObz/vWvf2H//v2YN28eotFo1u1RTZQKDGSiLOXz+fDNb34TK1asaPPam2++iXPOOQey3D2DZEuXLkVxcTEmT56M6dOnY/Hixd3yvkSZhIFMlMXOPfdcVFZWYv/+/c3HQqEQ/v73v+P888/v9PVt24ZhGG1+tdyzpqGhAatWrWre0ejiiy/G1q1bUVlZ2en3J8omDGSiLDZz5kz4fL5Ww9YrV65E7969MWnSpE5ff/369Rg7dmybX6+++mrzOa+99hpM08SFF14IAJg9ezby8vLYSyY6Dh/qIspiHo8HZWVlWLFiBa666ioAwBtvvIFzzz0XgiAkfb3jv2bs2LG4++6725w3ePDg5t8vXboUU6ZMgdfrRTAYBACcddZZeOONN/DLX/4SgUAg6TqIshEDmSjLfec738ENN9yAffv2we/3Y82aNbjpppvanOfz+aBpWrvXiB/3er2tjvv9fowbN+6E771t2zZs2bIFADBlypQ2r7/88svtbmpP1BMxkImy3Jlnnonc3Fy8/fbbyM3NxaBBg/CNb3yjzXl9+vRBXV1du9c4ePBg8znJqKiogNfrxZ/+9CeIYus7ZHfffTeWLFnCQCZqwkAmynIulwuzZs3CO++8A5/Ph/POO6/d86ZOnYrHH38c+/fvx0knndTqtbfeeguSJLXbyz0RTdPw2muvoaysDNOnT2/z+sUXX4wHHngA69atw9SpU5NrFFEW4kNdRD1A/GnrtWvXnjCQL7vsMhQUFOCyyy7DCy+8gPXr1+ODDz7A73//e8yfPx/XXHNNm6D+KqtWrcKRI0dO+H4XXHABJEniw11ETdhDJuoBTjvtNAQCAQwcOBDDhg1r95xAIICKigosXLgQf/nLX1BXVwePx4Pi4mLcc889+N73vpfUey5btgx5eXk444wz2n29X79+OO200/DOO++gvr4eBQUFyTaLKKsIdssJg0RERJQWHLImIiJyAAYyERGRAzCQiYiIHICBTERE5AAMZCIiIgdgIBMRETkAA5mIiMgBGMhEREQOwEAmIiJyAAYyERGRAzCQiYiIHOD/A4HpNoPN/j0GAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 500x500 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "x = 'MUFA'\n",
    "y = 'WOMAC_pain'\n",
    "df_sub = df_merge[[x,y]].dropna()\n",
    "sns.lmplot(x=x,y=y,data=df_sub)\n",
    "\n",
    "x = df_sub[x]\n",
    "y = df_sub[y]\n",
    "\n",
    "print(scipy.stats.pearsonr(x,y))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b3b3a55-d05d-4a16-992e-78ac2d3a3242",
   "metadata": {},
   "source": [
    "###### SCRATCH ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "c1201df2-f2bf-422f-b462-1e8e36d172d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SignificanceResult(statistic=0.2922404895854228, pvalue=0.38318565388964687)\n",
      "SignificanceResult(statistic=nan, pvalue=nan)\n",
      "Akker\n",
      "g__Butyribacter\n",
      "Anaerostipes\n",
      "Limivivens\n",
      "['d__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia', 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Butyribacter', 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Anaerostipes', 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens']\n",
      "PearsonRResult(statistic=0.13534457680522186, pvalue=0.6915318450819884)\n",
      "SignificanceResult(statistic=-0.21198381771691024, pvalue=0.531473949199541)\n",
      "PearsonRResult(statistic=-0.45422619985696916, pvalue=0.1604762923340049)\n",
      "SignificanceResult(statistic=-0.26027668603701715, pvalue=0.4395397689851793)\n",
      "PearsonRResult(statistic=0.5910664538572264, pvalue=0.05549416471219779)\n",
      "SignificanceResult(statistic=0.5616496909219845, pvalue=0.07217236648510983)\n"
     ]
    }
   ],
   "source": [
    "# does diet correlate with pain\n",
    "df_merge = pd.concat([df_diff, gt_to_df_diet['stool']['diff'], df_dx],axis=1)\n",
    "#print(df_merge.corr())\n",
    "#print(scipy.stats.pearsonr(df_merge['anti_inflammatory'],df_merge['WOMAC_pain']))\n",
    "# sns.scatterplot(data=df_merge,x='anti_inflammatory',y='WOMAC_pain')\n",
    "# df_merge = df_merge.dropna()\n",
    "print(scipy.stats.spearmanr(df_merge['anti_inflammatory'],df_merge['WOMAC_pain'], nan_policy='omit'))\n",
    "# the answer is not really\n",
    "\n",
    "# create difference of alpha divs\n",
    "df = df_merge.copy()\n",
    "for g in ['stool','saliva']:\n",
    "    df[g + '_taxa_alpha_diff'] = g_to_dfd[g + '_adh']['paired_alpha']['shannon_entropy_diff']\n",
    "\n",
    "for g in ['stool','saliva','plasma']:\n",
    "    df[g + '_meta_alpha_diff'] = gt_to_meta_norm[g]['post']['shannon_entropy'] - gt_to_meta_norm[g]['pre']['shannon_entropy']\n",
    "\n",
    "\n",
    "# do changes in diversity correlate with pain? YES\n",
    "# print(scipy.stats.pearsonr(df['stool_taxa_alpha_diff'],df['WOMAC_pain']))\n",
    "print(scipy.stats.spearmanr(df['stool_taxa_alpha_diff'],df['WOMAC_pain']))\n",
    "# sns.scatterplot(data=df,x='stool_taxa_alpha_diff',y='WOMAC_pain')\n",
    "\n",
    "taxa_int = []\n",
    "queries = ['Akker','g__Butyribacter','Anaerostipes','Limivivens']\n",
    "for q in queries:\n",
    "    print(q)\n",
    "    \n",
    "    for x in df_otu.columns.values:\n",
    "        if q in x:\n",
    "            taxa.append(x)\n",
    "            taxa_int.append(x)\n",
    "            \n",
    "# focus on taxa of interest\n",
    "print(taxa_int)\n",
    "df2 = pd.concat([df,gt_to_otu['stool']['diff'][taxa_int]],axis=1)\n",
    "df2=df2.dropna()\n",
    "df2.head()\n",
    "# df2.corr()\n",
    "\n",
    "# do changes in taxa of interest correlate with pain improvement?\n",
    "#t= 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens'\n",
    "t = 'd__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia'\n",
    "#t = 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Anaerostipes'\n",
    "#t = 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Butyribacter'\n",
    "df3 = pd.concat([df2],axis=1).dropna()\n",
    "print(scipy.stats.pearsonr(df3[t],df3['WOMAC_pain']))\n",
    "print(scipy.stats.spearmanr(df3[t],df3['WOMAC_pain']))\n",
    "# sns.scatterplot(data=df3,x=t,y='WOMAC_pain',hue='WOMAC_P_Response')\n",
    "# here and there\n",
    "\n",
    "# tack on metabolite changes\n",
    "df4 = df2.copy()\n",
    "# df4 = pd.concat([df2,df],axis=1)\n",
    "meta_int = ['H_Tetrahydroxylated_bile_acid',\n",
    "            # 'H_Glycocholate', # these two are the same so we comment one out\n",
    "            'H_Glycocholic_acid',\n",
    "            'H_Glycochenodeoxycholate',\n",
    "            'H_Deoxycholate',\n",
    "            'H_Pyridoxine'\n",
    "           ]\n",
    "for g in ['stool','saliva','plasma']:\n",
    "    subdf = gt_to_meta[g]['diff'].copy()\n",
    "    for b in meta_int:# + int_ba:\n",
    "        try:\n",
    "            df4[b + '_diff_' + g] = subdf[b]\n",
    "        except:\n",
    "            continue\n",
    "\n",
    "# specific corr of meta with pain\n",
    "t = 'H_Tetrahydroxylated_bile_acid_diff_stool'\n",
    "# t = 'H_Glycocholate_diff_plasma'\n",
    "t = 'H_Glycocholic_acid_diff_plasma'\n",
    "t = 'H_Glycochenodeoxycholate_diff_plasma'\n",
    "t = 'H_Deoxycholate_diff_plasma'\n",
    "df5 = pd.concat([df4],axis=1).dropna()\n",
    "print(scipy.stats.pearsonr(df5[t],df5['WOMAC_pain']))\n",
    "print(scipy.stats.spearmanr(df5[t],df5['WOMAC_pain']))\n",
    "# sns.scatterplot(data=df5,x=t,y='WOMAC_pain',hue='WOMAC_P_Response')\n",
    "\n",
    "# do changes in plasma meta diversities correlate with pain imprv? yes but less than stool taxa changes\n",
    "print(scipy.stats.pearsonr(df4['plasma_meta_alpha_diff'],df5['WOMAC_pain']))\n",
    "print(scipy.stats.spearmanr(df4['plasma_meta_alpha_diff'],df5['WOMAC_pain']))\n",
    "# sns.scatterplot(data=df5,x='plasma_meta_alpha_diff',y='WOMAC_pain',hue='WOMAC_P_Response')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "a15150aa-6a23-401c-9eb5-cc69413f1f03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WOMAC_Pain</th>\n",
       "      <th>Stool_Tetrahydroxylated_Bile_Acid</th>\n",
       "      <th>Stool_Pyridoxine</th>\n",
       "      <th>Plasma_Glycocholic_Acid</th>\n",
       "      <th>Plasma_Glycochenodeoxycholate</th>\n",
       "      <th>Plasma_Deoxycholate</th>\n",
       "      <th>Chicken</th>\n",
       "      <th>Eggs</th>\n",
       "      <th>Plant_Protein</th>\n",
       "      <th>PUFA</th>\n",
       "      <th>Omega6</th>\n",
       "      <th>Gut_Lachnospiraceae_Limivivens</th>\n",
       "      <th>Gut_Lachnospiraceae_Anaerostipes</th>\n",
       "      <th>Gut_Lachonospiraceae_Butyribacter</th>\n",
       "      <th>Gut_Akkermansia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>-1.5</td>\n",
       "      <td>0.070529</td>\n",
       "      <td>0.298280</td>\n",
       "      <td>-0.079546</td>\n",
       "      <td>-0.646926</td>\n",
       "      <td>-1.303523</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.4</td>\n",
       "      <td>54.8</td>\n",
       "      <td>-0.000426</td>\n",
       "      <td>-0.003440</td>\n",
       "      <td>-0.006675</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.159916</td>\n",
       "      <td>-0.199607</td>\n",
       "      <td>1.236516</td>\n",
       "      <td>0.999981</td>\n",
       "      <td>-1.234030</td>\n",
       "      <td>-15.6</td>\n",
       "      <td>10.4</td>\n",
       "      <td>0.4</td>\n",
       "      <td>14.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>-0.000971</td>\n",
       "      <td>-0.003430</td>\n",
       "      <td>0.001161</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>-2.5</td>\n",
       "      <td>-0.044181</td>\n",
       "      <td>0.418271</td>\n",
       "      <td>1.222726</td>\n",
       "      <td>0.508608</td>\n",
       "      <td>0.123904</td>\n",
       "      <td>-10.0</td>\n",
       "      <td>-1.6</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.4</td>\n",
       "      <td>68.8</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.002736</td>\n",
       "      <td>-0.000358</td>\n",
       "      <td>0.017790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD007</th>\n",
       "      <td>-6.0</td>\n",
       "      <td>-1.611473</td>\n",
       "      <td>-0.651325</td>\n",
       "      <td>0.102244</td>\n",
       "      <td>-0.551160</td>\n",
       "      <td>0.102244</td>\n",
       "      <td>-5.6</td>\n",
       "      <td>-11.2</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>25.7</td>\n",
       "      <td>-0.000320</td>\n",
       "      <td>0.005345</td>\n",
       "      <td>0.004751</td>\n",
       "      <td>0.003349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD009</th>\n",
       "      <td>-5.5</td>\n",
       "      <td>-0.379576</td>\n",
       "      <td>-0.134895</td>\n",
       "      <td>-0.241268</td>\n",
       "      <td>0.133767</td>\n",
       "      <td>0.801817</td>\n",
       "      <td>-6.0</td>\n",
       "      <td>-16.8</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.001166</td>\n",
       "      <td>0.004987</td>\n",
       "      <td>-0.000163</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        WOMAC_Pain  Stool_Tetrahydroxylated_Bile_Acid  Stool_Pyridoxine  \\\n",
       "OAD003        -1.5                           0.070529          0.298280   \n",
       "OAD004         1.0                           1.159916         -0.199607   \n",
       "OAD006        -2.5                          -0.044181          0.418271   \n",
       "OAD007        -6.0                          -1.611473         -0.651325   \n",
       "OAD009        -5.5                          -0.379576         -0.134895   \n",
       "\n",
       "        Plasma_Glycocholic_Acid  Plasma_Glycochenodeoxycholate  \\\n",
       "OAD003                -0.079546                      -0.646926   \n",
       "OAD004                 1.236516                       0.999981   \n",
       "OAD006                 1.222726                       0.508608   \n",
       "OAD007                 0.102244                      -0.551160   \n",
       "OAD009                -0.241268                       0.133767   \n",
       "\n",
       "        Plasma_Deoxycholate  Chicken  Eggs  Plant_Protein  PUFA  Omega6  \\\n",
       "OAD003            -1.303523      6.0   8.0            0.0   8.4    54.8   \n",
       "OAD004            -1.234030    -15.6  10.4            0.4  14.0    56.0   \n",
       "OAD006             0.123904    -10.0  -1.6            6.0   8.4    68.8   \n",
       "OAD007             0.102244     -5.6 -11.2            6.0   2.8    25.7   \n",
       "OAD009             0.801817     -6.0 -16.8            2.0   0.0    36.0   \n",
       "\n",
       "        Gut_Lachnospiraceae_Limivivens  Gut_Lachnospiraceae_Anaerostipes  \\\n",
       "OAD003                       -0.000426                         -0.003440   \n",
       "OAD004                       -0.000971                         -0.003430   \n",
       "OAD006                        0.000000                         -0.002736   \n",
       "OAD007                       -0.000320                          0.005345   \n",
       "OAD009                        0.000000                         -0.001166   \n",
       "\n",
       "        Gut_Lachonospiraceae_Butyribacter  Gut_Akkermansia  \n",
       "OAD003                          -0.006675         0.000000  \n",
       "OAD004                           0.001161         0.000000  \n",
       "OAD006                          -0.000358         0.017790  \n",
       "OAD007                           0.004751         0.003349  \n",
       "OAD009                           0.004987        -0.000163  "
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df for R\n",
    "taxa = ['d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens',\n",
    "      'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Anaerostipes',\n",
    "      'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Butyribacter',\n",
    "      'd__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia']\n",
    "keep = ['Sex','Age','DM','HTN','Race','Height','Weight']\n",
    "\n",
    "df_merge = df5.copy()\n",
    "df_map = type_to_df_map[g].copy()\n",
    "df_map = df_map[df_map.duplicated(subset='HostSubjectId', keep=False)]\n",
    "df_map = df_map.drop_duplicates(subset='HostSubjectId')\n",
    "df_map.index = df_map.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "\n",
    "df_map = df_map.loc[:,keep]\n",
    "\n",
    "df_merge = pd.concat([df_merge, df_map, gt_to_df_diet['stool']['diff']],axis=1)\n",
    "\n",
    "v_to_v ={\n",
    "   'WOMAC_pain': 'WOMAC_Pain',#_Improvement',\n",
    "    #'anti_inflammatory': 'Anti_Inflammatory_Diet_Score',#_Change',\n",
    "    #'plasma_meta_alpha_diff': 'Plasma_Metabolome_Alpha_Diversity',#_Difference',\n",
    "    #'stool_meta_alpha_diff': 'Gut_Metabolome_Alpha_Diversity',#Difference',\n",
    "    #'stool_taxa_alpha_diff': 'Gut_Microbiome_Alpha_Diversity',#Difference',\n",
    "    'H_Tetrahydroxylated_bile_acid_diff_stool': 'Stool_Tetrahydroxylated_Bile_Acid',\n",
    "    'H_Pyridoxine_diff_stool': 'Stool_Pyridoxine',\n",
    "    'H_Glycocholic_acid_diff_plasma': 'Plasma_Glycocholic_Acid', \n",
    "    'H_Glycochenodeoxycholate_diff_plasma':'Plasma_Glycochenodeoxycholate',\n",
    "    'H_Deoxycholate_diff_plasma':'Plasma_Deoxycholate',\n",
    "    'chicken.1':'Chicken',\n",
    "    'egg': 'Eggs',\n",
    "    'plant_protein': 'Plant_Protein',\n",
    "    'PUFA':'PUFA',\n",
    "    'Total_omega6':'Omega6',\n",
    "      'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens': 'Gut_Lachnospiraceae_Limivivens',\n",
    "      'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Anaerostipes': 'Gut_Lachnospiraceae_Anaerostipes',\n",
    "      'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Butyribacter': 'Gut_Lachonospiraceae_Butyribacter',\n",
    "      'd__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia': 'Gut_Akkermansia',\n",
    "    \n",
    "\n",
    "}\n",
    "    \n",
    "df_merge2 = df_merge.rename(columns=v_to_v)#.drop('Timepoint',axis=1)\n",
    "df_merge2 = df_merge.dropna()\n",
    "df_merge2.to_csv(path + 'inputs/df_curated_R.tsv', sep='\\t')\n",
    "\n",
    "# subset on corr vars\n",
    "corr_vars = list(v_to_v.values())\n",
    "\n",
    "#spec = ['plasma_meta_alpha_diff','stool_meta_alpha_diff','stool_taxa_alpha_diff']\n",
    "#df_merge = pd.concat([df_merge[spec], df_diet, df_meta, df_otu, df_path, df_outcomes],axis=1)\n",
    "df_merge = df_merge.rename(columns=v_to_v)#.drop('Timepoint',axis=1)\n",
    "# corr_vars = list(v_to_v.keys())\n",
    "\n",
    "# df_merge = pd.concat([df5[spec + ['WOMAC_P_Response']], df_diet, df_meta, df_otu, df_path, df_outcomes],axis=1)\n",
    "\n",
    "df_corr = df_merge.loc[:,corr_vars + ['WOMAC_P_Response']]\n",
    "# with label\n",
    "df_corr.to_csv(path + 'inputs/df_corr_R_label.tsv', sep='\\t')\n",
    "df_corr = df_merge.loc[:,corr_vars]\n",
    "df_corr = df_corr.dropna()\n",
    "df_corr = df_corr.loc[:,~df_corr.columns.duplicated()].copy()\n",
    "df_corr.to_csv(path + 'inputs/df_corr_R.tsv', sep='\\t')\n",
    "\n",
    "\n",
    "df_corr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "38839974-0383-4409-be25-52785305ed54",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'taxa'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[231], line 9\u001b[0m\n\u001b[1;32m      1\u001b[0m df_merge\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mvalues\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m# chicken.1\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# Nuts\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# egg\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# plant_protein\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;66;03m# PUFA\u001b[39;00m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m# Total_omega6\u001b[39;00m\n\u001b[0;32m----> 9\u001b[0m sig_feats \u001b[38;5;241m=\u001b[39m [d_to_sig[l] \u001b[38;5;28;01mfor\u001b[39;00m l \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdiet\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtaxa\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmeta\u001b[39m\u001b[38;5;124m'\u001b[39m]]\n\u001b[1;32m     11\u001b[0m all_feats \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m     12\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m sublist \u001b[38;5;129;01min\u001b[39;00m sig_feats:\n",
      "Cell \u001b[0;32mIn[231], line 9\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m      1\u001b[0m df_merge\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mvalues\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m# chicken.1\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# Nuts\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# egg\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# plant_protein\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;66;03m# PUFA\u001b[39;00m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m# Total_omega6\u001b[39;00m\n\u001b[0;32m----> 9\u001b[0m sig_feats \u001b[38;5;241m=\u001b[39m [d_to_sig[l] \u001b[38;5;28;01mfor\u001b[39;00m l \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdiet\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtaxa\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmeta\u001b[39m\u001b[38;5;124m'\u001b[39m]]\n\u001b[1;32m     11\u001b[0m all_feats \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m     12\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m sublist \u001b[38;5;129;01min\u001b[39;00m sig_feats:\n",
      "\u001b[0;31mKeyError\u001b[0m: 'taxa'"
     ]
    }
   ],
   "source": [
    "df_merge.columns.values\n",
    "# chicken.1\n",
    "# Nuts\n",
    "# egg\n",
    "# plant_protein\n",
    "# PUFA\n",
    "# Total_omega6\n",
    "\n",
    "sig_feats = [d_to_sig[l] for l in ['diet','taxa','meta']]\n",
    "\n",
    "all_feats = []\n",
    "for sublist in sig_feats:\n",
    "    all_feats.extend(sublist)\n",
    "\n",
    "spec = ['plasma_meta_alpha_diff', \n",
    "                          'stool_meta_alpha_diff',\n",
    "                          'stool_taxa_alpha_diff']\n",
    "\n",
    "all_feats = all_feats +  spec\n",
    "\n",
    "\n",
    "df_merge = df5.copy()\n",
    "df_merge = pd.concat([df5[spec + ['WOMAC_P_Response']], df_diet, df_meta, df_otu, df_path, df_outcomes],axis=1)\n",
    "\n",
    "#df_map = type_to_df_map[g].copy()\n",
    "#df_map = df_map[df_map.duplicated(subset='HostSubjectId', keep=False)]\n",
    "#df_map = df_map.drop_duplicates(subset='HostSubjectId')\n",
    "#df_map.index = df_map.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "\n",
    "# df_map = df_map.loc[:,all_feats]\n",
    "\n",
    "# df_merge = pd.concat([df_merge, df_map, gt_to_df_diet['stool']['diff']],axis=1)\n",
    "# df_merge = pd.concat([df_merge, gt_to_df_diet['stool']['diff']],axis=1)\n",
    "\n",
    "#df_merge = df_merge.rename(columns=v_to_v)#.drop('Timepoint',axis=1)\n",
    "df_merge = df_merge.dropna()\n",
    "df_merge.to_csv(path + 'inputs/df_curated_R2.tsv', sep='\\t')\n",
    "\n",
    "# subset on corr vars\n",
    "corr_vars = list(v_to_v.values())\n",
    "corr_vars = all_feats\n",
    "df_corr = df_merge.loc[:,corr_vars + ['WOMAC_P_Response']]\n",
    "# with label\n",
    "df_corr.to_csv(path + 'inputs/df_corr_R_label2.tsv', sep='\\t')\n",
    "df_corr = df_merge.loc[:,corr_vars + ['WOMAC_pain']]\n",
    "df_corr.columns = df_corr.columns.map(lambda x: x.split(';')[-1])\n",
    "df_corr.to_csv(path + 'inputs/df_corr_R2.tsv', sep='\\t')\n",
    "\n",
    "# df_corr['WOMAC_pain']\n",
    "# sig_feat\n",
    "df_corr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "7fe0e10e-c138-4e5a-82ec-a26846953603",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['plant_protein', 'PUFA', 'Nuts', 'egg', 'Total_omega6',\n",
       "       'g__Phocaeicola_A_858004', 'g__Bacteroides_H', 'g__Paramuribaculum',\n",
       "       'g__Blautia_A_141781', '__', 'g__Phascolarctobacterium_A',\n",
       "       'g__Anaerostipes', 'g__CAG-269', 'g__Butyricimonas', 'g__UMGS1071',\n",
       "       '__', 'g__Limivivens', 'g__Porcincola', 'g__CAG-1427',\n",
       "       'g__Clostridium_A', 'g__', 'g__Gordonibacter', 'g__Anaerotruncus',\n",
       "       'g__Corynebacterium', 'g__', 'g__Cloacibacterium', 'g__Staphylococcus',\n",
       "       'H_Pyridoxine', 'H_Lenticin', 'H_Tetrahydroxylated_bile_acid',\n",
       "       'H_Arg-C16:0', 'H_Hydroxydodecanoic_acid', 'H_Omega_Hydroxydodecanoate',\n",
       "       'H_Delta_Methyldodecenoic_Acid', 'H_Hydroxydecanoic_acid',\n",
       "       'H_Hydroxydecanoate', 'H_Methylpentanoic_acid',\n",
       "       'H_Monohydroxylated_bile_acid', 'plasma_meta_alpha_diff',\n",
       "       'stool_meta_alpha_diff', 'stool_taxa_alpha_diff', 'WOMAC_pain'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_corr.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "35b8cc76-db9e-4e84-b3be-238e4b490ead",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WOMAC_P_Response</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>DM</th>\n",
       "      <th>HTN</th>\n",
       "      <th>Race</th>\n",
       "      <th>Height</th>\n",
       "      <th>Weight</th>\n",
       "      <th>WOMAC_pain</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Butyribacter</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Anaerostipes</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens</th>\n",
       "      <th>H_Tetrahydroxylated_bile_acid_diff_stool</th>\n",
       "      <th>H_Glycocholic_acid_diff_stool</th>\n",
       "      <th>H_Glycochenodeoxycholate_diff_stool</th>\n",
       "      <th>H_Deoxycholate_diff_stool</th>\n",
       "      <th>H_Glycocholic_acid_diff_plasma</th>\n",
       "      <th>H_Glycochenodeoxycholate_diff_plasma</th>\n",
       "      <th>H_Deoxycholate_diff_plasma</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>No response</td>\n",
       "      <td>pre</td>\n",
       "      <td>F</td>\n",
       "      <td>69.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>White</td>\n",
       "      <td>158.0</td>\n",
       "      <td>71.50</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001161</td>\n",
       "      <td>-0.003430</td>\n",
       "      <td>-0.000971</td>\n",
       "      <td>1.159916</td>\n",
       "      <td>-0.800896</td>\n",
       "      <td>-0.311935</td>\n",
       "      <td>-0.627023</td>\n",
       "      <td>1.236516</td>\n",
       "      <td>0.999981</td>\n",
       "      <td>-1.234030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>Response</td>\n",
       "      <td>pre</td>\n",
       "      <td>F</td>\n",
       "      <td>59.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>White</td>\n",
       "      <td>167.0</td>\n",
       "      <td>79.60</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000358</td>\n",
       "      <td>-0.002736</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.044181</td>\n",
       "      <td>-1.942963</td>\n",
       "      <td>-1.856811</td>\n",
       "      <td>-0.105100</td>\n",
       "      <td>1.222726</td>\n",
       "      <td>0.508608</td>\n",
       "      <td>0.123904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD007</th>\n",
       "      <td>Response</td>\n",
       "      <td>pre</td>\n",
       "      <td>F</td>\n",
       "      <td>57.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>164.0</td>\n",
       "      <td>74.00</td>\n",
       "      <td>-6.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004751</td>\n",
       "      <td>0.005345</td>\n",
       "      <td>-0.000320</td>\n",
       "      <td>-1.611473</td>\n",
       "      <td>0.214088</td>\n",
       "      <td>-0.104483</td>\n",
       "      <td>0.239471</td>\n",
       "      <td>0.102244</td>\n",
       "      <td>-0.551160</td>\n",
       "      <td>0.102244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD009</th>\n",
       "      <td>Response</td>\n",
       "      <td>pre</td>\n",
       "      <td>F</td>\n",
       "      <td>61.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>White</td>\n",
       "      <td>162.0</td>\n",
       "      <td>55.30</td>\n",
       "      <td>-5.5</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004987</td>\n",
       "      <td>-0.001166</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.379576</td>\n",
       "      <td>-0.726880</td>\n",
       "      <td>-0.660090</td>\n",
       "      <td>0.224922</td>\n",
       "      <td>-0.241268</td>\n",
       "      <td>0.133767</td>\n",
       "      <td>0.801817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD012</th>\n",
       "      <td>Response</td>\n",
       "      <td>pre</td>\n",
       "      <td>F</td>\n",
       "      <td>71.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>White</td>\n",
       "      <td>167.0</td>\n",
       "      <td>59.82</td>\n",
       "      <td>-5.5</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000618</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.421385</td>\n",
       "      <td>0.491533</td>\n",
       "      <td>-0.041315</td>\n",
       "      <td>-0.087671</td>\n",
       "      <td>-0.054444</td>\n",
       "      <td>-0.036371</td>\n",
       "      <td>0.437509</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 82 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       WOMAC_P_Response Timepoint Sex   Age   DM  HTN      Race  Height  \\\n",
       "OAD004      No response       pre   F  69.0  0.0  1.0     White   158.0   \n",
       "OAD006         Response       pre   F  59.0  0.0  0.0     White   167.0   \n",
       "OAD007         Response       pre   F  57.0  0.0  0.0  Hispanic   164.0   \n",
       "OAD009         Response       pre   F  61.0  0.0  0.0     White   162.0   \n",
       "OAD012         Response       pre   F  71.0  0.0  0.0     White   167.0   \n",
       "\n",
       "        Weight  WOMAC_pain  ...  \\\n",
       "OAD004   71.50         1.0  ...   \n",
       "OAD006   79.60        -2.5  ...   \n",
       "OAD007   74.00        -6.0  ...   \n",
       "OAD009   55.30        -5.5  ...   \n",
       "OAD012   59.82        -5.5  ...   \n",
       "\n",
       "        d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Butyribacter  \\\n",
       "OAD004                                           0.001161                                                       \n",
       "OAD006                                          -0.000358                                                       \n",
       "OAD007                                           0.004751                                                       \n",
       "OAD009                                           0.004987                                                       \n",
       "OAD012                                           0.000000                                                       \n",
       "\n",
       "        d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Anaerostipes  \\\n",
       "OAD004                                          -0.003430                                                       \n",
       "OAD006                                          -0.002736                                                       \n",
       "OAD007                                           0.005345                                                       \n",
       "OAD009                                          -0.001166                                                       \n",
       "OAD012                                          -0.000618                                                       \n",
       "\n",
       "        d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens  \\\n",
       "OAD004                                          -0.000971                                                     \n",
       "OAD006                                           0.000000                                                     \n",
       "OAD007                                          -0.000320                                                     \n",
       "OAD009                                           0.000000                                                     \n",
       "OAD012                                           0.000000                                                     \n",
       "\n",
       "        H_Tetrahydroxylated_bile_acid_diff_stool  \\\n",
       "OAD004                                  1.159916   \n",
       "OAD006                                 -0.044181   \n",
       "OAD007                                 -1.611473   \n",
       "OAD009                                 -0.379576   \n",
       "OAD012                                  0.421385   \n",
       "\n",
       "        H_Glycocholic_acid_diff_stool  H_Glycochenodeoxycholate_diff_stool  \\\n",
       "OAD004                      -0.800896                            -0.311935   \n",
       "OAD006                      -1.942963                            -1.856811   \n",
       "OAD007                       0.214088                            -0.104483   \n",
       "OAD009                      -0.726880                            -0.660090   \n",
       "OAD012                       0.491533                            -0.041315   \n",
       "\n",
       "        H_Deoxycholate_diff_stool  H_Glycocholic_acid_diff_plasma  \\\n",
       "OAD004                  -0.627023                        1.236516   \n",
       "OAD006                  -0.105100                        1.222726   \n",
       "OAD007                   0.239471                        0.102244   \n",
       "OAD009                   0.224922                       -0.241268   \n",
       "OAD012                  -0.087671                       -0.054444   \n",
       "\n",
       "        H_Glycochenodeoxycholate_diff_plasma  H_Deoxycholate_diff_plasma  \n",
       "OAD004                              0.999981                   -1.234030  \n",
       "OAD006                              0.508608                    0.123904  \n",
       "OAD007                             -0.551160                    0.102244  \n",
       "OAD009                              0.133767                    0.801817  \n",
       "OAD012                             -0.036371                    0.437509  \n",
       "\n",
       "[5 rows x 82 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# prepare a large handy df for R\n",
    "g='stool'\n",
    "df_taxa = gt_to_otu[g]['diff'].copy()\n",
    "taxa = ['d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Limivivens',\n",
    "      'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Anaerostipes',\n",
    "      'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Butyribacter',\n",
    "      'd__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia']\n",
    "\n",
    "df_taxa = df_taxa.loc[:,taxa]\n",
    "df_taxa.columns = df_taxa.columns.map(lambda x: x.split('f__')[-1].replace(';g_',''))\n",
    "\n",
    "df_map = type_to_df_map[g].copy()\n",
    "df_map = df_map[df_map.duplicated(subset='HostSubjectId', keep=False)]\n",
    "df_map = df_map.drop_duplicates(subset='HostSubjectId')\n",
    "df_map.index = df_map.index.map(lambda x: x.split('.p')[0].replace('-',''))\n",
    "\n",
    "# merge\n",
    "df_merge = pd.concat([df_map, df_taxa],axis=1)\n",
    "\n",
    "# subset on interest\n",
    "keep = ['WOMAC_P_Response','Timepoint','Sex','Age','DM','HTN','Race','Height','Weight']\n",
    "\n",
    "# taxa = [y]\n",
    "df_merge = df_merge.loc[:,keep]\n",
    "\n",
    "# dropna\n",
    "df_merge = df_merge.dropna()\n",
    "\n",
    "# merge again\n",
    "df_merge = pd.concat([df_merge,df_diff,df5],axis=1)\n",
    "\n",
    "df_merge.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2c459a7-763e-45b0-be0c-be4fdb5e06f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# manual mapping for microbemasst of taxa to correlations with bile acids\n",
    "\n",
    "taxa_to_meta_to_match = {\n",
    "    'Clostridium': \n",
    "    {\n",
    "        'Glycocholic': 3,\n",
    "        'Glycocholate': 5,\n",
    "        'Dihydroxylated': -30,\n",
    "    },\n",
    "    'Pseudomonas':\n",
    "    {\n",
    "        'Glycocholate': 33, # or -33?\n",
    "        'Monohydroxylated': 5,\n",
    "    },\n",
    "    'Paraprevotella': \n",
    "    {\n",
    "        'Deoxycholate': 29,\n",
    "        'Deoxycholic': 2,\n",
    "    },\n",
    "    'Bacteroides':\n",
    "    {\n",
    "        'Tetrahydroxylated': -6,\n",
    "    },\n",
    "    'Collinsella':\n",
    "    {\n",
    "        'Dihydroxylated': -8\n",
    "    },\n",
    "    'Parabaceteroides':\n",
    "    {\n",
    "        'Dihydroxylated': 32\n",
    "    },\n",
    "    'Akkermansia':\n",
    "    {\n",
    "        'Glycholic': -4,\n",
    "        'Deoxycholate': 29,\n",
    "    },\n",
    "    'Bifidobacterium':\n",
    "    {\n",
    "        'Deoxycholic': -81,\n",
    "        'Glycocholic': -2,\n",
    "        'Dihydroxylated': -93,\n",
    "    },\n",
    "    'Sutterella':\n",
    "    {\n",
    "        'Deoxycholic': 4\n",
    "    },\n",
    "    'Veilonella':\n",
    "    {\n",
    "        'Dihydroxylated': 9, # or -9?\n",
    "        'Deoxycholate': -9,\n",
    "        'Deoxycholic': -9,\n",
    "    },\n",
    "    'Faecalibacterium':\n",
    "    {\n",
    "        'Deoxycholate': -2,\n",
    "        'Deoxycholic': -2\n",
    "        \n",
    "    },\n",
    "    'Prevotella':\n",
    "    {\n",
    "        'Deoxycholic': 65,\n",
    "        'Glycocholate': 4\n",
    "    },\n",
    "    'Christensenella':\n",
    "    {\n",
    "        'Deoxycholic': 3,\n",
    "        'Dihydroxylated': 4\n",
    "    },\n",
    "    'Oribacterium':\n",
    "    {\n",
    "        'Dihydroxylated': -4,\n",
    "    },\n",
    "    'Peptostreptococcus':\n",
    "    {\n",
    "        'Dihydroxylated': 5,\n",
    "        'Deoxycholate': 5,\n",
    "        'Deoxycholic': 5\n",
    "    },\n",
    "    'Neisseria':\n",
    "    {\n",
    "        'Dihydroxylated': 2\n",
    "    },\n",
    "    'Coprobacillus':\n",
    "    {\n",
    "        'Deoxycholic': 9\n",
    "    }\n",
    "}\n",
    "metas = list(set(df_merge['var1']))\n",
    "taxas =list(set(df_merge['var2']))\n",
    "\n",
    "for t in taxa_to_meta_to_match:\n",
    "    print(t)\n",
    "    meta_to_match = taxa_to_meta_to_match[t]\n",
    "    for y in taxas:\n",
    "        if t in y:\n",
    "            print(y)\n",
    "    for m in meta_to_match:\n",
    "        for x in metas:\n",
    "            if m in x:\n",
    "                print(x)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "082111cb-07ca-416c-94b2-adda8b22a9e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for row in corr\n",
    "# for row in df_merge:\n",
    "df_merge_new = df_merge.copy()\n",
    "df_merge_new['MM'] = 0\n",
    "int_rows = []\n",
    "\n",
    "for i in range(len(df_merge)):\n",
    "    row = df_merge.iloc[i,:]\n",
    "    \n",
    "    for t in taxa_to_meta_to_match:\n",
    "        meta_to_match = taxa_to_meta_to_match[t]\n",
    "        if t in row['var2']:    \n",
    "            for m in meta_to_match:\n",
    "                if m in row['var1']:\n",
    "                    matches = meta_to_match[m]\n",
    "                    row['MM'] = matches\n",
    "                    int_rows.append(row)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00cf15b4-ece2-49d2-9b16-4fff87be5a02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# int rows\n",
    "int_ba"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74801029-460c-4a60-9c85-3f7cd790b79d",
   "metadata": {},
   "outputs": [],
   "source": [
    "row = int_rows[0]\n",
    "print(row['var1'], row['var2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6f3aa88-2e19-42b3-bd79-3e0b66119a1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlations of taxa of interest with metabolites\n",
    "gtr_to_otucorr = {}\n",
    "for g in ['saliva','stool']:\n",
    "    print(g)\n",
    "    gtr_to_otucorr[g] = {}\n",
    "    for time in ['diff','all']:\n",
    "        print(time)\n",
    "        gtr_to_otucorr[g][time] = {}\n",
    "        # grab otu table of differences or of all samples\n",
    "        df_otu = gt_to_otu[g][time]\n",
    "        for comp in ['horizontal', 'vertical']:\n",
    "            if comp == 'vertical':\n",
    "                gr_to_sigotu = to_sigotu[comp]\n",
    "                rvs = ['pre', 'post']\n",
    "            elif comp == 'horizontal':\n",
    "                gr_to_sigotu = to_sigotu[comp]\n",
    "                rvs = ['Response', 'No response']\n",
    "            for rv in rvs: # ['pre', 'post']: # Response No response\n",
    "                print(rv)\n",
    "                sigcorr = []\n",
    "                # grab taxa of interest\n",
    "                taxa_int = gr_to_sigotu[g][rv]\n",
    "    \n",
    "                # merge otu table with metabolites            \n",
    "                df_merge = pd.concat([df_otu.loc[:,taxa_int], gt_to_meta[g][time]], axis=1)\n",
    "                meta = gt_to_meta[g][time].columns\n",
    "                \n",
    "                # look at taxa int and correlate with metabolites\n",
    "                # df_merge \n",
    "                for t in taxa_int:\n",
    "                    for m in meta:\n",
    "                        df_sub = df_merge.loc[:,[t,m]]\n",
    "                        df = type_to_df_map[g].copy()\n",
    "                        if time is 'all':\n",
    "                            df = df.set_index('Together')\n",
    "                            if g == 'stool':\n",
    "                                df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "                        elif time is 'diff':\n",
    "                            df = df[df['Timepoint'] == 'pre']\n",
    "                        df_sub = pd.concat([df_sub, df['WOMAC_P_Response']],axis=1)\n",
    "                        df_sub = df_sub.dropna(subset=[t,m])\n",
    "                        r, p = scipy.stats.spearmanr(df_sub[t].values, df_sub[m].values)\n",
    "                        if p < 0.05:\n",
    "                            sigcorr.append([t,m])\n",
    "                            sns.scatterplot(data=df_sub, x=t, y=m, hue='WOMAC_P_Response')\n",
    "                            sns.despine()\n",
    "                        \n",
    "                            plt.tight_layout()\n",
    "                            plt.savefig(path + 'outputs/' + g_to_job[g + '_adh'] + '/corr_taxa_' + comp + '_' + time + '_' + t.split(';g__')[-1] + '_' + m.replace('/','fslash') + '.pdf')\n",
    "                            plt.close()   \n",
    "                gtr_to_otucorr[g][time][rv] = sigcorr\n",
    "                print(len(sigcorr))\n",
    "           \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f6c73d6-71f4-44ee-ac66-5fe0711fe3f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlations of pathways of interest with metabolites\n",
    "gtr_to_pathcorr = {}\n",
    "for g in ['saliva','stool']:\n",
    "    print(g)\n",
    "    gtr_to_pathcorr[g] = {}\n",
    "    for time in ['diff','all']:\n",
    "        print(time)\n",
    "        gtr_to_pathcorr[g][time] = {}\n",
    "        # grab otu table of differences or of all samples\n",
    "        df_path = gt_to_path[g][time]\n",
    "        for comp in ['horizontal', 'vertical']:\n",
    "            if comp == 'vertical':\n",
    "                gr_to_sigpath = to_sigpath[comp]\n",
    "                rvs = ['pre', 'post']\n",
    "            elif comp == 'horizontal':\n",
    "                gr_to_sigpath = to_sigpath[comp]\n",
    "                rvs = ['Response', 'No response']\n",
    "            for rv in rvs: # ['pre', 'post']: # Response No response\n",
    "                print(rv)\n",
    "                sigcorr = []\n",
    "                # grab taxa of interest\n",
    "                taxa_int = gr_to_sigpath[g][rv]\n",
    "    \n",
    "                # merge otu table with metabolites            \n",
    "                df_merge = pd.concat([df_path.loc[:,taxa_int], gt_to_meta[g][time]], axis=1)\n",
    "                meta = gt_to_meta[g][time].columns\n",
    "                \n",
    "                # look at taxa int and correlate with metabolites\n",
    "                # df_merge \n",
    "                for t in taxa_int:\n",
    "                    for m in meta:\n",
    "                        df_sub = df_merge.loc[:,[t,m]]\n",
    "                        df = type_to_df_map[g].copy()\n",
    "                        if time is 'all':\n",
    "                            df = df.set_index('Together')\n",
    "                            if g == 'stool':\n",
    "                                df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "                        elif time is 'diff':\n",
    "                            df = df[df['Timepoint'] == 'pre']\n",
    "                        df_sub = pd.concat([df_sub, df['WOMAC_P_Response']],axis=1)\n",
    "                        df_sub = df_sub.dropna(subset=[t,m])\n",
    "                        r, p = scipy.stats.spearmanr(df_sub[t].values, df_sub[m].values)\n",
    "                        if p < 0.05:\n",
    "                            sigcorr.append([t,m])\n",
    "                            sns.scatterplot(data=df_sub, x=t, y=m, hue='WOMAC_P_Response')\n",
    "                            sns.despine()\n",
    "                        \n",
    "                            plt.tight_layout()\n",
    "                            plt.savefig(path + 'outputs/' + g_to_job[g + '_adh'] + '/corr_path_' + comp + '_' + time + '_' + t.split(';g__')[-1] + '_' + m.replace('/','fslash') + '.pdf')\n",
    "                            plt.close()   \n",
    "                gtr_to_pathcorr[g][time][rv] = sigcorr\n",
    "                print(len(sigcorr))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5abd2d7-1c78-4e99-a89e-bb864ac5547f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# L Methionine\n",
    "# Kaempferol\n",
    "# Tryptophan\n",
    "# Tyrosine Stool\n",
    "# Usnic acid\n",
    "# Xanthurenic acid\n",
    "# Tryhydroxycholestenoic acid Stool\n",
    "# Isofraxadine\n",
    "int = ['Isofrax', 'Xanth', 'ydroxychol', 'Usnic', 'Tyrosine', 'Tryptophan', 'Kaempferol', 'Methion']\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    for time in ['diff','all']:\n",
    "        for r in ['Response', 'No response']:\n",
    "            test = gtr_to_pathcorr[g][time][r]\n",
    "            for t in test:\n",
    "                p, m = t # unpack pathway, metabolite, from tuple representing sig corr\n",
    "                for i in int:\n",
    "                    if i in m:\n",
    "                        print(t)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f30bf9df-2904-4ca5-b179-4a4eaa5e71f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "int = ['Isofrax', 'Xanth', 'ydroxychol', 'Usnic', 'Tyrosine', 'Tryptophan', 'Kaempferol', 'Methion']\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    for time in ['diff','all']:\n",
    "        for r in ['Response', 'No response']:\n",
    "            test = gtr_to_otucorr[g][time][r]\n",
    "            for t in test:\n",
    "                p, m = t # unpack pathway, metabolite, from tuple representing sig corr\n",
    "                for i in int:\n",
    "                    if i in m:\n",
    "                        print(t)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fdbda6d-641c-47ab-9c83-d392cf0299a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# anaerobic vs aerobic in response\n",
    "# bile acids -> list? saliva and stool of interest \n",
    "#\n",
    "# Bile acids found in Stool: \n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid', 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide']\n",
    "         }\n",
    "\n",
    "mod_ba = {\n",
    "    'stool': [],\n",
    "    'saliva': []\n",
    "}\n",
    "full = list(gt_to_meta[g]['After diet'].columns.values)\n",
    "for g in ['stool','saliva']:\n",
    "    ba = int_ba[g]\n",
    "    for b in ba:\n",
    "        for f in full:\n",
    "            if b in f:\n",
    "                mod_ba[g].append(f)\n",
    "\n",
    "g_to_cutiejob = {\n",
    "    'saliva': 'jobs15',\n",
    "    'stool': 'jobs16'\n",
    "}\n",
    "\n",
    "\n",
    "# from difflib import SequenceMatcher\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    print(g)\n",
    "    n_corr = 0\n",
    "    # import CUITIE results\n",
    "    df_qt = pd.read_csv(path + 'outputs/' + g_to_cutiejob[g] + '/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "    df_qt = df_qt[df_qt['class'] == 'TP']\n",
    "    # iterate through rows\n",
    "    for r in range(len(df_qt)):\n",
    "        meta = df_qt.iloc[r,0]\n",
    "        abun = df_qt.iloc[r,1]\n",
    "        \n",
    "        #for i in int_ba:\n",
    "        #    s = SequenceMatcher(None, meta, i)\n",
    "        #    if s.ratio() > 0.9:\n",
    "        #        print(i)\n",
    "        \n",
    "        #if 'bile' in meta:\n",
    "        #    print(meta)\n",
    "        output = [r]\n",
    "        \n",
    "        for i in int_ba[g]:\n",
    "            if i in meta:\n",
    "                output.append(meta)\n",
    "\n",
    "        for c in ['horizontal','vertical']:\n",
    "            if c == 'vertical':\n",
    "                queries = ['pre', 'post']\n",
    "            if c == 'horizontal':\n",
    "                queries = ['Response','No response']\n",
    "            output.append(c)\n",
    "            for r in queries:\n",
    "                int_taxa = to_sigotu[c][g][r]\n",
    "                for i in int_taxa:\n",
    "                    if i in abun:\n",
    "                        output.append('taxa')\n",
    "                        output.append(i)\n",
    "                int_path = to_sigpath[c][g][r]\n",
    "                for i in int_path:\n",
    "                    if i in abun:\n",
    "                        output.append('path')\n",
    "                        output.append(i)\n",
    "        #output.append()\n",
    "        #output.append()\n",
    "        if len(output) > 5:\n",
    "            # print(output)\n",
    "            n_corr += 1\n",
    "\n",
    "    print(n_corr)\n",
    "\n",
    "# pathway prediction of response vs non response @ baseline\n",
    "# specific taxa at baseline\n",
    "\n",
    "\n",
    "# pathways of interest\n",
    "# ex inflammation in non responders\n",
    "# heatmap of differential pathways in responders/nonresponders \n",
    "# and correlation of those pathways with bacteria\n",
    "# send list of upregulated pathways \n",
    "\n",
    "int_taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76722aff-f034-4fff-a4da-c63dd1fd75ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotations for pathways\n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid', 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide']\n",
    "         }\n",
    "\n",
    "for g in ['saliva','stool']:\n",
    "    # int_ba[g] for metabolite (row) annotations\n",
    "    #df_map = type_to_df_map[g].set_index('Together')\n",
    "    #df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    #df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    #df_meta = pd.concat([gt_to_meta[g]['After diet'],df_map], axis=1)\n",
    "    df_meta = gt_to_meta[g]['After diet']\n",
    "    df_meta_R = df_meta.T\n",
    "    df_meta_R['var_type'] = df_meta_R.index.map(lambda x: 'BA' if x in mod_ba[g] else 'NonBA')\n",
    "    df_meta_R = df_meta_R['var_type'].reset_index()\n",
    "    # df_meta_R['index'] = df_meta_R['index'].str.lower()\n",
    "    df_meta_R = df_meta_R.sort_values(by='index',ascending=True)\n",
    "    df_meta_R.to_csv(path + 'inputs/row_' + g + '.tsv', sep='\\t')\n",
    "    \n",
    "    # dcgr_to_sig for column (taxa) annotations\n",
    "    df_abun = gt_to_otu[g]['post']\n",
    "    df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_abun_R = df_abun.T\n",
    "    df_abun_R['var_type'] = df_abun_R.index.map(lambda x: 'Sig' if x in dcgr_to_sig['taxa']['rvnr'][g]['Response'] else 'Nonsig')\n",
    "    df_abun_R = df_abun_R['var_type'].reset_index()\n",
    "    #df_abun_R['index'] = df_abun_R['index'].str.lower()\n",
    "    df_abun_R = df_abun_R.sort_values(by='index',ascending=True)\n",
    "    df_abun_R.to_csv(path + 'inputs/col_' + g + '.tsv', sep='\\t')\n",
    "\n",
    "df_abun_R.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0b6c598-88f0-40a6-abee-f7b517af1a11",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_diff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d030b184-a32d-4d7b-a541-99192c50deaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlation between datatypes \n",
    "# stool saliva, otu and meta, pre and post and all and diff\n",
    "# stool saliva plasma meta, pre and post and all and diff\n",
    "\n",
    "comparisons = [['taxa','stool','saliva'],['meta','stool','saliva'],\n",
    "               ['meta','stool','plasma'],['meta','saliva','plasma']]\n",
    "int_feat = [\n",
    "'d__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia',\n",
    " 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;__',\n",
    "    'H_N_Acetylneuraminic_acid','H_Glycocholic_acid'\n",
    "    \n",
    "]\n",
    "\n",
    "for c in comparisons:\n",
    "    print(c)\n",
    "    d,g1,g2 = c\n",
    "    if d == 'taxa':\n",
    "        base1,base2 = gt_to_otu, gt_to_otu\n",
    "    elif d == 'meta':\n",
    "        base1, base2 = gt_to_meta, gt_to_meta\n",
    "    for t in ['diff']:        \n",
    "        df1 = base1[g1][t].copy()\n",
    "        df2 = base2[g2][t].copy()\n",
    "\n",
    "        # determine shared indices\n",
    "        shared_samples = list(set(df1.index).intersection(set(df2.index)))\n",
    "        \n",
    "        # determine shared variables\n",
    "        shared_features = list(set(df1.columns).intersection(set(df2.columns)))\n",
    "        \n",
    "        df1 = df1.loc[shared_samples,shared_features]\n",
    "        df2 = df2.loc[shared_samples,shared_features]\n",
    "        \n",
    "        fs = []\n",
    "        rs = []\n",
    "        ps = []\n",
    "        \n",
    "        for f in shared_features:\n",
    "            df1f = df1.copy()\n",
    "            df2f = df2.copy()\n",
    "            df1f.columns = df1f.columns.map(lambda x: x + '_' + g1)\n",
    "            df2f.columns = df2f.columns.map(lambda x: x + '_' + g2)\n",
    "            f1 = f + '_' + g1\n",
    "            f2 = f + '_' + g2\n",
    "\n",
    "            df_merged = pd.concat([df1f,df2f,df_diff],axis=1).loc[:,[f1,f2,'WOMAC_pain']]\n",
    "            df_merged = df_merged.dropna()\n",
    "            x = df_merged[f1]\n",
    "            y = df_merged[f2]\n",
    "            \n",
    "            r,p=scipy.stats.pearsonr(x=x,y=y)\n",
    "            r,p=scipy.stats.spearmanr(x,y,nan_policy='omit')\n",
    "            if np.sum(x) == np.sum(y):\n",
    "                r = 1\n",
    "                p = 0\n",
    "            \n",
    "            fs.append(f)\n",
    "            rs.append(r)\n",
    "            ps.append(p)\n",
    "\n",
    "            if p<0.25:\n",
    "                if f in int_feat:\n",
    "                    plt.figure(figsize=(4,3))\n",
    "                    sns.lmplot(data=df_merged, x=f1, y=f2)\n",
    "                    sns.despine()\n",
    "                    plt.tight_layout()\n",
    "                    plt.savefig(path + 'outputs/jobs34/' + f + '_' + g1 + '_' + g2 + '.pdf')\n",
    "                    plt.close()\n",
    "            \n",
    "            \n",
    "        df_res = pd.DataFrame({'features':fs,'r': rs,'p':ps})\n",
    "        df_res = df_res.sort_values('p')\n",
    "        df_res = df_res[df_res['p'] < 0.25]\n",
    "        print(df_res.features.values)\n",
    "\n",
    "df_res.head() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fbf3e56-2d5f-41fa-8ed7-718d92ed6400",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlation between stool and saliva otu\n",
    "\n",
    "# stool saliva, otu and meta, pre and post and all and diff\n",
    "# stool saliva plasma meta, pre and post and all and diff\n",
    "\n",
    "\n",
    "\n",
    "df_stool = gt_to_otu['stool']['pre'].copy()\n",
    "df_saliva = gt_to_otu['saliva']['pre'].copy()\n",
    "\n",
    "# determine shared indices\n",
    "shared_samples = list(set(df_stool.index).intersection(set(df_saliva.index)))\n",
    "\n",
    "# determine shared variables\n",
    "shared_taxa = list(set(df_stool.columns).intersection(set(df_saliva.columns)))\n",
    "\n",
    "df_stool = df_stool.loc[shared_samples,shared_taxa]\n",
    "df_saliva = df_saliva.loc[shared_samples,shared_taxa]\n",
    "\n",
    "# 16 samples, 90 taxa\n",
    "ts = []\n",
    "rs = []\n",
    "ps = []\n",
    "\n",
    "for t in shared_taxa:\n",
    "    x = df_stool[t]\n",
    "    y = df_saliva[t]\n",
    "    r,p=scipy.stats.pearsonr(x=x,y=y)\n",
    "    if np.sum(x) == np.sum(y):\n",
    "        r = 1\n",
    "        p = 0\n",
    "    \n",
    "    ts.append(t)\n",
    "    rs.append(r)\n",
    "    ps.append(p)\n",
    "    \n",
    "    \n",
    "df_res = pd.DataFrame({'taxa':ts,'r': rs,'p':ps})\n",
    "df_res = df_res.sort_values('p')\n",
    "df_res = df_res[df_res['p'] < 0.25]\n",
    "df_res.taxa.values\n",
    "# df_stool\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "11360b5e-dd8b-4857-b205-4dd8c5c685b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PearsonRResult(statistic=nan, pvalue=nan)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scipy.stats.pearsonr([0,0],[0,0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2707d857-ee9c-466f-983c-3d94600478d3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a341d51b-1db7-4c98-b4e8-65a312f999e2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
