{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3f62dc73-df18-4548-bc0d-aea2a125fef0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "import pandas as pd\n",
    "import scipy.stats\n",
    "import statsmodels.stats.multitest\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "\n",
    "# disable warnings, use w caution\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "# project specific libs\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import pathlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a8b42930-bcde-48f3-9753-8779e1aa11fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# project specific path\n",
    "path = '/Users/KevinBu/Desktop/clemente_lab/Projects/oa/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0343d6a9-1c06-4ed7-90a1-ed193002a1dd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Process Mapping Files ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a3013747-2e0e-412a-b719-b2a6a28afbf2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>Separate</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Together</th>\n",
       "      <th>ContactEmail</th>\n",
       "      <th>ContactName</th>\n",
       "      <th>PrimaryInvestigator</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>RawDataNotes</th>\n",
       "      <th>...</th>\n",
       "      <th>broccoli</th>\n",
       "      <th>Garbanzo_beans</th>\n",
       "      <th>pork</th>\n",
       "      <th>beef</th>\n",
       "      <th>burger</th>\n",
       "      <th>Total_omega3</th>\n",
       "      <th>Adherence_omega3</th>\n",
       "      <th>Total_omega6</th>\n",
       "      <th>Adherence_omega6</th>\n",
       "      <th>Total_o3_o6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#SampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>TTCAGTTCGTTA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-001.pre.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>CGGCCAGAAGCA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>post</td>\n",
       "      <td>OAD-001.post.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-001.post.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>72.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>131.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>GACGTTAAGAAT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-003.pre.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>53.2</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>TCGCTACAGATG</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>post</td>\n",
       "      <td>OAD-003.post.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-003.post.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>108.0</td>\n",
       "      <td>High adherence</td>\n",
       "      <td>171.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>ATGGGACCTTCA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-004.pre.stool.guma.plate313</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>OAD-004.pre.stool.guma.plate313</td>\n",
       "      <td>...</td>\n",
       "      <td>2.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.8</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>28.0</td>\n",
       "      <td>Low adherence</td>\n",
       "      <td>35.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 188 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   BarcodeSequence    LinkerPrimerSequence Separate Timepoint  \\\n",
       "#SampleID                                                                       \n",
       "OAD-001.pre.stool     TTCAGTTCGTTA  CCGGACTACHVGGGTWTCTAAT      All       pre   \n",
       "OAD-001.post.stool    CGGCCAGAAGCA  CCGGACTACHVGGGTWTCTAAT      All      post   \n",
       "OAD-003.pre.stool     GACGTTAAGAAT  CCGGACTACHVGGGTWTCTAAT      All       pre   \n",
       "OAD-003.post.stool    TCGCTACAGATG  CCGGACTACHVGGGTWTCTAAT      All      post   \n",
       "OAD-004.pre.stool     ATGGGACCTTCA  CCGGACTACHVGGGTWTCTAAT      All       pre   \n",
       "\n",
       "                                            Together  \\\n",
       "#SampleID                                              \n",
       "OAD-001.pre.stool    OAD-001.pre.stool.guma.plate313   \n",
       "OAD-001.post.stool  OAD-001.post.stool.guma.plate313   \n",
       "OAD-003.pre.stool    OAD-003.pre.stool.guma.plate313   \n",
       "OAD-003.post.stool  OAD-003.post.stool.guma.plate313   \n",
       "OAD-004.pre.stool    OAD-004.pre.stool.guma.plate313   \n",
       "\n",
       "                                    ContactEmail    ContactName  \\\n",
       "#SampleID                                                         \n",
       "OAD-001.pre.stool   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-001.post.stool  rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-003.pre.stool   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-003.post.stool  rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "OAD-004.pre.stool   rebecca.blank@nyulangone.org  Rebecca Blank   \n",
       "\n",
       "                   PrimaryInvestigator Cohort  \\\n",
       "#SampleID                                       \n",
       "OAD-001.pre.stool           Jose Scher  NonVA   \n",
       "OAD-001.post.stool          Jose Scher  NonVA   \n",
       "OAD-003.pre.stool           Jose Scher  NonVA   \n",
       "OAD-003.post.stool          Jose Scher  NonVA   \n",
       "OAD-004.pre.stool           Jose Scher  NonVA   \n",
       "\n",
       "                                        RawDataNotes  ... broccoli  \\\n",
       "#SampleID                                             ...            \n",
       "OAD-001.pre.stool    OAD-001.pre.stool.guma.plate313  ...      0.0   \n",
       "OAD-001.post.stool  OAD-001.post.stool.guma.plate313  ...      4.0   \n",
       "OAD-003.pre.stool    OAD-003.pre.stool.guma.plate313  ...      0.0   \n",
       "OAD-003.post.stool  OAD-003.post.stool.guma.plate313  ...      4.0   \n",
       "OAD-004.pre.stool    OAD-004.pre.stool.guma.plate313  ...      2.8   \n",
       "\n",
       "                   Garbanzo_beans pork  beef burger Total_omega3  \\\n",
       "#SampleID                                                          \n",
       "OAD-001.pre.stool             0.0  0.0   0.0    0.0          NaN   \n",
       "OAD-001.post.stool            0.0  0.0   0.0    0.0         48.0   \n",
       "OAD-003.pre.stool             0.0  0.0  14.0    0.0         16.8   \n",
       "OAD-003.post.stool            0.0  0.0   0.0    0.0         48.0   \n",
       "OAD-004.pre.stool             0.0  0.0   0.0    0.0          2.8   \n",
       "\n",
       "                   Adherence_omega3 Total_omega6 Adherence_omega6 Total_o3_o6  \n",
       "#SampleID                                                                      \n",
       "OAD-001.pre.stool     Low adherence          NaN    Low adherence         0.0  \n",
       "OAD-001.post.stool    Low adherence         72.0    Low adherence       131.0  \n",
       "OAD-003.pre.stool     Low adherence         53.2   High adherence        75.0  \n",
       "OAD-003.post.stool    Low adherence        108.0   High adherence       171.0  \n",
       "OAD-004.pre.stool     Low adherence         28.0    Low adherence        35.8  \n",
       "\n",
       "[5 rows x 188 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# from AC Q2 run of merged saliva stool\n",
    "df_map = pd.read_csv(path + 'inputs/Qiime2_0/qiime_mapping_file.tsv', sep='\\t', index_col=0)\n",
    "q2_row = df_map.loc['#q2:types',:]\n",
    "df_map = df_map.drop('#q2:types')\n",
    "\n",
    "# change index so it matches metadata file\n",
    "df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "\n",
    "# drop MOC and elution buffer\n",
    "df_map = df_map.drop(['MOC.320','elutionbuffer.plate313'])\n",
    "\n",
    "# grab metadata\n",
    "df_meta = pd.read_csv(path + 'inputs/Metadata_OA.csv')\n",
    "\n",
    "# rename 'Run_ID_Saliva' to be correct\n",
    "df_meta['Timepoints'] = df_meta['Timepoints'].apply(lambda x: 'pre' if x == '0' else 'post')\n",
    "df_meta['Patient_ID'] = df_meta['Patient_ID'].apply(lambda x: x[:-3])  \n",
    "df_meta['Study_ID'] = df_meta['Study_ID'].apply(lambda x: x.split('_')[0][-3:]) \n",
    "df_outcomes = df_meta.copy()\n",
    "\n",
    "# create per sample type mapping files\n",
    "type_to_ST = {'saliva':'Saliva','stool':'fecal'}\n",
    "type_to_df_map = {}\n",
    "\n",
    "# split into specimen type\n",
    "for t in type_to_ST:\n",
    "    # subset on specimen type\n",
    "    df_map_type = df_map[df_map['SpecimenType'] == type_to_ST[t]]\n",
    "\n",
    "    # as to not overwrite df meta\n",
    "    df_meta_type = df_meta.copy()\n",
    "\n",
    "    # create new sample ID for specimen type and set as index\n",
    "    df_meta_type['#SampleID'] = df_meta['Patient_ID'] + '-' + df_meta['Study_ID'] + '.' + df_meta['Timepoints'] + '.' + t\n",
    "    df_meta_type = df_meta_type.set_index('#SampleID')\n",
    "\n",
    "    # create full mapping file\n",
    "    df_map_type = pd.concat([df_map_type, df_meta_type],axis=1)\n",
    "\n",
    "    # use only sequenced samples\n",
    "    df_map_type = df_map_type.dropna(how='any',subset='BarcodeSequence')\n",
    "\n",
    "    # drop all na columns\n",
    "    df_map_type = df_map_type.dropna(how='all',axis=1)\n",
    "\n",
    "    # drop VAD OA 015 because misdx with PsA not OA\n",
    "    if t == 'saliva':\n",
    "        df_map_type = df_map_type.drop(['VAOAD-015.pre.saliva','VAOAD-015.post.saliva'])\n",
    "    if t == 'stool':\n",
    "        df_map_type = df_map_type.drop(['VAOAD-015.pre.stool','VAOAD-015.post.stool'])\n",
    "\n",
    "    # populate dict of mapping files\n",
    "    type_to_df_map[t] = df_map_type\n",
    "\n",
    "    # export for q2\n",
    "    df_q2_type = pd.concat([q2_row.to_frame().T, df_map_type])\n",
    "    df_q2_type.index.name = '#SampleID'\n",
    "    df_q2_type.iloc[0,:] = 'categorical'\n",
    "    df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + t + '.tsv', sep='\\t')\n",
    "    df_q2_type = df_q2_type[df_q2_type['Adherece_antiinflam'].isin(['Moderate adherence', 'High adherence','categorical'])]\n",
    "    df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh.tsv', sep='\\t')\n",
    "    df_q2_type = df_q2_type[df_q2_type['Adherece_antiinflam'].isin(['Moderate adherence', 'High adherence','categorical'])]\n",
    "    df_q2_resp = df_q2_type[df_q2_type['WOMAC_P_Response'].isin(['categorical', 'Response'])]\n",
    "    df_q2_resp.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_response.tsv', sep='\\t')\n",
    "    df_q2_nonresp = df_q2_type[df_q2_type['WOMAC_P_Response'].isin(['categorical', 'No response'])]\n",
    "    df_q2_nonresp.to_csv(path + 'inputs/qiime_mapping_file_' + t + '_adh_noresponse.tsv', sep='\\t')\n",
    "\n",
    "type_to_df_map['stool'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "248b05fc-430d-4a56-9d93-bfc3334b840a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Assess Outcomes ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "95e4ef90-0416-4803-a5b2-d730031630c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n",
      "['VAS_Pt', 'VAS_overall', 'WOMAC_pain', 'WOMAC_stiffness', 'WOMAC_activity', 'WOMAC_total', 'Pain_DETECT', 'CES_D', 'Helplesness', 'Magnification', 'Rumination', 'PCS_EN', 'Sleep_distrubance', 'PASE_walk', 'PASE_light', 'BMI']\n",
      "32\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var</th>\n",
       "      <th>effect</th>\n",
       "      <th>pval</th>\n",
       "      <th>stat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VAS_Pt</td>\n",
       "      <td>198.5</td>\n",
       "      <td>0.008316</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VAS_overall</td>\n",
       "      <td>168.5</td>\n",
       "      <td>0.131314</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_pain</td>\n",
       "      <td>170.0</td>\n",
       "      <td>0.017900</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_stiffness</td>\n",
       "      <td>172.5</td>\n",
       "      <td>0.013052</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WOMAC_activity</td>\n",
       "      <td>164.5</td>\n",
       "      <td>0.032592</td>\n",
       "      <td>MWU</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               var  effect      pval stat\n",
       "0           VAS_Pt   198.5  0.008316  MWU\n",
       "0      VAS_overall   168.5  0.131314  MWU\n",
       "0       WOMAC_pain   170.0  0.017900  MWU\n",
       "0  WOMAC_stiffness   172.5  0.013052  MWU\n",
       "0   WOMAC_activity   164.5  0.032592  MWU"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###\n",
    "# Hypothesis 1: There will be a measurable difference in WOMAC pain response scores and \n",
    "# other outcomes from baseline to after the dietary intervention.\n",
    "###\n",
    "\n",
    "# outcome variables\n",
    "outcomes = ['VAS_Pt', 'VAS_overall', 'WOMAC_pain', 'WOMAC_stiffness', 'WOMAC_activity', 'WOMAC_total', 'Pain_DETECT', \n",
    "            'CES_D', 'Helplesness', 'Magnification', 'Rumination', 'PCS_EN', 'Sleep_distrubance', 'PASE_walk', 'PASE_light', \n",
    "            #'PASE_gardening', # Where did this go? gardening_improve is binary\n",
    "            'BMI']\n",
    "\n",
    "\n",
    "# hypothesis 1\n",
    "# create a new df_meta\n",
    "df_meta = pd.read_csv(path + 'inputs/Metadata_OA.csv')\n",
    "\n",
    "# rename 'Run_ID_Saliva' to be correct\n",
    "df_meta['Timepoints'] = df_meta['Timepoints'].apply(lambda x: 'pre' if x == '0' else 'post')\n",
    "df_meta['Patient_ID'] = df_meta['Patient_ID'].apply(lambda x: x[:-3])  \n",
    "df_meta['Study_ID'] = df_meta['Study_ID'].apply(lambda x: x.split('_')[0][-3:]) \n",
    "df_meta['#SampleID'] = df_meta['Patient_ID'] + '-' + df_meta['Study_ID'] + '.' + df_meta['Timepoints'] + '.stool'\n",
    "df_meta = df_meta.set_index('#SampleID')\n",
    "\n",
    "# convert % to floats for calculations down the road\n",
    "bin = []\n",
    "cont = []\n",
    "df_paired_os = []\n",
    "for w in outcomes:\n",
    "    df_w = df_meta[w]\n",
    "    if df_w.nunique() > 2: # do spearman\n",
    "        df_meta[w] = df_meta[w].astype(str).str.replace('%','').astype(float).values\n",
    "        cont.append(w)\n",
    "    else:\n",
    "        bin.append(w)\n",
    "\n",
    "    # compute difference and store it\n",
    "    df_md = df_meta.copy()\n",
    "    df_md['SubjectID'] = df_md['Patient_ID'] + df_md['Study_ID']\n",
    "    \n",
    "    # first drop unpaired samples\n",
    "    s_remove = []\n",
    "    for s in list(df_md['SubjectID'].values):\n",
    "        if len(df_md[df_md['SubjectID'] == s]) != 2:\n",
    "            s_remove.append(s)\n",
    "    df_md = df_md.loc[~df_md['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "    \n",
    "    # set vars\n",
    "    group_var = 'Timepoints'\n",
    "    pair_var = 'SubjectID'\n",
    "    groups = ['pre','post']\n",
    "    \n",
    "    # get paired per indiv pair\n",
    "    pair_to_diff = {}\n",
    "    for p in list(df_md[pair_var].values):\n",
    "        df = df_md[df_md[pair_var] == p]\n",
    "        t0 = float(df[df[group_var] == groups[0]][w].values)\n",
    "        tf = float(df[df[group_var] == groups[1]][w].values)\n",
    "        pair_to_diff[p] = tf - t0\n",
    "    \n",
    "    df_paired_o = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[w + '_diff'])\n",
    "    df_paired_os.append(df_paired_o)\n",
    "\n",
    "df_meta_paired = pd.concat([*df_paired_os], axis=1)    \n",
    "\n",
    "print(bin)\n",
    "print(cont)\n",
    "\n",
    "# split into all and mod high only\n",
    "a = 'modhigh'\n",
    "job = 'jobs02'\n",
    "df_meta = df_meta[df_meta['Adherece_antiinflam'].isin(['Moderate adherence', 'High adherence'])]\n",
    "        \n",
    "print(len(df_meta))\n",
    "\n",
    "df_results = pd.DataFrame(columns=['var','effect','pval','stat'])\n",
    "# do post treatment vals of binary vars differ from pre treatment 'unpaired'\n",
    "for b in bin:\n",
    "    ct_table_ind=pd.crosstab(df_meta[\"Timepoints\"],df_meta[b])\n",
    "    chi2_stat, p, dof, expected = scipy.stats.chi2_contingency(ct_table_ind)\n",
    "    row=pd.DataFrame.from_dict({'var': [b],'effect':[chi2_stat],'pval':[p],'stat':['chi2']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "# fishers exact\n",
    "for b in bin:\n",
    "    ct_table_ind=pd.crosstab(df_meta[\"Timepoints\"],df_meta[b])\n",
    "    fisher, p = scipy.stats.fisher_exact(ct_table_ind)\n",
    "    row=pd.DataFrame.from_dict({'var': [b],'effect':[t],'pval':[p],'stat':['fisher']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "    \n",
    "# do post treatment vals of continuous vars differ from pre treatment unpaired\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    try:\n",
    "        W,p = scipy.stats.mannwhitneyu(x=df_pre[c].values,y=df_post[c].values, nan_policy='omit')\n",
    "    except:\n",
    "        W,p = 0, 1\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['mwu']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v')\n",
    "        sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette='dark:grey', hue=None, orient='v')\n",
    "    \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/mwu_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['MWU']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "    \n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    t,p = scipy.stats.ttest_ind(a=df_pre[c].values,b=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['ttest']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v')\n",
    "        sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette='dark:grey', hue=None, orient='v')\n",
    "    \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/tt_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['ttest']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "# unpaired and then paired\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    W,p = scipy.stats.wilcoxon(x=df_pre[c].values,y=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['WSR']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.swarmplot(data=df_meta, x='Timepoints', y=c, hue='WOMAC_P_Response', palette=['b','r'], orient='v')\n",
    "        for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "            ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "        sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v', hue=None, boxprops=dict(alpha=.3))\n",
    "        \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/wsr_' + c  + '.pdf')\n",
    "        plt.close()            \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['WSR']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "df_pre = df_meta[df_meta['Timepoints'] == 'pre']\n",
    "df_post = df_meta[df_meta['Timepoints'] == 'post']\n",
    "for c in cont:\n",
    "    t,p = scipy.stats.ttest_rel(a=df_pre[c].values,b=df_post[c].values, nan_policy='omit')\n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[W],'pval':[p],'stat':['pairedt']})\n",
    "    if p < 0.05:\n",
    "        ax = sns.swarmplot(data=df_meta, x='Timepoints', y=c, palette=['b','r'], hue='WOMAC_P_Response',orient='v')\n",
    "        for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "            ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "        sns.boxplot(data=df_meta, x='Timepoints', y=c, orient='v', hue=None, boxprops=dict(alpha=0.3))\n",
    "        \n",
    "        # ax.axes.set_title(\"Title\",fontsize=48)\n",
    "        ax.set_ylabel(c,fontsize=16)\n",
    "        ax.set_xlabel('Timepoints',fontsize=16)                \n",
    "        ax.tick_params(labelsize=16)\n",
    "        sns.despine()\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/' + job + '/pairedt_' + c  + '.pdf')\n",
    "        plt.close()          \n",
    "    row=pd.DataFrame.from_dict({'var': [c],'effect':[t],'pval':[p],'stat':['pairedt']})\n",
    "    df_results = pd.concat([df_results, row])\n",
    "\n",
    "df_results.to_csv(path + 'outputs/' + job + '/outcome_testing.tsv', sep='\\t')\n",
    "df_results.head()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0379f4d-6b07-489a-bfae-75b6085b9a15",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Grab Paired Alpha ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "298b4824-3c24-485e-8705-5daf4a85d6d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unweighted_Unifrac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>0.6672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>0.2837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>0.4237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD005</th>\n",
       "      <td>0.3257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD006</th>\n",
       "      <td>0.1022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unweighted_Unifrac\n",
       "OAD001              0.6672\n",
       "OAD003              0.2837\n",
       "OAD004              0.4237\n",
       "OAD005              0.3257\n",
       "OAD006              0.1022"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###\n",
    "# Hypothesis 2: There will be an association between oral and gut microbiome and pain outcomes\n",
    "###\n",
    "# construct alpha, beta and paired alpha dataframes\n",
    "g_to_dfd = {}\n",
    "# g_test = ['stool','saliva_adh', 'stool_adh', 'saliva']\n",
    "g_test = ['saliva_adh', 'stool_adh']\n",
    "\n",
    "for g in g_test:\n",
    "    # maps diversity type to dataframe\n",
    "    g_to_dfd[g] = {}\n",
    "\n",
    "    # get alpha diversities\n",
    "    df_alpha = pd.read_csv(path + 'outputs/Qiime2_' + g + '/metadata.tsv', sep='\\t', index_col=0)\n",
    "    df_alpha = df_alpha.drop('#q2:types')\n",
    "    df_alpha['SubjectID'] = df_alpha['Patient_ID'] + df_alpha['Study_ID']\n",
    "    df_alpha = df_alpha[['SubjectID', 'Timepoints', 'shannon_entropy']]\n",
    "    g_to_dfd[g]['alpha'] = df_alpha\n",
    "\n",
    "    # get paired alpha div, first drop unpaired samples\n",
    "    s_remove = []\n",
    "    for s in list(df_alpha['SubjectID'].values):\n",
    "        if len(df_alpha[df_alpha['SubjectID'] == s]) != 2:\n",
    "            s_remove.append(s)\n",
    "    df_alpha = df_alpha.loc[~df_alpha['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "    \n",
    "    # set vars\n",
    "    alpha_metric = 'shannon_entropy'\n",
    "    group_var = 'Timepoints'\n",
    "    pair_var = 'SubjectID'\n",
    "    groups = ['pre','post']\n",
    "    \n",
    "    # get paired per indiv pair\n",
    "    pair_to_diff = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        df = df_alpha[df_alpha[pair_var] == p]\n",
    "        alpha_0 = float(df[df[group_var] == groups[0]][alpha_metric].values)\n",
    "        alpha_1 = float(df[df[group_var] == groups[1]][alpha_metric].values)\n",
    "        pair_to_diff[p] = alpha_1 - alpha_0\n",
    "    \n",
    "    df_paired_alpha = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[alpha_metric + '_diff'])\n",
    "    g_to_dfd[g]['paired_alpha'] = df_paired_alpha\n",
    "\n",
    "    # get beta div\n",
    "    df_beta = pd.read_csv(path + 'outputs/Qiime2_' + g + '/core_metrics_results/distance-matrix.tsv',\n",
    "                              sep='\\t', index_col=0)\n",
    "        \n",
    "    # grab twin to pair dict\n",
    "    pair_to_ids = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        df = df_alpha[df_alpha[pair_var] == p]\n",
    "        id_0 = str(df[df[group_var] == groups[0]].index.values[0])\n",
    "        id_1 = str(df[df[group_var] == groups[1]].index.values[0])\n",
    "        pair_to_ids[p] = (id_0, id_1)\n",
    "    \n",
    "    # get distances for each twin pair per beta div matrix    \n",
    "    pair_to_dist = {}\n",
    "    for p in list(df_alpha[pair_var].values):\n",
    "        id_0, id_1 = pair_to_ids[p]\n",
    "        pair_to_dist[p] = df_beta.loc[id_0, id_1]\n",
    "    \n",
    "    df_paired_beta = pd.DataFrame.from_dict(pair_to_dist, orient='index', columns=['Unweighted_Unifrac'])\n",
    "    g_to_dfd[g]['paired_beta'] = df_paired_beta\n",
    "\n",
    "g_to_dfd['stool_adh']['paired_beta'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e018facb-88c3-4c10-be22-dbe70ee41351",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Chi2 Association of Outcome with Alpha ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9d4f38ff-41d6-4f31-94ab-69879de56109",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>div</th>\n",
       "      <th>outcome</th>\n",
       "      <th>statistic</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.232031</td>\n",
       "      <td>0.630022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>1.153857</td>\n",
       "      <td>0.788413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.756449</td>\n",
       "      <td>0.456450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>metadata</td>\n",
       "      <td>Adherece_antiinflam</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>chisq</td>\n",
       "      <td>0.232031</td>\n",
       "      <td>0.630022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.756449</td>\n",
       "      <td>0.980536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.182238</td>\n",
       "      <td>0.856866</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        group                  div           outcome statistic  test_stat  \\\n",
       "0    metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.232031   \n",
       "1  saliva_adh                alpha  WOMAC_P_Response       mwu   1.153857   \n",
       "2  saliva_adh                alpha  WOMAC_P_Response        tt   0.756449   \n",
       "3    metadata  Adherece_antiinflam  WOMAC_P_Response     chisq   0.232031   \n",
       "4   stool_adh                alpha  WOMAC_P_Response       mwu   0.756449   \n",
       "5   stool_adh                alpha  WOMAC_P_Response        tt   0.182238   \n",
       "\n",
       "       pval  \n",
       "0  0.630022  \n",
       "1  0.788413  \n",
       "2  0.456450  \n",
       "3  0.630022  \n",
       "4  0.980536  \n",
       "5  0.856866  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Testing for each sample type for (1) all and (2) high adh only\n",
    "# (A) Chisq of quartiles with adherence \n",
    "# (B) MWU/TT unpaired of outcomes against distance\n",
    "\n",
    "def chisq_of_df_cols(df, c1, c2):\n",
    "    groupsizes = df.groupby([c1, c2]).size()\n",
    "    ctsum = groupsizes.unstack(c1)\n",
    "    # fillna(0) is necessary to remove any NAs which will cause exceptions\n",
    "    return(scipy.stats.chi2_contingency(ctsum.fillna(0)))\n",
    "\n",
    "d_to_metric = {\n",
    "    'alpha': 'shannon_entropy',\n",
    "}\n",
    "group_var = 'Adherece_antiinflam'\n",
    "\n",
    "dfd_to_merge = {}\n",
    "g_test = ['saliva_adh', 'stool_adh']\n",
    "\n",
    "for g in g_test:\n",
    "    dfd_to_merge[g] = {}\n",
    "\n",
    "gs = []\n",
    "ds = []\n",
    "os = []\n",
    "stats = []\n",
    "ts = []\n",
    "ps = []\n",
    "\n",
    "arr = [gs,ds,os,stats,ts,ps]\n",
    "\n",
    "def append_results(arr, val):\n",
    "    for a,v in zip(arr,val):\n",
    "        a.append(v)\n",
    "    return arr\n",
    "\n",
    "# for each sample type, grab relevant mapping file\n",
    "# g_test = saliva, saliva_adh, stool, etc.\n",
    "for g in g_test:\n",
    "    for d in d_to_metric:\n",
    "        # grab relevant diversity df\n",
    "        df_div = g_to_dfd[g][d]\n",
    "\n",
    "        if d == 'alpha':\n",
    "            # df_div = df_div[df_div['Timepoints'] == 'pre']\n",
    "            # df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            # df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div['shannon_entropy'] = df_div['shannon_entropy'].astype(float)\n",
    "\n",
    "        # merge with df of metadata var        \n",
    "        # grab relevant sample IDs\n",
    "        # g = saliva_adh\n",
    "        idx = 'Run_ID_' + g.split('_')[0].capitalize()\n",
    "        df_meta_sub = df_meta.dropna(subset=idx)\n",
    "        df_meta_sub = df_meta_sub.set_index(idx)\n",
    "        df_merge = pd.concat([df_meta_sub,df_div],axis=1)\n",
    "\n",
    "        # subset on adh only\n",
    "        df_merge = df_merge[df_merge[group_var].isin(['Moderate adherence', 'High adherence'])]\n",
    "\n",
    "        # test association of div with outcomes\n",
    "        for o in ['WOMAC_P_Response']: #  outcomes:                    \n",
    "            # test association of adherence with pain outcomes\n",
    "            div_metric = d_to_metric[d]\n",
    "            x,p,dof,ef = chisq_of_df_cols(df_merge, group_var, o)\n",
    "            arr = append_results(arr, ['metadata',group_var,o,'chisq',x,p])\n",
    "            \n",
    "            ax = sns.boxplot(data=df_merge, x=o, y=div_metric)\n",
    "            sns.swarmplot(data=df_merge, x=o, y=div_metric, palette='dark:grey')\n",
    "            sns.despine()\n",
    "        \n",
    "            plt.tight_layout()\n",
    "            plt.savefig(path + 'outputs/jobs06/rvnr_nondiff_' + g + '_' + o + '_' + d + '.pdf')\n",
    "            plt.close()          \n",
    "        \n",
    "            u, p = scipy.stats.mannwhitneyu(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'mwu',t,p])\n",
    "\n",
    "            t, p = scipy.stats.ttest_ind(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'tt',t,p])\n",
    "\n",
    "        # save results\n",
    "        dfd_to_merge[g][d] = df_merge\n",
    "\n",
    "        # export to Q2\n",
    "        # df_q2_type = df_merge.set_index(['Together'])\n",
    "        df_q2_type = df_merge.copy()\n",
    "        q2_row = pd.Series(data=['categorical' for i in range(len(df_merge.columns))], \n",
    "                           index=list(df_merge.columns.values), dtype=str, name='#q2:types')\n",
    "        df_q2_type = pd.concat([q2_row.to_frame().T, df_q2_type])\n",
    "        df_q2_type.index.name = '#SampleID'\n",
    "        df_q2_type.index = df_q2_type.index.map(lambda x: x.split('.guma')[0])\n",
    "        # df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + d + '_' + g + 'aggregate_outcomes.tsv', sep='\\t')\n",
    "\n",
    "df_results = pd.DataFrame.from_dict({\n",
    "    'group': gs,\n",
    "    'div': ds,\n",
    "    'outcome': os,\n",
    "    'statistic': stats,\n",
    "    'test_stat': ts,\n",
    "    'pval': ps\n",
    "})\n",
    "# df_results.to_csv(path + 'outputs/df_results_aggregate.tsv', sep='\\t')\n",
    "# df_results.head()\n",
    "# df_results[df_results['pval'] < 0.05]\n",
    "df_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "22f26972-623f-459c-93fd-d37eb65d17eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>div</th>\n",
       "      <th>outcome</th>\n",
       "      <th>statistic</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>1.153857</td>\n",
       "      <td>0.343434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>-0.594256</td>\n",
       "      <td>0.565542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>-0.594256</td>\n",
       "      <td>0.638889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.769556</td>\n",
       "      <td>0.459355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.769556</td>\n",
       "      <td>0.431818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>1.289488</td>\n",
       "      <td>0.226254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>1.289488</td>\n",
       "      <td>0.876263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>stool_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>-1.033580</td>\n",
       "      <td>0.325681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>-1.033580</td>\n",
       "      <td>0.755051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>pre_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.565329</td>\n",
       "      <td>0.584310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.565329</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>post_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.682542</td>\n",
       "      <td>0.510397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.682542</td>\n",
       "      <td>0.876263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_alpha</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>0.155526</td>\n",
       "      <td>0.879502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>mwu</td>\n",
       "      <td>0.155526</td>\n",
       "      <td>0.267677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>saliva_adh</td>\n",
       "      <td>paired_beta</td>\n",
       "      <td>WOMAC_P_Response</td>\n",
       "      <td>tt</td>\n",
       "      <td>1.182922</td>\n",
       "      <td>0.264200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         group           div           outcome statistic  test_stat      pval\n",
       "0    stool_adh     pre_alpha  WOMAC_P_Response       mwu   1.153857  0.343434\n",
       "1    stool_adh     pre_alpha  WOMAC_P_Response        tt  -0.594256  0.565542\n",
       "2    stool_adh    post_alpha  WOMAC_P_Response       mwu  -0.594256  0.638889\n",
       "3    stool_adh    post_alpha  WOMAC_P_Response        tt   0.769556  0.459355\n",
       "4    stool_adh  paired_alpha  WOMAC_P_Response       mwu   0.769556  0.431818\n",
       "5    stool_adh  paired_alpha  WOMAC_P_Response        tt   1.289488  0.226254\n",
       "6    stool_adh   paired_beta  WOMAC_P_Response       mwu   1.289488  0.876263\n",
       "7    stool_adh   paired_beta  WOMAC_P_Response        tt  -1.033580  0.325681\n",
       "8   saliva_adh     pre_alpha  WOMAC_P_Response       mwu  -1.033580  0.755051\n",
       "9   saliva_adh     pre_alpha  WOMAC_P_Response        tt   0.565329  0.584310\n",
       "10  saliva_adh    post_alpha  WOMAC_P_Response       mwu   0.565329  1.000000\n",
       "11  saliva_adh    post_alpha  WOMAC_P_Response        tt   0.682542  0.510397\n",
       "12  saliva_adh  paired_alpha  WOMAC_P_Response       mwu   0.682542  0.876263\n",
       "13  saliva_adh  paired_alpha  WOMAC_P_Response        tt   0.155526  0.879502\n",
       "14  saliva_adh   paired_beta  WOMAC_P_Response       mwu   0.155526  0.267677\n",
       "15  saliva_adh   paired_beta  WOMAC_P_Response        tt   1.182922  0.264200"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# expand previous to all data, paired and unpaired diversities\n",
    "d_to_metric = {\n",
    "    'pre_alpha': 'shannon_entropy',\n",
    "    'post_alpha': 'shannon_entropy',\n",
    "    'paired_alpha': 'shannon_entropy_diff',\n",
    "    'paired_beta': 'Unweighted_Unifrac',\n",
    "}\n",
    "group_var = 'Adherece_antiinflam'\n",
    "\n",
    "dfd_to_merge = {}\n",
    "g_test = ['stool_adh','saliva_adh']\n",
    "for g in g_test:\n",
    "    dfd_to_merge[g] = {}\n",
    "\n",
    "gs = []\n",
    "ds = []\n",
    "os = []\n",
    "stats = []\n",
    "ts = []\n",
    "ps = []\n",
    "\n",
    "arr = [gs,ds,os,stats,ts,ps]\n",
    "\n",
    "def append_results(arr, val):\n",
    "    for a,v in zip(arr,val):\n",
    "        a.append(v)\n",
    "    return arr\n",
    "\n",
    "# for each sample type, grab relevant mapping file\n",
    "# g_test = saliva, saliva_adh, stool, etc.\n",
    "d_to_map = {}\n",
    "\n",
    "for g in g_test:\n",
    "    # drop duplicates so you have sample mapping to adh\n",
    "    df_map_sub = type_to_df_map[g.split('_')[0]]\n",
    "    df_map_sub.index = df_map_sub.index.map(lambda x: x.split('.')[0].replace('-',''))\n",
    "    df_map_sub = df_map_sub.dropna(how='any',subset=group_var,axis=0)           \n",
    "    \n",
    "    # figure out which samples to keep\n",
    "    # i.e. samples that have a pre and post time point\n",
    "    keep = []\n",
    "    for i in list(df_map_sub.index.values):\n",
    "        if len(df_map_sub.loc[i,:]) == 2:\n",
    "            keep.append(i)\n",
    "\n",
    "    # get unique entires in sorted order\n",
    "    # at this point we are only concerned with differences in values, \n",
    "    # as we've dropped samples with only one endpoint val\n",
    "    save = []\n",
    "    [save.append(x) for x in keep if x not in save]\n",
    "    df_map_sub = df_map_sub.loc[save,:]\n",
    "\n",
    "    # this double populates as OAD001 is an index twice, so the diff fills to both the pre and post col\n",
    "    for o in outcomes:\n",
    "        df_map_sub[o + '_diff'] = df_map_sub[df_map_sub['Timepoint'] == 'post'][o] - df_map_sub[df_map_sub['Timepoint'] == 'pre'][o] \n",
    "\n",
    "    # here we keep only the pre, but everything is identical b/w pre and post\n",
    "    df_dropdup = df_map_sub[~df_map_sub.index.duplicated(keep='first')]\n",
    "    d_to_map[g] = df_dropdup\n",
    "    \n",
    "    for d in d_to_metric:\n",
    "        # grab relevant diversity df\n",
    "        if d == 'pre_alpha' or d == 'post_alpha':\n",
    "            df_div = g_to_dfd[g]['alpha']\n",
    "        else:\n",
    "            df_div = g_to_dfd[g][d]\n",
    "        div_metric = d_to_metric[d]\n",
    "\n",
    "        # when associating alpha div vs outcomes, look at if starting adiv predicts outcome\n",
    "        if d == 'pre_alpha':\n",
    "            df_div = df_div[df_div['Timepoints'] == 'pre']\n",
    "            df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div = df_div.astype(float)\n",
    "\n",
    "        if d == 'post_alpha':\n",
    "            df_div = df_div[df_div['Timepoints'] == 'post']\n",
    "            df_div = df_div.set_index('SubjectID').drop('Timepoints',axis=1)\n",
    "            df_div = df_div.astype(float)\n",
    "\n",
    "        # merge with df of metadata var        \n",
    "        df_merge = pd.concat([df_dropdup,df_div],axis=1)\n",
    "\n",
    "        # drop na in barcodes if samples not sequenced both pre and post\n",
    "        df_merge = df_merge.dropna(how='any',subset='BarcodeSequence')\n",
    "        \n",
    "        # test association of div with outcomes\n",
    "        for o in ['WOMAC_P_Response']:                     \n",
    "            # test association of adherence with pain outcomes\n",
    "            # flip x and y for horizontal\n",
    "            ax = sns.boxplot(data=df_merge, x=o, y=div_metric)\n",
    "            sns.swarmplot(data=df_merge, x=o, y=div_metric, palette='dark:grey')\n",
    "            sns.despine()\n",
    "        \n",
    "            plt.tight_layout()\n",
    "            plt.savefig(path + 'outputs/jobs06/WOMAC_barplot_' + g + '_' + o + '_' + d + '.pdf')\n",
    "            plt.close()          \n",
    "\n",
    "            # MWU and ttest\n",
    "            u, p = scipy.stats.mannwhitneyu(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'mwu',t,p])\n",
    "\n",
    "            t, p = scipy.stats.ttest_ind(df_merge[df_merge[o] == 'Response'][div_metric].values, \n",
    "                                            df_merge[df_merge[o] == 'No response'][div_metric].values, \n",
    "                                            nan_policy='omit')\n",
    "\n",
    "            arr = append_results(arr, [g,d,o,'tt',t,p])\n",
    "         \n",
    "        # save results\n",
    "        dfd_to_merge[g][d] = df_merge\n",
    "\n",
    "        # export to Q2\n",
    "        df_q2_type = df_merge.set_index(['Together'])\n",
    "        q2_row = pd.Series(data=['categorical' for i in range(len(df_merge.columns))], \n",
    "                           index=list(df_merge.columns.values), dtype=str, name='#q2:types')\n",
    "        df_q2_type = pd.concat([q2_row.to_frame().T, df_q2_type])\n",
    "        df_q2_type.index.name = '#SampleID'\n",
    "        df_q2_type.index = df_q2_type.index.map(lambda x: x.split('.guma')[0])\n",
    "        # df_q2_type.to_csv(path + 'inputs/qiime_mapping_file_' + d + '_' + g + '_outcomes.tsv', sep='\\t')\n",
    "\n",
    "df_results = pd.DataFrame.from_dict({\n",
    "    'group': gs,\n",
    "    'div': ds,\n",
    "    'outcome': os,\n",
    "    'statistic': stats,\n",
    "    'test_stat': ts,\n",
    "    'pval': ps\n",
    "})\n",
    "# df_results.to_csv(path + 'outputs/df_results_diff.tsv', sep='\\t')\n",
    "\n",
    "# df_results.head()\n",
    "# the div==alpha results test whether pre-alpha div state associates with pain outcome changes (differences) quartiles\n",
    "# the div==paired_alpha and paired_beta test whether the alpha and betas change in a similar way with the pain outcome\n",
    "print(len(df_results[df_results['pval'] < 0.05]))\n",
    "df_results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5db737b8-b788-4d7a-86a6-39e82ac57eb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_adh\n",
      "['Response', 'No response']\n",
      "WilcoxonResult(statistic=25.0, pvalue=0.30126953125)\n",
      "['Response']\n",
      "WilcoxonResult(statistic=4.0, pvalue=0.109375)\n",
      "['No response']\n",
      "WilcoxonResult(statistic=7.0, pvalue=1.0)\n",
      "saliva_adh\n",
      "['Response', 'No response']\n",
      "WilcoxonResult(statistic=24.0, pvalue=0.26611328125)\n",
      "['Response']\n",
      "WilcoxonResult(statistic=9.0, pvalue=0.46875)\n",
      "['No response']\n",
      "WilcoxonResult(statistic=5.0, pvalue=0.625)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>shannon_entropy_diff</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>Separate</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Together</th>\n",
       "      <th>ContactEmail</th>\n",
       "      <th>ContactName</th>\n",
       "      <th>PrimaryInvestigator</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>...</th>\n",
       "      <th>Pain_DETECT_diff</th>\n",
       "      <th>CES_D_diff</th>\n",
       "      <th>Helplesness_diff</th>\n",
       "      <th>Magnification_diff</th>\n",
       "      <th>Rumination_diff</th>\n",
       "      <th>PCS_EN_diff</th>\n",
       "      <th>Sleep_distrubance_diff</th>\n",
       "      <th>PASE_walk_diff</th>\n",
       "      <th>PASE_light_diff</th>\n",
       "      <th>BMI_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>-0.116436</td>\n",
       "      <td>AGGACAAACTAT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>Text</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-001.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>5.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>11.51</td>\n",
       "      <td>14.29</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD003</th>\n",
       "      <td>-0.327749</td>\n",
       "      <td>TCTTCCTAAAGT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-003.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-5.5</td>\n",
       "      <td>-14.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD004</th>\n",
       "      <td>0.866474</td>\n",
       "      <td>CTTGTGCGACAA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>OAD-004.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>NonVA</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.264032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>0.007278</td>\n",
       "      <td>CTTACACTGCTT</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-001.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>5.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.081446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD012</th>\n",
       "      <td>0.535614</td>\n",
       "      <td>GCGTGGTCATTA</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-012.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-0.5</td>\n",
       "      <td>-1.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.72</td>\n",
       "      <td>-1.072664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TTGGACGTCCAC</td>\n",
       "      <td>CCGGACTACHVGGGTWTCTAAT</td>\n",
       "      <td>All</td>\n",
       "      <td>pre</td>\n",
       "      <td>VAOAD-009.pre.saliva</td>\n",
       "      <td>rebecca.blank@nyulangone.org</td>\n",
       "      <td>Rebecca Blank</td>\n",
       "      <td>Jose Scher</td>\n",
       "      <td>VA</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>2.5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.139974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows Ã— 205 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          shannon_entropy_diff BarcodeSequence    LinkerPrimerSequence  \\\n",
       "OAD001               -0.116436    AGGACAAACTAT  CCGGACTACHVGGGTWTCTAAT   \n",
       "OAD003               -0.327749    TCTTCCTAAAGT  CCGGACTACHVGGGTWTCTAAT   \n",
       "OAD004                0.866474    CTTGTGCGACAA  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD001              0.007278    CTTACACTGCTT  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD012              0.535614    GCGTGGTCATTA  CCGGACTACHVGGGTWTCTAAT   \n",
       "VAOAD009                   NaN    TTGGACGTCCAC  CCGGACTACHVGGGTWTCTAAT   \n",
       "\n",
       "         Separate Timepoint              Together  \\\n",
       "OAD001       Text       pre    OAD-001.pre.saliva   \n",
       "OAD003        All       pre    OAD-003.pre.saliva   \n",
       "OAD004        All       pre    OAD-004.pre.saliva   \n",
       "VAOAD001      All       pre  VAOAD-001.pre.saliva   \n",
       "VAOAD012      All       pre  VAOAD-012.pre.saliva   \n",
       "VAOAD009      All       pre  VAOAD-009.pre.saliva   \n",
       "\n",
       "                          ContactEmail    ContactName PrimaryInvestigator  \\\n",
       "OAD001    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "OAD003    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "OAD004    rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD001  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD012  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "VAOAD009  rebecca.blank@nyulangone.org  Rebecca Blank          Jose Scher   \n",
       "\n",
       "         Cohort  ... Pain_DETECT_diff CES_D_diff Helplesness_diff  \\\n",
       "OAD001    NonVA  ...             -2.0       -1.0             -2.5   \n",
       "OAD003    NonVA  ...             -4.5        1.0             -1.0   \n",
       "OAD004    NonVA  ...             -1.0        3.0             -1.0   \n",
       "VAOAD001     VA  ...             -0.5        5.5              0.0   \n",
       "VAOAD012     VA  ...             -1.5        0.0              2.0   \n",
       "VAOAD009     VA  ...             -4.0        2.0              1.0   \n",
       "\n",
       "         Magnification_diff Rumination_diff PCS_EN_diff  \\\n",
       "OAD001                  5.5             1.0         4.0   \n",
       "OAD003                 -1.0            -2.0        -5.5   \n",
       "OAD004                 -1.0            -1.0        -3.0   \n",
       "VAOAD001                0.0             0.0         0.0   \n",
       "VAOAD012               -2.0            -0.5        -0.5   \n",
       "VAOAD009                0.0             0.5         1.5   \n",
       "\n",
       "         Sleep_distrubance_diff PASE_walk_diff PASE_light_diff  BMI_diff  \n",
       "OAD001                      9.0          11.51           14.29       NaN  \n",
       "OAD003                    -14.0           0.00            0.00       NaN  \n",
       "OAD004                      0.0           0.00            0.00  3.264032  \n",
       "VAOAD001                   -2.5           0.00            0.00 -0.081446  \n",
       "VAOAD012                   -1.5           0.00            6.72 -1.072664  \n",
       "VAOAD009                    2.5           0.00            0.00 -0.139974  \n",
       "\n",
       "[6 rows x 205 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test if the differences in diversity are significantly different from 0 for paired alpha\n",
    "for g in g_test:\n",
    "    print(g)\n",
    "    for d in ['paired_alpha']:\n",
    "        df_div = g_to_dfd[g][d] \n",
    "        df_map = d_to_map[g]\n",
    "\n",
    "        df_merge = pd.concat([df_div, df_map], axis=1)\n",
    "        for o in [['Response','No response'],['Response'], ['No response']]:\n",
    "            df = df_merge[df_merge['WOMAC_P_Response'].isin(o)]\n",
    "            print(o)\n",
    "            print(scipy.stats.wilcoxon(df['shannon_entropy_diff'], nan_policy='omit'))\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bef8c05d-208b-4b13-bac7-01bf00e4e80e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool_adh\n",
      "All\n",
      "paired alpha wilcoxon\n",
      "33.0 0.414306640625\n",
      "Response\n",
      "paired alpha wilcoxon\n",
      "4.0 0.109375\n",
      "No response\n",
      "paired alpha wilcoxon\n",
      "7.0 1.0\n",
      "saliva_adh\n",
      "All\n",
      "paired alpha wilcoxon\n",
      "26.0 0.19091796875\n",
      "Response\n",
      "paired alpha wilcoxon\n",
      "9.0 0.46875\n",
      "No response\n",
      "paired alpha wilcoxon\n",
      "5.0 0.625\n"
     ]
    }
   ],
   "source": [
    "# Alpha div paired\n",
    "# paired beta, comparing intra-indiv difference pre_post to inter pre and inter post\n",
    "g_test = ['stool_adh','saliva_adh']\n",
    "subgroups = ['All', 'Response', 'No response']\n",
    "\n",
    "for g in g_test: #subgroups:\n",
    "    print(g)\n",
    "    df_a = pd.read_csv(path + 'outputs/Qiime2_' + g + '/metadata.tsv', sep='\\t', index_col=0)\n",
    "    df_a = df_a.drop('#q2:types')\n",
    "\n",
    "    for sg in subgroups:\n",
    "        print(sg)\n",
    "        if sg == 'All':\n",
    "            df_alpha = df_a.copy()\n",
    "        else:\n",
    "            df_alpha = df_a[df_a['WOMAC_P_Response'] == sg]\n",
    "        \n",
    "        df_alpha['SubjectID'] = df_alpha['Patient_ID'] + df_alpha['Study_ID']\n",
    "        df_alpha = df_alpha[['SubjectID', 'Timepoints', 'shannon_entropy']]\n",
    "        \n",
    "        # drop unpaired samples\n",
    "        s_remove = []\n",
    "        for s in list(df_alpha['SubjectID'].values):\n",
    "            if len(df_alpha[df_alpha['SubjectID'] == s]) != 2:\n",
    "                s_remove.append(s)\n",
    "        df_alpha = df_alpha.loc[~df_alpha['SubjectID'].isin(s_remove),:] # careful not to use ([s_remove])\n",
    "        \n",
    "        # set vars\n",
    "        alpha_metric = 'shannon_entropy'\n",
    "        group_var = 'Timepoints'\n",
    "        pair_var = 'SubjectID'\n",
    "        groups = ['pre','post']\n",
    "        \n",
    "        # get paired per indiv pair\n",
    "        pair_to_diff = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            df = df_alpha[df_alpha[pair_var] == p]\n",
    "            alpha_0 = float(df[df[group_var] == groups[0]][alpha_metric].values)\n",
    "            alpha_1 = float(df[df[group_var] == groups[1]][alpha_metric].values)\n",
    "            pair_to_diff[p] = alpha_0 - alpha_1\n",
    "        \n",
    "        df_paired_alpha = pd.DataFrame.from_dict(pair_to_diff, orient='index', columns=[alpha_metric + '_diff'])\n",
    "        \n",
    "        # one-sided t-test, n.s.; RA-UA values \n",
    "        # t, p = scipy.stats.ttest_1samp(df_paired_alpha[alpha_metric + '_diff'],popmean=0)\n",
    "    \n",
    "        print('paired alpha wilcoxon')\n",
    "        s, p = scipy.stats.wilcoxon(df_paired_alpha[alpha_metric + '_diff'])\n",
    "        print(s, p)\n",
    "        \n",
    "        # separate\n",
    "        df_alpha[alpha_metric] = df_alpha[alpha_metric].map(lambda x: float(x))\n",
    "        ax = sns.boxplot(data=df_alpha, x=group_var, y=alpha_metric)\n",
    "        sns.swarmplot(data=df_alpha, x=group_var, y=alpha_metric, palette='dark:grey')\n",
    "        sns.despine()\n",
    "    \n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/Qiime2_' + g + '/' + sg + '_alpha.pdf')\n",
    "        plt.close()          \n",
    "    \n",
    "        # now do beta\n",
    "        df_beta = pd.read_csv(path + 'outputs/Qiime2_' + g + '/core_metrics_results/distance-matrix.tsv',\n",
    "                              sep='\\t', index_col=0)\n",
    "        \n",
    "        # set vars\n",
    "        alpha_metric = 'shannon_entropy'\n",
    "        group_var = 'Timepoints'\n",
    "        pair_var = 'SubjectID'\n",
    "        groups = ['pre','post']\n",
    "        g0, g1 = groups[0], groups[1]\n",
    "        \n",
    "        # grab twin to pair dict\n",
    "        pair_to_ids = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            df = df_alpha[df_alpha[pair_var] == p]\n",
    "            id_0 = str(df[df[group_var] == g0].index.values[0])\n",
    "            id_1 = str(df[df[group_var] == g1].index.values[0])\n",
    "            pair_to_ids[p] = (id_0, id_1)\n",
    "        \n",
    "        # get distances for each twin pair per beta div matrix    \n",
    "        pair_to_dist = {}\n",
    "        for p in list(df_alpha[pair_var].values):\n",
    "            id_0, id_1 = pair_to_ids[p]\n",
    "            pair_to_dist[p] = df_beta.loc[id_0, id_1]\n",
    "        \n",
    "        df_paired_beta = pd.DataFrame.from_dict(pair_to_dist, orient='index', columns=['Unweighted_Unifrac'])\n",
    "        \n",
    "        # grab inter RA distances\n",
    "        # this is from unweighted_Timepoint_significance.qzv -> download as tsv\n",
    "        df_raw = pd.read_csv(path + 'outputs/Qiime2_' + g + '/raw_data.tsv', \n",
    "                             sep='\\t', index_col=0)\n",
    "        df_0 = df_raw[df_raw['Group1'] == g0]\n",
    "        df_0 = df_0[df_0['Group2'] == g0]\n",
    "        df_1 = df_raw[df_raw['Group1'] == g1]\n",
    "        df_1 = df_1[df_1['Group2'] == g1]\n",
    "        \n",
    "        # compare distances\n",
    "        inter_twin = df_paired_beta['Unweighted_Unifrac'].values\n",
    "        inter_0 = df_0['Distance'].values\n",
    "        inter_1 = df_1['Distance'].values\n",
    "        \n",
    "        u, p = scipy.stats.mannwhitneyu(inter_twin, inter_0)\n",
    "        #print(u, p)\n",
    "        \n",
    "        t, p = scipy.stats.ttest_ind(inter_twin, inter_1)\n",
    "        #print(t, p)\n",
    "        \n",
    "        t, p = scipy.stats.ttest_ind(inter_0, inter_1)\n",
    "        # print(t, p)\n",
    "        \n",
    "        f, p = scipy.stats.f_oneway(inter_0, inter_1, inter_twin)\n",
    "        # print('inter intra anova')\n",
    "        p# rint(f, p)\n",
    "        \n",
    "        category = ['intra_twin_pair']*len(inter_twin) + ['inter_' + g0 + '_only']*len(inter_0) + ['inter_' + g1 + '_only']*len(inter_1)\n",
    "        distances = list(inter_twin) + list(inter_0) + list(inter_1)\n",
    "        df_dist = pd.DataFrame(data=np.array([category,distances]).T, columns=['category','distance'])\n",
    "        df_dist['distance'] = df_dist['distance'].astype(float)\n",
    "        df_dist.to_csv(path + 'outputs/Qiime2_' + g + '/' + sg + '_inter_intra_beta_dist.tsv',sep='\\t')\n",
    "                             \n",
    "        sns.boxplot(data=df_dist, x='category', y='distance')\n",
    "        sns.swarmplot(data=df_dist, x='category', y='distance', color='black')\n",
    "        sns.despine()\n",
    "    \n",
    "        plt.tight_layout()\n",
    "        plt.savefig(path + 'outputs/Qiime2_' + g + '/' + sg + '_beta.pdf')\n",
    "        plt.close()          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f2ebdb4d-0d17-4c0d-b282-cd78e27f7bb7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectID</th>\n",
       "      <th>Timepoints</th>\n",
       "      <th>shannon_entropy</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>WOMAC_P_Response</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>OAD001</td>\n",
       "      <td>pre</td>\n",
       "      <td>5.097506</td>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>OAD001</td>\n",
       "      <td>post</td>\n",
       "      <td>5.838205</td>\n",
       "      <td>post</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>OAD003</td>\n",
       "      <td>pre</td>\n",
       "      <td>5.190428</td>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>OAD003</td>\n",
       "      <td>post</td>\n",
       "      <td>4.679640</td>\n",
       "      <td>post</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>OAD004</td>\n",
       "      <td>pre</td>\n",
       "      <td>5.299560</td>\n",
       "      <td>pre</td>\n",
       "      <td>No response</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   SubjectID Timepoints  shannon_entropy Timepoint  \\\n",
       "OAD-001.pre.stool     OAD001        pre         5.097506       pre   \n",
       "OAD-001.post.stool    OAD001       post         5.838205      post   \n",
       "OAD-003.pre.stool     OAD003        pre         5.190428       pre   \n",
       "OAD-003.post.stool    OAD003       post         4.679640      post   \n",
       "OAD-004.pre.stool     OAD004        pre         5.299560       pre   \n",
       "\n",
       "                   WOMAC_P_Response  \n",
       "OAD-001.pre.stool       No response  \n",
       "OAD-001.post.stool      No response  \n",
       "OAD-003.pre.stool       No response  \n",
       "OAD-003.post.stool      No response  \n",
       "OAD-004.pre.stool       No response  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# alpha plots\n",
    "for g in ['saliva','stool']:\n",
    "    df_alpha = g_to_dfd[g + '_adh']['alpha']\n",
    "    df_map = type_to_df_map[g].copy()\n",
    "    df_map = df_map.set_index('Together')\n",
    "    df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.g')[0])\n",
    "    \n",
    "    df_merge = pd.concat([df_alpha, df_map], axis=1)\n",
    "    df_merge = df_merge.dropna()\n",
    "    \n",
    "    # sns.set_theme()\n",
    "    custom_params = {\"axes.spines.right\": False, \"axes.spines.top\": False}\n",
    "    sns.set_theme(style=\"ticks\", rc=custom_params)\n",
    "    plt.figure(figsize=(6,4))\n",
    "    sns.boxplot(x='Timepoint', y='shannon_entropy', data=df_merge,palette=['b','r'], boxprops=dict(alpha=0.5))#, hue='Timepoint', palette=['b','r'])\n",
    "    ax = sns.swarmplot(x='Timepoint', y='shannon_entropy', data=df_merge,color='black')\n",
    "    for (x0, y0), (x1, y1) in zip(ax.collections[0].get_offsets(), ax.collections[1].get_offsets()):\n",
    "        ax.plot([x0, x1], [y0, y1], color='black', ls=':', zorder=0)\n",
    "    sns.despine()#(left=True, bottom=True)\n",
    "    \n",
    "    # ax.axes.set_title(\"\",fontsize=48)\\n\",\n",
    "    plt.tight_layout()\n",
    "    ax.set_ylabel(\"Shannon Entropy\",fontsize=16)\n",
    "    ax.set_xlabel(\"Diagnosis\",fontsize=16)\n",
    "    ax.tick_params(labelsize=16)\n",
    "    \n",
    "    # plt.tight_layout()\n",
    "    plt.savefig(path + 'outputs/Qiime2_' + g + '_adh/alpha_all.pdf')\n",
    "    plt.close()\n",
    "\n",
    "df_merge.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51757cac-d364-49d9-88b5-d27af66b38fd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Process Taxa DF ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "8d8a7f96-5705-47fb-8d64-6ddcc4cd5d72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400</th>\n",
       "      <th>...</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;__;__;__;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum</th>\n",
       "      <th>d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__</th>\n",
       "      <th>d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704</th>\n",
       "      <th>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__</th>\n",
       "      <th>d__Bacteria;p__Actinobacteriota;__;__;__;__</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.stool</th>\n",
       "      <td>0.014084</td>\n",
       "      <td>0.039046</td>\n",
       "      <td>0.022027</td>\n",
       "      <td>0.001515</td>\n",
       "      <td>0.173283</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021230</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.152452</td>\n",
       "      <td>0.033001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.022576</td>\n",
       "      <td>0.007098</td>\n",
       "      <td>0.005128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.292202</td>\n",
       "      <td>0.001994</td>\n",
       "      <td>0.013555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000783</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.stool</th>\n",
       "      <td>0.273829</td>\n",
       "      <td>0.067421</td>\n",
       "      <td>0.184572</td>\n",
       "      <td>0.026929</td>\n",
       "      <td>0.053673</td>\n",
       "      <td>0.037783</td>\n",
       "      <td>0.009619</td>\n",
       "      <td>0.023935</td>\n",
       "      <td>0.005789</td>\n",
       "      <td>0.012060</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.181156</td>\n",
       "      <td>0.067844</td>\n",
       "      <td>0.150895</td>\n",
       "      <td>0.063229</td>\n",
       "      <td>0.051360</td>\n",
       "      <td>0.055600</td>\n",
       "      <td>0.012363</td>\n",
       "      <td>0.051513</td>\n",
       "      <td>0.006045</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.post.stool</th>\n",
       "      <td>0.174773</td>\n",
       "      <td>0.040785</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001855</td>\n",
       "      <td>0.124402</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001400</td>\n",
       "      <td>0.000371</td>\n",
       "      <td>0.014265</td>\n",
       "      <td>0.090594</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 374 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola_A_858004  \\\n",
       "index                                                                                                                       \n",
       "OAD-001.post.stool                                           0.014084                                                       \n",
       "OAD-001.pre.stool                                            0.022576                                                       \n",
       "OAD-003.post.stool                                           0.273829                                                       \n",
       "OAD-003.pre.stool                                            0.181156                                                       \n",
       "OAD-004.post.stool                                           0.174773                                                       \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Faecalibacterium  \\\n",
       "index                                                                                                                            \n",
       "OAD-001.post.stool                                           0.039046                                                            \n",
       "OAD-001.pre.stool                                            0.007098                                                            \n",
       "OAD-003.post.stool                                           0.067421                                                            \n",
       "OAD-003.pre.stool                                            0.067844                                                            \n",
       "OAD-004.post.stool                                           0.040785                                                            \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Prevotella  \\\n",
       "index                                                                                                             \n",
       "OAD-001.post.stool                                           0.022027                                             \n",
       "OAD-001.pre.stool                                            0.005128                                             \n",
       "OAD-003.post.stool                                           0.184572                                             \n",
       "OAD-003.pre.stool                                            0.150895                                             \n",
       "OAD-004.post.stool                                           0.000000                                             \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter_164117  \\\n",
       "index                                                                                                                              \n",
       "OAD-001.post.stool                                           0.001515                                                              \n",
       "OAD-001.pre.stool                                            0.000000                                                              \n",
       "OAD-003.post.stool                                           0.026929                                                              \n",
       "OAD-003.pre.stool                                            0.063229                                                              \n",
       "OAD-004.post.stool                                           0.001855                                                              \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides_H  \\\n",
       "index                                                                                                                \n",
       "OAD-001.post.stool                                           0.173283                                                \n",
       "OAD-001.pre.stool                                            0.292202                                                \n",
       "OAD-003.post.stool                                           0.053673                                                \n",
       "OAD-003.pre.stool                                            0.051360                                                \n",
       "OAD-004.post.stool                                           0.124402                                                \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;g__Paramuribaculum  \\\n",
       "index                                                                                                                  \n",
       "OAD-001.post.stool                                           0.000000                                                  \n",
       "OAD-001.pre.stool                                            0.001994                                                  \n",
       "OAD-003.post.stool                                           0.037783                                                  \n",
       "OAD-003.pre.stool                                            0.055600                                                  \n",
       "OAD-004.post.stool                                           0.000000                                                  \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger_A_73129  \\\n",
       "index                                                                                                                            \n",
       "OAD-001.post.stool                                           0.021230                                                            \n",
       "OAD-001.pre.stool                                            0.013555                                                            \n",
       "OAD-003.post.stool                                           0.009619                                                            \n",
       "OAD-003.pre.stool                                            0.012363                                                            \n",
       "OAD-004.post.stool                                           0.001400                                                            \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus  \\\n",
       "index                                                                                                                \n",
       "OAD-001.post.stool                                           0.000160                                                \n",
       "OAD-001.pre.stool                                            0.000000                                                \n",
       "OAD-003.post.stool                                           0.023935                                                \n",
       "OAD-003.pre.stool                                            0.051513                                                \n",
       "OAD-004.post.stool                                           0.000371                                                \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;g__Acetatifactor  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.post.stool                                           0.152452                                                        \n",
       "OAD-001.pre.stool                                            0.000783                                                        \n",
       "OAD-003.post.stool                                           0.005789                                                        \n",
       "OAD-003.pre.stool                                            0.006045                                                        \n",
       "OAD-004.post.stool                                           0.014265                                                        \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes_A_871400  \\\n",
       "index                                                                                                                    \n",
       "OAD-001.post.stool                                           0.033001                                                    \n",
       "OAD-001.pre.stool                                            0.026849                                                    \n",
       "OAD-003.post.stool                                           0.012060                                                    \n",
       "OAD-003.pre.stool                                            0.025918                                                    \n",
       "OAD-004.post.stool                                           0.090594                                                    \n",
       "\n",
       "                    ...  d__Bacteria;p__Firmicutes_A;__;__;__;__  \\\n",
       "index               ...                                            \n",
       "OAD-001.post.stool  ...                                      0.0   \n",
       "OAD-001.pre.stool   ...                                      0.0   \n",
       "OAD-003.post.stool  ...                                      0.0   \n",
       "OAD-003.pre.stool   ...                                      0.0   \n",
       "OAD-004.post.stool  ...                                      0.0   \n",
       "\n",
       "                    d__Bacteria;p__Actinobacteriota;c__Actinomycetia;o__Actinomycetales;f__Micrococcaceae;g__Nesterenkonia  \\\n",
       "index                                                                                                                        \n",
       "OAD-001.post.stool                                                0.0                                                        \n",
       "OAD-001.pre.stool                                                 0.0                                                        \n",
       "OAD-003.post.stool                                                0.0                                                        \n",
       "OAD-003.pre.stool                                                 0.0                                                        \n",
       "OAD-004.post.stool                                                0.0                                                        \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Lactobacillales;__;__  \\\n",
       "index                                                                                 \n",
       "OAD-001.post.stool                                                0.0                 \n",
       "OAD-001.pre.stool                                                 0.0                 \n",
       "OAD-003.post.stool                                                0.0                 \n",
       "OAD-003.pre.stool                                                 0.0                 \n",
       "OAD-004.post.stool                                                0.0                 \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Peptostreptococcales;f__Anaerovoracaceae;g__S5-A14a  \\\n",
       "index                                                                                                                         \n",
       "OAD-001.post.stool                                                0.0                                                         \n",
       "OAD-001.pre.stool                                                 0.0                                                         \n",
       "OAD-003.post.stool                                                0.0                                                         \n",
       "OAD-003.pre.stool                                                 0.0                                                         \n",
       "OAD-004.post.stool                                                0.0                                                         \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Clostridiales;f__Clostridiaceae_222000;__  \\\n",
       "index                                                                                                               \n",
       "OAD-001.post.stool                                                0.0                                               \n",
       "OAD-001.pre.stool                                                 0.0                                               \n",
       "OAD-003.post.stool                                                0.0                                               \n",
       "OAD-003.pre.stool                                                 0.0                                               \n",
       "OAD-004.post.stool                                                0.0                                               \n",
       "\n",
       "                    d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Azospirillales_507929;f__Azospirillaceae_507917;g__Azospirillum  \\\n",
       "index                                                                                                                                         \n",
       "OAD-001.post.stool                                                0.0                                                                         \n",
       "OAD-001.pre.stool                                                 0.0                                                                         \n",
       "OAD-003.post.stool                                                0.0                                                                         \n",
       "OAD-003.pre.stool                                                 0.0                                                                         \n",
       "OAD-004.post.stool                                                0.0                                                                         \n",
       "\n",
       "                    d__Bacteria;p__Firmicutes_D;c__Bacilli;o__Erysipelotrichales;f__Coprobacillaceae;__  \\\n",
       "index                                                                                                     \n",
       "OAD-001.post.stool                                                0.0                                     \n",
       "OAD-001.pre.stool                                                 0.0                                     \n",
       "OAD-003.post.stool                                                0.0                                     \n",
       "OAD-003.pre.stool                                                 0.0                                     \n",
       "OAD-004.post.stool                                                0.0                                     \n",
       "\n",
       "                    d__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingomonas_L_486704  \\\n",
       "index                                                                                                                                        \n",
       "OAD-001.post.stool                                                0.0                                                                        \n",
       "OAD-001.pre.stool                                                 0.0                                                                        \n",
       "OAD-003.post.stool                                                0.0                                                                        \n",
       "OAD-003.pre.stool                                                 0.0                                                                        \n",
       "OAD-004.post.stool                                                0.0                                                                        \n",
       "\n",
       "                    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Muribaculaceae;__  \\\n",
       "index                                                                                                  \n",
       "OAD-001.post.stool                                                0.0                                  \n",
       "OAD-001.pre.stool                                                 0.0                                  \n",
       "OAD-003.post.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                 0.0                                  \n",
       "OAD-004.post.stool                                                0.0                                  \n",
       "\n",
       "                    d__Bacteria;p__Actinobacteriota;__;__;__;__  \n",
       "index                                                            \n",
       "OAD-001.post.stool                                          0.0  \n",
       "OAD-001.pre.stool                                           0.0  \n",
       "OAD-003.post.stool                                          0.0  \n",
       "OAD-003.pre.stool                                           0.0  \n",
       "OAD-004.post.stool                                          0.0  \n",
       "\n",
       "[5 rows x 374 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specific taxa profiling\n",
    "GL_to_otu = {}\n",
    "\n",
    "groups = ['saliva_adh', 'stool_adh']\n",
    "levels = ['6','7']\n",
    "for g in groups:\n",
    "    GL_to_otu[g] = {}\n",
    "    for l in levels:\n",
    "        # read in csv\n",
    "        df_otu = pd.read_csv(path + 'outputs/Qiime2_' + g + '/level-' + l + '.csv', index_col=0)\n",
    "\n",
    "        # filter on OTU only \n",
    "        keep = []\n",
    "        for x in df_otu.columns:\n",
    "            if 'd__' in x:\n",
    "                keep.append(x)\n",
    "        df_otu = df_otu.loc[:,keep]\n",
    "\n",
    "        # normalize by dividing by row sum\n",
    "        df_otu = df_otu.div(df_otu.sum(axis=1),axis=0)\n",
    "\n",
    "        # save to dict\n",
    "        GL_to_otu[g]['L' + l] = df_otu\n",
    "\n",
    "                                       \n",
    "GL_to_otu['stool_adh']['L6'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "515d9fc7-6c65-439d-8733-6c92c35fc5b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Microbiome testing\n",
    "gt_to_otu = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gt_to_otu[g] = {}\n",
    "    df_otu = GL_to_otu[g + '_adh']['L6'].copy()\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['post','pre']:\n",
    "        df = df_otu[df_otu.index.to_series().str.contains(t)]\n",
    "        gt_to_otu[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_otu[g]['pre'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.split('.pre')[0])\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('-',''))\n",
    "    post_df = gt_to_otu[g]['post'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.split('.post')[0])\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('-',''))\n",
    "    \n",
    "    gt_to_otu[g]['diff'] = post_df - pre_df\n",
    "    gt_to_otu[g]['all'] = pd.concat([gt_to_otu[g]['pre'],gt_to_otu[g]['post']])\n",
    "\n",
    "gt_to_otu[g][t].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faf9a2fe-9f68-4cee-9e40-4d5b749b9f0e",
   "metadata": {},
   "source": [
    "###### Process Metabolite DF ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6bf1fdd5-b592-4a4e-a5d4-3cefc2d28c9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_Thiamine</th>\n",
       "      <th>H_Melezitose</th>\n",
       "      <th>H_Phosphocholine</th>\n",
       "      <th>H_SN_Glycero_3_Phophocholine</th>\n",
       "      <th>H_Gamma_Valerobetaine</th>\n",
       "      <th>H_N_Acetylneuraminic_acid</th>\n",
       "      <th>H_Pyridoxine</th>\n",
       "      <th>H_N_Acetylneuraminic_Acid</th>\n",
       "      <th>H_N_Acetylmuramic_Acid</th>\n",
       "      <th>...</th>\n",
       "      <th>H_Nonhydroxylated_bile_acid</th>\n",
       "      <th>H_Hydroxydodecanoic_acid</th>\n",
       "      <th>H_Omega_Hydroxydodecanoate</th>\n",
       "      <th>H_Delta_Methyldodecenoic_Acid</th>\n",
       "      <th>H_Hydroxydecanoic_acid</th>\n",
       "      <th>H_Hydroxydecanoate</th>\n",
       "      <th>H_Methylpentanoic_acid</th>\n",
       "      <th>H_Monohydroxylated_bile_acid</th>\n",
       "      <th>H_Palmitoyl_ethanolamide</th>\n",
       "      <th>H_Linoleic_Acid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>0.312342</td>\n",
       "      <td>-0.312477</td>\n",
       "      <td>0.119853</td>\n",
       "      <td>-0.492876</td>\n",
       "      <td>-0.671376</td>\n",
       "      <td>-0.202952</td>\n",
       "      <td>0.203945</td>\n",
       "      <td>-0.455869</td>\n",
       "      <td>0.352628</td>\n",
       "      <td>0.151464</td>\n",
       "      <td>...</td>\n",
       "      <td>0.054336</td>\n",
       "      <td>-0.330037</td>\n",
       "      <td>-0.886045</td>\n",
       "      <td>-0.314427</td>\n",
       "      <td>-0.322677</td>\n",
       "      <td>-0.650357</td>\n",
       "      <td>-0.361695</td>\n",
       "      <td>0.132794</td>\n",
       "      <td>-0.067464</td>\n",
       "      <td>-0.112641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>0.000428</td>\n",
       "      <td>-0.481251</td>\n",
       "      <td>-0.179337</td>\n",
       "      <td>0.983217</td>\n",
       "      <td>0.167836</td>\n",
       "      <td>0.067823</td>\n",
       "      <td>0.364897</td>\n",
       "      <td>-0.371942</td>\n",
       "      <td>0.545435</td>\n",
       "      <td>-0.007664</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.022614</td>\n",
       "      <td>-0.334936</td>\n",
       "      <td>-0.802118</td>\n",
       "      <td>-0.929470</td>\n",
       "      <td>-0.369578</td>\n",
       "      <td>-0.566431</td>\n",
       "      <td>0.075375</td>\n",
       "      <td>-0.519189</td>\n",
       "      <td>-0.180631</td>\n",
       "      <td>-0.418351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>-0.285212</td>\n",
       "      <td>-0.556445</td>\n",
       "      <td>-0.651325</td>\n",
       "      <td>0.357509</td>\n",
       "      <td>-0.401875</td>\n",
       "      <td>-0.295787</td>\n",
       "      <td>0.137736</td>\n",
       "      <td>-0.165902</td>\n",
       "      <td>0.112170</td>\n",
       "      <td>0.096791</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.923477</td>\n",
       "      <td>-0.589028</td>\n",
       "      <td>-1.535141</td>\n",
       "      <td>-0.525277</td>\n",
       "      <td>-0.278312</td>\n",
       "      <td>-0.600484</td>\n",
       "      <td>0.019657</td>\n",
       "      <td>-0.576831</td>\n",
       "      <td>-0.901479</td>\n",
       "      <td>-1.089906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD004</th>\n",
       "      <td>0.140904</td>\n",
       "      <td>1.089055</td>\n",
       "      <td>-0.668077</td>\n",
       "      <td>-1.061336</td>\n",
       "      <td>-0.370602</td>\n",
       "      <td>-0.329379</td>\n",
       "      <td>-0.654931</td>\n",
       "      <td>0.142971</td>\n",
       "      <td>-1.006689</td>\n",
       "      <td>-0.100373</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.354297</td>\n",
       "      <td>0.449897</td>\n",
       "      <td>0.489189</td>\n",
       "      <td>0.413723</td>\n",
       "      <td>0.474250</td>\n",
       "      <td>0.661963</td>\n",
       "      <td>0.245553</td>\n",
       "      <td>0.279290</td>\n",
       "      <td>-0.444319</td>\n",
       "      <td>-0.912164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD011</th>\n",
       "      <td>-0.346855</td>\n",
       "      <td>1.028262</td>\n",
       "      <td>-0.450323</td>\n",
       "      <td>-0.025939</td>\n",
       "      <td>-0.223029</td>\n",
       "      <td>-0.766374</td>\n",
       "      <td>-0.538410</td>\n",
       "      <td>-1.091423</td>\n",
       "      <td>-0.328522</td>\n",
       "      <td>-0.524686</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.385428</td>\n",
       "      <td>0.419748</td>\n",
       "      <td>0.429119</td>\n",
       "      <td>0.333756</td>\n",
       "      <td>0.448461</td>\n",
       "      <td>0.598846</td>\n",
       "      <td>0.390035</td>\n",
       "      <td>0.287682</td>\n",
       "      <td>-0.532141</td>\n",
       "      <td>-0.752280</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 88 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          H_Pyridoxamine  H_Thiamine  H_Melezitose  H_Phosphocholine  \\\n",
       "Study_ID                                                               \n",
       "VAOAD001        0.312342   -0.312477      0.119853         -0.492876   \n",
       "VAOAD009        0.000428   -0.481251     -0.179337          0.983217   \n",
       "OAD001         -0.285212   -0.556445     -0.651325          0.357509   \n",
       "VAOAD004        0.140904    1.089055     -0.668077         -1.061336   \n",
       "VAOAD011       -0.346855    1.028262     -0.450323         -0.025939   \n",
       "\n",
       "          H_SN_Glycero_3_Phophocholine  H_Gamma_Valerobetaine  \\\n",
       "Study_ID                                                        \n",
       "VAOAD001                     -0.671376              -0.202952   \n",
       "VAOAD009                      0.167836               0.067823   \n",
       "OAD001                       -0.401875              -0.295787   \n",
       "VAOAD004                     -0.370602              -0.329379   \n",
       "VAOAD011                     -0.223029              -0.766374   \n",
       "\n",
       "          H_N_Acetylneuraminic_acid  H_Pyridoxine  H_N_Acetylneuraminic_Acid  \\\n",
       "Study_ID                                                                       \n",
       "VAOAD001                   0.203945     -0.455869                   0.352628   \n",
       "VAOAD009                   0.364897     -0.371942                   0.545435   \n",
       "OAD001                     0.137736     -0.165902                   0.112170   \n",
       "VAOAD004                  -0.654931      0.142971                  -1.006689   \n",
       "VAOAD011                  -0.538410     -1.091423                  -0.328522   \n",
       "\n",
       "          H_N_Acetylmuramic_Acid  ...  H_Nonhydroxylated_bile_acid  \\\n",
       "Study_ID                          ...                                \n",
       "VAOAD001                0.151464  ...                     0.054336   \n",
       "VAOAD009               -0.007664  ...                    -0.022614   \n",
       "OAD001                  0.096791  ...                    -0.923477   \n",
       "VAOAD004               -0.100373  ...                    -0.354297   \n",
       "VAOAD011               -0.524686  ...                    -0.385428   \n",
       "\n",
       "          H_Hydroxydodecanoic_acid  H_Omega_Hydroxydodecanoate  \\\n",
       "Study_ID                                                         \n",
       "VAOAD001                 -0.330037                   -0.886045   \n",
       "VAOAD009                 -0.334936                   -0.802118   \n",
       "OAD001                   -0.589028                   -1.535141   \n",
       "VAOAD004                  0.449897                    0.489189   \n",
       "VAOAD011                  0.419748                    0.429119   \n",
       "\n",
       "          H_Delta_Methyldodecenoic_Acid  H_Hydroxydecanoic_acid  \\\n",
       "Study_ID                                                          \n",
       "VAOAD001                      -0.314427               -0.322677   \n",
       "VAOAD009                      -0.929470               -0.369578   \n",
       "OAD001                        -0.525277               -0.278312   \n",
       "VAOAD004                       0.413723                0.474250   \n",
       "VAOAD011                       0.333756                0.448461   \n",
       "\n",
       "          H_Hydroxydecanoate  H_Methylpentanoic_acid  \\\n",
       "Study_ID                                               \n",
       "VAOAD001           -0.650357               -0.361695   \n",
       "VAOAD009           -0.566431                0.075375   \n",
       "OAD001             -0.600484                0.019657   \n",
       "VAOAD004            0.661963                0.245553   \n",
       "VAOAD011            0.598846                0.390035   \n",
       "\n",
       "          H_Monohydroxylated_bile_acid  H_Palmitoyl_ethanolamide  \\\n",
       "Study_ID                                                           \n",
       "VAOAD001                      0.132794                 -0.067464   \n",
       "VAOAD009                     -0.519189                 -0.180631   \n",
       "OAD001                       -0.576831                 -0.901479   \n",
       "VAOAD004                      0.279290                 -0.444319   \n",
       "VAOAD011                      0.287682                 -0.532141   \n",
       "\n",
       "          H_Linoleic_Acid  \n",
       "Study_ID                   \n",
       "VAOAD001        -0.112641  \n",
       "VAOAD009        -0.418351  \n",
       "OAD001          -1.089906  \n",
       "VAOAD004        -0.912164  \n",
       "VAOAD011        -0.752280  \n",
       "\n",
       "[5 rows x 88 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Metabolome testing\n",
    "gdt_to_df = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gdt_to_df[g] = {}\n",
    "    for d in ['paired_beta']: # just arbitrarily \n",
    "        gdt_to_df[g][d] = {}\n",
    "        df_merge = dfd_to_merge[g + '_adh'][d]\n",
    "        df_meta = pd.read_csv(path + 'inputs/' + g + '_normalized.csv')\n",
    "\n",
    "        # split on timepoint\n",
    "        for t in ['Baseline','After diet']:\n",
    "            df = df_meta[df_meta['Time'] == t]\n",
    "            df = df.set_index('Study_ID')\n",
    "            df.index = df.index.map(lambda x: x.split('_')[0])\n",
    "            df = df.drop('Time',axis=1)\n",
    "            gdt_to_df[g][d][t] = df\n",
    "\n",
    "        gdt_to_df[g][d]['diff'] = gdt_to_df[g][d]['After diet'] - gdt_to_df[g][d]['Baseline']\n",
    "\n",
    "gdt_to_df[g][d][t].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "dbad4aa2-7c35-490c-900b-c607c4ff549d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_Pyridoxamine</th>\n",
       "      <th>H_Thiamine</th>\n",
       "      <th>H_Melezitose</th>\n",
       "      <th>H_Phosphocholine</th>\n",
       "      <th>H_SN_Glycero_3_Phophocholine</th>\n",
       "      <th>H_Gamma_Valerobetaine</th>\n",
       "      <th>H_N_Acetylneuraminic_acid</th>\n",
       "      <th>H_Pyridoxine</th>\n",
       "      <th>H_N_Acetylneuraminic_Acid</th>\n",
       "      <th>H_N_Acetylmuramic_Acid</th>\n",
       "      <th>...</th>\n",
       "      <th>H_Nonhydroxylated_bile_acid</th>\n",
       "      <th>H_Hydroxydodecanoic_acid</th>\n",
       "      <th>H_Omega_Hydroxydodecanoate</th>\n",
       "      <th>H_Delta_Methyldodecenoic_Acid</th>\n",
       "      <th>H_Hydroxydecanoic_acid</th>\n",
       "      <th>H_Hydroxydecanoate</th>\n",
       "      <th>H_Methylpentanoic_acid</th>\n",
       "      <th>H_Monohydroxylated_bile_acid</th>\n",
       "      <th>H_Palmitoyl_ethanolamide</th>\n",
       "      <th>H_Linoleic_Acid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Study_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VAOAD001</th>\n",
       "      <td>0.312342</td>\n",
       "      <td>-0.312477</td>\n",
       "      <td>0.119853</td>\n",
       "      <td>-0.492876</td>\n",
       "      <td>-0.671376</td>\n",
       "      <td>-0.202952</td>\n",
       "      <td>0.203945</td>\n",
       "      <td>-0.455869</td>\n",
       "      <td>0.352628</td>\n",
       "      <td>0.151464</td>\n",
       "      <td>...</td>\n",
       "      <td>0.054336</td>\n",
       "      <td>-0.330037</td>\n",
       "      <td>-0.886045</td>\n",
       "      <td>-0.314427</td>\n",
       "      <td>-0.322677</td>\n",
       "      <td>-0.650357</td>\n",
       "      <td>-0.361695</td>\n",
       "      <td>0.132794</td>\n",
       "      <td>-0.067464</td>\n",
       "      <td>-0.112641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD009</th>\n",
       "      <td>0.000428</td>\n",
       "      <td>-0.481251</td>\n",
       "      <td>-0.179337</td>\n",
       "      <td>0.983217</td>\n",
       "      <td>0.167836</td>\n",
       "      <td>0.067823</td>\n",
       "      <td>0.364897</td>\n",
       "      <td>-0.371942</td>\n",
       "      <td>0.545435</td>\n",
       "      <td>-0.007664</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.022614</td>\n",
       "      <td>-0.334936</td>\n",
       "      <td>-0.802118</td>\n",
       "      <td>-0.929470</td>\n",
       "      <td>-0.369578</td>\n",
       "      <td>-0.566431</td>\n",
       "      <td>0.075375</td>\n",
       "      <td>-0.519189</td>\n",
       "      <td>-0.180631</td>\n",
       "      <td>-0.418351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD001</th>\n",
       "      <td>-0.285212</td>\n",
       "      <td>-0.556445</td>\n",
       "      <td>-0.651325</td>\n",
       "      <td>0.357509</td>\n",
       "      <td>-0.401875</td>\n",
       "      <td>-0.295787</td>\n",
       "      <td>0.137736</td>\n",
       "      <td>-0.165902</td>\n",
       "      <td>0.112170</td>\n",
       "      <td>0.096791</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.923477</td>\n",
       "      <td>-0.589028</td>\n",
       "      <td>-1.535141</td>\n",
       "      <td>-0.525277</td>\n",
       "      <td>-0.278312</td>\n",
       "      <td>-0.600484</td>\n",
       "      <td>0.019657</td>\n",
       "      <td>-0.576831</td>\n",
       "      <td>-0.901479</td>\n",
       "      <td>-1.089906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD004</th>\n",
       "      <td>0.140904</td>\n",
       "      <td>1.089055</td>\n",
       "      <td>-0.668077</td>\n",
       "      <td>-1.061336</td>\n",
       "      <td>-0.370602</td>\n",
       "      <td>-0.329379</td>\n",
       "      <td>-0.654931</td>\n",
       "      <td>0.142971</td>\n",
       "      <td>-1.006689</td>\n",
       "      <td>-0.100373</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.354297</td>\n",
       "      <td>0.449897</td>\n",
       "      <td>0.489189</td>\n",
       "      <td>0.413723</td>\n",
       "      <td>0.474250</td>\n",
       "      <td>0.661963</td>\n",
       "      <td>0.245553</td>\n",
       "      <td>0.279290</td>\n",
       "      <td>-0.444319</td>\n",
       "      <td>-0.912164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD011</th>\n",
       "      <td>-0.346855</td>\n",
       "      <td>1.028262</td>\n",
       "      <td>-0.450323</td>\n",
       "      <td>-0.025939</td>\n",
       "      <td>-0.223029</td>\n",
       "      <td>-0.766374</td>\n",
       "      <td>-0.538410</td>\n",
       "      <td>-1.091423</td>\n",
       "      <td>-0.328522</td>\n",
       "      <td>-0.524686</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.385428</td>\n",
       "      <td>0.419748</td>\n",
       "      <td>0.429119</td>\n",
       "      <td>0.333756</td>\n",
       "      <td>0.448461</td>\n",
       "      <td>0.598846</td>\n",
       "      <td>0.390035</td>\n",
       "      <td>0.287682</td>\n",
       "      <td>-0.532141</td>\n",
       "      <td>-0.752280</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 88 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          H_Pyridoxamine  H_Thiamine  H_Melezitose  H_Phosphocholine  \\\n",
       "Study_ID                                                               \n",
       "VAOAD001        0.312342   -0.312477      0.119853         -0.492876   \n",
       "VAOAD009        0.000428   -0.481251     -0.179337          0.983217   \n",
       "OAD001         -0.285212   -0.556445     -0.651325          0.357509   \n",
       "VAOAD004        0.140904    1.089055     -0.668077         -1.061336   \n",
       "VAOAD011       -0.346855    1.028262     -0.450323         -0.025939   \n",
       "\n",
       "          H_SN_Glycero_3_Phophocholine  H_Gamma_Valerobetaine  \\\n",
       "Study_ID                                                        \n",
       "VAOAD001                     -0.671376              -0.202952   \n",
       "VAOAD009                      0.167836               0.067823   \n",
       "OAD001                       -0.401875              -0.295787   \n",
       "VAOAD004                     -0.370602              -0.329379   \n",
       "VAOAD011                     -0.223029              -0.766374   \n",
       "\n",
       "          H_N_Acetylneuraminic_acid  H_Pyridoxine  H_N_Acetylneuraminic_Acid  \\\n",
       "Study_ID                                                                       \n",
       "VAOAD001                   0.203945     -0.455869                   0.352628   \n",
       "VAOAD009                   0.364897     -0.371942                   0.545435   \n",
       "OAD001                     0.137736     -0.165902                   0.112170   \n",
       "VAOAD004                  -0.654931      0.142971                  -1.006689   \n",
       "VAOAD011                  -0.538410     -1.091423                  -0.328522   \n",
       "\n",
       "          H_N_Acetylmuramic_Acid  ...  H_Nonhydroxylated_bile_acid  \\\n",
       "Study_ID                          ...                                \n",
       "VAOAD001                0.151464  ...                     0.054336   \n",
       "VAOAD009               -0.007664  ...                    -0.022614   \n",
       "OAD001                  0.096791  ...                    -0.923477   \n",
       "VAOAD004               -0.100373  ...                    -0.354297   \n",
       "VAOAD011               -0.524686  ...                    -0.385428   \n",
       "\n",
       "          H_Hydroxydodecanoic_acid  H_Omega_Hydroxydodecanoate  \\\n",
       "Study_ID                                                         \n",
       "VAOAD001                 -0.330037                   -0.886045   \n",
       "VAOAD009                 -0.334936                   -0.802118   \n",
       "OAD001                   -0.589028                   -1.535141   \n",
       "VAOAD004                  0.449897                    0.489189   \n",
       "VAOAD011                  0.419748                    0.429119   \n",
       "\n",
       "          H_Delta_Methyldodecenoic_Acid  H_Hydroxydecanoic_acid  \\\n",
       "Study_ID                                                          \n",
       "VAOAD001                      -0.314427               -0.322677   \n",
       "VAOAD009                      -0.929470               -0.369578   \n",
       "OAD001                        -0.525277               -0.278312   \n",
       "VAOAD004                       0.413723                0.474250   \n",
       "VAOAD011                       0.333756                0.448461   \n",
       "\n",
       "          H_Hydroxydecanoate  H_Methylpentanoic_acid  \\\n",
       "Study_ID                                               \n",
       "VAOAD001           -0.650357               -0.361695   \n",
       "VAOAD009           -0.566431                0.075375   \n",
       "OAD001             -0.600484                0.019657   \n",
       "VAOAD004            0.661963                0.245553   \n",
       "VAOAD011            0.598846                0.390035   \n",
       "\n",
       "          H_Monohydroxylated_bile_acid  H_Palmitoyl_ethanolamide  \\\n",
       "Study_ID                                                           \n",
       "VAOAD001                      0.132794                 -0.067464   \n",
       "VAOAD009                     -0.519189                 -0.180631   \n",
       "OAD001                       -0.576831                 -0.901479   \n",
       "VAOAD004                      0.279290                 -0.444319   \n",
       "VAOAD011                      0.287682                 -0.532141   \n",
       "\n",
       "          H_Linoleic_Acid  \n",
       "Study_ID                   \n",
       "VAOAD001        -0.112641  \n",
       "VAOAD009        -0.418351  \n",
       "OAD001          -1.089906  \n",
       "VAOAD004        -0.912164  \n",
       "VAOAD011        -0.752280  \n",
       "\n",
       "[5 rows x 88 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Metabolome testing\n",
    "gt_to_meta = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gt_to_meta[g] = {}\n",
    "    df_meta = pd.read_csv(path + 'inputs/' + g + '_normalized.csv')\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['Baseline','After diet']:\n",
    "        df = df_meta[df_meta['Time'] == t]\n",
    "        df = df.set_index('Study_ID')\n",
    "        df.index = df.index.map(lambda x: x.split('_')[0])\n",
    "        df = df.drop('Time',axis=1)\n",
    "        gt_to_meta[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_meta[g]['Baseline'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('D0','D-0'))\n",
    "    pre_df.index = pre_df.index.map(lambda x: x + '.pre.' + g)\n",
    "    post_df = gt_to_meta[g]['After diet'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('D0','D-0'))\n",
    "    post_df.index = post_df.index.map(lambda x: x + '.post.' + g)\n",
    "    \n",
    "    gt_to_meta[g]['diff'] = gt_to_meta[g]['After diet'] - gt_to_meta[g]['Baseline']\n",
    "    gt_to_meta[g]['all'] = pd.concat([pre_df,post_df])\n",
    "    \n",
    "\n",
    "gt_to_meta[g][t].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "801b5fe1-2f5b-465e-b21a-943c574d1ee7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Pathway DF Processing ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "8b418d85-6211-4131-9007-92d2b5f1a3ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - globo and isoglobo series</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid elongation</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of 12-, 14- and 16-membered macrolides</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of enediyne antibiotics</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.post.saliva</th>\n",
       "      <td>0.155479</td>\n",
       "      <td>0.025863</td>\n",
       "      <td>0.001866</td>\n",
       "      <td>0.003177</td>\n",
       "      <td>0.008013</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.026699</td>\n",
       "      <td>0.000547</td>\n",
       "      <td>0.031793</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>2.433861e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.216931e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.saliva</th>\n",
       "      <td>0.159679</td>\n",
       "      <td>0.027060</td>\n",
       "      <td>0.001949</td>\n",
       "      <td>0.003082</td>\n",
       "      <td>0.007239</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.027470</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.034357</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>1.911111e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.539150e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.post.saliva</th>\n",
       "      <td>0.151338</td>\n",
       "      <td>0.027415</td>\n",
       "      <td>0.002223</td>\n",
       "      <td>0.002702</td>\n",
       "      <td>0.009135</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.021911</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.030358</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.saliva</th>\n",
       "      <td>0.152707</td>\n",
       "      <td>0.026974</td>\n",
       "      <td>0.002172</td>\n",
       "      <td>0.002788</td>\n",
       "      <td>0.008887</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.022846</td>\n",
       "      <td>0.000257</td>\n",
       "      <td>0.031226</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.post.saliva</th>\n",
       "      <td>0.159885</td>\n",
       "      <td>0.027003</td>\n",
       "      <td>0.001989</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>0.007319</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.026641</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.034499</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>6.002574e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.594595e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 168 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "KEGG_Pathway         Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.post.saliva    0.155479                            0.025863   \n",
       "OAD-001.pre.saliva     0.159679                            0.027060   \n",
       "OAD-003.post.saliva    0.151338                            0.027415   \n",
       "OAD-003.pre.saliva     0.152707                            0.026974   \n",
       "OAD-004.post.saliva    0.159885                            0.027003   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.post.saliva                                           0.001866                 \n",
       "OAD-001.pre.saliva                                            0.001949                 \n",
       "OAD-003.post.saliva                                           0.002223                 \n",
       "OAD-003.pre.saliva                                            0.002172                 \n",
       "OAD-004.post.saliva                                           0.001989                 \n",
       "\n",
       "KEGG_Pathway         Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.post.saliva                                           0.003177        \n",
       "OAD-001.pre.saliva                                            0.003082        \n",
       "OAD-003.post.saliva                                           0.002702        \n",
       "OAD-003.pre.saliva                                            0.002788        \n",
       "OAD-004.post.saliva                                           0.002953        \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism  \\\n",
       "OAD-001.post.saliva                     0.008013   \n",
       "OAD-001.pre.saliva                      0.007239   \n",
       "OAD-003.post.saliva                     0.009135   \n",
       "OAD-003.pre.saliva                      0.008887   \n",
       "OAD-004.post.saliva                     0.007319   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.post.saliva                                           0.000273    \n",
       "OAD-001.pre.saliva                                            0.000166    \n",
       "OAD-003.post.saliva                                           0.000180    \n",
       "OAD-003.pre.saliva                                            0.000176    \n",
       "OAD-004.post.saliva                                           0.000132    \n",
       "\n",
       "KEGG_Pathway         Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.post.saliva                          0.026699   \n",
       "OAD-001.pre.saliva                           0.027470   \n",
       "OAD-003.post.saliva                          0.021911   \n",
       "OAD-003.pre.saliva                           0.022846   \n",
       "OAD-004.post.saliva                          0.026641   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.post.saliva                                           0.000547      \n",
       "OAD-001.pre.saliva                                            0.000309      \n",
       "OAD-003.post.saliva                                           0.000261      \n",
       "OAD-003.pre.saliva                                            0.000257      \n",
       "OAD-004.post.saliva                                           0.000200      \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.post.saliva                                         0.031793   \n",
       "OAD-001.pre.saliva                                          0.034357   \n",
       "OAD-003.post.saliva                                         0.030358   \n",
       "OAD-003.pre.saliva                                          0.031226   \n",
       "OAD-004.post.saliva                                         0.034499   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.post.saliva                                           0.000046                    \n",
       "OAD-001.pre.saliva                                            0.000045                    \n",
       "OAD-003.post.saliva                                           0.000042                    \n",
       "OAD-003.pre.saliva                                            0.000041                    \n",
       "OAD-004.post.saliva                                           0.000038                    \n",
       "\n",
       "KEGG_Pathway         ...  \\\n",
       "OAD-001.post.saliva  ...   \n",
       "OAD-001.pre.saliva   ...   \n",
       "OAD-003.post.saliva  ...   \n",
       "OAD-003.pre.saliva   ...   \n",
       "OAD-004.post.saliva  ...   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - globo and isoglobo series  \\\n",
       "OAD-001.post.saliva                                           0.000021                                                          \n",
       "OAD-001.pre.saliva                                            0.000045                                                          \n",
       "OAD-003.post.saliva                                           0.000103                                                          \n",
       "OAD-003.pre.saliva                                            0.000080                                                          \n",
       "OAD-004.post.saliva                                           0.000062                                                          \n",
       "\n",
       "KEGG_Pathway         Metabolism|Lipid metabolism|Fatty acid elongation  \\\n",
       "OAD-001.post.saliva                                                0.0   \n",
       "OAD-001.pre.saliva                                                 0.0   \n",
       "OAD-003.post.saliva                                                0.0   \n",
       "OAD-003.pre.saliva                                                 0.0   \n",
       "OAD-004.post.saliva                                                0.0   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis  \\\n",
       "OAD-001.post.saliva                                           0.000189                                    \n",
       "OAD-001.pre.saliva                                            0.000095                                    \n",
       "OAD-003.post.saliva                                           0.000022                                    \n",
       "OAD-003.pre.saliva                                            0.000018                                    \n",
       "OAD-004.post.saliva                                           0.000039                                    \n",
       "\n",
       "KEGG_Pathway         Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series  \\\n",
       "OAD-001.post.saliva                                           0.000010                                               \n",
       "OAD-001.pre.saliva                                            0.000019                                               \n",
       "OAD-003.post.saliva                                           0.000060                                               \n",
       "OAD-003.pre.saliva                                            0.000043                                               \n",
       "OAD-004.post.saliva                                           0.000033                                               \n",
       "\n",
       "KEGG_Pathway         Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.post.saliva                                       2.433861e-07                                   \n",
       "OAD-001.pre.saliva                                        1.911111e-06                                   \n",
       "OAD-003.post.saliva                                       0.000000e+00                                   \n",
       "OAD-003.pre.saliva                                        0.000000e+00                                   \n",
       "OAD-004.post.saliva                                       6.002574e-07                                   \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.post.saliva                                                0.0                                  \n",
       "OAD-001.pre.saliva                                                 0.0                                  \n",
       "OAD-003.post.saliva                                                0.0                                  \n",
       "OAD-003.pre.saliva                                                 0.0                                  \n",
       "OAD-004.post.saliva                                                0.0                                  \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of 12-, 14- and 16-membered macrolides  \\\n",
       "OAD-001.post.saliva                                                0.0                                                         \n",
       "OAD-001.pre.saliva                                                 0.0                                                         \n",
       "OAD-003.post.saliva                                                0.0                                                         \n",
       "OAD-003.pre.saliva                                                 0.0                                                         \n",
       "OAD-004.post.saliva                                                0.0                                                         \n",
       "\n",
       "KEGG_Pathway         Metabolism|Metabolism of terpenoids and polyketides|Biosynthesis of enediyne antibiotics  \\\n",
       "OAD-001.post.saliva                                                0.0                                          \n",
       "OAD-001.pre.saliva                                                 0.0                                          \n",
       "OAD-003.post.saliva                                                0.0                                          \n",
       "OAD-003.pre.saliva                                                 0.0                                          \n",
       "OAD-004.post.saliva                                                0.0                                          \n",
       "\n",
       "KEGG_Pathway         Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.post.saliva                                       0.000000e+00                           \n",
       "OAD-001.pre.saliva                                        0.000000e+00                           \n",
       "OAD-003.post.saliva                                       0.000000e+00                           \n",
       "OAD-003.pre.saliva                                        0.000000e+00                           \n",
       "OAD-004.post.saliva                                       8.594595e-07                           \n",
       "\n",
       "KEGG_Pathway         Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \n",
       "OAD-001.post.saliva                                       1.216931e-07                                         \n",
       "OAD-001.pre.saliva                                        1.539150e-07                                         \n",
       "OAD-003.post.saliva                                       0.000000e+00                                         \n",
       "OAD-003.pre.saliva                                        0.000000e+00                                         \n",
       "OAD-004.post.saliva                                       0.000000e+00                                         \n",
       "\n",
       "[5 rows x 168 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# specific pathway profiling\n",
    "g_to_path = {}\n",
    "\n",
    "groups = ['saliva_adh', 'stool_adh']\n",
    "g_to_job = {\n",
    "    'saliva_adh': 'jobs07',\n",
    "    'stool_adh': 'jobs08'\n",
    "}\n",
    "for g in groups:\n",
    "    # read in csv\n",
    "    df_path = pd.read_csv(path + 'outputs/' + g_to_job[g] + '/pred_metagenome_unstrat_kegg.tsv', sep ='\\t', index_col=0)\n",
    "\n",
    "    # transpose\n",
    "    df_path = df_path.T\n",
    "    \n",
    "    # normalize by dividing by row sum\n",
    "    df_path = df_path.div(df_path.sum(axis=1),axis=0)\n",
    "\n",
    "    # filter on metabolism only \n",
    "    keep = []\n",
    "    for p in df_path.columns:\n",
    "        if 'Metabolism' in p: # or 'metabolism' in p: 168 vs 170 if you incl both for saliva\n",
    "            keep.append(p)\n",
    "    df_path = df_path[keep]\n",
    "    \n",
    "    # save to dict\n",
    "    g_to_path[g] = df_path\n",
    "                                 \n",
    "g_to_path['saliva_adh'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "34ed8f12-5226-47f7-b748-07609e554b8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Degradation of flavonoids</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis</th>\n",
       "      <th>Metabolism|Lipid metabolism|Steroid hormone biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.159285</td>\n",
       "      <td>0.035029</td>\n",
       "      <td>0.002320</td>\n",
       "      <td>0.003018</td>\n",
       "      <td>0.007937</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>0.027626</td>\n",
       "      <td>0.000229</td>\n",
       "      <td>0.026349</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000469</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>3.060096e-06</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.510913e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.161049</td>\n",
       "      <td>0.033134</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.002538</td>\n",
       "      <td>0.008606</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.029092</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.027987</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>9.470469e-06</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.511851e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.161760</td>\n",
       "      <td>0.034016</td>\n",
       "      <td>0.002254</td>\n",
       "      <td>0.002773</td>\n",
       "      <td>0.008658</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.029185</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>0.027522</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>2.489384e-06</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.100744e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-005.pre.stool</th>\n",
       "      <td>0.158015</td>\n",
       "      <td>0.033087</td>\n",
       "      <td>0.002185</td>\n",
       "      <td>0.002702</td>\n",
       "      <td>0.008593</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.028749</td>\n",
       "      <td>0.000216</td>\n",
       "      <td>0.027538</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000405</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>2.762671e-06</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>1.103744e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.297970e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-006.pre.stool</th>\n",
       "      <td>0.157791</td>\n",
       "      <td>0.032210</td>\n",
       "      <td>0.002112</td>\n",
       "      <td>0.002819</td>\n",
       "      <td>0.008727</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.030094</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.026678</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000329</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>2.184941e-07</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.127712e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 159 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "KEGG_Pathway       Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.pre.stool    0.159285                            0.035029   \n",
       "OAD-003.pre.stool    0.161049                            0.033134   \n",
       "OAD-004.pre.stool    0.161760                            0.034016   \n",
       "OAD-005.pre.stool    0.158015                            0.033087   \n",
       "OAD-006.pre.stool    0.157791                            0.032210   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.pre.stool                                           0.002320                 \n",
       "OAD-003.pre.stool                                           0.002184                 \n",
       "OAD-004.pre.stool                                           0.002254                 \n",
       "OAD-005.pre.stool                                           0.002185                 \n",
       "OAD-006.pre.stool                                           0.002112                 \n",
       "\n",
       "KEGG_Pathway       Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.pre.stool                                           0.003018        \n",
       "OAD-003.pre.stool                                           0.002538        \n",
       "OAD-004.pre.stool                                           0.002773        \n",
       "OAD-005.pre.stool                                           0.002702        \n",
       "OAD-006.pre.stool                                           0.002819        \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism  \\\n",
       "OAD-001.pre.stool                     0.007937   \n",
       "OAD-003.pre.stool                     0.008606   \n",
       "OAD-004.pre.stool                     0.008658   \n",
       "OAD-005.pre.stool                     0.008593   \n",
       "OAD-006.pre.stool                     0.008727   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.pre.stool                                           0.000192    \n",
       "OAD-003.pre.stool                                           0.000180    \n",
       "OAD-004.pre.stool                                           0.000157    \n",
       "OAD-005.pre.stool                                           0.000186    \n",
       "OAD-006.pre.stool                                           0.000199    \n",
       "\n",
       "KEGG_Pathway       Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.pre.stool                          0.027626   \n",
       "OAD-003.pre.stool                          0.029092   \n",
       "OAD-004.pre.stool                          0.029185   \n",
       "OAD-005.pre.stool                          0.028749   \n",
       "OAD-006.pre.stool                          0.030094   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.pre.stool                                           0.000229      \n",
       "OAD-003.pre.stool                                           0.000194      \n",
       "OAD-004.pre.stool                                           0.000173      \n",
       "OAD-005.pre.stool                                           0.000216      \n",
       "OAD-006.pre.stool                                           0.000220      \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.pre.stool                                         0.026349   \n",
       "OAD-003.pre.stool                                         0.027987   \n",
       "OAD-004.pre.stool                                         0.027522   \n",
       "OAD-005.pre.stool                                         0.027538   \n",
       "OAD-006.pre.stool                                         0.026678   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.pre.stool                                           0.000012                    \n",
       "OAD-003.pre.stool                                           0.000019                    \n",
       "OAD-004.pre.stool                                           0.000016                    \n",
       "OAD-005.pre.stool                                           0.000016                    \n",
       "OAD-006.pre.stool                                           0.000013                    \n",
       "\n",
       "KEGG_Pathway       ...  \\\n",
       "OAD-001.pre.stool  ...   \n",
       "OAD-003.pre.stool  ...   \n",
       "OAD-004.pre.stool  ...   \n",
       "OAD-005.pre.stool  ...   \n",
       "OAD-006.pre.stool  ...   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Degradation of flavonoids  \\\n",
       "OAD-001.pre.stool                                           0.000469                                  \n",
       "OAD-003.pre.stool                                           0.000482                                  \n",
       "OAD-004.pre.stool                                           0.000635                                  \n",
       "OAD-005.pre.stool                                           0.000405                                  \n",
       "OAD-006.pre.stool                                           0.000329                                  \n",
       "\n",
       "KEGG_Pathway       Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis  \\\n",
       "OAD-001.pre.stool                                           0.000028                                      \n",
       "OAD-003.pre.stool                                           0.000090                                      \n",
       "OAD-004.pre.stool                                           0.000103                                      \n",
       "OAD-005.pre.stool                                           0.000055                                      \n",
       "OAD-006.pre.stool                                           0.000087                                      \n",
       "\n",
       "KEGG_Pathway       Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis  \\\n",
       "OAD-001.pre.stool                                       3.060096e-06                                    \n",
       "OAD-003.pre.stool                                       9.470469e-06                                    \n",
       "OAD-004.pre.stool                                       2.489384e-06                                    \n",
       "OAD-005.pre.stool                                       2.762671e-06                                    \n",
       "OAD-006.pre.stool                                       2.184941e-07                                    \n",
       "\n",
       "KEGG_Pathway       Metabolism|Lipid metabolism|Steroid hormone biosynthesis  \\\n",
       "OAD-001.pre.stool                                           0.000026          \n",
       "OAD-003.pre.stool                                           0.000118          \n",
       "OAD-004.pre.stool                                           0.000108          \n",
       "OAD-005.pre.stool                                           0.000092          \n",
       "OAD-006.pre.stool                                           0.000088          \n",
       "\n",
       "KEGG_Pathway       Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series  \\\n",
       "OAD-001.pre.stool                                           0.000026                                               \n",
       "OAD-003.pre.stool                                           0.000090                                               \n",
       "OAD-004.pre.stool                                           0.000103                                               \n",
       "OAD-005.pre.stool                                           0.000055                                               \n",
       "OAD-006.pre.stool                                           0.000086                                               \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.pre.stool                                       0.000000e+00                                   \n",
       "OAD-003.pre.stool                                       0.000000e+00                                   \n",
       "OAD-004.pre.stool                                       0.000000e+00                                   \n",
       "OAD-005.pre.stool                                       1.103744e-07                                   \n",
       "OAD-006.pre.stool                                       0.000000e+00                                   \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.pre.stool                                                0.0                                  \n",
       "OAD-003.pre.stool                                                0.0                                  \n",
       "OAD-004.pre.stool                                                0.0                                  \n",
       "OAD-005.pre.stool                                                0.0                                  \n",
       "OAD-006.pre.stool                                                0.0                                  \n",
       "\n",
       "KEGG_Pathway       Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.pre.stool                                       1.510913e-06                           \n",
       "OAD-003.pre.stool                                       8.511851e-07                           \n",
       "OAD-004.pre.stool                                       6.100744e-07                           \n",
       "OAD-005.pre.stool                                       5.297970e-07                           \n",
       "OAD-006.pre.stool                                       1.127712e-07                           \n",
       "\n",
       "KEGG_Pathway       Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis  \\\n",
       "OAD-001.pre.stool                                                0.0                               \n",
       "OAD-003.pre.stool                                                0.0                               \n",
       "OAD-004.pre.stool                                                0.0                               \n",
       "OAD-005.pre.stool                                                0.0                               \n",
       "OAD-006.pre.stool                                                0.0                               \n",
       "\n",
       "KEGG_Pathway       Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \n",
       "OAD-001.pre.stool                                                0.0                                         \n",
       "OAD-003.pre.stool                                                0.0                                         \n",
       "OAD-004.pre.stool                                                0.0                                         \n",
       "OAD-005.pre.stool                                                0.0                                         \n",
       "OAD-006.pre.stool                                                0.0                                         \n",
       "\n",
       "[5 rows x 159 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pathway testing\n",
    "gt_to_path = {}\n",
    "for g in ['saliva','stool']:\n",
    "    gt_to_path[g] = {}\n",
    "    df_path = g_to_path[g + '_adh'].copy()\n",
    "\n",
    "    # split on timepoint\n",
    "    for t in ['post','pre']:\n",
    "        df = df_path[df_path.index.to_series().str.contains(t)]\n",
    "        gt_to_path[g][t] = df\n",
    "\n",
    "    pre_df = gt_to_path[g]['pre'].copy()\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.split('.pre')[0])\n",
    "    pre_df.index = pre_df.index.map(lambda x: x.replace('-',''))\n",
    "    post_df = gt_to_path[g]['post'].copy()\n",
    "    post_df.index = post_df.index.map(lambda x: x.split('.post')[0])\n",
    "    post_df.index = post_df.index.map(lambda x: x.replace('-',''))\n",
    "    \n",
    "    gt_to_path[g]['diff'] = post_df - pre_df\n",
    "    gt_to_path[g]['all'] = pd.concat([gt_to_path[g]['pre'],gt_to_path[g]['post']])\n",
    "\n",
    "gt_to_path[g][t].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81bd68ab-5702-4758-ad7a-34bf139443ea",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Diet DF Processing ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "77c18bcf-830a-458d-88c1-70f509e7c9c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkMAAAHJCAYAAACG+j24AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABlg0lEQVR4nO3deVhUdfsG8HsYmGFfRBYFFRSBUBEEBBQ1NSuz3LI3zX1LLcVMUUtTUMpM3HDP5XUrsdw1KpfKfcEVC7BSURQERBbZZmBmfn/4Y94mGAWFmWHm/lwXl3G+55x55oHg5nzPIlAoFAoQERERGSgjbRdAREREpE0MQ0RERGTQGIaIiIjIoDEMERERkUFjGCIiIiKDxjBEREREBo1hiIiIiAwawxAREREZNIYhIqIXxHvXEtVvDENEVKv27NkDLy8v3Lt3DwAwc+ZMdOvWTSu1DB06FF5eXsoPb29v+Pv7o3///ti2bRtkMpnK+t26dcPMmTNr9BrHjh3DjBkzXrg2Ly8vtG7dGi+//DKioqKQn59faZvbt28jMjISr7zyCnx9ffHyyy9jypQpSElJUfs6u3btgpeXF8aMGVPjGokMhbG2CyAi/fbBBx9g2LBhWnt9Hx8fzJ07FwAgk8mQn5+P48eP44svvsClS5ewdOlSCAQCAMDKlSthaWlZo/1v3ry5VmoDgLKyMvzxxx9YsmQJkpOTsWPHDmVtR44cQUREBFq2bIkJEybA1dUVDx48wLZt2/DOO+9g1apV6Ny5c6XX2L17Nzw9PXH69GmkpaWhSZMmz10vkb5iGCKiOtW0aVOtvr6lpSX8/PxUlnXr1g3u7u5YsGABunXrht69ewN4Ek60XVtQUBCKiooQGxuLa9euwc/PD3fv3sX06dPRqVMnLFu2DEKhULn+a6+9hvfeew8zZ87EL7/8AlNTU+XY7du3cfnyZXz99deIiIjAd999h6lTp2rq7RHVG5wmIyIVf/zxB4YPH46AgAD4+/tjxIgRuHbtmnL8+++/R//+/eHn5wdfX1/06dMH8fHxavf3z2myzz77DCEhISgvL1dZZ9GiRWjfvj2kUikA4OjRo3jvvffg7++P1q1b4/XXX8f27dtr9X0OHToUjo6OiIuLUy779zSZRCLBV199hS5duqB169Z46623VN7r0KFDceHCBVy4cAFeXl44f/58rdTWunVrAEB6ejoAYNu2bZBKpZg9e7ZKEAIAU1NTzJgxAwMGDEBBQYHK2O7du2FlZYXQ0FC8/vrr2L17t7LHRPQ/DENEpFRYWIgxY8bAzs4OsbGxWLp0KUpKSjB69Gg8fvwY33zzDebMmYPu3btj3bp1WLRoEUxMTBAREaH8xf00ffr0QW5uLs6ePatcplAoEB8fj9dffx0ikQi//fYbPvzwQ7Rq1QqrV6/GihUr4OLigvnz5+Py5cu19l6FQiFCQ0ORmJhYKZxV1PXhhx8iLi4OI0eOxJo1a+Dv748pU6Zg3759AIC5c+fCx8cHPj4+2LlzJ1q1alUrtd2+fRsAlFNaJ0+ehI+PD5ycnKpcPzg4GB9//DEcHR2Vy2QyGfbv34833ngDIpEI/fv3R05ODo4ePVorNRLpE06TEZHS33//jUePHmHo0KEICAgAADRv3hxxcXEoLCxEWloaRo0ahQ8//FC5jaurK/r374/Lly+jcePGT91/QEAAXF1dER8fj06dOgEALl26hPT0dPTp00dZQ9++fTFr1izldv7+/ggODkZCQgLatWtXa++3YcOGKCsrQ15eHho2bKgydubMGZw8eRJLly7FG2+8AQDo1KkTSkpKEBMTgzfffBMeHh7Kc4z+Pd1VHQqFQiWI5efn48KFC1izZg38/PyUR4gyMzPx0ksv1WjfJ06cQFZWFt5++21lfR4eHtixY4fy/RDREwxDRKTUsmVLNGjQABMmTEDPnj3RpUsXhIaGYvr06QCgnEJ6/PgxUlNTkZqaqjzKU1ZW9sz9CwQC9O7dG9u2bUNUVBREIhEOHTqEJk2aKMNXxVVPxcXFuHv3Lm7fvo3r169X+zWeR8VJyv909uxZCAQCdOnSRSWwdOvWDQcOHMBff/1V44DybwkJCZWOJhkZGSE0NBTz589X1iUQCCpd+fYsu3fvRrNmzeDu7q6cPuvZsydWrFiBmzdvokWLFi9UO5E+YRgiIiULCwt88803WLNmDeLj4xEXFwczMzP07t0bs2bNQmZmJubMmYNz587B2NgYzZs3h5eXF4Dq32unb9++WL16NU6cOIGXX34ZP/30E9577z3l+KNHjzB37lwcPXoUAoEAzZo1Uwal2r6fT2ZmJkxNTWFra1tpLC8vDwqFQu2RqKysrBcOQ61atUJUVBSAJ4FHLBajUaNGla5oc3Fxeeo0ZHl5OR49eqScJnv06BF+++03lJWVISgoqNL6O3fuxKeffvpCtRPpE4YhIlLRvHlzLFq0CDKZDImJidi/fz927NgBV1dX7NmzByYmJvjuu+/g4+MDY2Nj/P333zhw4EC199+sWTP4+fnhxx9/hImJCXJzc5VXcwHAtGnTcPPmTfz3v/9Fu3btIBKJUFJSgu+//75W36dMJsOFCxfQrl27SiclA4CVlRXMzc2xdetWte/jRVlYWKBNmzbPXC8sLAxbtmxBdnY2HBwcKo2fPHkS48ePx5IlS9CrVy/s378fZWVlWLlyJaytrVXWXbVqFfbt24ePP/5Y5cozIkPGE6iJSOmnn35CSEgIsrOzIRQK4e/vj8jISFhbWyM9PR23b9/GgAED4OvrC2PjJ39LnThxAgAgl8ur/Tq9e/fGiRMncOjQIfj5+cHNzU05dunSJbz22msICQmBSCR67td4lri4OGRlZWHQoEFVjrdv3x7FxcVQKBRo06aN8uOvv/7CqlWrlFNnRkZ1/2N08ODBMDExQXR0dKXpspKSEsTGxsLGxgZdu3YF8OTGl35+fujRoweCg4NVPgYNGoT8/Hz8+OOPdV43UX3BI0NEpNSuXTvI5XJ8+OGHeP/992FhYYEff/wRjx8/xuuvv44TJ07gm2++gbOzM6ytrXHq1Cls2bIFwJNfytXVq1cvLFiwAD/88IPKidIA4Ovri4MHD6JVq1ZwdnbGlStXsG7dOggEghq9RoXCwkJcvXoVwJMwlZubi1OnTmHnzp3o3bs3Xn311Sq369KlC4KCgvDBBx/ggw8+QIsWLZCYmIgVK1YgLCwMDRo0AABYW1vjypUrOHv2LHx8fGBjY1PjGp/F1dUVkZGRmDVrFgYPHoyBAweiUaNGuHv3LjZv3ow7d+5g/fr1MDc3R2JiIv78889Kfa3QvXt32NjYIC4uDv369av1WonqI4YhIlJydHTEhg0bsHz5csyaNQslJSVo2bIlVqxYgZCQEKxevRqff/45Zs6cCZFIBA8PD6xZswZffPEFLl68iKFDh1brdWxtbdGlSxccP3680pVNX375JebPn4/58+cDANzc3BAVFYUDBw7g4sWLNX5PSUlJePfddwE8OYpjb28Pd3d3fPnll3jrrbfUbmdkZISvv/4ay5cvx7p165CTkwMnJyeMGDFC5Wq6wYMH4/fff8fYsWOxYMGCp+7zRfTr1w/NmjXDli1bsGzZMuTk5MDBwQH+/v5Yvnw5PDw8ADw5cVooFKq9YkwkEqFnz56Ii4tDcnLyC5/3RKQPBAo+YZCIiIgMGI8MEVG9IpfLq3XuUMU5TZqky7URkXo8MkRE9crMmTOxd+/eZ6537NgxuLq6aqCi/9Hl2ohIPYYhIqpX7t27h9zc3Geu5+XlpbwaTVN0uTYiUo9hiIiIiAwa7zNEREREBo1hiIiIiAwaw1A13LhxAzdu3NB2GURERFQHGIaqQSqVQiqVarsMnSaRSHDp0iVIJBJtl2JQ2HftYe+1h73XHn3tPcMQ1YqK5yX9+7lJVLfYd+1h77WHvdcefe09wxAREREZNIYhIiIiMmgMQ0RERGTQGIaIiIjIoPFpgbVMJpOhrKxM22VoXMWVBRKJBEZGzNiaUld9NzExgVAorLX9ERHpMoahWqJQKPDgwQPk5eVpuxStkMvlMDY2Rnp6OsOQBtVl321tbeHs7AyBQFCr+yUi0jUMQ7WkIgg5OjrC3Nzc4H6ByGQySCQSiMViHlHQoLrou0KhQHFxMbKysgAAjRo1qpX9EhHpKoahWiCTyZRByN7eXtvlaEXFPSdMTU0ZhjSorvpuZmYGAMjKyoKjoyO/pkSk1zifUQsqzhEyNzfXciVEtafi+9kQz4EjIsPCMFSLDG1qjPQbv5+JyFBwmoyI6DnZWVtCWJqPkoe3IRCaQGjZAMZWdhAYcVqRqD5hGCIieg7CcgmsM67h/q7dgLwcAGBkagnH/lNh2sQbRsYiLVdIRNXFabI6Fh4ejqCgICgUCpXlycnJ8PLyQtu2bSs9/ffPP/+El5cXduzYAeDJ5dP79u3D0KFDERgYCD8/P7z55ptYsWIFcnNzVba9d+8evLy84OXlhZ07d1ZZ0+PHj9GmTRt4eXnh/PnzlcbT0tLg7e2N9u3bo7S0VO17k8vl+P777zF48GCEhoaiU6dOePvtt7F161ZIpdJq9eefZs6cqay94qNVq1YICwtDREQEMjIyarS/f/binx+tW7dGWFgYwsPDcefOnRrXSQQAZRl/ouDkTmUQAgB5aSEexH2O8vyHWqyMiGqKR4bqWIcOHfDzzz/j77//RsuWLZXLT548CVtbW+Tl5eHChQvo1KmTciwhIQEAEBYWhrKyMkyePBnHjx9H//79MWrUKJiamuL69evYunUr9uzZg7Vr18LLy0vldY2MjPDjjz/i3XffrVTTkSNHnhpWdu3aBTc3N9y7dw8//vgj+vXrV2mdkpISjB8/HteuXcOgQYMwatQoKBQKXLlyBTExMTh+/DjWrFkDkahmfx07ODhg5cqVys/Ly8tx+/ZtxMTE4MqVKzh06BBMTU1rtM8JEybg5ZdfVqn9jz/+wNq1a/H777/jp59+qnGdZNhkxQXIO/l91YPychT+fhINulT+f4+IdBPDUB3r0KEDAODy5cuVwtCrr76Ks2fP4uTJkyph6OLFi2jatCmaNGmChQsX4vjx41i7dq3KOqGhoejXrx+GDBmC8PBw7N+/XyUktGvXDhcuXMCjR4/QoEEDlZp++OEHvPTSS0hOTq5Ur1wux/79+9G3b18kJSUhLi6uyjC0YMECXL58Gdu2bYOfnx9kMhlKS0vRtWtXtG7dGh999BG++eYbjBw5skb9EolE8PPzU1kWGBgIExMTzJgxA8eOHUOvXr1qtM+mTZtW2mdoaCgsLS0xd+5cnD17Fl26dKnRPsmwKWRlKMvLVDsuzU6FQi7juUNE9QSnyepY06ZN4eLigsuXLyuXFRUV4cqVKwgNDUXHjh1x6tQplW0uXryIjh07Ijc3F9988w369++vEoQqODg4YNasWUhNTcWhQ4dUxnr06AEjIyP8/PPPKstzc3Nx7tw5tYHi1KlTyMjIQNeuXdG7d29cvXoVKSkpKus8evQIu3fvxttvv10pZABAz549MXr0aDg7Oz+1NzXRpk0bAMD9+/drbZ9WVlaVluXl5WHOnDno0KED2rRpg//85z84e/asyjpnzpzBu+++C39/fwQFBeGDDz7ArVu3lONDhw7FzJkzsW7dOnTs2BHt2rXDhAkTkJaWprKf69evY/To0QgODka7du0wfvx4/PXXX8rx8+fPw8vLC2fPnsWoUaPQtm1bdOjQAQsXLkR5ebnKeoMGDVJbDwAcPXoU/fv3R5s2bdCxY0dER0ejuLj4hfpnyATGIogaNlE7burizSBEVI8wDGlAaGioShg6d+4cZDIZOnTogLCwMNy8eRPp6ekAgDt37iArKwthYWG4cOECJBIJunfvrnbfYWFhsLW1xbFjx1SWW1tbo2PHjvjxxx9Vlv/8889o1KgRfH19q9zf7t274e7ujrZt26JHjx6wtrZWnrtU4ezZsygvL0fXrl3V1jV9+nT07NlT7XhN3b59G8CTcFlTcrkc5eXlyo/CwkKcPXsWS5cuhYuLCwIDAwE8eb7X8OHDcezYMUyZMgUrV66Es7MzxowZowxEaWlpmDBhAlq1aoU1a9YgOjoat27dwvvvvw+5XK58zWPHjmH37t2YNWsW5s2bh5SUFAwbNkwZQM6dO4dBgwZBLpfj888/R3R0NDIyMjBw4EDcvHlTpf5p06YhICAAa9euxVtvvYVNmzZh165dynqmTJny1HoOHjyIDz/8EM2bN8eqVaswceJEHDhwAB988EGlc9moeoRmVrDrMqjKMYGJKSy8gzVcERG9CE6TaUCHDh2wa9cuZGdnw8HBASdPnkSbNm1ga2uL0NBQGBsb4+TJk3j33XeRkJAAY2NjhISEKE+AdnV1VbtvIyMjuLi4VHnEpGfPnvj000+Rk5OjvDP2Dz/8oPaoUF5eHn755RdMmjQJACAWi9GrVy8cOHAA06dPh4WFBYAnjx55Vl0v4p9HPQoLC3H9+nUsWLAALi4uzzWdNWvWLMyaNUtlmbm5OTp27IgZM2Yo39f+/fuRkpKC7777Dm3btgUAdO7cGUOHDkVMTAx2796NxMRElJaWYty4cXBycgLw5HEVx44dQ3FxMSwtLQEAxcXF2L17tzK8NW/eHP369cPevXsxePBgLF68GE2aNMGGDRuUd3cOCwtDjx49sGLFCixbtkxZ6zvvvIMPP/wQwJNgffToUfz2228YOHAgrl+/jtLSUowdOxaNGzeuVI+FhQViYmLQqVMnxMTEKPfp5uaGESNG4Pjx4yrnU1H1GTVsArs3PkD+L1sgLy0CABjbOcOx7xQY2zhouToiqgkeGdKAkJAQCAQCXLlyBcCTqaiwsDAAgKWlJXx9fXHmzBkAT06e9vX1haWlpfKvdmPjp2dWY2PjKv/Cf+WVVyAUCpVTZVlZWbh48aLaMHTgwAGUl5ejW7duKCgoQEFBAV577TUUFRXh4MGDyvUqHgj6zyMhteX+/fto1aqV8iM4OBhjxoyBvb09Vq9erXxMRE1MnDgRu3btwnfffYdp06ZBJBLhjTfewLJly9Ckyf+mOs6ePQsHBwe0atVKeRRJJpOha9eu+P3335Gfn4+2bdtCLBZjwIABWLBgAc6cOQNvb29MmTJFGYQAwN/fX+Uolo+PD5o0aYKLFy+iuLgY169fxxtvvKHymAtra2t07dq10hV+/v7+Kp87OzsrjzBV1PPuu+9WWc+tW7fw4MEDdOvWTeXoWFBQECwtLXH69Oka95OekAtFyDJrjEajFsFl9CK4jl2CxsOiYdrYg1NkRPUMjwxpgL29PTw9PXH58mV4enoiLS1NGYaAJ0cEtm/fDoVCgYsXLypPWHZxcQHwJCC4ubmp3X9aWprySMY/WVpaonPnzvjxxx/x3nvv4aeffoKHhwc8PT2rvKR+z549kMvlVYaluLg4DBw4UKWu9PR0lZPC/yk7Oxt2dnbPDHL/5uDggDVr1ig/F4lEcHZ2ho2NTY32808uLi7Kc47atm0LBwcHzJgxA0KhEPPmzVOul5eXh+zsbLRq1arK/WRnZ8PDwwPbt2/H119/je+++w6bN2+GtbU13nvvPUyePFkZFB0dHSttb29vj4KCAjx+/BgKhQINGzastE7Dhg3x+PFjlWX/vnrOyMhIGX5dXFywfv16bN26tcp68vLyAABRUVGIioqq9HoVD2Ol51NYVAyZuBnM7Zy0XQoRvQCGIQ3p0KEDrl27hiZNmsDKykolvISFhSE2Nhbnzp3DvXv3lEEpLCwMIpEIP//8Mzp27FjlfiuuGFN3XtEbb7yBadOm4eHDh4iPj1d7VCgpKQnJycmYOHEi2rdvrzL2yy+/YPPmzbh27Rratm2LkJAQmJiY4Pjx42qnrcaNG4eSkpJK5yw9i0gkUgaXutK3b1/89NNP2LlzJ3r06KE8Od3Kygpubm4q00n/VDEt6Ovri5UrV0IqleLSpUvYuXOn8vYGb7zxBgAoQ8g/PXz4EE2bNoWVlRUEAgEePqx8L5rs7GzY2trW6P20bt0asbGxkMlkleqpCKvTp0+v9HUF8EIhk4hIX3CaTENCQ0ORnJyMc+fOoUOHDirTIxXnD8XFxcHa2loZBqysrDBy5Ejs2rULJ06cqLTP3NxcREVFoWnTpmpDTteuXSEWi7Ft2zZcvXpV7Xq7du2CSCTCiBEjEBwcrPIxevRoCIVCxMXFAXgynTNgwAB89913SExMrLSvQ4cO4Y8//kCfPn1q3CdNmT17NsRiMaKjo5X3XGrfvj0yMjJgb2+PNm3aKD/Onj2rPLdn8+bN6NatG6RSKUQiEUJDQzF//nwAULkp5JUrV/Do0SPl53/88Qfu3buH0NBQmJubo3Xr1oiPj1c+dR54cjPM3377DQEBAdV+H1u3bkWvXr3U1tO8eXPY29vj3r17Ku/J2dkZixcvRlJS0gv1kYhIH/DIkIYEBQWhvLwcv/76K+bMmaMyZmRkhJCQEBw7dgzdunVTCUqTJk3CnTt3MGHCBAwYMADdunWDmZkZkpKSsHnzZigUCqxdu1b5hPF/Mzc3R5cuXbBx40b4+vqqnCNTQSqV4ocffkCXLl2qvNzc0dERHTt2RHx8PD755BNYW1vj448/xvXr1zF8+HAMHjwYQUFBKC4uxoULF/D999+jc+fOGDNmzAt2re64urpi9OjRWL16NTZv3oz3338f/fv3x/bt2zFy5EiMHz8ejRo1wpkzZ7B+/XoMGTIEJiYmCAkJQUxMDD788EMMGTJEGRJFIpHK1XUlJSUYO3YsJkyYgKKiIixduhSenp548803AQBTp07F6NGjMWbMGAwZMgRlZWX4+uuvIZVKMXHixGq/j+DgYCxevBiTJk3C0KFDK9UjFAoxZcoUzJkzB0KhEF27dkVBQQFWr16NzMxMtVOCRESGhGFIQ8zNzdG2bVvlPYT+LSwsDD/99FOlMRMTEyxfvhzx8fHYuXMnZsyYgdLSUri6uqJ///4YOnQo7Ozsnvrab7zxBn766SflFM6/HT16FHl5ecpf1FXp168fTpw4gb1792L48OGwtrbGtm3bsH37dsTHxyMuLg4KhQLNmjXDJ598gnfeeafG5wtp2rhx47Bv3z6sWbMGffr0gZOTE7755hssXrwYixYtwuPHj+Hi4oKpU6di1KhRAABvb2+sXbsWq1atwscffwyZTIbWrVtj06ZNaN68uXLfgYGBCAkJUV7F1q1bN0yfPl15p+vQ0FD897//RWxsLD7++GOIRCIEBgZi4cKFas/DqoqXlxeWLVuGjRs3qq3nnXfegYWFBTZs2ICdO3fC3Nwc7dq1Q0xMTJXhmIjI0AgUvNHIM12/fh0A1J7LUlpaitu3b8Pd3b3Gj4rQFxV3oDY1NVU5smWIhg4dCgDYtm1bnb9WXfad39dPV1xcjOTkZLz00ktqj8xS3WDvtUdfe6/bf7pTvSeXy6t1Cb5QKIRAIKjWPv95HyJ1BAKBwYcyIiKqHoYhqlOrVq1SefCqOlu3bkVwcPXu2lud81zat2+vkSMzRERU/zEMUZ36z3/+U607HLu7u1d7nxWPoniairtKawNDGBFR/cIwRHXKyclJ+diK2lLX9yEiIiLDUi/uMySXyxEbG4tOnTqhbdu2GDVqFO7cuVOtbQ8ePAgvLy/cu3evjqskIiKi+qhehKHVq1cjLi4O0dHR2LlzJwQCAcaOHau8WZ469+/fr/IRBEREREQVdD4MSaVSbNq0CZMmTUKXLl3g7e2NpUuXIjMzE0eOHFG7nVwuR0REBG8qR0RERE+l82EoJSUFRUVFCAkJUS6ztraGj48PEhIS1G63du1alJWVYdy4cZook4iIiOopnT+B+sGDBwCARo0aqSx3dHRUeRbUPyUmJmLTpk3YtWsXMjMza6UOhUKB4uLiKsckEgnkcjlkMpnKs6YMScW9OxUKhcH2QBvqsu8ymQxyuRwlJSXVuleUoSkpKVH5lzSHvdee+tj76twcUufDUEXDKx5jUEEsFiM/P7/S+sXFxZg2bRqmTZsGNze3WgtDZWVlSE5OVjtubGwMiURSK69Vn7EH2lEXfZdIJCgvL8etW7dqfd/6JDU1VdslGCz2XnvqU++r8/BrnQ9DFY8BkEqlKo8EkEgkMDMzq7R+dHQ03NzcMHDgwFqtw8TEBB4eHlWOSSQSpKenQywWG+xjCxQKBSQSCcRiscqdpF955RXI5XIcOHAAlpaWKtt8+umnuH//PrZs2aLpcvWGur7XFmNjYzRt2hRisbjW913flZSUIDU1FW5ublX+LKK6w95rj772XufDUMX0WFZWFpo2bapcnpWVBW9v70rr7969GyKRCP7+/gCgnDp488030bt3b8ybN++56hAIBGoPtRkZGcHIyAhCoVAnHwEhkyuQdCsHjwpK0cDaFD7N7SE0qt1fnBV9/vdjMAQCAR48eIBFixYhOjpaZRuBQMDHZrwgdX2vDUKhEEZGRjAzMzPYkF8dZmZmevWMpvqEvdcefeu9zochb29vWFpa4vz588owVFBQgKSkJAwZMqTS+ocPH1b5/Nq1a4iIiMDXX3+NFi1aaKRmXXImMR1f77uOnPxS5TJ7G1O837cNOvg21kgNTZo0wffff4/XXnsNnTp10shrEhERVZfOX00mEokwZMgQxMTE4NixY0hJScGUKVPg7OyMHj16QCaTITs7G6WlT37ZN2vWTOWj4u7HjRs3hr29vTbfisadSUzHgi0JKkEIAHLyS7FgSwLOJKZrpI7evXsjNDQUn332GQoLC9Wul5eXh6ioKHTp0gW+vr4YNGgQLl68qHb98+fPw8vLC+vXr0dwcDD69esHmUyGzMxMTJkyBYGBgQgODsb48eNV5rdzcnIQHh6O4OBg+Pr6YuDAgbhw4YJyvFu3bli7di3GjRsHX19f9OjRA99//73Ka1+5cgXDhg1DQEAAgoOD8emnn6qcw9atWzd8/fXXmDRpEvz9/REcHIwvvvhC+ZBZmUyGRYsWoUuXLmjdujVef/117NixQ+U1du/ejZ49e8LX1xc9e/bEli1beCIzUTUUlZThfnYh/r6Xh4yHRSgpLdN2SaTjdD4MAUB4eDgGDBiA2bNnY9CgQRAKhdi4cSNEIhEyMjIQFhaG+Ph4bZepU2RyBb7ed/2p66zf/ztkckWd1yIQCPD555+joKAACxYsqHIdmUyGUaNG4eLFi1i4cCH27t0Lb29vjBgxAtevP/19/Pbbb9i5cye++OILSCQSDB06FDKZDNu3b8e2bdtgZ2eH//znP8qT6SMjI1FaWort27fj4MGDcHd3xwcffKByteCqVavQpk0b7Nu3D4MHD8acOXOU32OJiYkYOnQoPDw8sHPnTsTGxiIxMRGjRo1SCSsrVqxAUFAQ9u7di0mTJmHr1q04dOgQAODbb7/FTz/9hKVLl+Lnn3/GkCFDEBkZqQx/O3fuxMKFC/Hhhx/ihx9+wEcffYT169cjJibm+b8QRAbgYV4Jlnx7CRMWHsOUpccxfuExrN17HY/y68/VT6R5Oj9NBjw5dyEiIgIRERGVxlxdXXHjxg212wYHBz91XF8l3cqpdETo3x7mlSDpVg7aeDSs83pcXFwQERGByMhIvP7665Wmy06dOoU//vgDBw8ehKenJwBgzpw5uHbtGjZu3Ihly5ap3feoUaPg5uYGAPj++++Rm5uLxYsXw8TEBADw+eef4/z58/juu+8wadIk3L17F56ensoTg2fNmoW33npL5Zybjh07YuLEiQCA5s2b49q1a9iyZQveeOMNbNq0CV5eXpgzZw4AwMPDA4sXL0bv3r1x8uRJdOnSBQDQqVMnDBs2DADg5uaGXbt24fLly+jbty/u3r0Lc3NzNGnSBA4ODhgyZAiaN2+ufGDt6tWrMW7cOLz55psAnkw1FhYWIioqCpMnT+YJzURVKCiSYOmOy0j8+6FymVyuwC8X0wABML5fG5iJTbRYIemqenFkiGruUcHTg1BN16sNAwcOVDtd9ueff8LKykoZhIAnR5QCAwOfGWYrghAAJCUlobCwEO3bt4e/vz/8/f0REBCABw8e4ObNmwCAiRMn4siRIwgODsaYMWOwa9cueHh4qASM4OBgldfw8/PDn3/+qay1Xbt2KuNeXl6wtrZWqfXf56hZWVmhrOzJ4frBgwejsLAQnTt3xjvvvINly5bB3t4e9vb2ePToER48eIDly5cr34O/vz/mzZsHiUTC5+wRqZFfKFUJQv/026V7yHv89Ec4keGqF0eGqOYaWFfv6p/qrlcbKqbL3nrrrUrTZQqFospLw+VyOYyNn/5t+s8QI5fL4e7ujjVr1lRar+LKhx49euDkyZM4efIkzpw5gw0bNmD58uX47rvv0LJlSwCo9JoKhQJGRkbPrLXiaBRQ+d5YFdsCTwLc4cOHceHCBZw+fRrHjh3D2rVrsWDBAuVRs08++QQdOnSotI9/34CUiJ7Ifaz+fltyuQJFJTx3iKrGI0N6yqe5Pextnh50Gtqawae5Zk8qd3FxwfTp07Fr1y6Vk6O9vLxQUFCgPPpS4dKlS2rv71QVT09PpKenw8rKSnkSvYuLCxYvXoyEhARIpVIsWLAAaWlpeOONNxAdHY0jR47AyMgIv/32m3I//z5P6fLly/Dx8VG+xr9P7E5JSUFhYWG1r1jcunUrDh8+jI4dO2L69Ok4ePAgQkNDER8frzxCdPfuXZWLAf7444+nThcSGTprC/VTYAIBYGbKv/+pagxDekpoJMD7fds8dZ2xfVrX+v2GqmPgwIHo0KED0tLSlMs6duwILy8vTJ06FefPn8fNmzcRFRWFP//8E8OHD6/2vnv37g0bGxtMnDgRV69exc2bN/HJJ5/g+PHjaNmyJUQiEa5du4bPPvsMV69exb1797Bnzx4UFRUp700FAD/88AO2b9+O1NRUbNiwAUeOHMGYMWMAACNGjEBKSgrmzZuHmzdv4sKFC5g2bRp8fHwQGhparTpzcnIwb948HDt2DPfv38eJEyeQlJQEf39/CAQCjBkzBtu2bcO2bdtw9+5dHD16FFFRURCJRFUecSIiwNZSjOYuNlWOBb3kBFtLnmtHVWNM1mMdfBvjk+FBle4z1NDWDGP7tNbYfYaqEh0djbfeekv5ubGxMf773/9i4cKFmDRpEqRSKVq1aoXNmzfDz8+v2vu1srLC9u3b8dVXX2HMmDGQyWR46aWXsHHjRuUU2PLly7FgwQJMmDABjx8/RvPmzbF48WIEBgYq99O3b18cPnwYCxcuhJubG5YtW6Y8Mdrf3x/r16/H8uXL0bdvX1haWuKVV17B1KlTVabJnmbixIkoLy/H/Pnz8fDhQzg4OOC9995TPlh41KhREIvF2LZtGxYuXAh7e3v0798fU6ZMqXYviAyNrZUpPhkehAVbEnDr/v9uddHGoyEmvN0WFmY8eZqqJlBUnMRAalVMmbRpU/WRltLSUty+fRvu7u46eadeTd2BurS0FKampvX+jtLdunVDv379MGnSJG2X8kx12Xdd/77WtuLiYiQnJ+Oll17Sqzvx1gfP6n3e41LkFUpQUCSFraUpbCxFsOFRoVqhr9/3PDJkAIRGAo1cPk9EpAtsrUxha8UAT9XHc4aIiIjIoPHIENG//PLLL9ougYiINIhHhoiIiMigMQwRERGRQWMYIiIiIoPGMEREREQGjWGIiIiIDBrDEBERERk0hiEiIiIyaLzPENWpoUOH4sKFCyrLTExM4OjoiO7du2Pq1Kl81AMREWkVw5ABUMhlKE1LhqwwF0JLO5g2eQkCI809P6xnz56YNWuW8vPi4mKcOnUKCxYsgEwmw5w5czRWCxER0b8xDOm5opRzeHh4E2SPc5TLhFb2aPjqKFh4h2ikBlNTUzg4OKgsa9asGX7//Xf88MMPDENERKRVDEN6rCjlHDJ3L6q0XPY4B5m7F8Hp7QiNBaKqiMViGBk9OW1NKpVi+fLlOHDgAAoLC9GyZUuEh4cjLCzsSc0yGZYsWYJDhw4hJycHrq6uGD58OAYNGgQAmDlzJqRSKRo2bIjdu3dDJBKhT58++PjjjyESiQAAGRkZWLx4Mc6ePYuioiIEBARg+vTp8PLyUu5DJpOhYcOG2LdvH4qLi9GxY0dERUUpw9y+ffuwfv163L17F7a2tnj99dcRERGhfI3Lly9j8eLFuH79Oho0aICuXbti6tSpsLS01GhviYio+ngCtZ5SyGV4eHjTU9d5eGQTFHKZhir6n/Lycvz222/Yv38/+vTpAwD45JNPcPLkSSxatAh79+5Fz549MX78ePz2228AgG+//RY//fQTli5dip9//hlDhgxBZGQkLl68qNzv4cOHkZ6ejh07diA6Ohr79u3D559/DgAoLCzEoEGDkJmZiTVr1iAuLg7m5uYYMmQI0tPTlfv48ccfkZeXh+3bt2PlypW4dOkSli5dCgBISUnB7NmzMWnSJPz888/44osvsH//fmzYsEE5PmLECHTs2BEHDhxATEwM/vjjD4waNQoKhUITrSUioufAI0N6qjQtWWVqrCqyghyUpiXDrFnrOq3l4MGD+Pnnn/9XW2kpGjdujNGjR2P8+PG4c+cODh06hF27dqFNmzYAgJEjRyIlJQUbN27Eyy+/jLt378Lc3BxNmjSBg4MDhgwZgubNm8Pd3V25XxsbGyxatAhmZmbw9PREVlYWPv/8c0RERODAgQPIzc3Fnj170KBBAwBATEwMXnnlFXzzzTeIiIgAAFhaWmLevHkwMTFBixYt0KdPHxw/fhwAcO/ePQgEAri6uqJx48Zo3LgxNm7cqDzqs3HjRoSGhuKDDz4AALi5uWHx4sV45ZVXcOHCBQQHB9dpn4mI6PkwDOkpWWFura73Irp164Zp06ZBLpfj2rVrWLBgATp06IDx48fD2NgYSUlJAIBhw4apbFdWVgZra2sAwODBg3H06FF07twZrVu3RseOHdGzZ0/Y29sr12/Tpg3MzMyUn/v7+6OsrAy3b9/Gn3/+CTc3N2UQAp5M0/n6+uLGjRvKZc2aNYOJiYnycysrK5SVlQEAOnXqBH9/f7z99ttwc3NDhw4d0L17d7Ru/SRMJiUl4c6dO/D396/Ug5s3bzIMERHpKIYhPSW0tKvV9V6EhYUFmjVrBgBwd3eHs7MzRo4cCaFQiMjISOUU0jfffAMLCwuVbSvOKXJzc8Phw4dx4cIFnD59GseOHcPatWuxYMEC9OvXDwBUQgwAyOVyAIBQKIRCoYBAIKhUm0wmg7Hx//43qDj3pypisRhbt25FUlISTp06hVOnTiEuLg59+/bFggULIJfL8dZbb2H8+PGVtv1nCCMiIt3Cc4b0lGmTlyC0sn/qOkJre5g2eUlDFf1PSEgIRo4ciR07duDEiRNo2bIlACArKwvNmjVTfuzZswe7d+8GAGzduhWHDx9Gx44dMX36dBw8eBChoaGIj49X7vePP/6ATPa/c6CuXLkCMzMzuLu7w9PTE7dv30ZOzv+mDiUSCX7//Xd4eHhUq+7jx49j5cqV8PHxwfvvv4+tW7ciPDxcWUPLli3x119/qbwHmUyGBQsWICMj44X7RkREdYNhSE8JjIRo+Oqop67TsMcojd5v6J8mT54MNzc3zJ07F40bN0bXrl0xd+5cHDt2DGlpadi4cSPWrVuHJk2aAABycnIwb948HDt2DPfv38eJEyeQlJSkMiV1//59REZG4ubNmzhy5AhiY2MxZMgQmJmZ4a233oK1tTU++ugjJCYmIiUlBRERESguLsa7775brZqNjY2xatUqbN68GWlpabh+/Tp+/fVXZQ2jRo1CcnIy5syZg7///hvXrl3DtGnTcPv2bbi5udV6D4mIqHZwmkyPWXiHwOntiMr3GbK2R8MemrvPUFXEYjHmz5+PYcOGYenSpcqPuXPnIj8/H02aNMH8+fPx9ttvAwAmTpyI8vJyzJ8/Hw8fPoSDgwPee+89jBs3TrlPPz8/CAQCvP3227C2tsawYcMwYcIEAIC1tTW2b9+OhQsXYsSIEQCAgIAA7NixQxm4nqVjx474/PPPsWnTJixduhSmpqbo0qULZs6cqXz9DRs2YPny5ejfvz/MzMwQEhKCGTNmPHX6jYiItEug4DW/z3T9+nUAUF7p9G+lpaW4ffs23N3ddfLREpq4A7VMJkNpaSlMTU0hFGr+aNPMmTNx//59bNu2TeOvrU112Xdd/77WtuLiYiQnJ+Oll16Cubm5tssxKOy99uhr73lkyAAIjIR1fvk8ERFRfVUvzhmSy+WIjY1Fp06d0LZtW4waNQp37txRu/7vv/+O4cOHw9/fHyEhIZgzZw4KCgo0WDERERHVF/UiDK1evRpxcXGIjo7Gzp07IRAIMHbsWEil0krrZmVlYeTIkWjatCn27t2L1atX4/Lly5gxY4YWKidN+fLLLw1uioyIiGqHzochqVSKTZs2YdKkSejSpQu8vb2xdOlSZGZm4siRI5XWv3//Pjp16oS5c+fCzc0N7dq1wzvvvIOzZ89qoXoiIiLSdTofhlJSUlBUVISQkP9d+WRtbQ0fHx8kJCRUWt/f3x9LlixR3kjv77//xt69e9GxY8c6r5XnopM+4fczERkKnT+B+sGDBwCARo0aqSx3dHR85o3sXnvtNaSmpsLFxQWrV69+oToUCgWKi4urHJPJZJDL5SgsLDTYS6grfnEqFAqVGx9S3arLvhcWFkIul6OsrEx5N2/6n5KSEpV/SXPYe+2pj72vzlVvOh+GKhr+75AhFouRn5//1G1jYmJQWlqKmJgYDBs2DPv376/0uIfqKisrQ3Jy8lPXycjIQFlZGcRicZWPfjAEEolE2yUYpNrsu0KhgEQiwcOHD1FWVoY///yz1vatj1JTU7VdgsFi77WnPvU+ICDgmevofBiquL+JVCpVudeJRCJReShnVSruC7RixQp06dIFR44cQd++fZ+rDhMTk6c+tkGhUCAnJwe5uXX/4FNdVHFkQigUGmwQ1Ia67LutrS3s7e359VSjpKQEqampcHNze+bPIqpd7L326GvvdT4MVUyPZWVloWnTpsrlWVlZ8Pb2rrT+zZs3ce/ePXTp0kW5zNHRETY2NsjMzHzuOgQCwTMPtVlYWEAmkymfcm5ISkpKcOvWLTRt2lSv/gfRdXXVdxMTE63cPLM+MjMz06ubz9Un7L326FvvdT4MeXt7w9LSEufPn1eGoYKCAiQlJWHIkCGV1j958iSWLVuGU6dOwdLSEgBw9+5d5ObmokWLFnVer1AoNMhfIhXnlIjFYt6tWIPYdyKiF6fzV5OJRCIMGTIEMTExOHbsGFJSUjBlyhQ4OzujR48ekMlkyM7ORmlpKQCgT58+sLKyQkREBP766y9cvHgR4eHh8PX1RdeuXbX8boiIiEjX6HwYAoDw8HAMGDAAs2fPxqBBgyAUCrFx40aIRCJkZGQgLCwM8fHxAAA7Ozts3boVcrkcgwYNwocffggfHx9s3LjRII/YEBER0dPp/DQZ8GTqKSIiAhEREZXGXF1dcePGDZVl7u7uWLdunabKIyIionqsXhwZIiIiIqorDENERERk0BiGiIiIyKAxDBEREZFBYxgiIiIig8YwRERERAaNYYiIiIgMGsMQERERGTSGISIiIjJoDENERERk0BiGiIiIyKAxDBEREZFBYxgiIiIig8YwRERERAaNYYiIiIgMGsMQERERGTSGISIiIjJoDENERERk0BiGiIiIyKAxDBEREZFBYxgiMgByuQKFxWUokZRpuxQiIp1jrO0CiKhuZeUW40xiOk5dS4eZ2Bh9OreAh6stbK3E2i6NiEgnMAwR6bHMR8WYsfIkcvJLlcuu/pmNLu1cMLZPG9hYMhAREXGajEhPSctk2PPrXypBqMLxy/eR8bBIC1UREekehiEiPVVQJMWvl+6pHT928a4GqyEi0l0MQ0R6TKFQqB2Ty9SPEREZEoYhIj1lZW6CTn4uase7BTXVYDVERLqLYYhIT4lFxninuydsLEWVxtr7OMHFwVILVRER6R5eTUakxxo1tMDiyZ1x9MJdnEnM+P9L65ujdYuGvLSeiOj/MQwR6TmnBhYY2MMLvcKaQygQwMqi8pEiIiJDVi+myeRyOWJjY9GpUye0bdsWo0aNwp07d9Su/9dff+H9999HcHAwQkNDER4ejvT0dA1WTKRbhEIj2FqKGYSIiKpQL8LQ6tWrERcXh+joaOzcuRMCgQBjx46FVCqttG5ubi5GjhwJCwsLbN++HevXr0dubi7GjBkDiUSiheqJiIhIl+l8GJJKpdi0aRMmTZqELl26wNvbG0uXLkVmZiaOHDlSaf2jR4+ipKQEX375JVq2bInWrVtj0aJFuHnzJi5fvqyFd0BERES6TOfDUEpKCoqKihASEqJcZm1tDR8fHyQkJFRaPzQ0FKtWrYJYXPnk0Pz8/DqtlYiIiOofnT+B+sGDBwCARo0aqSx3dHRERkZGpfVdXV3h6uqqsmzdunUQi8UICgp67joUCgWKi4ufe3t9V1JSovIvaQb7rj3svfaw99pTH3tvbm7+zHV0PgxVNFwkUj3xUywWV+tIz9atW/Htt9/ik08+gb29/XPXUVZWhuTk5Ofe3lCkpqZquwSDxL5rD3uvPey99tSn3gcEBDxzHZ0PQ6ampgCenDtU8d8AIJFIYGZmpnY7hUKB5cuXY82aNRg3bhxGjBjxQnWYmJjAw8Pjhfahz0pKSpCamgo3N7enfl2odrHv2sPeaw97rz362nudD0MV02NZWVlo2vR/jw/IysqCt7d3lduUlZXhk08+waFDhzB9+nSMHj36hesQCATVOtRm6MzMzNgnLWDftYe91x72Xnv0rfc6fwK1t7c3LC0tcf78eeWygoICJCUlITAwsMptpk+fjp9++gmLFy+ulSBERERE+kvnjwyJRCIMGTIEMTExaNCgAVxcXLBo0SI4OzujR48ekMlkePToEaysrGBqaoo9e/YgPj4e06dPR/v27ZGdna3cV8U6RERERBV0/sgQAISHh2PAgAGYPXs2Bg0aBKFQiI0bN0IkEiEjIwNhYWGIj48HABw6dAgA8NVXXyEsLEzlo2IdIiIiogo6f2QIAIRCISIiIhAREVFpzNXVFTdu3FB+vmnTJk2WRkRERPVcvTgyRERERFRXGIaIiIjIoDEMERERkUFjGCIiIiKDxjBEREREBo1hiIiIiAwawxAREREZNIYhIiIiMmgMQ0RERGTQGIaIiIjIoDEMERERkUFjGCIiIiKDxjBEREREBo1hiIiIiAwawxAREREZNIYhIiIiMmgMQ0RERGTQGIaIiIjIoDEMERERkUFjGCIiIiKDxjBEREREBs1Y2wUQEekbWUkhZMX5kBc/hpGpOYzMbWBsYaPtsohIDYYhIqJaVP74EbJ/XIeSvy4ql4mcm8Op/zSY2DlpsTIiUofTZEREtUQuKUHOsa0qQQgApA9uIXPXQpQX5mmnMCJ6KoYhIqJaIivOQ1HS6SrHpFl3ICvM1XBFRFQdDENERLVELikFFHK14wxDRLqJYYiIqJYYic0AI6HacaFVAw1WQ0TVxTBERFRLhBa2sGrTtcoxcSMPCC3tNFwREVUHryYjIqolRiJT2HV5Fwp5GQp/P6mcMjNt1hoOb03k5fVEOqpehCG5XI6VK1fi+++/R0FBAQICAjB37lw0a9bsmduNHTsWfn5+mDRpkoaqJSJDZmzVAPavjYFd2DuQlRbBSGQGoYU1hGZW2i6NiNSoF9Nkq1evRlxcHKKjo7Fz504IBAKMHTsWUqlU7TalpaWIiIjAqVOnNFgpEREgFJvDpEEjmDb2gKihC4MQkY7T+TAklUqxadMmTJo0CV26dIG3tzeWLl2KzMxMHDlypMptLl++jH79+uHatWuwtrbWcMVERERUn+h8GEpJSUFRURFCQkKUy6ytreHj44OEhIQqtzl58iR69OiBffv2wcqKf5ERERGRejp/ztCDBw8AAI0aNVJZ7ujoiIyMjCq3mTx5cq3XoVAoUFxcXOv71RclJSUq/5JmsO/aw95rD3uvPfWx9+bm5s9cR+fDUEXDRSKRynKxWIz8/HyN1VFWVobk5GSNvV59lZqaqu0SDBL7rj3svfaw99pTn3ofEBDwzHV0PgyZmpoCeHLuUMV/A4BEIoGZmZnG6jAxMYGHh4fGXq++KSkpQWpqKtzc3DT6dTF07Lv2sPfaw95rj772XufDUMX0WFZWFpo2bapcnpWVBW9vb43VIRAIqnWozdCZmZmxT1rAvr+48qI8yArzIHv8CEJLOwgt7WBsafvM7dh77WHvtUffeq/zYcjb2xuWlpY4f/68MgwVFBQgKSkJQ4YM0XJ1RKQPyvKykLk7BtIHN5XLRI7N4PTODJjYOmmxMiLSBJ2/mkwkEmHIkCGIiYnBsWPHkJKSgilTpsDZ2Rk9evSATCZDdnY2SktLtV0qEdVDspLHyD6wQiUIAU+eMp+1ZzFkRZo7N5GItEPnwxAAhIeHY8CAAZg9ezYGDRoEoVCIjRs3QiQSISMjA2FhYYiPj9d2mURUD8mK8lGallTlmCTjJmTFBRquiIg0TeenyQBAKBQiIiICERERlcZcXV1x48YNtdv+8ssvdVkaEQF4lF+CnAIJ8gslcLA1g62VGDaWYm2XVS1y6dOPKsslvKUGkb6rF2GIiHTX/axCRG08h4yHRcplrVvYY+p7AWhoq/tXmxiZWgAQAFBUPW5mqdF6iEjz6sU0GRHppkf5pYjccFYlCAHA7zdzsGH/7yguLdNSZdUntLCBRauOVY6Ze7aH0JxPmifSdwxDRPTcHuaX4EFO1dNIZ6+nI79QouGKak4oNod99+GwbNsNEPz/j0SBESxbd0bD18dCyCNDRHqP02RE9NzyHqsPO3IFUCqVabCa52ds1QANXx0Nu45vQy4pgZHYFEILWxiJdH+aj4heHMMQET03Bzv1YcHE2Ahm4vrzI8ZIZAojkbO2yyAiLajRNFleXh7++9//IioqCjt27KjyQW1paWn45JNPaq1AItJdDaxN8ZKbXZVjvTq6o4G1aZVj9UFuQSnuZxfiQU4RSiS6f+4TET2/av/ZdvfuXbz33nvIzc2FpaUlduzYgfXr12PFihVo1aqVcr1Hjx5h3759WLBgQZ0UTES6w8ZSjOlDA7Fu73Wc/+MBFIonR4R6dXBH/64eEJkItV1ijZVIypGc+ghr9yQi42ERjARAcOtGGP1WKzjZW2i7PCKqA9UOQ4sWLULDhg2xd+9eODg44Ny5c/jss88wfPhwbNy4EW3btq3LOolIRzW0NcdHg9oh/7EEpVIZLEyNYWslhlhUf6bI/unW/XzM/fqs8nO5Ajh7PQO37ufjyw/D6sXtAoioZqo9TXblyhV88MEHcHBwAACEhIRg586daNSoEd5//33cvHnzGXsgIn1lYWqCxg6WaO5iAyd7i3obhPILJdh44PcqxzIfFePve3maLYiINKLaYUgqlcLUVHX+v0GDBtiwYQPMzMwwduxYZGdn13qBRESaIi2T4a+0PLXjV25kaa4YItKYaochLy8v7N69u9JyJycnrFu3Dvn5+Rg9ejQyMzNrtUAiIk0xMhLA2kKkdtyxgbkGqyEiTal2GJo4cSKOHTuGPn36YM+ePSpjXl5eWLlyJdLS0jB16tRaL5KISBNsrUzRt0uLKseMBEBI60YaroiINKHaYSg4OBjbtm1D48aNkZVV+VBxaGgovvnmG3h4eNRqgUREmiI0EqB7UFMEveSkstxYKMCMYUFoaFN/bxVAROrV6CxHf39/rFmzRu24j48P9u7di7S0tBcujIhIGxpYm2LyQH/k5JciOfURLM1N4NXUDg2sTevlrQKI6NlqfMmHVCpFQUEBGjZsqLI8Pj4e3bp1g6mpKZo0aVJrBRIRaZqNpRg2lmI0d+FDWokMQY3uQH3y5El069YN27ZtU1menZ2Njz/+GC+//DIuXLhQqwUSERER1aVqh6GUlBR8+OGHcHR0RFhYmMpYgwYNsGbNGjRq1AhjxozhPYeIiIio3qh2GPr666/h7e2NuLg4BAUFqYwJhUJ07doV3377LZo2bYq1a9fWeqFEREREdaFGd6AeOnQoRCL19+AwMzPD8OHDcfny5VopjoiIiKiuVTsMPXr0CM7Ozs9cr1mzZnj48OELFUVERESkKdUOQ46Ojrh3794z10tPT4e9vf0LFUVERESkKdUOQx07dkRcXBwUCoXadeRyOeLi4vgEeyIiIqo3qh2GRowYgT///BMfffRRldNgOTk5mDp1Kn7//XcMHz68VoskIiIiqivVvumim5sbvvrqK0yfPh0vv/wyWrVqBVdXV8hkMqSnpyMpKQnGxsaIjo6Gn59fHZZMREREVHtqdAfqHj164MCBA9i6dStOnTqFX375BUZGRnBxccGwYcMwePBguLi41FWtRERERLWuxo/jyM3NRbt27dC3b1+0atWqLmoiIiIi0phqh6GCggKMGzcOV69eVS7z8/PD0qVLq3XJPREREZEuqvYJ1MuWLUNSUhImTZqEdevWYcaMGbh9+zZmz55dl/URERER1alqHxn69ddf8fHHHyuvFOvcuTOcnJwwbdo0FBcXw9zcvM6KJCIiIqor1T4ylJ2dXekcoeDgYMhkMmRkZNR6Yf8kl8sRGxuLTp06oW3bthg1ahTu3Lmjdv3c3FxMnToVQUFBCAoKwmeffYbi4uI6rZGIiIjqp2qHofLy8krPJbOxsQEASCSS2q3qX1avXo24uDhER0dj586dEAgEGDt2LKRSaZXrh4eHIy0tDZs3b0ZsbCxOnz6NqKioOq2RiIiI6qdqh6GnedpdqV+UVCrFpk2bMGnSJHTp0gXe3t5YunQpMjMzceTIkUrrX7lyBRcuXMCCBQvQqlUrhIaGYt68edi/fz8yMzPrrE4iIiKqn2olDAkEgtrYTZVSUlJQVFSEkJAQ5TJra2v4+PggISGh0voXL16Eg4MDWrRooVzWvn17CAQCXLp0qc7qJCIiovqpRvcZioyMhKWlpfLziiNCn332GSwsLJTLBQIBtmzZUisFPnjwAADQqFEjleWOjo5VnquUmZlZaV2RSARbW9sXOrdJoVDwvKOnKCkpUfmXNIN91x72XnvYe+2pj72vzgVe1Q5DQUFBACpPiVW1vDanzSoa/u/zlcRiMfLz86tc/9/rVqz/Iuc2lZWVITk5+bm3NxSpqanaLsEgse/aw95rD3uvPfWp9wEBAc9cp9phaNu2bS9UzPMyNTUF8OTcoYr/Bp6ctG1mZlbl+lWdWC2RSF7o8n8TExN4eHg89/b6rqSkBKmpqXBzc6vy60J1g33XHvZee9h77dHX3tf4cRyaVjHllZWVhaZNmyqXZ2Vlwdvbu9L6zs7OOHr0qMoyqVSKvLw8ODk5PXcdAoGA91KqBjMzM/ZJC9h37WHvtYe91x59632tnEBdl7y9vWFpaYnz588rlxUUFCApKQmBgYGV1g8KCsKDBw9U7kNUsW27du3qvmAiIiKqV3T+yJBIJMKQIUMQExODBg0awMXFBYsWLYKzszN69OgBmUyGR48ewcrKCqampmjbti3atWuHKVOmIDIyEsXFxZg7dy769u37QkeGiIiISD/p/JEh4MlNFAcMGIDZs2dj0KBBEAqF2LhxI0QiETIyMhAWFob4+HgAT6azVq5cCVdXVwwfPhwfffQROnfujMjISO2+CSIiItJJOn9kCACEQiEiIiIQERFRaczV1RU3btxQWWZvb4/Y2FhNlUdERET1WL04MkRERERUVxiGiIiIyKAxDBEREZFBYxgiIiIig8YwRERERAaNYYiIiIgMGsMQERERGTSGISIiIjJoDENERERk0BiGiIiIyKAxDBEREZFBYxgiIiIig8YwRERERAaNYYiIiIgMGsMQERERGTSGISIiIjJoDENERERk0BiGiIiIyKAxDBEREZFBYxgiIiIig8YwRERERAaNYYiIiIgMGsMQERERGTSGISIiIjJoDENERERk0BiGiIiIyKAxDBEREZFBYxgiIiIig8YwRERERAaNYYiIiIgMms6HIYlEgqioKISGhsLf3x/h4eHIycmp1ralpaV46623sGfPnjqukoiIiOornQ9DkZGROH36NFasWIEtW7YgLS0NkydPfuZ2eXl5GDduHP78808NVElERET1lU6HoczMTOzbtw+zZ89GYGAgfH19sWTJEiQkJODq1atqtzt27Bj69OmDx48fa65YIiIiqpeMtV3A01y6dAkAEBwcrFzm7u4OJycnJCQkwM/Pr8rtfvvtNwwdOhTDhg1DmzZtaqUWhUKB4uLiWtmXPiopKVH5lzSDfdce9l572HvtqY+9Nzc3f+Y6Oh2GMjMzYWdnB7FYrLLc0dERGRkZarebP39+rddSVlaG5OTkWt+vvklNTdV2CQaJfdce9l572HvtqU+9DwgIeOY6Wg1D9+7dQ/fu3dWOT548GSKRqNJysVgMiURSl6VVYmJiAg8PD42+Zn1SUlKC1NRUuLm5wczMTNvlGAz2XXvYe+1h77VHX3uv1TDk5OSE+Ph4tePHjx+HVCqttFwikWj8iyAQCKp1qM3QmZmZsU9awL5rD3uvPey99uhb77UahkxMTNCiRQu14zdu3EBeXh6kUqnKEaKsrCw4OztrokQiIiLSczp9NVlAQADkcrnyRGoAuHXrFjIzMxEYGKjFyoiIiEhf6HQYcnJyQq9evTB79mycP38eiYmJmDp1Ktq3b6+8kkwqlSI7O7vK6TQiIiKiZ9HpMAQ8uTIsNDQUEydOxOjRo9G8eXPExsYqx69cuYKwsDBcuXJFi1USERFRfaXTl9YDT+4PEB0djejo6CrHg4ODcePGDbXbP22MiIiISOePDBERERHVJYYhIiIiMmgMQ0RERGTQGIaIiIjIoDEMERERkUFjGCIiIiKDxjBEREREBo1hiIiIiAwawxAREREZNIYhIiIiMmgMQ0RERGTQGIaIiIjIoDEMERERkUFjGCIiIiKDxjCkRXK5AtIyGRQKhbZLISIi0gp5mQSyksdQyMq1VoOx1l7ZgEmk5cjKLcGR83dwN/MxWrnbI8zPBY52ZhAKmU+JiEj/yUoKIX14D/nn9kP2OAemzVrD2r8HjG0dITASarQWhiENK5fJcPXPbHyx+QLk/39A6FJKFr479ie++KAjWjax026BpHWy4gIo5DIYiS1gZCLSdjlERLVOLilGwZUjyP11u3KZJOMmCi79jMbDoyF2ctdoPTwMoWGP8iVY/O1lZRCqUCqVYcm3l5H7uFQ7hZHWlRfm4fH148j4JhL3N83Aw5/XQ5qTrtVDx0REdUFWlI/cX7+ptFxRVoqH8WshKy7QaD08MqRhWbnFKJFU/cvtXlYhCoqksLMy1XBVpG2y4nzkHN6AouSzymWF135B0R+n0HjklxA7NtNidUREtav0/p8Aqj5fVpL+N2SlhRCaW2usHh4Z0rCycvlTx+X/PmREBqEs/6FKEKqgKJci5+hmyEqLtFAVEVEdedaFQxr+VcgwpGGNGlpAaCSocszWSgwrc54jYoiK/7qodqz0diLkDENEpEfELp4Aqv5dKHJ2h9DMQqP1MAxpmI2lGINe9apy7IO3fdHAmlNkhkhgbKJ+0EgICKr+oUFEVB8JLWxgGzag0nKBsQgN3xgPobmNRuvhOUMaZiY2xhsd3NHcxQY7Dt9AVm4x3BrZYGhPbzR1toaRmqNGpN8sWgZWeTIhAFh4h0JoZqXhioiI6o7Q1AI27XvBtFkr5J/eg/KiXJg19YF1UC+Y2DppvB6GIS2wshAhyMcZXs3sUFYuh6nIGBZmTzkyQHpPaGUP27AByDu1S3W5pS0avDwIRiIeMSQi/SI0s4K5WxuIG7UAyssgEJvD6GlHyesQw5AWWVuItV0C6YgnfyW9CXOPAOQnxENelA9zr/YwbxkIExsHbZdHRFRnhGJzQMu/DhmGiHSE0MwKQhcriJybA7JyHg0iItIQhiEiHWMkNAaE/F+TiEhTeDUZERERGTSGISIiIjJoOh+GJBIJoqKiEBoaCn9/f4SHhyMnJ+ep21y+fBlDhw5FQEAAOnXqhFmzZiEvL08zBRMREVG9ovNhKDIyEqdPn8aKFSuwZcsWpKWlYfLkyWrXv337NkaPHg1vb298//33WLp0KRITExEeHq7BqomIiKi+0OkwlJmZiX379mH27NkIDAyEr68vlixZgoSEBFy9erXKbfbt2wdHR0d8+umnaN68OQIDAzF37lycP38eaWlpmn0DREREpPN0OgxdunQJABAcHKxc5u7uDicnJyQkJFS5Te/evbFw4UIIqnh8AafKiIiI6N90+vrdzMxM2NnZQSxWvRuTo6MjMjIyqtymRYsWlZatX78eDg4O8Pb2fu5aFAoFiouLn3t7fVdSUqLyL2kG+6497L32sPfPZmQkgEChgBwCKJ71hPgaqI+9Nzc3f+Y6Wg1D9+7dQ/fu3dWOT548GSJR5ae4i8ViSCSSar3Gl19+iePHjyM2NhYmJs9/m++ysjIkJyc/9/aGIjU1VdslGCT2XXvYe+1h7ytztLWGrYkMhdd/hbwwF6Lm/hC6+uDuwwJIpdJae5361PuAgIBnrqPVMOTk5IT4+Hi148ePH6/yiyeRSGBmZvbUfZeVlWHOnDnYu3cv5s6di1dfffWFajUxMYGHh8cL7UOflZSUIDU1FW5ubs/82lDtYd+1h73XHva+akayMpQmnUD20c3KZcV/X4LQ0g6eQ+ZBZmb7wq+hr73XahgyMTGpclqrwo0bN5CXlwepVKpyhCgrKwvOzs5qtyssLMTEiRNx8eJFLF68GL169XrhWgUCQbUOtRk6MzMz9kkL2HftYe+1h71XJc1Jx6N/BKEKssJc5P32DRzenAgjce0EGH3rvU6fQB0QEAC5XK48kRoAbt26hczMTAQGBla5jVQqxbhx43D9+nVs2LChVoIQERGRriu5fU3tWNGNC5CVPNZgNfWLTochJycn9OrVC7Nnz8b58+eRmJiIqVOnon379vDz8wPwJPxkZ2crp9PWrVuHS5cuYd68eWjRogWys7OVH7U5X0pERKRLFNLSpwzKgVo8kVrf6HQYAoD58+cjNDQUEydOxOjRo9G8eXPExsYqx69cuYKwsDBcuXIFAHDo0CEoFAp8/PHHCAsLU/moWIeIiEjfmDVvq3ZM3NgTRmL9mdaqbTp9aT3w5JK46OhoREdHVzkeHByMGzduKD//+eefNVUaERGRzjC2bghz71AUp5xVHTAyhv3royE0t9JOYfWAzochIiIiejahuTUavjYaxS38kX9uP2Qlj2HW1Ad2nd6FsX0jbZen0xiGiIiI9ISxpR2s/brD3CMAkMtgZGoOI5H+XAJfVxiGiJ5CVvIYsqI8lOdmwcjcCsZW9jC2ttd2WURET2VsaavtEuoVhiEiNcofP8LD+HUo/vuicpnQqgGcB86G2LGZFisjIqLapPNXkxFpg7xcitzTu1WCEADIHj9CxjeRKCt4qKXKiIiotjEMEVVBVpiHwqvHqhyTFxegLCddwxUREVFdYRgiqoKiXAqFrEzteHlepgarISKiusQwRFQFgYkYRqYWasdFDk01WA0REdUlhiGiKhhbNoBth7erHDNp6ApjW0cNV0RERHWFV5MRVUEgFMLS92UoyiXIO7sPijIJAMDU3RcOb0yAsaWdliskIqLawjBEpIaxhQ1sQvvBss3LkEuKITAWQWhuDaGZpbZLIyKiWsQwRPQURsYmMOKUGBGRXuM5Q0RERGTQGIaIiIjIoDEMERERkUFjGCIiIiKDxjBEREREBo1hiIiIiAwawxAREREZNIYhIiIiMmi86SIZpMJiKfILpcgvksDc1Bi2lqawtRJruywiItIChiEyOI/yS7FubyLOXM9QLmvqbIVZI9ujcUM+aoOIyNBwmowMSqm0HN8eTlYJQgBw98FjRK4/h0cFpVqqjIiItIVhiAxK3mMJjiWkVTmW8bAID3NLNFwRERFpG6fJyKCUSstRLlOoHc/KK4ZnMzsNVkT6LDu3GLfTC3ArPR9Nnazg0cQWDrZmEAgE2i6NiP6BYYgMipnYGCbGRigrl1c57tTAQsMVkb5Ky3yMWWtOI/exRLnM0swEX3zQEe6NbbRYGRH9G6fJyKDYWZmiZ6hblWPNnK3Q0NZUswWRXsp7LMFX2y6qBCEAKCwpw/xN55GTz+lYIl3CMEQGRWQixIBuLfFaSDMYGf1vqsLHvQE+Gx0MOyuGIXpx+UUSpGYUVDmWnVuCvH+FJCLSLk6TkcGxszbF6N6t8XbXligslsJUbAxbSzGsLETaLo30RFmZ7KnjkmeME5Fm6fyRIYlEgqioKISGhsLf3x/h4eHIycl56jZnzpzBO++8g7Zt26Jz585YtGgRpFKphiqm+sBMbIxGDS3QsqkdmjhZMQhRrbKyEENkXPWPVyMBYMcbfBLpFJ0PQ5GRkTh9+jRWrFiBLVu2IC0tDZMnT1a7fkpKCt5//3106tQJhw4dwoIFC3DgwAF89dVXGqyaiAyZnZUYb3drWeVYzw7usLVkGCLSJTo9TZaZmYl9+/Zh3bp1CAwMBAAsWbIEr7/+Oq5evQo/P79K26Snp6N///4IDw8HADRp0gQ9e/bE6dOnNVk6ERkwkYkQvTq6w85KjB2HbyD3sQTWFiK83dUD3QKbwMzURNslEtE/6HQYunTpEgAgODhYuczd3R1OTk5ISEioMgx169YN3bp1AwAoFAokJibiyJEj6NGjh0ZqJiICABtLMV4PdUN7H2dIy+UwMTaCnbUphEa8xxCRrtHpMJSZmQk7OzuIxaqHlB0dHZGRkaFmqydkMhnatWuH0tJS+Pj4YOLEiS9Ui0KhQHFx8QvtQ5+VlJSo/Euawb5rT3V7byYCzEQCAApISvl1qg38vtee+th7c3PzZ66j1TB07949dO/eXe345MmTIRJVPrFVLBZDInn6palyuRzbt29Hbm4uvvjiC4wdOxY7duyAkdHznSZVVlaG5OTk59rWkKSmpmq7BIPEvmsPe6897L321KfeBwQEPHMdrYYhJycnxMfHqx0/fvx4lVeBSSQSmJmZPXXfJiYmaNOmDQDAxsYG//nPf3Dp0iUEBQU9V60mJibw8PB4rm0NQUlJCVJTU+Hm5vbMrw3VHvZde9h77WHvtUdfe6/VMGRiYoIWLVqoHb9x4wby8vIglUpVjhBlZWXB2dm5ym1+//13FBcXo3379splLVs+uaojMzPzuWsVCATVOtRm6MzMzNgnLWDftYe91x72Xnv0rfc6fWl9QEAA5HK58kRqALh16xYyMzOVV5f92/79+zFz5kzIZP+7qdm1a9cAgEd2iIiIqBKdDkNOTk7o1asXZs+ejfPnzyMxMRFTp05F+/btlVeSSaVSZGdnK6fTBg8ejNzcXERGRuL27ds4ceIEZs2ahddeew3e3t5afDdERESki3Q6DAHA/PnzERoaiokTJ2L06NFo3rw5YmNjleNXrlxBWFgYrly5AgBwc3PDli1bkJqaiv79++PTTz/Fa6+9hpiYGG29BSIiItJhOn1pPfDkkrjo6GhER0dXOR4cHIwbN26oLPP19cW2bds0UR4RERHVczp/ZIiIiIioLjEMERERkUFjGCIiIiKDxjBEREREBo1hiIiIiAwawxAREREZNIYhIiIiMmgMQ0RERGTQGIaIiIjIoDEMERERkUFjGCIiIiKDxjBEREREBo1hiIiIiAwawxAREREZNIYhIiIiMmgMQ0RERGTQjLVdABERkS5TKBSQPX6E8sJcKMpKYWzdEEJzGxiJzbRdGtUShiEiIiI1FHIZJA9uIfO7LyErynuyUGAE66CesOvQH0ILW22WR7WE02RERERqlBc8RMb2yP8FIQBQyFFw4QcUpZyHQqHQWm1UexiGiIiI1Ci9kwRFWWmVY7mnvoesMFfDFVFdYBgiIiJSQ5p9R+2YrDAXCplMg9VQXWEYIiIiUkPcuKXaMWNbJwiMeeqtPmAYIiIiUkPs4gkjc+sqxxq8/B6MLe00XBHVBYYhIiIiNUxsHNB4yDyYODRVLhOYmKLBKyNg1rytFiuj2sTje0RERE8hcmiCRoMjIS8ugKK8DEbmlhBa2sFIaKLt0qiWMAwRERE9g7GFDWBho+0yqI5wmoyIiIgMGsMQERERGTSGISIiIjJoDENERERk0HQ+DEkkEkRFRSE0NBT+/v4IDw9HTk5Otbdfs2YNvLy86rBCIiIiqs90PgxFRkbi9OnTWLFiBbZs2YK0tDRMnjy5WtsmJiZi5cqVdVwhERER1Wc6HYYyMzOxb98+zJ49G4GBgfD19cWSJUuQkJCAq1evPnXb4uJiREREIDAwUDPFEhERUb2k0/cZunTpEgAgODhYuczd3R1OTk5ISEiAn5+f2m0///xzeHp6omvXrjh37twL16JQKFBcXPzC+9FXJSUlKv+SZrDv2sPeaw97rz31sffm5ubPXEenw1BmZibs7OwgFotVljs6OiIjI0PtdkeOHMHx48dx8OBB/Prrr7VSS1lZGZKTk2tlX/osNTVV2yUYJPZde9h77WHvtac+9T4gIOCZ62g1DN27dw/du3dXOz558mSIRKJKy8ViMSQSSZXbZGZm4rPPPsNXX30FO7vae4CeiYkJPDw8am1/+qakpASpqalwc3ODmZmZtssxGOy79rD32sPea4++9l6rYcjJyQnx8fFqx48fPw6pVFppuUQiqfKLoFAoMHPmTPTs2ROdO3eutTrLysqgUChw8+bNWtunvlEoFACA+/fvQyAQaLkaw8G+aw97rz3svfbUx96LRKJnXlWu1TBkYmKCFi1aqB2/ceMG8vLyIJVKVY4QZWVlwdnZudL66enpOHPmDC5fvox9+/YBAMrLywEA/v7+GDduHMaPH1/jOuvLF1ybBAJBlUfxqG6x79rD3msPe689+tp7nT5nKCAgAHK5HJcuXUJoaCgA4NatW8jMzKzyKjEnJyccPnxYZdnhw4cRExODffv2wcbm+R6y5+/v/1zbERERke7T6TDk5OSEXr16Yfbs2fjiiy9gZmaGuXPnon379soryaRSKfLz82FjYwORSIRmzZqp7MPe3h4AKi0nIiIiAnT8PkMAMH/+fISGhmLixIkYPXo0mjdvjtjYWOX4lStXEBYWhitXrmixSiIiIqqvBIqKs6GIiIiIDJDOHxkiIiIiqksMQ0RERGTQGIaIiIjIoDEMERERkUFjGCIiIiKDxjBEREREBo1hiIiIiAwawxAREREZNIYhIiIiMmgMQ/RcJBIJoqKiEBoaCn9/f4SHhyMnJ6fa269ZswZeXl51WKF+ep6+X758GUOHDkVAQAA6deqEWbNmIS8vTzMF12NyuRyxsbHo1KkT2rZti1GjRuHOnTtq18/NzcXUqVMRFBSEoKAgfPbZZyguLtZgxfqjpr3/66+/8P777yM4OBihoaEIDw9Henq6BivWHzXt/T8dPHgQXl5euHfvXh1XWfsYhui5REZG4vTp01ixYgW2bNmCtLQ0TJ48uVrbJiYmYuXKlXVcoX6qad9v376N0aNHw9vbG99//z2WLl2KxMREhIeHa7Dq+mn16tWIi4tDdHQ0du7cCYFAgLFjx0IqlVa5fnh4ONLS0rB582bExsbi9OnTiIqK0nDV+qEmvc/NzcXIkSNhYWGB7du3Y/369cjNzcWYMWMgkUi0UH39VtPv+wr379+v39/vCqIaevDggcLb21tx/Phx5bJbt24pPD09FVeuXHnqtkVFRYpXX31VMWzYMIWnp2cdV6pfnqfvS5YsUbz66qsKuVyuXJaQkKDw9PRU3L17t65LrrckEonC399f8e233yqX5efnK3x9fRWHDh2qtP7ly5cVnp6eir///lu57OTJkwovLy/FgwcPNFKzvqhp77/77jtFu3btFKWlpcplGRkZCk9PT8WZM2c0UrO+qGnvK8hkMsWgQYOUP9fT0tI0UW6t4pEhqrFLly4BAIKDg5XL3N3d4eTkhISEhKdu+/nnn8PT0xN9+vSp0xr10fP0vXfv3li4cCEEAkGlMU6VqZeSkoKioiKEhIQol1lbW8PHx6fKXl+8eBEODg5o0aKFcln79u0hEAiUXzeqnpr2PjQ0FKtWrYJYLK40lp+fX6e16pua9r7C2rVrUVZWhnHjxmmizDphrO0CqP7JzMyEnZ1dpR8+jo6OyMjIULvdkSNHcPz4cRw8eBC//vprXZepd56n7//85Vxh/fr1cHBwgLe3d53UqQ8ePHgAAGjUqJHKcnW9zszMrLSuSCSCra3tU/+foMpq2ntXV1e4urqqLFu3bh3EYjGCgoLqrlA9VNPeA09Oe9i0aRN27dqFzMzMOq+xrjAMUSX37t1D9+7d1Y5PnjwZIpGo0nKxWKx2jj4zMxOfffYZvvrqK9jZ2dVarfqkLvr+b19++SWOHz+O2NhYmJiYPHet+q6kpAQAKvVbLBZXebShpKTkhb829ERNe/9vW7duxbfffotPPvkE9vb2dVKjvqpp74uLizFt2jRMmzYNbm5uDEOkX5ycnBAfH692/Pjx41WeTCeRSGBmZlZpuUKhwMyZM9GzZ0907ty5VmvVJ7Xd938qKyvDnDlzsHfvXsydOxevvvrqC9erz0xNTQEAUqlU+d+A+l6bmpqq/dqYm5vXXaF6qKa9r6BQKLB8+XKsWbMG48aNw4gRI+q6VL1T095HR0fDzc0NAwcO1FiNdYVhiCoxMTGpcnqlwo0bN5CXlwepVKryF0RWVhacnZ0rrZ+eno4zZ87g8uXL2LdvHwCgvLwcAODv749x48Zh/Pjxtfsm6qHa7nuFwsJCTJw4ERcvXsTixYvRq1evWq1bH1VME2RlZaFp06bK5VlZWVVOLzo7O+Po0aMqy6RSKfLy8uDk5FS3xeqZmvYeeBL2P/nkExw6dAjTp0/H6NGjNVKrvqlp73fv3g2RSAR/f38AgEwmAwC8+eab6N27N+bNm6eBqmsHwxDVWEBAAORyOS5duoTQ0FAAwK1bt5CZmYnAwMBK6zs5OeHw4cMqyw4fPoyYmBjs27cPNjY2Gqm7vqtp34Env5DHjRuHlJQUbNiwQeXESFLP29sblpaWOH/+vPKXQkFBAZKSkjBkyJBK6wcFBSEmJgZ37txBs2bNAADnz58HALRr105zheuBmvYeAKZPn44jR44w7L+gmvb+3z/Xr127hoiICHz99ddP/cNOFzEMUY05OTmhV69emD17Nr744guYmZlh7ty5aN++Pfz8/AA8+SWcn58PGxsbiEQi5S+IChVz+f9eTuo9T9/XrVuHS5cuYfHixWjRogWys7OV+6tYhyoTiUQYMmQIYmJi0KBBA7i4uGDRokVwdnZGjx49IJPJ8OjRI1hZWcHU1BRt27ZFu3btMGXKFERGRqK4uBhz585F3759eWSohmra+z179iA+Ph7Tp09H+/btVb7HK9ah6qlp7//987viBOzGjRvXv/O1tH1tP9VPRUVFilmzZikCAwMVgYGBio8//ljx6NEj5fi5c+cUnp6einPnzlW5/e7du3mfoedQ076/+uqrCk9Pzyo/1H1t6Iny8nLFV199pQgJCVH4+fkpxo4dq7x/SlpamsLT01Oxe/du5foPHz5UTJo0SeHn56cIDg5WzJ07V+XeN1R9Nen9yJEj1X6P//PrQ9VT0+/7f6r4+VMf7zMkUCgUCm0HMiIiIiJt4U0XiYiIyKAxDBEREZFBYxgiIiIig8YwRERERAaNYYiIiIgMGsMQERERGTSGISIiIjJoDENERERk0Pg4DiLSC0OHDsWFCxdUlpmYmKBhw4bo2rUrPvroI5Xn4N2+fRtbtmzBqVOnkJWVhQYNGigfHPy0B4K+99576Ny5MyZNmlSn74eINIdhiIj0ho+PD+bOnav8vKysDH/88QeWLFmC5ORk7NixAwKBAEeOHEFERARatmyJCRMmwNXVFQ8ePMC2bdvwzjvvYNWqVejcubPKvktLSxEREYHExMRKY0RUvzEMEZHesLS0VD60tkJQUBCKiooQGxuLa9euoUGDBpg+fTo6deqEZcuWQSgUKtd97bXX8N5772HmzJn45ZdflA/5vHjxIqKiopCVlaXJt0NEGsJzhohI77Vu3RoAkJ6ejm3btkEqlWL27NkqQQgATE1NMWPGDAwYMAAFBQXK5RMmTICLiwv27Nmj0bqJSDN4ZIiI9N7t27cBAE2aNEFsbCx8fHzg5ORU5brBwcEIDg5WWbZ9+3Z4eXnVeZ1EpB0MQ0SkNxQKBcrLy5Wf5+fn48KFC1izZg38/PzQunVrZGZm4qWXXqrRfhmEiPQbwxAR6Y2EhAS0atVKZZmRkRFCQ0Mxf/58CAQCCAQCyGQyLVVIRLqIYYiI9EarVq0QFRUFABAIBBCLxWjUqBEsLS2V67i4uCA9PV3tPsrLy/Ho0SM4OjrWeb1EpBt4AjUR6Q0LCwu0adMGbdq0QevWrdGyZUuVIAQAYWFhSEpKQnZ2dpX7OHnyJDp16oQffvhBEyUTkQ5gGCIigzJ48GCYmJggOjq60nRZSUkJYmNjYWNjg65du2qpQiLSNE6TEZFBcXV1RWRkJGbNmoXBgwdj4MCBaNSoEe7evYvNmzfjzp07WL9+PczNzbVdKhFpCMMQERmcfv36oVmzZtiyZQuWLVuGnJwcODg4wN/fH8uXL4eHh4e2SyQiDRIoFAqFtosgIiIi0haeM0REREQGjWGIiIiIDBrDEBERERk0hiEiIiIyaAxDREREZNAYhoiIiMigMQwRERGRQWMYIiIiIoPGMEREREQGjWGIiIiIDBrDEBERERm0/wPAhsuEjySrrgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Dietary variables\n",
    "diet = ['animal_proteins_red',\n",
    "       'refined_grains', 'solanaceae', 'saturated', 'pastries',\n",
    "       'milk_derivates', 'beverages_prohibited', 'pro_inflammatory',\n",
    "       'chicken', 'plant_protein', 'whole_grains_pseudocereals_oats',\n",
    "       'gluten_free', 'vegetables', 'cruciferous',\n",
    "       'berries_and_enzymatic_fruit', 'other_fruit', 'fish', 'milk_subs',\n",
    "       'PUFA', 'MUFA', 'Walnuts', 'Nuts', 'probiotics', 'Tahini_avo',\n",
    "       'green_tea', 'anti_inflam_spices', 'inflammatory_spices',\n",
    "       'inflammatory', 'anti_inflammatory', 'total_antiinflammatory',\n",
    "       'Adherece_antiinflam', 'oats', 'almond_milk', 'fatty_fish',\n",
    "       'sesame_seeds', 'egg', 'chia', 'flaxseed_oil', 'walnuts',\n",
    "       'spinach', 'corn_tortilla', 'tahini', 'avocado', 'chicken.1',\n",
    "       'brown_rice', 'corn', 'quinoa', 'broccoli', 'Garbanzo_beans',\n",
    "       'pork', 'beef', 'burger', 'Total_omega3', 'Adherence_omega3',\n",
    "       'Total_omega6', 'Adherence_omega6', 'Total_o3_o6']\n",
    "\n",
    "for d in diet:\n",
    "    if 'Adh' in d:\n",
    "        diet.remove(d)\n",
    "\n",
    "g = 'saliva'\n",
    "df_map = type_to_df_map[g].reset_index().drop_duplicates(subset='#SampleID').set_index('#SampleID')\n",
    "df_diet = df_map[diet].dropna()\n",
    "\n",
    "from sklearn.decomposition import PCA\n",
    "npcs = 3\n",
    "pca = PCA(n_components=3, whiten=True)\n",
    "pca.fit(df_diet.T)\n",
    "# Reformat and view results\n",
    "loadings = pd.DataFrame(pca.components_,\n",
    "                        index=['PC%s' % _ for _ in range(npcs)], \n",
    "                        columns=df_diet.index).T\n",
    "\n",
    "df_merge = pd.concat([df_map, df_diet, loadings], axis=1)\n",
    "plt.title(g + '_Diet_PCA')\n",
    "sns.scatterplot(data=df_merge, x='PC1', y='PC2', hue='WOMAC_P_Response')\n",
    "sns.despine()\n",
    "plt.savefig(path + 'outputs/jobs06/' + c + '_' + dt + '_barplot_' + g + '_' + r + '_' + x.split('g__')[-1] + '.pdf')\n",
    "plt.close() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "d98a47aa-ba40-47eb-8258-d87d001e000c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.22984773285789006\n"
     ]
    }
   ],
   "source": [
    "# Procrustes\n",
    "import scipy.spatial\n",
    "\n",
    "# \n",
    "# df1 = gt_to_meta['saliva']['After diet']\n",
    "df1 = gt_to_meta['stool']['Baseline']\n",
    "df2 = gt_to_meta['saliva']['Baseline']\n",
    "\n",
    "def convert_to_pcs(df, npcs):\n",
    "    pca = PCA(n_components=npcs, whiten=True)\n",
    "    pca.fit(df.T)\n",
    "    return pd.DataFrame(pca.components_, \n",
    "                        index=['PC%s' % _ for _ in range(npcs)], \n",
    "                        columns=df.index).T\n",
    "\n",
    "df1 = convert_to_pcs(df1, 10)\n",
    "df2 = convert_to_pcs(df2, 10)\n",
    "df1p, df2p, m2 = scipy.spatial.procrustes(df1.values,df2.values)\n",
    "print(m2)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "a2ca7774-e244-4b35-b3ec-1fcc32cfbfb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total_effect_p\n",
      "0.051199999999999996 0.03326499661806687\n",
      "direct_effect_p\n",
      "0.6058 0.09083149233608354\n",
      "indirect_effect_p\n",
      "0.10980000000000001 0.04004946940971876\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "indirect_effect_p\n",
       "0.10    23\n",
       "0.08    17\n",
       "0.12    14\n",
       "0.14    13\n",
       "0.06    11\n",
       "0.16    10\n",
       "0.04     5\n",
       "0.18     4\n",
       "0.20     2\n",
       "0.24     1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mediation stability\n",
    "df = pd.read_csv(path + 'outputs/jobs20/df100.csv')\n",
    "for y in ['total_effect_p','direct_effect_p','indirect_effect_p']:\n",
    "    x = df[y]\n",
    "    print(y)\n",
    "    print(np.mean(x), np.std(x))\n",
    "\n",
    "x.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e22033b-a7a9-406a-8cb1-c1074ae78087",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Complete DA Analysis ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "856e9286-7a26-4639-a599-53c699043e5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taxa\n",
      "pvp\n",
      "saliva\n",
      "Number of sig var\n",
      "2\n",
      "Number of sig var\n",
      "5\n",
      "stool\n",
      "Number of sig var\n",
      "6\n",
      "Number of sig var\n",
      "4\n",
      "rvnr\n",
      "saliva\n",
      "Number of sig var\n",
      "9\n",
      "Number of sig var\n",
      "0\n",
      "stool\n",
      "Number of sig var\n",
      "6\n",
      "Number of sig var\n",
      "0\n",
      "path\n",
      "pvp\n",
      "saliva\n",
      "Number of sig var\n",
      "3\n",
      "Number of sig var\n",
      "5\n",
      "stool\n",
      "Number of sig var\n",
      "4\n",
      "Number of sig var\n",
      "6\n",
      "rvnr\n",
      "saliva\n",
      "Number of sig var\n",
      "22\n",
      "Number of sig var\n",
      "0\n",
      "stool\n",
      "Number of sig var\n",
      "11\n",
      "Number of sig var\n",
      "0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis</th>\n",
       "      <th>Metabolism|Carbohydrate metabolism|Pyruvate metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism</th>\n",
       "      <th>Metabolism|Lipid metabolism|Fatty acid degradation</th>\n",
       "      <th>Metabolism|Amino acid metabolism</th>\n",
       "      <th>Metabolism|Amino acid metabolism|Tyrosine metabolism</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins</th>\n",
       "      <th>Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism</th>\n",
       "      <th>...</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis</th>\n",
       "      <th>Metabolism|Lipid metabolism|Steroid hormone biosynthesis</th>\n",
       "      <th>Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures</th>\n",
       "      <th>Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation</th>\n",
       "      <th>Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis</th>\n",
       "      <th>Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids</th>\n",
       "      <th>Timepoint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OAD-001.pre.stool</th>\n",
       "      <td>0.159285</td>\n",
       "      <td>0.035029</td>\n",
       "      <td>0.002320</td>\n",
       "      <td>0.003018</td>\n",
       "      <td>0.007937</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>0.027626</td>\n",
       "      <td>0.000229</td>\n",
       "      <td>0.026349</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>3.060096e-06</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.510913e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-003.pre.stool</th>\n",
       "      <td>0.161049</td>\n",
       "      <td>0.033134</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.002538</td>\n",
       "      <td>0.008606</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.029092</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.027987</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>9.470469e-06</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.511851e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OAD-004.pre.stool</th>\n",
       "      <td>0.161760</td>\n",
       "      <td>0.034016</td>\n",
       "      <td>0.002254</td>\n",
       "      <td>0.002773</td>\n",
       "      <td>0.008658</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.029185</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>0.027522</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>2.489384e-06</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.100744e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-001.pre.stool</th>\n",
       "      <td>0.163414</td>\n",
       "      <td>0.036621</td>\n",
       "      <td>0.002214</td>\n",
       "      <td>0.002779</td>\n",
       "      <td>0.008978</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>0.028735</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>0.026849</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>8.806406e-06</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.026500e-06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.425694e-07</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAOAD-012.pre.stool</th>\n",
       "      <td>0.164662</td>\n",
       "      <td>0.036528</td>\n",
       "      <td>0.002363</td>\n",
       "      <td>0.002908</td>\n",
       "      <td>0.009032</td>\n",
       "      <td>0.000170</td>\n",
       "      <td>0.028994</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>0.027414</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>1.871888e-07</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.499616e-07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>pre</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 160 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Metabolism  Metabolism|Carbohydrate metabolism  \\\n",
       "OAD-001.pre.stool      0.159285                            0.035029   \n",
       "OAD-003.pre.stool      0.161049                            0.033134   \n",
       "OAD-004.pre.stool      0.161760                            0.034016   \n",
       "VAOAD-001.pre.stool    0.163414                            0.036621   \n",
       "VAOAD-012.pre.stool    0.164662                            0.036528   \n",
       "\n",
       "                     Metabolism|Carbohydrate metabolism|Glycolysis / Gluconeogenesis  \\\n",
       "OAD-001.pre.stool                                             0.002320                 \n",
       "OAD-003.pre.stool                                             0.002184                 \n",
       "OAD-004.pre.stool                                             0.002254                 \n",
       "VAOAD-001.pre.stool                                           0.002214                 \n",
       "VAOAD-012.pre.stool                                           0.002363                 \n",
       "\n",
       "                     Metabolism|Carbohydrate metabolism|Pyruvate metabolism  \\\n",
       "OAD-001.pre.stool                                             0.003018        \n",
       "OAD-003.pre.stool                                             0.002538        \n",
       "OAD-004.pre.stool                                             0.002773        \n",
       "VAOAD-001.pre.stool                                           0.002779        \n",
       "VAOAD-012.pre.stool                                           0.002908        \n",
       "\n",
       "                     Metabolism|Lipid metabolism  \\\n",
       "OAD-001.pre.stool                       0.007937   \n",
       "OAD-003.pre.stool                       0.008606   \n",
       "OAD-004.pre.stool                       0.008658   \n",
       "VAOAD-001.pre.stool                     0.008978   \n",
       "VAOAD-012.pre.stool                     0.009032   \n",
       "\n",
       "                     Metabolism|Lipid metabolism|Fatty acid degradation  \\\n",
       "OAD-001.pre.stool                                             0.000192    \n",
       "OAD-003.pre.stool                                             0.000180    \n",
       "OAD-004.pre.stool                                             0.000157    \n",
       "VAOAD-001.pre.stool                                           0.000195    \n",
       "VAOAD-012.pre.stool                                           0.000170    \n",
       "\n",
       "                     Metabolism|Amino acid metabolism  \\\n",
       "OAD-001.pre.stool                            0.027626   \n",
       "OAD-003.pre.stool                            0.029092   \n",
       "OAD-004.pre.stool                            0.029185   \n",
       "VAOAD-001.pre.stool                          0.028735   \n",
       "VAOAD-012.pre.stool                          0.028994   \n",
       "\n",
       "                     Metabolism|Amino acid metabolism|Tyrosine metabolism  \\\n",
       "OAD-001.pre.stool                                             0.000229      \n",
       "OAD-003.pre.stool                                             0.000194      \n",
       "OAD-004.pre.stool                                             0.000173      \n",
       "VAOAD-001.pre.stool                                           0.000203      \n",
       "VAOAD-012.pre.stool                                           0.000194      \n",
       "\n",
       "                     Metabolism|Metabolism of cofactors and vitamins  \\\n",
       "OAD-001.pre.stool                                           0.026349   \n",
       "OAD-003.pre.stool                                           0.027987   \n",
       "OAD-004.pre.stool                                           0.027522   \n",
       "VAOAD-001.pre.stool                                         0.026849   \n",
       "VAOAD-012.pre.stool                                         0.027414   \n",
       "\n",
       "                     Metabolism|Metabolism of cofactors and vitamins|Retinol metabolism  \\\n",
       "OAD-001.pre.stool                                             0.000012                    \n",
       "OAD-003.pre.stool                                             0.000019                    \n",
       "OAD-004.pre.stool                                             0.000016                    \n",
       "VAOAD-001.pre.stool                                           0.000027                    \n",
       "VAOAD-012.pre.stool                                           0.000013                    \n",
       "\n",
       "                     ...  \\\n",
       "OAD-001.pre.stool    ...   \n",
       "OAD-003.pre.stool    ...   \n",
       "OAD-004.pre.stool    ...   \n",
       "VAOAD-001.pre.stool  ...   \n",
       "VAOAD-012.pre.stool  ...   \n",
       "\n",
       "                     Metabolism|Glycan biosynthesis and metabolism|Various types of N-glycan biosynthesis  \\\n",
       "OAD-001.pre.stool                                             0.000028                                      \n",
       "OAD-003.pre.stool                                             0.000090                                      \n",
       "OAD-004.pre.stool                                             0.000103                                      \n",
       "VAOAD-001.pre.stool                                           0.000102                                      \n",
       "VAOAD-012.pre.stool                                           0.000106                                      \n",
       "\n",
       "                     Metabolism|Glycan biosynthesis and metabolism|Lipoarabinomannan (LAM) biosynthesis  \\\n",
       "OAD-001.pre.stool                                         3.060096e-06                                    \n",
       "OAD-003.pre.stool                                         9.470469e-06                                    \n",
       "OAD-004.pre.stool                                         2.489384e-06                                    \n",
       "VAOAD-001.pre.stool                                       8.806406e-06                                    \n",
       "VAOAD-012.pre.stool                                       1.871888e-07                                    \n",
       "\n",
       "                     Metabolism|Lipid metabolism|Steroid hormone biosynthesis  \\\n",
       "OAD-001.pre.stool                                             0.000026          \n",
       "OAD-003.pre.stool                                             0.000118          \n",
       "OAD-004.pre.stool                                             0.000108          \n",
       "VAOAD-001.pre.stool                                           0.000105          \n",
       "VAOAD-012.pre.stool                                           0.000071          \n",
       "\n",
       "                     Metabolism|Glycan biosynthesis and metabolism|Glycosphingolipid biosynthesis - ganglio series  \\\n",
       "OAD-001.pre.stool                                             0.000026                                               \n",
       "OAD-003.pre.stool                                             0.000090                                               \n",
       "OAD-004.pre.stool                                             0.000103                                               \n",
       "VAOAD-001.pre.stool                                           0.000102                                               \n",
       "VAOAD-012.pre.stool                                           0.000106                                               \n",
       "\n",
       "                     Metabolism|Biosynthesis of other secondary metabolites|Staurosporine biosynthesis  \\\n",
       "OAD-001.pre.stool                                                  0.0                                   \n",
       "OAD-003.pre.stool                                                  0.0                                   \n",
       "OAD-004.pre.stool                                                  0.0                                   \n",
       "VAOAD-001.pre.stool                                                0.0                                   \n",
       "VAOAD-012.pre.stool                                                0.0                                   \n",
       "\n",
       "                     Metabolism|Metabolism of terpenoids and polyketides|Type I polyketide structures  \\\n",
       "OAD-001.pre.stool                                                  0.0                                  \n",
       "OAD-003.pre.stool                                                  0.0                                  \n",
       "OAD-004.pre.stool                                                  0.0                                  \n",
       "VAOAD-001.pre.stool                                                0.0                                  \n",
       "VAOAD-012.pre.stool                                                0.0                                  \n",
       "\n",
       "                     Metabolism|Xenobiotics biodegradation and metabolism|Furfural degradation  \\\n",
       "OAD-001.pre.stool                                         1.510913e-06                           \n",
       "OAD-003.pre.stool                                         8.511851e-07                           \n",
       "OAD-004.pre.stool                                         6.100744e-07                           \n",
       "VAOAD-001.pre.stool                                       1.026500e-06                           \n",
       "VAOAD-012.pre.stool                                       3.499616e-07                           \n",
       "\n",
       "                     Metabolism|Metabolism of terpenoids and polyketides|Tetracycline biosynthesis  \\\n",
       "OAD-001.pre.stool                                                  0.0                               \n",
       "OAD-003.pre.stool                                                  0.0                               \n",
       "OAD-004.pre.stool                                                  0.0                               \n",
       "VAOAD-001.pre.stool                                                0.0                               \n",
       "VAOAD-012.pre.stool                                                0.0                               \n",
       "\n",
       "                     Metabolism|Biosynthesis of other secondary metabolites|Biosynthesis of various alkaloids  \\\n",
       "OAD-001.pre.stool                                         0.000000e+00                                          \n",
       "OAD-003.pre.stool                                         0.000000e+00                                          \n",
       "OAD-004.pre.stool                                         0.000000e+00                                          \n",
       "VAOAD-001.pre.stool                                       1.425694e-07                                          \n",
       "VAOAD-012.pre.stool                                       0.000000e+00                                          \n",
       "\n",
       "                     Timepoint  \n",
       "OAD-001.pre.stool          pre  \n",
       "OAD-003.pre.stool          pre  \n",
       "OAD-004.pre.stool          pre  \n",
       "VAOAD-001.pre.stool        pre  \n",
       "VAOAD-012.pre.stool        pre  \n",
       "\n",
       "[5 rows x 160 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# complete block for paired testing in r v nr and non paired testing pre and post\n",
    "\n",
    "dcgr_to_sig = {}\n",
    "\n",
    "for dt in ['taxa', 'path']:\n",
    "    print(dt)\n",
    "    dcgr_to_sig[dt] = {}\n",
    "    for c in ['pvp', 'rvnr']: # pvp is pre AND post, comparing rvnr at pre and post time points\n",
    "        print(c)\n",
    "        dcgr_to_sig[dt][c] = {}        \n",
    "    \n",
    "        # define df arrays\n",
    "        xs = []\n",
    "        gs = []\n",
    "        ws = []\n",
    "        ps = []\n",
    "        os = []\n",
    "        dirs = []\n",
    "        adj_rs = []\n",
    "        \n",
    "        # define analysis specific params\n",
    "        if c == 'pvp':\n",
    "            outer_loop = ['pre', 'post']\n",
    "            comp = ['Response', 'No response']\n",
    "            var = 'WOMAC_P_Response'\n",
    "            drop = 'Timepoint'\n",
    "        elif c == 'rvnr': # pvp is pre and post, comparing rvnr at pre and posttime points\n",
    "            outer_loop = ['Response', 'No response']\n",
    "            comp = ['pre', 'post']\n",
    "            var = 'Timepoint'\n",
    "            drop = 'WOMAC_P_Response'\n",
    "\n",
    "        # for each sample type\n",
    "        for g in ['saliva','stool']:\n",
    "            print(g)\n",
    "            dcgr_to_sig[dt][c][g] = {}\n",
    "            \n",
    "            # grab relevant table table\n",
    "            if dt == 'taxa':\n",
    "                df_abun = gt_to_otu[g]['all']\n",
    "            elif dt == 'path':\n",
    "                df_abun = gt_to_path[g]['all']\n",
    "\n",
    "            df_map = type_to_df_map[g].copy()\n",
    "            df_map = df_map.set_index('Together')\n",
    "            if g == 'stool':\n",
    "                df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "            df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "\n",
    "            for r in outer_loop:\n",
    "                df_merge = pd.concat([df_abun, df_map[df_map[drop] == r]], axis=1)\n",
    "                df_merge = df_merge.dropna()\n",
    "                df_merge['SampleID'] = df_merge.index.map(lambda x: x.split('.p')[0])\n",
    "                if c == 'rvnr':\n",
    "                    df_vc = df_merge['SampleID'].value_counts()\n",
    "                    keep = df_vc[df_vc == 2].index.values\n",
    "                    df_merge = df_merge[df_merge['SampleID'].isin(keep)] \n",
    "                df_merge = df_merge.drop(['SampleID',drop],axis=1)\n",
    "                sig = []\n",
    "                for x in list(df_merge.columns.values)[:-1]:\n",
    "                    df_a = df_merge[df_merge[var] == comp[0]][x]\n",
    "                    df_b = df_merge[df_merge[var] == comp[1]][x]\n",
    "\n",
    "                    try:\n",
    "                        if c == 'rvnr':\n",
    "                            w,p = scipy.stats.wilcoxon(df_a.values, df_b.values, nan_policy='omit')\n",
    "                        elif c == 'pvp':\n",
    "                            w,p = scipy.stats.mannwhitneyu(df_a.values, df_b.values, nan_policy='omit')\n",
    "\n",
    "                        a = df_a.values\n",
    "                        b = df_b.values\n",
    "                        # if median in A is greater\n",
    "                        abr, abp = scipy.stats.mannwhitneyu(a,b,nan_policy='omit')\n",
    "                        # if median in B is greater\n",
    "                        bar, bap = scipy.stats.mannwhitneyu(b,a,nan_policy='omit')\n",
    "                        if abr > bar:\n",
    "                            dirs.append(comp[0])\n",
    "                            adj_rs.append(abr)\n",
    "                        elif abr < bar:\n",
    "                            dirs.append(comp[1])\n",
    "                            adj_rs.append(bar * -1)\n",
    "                        else:\n",
    "                            dirs.append('same')\n",
    "                            adj_rs.append(abr)\n",
    "\n",
    "                    except:\n",
    "                        w,p = 0, 1\n",
    "                        dirs.append(np.nan)\n",
    "                        adj_rs.append(np.nan)\n",
    "\n",
    "                    os.append(r)\n",
    "                    xs.append(x)\n",
    "                    gs.append(g)\n",
    "                    ws.append(w)\n",
    "                    ps.append(p)\n",
    "\n",
    "                    if p < 0.05:\n",
    "                        ax = sns.boxplot(data=df_merge, y=x, x=var)\n",
    "                        sns.swarmplot(data=df_merge, y=x, x=var, palette='dark:grey')\n",
    "                        sns.despine()\n",
    "                    \n",
    "                        plt.tight_layout()\n",
    "                        plt.savefig(path + 'outputs/jobs06/' + c + '_' + dt + '_barplot_' + g + '_' + r + '_' + x.split('g__')[-1] + '.pdf')\n",
    "                        plt.close() \n",
    "                        sig.append(x)\n",
    "                print('Number of sig var')\n",
    "                print(len(sig))\n",
    "                dcgr_to_sig[dt][c][g][r] = sig\n",
    "                \n",
    "        df_res = pd.DataFrame({'SampleType':gs, # saliva v stool\n",
    "                      'fixed': os,\n",
    "                      'var': xs, # example pathway or taxa\n",
    "                      'direction': dirs, # direction of enrichment\n",
    "                      'Statistic': ws, \n",
    "                      'pvalue': ps,\n",
    "                      'adjusted': adj_rs                      \n",
    "                     })\n",
    "        df_res[df_res['pvalue'] < 0.05].to_csv(path + 'outputs/jobs06/' + dt + '_' + c + '.tsv', sep='\\t')\n",
    "df_merge.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "43bca09e-c560-43cf-975b-7b58541ae974",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stool\n",
      "meta\n",
      "otu\n",
      "no plot for :stool_otu_['No response']\n",
      "path\n",
      "no plot for :stool_path_['No response']\n",
      "saliva\n",
      "meta\n",
      "no plot for :saliva_meta_['No response']\n",
      "otu\n",
      "no plot for :saliva_otu_['No response']\n",
      "path\n",
      "no plot for :saliva_path_['No response']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>path</th>\n",
       "      <th>test_stat</th>\n",
       "      <th>adj_test_stat</th>\n",
       "      <th>pvals</th>\n",
       "      <th>direction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [path, test_stat, adj_test_stat, pvals, direction]\n",
       "Index: []"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# horizontal barplot for features\n",
    "\n",
    "d_to_df = {\n",
    "    'meta': gt_to_meta,\n",
    "    'otu': gt_to_otu,\n",
    "    'path': gt_to_path\n",
    "}\n",
    "d_to_label = {\n",
    "    'meta': 'Metabolite',\n",
    "    'otu': 'Taxa',\n",
    "    'path': 'Pathway'\n",
    "}\n",
    "t = 'all'\n",
    "\n",
    "gd_to_res = {}\n",
    "for g in ['stool','saliva']:\n",
    "    print(g)\n",
    "    gd_to_res[g] = {}\n",
    "    for d in ['meta','otu','path']:\n",
    "        print(d)\n",
    "        # grab original dfs\n",
    "        df_feat = d_to_df[d][g][t]\n",
    "\n",
    "        df_map = type_to_df_map[g].set_index('Together')\n",
    "        df_map.index = df_map.index.map(lambda x: x.split('.guma')[0])\n",
    "        \n",
    "        df_map = df_map[['Timepoint','WOMAC_P_Response','HostSubjectId']]\n",
    "        \n",
    "        # merge df\n",
    "        df_merge = pd.concat([df_feat, df_map],axis=1)\n",
    "        df_merge = df_merge.dropna()\n",
    "        \n",
    "        # do paired test\n",
    "        # grab all sample IDs that are duplicated\n",
    "        df_vc = df_merge['HostSubjectId'].value_counts() == 2\n",
    "        df_vc = df_vc[df_vc == True]\n",
    "        keep_id = df_vc.index.values\n",
    "        \n",
    "        df_merge_complete = df_merge[df_merge['HostSubjectId'].isin(keep_id)]\n",
    "\n",
    "        for keep in [['Response','No response'],['Response'],['No response']]:\n",
    "            df_merge = df_merge_complete[df_merge_complete['WOMAC_P_Response'].isin(keep)]\n",
    "        \n",
    "            # samples are in order when subsetted on Timepoint\n",
    "            df_pre = df_merge[df_merge['Timepoint'] == 'pre'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "            df_post = df_merge[df_merge['Timepoint'] == 'post'].drop(['Timepoint','WOMAC_P_Response','HostSubjectId'], axis=1)\n",
    "            \n",
    "            # compute rank sum\n",
    "            feat = []\n",
    "            rs = []\n",
    "            adj_rs = []\n",
    "            ps = []\n",
    "            dir = []\n",
    "            for x in df_pre.columns.values:\n",
    "                try:\n",
    "                    r, p = scipy.stats.wilcoxon(df_post[x].values-df_pre[x].values,nan_policy='omit')\n",
    "                except ValueError:\n",
    "                    r, p = 0, 1\n",
    "                if d == 'otu':\n",
    "                    feat.append(x.split('__')[-1])\n",
    "                elif d == 'path':\n",
    "                    feat.append(x.split('|')[-1])\n",
    "                else:\n",
    "                    feat.append(x)\n",
    "                # rs.append(r)\n",
    "                ps.append(p)\n",
    "                # determine direction via MWU\n",
    "                a = df_post[x].values\n",
    "                b = df_pre[x].values\n",
    "                abr, abp = scipy.stats.mannwhitneyu(a,b,nan_policy='omit')\n",
    "                bar, bap = scipy.stats.mannwhitneyu(b,a,nan_policy='omit')\n",
    "                if abr > bar:\n",
    "                    dir.append('post')\n",
    "                    adj_rs.append(abr)\n",
    "                    rs.append(bar)\n",
    "                elif abr < bar:\n",
    "                    dir.append('pre')\n",
    "                    adj_rs.append(bar * -1)\n",
    "                    rs.append(bar)\n",
    "                else:\n",
    "                    dir.append('same')\n",
    "                    adj_rs.append(abr)\n",
    "                    rs.append(abr)\n",
    "            # compile results df\n",
    "            df_results = pd.DataFrame({d: feat, 'test_stat': rs, 'adj_test_stat': adj_rs,'pvals': ps, 'direction': dir})\n",
    "            gd_to_res[g][d] = df_results\n",
    "            df = df_results[df_results['pvals'] < 0.05]\n",
    "            df = df.sort_values(by=['adj_test_stat'],ascending=False)\n",
    "            # plot\n",
    "            # create plot\n",
    "            try:\n",
    "                plt.figure(figsize=(8,6))\n",
    "                sns.set_theme(style=\"whitegrid\")\n",
    "                \n",
    "                sns.set_color_codes(\"pastel\")\n",
    "                sns.barplot(x='adj_test_stat', y=d, data=df,\n",
    "                            hue=\"direction\",orient='horizontal')#,native_scale=True)#width=1)\n",
    "                \n",
    "                ax.set_ylabel(d_to_label[d],fontsize=16)\n",
    "                ax.set_xlabel(\"Effect Size / Adjusted Test Statistic\",fontsize=16)\n",
    "                ax.tick_params(labelsize=16)\n",
    "                sns.despine(left=True, bottom=True)\n",
    "                plt.tight_layout()\n",
    "                plt.title(g + '_' + d + '_' + str(keep))\n",
    "                plt.savefig(path + 'outputs/jobs18/DA_hbarplot_' + g + '_' + d + '_' + str(keep) + '.pdf')\n",
    "                plt.close()\n",
    "                df.to_csv(path + 'outputs/jobs18/' + g + '_' + d + '_' + str(keep) + '.tsv', sep='\\t')\n",
    "            \n",
    "            except: \n",
    "                print('no plot for :' + g + '_' + d + '_' + str(keep))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70c6b863-4344-44ac-9ac9-949de2808948",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Mediation ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "90768c43-3b95-494b-93d3-46d7b1a4222e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X</th>\n",
       "      <th>M</th>\n",
       "      <th>Y</th>\n",
       "      <th>split_value</th>\n",
       "      <th>total_effect</th>\n",
       "      <th>total_effect_p</th>\n",
       "      <th>total_effect_ci_lower</th>\n",
       "      <th>total_effect_ci_upper</th>\n",
       "      <th>direct_effect</th>\n",
       "      <th>direct_effect_p</th>\n",
       "      <th>...</th>\n",
       "      <th>b2_effect_ci_lower</th>\n",
       "      <th>b2_effect_ci_upper</th>\n",
       "      <th>b3_effect</th>\n",
       "      <th>b3_effect_p</th>\n",
       "      <th>b3_effect_ci_lower</th>\n",
       "      <th>b3_effect_ci_upper</th>\n",
       "      <th>X_sparsity</th>\n",
       "      <th>M_sparsity</th>\n",
       "      <th>Y_sparsity</th>\n",
       "      <th>n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [X, M, Y, split_value, total_effect, total_effect_p, total_effect_ci_lower, total_effect_ci_upper, direct_effect, direct_effect_p, direct_effect_ci_lower, direct_effect_ci_upper, indirect_effect, indirect_effect_p, indirect_effect_ci_lower, indirect_effect_ci_upper, prop_mediated, prop_mediated_p, prop_mediated_ci_lower, prop_mediated_ci_upper, b2_effect, b2_effect_p, b2_effect_ci_lower, b2_effect_ci_upper, b3_effect, b3_effect_p, b3_effect_ci_lower, b3_effect_ci_upper, X_sparsity, M_sparsity, Y_sparsity, n]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 32 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Mediation analyses\n",
    "# check for int_ba\n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid'] + ['Deoxycholic_acid','Lithocholic_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid', 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide']+ ['Deoxycholic_acid','Lithocholic_acid']\n",
    "         }\n",
    "\n",
    "# filter on CUTIE results\n",
    "g_to_cutiejob = {\n",
    "    'saliva': 'jobs15',\n",
    "    'stool': 'jobs16'\n",
    "}\n",
    "\n",
    "g_to_medjob = {\n",
    "    'saliva': '20240917',\n",
    "    'stool': '20240913'\n",
    "}\n",
    "# figure out cutie plots\n",
    "#X = list(df_X.columns.values)\n",
    "#M = list(df_M.columns.values)\n",
    "#x,m = df.iloc[0,0:2].values\n",
    "#X.index(x), M.index(m)\n",
    "\n",
    "# define outcome\n",
    "y = 'WOMAC_pain_diff'\n",
    "\n",
    "\n",
    "\n",
    "for g in ['stool', 'saliva']:\n",
    "    # import mediation DFs\n",
    "    df_M = pd.read_csv(path + 'inputs/cutie_df_meta_' + g + '.tsv', sep='\\t', index_col=0)\n",
    "    df_M.iloc[:,:-1].to_csv(path + 'inputs/mediation_meta_' + g + '_post_test.txt', sep='\\t')\n",
    "    \n",
    "    df_Y = df_meta_paired[['VAS_Pt_diff', 'WOMAC_pain_diff']]\n",
    "    df_Y.to_csv(path + 'inputs/mediation_WMdiff_test.txt', sep='\\t')\n",
    "    \n",
    "    df_X = pd.read_csv(path + 'inputs/cutie_df_taxa_' + g + '.tsv', sep='\\t', index_col=0)\n",
    "    df_X.index = df_X.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    df_X.to_csv(path + 'inputs/mediation_taxa_' + g + '_post_test.txt', sep='\\t')\n",
    "\n",
    "    # import mediation results\n",
    "    df_med = pd.read_csv(path + 'outputs/jobs19/' + g_to_medjob[g] + '/outputs/mediation_taxa_' + g + '_post_mediation_meta_' + g + '_post_mediation_WMdiff_test/mediation_results_all.tsv', \n",
    "                         sep='\\t')\n",
    "\n",
    "    for t in ['partial','complete']:\n",
    "        # copy df\n",
    "        df = df_med.copy()\n",
    "        \n",
    "        # filter on significant total\n",
    "        df = df[df['total_effect_p'] < 0.05]\n",
    "\n",
    "        if t == 'complete':\n",
    "            # find complete mediations\n",
    "            df = df[df['direct_effect_p'] > 0.05]\n",
    "\n",
    "        elif t == 'partial':\n",
    "            # find partial mediations\n",
    "            df = df[df['direct_effect_p'] < 0.05]\n",
    "        \n",
    "        # check how many indirect; this reduces it by alot again\n",
    "        df = df[df['indirect_effect_p'] < 0.05]\n",
    "\n",
    "        # create column for metabolites\n",
    "        df = df[df['M'].isin(['H_' + x for x in int_ba[g]])]\n",
    "        \n",
    "        # grab CUTIE results\n",
    "        df_qt = pd.read_csv(path + 'outputs/' + g_to_cutiejob[g] + '/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "        df_qt = df_qt[df_qt['class'] == 'TP']\n",
    "\n",
    "        for r in range(len(df)):\n",
    "            x,m,y = df.iloc[r,0:3].values\n",
    "            \n",
    "            # check if it's in CUTIE\n",
    "            df_q = df_qt.copy()\n",
    "            # df_q = df_q[df_q['var1'] == m]\n",
    "            df_q = df_q[df_q['var2'] == x]\n",
    "        \n",
    "            if len(df_q) > 0: #len(df_q2) > 0: # len(df_q1) > 0 or len(df_q2) > 0:\n",
    "                # print('plotting ' + m + ' and ' + x)\n",
    "                # create plotting df\n",
    "                df_plot = pd.concat([df_X,df_M,df_Y],axis=1)\n",
    "                # path + 'outputs/jobs19/plots/'\n",
    "        \n",
    "                plt.figure(figsize=(8,6))\n",
    "                sns.set_theme(style=\"whitegrid\")\n",
    "                # sns.scatterplot(data=df_plot,x=x,y=m,hue=y)\n",
    "                sns.scatterplot(data=df_plot,x=x,y=y,hue=m,palette='coolwarm')#,size=6)\n",
    "                plt.tight_layout()\n",
    "                sns.despine()\n",
    "                plt.savefig(path + 'outputs/jobs19/plots/' + g + '/' + x.split(';g__')[-1] + '_' + m.replace('/','fslash') + '_' + t + '_' + y + '_MX.pdf')\n",
    "                plt.close()\n",
    "        \n",
    "                plt.figure(figsize=(8,6))\n",
    "                sns.set_theme(style=\"whitegrid\")        \n",
    "                sns.scatterplot(data=df_plot,x=x,y=m,hue=m,palette='coolwarm')#,size=6)\n",
    "                sns.despine()\n",
    "                plt.tight_layout()\n",
    "                plt.savefig(path + 'outputs/jobs19/plots/' + g + '/' + x.split(';g__')[-1] + '_' + m.replace('/','fslash') + '_' + t + '_' + y + '_YMX.pdf')\n",
    "                plt.close()\n",
    "        \n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8303f089-f331-4356-b6df-487184f8f8e3",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "###### Scratch ######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f932f988-b70e-4591-8e51-a1f70690fc1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = df.copy()\n",
    "\n",
    "m = 'H_Nonhydroxylated_bile_acid'\n",
    "y = 'WOMAC_pain_diff'\n",
    "y = 'VAS_Pt_diff'\n",
    "x = 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;__'\n",
    "\n",
    "d = d[d['X'] == x]\n",
    "d = d[d['Y'] == y]\n",
    "d = d[d['M'] == m]\n",
    "print(d['indirect_effect_p'])\n",
    "print(d['indirect_effect_ci_lower'])\n",
    "print(d['indirect_effect_ci_upper'])\n",
    "d\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6f3aa88-2e19-42b3-bd79-3e0b66119a1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlations of taxa of interest with metabolites\n",
    "gtr_to_otucorr = {}\n",
    "for g in ['saliva','stool']:\n",
    "    print(g)\n",
    "    gtr_to_otucorr[g] = {}\n",
    "    for time in ['diff','all']:\n",
    "        print(time)\n",
    "        gtr_to_otucorr[g][time] = {}\n",
    "        # grab otu table of differences or of all samples\n",
    "        df_otu = gt_to_otu[g][time]\n",
    "        for comp in ['horizontal', 'vertical']:\n",
    "            if comp == 'vertical':\n",
    "                gr_to_sigotu = to_sigotu[comp]\n",
    "                rvs = ['pre', 'post']\n",
    "            elif comp == 'horizontal':\n",
    "                gr_to_sigotu = to_sigotu[comp]\n",
    "                rvs = ['Response', 'No response']\n",
    "            for rv in rvs: # ['pre', 'post']: # Response No response\n",
    "                print(rv)\n",
    "                sigcorr = []\n",
    "                # grab taxa of interest\n",
    "                taxa_int = gr_to_sigotu[g][rv]\n",
    "    \n",
    "                # merge otu table with metabolites            \n",
    "                df_merge = pd.concat([df_otu.loc[:,taxa_int], gt_to_meta[g][time]], axis=1)\n",
    "                meta = gt_to_meta[g][time].columns\n",
    "                \n",
    "                # look at taxa int and correlate with metabolites\n",
    "                # df_merge \n",
    "                for t in taxa_int:\n",
    "                    for m in meta:\n",
    "                        df_sub = df_merge.loc[:,[t,m]]\n",
    "                        df = type_to_df_map[g].copy()\n",
    "                        if time is 'all':\n",
    "                            df = df.set_index('Together')\n",
    "                            if g == 'stool':\n",
    "                                df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "                        elif time is 'diff':\n",
    "                            df = df[df['Timepoint'] == 'pre']\n",
    "                        df_sub = pd.concat([df_sub, df['WOMAC_P_Response']],axis=1)\n",
    "                        df_sub = df_sub.dropna(subset=[t,m])\n",
    "                        r, p = scipy.stats.spearmanr(df_sub[t].values, df_sub[m].values)\n",
    "                        if p < 0.05:\n",
    "                            sigcorr.append([t,m])\n",
    "                            sns.scatterplot(data=df_sub, x=t, y=m, hue='WOMAC_P_Response')\n",
    "                            sns.despine()\n",
    "                        \n",
    "                            plt.tight_layout()\n",
    "                            plt.savefig(path + 'outputs/' + g_to_job[g + '_adh'] + '/corr_taxa_' + comp + '_' + time + '_' + t.split(';g__')[-1] + '_' + m.replace('/','fslash') + '.pdf')\n",
    "                            plt.close()   \n",
    "                gtr_to_otucorr[g][time][rv] = sigcorr\n",
    "                print(len(sigcorr))\n",
    "           \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f6c73d6-71f4-44ee-ac66-5fe0711fe3f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlations of pathways of interest with metabolites\n",
    "gtr_to_pathcorr = {}\n",
    "for g in ['saliva','stool']:\n",
    "    print(g)\n",
    "    gtr_to_pathcorr[g] = {}\n",
    "    for time in ['diff','all']:\n",
    "        print(time)\n",
    "        gtr_to_pathcorr[g][time] = {}\n",
    "        # grab otu table of differences or of all samples\n",
    "        df_path = gt_to_path[g][time]\n",
    "        for comp in ['horizontal', 'vertical']:\n",
    "            if comp == 'vertical':\n",
    "                gr_to_sigpath = to_sigpath[comp]\n",
    "                rvs = ['pre', 'post']\n",
    "            elif comp == 'horizontal':\n",
    "                gr_to_sigpath = to_sigpath[comp]\n",
    "                rvs = ['Response', 'No response']\n",
    "            for rv in rvs: # ['pre', 'post']: # Response No response\n",
    "                print(rv)\n",
    "                sigcorr = []\n",
    "                # grab taxa of interest\n",
    "                taxa_int = gr_to_sigpath[g][rv]\n",
    "    \n",
    "                # merge otu table with metabolites            \n",
    "                df_merge = pd.concat([df_path.loc[:,taxa_int], gt_to_meta[g][time]], axis=1)\n",
    "                meta = gt_to_meta[g][time].columns\n",
    "                \n",
    "                # look at taxa int and correlate with metabolites\n",
    "                # df_merge \n",
    "                for t in taxa_int:\n",
    "                    for m in meta:\n",
    "                        df_sub = df_merge.loc[:,[t,m]]\n",
    "                        df = type_to_df_map[g].copy()\n",
    "                        if time is 'all':\n",
    "                            df = df.set_index('Together')\n",
    "                            if g == 'stool':\n",
    "                                df.index = df.index.map(lambda x: x.split('.guma')[0])\n",
    "                        elif time is 'diff':\n",
    "                            df = df[df['Timepoint'] == 'pre']\n",
    "                        df_sub = pd.concat([df_sub, df['WOMAC_P_Response']],axis=1)\n",
    "                        df_sub = df_sub.dropna(subset=[t,m])\n",
    "                        r, p = scipy.stats.spearmanr(df_sub[t].values, df_sub[m].values)\n",
    "                        if p < 0.05:\n",
    "                            sigcorr.append([t,m])\n",
    "                            sns.scatterplot(data=df_sub, x=t, y=m, hue='WOMAC_P_Response')\n",
    "                            sns.despine()\n",
    "                        \n",
    "                            plt.tight_layout()\n",
    "                            plt.savefig(path + 'outputs/' + g_to_job[g + '_adh'] + '/corr_path_' + comp + '_' + time + '_' + t.split(';g__')[-1] + '_' + m.replace('/','fslash') + '.pdf')\n",
    "                            plt.close()   \n",
    "                gtr_to_pathcorr[g][time][rv] = sigcorr\n",
    "                print(len(sigcorr))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5abd2d7-1c78-4e99-a89e-bb864ac5547f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# L Methionine\n",
    "# Kaempferol\n",
    "# Tryptophan\n",
    "# Tyrosine Stool\n",
    "# Usnic acid\n",
    "# Xanthurenic acid\n",
    "# Tryhydroxycholestenoic acid Stool\n",
    "# Isofraxadine\n",
    "int = ['Isofrax', 'Xanth', 'ydroxychol', 'Usnic', 'Tyrosine', 'Tryptophan', 'Kaempferol', 'Methion']\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    for time in ['diff','all']:\n",
    "        for r in ['Response', 'No response']:\n",
    "            test = gtr_to_pathcorr[g][time][r]\n",
    "            for t in test:\n",
    "                p, m = t # unpack pathway, metabolite, from tuple representing sig corr\n",
    "                for i in int:\n",
    "                    if i in m:\n",
    "                        print(t)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f30bf9df-2904-4ca5-b179-4a4eaa5e71f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "int = ['Isofrax', 'Xanth', 'ydroxychol', 'Usnic', 'Tyrosine', 'Tryptophan', 'Kaempferol', 'Methion']\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    for time in ['diff','all']:\n",
    "        for r in ['Response', 'No response']:\n",
    "            test = gtr_to_otucorr[g][time][r]\n",
    "            for t in test:\n",
    "                p, m = t # unpack pathway, metabolite, from tuple representing sig corr\n",
    "                for i in int:\n",
    "                    if i in m:\n",
    "                        print(t)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d51feb2-1356-4615-81a9-ae357b3f93e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CUTIE\n",
    "# In POST samples only\n",
    "# Color by Resp and Nonresp\n",
    "\n",
    "for g in ['saliva', 'stool']:\n",
    "\n",
    "    df_taxa = gt_to_otu[g]['post']\n",
    "    df_taxa.to_csv(path + 'inputs/cutie_df_taxa_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "    \n",
    "    df_abun = pd.concat([gt_to_otu[g]['post'], gt_to_path[g]['post']],axis=1)\n",
    "    df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_map = type_to_df_map[g].set_index('Together')\n",
    "    df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_meta = pd.concat([gt_to_meta[g]['After diet'],df_map], axis=1)\n",
    "\n",
    "    df_abun.to_csv(path + 'inputs/cutie_df_abun_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "    df_meta.to_csv(path + 'inputs/cutie_df_meta_' + g + '.tsv', sep='\\t', index_label='SampleID')\n",
    "\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fdbda6d-641c-47ab-9c83-d392cf0299a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# anaerobic vs aerobic in response\n",
    "# bile acids -> list? saliva and stool of interest \n",
    "#\n",
    "# Bile acids found in Stool: \n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid', 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide']\n",
    "         }\n",
    "\n",
    "mod_ba = {\n",
    "    'stool': [],\n",
    "    'saliva': []\n",
    "}\n",
    "full = list(gt_to_meta[g]['After diet'].columns.values)\n",
    "for g in ['stool','saliva']:\n",
    "    ba = int_ba[g]\n",
    "    for b in ba:\n",
    "        for f in full:\n",
    "            if b in f:\n",
    "                mod_ba[g].append(f)\n",
    "\n",
    "g_to_cutiejob = {\n",
    "    'saliva': 'jobs15',\n",
    "    'stool': 'jobs16'\n",
    "}\n",
    "\n",
    "\n",
    "# from difflib import SequenceMatcher\n",
    "\n",
    "for g in ['stool','saliva']:\n",
    "    print(g)\n",
    "    n_corr = 0\n",
    "    # import CUITIE results\n",
    "    df_qt = pd.read_csv(path + 'outputs/' + g_to_cutiejob[g] + '/data_processing/summary_df_resample_1.txt', sep='\\t')\n",
    "    df_qt = df_qt[df_qt['class'] == 'TP']\n",
    "    # iterate through rows\n",
    "    for r in range(len(df_qt)):\n",
    "        meta = df_qt.iloc[r,0]\n",
    "        abun = df_qt.iloc[r,1]\n",
    "        \n",
    "        #for i in int_ba:\n",
    "        #    s = SequenceMatcher(None, meta, i)\n",
    "        #    if s.ratio() > 0.9:\n",
    "        #        print(i)\n",
    "        \n",
    "        #if 'bile' in meta:\n",
    "        #    print(meta)\n",
    "        output = [r]\n",
    "        \n",
    "        for i in int_ba[g]:\n",
    "            if i in meta:\n",
    "                output.append(meta)\n",
    "\n",
    "        for c in ['horizontal','vertical']:\n",
    "            if c == 'vertical':\n",
    "                queries = ['pre', 'post']\n",
    "            if c == 'horizontal':\n",
    "                queries = ['Response','No response']\n",
    "            output.append(c)\n",
    "            for r in queries:\n",
    "                int_taxa = to_sigotu[c][g][r]\n",
    "                for i in int_taxa:\n",
    "                    if i in abun:\n",
    "                        output.append('taxa')\n",
    "                        output.append(i)\n",
    "                int_path = to_sigpath[c][g][r]\n",
    "                for i in int_path:\n",
    "                    if i in abun:\n",
    "                        output.append('path')\n",
    "                        output.append(i)\n",
    "        #output.append()\n",
    "        #output.append()\n",
    "        if len(output) > 5:\n",
    "            # print(output)\n",
    "            n_corr += 1\n",
    "\n",
    "    print(n_corr)\n",
    "\n",
    "# pathway prediction of response vs non response @ baseline\n",
    "# specific taxa at baseline\n",
    "\n",
    "\n",
    "# pathways of interest\n",
    "# ex inflammation in non responders\n",
    "# heatmap of differential pathways in responders/nonresponders \n",
    "# and correlation of those pathways with bacteria\n",
    "# send list of upregulated pathways \n",
    "\n",
    "int_taxa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76722aff-f034-4fff-a4da-c63dd1fd75ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotations for pathways\n",
    "int_ba = {'stool': ['L_methionine', 'Pentahydroxylated_bile_acid', 'Dihydroxylated_bile_acid',\n",
    "                    'Trihydroxylated_bile_acid', 'Glycocholic_acid', 'Glycocholate',\n",
    "                    'Glycocholic_acid', 'Tetrahydroxylated_bile_acid', 'Glycochenodeoxycholate',\n",
    "                    'Deoxycholate', 'Hydroxyhexadecanoyl_lysine', 'Deoxycholic_acid'\n",
    "                    'Trihydroxycholestanoic_acid', 'Phenylalanine_conjugated_chenodeoxycholic_acid', 'Nonhydroxylated_bile_acid',\n",
    "                    'Monohydroxylated_bile_acid'],\n",
    "          'saliva': ['Trihydroxylated_bile_acid', 'Palmitoylcarnitine', 'Palmitoyl_ethanolamide']\n",
    "         }\n",
    "\n",
    "for g in ['saliva','stool']:\n",
    "    # int_ba[g] for metabolite (row) annotations\n",
    "    #df_map = type_to_df_map[g].set_index('Together')\n",
    "    #df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    #df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    #df_meta = pd.concat([gt_to_meta[g]['After diet'],df_map], axis=1)\n",
    "    df_meta = gt_to_meta[g]['After diet']\n",
    "    df_meta_R = df_meta.T\n",
    "    df_meta_R['var_type'] = df_meta_R.index.map(lambda x: 'BA' if x in mod_ba[g] else 'NonBA')\n",
    "    df_meta_R = df_meta_R['var_type'].reset_index()\n",
    "    # df_meta_R['index'] = df_meta_R['index'].str.lower()\n",
    "    df_meta_R = df_meta_R.sort_values(by='index',ascending=True)\n",
    "    df_meta_R.to_csv(path + 'inputs/row_' + g + '.tsv', sep='\\t')\n",
    "    \n",
    "    # dcgr_to_sig for column (taxa) annotations\n",
    "    df_abun = gt_to_otu[g]['post']\n",
    "    df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_abun_R = df_abun.T\n",
    "    df_abun_R['var_type'] = df_abun_R.index.map(lambda x: 'Sig' if x in dcgr_to_sig['taxa']['rvnr'][g]['Response'] else 'Nonsig')\n",
    "    df_abun_R = df_abun_R['var_type'].reset_index()\n",
    "    #df_abun_R['index'] = df_abun_R['index'].str.lower()\n",
    "    df_abun_R = df_abun_R.sort_values(by='index',ascending=True)\n",
    "    df_abun_R.to_csv(path + 'inputs/col_' + g + '.tsv', sep='\\t')\n",
    "\n",
    "df_abun_R.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fbf3e56-2d5f-41fa-8ed7-718d92ed6400",
   "metadata": {},
   "outputs": [],
   "source": [
    "# play testing mediation\n",
    "g = 'stool'\n",
    "df_M = pd.read_csv(path + 'inputs/cutie_df_meta_stool.tsv', sep='\\t', index_col=0)\n",
    "df_M.iloc[:,:-1].to_csv(path + 'inputs/mediation_meta_stool_post_test.txt', sep='\\t')\n",
    "\n",
    "df_Y = df_meta_paired[['VAS_Pt_diff', 'WOMAC_pain_diff']]\n",
    "df_Y.to_csv(path + 'inputs/mediation_WMdiff_test.txt', sep='\\t')\n",
    "\n",
    "df_X = pd.read_csv(path + 'inputs/cutie_df_taxa_stool.tsv', sep='\\t', index_col=0)\n",
    "df_X.index = df_X.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "df_X.to_csv(path + 'inputs/mediation_taxa_stool_post_test.txt', sep='\\t')\n",
    "df_X.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11360b5e-dd8b-4857-b205-4dd8c5c685b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for corr heatmaps \n",
    "for g in ['saliva','stool']:\n",
    "    df_abun = pd.concat([gt_to_otu[g]['post'], gt_to_path[g]['post']],axis=1)\n",
    "    df_abun.index = df_abun.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_map = type_to_df_map[g].set_index('Together')\n",
    "    df_map = df_map[df_map['Timepoint'] == 'post']['WOMAC_P_Response']\n",
    "    df_map.index = df_map.index.map(lambda x: x.split('.post')[0].replace('-',''))\n",
    "    \n",
    "    df_meta = pd.concat([gt_to_meta[g]['After diet'],df_map], axis=1)\n",
    "\n",
    "    df_quant_R = df_abun.copy()\n",
    "    #df_quant_R = pd.concat([df_abun, df_meta],axis=1)# .copy() #([df_quant, df_asv100, df_path_top100], axis=1)\n",
    "    # df_quant_R = df_quant_R.drop(['WOMAC_P_Response'],axis=1)\n",
    "    # df_quant_R.to_csv(path + 'inputs/df_quant_' + g + '.tsv', sep='\\t')\n",
    "    df_quant_R = df_quant_R.T\n",
    "    # this is in the order they were entered i.e. \n",
    "    # [df_olink, df_fa, df_acpa_fecal, df_acpa_plasma, df_plasma, df_rbfa, df_mb]\n",
    "     # ['rbfa']*len(df_rbfa.columns) + \\  \n",
    "    df_quant_R['var_type'] = ['Taxa']*len(gt_to_otu[g]['post'].columns) + \\\n",
    "                             ['Pathways']*len(gt_to_path[g]['post'].columns) # + \\\n",
    "                             # ['Metabolite']*len(gt_to_meta[g]['After diet'].columns)\n",
    "    # df_quant_R = df_quant_R[~df_quant_R['var_type'].isin(['Metagenomic_ASVs','Metagenomic_Pathways'])]\n",
    "    df_quant_R = df_quant_R['var_type'].reset_index()\n",
    "    df_quant_R['index'] = df_quant_R['index'].str.lower()\n",
    "    df_quant_R = df_quant_R.sort_values(by='index',ascending=True)\n",
    "    df_quant_R.to_csv(path + 'inputs/df_quant_R_corr_labels_' + g + '.tsv', sep='\\t')\n",
    "    print(len(df_quant_R))\n",
    "df_quant_R.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2707d857-ee9c-466f-983c-3d94600478d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4 way paired boxplot trial\n",
    "g = 'saliva'\n",
    "df_alpha = g_to_dfd[g + '_adh']['alpha']\n",
    "\n",
    "df_map = type_to_df_map[g].copy()\n",
    "df_map = df_map.set_index('Together')\n",
    "df_map = df_map[['Timepoint','WOMAC_P_Response']]\n",
    "\n",
    "df_merge = pd.concat([df_alpha, df_map], axis=1)\n",
    "df_merge = df_merge.dropna()\n",
    "\n",
    "sns.boxplot(x='Timepoint', y='shannon_entropy',\n",
    "            hue='WOMAC_P_Response', palette=[\"r\", \"g\"],\n",
    "            data=df_merge)\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a341d51b-1db7-4c98-b4e8-65a312f999e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test if each sub correlation is sig\n",
    "m = 'H_Nonhydroxylated_bile_acid'\n",
    "y = 'WOMAC_pain_diff'\n",
    "y = 'VAS_Pt_diff'\n",
    "x = 'd__Bacteria;p__Firmicutes_A;c__Clostridia_258483;o__Lachnospirales;f__Lachnospiraceae;__'\n",
    "\n",
    "df = df_plot[[x,y,m]]\n",
    "\n",
    "df_corr = df.dropna()\n",
    "print('Y on X')\n",
    "print(scipy.stats.pearsonr(df_corr[y].values, df_corr[x].values))\n",
    "print('M on X')\n",
    "print(scipy.stats.pearsonr(df_corr[m].values, df_corr[x].values))\n",
    "\n",
    "df['meta'] = df[m].apply(lambda x: 'high' if -1.25 <= x <= 0.25 else 'low')\n",
    "\n",
    "# do each linear model separately\n",
    "df_corr = df[df['meta'] == 'low']\n",
    "df_corr = df_corr.dropna()\n",
    "print(scipy.stats.pearsonr(df_corr[y].values, df_corr[x].values))\n",
    "\n",
    "df_corr = df[df['meta'] == 'high']\n",
    "df_corr = df_corr.dropna()\n",
    "print(scipy.stats.pearsonr(df_corr[y].values, df_corr[x].values))\n",
    "# df_corr\n",
    "df.columns = ['Lachno','Pain','H_Nonhydroxylated_BA', 'meta']\n",
    "df.to_csv(path + 'inputs/df_med_R.tsv', sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "101b41e6-d1b0-4d18-a49c-7c338f54078d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://github.com/theJohnnyBrown/permanova/blob/master/permanova.py\n",
    "\n",
    "#!/usr/bin/python\n",
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "# Copyright (C) 2010 - 2012, University of New Orleans\n",
    "#\n",
    "# This program is free software; you can redistribute it and/or modify it under\n",
    "# the terms of the GNU General Public License as published by the Free\n",
    "# Software Foundation; either version 2 of the License, or (at your option)\n",
    "# any later version.\n",
    "#\n",
    "# Author: Johnny Brown\n",
    "\n",
    "import numpy as np\n",
    "import random as r\n",
    "from itertools import permutations, product, chain\n",
    "\n",
    "from scipy import stats\n",
    "#stats.ss chokes on generators. Might be worth fixing that.\n",
    "stats.ss = lambda l: sum(a*a for a in l)\n",
    "\n",
    "def above_diagonal(n):\n",
    "    row = range(n)\n",
    "    for i in row:\n",
    "        for j in range(i+1,n):\n",
    "            yield i,j\n",
    "    \n",
    "\n",
    "def permanova_oneway(dm, levels, permutations = 200):\n",
    "    \"\"\" \n",
    "    Performs one-way permutational ANOVA on the given distance matrix.\n",
    "\n",
    "    One-way permanova tests the null hypothesis that distances between levels of\n",
    "    the variable are not significantly different from distances within levels of\n",
    "    the variable.\n",
    "\n",
    "    The test-statistic is a multivariate analogue to Fisher's F-ratio and is \n",
    "    calculated directly from any symmetric distance or dissimilarity matrix. \n",
    "    P-values are then obtained using permutations.\n",
    "\n",
    "    In the case of a Euclidean distance matrix calculated from only one \n",
    "    variable, the F-value is the same as the traditional parametric univariate \n",
    "    F-statistic.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    dm : array_like\n",
    "        The distance matrix of observations x observations. Represents a \n",
    "        symmetric n x n matrix with zeros on the diagonal.\n",
    "\n",
    "    levels : array_like\n",
    "        An array indicating the levels of the variable at each observation, such\n",
    "        that levels[i] == levels[j] means that dm[i] and dm[j] are rows \n",
    "        corresponding to observations in the same level or treatment.\n",
    "\n",
    "    permutations : int\n",
    "        The number of permutations used to compute the p-value. Default is 200.\n",
    "        If there are less than ``permutations`` unique permutations, then all of \n",
    "        them will be used\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    F-value : float\n",
    "        The computed F-value of the test.\n",
    "    p-value : float\n",
    "        The associated p-value from the F-distribution, generated by permuting \n",
    "        the levels\n",
    "\n",
    "    Notes\n",
    "    -----\n",
    "    It is assumed that all sample groups are the same size, n. \n",
    "\n",
    "    For example, if the values in levels are placebo, 5mg, and 10mg, then each \n",
    "    value must occur n times in the levels array, and \n",
    "    3n == len(levels) == len(dm) == len(dm[i]) (for all 0 <= i < len(dm))\n",
    "    \n",
    "    The algorithm is from Anderson[2] \n",
    "\n",
    "    References\n",
    "    ----------\n",
    "    .. [1] Lowry, Richard.  \"Concepts and Applications of Inferential\n",
    "           Statistics\". Chapter 14. http://faculty.vassar.edu/lowry/ch14pt1.html\n",
    "\n",
    "    .. [2] Anderson, Marti. A new method for non-parametric multivariate analysis \n",
    "           of variance. 2001. \n",
    "           http://www.entsoc.org/PDF/MUVE/6_NewMethod_MANOVA1_2.pdf\n",
    "    \n",
    "    .. [3] Jones, James. Lecture notes for Concepts of Statistics\n",
    "           http://people.richland.edu/james/lecture/m170/ch13-2wy.html\n",
    "    \"\"\"\n",
    "    bigf = f_oneway(dm,levels)\n",
    "\n",
    "    above = below = 0\n",
    "    nf = 0\n",
    "\n",
    "    #TODO make this pretty with math and functions\n",
    "    #perms = r.sample(list(perm_unique(levels)),permutations)\n",
    "    shuffledlevels = list(levels)#copy list so we can shuffle it\n",
    "    \n",
    "    for i in range(permutations):\n",
    "        r.shuffle(shuffledlevels)\n",
    "        f = f_oneway(dm,shuffledlevels)\n",
    "        \n",
    "        if f >= bigf:\n",
    "            above += 1\n",
    "\n",
    "        #debug\n",
    "        ## print shuffledlevels\n",
    "        ## print f\n",
    "\n",
    "    p = above/float(permutations)\n",
    "\n",
    "    return (bigf,p)\n",
    "\n",
    "#FIXME the right way to reuse code for one-way and n-way is to add an arg f such that\n",
    "#f(levels[i],levels[j]) returns True iff dm[i][j] should be included in the\n",
    "#desired sum of squares\n",
    "def f_oneway(dm,levels):\n",
    "    bign = len(levels)#number of observations\n",
    "    dm = np.asarray(dm)#distance matrix\n",
    "    a = len(set(levels))#number of levels\n",
    "    n = bign/a#number of observations per level\n",
    "    \n",
    "    assert dm.shape == (bign,bign) #check the dist matrix is square and the size\n",
    "                                   #corresponds to the length of levels\n",
    "\n",
    "    #total sum of squared distances                                   \n",
    "    sst = np.sum(stats.ss(r) for r in \n",
    "              (s[n+1:] for n,s in enumerate(dm[:-1])) )/float(bign)\n",
    "\n",
    "    #sum of within-group squares\n",
    "    #itertools.combinations(range(len(dm)),2)#top half of dm\n",
    "    ssw = np.sum((dm[i][j]**2 for i,j in  \n",
    "               product(range(len(dm)),range(1,len(dm)))\n",
    "               if i<j and levels[i] == levels[j]))/float(n)\n",
    "\n",
    "    ssa = sst - ssw\n",
    "\n",
    "    fstat = (ssa/float(a-1))/(ssw/float(bign-a))\n",
    "    #print (fstat,sst,ssa,ssw,a,bign,n)\n",
    "\n",
    "    return fstat\n",
    "\n",
    "def permanova_twoway(dm,levels,permutations=200):\n",
    "    \"\"\"Performs one-way permutational ANOVA on the given distance matrix.\n",
    "\n",
    "    One-way permanova tests the null hypothesis that distances between levels of\n",
    "    the variable are not significantly different from distances within levels of\n",
    "    the variable.\n",
    "\n",
    "    The test-statistic is a multivariate analogue to Fisher's F-ratio and is \n",
    "    calculated directly from any symmetric distance or dissimilarity matrix. \n",
    "    P-values are then obtained using permutations.\n",
    "\n",
    "    In the case of a Euclidean distance matrix calculated from only one \n",
    "    variable, the F-value is the same as the traditional parametric univariate \n",
    "    F-statistic.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    dm : array_like\n",
    "        The distance matrix of observations x observations. Represents a \n",
    "        symmetric n x n matrix with zeros on the diagonal.\n",
    "\n",
    "    levels : array_like\n",
    "        An array of pairs indicating the levels of the variable at each \n",
    "        observation, such that levels[i] == levels[j] means that dm[i] and dm[j] \n",
    "        are rows corresponding to observations in the same level or treatment.\n",
    "\n",
    "    permutations : int\n",
    "        The number of permutations used to compute the p-value. Default is 200.\n",
    "        If there are less than ``permutations`` unique permutations, then all of \n",
    "        them will be used\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    F-values : 3-tuple\n",
    "        The computed F-values of each test. F-values[0] corresponds to the \n",
    "        interaction of the two levels, F-values[1] corresponds to the effect\n",
    "        of variable a, and F-values[2] corresponds to the effect of variable\n",
    "        b. \n",
    "    p-value : 3-tuple\n",
    "        The associated p-values from the F-distribution, generated by permuting \n",
    "        the levels. The values correspond to (interaction, a, b) as with the \n",
    "        F-values.\n",
    "\n",
    "    Notes\n",
    "    -----\n",
    "    All sample groups must be the same size, n. \n",
    "\n",
    "    For example, if the variables are drug and dose, with levels 5mg, 10mg, 15mg\n",
    "    and drug1, drug2, then each combination of drug type and dose level must \n",
    "    be of size n, i.e. (5mg, drug1) must occur n times in the levels array and\n",
    "    6n == len(levels) == len(dm)\n",
    "\n",
    "    The algorithm is from Anderson[2] \n",
    "\n",
    "    References\n",
    "    ----------\n",
    "    .. [1] Lowry, Richard.  \"Concepts and Applications of Inferential\n",
    "           Statistics\". Chapter 14. http://faculty.vassar.edu/lowry/ch14pt1.html\n",
    "\n",
    "    .. [2] Anderson, Marti. A new method for non-parametric multivariate analysis \n",
    "           of variance. 2001. \n",
    "           http://stg-entsoc.bivings.com/PDF/MUVE/6_NewMethod_MANOVA1_2.pdf\n",
    "\n",
    "    .. [3] Jones, James. Lecture notes for Concepts of Statistics\n",
    "           http://people.richland.edu/james/lecture/m170/ch13-2wy.html\n",
    "    \"\"\"\n",
    "\n",
    "    bigf_i, bigf_a, bigf_b = f_twoway(dm,levels)\n",
    "\n",
    "    above_i = above_a = above_b = 0\n",
    "\n",
    "    #TODO make this pretty with math and functions\n",
    "    #perms = r.sample(list(perm_unique(levels)),permutations)\n",
    "    shuffledlevels = list(levels)#copy list so we can shuffle it\n",
    "\n",
    "    a_levels = list([l[0] for l in levels])\n",
    "    b_levels = list([l[1] for l in levels])\n",
    "    \n",
    "    #permutations\n",
    "    for i in range(permutations):\n",
    "        #All these are probably the wrong way to do the permutations. Wrong as in\n",
    "        #incorrect\n",
    "        ## r.shuffle(a_levels)\n",
    "        ## r.shuffle(b_levels)\n",
    "        ## shuffledlevels = zip(a_levels,b_levels)\n",
    "\n",
    "        r.shuffle(shuffledlevels)\n",
    "        \n",
    "        f_i, f_a, f_b = f_twoway(dm,shuffledlevels)\n",
    "\n",
    "        if f_i > bigf_i:\n",
    "            above_i += 1\n",
    "\n",
    "    for i in range(permutations):\n",
    "        r.shuffle(a_levels)\n",
    "\n",
    "        f_i, f_a, f_b = f_twoway(dm,zip(a_levels, [l[1] for l in levels]))\n",
    "\n",
    "        if f_a > bigf_a:\n",
    "            above_a += 1\n",
    "\n",
    "    for i in range(permutations):\n",
    "        r.shuffle(b_levels)\n",
    "\n",
    "        f_i, f_a, f_b = f_twoway(dm,zip([l[0] for l in levels], b_levels))\n",
    "            \n",
    "        if f_b > bigf_b:\n",
    "            above_b += 1\n",
    "\n",
    "\n",
    "\n",
    "    p_i,p_a,p_b = [ above/float(permutations) for above in \n",
    "                    [above_i, above_a, above_b]]\n",
    "\n",
    "    return ((bigf_i, bigf_a, bigf_b), (p_i, p_a, p_b))\n",
    "\n",
    "    \n",
    "    \n",
    "def f_twoway(dm, levels):\n",
    "    \n",
    "    bign = len(levels)#number of observations\n",
    "    dm = np.asarray(dm)#distance matrix\n",
    "    l = len(set(levels))#number of levels\n",
    "    a = len(set([l[0] for l in levels]))#number of a-levels\n",
    "    b = len(set([l[1] for l in levels]))#number of b-levels\n",
    "    n = bign/float(a*b)#number of observations per level\n",
    "\n",
    "    #sum of all distances\n",
    "    ## sst = np.sum(stats.ss(r) for r in \n",
    "    ##         (s[n+1:] for n,s in enumerate(dm[:-1])) )/float(bign)\n",
    "    sst = stats.ss(chain(*(r[i+1:] for i,r in enumerate(dm))))/float(bign)\n",
    "\n",
    "    #same level of both a and b (error, within-group)\n",
    "    ssr = select_ss(dm, levels, lambda a,b: a==b)/float(n)\n",
    "\n",
    "    #same level of a\n",
    "    sswa = select_ss(dm, levels, lambda a,b: a[0] == b[0])/float(b*n)\n",
    "\n",
    "    #same level of b\n",
    "    sswb = select_ss(dm, levels, lambda a,b: a[1] == b[1])/float(a*n)\n",
    "\n",
    "    \n",
    "    ssa = sst - sswa#effect of a\n",
    "    ssb = sst - sswb#effect of b\n",
    "    ssab = sst - ssa - ssb - ssr #interaction sum-of-squares\n",
    "\n",
    "    #these should each be separate functions?\n",
    "    f_interaction = (ssab/float((a-1)*(b-1)))/(ssr/float(bign - a*b))\n",
    "    f_a = (ssa/float((a-1)))/(ssr/float(bign - a*b))\n",
    "    f_b = (ssb/float((b-1)))/(ssr/float(bign - a*b))\n",
    "\n",
    "    return (f_interaction,f_a,f_b)\n",
    "\n",
    "def select_ss(dm, levels,  included):\n",
    "    \n",
    "    bign = len(dm)\n",
    "\n",
    "    distances = (dm[i][j] for i,j in above_diagonal(bign) \n",
    "                           if included(levels[i], levels[j]))\n",
    "\n",
    "    return stats.ss(distances)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "061973e1-2fae-4d91-8847-ea6ce3450a9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(621.576923076923, 0.11)"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "permanova_oneway([[1,2,30,30,30],[2,1,30,30,30],[30,30,1,1,2],[30,30,30, 1,2], [30,30,30,2,1]], levels=[1,1,0,0,0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ad3a4bc-37a9-4ae4-9075-376b97b35337",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
